# Multi-table methods {#multitable}

```{r setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble()
```

This chapter focuses on multi-table methods.



## Representation of multiple data tables

Many microbiome experiments include multiple versions and types of
data. We start by providing recommendations on how to represent
different varieties of multi-table data within the
TreeSummarizedExperiment framework.

**Transformations** The original count-based taxonomic abundance
tables may have different transformations, such as log10p, clr, or
relative abundance. These are typically stored in _assays_.

TODO: Example - OR refer to the data basics/manipulation section.


**Alternative experiments** differ from transformations as they can
  contain complementary data. One can have taxonomic abundance
  profiles quantified with different measurement technologies, for
  instance, such as phylogenetic microarrays, amplicon sequencing, or
  metagenomic sequencing. Assuming that the taxonomic information can
  be mapped between these types. Alternative experiments can be
  represented as 1) separate _assays_; or 2) using the altExp() option in 

TODO: Example on assays - OR refer to the data basics/manipulation section.

TODO: Example on SingleCellExperiment::altExp() [examples](https://bioconductor.org/packages/3.12/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)


**Multiple experiments** relate to complementary measurement types,
  such as transcriptomic or metabolomic profiling of the microbiome or
  the host. Multiple experiments can be represented using the same
  options as alternative exaperiments, or by using the
  MultiAssayExperiment class.

TODO: Example on MultiAssayExperiment()


## Session Info {-}

```{r sessionInfo, echo=FALSE, results='asis'}
prettySessionInfo()
```
