# Differential Abundance {#differential-abundance}

```{r setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble()
```

This section provides an introduction to Differential Abundance Analysis (DAA),
which is used to identify differences in the abundances of individual taxa (at
any taxonomic level) between two or more groups, such as treatment and control.
Here, we demonstrate its implementation on Tengeler2020, described in chapter \@ref(tengeler-desc).

The goal of DAA is to identify biomarkers of a certain phenotype or condition,
and gain understanding of a complex system by looking at its isolated components.
For example, the identification of a bacterial taxon that is differentially
abundant between healthy patients and diseased patients can lead to important
insights into the pathophysiology of the disease. In other words, differentially
abundant taxa can be involved in the dynamics of the disease, which in turn helps
understand the system as a whole. Despite its relevance in current research,
the DAA approach has also been subject to debate [@Quinn2021].


## Statistical challenges of microbiome data 

Microbiome data display unique properties that are exclusively addressed by DAA
tools developed for microbiome analysis [@Yang2022]. Specifically, microbiome
data are characterized by high variability, zero-inflation and compositionality.
**High variability** expresses that abundance of taxa often varies by several
orders of magnitude from sample to sample. **Zero-inflation** means that
typically more than 70% of the values are zeros, which could be due to either
physical absence (structural zeros) or insufficient sampling effort (sampling zeros).
**Compositionality** implies that a change in the absolute abundance of one taxon
will lead to apparent variations in the relative abundances of other taxa in the
same sample. If neglected, such properties may cause significant bias in the
results of DAA. Therefore, several approaches have been developed to address the
unique properties of microbiome data and provide statistically useful results.

The first approach to target zero-inflated data consists of specialized models,
such as **over-dispersed count models** and **zero-inflated mixture models**.
DESeq2 [@Love2014], edgeR [@Chen2016] and corncorb [@Martin2021] are based on
over-dispersed count models, whereas metagenomeSeq [@Paulson2017], RAIDA
[@Sohn2015], ZIBB [@ZIBB2018] and Omnibus [@Omnibus2018] implement zero-inflated
mixture models to address zero-inflation. Typically, these models assume a
negative binomial, beta-binomial or normal/log-normal distribution. Alternatively, **zero imputation** also represents a valid approach to deal with zero-inflated
data. ALDEx2 [@Gloor2016] and eBay [@Liu2020] apply a Bayesian model to impute
the zeros when working with proportion data, accounting for sampling variability
and sequencing depth variation. Other methods, such as MaAsLin2 [@Mallick2020]
and ANCOMBC [@ancombc2020] impute the zeros with a pseudo-count strategy.

Regarding the compositionality of microbiome data, several approaches have been
developed to perform **robust normalization** with methods specifically designed
to reduce the bias found in compositional data. Some examples include trimmed mean
of M-values (TMM) normalization used by edgeR, relative log expression (RLE) normalization used by DESeq2, cumulative sum scaling (CSS) normalization used by metagenomeSeq, centered log-ratio transformation (CLR) normalization used
by ALDEx2 and geometric mean of pairwise ratios (GMPR) normalization used by Omnibus and Wrench normalization [@Kumar2018], which corrects the compositional bias by an
empirical Bayes approach. Other methods to deal with compositional data entail
reference taxa approach used by DACOMP [@Brill2019] and RAIDA, analyzing the
pattern of pairwise log ratios as done by ANCOM [@Mandal2015] and bias-correction
applied by ANCOMBC.

We recommend to have a look at @Nearing2022. In this study, multiple DAA methods
were applied to 38 different datasets and their results were compared to one
another. Because each method follows a slightly different approach in terms of
assumptions and normalization techniques, it was shown that results on the same
dataset may differ substantially depending on the method. More recently, @Yang2022 comprehensively evaluated DAA methods via a semi-parametric framework and 106 real datasets. This study also concluded that different methods can produce contradictory results, creating the risk of cherry-picking the most favorable options for oneâ€™s
own hypothesis. Therefore, it is highly recommended to perform DAA with multiple
methods to determine whether the findings can be reproduced by different approaches.

## Using the tools

In this section we demonstrate the use of four methods that can be
recommended based on recent literature (ANCOM-BC [@ancombc2020],
_ALDEx2_ [@Gloor2016], _Maaslin2_ [@Mallick2020], _LinDA_ [@Zhou2022]
and _ZicoSeq_ [@Yang2022]).

The purpose of this section is to show how to perform DAA in R, not
how to correctly do causal inference. Depending on your experimental
setup and your theory, you must determine how to specify any model
exactly.  E.g., there might be confounding factors that might drive
(the absence of) differences between the shown groups that we ignore
here for simplicity. Or your dataset is repeated sampling design,
matched-pair design or the general longitudianl design.  We will
demonstrate how to include covariates in those models. We picked a
dataset that merely has microbial abundances in a TSE object as well
as a grouping variable in the sample data. We simplify the examples by
only including two of the three groups.

```{r install-pkg, include = FALSE}
if(!require(GUniFrac)){
  install.packages("GUniFrac")
}
```


```{r load-pkg-data}
library(tidySummarizedExperiment)
library(tidyverse)

# Import dataset
library(mia)
data("Tengeler2020", package = "mia")
tse <- Tengeler2020

# Show patient status by cohort
table(tse$patient_status, tse$cohort) %>%
  knitr::kable()
```

### Preparing the data for DAA 

Before starting the analysis, it is recommended to reduce the size and
complexity of the data to make the results more reproducible. For this
purpose, we agglomerate the features by genus and filter them by a prevalence
threshold of 10%.

```{r datasetup}
# Agglomerate by genus and subset by prevalence
tse <- subsetByPrevalentTaxa(tse,
                             rank = "Genus",
                             prevalence = 10 / 100)

# Apply relative abundance transformation
tse <- transformAssay(tse,
                      assay.type = "counts",
                      method = "relabundance")

# Apply clr transformation
tse <- transformAssay(tse,
                      assay.type = "counts",
                      method = "clr",
                      pseudocount = 1)
```

While some DAA tools provide optional arguments for prevalence filtering, here
we filtered the tse object directly. This way, we ensure that the input data
remains the same when multiple tools are used.


### ALDEx2

In this section, we will show how to perform a simple ALDEx2 analysis. 
If you wanted to pick a single method, this method could be recommended to use.
According to the developers experience, it tends to identify the common
features identified by other methods. This statement is in line with a recent
independent evaluation by @Nearing2022.  
Please also have a look at the more extensive 
[vignette](https://bioconductor.org/packages/release/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.html) 
that covers this flexible tool in more depth. ALDEx2 estimates technical
variation within each sample per taxon by utilizing the Dirichlet distribution.
It furthermore applies the centered-log-ratio transformation (or closely
related log-ratio transforms). Depending on the experimental setup, it will
perform a two sample Welch's T-test and Wilcoxon-test or a one-way ANOVA and
Kruskal-Wallis-test. For more complex study designs, there is a possibility to 
utilize the `glm` functionality within ALDEx2.

The Benjamini-Hochberg procedure is applied by default to correct for
multiple testing. Below we show a simple example that illustrates the
workflow.


```{r, aldex2, eval=TRUE}
# Load package
library(ALDEx2)

# Generate Monte Carlo samples of the Dirichlet distribution for each sample.
# Convert each instance using the centered log-ratio transform.
# This is the input for all further analyses.
set.seed(123)
x <- aldex.clr(assay(tse), tse$patient_status)     
```


The t-test:

```{r, aldex2_ttest, eval=TRUE}
# calculates expected values of the Welch's t-test and Wilcoxon rank
# test on the data returned by aldex.clr
x_tt <- aldex.ttest(x, paired.test = FALSE, verbose = FALSE)
```


Effect sizes:

```{r, aldex2_efs, eval=TRUE}
# Determines the median clr abundance of the feature in all samples and in
# groups, the median difference between the two groups, the median variation
# within each group and the effect size, which is the median of the ratio
# of the between group difference and the larger of the variance within groups
x_effect <- aldex.effect(x, CI = TRUE, verbose = FALSE)

# combine all outputs 
aldex_out <- data.frame(x_tt, x_effect)
```

Now, we can create a so called Bland-Altman or MA plot (left). It shows the
association between the relative abundance and the magnitude of the difference
per sample. Next to that, we can also create a plot that shows the dispersion
on the x-axis instead of log-ratio abundance. Red dots represent genera that are
differentially abundant ($q \leq 0.1$) between the 2 groups. Black points are
rare taxa and grey ones are abundant taxa. The dashed line represent an effect
size of 1. See @Gloor2016 to learn more about these plots.

```{r, eval=TRUE}
par(mfrow = c(1, 2))

aldex.plot(aldex_out,
           type = "MA",
           test = "welch",
           xlab = "Log-ratio abundance",
           ylab = "Difference",
           cutoff = 0.05)

aldex.plot(aldex_out,
           type = "MW",
           test = "welch",
           xlab = "Dispersion",
           ylab = "Difference",
           cutoff = 0.05)
```

The evaluation as differential abundant in above plots is based on the
corrected p-value. According to the ALDEx2 developers, the safest
approach is to identify those features where the 95% CI of the effect
size does not cross 0. As we can see in below table, this is not the
case for any of the identified genera (see overlap column, which
indicates the proportion of overlap). Also, the authors recommend to
focus on effect sizes and CIs rather than interpreting the p-value. To
keep the comparison simple, we will here use the p-value as decision
criterion. But please be aware that the effect size together with the
CI is a better answer to the question we are typically interested in
(see also [this
article](https://www.nature.com/articles/d41586-019-00857-9)).


```{r, eval=TRUE}
aldex_out %>%
  rownames_to_column(var = "genus") %>%
  # here we choose the wilcoxon output rather than t-test output
  filter(wi.eBH <= 0.05)  %>%
  dplyr::select(genus, we.eBH, wi.eBH, effect, overlap) %>%
  knitr::kable()
```

### ANCOM-BC

The analysis of composition of microbiomes with bias correction
(ANCOM-BC) [@ancombc2020] is a recently developed method for differential
abundance testing. It is based on an earlier published approach
[@Mandal2015].  The previous version of ANCOM was among the methods
that produced the most consistent results and is probably a
conservative approach [@Nearing2022].  However, the new ANCOM-BC
method operates quite differently compared to the former ANCOM method.

As the only method, ANCOM-BC incorporates the so called *sampling
fraction* into the model. The latter term could be empirically
estimated by the ratio of the library size to the microbial
load. According to the authors, ignoring variations
in this sampling fraction would bias DAA results. Furthermore,
this method provides p-values and confidence intervals for each
taxon. It also controls the FDR and it is computationally simple to
implement.

Note that the original method was implemented in the `ancombc()` function (see 
[extended tutorial](https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html)).
The method has since then been updated and new features have been added to enable
multi-group comparisons and repeated measurements among other improvements. 
We do not cover the more advanced features of ANCOMBC in this tutorial 
as these features are documented in detail in this 
[tutorial](https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html).

We now proceed with a simple example.  First, we specify a formula. In this 
formula, other covariates could potentially be included to adjust for 
confounding. We show this further below. Again, please make sure to check the 
[function documentation](https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html)
as well as the linked tutorials to learn about the additional arguments 
that we specify.


```{r ancombc2, warning = FALSE, eval=TRUE}
# Load package
library(ANCOMBC)

# Run ANCOM-BC at the genus level and only including the prevalent genera
out <- ancombc2(data = tse,
                assay_name = "counts",
                fix_formula = "patient_status",
                p_adj_method = "fdr",
                prv_cut = 0,
                group = "patient_status",
                struc_zero = TRUE,
                neg_lb = TRUE,
                # multi group comparison is deactivated automatically
                global = TRUE)
```

```{r ancombc2b, warning = FALSE, eval=TRUE}
# store the FDR adjusted results [test on v2.0.3] 
ancombc_result <- cbind.data.frame(taxid = out$res$taxon,
                       ancombc = as.vector(out$res$q_patient_statusControl))
```

```{r ancombc2c, warning = FALSE, eval=TRUE}
# store the FDR adjusted results [test on v1.2.2]
ancombc_result <- out$res %>%
  dplyr::select(starts_with(c("taxon", "lfc", "q")))
```

The object `out` contains all model output. Again, see the
[documentation of the
function](https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html)
under **Value** for details. Our question whether taxa are
differentially abundant can be answered by looking at the `res`
object, which contains dataframes with the coefficients, standard
errors, p-values and q-values. Below we show the first entries of this
dataframe.

```{r ancom_kable, eval=TRUE}
ancombc_result %>%
  head() %>%
  knitr::kable()
```



### MaAsLin2 

Let us next illustrate MaAsLin2 [@Mallick2020]. This method is based on
generalized linear models and flexible for different study designs
and covariate structures. For details, check their
[Biobakery tutorial](https://github.com/biobakery/biobakery/wiki/maaslin2).

```{r maaslin2, results = "hide", eval=TRUE}
# Load package
library(Maaslin2)

# maaslin expects features as columns and samples as rows 
# for both the abundance table as well as metadata 

# We can specify different GLMs/normalizations/transforms.
# specifying a ref is especially important if you have more than 2 levels
maaslin2_out <- Maaslin2(input_data = t(assay(tse)),
                         input_metadata = as.data.frame(colData(tse)),
                         output = "DAA example",
                         transform = "AST",
                         fixed_effects = "patient_status",
                         # you can also fit MLM by specifying random effects
                         # random_effects = c(...),
                         reference = "patient_status,Control",
                         normalization = "TSS",
                         standardize = FALSE,
                         # filtering was previously performed
                         min_prevalence = 0)
```

Which genera are identified as differentially abundant? (leave out "head" to see all).

```{r, maaslin2kable, eval=TRUE}
maaslin2_out$results %>%
  filter(qval <= 0.05) %>%
  head() %>%
  knitr::kable()
```

This will create a folder that is called like in the output specified
above. It contains also figures to visualize difference between
significant taxa.


### LinDA 

Lastly, we cover linear models for differential abundance analysis of
microbiome compositional data (@Zhou2022). This is very similar to
ANCOMBC with few differences: 1) LinDA correct for the compositional
bias differently using the mode of all regression coefficients. 2) it
is faster (100x-1000x than ANCOMBC and according to the authors); 3)
it supports hierarchical models. The latest ANCOMBC versions are also
supporting hierarchical models. Nevertheless, LinDA seems a promising
tool that achieves a very good power/fdr trade-off together with
ANCOMBC according to the review. The speed improvements might make it
critical especially for datasets that have higher sample or feature
set sizes.


```{r linda, eval=TRUE}
# Load package
library(MicrobiomeStat)

# Store independent variables into a data.frame
meta.dat <- colData(tse) %>%
  as.data.frame() %>%
  dplyr::select(patient_status)

# Run LinDA
linda.res <- linda(feature.dat = as.data.frame(assay(tse)),
                   meta.dat = meta.dat,
                   formula = "~ patient_status",
                   alpha = 0.05,
                   prev.filter = 0,
                   mean.abund.filter = 0)

linda_out <- linda.res$output$patient_statusControl
```

```{r linda2, eval=TRUE}
# to scan the table for genera where H0 could be rejected:
linda_out %>%
  as.data.frame() %>%
  filter(reject) %>%
  head() %>%
  knitr::kable()
```

### ZicoSeq

Subsequently, we add a linear model and permutation-based method, see
details at [tutorial](https://cran.r-project.org/web/packages/GUniFrac/vignettes/ZicoSeq.html).

This approach has been assessed to exhibit high power and a low false
discovery rate, which has the following components:

  1. Winsorization to decrease the influence of outliers;
  
  2. Posterior sampling based on a beta mixture prior to address
  sampling variability and zero inflation;
  
  3. Reference-based multiple-stage normalization to address
  compositional effects;


```{r ZicoSeq, eval=TRUE}
# Load package
library(GUniFrac)

set.seed(123)
zicoseq.obj <- ZicoSeq(feature.dat = as.matrix(assay(tse)),
                       meta.dat = as.data.frame(colData(tse)),
                       grp.name = "patient_status",
                       feature.dat.type = "count",
                       return.feature.dat = TRUE,
                       prev.filter = 0,
                       mean.abund.filter = 0,
                       max.abund.filter = 0,
                       perm.no = 999)
```

```{r ZicoSeq_kable, eval=TRUE}
zicoseq_out <- cbind.data.frame(p.raw = zicoseq.obj$p.raw,
                                p.adj.fdr = zicoseq.obj$p.adj.fdr)

zicoseq_out %>%
  filter(p.adj.fdr < 0.05) %>%
  head() %>%
  knitr::kable()
```


```{r ZicoSeqplot, eval=TRUE}
## x-axis is the effect size: R2 * direction of coefficient
ZicoSeq.plot(ZicoSeq.obj = zicoseq.obj,
             meta.dat = meta.dat,
             pvalue.type ='p.adj.fdr')
```


### Comparison of methods

The different methods yield slightly different results but they could
be also expected to overlap to a substantial degree. As an exercirse,
you can compare the outcomes between the different methods in terms of
effect sizes, significances, or other aspects that are comparable
between the methods.


## Confounding variables

Confounders are common in experimental studies. Generally, they can be
classified into 3 groups:

- Biological confounders, such as age, sex, etc. 

- Technical confounders produced by data collection, storage, DNA
  extraction, sequencing process, etc.

- Confounders resulting from experimental models, such as cage effect,
  sample background, etc.

Adjusting confounders is an important practice to reach an unbiased
conclusion. To perform causal inference, it is crucial that the method
is able to include covariates in the model. This is not possible with
statistical tests of general use, such as the Wilcoxon test. In contrast,
methods that target DAA, such as those described in this chapter, allow controlling for confounders. Below we show how to include a confounders in ANCOMBC, LinDA and ZicoSeq.


### ANCOMBC

```{r ancombc_adj, eval=TRUE}
# perform the analysis 
ancombc_cov <- ancombc2(tse,
                        assay_name = "counts",
                        fix_formula = "patient_status + cohort",
                        p_adj_method = "fdr",
                        lib_cut = 0,
                        group = "patient_status", 
                        struc_zero = TRUE, 
                        neg_lb = TRUE,
                        alpha = 0.05,
                        # multi group comparison is deactivated automatically
                        global = TRUE)

# now the model answers the question: holding cohort constant, are bacterial
# taxa differentially abundant? Or holding patient status constant, is cohort
# associated with bacterial abundance?
# Again we only show the first 6 entries.
```

```{r ancombc_adj2, eval=TRUE}
ancombc_cov$res %>%
  dplyr::select(starts_with(c("taxon", "lfc", "q"))) %>%
  head() %>%
  knitr::kable()
```

### LinDA


```{r linda_adj, eval=TRUE}
linda_cov <- linda(as.data.frame(assay(tse, "counts")),
                   as.data.frame(colData(tse)),
                   formula = "~ patient_status + cohort",
                   alpha = 0.05,
                   prev.filter = 0,
                   mean.abund.filter = 0)

linda.res <- linda_cov$output$patient_statusControl
```

```{r linda_adj2, eval=TRUE}
linda.res %>%
  filter(reject) %>%
  head() %>%
  knitr::kable()
```



### ZicoSeq

```{r ZicoSeq_adj, eval=TRUE}
set.seed(123)
zicoseq.obj <- ZicoSeq(feature.dat = as.matrix(assay(tse)),
                       meta.dat = as.data.frame(colData(tse)),
                       grp.name = "patient_status",
                       adj.name = "cohort", 
                       feature.dat.type = "count",
                       return.feature.dat = TRUE,
                       prev.filter = 0,
                       mean.abund.filter = 0,
                       max.abund.filter = 0,
                       perm.no = 999)

zicoseq_out <- cbind.data.frame(p.raw = zicoseq.obj$p.raw,
                                p.adj.fdr = zicoseq.obj$p.adj.fdr) 
```

```{r ZicoSeq_adj2, eval=TRUE}
zicoseq_out %>%
  filter(p.adj.fdr < 0.05) %>%
  head() %>%
  knitr::kable()
```



## Tree-based methods

Let us next cover phylogeny-aware methods to perform group-wise
associations.

### Group-wise associations testing based on balances

For testing associations based on balances, check the philr
R/Bioconductor package.



## Additional resources

DAA can be performed by several means. Although most of them provide similar functionality, some may be more suitable than others given a certain study
design or data type. Commonly used DAA tools include:

- ALDEx2 [@Gloor2016]
- ANCOM [@Mandal2015]
- ANCOMBC [@ancombc2020]
- corncob [@Martin2021]
- DACOMP [@Brill2019]
- DESeq2 [@Love2014]
- eBay [@Liu2020]
- edgeR [@Chen2016]
- fastANCOM [@fastANCOM2022]
- LDM [@Hu2020]
- lefser [@Khlebrodova2021]
- limma [@Ritchie2015]
- LinDA [@Zhou2022]
- MaAsLin2 [@Mallick2020]
- metagenomeSeq [@Paulson2017]
- Omnibus [@Omnibus2018]
- RAIDA [@Sohn2015]
- t-test
- Wilcoxon test
- ZicoSeq [@Yang2022]
- ZINQ [@Ling2021]


