# (PART) Appendix {-}

# Appendix {#appendix}

```{r setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble()
```


## Acknowledgments

This work would not have been possible without the countless
contributions and interactions within the broader research
community. We express our gratitude to the entire Bioconductor
community for developing this high-quality open research software
repository for life science analytics, continuously pushing the limits
in emerging application fields [@Gentleman2004, @Huber2015].

The base ecosystem of data containers, packages, and tutorials was set
up as a collaborative effort by Tuomas Borman, Henrik Eckermann,
Chouaib Benchraka, Chandler Ross, Shigdel Rajesh, Yağmur Şimşek,
Giulio Benedetti, Sudarshan Shetty, Felix Ernst, and [Leo
Lahti](http://www.iki.fi/Leo.Lahti), with further support from the
COST Action network on Statistical and Machine Learning Techniques for
Human Microbiome Studies
([ML4microbiome](https://www.ml4microbiome.eu/))
[@MorenoIndias2021]. The framework is based on the
_TreeSummarizedExperiment_ data container created by Ruizhu Huang and
others [@R-TreeSummarizedExperiment]. The idea of using this container
as a new basis for microbiome data science was initially advanced by
the groundwork of Domenick Braccia, Héctor Corrada Bravo and others,
and brought together with other microbiome data science developers
[@Shetty2019].

Ample demonstration data resources have been made available as the
[curatedMetagenomicData](https://waldronlab.io/curatedMetagenomicData/)
project by Edoardo Pasolli, Lucas Schiffer, Levi Waldron and others
[@Pasolli2017] adding important early support for the emerging
framework. A number of other contributors have advanced the ecosystem
further, and will be acknowledged in the individual
packages, [pull
requests](https://github.com/microbiome/OMA/graphs/contributors),
[issues](https://github.com/microbiome/OMA/issues), and other work.

The work has drawn inspiration from many sources, most notably from
the work on _phyloseq_ by Paul McMurdie and Susan Holmes
[@McMurdie2013] who pioneered the work on rigorous and reproducible
microbiome data science ecosystems in R/Bioconductor. The phyloseq
framework continues to provide a strong complementary set of packages
and methods for microbiome studies, and we do our best to support full
interoperability.

The open source books by Susan Holmes and Wolfgang Huber, Modern
Statistics for Modern Biology [@Holmes2019] and by Garret Grolemund
and Hadley Wickham, the R for Data Science [@Grolemund2017], and
Richard McElreath's Statistical Rethinking and the associated online
resources by Solomon Kurz [@McElreath2020] are key references that
advanced reproducible data science training and dissemination. The
Orchestrating Single-Cell Analysis with Bioconductor, or _OSCA_ book
by Robert Amezquita, Aaron Lun, Stephanie Hicks, and Raphael Gottardo
[@Amezquita2020] has implemented closely related work on the
_SummarizedExperiment_ data container and its derivatives in the field
of single cell sequencing studies. Many approaches used in this book
have been derived from the [OSCA
framework](https://bioconductor.org/books/release/OSCA/), with various
adjustments and extensions dedicated to microbiome data science.





