# Developers {-}

```{r setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble()
```

### Core team {-}

Contributions to this Gitbook from the various developers are
coordinated by:

- *Leo Lahti, DSc*, professor in Data Science at the [Department of
   Computing, University of Turku,
   Finland](https://datascience.utu.fi/), with a focus on
   computational microbiome analysis. Lahti obtained doctoral degree
   (DSc) from Aalto University in Finland (2010), developing
   probabilistic machine learning with applications to high-throughput
   life science data integration. Since then he has focused on
   microbiome research and developed, for instance, the
   _phyloseq_-based [microbiome R
   package](https://bioconductor.org/packages/release/bioc/html/microbiome.html)
   before starting to develop the _TreeSummarizedExperiment_ /
   _MultiAssayExperiment_ framework and the mia family of Bioconductor
   packages for microbiome data science introduced in this
   gitbook. Lahti led the development of [national policy on open
   access to research methods in
   Finland](https://avointiede.fi/en/policies-materials/policies-open-science-and-research-finland/policy-open-research-data-and-methods).
   He is current member in the [Bioconductor Community Advisory
   Board](https://bioconductor.org/about/community-advisory-board/)
   and runs regular training workshops in microbiome data science.

- *Tuomas Borman*, PhD researcher and the lead developer of OMA/mia at
   the Department of Computing, University of Turku. 


### Contributors {-}

This work is a remarkably collaborative effort. The full list of
contributors is available via
[Github](https://github.com/microbiome/OMA/graphs/contributors). Some
key authors/contributors include:

- *Felix Ernst, PhD*, among the first developers of R/Bioc methods for
   microbiome research based on the _SummarizedExperiment_ class and
   its derivatives.

- *Giulio Benedetti*, scientific programmer at the Department of
   Computing, University of Turku. His research interest is mostly
   related to Data Science. He has also helped to expand the
   SummarizedExperiment-based microbiome analysis framework to the
   Julia language, implementing
   [MicrobiomeAnalysis.jl](https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl).

- *Sudarshan Shetty, PhD* has supported the establishment of the
   framework and associated tools. He also maintains a list of
   [microbiome R
   packages](https://microsud.github.io/Tools-Microbiome-Analysis/).

- *Henrik Eckermann*, in particular to the development of
   the differential abundance analyses

- *Chouaib Benchraka* provided various contributions to the package ecosystem and the OMA book

- *Yağmur Şimşek* converted the miaSim R package to support the Bioconductor framework

- *Basil Courbayre* provided various contributions to the package
   ecosystem and the OMA book, in particular on unsupervised machine
   learning

- *Matti Ruuskanen, PhD*, added machine learning techniques for
   microbiome analysis
   
- *Stefanie Peschel* has contributed chapters on the construction, analysis, and comparison of microbial association networks.

- *Christian L. Müller*, group leader at the [Computational Health Center, Helmholtz Zentrum München, Germany](https://www.helmholtz-munich.de/en/computational-health-center) and a Professor for Biomedical Statistics and Data Science at [LMU Munich](https://www.en.statistik.uni-muenchen.de/index.html). He assisted in writing the chapters on network learning and comparison.

- *Shigdel Rajesh, PhD* 

- *Artur Sannikov* 

- *Jeba Akewak* 

- *Himmi Lindgren*

- *Lu Yang* 



### Acknowledgments {-}

This work would not have been possible without the countless
contributions and interactions with other researchers, developers, and
users. We express our gratitude to the entire Bioconductor community
for developing this high-quality open research software repository for
life science analytics, continuously pushing the limits in emerging
fields [@Gentleman2004], [@Huber2015].

The presented framework for microbiome data science is based on the
_TreeSummarizedExperiment_ data container created by Ruizhu Huang and
others [@R_TreeSummarizedExperiment], [@Ernst2020], and on the
_MultiAssayExperiment_ by Marcel Ramos et al. [@Ramos2017]. The idea
of using these containers as a basis for microbiome data science was
initially advanced by the groundwork of Domenick Braccia, Héctor
Corrada Bravo and others and brought together with other microbiome
data science developers [@Shetty2019]. Setting up the base ecosystem
of packages and tutorials was then subsequently led by Tuomas Borman,
Felix Ernst, and [Leo Lahti](http://www.iki.fi/Leo.Lahti). We would
specifically like to thank everyone who contributed to the work
supporting the _TreeSummarizedExperiment_ ecosystem for microbiome
research, including but not limited to the R packages mia, miaViz,
miaTime, miaSim, philr, ANCOMBC, curatedMetagenomicData, scater,
scuttle, and other packages, some of which are listed in Section
\@ref(ecosystem). A number of other contributors have advanced the
ecosystem further, and will be acknowledged in the individual
packages, [pull
requests](https://github.com/microbiome/OMA/graphs/contributors),
[issues](https://github.com/microbiome/OMA/issues), and other work.

Ample demonstration data resources supporting this framework have been
made available through the
[curatedMetagenomicData](https://waldronlab.io/curatedMetagenomicData/)
project by Edoardo Pasolli, Lucas Schiffer, Levi Waldron and others
[@Pasolli2017].

The work has drawn initial inspiration from many sources, most notably
from the work on _phyloseq_ by Paul McMurdie and Susan Holmes
[@McMurdie2013] who pioneered the work on rigorous and reproducible
microbiome data science ecosystems in R/Bioconductor. The phyloseq
framework continues to provide a vast array of complementary packages
and methods for microbiome studies. The Orchestrating Single-Cell
Analysis with Bioconductor, or
[_OSCA_](https://bioconductor.org/books/release/OSCA/) book by Robert
Amezquita, Aaron Lun, Stephanie Hicks, and Raphael Gottardo
[@Amezquita2020natmeth] has implemented closely related work on the
_SummarizedExperiment_ data container and its derivatives in the field
of single cell sequencing studies that have inspired this work.

In the background, the open source books by Susan Holmes and Wolfgang
Huber, Modern Statistics for Modern Biology [@Holmes2019] and by
Garret Grolemund and Hadley Wickham, the R for Data Science
[@Grolemund2017], and Richard McElreath's Statistical Rethinking and
the associated online resources by Solomon Kurz [@McElreath2020] are
key references that have advanced reproducible data science training
and dissemination.

### How to contribute

To contribute reports, follow the Git flow procedure (you can see instructions
to [getting started with Github](https://docs.github.com/en/get-started)):

1. Fork the project
2. Clone your fork
3. Modify the material
4. Check locally that the changes render successfully (see above)
5. Add and commit the changes to your fork
6. Create a pull request (PR) from your fork back to the original repo
7. Fix and discuss issues in the review process

More detailed instructions for contributing can be found on OMA [README](https://github.com/microbiome/OMA/blob/master/README.md).

### Support {-}

This work has been supported by:

* [Research Council of Finland](https://www.aka.fi/)

* [FindingPheno](https://www.findingpheno.eu/) European Union’s Horizon 2020 research and innovation programme under grant agreement No 952914

* COST Action network on Statistical and Machine Learning Techniques for Human Microbiome Studies
  ([ML4microbiome](https://www.ml4microbiome.eu/)) [@MorenoIndias2021].

* Computational Life Science Research Program, [Biocity Turku](https://biocityturku.fi/)

* [Turku University Foundation](https://www.yliopistosaatio.fi/en/)

