# Extras {#extras}

## Interactive 3D Plots

```{r, message=FALSE, warning=FALSE}
# Installing required packages
if (!require(rgl)){
  BiocManager::install("rgl")  
}
if (!require(plotly)){
  BiocManager::install("plotly")  
}
```

```{r, setup, warning=FALSE, message=FALSE}
library(knitr)
library(rgl)
knitr::knit_hooks$set(webgl = hook_webgl)
```


In this section we make a 3D version of the earlier [ Visualizing the most dominant genus on PCoA](https://microbiome.github.io/OMA/microbiome-exploration.html#visualizing-the-most-dominant-genus-on-pcoa), with the help of the [plotly](https://plotly.com/r/) R package.

```{r, message=FALSE, warning=FALSE}
# Installing the package
if (!require(curatedMetagenomicData)){
  BiocManager::install("curatedMetagenomicData")  
}
# Importing necessary libraries
library(curatedMetagenomicData)
library(dplyr)
library(DT)
library(mia)
library(scater)

# Querying the data
tse <- sampleMetadata %>%
    filter(age >= 18) %>% # taking only data of age 18 or above
    filter(!is.na(alcohol)) %>% # excluding missing values
    returnSamples("relative_abundance")

tse_Genus <- agglomerateByRank(tse, rank="Genus")
tse_Genus <- addPerSampleDominantTaxa(tse_Genus,abund_values="relative_abundance", name = "dominant_taxa")

# Performing PCoA with Bray-Curtis dissimilarity.
tse_Genus <- runMDS(tse_Genus, FUN = vegan::vegdist, ncomponents = 3,
              name = "PCoA_BC", exprs_values = "relative_abundance")

# Getting the 6 top taxa
top_taxa <- getTopTaxa(tse_Genus,top = 6, abund_values = "relative_abundance")

# Naming all the rest of non top-taxa as "Other"
most_abundant <- lapply(colData(tse_Genus)$dominant_taxa,
                   function(x){if (x %in% top_taxa) {x} else {"Other"}})

# Storing the previous results as a new column within colData
colData(tse_Genus)$most_abundant <- as.character(most_abundant)

# Calculating percentage of the most abundant
most_abundant_freq <- table(as.character(most_abundant))
most_abundant_percent <- round(most_abundant_freq/sum(most_abundant_freq)*100, 1)

# Retrieving the explained variance
e <- attr(reducedDim(tse_Genus, "PCoA_BC"), "eig");
var_explained <- e/sum(e[e>0])*100
```

Interactive 3D visualization of the most dominant genus on PCoA.
Note that labels at legend can be used to visualize one or more Genus separately (double click to isolate one from the others, or toggle to select multiple ones).

```{r, test-rgl,webgl=TRUE, warning=FALSE, message=FALSE}
library(plotly)

# 3D Visualization
reduced_data  <- as.data.frame(reducedDim(tse_Genus)[,])
names(reduced_data) <- c("PC1","PC2","PC3")
plot_ly(reduced_data, x=~PC1,y=~PC2,z=~PC3)%>%
  add_markers(color=colData(tse_Genus)$most_abundant, size=5,
              colors=c("black", "blue", "lightblue", "darkgray", "magenta", "darkgreen", "red")) %>%
  layout(scene=list(xaxis=list(title = paste("PC1 (",round(var_explained[1],1),"%)")),
                    yaxis=list(title = paste("PC2 (",round(var_explained[2],1),"%)")),
                    zaxis=list(title = paste("PC3 (",round(var_explained[3],1),"%)"))))

```

## Speed comparisons between `tse` and `pseq`containers

### Methods under inspection

Tree summarized experiment objects (tse) and phyloseq (pseq) objects can be viewed as two distinct containers with the same content organized in two different ways. Consequently, these two classes of objects do not perform equally when subject to analogous computational tasks. While the time execution might not diverge significantly in the case of smaller data sets, it could be of matter for larger data sets.

To estimate the time efficiency of homologous tasks but with different targets (`tse` or `pseq`), the following working routines are performed on data sets of various size and compared in terms of the targeted object class:

* Melting
* Transformation
* Agglomeration
* Alpha diversity estimation
* Beta diversity estimation

The analyzed data sets include:

* AsnicarF_2017.relative_abundance
* GlobalPatterns
* VincentC_2016
* SilvermanAGutData
* SongQAData
* SprockettTHData

```{r, message=FALSE, warning=FALSE}
# load libraries
library(ggplot2)
library(dplyr)
```

```{r}
# load data
execution_times <- read.csv("execution_times.csv")
```

```{r}
# filter data
execution_times <- execution_times %>% filter(!is.na(Samples)) 
execution_times
```

### Execution time of melting

```{r}
# compare melting
ggplot(execution_times, aes(x = Samples, y = Melt, color = MeltCommand, shape = as.factor(Features))) + 
  geom_point() +
  labs(title = "Melting comparison", x = "Number of samples", y = "Execution time in s", color = "method:", shape = "number of features:", caption = "Execution time of melting as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare melting
ggplot(execution_times, aes(x = Features, y = Melt, color = MeltCommand, shape = as.factor(Samples))) + 
  geom_point() +
  labs(title = "Melting comparison", x = "Number of features", y = "Execution time in s", color = "method:", shape = "number of samples:", caption = "Execution time of melting as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

### Execution time of transformation

```{r}
# compare transformation in terms of samples
ggplot(execution_times, aes(x = Samples, y = Transform, color = TransformCommand, shape = as.factor(Features))) + 
  geom_point() +
  labs(title = "Transformation comparison", x = "Number of samples", y = "Execution time in s", color = "method:", shape = "number of features:", caption = "Execution time of transformation as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare transformation in terms of features
ggplot(execution_times, aes(x = Features, y = Transform, color = TransformCommand, shape = as.factor(Samples))) + 
  geom_point() +
  labs(title = "Transformation comparison", x = "Number of features", y = "Execution time in s", color = "method:", shape = "number of samples:", caption = "Execution time of transformation as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

### Execution time of agglomeration

```{r}
# compare agglomeration in terms of samples
ggplot(execution_times, aes(x = Samples, y = Agglomerate, color = AgglomerateCommand, shape = as.factor(Features))) + 
  geom_point() +
  labs(title = "Agglomeration comparison", x = "Number of samples", y = "Execution time in s", color = "method:", shape = "number of features:", caption = "Execution time of agglomeration as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare agglomeration in terms of features
ggplot(execution_times, aes(x = Features, y = Agglomerate, color = AgglomerateCommand, shape = as.factor(Samples))) + 
  geom_point() +
  labs(title = "Agglomeration comparison", x = "Number of features", y = "Execution time in s", color = "method:", shape = "number of samples:", caption = "Execution time of agglomeration as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

### Execution time of alpha diversity estimation

```{r}
# compare estimation of alpha diversity in terms of samples
ggplot(execution_times, aes(x = Samples, y = AlphaEstimation, color = AlphaCommand, shape = as.factor(Features))) + 
  geom_point() +
  labs(title = "Alpha Diversity Estimation comparison", x = "Number of samples", y = "Execution time in s", color = "method:", shape = "number of features:", caption = "Execution time of estimating alpha diversity as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare estimation of alpha diversity in terms of features
ggplot(execution_times, aes(x = Features, y = AlphaEstimation, color = AlphaCommand, shape = as.factor(Samples))) + 
  geom_point() +
  labs(title = "Alpha Diversity Estimation comparison", x = "Number of features", y = "Execution time in s", color = "method:", shape = "number of samples:", caption = "Execution time of estimating alpha diversity as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

### Execution time of beta diversity estimation

```{r}
# compare estimation of beta diversity in terms of samples
ggplot(execution_times, aes(x = Samples, y = BetaEstimation, color = BetaCommand, shape = as.factor(Features))) + 
  geom_point() +
  labs(title = "Beta Diversity Estimation comparison", x = "Number of samples", y = "Execution time in s", color = "method:", shape = "number of features:", caption = "Execution time of estimating beta diversity as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare estimation of beta diversity in terms of features
ggplot(execution_times, aes(x = Features, y = BetaEstimation, color = BetaCommand, shape = as.factor(Samples))) + 
  geom_point() +
  labs(title = "Beta Diversity Estimation comparison", x = "Number of features", y = "Execution time in s", color = "method:", shape = "number of samples:", caption = "Execution time of estimating beta diversity as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
mean_time <- execution_times %>% group_by(ObjectType) %>% summarize(mean_melt = mean(Melt), mean_transform = mean(Transform), mean_agglomerate = mean(Agglomerate), mean_alpha = mean(AlphaEstimation), mean_beta = mean(BetaEstimation, na.rm = TRUE))
mean_time
```
