# Extras {#extras}

## Interactive 3D Plots

```{r, message=FALSE, warning=FALSE}
# Installing required packages
if (!require(rgl)){
  BiocManager::install("rgl")  
}
if (!require(plotly)){
  BiocManager::install("plotly")  
}
```

```{r, setup, warning=FALSE, message=FALSE}
library(knitr)
library(rgl)
knitr::knit_hooks$set(webgl = hook_webgl)
```


In this section we make a 3D version of the earlier [ Visualizing the most dominant genus on PCoA](https://microbiome.github.io/OMA/microbiome-exploration.html#visualizing-the-most-dominant-genus-on-pcoa), with the help of the [plotly](https://plotly.com/r/) R package.

```{r, message=FALSE, warning=FALSE}
# Installing the package
if (!require(curatedMetagenomicData)){
  BiocManager::install("curatedMetagenomicData")  
}
# Importing necessary libraries
library(curatedMetagenomicData)
library(dplyr)
library(DT)
library(mia)
library(scater)

# Querying the data
tse <- sampleMetadata %>%
    filter(age >= 18) %>% # taking only data of age 18 or above
    filter(!is.na(alcohol)) %>% # excluding missing values
    returnSamples("relative_abundance")

tse_Genus <- agglomerateByRank(tse, rank="Genus")
tse_Genus <- addPerSampleDominantTaxa(tse_Genus,abund_values="relative_abundance", name = "dominant_taxa")

# Performing PCoA with Bray-Curtis dissimilarity.
tse_Genus <- runMDS(tse_Genus, FUN = vegan::vegdist, ncomponents = 3,
              name = "PCoA_BC", exprs_values = "relative_abundance")

# Getting the 6 top taxa
top_taxa <- getTopTaxa(tse_Genus,top = 6, abund_values = "relative_abundance")

# Naming all the rest of non top-taxa as "Other"
most_abundant <- lapply(colData(tse_Genus)$dominant_taxa,
                   function(x){if (x %in% top_taxa) {x} else {"Other"}})

# Storing the previous results as a new column within colData
colData(tse_Genus)$most_abundant <- as.character(most_abundant)

# Calculating percentage of the most abundant
most_abundant_freq <- table(as.character(most_abundant))
most_abundant_percent <- round(most_abundant_freq/sum(most_abundant_freq)*100, 1)

# Retrieving the explained variance
e <- attr(reducedDim(tse_Genus, "PCoA_BC"), "eig");
var_explained <- e/sum(e[e>0])*100
```

Interactive 3D visualization of the most dominant genus on PCoA.
Note that labels at legend can be used to visualize one or more Genus separately (double click to isolate one from the others, or toggle to select multiple ones).

```{r, test-rgl,webgl=TRUE, warning=FALSE, message=FALSE}
library(plotly)

# 3D Visualization
reduced_data  <- as.data.frame(reducedDim(tse_Genus)[,])
names(reduced_data) <- c("PC1","PC2","PC3")
plot_ly(reduced_data, x=~PC1,y=~PC2,z=~PC3)%>%
  add_markers(color=colData(tse_Genus)$most_abundant, size=5,
              colors=c("black", "blue", "lightblue", "darkgray", "magenta", "darkgreen", "red")) %>%
  layout(scene=list(xaxis=list(title = paste("PC1 (",round(var_explained[1],1),"%)")),
                    yaxis=list(title = paste("PC2 (",round(var_explained[2],1),"%)")),
                    zaxis=list(title = paste("PC3 (",round(var_explained[3],1),"%)"))))

```

## Speed comparisons between `tse`and `pseq`containers

### Methods under inspection

```{r, message=FALSE, warning=FALSE}
# load libraries
library(ggplot2)
library(dplyr)
```

```{r}
# load data
execution_times <- read.csv("data/execution_times.csv")
```

```{r}
# filter data
execution_times <- execution_times %>% filter(!is.na(samples)) 
```

### Execution time as a function of number of samples

```{r}
# compare melting
ggplot(execution_times, aes(x = samples, y = melt, color = melt_command)) + 
  geom_point() +
  labs(title = "Melting comparison", x = "Number of samples", y = "Execution time in s", color = "melt command", caption = "Execution time of melting as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare transform
ggplot(execution_times, aes(x = samples, y = transform, color = transform_command)) + 
  geom_point() +
  labs(title = "Transformation comparison", x = "Number of samples", y = "Execution time in s", color = "transform command", caption = "Execution time of transformation as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare agglomerate
ggplot(execution_times, aes(x = samples, y = agglomerate, color = agglomerate_command)) + 
  geom_point() +
  labs(title = "Agglomeration comparison", x = "Number of samples", y = "Execution time in s", color = "agglomerate method", caption = "Execution time of agglomeration as a function of number of samples") +
  theme_classic() +
  theme(legend.position = "bottom")
```

### Execution time as a function of number of features

```{r}
# compare melting
ggplot(execution_times, aes(x = features, y = melt, color = melt_command)) + 
  geom_point() +
  labs(title = "Melting comparison", x = "Number of features", y = "Execution time in s", color = "melt command", caption = "Execution time of melting as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare transform
ggplot(execution_times, aes(x = features, y = transform, color = transform_command)) + 
  geom_point() +
  labs(title = "Transformation comparison", x = "Number of features", y = "Execution time in s", color = "transform command", caption = "Execution time of transformation as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```

```{r}
# compare agglomerate
ggplot(execution_times, aes(x = features, y = agglomerate, color = agglomerate_command)) + 
  geom_point() +
  labs(title = "Agglomeration comparison", x = "Number of features", y = "Execution time in s", color = "agglomerate method", caption = "Execution time of agglomeration as a function of number of features") +
  theme_classic() +
  theme(legend.position = "bottom")
```
