<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Biclustering | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 11 Biclustering | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Biclustering | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Biclustering | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="community-typing.html"/>
<link rel="next" href="differential-abundance.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.2.2</b> colData</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.2.3</b> rowData</a></li>
<li class="chapter" data-level="2.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.2.4</b> rowTree</a></li>
<li class="chapter" data-level="2.2.5" data-path="containers.html"><a href="containers.html#alternative-experiments"><i class="fa fa-check"></i><b>2.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.6" data-path="containers.html"><a href="containers.html#multiassayexperiments"><i class="fa fa-check"></i><b>2.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>2.3.1</b> 16S workflow</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.2</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.3" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.3</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#some-available-packages"><i class="fa fa-check"></i><b>3.2</b> Some available packages</a></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
<li class="chapter" data-level="4.1.4" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.1.4</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
<li class="chapter" data-level="" data-path="community-diversity.html"><a href="community-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#community-comparisons-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.2</b> Community comparisons by beta diversity analysis</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="community-similarity.html"><a href="community-similarity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.4.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.4.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.4.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="community-similarity.html"><a href="community-similarity.html#further-reading"><i class="fa fa-check"></i><b>8.5</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="community-similarity.html"><a href="community-similarity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>10</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.1</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.1.1</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>10.2</b> Community Detection</a></li>
<li class="chapter" data-level="10.3" data-path="community-typing.html"><a href="community-typing.html#additional-community-typing"><i class="fa fa-check"></i><b>10.3</b> Additional Community Typing</a></li>
<li class="chapter" data-level="" data-path="community-typing.html"><a href="community-typing.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="biclustering.html"><a href="biclustering.html"><i class="fa fa-check"></i><b>11</b> Biclustering</a>
<ul>
<li class="chapter" data-level="11.1" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>11.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="11.2" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>11.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="11.3" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>11.3</b> Taxa vs taxa</a></li>
<li class="chapter" data-level="" data-path="biclustering.html"><a href="biclustering.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>12</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="12.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>12.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>12.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="12.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>12.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="12.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>12.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="12.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>12.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="12.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>12.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="12.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>12.1.6</b> Comparison of the methods</a></li>
<li class="chapter" data-level="12.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>12.1.7</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>12.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>12.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#multi-omics-factor-analysis"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
<li class="chapter" data-level="" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>14</b> Visualization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>14.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>14.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="14.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>14.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>14.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>14.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="14.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>14.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>14.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>14.3.1</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="viz-chapter.html"><a href="viz-chapter.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="15" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>15</b> Resources</a>
<ul>
<li class="chapter" data-level="15.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>15.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="15.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>15.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="15.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>15.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="15.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>15.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="15.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>15.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="15.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>15.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>15.2.1</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="resources.html"><a href="resources.html#reproducible-reporting-with-rmarkdown"><i class="fa fa-check"></i><b>15.3</b> Reproducible reporting with Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>16</b> Extra material</a>
<ul>
<li class="chapter" data-level="16.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>16.1</b> Interactive 3D Plots</a></li>
<li class="chapter" data-level="16.2" data-path="extras.html"><a href="extras.html#permanova-comparison"><i class="fa fa-check"></i><b>16.2</b> PERMANOVA comparison</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>17</b> Acknowledgments</a></li>
<li class="chapter" data-level="18" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>18</b> Exercises</a>
<ul>
<li class="chapter" data-level="18.1" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>18.1</b> Workflows</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting"><i class="fa fa-check"></i><b>18.1.1</b> Reproducible reporting</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="exercises.html"><a href="exercises.html#data-containers-2"><i class="fa fa-check"></i><b>18.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-data-exploration"><i class="fa fa-check"></i><b>18.2.1</b> TreeSummarizedExperiment data exploration</a></li>
<li class="chapter" data-level="18.2.2" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-treese-data-import"><i class="fa fa-check"></i><b>18.2.2</b> TreeSummarizedExperiment (TreeSE) data import</a></li>
<li class="chapter" data-level="18.2.3" data-path="exercises.html"><a href="exercises.html#introduction-to-multiassayexperient-mae"><i class="fa fa-check"></i><b>18.2.3</b> Introduction to MultiAssayExperient (MAE)</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>18.3</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-features"><i class="fa fa-check"></i><b>18.3.1</b> Prevalent and core features</a></li>
<li class="chapter" data-level="18.3.2" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>18.3.2</b> Taxonomic levels</a></li>
</ul></li>
<li class="chapter" data-level="18.4" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>18.4</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="18.4.1" data-path="exercises.html"><a href="exercises.html#univariate-analyses"><i class="fa fa-check"></i><b>18.4.1</b> Univariate analyses</a></li>
<li class="chapter" data-level="18.4.2" data-path="exercises.html"><a href="exercises.html#differential-abundance-analysis-1"><i class="fa fa-check"></i><b>18.4.2</b> Differential abundance analysis</a></li>
</ul></li>
<li class="chapter" data-level="18.5" data-path="exercises.html"><a href="exercises.html#alpha-diversity"><i class="fa fa-check"></i><b>18.5</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="18.5.1" data-path="exercises.html"><a href="exercises.html#alpha-diversity-basics"><i class="fa fa-check"></i><b>18.5.1</b> Alpha diversity basics</a></li>
<li class="chapter" data-level="18.5.2" data-path="exercises.html"><a href="exercises.html#alpha-diversity-extra"><i class="fa fa-check"></i><b>18.5.2</b> Alpha diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="18.6" data-path="exercises.html"><a href="exercises.html#community-composition"><i class="fa fa-check"></i><b>18.6</b> Community composition</a>
<ul>
<li class="chapter" data-level="18.6.1" data-path="exercises.html"><a href="exercises.html#beta-diversity-basics"><i class="fa fa-check"></i><b>18.6.1</b> Beta diversity basics</a></li>
<li class="chapter" data-level="18.6.2" data-path="exercises.html"><a href="exercises.html#beta-diversity-extra"><i class="fa fa-check"></i><b>18.6.2</b> Beta diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="18.7" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>18.7</b> Visualization</a>
<ul>
<li class="chapter" data-level="18.7.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>18.7.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="18.7.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>18.7.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="18.8" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>18.8</b> Multiomics</a>
<ul>
<li class="chapter" data-level="18.8.1" data-path="exercises.html"><a href="exercises.html#introduction-to-multiomics"><i class="fa fa-check"></i><b>18.8.1</b> Introduction to multiomics</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biclustering" class="section level1 hasAnchor" number="11">
<h1><span class="header-section-number">Chapter 11</span> Biclustering<a href="biclustering.html#biclustering" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Biclustering methods cluster rows and columns simultaneously in order
to find subsets of correlated features/samples.</p>
<p>Here, we use following packages:
- <a href="https://cran.r-project.org/web/packages/biclust/index.html">biclust</a>
- <a href="https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13582">cobiclust</a></p>
<p><em>cobiclust</em> is especially developed for microbiome data whereas <em>biclust</em> is more
general method. In this section, we show three different cases and example
solutions to apply biclustering to them.</p>
<ol style="list-style-type: decimal">
<li>Taxa vs samples</li>
<li>Taxa vs biomolecule/biomarker</li>
<li>Taxa vs taxa</li>
</ol>
<p>Biclusters can be visualized using heatmap or boxplot, for instance. For checking purposes,
also scatter plot might be valid choice.</p>
<p>Check more ideas for heatmaps from chapters <a href="viz-chapter.html#viz-chapter">14</a> and @ref(microbiome-community.</p>
<div id="taxa-vs-samples" class="section level2 hasAnchor" number="11.1">
<h2><span class="header-section-number">11.1</span> Taxa vs samples<a href="biclustering.html#taxa-vs-samples" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>When you have microbial abundance matrices, we suggest to use <em>cobiclust</em> which is
designed for microbial data.</p>
<p>Load example data</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="biclustering.html#cb272-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb272-2"><a href="biclustering.html#cb272-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb272-3"><a href="biclustering.html#cb272-3" aria-hidden="true" tabindex="-1"></a>mae <span class="ot">&lt;-</span> <span class="fu">HintikkaXOData</span>()</span></code></pre></div>
<p>Only the most prevalent taxa are included in analysis.</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="biclustering.html#cb273-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data in the first experiment</span></span>
<span id="cb273-2"><a href="biclustering.html#cb273-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">subsetByPrevalentTaxa</span>(mae[[<span class="dv">1</span>]], <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">prevalence =</span> <span class="fl">0.2</span>, <span class="at">detection =</span> <span class="fl">0.001</span>)</span>
<span id="cb273-3"><a href="biclustering.html#cb273-3" aria-hidden="true" tabindex="-1"></a><span class="co"># clr-transform in the first experiment</span></span>
<span id="cb273-4"><a href="biclustering.html#cb273-4" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformSamples</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb273-5"><a href="biclustering.html#cb273-5" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformSamples</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;relabundance&quot;</span>, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>)</span></code></pre></div>
<p><em>cobiclust</em> takes counts table as an input and gives <em>cobiclust</em> object as an output.
It includes clusters for taxa and samples.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="biclustering.html#cb274-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(cobiclust)){</span>
<span id="cb274-2"><a href="biclustering.html#cb274-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;cobiclust&quot;</span>)</span>
<span id="cb274-3"><a href="biclustering.html#cb274-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(cobiclust)</span>
<span id="cb274-4"><a href="biclustering.html#cb274-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb274-5"><a href="biclustering.html#cb274-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-6"><a href="biclustering.html#cb274-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Do clustering; use counts tableÂ´</span></span>
<span id="cb274-7"><a href="biclustering.html#cb274-7" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">cobiclust</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;counts&quot;</span>))</span>
<span id="cb274-8"><a href="biclustering.html#cb274-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-9"><a href="biclustering.html#cb274-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Get clusters</span></span>
<span id="cb274-10"><a href="biclustering.html#cb274-10" aria-hidden="true" tabindex="-1"></a>row_clusters <span class="ot">&lt;-</span> clusters<span class="sc">$</span>classification<span class="sc">$</span>rowclass</span>
<span id="cb274-11"><a href="biclustering.html#cb274-11" aria-hidden="true" tabindex="-1"></a>col_clusters <span class="ot">&lt;-</span> clusters<span class="sc">$</span>classification<span class="sc">$</span>colclass</span>
<span id="cb274-12"><a href="biclustering.html#cb274-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-13"><a href="biclustering.html#cb274-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Add clusters to rowdata and coldata</span></span>
<span id="cb274-14"><a href="biclustering.html#cb274-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(row_clusters)</span>
<span id="cb274-15"><a href="biclustering.html#cb274-15" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(col_clusters)</span>
<span id="cb274-16"><a href="biclustering.html#cb274-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-17"><a href="biclustering.html#cb274-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Order data based on clusters</span></span>
<span id="cb274-18"><a href="biclustering.html#cb274-18" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="fu">order</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters), <span class="fu">order</span>(<span class="fu">colData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters)]</span>
<span id="cb274-19"><a href="biclustering.html#cb274-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-20"><a href="biclustering.html#cb274-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Print clusters</span></span>
<span id="cb274-21"><a href="biclustering.html#cb274-21" aria-hidden="true" tabindex="-1"></a>clusters<span class="sc">$</span>classification</span></code></pre></div>
<pre><code>## $rowclass
##  [1] 1 1 1 1 2 2 1 1 1 1 1 1 2 2 2 2 1 2 1 1 2 1 2 2 1 1 2 1 1 1 1 1 2 1 1 2 1 1
## [39] 1 1 1 1 1 1 1 1 1 2 1 2 1 1 1 2 1 1 1
## 
## $colclass
##  C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 C20 
##   1   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 
## C21 C22 C23 C24 C25 C26 C27 C28 C29 C30 C31 C32 C33 C34 C35 C36 C37 C38 C39 C40 
##   2   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   1</code></pre>
<p>Next we can plot clusters. Commonly used plot is heatmap with annotations.</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="biclustering.html#cb276-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(pheatmap)){</span>
<span id="cb276-2"><a href="biclustering.html#cb276-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;pheatmap&quot;</span>)</span>
<span id="cb276-3"><a href="biclustering.html#cb276-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pheatmap)</span>
<span id="cb276-4"><a href="biclustering.html#cb276-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb276-5"><a href="biclustering.html#cb276-5" aria-hidden="true" tabindex="-1"></a><span class="co"># z-transform for heatmap</span></span>
<span id="cb276-6"><a href="biclustering.html#cb276-6" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformFeatures</span>(mae[[<span class="dv">1</span>]], <span class="at">abund_values =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">name =</span> <span class="st">&quot;clr_z&quot;</span>)</span>
<span id="cb276-7"><a href="biclustering.html#cb276-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb276-8"><a href="biclustering.html#cb276-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create annotations. When column names are equal, they should share levels. </span></span>
<span id="cb276-9"><a href="biclustering.html#cb276-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Here samples include 3 clusters, and taxa 2. That is why we have to make </span></span>
<span id="cb276-10"><a href="biclustering.html#cb276-10" aria-hidden="true" tabindex="-1"></a><span class="co"># column names unique. </span></span>
<span id="cb276-11"><a href="biclustering.html#cb276-11" aria-hidden="true" tabindex="-1"></a>annotation_col <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(mae[[<span class="dv">1</span>]])[, <span class="st">&quot;clusters&quot;</span>, <span class="at">drop =</span> F])</span>
<span id="cb276-12"><a href="biclustering.html#cb276-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annotation_col) <span class="ot">&lt;-</span> <span class="st">&quot;col_clusters&quot;</span></span>
<span id="cb276-13"><a href="biclustering.html#cb276-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb276-14"><a href="biclustering.html#cb276-14" aria-hidden="true" tabindex="-1"></a>annotation_row <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])[, <span class="st">&quot;clusters&quot;</span>, <span class="at">drop =</span> F])</span>
<span id="cb276-15"><a href="biclustering.html#cb276-15" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annotation_row) <span class="ot">&lt;-</span> <span class="st">&quot;row_clusters&quot;</span></span>
<span id="cb276-16"><a href="biclustering.html#cb276-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb276-17"><a href="biclustering.html#cb276-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap</span></span>
<span id="cb276-18"><a href="biclustering.html#cb276-18" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;clr_z&quot;</span>), <span class="at">cluster_rows =</span> F, <span class="at">cluster_cols =</span> F, </span>
<span id="cb276-19"><a href="biclustering.html#cb276-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> annotation_col,</span>
<span id="cb276-20"><a href="biclustering.html#cb276-20" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> annotation_row)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/cobiclust_3-1.png" width="1344" /></p>
<p>Boxplot is commonly used to summarize the results:</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="biclustering.html#cb277-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(ggplot2)){</span>
<span id="cb277-2"><a href="biclustering.html#cb277-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb277-3"><a href="biclustering.html#cb277-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb277-4"><a href="biclustering.html#cb277-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb277-5"><a href="biclustering.html#cb277-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(patchwork)){</span>
<span id="cb277-6"><a href="biclustering.html#cb277-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;patchwork&quot;</span>)</span>
<span id="cb277-7"><a href="biclustering.html#cb277-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb277-8"><a href="biclustering.html#cb277-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb277-9"><a href="biclustering.html#cb277-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb277-10"><a href="biclustering.html#cb277-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot requires data in melted format</span></span>
<span id="cb277-11"><a href="biclustering.html#cb277-11" aria-hidden="true" tabindex="-1"></a>melt_assay <span class="ot">&lt;-</span> <span class="fu">meltAssay</span>(mae[[<span class="dv">1</span>]], <span class="at">abund_values =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">add_col_data =</span> T, <span class="at">add_row_data =</span> T)</span>
<span id="cb277-12"><a href="biclustering.html#cb277-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb277-13"><a href="biclustering.html#cb277-13" aria-hidden="true" tabindex="-1"></a><span class="co"># patchwork two plots side-by-side</span></span>
<span id="cb277-14"><a href="biclustering.html#cb277-14" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(melt_assay) <span class="sc">+</span></span>
<span id="cb277-15"><a href="biclustering.html#cb277-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> clusters.x, <span class="at">y =</span> clr)) <span class="sc">+</span></span>
<span id="cb277-16"><a href="biclustering.html#cb277-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Taxa clusters&quot;</span>)</span>
<span id="cb277-17"><a href="biclustering.html#cb277-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb277-18"><a href="biclustering.html#cb277-18" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(melt_assay) <span class="sc">+</span></span>
<span id="cb277-19"><a href="biclustering.html#cb277-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> clusters.y, <span class="at">y =</span> clr)) <span class="sc">+</span></span>
<span id="cb277-20"><a href="biclustering.html#cb277-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Sample clusters&quot;</span>)</span>
<span id="cb277-21"><a href="biclustering.html#cb277-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb277-22"><a href="biclustering.html#cb277-22" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/cobiclust_4-1.png" width="672" /></p>
</div>
<div id="taxa-vs-biomolecules" class="section level2 hasAnchor" number="11.2">
<h2><span class="header-section-number">11.2</span> Taxa vs biomolecules<a href="biclustering.html#taxa-vs-biomolecules" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here, we analyze cross-correlation between taxa and metabolites. This is a case, where
we use <em>biclust</em> method which is suitable for numeric matrices in general.</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="biclustering.html#cb278-1" aria-hidden="true" tabindex="-1"></a><span class="do">############################ REMOVE THIS WHEN </span><span class="al">BUG</span><span class="do"> IN GETEXPERIMENTCROSSCORR IS FIXED</span></span>
<span id="cb278-2"><a href="biclustering.html#cb278-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(<span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]))</span>
<span id="cb278-3"><a href="biclustering.html#cb278-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb278-4"><a href="biclustering.html#cb278-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate correlations</span></span>
<span id="cb278-5"><a href="biclustering.html#cb278-5" aria-hidden="true" tabindex="-1"></a>corr <span class="ot">&lt;-</span> <span class="fu">getExperimentCrossCorrelation</span>(mae, <span class="dv">1</span>, <span class="dv">2</span>, </span>
<span id="cb278-6"><a href="biclustering.html#cb278-6" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">abund_values1 =</span> <span class="st">&quot;clr&quot;</span>, </span>
<span id="cb278-7"><a href="biclustering.html#cb278-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">abund_values2 =</span> <span class="st">&quot;nmr&quot;</span>, </span>
<span id="cb278-8"><a href="biclustering.html#cb278-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>, </span>
<span id="cb278-9"><a href="biclustering.html#cb278-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">cor_threshold =</span> <span class="fl">0.2</span>)</span></code></pre></div>
<p><em>biclust</em> takes matrix as an input and returns <em>biclust</em> object.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="biclustering.html#cb279-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb279-2"><a href="biclustering.html#cb279-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(biclust)){</span>
<span id="cb279-3"><a href="biclustering.html#cb279-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;biclust&quot;</span>)</span>
<span id="cb279-4"><a href="biclustering.html#cb279-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(biclust)</span>
<span id="cb279-5"><a href="biclustering.html#cb279-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb279-6"><a href="biclustering.html#cb279-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb279-7"><a href="biclustering.html#cb279-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Set seed for reproducibility</span></span>
<span id="cb279-8"><a href="biclustering.html#cb279-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3973</span>)</span>
<span id="cb279-9"><a href="biclustering.html#cb279-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb279-10"><a href="biclustering.html#cb279-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Find biclusters</span></span>
<span id="cb279-11"><a href="biclustering.html#cb279-11" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> <span class="fu">biclust</span>(corr, <span class="at">method=</span><span class="fu">BCPlaid</span>(), <span class="at">fit.model =</span> y <span class="sc">~</span> m,</span>
<span id="cb279-12"><a href="biclustering.html#cb279-12" aria-hidden="true" tabindex="-1"></a>              <span class="at">background =</span> <span class="cn">TRUE</span>, <span class="at">shuffle =</span> <span class="dv">100</span>, <span class="at">back.fit =</span> <span class="dv">0</span>, <span class="at">max.layers =</span> <span class="dv">10</span>,</span>
<span id="cb279-13"><a href="biclustering.html#cb279-13" aria-hidden="true" tabindex="-1"></a>              <span class="at">iter.startup =</span> <span class="dv">10</span>, <span class="at">iter.layer =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb279-14"><a href="biclustering.html#cb279-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb279-15"><a href="biclustering.html#cb279-15" aria-hidden="true" tabindex="-1"></a>bc</span></code></pre></div>
<pre><code>## 
## An object of class Biclust 
## 
## call:
##  biclust(x = corr, method = BCPlaid(), fit.model = y ~ m, background = TRUE, 
##      shuffle = 100, back.fit = 0, max.layers = 10, iter.startup = 10, 
##      iter.layer = 100, verbose = FALSE)
## 
## There was no cluster found</code></pre>
<p>The object includes cluster information. However compared to <em>cobiclust</em>,
<em>biclust</em> object includes only information about clusters that were found, not general cluster.</p>
<p>Meaning that if one cluster size of 5 features was found out of 20 features,
those 15 features do not belong to any cluster. That is why we have to create an
additional cluster for features/samples that are not assigned into any cluster.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="biclustering.html#cb281-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Functions for obtaining biclust information</span></span>
<span id="cb281-2"><a href="biclustering.html#cb281-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb281-3"><a href="biclustering.html#cb281-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get clusters for rows and columns</span></span>
<span id="cb281-4"><a href="biclustering.html#cb281-4" aria-hidden="true" tabindex="-1"></a>.get_biclusters_from_biclust <span class="ot">&lt;-</span> <span class="cf">function</span>(bc, assay){</span>
<span id="cb281-5"><a href="biclustering.html#cb281-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get cluster information for columns and rows</span></span>
<span id="cb281-6"><a href="biclustering.html#cb281-6" aria-hidden="true" tabindex="-1"></a>  bc_columns <span class="ot">&lt;-</span> <span class="fu">t</span>(bc<span class="sc">@</span>NumberxCol)</span>
<span id="cb281-7"><a href="biclustering.html#cb281-7" aria-hidden="true" tabindex="-1"></a>  bc_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(bc_columns)</span>
<span id="cb281-8"><a href="biclustering.html#cb281-8" aria-hidden="true" tabindex="-1"></a>  bc_rows <span class="ot">&lt;-</span> bc<span class="sc">@</span>RowxNumber</span>
<span id="cb281-9"><a href="biclustering.html#cb281-9" aria-hidden="true" tabindex="-1"></a>  bc_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(bc_rows)</span>
<span id="cb281-10"><a href="biclustering.html#cb281-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb281-11"><a href="biclustering.html#cb281-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get data into right format</span></span>
<span id="cb281-12"><a href="biclustering.html#cb281-12" aria-hidden="true" tabindex="-1"></a>  bc_columns <span class="ot">&lt;-</span> <span class="fu">.manipulate_bc_data</span>(bc_columns, assay, <span class="st">&quot;col&quot;</span>)</span>
<span id="cb281-13"><a href="biclustering.html#cb281-13" aria-hidden="true" tabindex="-1"></a>  bc_rows <span class="ot">&lt;-</span> <span class="fu">.manipulate_bc_data</span>(bc_rows, assay, <span class="st">&quot;row&quot;</span>)</span>
<span id="cb281-14"><a href="biclustering.html#cb281-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb281-15"><a href="biclustering.html#cb281-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">bc_columns =</span> bc_columns, <span class="at">bc_rows =</span> bc_rows))</span>
<span id="cb281-16"><a href="biclustering.html#cb281-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb281-17"><a href="biclustering.html#cb281-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb281-18"><a href="biclustering.html#cb281-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Input clusters, and how many observations there should be, i.e., the number of samples or features</span></span>
<span id="cb281-19"><a href="biclustering.html#cb281-19" aria-hidden="true" tabindex="-1"></a>.manipulate_bc_data <span class="ot">&lt;-</span> <span class="cf">function</span>(bc_clusters, assay, row_col){</span>
<span id="cb281-20"><a href="biclustering.html#cb281-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get right dimension</span></span>
<span id="cb281-21"><a href="biclustering.html#cb281-21" aria-hidden="true" tabindex="-1"></a>  dim <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(row_col <span class="sc">==</span> <span class="st">&quot;col&quot;</span>, <span class="fu">ncol</span>(assay), <span class="fu">nrow</span>(assay))</span>
<span id="cb281-22"><a href="biclustering.html#cb281-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get column/row names</span></span>
<span id="cb281-23"><a href="biclustering.html#cb281-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>( row_col <span class="sc">==</span> <span class="st">&quot;col&quot;</span> ){</span>
<span id="cb281-24"><a href="biclustering.html#cb281-24" aria-hidden="true" tabindex="-1"></a>    names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(assay)</span>
<span id="cb281-25"><a href="biclustering.html#cb281-25" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span>{</span>
<span id="cb281-26"><a href="biclustering.html#cb281-26" aria-hidden="true" tabindex="-1"></a>    names <span class="ot">&lt;-</span> <span class="fu">rownames</span>(assay)</span>
<span id="cb281-27"><a href="biclustering.html#cb281-27" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb281-28"><a href="biclustering.html#cb281-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb281-29"><a href="biclustering.html#cb281-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># If no clusters were found, create one. Otherwise create additional cluster which</span></span>
<span id="cb281-30"><a href="biclustering.html#cb281-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># contain those samples that are not included in clusters that were found.</span></span>
<span id="cb281-31"><a href="biclustering.html#cb281-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>( <span class="fu">nrow</span>(bc_clusters) <span class="sc">!=</span> dim ){</span>
<span id="cb281-32"><a href="biclustering.html#cb281-32" aria-hidden="true" tabindex="-1"></a>      bc_clusters <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">cluster =</span> <span class="fu">rep</span>(<span class="cn">TRUE</span>, dim))</span>
<span id="cb281-33"><a href="biclustering.html#cb281-33" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb281-34"><a href="biclustering.html#cb281-34" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Create additional cluster that includes those samples/features that</span></span>
<span id="cb281-35"><a href="biclustering.html#cb281-35" aria-hidden="true" tabindex="-1"></a>      <span class="co"># are not included in other clusters.</span></span>
<span id="cb281-36"><a href="biclustering.html#cb281-36" aria-hidden="true" tabindex="-1"></a>      vec <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">rowSums</span>(bc_clusters) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>)</span>
<span id="cb281-37"><a href="biclustering.html#cb281-37" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If additional cluster contains samples, then add it</span></span>
<span id="cb281-38"><a href="biclustering.html#cb281-38" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> ( <span class="fu">any</span>(vec) ){</span>
<span id="cb281-39"><a href="biclustering.html#cb281-39" aria-hidden="true" tabindex="-1"></a>          bc_clusters <span class="ot">&lt;-</span> <span class="fu">cbind</span>(bc_clusters, vec)</span>
<span id="cb281-40"><a href="biclustering.html#cb281-40" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb281-41"><a href="biclustering.html#cb281-41" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb281-42"><a href="biclustering.html#cb281-42" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Adjust row and column names</span></span>
<span id="cb281-43"><a href="biclustering.html#cb281-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(bc_clusters) <span class="ot">&lt;-</span> names</span>
<span id="cb281-44"><a href="biclustering.html#cb281-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(bc_clusters) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;cluster_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(bc_clusters))</span>
<span id="cb281-45"><a href="biclustering.html#cb281-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(bc_clusters)</span>
<span id="cb281-46"><a href="biclustering.html#cb281-46" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="biclustering.html#cb282-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get biclusters</span></span>
<span id="cb282-2"><a href="biclustering.html#cb282-2" aria-hidden="true" tabindex="-1"></a>bcs <span class="ot">&lt;-</span> <span class="fu">.get_biclusters_from_biclust</span>(bc, corr)</span>
<span id="cb282-3"><a href="biclustering.html#cb282-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb282-4"><a href="biclustering.html#cb282-4" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_rows</span>
<span id="cb282-5"><a href="biclustering.html#cb282-5" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_columns</span>
<span id="cb282-6"><a href="biclustering.html#cb282-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb282-7"><a href="biclustering.html#cb282-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Print biclusters for rows</span></span>
<span id="cb282-8"><a href="biclustering.html#cb282-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bicluster_rows)</span></code></pre></div>
<pre><code>##                           cluster_1
## D_5__Ruminiclostridium 5       TRUE
## D_5__Lactobacillus             TRUE
## D_5__uncultured                TRUE
## D_5__uncultured bacterium      TRUE
## D_5__Lactococcus               TRUE
## D_5__Lachnoclostridium         TRUE</code></pre>
<p>Letâs collect information for the scatter plot.</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="biclustering.html#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Function for obtaining sample-wise sum, mean, median, and mean variance for each cluster</span></span>
<span id="cb284-2"><a href="biclustering.html#cb284-2" aria-hidden="true" tabindex="-1"></a>.sum_mean_median_var <span class="ot">&lt;-</span> <span class="cf">function</span>(tse1, tse2, abund_values1, abund_values2, clusters1, clusters2){</span>
<span id="cb284-3"><a href="biclustering.html#cb284-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb284-4"><a href="biclustering.html#cb284-4" aria-hidden="true" tabindex="-1"></a>  list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb284-5"><a href="biclustering.html#cb284-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create a data frame that includes all the information</span></span>
<span id="cb284-6"><a href="biclustering.html#cb284-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(clusters1) ){</span>
<span id="cb284-7"><a href="biclustering.html#cb284-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Subset data based on cluster</span></span>
<span id="cb284-8"><a href="biclustering.html#cb284-8" aria-hidden="true" tabindex="-1"></a>    tse_subset1 <span class="ot">&lt;-</span> tse1[clusters1[,i], ]</span>
<span id="cb284-9"><a href="biclustering.html#cb284-9" aria-hidden="true" tabindex="-1"></a>    tse_subset2 <span class="ot">&lt;-</span> tse2[clusters2[,i], ]</span>
<span id="cb284-10"><a href="biclustering.html#cb284-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get assay</span></span>
<span id="cb284-11"><a href="biclustering.html#cb284-11" aria-hidden="true" tabindex="-1"></a>    assay1 <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_subset1, abund_values1)</span>
<span id="cb284-12"><a href="biclustering.html#cb284-12" aria-hidden="true" tabindex="-1"></a>    assay2 <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_subset2, abund_values2)</span>
<span id="cb284-13"><a href="biclustering.html#cb284-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate sum, mean, median, and mean variance</span></span>
<span id="cb284-14"><a href="biclustering.html#cb284-14" aria-hidden="true" tabindex="-1"></a>    sum1 <span class="ot">&lt;-</span> <span class="fu">colSums2</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb284-15"><a href="biclustering.html#cb284-15" aria-hidden="true" tabindex="-1"></a>    mean1 <span class="ot">&lt;-</span> <span class="fu">colMeans2</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb284-16"><a href="biclustering.html#cb284-16" aria-hidden="true" tabindex="-1"></a>    median1 <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb284-17"><a href="biclustering.html#cb284-17" aria-hidden="true" tabindex="-1"></a>    var1 <span class="ot">&lt;-</span> <span class="fu">colVars</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb284-18"><a href="biclustering.html#cb284-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb284-19"><a href="biclustering.html#cb284-19" aria-hidden="true" tabindex="-1"></a>    sum2 <span class="ot">&lt;-</span> <span class="fu">colSums2</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb284-20"><a href="biclustering.html#cb284-20" aria-hidden="true" tabindex="-1"></a>    mean2 <span class="ot">&lt;-</span> <span class="fu">colMeans2</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb284-21"><a href="biclustering.html#cb284-21" aria-hidden="true" tabindex="-1"></a>    median2 <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb284-22"><a href="biclustering.html#cb284-22" aria-hidden="true" tabindex="-1"></a>    var2 <span class="ot">&lt;-</span> <span class="fu">colVars</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb284-23"><a href="biclustering.html#cb284-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb284-24"><a href="biclustering.html#cb284-24" aria-hidden="true" tabindex="-1"></a>    list[[i]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(tse1), sum1, sum2, mean1, mean2, </span>
<span id="cb284-25"><a href="biclustering.html#cb284-25" aria-hidden="true" tabindex="-1"></a>                     median1, median2, var1, var2)</span>
<span id="cb284-26"><a href="biclustering.html#cb284-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb284-27"><a href="biclustering.html#cb284-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb284-28"><a href="biclustering.html#cb284-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list)</span>
<span id="cb284-29"><a href="biclustering.html#cb284-29" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb284-30"><a href="biclustering.html#cb284-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb284-31"><a href="biclustering.html#cb284-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate info</span></span>
<span id="cb284-32"><a href="biclustering.html#cb284-32" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">.sum_mean_median_var</span>(mae[[<span class="dv">1</span>]], mae[[<span class="dv">2</span>]], <span class="st">&quot;clr&quot;</span>, <span class="st">&quot;nmr&quot;</span>, bicluster_rows, bicluster_columns)</span></code></pre></div>
<p>Now we can create a scatter plot. X-axis includes median clr abundance of microbiome
and y-axis median absolute concentration of each metabolite. Each data point represents
a single sample.</p>
<p>From the plots, we can see that there is low negative correlation in both cluster 1 and 3.
This means that when abundance of bacteria belonging to cluster 1 or 3 is higher,
the concentration of metabolites of cluster 1 or 3 is lower, and vice versa.</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="biclustering.html#cb285-1" aria-hidden="true" tabindex="-1"></a>pics <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb285-2"><a href="biclustering.html#cb285-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(df)){</span>
<span id="cb285-3"><a href="biclustering.html#cb285-3" aria-hidden="true" tabindex="-1"></a>  pics[[i]] <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df[[i]])  <span class="sc">+</span></span>
<span id="cb285-4"><a href="biclustering.html#cb285-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> median1, <span class="at">y =</span> median2)) <span class="sc">+</span> </span>
<span id="cb285-5"><a href="biclustering.html#cb285-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;Cluster &quot;</span>, i),</span>
<span id="cb285-6"><a href="biclustering.html#cb285-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="st">&quot;Taxa (clr median)&quot;</span>,</span>
<span id="cb285-7"><a href="biclustering.html#cb285-7" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="st">&quot;Metabolites (abs. median)&quot;</span>)</span>
<span id="cb285-8"><a href="biclustering.html#cb285-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(pics[[i]])</span>
<span id="cb285-9"><a href="biclustering.html#cb285-9" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/biclust_6-1.png" width="33%" /></p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="biclustering.html#cb286-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pics[[1]] + pics[[2]] + pics[[3]]</span></span></code></pre></div>
<p><em>pheatmap</em> does not allow boolean values, so they must be converted into factors.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="biclustering.html#cb287-1" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_columns, <span class="dv">2</span>, as.factor))</span>
<span id="cb287-2"><a href="biclustering.html#cb287-2" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_rows, <span class="dv">2</span>, as.factor))</span></code></pre></div>
<p>Again, we can plot clusters with heatmap.</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="biclustering.html#cb288-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Adjust colors for all clusters</span></span>
<span id="cb288-2"><a href="biclustering.html#cb288-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( <span class="fu">ncol</span>(bicluster_rows) <span class="sc">&gt;</span> <span class="fu">ncol</span>(bicluster_columns) ){</span>
<span id="cb288-3"><a href="biclustering.html#cb288-3" aria-hidden="true" tabindex="-1"></a>  cluster_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bicluster_rows)</span>
<span id="cb288-4"><a href="biclustering.html#cb288-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb288-5"><a href="biclustering.html#cb288-5" aria-hidden="true" tabindex="-1"></a>  cluster_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bicluster_columns)</span>
<span id="cb288-6"><a href="biclustering.html#cb288-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb288-7"><a href="biclustering.html#cb288-7" aria-hidden="true" tabindex="-1"></a>annotation_colors <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb288-8"><a href="biclustering.html#cb288-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> cluster_names){</span>
<span id="cb288-9"><a href="biclustering.html#cb288-9" aria-hidden="true" tabindex="-1"></a>  annotation_colors[[name]] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;TRUE&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span>, <span class="st">&quot;FALSE&quot;</span> <span class="ot">=</span> <span class="st">&quot;white&quot;</span>)</span>
<span id="cb288-10"><a href="biclustering.html#cb288-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb288-11"><a href="biclustering.html#cb288-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb288-12"><a href="biclustering.html#cb288-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap</span></span>
<span id="cb288-13"><a href="biclustering.html#cb288-13" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(corr, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F,</span>
<span id="cb288-14"><a href="biclustering.html#cb288-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> bicluster_columns, </span>
<span id="cb288-15"><a href="biclustering.html#cb288-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> bicluster_rows,</span>
<span id="cb288-16"><a href="biclustering.html#cb288-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_colors =</span> annotation_colors)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/biclust_8-1.png" width="960" /></p>
</div>
<div id="taxa-vs-taxa" class="section level2 hasAnchor" number="11.3">
<h2><span class="header-section-number">11.3</span> Taxa vs taxa<a href="biclustering.html#taxa-vs-taxa" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Third and final example deals with situation where we want to analyze correlation
between taxa. <em>biclust</em> is suitable for this.</p>
<div class="sourceCode" id="cb289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="biclustering.html#cb289-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate cross-correlation</span></span>
<span id="cb289-2"><a href="biclustering.html#cb289-2" aria-hidden="true" tabindex="-1"></a>corr <span class="ot">&lt;-</span> <span class="fu">getExperimentCrossCorrelation</span>(mae, <span class="dv">1</span>, <span class="dv">1</span>, </span>
<span id="cb289-3"><a href="biclustering.html#cb289-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">abund_values1 =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">abund_values2 =</span> <span class="st">&quot;clr&quot;</span>, </span>
<span id="cb289-4"><a href="biclustering.html#cb289-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>,</span>
<span id="cb289-5"><a href="biclustering.html#cb289-5" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">cor_threshold =</span> <span class="fl">0.2</span>, <span class="at">verbose =</span> F, <span class="at">show_warning =</span> F)</span>
<span id="cb289-6"><a href="biclustering.html#cb289-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb289-7"><a href="biclustering.html#cb289-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Find biclusters</span></span>
<span id="cb289-8"><a href="biclustering.html#cb289-8" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> <span class="fu">biclust</span>(corr, <span class="at">method=</span><span class="fu">BCPlaid</span>(), <span class="at">fit.model =</span> y <span class="sc">~</span> m,</span>
<span id="cb289-9"><a href="biclustering.html#cb289-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">background =</span> <span class="cn">TRUE</span>, <span class="at">shuffle =</span> <span class="dv">100</span>, <span class="at">back.fit =</span> <span class="dv">0</span>, <span class="at">max.layers =</span> <span class="dv">10</span>,</span>
<span id="cb289-10"><a href="biclustering.html#cb289-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">iter.startup =</span> <span class="dv">10</span>, <span class="at">iter.layer =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="biclustering.html#cb290-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get biclusters</span></span>
<span id="cb290-2"><a href="biclustering.html#cb290-2" aria-hidden="true" tabindex="-1"></a>bcs <span class="ot">&lt;-</span> <span class="fu">.get_biclusters_from_biclust</span>(bc, corr)</span>
<span id="cb290-3"><a href="biclustering.html#cb290-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb290-4"><a href="biclustering.html#cb290-4" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_rows</span>
<span id="cb290-5"><a href="biclustering.html#cb290-5" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_columns</span></code></pre></div>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="biclustering.html#cb291-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a column that combines information</span></span>
<span id="cb291-2"><a href="biclustering.html#cb291-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If row/column includes in multiple clusters, cluster numbers are separated with &quot;_&amp;_&quot;</span></span>
<span id="cb291-3"><a href="biclustering.html#cb291-3" aria-hidden="true" tabindex="-1"></a>bicluster_columns<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">apply</span>(bicluster_columns, <span class="dv">1</span>, </span>
<span id="cb291-4"><a href="biclustering.html#cb291-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="cf">function</span>(x){<span class="fu">paste</span>(<span class="fu">paste</span>(<span class="fu">which</span>(x)), <span class="at">collapse =</span> <span class="st">&quot;_&amp;_&quot;</span>) })</span>
<span id="cb291-5"><a href="biclustering.html#cb291-5" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bicluster_columns[, <span class="st">&quot;clusters&quot;</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb291-6"><a href="biclustering.html#cb291-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb291-7"><a href="biclustering.html#cb291-7" aria-hidden="true" tabindex="-1"></a>bicluster_rows<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">apply</span>(bicluster_rows, <span class="dv">1</span>, </span>
<span id="cb291-8"><a href="biclustering.html#cb291-8" aria-hidden="true" tabindex="-1"></a>                                 <span class="cf">function</span>(x){<span class="fu">paste</span>(<span class="fu">paste</span>(<span class="fu">which</span>(x)), <span class="at">collapse =</span> <span class="st">&quot;_&amp;_&quot;</span>) })</span>
<span id="cb291-9"><a href="biclustering.html#cb291-9" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bicluster_rows[, <span class="st">&quot;clusters&quot;</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span></code></pre></div>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="biclustering.html#cb292-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert boolean values into factor</span></span>
<span id="cb292-2"><a href="biclustering.html#cb292-2" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_columns, <span class="dv">2</span>, as.factor))</span>
<span id="cb292-3"><a href="biclustering.html#cb292-3" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_rows, <span class="dv">2</span>, as.factor))</span>
<span id="cb292-4"><a href="biclustering.html#cb292-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb292-5"><a href="biclustering.html#cb292-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(corr, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F,</span>
<span id="cb292-6"><a href="biclustering.html#cb292-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> bicluster_columns, </span>
<span id="cb292-7"><a href="biclustering.html#cb292-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> bicluster_rows)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/biclust_12-1.png" width="1344" /></p>
</div>
<div id="session-info-7" class="section level2 unnumbered hasAnchor">
<h2>Session Info<a href="biclustering.html#session-info-7" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] biclust_2.0.3                  lattice_0.20-45               
 [3] colorspace_2.0-3               MASS_7.3-57                   
 [5] patchwork_1.1.1                ggplot2_3.3.6                 
 [7] pheatmap_1.0.12                cobiclust_0.1.0               
 [9] microbiomeDataSets_1.1.5       mia_1.3.27                    
[11] MultiAssayExperiment_1.22.0    TreeSummarizedExperiment_2.1.4
[13] Biostrings_2.64.0              XVector_0.36.0                
[15] SingleCellExperiment_1.18.0    SummarizedExperiment_1.26.1   
[17] Biobase_2.56.0                 GenomicRanges_1.48.0          
[19] GenomeInfoDb_1.32.2            IRanges_2.30.0                
[21] S4Vectors_0.34.0               BiocGenerics_0.42.0           
[23] MatrixGenerics_1.8.1           matrixStats_0.62.0-9000       
[25] BiocStyle_2.24.0               rebook_1.6.0                  

loaded via a namespace (and not attached):
  [1] AnnotationHub_3.4.0           BiocFileCache_2.4.0          
  [3] plyr_1.8.7                    lazyeval_0.2.2               
  [5] splines_4.2.0                 BiocParallel_1.30.3          
  [7] scater_1.24.0                 digest_0.6.29                
  [9] yulab.utils_0.0.5             htmltools_0.5.2              
 [11] viridis_0.6.2                 fansi_1.0.3                  
 [13] magrittr_2.0.3                memoise_2.0.1                
 [15] ScaledMatrix_1.4.0            cluster_2.1.3                
 [17] DECIPHER_2.24.0               blob_1.2.3                   
 [19] rappdirs_0.3.3                ggrepel_0.9.1                
 [21] xfun_0.31                     dplyr_1.0.9                  
 [23] crayon_1.5.1                  RCurl_1.98-1.7               
 [25] jsonlite_1.8.0                graph_1.74.0                 
 [27] ape_5.6-2                     glue_1.6.2                   
 [29] gtable_0.3.0                  zlibbioc_1.42.0              
 [31] DelayedArray_0.22.0           additivityTests_1.1-4.1      
 [33] BiocSingular_1.12.0           scales_1.2.0                 
 [35] DBI_1.1.3                     Rcpp_1.0.8.3                 
 [37] viridisLite_0.4.0             xtable_1.8-4                 
 [39] decontam_1.16.0               tidytree_0.3.9               
 [41] bit_4.0.4                     rsvd_1.0.5                   
 [43] httr_1.4.3                    RColorBrewer_1.1-3           
 [45] dir.expiry_1.4.0              modeltools_0.2-23            
 [47] ellipsis_0.3.2                farver_2.1.1                 
 [49] pkgconfig_2.0.3               XML_3.99-0.10                
 [51] scuttle_1.6.2                 CodeDepends_0.6.5            
 [53] sass_0.4.1                    dbplyr_2.2.1                 
 [55] utf8_1.2.2                    labeling_0.4.2               
 [57] tidyselect_1.1.2              rlang_1.0.3                  
 [59] reshape2_1.4.4                later_1.3.0                  
 [61] AnnotationDbi_1.58.0          munsell_0.5.0                
 [63] BiocVersion_3.15.2            tools_4.2.0                  
 [65] cachem_1.0.6                  cli_3.3.0                    
 [67] DirichletMultinomial_1.38.0   generics_0.1.3               
 [69] RSQLite_2.2.14                ExperimentHub_2.4.0          
 [71] evaluate_0.15                 stringr_1.4.0                
 [73] fastmap_1.1.0                 yaml_2.3.5                   
 [75] knitr_1.39                    bit64_4.0.5                  
 [77] purrr_0.3.4                   KEGGREST_1.36.2              
 [79] nlme_3.1-158                  sparseMatrixStats_1.8.0      
 [81] mime_0.12                     flexclust_1.4-1              
 [83] compiler_4.2.0                png_0.1-7                    
 [85] beeswarm_0.4.0                filelock_1.0.2               
 [87] curl_4.3.2                    interactiveDisplayBase_1.34.0
 [89] treeio_1.20.0                 tibble_3.1.7                 
 [91] bslib_0.3.1                   stringi_1.7.6                
 [93] highr_0.9                     Matrix_1.4-1                 
 [95] vegan_2.6-2                   permute_0.9-7                
 [97] vctrs_0.4.1                   pillar_1.7.0                 
 [99] lifecycle_1.0.1               BiocManager_1.30.18          
[101] jquerylib_0.1.4               BiocNeighbors_1.14.0         
[103] bitops_1.0-7                  irlba_2.3.5                  
[105] httpuv_1.6.5                  R6_2.5.1                     
[107] bookdown_0.27                 promises_1.2.0.1             
[109] gridExtra_2.3                 vipor_0.4.5                  
[111] codetools_0.2-18              assertthat_0.2.1             
[113] withr_2.5.0                   GenomeInfoDbData_1.2.8       
[115] mgcv_1.8-40                   parallel_4.2.0               
[117] beachmat_2.12.0               class_7.3-20                 
[119] tidyr_1.2.0                   rmarkdown_2.14               
[121] DelayedMatrixStats_1.18.0     shiny_1.7.1                  
[123] ggbeeswarm_0.6.0             </code></pre>
</div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="community-typing.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="differential-abundance.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
