<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Biclustering | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 13 Biclustering | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Biclustering | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Biclustering | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="multi-assay_analyses.html"/>
<link rel="next" href="resources.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.2.2</b> colData</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.2.3</b> rowData</a></li>
<li class="chapter" data-level="2.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.2.4</b> rowTree</a></li>
<li class="chapter" data-level="2.2.5" data-path="containers.html"><a href="containers.html#alternative-experiments"><i class="fa fa-check"></i><b>2.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.6" data-path="containers.html"><a href="containers.html#multiassayexperiments"><i class="fa fa-check"></i><b>2.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#some-available-packages"><i class="fa fa-check"></i><b>3.2</b> Some available packages</a></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#quality-control-1"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.3.2</b> Library size</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.3</b> Evenness</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.4</b> Dominance</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.5</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
<li class="chapter" data-level="" data-path="community-diversity.html"><a href="community-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#community-comparisons-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.2</b> Community comparisons by beta diversity analysis</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="community-similarity.html"><a href="community-similarity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.4.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.4.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.4.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="community-similarity.html"><a href="community-similarity.html#further-reading"><i class="fa fa-check"></i><b>8.5</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="community-similarity.html"><a href="community-similarity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visualizing-taxonomic-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>10</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.0.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.0.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="10.0.2" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.0.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
<li class="chapter" data-level="10.1" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>10.1</b> Community Detection</a></li>
<li class="chapter" data-level="10.2" data-path="community-typing.html"><a href="community-typing.html#additional-community-typing"><i class="fa fa-check"></i><b>10.2</b> Additional Community Typing</a></li>
<li class="chapter" data-level="" data-path="community-typing.html"><a href="community-typing.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>11.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.6</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.7</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html"><i class="fa fa-check"></i><b>12</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="biclustering.html"><a href="biclustering.html"><i class="fa fa-check"></i><b>13</b> Biclustering</a>
<ul>
<li class="chapter" data-level="13.1" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>13.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="13.2" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>13.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="13.3" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>13.3</b> Taxa vs taxa</a></li>
<li class="chapter" data-level="" data-path="biclustering.html"><a href="biclustering.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="14" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>14</b> Resources</a>
<ul>
<li class="chapter" data-level="14.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>14.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>14.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="14.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>14.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="14.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>14.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="14.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>14.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>14.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>14.2.1</b> Bioconductor Classes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>15</b> Extra material</a>
<ul>
<li class="chapter" data-level="15.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>15.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>16</b> Acknowledgments</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biclustering" class="section level1" number="13">
<h1><span class="header-section-number">Chapter 13</span> Biclustering</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Biclustering is a clustering method, which simultaneously clusters rows and columns.
The idea is to find clusters where subset of features/samples share similar pattern
over samples/features.</p>
<p>There are multiple biclustering packages available, In this example, we use following methods
- <a href="https://cran.r-project.org/web/packages/biclust/index.html">biclust</a>
- <a href="https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13582">cobiclust</a></p>
<p><em>cobiclust</em> is especially developed for microbiome data whereas <em>biclust</em> is more
general method. In this section, we show three different cases and example
solutions to apply biclustering to them.</p>
<ol style="list-style-type: decimal">
<li>Taxa vs samples</li>
<li>Taxa vs biomolecule/biomarker</li>
<li>Taxa vs taxa</li>
</ol>
<p>Biclusters can be visualized using heatmap or boxplot, for instance. For checking purposes,
also scatter plot might be valid choice.</p>
<p>Check more ideas for heatmaps from
<a href="https://microbiome.github.io/OMA/microbiome-community.html#composition-heatmap">here</a>.</p>
<div id="taxa-vs-samples" class="section level2" number="13.1">
<h2><span class="header-section-number">13.1</span> Taxa vs samples</h2>
<p>When you have microbial abundance matrices, we suggest to use <em>cobiclust</em> which is
designed for microbial data.</p>
<p>Load example data</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="biclustering.html#cb313-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb313-2"><a href="biclustering.html#cb313-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(microbiomeDataSets)){</span>
<span id="cb313-3"><a href="biclustering.html#cb313-3" aria-hidden="true" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;microbiomeDataSets&quot;</span>)</span>
<span id="cb313-4"><a href="biclustering.html#cb313-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb313-5"><a href="biclustering.html#cb313-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb313-6"><a href="biclustering.html#cb313-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb313-7"><a href="biclustering.html#cb313-7" aria-hidden="true" tabindex="-1"></a>mae <span class="ot">&lt;-</span> <span class="fu">HintikkaXOData</span>()</span>
<span id="cb313-8"><a href="biclustering.html#cb313-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb313-9"><a href="biclustering.html#cb313-9" aria-hidden="true" tabindex="-1"></a>mae</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] microbiota: SummarizedExperiment with 12706 rows and 40 columns
##  [2] metabolites: SummarizedExperiment with 38 rows and 40 columns
##  [3] biomarkers: SummarizedExperiment with 39 rows and 40 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files</code></pre>
<p>Only the most prevalent taxa are included in analysis.</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="biclustering.html#cb315-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data</span></span>
<span id="cb315-2"><a href="biclustering.html#cb315-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">subsetByPrevalentTaxa</span>(mae[[<span class="dv">1</span>]], <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">prevalence =</span> <span class="fl">0.2</span>, <span class="at">detection =</span> <span class="fl">0.001</span>)</span>
<span id="cb315-3"><a href="biclustering.html#cb315-3" aria-hidden="true" tabindex="-1"></a><span class="co"># clr-transform</span></span>
<span id="cb315-4"><a href="biclustering.html#cb315-4" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformSamples</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb315-5"><a href="biclustering.html#cb315-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb315-6"><a href="biclustering.html#cb315-6" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 57 40 
## metadata(0):
## assays(2): counts clr
## rownames(57): D_5__Escherichia-Shigella D_5__Ruminiclostridium 5 ...
##   D_5__[Ruminococcus] gauvreauii group D_5__Defluviitaleaceae UCG-011
## rowData names(7): Phylum Class ... Species OTU
## colnames(40): C1 C2 ... C39 C40
## colData names(0):</code></pre>
<p><em>cobiclust</em> takes counts table as an input and gives <em>cobiclust</em> object as an output.
It includes clusters for taxa and samples.</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="biclustering.html#cb317-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(cobiclust)){</span>
<span id="cb317-2"><a href="biclustering.html#cb317-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;cobiclust&quot;</span>)</span>
<span id="cb317-3"><a href="biclustering.html#cb317-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(cobiclust)</span>
<span id="cb317-4"><a href="biclustering.html#cb317-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb317-5"><a href="biclustering.html#cb317-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb317-6"><a href="biclustering.html#cb317-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Do clustering; use counts tableÂ´</span></span>
<span id="cb317-7"><a href="biclustering.html#cb317-7" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">cobiclust</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;counts&quot;</span>))</span>
<span id="cb317-8"><a href="biclustering.html#cb317-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb317-9"><a href="biclustering.html#cb317-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Get clusters</span></span>
<span id="cb317-10"><a href="biclustering.html#cb317-10" aria-hidden="true" tabindex="-1"></a>row_clusters <span class="ot">&lt;-</span> clusters<span class="sc">$</span>classification<span class="sc">$</span>rowclass</span>
<span id="cb317-11"><a href="biclustering.html#cb317-11" aria-hidden="true" tabindex="-1"></a>col_clusters <span class="ot">&lt;-</span> clusters<span class="sc">$</span>classification<span class="sc">$</span>colclass</span>
<span id="cb317-12"><a href="biclustering.html#cb317-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb317-13"><a href="biclustering.html#cb317-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Add clusters to rowdata and coldata</span></span>
<span id="cb317-14"><a href="biclustering.html#cb317-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(row_clusters)</span>
<span id="cb317-15"><a href="biclustering.html#cb317-15" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(col_clusters)</span>
<span id="cb317-16"><a href="biclustering.html#cb317-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb317-17"><a href="biclustering.html#cb317-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Order data based on clusters</span></span>
<span id="cb317-18"><a href="biclustering.html#cb317-18" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="fu">order</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters), <span class="fu">order</span>(<span class="fu">colData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters)]</span>
<span id="cb317-19"><a href="biclustering.html#cb317-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb317-20"><a href="biclustering.html#cb317-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Print clusters</span></span>
<span id="cb317-21"><a href="biclustering.html#cb317-21" aria-hidden="true" tabindex="-1"></a>clusters<span class="sc">$</span>classification</span></code></pre></div>
<pre><code>## $rowclass
##  [1] 1 1 1 1 2 2 1 1 1 1 1 1 2 2 2 2 1 2 1 1 2 1 2 2 1 1 2 1 1 1 1 1 2 1 1 2 1 1
## [39] 1 1 1 1 1 1 1 1 1 2 1 2 1 1 1 2 1 1 1
## 
## $colclass
##  C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 C20 
##   1   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 
## C21 C22 C23 C24 C25 C26 C27 C28 C29 C30 C31 C32 C33 C34 C35 C36 C37 C38 C39 C40 
##   2   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   1</code></pre>
<p>Next we can plot clusters. Commonly used plot is heatmap with annotations.</p>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="biclustering.html#cb319-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(pheatmap)){</span>
<span id="cb319-2"><a href="biclustering.html#cb319-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;pheatmap&quot;</span>)</span>
<span id="cb319-3"><a href="biclustering.html#cb319-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pheatmap)</span>
<span id="cb319-4"><a href="biclustering.html#cb319-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb319-5"><a href="biclustering.html#cb319-5" aria-hidden="true" tabindex="-1"></a><span class="co"># z-transform for heatmap</span></span>
<span id="cb319-6"><a href="biclustering.html#cb319-6" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformFeatures</span>(mae[[<span class="dv">1</span>]], <span class="at">abund_values =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">name =</span> <span class="st">&quot;clr_z&quot;</span>)</span>
<span id="cb319-7"><a href="biclustering.html#cb319-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-8"><a href="biclustering.html#cb319-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create annotations. When column names are equal, they should share levels. </span></span>
<span id="cb319-9"><a href="biclustering.html#cb319-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Here samples include 3 clusters, and taxa 2. That is why we have to make </span></span>
<span id="cb319-10"><a href="biclustering.html#cb319-10" aria-hidden="true" tabindex="-1"></a><span class="co"># column names unique. </span></span>
<span id="cb319-11"><a href="biclustering.html#cb319-11" aria-hidden="true" tabindex="-1"></a>annotation_col <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(mae[[<span class="dv">1</span>]])[, <span class="st">&quot;clusters&quot;</span>, <span class="at">drop =</span> F])</span>
<span id="cb319-12"><a href="biclustering.html#cb319-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annotation_col) <span class="ot">&lt;-</span> <span class="st">&quot;col_clusters&quot;</span></span>
<span id="cb319-13"><a href="biclustering.html#cb319-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-14"><a href="biclustering.html#cb319-14" aria-hidden="true" tabindex="-1"></a>annotation_row <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])[, <span class="st">&quot;clusters&quot;</span>, <span class="at">drop =</span> F])</span>
<span id="cb319-15"><a href="biclustering.html#cb319-15" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annotation_row) <span class="ot">&lt;-</span> <span class="st">&quot;row_clusters&quot;</span></span>
<span id="cb319-16"><a href="biclustering.html#cb319-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb319-17"><a href="biclustering.html#cb319-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap</span></span>
<span id="cb319-18"><a href="biclustering.html#cb319-18" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;clr_z&quot;</span>), <span class="at">cluster_rows =</span> F, <span class="at">cluster_cols =</span> F, </span>
<span id="cb319-19"><a href="biclustering.html#cb319-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> annotation_col,</span>
<span id="cb319-20"><a href="biclustering.html#cb319-20" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> annotation_row)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/cobiclust_3-1.png" width="1344" /></p>
<p>Another common plot is a boxplot.</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="biclustering.html#cb320-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(ggplot2)){</span>
<span id="cb320-2"><a href="biclustering.html#cb320-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb320-3"><a href="biclustering.html#cb320-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb320-4"><a href="biclustering.html#cb320-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb320-5"><a href="biclustering.html#cb320-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(gridExtra)){</span>
<span id="cb320-6"><a href="biclustering.html#cb320-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;gridExtra&quot;</span>)</span>
<span id="cb320-7"><a href="biclustering.html#cb320-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(gridExtra)</span>
<span id="cb320-8"><a href="biclustering.html#cb320-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb320-9"><a href="biclustering.html#cb320-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb320-10"><a href="biclustering.html#cb320-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot requires data in melted format</span></span>
<span id="cb320-11"><a href="biclustering.html#cb320-11" aria-hidden="true" tabindex="-1"></a>melt_assay <span class="ot">&lt;-</span> <span class="fu">meltAssay</span>(mae[[<span class="dv">1</span>]], <span class="at">abund_values =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">add_col_data =</span> T, <span class="at">add_row_data =</span> T)</span>
<span id="cb320-12"><a href="biclustering.html#cb320-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb320-13"><a href="biclustering.html#cb320-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Use grid.arrange to arrange two plots side-by-side</span></span>
<span id="cb320-14"><a href="biclustering.html#cb320-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grid.arrange</span>(</span>
<span id="cb320-15"><a href="biclustering.html#cb320-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(melt_assay) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> clusters.x, <span class="at">y =</span> clr)) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Taxa clusters&quot;</span>),</span>
<span id="cb320-16"><a href="biclustering.html#cb320-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(melt_assay) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> clusters.y, <span class="at">y =</span> clr)) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Sample clusters&quot;</span>),</span>
<span id="cb320-17"><a href="biclustering.html#cb320-17" aria-hidden="true" tabindex="-1"></a><span class="at">nrow =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/cobiclust_4-1.png" width="672" /></p>
</div>
<div id="taxa-vs-biomolecules" class="section level2" number="13.2">
<h2><span class="header-section-number">13.2</span> Taxa vs biomolecules</h2>
<p>Here, we analyze cross-correlation between taxa and metabolites. This is a case, where
we use <em>biclust</em> method which is suitable for numeric matrices in general.</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="biclustering.html#cb321-1" aria-hidden="true" tabindex="-1"></a><span class="do">############################ REMOVE THIS WHEN </span><span class="al">BUG</span><span class="do"> IN GETEXPERIMENTCROSSCORR IS FIXED</span></span>
<span id="cb321-2"><a href="biclustering.html#cb321-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(<span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]))</span>
<span id="cb321-3"><a href="biclustering.html#cb321-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb321-4"><a href="biclustering.html#cb321-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate correlations</span></span>
<span id="cb321-5"><a href="biclustering.html#cb321-5" aria-hidden="true" tabindex="-1"></a>corr <span class="ot">&lt;-</span> <span class="fu">getExperimentCrossCorrelation</span>(mae, <span class="dv">1</span>, <span class="dv">2</span>, </span>
<span id="cb321-6"><a href="biclustering.html#cb321-6" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">abund_values1 =</span> <span class="st">&quot;clr&quot;</span>, </span>
<span id="cb321-7"><a href="biclustering.html#cb321-7" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">abund_values2 =</span> <span class="st">&quot;nmr&quot;</span>, </span>
<span id="cb321-8"><a href="biclustering.html#cb321-8" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>, </span>
<span id="cb321-9"><a href="biclustering.html#cb321-9" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">cor_threshold =</span> <span class="fl">0.2</span>)</span></code></pre></div>
<p><em>biclust</em> takes matrix as an input and returns <em>biclust</em> object.</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb322-1"><a href="biclustering.html#cb322-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb322-2"><a href="biclustering.html#cb322-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(biclust)){</span>
<span id="cb322-3"><a href="biclustering.html#cb322-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;biclust&quot;</span>)</span>
<span id="cb322-4"><a href="biclustering.html#cb322-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(biclust)</span>
<span id="cb322-5"><a href="biclustering.html#cb322-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb322-6"><a href="biclustering.html#cb322-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-7"><a href="biclustering.html#cb322-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Find biclusters</span></span>
<span id="cb322-8"><a href="biclustering.html#cb322-8" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> <span class="fu">biclust</span>(corr, <span class="at">method=</span><span class="fu">BCPlaid</span>(), <span class="at">fit.model =</span> y <span class="sc">~</span> m,</span>
<span id="cb322-9"><a href="biclustering.html#cb322-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">background =</span> <span class="cn">TRUE</span>, <span class="at">shuffle =</span> <span class="dv">100</span>, <span class="at">back.fit =</span> <span class="dv">0</span>, <span class="at">max.layers =</span> <span class="dv">10</span>,</span>
<span id="cb322-10"><a href="biclustering.html#cb322-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">iter.startup =</span> <span class="dv">10</span>, <span class="at">iter.layer =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb322-11"><a href="biclustering.html#cb322-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-12"><a href="biclustering.html#cb322-12" aria-hidden="true" tabindex="-1"></a>bc</span></code></pre></div>
<pre><code>## 
## An object of class Biclust 
## 
## call:
##  biclust(x = corr, method = BCPlaid(), fit.model = y ~ m, background = TRUE, 
##      shuffle = 100, back.fit = 0, max.layers = 10, iter.startup = 10, 
##      iter.layer = 100, verbose = FALSE)
## 
## Number of Clusters found:  2 
## 
## First  2  Cluster sizes:
##                    BC 1 BC 2
## Number of Rows:       8    3
## Number of Columns:    7    8</code></pre>
<p>The object includes cluster information. However compared to <em>cobiclust</em>,
<em>biclust</em> object includes only information about clusters that were found, not general cluster.</p>
<p>Meaning that if one cluster size of 5 features was found out of 20 features,
those 15 features do not belong to any cluster. That is why we have to create an
additional cluster for features/samples that are not assigned into any cluster.</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="biclustering.html#cb324-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Functions for obtaining biclust information</span></span>
<span id="cb324-2"><a href="biclustering.html#cb324-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb324-3"><a href="biclustering.html#cb324-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get clusters for rows and columns</span></span>
<span id="cb324-4"><a href="biclustering.html#cb324-4" aria-hidden="true" tabindex="-1"></a>.get_biclusters_from_biclust <span class="ot">&lt;-</span> <span class="cf">function</span>(bc, assay){</span>
<span id="cb324-5"><a href="biclustering.html#cb324-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get cluster information for columns and rows</span></span>
<span id="cb324-6"><a href="biclustering.html#cb324-6" aria-hidden="true" tabindex="-1"></a>  bc_columns <span class="ot">&lt;-</span> <span class="fu">t</span>(bc<span class="sc">@</span>NumberxCol)</span>
<span id="cb324-7"><a href="biclustering.html#cb324-7" aria-hidden="true" tabindex="-1"></a>  bc_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(bc_columns)</span>
<span id="cb324-8"><a href="biclustering.html#cb324-8" aria-hidden="true" tabindex="-1"></a>  bc_rows <span class="ot">&lt;-</span> bc<span class="sc">@</span>RowxNumber</span>
<span id="cb324-9"><a href="biclustering.html#cb324-9" aria-hidden="true" tabindex="-1"></a>  bc_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(bc_rows)</span>
<span id="cb324-10"><a href="biclustering.html#cb324-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb324-11"><a href="biclustering.html#cb324-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get data into right format</span></span>
<span id="cb324-12"><a href="biclustering.html#cb324-12" aria-hidden="true" tabindex="-1"></a>  bc_columns <span class="ot">&lt;-</span> <span class="fu">.manipulate_bc_data</span>(bc_columns, assay, <span class="st">&quot;col&quot;</span>)</span>
<span id="cb324-13"><a href="biclustering.html#cb324-13" aria-hidden="true" tabindex="-1"></a>  bc_rows <span class="ot">&lt;-</span> <span class="fu">.manipulate_bc_data</span>(bc_rows, assay, <span class="st">&quot;row&quot;</span>)</span>
<span id="cb324-14"><a href="biclustering.html#cb324-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb324-15"><a href="biclustering.html#cb324-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">bc_columns =</span> bc_columns, <span class="at">bc_rows =</span> bc_rows))</span>
<span id="cb324-16"><a href="biclustering.html#cb324-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb324-17"><a href="biclustering.html#cb324-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb324-18"><a href="biclustering.html#cb324-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Input clusters, and how many observations there should be, i.e., the number of samples or features</span></span>
<span id="cb324-19"><a href="biclustering.html#cb324-19" aria-hidden="true" tabindex="-1"></a>.manipulate_bc_data <span class="ot">&lt;-</span> <span class="cf">function</span>(bc_clusters, assay, row_col){</span>
<span id="cb324-20"><a href="biclustering.html#cb324-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get right dimension</span></span>
<span id="cb324-21"><a href="biclustering.html#cb324-21" aria-hidden="true" tabindex="-1"></a>  dim <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(row_col <span class="sc">==</span> <span class="st">&quot;col&quot;</span>, <span class="fu">ncol</span>(assay), <span class="fu">nrow</span>(assay))</span>
<span id="cb324-22"><a href="biclustering.html#cb324-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get column/row names</span></span>
<span id="cb324-23"><a href="biclustering.html#cb324-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>( row_col <span class="sc">==</span> <span class="st">&quot;col&quot;</span> ){</span>
<span id="cb324-24"><a href="biclustering.html#cb324-24" aria-hidden="true" tabindex="-1"></a>    names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(assay)</span>
<span id="cb324-25"><a href="biclustering.html#cb324-25" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span>{</span>
<span id="cb324-26"><a href="biclustering.html#cb324-26" aria-hidden="true" tabindex="-1"></a>    names <span class="ot">&lt;-</span> <span class="fu">rownames</span>(assay)</span>
<span id="cb324-27"><a href="biclustering.html#cb324-27" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb324-28"><a href="biclustering.html#cb324-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb324-29"><a href="biclustering.html#cb324-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># If no clusters were found, create one. Otherwise create additional cluster which</span></span>
<span id="cb324-30"><a href="biclustering.html#cb324-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># contain those samples that are not included in clusters that were found.</span></span>
<span id="cb324-31"><a href="biclustering.html#cb324-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>( <span class="fu">nrow</span>(bc_clusters) <span class="sc">!=</span> dim ){</span>
<span id="cb324-32"><a href="biclustering.html#cb324-32" aria-hidden="true" tabindex="-1"></a>      bc_clusters <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">cluster =</span> <span class="fu">rep</span>(<span class="cn">TRUE</span>, dim))</span>
<span id="cb324-33"><a href="biclustering.html#cb324-33" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb324-34"><a href="biclustering.html#cb324-34" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Create additional cluster that includes those samples/features that</span></span>
<span id="cb324-35"><a href="biclustering.html#cb324-35" aria-hidden="true" tabindex="-1"></a>      <span class="co"># are not included in other clusters.</span></span>
<span id="cb324-36"><a href="biclustering.html#cb324-36" aria-hidden="true" tabindex="-1"></a>      vec <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">rowSums</span>(bc_clusters) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>)</span>
<span id="cb324-37"><a href="biclustering.html#cb324-37" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If additional cluster contains samples, then add it</span></span>
<span id="cb324-38"><a href="biclustering.html#cb324-38" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> ( <span class="fu">any</span>(vec) ){</span>
<span id="cb324-39"><a href="biclustering.html#cb324-39" aria-hidden="true" tabindex="-1"></a>          bc_clusters <span class="ot">&lt;-</span> <span class="fu">cbind</span>(bc_clusters, vec)</span>
<span id="cb324-40"><a href="biclustering.html#cb324-40" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb324-41"><a href="biclustering.html#cb324-41" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb324-42"><a href="biclustering.html#cb324-42" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Adjust row and column names</span></span>
<span id="cb324-43"><a href="biclustering.html#cb324-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(bc_clusters) <span class="ot">&lt;-</span> names</span>
<span id="cb324-44"><a href="biclustering.html#cb324-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(bc_clusters) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;cluster_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(bc_clusters))</span>
<span id="cb324-45"><a href="biclustering.html#cb324-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(bc_clusters)</span>
<span id="cb324-46"><a href="biclustering.html#cb324-46" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="biclustering.html#cb325-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get biclusters</span></span>
<span id="cb325-2"><a href="biclustering.html#cb325-2" aria-hidden="true" tabindex="-1"></a>bcs <span class="ot">&lt;-</span> <span class="fu">.get_biclusters_from_biclust</span>(bc, corr)</span>
<span id="cb325-3"><a href="biclustering.html#cb325-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb325-4"><a href="biclustering.html#cb325-4" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_rows</span>
<span id="cb325-5"><a href="biclustering.html#cb325-5" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_columns</span>
<span id="cb325-6"><a href="biclustering.html#cb325-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb325-7"><a href="biclustering.html#cb325-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Print biclusters for rows</span></span>
<span id="cb325-8"><a href="biclustering.html#cb325-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bicluster_rows)</span></code></pre></div>
<pre><code>##                           cluster_1 cluster_2 cluster_3
## D_5__Escherichia-Shigella     FALSE     FALSE      TRUE
## D_5__Ruminiclostridium 5       TRUE     FALSE     FALSE
## D_5__Lactobacillus            FALSE     FALSE      TRUE
## D_5__uncultured               FALSE     FALSE      TRUE
## D_5__uncultured bacterium     FALSE     FALSE      TRUE
## D_5__Lactococcus              FALSE     FALSE      TRUE</code></pre>
<p>Letâs collect information for the scatter plot.</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="biclustering.html#cb327-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Function for obtaining sample-wise sum, mean, median, and mean variance for each cluster</span></span>
<span id="cb327-2"><a href="biclustering.html#cb327-2" aria-hidden="true" tabindex="-1"></a>.sum_mean_median_var <span class="ot">&lt;-</span> <span class="cf">function</span>(tse1, tse2, abund_values1, abund_values2, clusters1, clusters2){</span>
<span id="cb327-3"><a href="biclustering.html#cb327-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb327-4"><a href="biclustering.html#cb327-4" aria-hidden="true" tabindex="-1"></a>  list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb327-5"><a href="biclustering.html#cb327-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create a data frame that includes all the information</span></span>
<span id="cb327-6"><a href="biclustering.html#cb327-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(clusters1) ){</span>
<span id="cb327-7"><a href="biclustering.html#cb327-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Subset data based on cluster</span></span>
<span id="cb327-8"><a href="biclustering.html#cb327-8" aria-hidden="true" tabindex="-1"></a>    tse_subset1 <span class="ot">&lt;-</span> tse1[clusters1[,i], ]</span>
<span id="cb327-9"><a href="biclustering.html#cb327-9" aria-hidden="true" tabindex="-1"></a>    tse_subset2 <span class="ot">&lt;-</span> tse2[clusters2[,i], ]</span>
<span id="cb327-10"><a href="biclustering.html#cb327-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get assay</span></span>
<span id="cb327-11"><a href="biclustering.html#cb327-11" aria-hidden="true" tabindex="-1"></a>    assay1 <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_subset1, abund_values1)</span>
<span id="cb327-12"><a href="biclustering.html#cb327-12" aria-hidden="true" tabindex="-1"></a>    assay2 <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_subset2, abund_values2)</span>
<span id="cb327-13"><a href="biclustering.html#cb327-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate sum, mean, median, and mean variance</span></span>
<span id="cb327-14"><a href="biclustering.html#cb327-14" aria-hidden="true" tabindex="-1"></a>    sum1 <span class="ot">&lt;-</span> <span class="fu">colSums2</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb327-15"><a href="biclustering.html#cb327-15" aria-hidden="true" tabindex="-1"></a>    mean1 <span class="ot">&lt;-</span> <span class="fu">colMeans2</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb327-16"><a href="biclustering.html#cb327-16" aria-hidden="true" tabindex="-1"></a>    median1 <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb327-17"><a href="biclustering.html#cb327-17" aria-hidden="true" tabindex="-1"></a>    var1 <span class="ot">&lt;-</span> <span class="fu">colVars</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb327-18"><a href="biclustering.html#cb327-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb327-19"><a href="biclustering.html#cb327-19" aria-hidden="true" tabindex="-1"></a>    sum2 <span class="ot">&lt;-</span> <span class="fu">colSums2</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb327-20"><a href="biclustering.html#cb327-20" aria-hidden="true" tabindex="-1"></a>    mean2 <span class="ot">&lt;-</span> <span class="fu">colMeans2</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb327-21"><a href="biclustering.html#cb327-21" aria-hidden="true" tabindex="-1"></a>    median2 <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb327-22"><a href="biclustering.html#cb327-22" aria-hidden="true" tabindex="-1"></a>    var2 <span class="ot">&lt;-</span> <span class="fu">colVars</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb327-23"><a href="biclustering.html#cb327-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb327-24"><a href="biclustering.html#cb327-24" aria-hidden="true" tabindex="-1"></a>    list[[i]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(tse1), sum1, sum2, mean1, mean2, </span>
<span id="cb327-25"><a href="biclustering.html#cb327-25" aria-hidden="true" tabindex="-1"></a>                     median1, median2, var1, var2)</span>
<span id="cb327-26"><a href="biclustering.html#cb327-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb327-27"><a href="biclustering.html#cb327-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-28"><a href="biclustering.html#cb327-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list)</span>
<span id="cb327-29"><a href="biclustering.html#cb327-29" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb327-30"><a href="biclustering.html#cb327-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb327-31"><a href="biclustering.html#cb327-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate info</span></span>
<span id="cb327-32"><a href="biclustering.html#cb327-32" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">.sum_mean_median_var</span>(mae[[<span class="dv">1</span>]], mae[[<span class="dv">2</span>]], <span class="st">&quot;clr&quot;</span>, <span class="st">&quot;nmr&quot;</span>, bicluster_rows, bicluster_columns)</span></code></pre></div>
<p>Now we can create a scatter plot. X-axis includes median clr abundance of microbiome
and y-axis median absolute concentration of each metabolite. Each data point represents
a single sample.</p>
<p>From the plots, we can see that there is low negative correlation in both cluster 1 and 3.
This means that when abundance of bacteria belonging to cluster 1 or 3 is higher,
the concentration of metabolites of cluster 1 or 3 is lower, and vice versa.</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="biclustering.html#cb328-1" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb328-2"><a href="biclustering.html#cb328-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb328-3"><a href="biclustering.html#cb328-3" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb328-4"><a href="biclustering.html#cb328-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>( data <span class="cf">in</span> df ){</span>
<span id="cb328-5"><a href="biclustering.html#cb328-5" aria-hidden="true" tabindex="-1"></a>  i <span class="ot">&lt;-</span> i <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb328-6"><a href="biclustering.html#cb328-6" aria-hidden="true" tabindex="-1"></a>  plot_list[[i]] <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data)  <span class="sc">+</span></span>
<span id="cb328-7"><a href="biclustering.html#cb328-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> median1, <span class="at">y =</span> median2)) <span class="sc">+</span> </span>
<span id="cb328-8"><a href="biclustering.html#cb328-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;Cluster &quot;</span>, i), <span class="at">x =</span> <span class="st">&quot;Taxa (clr median)&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Metabolites (abs. median)&quot;</span>)</span>
<span id="cb328-9"><a href="biclustering.html#cb328-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb328-10"><a href="biclustering.html#cb328-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb328-11"><a href="biclustering.html#cb328-11" aria-hidden="true" tabindex="-1"></a><span class="fu">do.call</span>(grid.arrange, <span class="fu">c</span>(plot_list, <span class="at">nrow =</span> <span class="dv">1</span>))</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/biclust_6-1.png" width="1344" /></p>
<p><em>pheatmap</em> does not allow boolean values, so they must be converted into factors.</p>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb329-1"><a href="biclustering.html#cb329-1" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_columns, <span class="dv">2</span>, as.factor))</span>
<span id="cb329-2"><a href="biclustering.html#cb329-2" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_rows, <span class="dv">2</span>, as.factor))</span></code></pre></div>
<p>Again, we can plot clusters with heatmap.</p>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb330-1"><a href="biclustering.html#cb330-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Adjust colors for all clusters</span></span>
<span id="cb330-2"><a href="biclustering.html#cb330-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( <span class="fu">ncol</span>(bicluster_rows) <span class="sc">&gt;</span> <span class="fu">ncol</span>(bicluster_columns) ){</span>
<span id="cb330-3"><a href="biclustering.html#cb330-3" aria-hidden="true" tabindex="-1"></a>  cluster_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bicluster_rows)</span>
<span id="cb330-4"><a href="biclustering.html#cb330-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb330-5"><a href="biclustering.html#cb330-5" aria-hidden="true" tabindex="-1"></a>  cluster_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bicluster_columns)</span>
<span id="cb330-6"><a href="biclustering.html#cb330-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb330-7"><a href="biclustering.html#cb330-7" aria-hidden="true" tabindex="-1"></a>annotation_colors <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb330-8"><a href="biclustering.html#cb330-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> cluster_names){</span>
<span id="cb330-9"><a href="biclustering.html#cb330-9" aria-hidden="true" tabindex="-1"></a>  annotation_colors[[name]] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;TRUE&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span>, <span class="st">&quot;FALSE&quot;</span> <span class="ot">=</span> <span class="st">&quot;white&quot;</span>)</span>
<span id="cb330-10"><a href="biclustering.html#cb330-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb330-11"><a href="biclustering.html#cb330-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb330-12"><a href="biclustering.html#cb330-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap</span></span>
<span id="cb330-13"><a href="biclustering.html#cb330-13" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(corr, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F,</span>
<span id="cb330-14"><a href="biclustering.html#cb330-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> bicluster_columns, </span>
<span id="cb330-15"><a href="biclustering.html#cb330-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> bicluster_rows,</span>
<span id="cb330-16"><a href="biclustering.html#cb330-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_colors =</span> annotation_colors)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/biclust_8-1.png" width="960" /></p>
</div>
<div id="taxa-vs-taxa" class="section level2" number="13.3">
<h2><span class="header-section-number">13.3</span> Taxa vs taxa</h2>
<p>Third and final example deals with situation where we want to analyze correlation
between taxa. <em>biclust</em> is suitable for this.</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb331-1"><a href="biclustering.html#cb331-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate cross-correlation</span></span>
<span id="cb331-2"><a href="biclustering.html#cb331-2" aria-hidden="true" tabindex="-1"></a>corr <span class="ot">&lt;-</span> <span class="fu">getExperimentCrossCorrelation</span>(mae, <span class="dv">1</span>, <span class="dv">1</span>, </span>
<span id="cb331-3"><a href="biclustering.html#cb331-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">abund_values1 =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">abund_values2 =</span> <span class="st">&quot;clr&quot;</span>, </span>
<span id="cb331-4"><a href="biclustering.html#cb331-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>,</span>
<span id="cb331-5"><a href="biclustering.html#cb331-5" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">cor_threshold =</span> <span class="fl">0.2</span>, <span class="at">verbose =</span> F, <span class="at">show_warning =</span> F)</span>
<span id="cb331-6"><a href="biclustering.html#cb331-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb331-7"><a href="biclustering.html#cb331-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Find biclusters</span></span>
<span id="cb331-8"><a href="biclustering.html#cb331-8" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> <span class="fu">biclust</span>(corr, <span class="at">method=</span><span class="fu">BCPlaid</span>(), <span class="at">fit.model =</span> y <span class="sc">~</span> m,</span>
<span id="cb331-9"><a href="biclustering.html#cb331-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">background =</span> <span class="cn">TRUE</span>, <span class="at">shuffle =</span> <span class="dv">100</span>, <span class="at">back.fit =</span> <span class="dv">0</span>, <span class="at">max.layers =</span> <span class="dv">10</span>,</span>
<span id="cb331-10"><a href="biclustering.html#cb331-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">iter.startup =</span> <span class="dv">10</span>, <span class="at">iter.layer =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb332-1"><a href="biclustering.html#cb332-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get biclusters</span></span>
<span id="cb332-2"><a href="biclustering.html#cb332-2" aria-hidden="true" tabindex="-1"></a>bcs <span class="ot">&lt;-</span> <span class="fu">.get_biclusters_from_biclust</span>(bc, corr)</span>
<span id="cb332-3"><a href="biclustering.html#cb332-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb332-4"><a href="biclustering.html#cb332-4" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_rows</span>
<span id="cb332-5"><a href="biclustering.html#cb332-5" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_columns</span></code></pre></div>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="biclustering.html#cb333-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a column that combines information</span></span>
<span id="cb333-2"><a href="biclustering.html#cb333-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If row/column includes in multiple clusters, cluster numbers are separated with &quot;_&amp;_&quot;</span></span>
<span id="cb333-3"><a href="biclustering.html#cb333-3" aria-hidden="true" tabindex="-1"></a>bicluster_columns<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">apply</span>(bicluster_columns, <span class="dv">1</span>, </span>
<span id="cb333-4"><a href="biclustering.html#cb333-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="cf">function</span>(x){<span class="fu">paste</span>(<span class="fu">paste</span>(<span class="fu">which</span>(x)), <span class="at">collapse =</span> <span class="st">&quot;_&amp;_&quot;</span>) })</span>
<span id="cb333-5"><a href="biclustering.html#cb333-5" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bicluster_columns[, <span class="st">&quot;clusters&quot;</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb333-6"><a href="biclustering.html#cb333-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb333-7"><a href="biclustering.html#cb333-7" aria-hidden="true" tabindex="-1"></a>bicluster_rows<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">apply</span>(bicluster_rows, <span class="dv">1</span>, </span>
<span id="cb333-8"><a href="biclustering.html#cb333-8" aria-hidden="true" tabindex="-1"></a>                                 <span class="cf">function</span>(x){<span class="fu">paste</span>(<span class="fu">paste</span>(<span class="fu">which</span>(x)), <span class="at">collapse =</span> <span class="st">&quot;_&amp;_&quot;</span>) })</span>
<span id="cb333-9"><a href="biclustering.html#cb333-9" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bicluster_rows[, <span class="st">&quot;clusters&quot;</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span></code></pre></div>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb334-1"><a href="biclustering.html#cb334-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert boolean values into factor</span></span>
<span id="cb334-2"><a href="biclustering.html#cb334-2" aria-hidden="true" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_columns, <span class="dv">2</span>, as.factor))</span>
<span id="cb334-3"><a href="biclustering.html#cb334-3" aria-hidden="true" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_rows, <span class="dv">2</span>, as.factor))</span>
<span id="cb334-4"><a href="biclustering.html#cb334-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb334-5"><a href="biclustering.html#cb334-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(corr, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F,</span>
<span id="cb334-6"><a href="biclustering.html#cb334-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> bicluster_columns, </span>
<span id="cb334-7"><a href="biclustering.html#cb334-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> bicluster_rows)</span></code></pre></div>
<p><img src="24_biclustering_files/figure-html/biclust_12-1.png" width="1344" /></p>
</div>
<div id="session-info-9" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] biclust_2.0.3                  lattice_0.20-45               
 [3] colorspace_2.0-2               MASS_7.3-54                   
 [5] gridExtra_2.3                  ggplot2_3.3.5                 
 [7] pheatmap_1.0.12                cobiclust_0.1.0               
 [9] microbiomeDataSets_1.1.5       mia_1.3.13                    
[11] MultiAssayExperiment_1.20.0    TreeSummarizedExperiment_2.1.4
[13] Biostrings_2.62.0              XVector_0.34.0                
[15] SingleCellExperiment_1.16.0    SummarizedExperiment_1.24.0   
[17] Biobase_2.54.0                 GenomicRanges_1.46.1          
[19] GenomeInfoDb_1.30.0            IRanges_2.28.0                
[21] S4Vectors_0.32.3               BiocGenerics_0.40.0           
[23] MatrixGenerics_1.6.0           matrixStats_0.61.0-9001       
[25] BiocStyle_2.22.0               rebook_1.4.0                  

loaded via a namespace (and not attached):
  [1] AnnotationHub_3.2.0           BiocFileCache_2.2.0          
  [3] plyr_1.8.6                    lazyeval_0.2.2               
  [5] splines_4.1.2                 BiocParallel_1.28.3          
  [7] scater_1.22.0                 digest_0.6.29                
  [9] yulab.utils_0.0.4             htmltools_0.5.2              
 [11] viridis_0.6.2                 fansi_0.5.0                  
 [13] magrittr_2.0.1                memoise_2.0.1                
 [15] ScaledMatrix_1.2.0            cluster_2.1.2                
 [17] DECIPHER_2.22.0               blob_1.2.2                   
 [19] rappdirs_0.3.3                ggrepel_0.9.1                
 [21] xfun_0.29                     dplyr_1.0.7                  
 [23] crayon_1.4.2                  RCurl_1.98-1.5               
 [25] jsonlite_1.7.2                graph_1.72.0                 
 [27] ape_5.6                       glue_1.6.0                   
 [29] gtable_0.3.0                  zlibbioc_1.40.0              
 [31] DelayedArray_0.20.0           additivityTests_1.1-4        
 [33] BiocSingular_1.10.0           scales_1.1.1                 
 [35] DBI_1.1.2                     Rcpp_1.0.7                   
 [37] viridisLite_0.4.0             xtable_1.8-4                 
 [39] decontam_1.14.0               tidytree_0.3.6               
 [41] bit_4.0.4                     rsvd_1.0.5                   
 [43] httr_1.4.2                    RColorBrewer_1.1-2           
 [45] dir.expiry_1.2.0              modeltools_0.2-23            
 [47] ellipsis_0.3.2                farver_2.1.0                 
 [49] pkgconfig_2.0.3               XML_3.99-0.8                 
 [51] scuttle_1.4.0                 CodeDepends_0.6.5            
 [53] sass_0.4.0                    dbplyr_2.1.1                 
 [55] utf8_1.2.2                    labeling_0.4.2               
 [57] tidyselect_1.1.1              rlang_0.4.12                 
 [59] reshape2_1.4.4                later_1.3.0                  
 [61] AnnotationDbi_1.56.2          munsell_0.5.0                
 [63] BiocVersion_3.14.0            tools_4.1.2                  
 [65] cachem_1.0.6                  DirichletMultinomial_1.36.0  
 [67] generics_0.1.1                RSQLite_2.2.9                
 [69] ExperimentHub_2.2.0           evaluate_0.14                
 [71] stringr_1.4.0                 fastmap_1.1.0                
 [73] yaml_2.2.1                    knitr_1.37                   
 [75] bit64_4.0.5                   purrr_0.3.4                  
 [77] KEGGREST_1.34.0               nlme_3.1-153                 
 [79] sparseMatrixStats_1.6.0       mime_0.12                    
 [81] flexclust_1.4-0               compiler_4.1.2               
 [83] beeswarm_0.4.0                filelock_1.0.2               
 [85] curl_4.3.2                    png_0.1-7                    
 [87] interactiveDisplayBase_1.32.0 treeio_1.18.1                
 [89] tibble_3.1.6                  bslib_0.3.1                  
 [91] stringi_1.7.6                 highr_0.9                    
 [93] Matrix_1.4-0                  vegan_2.5-7                  
 [95] permute_0.9-5                 vctrs_0.3.8                  
 [97] pillar_1.6.4                  lifecycle_1.0.1              
 [99] BiocManager_1.30.16           jquerylib_0.1.4              
[101] BiocNeighbors_1.12.0          bitops_1.0-7                 
[103] irlba_2.3.5                   httpuv_1.6.5                 
[105] R6_2.5.1                      bookdown_0.24                
[107] promises_1.2.0.1              vipor_0.4.5                  
[109] codetools_0.2-18              assertthat_0.2.1             
[111] withr_2.4.3                   GenomeInfoDbData_1.2.7       
[113] mgcv_1.8-38                   parallel_4.1.2               
[115] beachmat_2.10.0               class_7.3-19                 
[117] tidyr_1.1.4                   rmarkdown_2.11               
[119] DelayedMatrixStats_1.16.0     shiny_1.7.1                  
[121] ggbeeswarm_0.6.0             </code></pre>
</div>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="multi-assay_analyses.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="resources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
