<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Community Typing (Clustering) | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 10 Community Typing (Clustering) | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Community Typing (Clustering) | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Community Typing (Clustering) | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="microbiome-community.html"/>
<link rel="next" href="differential-abundance.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#human-microbiome-compendium"><i class="fa fa-check"></i><b>3.3.4</b> Human microbiome compendium</a></li>
<li class="chapter" data-level="3.3.5" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.5</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis with igraph</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-plot"><i class="fa fa-check"></i><b>12.2.1</b> Network plot</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#centrality-measures"><i class="fa fa-check"></i><b>12.2.2</b> Centrality measures</a></li>
<li class="chapter" data-level="12.2.3" data-path="network-learning.html"><a href="network-learning.html#scale-node-sizes-by-centrality"><i class="fa fa-check"></i><b>12.2.3</b> Scale node sizes by centrality</a></li>
<li class="chapter" data-level="12.2.4" data-path="network-learning.html"><a href="network-learning.html#degree-distribution"><i class="fa fa-check"></i><b>12.2.4</b> Degree distribution</a></li>
<li class="chapter" data-level="12.2.5" data-path="network-learning.html"><a href="network-learning.html#clustered-heatmaps"><i class="fa fa-check"></i><b>12.2.5</b> Clustered heatmaps</a></li>
<li class="chapter" data-level="12.2.6" data-path="network-learning.html"><a href="network-learning.html#global-network-measures"><i class="fa fa-check"></i><b>12.2.6</b> Global network measures</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.3</b> Network analysis with NetCoMi</a></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.4</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.4.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.4.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.4.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.5</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.6</b> More about association measures</a></li>
<li class="chapter" data-level="12.7" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.7</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.7.1</b> SparCC</a></li>
<li class="chapter" data-level="12.7.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.7.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.7.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.7.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.7.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.7.4</b> Network plots</a></li>
<li class="chapter" data-level="12.7.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.7.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="clustering" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Community Typing (Clustering)<a href="clustering.html#clustering" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Clustering techniques are unsupervised machine learning that aims to find groups, called clusters, that share a pattern in the data. In the microbiome context, clustering techniques are included in microbiome community typing methods. For example, clustering allow samples to be distinguished from each other based on their microbiome community composition. There are multiple clustering algorithms available.</p>
<p>The data can be clustered either based on features or samples. Hence, depending on the analysis goal the data might require transformation. The examples below are focused on sample clustering. To learn about feature clustering, check out chapter <a href="taxonomic-information.html#taxa-clustering">6.3.1</a>.</p>
<div id="custom-tools" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> Custom tools<a href="clustering.html#custom-tools" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><em>bluster</em> is a Bioconductor package providing tools for clustering data in
in the <code>SummarizedExperiment</code> container. It offers multiple algorithms such
as hierarchical clustering, DBSCAN, and K-means.</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="clustering.html#cb275-1" tabindex="-1"></a><span class="co"># Load dependencies</span></span>
<span id="cb275-2"><a href="clustering.html#cb275-2" tabindex="-1"></a><span class="fu">library</span>(bluster)</span>
<span id="cb275-3"><a href="clustering.html#cb275-3" tabindex="-1"></a><span class="fu">library</span>(kableExtra)</span></code></pre></div>
<p>In the first examples of microbiome community typing we use <code>enterotype</code> data.</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="clustering.html#cb276-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb276-2"><a href="clustering.html#cb276-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;enterotype&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb276-3"><a href="clustering.html#cb276-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> enterotype</span>
<span id="cb276-4"><a href="clustering.html#cb276-4" tabindex="-1"></a></span>
<span id="cb276-5"><a href="clustering.html#cb276-5" tabindex="-1"></a><span class="co"># Apply transformation</span></span>
<span id="cb276-6"><a href="clustering.html#cb276-6" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span></code></pre></div>
<p>The main focus in this example is to show how to use mia’s <code>cluster</code> function to
cluster enterotype data. <code>cluster</code> function allows to choose a clustering algorithm and offers multiple parameters to shape the result.</p>
<p>In this example, HclustParam() parameter is chosen for hierarchical clustering.
HclustParam() parameter itself has parameters on its own
<a href="https://rdrr.io/github/LTLA/bluster/man/HclustParam-class.html">HclustParam documentation</a>.
A parameter is <code>MARGIN</code> defines whether to cluster features or samples .</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="clustering.html#cb277-1" tabindex="-1"></a><span class="co"># Simple use of the hierarchical clustering. Here, the default parameters</span></span>
<span id="cb277-2"><a href="clustering.html#cb277-2" tabindex="-1"></a><span class="co"># Set the cut height to half of the dendrogram height</span></span>
<span id="cb277-3"><a href="clustering.html#cb277-3" tabindex="-1"></a></span>
<span id="cb277-4"><a href="clustering.html#cb277-4" tabindex="-1"></a><span class="co"># Save as an alternative experiment that contains clustering information</span></span>
<span id="cb277-5"><a href="clustering.html#cb277-5" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;hclust&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>, </span>
<span id="cb277-6"><a href="clustering.html#cb277-6" tabindex="-1"></a>               <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>, <span class="fu">HclustParam</span>())</span>
<span id="cb277-7"><a href="clustering.html#cb277-7" tabindex="-1"></a></span>
<span id="cb277-8"><a href="clustering.html#cb277-8" tabindex="-1"></a><span class="co"># The result can be found in &#39;clusters&#39; column of colData</span></span>
<span id="cb277-9"><a href="clustering.html#cb277-9" tabindex="-1"></a></span>
<span id="cb277-10"><a href="clustering.html#cb277-10" tabindex="-1"></a><span class="co"># The number of samples included in each cluster</span></span>
<span id="cb277-11"><a href="clustering.html#cb277-11" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">colData</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;hclust&quot;</span>))<span class="sc">$</span>clusters)</span></code></pre></div>
<pre><code>##  1  2  3  4  5  6  7  8  9 
## 92 25 21  2 55 41 20 23  1</code></pre>
<p>Once the clustering on the samples is done, we can also plot the clusters.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="clustering.html#cb279-1" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb279-2"><a href="clustering.html#cb279-2" tabindex="-1"></a></span>
<span id="cb279-3"><a href="clustering.html#cb279-3" tabindex="-1"></a><span class="co"># Add the MDS dimensions for plotting</span></span>
<span id="cb279-4"><a href="clustering.html#cb279-4" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;hclust&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;hclust&quot;</span>), <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>, </span>
<span id="cb279-5"><a href="clustering.html#cb279-5" tabindex="-1"></a>              <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist, <span class="at">method =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb279-6"><a href="clustering.html#cb279-6" tabindex="-1"></a></span>
<span id="cb279-7"><a href="clustering.html#cb279-7" tabindex="-1"></a><span class="co"># Plot the clusters</span></span>
<span id="cb279-8"><a href="clustering.html#cb279-8" tabindex="-1"></a><span class="fu">plotReducedDim</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;hclust&quot;</span>), <span class="st">&quot;MDS&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;clusters&quot;</span>)</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/bluster_sample_plot-1.png" width="672" /></p>
</div>
<div id="hierarchical-clustering" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> Hierarchical clustering<a href="clustering.html#hierarchical-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The hierarchical clustering algorithm aims to find hierarchy between
samples/features. There are to approaches: agglomerative (“bottom-up”)
and divisive (“top-down”). In agglomerative approach, each observation is first in a unique cluster. The algorithm continues by agglomerating similar clusters. The divisive approach, instead, starts with one cluster that contains all observations. Clusters are split recursively into clusters that differ the most. The clustering ends when each cluster contains only one observation. In this algorithm, the similarity of two clusters is based on the distance
between them.</p>
<p>Hierarchical clustering can be visualized with a dendrogram tree. In each
splitting point, the tree is divided into two clusters leading to the
hierarchy.</p>
<p>Hierarchical clustering requires two steps.
1. Computation of the dissimilarities with a given distance.<br />
2. Clustering based on dissimilarities.</p>
<p>Additionally, since sequencing data is compositional, we apply relative
transformation (as seen in the previous example).</p>
<p>In this example, we want to add information on the clustering. To do so,
we use the <code>full</code> parameter. We also compute the dissimilarities with the
bray distance. Finally, the <code>clust.col</code> parameter allows us to choose the name
of the column in the colData (default name is <code>clusters</code>).</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="clustering.html#cb280-1" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb280-2"><a href="clustering.html#cb280-2" tabindex="-1"></a></span>
<span id="cb280-3"><a href="clustering.html#cb280-3" tabindex="-1"></a><span class="co"># Save another alternative experiment that contains full clustering information</span></span>
<span id="cb280-4"><a href="clustering.html#cb280-4" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;hclust_full&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse,</span>
<span id="cb280-5"><a href="clustering.html#cb280-5" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb280-6"><a href="clustering.html#cb280-6" tabindex="-1"></a>               <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>,</span>
<span id="cb280-7"><a href="clustering.html#cb280-7" tabindex="-1"></a>               <span class="fu">HclustParam</span>(<span class="at">method =</span> <span class="st">&quot;complete&quot;</span>,</span>
<span id="cb280-8"><a href="clustering.html#cb280-8" tabindex="-1"></a>                           <span class="at">dist.fun =</span> vegdist,</span>
<span id="cb280-9"><a href="clustering.html#cb280-9" tabindex="-1"></a>                           <span class="at">metric =</span> <span class="st">&quot;bray&quot;</span>),</span>
<span id="cb280-10"><a href="clustering.html#cb280-10" tabindex="-1"></a>               <span class="at">full =</span> <span class="cn">TRUE</span>,</span>
<span id="cb280-11"><a href="clustering.html#cb280-11" tabindex="-1"></a>               <span class="at">clust.col =</span> <span class="st">&quot;Hclust&quot;</span>)</span></code></pre></div>
<p>We plot the dendrogram, which is possible since we got the
additional information from the clustering.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="clustering.html#cb281-1" tabindex="-1"></a><span class="fu">library</span>(dendextend)</span>
<span id="cb281-2"><a href="clustering.html#cb281-2" tabindex="-1"></a></span>
<span id="cb281-3"><a href="clustering.html#cb281-3" tabindex="-1"></a><span class="co"># Get hclust data from metadata</span></span>
<span id="cb281-4"><a href="clustering.html#cb281-4" tabindex="-1"></a>hclust_data <span class="ot">&lt;-</span> <span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;hclust_full&quot;</span>))<span class="sc">$</span>clusters<span class="sc">$</span>hclust</span>
<span id="cb281-5"><a href="clustering.html#cb281-5" tabindex="-1"></a></span>
<span id="cb281-6"><a href="clustering.html#cb281-6" tabindex="-1"></a><span class="co"># Get the dendrogram object</span></span>
<span id="cb281-7"><a href="clustering.html#cb281-7" tabindex="-1"></a>dendro <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(hclust_data)</span>
<span id="cb281-8"><a href="clustering.html#cb281-8" tabindex="-1"></a></span>
<span id="cb281-9"><a href="clustering.html#cb281-9" tabindex="-1"></a><span class="co"># Plot dendrogram</span></span>
<span id="cb281-10"><a href="clustering.html#cb281-10" tabindex="-1"></a>dendro <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">&quot;labels&quot;</span>, <span class="cn">NULL</span>) <span class="sc">%&gt;%</span> <span class="fu">plot</span>()</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/hclust3-1.png" width="672" /></p>
<p>In our case, we cut the dendrogram in half by default. To know how many clusters
we have, we can check the colData.</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="clustering.html#cb282-1" tabindex="-1"></a><span class="co"># Get the clusters</span></span>
<span id="cb282-2"><a href="clustering.html#cb282-2" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">colData</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;hclust_full&quot;</span>))<span class="sc">$</span>Hclust)</span></code></pre></div>
<pre><code>##  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 
##  3 65 35 12  9  2 23 23 27 15  4 11  6 11  2  1  9  7  4  4  1  2  1  1  1  1</code></pre>
<p>We can see that there are 26 clusters, but that probably is not optimal since the
the number of clusters has been chosen arbitrarily. To determine the number of
clusters, we can use the dendrogram. Usually the tree is split where the branch
length is the largest. However, as we can see from the dendrogram, clusters are
not clear. There are algorithms to identify the optimal number of clusters.</p>
<p>The <code>NbClust</code> library is useful to that end as it offers multiple methods to
determine the optimal number of clusters. Here we will use the silhouette
analysis to determine the optimal number of clusters. For each data point, this
analysis measures the distance to other data points in the same cluster
(cohesion), and the distance to the other clusters (separation), establishing a
score. That score is then combined across the data points. <code>NbClust</code> does
this for multiple number of clusters and the best score corresponds to the
optimal number of clusters.</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="clustering.html#cb284-1" tabindex="-1"></a><span class="fu">library</span>(NbClust)</span>
<span id="cb284-2"><a href="clustering.html#cb284-2" tabindex="-1"></a>diss <span class="ot">&lt;-</span> <span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;hclust_full&quot;</span>))<span class="sc">$</span>clusters<span class="sc">$</span>dist</span>
<span id="cb284-3"><a href="clustering.html#cb284-3" tabindex="-1"></a></span>
<span id="cb284-4"><a href="clustering.html#cb284-4" tabindex="-1"></a><span class="co"># Apply the silhouette analysis on the distance matrix</span></span>
<span id="cb284-5"><a href="clustering.html#cb284-5" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">NbClust</span>(<span class="at">diss =</span> diss, <span class="at">distance =</span> <span class="cn">NULL</span>, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>,</span>
<span id="cb284-6"><a href="clustering.html#cb284-6" tabindex="-1"></a>               <span class="at">index =</span> <span class="st">&quot;silhouette&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  Only frey, mcclain, cindex, sihouette and dunn can be computed. To compute the other indices, data matrix is needed</code></pre>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="clustering.html#cb286-1" tabindex="-1"></a>res<span class="sc">$</span>Best.nc</span></code></pre></div>
<pre><code>## Number_clusters     Value_Index 
##          2.0000          0.4783</code></pre>
<p>Based on the result, let’s divide observations into 2 clusters.</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="clustering.html#cb288-1" tabindex="-1"></a><span class="fu">library</span>(dendextend)</span>
<span id="cb288-2"><a href="clustering.html#cb288-2" tabindex="-1"></a></span>
<span id="cb288-3"><a href="clustering.html#cb288-3" tabindex="-1"></a><span class="co"># Get optimal number of clusters</span></span>
<span id="cb288-4"><a href="clustering.html#cb288-4" tabindex="-1"></a>k <span class="ot">&lt;-</span> res<span class="sc">$</span>Best.nc[<span class="dv">1</span>]</span>
<span id="cb288-5"><a href="clustering.html#cb288-5" tabindex="-1"></a></span>
<span id="cb288-6"><a href="clustering.html#cb288-6" tabindex="-1"></a><span class="co"># Making colors for 2 clusters</span></span>
<span id="cb288-7"><a href="clustering.html#cb288-7" tabindex="-1"></a>col_val_map <span class="ot">&lt;-</span> randomcoloR<span class="sc">::</span><span class="fu">distinctColorPalette</span>(k) <span class="sc">%&gt;%</span></span>
<span id="cb288-8"><a href="clustering.html#cb288-8" tabindex="-1"></a>    <span class="fu">as.list</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb288-9"><a href="clustering.html#cb288-9" tabindex="-1"></a>    <span class="fu">setNames</span>(<span class="fu">paste0</span>(<span class="st">&quot;clust_&quot;</span>, <span class="fu">seq</span>(k)))</span>
<span id="cb288-10"><a href="clustering.html#cb288-10" tabindex="-1"></a></span>
<span id="cb288-11"><a href="clustering.html#cb288-11" tabindex="-1"></a>dend <span class="ot">&lt;-</span> <span class="fu">color_branches</span>(dendro, <span class="at">k =</span> k, <span class="at">col =</span> <span class="fu">unlist</span>(col_val_map))</span>
<span id="cb288-12"><a href="clustering.html#cb288-12" tabindex="-1"></a><span class="fu">labels</span>(dend) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb288-13"><a href="clustering.html#cb288-13" tabindex="-1"></a><span class="fu">plot</span>(dend)</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/hclust6-1.png" width="672" /></p>
</div>
<div id="dirichlet-multinomial-mixtures-dmm" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> Dirichlet Multinomial Mixtures (DMM)<a href="clustering.html#dirichlet-multinomial-mixtures-dmm" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This section focus on <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0030126">Dirichlet-Multinomial Mixture
Model</a> analysis. It is a probabilistic technique that allows to search for sample patterns that
reflect sample similarity in the data. DMM has a property of determining an optimal number of clusters (k) to obtain the best model.
The minimum value of Laplace approximation to the negative log model evidence for DMM models as a function of k, determines an optimal k. The optimal k suggests to fit a model with k mixtures of Dirichlet distributions. For the best model, k probabilities for each sample to belong to each cluster are obtained.</p>
<p>In this example, we cluster the data with DMM clustering. Since the data set is large, the
algorithm requires a lot of computational capacity. Therefore, we use only a subset
of the data that is agglomerated by Phylum as a rank.</p>
<p>In the example of DMM we use <code>GlobalPatterns</code> data.</p>
<div class="sourceCode" id="cb289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="clustering.html#cb289-1" tabindex="-1"></a><span class="co"># Get the data</span></span>
<span id="cb289-2"><a href="clustering.html#cb289-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb289-3"><a href="clustering.html#cb289-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb289-4"><a href="clustering.html#cb289-4" tabindex="-1"></a></span>
<span id="cb289-5"><a href="clustering.html#cb289-5" tabindex="-1"></a><span class="co"># Agglomerate by rank</span></span>
<span id="cb289-6"><a href="clustering.html#cb289-6" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">agglomerateTree =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>In the example below, we calculate model fit using Laplace approximation. The cluster information is added in the metadata with an optional <code>name</code>.</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="clustering.html#cb290-1" tabindex="-1"></a><span class="co"># Run the model and calculates the most likely number of clusters from 1 to 7</span></span>
<span id="cb290-2"><a href="clustering.html#cb290-2" tabindex="-1"></a></span>
<span id="cb290-3"><a href="clustering.html#cb290-3" tabindex="-1"></a><span class="co"># Save as an alternative experiment that contains clustering information</span></span>
<span id="cb290-4"><a href="clustering.html#cb290-4" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse, <span class="at">name =</span> <span class="st">&quot;DMM&quot;</span>, <span class="fu">DmmParam</span>(<span class="at">k =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, <span class="at">type =</span> <span class="st">&quot;laplace&quot;</span>), </span>
<span id="cb290-5"><a href="clustering.html#cb290-5" tabindex="-1"></a>                   <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>, <span class="at">full =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="clustering.html#cb291-1" tabindex="-1"></a><span class="co"># The dmm information is stored in the metadata under the &#39;DMM&#39; column that includes information about all seven models </span></span>
<span id="cb291-2"><a href="clustering.html#cb291-2" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>)</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 67 26 
## metadata(2): agglomerated_by_rank DMM
## assays(1): counts
## rownames(67): Phylum:Crenarchaeota Phylum:Euryarchaeota ...
##   Phylum:Synergistetes Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(8): X.SampleID Primer ... Description clusters
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (67 rows)
## rowTree: 1 phylo tree(s) (66 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>The plot below represents the Laplace approximation to the model evidence for each of the k models. We can see that the best number of clusters is two.</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="clustering.html#cb293-1" tabindex="-1"></a><span class="fu">library</span>(miaViz)</span>
<span id="cb293-2"><a href="clustering.html#cb293-2" tabindex="-1"></a><span class="fu">plotDMNFit</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>), <span class="at">type =</span> <span class="st">&quot;laplace&quot;</span>, <span class="at">name =</span> <span class="st">&quot;DMM&quot;</span>)</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/dmm4-1.png" width="672" /></p>
<p>The best model can be confirmed with the following operation.</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="clustering.html#cb294-1" tabindex="-1"></a><span class="co"># Get the the best model</span></span>
<span id="cb294-2"><a href="clustering.html#cb294-2" tabindex="-1"></a>bestFit <span class="ot">&lt;-</span> <span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>DMM<span class="sc">$</span>dmm[[<span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>DMM<span class="sc">$</span>best]]</span>
<span id="cb294-3"><a href="clustering.html#cb294-3" tabindex="-1"></a>bestFit</span></code></pre></div>
<pre><code>## class: DMN 
## k: 2 
## samples x taxa: 26 x 67 
## Laplace: 7673 BIC: 7927 AIC: 7842</code></pre>
<p>The clusters for the best model are saved in the colData under ‘cluster’ column.</p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="clustering.html#cb296-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>clusters, <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong 
##       1       1       1       2       2       2       2       2       2       2 
## Levels: 1 2</code></pre>
<p>More detailed information about the clusters can be accessed in the metadata. The metadata contains samples-cluster assignment probabilities that tell us the likelihood for each sample to belong to each cluster.</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="clustering.html#cb298-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>DMM<span class="sc">$</span>prob, <span class="dv">10</span>) </span></code></pre></div>
<pre><code>##                 1         2
## CL3     1.000e+00 5.009e-17
## CC1     1.000e+00 3.801e-22
## SV1     1.000e+00 2.017e-12
## M31Fcsw 7.316e-26 1.000e+00
## M11Fcsw 1.062e-16 1.000e+00
## M31Plmr 9.971e-14 1.000e+00
## M11Plmr 2.184e-06 1.000e+00
## F21Plmr 2.840e-11 1.000e+00
## M31Tong 1.304e-08 1.000e+00
## M11Tong 2.082e-06 1.000e+00</code></pre>
<p>Once the optimal model have been confirmed, we can find out which samples are grouped with each other. The table below shows one sample of each sample type clustered in either of the groups. We can notice that DMM can distinguish environmental samples into one group, and mock and human samples into another. For clarity, in this example, the probabilities for each sample to belong in each cluster have been rounded.</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="clustering.html#cb300-1" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb300-2"><a href="clustering.html#cb300-2" tabindex="-1"></a></span>
<span id="cb300-3"><a href="clustering.html#cb300-3" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>DMM<span class="sc">$</span>prob, <span class="dv">1</span>)</span>
<span id="cb300-4"><a href="clustering.html#cb300-4" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(clusters, <span class="fu">levels</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>)<span class="sc">$</span>SampleType)[<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>)<span class="sc">$</span>SampleType])) <span class="co"># add sample type information</span></span>
<span id="cb300-5"><a href="clustering.html#cb300-5" tabindex="-1"></a><span class="fu">colnames</span>(clusters) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Group1&quot;</span>, <span class="st">&quot;Group2&quot;</span>, <span class="st">&quot;SampleType&quot;</span>)</span>
<span id="cb300-6"><a href="clustering.html#cb300-6" tabindex="-1"></a></span>
<span id="cb300-7"><a href="clustering.html#cb300-7" tabindex="-1"></a>clusters <span class="sc">%&gt;%</span></span>
<span id="cb300-8"><a href="clustering.html#cb300-8" tabindex="-1"></a>  <span class="fu">group_by</span>(SampleType) <span class="sc">%&gt;%</span></span>
<span id="cb300-9"><a href="clustering.html#cb300-9" tabindex="-1"></a>  <span class="fu">arrange</span>(Group1) <span class="sc">%&gt;%</span></span>
<span id="cb300-10"><a href="clustering.html#cb300-10" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">row_number</span>()<span class="sc">==</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## # A tibble: 9 × 3
## # Groups:   SampleType [9]
##   Group1 Group2 SampleType        
##   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;             
## 1 0      1      Feces             
## 2 0      1      Skin              
## 3 0      1      Tongue            
## 4 0      1      Mock              
## 5 1      0      Soil              
## 6 1      0      Freshwater        
## 7 1      0      Freshwater (creek)
## 8 1      0      Ocean             
## 9 1      0      Sediment (estuary)</code></pre>
<p>We can also plot the driver Phyla in each group. In this case, it reflects the differences between environmental and human samples.</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="clustering.html#cb302-1" tabindex="-1"></a><span class="co"># Get the estimates on how much each phyla contributes on each cluster</span></span>
<span id="cb302-2"><a href="clustering.html#cb302-2" tabindex="-1"></a>best_model <span class="ot">&lt;-</span> <span class="fu">metadata</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>DMM<span class="sc">$</span>dmm[<span class="dv">2</span>]</span>
<span id="cb302-3"><a href="clustering.html#cb302-3" tabindex="-1"></a>drivers <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(best_model[[<span class="dv">1</span>]]<span class="sc">@</span>fit<span class="sc">$</span>Estimate)</span>
<span id="cb302-4"><a href="clustering.html#cb302-4" tabindex="-1"></a></span>
<span id="cb302-5"><a href="clustering.html#cb302-5" tabindex="-1"></a>drivers<span class="sc">$</span>phyla <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;Phylum:&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">rownames</span>(drivers)) <span class="co"># Clean phylum names</span></span>
<span id="cb302-6"><a href="clustering.html#cb302-6" tabindex="-1"></a></span>
<span id="cb302-7"><a href="clustering.html#cb302-7" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb302-8"><a href="clustering.html#cb302-8" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>) {</span>
<span id="cb302-9"><a href="clustering.html#cb302-9" tabindex="-1"></a>  drivers <span class="ot">&lt;-</span> drivers[<span class="fu">order</span>(drivers[[i]], <span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb302-10"><a href="clustering.html#cb302-10" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">head</span>(drivers, <span class="dv">10</span>), <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(<span class="fu">head</span>(phyla, <span class="dv">10</span>), <span class="sc">+</span> <span class="fu">head</span>(drivers[[i]], <span class="dv">10</span>)), <span class="at">y =</span> <span class="fu">head</span>(drivers[[i]], <span class="dv">10</span>))) <span class="sc">+</span></span>
<span id="cb302-11"><a href="clustering.html#cb302-11" tabindex="-1"></a>          <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;deeppink4&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb302-12"><a href="clustering.html#cb302-12" tabindex="-1"></a>          <span class="fu">coord_flip</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">&quot;Top phyla in group&quot;</span>, i)) <span class="sc">+</span></span>
<span id="cb302-13"><a href="clustering.html#cb302-13" tabindex="-1"></a>          <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;&quot;</span>, <span class="at">y=</span><span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">limits=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">7</span>))</span>
<span id="cb302-14"><a href="clustering.html#cb302-14" tabindex="-1"></a></span>
<span id="cb302-15"><a href="clustering.html#cb302-15" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb302-16"><a href="clustering.html#cb302-16" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/dmm9-1.png" width="768" /><img src="24_clustering_files/figure-html/dmm9-2.png" width="768" /></p>
<p>We use <code>calculateDMNgroup</code> function to have an overview of the best model. The function groups samples by <code>SampleType</code> column from colData and returns DMNGroup object that contains a summary.</p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="clustering.html#cb303-1" tabindex="-1"></a>dmm_group <span class="ot">&lt;-</span> <span class="fu">calculateDMNgroup</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>), <span class="at">variable =</span> <span class="st">&quot;SampleType&quot;</span>, </span>
<span id="cb303-2"><a href="clustering.html#cb303-2" tabindex="-1"></a>                               <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">k =</span> <span class="dv">2</span>, </span>
<span id="cb303-3"><a href="clustering.html#cb303-3" tabindex="-1"></a>                               <span class="at">seed =</span> .Machine<span class="sc">$</span>integer.max)</span>
<span id="cb303-4"><a href="clustering.html#cb303-4" tabindex="-1"></a></span>
<span id="cb303-5"><a href="clustering.html#cb303-5" tabindex="-1"></a>dmm_group</span></code></pre></div>
<pre><code>## class: DMNGroup 
## summary:
##                    k samples taxa    NLE  LogDet Laplace    BIC  AIC
## Feces              2       4   67 1078.3 -106.26   901.1 1171.9 1213
## Freshwater         2       2   67  889.6  -97.23   716.9  936.4 1025
## Freshwater (creek) 2       3   67 1600.3  862.19  1907.3 1674.5 1735
## Mock               2       3   67 1008.4  -55.40   856.6 1082.5 1143
## Ocean              2       3   67 1096.7  -56.66   944.3 1170.9 1232
## Sediment (estuary) 2       3   67 1195.5   18.63  1080.8 1269.7 1331
## Skin               2       3   67  992.6  -85.05   826.1 1066.8 1128
## Soil               2       3   67 1380.3   11.20  1261.8 1454.5 1515
## Tongue             2       2   67  783.0 -107.79   605.0  829.8  918</code></pre>
<p>Mixture weights can be used for having a rough approximation of the cluster size.</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="clustering.html#cb305-1" tabindex="-1"></a>DirichletMultinomial<span class="sc">::</span><span class="fu">mixturewt</span>(bestFit)</span></code></pre></div>
<pre><code>##       pi theta
## 1 0.5385 20.60
## 2 0.4615 15.29</code></pre>
<div id="pcoa-with-dmm-clusters" class="section level3 hasAnchor" number="10.3.1">
<h3><span class="header-section-number">10.3.1</span> PCoA with DMM clusters<a href="clustering.html#pcoa-with-dmm-clusters" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this section we show how to calculate principal coordinates for clr transformed abundance data. To calculate PCoA, we use Aitchison distance as a distance metrics that calculates Euclidean distances for clr transformed compositions.</p>
<p>In the visualization section, we project the sample distances on two dimensional space of first two principal coordinates. We colour the samples based on their DMM clusters. The visualization demonstrates that the DMM clusters can be distinguished on a PCoA plot, although the clusters are not coherent. This means that two-dimensional representation of the data created by PCoA preserves similar information that drives the DMM cluster division.</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="clustering.html#cb307-1" tabindex="-1"></a><span class="co"># add pseudocount, because data contains zeros</span></span>
<span id="cb307-2"><a href="clustering.html#cb307-2" tabindex="-1"></a><span class="fu">assay</span>(tse, <span class="st">&quot;pseudo&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>) <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb307-3"><a href="clustering.html#cb307-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;pseudo&quot;</span>, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb307-4"><a href="clustering.html#cb307-4" tabindex="-1"></a></span>
<span id="cb307-5"><a href="clustering.html#cb307-5" tabindex="-1"></a><span class="co"># clr transformation</span></span>
<span id="cb307-6"><a href="clustering.html#cb307-6" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="st">&quot;relabundance&quot;</span>, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>)</span>
<span id="cb307-7"><a href="clustering.html#cb307-7" tabindex="-1"></a></span>
<span id="cb307-8"><a href="clustering.html#cb307-8" tabindex="-1"></a><span class="co"># principal coordinate analysis</span></span>
<span id="cb307-9"><a href="clustering.html#cb307-9" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">calculateMDS</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>)</span>
<span id="cb307-10"><a href="clustering.html#cb307-10" tabindex="-1"></a></span>
<span id="cb307-11"><a href="clustering.html#cb307-11" tabindex="-1"></a><span class="co"># Create a data frame from principal coordinates</span></span>
<span id="cb307-12"><a href="clustering.html#cb307-12" tabindex="-1"></a>euclidean_pcoa_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pcoa1 =</span> df[, <span class="dv">1</span>], <span class="at">pcoa2 =</span> df[, <span class="dv">2</span>])</span></code></pre></div>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="clustering.html#cb308-1" tabindex="-1"></a><span class="co"># Create a data frame that contains principal coordinates and DMM information</span></span>
<span id="cb308-2"><a href="clustering.html#cb308-2" tabindex="-1"></a>euclidean_dmm_pcoa_df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(euclidean_pcoa_df,</span>
<span id="cb308-3"><a href="clustering.html#cb308-3" tabindex="-1"></a>                               <span class="at">dmm_component =</span> <span class="fu">colData</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;dmm&quot;</span>))<span class="sc">$</span>clusters)</span>
<span id="cb308-4"><a href="clustering.html#cb308-4" tabindex="-1"></a></span>
<span id="cb308-5"><a href="clustering.html#cb308-5" tabindex="-1"></a><span class="co"># Create a plot</span></span>
<span id="cb308-6"><a href="clustering.html#cb308-6" tabindex="-1"></a>euclidean_dmm_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> euclidean_dmm_pcoa_df,</span>
<span id="cb308-7"><a href="clustering.html#cb308-7" tabindex="-1"></a>                             <span class="fu">aes</span>(<span class="at">x =</span> pcoa1, <span class="at">y =</span> pcoa2, <span class="at">color =</span> dmm_component)) <span class="sc">+</span></span>
<span id="cb308-8"><a href="clustering.html#cb308-8" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb308-9"><a href="clustering.html#cb308-9" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Coordinate 1&quot;</span>,<span class="at">y =</span> <span class="st">&quot;Coordinate 2&quot;</span>, </span>
<span id="cb308-10"><a href="clustering.html#cb308-10" tabindex="-1"></a>         <span class="at">title =</span> <span class="st">&quot;PCoA with Aitchison distances&quot;</span>) <span class="sc">+</span></span>
<span id="cb308-11"><a href="clustering.html#cb308-11" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="co"># makes titles smaller</span></span>
<span id="cb308-12"><a href="clustering.html#cb308-12" tabindex="-1"></a>                                    <span class="at">hjust =</span> <span class="fl">0.5</span>)) </span>
<span id="cb308-13"><a href="clustering.html#cb308-13" tabindex="-1"></a></span>
<span id="cb308-14"><a href="clustering.html#cb308-14" tabindex="-1"></a>euclidean_dmm_plot</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/dmm12-1.png" width="672" /></p>
</div>
</div>
<div id="biclustering" class="section level2 hasAnchor" number="10.4">
<h2><span class="header-section-number">10.4</span> Biclustering<a href="clustering.html#biclustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Biclustering methods cluster rows and columns simultaneously in order
to find subsets of correlated features/samples.</p>
<p>Here, we use following packages:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/biclust/index.html">biclust</a></li>
<li><a href="https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13582">cobiclust</a></li>
</ul>
<p><em>cobiclust</em> is especially developed for microbiome data whereas <em>biclust</em> is more
general method. In this section, we show two different cases and example
solutions to apply biclustering to them.</p>
<ol style="list-style-type: decimal">
<li>Taxa vs samples</li>
<li>Taxa vs biomolecule/biomarker</li>
</ol>
<p>Biclusters can be visualized using heatmap or boxplot, for
instance. For checking purposes, also scatter plot might be valid
choice.</p>
<p>Check more ideas for heatmaps from chapters <a href="viz-chapter.html#viz-chapter">16</a> and
<a href="microbiome-community.html#microbiome-community">9</a>.</p>
<div id="taxa-vs-samples" class="section level3 hasAnchor" number="10.4.1">
<h3><span class="header-section-number">10.4.1</span> Taxa vs samples<a href="clustering.html#taxa-vs-samples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>When you have microbial abundance matrices, we suggest to use
<em>cobiclust</em> which is designed for microbial data.</p>
<p>Load example data</p>
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="clustering.html#cb309-1" tabindex="-1"></a><span class="fu">library</span>(cobiclust)</span>
<span id="cb309-2"><a href="clustering.html#cb309-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;HintikkaXOData&quot;</span>)</span>
<span id="cb309-3"><a href="clustering.html#cb309-3" tabindex="-1"></a>mae <span class="ot">&lt;-</span> HintikkaXOData</span></code></pre></div>
<p>Only the most prevalent taxa are included in analysis.</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="clustering.html#cb310-1" tabindex="-1"></a><span class="co"># Subset data in the first experiment</span></span>
<span id="cb310-2"><a href="clustering.html#cb310-2" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">subsetByPrevalentFeatures</span>(mae[[<span class="dv">1</span>]], <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>, </span>
<span id="cb310-3"><a href="clustering.html#cb310-3" tabindex="-1"></a>                                      <span class="at">prevalence =</span> <span class="fl">0.2</span>, </span>
<span id="cb310-4"><a href="clustering.html#cb310-4" tabindex="-1"></a>                                      <span class="at">detection =</span> <span class="fl">0.001</span>)</span>
<span id="cb310-5"><a href="clustering.html#cb310-5" tabindex="-1"></a></span>
<span id="cb310-6"><a href="clustering.html#cb310-6" tabindex="-1"></a><span class="co"># rclr-transform in the first experiment</span></span>
<span id="cb310-7"><a href="clustering.html#cb310-7" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;rclr&quot;</span>)</span></code></pre></div>
<p><em>cobiclust</em> takes counts table as an input and gives <em>cobiclust</em> object as an output.
It includes clusters for taxa and samples.</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="clustering.html#cb311-1" tabindex="-1"></a><span class="co"># Do clustering using counts table</span></span>
<span id="cb311-2"><a href="clustering.html#cb311-2" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">cobiclust</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;counts&quot;</span>))</span>
<span id="cb311-3"><a href="clustering.html#cb311-3" tabindex="-1"></a></span>
<span id="cb311-4"><a href="clustering.html#cb311-4" tabindex="-1"></a><span class="co"># Get clusters</span></span>
<span id="cb311-5"><a href="clustering.html#cb311-5" tabindex="-1"></a>row_clusters <span class="ot">&lt;-</span> clusters<span class="sc">$</span>classification<span class="sc">$</span>rowclass</span>
<span id="cb311-6"><a href="clustering.html#cb311-6" tabindex="-1"></a>col_clusters <span class="ot">&lt;-</span> clusters<span class="sc">$</span>classification<span class="sc">$</span>colclass</span>
<span id="cb311-7"><a href="clustering.html#cb311-7" tabindex="-1"></a></span>
<span id="cb311-8"><a href="clustering.html#cb311-8" tabindex="-1"></a><span class="co"># Add clusters to rowdata and coldata</span></span>
<span id="cb311-9"><a href="clustering.html#cb311-9" tabindex="-1"></a><span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(row_clusters)</span>
<span id="cb311-10"><a href="clustering.html#cb311-10" tabindex="-1"></a><span class="fu">colData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(col_clusters)</span>
<span id="cb311-11"><a href="clustering.html#cb311-11" tabindex="-1"></a></span>
<span id="cb311-12"><a href="clustering.html#cb311-12" tabindex="-1"></a><span class="co"># Order data based on clusters</span></span>
<span id="cb311-13"><a href="clustering.html#cb311-13" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="fu">order</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters),</span>
<span id="cb311-14"><a href="clustering.html#cb311-14" tabindex="-1"></a>                     <span class="fu">order</span>(<span class="fu">colData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>clusters)]</span>
<span id="cb311-15"><a href="clustering.html#cb311-15" tabindex="-1"></a></span>
<span id="cb311-16"><a href="clustering.html#cb311-16" tabindex="-1"></a><span class="co"># Print clusters</span></span>
<span id="cb311-17"><a href="clustering.html#cb311-17" tabindex="-1"></a>clusters<span class="sc">$</span>classification</span></code></pre></div>
<pre><code>## $rowclass
##   [1] 1 1 1 1 1 1 1 1 1 2 2 1 2 1 1 1 2 1 2 1 1 1 2 2 1 1 2 2 1 2 2 1 1 2 1 1 2
##  [38] 1 2 2 2 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 2 1 1 2 1 2 2 1 1 1 1 1 1 1 1 1 1 1
##  [75] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 2 2 2 2 1 1 1 1 1 1 1 2 2 1 1 1 1
## [112] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $colclass
##  C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 C20 
##   1   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 
## C21 C22 C23 C24 C25 C26 C27 C28 C29 C30 C31 C32 C33 C34 C35 C36 C37 C38 C39 C40 
##   2   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   1</code></pre>
<p>Next we can plot clusters. Annotated heatmap is a common choice.</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="clustering.html#cb313-1" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb313-2"><a href="clustering.html#cb313-2" tabindex="-1"></a><span class="co"># z-transform for heatmap</span></span>
<span id="cb313-3"><a href="clustering.html#cb313-3" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(mae[[<span class="dv">1</span>]], <span class="at">assay.type =</span> <span class="st">&quot;rclr&quot;</span>,</span>
<span id="cb313-4"><a href="clustering.html#cb313-4" tabindex="-1"></a>                            <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>, <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">name =</span> <span class="st">&quot;rclr_z&quot;</span>)</span>
<span id="cb313-5"><a href="clustering.html#cb313-5" tabindex="-1"></a></span>
<span id="cb313-6"><a href="clustering.html#cb313-6" tabindex="-1"></a><span class="co"># Create annotations. When column names are equal, they should share levels.</span></span>
<span id="cb313-7"><a href="clustering.html#cb313-7" tabindex="-1"></a><span class="co"># Here samples include 3 clusters, and taxa 2. That is why we have to make</span></span>
<span id="cb313-8"><a href="clustering.html#cb313-8" tabindex="-1"></a><span class="co"># column names unique.</span></span>
<span id="cb313-9"><a href="clustering.html#cb313-9" tabindex="-1"></a>annotation_col <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(mae[[<span class="dv">1</span>]])[, <span class="st">&quot;clusters&quot;</span>, <span class="at">drop =</span> F])</span>
<span id="cb313-10"><a href="clustering.html#cb313-10" tabindex="-1"></a><span class="fu">colnames</span>(annotation_col) <span class="ot">&lt;-</span> <span class="st">&quot;col_clusters&quot;</span></span>
<span id="cb313-11"><a href="clustering.html#cb313-11" tabindex="-1"></a></span>
<span id="cb313-12"><a href="clustering.html#cb313-12" tabindex="-1"></a>annotation_row <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])[, <span class="st">&quot;clusters&quot;</span>, <span class="at">drop =</span> F])</span>
<span id="cb313-13"><a href="clustering.html#cb313-13" tabindex="-1"></a><span class="fu">colnames</span>(annotation_row) <span class="ot">&lt;-</span> <span class="st">&quot;row_clusters&quot;</span></span></code></pre></div>
<p>Plot the heatmap.</p>
<div class="sourceCode" id="cb314"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb314-1"><a href="clustering.html#cb314-1" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;rclr_z&quot;</span>), <span class="at">cluster_rows =</span> F, <span class="at">cluster_cols =</span> F,</span>
<span id="cb314-2"><a href="clustering.html#cb314-2" tabindex="-1"></a>         <span class="at">annotation_col =</span> annotation_col, <span class="at">annotation_row =</span> annotation_row)</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/cobiclust_3b-1.png" width="1344" /></p>
<p>Boxplot is commonly used to summarize the results:</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="clustering.html#cb315-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb315-2"><a href="clustering.html#cb315-2" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb315-3"><a href="clustering.html#cb315-3" tabindex="-1"></a></span>
<span id="cb315-4"><a href="clustering.html#cb315-4" tabindex="-1"></a><span class="co"># ggplot requires data in melted format</span></span>
<span id="cb315-5"><a href="clustering.html#cb315-5" tabindex="-1"></a>melt_assay <span class="ot">&lt;-</span> <span class="fu">meltAssay</span>(mae[[<span class="dv">1</span>]], <span class="at">assay.type =</span> <span class="st">&quot;rclr&quot;</span>, </span>
<span id="cb315-6"><a href="clustering.html#cb315-6" tabindex="-1"></a>                        <span class="at">add_col_data =</span> T, <span class="at">add_row_data =</span> T)</span>
<span id="cb315-7"><a href="clustering.html#cb315-7" tabindex="-1"></a></span>
<span id="cb315-8"><a href="clustering.html#cb315-8" tabindex="-1"></a><span class="co"># patchwork two plots side-by-side</span></span>
<span id="cb315-9"><a href="clustering.html#cb315-9" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(melt_assay) <span class="sc">+</span></span>
<span id="cb315-10"><a href="clustering.html#cb315-10" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> clusters.x, <span class="at">y =</span> rclr)) <span class="sc">+</span></span>
<span id="cb315-11"><a href="clustering.html#cb315-11" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Taxa clusters&quot;</span>)</span>
<span id="cb315-12"><a href="clustering.html#cb315-12" tabindex="-1"></a></span>
<span id="cb315-13"><a href="clustering.html#cb315-13" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(melt_assay) <span class="sc">+</span></span>
<span id="cb315-14"><a href="clustering.html#cb315-14" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> clusters.y, <span class="at">y =</span> rclr)) <span class="sc">+</span></span>
<span id="cb315-15"><a href="clustering.html#cb315-15" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Sample clusters&quot;</span>)</span>
<span id="cb315-16"><a href="clustering.html#cb315-16" tabindex="-1"></a></span>
<span id="cb315-17"><a href="clustering.html#cb315-17" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/cobiclust_4-1.png" width="672" /></p>
</div>
<div id="taxa-vs-biomolecules" class="section level3 hasAnchor" number="10.4.2">
<h3><span class="header-section-number">10.4.2</span> Taxa vs biomolecules<a href="clustering.html#taxa-vs-biomolecules" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here, we analyze cross-correlation between taxa and metabolites. This
is a case, where we use <em>biclust</em> method which is suitable for numeric
matrices in general. First we pre-process the data.</p>
<div class="sourceCode" id="cb316"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb316-1"><a href="clustering.html#cb316-1" tabindex="-1"></a><span class="co"># Samples must be in equal order</span></span>
<span id="cb316-2"><a href="clustering.html#cb316-2" tabindex="-1"></a><span class="co"># (Only 1st experiment was ordered in cobiclust step leading to unequal order)</span></span>
<span id="cb316-3"><a href="clustering.html#cb316-3" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][, <span class="fu">colnames</span>(mae[[<span class="dv">2</span>]])]</span>
<span id="cb316-4"><a href="clustering.html#cb316-4" tabindex="-1"></a></span>
<span id="cb316-5"><a href="clustering.html#cb316-5" tabindex="-1"></a><span class="co"># Make rownames unique since it is required by other steps</span></span>
<span id="cb316-6"><a href="clustering.html#cb316-6" tabindex="-1"></a><span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(<span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]))</span>
<span id="cb316-7"><a href="clustering.html#cb316-7" tabindex="-1"></a></span>
<span id="cb316-8"><a href="clustering.html#cb316-8" tabindex="-1"></a><span class="co"># Transform the metabolites to be in log basis</span></span>
<span id="cb316-9"><a href="clustering.html#cb316-9" tabindex="-1"></a>mae[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(mae[[<span class="dv">2</span>]], <span class="at">assay.type =</span> <span class="st">&quot;nmr&quot;</span>, <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>)</span>
<span id="cb316-10"><a href="clustering.html#cb316-10" tabindex="-1"></a></span>
<span id="cb316-11"><a href="clustering.html#cb316-11" tabindex="-1"></a><span class="co"># Add missing data to the metabolites</span></span>
<span id="cb316-12"><a href="clustering.html#cb316-12" tabindex="-1"></a>replace_na <span class="ot">&lt;-</span> <span class="cf">function</span>(row) {</span>
<span id="cb316-13"><a href="clustering.html#cb316-13" tabindex="-1"></a>    na_indices <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">is.na</span>(row))</span>
<span id="cb316-14"><a href="clustering.html#cb316-14" tabindex="-1"></a>    non_na_values <span class="ot">&lt;-</span> row[<span class="sc">!</span><span class="fu">is.na</span>(row)]</span>
<span id="cb316-15"><a href="clustering.html#cb316-15" tabindex="-1"></a>    row[na_indices] <span class="ot">&lt;-</span> <span class="fu">sample</span>(non_na_values, <span class="fu">length</span>(na_indices), <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb316-16"><a href="clustering.html#cb316-16" tabindex="-1"></a>    row</span>
<span id="cb316-17"><a href="clustering.html#cb316-17" tabindex="-1"></a>}</span>
<span id="cb316-18"><a href="clustering.html#cb316-18" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">2</span>]], <span class="st">&quot;log10&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(<span class="fu">assay</span>(mae[[<span class="dv">2</span>]], <span class="st">&quot;log10&quot;</span>), <span class="dv">1</span>, replace_na))</span></code></pre></div>
<p>Next, we compute the spearman correlation matrix.</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="clustering.html#cb317-1" tabindex="-1"></a><span class="co"># Calculate correlations</span></span>
<span id="cb317-2"><a href="clustering.html#cb317-2" tabindex="-1"></a>corr <span class="ot">&lt;-</span> <span class="fu">getExperimentCrossCorrelation</span>(mae, <span class="dv">1</span>, <span class="dv">2</span>, <span class="at">assay.type1 =</span> <span class="st">&quot;rclr&quot;</span>,</span>
<span id="cb317-3"><a href="clustering.html#cb317-3" tabindex="-1"></a>                                      <span class="at">assay.type2 =</span> <span class="st">&quot;log10&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>,</span>
<span id="cb317-4"><a href="clustering.html#cb317-4" tabindex="-1"></a>                                      <span class="at">correlation =</span> <span class="st">&quot;spearman&quot;</span>)</span></code></pre></div>
<p><em>biclust</em> takes a matrix as an input and returns a <em>biclust</em> object.</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb318-1"><a href="clustering.html#cb318-1" tabindex="-1"></a><span class="fu">library</span>(biclust)</span>
<span id="cb318-2"><a href="clustering.html#cb318-2" tabindex="-1"></a><span class="co"># Set seed for reproducibility</span></span>
<span id="cb318-3"><a href="clustering.html#cb318-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3973</span>)</span>
<span id="cb318-4"><a href="clustering.html#cb318-4" tabindex="-1"></a></span>
<span id="cb318-5"><a href="clustering.html#cb318-5" tabindex="-1"></a><span class="co"># Find biclusters</span></span>
<span id="cb318-6"><a href="clustering.html#cb318-6" tabindex="-1"></a>bc <span class="ot">&lt;-</span> <span class="fu">biclust</span>(corr, <span class="at">method =</span> <span class="fu">BCPlaid</span>(), <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb318-7"><a href="clustering.html#cb318-7" tabindex="-1"></a></span>
<span id="cb318-8"><a href="clustering.html#cb318-8" tabindex="-1"></a>bc</span></code></pre></div>
<pre><code>## 
## An object of class Biclust 
## 
## call:
##  biclust(x = corr, method = BCPlaid(), verbose = FALSE)
## 
## Number of Clusters found:  5 
## 
## First  5  Cluster sizes:
##                    BC 1 BC 2 BC 3 BC 4 BC 5
## Number of Rows:      21   18    7    3    3
## Number of Columns:   13   11    9    8    6</code></pre>
<p>The object includes cluster information. However compared to
<em>cobiclust</em>, <em>biclust</em> object includes only information about clusters
that were found, not general cluster.</p>
<p>Meaning that if one cluster size of 5 features was found out of 20 features,
those 15 features do not belong to any cluster. That is why we have to create an
additional cluster for features/samples that are not assigned into any cluster.</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="clustering.html#cb320-1" tabindex="-1"></a><span class="co"># Functions for obtaining biclust information</span></span>
<span id="cb320-2"><a href="clustering.html#cb320-2" tabindex="-1"></a></span>
<span id="cb320-3"><a href="clustering.html#cb320-3" tabindex="-1"></a><span class="co"># Get clusters for rows and columns</span></span>
<span id="cb320-4"><a href="clustering.html#cb320-4" tabindex="-1"></a>.get_biclusters_from_biclust <span class="ot">&lt;-</span> <span class="cf">function</span>(bc, assay) {</span>
<span id="cb320-5"><a href="clustering.html#cb320-5" tabindex="-1"></a>    <span class="co"># Get cluster information for columns and rows</span></span>
<span id="cb320-6"><a href="clustering.html#cb320-6" tabindex="-1"></a>    bc_columns <span class="ot">&lt;-</span> <span class="fu">t</span>(bc<span class="sc">@</span>NumberxCol)</span>
<span id="cb320-7"><a href="clustering.html#cb320-7" tabindex="-1"></a>    bc_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(bc_columns)</span>
<span id="cb320-8"><a href="clustering.html#cb320-8" tabindex="-1"></a>    bc_rows <span class="ot">&lt;-</span> bc<span class="sc">@</span>RowxNumber</span>
<span id="cb320-9"><a href="clustering.html#cb320-9" tabindex="-1"></a>    bc_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(bc_rows)</span>
<span id="cb320-10"><a href="clustering.html#cb320-10" tabindex="-1"></a></span>
<span id="cb320-11"><a href="clustering.html#cb320-11" tabindex="-1"></a>    <span class="co"># Get data into right format</span></span>
<span id="cb320-12"><a href="clustering.html#cb320-12" tabindex="-1"></a>    bc_columns <span class="ot">&lt;-</span> <span class="fu">.manipulate_bc_data</span>(bc_columns, assay, <span class="st">&quot;col&quot;</span>)</span>
<span id="cb320-13"><a href="clustering.html#cb320-13" tabindex="-1"></a>    bc_rows <span class="ot">&lt;-</span> <span class="fu">.manipulate_bc_data</span>(bc_rows, assay, <span class="st">&quot;row&quot;</span>)</span>
<span id="cb320-14"><a href="clustering.html#cb320-14" tabindex="-1"></a>    </span>
<span id="cb320-15"><a href="clustering.html#cb320-15" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">bc_columns =</span> bc_columns, <span class="at">bc_rows =</span> bc_rows))</span>
<span id="cb320-16"><a href="clustering.html#cb320-16" tabindex="-1"></a>}</span>
<span id="cb320-17"><a href="clustering.html#cb320-17" tabindex="-1"></a></span>
<span id="cb320-18"><a href="clustering.html#cb320-18" tabindex="-1"></a><span class="co"># Input clusters, and how many observations there should be, i.e.,</span></span>
<span id="cb320-19"><a href="clustering.html#cb320-19" tabindex="-1"></a><span class="co"># the number of samples or features</span></span>
<span id="cb320-20"><a href="clustering.html#cb320-20" tabindex="-1"></a>.manipulate_bc_data <span class="ot">&lt;-</span> <span class="cf">function</span>(bc_clusters, assay, row_col) {</span>
<span id="cb320-21"><a href="clustering.html#cb320-21" tabindex="-1"></a>    <span class="co"># Get right dimension</span></span>
<span id="cb320-22"><a href="clustering.html#cb320-22" tabindex="-1"></a>    dim <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(row_col <span class="sc">==</span> <span class="st">&quot;col&quot;</span>, <span class="fu">ncol</span>(assay), <span class="fu">nrow</span>(assay))</span>
<span id="cb320-23"><a href="clustering.html#cb320-23" tabindex="-1"></a>    <span class="co"># Get column/row names</span></span>
<span id="cb320-24"><a href="clustering.html#cb320-24" tabindex="-1"></a>    <span class="cf">if</span> (row_col <span class="sc">==</span> <span class="st">&quot;col&quot;</span>) {</span>
<span id="cb320-25"><a href="clustering.html#cb320-25" tabindex="-1"></a>        names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(assay)</span>
<span id="cb320-26"><a href="clustering.html#cb320-26" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb320-27"><a href="clustering.html#cb320-27" tabindex="-1"></a>        names <span class="ot">&lt;-</span> <span class="fu">rownames</span>(assay)</span>
<span id="cb320-28"><a href="clustering.html#cb320-28" tabindex="-1"></a>    }</span>
<span id="cb320-29"><a href="clustering.html#cb320-29" tabindex="-1"></a></span>
<span id="cb320-30"><a href="clustering.html#cb320-30" tabindex="-1"></a>    <span class="co"># If no clusters were found, create one. Otherwise create additional</span></span>
<span id="cb320-31"><a href="clustering.html#cb320-31" tabindex="-1"></a>    <span class="co"># cluster which</span></span>
<span id="cb320-32"><a href="clustering.html#cb320-32" tabindex="-1"></a>    <span class="co"># contain those samples that are not included in clusters that were found.</span></span>
<span id="cb320-33"><a href="clustering.html#cb320-33" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(bc_clusters) <span class="sc">!=</span> dim) {</span>
<span id="cb320-34"><a href="clustering.html#cb320-34" tabindex="-1"></a>        bc_clusters <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">cluster =</span> <span class="fu">rep</span>(<span class="cn">TRUE</span>, dim))</span>
<span id="cb320-35"><a href="clustering.html#cb320-35" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb320-36"><a href="clustering.html#cb320-36" tabindex="-1"></a>        <span class="co"># Create additional cluster that includes those samples/features that</span></span>
<span id="cb320-37"><a href="clustering.html#cb320-37" tabindex="-1"></a>        <span class="co"># are not included in other clusters.</span></span>
<span id="cb320-38"><a href="clustering.html#cb320-38" tabindex="-1"></a>        vec <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">rowSums</span>(bc_clusters) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>)</span>
<span id="cb320-39"><a href="clustering.html#cb320-39" tabindex="-1"></a></span>
<span id="cb320-40"><a href="clustering.html#cb320-40" tabindex="-1"></a>        <span class="co"># If additional cluster contains samples, then add it</span></span>
<span id="cb320-41"><a href="clustering.html#cb320-41" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">any</span>(vec)) {</span>
<span id="cb320-42"><a href="clustering.html#cb320-42" tabindex="-1"></a>            bc_clusters <span class="ot">&lt;-</span> <span class="fu">cbind</span>(bc_clusters, vec)</span>
<span id="cb320-43"><a href="clustering.html#cb320-43" tabindex="-1"></a>        }</span>
<span id="cb320-44"><a href="clustering.html#cb320-44" tabindex="-1"></a>    }</span>
<span id="cb320-45"><a href="clustering.html#cb320-45" tabindex="-1"></a>    </span>
<span id="cb320-46"><a href="clustering.html#cb320-46" tabindex="-1"></a>    <span class="co"># Adjust row and column names</span></span>
<span id="cb320-47"><a href="clustering.html#cb320-47" tabindex="-1"></a>    <span class="fu">rownames</span>(bc_clusters) <span class="ot">&lt;-</span> names</span>
<span id="cb320-48"><a href="clustering.html#cb320-48" tabindex="-1"></a>    <span class="fu">colnames</span>(bc_clusters) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;cluster_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(bc_clusters))</span>
<span id="cb320-49"><a href="clustering.html#cb320-49" tabindex="-1"></a>    <span class="fu">return</span>(bc_clusters)</span>
<span id="cb320-50"><a href="clustering.html#cb320-50" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="clustering.html#cb321-1" tabindex="-1"></a><span class="co"># Get biclusters</span></span>
<span id="cb321-2"><a href="clustering.html#cb321-2" tabindex="-1"></a>bcs <span class="ot">&lt;-</span> <span class="fu">.get_biclusters_from_biclust</span>(bc, corr)</span>
<span id="cb321-3"><a href="clustering.html#cb321-3" tabindex="-1"></a></span>
<span id="cb321-4"><a href="clustering.html#cb321-4" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_rows</span>
<span id="cb321-5"><a href="clustering.html#cb321-5" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> bcs<span class="sc">$</span>bc_columns</span>
<span id="cb321-6"><a href="clustering.html#cb321-6" tabindex="-1"></a></span>
<span id="cb321-7"><a href="clustering.html#cb321-7" tabindex="-1"></a><span class="co"># Print biclusters for rows</span></span>
<span id="cb321-8"><a href="clustering.html#cb321-8" tabindex="-1"></a><span class="fu">head</span>(bicluster_rows)</span></code></pre></div>
<pre><code>##                                                                                                                       cluster_1
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_2
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                           TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_3
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                           TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_4
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_5
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                 TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_6
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                    TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella                TRUE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella      TRUE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                   TRUE</code></pre>
<p>Let’s collect information for the scatter plot.</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="clustering.html#cb323-1" tabindex="-1"></a><span class="co"># Function for obtaining sample-wise sum, mean, median, and mean variance</span></span>
<span id="cb323-2"><a href="clustering.html#cb323-2" tabindex="-1"></a><span class="co"># for each cluster</span></span>
<span id="cb323-3"><a href="clustering.html#cb323-3" tabindex="-1"></a></span>
<span id="cb323-4"><a href="clustering.html#cb323-4" tabindex="-1"></a>.sum_mean_median_var <span class="ot">&lt;-</span> <span class="cf">function</span>(tse1, tse2, assay.type1, assay.type2, clusters1, clusters2) {</span>
<span id="cb323-5"><a href="clustering.html#cb323-5" tabindex="-1"></a>    list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb323-6"><a href="clustering.html#cb323-6" tabindex="-1"></a>    <span class="co"># Create a data frame that includes all the information</span></span>
<span id="cb323-7"><a href="clustering.html#cb323-7" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(clusters1)) {</span>
<span id="cb323-8"><a href="clustering.html#cb323-8" tabindex="-1"></a>        <span class="co"># Subset data based on cluster</span></span>
<span id="cb323-9"><a href="clustering.html#cb323-9" tabindex="-1"></a>        tse_subset1 <span class="ot">&lt;-</span> tse1[clusters1[, i], ]</span>
<span id="cb323-10"><a href="clustering.html#cb323-10" tabindex="-1"></a>        tse_subset2 <span class="ot">&lt;-</span> tse2[clusters2[, i], ]</span>
<span id="cb323-11"><a href="clustering.html#cb323-11" tabindex="-1"></a>        <span class="co"># Get assay</span></span>
<span id="cb323-12"><a href="clustering.html#cb323-12" tabindex="-1"></a>        assay1 <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_subset1, assay.type1)</span>
<span id="cb323-13"><a href="clustering.html#cb323-13" tabindex="-1"></a>        assay2 <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_subset2, assay.type2)</span>
<span id="cb323-14"><a href="clustering.html#cb323-14" tabindex="-1"></a>        <span class="co"># Calculate sum, mean, median, and mean variance</span></span>
<span id="cb323-15"><a href="clustering.html#cb323-15" tabindex="-1"></a>        sum1 <span class="ot">&lt;-</span> <span class="fu">colSums2</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb323-16"><a href="clustering.html#cb323-16" tabindex="-1"></a>        mean1 <span class="ot">&lt;-</span> <span class="fu">colMeans2</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb323-17"><a href="clustering.html#cb323-17" tabindex="-1"></a>        median1 <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb323-18"><a href="clustering.html#cb323-18" tabindex="-1"></a>        var1 <span class="ot">&lt;-</span> <span class="fu">colVars</span>(assay1, <span class="at">na.rm =</span> T)</span>
<span id="cb323-19"><a href="clustering.html#cb323-19" tabindex="-1"></a>        </span>
<span id="cb323-20"><a href="clustering.html#cb323-20" tabindex="-1"></a>        sum2 <span class="ot">&lt;-</span> <span class="fu">colSums2</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb323-21"><a href="clustering.html#cb323-21" tabindex="-1"></a>        mean2 <span class="ot">&lt;-</span> <span class="fu">colMeans2</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb323-22"><a href="clustering.html#cb323-22" tabindex="-1"></a>        median2 <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb323-23"><a href="clustering.html#cb323-23" tabindex="-1"></a>        var2 <span class="ot">&lt;-</span> <span class="fu">colVars</span>(assay2, <span class="at">na.rm =</span> T)</span>
<span id="cb323-24"><a href="clustering.html#cb323-24" tabindex="-1"></a>        </span>
<span id="cb323-25"><a href="clustering.html#cb323-25" tabindex="-1"></a>        list[[i]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(tse1), sum1, sum2, mean1, </span>
<span id="cb323-26"><a href="clustering.html#cb323-26" tabindex="-1"></a>                                 mean2, median1, median2, var1, var2)</span>
<span id="cb323-27"><a href="clustering.html#cb323-27" tabindex="-1"></a>    }</span>
<span id="cb323-28"><a href="clustering.html#cb323-28" tabindex="-1"></a>    <span class="fu">return</span>(list)</span>
<span id="cb323-29"><a href="clustering.html#cb323-29" tabindex="-1"></a>}</span>
<span id="cb323-30"><a href="clustering.html#cb323-30" tabindex="-1"></a></span>
<span id="cb323-31"><a href="clustering.html#cb323-31" tabindex="-1"></a><span class="co"># Calculate info</span></span>
<span id="cb323-32"><a href="clustering.html#cb323-32" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">.sum_mean_median_var</span>(mae[[<span class="dv">1</span>]], mae[[<span class="dv">2</span>]], <span class="st">&quot;rclr&quot;</span>, <span class="st">&quot;log10&quot;</span>, bicluster_rows, bicluster_columns)</span></code></pre></div>
<p>Now we can create a scatter plot. X-axis includes median clr abundance
of microbiome and y-axis median absolute concentration of each
metabolite. Each data point represents a single sample.</p>
<p>From the plots, we can see that there is low negative correlation in
both cluster 1 and 3. This means that when abundance of bacteria
belonging to cluster 1 or 3 is higher, the concentration of
metabolites of cluster 1 or 3 is lower, and vice versa.</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="clustering.html#cb324-1" tabindex="-1"></a>pics <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb324-2"><a href="clustering.html#cb324-2" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(df)) {</span>
<span id="cb324-3"><a href="clustering.html#cb324-3" tabindex="-1"></a>    pics[[i]] <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df[[i]]) <span class="sc">+</span></span>
<span id="cb324-4"><a href="clustering.html#cb324-4" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> median1, <span class="at">y =</span> median2)) <span class="sc">+</span></span>
<span id="cb324-5"><a href="clustering.html#cb324-5" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;Cluster &quot;</span>, i), <span class="at">x =</span> <span class="st">&quot;Taxa (rclr median)&quot;</span>,</span>
<span id="cb324-6"><a href="clustering.html#cb324-6" tabindex="-1"></a>             <span class="at">y =</span> <span class="st">&quot;Metabolites (abs. median)&quot;</span>)</span>
<span id="cb324-7"><a href="clustering.html#cb324-7" tabindex="-1"></a>    <span class="fu">print</span>(pics[[i]])</span>
<span id="cb324-8"><a href="clustering.html#cb324-8" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/biclust_7-1.png" width="33%" /><img src="24_clustering_files/figure-html/biclust_7-2.png" width="33%" /><img src="24_clustering_files/figure-html/biclust_7-3.png" width="33%" /><img src="24_clustering_files/figure-html/biclust_7-4.png" width="33%" /><img src="24_clustering_files/figure-html/biclust_7-5.png" width="33%" /><img src="24_clustering_files/figure-html/biclust_7-6.png" width="33%" /></p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="clustering.html#cb325-1" tabindex="-1"></a>pics[[<span class="dv">1</span>]] <span class="sc">+</span> pics[[<span class="dv">2</span>]] <span class="sc">+</span> pics[[<span class="dv">3</span>]]</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/biclust_7-7.png" width="33%" /></p>
<p><em>pheatmap</em> does not allow boolean values, so they must be converted into factors.</p>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb326-1"><a href="clustering.html#cb326-1" tabindex="-1"></a>bicluster_columns <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_columns, <span class="dv">2</span>, as.factor))</span>
<span id="cb326-2"><a href="clustering.html#cb326-2" tabindex="-1"></a>bicluster_rows <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">apply</span>(bicluster_rows, <span class="dv">2</span>, as.factor))</span></code></pre></div>
<p>Again, we can plot clusters with heatmap.</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="clustering.html#cb327-1" tabindex="-1"></a><span class="co"># Adjust colors for all clusters</span></span>
<span id="cb327-2"><a href="clustering.html#cb327-2" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">ncol</span>(bicluster_rows) <span class="sc">&gt;</span> <span class="fu">ncol</span>(bicluster_columns)) {</span>
<span id="cb327-3"><a href="clustering.html#cb327-3" tabindex="-1"></a>    cluster_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bicluster_rows)</span>
<span id="cb327-4"><a href="clustering.html#cb327-4" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb327-5"><a href="clustering.html#cb327-5" tabindex="-1"></a>    cluster_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bicluster_columns)</span>
<span id="cb327-6"><a href="clustering.html#cb327-6" tabindex="-1"></a>}</span>
<span id="cb327-7"><a href="clustering.html#cb327-7" tabindex="-1"></a>annotation_colors <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb327-8"><a href="clustering.html#cb327-8" tabindex="-1"></a><span class="cf">for</span> (name <span class="cf">in</span> cluster_names) {</span>
<span id="cb327-9"><a href="clustering.html#cb327-9" tabindex="-1"></a>    annotation_colors[[name]] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;TRUE&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span>, <span class="st">&quot;FALSE&quot;</span> <span class="ot">=</span> <span class="st">&quot;white&quot;</span>)</span>
<span id="cb327-10"><a href="clustering.html#cb327-10" tabindex="-1"></a>}</span>
<span id="cb327-11"><a href="clustering.html#cb327-11" tabindex="-1"></a></span>
<span id="cb327-12"><a href="clustering.html#cb327-12" tabindex="-1"></a><span class="co"># Create a heatmap</span></span>
<span id="cb327-13"><a href="clustering.html#cb327-13" tabindex="-1"></a><span class="fu">pheatmap</span>(corr, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> F,</span>
<span id="cb327-14"><a href="clustering.html#cb327-14" tabindex="-1"></a>         <span class="at">annotation_col =</span> bicluster_columns, <span class="at">annotation_row =</span> bicluster_rows,</span>
<span id="cb327-15"><a href="clustering.html#cb327-15" tabindex="-1"></a>         <span class="at">annotation_colors =</span> annotation_colors)</span></code></pre></div>
<p><img src="24_clustering_files/figure-html/biclust_9-1.png" width="960" /></p>
</div>
</div>
<div id="additional-community-typing" class="section level2 hasAnchor" number="10.5">
<h2><span class="header-section-number">10.5</span> Additional Community Typing<a href="clustering.html#additional-community-typing" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For more community typing techniques applied to the ‘SprockettTHData’
data set, see the attached .Rmd file.</p>
<p>Link:</p>
<ul>
<li><a href="add-comm-typing.Rmd">Rmd</a></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="microbiome-community.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="differential-abundance.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/24_clustering.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/24_clustering.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/24_clustering.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
