<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Community Similarity | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 8 Community Similarity | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Community Similarity | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Community Similarity | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="community-diversity.html"/>
<link rel="next" href="microbiome-community.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#human-microbiome-compendium"><i class="fa fa-check"></i><b>3.3.4</b> Human microbiome compendium</a></li>
<li class="chapter" data-level="3.3.5" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.5</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis with igraph</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-plot"><i class="fa fa-check"></i><b>12.2.1</b> Network plot</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#centrality-measures"><i class="fa fa-check"></i><b>12.2.2</b> Centrality measures</a></li>
<li class="chapter" data-level="12.2.3" data-path="network-learning.html"><a href="network-learning.html#scale-node-sizes-by-centrality"><i class="fa fa-check"></i><b>12.2.3</b> Scale node sizes by centrality</a></li>
<li class="chapter" data-level="12.2.4" data-path="network-learning.html"><a href="network-learning.html#degree-distribution"><i class="fa fa-check"></i><b>12.2.4</b> Degree distribution</a></li>
<li class="chapter" data-level="12.2.5" data-path="network-learning.html"><a href="network-learning.html#clustered-heatmaps"><i class="fa fa-check"></i><b>12.2.5</b> Clustered heatmaps</a></li>
<li class="chapter" data-level="12.2.6" data-path="network-learning.html"><a href="network-learning.html#global-network-measures"><i class="fa fa-check"></i><b>12.2.6</b> Global network measures</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.3</b> Network analysis with NetCoMi</a></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.4</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.4.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.4.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.4.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.5</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.6</b> More about association measures</a></li>
<li class="chapter" data-level="12.7" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.7</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.7.1</b> SparCC</a></li>
<li class="chapter" data-level="12.7.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.7.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.7.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.7.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.7.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.7.4</b> Network plots</a></li>
<li class="chapter" data-level="12.7.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.7.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="community-similarity" class="section level1 hasAnchor" number="8">
<h1><span class="header-section-number">Chapter 8</span> Community Similarity<a href="community-similarity.html#community-similarity" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Beta diversity quantifies the dissimilarity between communities (multiple
samples), as opposed to alpha diversity which focuses on variation within a
community (one sample). In microbiome research, commonly used metrics of beta
diversity include the Bray-Curtis index (for compositional data), Jaccard index
(for presence/absence data, ignoring abundance information), Aitchison distance
(Euclidean distance for clr transformed abundances, aiming to avoid the
compositionality bias), and the Unifrac distance (that takes into account the
phylogenetic tree information). Notably, only some of these measures are actual
<em>distances</em>, as this is a mathematical concept whose definition is not satisfied
by certain ecological measure, such as the Bray-Curtis index. Therefore, the terms
dissimilarity and beta diversity are preferred.</p>
<table>
<colgroup>
<col width="39%" />
<col width="28%" />
<col width="31%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Method description</th>
<th align="center">Assay type</th>
<th align="center">Beta diversity metric</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Quantitative profiling</td>
<td align="center">Absolute counts</td>
<td align="center">Bray-Curtis</td>
</tr>
<tr class="even">
<td align="center">Relative profiling</td>
<td align="center">Relative abundances</td>
<td align="center">Bray-Curtis</td>
</tr>
<tr class="odd">
<td align="center">Aitchison distance</td>
<td align="center">Absolute counts</td>
<td align="center">Aitchison</td>
</tr>
<tr class="even">
<td align="center">Aitchison distance</td>
<td align="center">clr</td>
<td align="center">Euclidean</td>
</tr>
<tr class="odd">
<td align="center">Robust Aitchison distance</td>
<td align="center">rclr</td>
<td align="center">Euclidean</td>
</tr>
<tr class="even">
<td align="center">Presence/Absence similarity</td>
<td align="center">Relative abundances</td>
<td align="center">Jaccard</td>
</tr>
<tr class="odd">
<td align="center">Presence/Absence similarity</td>
<td align="center">Absolute counts</td>
<td align="center">Jaccard</td>
</tr>
<tr class="even">
<td align="center">Phylogenetic distance</td>
<td align="center">Rarefied counts</td>
<td align="center">Unifrac</td>
</tr>
</tbody>
</table>
<p>In practice, beta diversity is usually represented as a <code>dist</code> object, a
triangular matrix where the distance between each pair of samples is encoded by
a specific cell. This distance matrix can then undergo ordination, which is an
important ecological tool to reduce the dimensionality of data for a more
efficient analysis and visualization. Ordination techniques aim to capture as
much essential information from the data as possible and turn it into a lower
dimensional representation. Dimension reduction is bound to lose information but
commonly used ordination techniques can preserve relevant information of sample
similarities in an optimal way, which is defined in different ways by different
methods.</p>
<p>Based on the type of algorithm, ordination methods in microbiome research can
be generally divided in two categories: unsupervised and supervised ordination.
The former includes Principal Coordinate Analysis (PCoA), Principal Component
Analysis (PCA) and Uniform Manifold Approximation and Projection for Dimension
Reduction (UMAP), whereas the latter is mainly represented by distance-based
Redundancy Analysis (dbRDA). We will first discuss unsupervised ordination
methods and then proceed to supervised ones.</p>
<div id="unsupervised-ordination" class="section level2 hasAnchor" number="8.1">
<h2><span class="header-section-number">8.1</span> Unsupervised ordination<a href="community-similarity.html#unsupervised-ordination" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Unsupervised ordination methods variation in the data without additional
information on covariates or other supervision of the model. Among the different
approaches, Multi-Dimensional Scaling (MDS) and non-metric MDS (NMDS) can be
regarded as the standard. They are jointly referred to as PCoA. For this
demonstration we will analyse beta diversity in GlobalPatterns, and observe the
variation between stool samples and those with a different origin.</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="community-similarity.html#cb233-1" tabindex="-1"></a><span class="co"># Load mia and import sample dataset</span></span>
<span id="cb233-2"><a href="community-similarity.html#cb233-2" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb233-3"><a href="community-similarity.html#cb233-3" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb233-4"><a href="community-similarity.html#cb233-4" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb233-5"><a href="community-similarity.html#cb233-5" tabindex="-1"></a></span>
<span id="cb233-6"><a href="community-similarity.html#cb233-6" tabindex="-1"></a><span class="co"># Beta diversity metrics like Bray-Curtis are often applied to relabundances</span></span>
<span id="cb233-7"><a href="community-similarity.html#cb233-7" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse,</span>
<span id="cb233-8"><a href="community-similarity.html#cb233-8" tabindex="-1"></a>                      <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb233-9"><a href="community-similarity.html#cb233-9" tabindex="-1"></a>                      <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb233-10"><a href="community-similarity.html#cb233-10" tabindex="-1"></a></span>
<span id="cb233-11"><a href="community-similarity.html#cb233-11" tabindex="-1"></a><span class="co"># Other metrics like Aitchison to clr-transformed data</span></span>
<span id="cb233-12"><a href="community-similarity.html#cb233-12" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse,</span>
<span id="cb233-13"><a href="community-similarity.html#cb233-13" tabindex="-1"></a>                      <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb233-14"><a href="community-similarity.html#cb233-14" tabindex="-1"></a>                      <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb233-15"><a href="community-similarity.html#cb233-15" tabindex="-1"></a>                      <span class="at">pseudocount =</span> <span class="cn">TRUE</span>)</span>
<span id="cb233-16"><a href="community-similarity.html#cb233-16" tabindex="-1"></a></span>
<span id="cb233-17"><a href="community-similarity.html#cb233-17" tabindex="-1"></a><span class="co"># Add group information Feces yes/no</span></span>
<span id="cb233-18"><a href="community-similarity.html#cb233-18" tabindex="-1"></a>tse<span class="sc">$</span>Group <span class="ot">&lt;-</span> tse<span class="sc">$</span>SampleType <span class="sc">==</span> <span class="st">&quot;Feces&quot;</span></span></code></pre></div>
<div id="comparing-communities-by-beta-diversity-analysis" class="section level3 hasAnchor" number="8.1.1">
<h3><span class="header-section-number">8.1.1</span> Comparing communities by beta diversity analysis<a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A typical comparison of community compositions starts with a visual
representation of the groups by a 2D ordination. Then we estimate relative
abundances and MDS ordination based on Bray-Curtis index between the groups,
and visualize the results.</p>
<p>In the following examples dissimilarity is calculated with the function supplied
to the <code>FUN</code> argument. Several metrics of beta diversity are defined by the <code>vegdist</code>
function of the vegan package, which is often used in this context. However, such
custom functions created by the user also work, as long as they return a <code>dist</code>
object. In either case, this function is then applied to calculate reduced
dimensions via an ordination method, the results of which can be stored in the
<code>reducedDim</code> slot of the TreeSE. This entire process is contained by the <code>runMDS</code>
and <code>runNMDS</code> functions.</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="community-similarity.html#cb234-1" tabindex="-1"></a><span class="co"># Load package to plot reducedDim</span></span>
<span id="cb234-2"><a href="community-similarity.html#cb234-2" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb234-3"><a href="community-similarity.html#cb234-3" tabindex="-1"></a></span>
<span id="cb234-4"><a href="community-similarity.html#cb234-4" tabindex="-1"></a><span class="co"># Run PCoA on relabundance assay with Bray-Curtis distances</span></span>
<span id="cb234-5"><a href="community-similarity.html#cb234-5" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(tse,</span>
<span id="cb234-6"><a href="community-similarity.html#cb234-6" tabindex="-1"></a>              <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb234-7"><a href="community-similarity.html#cb234-7" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;bray&quot;</span>,</span>
<span id="cb234-8"><a href="community-similarity.html#cb234-8" tabindex="-1"></a>              <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb234-9"><a href="community-similarity.html#cb234-9" tabindex="-1"></a>              <span class="at">name =</span> <span class="st">&quot;MDS_bray&quot;</span>)</span></code></pre></div>
<p>Sample dissimilarity can be visualized on a lower-dimensional display (typically
2D) using the <code>plotReducedDim</code> function from the scater package. This also
provides tools to incorporate additional information encoded by color, shape,
size and other aesthetics. Can you find any difference between the groups?</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="community-similarity.html#cb235-1" tabindex="-1"></a><span class="co"># Create ggplot object</span></span>
<span id="cb235-2"><a href="community-similarity.html#cb235-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plotReducedDim</span>(tse, <span class="st">&quot;MDS_bray&quot;</span>,</span>
<span id="cb235-3"><a href="community-similarity.html#cb235-3" tabindex="-1"></a>                    <span class="at">colour_by =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb235-4"><a href="community-similarity.html#cb235-4" tabindex="-1"></a></span>
<span id="cb235-5"><a href="community-similarity.html#cb235-5" tabindex="-1"></a><span class="co"># Calculate explained variance</span></span>
<span id="cb235-6"><a href="community-similarity.html#cb235-6" tabindex="-1"></a>e <span class="ot">&lt;-</span> <span class="fu">attr</span>(<span class="fu">reducedDim</span>(tse, <span class="st">&quot;MDS_bray&quot;</span>), <span class="st">&quot;eig&quot;</span>)</span>
<span id="cb235-7"><a href="community-similarity.html#cb235-7" tabindex="-1"></a>rel_eig <span class="ot">&lt;-</span> e <span class="sc">/</span> <span class="fu">sum</span>(e[e <span class="sc">&gt;</span> <span class="dv">0</span>])</span>
<span id="cb235-8"><a href="community-similarity.html#cb235-8" tabindex="-1"></a></span>
<span id="cb235-9"><a href="community-similarity.html#cb235-9" tabindex="-1"></a><span class="co"># Add explained variance for each axis</span></span>
<span id="cb235-10"><a href="community-similarity.html#cb235-10" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste</span>(<span class="st">&quot;PCoA 1 (&quot;</span>, <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> rel_eig[[<span class="dv">1</span>]], <span class="dv">1</span>), <span class="st">&quot;%&quot;</span>, <span class="st">&quot;)&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb235-11"><a href="community-similarity.html#cb235-11" tabindex="-1"></a>              <span class="at">y =</span> <span class="fu">paste</span>(<span class="st">&quot;PCoA 2 (&quot;</span>, <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> rel_eig[[<span class="dv">2</span>]], <span class="dv">1</span>), <span class="st">&quot;%&quot;</span>, <span class="st">&quot;)&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span>
<span id="cb235-12"><a href="community-similarity.html#cb235-12" tabindex="-1"></a></span>
<span id="cb235-13"><a href="community-similarity.html#cb235-13" tabindex="-1"></a>p</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-mds-bray-curtis"></span>
<img src="20_beta_diversity_files/figure-html/plot-mds-bray-curtis-1.png" alt="MDS plot based on the Bray-Curtis distances on the GlobalPattern dataset." width="3000" />
<p class="caption">
Figure 8.1: MDS plot based on the Bray-Curtis distances on the GlobalPattern dataset.
</p>
</div>
<p>A few combinations of beta diversity metrics and assay types are typically
used. For instance, Bray-Curtis dissimilarity and Euclidean distance are often
applied to the relative abundance and the clr assays, respectively. Besides
<strong>beta diversity metric</strong> and <strong>assay type</strong>, the <strong>PCoA algorithm</strong> is also a
variable that should be considered. Below, we show how the choice of these three
factors can affect the resulting lower-dimensional data.</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="community-similarity.html#cb236-1" tabindex="-1"></a><span class="co"># Run NMDS on relabundance assay with Bray-Curtis distances</span></span>
<span id="cb236-2"><a href="community-similarity.html#cb236-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runNMDS</span>(tse,</span>
<span id="cb236-3"><a href="community-similarity.html#cb236-3" tabindex="-1"></a>               <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb236-4"><a href="community-similarity.html#cb236-4" tabindex="-1"></a>               <span class="at">method =</span> <span class="st">&quot;bray&quot;</span>,</span>
<span id="cb236-5"><a href="community-similarity.html#cb236-5" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb236-6"><a href="community-similarity.html#cb236-6" tabindex="-1"></a>               <span class="at">name =</span> <span class="st">&quot;NMDS_bray&quot;</span>)</span>
<span id="cb236-7"><a href="community-similarity.html#cb236-7" tabindex="-1"></a></span>
<span id="cb236-8"><a href="community-similarity.html#cb236-8" tabindex="-1"></a><span class="co"># Run MDS on clr assay with Aitchison distances</span></span>
<span id="cb236-9"><a href="community-similarity.html#cb236-9" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(tse,</span>
<span id="cb236-10"><a href="community-similarity.html#cb236-10" tabindex="-1"></a>              <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb236-11"><a href="community-similarity.html#cb236-11" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>,</span>
<span id="cb236-12"><a href="community-similarity.html#cb236-12" tabindex="-1"></a>              <span class="at">assay.type =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb236-13"><a href="community-similarity.html#cb236-13" tabindex="-1"></a>              <span class="at">name =</span> <span class="st">&quot;MDS_aitchison&quot;</span>)</span>
<span id="cb236-14"><a href="community-similarity.html#cb236-14" tabindex="-1"></a></span>
<span id="cb236-15"><a href="community-similarity.html#cb236-15" tabindex="-1"></a><span class="co"># Run NMDS on clr assay with Euclidean distances</span></span>
<span id="cb236-16"><a href="community-similarity.html#cb236-16" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runNMDS</span>(tse,</span>
<span id="cb236-17"><a href="community-similarity.html#cb236-17" tabindex="-1"></a>               <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb236-18"><a href="community-similarity.html#cb236-18" tabindex="-1"></a>               <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>,</span>
<span id="cb236-19"><a href="community-similarity.html#cb236-19" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb236-20"><a href="community-similarity.html#cb236-20" tabindex="-1"></a>               <span class="at">name =</span> <span class="st">&quot;NMDS_aitchison&quot;</span>)</span></code></pre></div>
<p>Multiple ordination plots are combined into a multi-panel plot with the
patchwork package, so that different methods can be compared to find similarities
between them or select the most suitable one to visualize beta diversity in the
light of the research question.</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="community-similarity.html#cb237-1" tabindex="-1"></a><span class="co"># Load package for multi-panel plotting</span></span>
<span id="cb237-2"><a href="community-similarity.html#cb237-2" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb237-3"><a href="community-similarity.html#cb237-3" tabindex="-1"></a></span>
<span id="cb237-4"><a href="community-similarity.html#cb237-4" tabindex="-1"></a><span class="co"># Generate plots for all 4 reducedDims</span></span>
<span id="cb237-5"><a href="community-similarity.html#cb237-5" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;MDS_bray&quot;</span>, <span class="st">&quot;MDS_aitchison&quot;</span>,</span>
<span id="cb237-6"><a href="community-similarity.html#cb237-6" tabindex="-1"></a>                  <span class="st">&quot;NMDS_bray&quot;</span>, <span class="st">&quot;NMDS_aitchison&quot;</span>),</span>
<span id="cb237-7"><a href="community-similarity.html#cb237-7" tabindex="-1"></a>                plotReducedDim,</span>
<span id="cb237-8"><a href="community-similarity.html#cb237-8" tabindex="-1"></a>                <span class="at">object =</span> tse,</span>
<span id="cb237-9"><a href="community-similarity.html#cb237-9" tabindex="-1"></a>                <span class="at">colour_by =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb237-10"><a href="community-similarity.html#cb237-10" tabindex="-1"></a></span>
<span id="cb237-11"><a href="community-similarity.html#cb237-11" tabindex="-1"></a><span class="co"># Generate multi-panel plot</span></span>
<span id="cb237-12"><a href="community-similarity.html#cb237-12" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots) <span class="sc">+</span></span>
<span id="cb237-13"><a href="community-similarity.html#cb237-13" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">&quot;collect&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-2"></span>
<img src="20_beta_diversity_files/figure-html/unnamed-chunk-2-1.png" alt="Comparison of MDS and NMDS plots based on the Bray-Curtis or Aitchison distances on the GlobalPattern dataset." width="3000" />
<p class="caption">
Figure 8.2: Comparison of MDS and NMDS plots based on the Bray-Curtis or Aitchison distances on the GlobalPattern dataset.
</p>
</div>
<p>The <em>Unifrac</em> method is a special case, as it requires data on the
relationship of features in form on a <code>phylo</code> tree. <code>calculateUnifrac</code>
performs the calculation to return a <code>dist</code> object, which can again be
used within <code>runMDS</code>.</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="community-similarity.html#cb238-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(tse,</span>
<span id="cb238-2"><a href="community-similarity.html#cb238-2" tabindex="-1"></a>              <span class="at">FUN =</span> mia<span class="sc">::</span>calculateUnifrac,</span>
<span id="cb238-3"><a href="community-similarity.html#cb238-3" tabindex="-1"></a>              <span class="at">name =</span> <span class="st">&quot;Unifrac&quot;</span>,</span>
<span id="cb238-4"><a href="community-similarity.html#cb238-4" tabindex="-1"></a>              <span class="at">tree =</span> <span class="fu">rowTree</span>(tse),</span>
<span id="cb238-5"><a href="community-similarity.html#cb238-5" tabindex="-1"></a>              <span class="at">ntop =</span> <span class="fu">nrow</span>(tse),</span>
<span id="cb238-6"><a href="community-similarity.html#cb238-6" tabindex="-1"></a>              <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb238-7"><a href="community-similarity.html#cb238-7" tabindex="-1"></a></span>
<span id="cb238-8"><a href="community-similarity.html#cb238-8" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse, <span class="st">&quot;Unifrac&quot;</span>,</span>
<span id="cb238-9"><a href="community-similarity.html#cb238-9" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-unifrac"></span>
<img src="20_beta_diversity_files/figure-html/plot-unifrac-1.png" alt="Unifrac distances scaled by MDS of the GlobalPattern dataset." width="3000" />
<p class="caption">
Figure 8.3: Unifrac distances scaled by MDS of the GlobalPattern dataset.
</p>
</div>
</div>
<div id="other-ord-methods" class="section level3 hasAnchor" number="8.1.2">
<h3><span class="header-section-number">8.1.2</span> Other ordination methods<a href="community-similarity.html#other-ord-methods" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Other dimension reduction methods, such as PCA and UMAP, are inherited from the
scater package.</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="community-similarity.html#cb239-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(tse,</span>
<span id="cb239-2"><a href="community-similarity.html#cb239-2" tabindex="-1"></a>              <span class="at">name =</span> <span class="st">&quot;PCA&quot;</span>,</span>
<span id="cb239-3"><a href="community-similarity.html#cb239-3" tabindex="-1"></a>              <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb239-4"><a href="community-similarity.html#cb239-4" tabindex="-1"></a>              <span class="at">ncomponents =</span> <span class="dv">10</span>)</span>
<span id="cb239-5"><a href="community-similarity.html#cb239-5" tabindex="-1"></a></span>
<span id="cb239-6"><a href="community-similarity.html#cb239-6" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse, <span class="st">&quot;PCA&quot;</span>,</span>
<span id="cb239-7"><a href="community-similarity.html#cb239-7" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-pca"></span>
<img src="20_beta_diversity_files/figure-html/plot-pca-1.png" alt="PCA plot on the GlobalPatterns data set containing sample from different sources." width="3000" />
<p class="caption">
Figure 8.4: PCA plot on the GlobalPatterns data set containing sample from different sources.
</p>
</div>
<p>As mentioned before, applicability of the different methods depends on your
sample set and research question.</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="community-similarity.html#cb240-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(tse,</span>
<span id="cb240-2"><a href="community-similarity.html#cb240-2" tabindex="-1"></a>               <span class="at">name =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb240-3"><a href="community-similarity.html#cb240-3" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb240-4"><a href="community-similarity.html#cb240-4" tabindex="-1"></a>               <span class="at">ncomponents =</span> <span class="dv">3</span>)</span>
<span id="cb240-5"><a href="community-similarity.html#cb240-5" tabindex="-1"></a></span>
<span id="cb240-6"><a href="community-similarity.html#cb240-6" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse, <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb240-7"><a href="community-similarity.html#cb240-7" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">&quot;Group&quot;</span>,</span>
<span id="cb240-8"><a href="community-similarity.html#cb240-8" tabindex="-1"></a>               <span class="at">ncomponents =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>))</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-umap"></span>
<img src="20_beta_diversity_files/figure-html/plot-umap-1.png" alt="UMAP plot on the GlobalPatterns data set containing sample from different sources." width="3000" />
<p class="caption">
Figure 8.5: UMAP plot on the GlobalPatterns data set containing sample from different sources.
</p>
</div>
</div>
<div id="explained-variance" class="section level3 hasAnchor" number="8.1.3">
<h3><span class="header-section-number">8.1.3</span> Explained variance<a href="community-similarity.html#explained-variance" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The percentage of explained variance is typically shown for PCA
ordination plots. This quantifies the proportion of overall variance
in the data that is captured by the PCA axes, or how well the
ordination axes reflect the original distances.</p>
<p>Sometimes a similar measure is shown for MDS/PCoA. The interpretation
is generally different, however, and hence we do not recommend using
it. PCA is a special case of PCoA with Euclidean distances. With
non-Euclidean dissimilarities PCoA uses a trick where the pointwise
dissimilarities are first cast into similarities in a Euclidean space
(with some information loss i.e. stress) and then projected to the
maximal variance axes. In this case, the maximal variance axes do not
directly reflect the correspondence of the projected distances and
original distances, as they do for PCA.</p>
<p>In typical use cases, we would like to know how well the ordination
reflects the original similarity structures; then the quantity of
interest is the so-called “stress” function, which measures the
difference in pairwise similarities between the data points in the
original (high-dimensional) vs. projected (low-dimensional) space.</p>
<p>Hence, we propose that for PCoA and other ordination methods, users
would report relative stress, which varies in the unit interval and is better
if smaller. This can be calculated as shown below.</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="community-similarity.html#cb241-1" tabindex="-1"></a><span class="co"># Load vegan package</span></span>
<span id="cb241-2"><a href="community-similarity.html#cb241-2" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb241-3"><a href="community-similarity.html#cb241-3" tabindex="-1"></a></span>
<span id="cb241-4"><a href="community-similarity.html#cb241-4" tabindex="-1"></a><span class="co"># Quantify dissimilarities in the original feature space</span></span>
<span id="cb241-5"><a href="community-similarity.html#cb241-5" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse, <span class="st">&quot;relabundance&quot;</span>) <span class="co"># Pick relabunance assay separately</span></span>
<span id="cb241-6"><a href="community-similarity.html#cb241-6" tabindex="-1"></a>d0 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">vegdist</span>(<span class="fu">t</span>(x), <span class="st">&quot;bray&quot;</span>))</span>
<span id="cb241-7"><a href="community-similarity.html#cb241-7" tabindex="-1"></a></span>
<span id="cb241-8"><a href="community-similarity.html#cb241-8" tabindex="-1"></a><span class="co"># PCoA Ordination</span></span>
<span id="cb241-9"><a href="community-similarity.html#cb241-9" tabindex="-1"></a>pcoa <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cmdscale</span>(d0, <span class="at">k =</span> <span class="dv">2</span>))</span>
<span id="cb241-10"><a href="community-similarity.html#cb241-10" tabindex="-1"></a><span class="fu">names</span>(pcoa) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PCoA1&quot;</span>, <span class="st">&quot;PCoA2&quot;</span>)</span>
<span id="cb241-11"><a href="community-similarity.html#cb241-11" tabindex="-1"></a></span>
<span id="cb241-12"><a href="community-similarity.html#cb241-12" tabindex="-1"></a><span class="co"># Quantify dissimilarities in the ordination space</span></span>
<span id="cb241-13"><a href="community-similarity.html#cb241-13" tabindex="-1"></a>dp <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">dist</span>(pcoa))</span>
<span id="cb241-14"><a href="community-similarity.html#cb241-14" tabindex="-1"></a></span>
<span id="cb241-15"><a href="community-similarity.html#cb241-15" tabindex="-1"></a><span class="co"># Calculate stress i.e. relative difference in the original and</span></span>
<span id="cb241-16"><a href="community-similarity.html#cb241-16" tabindex="-1"></a><span class="co"># projected dissimilarities</span></span>
<span id="cb241-17"><a href="community-similarity.html#cb241-17" tabindex="-1"></a>stress <span class="ot">&lt;-</span> <span class="fu">sum</span>((dp <span class="sc">-</span> d0)<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> <span class="fu">sum</span>(d0<span class="sc">^</span><span class="dv">2</span>)</span></code></pre></div>
<p>A Shepard plot visualizes the original versus the ordinated dissimilarity
between the observations.</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="community-similarity.html#cb242-1" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">as.vector</span>(d0))</span>
<span id="cb242-2"><a href="community-similarity.html#cb242-2" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">d0 =</span> <span class="fu">as.vector</span>(d0)[ord],</span>
<span id="cb242-3"><a href="community-similarity.html#cb242-3" tabindex="-1"></a>                 <span class="at">dmds =</span> <span class="fu">as.vector</span>(dp)[ord])</span>
<span id="cb242-4"><a href="community-similarity.html#cb242-4" tabindex="-1"></a></span>
<span id="cb242-5"><a href="community-similarity.html#cb242-5" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> d0, <span class="at">y =</span> dmds)) <span class="sc">+</span></span>
<span id="cb242-6"><a href="community-similarity.html#cb242-6" tabindex="-1"></a>  <span class="fu">geom_smooth</span>() <span class="sc">+</span></span>
<span id="cb242-7"><a href="community-similarity.html#cb242-7" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span>    </span>
<span id="cb242-8"><a href="community-similarity.html#cb242-8" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Shepard plot&quot;</span>,</span>
<span id="cb242-9"><a href="community-similarity.html#cb242-9" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;Original distance&quot;</span>,</span>
<span id="cb242-10"><a href="community-similarity.html#cb242-10" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;MDS distance&quot;</span>,   </span>
<span id="cb242-11"><a href="community-similarity.html#cb242-11" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="fu">paste</span>(<span class="st">&quot;Stress:&quot;</span>, <span class="fu">round</span>(stress, <span class="dv">2</span>))) <span class="sc">+</span></span>
<span id="cb242-12"><a href="community-similarity.html#cb242-12" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="20_beta_diversity_files/figure-html/shepard-1.png" width="3000" /></p>
</div>
</div>
<div id="supervised-ordination" class="section level2 hasAnchor" number="8.2">
<h2><span class="header-section-number">8.2</span> Supervised ordination<a href="community-similarity.html#supervised-ordination" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>dbRDA is a supervised counterpart of PCoA, that is, it takes into account the
covariates specified by the user to maximize the variance with respect to the
them. The result shows how much each covariate affects beta diversity. The table
below illustrates the relation between supervised and unsupervised ordination
methods.</p>
<table>
<colgroup>
<col width="34%" />
<col width="31%" />
<col width="34%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"></th>
<th align="center">supervised ordination</th>
<th align="center">unsupervised ordination</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Euclidean distance</td>
<td align="center">RDA</td>
<td align="center">PCA</td>
</tr>
<tr class="even">
<td align="center">non-Euclidean distance</td>
<td align="center">dbRDA</td>
<td align="center">PCoA/MDS, NMDS and UMAP</td>
</tr>
</tbody>
</table>
<p>We demonstrate the usage of dbRDA with the enterotype dataset, where samples
correspond to patients. The colData contains the clinical status of each patient
and a few covariates such as their gender and age.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="community-similarity.html#cb243-1" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb243-2"><a href="community-similarity.html#cb243-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;enterotype&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb243-3"><a href="community-similarity.html#cb243-3" tabindex="-1"></a>tse2 <span class="ot">&lt;-</span> enterotype</span>
<span id="cb243-4"><a href="community-similarity.html#cb243-4" tabindex="-1"></a></span>
<span id="cb243-5"><a href="community-similarity.html#cb243-5" tabindex="-1"></a><span class="co"># Apply relative transform</span></span>
<span id="cb243-6"><a href="community-similarity.html#cb243-6" tabindex="-1"></a>tse2 <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse2,</span>
<span id="cb243-7"><a href="community-similarity.html#cb243-7" tabindex="-1"></a>                       <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span></code></pre></div>
<p>dbRDA can be perfomed with the <code>runRDA</code> function. In addition to the arguments
previously defined for <a href="community-similarity.html#unsupervised-ordination">unsupervised ordination</a>, this
function takes a formula to control for variables and an action to treat missing
values. Along with clinical status, which is the main outcome, we control for
gender and age, and exclude observations where one of these variables is missing.</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="community-similarity.html#cb244-1" tabindex="-1"></a><span class="co"># Perform RDA</span></span>
<span id="cb244-2"><a href="community-similarity.html#cb244-2" tabindex="-1"></a>tse2 <span class="ot">&lt;-</span> <span class="fu">runRDA</span>(tse2,</span>
<span id="cb244-3"><a href="community-similarity.html#cb244-3" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb244-4"><a href="community-similarity.html#cb244-4" tabindex="-1"></a>               <span class="at">formula =</span> assay <span class="sc">~</span> ClinicalStatus <span class="sc">+</span> Gender <span class="sc">+</span> Age,</span>
<span id="cb244-5"><a href="community-similarity.html#cb244-5" tabindex="-1"></a>               <span class="at">distance =</span> <span class="st">&quot;bray&quot;</span>,</span>
<span id="cb244-6"><a href="community-similarity.html#cb244-6" tabindex="-1"></a>               <span class="at">na.action =</span> na.exclude)</span>
<span id="cb244-7"><a href="community-similarity.html#cb244-7" tabindex="-1"></a></span>
<span id="cb244-8"><a href="community-similarity.html#cb244-8" tabindex="-1"></a><span class="co"># Store results of PERMANOVA test</span></span>
<span id="cb244-9"><a href="community-similarity.html#cb244-9" tabindex="-1"></a>rda_info <span class="ot">&lt;-</span> <span class="fu">attr</span>(<span class="fu">reducedDim</span>(tse2, <span class="st">&quot;RDA&quot;</span>), <span class="st">&quot;significance&quot;</span>)</span></code></pre></div>
<p>The importance of each variable on the similarity between samples can be
assessed from the results of PERMANOVA, automatically provided by the <code>runRDA</code>
function. We see that both clinical status and age explain more than 10% of the
variance, but only age shows statistical significance.</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="community-similarity.html#cb245-1" tabindex="-1"></a>rda_info<span class="sc">$</span>permanova <span class="sc">|&gt;</span></span>
<span id="cb245-2"><a href="community-similarity.html#cb245-2" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code></pre></div>
<table>
<colgroup>
<col width="20%" />
<col width="4%" />
<col width="12%" />
<col width="8%" />
<col width="9%" />
<col width="20%" />
<col width="25%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Df</th>
<th align="right">SumOfSqs</th>
<th align="right">F</th>
<th align="right">Pr(&gt;F)</th>
<th align="right">Total variance</th>
<th align="right">Explained variance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Model</td>
<td align="right">6</td>
<td align="right">1.1157</td>
<td align="right">1.940</td>
<td align="right">0.030</td>
<td align="right">3.991</td>
<td align="right">0.2795</td>
</tr>
<tr class="even">
<td align="left">ClinicalStatus</td>
<td align="right">4</td>
<td align="right">0.5837</td>
<td align="right">1.522</td>
<td align="right">0.128</td>
<td align="right">3.991</td>
<td align="right">0.1463</td>
</tr>
<tr class="odd">
<td align="left">Gender</td>
<td align="right">1</td>
<td align="right">0.1679</td>
<td align="right">1.751</td>
<td align="right">0.101</td>
<td align="right">3.991</td>
<td align="right">0.0421</td>
</tr>
<tr class="even">
<td align="left">Age</td>
<td align="right">1</td>
<td align="right">0.5245</td>
<td align="right">5.471</td>
<td align="right">0.001</td>
<td align="right">3.991</td>
<td align="right">0.1314</td>
</tr>
<tr class="odd">
<td align="left">Residual</td>
<td align="right">30</td>
<td align="right">2.8757</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">3.991</td>
<td align="right">0.7205</td>
</tr>
</tbody>
</table>
<p>To ensure that the homogeneity assumption holds, we retrieve the corresponding
information from the results of RDA. In this case, none of the p-values is lower
than the significance threshold, and thus homogeneity is observed.</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="community-similarity.html#cb246-1" tabindex="-1"></a>rda_info<span class="sc">$</span>homogeneity <span class="sc">|&gt;</span></span>
<span id="cb246-2"><a href="community-similarity.html#cb246-2" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="16%" />
<col width="3%" />
<col width="7%" />
<col width="8%" />
<col width="8%" />
<col width="7%" />
<col width="7%" />
<col width="16%" />
<col width="21%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Df</th>
<th align="right">Sum Sq</th>
<th align="right">Mean Sq</th>
<th align="right">F</th>
<th align="right">N.Perm</th>
<th align="right">Pr(&gt;F)</th>
<th align="right">Total variance</th>
<th align="right">Explained variance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ClinicalStatus</td>
<td align="right">4</td>
<td align="right">0.2511</td>
<td align="right">0.0628</td>
<td align="right">2.7440</td>
<td align="right">999</td>
<td align="right">0.113</td>
<td align="right">1.0288</td>
<td align="right">0.2440</td>
</tr>
<tr class="even">
<td align="left">Gender</td>
<td align="right">1</td>
<td align="right">0.0103</td>
<td align="right">0.0103</td>
<td align="right">0.4158</td>
<td align="right">999</td>
<td align="right">0.525</td>
<td align="right">0.9283</td>
<td align="right">0.0111</td>
</tr>
<tr class="odd">
<td align="left">Age</td>
<td align="right">29</td>
<td align="right">0.3272</td>
<td align="right">0.0113</td>
<td align="right">17.0255</td>
<td align="right">999</td>
<td align="right">0.415</td>
<td align="right">0.3319</td>
<td align="right">0.9860</td>
</tr>
</tbody>
</table>
<p>Next, we proceed to visualize the weight and significance of each variable on
the similarity between samples with an RDA plot, which can be generated with
the <code>plotRDA</code> function from the miaViz package.</p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="community-similarity.html#cb247-1" tabindex="-1"></a><span class="co"># Load packages for plotting function</span></span>
<span id="cb247-2"><a href="community-similarity.html#cb247-2" tabindex="-1"></a><span class="fu">library</span>(miaViz)</span>
<span id="cb247-3"><a href="community-similarity.html#cb247-3" tabindex="-1"></a></span>
<span id="cb247-4"><a href="community-similarity.html#cb247-4" tabindex="-1"></a><span class="co"># Generate RDA plot coloured by clinical status</span></span>
<span id="cb247-5"><a href="community-similarity.html#cb247-5" tabindex="-1"></a><span class="fu">plotRDA</span>(tse2, <span class="st">&quot;RDA&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;ClinicalStatus&quot;</span>)</span></code></pre></div>
<p><img src="20_beta_diversity_files/figure-html/plot-rda-1.png" width="3000" /></p>
<p>From the plot above, we can see that only age significantly describes
differences between the microbial profiles of different samples. Such visual
approach complements the previous results of PERMANOVA.</p>
</div>
<div id="case-studies" class="section level2 hasAnchor" number="8.3">
<h2><span class="header-section-number">8.3</span> Case studies<a href="community-similarity.html#case-studies" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="pcoa-genus" class="section level4 hasAnchor" number="8.3.0.1">
<h4><span class="header-section-number">8.3.0.1</span> Visualizing the most dominant genus on PCoA<a href="community-similarity.html#pcoa-genus" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In this section, we visualize the most dominant genus on PCoA. A similar
visualization was proposed by <span class="citation">(2021)</span>. First, we agglomerate the
data at the Genus level and get the dominant taxa per sample.</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="community-similarity.html#cb248-1" tabindex="-1"></a><span class="co"># Agglomerate to genus level</span></span>
<span id="cb248-2"><a href="community-similarity.html#cb248-2" tabindex="-1"></a>tse_genus <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse,</span>
<span id="cb248-3"><a href="community-similarity.html#cb248-3" tabindex="-1"></a>                                 <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb248-4"><a href="community-similarity.html#cb248-4" tabindex="-1"></a></span>
<span id="cb248-5"><a href="community-similarity.html#cb248-5" tabindex="-1"></a><span class="co"># Convert to relative abundances</span></span>
<span id="cb248-6"><a href="community-similarity.html#cb248-6" tabindex="-1"></a>tse_genus <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse,</span>
<span id="cb248-7"><a href="community-similarity.html#cb248-7" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb248-8"><a href="community-similarity.html#cb248-8" tabindex="-1"></a>                            <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb248-9"><a href="community-similarity.html#cb248-9" tabindex="-1"></a></span>
<span id="cb248-10"><a href="community-similarity.html#cb248-10" tabindex="-1"></a><span class="co"># Add info on dominant genus per sample</span></span>
<span id="cb248-11"><a href="community-similarity.html#cb248-11" tabindex="-1"></a>tse_genus <span class="ot">&lt;-</span> <span class="fu">addPerSampleDominantFeatures</span>(tse_genus,</span>
<span id="cb248-12"><a href="community-similarity.html#cb248-12" tabindex="-1"></a>                                          <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb248-13"><a href="community-similarity.html#cb248-13" tabindex="-1"></a>                                          <span class="at">name =</span> <span class="st">&quot;dominant_taxa&quot;</span>)</span>
<span id="cb248-14"><a href="community-similarity.html#cb248-14" tabindex="-1"></a><span class="co"># Overview</span></span>
<span id="cb248-15"><a href="community-similarity.html#cb248-15" tabindex="-1"></a><span class="fu">countDominantFeatures</span>(tse_genus, <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">digits =</span> <span class="dv">3</span>, <span class="at">name =</span> <span class="st">&quot;dominant_taxa&quot;</span>)</span></code></pre></div>
<pre><code>## # A tibble: 17 × 3
##    dominant_taxa                  n rel_freq
##    &lt;chr&gt;                      &lt;int&gt;    &lt;dbl&gt;
##  1 Genus:Bacteroides              5    0.192
##  2 Order:Stramenopiles            4    0.154
##  3 Family:Desulfobulbaceae        2    0.077
##  4 Genus:Streptococcus            2    0.077
##  5 Class:Chloracidobacteria       1    0.038
##  6 Family:ACK-M1                  1    0.038
##  7 Family:Flavobacteriaceae       1    0.038
##  8 Family:Moraxellaceae           1    0.038
##  9 Family:Ruminococcaceae         1    0.038
## 10 Genus:CandidatusSolibacter     1    0.038
## 11 Genus:Dolichospermum           1    0.038
## 12 Genus:Faecalibacterium         1    0.038
## 13 Genus:MC18                     1    0.038
## 14 Genus:Neisseria                1    0.038
## 15 Genus:Prochlorococcus          1    0.038
## 16 Genus:Veillonella              1    0.038
## 17 Order:Chromatiales             1    0.038</code></pre>
<p>Next, we perform PCoA with Bray-Curtis dissimilarity.</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="community-similarity.html#cb250-1" tabindex="-1"></a>tse_genus <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(tse_genus,</span>
<span id="cb250-2"><a href="community-similarity.html#cb250-2" tabindex="-1"></a>                    <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb250-3"><a href="community-similarity.html#cb250-3" tabindex="-1"></a>                    <span class="at">name =</span> <span class="st">&quot;PCoA_BC&quot;</span>,</span>
<span id="cb250-4"><a href="community-similarity.html#cb250-4" tabindex="-1"></a>                    <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>)</span></code></pre></div>
<p>Finally, we get the top taxa and and visualize their abundances on PCoA. Note
that A 3D interactive version of the plot below can be found in <a href="extras.html#extras">20</a>.</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="community-similarity.html#cb251-1" tabindex="-1"></a><span class="co"># Getting the top taxa</span></span>
<span id="cb251-2"><a href="community-similarity.html#cb251-2" tabindex="-1"></a>top_taxa <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(tse_genus,</span>
<span id="cb251-3"><a href="community-similarity.html#cb251-3" tabindex="-1"></a>                           <span class="at">top =</span> <span class="dv">6</span>,</span>
<span id="cb251-4"><a href="community-similarity.html#cb251-4" tabindex="-1"></a>                           <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb251-5"><a href="community-similarity.html#cb251-5" tabindex="-1"></a></span>
<span id="cb251-6"><a href="community-similarity.html#cb251-6" tabindex="-1"></a><span class="co"># Naming all the rest of non top-taxa as &quot;Other&quot;</span></span>
<span id="cb251-7"><a href="community-similarity.html#cb251-7" tabindex="-1"></a>most_abundant <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">colData</span>(tse_genus)<span class="sc">$</span>dominant_taxa,</span>
<span id="cb251-8"><a href="community-similarity.html#cb251-8" tabindex="-1"></a>                        <span class="cf">function</span>(x) {<span class="cf">if</span> (x <span class="sc">%in%</span> top_taxa) {x} <span class="cf">else</span> {<span class="st">&quot;Other&quot;</span>}})</span>
<span id="cb251-9"><a href="community-similarity.html#cb251-9" tabindex="-1"></a></span>
<span id="cb251-10"><a href="community-similarity.html#cb251-10" tabindex="-1"></a><span class="co"># Storing the previous results as a new column within colData</span></span>
<span id="cb251-11"><a href="community-similarity.html#cb251-11" tabindex="-1"></a><span class="fu">colData</span>(tse_genus)<span class="sc">$</span>most_abundant <span class="ot">&lt;-</span> <span class="fu">as.character</span>(most_abundant)</span>
<span id="cb251-12"><a href="community-similarity.html#cb251-12" tabindex="-1"></a></span>
<span id="cb251-13"><a href="community-similarity.html#cb251-13" tabindex="-1"></a><span class="co"># Calculating percentage of the most abundant</span></span>
<span id="cb251-14"><a href="community-similarity.html#cb251-14" tabindex="-1"></a>most_abundant_freq <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">as.character</span>(most_abundant))</span>
<span id="cb251-15"><a href="community-similarity.html#cb251-15" tabindex="-1"></a>most_abundant_percent <span class="ot">&lt;-</span> <span class="fu">round</span>(most_abundant_freq <span class="sc">/</span> <span class="fu">sum</span>(most_abundant_freq) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb251-16"><a href="community-similarity.html#cb251-16" tabindex="-1"></a></span>
<span id="cb251-17"><a href="community-similarity.html#cb251-17" tabindex="-1"></a><span class="co"># Retrieving the explained variance</span></span>
<span id="cb251-18"><a href="community-similarity.html#cb251-18" tabindex="-1"></a>e <span class="ot">&lt;-</span> <span class="fu">attr</span>(<span class="fu">reducedDim</span>(tse_genus, <span class="st">&quot;PCoA_BC&quot;</span>), <span class="st">&quot;eig&quot;</span>)</span>
<span id="cb251-19"><a href="community-similarity.html#cb251-19" tabindex="-1"></a>var_explained <span class="ot">&lt;-</span> e <span class="sc">/</span> <span class="fu">sum</span>(e[e <span class="sc">&gt;</span> <span class="dv">0</span>]) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb251-20"><a href="community-similarity.html#cb251-20" tabindex="-1"></a></span>
<span id="cb251-21"><a href="community-similarity.html#cb251-21" tabindex="-1"></a><span class="co"># Define colors for visualization</span></span>
<span id="cb251-22"><a href="community-similarity.html#cb251-22" tabindex="-1"></a>my_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;darkgray&quot;</span>, <span class="st">&quot;magenta&quot;</span>, <span class="st">&quot;darkgreen&quot;</span>, <span class="st">&quot;red&quot;</span>)</span>
<span id="cb251-23"><a href="community-similarity.html#cb251-23" tabindex="-1"></a></span>
<span id="cb251-24"><a href="community-similarity.html#cb251-24" tabindex="-1"></a><span class="co"># Visualization</span></span>
<span id="cb251-25"><a href="community-similarity.html#cb251-25" tabindex="-1"></a>plot <span class="ot">&lt;-</span><span class="fu">plotReducedDim</span>(tse_genus, <span class="st">&quot;PCoA_BC&quot;</span>,</span>
<span id="cb251-26"><a href="community-similarity.html#cb251-26" tabindex="-1"></a>                      <span class="at">colour_by =</span> <span class="st">&quot;most_abundant&quot;</span>) <span class="sc">+</span></span>
<span id="cb251-27"><a href="community-similarity.html#cb251-27" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> my_colors,</span>
<span id="cb251-28"><a href="community-similarity.html#cb251-28" tabindex="-1"></a>                      <span class="at">labels =</span> <span class="fu">paste0</span>(<span class="fu">names</span>(most_abundant_percent), <span class="st">&quot;(&quot;</span>, most_abundant_percent, <span class="st">&quot;%)&quot;</span>)) <span class="sc">+</span></span>
<span id="cb251-29"><a href="community-similarity.html#cb251-29" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste</span>(<span class="st">&quot;PC 1 (&quot;</span>, <span class="fu">round</span>(var_explained[<span class="dv">1</span>], <span class="dv">1</span>), <span class="st">&quot;%)&quot;</span>),</span>
<span id="cb251-30"><a href="community-similarity.html#cb251-30" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">paste</span>(<span class="st">&quot;PC 2 (&quot;</span>, <span class="fu">round</span>(var_explained[<span class="dv">2</span>], <span class="dv">1</span>), <span class="st">&quot;%)&quot;</span>),</span>
<span id="cb251-31"><a href="community-similarity.html#cb251-31" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb251-32"><a href="community-similarity.html#cb251-32" tabindex="-1"></a></span>
<span id="cb251-33"><a href="community-similarity.html#cb251-33" tabindex="-1"></a>plot</span></code></pre></div>
<p><img src="20_beta_diversity_files/figure-html/unnamed-chunk-5-1.png" width="3000" /></p>
<p>Similarly, we visualize and compare the sub-population.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="community-similarity.html#cb252-1" tabindex="-1"></a><span class="co"># Calculating the frequencies and percentages for both categories</span></span>
<span id="cb252-2"><a href="community-similarity.html#cb252-2" tabindex="-1"></a>freq_TRUE <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">as.character</span>(most_abundant[<span class="fu">colData</span>(tse_genus)<span class="sc">$</span>Group <span class="sc">==</span> <span class="cn">TRUE</span>]))</span>
<span id="cb252-3"><a href="community-similarity.html#cb252-3" tabindex="-1"></a>freq_FALSE <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">as.character</span>(most_abundant[<span class="fu">colData</span>(tse_genus)<span class="sc">$</span>Group <span class="sc">==</span> <span class="cn">FALSE</span>]))</span>
<span id="cb252-4"><a href="community-similarity.html#cb252-4" tabindex="-1"></a>percent_TRUE <span class="ot">&lt;-</span> <span class="fu">round</span>(freq_TRUE <span class="sc">/</span> <span class="fu">sum</span>(freq_TRUE) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb252-5"><a href="community-similarity.html#cb252-5" tabindex="-1"></a>percent_FALSE <span class="ot">&lt;-</span> <span class="fu">round</span>(freq_FALSE <span class="sc">/</span> <span class="fu">sum</span>(freq_FALSE) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb252-6"><a href="community-similarity.html#cb252-6" tabindex="-1"></a></span>
<span id="cb252-7"><a href="community-similarity.html#cb252-7" tabindex="-1"></a><span class="co"># Visualization</span></span>
<span id="cb252-8"><a href="community-similarity.html#cb252-8" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse_genus[ , <span class="fu">colData</span>(tse_genus)<span class="sc">$</span>Group <span class="sc">==</span> <span class="cn">TRUE</span>], <span class="st">&quot;PCoA_BC&quot;</span>,</span>
<span id="cb252-9"><a href="community-similarity.html#cb252-9" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">&quot;most_abundant&quot;</span>) <span class="sc">+</span></span>
<span id="cb252-10"><a href="community-similarity.html#cb252-10" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> my_colors,</span>
<span id="cb252-11"><a href="community-similarity.html#cb252-11" tabindex="-1"></a>                      <span class="at">labels =</span> <span class="fu">paste0</span>(<span class="fu">names</span>(percent_TRUE), <span class="st">&quot;(&quot;</span>, percent_TRUE, <span class="st">&quot;%)&quot;</span>)) <span class="sc">+</span></span>
<span id="cb252-12"><a href="community-similarity.html#cb252-12" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste</span>(<span class="st">&quot;PC 1 (&quot;</span>, <span class="fu">round</span>(var_explained[<span class="dv">1</span>], <span class="dv">1</span>), <span class="st">&quot;%)&quot;</span>),</span>
<span id="cb252-13"><a href="community-similarity.html#cb252-13" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">paste</span>(<span class="st">&quot;PC 2 (&quot;</span>, <span class="fu">round</span>(var_explained[<span class="dv">2</span>], <span class="dv">1</span>), <span class="st">&quot;%)&quot;</span>),</span>
<span id="cb252-14"><a href="community-similarity.html#cb252-14" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Group = TRUE&quot;</span>, <span class="at">color =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="20_beta_diversity_files/figure-html/unnamed-chunk-6-1.png" width="3000" /></p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="community-similarity.html#cb253-1" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse_genus[ , <span class="fu">colData</span>(tse_genus)<span class="sc">$</span>Group <span class="sc">==</span> <span class="cn">FALSE</span>], <span class="st">&quot;PCoA_BC&quot;</span>,</span>
<span id="cb253-2"><a href="community-similarity.html#cb253-2" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">&quot;most_abundant&quot;</span>) <span class="sc">+</span></span>
<span id="cb253-3"><a href="community-similarity.html#cb253-3" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> my_colors,</span>
<span id="cb253-4"><a href="community-similarity.html#cb253-4" tabindex="-1"></a>                      <span class="at">labels =</span> <span class="fu">paste0</span>(<span class="fu">names</span>(percent_FALSE), <span class="st">&quot;(&quot;</span>, percent_FALSE, <span class="st">&quot;%)&quot;</span>)) <span class="sc">+</span></span>
<span id="cb253-5"><a href="community-similarity.html#cb253-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste</span>(<span class="st">&quot;PC 1 (&quot;</span>, <span class="fu">round</span>(var_explained[<span class="dv">1</span>], <span class="dv">1</span>), <span class="st">&quot;%)&quot;</span>),</span>
<span id="cb253-6"><a href="community-similarity.html#cb253-6" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">paste</span>(<span class="st">&quot;PC 2 (&quot;</span>, <span class="fu">round</span>(var_explained[<span class="dv">2</span>], <span class="dv">1</span>), <span class="st">&quot;%)&quot;</span>),</span>
<span id="cb253-7"><a href="community-similarity.html#cb253-7" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Group = FALSE&quot;</span>, <span class="at">color =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="20_beta_diversity_files/figure-html/unnamed-chunk-6-2.png" width="3000" /></p>
</div>
<div id="dbrda-workflow" class="section level3 hasAnchor" number="8.3.1">
<h3><span class="header-section-number">8.3.1</span> Testing differences in community composition between sample groups<a href="community-similarity.html#dbrda-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Permutational Analysis of Variance (PERMANOVA; <span class="citation">(2001)</span>) is
a widely used non-parametric multivariate method that aims to
estimate the actual statistical significance of differences in the
observed community composition between two groups of samples.</p>
<p>PERMANOVA tests the hypothesis that the centroids and dispersion
of the community are equivalent between the compared groups. A p-value
smaller than the significance threshold indicates that the groups have a
different community composition. This method is implemented with the
<a href="https://www.rdocumentation.org/packages/vegan/versions/2.4-2/topics/adonis"><code>adonis2</code></a>
function from the vegan package.</p>
<p>By default, the argument <code>by</code> is set to <code>"terms"</code>, in which the order of variables
in the formula matters. In this case, each variable is analyzed sequentially, and
the result is different when more than 1 variable is introduced and their order differs.
Therefore, it is recommended to set <code>by = "margin"</code>, which specifies that the
marginal effect of each variable is analyzed individually. You can view a
comparison between the two designs in chapter <a href="extras.html#compare-permanova">20.2</a>.</p>
<p>We can perform PERMANOVA either with <code>adonis2</code> function or by first performing
dbRDA and then applying permutational test its results. An advantage of the latter
approach is that by doing so we can get coefficients: how much each taxa affects
the variation between communities.</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="community-similarity.html#cb254-1" tabindex="-1"></a><span class="co"># Agglomerate data to Species level</span></span>
<span id="cb254-2"><a href="community-similarity.html#cb254-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse,</span>
<span id="cb254-3"><a href="community-similarity.html#cb254-3" tabindex="-1"></a>                           <span class="at">rank =</span> <span class="st">&quot;Species&quot;</span>)</span>
<span id="cb254-4"><a href="community-similarity.html#cb254-4" tabindex="-1"></a></span>
<span id="cb254-5"><a href="community-similarity.html#cb254-5" tabindex="-1"></a><span class="co"># Set seed for reproducibility</span></span>
<span id="cb254-6"><a href="community-similarity.html#cb254-6" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1576</span>)</span>
<span id="cb254-7"><a href="community-similarity.html#cb254-7" tabindex="-1"></a><span class="co"># We choose 99 random permutations. Consider applying more (999 or 9999) in your</span></span>
<span id="cb254-8"><a href="community-similarity.html#cb254-8" tabindex="-1"></a><span class="co"># analysis. </span></span>
<span id="cb254-9"><a href="community-similarity.html#cb254-9" tabindex="-1"></a>permanova <span class="ot">&lt;-</span> <span class="fu">adonis2</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;relabundance&quot;</span>)) <span class="sc">~</span> Group,</span>
<span id="cb254-10"><a href="community-similarity.html#cb254-10" tabindex="-1"></a>                     <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>, <span class="co"># each term (here only &#39;Group&#39;) analyzed individually</span></span>
<span id="cb254-11"><a href="community-similarity.html#cb254-11" tabindex="-1"></a>                     <span class="at">data =</span> <span class="fu">colData</span>(tse),</span>
<span id="cb254-12"><a href="community-similarity.html#cb254-12" tabindex="-1"></a>                     <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>,</span>
<span id="cb254-13"><a href="community-similarity.html#cb254-13" tabindex="-1"></a>                     <span class="at">permutations =</span> <span class="dv">99</span>)</span>
<span id="cb254-14"><a href="community-similarity.html#cb254-14" tabindex="-1"></a></span>
<span id="cb254-15"><a href="community-similarity.html#cb254-15" tabindex="-1"></a><span class="co"># Set seed for reproducibility</span></span>
<span id="cb254-16"><a href="community-similarity.html#cb254-16" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1576</span>)</span>
<span id="cb254-17"><a href="community-similarity.html#cb254-17" tabindex="-1"></a><span class="co"># Perform dbRDA</span></span>
<span id="cb254-18"><a href="community-similarity.html#cb254-18" tabindex="-1"></a>dbrda <span class="ot">&lt;-</span> <span class="fu">dbrda</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse,<span class="st">&quot;relabundance&quot;</span>)) <span class="sc">~</span> Group, </span>
<span id="cb254-19"><a href="community-similarity.html#cb254-19" tabindex="-1"></a>               <span class="at">data =</span> <span class="fu">colData</span>(tse))</span>
<span id="cb254-20"><a href="community-similarity.html#cb254-20" tabindex="-1"></a><span class="co"># Perform permutational analysis</span></span>
<span id="cb254-21"><a href="community-similarity.html#cb254-21" tabindex="-1"></a>permanova2 <span class="ot">&lt;-</span> <span class="fu">anova.cca</span>(dbrda,</span>
<span id="cb254-22"><a href="community-similarity.html#cb254-22" tabindex="-1"></a>                        <span class="at">by =</span> <span class="st">&quot;margin&quot;</span>, <span class="co"># each term (here only &#39;Group&#39;) analyzed individually</span></span>
<span id="cb254-23"><a href="community-similarity.html#cb254-23" tabindex="-1"></a>                        <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>,</span>
<span id="cb254-24"><a href="community-similarity.html#cb254-24" tabindex="-1"></a>                        <span class="at">permutations =</span> <span class="dv">99</span>)</span>
<span id="cb254-25"><a href="community-similarity.html#cb254-25" tabindex="-1"></a></span>
<span id="cb254-26"><a href="community-similarity.html#cb254-26" tabindex="-1"></a><span class="co"># Get p-values</span></span>
<span id="cb254-27"><a href="community-similarity.html#cb254-27" tabindex="-1"></a>p_values <span class="ot">&lt;-</span> <span class="fu">c</span>(permanova[<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Pr(&gt;F)&quot;</span>], permanova2[<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Pr(&gt;F)&quot;</span>])</span>
<span id="cb254-28"><a href="community-similarity.html#cb254-28" tabindex="-1"></a>p_values <span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(p_values)</span>
<span id="cb254-29"><a href="community-similarity.html#cb254-29" tabindex="-1"></a><span class="fu">rownames</span>(p_values) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;adonis2&quot;</span>, <span class="st">&quot;dbRDA+anova.cca&quot;</span>)</span>
<span id="cb254-30"><a href="community-similarity.html#cb254-30" tabindex="-1"></a>p_values</span></code></pre></div>
<pre><code>##                 p_values
## adonis2             0.02
## dbRDA+anova.cca     0.02</code></pre>
<p>As we can see, the community composition is significantly different
between the groups (p &lt; 0.05), and these two methods give equal p-values.</p>
<p>Let us visualize the model coefficients for species that exhibit the
largest differences between the groups. This gives some insights into
how the groups tend to differ from each other in terms of community
composition.</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="community-similarity.html#cb256-1" tabindex="-1"></a><span class="co"># Add taxa info</span></span>
<span id="cb256-2"><a href="community-similarity.html#cb256-2" tabindex="-1"></a><span class="fu">sppscores</span>(dbrda) <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;relabundance&quot;</span>))</span>
<span id="cb256-3"><a href="community-similarity.html#cb256-3" tabindex="-1"></a><span class="co"># Get coefficients</span></span>
<span id="cb256-4"><a href="community-similarity.html#cb256-4" tabindex="-1"></a>coef <span class="ot">&lt;-</span> dbrda<span class="sc">$</span>CCA<span class="sc">$</span>v</span>
<span id="cb256-5"><a href="community-similarity.html#cb256-5" tabindex="-1"></a><span class="co"># Get the taxa with biggest weights</span></span>
<span id="cb256-6"><a href="community-similarity.html#cb256-6" tabindex="-1"></a>top.coef <span class="ot">&lt;-</span> <span class="fu">head</span>(coef[<span class="fu">rev</span>(<span class="fu">order</span>(<span class="fu">abs</span>(coef))), , <span class="at">drop =</span> <span class="cn">FALSE</span>], <span class="dv">20</span>)</span>
<span id="cb256-7"><a href="community-similarity.html#cb256-7" tabindex="-1"></a><span class="co"># Sort weights in increasing order</span></span>
<span id="cb256-8"><a href="community-similarity.html#cb256-8" tabindex="-1"></a>top.coef <span class="ot">&lt;-</span> top.coef[<span class="fu">order</span>(top.coef), ]</span>
<span id="cb256-9"><a href="community-similarity.html#cb256-9" tabindex="-1"></a><span class="co"># Get top names</span></span>
<span id="cb256-10"><a href="community-similarity.html#cb256-10" tabindex="-1"></a>top_names <span class="ot">&lt;-</span> <span class="fu">names</span>(top.coef)[<span class="fu">order</span>(<span class="fu">abs</span>(top.coef), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)]</span></code></pre></div>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="community-similarity.html#cb257-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> top.coef,</span>
<span id="cb257-2"><a href="community-similarity.html#cb257-2" tabindex="-1"></a>                 <span class="at">y =</span> <span class="fu">factor</span>(<span class="fu">names</span>(top.coef), <span class="fu">unique</span>(<span class="fu">names</span>(top.coef))))</span>
<span id="cb257-3"><a href="community-similarity.html#cb257-3" tabindex="-1"></a></span>
<span id="cb257-4"><a href="community-similarity.html#cb257-4" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb257-5"><a href="community-similarity.html#cb257-5" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb257-6"><a href="community-similarity.html#cb257-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;&quot;</span>, <span class="at">y=</span> <span class="st">&quot;&quot;</span>, <span class="at">title =</span> <span class="st">&quot;Top Taxa&quot;</span>) <span class="sc">+</span></span>
<span id="cb257-7"><a href="community-similarity.html#cb257-7" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="20_beta_diversity_files/figure-html/plot-top-coef-anova-1.png" width="3000" /></p>
<p>In the example above, the largest differences between the two groups
can be attributed to <em>Genus:Bacteroides</em> (elevated in the first
group) and <em>Family:Ruminococcaceae</em> (elevated in the second
group), and many other co-varying species.</p>
</div>
<div id="checking-the-homogeneity-condition" class="section level3 hasAnchor" number="8.3.2">
<h3><span class="header-section-number">8.3.2</span> Checking the homogeneity condition<a href="community-similarity.html#checking-the-homogeneity-condition" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It is important to note that the application of PERMANOVA assumes
homogeneous group dispersions (variances). This can be tested with the
PERMDISP2 method <span class="citation">(Anderson 2006)</span> by using the same assay and distance
method than in PERMANOVA.</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="community-similarity.html#cb258-1" tabindex="-1"></a><span class="fu">anova</span>(<span class="fu">betadisper</span>(<span class="fu">vegdist</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>))), <span class="fu">colData</span>(tse)<span class="sc">$</span>Group))</span></code></pre></div>
<pre><code>## Analysis of Variance Table
## 
## Response: Distances
##           Df Sum Sq Mean Sq F value  Pr(&gt;F)    
## Groups     1 0.2385  0.2385     103 3.6e-10 ***
## Residuals 24 0.0554  0.0023                    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>If the groups have similar dispersion, PERMANOVA can be seen as an
appropriate choice for comparing community compositions.</p>
</div>
</div>
<div id="summary" class="section level2 hasAnchor" number="8.4">
<h2><span class="header-section-number">8.4</span> Summary<a href="community-similarity.html#summary" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>As a final note, we provide a comprehensive list of functions for the evaluation
of dissimilarity indices available in the mia and scater packages. The
<code>calculate</code> methods return a reducedDim object as an output, whereas the <code>run</code>
methods store the reducedDim object into the specified TreeSE.</p>
<ul>
<li>Canonical Correspondence Analysis (CCA): <code>calculateCCA</code> and <code>runCCA</code></li>
<li>dbRDA: <code>calculateRDA</code> and <code>runRDA</code></li>
<li>Double Principal Coordinate Analysis (DPCoA): <code>calculateDPCoA</code> and <code>runDPCoA</code></li>
<li>Jensen-Shannon Divergence (JSD): <code>calculateJSD</code> and <code>runJSD</code></li>
<li>MDS: <code>calculateMDS</code> and <code>runMDS</code></li>
<li>NMDS: <code>calculateNMDS</code> and <code>runNMDS</code></li>
<li>Overlap: <code>calculateOverlap</code> and <code>runOverlap</code></li>
<li>t-distributed Stochastic Neighbor Embedding (t-SNE): <code>calculateTSNE</code> and <code>runTSNE</code></li>
<li>UMAP: <code>calculateUMAP</code> and <code>runUMAP</code></li>
</ul>
<p>For more information on clustering samples by beta diversity, you can refer to:</p>
<ul>
<li><a href="http://bioconductor.org/books/release/OSCA/clustering.html">How to extract information from clusters</a></li>
<li>Chapter <a href="clustering.html#clustering">10</a> on community typing</li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="community-diversity.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="microbiome-community.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/20_beta_diversity.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/20_beta_diversity.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/20_beta_diversity.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
