<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Community diversity | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 7 Community diversity | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Community diversity | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Community diversity | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="taxonomic-information.html"/>
<link rel="next" href="community-similarity.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.2.2</b> colData</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.2.3</b> rowData</a></li>
<li class="chapter" data-level="2.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.2.4</b> rowTree</a></li>
<li class="chapter" data-level="2.2.5" data-path="containers.html"><a href="containers.html#alternative-experiments"><i class="fa fa-check"></i><b>2.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.6" data-path="containers.html"><a href="containers.html#multiassayexperiments"><i class="fa fa-check"></i><b>2.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#some-available-packages"><i class="fa fa-check"></i><b>3.2</b> Some available packages</a></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#quality-control-1"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.3.2</b> Library size</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-phylogenetic-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
<li class="chapter" data-level="" data-path="community-diversity.html"><a href="community-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#community-comparisons-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.2</b> Community comparisons by beta diversity analysis</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="community-similarity.html"><a href="community-similarity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.4.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.4.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.4.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="community-similarity.html"><a href="community-similarity.html#further-reading"><i class="fa fa-check"></i><b>8.5</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="community-similarity.html"><a href="community-similarity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visualizing-taxonomic-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>10</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.1</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.1.1</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>10.2</b> Community Detection</a></li>
<li class="chapter" data-level="10.3" data-path="community-typing.html"><a href="community-typing.html#additional-community-typing"><i class="fa fa-check"></i><b>10.3</b> Additional Community Typing</a></li>
<li class="chapter" data-level="" data-path="community-typing.html"><a href="community-typing.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="biclustering.html"><a href="biclustering.html"><i class="fa fa-check"></i><b>11</b> Biclustering</a>
<ul>
<li class="chapter" data-level="11.1" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>11.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="11.2" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>11.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="11.3" data-path="biclustering.html"><a href="biclustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>11.3</b> Taxa vs taxa</a></li>
<li class="chapter" data-level="" data-path="biclustering.html"><a href="biclustering.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>12</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="12.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>12.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>12.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="12.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>12.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="12.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>12.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="12.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>12.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="12.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>12.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="12.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>12.1.6</b> Comparison of the methods</a></li>
<li class="chapter" data-level="12.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>12.1.7</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>12.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>12.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#multi-omics-factor-analysis"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
<li class="chapter" data-level="" data-path="multi-assay_analyses.html"><a href="multi-assay_analyses.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="14" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>14</b> Resources</a>
<ul>
<li class="chapter" data-level="14.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>14.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>14.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="14.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>14.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="14.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>14.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="14.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>14.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>14.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>14.2.1</b> Bioconductor Classes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>15</b> Extra material</a>
<ul>
<li class="chapter" data-level="15.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>15.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>16</b> Acknowledgments</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="community-diversity" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> Community diversity</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Diversity estimates are a central topic in microbiome data analysis.</p>
<p>There are three commonly employed levels of diversity measurements,
which are trying to put a number on different aspects of the questions
associated with diversity <span class="citation">(<a href="#ref-Whittaker1960" role="doc-biblioref">Whittaker 1960</a>)</span>.</p>
<p>Many different ways for estimating such diversity measurements have been
described in the literature. Which measurement is best or applicable for your
samples, is not the aim of the following sections.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="community-diversity.html#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb175-2"><a href="community-diversity.html#cb175-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb175-3"><a href="community-diversity.html#cb175-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span></code></pre></div>
<p><strong><em>Alpha diversity</em></strong>, also sometimes interchangeably used with the
term <strong><em>species diversity</em></strong>, summarizes the distribution of species
abundances in a given sample into a single number that depends on
species richness and evenness. Diversity indices measure the overall
community heterogeneity. A number of ecological diversity measures are
available. The Hill coefficient combines many standard indices into a
single equation that provides observed richness, inverse Simpson, and
Shannon diversity, and generalized diversity as special cases. In
general, diversity increases together with increasing richness and
evenness. Sometimes richness, phylogenetic diversity, evenness, dominance,
and rarity are considered to be variants of alpha diversity.</p>
<p><strong>Richness</strong> refers to the total number of species in a community
(sample). The simplest richness index is the number of observed
species (observed richness). Assuming limited sampling from the
community, however, this may underestimate the true species
richness. Several estimators are available, including for instance ACE
<span class="citation">(<a href="#ref-Chao1992" role="doc-biblioref">A and SM 1992</a>)</span> and Chao1 <span class="citation">(<a href="#ref-Chao1984" role="doc-biblioref">A 1984</a>)</span>. Richness estimates are unaffected
by species abundances.</p>
<p><strong>Phylogenetic diversity</strong> was first proposed by <span class="citation">(<a href="#ref-Faith1992" role="doc-biblioref">Faith 1992</a>)</span>. Unlike the
diversity measures mentioned above, Phylogenetic diversity (PD)
measure incorporates information from phylogenetic relationships
stored in <code>phylo</code> tree between species in a community (sample). The
Faith’s PD is calculated as the sum of branch length of all species in
a community (sample).</p>
<p><strong>Evenness</strong> focuses on species abundances, and can thus complement
the number of species. A typical evenness index is the Pielou’s
evenness, which is Shannon diversity normalized by the observed
richness.</p>
<p><strong>Dominance</strong> indices are in general negatively correlated with
diversity, and sometimes used in ecological literature. High
dominance is obtained when one or few species have a high share of
the total species abundance in the community.</p>
<p><strong>Rarity</strong> indices characterize the concentration of taxa at low abundance.
Prevalence and detection thresholds determine rare taxa whose total concentration
is represented as a rarity index.</p>
<div id="estimation" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Estimation</h2>
<p>Alpha diversity can be estimated with wrapper functions that interact
with other packages implementing the calculation, such as <em><code>vegan</code></em>
<span class="citation">(<a href="#ref-R-vegan" role="doc-biblioref">Oksanen et al. 2020</a>)</span>.</p>
<div id="richness" class="section level3" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Richness</h3>
<p>Richness gives the number of features present within a community and can be calculated with <code>estimateRichness</code>. Each of the estimate diversity/richness/evenness/dominance functions adds the calculated measure(s) to the <code>colData</code> of the <code>SummarizedExperiment</code> under the given column <code>name</code>. Here, we calculate <code>observed</code> features as a measure of richness.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="community-diversity.html#cb176-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateRichness</span>(tse, </span>
<span id="cb176-2"><a href="community-diversity.html#cb176-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb176-3"><a href="community-diversity.html#cb176-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">index =</span> <span class="st">&quot;observed&quot;</span>, </span>
<span id="cb176-4"><a href="community-diversity.html#cb176-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">name=</span><span class="st">&quot;observed&quot;</span>)</span>
<span id="cb176-5"><a href="community-diversity.html#cb176-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-6"><a href="community-diversity.html#cb176-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>observed)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
##    6964    7679    5729    2667    2574    3214</code></pre>
<p>This allows access to the values to be analyzed directly from the <code>colData</code>, for example
by plotting them using <code>plotColData</code> from the <em><code>scater</code></em> package <span class="citation">(<a href="#ref-R-scater" role="doc-biblioref">McCarthy et al. 2020</a>)</span>.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="community-diversity.html#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb178-2"><a href="community-diversity.html#cb178-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(tse, </span>
<span id="cb178-3"><a href="community-diversity.html#cb178-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;observed&quot;</span>, </span>
<span id="cb178-4"><a href="community-diversity.html#cb178-4" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;SampleType&quot;</span>, </span>
<span id="cb178-5"><a href="community-diversity.html#cb178-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>) <span class="sc">+</span></span>
<span id="cb178-6"><a href="community-diversity.html#cb178-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>,<span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb178-7"><a href="community-diversity.html#cb178-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(Richness[Observed]))</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-div-shannon"></span>
<img src="14_alpha_diversity_files/figure-html/plot-div-shannon-1.png" alt="Shannon diversity estimates plotted grouped by sample type." width="672" />
<p class="caption">
Figure 7.1: Shannon diversity estimates plotted grouped by sample type.
</p>
</div>
</div>
<div id="diversity" class="section level3" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Diversity</h3>
<p><strong>Non-Phylogenetic measures</strong><br />
The main function, <code>estimateDiversity</code>, calculates the selected
diversity index based on the selected assay data.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="community-diversity.html#cb179-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateDiversity</span>(tse, </span>
<span id="cb179-2"><a href="community-diversity.html#cb179-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb179-3"><a href="community-diversity.html#cb179-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">index =</span> <span class="st">&quot;shannon&quot;</span>, </span>
<span id="cb179-4"><a href="community-diversity.html#cb179-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">name =</span> <span class="st">&quot;shannon&quot;</span>)</span>
<span id="cb179-5"><a href="community-diversity.html#cb179-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>shannon)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
##   6.577   6.777   6.498   3.828   3.288   4.289</code></pre>
<p>Alpha diversities can be visualized with boxplot. Here, Shannon index is compared
between different sample type groups. Individual data points are visualized by
plotting them as points with <code>geom_jitter</code>.</p>
<p><code>geom_signif</code> is used to test whether these differences are statistically significant.
It adds p-values to plot.</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="community-diversity.html#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( <span class="sc">!</span><span class="fu">require</span>(ggsignif) ){</span>
<span id="cb181-2"><a href="community-diversity.html#cb181-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(ggsignif)</span>
<span id="cb181-3"><a href="community-diversity.html#cb181-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb181-4"><a href="community-diversity.html#cb181-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb181-5"><a href="community-diversity.html#cb181-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb181-6"><a href="community-diversity.html#cb181-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsignif)</span>
<span id="cb181-7"><a href="community-diversity.html#cb181-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-8"><a href="community-diversity.html#cb181-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsets the data. Takes only those samples that are from feces, skin, or tongue,</span></span>
<span id="cb181-9"><a href="community-diversity.html#cb181-9" aria-hidden="true" tabindex="-1"></a><span class="co"># and creates data frame from the collected data</span></span>
<span id="cb181-10"><a href="community-diversity.html#cb181-10" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)[<span class="fu">colData</span>(tse)<span class="sc">$</span>SampleType <span class="sc">%in%</span> </span>
<span id="cb181-11"><a href="community-diversity.html#cb181-11" aria-hidden="true" tabindex="-1"></a>                                  <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>), ])</span>
<span id="cb181-12"><a href="community-diversity.html#cb181-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-13"><a href="community-diversity.html#cb181-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Changes old levels with new levels</span></span>
<span id="cb181-14"><a href="community-diversity.html#cb181-14" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>SampleType <span class="ot">&lt;-</span> <span class="fu">factor</span>(df<span class="sc">$</span>SampleType)</span>
<span id="cb181-15"><a href="community-diversity.html#cb181-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-16"><a href="community-diversity.html#cb181-16" aria-hidden="true" tabindex="-1"></a><span class="co"># For significance testing, all different combinations are determined</span></span>
<span id="cb181-17"><a href="community-diversity.html#cb181-17" aria-hidden="true" tabindex="-1"></a>comb <span class="ot">&lt;-</span> <span class="fu">split</span>(<span class="fu">t</span>(<span class="fu">combn</span>(<span class="fu">levels</span>(df<span class="sc">$</span>SampleType), <span class="dv">2</span>)), </span>
<span id="cb181-18"><a href="community-diversity.html#cb181-18" aria-hidden="true" tabindex="-1"></a>           <span class="fu">seq</span>(<span class="fu">nrow</span>(<span class="fu">t</span>(<span class="fu">combn</span>(<span class="fu">levels</span>(df<span class="sc">$</span>SampleType), <span class="dv">2</span>)))))</span>
<span id="cb181-19"><a href="community-diversity.html#cb181-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-20"><a href="community-diversity.html#cb181-20" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> SampleType, <span class="at">y =</span> shannon)) <span class="sc">+</span></span>
<span id="cb181-21"><a href="community-diversity.html#cb181-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Outliers are removed, because otherwise each data point would be plotted twice; </span></span>
<span id="cb181-22"><a href="community-diversity.html#cb181-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># as an outlier of boxplot and as a point of dotplot.</span></span>
<span id="cb181-23"><a href="community-diversity.html#cb181-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span> </span>
<span id="cb181-24"><a href="community-diversity.html#cb181-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb181-25"><a href="community-diversity.html#cb181-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_signif</span>(<span class="at">comparisons =</span> comb, <span class="at">map_signif_level =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb181-26"><a href="community-diversity.html#cb181-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>))</span></code></pre></div>
<p><img src="14_alpha_diversity_files/figure-html/visualize-shannon-1.png" width="672" /></p>
</div>
<div id="faith-phylogenetic-diversity" class="section level3" number="7.1.3">
<h3><span class="header-section-number">7.1.3</span> Faith phylogenetic diversity</h3>
<p>The Faith index is returned by the function <code>estimateFaith</code>.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="community-diversity.html#cb182-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateFaith</span>(tse,</span>
<span id="cb182-2"><a href="community-diversity.html#cb182-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb182-3"><a href="community-diversity.html#cb182-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>faith)</span></code></pre></div>
<pre><code>## [1] 250.5 262.3 208.5 117.9 119.8 135.8</code></pre>
<p><strong>Note</strong>: because <code>tse</code> is a <code>TreeSummarizedExperiment</code> object, its phylogenetic tree is used by default. However, the optional argument <code>tree</code> must be provided if <code>tse</code> does not contain one.</p>
<p>Below a visual comparison between shannon and faith indices is shown with a violin plot.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="community-diversity.html#cb184-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;shannon&quot;</span>, <span class="st">&quot;faith&quot;</span>),</span>
<span id="cb184-2"><a href="community-diversity.html#cb184-2" aria-hidden="true" tabindex="-1"></a>                plotColData,</span>
<span id="cb184-3"><a href="community-diversity.html#cb184-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">object =</span> tse)</span>
<span id="cb184-4"><a href="community-diversity.html#cb184-4" aria-hidden="true" tabindex="-1"></a>plots[[<span class="dv">1</span>]] <span class="sc">+</span> plots[[<span class="dv">2</span>]]</span></code></pre></div>
<p><img src="14_alpha_diversity_files/figure-html/phylo-div-2-1.png" width="672" /></p>
<p>Alternatively, the phylogenetic diversity can be calculated by <code>mia::estimateDiversity</code>. This is a faster re-implementation of<br />
the widely used function in <em><code>picante</code></em> <span class="citation"><a href="#ref-Kembel2010" role="doc-biblioref">W et al.</a> (<a href="#ref-Kembel2010" role="doc-biblioref">2010</a>)</span>.</p>
<p>Load <code>picante</code> R package and get the <code>phylo</code> stored in <code>rowTree</code>.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="community-diversity.html#cb185-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateDiversity</span>(tse, </span>
<span id="cb185-2"><a href="community-diversity.html#cb185-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb185-3"><a href="community-diversity.html#cb185-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">index =</span> <span class="st">&quot;faith&quot;</span>, </span>
<span id="cb185-4"><a href="community-diversity.html#cb185-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">name =</span> <span class="st">&quot;faith&quot;</span>)</span></code></pre></div>
</div>
<div id="evenness" class="section level3" number="7.1.4">
<h3><span class="header-section-number">7.1.4</span> Evenness</h3>
<p>Evenness can be calculated with <code>estimateEvenness</code>.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="community-diversity.html#cb186-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">estimateEvenness</span>(tse, </span>
<span id="cb186-2"><a href="community-diversity.html#cb186-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb186-3"><a href="community-diversity.html#cb186-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">index=</span><span class="st">&quot;simpson&quot;</span>)</span>
<span id="cb186-4"><a href="community-diversity.html#cb186-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>simpson)</span></code></pre></div>
<pre><code>## [1] 0.026871 0.027197 0.047049 0.005179 0.004304 0.005011</code></pre>
</div>
<div id="dominance" class="section level3" number="7.1.5">
<h3><span class="header-section-number">7.1.5</span> Dominance</h3>
<p>Dominance can be calculated with <code>estimateDominance</code>. Here, the <code>Relative index</code> is calculated which is the relative abundance of the most dominant species in the sample.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="community-diversity.html#cb188-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">estimateDominance</span>(tse, </span>
<span id="cb188-2"><a href="community-diversity.html#cb188-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb188-3"><a href="community-diversity.html#cb188-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">index=</span><span class="st">&quot;relative&quot;</span>)</span>
<span id="cb188-4"><a href="community-diversity.html#cb188-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-5"><a href="community-diversity.html#cb188-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>relative)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
## 0.03910 0.03226 0.01690 0.22981 0.21778 0.22329</code></pre>
</div>
<div id="rarity" class="section level3" number="7.1.6">
<h3><span class="header-section-number">7.1.6</span> Rarity</h3>
<p><code>mia</code> package provides one rarity index called log-modulo skewness. It can be
calculated with <code>estimateDiversity</code>.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="community-diversity.html#cb190-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateDiversity</span>(tse, </span>
<span id="cb190-2"><a href="community-diversity.html#cb190-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb190-3"><a href="community-diversity.html#cb190-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">index =</span> <span class="st">&quot;log_modulo_skewness&quot;</span>)</span>
<span id="cb190-4"><a href="community-diversity.html#cb190-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-5"><a href="community-diversity.html#cb190-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>log_modulo_skewness)</span></code></pre></div>
<pre><code>## [1] 2.061 2.061 2.061 2.061 2.061 2.061</code></pre>
</div>
</div>
<div id="visualization" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Visualization</h2>
<p>A plot comparing all the diversity measures calculated above and stored in <code>colData</code> can then be constructed directly.</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="community-diversity.html#cb192-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;observed&quot;</span>, <span class="st">&quot;shannon&quot;</span>, <span class="st">&quot;simpson&quot;</span>, <span class="st">&quot;relative&quot;</span>, <span class="st">&quot;faith&quot;</span>, <span class="st">&quot;log_modulo_skewness&quot;</span>),</span>
<span id="cb192-2"><a href="community-diversity.html#cb192-2" aria-hidden="true" tabindex="-1"></a>                plotColData,</span>
<span id="cb192-3"><a href="community-diversity.html#cb192-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">object =</span> tse,</span>
<span id="cb192-4"><a href="community-diversity.html#cb192-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">x =</span> <span class="st">&quot;SampleType&quot;</span>,</span>
<span id="cb192-5"><a href="community-diversity.html#cb192-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span>
<span id="cb192-6"><a href="community-diversity.html#cb192-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-7"><a href="community-diversity.html#cb192-7" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plots, <span class="st">&quot;+&quot;</span>, </span>
<span id="cb192-8"><a href="community-diversity.html#cb192-8" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb192-9"><a href="community-diversity.html#cb192-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb192-10"><a href="community-diversity.html#cb192-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>()))</span>
<span id="cb192-11"><a href="community-diversity.html#cb192-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-12"><a href="community-diversity.html#cb192-12" aria-hidden="true" tabindex="-1"></a>((plots[[<span class="dv">1</span>]] <span class="sc">|</span> plots[[<span class="dv">2</span>]] <span class="sc">|</span> plots[[<span class="dv">3</span>]]) <span class="sc">/</span> </span>
<span id="cb192-13"><a href="community-diversity.html#cb192-13" aria-hidden="true" tabindex="-1"></a>(plots[[<span class="dv">4</span>]] <span class="sc">|</span> plots[[<span class="dv">5</span>]] <span class="sc">|</span> plots[[<span class="dv">6</span>]])) <span class="sc">+</span></span>
<span id="cb192-14"><a href="community-diversity.html#cb192-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">&quot;collect&quot;</span>)</span></code></pre></div>
<p><img src="14_alpha_diversity_files/figure-html/plot-all-diversities-1.png" width="672" /></p>
</div>
<div id="session-info-4" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] patchwork_1.1.1                ggsignif_0.6.3                
 [3] scater_1.22.0                  ggplot2_3.3.5                 
 [5] scuttle_1.4.0                  mia_1.3.15                    
 [7] MultiAssayExperiment_1.20.0    TreeSummarizedExperiment_2.1.4
 [9] Biostrings_2.62.0              XVector_0.34.0                
[11] SingleCellExperiment_1.16.0    SummarizedExperiment_1.24.0   
[13] Biobase_2.54.0                 GenomicRanges_1.46.1          
[15] GenomeInfoDb_1.30.1            IRanges_2.28.0                
[17] S4Vectors_0.32.3               BiocGenerics_0.40.0           
[19] MatrixGenerics_1.6.0           matrixStats_0.61.0-9001       
[21] BiocStyle_2.22.0               rebook_1.4.0                  

loaded via a namespace (and not attached):
  [1] ggbeeswarm_0.6.0            colorspace_2.0-2           
  [3] ellipsis_0.3.2              BiocNeighbors_1.12.0       
  [5] farver_2.1.0                ggrepel_0.9.1              
  [7] bit64_4.0.5                 fansi_1.0.2                
  [9] decontam_1.14.0             splines_4.1.2              
 [11] codetools_0.2-18            sparseMatrixStats_1.6.0    
 [13] cachem_1.0.6                knitr_1.37                 
 [15] jsonlite_1.7.3              cluster_2.1.2              
 [17] graph_1.72.0                BiocManager_1.30.16        
 [19] compiler_4.1.2              assertthat_0.2.1           
 [21] Matrix_1.4-0                fastmap_1.1.0              
 [23] lazyeval_0.2.2              cli_3.1.1                  
 [25] BiocSingular_1.10.0         htmltools_0.5.2            
 [27] tools_4.1.2                 rsvd_1.0.5                 
 [29] gtable_0.3.0                glue_1.6.1                 
 [31] GenomeInfoDbData_1.2.7      reshape2_1.4.4             
 [33] dplyr_1.0.7                 Rcpp_1.0.8                 
 [35] jquerylib_0.1.4             vctrs_0.3.8                
 [37] ape_5.6-1                   nlme_3.1-155               
 [39] DECIPHER_2.22.0             DelayedMatrixStats_1.16.0  
 [41] xfun_0.29                   stringr_1.4.0              
 [43] beachmat_2.10.0             lifecycle_1.0.1            
 [45] irlba_2.3.5                 XML_3.99-0.8               
 [47] zlibbioc_1.40.0             MASS_7.3-55                
 [49] scales_1.1.1                parallel_4.1.2             
 [51] yaml_2.2.2                  memoise_2.0.1              
 [53] gridExtra_2.3               yulab.utils_0.0.4          
 [55] sass_0.4.0                  stringi_1.7.6              
 [57] RSQLite_2.2.9               highr_0.9                  
 [59] ScaledMatrix_1.2.0          permute_0.9-7              
 [61] tidytree_0.3.7              filelock_1.0.2             
 [63] BiocParallel_1.28.3         rlang_1.0.0                
 [65] pkgconfig_2.0.3             bitops_1.0-7               
 [67] evaluate_0.14               lattice_0.20-45            
 [69] purrr_0.3.4                 labeling_0.4.2             
 [71] treeio_1.18.1               CodeDepends_0.6.5          
 [73] cowplot_1.1.1               bit_4.0.4                  
 [75] tidyselect_1.1.1            plyr_1.8.6                 
 [77] magrittr_2.0.2              bookdown_0.24              
 [79] R6_2.5.1                    generics_0.1.2             
 [81] DelayedArray_0.20.0         DBI_1.1.2                  
 [83] withr_2.4.3                 mgcv_1.8-38                
 [85] pillar_1.7.0                RCurl_1.98-1.5             
 [87] tibble_3.1.6                dir.expiry_1.2.0           
 [89] crayon_1.4.2                utf8_1.2.2                 
 [91] rmarkdown_2.11              viridis_0.6.2              
 [93] grid_4.1.2                  vegan_2.5-7                
 [95] blob_1.2.2                  digest_0.6.29              
 [97] tidyr_1.2.0                 munsell_0.5.0              
 [99] DirichletMultinomial_1.36.0 beeswarm_0.4.0             
[101] viridisLite_0.4.0           vipor_0.4.5                
[103] bslib_0.3.1                </code></pre>
</div>

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
</div>
<h3>Bibliography</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Chao1984" class="csl-entry">
A, Chao. 1984. <span>“Non-Parametric Estimation of the Number of Classes in a Population.”</span> <em>Scandinavian Journal of Statistics</em> 11 (4): 265–70. <a href="https://www.jstor.org/stable/4615964">https://www.jstor.org/stable/4615964</a>.
</div>
<div id="ref-Chao1992" class="csl-entry">
A, Chao, and Lee SM. 1992. <span>“Estimating the Number of Classes via Sample Coverage.”</span> <em>Journal of the American Statistical Association</em> 87 (417): 210–17. https://doi.org/<a href="https://doi.org/10.1080/01621459.1992.10475194">https://doi.org/10.1080/01621459.1992.10475194</a>.
</div>
<div id="ref-Faith1992" class="csl-entry">
Faith, D. P. 1992. <span>“Conservation Evaluation and Phylogenetic Diversity.”</span> <em>Biological Conservation</em> 61 (1): 10. https://doi.org/<a href="https://doi.org/10.1016/0006-3207(92)91201-3">https://doi.org/10.1016/0006-3207(92)91201-3</a>.
</div>
<div id="ref-R-scater" class="csl-entry">
McCarthy, Davis, Kieran Campbell, Aaron Lun, and Quin Wills. 2020. <em>Scater: Single-Cell Analysis Toolkit for Gene Expression Data in r</em>. <a href="http://bioconductor.org/packages/scater/">http://bioconductor.org/packages/scater/</a>.
</div>
<div id="ref-R-vegan" class="csl-entry">
Oksanen, Jari, F. Guillaume Blanchet, Michael Friendly, Roeland Kindt, Pierre Legendre, Dan McGlinn, Peter R. Minchin, et al. 2020. <em>Vegan: Community Ecology Package</em>. <a href="https://CRAN.R-project.org/package=vegan">https://CRAN.R-project.org/package=vegan</a>.
</div>
<div id="ref-Kembel2010" class="csl-entry">
W, Kembel Steven, Cowan Peter D, Helmus Matthew R, Cornwell William K, Morlon Helene, Ackerly David D, Blomberg Simon P, and Webb Campbell O. 2010. <span>“Picante: R Tools for Integrating Phylogenies and Ecology.”</span> <em>Bioinformatics</em> 26 (11): 1463–64. https://doi.org/<a href="https://doi.org/10.1093/bioinformatics/btq166">https://doi.org/10.1093/bioinformatics/btq166</a>.
</div>
<div id="ref-Whittaker1960" class="csl-entry">
Whittaker, R. H. 1960. <span>“Vegetation of the Siskiyou Mountains, Oregon and California.”</span> <em>Ecological Monographs</em> 30 (3): 279–338. https://doi.org/<a href="https://doi.org/10.2307/1943563">https://doi.org/10.2307/1943563</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="taxonomic-information.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="community-similarity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
