<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Microbiome Data | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 3 Microbiome Data | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Microbiome Data | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Microbiome Data | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="packages.html"/>
<link rel="next" href="datamanipulation.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.4</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis with igraph</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-plot"><i class="fa fa-check"></i><b>12.2.1</b> Network plot</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#centrality-measures"><i class="fa fa-check"></i><b>12.2.2</b> Centrality measures</a></li>
<li class="chapter" data-level="12.2.3" data-path="network-learning.html"><a href="network-learning.html#scale-node-sizes-by-centrality"><i class="fa fa-check"></i><b>12.2.3</b> Scale node sizes by centrality</a></li>
<li class="chapter" data-level="12.2.4" data-path="network-learning.html"><a href="network-learning.html#degree-distribution"><i class="fa fa-check"></i><b>12.2.4</b> Degree distribution</a></li>
<li class="chapter" data-level="12.2.5" data-path="network-learning.html"><a href="network-learning.html#clustered-heatmaps"><i class="fa fa-check"></i><b>12.2.5</b> Clustered heatmaps</a></li>
<li class="chapter" data-level="12.2.6" data-path="network-learning.html"><a href="network-learning.html#global-network-measures"><i class="fa fa-check"></i><b>12.2.6</b> Global network measures</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.3</b> Network analysis with NetCoMi</a></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.4</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.4.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.4.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.4.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.5</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.6</b> More about association measures</a></li>
<li class="chapter" data-level="12.7" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.7</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.7.1</b> SparCC</a></li>
<li class="chapter" data-level="12.7.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.7.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.7.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.7.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.7.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.7.4</b> Network plots</a></li>
<li class="chapter" data-level="12.7.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.7.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="containers" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Microbiome Data<a href="containers.html#containers" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="data-science-framework" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Data science framework<a href="containers.html#data-science-framework" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The building blocks of the framework are <strong>data container</strong>
(SummarizedExperiment and its derivatives), <strong>packages</strong> from various
developers using the TreeSE container, open <strong>demonstration data
sets</strong>, in a separate chapter <a href="containers.html#example-data">3.3</a>, and <strong>online
tutorials</strong> including this online book as well as the various package
vignettes and other materials.</p>
<p><img src="general/figures/FigureOverviewV2_mod.png" width="672" /></p>
</div>
<div id="data-containers" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Data containers<a href="containers.html#data-containers" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>SummarizedExperiment</code> (<code>SE</code>) <span class="citation">(Morgan et al. 2020)</span> is a generic and highly optimized container for complex data
structures. It has become a common choice for analysing various types
of biomedical profiling data, such as RNAseq, ChIp-Seq, microarrays,
flow cytometry, proteomics, and single-cell
sequencing.</p>
<p>[<code>TreeSummarizedExperiment</code>] (<code>TreeSE</code>) <span class="citation">(Huang 2020)</span> was developed as an extension to incorporate hierarchical
information (such as phylogenetic trees and sample hierarchies) and
reference sequences.</p>
<p>[<code>MultiAssayExperiment</code>] (<code>MAE</code>) <span class="citation">(Ramos et al. 2017)</span> provides an organized
way to bind several different data containers together in a single
object. For example, we can bind microbiome data (in <code>TreeSE</code>
container) with metabolomic profiling data (in <code>SE</code>) container, with
(partially) shared sample metadata. This is convenient and robust for
instance in subsetting and other data manipulation tasks. Microbiome
data can be part of multiomics experiments and analysis strategies. We
highlight how the methods used througout in this book relate to this
data framework by using the <code>TreeSummarizedExperiment</code>,
<code>MultiAssayExperiment</code>, and classes beyond.</p>
<p>This section provides an introductions to these data containers. In
microbiome data science, these containers link taxonomic abundance
tables with rich side information on the features and
samples. Taxonomic abundance data can be obtained by 16S rRNA amplicon
or metagenomic sequencing, phylogenetic microarrays, or by other
means. Many microbiome experiments include multiple versions and types
of data generated independently or derived from each other through
transformation or agglomeration. We start by providing recommendations
on how to represent different varieties of multi-table data within the
<code>TreeSummarizedExperiment</code> class.</p>
<p>The options and recommendations are summarized in Table <a href="containers.html#tab:options">3.1</a>.</p>
<div id="assay-slot" class="section level3 hasAnchor" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Assay data<a href="containers.html#assay-slot" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The original count-based taxonomic abundance tables may have different
transformations, such as logarithmic, Centered Log-Ratio (CLR), or relative
abundance. These are typically stored in <em><strong>assays</strong></em>.</p>
<p>Let us load example data and rename it as tse.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="containers.html#cb4-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb4-2"><a href="containers.html#cb4-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;hitchip1006&quot;</span>, <span class="at">package =</span> <span class="st">&quot;miaTime&quot;</span>)</span>
<span id="cb4-3"><a href="containers.html#cb4-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> hitchip1006</span></code></pre></div>
<p>The <code>assays</code> slot contains the experimental data as multiple count matrices. The result of <code>assays</code> is a list of matrices.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="containers.html#cb5-1" tabindex="-1"></a><span class="fu">assays</span>(tse)</span></code></pre></div>
<pre><code>## List of length 1
## names(1): counts</code></pre>
<p>Individual assays can be accessed via <code>assay</code></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="containers.html#cb7-1" tabindex="-1"></a><span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                              Sample-1 Sample-2 Sample-3 Sample-4 Sample-5
## Actinomycetaceae                    0        0        0        0        0
## Aerococcus                          0        0        0        0        0
## Aeromonas                           0        0        0        0        0
## Akkermansia                        21       36      475       61       34
## Alcaligenes faecalis et rel.        1        1        1        2        1
##                              Sample-6 Sample-7
## Actinomycetaceae                    0        0
## Aerococcus                          0        0
## Aeromonas                           0        0
## Akkermansia                        14       27
## Alcaligenes faecalis et rel.        1        1</code></pre>
<p>To illustrate the use of multiple assays, the relative abundance data can be
calculated and stored along the original count data using <code>transformAssay</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="containers.html#cb9-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb9-2"><a href="containers.html#cb9-2" tabindex="-1"></a><span class="fu">assays</span>(tse)</span></code></pre></div>
<pre><code>## List of length 2
## names(2): counts relabundance</code></pre>
<p>Now there are two assays available in the <code>tse</code> object, <code>counts</code> and
<code>relabundance</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="containers.html#cb11-1" tabindex="-1"></a><span class="fu">assay</span>(tse, <span class="st">&quot;relabundance&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                               Sample-1  Sample-2  Sample-3  Sample-4  Sample-5
## Actinomycetaceae             0.0000000 0.000e+00 0.0000000 0.0000000 0.000e+00
## Aerococcus                   0.0000000 0.000e+00 0.0000000 0.0000000 0.000e+00
## Aeromonas                    0.0000000 0.000e+00 0.0000000 0.0000000 0.000e+00
## Akkermansia                  0.0027657 3.547e-03 0.0666106 0.0056195 2.833e-03
## Alcaligenes faecalis et rel. 0.0001317 9.854e-05 0.0001402 0.0001842 8.333e-05
##                               Sample-6  Sample-7
## Actinomycetaceae             0.0000000 0.0000000
## Aerococcus                   0.0000000 0.0000000
## Aeromonas                    0.0000000 0.0000000
## Akkermansia                  0.0017690 0.0045570
## Alcaligenes faecalis et rel. 0.0001264 0.0001688</code></pre>
<p>Here the dimension of the count data remains unchanged in
transformation. This is in fact, a requirement for the assays.</p>
</div>
<div id="coldata" class="section level3 hasAnchor" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> colData<a href="containers.html#coldata" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>colData</code> contains data on the samples.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="containers.html#cb13-1" tabindex="-1"></a><span class="fu">colData</span>(tse)</span></code></pre></div>
<pre><code>## DataFrame with 1151 rows and 10 columns
##                   age      sex nationality DNA_extraction_method  project
##             &lt;integer&gt; &lt;factor&gt;    &lt;factor&gt;              &lt;factor&gt; &lt;factor&gt;
## Sample-1           28   male            US                    NA        1
## Sample-2           24   female          US                    NA        1
## Sample-3           52   male            US                    NA        1
## Sample-4           22   female          US                    NA        1
## Sample-5           25   female          US                    NA        1
## ...               ...      ...         ...                   ...      ...
## Sample-1168        50   female Scandinavia                     r       40
## Sample-1169        31   female Scandinavia                     r       40
## Sample-1170        31   female Scandinavia                     r       40
## Sample-1171        52   male   Scandinavia                     r       40
## Sample-1172        52   male   Scandinavia                     r       40
##             diversity   bmi_group  subject      time      sample
##             &lt;numeric&gt;    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt; &lt;character&gt;
## Sample-1         5.76 severeobese        1         0    Sample-1
## Sample-2         6.06 obese              2         0    Sample-2
## Sample-3         5.50 lean               3         0    Sample-3
## Sample-4         5.87 underweight        4         0    Sample-4
## Sample-5         5.89 lean               5         0    Sample-5
## ...               ...         ...      ...       ...         ...
## Sample-1168      5.87 severeobese      244       8.1 Sample-1168
## Sample-1169      5.87 overweight       245       2.3 Sample-1169
## Sample-1170      5.92 overweight       245       8.2 Sample-1170
## Sample-1171      6.04 overweight       246       2.1 Sample-1171
## Sample-1172      5.74 overweight       246       7.9 Sample-1172</code></pre>
</div>
<div id="rowdata" class="section level3 hasAnchor" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> rowData<a href="containers.html#rowdata" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>rowData</code> contains data on the features of the analyzed samples. Of particular
interest to the microbiome field, this is used to store taxonomic information.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="containers.html#cb15-1" tabindex="-1"></a><span class="fu">rowData</span>(tse)</span></code></pre></div>
<pre><code>## DataFrame with 130 rows and 3 columns
##                                       Phylum          Family
##                                  &lt;character&gt;     &lt;character&gt;
## Actinomycetaceae              Actinobacteria  Actinobacteria
## Aerococcus                        Firmicutes         Bacilli
## Aeromonas                     Proteobacteria  Proteobacteria
## Akkermansia                  Verrucomicrobia Verrucomicrobia
## Alcaligenes faecalis et rel.  Proteobacteria  Proteobacteria
## ...                                      ...             ...
## Vibrio                        Proteobacteria  Proteobacteria
## Weissella et rel.                 Firmicutes         Bacilli
## Wissella et rel.                  Firmicutes         Bacilli
## Xanthomonadaceae              Proteobacteria  Proteobacteria
## Yersinia et rel.              Proteobacteria  Proteobacteria
##                                               Genus
##                                         &lt;character&gt;
## Actinomycetaceae                   Actinomycetaceae
## Aerococcus                               Aerococcus
## Aeromonas                                 Aeromonas
## Akkermansia                             Akkermansia
## Alcaligenes faecalis et rel. Alcaligenes faecalis..
## ...                                             ...
## Vibrio                                       Vibrio
## Weissella et rel.                 Weissella et rel.
## Wissella et rel.                   Wissella et rel.
## Xanthomonadaceae                   Xanthomonadaceae
## Yersinia et rel.                   Yersinia et rel.</code></pre>
</div>
<div id="rowtree" class="section level3 hasAnchor" number="3.2.4">
<h3><span class="header-section-number">3.2.4</span> rowTree<a href="containers.html#rowtree" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Phylogenetic trees also play an important role in the microbiome field. The
<code>TreeSummarizedExperiment</code> class can keep track of features and node
relations via two functions, <code>rowTree</code> and <code>rowLinks</code>.</p>
<p>A tree can be accessed via <code>rowTree</code> as <code>phylo</code> object.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="containers.html#cb17-1" tabindex="-1"></a><span class="fu">rowTree</span>(tse)</span></code></pre></div>
<pre><code>## NULL</code></pre>
<p>The links to the individual features are available through <code>rowLinks</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="containers.html#cb19-1" tabindex="-1"></a><span class="fu">rowLinks</span>(tse)</span></code></pre></div>
<pre><code>## NULL</code></pre>
<p>Please note that there can be a 1:1 relationship between tree nodes and
features, but this is not a must-have. This means there can be features, which
are not linked to nodes, and nodes, which are not linked to features. To change
the links in an existing object, the <code>changeTree</code> function is available.</p>
</div>
<div id="alt-exp" class="section level3 hasAnchor" number="3.2.5">
<h3><span class="header-section-number">3.2.5</span> Alternative experiments<a href="containers.html#alt-exp" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><em><strong>Alternative experiments</strong></em> complement <em>assays</em>. They can contain
complementary data, which is no longer tied to the same dimensions as
the assay data. However, the number of samples (columns) must be the
same.</p>
<p>This can come into play, for instance, when one has taxonomic
abundance profiles quantified with different measurement technologies,
such as phylogenetic microarrays, amplicon sequencing, or metagenomic
sequencing. Another common use case is including abundance tables for
different taxonomic ranks. Such alternative experiments concerning the
same set of samples can be stored as</p>
<ol style="list-style-type: decimal">
<li>Separate <em>assays</em> assuming that the taxonomic information can be mapped
between features directly 1:1; or</li>
<li>Data in the <em>altExp</em> slot of the <code>TreeSummarizedExperiment</code>, if the feature
dimensions differ. Each element of the <em>altExp</em> slot is a <code>SummarizedExperiment</code>
or an object from a derived class with independent feature data.</li>
</ol>
<p>The following shows how to store taxonomic abundance tables
agglomerated at different taxonomic levels. However, the data could as
well originate from entirely different measurement sources as long as
the samples match.</p>
<p>Let us first agglomerate the data to Phylum level. This yields a new
TreeSE data object.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="containers.html#cb21-1" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse, <span class="st">&quot;Phylum&quot;</span>, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb21-2"><a href="containers.html#cb21-2" tabindex="-1"></a><span class="co"># Both have the same number of columns (samples)</span></span>
<span id="cb21-3"><a href="containers.html#cb21-3" tabindex="-1"></a><span class="fu">dim</span>(tse)</span></code></pre></div>
<pre><code>## [1]  130 1151</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="containers.html#cb23-1" tabindex="-1"></a><span class="fu">dim</span>(tse_phylum)</span></code></pre></div>
<pre><code>## [1]    8 1151</code></pre>
<p>Then we can add the new phylum-level data object as an alternative experiment in the original data.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="containers.html#cb25-1" tabindex="-1"></a><span class="co"># Add the new data object to the original data object as an alternative experiment with the name &quot;Phylum&quot;</span></span>
<span id="cb25-2"><a href="containers.html#cb25-2" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;Phylum&quot;</span>) <span class="ot">&lt;-</span> tse_phylum</span>
<span id="cb25-3"><a href="containers.html#cb25-3" tabindex="-1"></a></span>
<span id="cb25-4"><a href="containers.html#cb25-4" tabindex="-1"></a><span class="co"># Check the alternative experiment names available in the data</span></span>
<span id="cb25-5"><a href="containers.html#cb25-5" tabindex="-1"></a><span class="fu">altExpNames</span>(tse)</span></code></pre></div>
<pre><code>## [1] &quot;Phylum&quot;</code></pre>
<p>We can now subset the data, for instance, and this acts on both altExp and assay data.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="containers.html#cb27-1" tabindex="-1"></a>tse[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 130 10 
## metadata(0):
## assays(2): counts relabundance
## rownames(130): Actinomycetaceae Aerococcus ... Xanthomonadaceae
##   Yersinia et rel.
## rowData names(3): Phylum Family Genus
## colnames(10): Sample-1 Sample-2 ... Sample-9 Sample-10
## colData names(10): age sex ... time sample
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(1): Phylum
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="containers.html#cb29-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">altExp</span>(tse[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>],<span class="st">&quot;Phylum&quot;</span>))</span></code></pre></div>
<pre><code>## [1]  8 10</code></pre>
<p>For more details on <em>altExp</em>, you can check the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">introduction</a> to the <code>SingleCellExperiment</code> package <span class="citation">(Lun and Risso 2020)</span>.</p>
</div>
<div id="mae" class="section level3 hasAnchor" number="3.2.6">
<h3><span class="header-section-number">3.2.6</span> MultiAssayExperiments<a href="containers.html#mae" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><em><strong>Multiple experiments</strong></em> relate to complementary measurement types,
such as transcriptomic or metabolomic profiling of the microbiome or
the host. Multiple experiments can be represented using the same
options as alternative experiments, or by using the
<code>MultiAssayExperiment</code> class <span class="citation">(Ramos et al. 2017)</span>. Depending on how the
datasets relate to each other the data can be stored as:</p>
<ol style="list-style-type: decimal">
<li>Separate <em>altExp</em> if the samples can be matched directly 1:1; or</li>
<li>As <code>MultiAssayExperiment</code> objects, in which the connections between
samples are defined through a <code>sampleMap</code>. Each element on the
<code>experimentsList</code> of an <code>MultiAssayExperiment</code> is <code>matrix</code> or
<code>matrix</code>-like objects, including <code>SummarizedExperiment</code> objects, and
the number of samples can differ between the elements.</li>
</ol>
<!--

```r
#TODO: Find the right dataset to explain a non 1:1 sample relationship
```
-->
<p>For information have a look at the <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html">intro vignette</a> of the <code>MultiAssayExperiment</code> package.</p>
<table>
<caption><span id="tab:options">Table 3.1: </span> <strong>Recommended options for storing multiple data tables in microbiome studies</strong> The <em>assays</em> are best suited for data transformations (one-to-one match between samples and columns across the assays). The <em>alternative experiments</em> are particularly suitable for alternative versions of the data that are of same type but may have a different number of features (e.g.taxonomic groups); this is for instance the case with taxonomic abundance tables agglomerated at different levels (e.g.genus vs.phyla) or alternative profiling technologies (e.g.amplicon sequencing vs.shallow shotgun metagenomics). For alternative experiments one-to-one match between samples (cols) is libraryd but the alternative experiment tables can have different numbers of features (rows). Finally, elements of the <em>MultiAssayExperiment</em> provide the most flexible way to incorporate multi-omic data tables with flexible numbers of samples and features. We recommend these conventions as the basis for methods development and application in microbiome studies.</caption>
<thead>
<tr class="header">
<th align="right">Option</th>
<th>Rows (features)</th>
<th align="right">Cols (samples)</th>
<th align="right">Recommended</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">assays</td>
<td>match</td>
<td align="right">match</td>
<td align="right">Data transformations</td>
</tr>
<tr class="even">
<td align="right">altExp</td>
<td>free</td>
<td align="right">match</td>
<td align="right">Alternative experiments</td>
</tr>
<tr class="odd">
<td align="right">MultiAssay</td>
<td>free</td>
<td align="right">free (mapping)</td>
<td align="right">Multi-omic experiments</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="example-data" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Demonstration data<a href="containers.html#example-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Open demonstration data for testing and benchmarking purposes is
available from multiple locations. This chapter introduces some
options. The other chapters of this book provide ample examples about
the use of the data.</p>
<div id="package-data" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Package data<a href="containers.html#package-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <code>mia</code> R package contains example datasets that are direct
conversions from the alternative <code>phyloseq</code> container to the
<code>TreeSummarizedExperiment</code> container.</p>
<p>List the <a href="https://microbiome.github.io/mia/reference/mia-datasets.html">available
datasets</a> in
the <code>mia</code> package:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="containers.html#cb31-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb31-2"><a href="containers.html#cb31-2" tabindex="-1"></a><span class="fu">data</span>(<span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span></code></pre></div>
<p>Load the <code>GlobalPatterns</code> data from the <code>mia</code> package:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="containers.html#cb32-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb32-2"><a href="containers.html#cb32-2" tabindex="-1"></a>GlobalPatterns</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div id="tengeler-desc" class="section level4 hasAnchor" number="3.3.1.1">
<h4><span class="header-section-number">3.3.1.1</span> Tengeler2020<a href="containers.html#tengeler-desc" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Tengeler2020 is derived from a randomised blinded study on the effects of gut
microbiome on attention-deficit/hyperactivity disorder (ADHD) in humanised mice
<span class="citation">(C Tengeler et al. 2020)</span>. The dataset is briefly presented in
<a href="https://microbiome.github.io/outreach/tengeler2020_presentation.html">these slides</a>.</p>
</div>
<div id="hintikka-desc" class="section level4 hasAnchor" number="3.3.1.2">
<h4><span class="header-section-number">3.3.1.2</span> HintikkaXOData<a href="containers.html#hintikka-desc" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><a href="https://microbiome.github.io/microbiomeDataSets/reference/HintikkaXOData.html">HintikkaXOData</a>
is derived from a study about the effects of fat diet and prebiotics on the
microbiome of rat models <span class="citation">(Hintikka et al. 2021)</span>. It is available in the MAE data
container for R. The dataset is briefly summarized in
<a href="https://microbiome.github.io/outreach/hintikkaxo_presentation.html">these slides</a>.</p>
</div>
</div>
<div id="experimenthub-data" class="section level3 hasAnchor" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> ExperimentHub data<a href="containers.html#experimenthub-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/ExperimentHub/inst/doc/ExperimentHub.html">ExperimentHub</a>
provides a variety of data resources, including the
<a href="https://bioconductor.org/packages/release/data/experiment/html/microbiomeDataSets.html">microbiomeDataSets</a>
package <span class="citation">(Morgan and Shepherd 2021; Leo Lahti, Ernst, and Shetty 2021)</span>.</p>
<p>A table of the available datasets is available through the
<code>availableDataSets</code> function.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="containers.html#cb34-1" tabindex="-1"></a><span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb34-2"><a href="containers.html#cb34-2" tabindex="-1"></a><span class="fu">availableDataSets</span>()</span></code></pre></div>
<pre><code>##             Dataset
## 1  GrieneisenTSData
## 2    HintikkaXOData
## 3       LahtiMLData
## 4        LahtiMData
## 5       LahtiWAData
## 6      OKeefeDSData
## 7 SilvermanAGutData
## 8        SongQAData
## 9   SprockettTHData</code></pre>
<p>All data are downloaded from ExperimentHub and cached for local
re-use. Check the <a href="https://microbiome.github.io/microbiomeDataSets/reference/index.html">man pages of each
function</a>
for a detailed documentation of the data contents and references. Let
us retrieve a <em><a href="https://bioconductor.org/packages/3.17/MultiAssayExperiment">MultiAssayExperiment</a></em> dataset:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="containers.html#cb36-1" tabindex="-1"></a><span class="co"># mae &lt;- HintikkaXOData()</span></span>
<span id="cb36-2"><a href="containers.html#cb36-2" tabindex="-1"></a><span class="co"># Since HintikkaXOData is now added to mia, we can load it directly from there</span></span>
<span id="cb36-3"><a href="containers.html#cb36-3" tabindex="-1"></a><span class="co"># We suggest to check other datasets from microbiomeDataSets</span></span>
<span id="cb36-4"><a href="containers.html#cb36-4" tabindex="-1"></a><span class="fu">data</span>(HintikkaXOData, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb36-5"><a href="containers.html#cb36-5" tabindex="-1"></a>mae <span class="ot">&lt;-</span> HintikkaXOData</span></code></pre></div>
<p>Data is available in <em><a href="https://bioconductor.org/packages/3.17/SummarizedExperiment">SummarizedExperiment</a></em>, <code>r Biocpkg("TreeSummarizedExperiment")</code> and <code>r Biocpkg("MultiAssayExperiment")</code> data containers; see the separate
page on <a href="https://microbiome.github.io/OMA/multitable.html">alternative
containers</a> for more
details.</p>
</div>
<div id="curated-metagenomic-data" class="section level3 hasAnchor" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Curated metagenomic data<a href="containers.html#curated-metagenomic-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://bioconductor.org/packages/release/data/experiment/html/curatedMetagenomicData.html">curatedMetagenomicData</a>
is a large collection of curated human microbiome datasets, provided as
<code>(Tree)SummarizedExperiment</code> objects <span class="citation">(Pasolli et al. 2017)</span>. The resource
provides curated human microbiome data including gene families, marker
abundance, marker presence, pathway abundance, pathway coverage, and
relative abundance for samples from different body sites. See the
package homepage for more details on data availability and access.</p>
<p>As one example, let us retrieve the Vatanen (2016) <span class="citation">(Vatanen et al. 2016)</span> data
set. This is a larger collection with a bit longer download time.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="containers.html#cb37-1" tabindex="-1"></a><span class="fu">library</span>(curatedMetagenomicData)</span>
<span id="cb37-2"><a href="containers.html#cb37-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">curatedMetagenomicData</span>(<span class="st">&quot;Vatanen*&quot;</span>, <span class="at">dryrun =</span> <span class="cn">FALSE</span>, <span class="at">counts =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="other-data-sources" class="section level3 hasAnchor" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Other data sources<a href="containers.html#other-data-sources" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The current collections provide access to vast microbiome data
resources. The output has to be converted into TreeSE/MAE separately.</p>
<ul>
<li><a href="https://github.com/beadyallen/MGnifyR">MGnifyR</a> provides access to <a href="https://www.ebi.ac.uk/metagenomics/">EBI/MGnify</a></li>
<li><a href="https://github.com/cran/qiitr">qiitr</a> provides access to <a href="https://qiita.com/about">QIITA</a></li>
</ul>
</div>
</div>
<div id="loading-experimental-microbiome-data" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Loading experimental microbiome data<a href="containers.html#loading-experimental-microbiome-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="s-workflow" class="section level3 hasAnchor" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> 16S workflow<a href="containers.html#s-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Result of amplicon sequencing is a large number of files that include all the sequences
that were read from samples. Those sequences need to be matched with taxa. Additionally,
we need to know how many times each taxa were found from each sample.</p>
<p>There are several algorithms to do that, and DADA2 is one of the most common.
You can find DADA2 pipeline tutorial, for example,
<a href="https://benjjneb.github.io/dada2/tutorial.html">here</a>.
After the DADA2 portion of the tutorial is completed, the data is stored into <em>phyloseq</em> object
(Bonus: Handoff to phyloseq). To store the data to <em>TreeSummarizedExperiment</em>,
follow the example below.</p>
<p>You can find full workflow script without further explanations and comments from
<a href="https://github.com/microbiome/OMA/blob/master/dada2_workflow.Rmd">here</a></p>
<p>Load required packages.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="containers.html#cb38-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb38-2"><a href="containers.html#cb38-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb38-3"><a href="containers.html#cb38-3" tabindex="-1"></a><span class="fu">library</span>(BiocManager)</span>
<span id="cb38-4"><a href="containers.html#cb38-4" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span></code></pre></div>
<p>Create arbitrary example sample metadata like it was done in the tutorial. Usually,
sample metadata is imported as a file.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="containers.html#cb39-1" tabindex="-1"></a>samples.out <span class="ot">&lt;-</span> <span class="fu">rownames</span>(seqtab.nochim)</span>
<span id="cb39-2"><a href="containers.html#cb39-2" tabindex="-1"></a>subject <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(samples.out, <span class="st">&quot;D&quot;</span>), <span class="st">`</span><span class="at">[</span><span class="st">`</span>, <span class="dv">1</span>)</span>
<span id="cb39-3"><a href="containers.html#cb39-3" tabindex="-1"></a>gender <span class="ot">&lt;-</span> <span class="fu">substr</span>(subject,<span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb39-4"><a href="containers.html#cb39-4" tabindex="-1"></a>subject <span class="ot">&lt;-</span> <span class="fu">substr</span>(subject,<span class="dv">2</span>,<span class="dv">999</span>)</span>
<span id="cb39-5"><a href="containers.html#cb39-5" tabindex="-1"></a>day <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">sapply</span>(<span class="fu">strsplit</span>(samples.out, <span class="st">&quot;D&quot;</span>), <span class="st">`</span><span class="at">[</span><span class="st">`</span>, <span class="dv">2</span>))</span>
<span id="cb39-6"><a href="containers.html#cb39-6" tabindex="-1"></a>samdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Subject=</span>subject, <span class="at">Gender=</span>gender, <span class="at">Day=</span>day)</span>
<span id="cb39-7"><a href="containers.html#cb39-7" tabindex="-1"></a>samdf<span class="sc">$</span>When <span class="ot">&lt;-</span> <span class="st">&quot;Early&quot;</span></span>
<span id="cb39-8"><a href="containers.html#cb39-8" tabindex="-1"></a>samdf<span class="sc">$</span>When[samdf<span class="sc">$</span>Day<span class="sc">&gt;</span><span class="dv">100</span>] <span class="ot">&lt;-</span> <span class="st">&quot;Late&quot;</span></span>
<span id="cb39-9"><a href="containers.html#cb39-9" tabindex="-1"></a><span class="fu">rownames</span>(samdf) <span class="ot">&lt;-</span> samples.out</span></code></pre></div>
<p>Convert data into right format and create a <em>TreeSE</em> object.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="containers.html#cb40-1" tabindex="-1"></a><span class="co"># Create a list that contains assays</span></span>
<span id="cb40-2"><a href="containers.html#cb40-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">t</span>(seqtab.nochim)</span>
<span id="cb40-3"><a href="containers.html#cb40-3" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(counts)</span>
<span id="cb40-4"><a href="containers.html#cb40-4" tabindex="-1"></a>assays <span class="ot">&lt;-</span> <span class="fu">SimpleList</span>(<span class="at">counts =</span> counts)</span>
<span id="cb40-5"><a href="containers.html#cb40-5" tabindex="-1"></a></span>
<span id="cb40-6"><a href="containers.html#cb40-6" tabindex="-1"></a><span class="co"># Convert colData and rowData into DataFrame</span></span>
<span id="cb40-7"><a href="containers.html#cb40-7" tabindex="-1"></a>samdf <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(samdf)</span>
<span id="cb40-8"><a href="containers.html#cb40-8" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(taxa)</span>
<span id="cb40-9"><a href="containers.html#cb40-9" tabindex="-1"></a></span>
<span id="cb40-10"><a href="containers.html#cb40-10" tabindex="-1"></a><span class="co"># Create TreeSE</span></span>
<span id="cb40-11"><a href="containers.html#cb40-11" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">TreeSummarizedExperiment</span>(<span class="at">assays =</span> assays,</span>
<span id="cb40-12"><a href="containers.html#cb40-12" tabindex="-1"></a>                                <span class="at">colData =</span> samdf,</span>
<span id="cb40-13"><a href="containers.html#cb40-13" tabindex="-1"></a>                                <span class="at">rowData =</span> taxa</span>
<span id="cb40-14"><a href="containers.html#cb40-14" tabindex="-1"></a>                                )</span>
<span id="cb40-15"><a href="containers.html#cb40-15" tabindex="-1"></a></span>
<span id="cb40-16"><a href="containers.html#cb40-16" tabindex="-1"></a><span class="co"># Remove mock sample like it is also done in DADA2 pipeline tutorial</span></span>
<span id="cb40-17"><a href="containers.html#cb40-17" tabindex="-1"></a>tse <span class="ot">&lt;-</span> tse[ , <span class="fu">colnames</span>(tse) <span class="sc">!=</span> <span class="st">&quot;mock&quot;</span>]</span></code></pre></div>
<p>Add sequences into <em>referenceSeq</em> slot and convert rownames into simpler format.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="containers.html#cb41-1" tabindex="-1"></a><span class="co"># Convert sequences into right format</span></span>
<span id="cb41-2"><a href="containers.html#cb41-2" tabindex="-1"></a>dna <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">DNAStringSet</span>( <span class="fu">rownames</span>(tse) )</span>
<span id="cb41-3"><a href="containers.html#cb41-3" tabindex="-1"></a><span class="co"># Add sequences into referenceSeq slot</span></span>
<span id="cb41-4"><a href="containers.html#cb41-4" tabindex="-1"></a><span class="fu">referenceSeq</span>(tse) <span class="ot">&lt;-</span> dna</span>
<span id="cb41-5"><a href="containers.html#cb41-5" tabindex="-1"></a><span class="co"># Convert rownames into ASV_number format</span></span>
<span id="cb41-6"><a href="containers.html#cb41-6" tabindex="-1"></a><span class="fu">rownames</span>(tse) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;ASV&quot;</span>, <span class="fu">seq</span>( <span class="fu">nrow</span>(tse) ))</span>
<span id="cb41-7"><a href="containers.html#cb41-7" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 232 20 
## metadata(0):
## assays(1): counts
## rownames(232): ASV1 ASV2 ... ASV231 ASV232
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(20): F3D0 F3D1 ... F3D9 Mock
## colData names(4): Subject Gender Day When
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
## referenceSeq: a DNAStringSet (232 sequences)</code></pre>
</div>
<div id="import-from-file" class="section level3 hasAnchor" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> Import from external files<a href="containers.html#import-from-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Microbiome (taxonomic) profiling data is commonly distributed in
various file formats. You can import such external data files as a
(Tree)SummarizedExperiment object, but the details depend on the file
format. Here, we provide examples for common formats. Some datasets and raw
files to learn how to import raw data and construct TreeSE/MAE containers are
available in the
<a href="https://github.com/microbiome/data">microbiome data repository</a>.</p>
<div id="csv-import" class="section level4 hasAnchor" number="3.4.2.1">
<h4><span class="header-section-number">3.4.2.1</span> CSV import<a href="containers.html#csv-import" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>CSV data tables</strong> can be imported with the standard R functions,
then converted to the desired format. For detailed examples, you can
check the <a href="https://bioconductor.org/help/course-materials/2019/BSS2019/04_Practical_CoreApproachesInBioconductor.html">Bioconductor course
material</a>
by Martin Morgan. You can also check the <a href="https://github.com/microbiome/OMA/tree/master/data">example
files</a> and
construct your own CSV files accordingly.</p>
<p>Recommendations for the CSV files are the following. File names are
arbitrary; we refer here to the same names as in the examples:</p>
<ul>
<li><p>Abundance table (<code>assay_taxa.csv</code>): data matrix (features x
samples); first column provides feature IDs, the first row provides
sample IDs; other values should be numeric (abundances).</p></li>
<li><p>Row data (<code>rowdata_taxa.csv</code>): data table (features x info); first
column provides feature IDs, the first row provides column headers;
this file usually contains the taxonomic mapping between different
taxonomic levels. Ideally, the feature IDs (row names) match one-to-one with
the abundance table row names.</p></li>
<li><p>Column data (<code>coldata.csv</code>): data table (samples x info); first
column provides sample IDs, the first row provides column headers;
this file usually contains the sample metadata/phenodata (such as
subject age, health etc). Ideally, the sample IDs match one-to-one with
the abundance table column names.</p></li>
</ul>
<p>After you have set up the CSV files, you can read them in R:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="containers.html#cb43-1" tabindex="-1"></a>count_file  <span class="ot">&lt;-</span> <span class="st">&quot;data/assay_taxa.csv&quot;</span></span>
<span id="cb43-2"><a href="containers.html#cb43-2" tabindex="-1"></a>tax_file    <span class="ot">&lt;-</span> <span class="st">&quot;data/rowdata_taxa.csv&quot;</span></span>
<span id="cb43-3"><a href="containers.html#cb43-3" tabindex="-1"></a>sample_file <span class="ot">&lt;-</span> <span class="st">&quot;data/coldata.csv&quot;</span></span>
<span id="cb43-4"><a href="containers.html#cb43-4" tabindex="-1"></a></span>
<span id="cb43-5"><a href="containers.html#cb43-5" tabindex="-1"></a><span class="co"># Load files</span></span>
<span id="cb43-6"><a href="containers.html#cb43-6" tabindex="-1"></a>counts  <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(count_file, <span class="at">row.names=</span><span class="dv">1</span>)   <span class="co"># Abundance table (e.g. ASV data; to assay data)</span></span>
<span id="cb43-7"><a href="containers.html#cb43-7" tabindex="-1"></a>tax     <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(tax_file, <span class="at">row.names=</span><span class="dv">1</span>)     <span class="co"># Taxonomy table (to rowData)</span></span>
<span id="cb43-8"><a href="containers.html#cb43-8" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(sample_file, <span class="at">row.names=</span><span class="dv">1</span>)  <span class="co"># Sample data (to colData)</span></span></code></pre></div>
<p>After reading the data in R, ensure the following:</p>
<ul>
<li><p>abundance table (<code>counts</code>): numeric <code>matrix</code>, with feature IDs as
rownames and sample IDs as column names</p></li>
<li><p>rowdata (<code>tax</code>): <code>DataFrame</code>, with feature IDs as rownames. If this
is a <code>data.frame</code> you can use the function <code>DataFrame()</code> to change
the format. Column names are free but in microbiome analysis they
usually they refer to taxonomic ranks. The rownames in rowdata
should match with rownames in abundance table.</p></li>
<li><p>coldata (<code>samples</code>): <code>DataFrame</code>, with sample IDs as rownames. If
this is a <code>data.frame</code> you can use the function <code>DataFrame()</code> to
change the format. Column names are free. The rownames in coldata
should match with colnames in abundance table.</p></li>
</ul>
<p><strong>Always ensure that the tables have rownames!</strong> The <em>TreeSE</em> constructor compares
rownames and ensures that, for example, right samples are linked with right patient.</p>
<p>Also ensure that the row and column names match one-to-one between
abundance table, rowdata, and coldata:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="containers.html#cb44-1" tabindex="-1"></a><span class="co"># Match rows and columns</span></span>
<span id="cb44-2"><a href="containers.html#cb44-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> counts[<span class="fu">rownames</span>(tax), <span class="fu">rownames</span>(samples)]</span>
<span id="cb44-3"><a href="containers.html#cb44-3" tabindex="-1"></a></span>
<span id="cb44-4"><a href="containers.html#cb44-4" tabindex="-1"></a><span class="co"># Let us ensure that the data is in correct (numeric matrix) format:</span></span>
<span id="cb44-5"><a href="containers.html#cb44-5" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(counts)</span></code></pre></div>
<p>If you hesitate about the format of the data, you can compare to one
of the available demonstration datasets, and make sure that your data
components have the same format.</p>
<p>There are many different source files and many different ways to read
data in R. One can do data manipulation in R as well. Investigate the
entries as follows.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="containers.html#cb45-1" tabindex="-1"></a><span class="co"># coldata rownames match assay colnames</span></span>
<span id="cb45-2"><a href="containers.html#cb45-2" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(samples) <span class="sc">==</span> <span class="fu">colnames</span>(counts)) <span class="co"># our dataset</span></span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="containers.html#cb47-1" tabindex="-1"></a><span class="fu">class</span>(samples) <span class="co"># should be data.frame or DataFrame</span></span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="containers.html#cb49-1" tabindex="-1"></a><span class="co"># rowdata rownames match assay rownames</span></span>
<span id="cb49-2"><a href="containers.html#cb49-2" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(tax) <span class="sc">==</span> <span class="fu">rownames</span>(counts)) <span class="co"># our dataset</span></span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="containers.html#cb51-1" tabindex="-1"></a><span class="fu">class</span>(tax) <span class="co"># should be data.frame or DataFrame</span></span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="containers.html#cb53-1" tabindex="-1"></a><span class="co"># Counts </span></span>
<span id="cb53-2"><a href="containers.html#cb53-2" tabindex="-1"></a><span class="fu">class</span>(counts) <span class="co"># should be a numeric matrix</span></span></code></pre></div>
<pre><code>## [1] &quot;matrix&quot; &quot;array&quot;</code></pre>
</div>
</div>
<div id="constructing-treesummarizedexperiment" class="section level3 hasAnchor" number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> Constructing TreeSummarizedExperiment<a href="containers.html#constructing-treesummarizedexperiment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now let us create the TreeSE object from the input data tables. Here
we also convert the data objects in their preferred formats:</p>
<ul>
<li>counts &gt; numeric matrix</li>
<li>rowData &gt; DataFrame</li>
<li>colData &gt; DataFrame</li>
</ul>
<p>The <code>SimpleList</code> could be used to include multiple alternative assays, if
necessary.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="containers.html#cb55-1" tabindex="-1"></a><span class="co"># Create a TreeSE</span></span>
<span id="cb55-2"><a href="containers.html#cb55-2" tabindex="-1"></a>tse_taxa <span class="ot">&lt;-</span> <span class="fu">TreeSummarizedExperiment</span>(<span class="at">assays =</span>  <span class="fu">SimpleList</span>(<span class="at">counts =</span> counts),</span>
<span id="cb55-3"><a href="containers.html#cb55-3" tabindex="-1"></a>                                     <span class="at">colData =</span> <span class="fu">DataFrame</span>(samples),</span>
<span id="cb55-4"><a href="containers.html#cb55-4" tabindex="-1"></a>                                     <span class="at">rowData =</span> <span class="fu">DataFrame</span>(tax))</span>
<span id="cb55-5"><a href="containers.html#cb55-5" tabindex="-1"></a></span>
<span id="cb55-6"><a href="containers.html#cb55-6" tabindex="-1"></a>tse_taxa</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 12706 40 
## metadata(0):
## assays(1): counts
## rownames(12706): GAYR01026362.62.2014 CVJT01000011.50.2173 ...
##   JRJTB:03787:02429 JRJTB:03787:02478
## rowData names(7): Phylum Class ... Species OTU
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Now you should have a ready-made TreeSE data object that can be used in downstream analyses.</p>
</div>
<div id="constructing-multiassayexperiment" class="section level3 hasAnchor" number="3.4.4">
<h3><span class="header-section-number">3.4.4</span> Constructing MultiAssayExperiment<a href="containers.html#constructing-multiassayexperiment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To construct a <em>MultiAssayExperiment</em> object, just combine multiple <em>TreeSE</em> data containers.
Here we import metabolite data from the same study.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="containers.html#cb57-1" tabindex="-1"></a>count_file <span class="ot">&lt;-</span> <span class="st">&quot;data/assay_metabolites.csv&quot;</span></span>
<span id="cb57-2"><a href="containers.html#cb57-2" tabindex="-1"></a>sample_file <span class="ot">&lt;-</span> <span class="st">&quot;data/coldata.csv&quot;</span></span>
<span id="cb57-3"><a href="containers.html#cb57-3" tabindex="-1"></a></span>
<span id="cb57-4"><a href="containers.html#cb57-4" tabindex="-1"></a><span class="co"># Load files</span></span>
<span id="cb57-5"><a href="containers.html#cb57-5" tabindex="-1"></a>counts  <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(count_file, <span class="at">row.names=</span><span class="dv">1</span>)  </span>
<span id="cb57-6"><a href="containers.html#cb57-6" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(sample_file, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb57-7"><a href="containers.html#cb57-7" tabindex="-1"></a></span>
<span id="cb57-8"><a href="containers.html#cb57-8" tabindex="-1"></a><span class="co"># Create a TreeSE for the metabolite data</span></span>
<span id="cb57-9"><a href="containers.html#cb57-9" tabindex="-1"></a>tse_metabolite <span class="ot">&lt;-</span> <span class="fu">TreeSummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">SimpleList</span>(<span class="at">concs =</span> <span class="fu">as.matrix</span>(counts)),</span>
<span id="cb57-10"><a href="containers.html#cb57-10" tabindex="-1"></a>                                           <span class="at">colData =</span> <span class="fu">DataFrame</span>(samples))</span>
<span id="cb57-11"><a href="containers.html#cb57-11" tabindex="-1"></a></span>
<span id="cb57-12"><a href="containers.html#cb57-12" tabindex="-1"></a>tse_metabolite</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 38 40 
## metadata(0):
## assays(1): concs
## rownames(38): Butyrate Acetate ... Malonate 1,3-dihydroxyacetone
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Now we can combine these two experiments into <em>MAE</em>.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="containers.html#cb59-1" tabindex="-1"></a><span class="co"># Create an ExperimentList that includes experiments</span></span>
<span id="cb59-2"><a href="containers.html#cb59-2" tabindex="-1"></a>experiments <span class="ot">&lt;-</span> <span class="fu">ExperimentList</span>(<span class="at">microbiome =</span> tse_taxa, </span>
<span id="cb59-3"><a href="containers.html#cb59-3" tabindex="-1"></a>                              <span class="at">metabolite =</span> tse_metabolite)</span>
<span id="cb59-4"><a href="containers.html#cb59-4" tabindex="-1"></a></span>
<span id="cb59-5"><a href="containers.html#cb59-5" tabindex="-1"></a><span class="co"># Create a MAE</span></span>
<span id="cb59-6"><a href="containers.html#cb59-6" tabindex="-1"></a>mae <span class="ot">&lt;-</span> <span class="fu">MultiAssayExperiment</span>(<span class="at">experiments =</span> experiments)</span>
<span id="cb59-7"><a href="containers.html#cb59-7" tabindex="-1"></a></span>
<span id="cb59-8"><a href="containers.html#cb59-8" tabindex="-1"></a>mae</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 2:
##  [1] microbiome: TreeSummarizedExperiment with 12706 rows and 40 columns
##  [2] metabolite: TreeSummarizedExperiment with 38 rows and 40 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files</code></pre>
</div>
<div id="import-functions-for-standard-formats" class="section level3 hasAnchor" number="3.4.5">
<h3><span class="header-section-number">3.4.5</span> Import functions for standard formats<a href="containers.html#import-functions-for-standard-formats" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Specific import functions are provided for:</p>
<ul>
<li>Biom files (see <code>help(mia::loadFromBiom)</code>)</li>
<li>QIIME2 files (see <code>help(mia::loadFromQIIME2)</code>)</li>
<li>Mothur files (see <code>help(mia::loadFromMothur)</code>)</li>
</ul>
<div id="biom-import" class="section level4 hasAnchor" number="3.4.5.1">
<h4><span class="header-section-number">3.4.5.1</span> Biom import<a href="containers.html#biom-import" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Here we show how <a href="https://biom-format.org/">Biom files</a> are imported into
a TreeSE object using as an example Tengeler2020, which is further described in section <a href="containers.html#tengeler-desc">3.3.1.1</a>. This dataset consists of 3 files, which can be
fetched or downloaded from
<a href="https://github.com/microbiome/data/tree/main/Tengeler2020">this repository</a>:</p>
<ul>
<li>biom file: abundance table and taxonomy information</li>
<li>csv file: sample metadata</li>
<li>tree file: phylogenetic tree</li>
</ul>
<p>To begin with, we store the data in a local directory within the working
directory, such as <em>data/</em>, and define the source file paths.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="containers.html#cb61-1" tabindex="-1"></a>biom_file_path <span class="ot">&lt;-</span> <span class="st">&quot;data/Aggregated_humanization2.biom&quot;</span></span>
<span id="cb61-2"><a href="containers.html#cb61-2" tabindex="-1"></a>sample_meta_file_path <span class="ot">&lt;-</span> <span class="st">&quot;data/Mapping_file_ADHD_aggregated.csv&quot;</span></span>
<span id="cb61-3"><a href="containers.html#cb61-3" tabindex="-1"></a>tree_file_path <span class="ot">&lt;-</span> <span class="st">&quot;data/Data_humanization_phylo_aggregation.tre&quot;</span></span></code></pre></div>
<p>Now we can read in the biom file and convert it into a TreeSE object. In addition, we retrieve the rank names from the prefixes of the feature names and then remove them with the <code>rankFromPrefix</code> and <code>removeTaxaPrefixes</code> optional arguments.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="containers.html#cb62-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb62-2"><a href="containers.html#cb62-2" tabindex="-1"></a></span>
<span id="cb62-3"><a href="containers.html#cb62-3" tabindex="-1"></a><span class="co"># read biom and convert it to TreeSE</span></span>
<span id="cb62-4"><a href="containers.html#cb62-4" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">loadFromBiom</span>(biom_file_path,</span>
<span id="cb62-5"><a href="containers.html#cb62-5" tabindex="-1"></a>                    <span class="at">rankFromPrefix =</span> <span class="cn">TRUE</span>,</span>
<span id="cb62-6"><a href="containers.html#cb62-6" tabindex="-1"></a>                    <span class="at">removeTaxaPrefixes =</span> <span class="cn">TRUE</span>)</span>
<span id="cb62-7"><a href="containers.html#cb62-7" tabindex="-1"></a></span>
<span id="cb62-8"><a href="containers.html#cb62-8" tabindex="-1"></a><span class="co"># Check</span></span>
<span id="cb62-9"><a href="containers.html#cb62-9" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): taxonomy1 Phylum ... Family Genus
## colnames(27): A110 A111 ... A38 A39
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<p>The <code>assays</code> slot includes a list of abundance tables. The imported
abundance table is named as counts. Let us inspect only the first
cols and rows.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="containers.html#cb64-1" tabindex="-1"></a><span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>##           A110  A111  A12
## 1726470  17722 11630    0
## 1726471  12052     0 2679
## 17264731     0   970    0</code></pre>
<p>The <code>rowdata</code> includes taxonomic information from the biom file. The <code>head()</code> command shows just the beginning of the data table for an overview.</p>
<p><code>knitr::kable()</code> helps print the information more nicely.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="containers.html#cb66-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(tse))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##             taxonomy1          Phylum            Class              Order
##           &lt;character&gt;     &lt;character&gt;      &lt;character&gt;        &lt;character&gt;
## 1726470  &quot;k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726471  &quot;k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264731 &quot;k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264726 &quot;k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726472  &quot;k__Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales
## 17264724 &quot;k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
##                       Family            Genus
##                  &lt;character&gt;      &lt;character&gt;
## 1726470       Bacteroidaceae     Bacteroides&quot;
## 1726471       Bacteroidaceae     Bacteroides&quot;
## 17264731  Porphyromonadaceae Parabacteroides&quot;
## 17264726      Bacteroidaceae     Bacteroides&quot;
## 1726472  Verrucomicrobiaceae     Akkermansia&quot;
## 17264724      Bacteroidaceae     Bacteroides&quot;</code></pre>
<p>We further polish the feature names by removing unnecessary characters and then replace the original rowData with its updated version.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="containers.html#cb68-1" tabindex="-1"></a><span class="co"># Genus level has additional &#39;\&quot;&#39;, so let&#39;s delete that also</span></span>
<span id="cb68-2"><a href="containers.html#cb68-2" tabindex="-1"></a>rowdata_modified <span class="ot">&lt;-</span> BiocParallel<span class="sc">::</span><span class="fu">bplapply</span>(<span class="fu">rowData</span>(tse), </span>
<span id="cb68-3"><a href="containers.html#cb68-3" tabindex="-1"></a>                                           <span class="at">FUN =</span> stringr<span class="sc">::</span>str_remove, </span>
<span id="cb68-4"><a href="containers.html#cb68-4" tabindex="-1"></a>                                           <span class="at">pattern =</span> <span class="st">&#39;</span><span class="sc">\&quot;</span><span class="st">&#39;</span>)</span>
<span id="cb68-5"><a href="containers.html#cb68-5" tabindex="-1"></a></span>
<span id="cb68-6"><a href="containers.html#cb68-6" tabindex="-1"></a><span class="co"># rowdata_modified is a list, so convert this back to DataFrame format. </span></span>
<span id="cb68-7"><a href="containers.html#cb68-7" tabindex="-1"></a><span class="co"># and assign the cleaned data back to the TSE rowData</span></span>
<span id="cb68-8"><a href="containers.html#cb68-8" tabindex="-1"></a><span class="fu">rowData</span>(tse) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(rowdata_modified)</span>
<span id="cb68-9"><a href="containers.html#cb68-9" tabindex="-1"></a></span>
<span id="cb68-10"><a href="containers.html#cb68-10" tabindex="-1"></a><span class="co"># Now we have a nicer table</span></span>
<span id="cb68-11"><a href="containers.html#cb68-11" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(tse))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##            taxonomy1          Phylum            Class              Order
##          &lt;character&gt;     &lt;character&gt;      &lt;character&gt;        &lt;character&gt;
## 1726470  k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726471  k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264731 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264726 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726472  k__Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales
## 17264724 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
##                       Family           Genus
##                  &lt;character&gt;     &lt;character&gt;
## 1726470       Bacteroidaceae     Bacteroides
## 1726471       Bacteroidaceae     Bacteroides
## 17264731  Porphyromonadaceae Parabacteroides
## 17264726      Bacteroidaceae     Bacteroides
## 1726472  Verrucomicrobiaceae     Akkermansia
## 17264724      Bacteroidaceae     Bacteroides</code></pre>
<p>We notice that the imported biom file did not contain any colData yet,
so only an empty dataframe appears in this slot.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="containers.html#cb70-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 0 columns</code></pre>
<p>Let us add colData from the sample metadata, which is stored in a CSV file.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="containers.html#cb72-1" tabindex="-1"></a><span class="co"># CSV file with colnames in the first row and rownames in the first column</span></span>
<span id="cb72-2"><a href="containers.html#cb72-2" tabindex="-1"></a>sample_meta <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(sample_meta_file_path,</span>
<span id="cb72-3"><a href="containers.html#cb72-3" tabindex="-1"></a>                        <span class="at">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb72-4"><a href="containers.html#cb72-4" tabindex="-1"></a></span>
<span id="cb72-5"><a href="containers.html#cb72-5" tabindex="-1"></a><span class="co"># Add this sample data to colData of the taxonomic data object</span></span>
<span id="cb72-6"><a href="containers.html#cb72-6" tabindex="-1"></a><span class="co"># Note that the data must be given in a DataFrame format (required for our purposes)</span></span>
<span id="cb72-7"><a href="containers.html#cb72-7" tabindex="-1"></a><span class="fu">colData</span>(tse) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(sample_meta)</span></code></pre></div>
<p>Now the colData includes the sample metadata.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="containers.html#cb73-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(tse))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##        Treatment      Cohort TreatmentxCohort Description
##      &lt;character&gt; &lt;character&gt;      &lt;character&gt; &lt;character&gt;
## A110        ADHD    Cohort_1    ADHD_Cohort_1        A110
## A12         ADHD    Cohort_1    ADHD_Cohort_1         A12
## A15         ADHD    Cohort_1    ADHD_Cohort_1         A15
## A19         ADHD    Cohort_1    ADHD_Cohort_1         A19
## A21         ADHD    Cohort_2    ADHD_Cohort_2         A21
## A23         ADHD    Cohort_2    ADHD_Cohort_2         A23</code></pre>
<p>Finally, we add a phylogenetic tree to the rowData slot. Such feature is available only in TreeSE objects. Similarly, Trees specifying the sample hierarchy can be stored in the colTree slot.</p>
<p>Here, we read in the file containing the phylogenetic tree and insert it in corresponding slot of the TreeSE object.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="containers.html#cb75-1" tabindex="-1"></a><span class="co"># Reads the tree file</span></span>
<span id="cb75-2"><a href="containers.html#cb75-2" tabindex="-1"></a>tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(tree_file_path)</span>
<span id="cb75-3"><a href="containers.html#cb75-3" tabindex="-1"></a></span>
<span id="cb75-4"><a href="containers.html#cb75-4" tabindex="-1"></a><span class="co"># Add tree to rowTree</span></span>
<span id="cb75-5"><a href="containers.html#cb75-5" tabindex="-1"></a><span class="fu">rowTree</span>(tse) <span class="ot">&lt;-</span> tree</span>
<span id="cb75-6"><a href="containers.html#cb75-6" tabindex="-1"></a></span>
<span id="cb75-7"><a href="containers.html#cb75-7" tabindex="-1"></a><span class="co"># Check</span></span>
<span id="cb75-8"><a href="containers.html#cb75-8" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): taxonomy1 Phylum ... Family Genus
## colnames(27): A110 A12 ... A35 A38
## colData names(4): Treatment Cohort TreatmentxCohort Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (151 rows)
## rowTree: 1 phylo tree(s) (151 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Now the rowTree slot contains the phylogenetic tree:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="containers.html#cb77-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowTree</span>(tse))</span></code></pre></div>
</div>
</div>
<div id="conversions-between-data-formats-in-r" class="section level3 hasAnchor" number="3.4.6">
<h3><span class="header-section-number">3.4.6</span> Conversions between data formats in R<a href="containers.html#conversions-between-data-formats-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If the data has already been imported in R in another format, it
can be readily converted into <code>TreeSummarizedExperiment</code>, as shown in our next
example. Note that similar conversion functions to
<code>TreeSummarizedExperiment</code> are available for multiple data formats via
the <code>mia</code> package (see makeTreeSummarizedExperimentFrom* for phyloseq,
Biom, and DADA2).</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="containers.html#cb78-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb78-2"><a href="containers.html#cb78-2" tabindex="-1"></a></span>
<span id="cb78-3"><a href="containers.html#cb78-3" tabindex="-1"></a><span class="co"># phyloseq example data</span></span>
<span id="cb78-4"><a href="containers.html#cb78-4" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns, <span class="at">package=</span><span class="st">&quot;phyloseq&quot;</span>) </span>
<span id="cb78-5"><a href="containers.html#cb78-5" tabindex="-1"></a>GlobalPatterns_phyloseq <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb78-6"><a href="containers.html#cb78-6" tabindex="-1"></a>GlobalPatterns_phyloseq</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="containers.html#cb80-1" tabindex="-1"></a><span class="co"># convert phyloseq to TSE</span></span>
<span id="cb80-2"><a href="containers.html#cb80-2" tabindex="-1"></a>GlobalPatterns_TSE <span class="ot">&lt;-</span> <span class="fu">makeTreeSummarizedExperimentFromPhyloseq</span>(GlobalPatterns_phyloseq) </span>
<span id="cb80-3"><a href="containers.html#cb80-3" tabindex="-1"></a>GlobalPatterns_TSE</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>We can also convert <code>TreeSummarizedExperiment</code> objects into <code>phyloseq</code>
with respect to the shared components that are supported by both
formats (i.e.taxonomic abundance table, sample metadata, taxonomic
table, phylogenetic tree, sequence information). This is useful for
instance when additional methods are available for <code>phyloseq</code>.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="containers.html#cb82-1" tabindex="-1"></a><span class="co"># convert TSE to phyloseq</span></span>
<span id="cb82-2"><a href="containers.html#cb82-2" tabindex="-1"></a>GlobalPatterns_phyloseq2 <span class="ot">&lt;-</span> <span class="fu">makePhyloseqFromTreeSummarizedExperiment</span>(GlobalPatterns_TSE) </span>
<span id="cb82-3"><a href="containers.html#cb82-3" tabindex="-1"></a>GlobalPatterns_phyloseq2</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
<p>Conversion is possible between other data formats. Interested readers can refer to the following functions:
* <a href="https://microbiome.github.io/mia/reference/makeTreeSummarizedExperimentFromDADA2.html">makeTreeSummarizedExperimentFromDADA2</a>
* <a href="https://microbiome.github.io/mia/reference/makeSummarizedExperimentFromBiom.html">makeSummarizedExperimentFromBiom</a>
* <a href="https://microbiome.github.io/mia/reference/loadFromMetaphlan.html">loadFromMetaphlan</a>
* <a href="https://microbiome.github.io/mia/reference/loadFromQIIME2.html">readQZA</a></p>

</div>
</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="packages.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="datamanipulation.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/04_containers.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/04_containers.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/04_containers.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
