<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Data Manipulation | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 4 Data Manipulation | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.32 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Data Manipulation | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Data Manipulation | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="packages.html"/>
<link rel="next" href="quality-control.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.1/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.2.2</b> colData</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.2.3</b> rowData</a></li>
<li class="chapter" data-level="2.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.2.4</b> rowTree</a></li>
<li class="chapter" data-level="2.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>2.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>2.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>2.3.1</b> 16S workflow</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.2</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.3" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.3</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#some-available-packages"><i class="fa fa-check"></i><b>3.2</b> Some available packages</a></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.2</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.2.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
<li class="chapter" data-level="" data-path="community-diversity.html"><a href="community-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#community-comparisons-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.2</b> Community comparisons by beta diversity analysis</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="community-similarity.html"><a href="community-similarity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.4.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.4.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.4.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="community-similarity.html"><a href="community-similarity.html#further-reading"><i class="fa fa-check"></i><b>8.5</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="community-similarity.html"><a href="community-similarity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community typing (clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#hiearchical-clustering"><i class="fa fa-check"></i><b>10.1</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>10.2</b> K-means clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.3.1</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#community-detection"><i class="fa fa-check"></i><b>10.4</b> Community Detection</a></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.5</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.5.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.5.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>10.5.3</b> Taxa vs taxa</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.6</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>11.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.1.6</b> LinDA</a></li>
<li class="chapter" data-level="11.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.7</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.8" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.8</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>12</b> Machine learning</a>
<ul>
<li class="chapter" data-level="12.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>12.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="12.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>12.2</b> Unsupervised machine learning</a></li>
<li class="chapter" data-level="" data-path="machine_learning.html"><a href="machine_learning.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#multi-omics-factor-analysis"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
<li class="chapter" data-level="" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>14</b> Visualization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>14.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>14.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="14.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>14.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>14.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>14.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="14.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>14.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>14.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>14.3.1</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="viz-chapter.html"><a href="viz-chapter.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="15" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>15</b> Resources</a>
<ul>
<li class="chapter" data-level="15.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>15.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="15.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>15.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="15.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>15.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="15.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>15.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="15.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>15.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="15.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>15.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>15.2.1</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="resources.html"><a href="resources.html#reproducible-reporting-with-rmarkdown"><i class="fa fa-check"></i><b>15.3</b> Reproducible reporting with Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>16</b> Extra material</a>
<ul>
<li class="chapter" data-level="16.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>16.1</b> Interactive 3D Plots</a></li>
<li class="chapter" data-level="16.2" data-path="extras.html"><a href="extras.html#permanova-comparison"><i class="fa fa-check"></i><b>16.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="16.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>16.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>17</b> Acknowledgments</a></li>
<li class="chapter" data-level="18" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>18</b> Exercises</a>
<ul>
<li class="chapter" data-level="18.1" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>18.1</b> Workflows</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting"><i class="fa fa-check"></i><b>18.1.1</b> Reproducible reporting</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-basic-components"><i class="fa fa-check"></i><b>18.2</b> TreeSummarizedExperiment: basic components</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="exercises.html"><a href="exercises.html#basic-summaries"><i class="fa fa-check"></i><b>18.2.1</b> Basic summaries</a></li>
<li class="chapter" data-level="18.2.2" data-path="exercises.html"><a href="exercises.html#taxonomic-abundance-table-assay"><i class="fa fa-check"></i><b>18.2.2</b> Taxonomic abundance table (assay)</a></li>
<li class="chapter" data-level="18.2.3" data-path="exercises.html"><a href="exercises.html#sample-side-information"><i class="fa fa-check"></i><b>18.2.3</b> Sample side information</a></li>
<li class="chapter" data-level="18.2.4" data-path="exercises.html"><a href="exercises.html#feature-side-information"><i class="fa fa-check"></i><b>18.2.4</b> Feature side information</a></li>
<li class="chapter" data-level="18.2.5" data-path="exercises.html"><a href="exercises.html#other-components"><i class="fa fa-check"></i><b>18.2.5</b> Other components</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-data-import"><i class="fa fa-check"></i><b>18.3</b> TreeSummarizedExperiment: data import</a></li>
<li class="chapter" data-level="18.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>18.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="18.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>18.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="18.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>18.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="18.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>18.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="18.4.4" data-path="exercises.html"><a href="exercises.html#explore-the-data"><i class="fa fa-check"></i><b>18.4.4</b> Explore the data</a></li>
<li class="chapter" data-level="18.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>18.4.5</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="18.5" data-path="exercises.html"><a href="exercises.html#transformations"><i class="fa fa-check"></i><b>18.5</b> Transformations</a>
<ul>
<li class="chapter" data-level="18.5.1" data-path="exercises.html"><a href="exercises.html#transformations-1"><i class="fa fa-check"></i><b>18.5.1</b> Transformations</a></li>
</ul></li>
<li class="chapter" data-level="18.6" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>18.6</b> Taxonomic levels</a>
<ul>
<li class="chapter" data-level="18.6.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels-1"><i class="fa fa-check"></i><b>18.6.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="18.6.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments-altexp"><i class="fa fa-check"></i><b>18.6.2</b> Alternative experiments (altExp)</a></li>
</ul></li>
<li class="chapter" data-level="18.7" data-path="exercises.html"><a href="exercises.html#alpha-diversity"><i class="fa fa-check"></i><b>18.7</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="18.7.1" data-path="exercises.html"><a href="exercises.html#alpha-diversity-basics"><i class="fa fa-check"></i><b>18.7.1</b> Alpha diversity basics</a></li>
<li class="chapter" data-level="18.7.2" data-path="exercises.html"><a href="exercises.html#alpha-diversity-extra"><i class="fa fa-check"></i><b>18.7.2</b> Alpha diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="18.8" data-path="exercises.html"><a href="exercises.html#community-composition"><i class="fa fa-check"></i><b>18.8</b> Community composition</a>
<ul>
<li class="chapter" data-level="18.8.1" data-path="exercises.html"><a href="exercises.html#beta-diversity-basics"><i class="fa fa-check"></i><b>18.8.1</b> Beta diversity basics</a></li>
<li class="chapter" data-level="18.8.2" data-path="exercises.html"><a href="exercises.html#beta-diversity-extra"><i class="fa fa-check"></i><b>18.8.2</b> Beta diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="18.9" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>18.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="18.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>18.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="18.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>18.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="18.10" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>18.10</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="18.10.1" data-path="exercises.html"><a href="exercises.html#univariate-analyses"><i class="fa fa-check"></i><b>18.10.1</b> Univariate analyses</a></li>
<li class="chapter" data-level="18.10.2" data-path="exercises.html"><a href="exercises.html#differential-abundance-analysis-1"><i class="fa fa-check"></i><b>18.10.2</b> Differential abundance analysis</a></li>
</ul></li>
<li class="chapter" data-level="18.11" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>18.11</b> Multiomics</a>
<ul>
<li class="chapter" data-level="18.11.1" data-path="exercises.html"><a href="exercises.html#introduction-to-multiassayexperiment-mae"><i class="fa fa-check"></i><b>18.11.1</b> Introduction to MultiAssayExperiment (MAE)</a></li>
<li class="chapter" data-level="18.11.2" data-path="exercises.html"><a href="exercises.html#introduction-to-multiomics"><i class="fa fa-check"></i><b>18.11.2</b> Introduction to multiomics</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="datamanipulation" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Data Manipulation<a href="datamanipulation.html#datamanipulation" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="tidying-and-subsetting" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Tidying and subsetting<a href="datamanipulation.html#tidying-and-subsetting" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="tidy-data" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Tidy data<a href="datamanipulation.html#tidy-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For several custom analysis and visualization packages, such as those from the
<code>tidyverse</code>, the <code>SE</code> data can be converted to long data.frame format with
<code>meltAssay</code>.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="datamanipulation.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb81-2"><a href="datamanipulation.html#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb81-3"><a href="datamanipulation.html#cb81-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb81-4"><a href="datamanipulation.html#cb81-4" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(tse, <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>, <span class="at">method=</span><span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb81-5"><a href="datamanipulation.html#cb81-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-6"><a href="datamanipulation.html#cb81-6" aria-hidden="true" tabindex="-1"></a>molten_tse <span class="ot">&lt;-</span> <span class="fu">meltAssay</span>(tse,</span>
<span id="cb81-7"><a href="datamanipulation.html#cb81-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">add_row_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-8"><a href="datamanipulation.html#cb81-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">add_col_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-9"><a href="datamanipulation.html#cb81-9" aria-hidden="true" tabindex="-1"></a>                        <span class="at">assay_name =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb81-10"><a href="datamanipulation.html#cb81-10" aria-hidden="true" tabindex="-1"></a>molten_tse</span></code></pre></div>
<pre><code>## # A tibble: 499,616 × 17
##    FeatureID SampleID relabund…¹ Kingdom Phylum Class Order Family Genus Species
##    &lt;fct&gt;     &lt;fct&gt;         &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;  
##  1 549322    CL3               0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  2 549322    CC1               0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  3 549322    SV1               0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  4 549322    M31Fcsw           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  5 549322    M11Fcsw           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  6 549322    M31Plmr           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  7 549322    M11Plmr           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  8 549322    F21Plmr           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
##  9 549322    M31Tong           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
## 10 549322    M11Tong           0 Archaea Crena… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;   
## # … with 499,606 more rows, 7 more variables: X.SampleID &lt;fct&gt;, Primer &lt;fct&gt;,
## #   Final_Barcode &lt;fct&gt;, Barcode_truncated_plus_T &lt;fct&gt;,
## #   Barcode_full_length &lt;fct&gt;, SampleType &lt;fct&gt;, Description &lt;fct&gt;, and
## #   abbreviated variable name ¹​relabundance</code></pre>
</div>
<div id="subsetting" class="section level3 hasAnchor" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Subsetting<a href="datamanipulation.html#subsetting" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Subsetting</strong> data helps to draw the focus of analysis on particular
sets of samples and / or features. When dealing with large data
sets, the subset of interest can be extracted and investigated
separately. This might improve performance and reduce the
computational load.</p>
<p>Load:</p>
<ul>
<li>mia</li>
<li>dplyr</li>
<li>knitr</li>
<li>data <code>GlobalPatterns</code></li>
</ul>
<p>Let us store <code>GlobalPatterns</code> into <code>tse</code> and check its original number of features (rows) and samples (columns). <strong>Note</strong>: when subsetting by sample, expect the number of columns to decrease; when subsetting by feature, expect the number of rows to decrease.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="datamanipulation.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># store data into se and check dimensions</span></span>
<span id="cb83-2"><a href="datamanipulation.html#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb83-3"><a href="datamanipulation.html#cb83-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb83-4"><a href="datamanipulation.html#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions (features x samples)</span></span>
<span id="cb83-5"><a href="datamanipulation.html#cb83-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse) </span></code></pre></div>
<pre><code>## [1] 19216    26</code></pre>
<div id="subset-by-sample-column-wise" class="section level4 hasAnchor" number="4.1.2.1">
<h4><span class="header-section-number">4.1.2.1</span> Subset by sample (column-wise)<a href="datamanipulation.html#subset-by-sample-column-wise" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For the sake of demonstration, here we will extract a subset containing only the samples of human origin (feces, skin or tongue), stored as <code>SampleType</code> within <code>colData(tse)</code> and also in <code>tse</code>.</p>
<p>First, we would like to see all the possible values that <code>SampleType</code> can take on and how frequent those are:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="datamanipulation.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect possible values for SampleType</span></span>
<span id="cb85-2"><a href="datamanipulation.html#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(tse<span class="sc">$</span>SampleType)</span></code></pre></div>
<pre><code>## [1] Soil               Feces              Skin               Tongue            
## [5] Freshwater         Freshwater (creek) Ocean              Sediment (estuary)
## [9] Mock              
## 9 Levels: Feces Freshwater Freshwater (creek) Mock ... Tongue</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="datamanipulation.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show recurrence for each value</span></span>
<span id="cb87-2"><a href="datamanipulation.html#cb87-2" aria-hidden="true" tabindex="-1"></a>tse<span class="sc">$</span>SampleType <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
.
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Feces
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
Freshwater
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Freshwater (creek)
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Mock
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Ocean
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Sediment (estuary)
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Skin
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Soil
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Tongue
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table>
</div>
<p><strong>Note</strong>: after subsetting, expect the number of columns to equal the
sum of the recurrences of the samples that you are interested
in. For instance, <code>ncols = Feces + Skin + Tongue = 4 + 3 + 2 = 9</code>.</p>
<p>Next, we <em>logical index</em> across the columns of <code>tse</code> (make sure to
leave the first index empty to select all rows) and filter for the
samples of human origin. For this, we use the information on the
samples from the meta data <code>colData(tse)</code>.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="datamanipulation.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by sample</span></span>
<span id="cb88-2"><a href="datamanipulation.html#cb88-2" aria-hidden="true" tabindex="-1"></a>tse_subset_by_sample <span class="ot">&lt;-</span> tse[ , tse<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>)]</span>
<span id="cb88-3"><a href="datamanipulation.html#cb88-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-4"><a href="datamanipulation.html#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb88-5"><a href="datamanipulation.html#cb88-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_sample)</span></code></pre></div>
<pre><code>## [1] 19216     9</code></pre>
<p>As a sanity check, the new object <code>tse_subset_by_sample</code> should have
the original number of features (rows) and a number of samples
(columns) equal to the sum of the samples of interest (in this case
9).</p>
<p>Several characteristics can be used to subset by sample:</p>
<ul>
<li>origin</li>
<li>sampling time</li>
<li>sequencing method</li>
<li>DNA / RNA barcode</li>
<li>cohort</li>
</ul>
</div>
<div id="subset-by-feature-row-wise" class="section level4 hasAnchor" number="4.1.2.2">
<h4><span class="header-section-number">4.1.2.2</span> Subset by feature (row-wise)<a href="datamanipulation.html#subset-by-feature-row-wise" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Similarly, here we will extract a subset containing only the features
that belong to the Phyla “Actinobacteria” and “Chlamydiae”, stored as
<code>Phylum</code> within <code>rowData(tse)</code>. However, subsetting by feature implies
a few more obstacles, such as the presence of NA elements and the
possible need for agglomeration.</p>
<p>As previously, we would first like to see all the possible values that
<code>Phylum</code> can take on and how frequent those are:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="datamanipulation.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect possible values for Phylum</span></span>
<span id="cb90-2"><a href="datamanipulation.html#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum)</span></code></pre></div>
<pre><code>##  [1] &quot;Crenarchaeota&quot;    &quot;Euryarchaeota&quot;    &quot;Actinobacteria&quot;   &quot;Spirochaetes&quot;    
##  [5] &quot;MVP-15&quot;           &quot;Proteobacteria&quot;   &quot;SBR1093&quot;          &quot;Fusobacteria&quot;    
##  [9] &quot;Tenericutes&quot;      &quot;ZB3&quot;              &quot;Cyanobacteria&quot;    &quot;GOUTA4&quot;          
## [13] &quot;TG3&quot;              &quot;Chlorobi&quot;         &quot;Bacteroidetes&quot;    &quot;Caldithrix&quot;      
## [17] &quot;KSB1&quot;             &quot;SAR406&quot;           &quot;LCP-89&quot;           &quot;Thermi&quot;          
## [21] &quot;Gemmatimonadetes&quot; &quot;Fibrobacteres&quot;    &quot;GN06&quot;             &quot;AC1&quot;             
## [25] &quot;TM6&quot;              &quot;OP8&quot;              &quot;Elusimicrobia&quot;    &quot;NC10&quot;            
## [29] &quot;SPAM&quot;             NA                 &quot;Acidobacteria&quot;    &quot;CCM11b&quot;          
## [33] &quot;Nitrospirae&quot;      &quot;NKB19&quot;            &quot;BRC1&quot;             &quot;Hyd24-12&quot;        
## [37] &quot;WS3&quot;              &quot;PAUC34f&quot;          &quot;GN04&quot;             &quot;GN12&quot;            
## [41] &quot;Verrucomicrobia&quot;  &quot;Lentisphaerae&quot;    &quot;LD1&quot;              &quot;Chlamydiae&quot;      
## [45] &quot;OP3&quot;              &quot;Planctomycetes&quot;   &quot;Firmicutes&quot;       &quot;OP9&quot;             
## [49] &quot;WPS-2&quot;            &quot;Armatimonadetes&quot;  &quot;SC3&quot;              &quot;TM7&quot;             
## [53] &quot;GN02&quot;             &quot;SM2F11&quot;           &quot;ABY1_OD1&quot;         &quot;ZB2&quot;             
## [57] &quot;OP11&quot;             &quot;Chloroflexi&quot;      &quot;SC4&quot;              &quot;WS1&quot;             
## [61] &quot;GAL15&quot;            &quot;AD3&quot;              &quot;WS2&quot;              &quot;Caldiserica&quot;     
## [65] &quot;Thermotogae&quot;      &quot;Synergistetes&quot;    &quot;SR1&quot;</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="datamanipulation.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show recurrence for each value</span></span>
<span id="cb92-2"><a href="datamanipulation.html#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
.
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ABY1_OD1
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
AC1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Acidobacteria
</td>
<td style="text-align:right;">
1021
</td>
</tr>
<tr>
<td style="text-align:left;">
Actinobacteria
</td>
<td style="text-align:right;">
1631
</td>
</tr>
<tr>
<td style="text-align:left;">
AD3
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
Armatimonadetes
</td>
<td style="text-align:right;">
61
</td>
</tr>
<tr>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:right;">
2382
</td>
</tr>
<tr>
<td style="text-align:left;">
BRC1
</td>
<td style="text-align:right;">
13
</td>
</tr>
<tr>
<td style="text-align:left;">
Caldiserica
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Caldithrix
</td>
<td style="text-align:right;">
10
</td>
</tr>
<tr>
<td style="text-align:left;">
CCM11b
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Chlamydiae
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
Chlorobi
</td>
<td style="text-align:right;">
64
</td>
</tr>
<tr>
<td style="text-align:left;">
Chloroflexi
</td>
<td style="text-align:right;">
437
</td>
</tr>
<tr>
<td style="text-align:left;">
Crenarchaeota
</td>
<td style="text-align:right;">
106
</td>
</tr>
<tr>
<td style="text-align:left;">
Cyanobacteria
</td>
<td style="text-align:right;">
393
</td>
</tr>
<tr>
<td style="text-align:left;">
Elusimicrobia
</td>
<td style="text-align:right;">
31
</td>
</tr>
<tr>
<td style="text-align:left;">
Euryarchaeota
</td>
<td style="text-align:right;">
102
</td>
</tr>
<tr>
<td style="text-align:left;">
Fibrobacteres
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:right;">
4356
</td>
</tr>
<tr>
<td style="text-align:left;">
Fusobacteria
</td>
<td style="text-align:right;">
37
</td>
</tr>
<tr>
<td style="text-align:left;">
GAL15
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Gemmatimonadetes
</td>
<td style="text-align:right;">
191
</td>
</tr>
<tr>
<td style="text-align:left;">
GN02
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
GN04
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
GN06
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
GN12
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GOUTA4
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
Hyd24-12
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
KSB1
</td>
<td style="text-align:right;">
6
</td>
</tr>
<tr>
<td style="text-align:left;">
LCP-89
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
LD1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Lentisphaerae
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
MVP-15
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
NC10
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
Nitrospirae
</td>
<td style="text-align:right;">
74
</td>
</tr>
<tr>
<td style="text-align:left;">
NKB19
</td>
<td style="text-align:right;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
OP11
</td>
<td style="text-align:right;">
6
</td>
</tr>
<tr>
<td style="text-align:left;">
OP3
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
OP8
</td>
<td style="text-align:right;">
27
</td>
</tr>
<tr>
<td style="text-align:left;">
OP9
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
PAUC34f
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Planctomycetes
</td>
<td style="text-align:right;">
638
</td>
</tr>
<tr>
<td style="text-align:left;">
Proteobacteria
</td>
<td style="text-align:right;">
6416
</td>
</tr>
<tr>
<td style="text-align:left;">
SAR406
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
SBR1093
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
SC3
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
SC4
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
SM2F11
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
SPAM
</td>
<td style="text-align:right;">
22
</td>
</tr>
<tr>
<td style="text-align:left;">
Spirochaetes
</td>
<td style="text-align:right;">
124
</td>
</tr>
<tr>
<td style="text-align:left;">
SR1
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
Synergistetes
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
Tenericutes
</td>
<td style="text-align:right;">
143
</td>
</tr>
<tr>
<td style="text-align:left;">
TG3
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
Thermi
</td>
<td style="text-align:right;">
46
</td>
</tr>
<tr>
<td style="text-align:left;">
Thermotogae
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
TM6
</td>
<td style="text-align:right;">
27
</td>
</tr>
<tr>
<td style="text-align:left;">
TM7
</td>
<td style="text-align:right;">
32
</td>
</tr>
<tr>
<td style="text-align:left;">
Verrucomicrobia
</td>
<td style="text-align:right;">
470
</td>
</tr>
<tr>
<td style="text-align:left;">
WPS-2
</td>
<td style="text-align:right;">
20
</td>
</tr>
<tr>
<td style="text-align:left;">
WS1
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
WS2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
WS3
</td>
<td style="text-align:right;">
70
</td>
</tr>
<tr>
<td style="text-align:left;">
ZB2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
ZB3
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table>
</div>
<p><strong>Note</strong>: after subsetting, expect the number of columns to equal the
sum of the recurrences of the feature(s) that you are interested
in. For instance, <code>nrows = Actinobacteria + Chlamydiae = 1631 + 21 =   1652</code>.</p>
<p>Depending on your research question, you might need to or need not
agglomerate the data in the first place: if you want to find the
abundance of each and every feature that belongs to Actinobacteria and
Chlamydiae, agglomeration is not needed; if you want to find the total
abundance of all the features that belong to Actinobacteria or
Chlamydiae, agglomeration is recommended.</p>
<div id="non-agglomerated-data" class="section level5 hasAnchor" number="4.1.2.2.1">
<h5><span class="header-section-number">4.1.2.2.1</span> Non-agglomerated data<a href="datamanipulation.html#non-agglomerated-data" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Next, we <em>logical index</em> across the rows of <code>tse</code> (make sure to leave
the second index empty to select all columns) and filter for the
features that fall in either Actinobacteria or Chlamydiae. For this,
we use the information on the samples from the meta data
<code>rowData(tse)</code>.</p>
<p>The first term with the <code>%in%</code> operator are includes all the features
of interest, whereas the second term after the AND operator <code>&amp;</code>
filters out all the features that present a NA in place of Phylum.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="datamanipulation.html#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature</span></span>
<span id="cb93-2"><a href="datamanipulation.html#cb93-2" aria-hidden="true" tabindex="-1"></a>tse_subset_by_feature <span class="ot">&lt;-</span> tse[<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Actinobacteria&quot;</span>, <span class="st">&quot;Chlamydiae&quot;</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum), ]</span>
<span id="cb93-3"><a href="datamanipulation.html#cb93-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-4"><a href="datamanipulation.html#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb93-5"><a href="datamanipulation.html#cb93-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_feature)</span></code></pre></div>
<pre><code>## [1] 1652   26</code></pre>
<p>As a sanity check, the new object <code>tse_subset_by_feature</code> should have the original number of samples (columns) and a number of features (rows) equal to the sum of the features of interest (in this case 1652).</p>
</div>
<div id="agglomerated-data" class="section level5 hasAnchor" number="4.1.2.2.2">
<h5><span class="header-section-number">4.1.2.2.2</span> Agglomerated data<a href="datamanipulation.html#agglomerated-data" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>When total abundances of certain Phyla are of relevance, the data is initially agglomerated by Phylum. Then, similar steps as in the case of not agglomerated data are followed.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="datamanipulation.html#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># agglomerate by Phylum</span></span>
<span id="cb95-2"><a href="datamanipulation.html#cb95-2" aria-hidden="true" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> tse <span class="sc">%&gt;%</span> <span class="fu">agglomerateByRank</span>(<span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb95-3"><a href="datamanipulation.html#cb95-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-4"><a href="datamanipulation.html#cb95-4" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature and get rid of NAs</span></span>
<span id="cb95-5"><a href="datamanipulation.html#cb95-5" aria-hidden="true" tabindex="-1"></a>tse_phylum_subset_by_feature <span class="ot">&lt;-</span> tse_phylum[<span class="fu">rowData</span>(tse_phylum)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Actinobacteria&quot;</span>, <span class="st">&quot;Chlamydiae&quot;</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse_phylum)<span class="sc">$</span>Phylum), ]</span>
<span id="cb95-6"><a href="datamanipulation.html#cb95-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-7"><a href="datamanipulation.html#cb95-7" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb95-8"><a href="datamanipulation.html#cb95-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_phylum_subset_by_feature)</span></code></pre></div>
<pre><code>## [1]  2 26</code></pre>
<p><strong>Note</strong>: as data was agglomerated, the number of rows equal the
number of Phyla used to index (in this case, just 2)</p>
<p>Alternatively:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="datamanipulation.html#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co"># store features of interest into phyla</span></span>
<span id="cb97-2"><a href="datamanipulation.html#cb97-2" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Phylum:Actinobacteria&quot;</span>, <span class="st">&quot;Phylum:Chlamydiae&quot;</span>)</span>
<span id="cb97-3"><a href="datamanipulation.html#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature</span></span>
<span id="cb97-4"><a href="datamanipulation.html#cb97-4" aria-hidden="true" tabindex="-1"></a>tse_phylum_subset_by_feature <span class="ot">&lt;-</span> tse_phylum[phyla, ]</span>
<span id="cb97-5"><a href="datamanipulation.html#cb97-5" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb97-6"><a href="datamanipulation.html#cb97-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_feature)</span></code></pre></div>
<pre><code>## [1] 1652   26</code></pre>
<p>The code above returns the not agglomerated version of the data.</p>
<p>Fewer characteristics can be used to subset by feature:</p>
<ul>
<li>Taxonomic rank</li>
<li>Meta-taxonomic group</li>
</ul>
<p>For subsetting by Kingdom, agglomeration does not apply, whereas for
the other ranks it can be applied if necessary.</p>
</div>
</div>
<div id="subset-by-sample-and-feature" class="section level4 hasAnchor" number="4.1.2.3">
<h4><span class="header-section-number">4.1.2.3</span> Subset by sample and feature<a href="datamanipulation.html#subset-by-sample-and-feature" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Finally, we can subset data by sample and feature at once. The
resulting subset contains all the samples of human origin and all the
features of Phyla “Actinobacteria” or “Chlamydiae”.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="datamanipulation.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by sample and feature and get rid of NAs</span></span>
<span id="cb99-2"><a href="datamanipulation.html#cb99-2" aria-hidden="true" tabindex="-1"></a>tse_subset_by_sample_feature <span class="ot">&lt;-</span> tse[<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Actinobacteria&quot;</span>, <span class="st">&quot;Chlamydiae&quot;</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum), tse<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>)]</span>
<span id="cb99-3"><a href="datamanipulation.html#cb99-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-4"><a href="datamanipulation.html#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb99-5"><a href="datamanipulation.html#cb99-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_sample_feature)</span></code></pre></div>
<pre><code>## [1] 1652    9</code></pre>
<p><strong>Note</strong>: the dimensions of <code>tse_subset_by_sample_feature</code> agree with
those of the previous subsets (9 columns filtered by sample and 1652
rows filtered by feature).</p>
<p>If a study was to consider and quantify the presence of Actinobacteria
as well as Chlamydiae in different sites of the human body,
<code>tse_subset_by_sample_feature</code> might be a suitable subset to start
with.</p>
</div>
<div id="remove-empty-columns-and-rows" class="section level4 hasAnchor" number="4.1.2.4">
<h4><span class="header-section-number">4.1.2.4</span> Remove empty columns and rows<a href="datamanipulation.html#remove-empty-columns-and-rows" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Sometimes data might contain, e.g., features that are not present in any of the samples.
This might occur after subsetting for instance. In certain analyses we might want to
remove those instances.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="datamanipulation.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate data at Genus level </span></span>
<span id="cb101-2"><a href="datamanipulation.html#cb101-2" aria-hidden="true" tabindex="-1"></a>tse_genus <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb101-3"><a href="datamanipulation.html#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="co"># List bacteria that we want to include</span></span>
<span id="cb101-4"><a href="datamanipulation.html#cb101-4" aria-hidden="true" tabindex="-1"></a>genera <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Class:Thermoprotei&quot;</span>, <span class="st">&quot;Genus:Sulfolobus&quot;</span>, <span class="st">&quot;Genus:Sediminicola&quot;</span>)</span>
<span id="cb101-5"><a href="datamanipulation.html#cb101-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data</span></span>
<span id="cb101-6"><a href="datamanipulation.html#cb101-6" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus[genera, ]</span>
<span id="cb101-7"><a href="datamanipulation.html#cb101-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-8"><a href="datamanipulation.html#cb101-8" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 3 26 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(3): Class:Thermoprotei Genus:Sulfolobus Genus:Sediminicola
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (3 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="datamanipulation.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List total counts of each sample</span></span>
<span id="cb103-2"><a href="datamanipulation.html#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">&quot;counts&quot;</span>))</span></code></pre></div>
<pre><code>##      CL3      CC1      SV1  M31Fcsw  M11Fcsw  M31Plmr  M11Plmr  F21Plmr 
##        1        0        0        1        1        0        4        1 
##  M31Tong  M11Tong LMEpi24M SLEpi20M   AQC1cm   AQC4cm   AQC7cm      NP2 
##        7        3        0        2       64      105      136      222 
##      NP3      NP5  TRRsed1  TRRsed2  TRRsed3     TS28     TS29    Even1 
##     6433     1154        2        2        2        0        0        0 
##    Even2    Even3 
##        2        0</code></pre>
<p>Now we can see that certain samples do not include any bacteria. We can remove those.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="datamanipulation.html#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove samples that do not have present any bacteria</span></span>
<span id="cb105-2"><a href="datamanipulation.html#cb105-2" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus_sub[ , <span class="fu">colSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">&quot;counts&quot;</span>)) <span class="sc">!=</span> <span class="dv">0</span> ]</span>
<span id="cb105-3"><a href="datamanipulation.html#cb105-3" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 3 18 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(3): Class:Thermoprotei Genus:Sulfolobus Genus:Sediminicola
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(18): CL3 M31Fcsw ... TRRsed3 Even2
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (3 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Same thing can be done for features.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="datamanipulation.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take only those samples that are collected from feces, skin, or tongue</span></span>
<span id="cb107-2"><a href="datamanipulation.html#cb107-2" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus[ , <span class="fu">colData</span>(tse_genus)<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>)]</span>
<span id="cb107-3"><a href="datamanipulation.html#cb107-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-4"><a href="datamanipulation.html#cb107-4" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 1516 9 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(1516): Class:Thermoprotei Genus:Sulfolobus ...
##   Genus:Coprothermobacter Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(9): M31Fcsw M11Fcsw ... TS28 TS29
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (1516 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="datamanipulation.html#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many bacteria there are that are not present?</span></span>
<span id="cb109-2"><a href="datamanipulation.html#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">rowSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">&quot;counts&quot;</span>)) <span class="sc">==</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>## [1] 435</code></pre>
<p>We can see that there are bacteria that are not present in these samples that we chose.
We can remove those bacteria from the data.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="datamanipulation.html#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take only those bacteria that are present</span></span>
<span id="cb111-2"><a href="datamanipulation.html#cb111-2" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus_sub[<span class="fu">rowSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">&quot;counts&quot;</span>)) <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb111-3"><a href="datamanipulation.html#cb111-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-4"><a href="datamanipulation.html#cb111-4" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 1081 9 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(1081): Genus:Sulfolobus Order:NRP-J ...
##   Genus:Coprothermobacter Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(9): M31Fcsw M11Fcsw ... TS28 TS29
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (1081 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
</div>
</div>
<div id="splitting" class="section level3 hasAnchor" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Splitting<a href="datamanipulation.html#splitting" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can split data base on variables. There are available functions <code>splitByRanks</code>
and <code>splitOn</code>.</p>
<p><code>splitByRanks</code> splits the data based on rank. Since the elements of the output list
share columns, they can be stored into <code>altExp</code>.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="datamanipulation.html#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(tse) <span class="ot">&lt;-</span> <span class="fu">splitByRanks</span>(tse)</span>
<span id="cb113-2"><a href="datamanipulation.html#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(tse)</span></code></pre></div>
<pre><code>## List of length 7
## names(7): Kingdom Phylum Class Order Family Genus Species</code></pre>
<p>If you want to split the data based on other variable than taxonomy rank, use
<code>splitOn</code>. It works for row-wise and column-wise splitting.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="datamanipulation.html#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">splitOn</span>(tse, <span class="st">&quot;SampleType&quot;</span>)</span></code></pre></div>
<pre><code>## List of length 9
## names(9): Soil Feces Skin Tongue ... Ocean Sediment (estuary) Mock</code></pre>
</div>
</div>
<div id="merge-data" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> Merge data<a href="datamanipulation.html#merge-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>mia</code> package has <code>mergeSEs</code> function that merges multiple <code>SummarizedExperiment</code>
objects. For example, it is possible to combine multiple <code>TreeSE</code> objects which each
includes one sample.</p>
<p><code>mergeSEs</code> works like <code>dplyr</code> joining functions. In fact, there are available
<code>dplyr-like</code> aliases of <code>mergeSEs</code>, such as <code>full_join</code>.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="datamanipulation.html#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take subsets for demonstrative purpose</span></span>
<span id="cb117-2"><a href="datamanipulation.html#cb117-2" aria-hidden="true" tabindex="-1"></a>tse1 <span class="ot">&lt;-</span> tse[, <span class="dv">1</span>]</span>
<span id="cb117-3"><a href="datamanipulation.html#cb117-3" aria-hidden="true" tabindex="-1"></a>tse2 <span class="ot">&lt;-</span> tse[, <span class="dv">2</span>]</span>
<span id="cb117-4"><a href="datamanipulation.html#cb117-4" aria-hidden="true" tabindex="-1"></a>tse3 <span class="ot">&lt;-</span> tse[, <span class="dv">3</span>]</span>
<span id="cb117-5"><a href="datamanipulation.html#cb117-5" aria-hidden="true" tabindex="-1"></a>tse4 <span class="ot">&lt;-</span> tse[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="dv">4</span>]</span></code></pre></div>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="datamanipulation.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># With inner join, we want to include all shared rows. When using mergeSEs function</span></span>
<span id="cb118-2"><a href="datamanipulation.html#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all the samples are always preserved.</span></span>
<span id="cb118-3"><a href="datamanipulation.html#cb118-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">mergeSEs</span>(<span class="fu">list</span>(tse1, tse2, tse3, tse4), <span class="at">join =</span> <span class="st">&quot;inner&quot;</span>)</span>
<span id="cb118-4"><a href="datamanipulation.html#cb118-4" aria-hidden="true" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 100 4 
## metadata(0):
## assays(1): counts
## rownames(100): 239672 243675 ... 549322 951
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(4): CC1 CL3 M31Fcsw SV1
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (100 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="datamanipulation.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># left join preserves all the rows of the 1st object</span></span>
<span id="cb120-2"><a href="datamanipulation.html#cb120-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">left_join</span>(tse1, tse4, <span class="at">missing_values =</span> <span class="dv">0</span>)</span>
<span id="cb120-3"><a href="datamanipulation.html#cb120-3" aria-hidden="true" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 2 
## metadata(0):
## assays(1): counts
## rownames(19216): 239672 243675 ... 146168 594324
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(2): CL3 M31Fcsw
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div id="additional-functions" class="section level3 hasAnchor" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Additional functions<a href="datamanipulation.html#additional-functions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><a href="https://microbiome.github.io/mia/reference/taxonomy-methods.html">mapTaxonomy</a></li>
<li><a href="https://microbiome.github.io/mia/reference/merge-methods.html">mergeRows/mergeCols</a></li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="packages.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="quality-control.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
