<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Data operations | Microbiome Analysis</title>
  <meta name="description" content="Chapter 2 Data operations | Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Data operations | Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Data operations | Microbiome Analysis" />
  
  
  

<meta name="author" content="" />


<meta name="date" content="2020-11-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro.html"/>
<link rel="next" href="bibliography.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0/anchor-sections.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#package-data"><i class="fa fa-check"></i><b>1.2</b> Package data</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#tools"><i class="fa fa-check"></i><b>1.3</b> Tools</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Data-operations.html"><a href="Data-operations.html"><i class="fa fa-check"></i><b>2</b> Data operations</a><ul>
<li class="chapter" data-level="2.1" data-path="Data-operations.html"><a href="Data-operations.html#common-terminologies"><i class="fa fa-check"></i><b>2.1</b> Common terminologies</a></li>
<li class="chapter" data-level="2.2" data-path="Data-operations.html"><a href="Data-operations.html#retrieving-data-elements"><i class="fa fa-check"></i><b>2.2</b> Retrieving data elements</a></li>
<li class="chapter" data-level="2.3" data-path="Data-operations.html"><a href="Data-operations.html#extract-specific-elements"><i class="fa fa-check"></i><b>2.3</b> Extract specific elements</a><ul>
<li class="chapter" data-level="2.3.1" data-path="Data-operations.html"><a href="Data-operations.html#assays"><i class="fa fa-check"></i><b>2.3.1</b> Assays</a></li>
<li class="chapter" data-level="2.3.2" data-path="Data-operations.html"><a href="Data-operations.html#coldata"><i class="fa fa-check"></i><b>2.3.2</b> colData</a></li>
<li class="chapter" data-level="2.3.3" data-path="Data-operations.html"><a href="Data-operations.html#rowdata"><i class="fa fa-check"></i><b>2.3.3</b> rowData</a></li>
<li class="chapter" data-level="2.3.4" data-path="Data-operations.html"><a href="Data-operations.html#rowtree"><i class="fa fa-check"></i><b>2.3.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="Data-operations.html"><a href="Data-operations.html#library-size"><i class="fa fa-check"></i><b>2.4</b> Library size</a></li>
<li class="chapter" data-level="2.5" data-path="Data-operations.html"><a href="Data-operations.html#agglomerate"><i class="fa fa-check"></i><b>2.5</b> Agglomerate</a></li>
<li class="chapter" data-level="2.6" data-path="Data-operations.html"><a href="Data-operations.html#get-unique"><i class="fa fa-check"></i><b>2.6</b> Get unique</a></li>
<li class="chapter" data-level="2.7" data-path="Data-operations.html"><a href="Data-operations.html#pick-specific"><i class="fa fa-check"></i><b>2.7</b> Pick specific</a><ul>
<li class="chapter" data-level="2.7.1" data-path="Data-operations.html"><a href="Data-operations.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>2.7.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="2.7.2" data-path="Data-operations.html"><a href="Data-operations.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>2.7.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="Data-operations.html"><a href="Data-operations.html#join"><i class="fa fa-check"></i><b>2.8</b> Join</a><ul>
<li class="chapter" data-level="2.8.1" data-path="Data-operations.html"><a href="Data-operations.html#convert-to-long-data.frame"><i class="fa fa-check"></i><b>2.8.1</b> Convert to long data.frame</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="Data-operations.html"><a href="Data-operations.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>2.9</b> Get top taxa and taxonomy</a><ul>
<li class="chapter" data-level="2.9.1" data-path="Data-operations.html"><a href="Data-operations.html#features"><i class="fa fa-check"></i><b>2.9.1</b> Features</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="Data-operations" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Data operations</h1>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html"><code>SummarizedExperiment</code></a> (<code>SE</code>) is a widely used class for analyzing data obtained by common sequencing techniques. <code>SE</code> is common structure for several Bioconductor packages that are used for analyzing RNAseq, ChIp-Seq data. <code>SE</code> class is also used in R packages for analyzing microarrays, flow cytometry, proteomics, single-cell sequencing data and many more. The single-cell analysis is facilitated by <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingelCellExperiment class</a> (<code>SCE</code>), which allows the user to store results of dimensionality reduction and alternative experiments. Alternative experiments (<code>altExps</code>) can be differently processed data within the analysis workflows.</p>
<p>Recently, <a href="https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html"><code>TreeSummarizedExperiment</code></a> (<code>TSE</code>), <a href="https://github.com/FelixErnst/MicrobiomeExperiment"><code>MicrobiomeExperiment</code></a> were developed to extend the <code>SE</code> and <code>SCE</code> class for incorporating hierarchical information (including phylogenetic tree) and reference sequences.</p>
<p>The <code>mia</code> package implements tools using these classes for analysis of microbiome sequencing data.</p>
<div id="common-terminologies" class="section level2">
<h2><span class="header-section-number">2.1</span> Common terminologies</h2>
<p><code>assays</code> : This slot is similar to <code>otu_table</code> in <code>phyloseq</code>. In a <code>SummarizedExperiment</code>
object multiple assays, raw counts, transformed counts can be stored. See also
<a href="https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html"><code>MultiAssayExperiment</code></a> for storing data from multiple <code>experiments</code> such as RNASeq, Proteomics, etc.<br />
<code>rowData</code> : This slot is similar to <code>tax_table</code> in <code>phyloseq</code> to store taxonomic information.<br />
<code>colData</code> : This slot is similar to <code>sample_data</code> in <code>phyloseq</code> to store information related to samples.<br />
<code>rowTree</code> : This slot is similar to <code>phy_tree</code> in <code>phyloseq</code> to store phylogenetic tree.</p>
<p>In this book, you will come across terms like <code>FeatureIDs</code> and <code>SampleIDs</code>.<br />
<code>FeatureIDs</code> : These are basically OTU/ASV ids which are row names in <code>assays</code> and <code>rowData</code>.<br />
<code>SampleIDs</code> : As the name suggests, these are sample ids which are column names in <code>assays</code> and row names in <code>colData</code>.</p>
</div>
<div id="retrieving-data-elements" class="section level2">
<h2><span class="header-section-number">2.2</span> Retrieving data elements</h2>
<p>Extract specific elements from <code>TreeSummarizedExperiment</code> and <code>MicrobiomeExperiment</code> object.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="Data-operations.html#cb35-1"></a><span class="kw">library</span>(mia)</span>
<span id="cb35-2"><a href="Data-operations.html#cb35-2"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb35-3"><a href="Data-operations.html#cb35-3"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb35-4"><a href="Data-operations.html#cb35-4"></a><span class="kw">library</span>(tibble)</span></code></pre></div>
<p>Example data</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="Data-operations.html#cb36-1"></a><span class="kw">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>)</span>
<span id="cb36-2"><a href="Data-operations.html#cb36-2"></a>me &lt;-<span class="st"> </span>GlobalPatterns </span></code></pre></div>
<p>Create a relative abundance assay</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="Data-operations.html#cb37-1"></a>me &lt;-<span class="st"> </span><span class="kw">relAbundanceCounts</span>(GlobalPatterns)</span>
<span id="cb37-2"><a href="Data-operations.html#cb37-2"></a>me</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(2): counts relabundance
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: a phylo (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Above there are now two <code>assays</code><br />
1] counts<br />
2] relabundance</p>
</div>
<div id="extract-specific-elements" class="section level2">
<h2><span class="header-section-number">2.3</span> Extract specific elements</h2>
<div id="assays" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Assays</h3>
<p><code>Assay</code> slots hold the counts and transformed data for <code>FeatureIDs</code> and these can be extracted as follows:<br />
Extract count abundance data.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="Data-operations.html#cb39-1"></a>me_counts &lt;-<span class="st"> </span><span class="kw">assay</span>(me, <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb39-2"><a href="Data-operations.html#cb39-2"></a>me_counts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##        CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## 549322   0   0   0       0       0       0       0
## 522457   0   0   0       0       0       0       0
## 951      0   0   0       0       0       0       1
## 244423   0   0   0       0       0       0       0
## 586076   0   0   0       0       0       0       0</code></pre>
<p>Extract relative abundance data.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="Data-operations.html#cb41-1"></a>me_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me, <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb41-2"><a href="Data-operations.html#cb41-2"></a>me_reab[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##        CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr     M11Plmr
## 549322   0   0   0       0       0       0 0.00000e+00
## 522457   0   0   0       0       0       0 0.00000e+00
## 951      0   0   0       0       0       0 2.30471e-06
## 244423   0   0   0       0       0       0 0.00000e+00
## 586076   0   0   0       0       0       0 0.00000e+00</code></pre>
</div>
<div id="coldata" class="section level3">
<h3><span class="header-section-number">2.3.2</span> colData</h3>
<p>The sample information can be retrieved by accessing the <code>colData</code> as a data.frame.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="Data-operations.html#cb43-1"></a>me_sample_data &lt;-<span class="st"> </span><span class="kw">colData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb43-2"><a href="Data-operations.html#cb43-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>()</span>
<span id="cb43-3"><a href="Data-operations.html#cb43-3"></a><span class="co"># show select row and columns</span></span>
<span id="cb43-4"><a href="Data-operations.html#cb43-4"></a>me_sample_data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##         X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T
## CL3            CL3 ILBC_01        AACGCA                   TGCGTT
## CC1            CC1 ILBC_02        AACTCG                   CGAGTT
## SV1            SV1 ILBC_03        AACTGT                   ACAGTT
## M31Fcsw    M31Fcsw ILBC_04        AAGAGA                   TCTCTT
## M11Fcsw    M11Fcsw ILBC_05        AAGCTG                   CAGCTT
##         Barcode_full_length
## CL3             CTAGCGTGCGT
## CC1             CATCGACGAGT
## SV1             GTACGCACAGT
## M31Fcsw         TCGACATCTCT
## M11Fcsw         CGACTGCAGCT</code></pre>
</div>
<div id="rowdata" class="section level3">
<h3><span class="header-section-number">2.3.3</span> rowData</h3>
<p>The taxonomic classification of <code>FeatureIds</code> can be retrieved by accessing the <code>rowData</code> as a data.frame.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="Data-operations.html#cb45-1"></a>me_row_data &lt;-<span class="st"> </span><span class="kw">rowData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb45-2"><a href="Data-operations.html#cb45-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>()</span>
<span id="cb45-3"><a href="Data-operations.html#cb45-3"></a><span class="co"># show select row and columns</span></span>
<span id="cb45-4"><a href="Data-operations.html#cb45-4"></a>me_row_data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##        Kingdom        Phylum        Class        Order        Family
## 549322 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 522457 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 951    Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 244423 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;
## 586076 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;</code></pre>
</div>
<div id="rowtree" class="section level3">
<h3><span class="header-section-number">2.3.4</span> rowTree</h3>
<p>The phylogenetic tree for <code>FeatureIds</code> can be retrieved by accessing the <code>rowTree</code> which is a <code>phylo</code> object.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="Data-operations.html#cb47-1"></a>me_row_tree &lt;-<span class="st"> </span><span class="kw">rowTree</span>(me) </span>
<span id="cb47-2"><a href="Data-operations.html#cb47-2"></a>me_row_tree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 19216 tips and 19215 internal nodes.
## 
## Tip labels:
##   549322, 522457, 951, 244423, 586076, 246140, ...
## Node labels:
##   , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>Check link between rownames <code>FeatureIds</code>.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="Data-operations.html#cb49-1"></a><span class="co"># extract the linkData</span></span>
<span id="cb49-2"><a href="Data-operations.html#cb49-2"></a><span class="co"># on rows</span></span>
<span id="cb49-3"><a href="Data-operations.html#cb49-3"></a>rowL &lt;-<span class="st"> </span><span class="kw">rowLinks</span>(<span class="dt">x =</span> me)</span></code></pre></div>
</div>
</div>
<div id="library-size" class="section level2">
<h2><span class="header-section-number">2.4</span> Library size</h2>
<p>The total counts/sample can be calculated using the <code>colSums</code>.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="Data-operations.html#cb50-1"></a>reads_sample &lt;-<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">assay</span>(me, <span class="st">&quot;counts&quot;</span>)) </span>
<span id="cb50-2"><a href="Data-operations.html#cb50-2"></a><span class="co"># add it to colData</span></span>
<span id="cb50-3"><a href="Data-operations.html#cb50-3"></a><span class="kw">colData</span>(me)<span class="op">$</span>reads_sample &lt;-<span class="st"> </span>reads_sample</span></code></pre></div>
<p>Visualize library size</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="Data-operations.html#cb51-1"></a>me_sample_data &lt;-<span class="st"> </span><span class="kw">colData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb51-2"><a href="Data-operations.html#cb51-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb51-3"><a href="Data-operations.html#cb51-3"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;SampleId&quot;</span>)</span>
<span id="cb51-4"><a href="Data-operations.html#cb51-4"></a></span>
<span id="cb51-5"><a href="Data-operations.html#cb51-5"></a><span class="kw">ggplot</span>(me_sample_data, <span class="kw">aes</span>(reads_sample, SampleId)) <span class="op">+</span></span>
<span id="cb51-6"><a href="Data-operations.html#cb51-6"></a><span class="st">    </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>SampleType)) <span class="op">+</span></span>
<span id="cb51-7"><a href="Data-operations.html#cb51-7"></a><span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span><span class="st"> </span><span class="co"># prettify plot</span></span>
<span id="cb51-8"><a href="Data-operations.html#cb51-8"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">10</span>), </span>
<span id="cb51-9"><a href="Data-operations.html#cb51-9"></a>          <span class="dt">axis.line.x =</span> <span class="kw">element_line</span>(<span class="dt">color =</span> <span class="st">&quot;#242424&quot;</span>), </span>
<span id="cb51-10"><a href="Data-operations.html#cb51-10"></a>          <span class="dt">axis.line.y =</span> <span class="kw">element_line</span>(<span class="dt">color =</span> <span class="st">&quot;#242424&quot;</span>), </span>
<span id="cb51-11"><a href="Data-operations.html#cb51-11"></a>          <span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(), </span>
<span id="cb51-12"><a href="Data-operations.html#cb51-12"></a>          <span class="dt">panel.border =</span> <span class="kw">element_rect</span>(<span class="dt">colour =</span> <span class="st">&quot;#242424&quot;</span>), </span>
<span id="cb51-13"><a href="Data-operations.html#cb51-13"></a>          <span class="dt">legend.key =</span> <span class="kw">element_blank</span>(), <span class="dt">legend.text =</span> <span class="kw">element_text</span>(<span class="dt">color =</span> <span class="st">&quot;#353535&quot;</span>), </span>
<span id="cb51-14"><a href="Data-operations.html#cb51-14"></a>          <span class="dt">legend.background =</span> <span class="kw">element_rect</span>(<span class="dt">colour =</span> <span class="ot">NA</span>, <span class="dt">fill =</span> <span class="ot">NA</span>))</span></code></pre></div>
<p><img src="02-data-operations_files/figure-html/viz-lib-size-1.png" width="768" /></p>
</div>
<div id="agglomerate" class="section level2">
<h2><span class="header-section-number">2.5</span> Agglomerate</h2>
<p>Agglomerate at a specific taxonomic rank. If multiple assays (counts and relabundance) exists, both will be agglomerated.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="Data-operations.html#cb52-1"></a>me_fam &lt;-<span class="st"> </span><span class="kw">agglomerateByRank</span>(me, <span class="dt">rank =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb52-2"><a href="Data-operations.html#cb52-2"></a>                        <span class="dt">agglomerateTree =</span> <span class="ot">TRUE</span>)</span>
<span id="cb52-3"><a href="Data-operations.html#cb52-3"></a></span>
<span id="cb52-4"><a href="Data-operations.html#cb52-4"></a><span class="kw">print</span>(me_fam)</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 603 26 
## metadata(0):
## assays(2): counts relabundance
## rownames(603): Class::Thermoprotei Family:Sulfolobaceae ...
##   Family:Thermodesulfobiaceae Phylum::SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(8): X.SampleID Primer ... Description reads_sample
## reducedDimNames(0):
## altExpNames(0):
## rowLinks: a LinkDataFrame (603 rows)
## rowTree: a phylo (603 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="Data-operations.html#cb54-1"></a>me_fam_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me_fam, <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb54-2"><a href="Data-operations.html#cb54-2"></a>me_fam_reab[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                               CL3          CC1 SV1 M31Fcsw M11Fcsw M31Plmr
## Class::Thermoprotei  0.0000000000 0.000000e+00   0       0       0       0
## Family:Sulfolobaceae 0.0000000000 0.000000e+00   0       0       0       0
## Class::Sd-NA         0.0000000000 0.000000e+00   0       0       0       0
## Order::NRP-J         0.0001990563 2.069651e-04   0       0       0       0
## Family:SAGMA-X       0.0000000000 6.164919e-06   0       0       0       0
##                           M11Plmr
## Class::Thermoprotei  0.000000e+00
## Family:Sulfolobaceae 2.304710e-06
## Class::Sd-NA         0.000000e+00
## Order::NRP-J         6.914131e-06
## Family:SAGMA-X       0.000000e+00</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="Data-operations.html#cb56-1"></a>me_fam_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me_fam, <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb56-2"><a href="Data-operations.html#cb56-2"></a>me_fam_reab[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                      CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## Class::Thermoprotei    0   0   0       0       0       0       0
## Family:Sulfolobaceae   0   0   0       0       0       0       1
## Class::Sd-NA           0   0   0       0       0       0       0
## Order::NRP-J         172 235   0       0       0       0       3
## Family:SAGMA-X         0   7   0       0       0       0       0</code></pre>
<p>These newly created data can also be kept in <code>altExp</code> a feature provided by <code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="Data-operations.html#cb58-1"></a><span class="kw">altExp</span>(me, <span class="st">&quot;family&quot;</span>) &lt;-<span class="st"> </span>me_fam</span></code></pre></div>
<p><code>altExpNames</code> now consists of <code>family</code> level data. This can be extended to use any level present in Kingdom, Phylum, Class, Order, Family, Genus, Species.</p>
</div>
<div id="get-unique" class="section level2">
<h2><span class="header-section-number">2.6</span> Get unique</h2>
<p>Get which Phyla are present.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="Data-operations.html#cb59-1"></a><span class="kw">head</span>(<span class="kw">unique</span>(<span class="kw">rowData</span>(me)[,<span class="st">&quot;Phylum&quot;</span>]))</span></code></pre></div>
<pre><code>## [1] &quot;Crenarchaeota&quot;  &quot;Euryarchaeota&quot;  &quot;Actinobacteria&quot; &quot;Spirochaetes&quot;  
## [5] &quot;MVP-15&quot;         &quot;Proteobacteria&quot;</code></pre>
</div>
<div id="pick-specific" class="section level2">
<h2><span class="header-section-number">2.7</span> Pick specific</h2>
<p>Retrieving of specific elements are required for specific analysis. For instance, extracting abundances for a specific taxa in all samples or all taxa in one sample.</p>
<div id="abundances-of-all-taxa-in-specific-sample" class="section level3">
<h3><span class="header-section-number">2.7.1</span> Abundances of all taxa in specific sample</h3>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="Data-operations.html#cb61-1"></a>taxa.abund.cc1 &lt;-<span class="st"> </span><span class="kw">getAbundanceSample</span>(me, </span>
<span id="cb61-2"><a href="Data-operations.html#cb61-2"></a>                                     <span class="dt">sample_id =</span> <span class="st">&quot;CC1&quot;</span>,</span>
<span id="cb61-3"><a href="Data-operations.html#cb61-3"></a>                                     <span class="dt">abund_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb61-4"><a href="Data-operations.html#cb61-4"></a></span>
<span id="cb61-5"><a href="Data-operations.html#cb61-5"></a>taxa.abund.cc1[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 549322 522457    951 244423 586076 246140 143239 244960 255340 144887 
##      0      0      0      0      0      0      1      0    194      5</code></pre>
</div>
<div id="abundances-of-specific-taxa-in-all-samples" class="section level3">
<h3><span class="header-section-number">2.7.2</span> Abundances of specific taxa in all samples</h3>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="Data-operations.html#cb63-1"></a>taxa.abundances &lt;-<span class="st"> </span><span class="kw">getAbundanceFeature</span>(me, </span>
<span id="cb63-2"><a href="Data-operations.html#cb63-2"></a>                                      <span class="dt">feature_id =</span> <span class="st">&quot;255340&quot;</span>,</span>
<span id="cb63-3"><a href="Data-operations.html#cb63-3"></a>                                       <span class="dt">abund_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb63-4"><a href="Data-operations.html#cb63-4"></a>taxa.abundances[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong 
##     153     194       0       0       0       0       0       0       0       0</code></pre>
</div>
</div>
<div id="join" class="section level2">
<h2><span class="header-section-number">2.8</span> Join</h2>
<p>Joining <code>assay</code> and <code>rowData</code> in one data frame for exploring <code>featureIds</code> and their taxonomic identities.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="Data-operations.html#cb65-1"></a>me_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me, <span class="st">&quot;relabundance&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb65-2"><a href="Data-operations.html#cb65-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb65-3"><a href="Data-operations.html#cb65-3"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;FeatureID&quot;</span>)</span>
<span id="cb65-4"><a href="Data-operations.html#cb65-4"></a></span>
<span id="cb65-5"><a href="Data-operations.html#cb65-5"></a>me_joined &lt;-<span class="st"> </span><span class="kw">rowData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb65-6"><a href="Data-operations.html#cb65-6"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb65-7"><a href="Data-operations.html#cb65-7"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;FeatureID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb65-8"><a href="Data-operations.html#cb65-8"></a><span class="st">    </span><span class="kw">left_join</span>(me_reab) </span></code></pre></div>
<pre><code>## Joining, by = &quot;FeatureID&quot;</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="Data-operations.html#cb67-1"></a>me_joined[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>]</span></code></pre></div>
<pre><code>##   FeatureID Kingdom        Phylum        Class        Order        Family
## 1    549322 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 2    522457 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 3       951 Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
##        Genus                  Species
## 1       &lt;NA&gt;                     &lt;NA&gt;
## 2       &lt;NA&gt;                     &lt;NA&gt;
## 3 Sulfolobus Sulfolobusacidocaldarius</code></pre>
<div id="convert-to-long-data.frame" class="section level3">
<h3><span class="header-section-number">2.8.1</span> Convert to long data.frame</h3>
<p>For several custom analysis and visualization packages such as those from the <code>tidyverse</code> the <code>SE</code> data can be converted to long data.frame format with <code>meltAssay</code>.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="Data-operations.html#cb69-1"></a>molten_se &lt;-<span class="st"> </span><span class="kw">meltAssay</span>(me,</span>
<span id="cb69-2"><a href="Data-operations.html#cb69-2"></a>                       <span class="dt">add_row_data =</span> <span class="ot">TRUE</span>,</span>
<span id="cb69-3"><a href="Data-operations.html#cb69-3"></a>                       <span class="dt">add_col_data =</span> <span class="ot">TRUE</span>,</span>
<span id="cb69-4"><a href="Data-operations.html#cb69-4"></a>                       <span class="dt">abund_values =</span> <span class="st">&quot;relabundance&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="get-top-taxa-and-taxonomy" class="section level2">
<h2><span class="header-section-number">2.9</span> Get top taxa and taxonomy</h2>
<div id="features" class="section level3">
<h3><span class="header-section-number">2.9.1</span> Features</h3>
<p>The <code>getTopTaxa</code> can be used for identifying top taxa in the data.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="Data-operations.html#cb70-1"></a>top_features &lt;-<span class="st"> </span><span class="kw">getTopTaxa</span>(me, <span class="dt">method=</span><span class="st">&quot;median&quot;</span>, <span class="dt">top=</span><span class="dv">10</span>)</span>
<span id="cb70-2"><a href="Data-operations.html#cb70-2"></a>tax_data &lt;-<span class="st"> </span><span class="kw">rowData</span>(me)[top_features,<span class="kw">taxonomyRanks</span>(me)]</span>
<span id="cb70-3"><a href="Data-operations.html#cb70-3"></a>tax_data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>()</span></code></pre></div>
<pre><code>##         Kingdom         Phylum               Class             Order
## 549656 Bacteria  Cyanobacteria         Chloroplast     Stramenopiles
## 331820 Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
## 317182 Bacteria  Cyanobacteria         Chloroplast     Stramenopiles
## 94166  Bacteria Proteobacteria Gammaproteobacteria    Pasteurellales
## 279599 Bacteria  Cyanobacteria    Nostocophycideae        Nostocales
## 158660 Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
## 329744 Bacteria Actinobacteria      Actinobacteria   Actinomycetales
## 326977 Bacteria Actinobacteria      Actinobacteria Bifidobacteriales
## 248140 Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
## 550960 Bacteria Proteobacteria Gammaproteobacteria Enterobacteriales
##                    Family           Genus                     Species
## 549656               &lt;NA&gt;            &lt;NA&gt;                        &lt;NA&gt;
## 331820     Bacteroidaceae     Bacteroides                        &lt;NA&gt;
## 317182               &lt;NA&gt;            &lt;NA&gt;                        &lt;NA&gt;
## 94166     Pasteurellaceae     Haemophilus   Haemophilusparainfluenzae
## 279599        Nostocaceae  Dolichospermum                        &lt;NA&gt;
## 158660     Bacteroidaceae     Bacteroides                        &lt;NA&gt;
## 329744             ACK-M1            &lt;NA&gt;                        &lt;NA&gt;
## 326977 Bifidobacteriaceae Bifidobacterium Bifidobacteriumadolescentis
## 248140     Bacteroidaceae     Bacteroides           Bacteroidescaccae
## 550960 Enterobacteriaceae     Providencia                        &lt;NA&gt;</code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bibliography.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MiaBook.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
