<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Data Infrastructure | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 2 Data Infrastructure | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Data Infrastructure | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Data Infrastructure | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="example-data.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="" data-path="introduction-to-bioconductor-classes.html"><a href="introduction-to-bioconductor-classes.html"><i class="fa fa-check"></i>Introduction to Bioconductor Classes</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#community"><i class="fa fa-check"></i><b>1.1</b> Community</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#developers-and-contributors"><i class="fa fa-check"></i><b>1.2</b> Developers and contributors</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.3</b> Acknowledgements</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#package-ecosystem"><i class="fa fa-check"></i><b>1.4</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="introduction.html"><a href="introduction.html#data-container"><i class="fa fa-check"></i><b>1.4.1</b> Data container:</a></li>
<li class="chapter" data-level="1.4.2" data-path="introduction.html"><a href="introduction.html#packages"><i class="fa fa-check"></i><b>1.4.2</b> Packages</a></li>
<li class="chapter" data-level="1.4.3" data-path="introduction.html"><a href="introduction.html#demonstration-data"><i class="fa fa-check"></i><b>1.4.3</b> Demonstration data</a></li>
<li class="chapter" data-level="1.4.4" data-path="introduction.html"><a href="introduction.html#online-tutorials"><i class="fa fa-check"></i><b>1.4.4</b> Online tutorials</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-introduction.html"><a href="data-introduction.html"><i class="fa fa-check"></i><b>2</b> Data Infrastructure</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-introduction.html"><a href="data-introduction.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a></li>
<li class="chapter" data-level="2.2" data-path="data-introduction.html"><a href="data-introduction.html#background"><i class="fa fa-check"></i><b>2.2</b> Background</a></li>
<li class="chapter" data-level="2.3" data-path="data-introduction.html"><a href="data-introduction.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="data-introduction.html"><a href="data-introduction.html#importing-data-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Importing data from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="data-introduction.html"><a href="data-introduction.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-introduction.html"><a href="data-introduction.html#metadata"><i class="fa fa-check"></i><b>2.4</b> Metadata</a></li>
<li class="chapter" data-level="2.5" data-path="data-introduction.html"><a href="data-introduction.html#microbiome-and-tree-data-specific-aspects"><i class="fa fa-check"></i><b>2.5</b> Microbiome and tree data specific aspects</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="data-introduction.html"><a href="data-introduction.html#assays"><i class="fa fa-check"></i><b>2.5.1</b> Assays</a></li>
<li class="chapter" data-level="2.5.2" data-path="data-introduction.html"><a href="data-introduction.html#coldata"><i class="fa fa-check"></i><b>2.5.2</b> colData</a></li>
<li class="chapter" data-level="2.5.3" data-path="data-introduction.html"><a href="data-introduction.html#rowdata"><i class="fa fa-check"></i><b>2.5.3</b> rowData</a></li>
<li class="chapter" data-level="2.5.4" data-path="data-introduction.html"><a href="data-introduction.html#rowtree"><i class="fa fa-check"></i><b>2.5.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="data-introduction.html"><a href="data-introduction.html#data-manipulation"><i class="fa fa-check"></i><b>2.6</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="data-introduction.html"><a href="data-introduction.html#tidy-data"><i class="fa fa-check"></i><b>2.6.1</b> Tidy data</a></li>
<li class="chapter" data-level="2.6.2" data-path="data-introduction.html"><a href="data-introduction.html#subsetting"><i class="fa fa-check"></i><b>2.6.2</b> Subsetting</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="data-introduction.html"><a href="data-introduction.html#additional-reading"><i class="fa fa-check"></i><b>2.7</b> Additional Reading</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="data-introduction.html"><a href="data-introduction.html#r-programming-resources"><i class="fa fa-check"></i><b>2.7.1</b> R programming resources</a></li>
<li class="chapter" data-level="2.7.2" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>2.7.2</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="2.7.3" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>2.7.3</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="2.7.4" data-path="data-introduction.html"><a href="data-introduction.html#further-reading"><i class="fa fa-check"></i><b>2.7.4</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-introduction.html"><a href="data-introduction.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Appendix</b></span></li>
<li class="chapter" data-level="3" data-path="example-data.html"><a href="example-data.html"><i class="fa fa-check"></i><b>3</b> Demonstration Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="example-data.html"><a href="example-data.html#package-data"><i class="fa fa-check"></i><b>3.1</b> Package data</a></li>
<li class="chapter" data-level="3.2" data-path="example-data.html"><a href="example-data.html#experimenthub-data"><i class="fa fa-check"></i><b>3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3" data-path="example-data.html"><a href="example-data.html#other-data-sources"><i class="fa fa-check"></i><b>3.3</b> Other data sources</a></li>
<li class="chapter" data-level="" data-path="example-data.html"><a href="example-data.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>4</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="4.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>4.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>4.1.1</b> dada2</a></li>
<li class="chapter" data-level="4.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>4.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>4.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>4.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>4.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="4.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>4.4</b> Data transformation</a></li>
<li class="chapter" data-level="4.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>4.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>4.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="4.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>4.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>5.1</b> Get top taxa and taxonomy</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="quality-control.html"><a href="quality-control.html#features"><i class="fa fa-check"></i><b>5.1.1</b> Features</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.2</b> Library size</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html"><i class="fa fa-check"></i><b>6</b> Microbiome Exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalence"><i class="fa fa-check"></i><b>6.1</b> Prevalence</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>6.1.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="6.1.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#plotting-prevalence"><i class="fa fa-check"></i><b>6.1.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#abundance"><i class="fa fa-check"></i><b>6.2</b> Abundance</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>7</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#estimating-alpha-diversity"><i class="fa fa-check"></i><b>7.1</b> Estimating alpha diversity</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.3</b> Evenness</a></li>
<li class="chapter" data-level="7.1.4" data-path="alpha-diversity.html"><a href="alpha-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.4</b> Dominance</a></li>
<li class="chapter" data-level="7.1.5" data-path="alpha-diversity.html"><a href="alpha-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.5</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualize-alpha-diversities"><i class="fa fa-check"></i><b>7.2</b> Visualize alpha diversities</a></li>
<li class="chapter" data-level="" data-path="alpha-diversity.html"><a href="alpha-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>8</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="beta-diversity.html"><a href="beta-diversity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-beta-diversity"><i class="fa fa-check"></i><b>8.2</b> Estimating beta diversity</a></li>
<li class="chapter" data-level="8.3" data-path="beta-diversity.html"><a href="beta-diversity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="beta-diversity.html"><a href="beta-diversity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a></li>
<li class="chapter" data-level="8.5" data-path="beta-diversity.html"><a href="beta-diversity.html#community-comparisons-todo-combine-with-the-material-above-for-simplicity"><i class="fa fa-check"></i><b>8.5</b> Community comparisons [TODO combine with the material above for simplicity?]</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="beta-diversity.html"><a href="beta-diversity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.5.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.5.2" data-path="beta-diversity.html"><a href="beta-diversity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.5.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="beta-diversity.html"><a href="beta-diversity.html#further-reading-1"><i class="fa fa-check"></i><b>8.6</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="beta-diversity.html"><a href="beta-diversity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Microbiome Community</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#community-composition"><i class="fa fa-check"></i><b>9.1</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>10</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="community-typing.html"><a href="community-typing.html#community-typing-1"><i class="fa fa-check"></i><b>10.1</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.1.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="10.1.2" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.1.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>10.2</b> Community Detection</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#introduction-to-differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Introduction to differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.1</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.6</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multitable.html"><a href="multitable.html"><i class="fa fa-check"></i><b>12</b> Representation of Multiple Data Tables</a>
<ul>
<li class="chapter" data-level="12.1" data-path="multitable.html"><a href="multitable.html#assay-data"><i class="fa fa-check"></i><b>12.1</b> Assay data</a></li>
<li class="chapter" data-level="12.2" data-path="multitable.html"><a href="multitable.html#alternative-experiments"><i class="fa fa-check"></i><b>12.2</b> Alternative experiments</a></li>
<li class="chapter" data-level="12.3" data-path="multitable.html"><a href="multitable.html#multiassayexperiments"><i class="fa fa-check"></i><b>12.3</b> MultiAssayExperiments</a></li>
<li class="chapter" data-level="" data-path="multitable.html"><a href="multitable.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="function-ref.html"><a href="function-ref.html"><i class="fa fa-check"></i><b>13</b> Function Reference</a>
<ul>
<li class="chapter" data-level="13.1" data-path="function-ref.html"><a href="function-ref.html#objects"><i class="fa fa-check"></i><b>13.1</b> Objects</a></li>
<li class="chapter" data-level="13.2" data-path="function-ref.html"><a href="function-ref.html#reduced-dimension-pca-t-sne-umap-mds-etc"><i class="fa fa-check"></i><b>13.2</b> Reduced dimension (PCA, t-SNE, UMAP, MDS, etc)</a></li>
<li class="chapter" data-level="13.3" data-path="function-ref.html"><a href="function-ref.html#plotting"><i class="fa fa-check"></i><b>13.3</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>14</b> Extras</a>
<ul>
<li class="chapter" data-level="14.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>14.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-introduction" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Data Infrastructure</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p><a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html"><code>SummarizedExperiment</code></a>
(<code>SE</code>) is a generic and highly optimized container for complex data
structures. It has become a common choice for analysing various types
of biomedical profiling data, such as RNAseq, ChIp-Seq, microarrays,
flow cytometry, proteomics, and single-cell
sequencing. <a href="https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html"><code>TreeSummarizedExperiment</code></a>
(<code>TreeSE</code>) was developed as an extension to incorporate hierarchical
information (such as phylogenetic trees and sample hierarchies) and
reference sequences.</p>
<p>In microbiome data science, these containers can be used to link
taxonomic abundance tables with rich side information on the features
and samples. Taxonomic abundance data can be obtained by 16S rRNA
amplicon or metagenomic sequencing, phylogenetic microarrays, or by
other means.</p>
<div id="package-installation" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Package installation</h2>
<p>Several R packages are now available providing methods for the
analysis and manipulation of data sets in such containers. One of
these is <code>mia</code>.</p>
<p>Install the Biocondcuctor development version with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="data-introduction.html#cb1-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;microbiome/mia&quot;</span>, <span class="at">version=</span><span class="st">&quot;devel&quot;</span>)</span></code></pre></div>
</div>
<div id="background" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Background</h2>
<p>The widely used <code>phyloseq</code> package and class were around before the
<a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html"><code>SummarizedExperiment</code></a>
and the derived
<a href="https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html"><code>TreeSummarizedExperiment</code></a>
class. Many methods for taxonomic profiling data are readily available
for the <code>phyloseq</code> class structure.</p>
<p>In order to facilitate the transition, we provide here a short
description how <code>phyloseq</code> and <code>*Experiment</code> classes relate to each
other.</p>
<p><code>assays</code> : This slot is similar to <code>otu_table</code> in <code>phyloseq</code>. In a
<code>SummarizedExperiment</code> object multiple assays, raw
counts, transformed counts can be stored. See also
<a href="https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html"><code>MultiAssayExperiment</code></a>
for storing data from multiple <code>experiments</code> such as
RNASeq, Proteomics, etc. <code>rowData</code> : This slot is
similar to <code>tax_table</code> in <code>phyloseq</code> to store taxonomic
information. <code>colData</code> : This slot is similar to
<code>sample_data</code> in <code>phyloseq</code> to store information
related to samples. <code>rowTree</code> : This slot is similar
to <code>phy_tree</code> in <code>phyloseq</code> to store phylogenetic tree.</p>
<p>In this book, you will come across terms like <code>FeatureIDs</code> and
<code>SampleIDs</code>. <code>FeatureIDs</code> : These are basically OTU/ASV ids which are
row names in <code>assays</code> and <code>rowData</code>. <code>SampleIDs</code> : As the name
suggests, these are sample ids which are column names in <code>assays</code> and
row names in <code>colData</code>.</p>
<p><code>FeatureIDs</code> and <code>SampleIDs</code> are used but the technical terms
<code>rownames</code> and <code>colnames</code> are encouraged to be used, since they relate
to actual objects we work with.</p>
<p><img
src="https://raw.githubusercontent.com/FelixErnst/TreeSummarizedExperiment
/2293440c6e70ae4d6e978b6fdf2c42fdea7fb36a/vignettes/tse2.png"
width="100%"/></p>
<p><strong>Figure sources:</strong></p>
<p><strong>Original article</strong>
- Huang R <em>et al</em>. (2021) <a href="https://doi.org/10.12688/%20f1000research.26669.2">TreeSummarizedExperiment: a S4 class
for data with hierarchical structure</a>. F1000Research 9:1246.</p>
<p><strong>Reference Sequence slot extension</strong>
- Lahti L <em>et al</em>. (2020) <a href="https://doi.org/10.7490/%20f1000research.1118447.1">Upgrading the R/Bioconductor ecosystem for microbiome
research</a> F1000Research 9:1464 (slides).</p>
</div>
<div id="loading-experimental-microbiome-data" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Loading experimental microbiome data</h2>
<div id="importing-data-from-external-files" class="section level3" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> Importing data from external files</h3>
<p>Microbiome (taxonomic) profiling data is commonly distributed in
various file formats. You can import such external data files as a
(Tree)SummarizedExperiment object but the details depend on the file
format. Here, we provide examples for common formats.</p>
<p><strong>CSV data tables</strong> can be imported with the standard R functions,
then converted to the desired format. For detailed examples, you can
check the <a href="https://bioconductor.org/help/course-materials/2019/BSS2019/04_Practical_CoreApproachesInBioconductor.html">Bioconductor course
material</a>
by Martin Morgan. The following example reads abundance tables,
taxonomic mapping tables, and sample metadata, assuming that the
input data files are properly prepared with appropriate row and
column names.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="data-introduction.html#cb2-1" aria-hidden="true" tabindex="-1"></a>counts  <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(count_file)   <span class="co"># Abundance table (e.g. ASV data; to assay data)</span></span>
<span id="cb2-2"><a href="data-introduction.html#cb2-2" aria-hidden="true" tabindex="-1"></a>tax     <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(tax_file)     <span class="co"># Taxonomy table (to rowData)</span></span>
<span id="cb2-3"><a href="data-introduction.html#cb2-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(sample_file)  <span class="co"># Sample data (to colData)</span></span>
<span id="cb2-4"><a href="data-introduction.html#cb2-4" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> counts),</span>
<span id="cb2-5"><a href="data-introduction.html#cb2-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">colData =</span> samples,</span>
<span id="cb2-6"><a href="data-introduction.html#cb2-6" aria-hidden="true" tabindex="-1"></a>                           <span class="at">rowData =</span> tax)</span></code></pre></div>
<p>Specific import functions are provided for:</p>
<ul>
<li>Biom files (see <code>help(mia::loadFromBiom)</code>)</li>
<li>QIIME2 files (see <code>help(mia::loadFromQIIME2)</code>)</li>
<li>Mothur files (see <code>help(mia::loadFromMothur)</code>)</li>
</ul>
<div id="example-for-importing-biom-files" class="section level4" number="2.3.1.1">
<h4><span class="header-section-number">2.3.1.1</span> Example for importing Biom files</h4>
<p>This example shows how Biom files are imported into a <code>TreeSummarizedExperiment</code> object.</p>
<p>The data is from following publication:
Tengeler AC <em>et al.</em> (2020) <a href="https://doi.org/10.1186/s40168-020-00816-x"><strong>Gut microbiota from persons with
attention-deficit/hyperactivity disorder affects the brain in
mice</strong></a>.</p>
<p>The data set consists of 3 files:</p>
<ul>
<li>biom file: abundance table and taxonomy information</li>
<li>csv file: sample metadata</li>
<li>tree file: phylogenetic tree</li>
</ul>
<p>Store the data in your desired local directory (for instance, <em>data/</em> under the
working directory), and define source file paths</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="data-introduction.html#cb3-1" aria-hidden="true" tabindex="-1"></a>biom_file_path <span class="ot">&lt;-</span> <span class="st">&quot;data/Aggregated_humanization2.biom&quot;</span></span>
<span id="cb3-2"><a href="data-introduction.html#cb3-2" aria-hidden="true" tabindex="-1"></a>sample_meta_file_path <span class="ot">&lt;-</span> <span class="st">&quot;data/Mapping_file_ADHD_aggregated.csv&quot;</span></span>
<span id="cb3-3"><a href="data-introduction.html#cb3-3" aria-hidden="true" tabindex="-1"></a>tree_file_path <span class="ot">&lt;-</span> <span class="st">&quot;data/Data_humanization_phylo_aggregation.tre&quot;</span></span></code></pre></div>
<p>Now we can load the biom data into a SummarizedExperiment (SE) object.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="data-introduction.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb4-2"><a href="data-introduction.html#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="data-introduction.html#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Imports the data</span></span>
<span id="cb4-4"><a href="data-introduction.html#cb4-4" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">loadFromBiom</span>(biom_file_path)</span>
<span id="cb4-5"><a href="data-introduction.html#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="data-introduction.html#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Check</span></span>
<span id="cb4-7"><a href="data-introduction.html#cb4-7" aria-hidden="true" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): taxonomy1 taxonomy2 ... taxonomy5 taxonomy6
## colnames(27): A110 A111 ... A38 A39
## colData names(0):</code></pre>
<p>The <code>assays</code> slot includes a list of abundance tables. The imported
abundance table is named as “counts.” Let us inspect only the first
cols and rows.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="data-introduction.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)<span class="sc">$</span>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>##           A110  A111  A12
## 1726470  17722 11630    0
## 1726471  12052     0 2679
## 17264731     0   970    0</code></pre>
<p>The <code>rowdata</code> includes taxonomic information from the biom file. The <code>head()</code> command
shows just the beginning of the data table for an overview.</p>
<p><code>knitr::kable()</code> is for printing the information more nicely.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="data-introduction.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(se))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##             taxonomy1          taxonomy2           taxonomy3
##           &lt;character&gt;        &lt;character&gt;         &lt;character&gt;
## 1726470  &quot;k__Bacteria   p__Bacteroidetes      c__Bacteroidia
## 1726471  &quot;k__Bacteria   p__Bacteroidetes      c__Bacteroidia
## 17264731 &quot;k__Bacteria   p__Bacteroidetes      c__Bacteroidia
## 17264726 &quot;k__Bacteria   p__Bacteroidetes      c__Bacteroidia
## 1726472  &quot;k__Bacteria p__Verrucomicrobia c__Verrucomicrobiae
## 17264724 &quot;k__Bacteria   p__Bacteroidetes      c__Bacteroidia
##                      taxonomy4              taxonomy5           taxonomy6
##                    &lt;character&gt;            &lt;character&gt;         &lt;character&gt;
## 1726470       o__Bacteroidales      f__Bacteroidaceae     g__Bacteroides&quot;
## 1726471       o__Bacteroidales      f__Bacteroidaceae     g__Bacteroides&quot;
## 17264731      o__Bacteroidales  f__Porphyromonadaceae g__Parabacteroides&quot;
## 17264726      o__Bacteroidales      f__Bacteroidaceae     g__Bacteroides&quot;
## 1726472  o__Verrucomicrobiales f__Verrucomicrobiaceae     g__Akkermansia&quot;
## 17264724      o__Bacteroidales      f__Bacteroidaceae     g__Bacteroides&quot;</code></pre>
<p>These taxonomic rank names (column names) are not real rank
names. Let’s replace them with real rank names.</p>
<p>In addition to that, the taxa names include, e.g., ’"k__’ before the name, so let’s
make them cleaner by removing them.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="data-introduction.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">rowData</span>(se)) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Kingdom&quot;</span>, <span class="st">&quot;Phylum&quot;</span>, <span class="st">&quot;Class&quot;</span>, <span class="st">&quot;Order&quot;</span>, </span>
<span id="cb10-2"><a href="data-introduction.html#cb10-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">&quot;Family&quot;</span>, <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb10-3"><a href="data-introduction.html#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="data-introduction.html#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Goes through the whole DataFrame. Removes &#39;.*[kpcofg]__&#39; from strings, where [kpcofg] </span></span>
<span id="cb10-5"><a href="data-introduction.html#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># is any character from listed ones, and .* any character.</span></span>
<span id="cb10-6"><a href="data-introduction.html#cb10-6" aria-hidden="true" tabindex="-1"></a>rowdata_modified <span class="ot">&lt;-</span> BiocParallel<span class="sc">::</span><span class="fu">bplapply</span>(<span class="fu">rowData</span>(se), </span>
<span id="cb10-7"><a href="data-introduction.html#cb10-7" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">FUN =</span> stringr<span class="sc">::</span>str_remove, </span>
<span id="cb10-8"><a href="data-introduction.html#cb10-8" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">pattern =</span> <span class="st">&#39;.*[kpcofg]__&#39;</span>)</span>
<span id="cb10-9"><a href="data-introduction.html#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="data-introduction.html#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Genus level has additional &#39;\&quot;&#39;, so let&#39;s delete that also</span></span>
<span id="cb10-11"><a href="data-introduction.html#cb10-11" aria-hidden="true" tabindex="-1"></a>rowdata_modified <span class="ot">&lt;-</span> BiocParallel<span class="sc">::</span><span class="fu">bplapply</span>(rowdata_modified, </span>
<span id="cb10-12"><a href="data-introduction.html#cb10-12" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">FUN =</span> stringr<span class="sc">::</span>str_remove, </span>
<span id="cb10-13"><a href="data-introduction.html#cb10-13" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">pattern =</span> <span class="st">&#39;</span><span class="sc">\&quot;</span><span class="st">&#39;</span>)</span>
<span id="cb10-14"><a href="data-introduction.html#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="data-introduction.html#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="co"># rowdata_modified is a list, so it is converted back to DataFrame format. </span></span>
<span id="cb10-16"><a href="data-introduction.html#cb10-16" aria-hidden="true" tabindex="-1"></a>rowdata_modified <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(rowdata_modified)</span>
<span id="cb10-17"><a href="data-introduction.html#cb10-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-18"><a href="data-introduction.html#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="co"># And then assigned back to the SE object</span></span>
<span id="cb10-19"><a href="data-introduction.html#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se) <span class="ot">&lt;-</span> rowdata_modified</span>
<span id="cb10-20"><a href="data-introduction.html#cb10-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-21"><a href="data-introduction.html#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we have a nicer table</span></span>
<span id="cb10-22"><a href="data-introduction.html#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(se))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##              Kingdom          Phylum            Class              Order
##          &lt;character&gt;     &lt;character&gt;      &lt;character&gt;        &lt;character&gt;
## 1726470     Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726471     Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264731    Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264726    Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726472     Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales
## 17264724    Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
##                       Family           Genus
##                  &lt;character&gt;     &lt;character&gt;
## 1726470       Bacteroidaceae     Bacteroides
## 1726471       Bacteroidaceae     Bacteroides
## 17264731  Porphyromonadaceae Parabacteroides
## 17264726      Bacteroidaceae     Bacteroides
## 1726472  Verrucomicrobiaceae     Akkermansia
## 17264724      Bacteroidaceae     Bacteroides</code></pre>
<p>We notice that the imported biom file did not contain the sample meta data
yet, so it includes an empty data frame.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="data-introduction.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 0 columns</code></pre>
<p>Let us add a sample metadata file.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="data-introduction.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We use this to check what type of data it is</span></span>
<span id="cb14-2"><a href="data-introduction.html#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># read.table(sample_meta_file_path)</span></span>
<span id="cb14-3"><a href="data-introduction.html#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="data-introduction.html#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It seems like a comma separated file and it does not include headers</span></span>
<span id="cb14-5"><a href="data-introduction.html#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Let us read it and then convert from data.frame to DataFrame</span></span>
<span id="cb14-6"><a href="data-introduction.html#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co"># (required for our purposes)</span></span>
<span id="cb14-7"><a href="data-introduction.html#cb14-7" aria-hidden="true" tabindex="-1"></a>sample_meta <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="fu">read.table</span>(sample_meta_file_path, <span class="at">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="at">header =</span> <span class="cn">FALSE</span>))</span>
<span id="cb14-8"><a href="data-introduction.html#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="data-introduction.html#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Add sample names to rownames</span></span>
<span id="cb14-10"><a href="data-introduction.html#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sample_meta) <span class="ot">&lt;-</span> sample_meta[,<span class="dv">1</span>]</span>
<span id="cb14-11"><a href="data-introduction.html#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="data-introduction.html#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Delete column that included sample names</span></span>
<span id="cb14-13"><a href="data-introduction.html#cb14-13" aria-hidden="true" tabindex="-1"></a>sample_meta[,<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-14"><a href="data-introduction.html#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="data-introduction.html#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="co"># We can add headers</span></span>
<span id="cb14-16"><a href="data-introduction.html#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sample_meta) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;patient_status&quot;</span>, <span class="st">&quot;cohort&quot;</span>, <span class="st">&quot;patient_status_vs_cohort&quot;</span>, <span class="st">&quot;sample_name&quot;</span>)</span>
<span id="cb14-17"><a href="data-introduction.html#cb14-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-18"><a href="data-introduction.html#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Then it can be added to colData</span></span>
<span id="cb14-19"><a href="data-introduction.html#cb14-19" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se) <span class="ot">&lt;-</span> sample_meta</span></code></pre></div>
<p>Now <code>colData</code> includes the sample metadata.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="data-introduction.html#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##      patient_status      cohort patient_status_vs_cohort sample_name
##         &lt;character&gt; &lt;character&gt;              &lt;character&gt; &lt;character&gt;
## A110           ADHD    Cohort_1            ADHD_Cohort_1        A110
## A12            ADHD    Cohort_1            ADHD_Cohort_1         A12
## A15            ADHD    Cohort_1            ADHD_Cohort_1         A15
## A19            ADHD    Cohort_1            ADHD_Cohort_1         A19
## A21            ADHD    Cohort_2            ADHD_Cohort_2         A21
## A23            ADHD    Cohort_2            ADHD_Cohort_2         A23</code></pre>
<p>Now, let’s add a phylogenetic tree.</p>
<p>The current data object, se, is a SummarizedExperiment object. This
does not include a slot for adding a phylogenetic tree. In order to do
this, we can convert the SE object to an extended TreeSummarizedExperiment
object which includes also a <code>rowTree</code> slot.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="data-introduction.html#cb17-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">as</span>(se, <span class="st">&quot;TreeSummarizedExperiment&quot;</span>)</span>
<span id="cb17-2"><a href="data-introduction.html#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="data-introduction.html#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># tse includes same data as se</span></span>
<span id="cb17-4"><a href="data-introduction.html#cb17-4" aria-hidden="true" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): Kingdom Phylum ... Family Genus
## colnames(27): A110 A12 ... A35 A38
## colData names(4): patient_status cohort patient_status_vs_cohort
##   sample_name
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Next, let us read the tree data file and add it to the R data object (tse).</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="data-introduction.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reads the tree file</span></span>
<span id="cb19-2"><a href="data-introduction.html#cb19-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(tree_file_path)</span>
<span id="cb19-3"><a href="data-introduction.html#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="data-introduction.html#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add tree to rowTree</span></span>
<span id="cb19-5"><a href="data-introduction.html#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rowTree</span>(tse) <span class="ot">&lt;-</span> tree</span>
<span id="cb19-6"><a href="data-introduction.html#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="data-introduction.html#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Check</span></span>
<span id="cb19-8"><a href="data-introduction.html#cb19-8" aria-hidden="true" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): Kingdom Phylum ... Family Genus
## colnames(27): A110 A12 ... A35 A38
## colData names(4): patient_status cohort patient_status_vs_cohort
##   sample_name
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (151 rows)
## rowTree: 1 phylo tree(s) (151 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Now <code>rowTree</code> includes a phylogenetic tree:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="data-introduction.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowTree</span>(tse))</span></code></pre></div>
</div>
</div>
<div id="conversions-between-data-formats-in-r" class="section level3" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Conversions between data formats in R</h3>
<p>If the data has already been imported in R in another format, it
can be readily converted into <code>TreeSummarizedExperiment</code>, as shown in our next
example. Note that similar conversion functions to
<code>TreeSummarizedExperiment</code> are available for multiple data formats via
the <code>mia</code> package (see makeTreeSummarizedExperimentFrom* for phyloseq,
Biom, and DADA2).</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="data-introduction.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb22-2"><a href="data-introduction.html#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="data-introduction.html#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># phyloseq example data</span></span>
<span id="cb22-4"><a href="data-introduction.html#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns, <span class="at">package=</span><span class="st">&quot;phyloseq&quot;</span>) </span>
<span id="cb22-5"><a href="data-introduction.html#cb22-5" aria-hidden="true" tabindex="-1"></a>GlobalPatterns_phyloseq <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb22-6"><a href="data-introduction.html#cb22-6" aria-hidden="true" tabindex="-1"></a>GlobalPatterns_phyloseq</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="data-introduction.html#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert phyloseq to TSE</span></span>
<span id="cb24-2"><a href="data-introduction.html#cb24-2" aria-hidden="true" tabindex="-1"></a>GlobalPatterns_TSE <span class="ot">&lt;-</span> <span class="fu">makeTreeSummarizedExperimentFromPhyloseq</span>(GlobalPatterns_phyloseq) </span>
<span id="cb24-3"><a href="data-introduction.html#cb24-3" aria-hidden="true" tabindex="-1"></a>GlobalPatterns_TSE</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>We can also convert <code>TreeSummarizedExperiment</code> objects into <code>phyloseq</code>
with respect to the shared components that are supported by both
formats (i.e. taxonomic abundance table, sample metadata, taxonomic
table, phylogenetic tree, sequence information). This is useful for
instance when additional methods are available for <code>phyloseq</code>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="data-introduction.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert TSE to phyloseq</span></span>
<span id="cb26-2"><a href="data-introduction.html#cb26-2" aria-hidden="true" tabindex="-1"></a>GlobalPatterns_phyloseq2 <span class="ot">&lt;-</span> <span class="fu">makePhyloseqFromTreeSummarizedExperiment</span>(GlobalPatterns_TSE) </span>
<span id="cb26-3"><a href="data-introduction.html#cb26-3" aria-hidden="true" tabindex="-1"></a>GlobalPatterns_phyloseq2</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
</div>
<div id="metadata" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Metadata</h2>
</div>
<div id="microbiome-and-tree-data-specific-aspects" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Microbiome and tree data specific aspects</h2>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="data-introduction.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb28-2"><a href="data-introduction.html#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb28-3"><a href="data-introduction.html#cb28-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> GlobalPatterns </span>
<span id="cb28-4"><a href="data-introduction.html#cb28-4" aria-hidden="true" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div id="assays" class="section level3" number="2.5.1">
<h3><span class="header-section-number">2.5.1</span> Assays</h3>
<p>The <code>assays</code> slot contains the experimental data as count matrices. Multiple
matrices can be stored the result of <code>assays</code> is actually a list of matrices.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="data-introduction.html#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code></pre></div>
<pre><code>## List of length 1
## names(1): counts</code></pre>
<p>Individual assays can be accessed via <code>assay</code></p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="data-introduction.html#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">&quot;counts&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##        CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## 549322   0   0   0       0       0       0       0
## 522457   0   0   0       0       0       0       0
## 951      0   0   0       0       0       0       1
## 244423   0   0   0       0       0       0       0
## 586076   0   0   0       0       0       0       0</code></pre>
<p>To illustrate the use of multiple assays, the relative abundance data can be
calcualted and stored along the original count data using <code>relAbundanceCounts</code>.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="data-introduction.html#cb34-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">relAbundanceCounts</span>(se)</span>
<span id="cb34-2"><a href="data-introduction.html#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code></pre></div>
<pre><code>## List of length 2
## names(2): counts relabundance</code></pre>
<p>Now there are two assays available in the <code>se</code> object, <code>counts</code> and
<code>relabundance</code>.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="data-introduction.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">&quot;relabundance&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##        CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr   M11Plmr
## 549322   0   0   0       0       0       0 0.000e+00
## 522457   0   0   0       0       0       0 0.000e+00
## 951      0   0   0       0       0       0 2.305e-06
## 244423   0   0   0       0       0       0 0.000e+00
## 586076   0   0   0       0       0       0 0.000e+00</code></pre>
</div>
<div id="coldata" class="section level3" number="2.5.2">
<h3><span class="header-section-number">2.5.2</span> colData</h3>
<p><code>colData</code> contains data on the samples.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="data-introduction.html#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code></pre></div>
<pre><code>## DataFrame with 26 rows and 7 columns
##         X.SampleID   Primer Final_Barcode Barcode_truncated_plus_T
##           &lt;factor&gt; &lt;factor&gt;      &lt;factor&gt;                 &lt;factor&gt;
## CL3        CL3      ILBC_01        AACGCA                   TGCGTT
## CC1        CC1      ILBC_02        AACTCG                   CGAGTT
## SV1        SV1      ILBC_03        AACTGT                   ACAGTT
## M31Fcsw    M31Fcsw  ILBC_04        AAGAGA                   TCTCTT
## M11Fcsw    M11Fcsw  ILBC_05        AAGCTG                   CAGCTT
## ...            ...      ...           ...                      ...
## TS28         TS28   ILBC_25        ACCAGA                   TCTGGT
## TS29         TS29   ILBC_26        ACCAGC                   GCTGGT
## Even1        Even1  ILBC_27        ACCGCA                   TGCGGT
## Even2        Even2  ILBC_28        ACCTCG                   CGAGGT
## Even3        Even3  ILBC_29        ACCTGT                   ACAGGT
##         Barcode_full_length SampleType
##                    &lt;factor&gt;   &lt;factor&gt;
## CL3             CTAGCGTGCGT      Soil 
## CC1             CATCGACGAGT      Soil 
## SV1             GTACGCACAGT      Soil 
## M31Fcsw         TCGACATCTCT      Feces
## M11Fcsw         CGACTGCAGCT      Feces
## ...                     ...        ...
## TS28            GCATCGTCTGG      Feces
## TS29            CTAGTCGCTGG      Feces
## Even1           TGACTCTGCGG      Mock 
## Even2           TCTGATCGAGG      Mock 
## Even3           AGAGAGACAGG      Mock 
##                                        Description
##                                           &lt;factor&gt;
## CL3     Calhoun South Carolina Pine soil, pH 4.9  
## CC1     Cedar Creek Minnesota, grassland, pH 6.1  
## SV1     Sevilleta new Mexico, desert scrub, pH 8.3
## M31Fcsw M3, Day 1, fecal swab, whole body study   
## M11Fcsw M1, Day 1, fecal swab, whole body study   
## ...                                            ...
## TS28                                       Twin #1
## TS29                                       Twin #2
## Even1                                      Even1  
## Even2                                      Even2  
## Even3                                      Even3</code></pre>
</div>
<div id="rowdata" class="section level3" number="2.5.3">
<h3><span class="header-section-number">2.5.3</span> rowData</h3>
<p><code>rowData</code> contains data on the features of the analyzed samples. Of particular
interest for the microbiome field this is used to store taxonomic information.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="data-introduction.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)</span></code></pre></div>
<pre><code>## DataFrame with 19216 rows and 7 columns
##            Kingdom        Phylum        Class        Order        Family
##        &lt;character&gt;   &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;
## 549322     Archaea Crenarchaeota Thermoprotei           NA            NA
## 522457     Archaea Crenarchaeota Thermoprotei           NA            NA
## 951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 244423     Archaea Crenarchaeota        Sd-NA           NA            NA
## 586076     Archaea Crenarchaeota        Sd-NA           NA            NA
## ...            ...           ...          ...          ...           ...
## 278222    Bacteria           SR1           NA           NA            NA
## 463590    Bacteria           SR1           NA           NA            NA
## 535321    Bacteria           SR1           NA           NA            NA
## 200359    Bacteria           SR1           NA           NA            NA
## 271582    Bacteria           SR1           NA           NA            NA
##              Genus                Species
##        &lt;character&gt;            &lt;character&gt;
## 549322          NA                     NA
## 522457          NA                     NA
## 951     Sulfolobus Sulfolobusacidocalda..
## 244423          NA                     NA
## 586076          NA                     NA
## ...            ...                    ...
## 278222          NA                     NA
## 463590          NA                     NA
## 535321          NA                     NA
## 200359          NA                     NA
## 271582          NA                     NA</code></pre>
</div>
<div id="rowtree" class="section level3" number="2.5.4">
<h3><span class="header-section-number">2.5.4</span> rowTree</h3>
<p>Phylogenetic trees also play an important role for the microbiome field. The
<code>TreeSummarizedExperiment</code> class is able to keep track of feature and node
relations via two functions, <code>rowTree</code> and <code>rowLinks</code>.</p>
<p>A tree can be accessed via <code>rowTree</code> as <code>phylo</code> object.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="data-introduction.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowTree</span>(se)</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 19216 tips and 19215 internal nodes.
## 
## Tip labels:
##   549322, 522457, 951, 244423, 586076, 246140, ...
## Node labels:
##   , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>The links to the individual features are available through <code>rowLinks</code>.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="data-introduction.html#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowLinks</span>(se)</span></code></pre></div>
<pre><code>## LinkDataFrame with 19216 rows and 5 columns
##           nodeLab   nodeNum nodeLab_alias    isLeaf   whichTree
##       &lt;character&gt; &lt;integer&gt;   &lt;character&gt; &lt;logical&gt; &lt;character&gt;
## 1          549322         1       alias_1      TRUE       phylo
## 2          522457         2       alias_2      TRUE       phylo
## 3             951         3       alias_3      TRUE       phylo
## 4          244423         4       alias_4      TRUE       phylo
## 5          586076         5       alias_5      TRUE       phylo
## ...           ...       ...           ...       ...         ...
## 19212      278222     19212   alias_19212      TRUE       phylo
## 19213      463590     19213   alias_19213      TRUE       phylo
## 19214      535321     19214   alias_19214      TRUE       phylo
## 19215      200359     19215   alias_19215      TRUE       phylo
## 19216      271582     19216   alias_19216      TRUE       phylo</code></pre>
<p>Please note that there can be a 1:1 relationship between tree nodes and
features, but this is not a must have. This means there can be features, which
are not linked to nodes, and nodes, which are not linked to features. To change
the links in an existing object, the <code>changeTree</code> function is available.</p>
</div>
</div>
<div id="data-manipulation" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Data manipulation</h2>
<p>Sometimes custom solutions are needed for analyzing the data. <code>mia</code> contains a
few functions to help in these situations.</p>
<div id="tidy-data" class="section level3" number="2.6.1">
<h3><span class="header-section-number">2.6.1</span> Tidy data</h3>
<p>For several custom analysis and visualization packages, such as those from the
<code>tidyverse</code>, the <code>SE</code> data can be converted to long data.frame format with
<code>meltAssay</code>.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="data-introduction.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb46-2"><a href="data-introduction.html#cb46-2" aria-hidden="true" tabindex="-1"></a>molten_se <span class="ot">&lt;-</span> <span class="fu">meltAssay</span>(se,</span>
<span id="cb46-3"><a href="data-introduction.html#cb46-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add_row_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb46-4"><a href="data-introduction.html#cb46-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add_col_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb46-5"><a href="data-introduction.html#cb46-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">abund_values =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb46-6"><a href="data-introduction.html#cb46-6" aria-hidden="true" tabindex="-1"></a>molten_se</span></code></pre></div>
<pre><code>## # A tibble: 499,616 × 17
##    FeatureID SampleID relabundance Kingdom Phylum    Class    Order Family Genus
##    &lt;fct&gt;     &lt;fct&gt;           &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;
##  1 549322    CL3                 0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  2 549322    CC1                 0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  3 549322    SV1                 0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  4 549322    M31Fcsw             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  5 549322    M11Fcsw             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  6 549322    M31Plmr             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  7 549322    M11Plmr             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  8 549322    F21Plmr             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
##  9 549322    M31Tong             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
## 10 549322    M11Tong             0 Archaea Crenarch… Thermop… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
## # … with 499,606 more rows, and 8 more variables: Species &lt;chr&gt;,
## #   X.SampleID &lt;fct&gt;, Primer &lt;fct&gt;, Final_Barcode &lt;fct&gt;,
## #   Barcode_truncated_plus_T &lt;fct&gt;, Barcode_full_length &lt;fct&gt;,
## #   SampleType &lt;fct&gt;, Description &lt;fct&gt;</code></pre>
</div>
<div id="subsetting" class="section level3" number="2.6.2">
<h3><span class="header-section-number">2.6.2</span> Subsetting</h3>
<p><strong>Subsetting</strong> data helps to draw the focus of analysis on particular
sets of samples and / or features. When dealing with large data
sets, the subset of interest can be extracted and investigated
separately. This might improve performance and reduce the
computational load.</p>
<p>Load:</p>
<ul>
<li>mia</li>
<li>dplyr</li>
<li>knitr</li>
<li>data <code>GlobalPatterns</code></li>
</ul>
<p>Let us store <code>GlobalPatterns</code> into <code>se</code> and check its original number of features (rows) and samples (columns). <strong>Note</strong>: when subsetting by sample, expect the number of columns to decrease; when subsetting by feature, expect the number of rows to decrease.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="data-introduction.html#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># store data into se and check dimensions</span></span>
<span id="cb48-2"><a href="data-introduction.html#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb48-3"><a href="data-introduction.html#cb48-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb48-4"><a href="data-introduction.html#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb48-5"><a href="data-introduction.html#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se) <span class="do">## [1] n°features    n°samples</span></span></code></pre></div>
<pre><code>## [1] 19216    26</code></pre>
<div id="subset-by-sample-column-wise" class="section level4" number="2.6.2.1">
<h4><span class="header-section-number">2.6.2.1</span> Subset by sample (column-wise)</h4>
<p>For the sake of demonstration, here we will extract a subset containing only the samples of human origin (feces, skin or tongue), stored as <code>SampleType</code> within <code>colData(se)</code> and also in <code>se</code>.</p>
<p>First, we would like to see all the possible values that <code>SampleType</code> can take on and how frequent those are:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="data-introduction.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect possible values for SampleType</span></span>
<span id="cb50-2"><a href="data-introduction.html#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(se<span class="sc">$</span>SampleType)</span></code></pre></div>
<pre><code>## [1] Soil               Feces              Skin               Tongue            
## [5] Freshwater         Freshwater (creek) Ocean              Sediment (estuary)
## [9] Mock              
## 9 Levels: Feces Freshwater Freshwater (creek) Mock ... Tongue</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="data-introduction.html#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show recurrence for each value</span></span>
<span id="cb52-2"><a href="data-introduction.html#cb52-2" aria-hidden="true" tabindex="-1"></a>se<span class="sc">$</span>SampleType <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
.
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Feces
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
Freshwater
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Freshwater (creek)
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Mock
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Ocean
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Sediment (estuary)
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Skin
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Soil
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Tongue
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table>
</div>
<p><strong>Note</strong>: after subsetting, expect the number of columns to equal the
sum of the recurrences of the samples that you are interested
in. For instance, <code>ncols = Feces + Skin + Tongue = 4 + 3 + 2 = 9</code>.</p>
<p>Next, we <em>logical index</em> across the columns of <code>se</code> (make sure to
leave the first index empty to select all rows) and filter for the
samples of human origin. For this, we use the information on the
samples from the meta data <code>colData(se)</code>.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="data-introduction.html#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by sample</span></span>
<span id="cb53-2"><a href="data-introduction.html#cb53-2" aria-hidden="true" tabindex="-1"></a>se_subset_by_sample <span class="ot">&lt;-</span> se[ , se<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>)]</span>
<span id="cb53-3"><a href="data-introduction.html#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="data-introduction.html#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb53-5"><a href="data-introduction.html#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se_subset_by_sample)</span></code></pre></div>
<pre><code>## [1] 19216     9</code></pre>
<p>As a sanity check, the new object <code>se_subset_by_sample</code> should have
the original number of features (rows) and a number of samples
(columns) equal to the sum of the samples of interest (in this case
9).</p>
<p>Several characteristics can be used to subset by sample:</p>
<ul>
<li>origin</li>
<li>sampling time</li>
<li>sequencing method</li>
<li>DNA / RNA barcode</li>
<li>cohort</li>
</ul>
</div>
<div id="subset-by-feature-row-wise" class="section level4" number="2.6.2.2">
<h4><span class="header-section-number">2.6.2.2</span> Subset by feature (row-wise)</h4>
<p>Similarly, here we will extract a subset containing only the features
that belong to the Phyla “Actinobacteria” and “Chlamydiae,” stored as
<code>Phylum</code> within <code>rowData(se)</code>. However, subsetting by feature implies
a few more obstacles, such as the presence of NA elements and the
possible need for agglomeration.</p>
<p>As previously, we would first like to see all the possible values that
<code>Phylum</code> can take on and how frequent those are:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="data-introduction.html#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect possible values for Phylum</span></span>
<span id="cb55-2"><a href="data-introduction.html#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">rowData</span>(se)<span class="sc">$</span>Phylum)</span></code></pre></div>
<pre><code>##  [1] &quot;Crenarchaeota&quot;    &quot;Euryarchaeota&quot;    &quot;Actinobacteria&quot;   &quot;Spirochaetes&quot;    
##  [5] &quot;MVP-15&quot;           &quot;Proteobacteria&quot;   &quot;SBR1093&quot;          &quot;Fusobacteria&quot;    
##  [9] &quot;Tenericutes&quot;      &quot;ZB3&quot;              &quot;Cyanobacteria&quot;    &quot;GOUTA4&quot;          
## [13] &quot;TG3&quot;              &quot;Chlorobi&quot;         &quot;Bacteroidetes&quot;    &quot;Caldithrix&quot;      
## [17] &quot;KSB1&quot;             &quot;SAR406&quot;           &quot;LCP-89&quot;           &quot;Thermi&quot;          
## [21] &quot;Gemmatimonadetes&quot; &quot;Fibrobacteres&quot;    &quot;GN06&quot;             &quot;AC1&quot;             
## [25] &quot;TM6&quot;              &quot;OP8&quot;              &quot;Elusimicrobia&quot;    &quot;NC10&quot;            
## [29] &quot;SPAM&quot;             NA                 &quot;Acidobacteria&quot;    &quot;CCM11b&quot;          
## [33] &quot;Nitrospirae&quot;      &quot;NKB19&quot;            &quot;BRC1&quot;             &quot;Hyd24-12&quot;        
## [37] &quot;WS3&quot;              &quot;PAUC34f&quot;          &quot;GN04&quot;             &quot;GN12&quot;            
## [41] &quot;Verrucomicrobia&quot;  &quot;Lentisphaerae&quot;    &quot;LD1&quot;              &quot;Chlamydiae&quot;      
## [45] &quot;OP3&quot;              &quot;Planctomycetes&quot;   &quot;Firmicutes&quot;       &quot;OP9&quot;             
## [49] &quot;WPS-2&quot;            &quot;Armatimonadetes&quot;  &quot;SC3&quot;              &quot;TM7&quot;             
## [53] &quot;GN02&quot;             &quot;SM2F11&quot;           &quot;ABY1_OD1&quot;         &quot;ZB2&quot;             
## [57] &quot;OP11&quot;             &quot;Chloroflexi&quot;      &quot;SC4&quot;              &quot;WS1&quot;             
## [61] &quot;GAL15&quot;            &quot;AD3&quot;              &quot;WS2&quot;              &quot;Caldiserica&quot;     
## [65] &quot;Thermotogae&quot;      &quot;Synergistetes&quot;    &quot;SR1&quot;</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="data-introduction.html#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show recurrence for each value</span></span>
<span id="cb57-2"><a href="data-introduction.html#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)<span class="sc">$</span>Phylum <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
.
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ABY1_OD1
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
AC1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Acidobacteria
</td>
<td style="text-align:right;">
1021
</td>
</tr>
<tr>
<td style="text-align:left;">
Actinobacteria
</td>
<td style="text-align:right;">
1631
</td>
</tr>
<tr>
<td style="text-align:left;">
AD3
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
Armatimonadetes
</td>
<td style="text-align:right;">
61
</td>
</tr>
<tr>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:right;">
2382
</td>
</tr>
<tr>
<td style="text-align:left;">
BRC1
</td>
<td style="text-align:right;">
13
</td>
</tr>
<tr>
<td style="text-align:left;">
Caldiserica
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Caldithrix
</td>
<td style="text-align:right;">
10
</td>
</tr>
<tr>
<td style="text-align:left;">
CCM11b
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Chlamydiae
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
Chlorobi
</td>
<td style="text-align:right;">
64
</td>
</tr>
<tr>
<td style="text-align:left;">
Chloroflexi
</td>
<td style="text-align:right;">
437
</td>
</tr>
<tr>
<td style="text-align:left;">
Crenarchaeota
</td>
<td style="text-align:right;">
106
</td>
</tr>
<tr>
<td style="text-align:left;">
Cyanobacteria
</td>
<td style="text-align:right;">
393
</td>
</tr>
<tr>
<td style="text-align:left;">
Elusimicrobia
</td>
<td style="text-align:right;">
31
</td>
</tr>
<tr>
<td style="text-align:left;">
Euryarchaeota
</td>
<td style="text-align:right;">
102
</td>
</tr>
<tr>
<td style="text-align:left;">
Fibrobacteres
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:right;">
4356
</td>
</tr>
<tr>
<td style="text-align:left;">
Fusobacteria
</td>
<td style="text-align:right;">
37
</td>
</tr>
<tr>
<td style="text-align:left;">
GAL15
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Gemmatimonadetes
</td>
<td style="text-align:right;">
191
</td>
</tr>
<tr>
<td style="text-align:left;">
GN02
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
GN04
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
GN06
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
GN12
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GOUTA4
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
Hyd24-12
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
KSB1
</td>
<td style="text-align:right;">
6
</td>
</tr>
<tr>
<td style="text-align:left;">
LCP-89
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
LD1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Lentisphaerae
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
MVP-15
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
NC10
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
Nitrospirae
</td>
<td style="text-align:right;">
74
</td>
</tr>
<tr>
<td style="text-align:left;">
NKB19
</td>
<td style="text-align:right;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
OP11
</td>
<td style="text-align:right;">
6
</td>
</tr>
<tr>
<td style="text-align:left;">
OP3
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
OP8
</td>
<td style="text-align:right;">
27
</td>
</tr>
<tr>
<td style="text-align:left;">
OP9
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
PAUC34f
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Planctomycetes
</td>
<td style="text-align:right;">
638
</td>
</tr>
<tr>
<td style="text-align:left;">
Proteobacteria
</td>
<td style="text-align:right;">
6416
</td>
</tr>
<tr>
<td style="text-align:left;">
SAR406
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
SBR1093
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
SC3
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
SC4
</td>
<td style="text-align:right;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
SM2F11
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
SPAM
</td>
<td style="text-align:right;">
22
</td>
</tr>
<tr>
<td style="text-align:left;">
Spirochaetes
</td>
<td style="text-align:right;">
124
</td>
</tr>
<tr>
<td style="text-align:left;">
SR1
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
Synergistetes
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
Tenericutes
</td>
<td style="text-align:right;">
143
</td>
</tr>
<tr>
<td style="text-align:left;">
TG3
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
Thermi
</td>
<td style="text-align:right;">
46
</td>
</tr>
<tr>
<td style="text-align:left;">
Thermotogae
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
TM6
</td>
<td style="text-align:right;">
27
</td>
</tr>
<tr>
<td style="text-align:left;">
TM7
</td>
<td style="text-align:right;">
32
</td>
</tr>
<tr>
<td style="text-align:left;">
Verrucomicrobia
</td>
<td style="text-align:right;">
470
</td>
</tr>
<tr>
<td style="text-align:left;">
WPS-2
</td>
<td style="text-align:right;">
20
</td>
</tr>
<tr>
<td style="text-align:left;">
WS1
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
WS2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
WS3
</td>
<td style="text-align:right;">
70
</td>
</tr>
<tr>
<td style="text-align:left;">
ZB2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
ZB3
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table>
</div>
<p><strong>Note</strong>: after subsetting, expect the number of columns to equal the
sum of the recurrences of the feature(s) that you are interested
in. For instance, <code>nrows = Actinobacteria + Chlamydiae = 1631 + 21 =   1652</code>.</p>
<p>Depending on your research question, you might need to or need not
agglomerate the data in the first place: if you want to find the
abundance of each and every feature that belongs to Actinobacteria and
Chlamydiae, agglomeration is not needed; if you want to find the total
abundance of all the features that belong to Actinobacteria or
Chlamydiae, agglomeration is recommended.</p>
<div id="non-agglomerated-data" class="section level5" number="2.6.2.2.1">
<h5><span class="header-section-number">2.6.2.2.1</span> Non-agglomerated data</h5>
<p>Next, we <em>logical index</em> across the rows of <code>se</code> (make sure to leave
the second index empty to select all columns) and filter for the
features that fall in either Actinobacteria or Chlamydiae. For this,
we use the information on the samples from the meta data
<code>rowData(se)</code>.</p>
<p>The first term with the <code>%in%</code> operator are includes all the features
of interest, whereas the second term after the AND operator <code>&amp;</code>
filters out all the features that present a NA in place of Phylum.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="data-introduction.html#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature</span></span>
<span id="cb58-2"><a href="data-introduction.html#cb58-2" aria-hidden="true" tabindex="-1"></a>se_subset_by_feature <span class="ot">&lt;-</span> se[<span class="fu">rowData</span>(se)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Actinobacteria&quot;</span>, <span class="st">&quot;Chlamydiae&quot;</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(se)<span class="sc">$</span>Phylum), ]</span>
<span id="cb58-3"><a href="data-introduction.html#cb58-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-4"><a href="data-introduction.html#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb58-5"><a href="data-introduction.html#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se_subset_by_feature)</span></code></pre></div>
<pre><code>## [1] 1652   26</code></pre>
<p>As a sanity check, the new object <code>se_subset_by_feature</code> should have the original number of samples (columns) and a number of features (rows) equal to the sum of the features of interest (in this case 1652).</p>
</div>
<div id="agglomerated-data" class="section level5" number="2.6.2.2.2">
<h5><span class="header-section-number">2.6.2.2.2</span> Agglomerated data</h5>
<p>When total abundances of certain Phyla are of relevance, the data is initially agglomerated by Phylum. Then, similar steps as in the case of not agglomerated data are followed.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="data-introduction.html#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># agglomerate by Phylum</span></span>
<span id="cb60-2"><a href="data-introduction.html#cb60-2" aria-hidden="true" tabindex="-1"></a>se_phylum <span class="ot">&lt;-</span> se <span class="sc">%&gt;%</span> <span class="fu">agglomerateByRank</span>(<span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb60-3"><a href="data-introduction.html#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="data-introduction.html#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature and get rid of NAs</span></span>
<span id="cb60-5"><a href="data-introduction.html#cb60-5" aria-hidden="true" tabindex="-1"></a>se_phylum_subset_by_feature <span class="ot">&lt;-</span> se_phylum[<span class="fu">rowData</span>(se_phylum)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Actinobacteria&quot;</span>, <span class="st">&quot;Chlamydiae&quot;</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(se_phylum)<span class="sc">$</span>Phylum), ]</span>
<span id="cb60-6"><a href="data-introduction.html#cb60-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-7"><a href="data-introduction.html#cb60-7" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb60-8"><a href="data-introduction.html#cb60-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se_phylum_subset_by_feature)</span></code></pre></div>
<pre><code>## [1]  2 26</code></pre>
<p><strong>Note</strong>: as data was agglomerated, the number of rows equal the
number of Phyla used to index (in this case, just 2)</p>
<p>Alternatively:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="data-introduction.html#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># store features of interest into phyla</span></span>
<span id="cb62-2"><a href="data-introduction.html#cb62-2" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Phylum:Actinobacteria&quot;</span>, <span class="st">&quot;Phylum:Chlamydiae&quot;</span>)</span>
<span id="cb62-3"><a href="data-introduction.html#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature</span></span>
<span id="cb62-4"><a href="data-introduction.html#cb62-4" aria-hidden="true" tabindex="-1"></a>se_phylum_subset_by_feature <span class="ot">&lt;-</span> se_phylum[phyla, ]</span>
<span id="cb62-5"><a href="data-introduction.html#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb62-6"><a href="data-introduction.html#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se_subset_by_feature)</span></code></pre></div>
<pre><code>## [1] 1652   26</code></pre>
<p>The code above returns the not agglomerated version of the data.</p>
<p>Fewer characteristics can be used to subset by feature:</p>
<ul>
<li>Taxonomic rank</li>
<li>Meta-taxonomic group</li>
</ul>
<p>For subsetting by Kingdom, agglomeration does not apply, whereas for
the other ranks it can be applied if necessary.</p>
</div>
</div>
<div id="subset-by-sample-and-feature" class="section level4" number="2.6.2.3">
<h4><span class="header-section-number">2.6.2.3</span> Subset by sample and feature</h4>
<p>Finally, we can subset data by sample and feature at once. The
resulting subset contains all the samples of human origin and all the
features of Phyla “Actinobacteria” or “Chlamydiae.”</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="data-introduction.html#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by sample and feature and get rid of NAs</span></span>
<span id="cb64-2"><a href="data-introduction.html#cb64-2" aria-hidden="true" tabindex="-1"></a>se_subset_by_sample_feature <span class="ot">&lt;-</span> se[<span class="fu">rowData</span>(se)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Actinobacteria&quot;</span>, <span class="st">&quot;Chlamydiae&quot;</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(se)<span class="sc">$</span>Phylum), se<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>)]</span>
<span id="cb64-3"><a href="data-introduction.html#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="data-introduction.html#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show dimensions</span></span>
<span id="cb64-5"><a href="data-introduction.html#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se_subset_by_sample_feature)</span></code></pre></div>
<pre><code>## [1] 1652    9</code></pre>
<p><strong>Note</strong>: the dimensions of <code>se_subset_by_sample_feature</code> agree with
those of the previous subsets (9 columns filtered by sample and 1652
rows filtered by feature).</p>
<p>If a study was to consider and quantify the presence of Actinobacteria
as well as Chlamydiae in different sites of the human body,
<code>se_subset_by_sample_feature</code> might be a suitable subset to start
with.</p>
</div>
</div>
</div>
<div id="additional-reading" class="section level2" number="2.7">
<h2><span class="header-section-number">2.7</span> Additional Reading</h2>
<div id="r-programming-resources" class="section level3" number="2.7.1">
<h3><span class="header-section-number">2.7.1</span> R programming resources</h3>
<ul>
<li>R programming basics: <a href="https://www.rstudio.com/wp-content/uploads/2016/10/r-cheat-sheet-3.pdf">Base R</a></li>
<li>Basics of R programming: <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/master/base-r.pdf">Base R</a></li>
<li><a href="https://www.rstudio.com/resources/cheatsheets/">R cheat sheets</a></li>
<li>R visualization with <a href="https://www.rstudio.com/wp-content/uploads/2016/11/ggplot2-cheatsheet-2.1.pdf">ggplot2</a></li>
<li><a href="http://www.cookbook-r.com/Graphs/">R graphics cookbook</a></li>
</ul>
<p>Rmarkdown</p>
<ul>
<li><a href="https://rmarkdown.rstudio.com/">Rmarkdown tips</a></li>
</ul>
<p>RStudio</p>
<ul>
<li><a href="https://www.rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf">RStudio cheat sheet</a></li>
</ul>
</div>
<div id="resources-for-treesummarizedexperiment" class="section level3" number="2.7.2">
<h3><span class="header-section-number">2.7.2</span> Resources for TreeSummarizedExperiment</h3>
<ul>
<li>SingleCellExperiment
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">Publication</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">Project page</a></li>
</ul></li>
<li>SummarizedExperiment
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">Publication</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">Project page</a></li>
</ul></li>
<li>TreeSummarizedExperiment
<ul>
<li><a href="https://f1000research.com/articles/9-1246">Publication</a></li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html">Project page</a></li>
</ul></li>
</ul>
</div>
<div id="resources-for-phyloseq" class="section level3" number="2.7.3">
<h3><span class="header-section-number">2.7.3</span> Resources for phyloseq</h3>
<ul>
<li><a href="https://microsud.github.io/Tools-Microbiome-Analysis/">List of R tools for microbiome analysis</a></li>
<li><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">phyloseq</a></li>
<li><a href="http://microbiome.github.io/tutorials/">microbiome tutorial</a></li>
<li><a href="https://microsud.github.io/microbiomeutilities/">microbiomeutilities</a></li>
<li>Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses (<a href="https://f1000research.com/articles/5-1492/v2">Callahan et al. F1000, 2016</a>).</li>
</ul>
</div>
<div id="further-reading" class="section level3" number="2.7.4">
<h3><span class="header-section-number">2.7.4</span> Further reading</h3>
<ul>
<li><p><a href="https://datacarpentry.org/R-ecology-lesson/">Data Analysis and Visualization in R for Ecologists</a> by Data Carpentry</p></li>
<li><p><a href="http://web.stanford.edu/class/bios221/book/">Modern Statistics for Modern Biology. Holmes &amp; Huber (2018)</a> for background in statistical analysis</p></li>
<li><p><a href="https://openresearchlabs.github.io/publications/papers/2018-Shetty-Lahti-MDS.pdf">Microbiome Data Science. Shetty &amp; Lahti, 2019</a></p></li>
</ul>
</div>
</div>
<div id="session-info" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] knitr_1.33                     dplyr_1.0.7                   
 [3] phyloseq_1.37.0                mia_1.1.14                    
 [5] TreeSummarizedExperiment_2.1.4 Biostrings_2.61.2             
 [7] XVector_0.33.0                 SingleCellExperiment_1.15.2   
 [9] SummarizedExperiment_1.23.4    Biobase_2.53.0                
[11] GenomicRanges_1.45.0           GenomeInfoDb_1.29.8           
[13] IRanges_2.27.2                 S4Vectors_0.31.3              
[15] BiocGenerics_0.39.2            MatrixGenerics_1.5.4          
[17] matrixStats_0.60.1-9001        BiocStyle_2.21.3              
[19] rebook_1.3.1                  

loaded via a namespace (and not attached):
  [1] systemfonts_1.0.2           plyr_1.8.6                 
  [3] igraph_1.2.6                lazyeval_0.2.2             
  [5] splines_4.1.1               BiocParallel_1.27.6        
  [7] ggplot2_3.3.5               scater_1.21.3              
  [9] digest_0.6.27               foreach_1.5.1              
 [11] htmltools_0.5.2             viridis_0.6.1              
 [13] fansi_0.5.0                 magrittr_2.0.1             
 [15] memoise_2.0.0               ScaledMatrix_1.1.0         
 [17] cluster_2.1.2               DECIPHER_2.21.0            
 [19] svglite_2.0.0               colorspace_2.0-2           
 [21] blob_1.2.2                  rvest_1.0.1                
 [23] ggrepel_0.9.1               xfun_0.25                  
 [25] crayon_1.4.1                RCurl_1.98-1.4             
 [27] jsonlite_1.7.2              graph_1.71.2               
 [29] survival_3.2-13             iterators_1.0.13           
 [31] ape_5.5                     glue_1.4.2                 
 [33] kableExtra_1.3.4            gtable_0.3.0               
 [35] zlibbioc_1.39.0             webshot_0.5.2              
 [37] DelayedArray_0.19.2         BiocSingular_1.9.1         
 [39] Rhdf5lib_1.15.2             scales_1.1.1               
 [41] DBI_1.1.1                   Rcpp_1.0.7                 
 [43] viridisLite_0.4.0           decontam_1.13.0            
 [45] tidytree_0.3.5              bit_4.0.4                  
 [47] rsvd_1.0.5                  httr_1.4.2                 
 [49] dir.expiry_1.1.0            ellipsis_0.3.2             
 [51] pkgconfig_2.0.3             XML_3.99-0.7               
 [53] scuttle_1.3.1               CodeDepends_0.6.5          
 [55] sass_0.4.0                  utf8_1.2.2                 
 [57] tidyselect_1.1.1            rlang_0.4.11               
 [59] reshape2_1.4.4              munsell_0.5.0              
 [61] tools_4.1.1                 cachem_1.0.6               
 [63] cli_3.0.1                   DirichletMultinomial_1.35.0
 [65] generics_0.1.0              RSQLite_2.2.8              
 [67] ade4_1.7-17                 evaluate_0.14              
 [69] biomformat_1.21.0           stringr_1.4.0              
 [71] fastmap_1.1.0               yaml_2.2.1                 
 [73] bit64_4.0.5                 purrr_0.3.4                
 [75] nlme_3.1-153                sparseMatrixStats_1.5.3    
 [77] xml2_1.3.2                  compiler_4.1.1             
 [79] rstudioapi_0.13             beeswarm_0.4.0             
 [81] filelock_1.0.2              treeio_1.17.2              
 [83] tibble_3.1.4                bslib_0.3.0                
 [85] stringi_1.7.4               highr_0.9                  
 [87] lattice_0.20-44             Matrix_1.3-4               
 [89] vegan_2.5-7                 permute_0.9-5              
 [91] multtest_2.49.0             vctrs_0.3.8                
 [93] pillar_1.6.2                lifecycle_1.0.0            
 [95] rhdf5filters_1.5.0          BiocManager_1.30.16        
 [97] jquerylib_0.1.4             BiocNeighbors_1.11.0       
 [99] data.table_1.14.0           bitops_1.0-7               
[101] irlba_2.3.3                 R6_2.5.1                   
[103] bookdown_0.24               gridExtra_2.3              
[105] vipor_0.4.5                 codetools_0.2-18           
[107] MASS_7.3-54                 assertthat_0.2.1           
[109] rhdf5_2.37.3                GenomeInfoDbData_1.2.6     
[111] mgcv_1.8-36                 parallel_4.1.1             
[113] grid_4.1.1                  beachmat_2.9.1             
[115] tidyr_1.1.3                 rmarkdown_2.10             
[117] DelayedMatrixStats_1.15.4   ggbeeswarm_0.6.0           </code></pre>
</div>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="example-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
