<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Demonstration Data | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 3 Demonstration Data | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Demonstration Data | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Demonstration Data | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-introduction.html"/>
<link rel="next" href="taxonomic-information.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="" data-path="introduction-to-bioconductor-classes.html"><a href="introduction-to-bioconductor-classes.html"><i class="fa fa-check"></i>Introduction to Bioconductor Classes</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#community"><i class="fa fa-check"></i><b>1.1</b> Community</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#developers-and-contributors"><i class="fa fa-check"></i><b>1.2</b> Developers and contributors</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.3</b> Acknowledgements</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#package-ecosystem"><i class="fa fa-check"></i><b>1.4</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="introduction.html"><a href="introduction.html#data-container"><i class="fa fa-check"></i><b>1.4.1</b> Data container:</a></li>
<li class="chapter" data-level="1.4.2" data-path="introduction.html"><a href="introduction.html#packages"><i class="fa fa-check"></i><b>1.4.2</b> Packages</a></li>
<li class="chapter" data-level="1.4.3" data-path="introduction.html"><a href="introduction.html#demonstration-data"><i class="fa fa-check"></i><b>1.4.3</b> Demonstration data</a></li>
<li class="chapter" data-level="1.4.4" data-path="introduction.html"><a href="introduction.html#online-tutorials"><i class="fa fa-check"></i><b>1.4.4</b> Online tutorials</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-introduction.html"><a href="data-introduction.html"><i class="fa fa-check"></i><b>2</b> Data Infrastructure</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-introduction.html"><a href="data-introduction.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a></li>
<li class="chapter" data-level="2.2" data-path="data-introduction.html"><a href="data-introduction.html#background"><i class="fa fa-check"></i><b>2.2</b> Background</a></li>
<li class="chapter" data-level="2.3" data-path="data-introduction.html"><a href="data-introduction.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="data-introduction.html"><a href="data-introduction.html#importing-data-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Importing data from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="data-introduction.html"><a href="data-introduction.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-introduction.html"><a href="data-introduction.html#metadata"><i class="fa fa-check"></i><b>2.4</b> Metadata</a></li>
<li class="chapter" data-level="2.5" data-path="data-introduction.html"><a href="data-introduction.html#microbiome-and-tree-data-specific-aspects"><i class="fa fa-check"></i><b>2.5</b> Microbiome and tree data specific aspects</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="data-introduction.html"><a href="data-introduction.html#assays"><i class="fa fa-check"></i><b>2.5.1</b> Assays</a></li>
<li class="chapter" data-level="2.5.2" data-path="data-introduction.html"><a href="data-introduction.html#coldata"><i class="fa fa-check"></i><b>2.5.2</b> colData</a></li>
<li class="chapter" data-level="2.5.3" data-path="data-introduction.html"><a href="data-introduction.html#rowdata"><i class="fa fa-check"></i><b>2.5.3</b> rowData</a></li>
<li class="chapter" data-level="2.5.4" data-path="data-introduction.html"><a href="data-introduction.html#rowtree"><i class="fa fa-check"></i><b>2.5.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="data-introduction.html"><a href="data-introduction.html#data-manipulation"><i class="fa fa-check"></i><b>2.6</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="data-introduction.html"><a href="data-introduction.html#tidy-data"><i class="fa fa-check"></i><b>2.6.1</b> Tidy data</a></li>
<li class="chapter" data-level="2.6.2" data-path="data-introduction.html"><a href="data-introduction.html#subsetting"><i class="fa fa-check"></i><b>2.6.2</b> Subsetting</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="data-introduction.html"><a href="data-introduction.html#additional-reading"><i class="fa fa-check"></i><b>2.7</b> Additional Reading</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="data-introduction.html"><a href="data-introduction.html#r-programming-resources"><i class="fa fa-check"></i><b>2.7.1</b> R programming resources</a></li>
<li class="chapter" data-level="2.7.2" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>2.7.2</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="2.7.3" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>2.7.3</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="2.7.4" data-path="data-introduction.html"><a href="data-introduction.html#further-reading"><i class="fa fa-check"></i><b>2.7.4</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-introduction.html"><a href="data-introduction.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Appendix</b></span></li>
<li class="chapter" data-level="3" data-path="example-data.html"><a href="example-data.html"><i class="fa fa-check"></i><b>3</b> Demonstration Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="example-data.html"><a href="example-data.html#package-data"><i class="fa fa-check"></i><b>3.1</b> Package data</a></li>
<li class="chapter" data-level="3.2" data-path="example-data.html"><a href="example-data.html#experimenthub-data"><i class="fa fa-check"></i><b>3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3" data-path="example-data.html"><a href="example-data.html#other-data-sources"><i class="fa fa-check"></i><b>3.3</b> Other data sources</a></li>
<li class="chapter" data-level="" data-path="example-data.html"><a href="example-data.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>4</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="4.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>4.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>4.1.1</b> dada2</a></li>
<li class="chapter" data-level="4.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>4.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>4.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>4.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>4.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="4.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>4.4</b> Data transformation</a></li>
<li class="chapter" data-level="4.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>4.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>4.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="4.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>4.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>5.1</b> Get top taxa and taxonomy</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="quality-control.html"><a href="quality-control.html#features"><i class="fa fa-check"></i><b>5.1.1</b> Features</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.2</b> Library size</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html"><i class="fa fa-check"></i><b>6</b> Microbiome Exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalence"><i class="fa fa-check"></i><b>6.1</b> Prevalence</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>6.1.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="6.1.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#plotting-prevalence"><i class="fa fa-check"></i><b>6.1.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#abundance"><i class="fa fa-check"></i><b>6.2</b> Abundance</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>7</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#estimating-alpha-diversity"><i class="fa fa-check"></i><b>7.1</b> Estimating alpha diversity</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.3</b> Evenness</a></li>
<li class="chapter" data-level="7.1.4" data-path="alpha-diversity.html"><a href="alpha-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.4</b> Dominance</a></li>
<li class="chapter" data-level="7.1.5" data-path="alpha-diversity.html"><a href="alpha-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.5</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualize-alpha-diversities"><i class="fa fa-check"></i><b>7.2</b> Visualize alpha diversities</a></li>
<li class="chapter" data-level="" data-path="alpha-diversity.html"><a href="alpha-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>8</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="beta-diversity.html"><a href="beta-diversity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-beta-diversity"><i class="fa fa-check"></i><b>8.2</b> Estimating beta diversity</a></li>
<li class="chapter" data-level="8.3" data-path="beta-diversity.html"><a href="beta-diversity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="beta-diversity.html"><a href="beta-diversity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a></li>
<li class="chapter" data-level="8.5" data-path="beta-diversity.html"><a href="beta-diversity.html#community-comparisons-todo-combine-with-the-material-above-for-simplicity"><i class="fa fa-check"></i><b>8.5</b> Community comparisons [TODO combine with the material above for simplicity?]</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="beta-diversity.html"><a href="beta-diversity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.5.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.5.2" data-path="beta-diversity.html"><a href="beta-diversity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.5.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="beta-diversity.html"><a href="beta-diversity.html#further-reading-1"><i class="fa fa-check"></i><b>8.6</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="beta-diversity.html"><a href="beta-diversity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Microbiome Community</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#community-composition"><i class="fa fa-check"></i><b>9.1</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>10</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="community-typing.html"><a href="community-typing.html#community-typing-1"><i class="fa fa-check"></i><b>10.1</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.1.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="10.1.2" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.1.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>10.2</b> Community Detection</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#introduction-to-differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Introduction to differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.1</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.6</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multitable.html"><a href="multitable.html"><i class="fa fa-check"></i><b>12</b> Representation of Multiple Data Tables</a>
<ul>
<li class="chapter" data-level="12.1" data-path="multitable.html"><a href="multitable.html#assay-data"><i class="fa fa-check"></i><b>12.1</b> Assay data</a></li>
<li class="chapter" data-level="12.2" data-path="multitable.html"><a href="multitable.html#alternative-experiments"><i class="fa fa-check"></i><b>12.2</b> Alternative experiments</a></li>
<li class="chapter" data-level="12.3" data-path="multitable.html"><a href="multitable.html#multiassayexperiments"><i class="fa fa-check"></i><b>12.3</b> MultiAssayExperiments</a></li>
<li class="chapter" data-level="" data-path="multitable.html"><a href="multitable.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="function-ref.html"><a href="function-ref.html"><i class="fa fa-check"></i><b>13</b> Function Reference</a>
<ul>
<li class="chapter" data-level="13.1" data-path="function-ref.html"><a href="function-ref.html#objects"><i class="fa fa-check"></i><b>13.1</b> Objects</a></li>
<li class="chapter" data-level="13.2" data-path="function-ref.html"><a href="function-ref.html#reduced-dimension-pca-t-sne-umap-mds-etc"><i class="fa fa-check"></i><b>13.2</b> Reduced dimension (PCA, t-SNE, UMAP, MDS, etc)</a></li>
<li class="chapter" data-level="13.3" data-path="function-ref.html"><a href="function-ref.html#plotting"><i class="fa fa-check"></i><b>13.3</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>14</b> Extras</a>
<ul>
<li class="chapter" data-level="14.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>14.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="example-data" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Demonstration Data</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Open demonstration data for testing and benchmarking purposes is
available from multiple locations. This chapter introduces some
options. The other chapters of this book provide ample examples about
the use of the data.</p>
<div id="package-data" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Package data</h2>
<p>The <code>mia</code> R package contains example data sets that are direct
conversions from the alternative <code>phyloseq</code> container to the
<code>TreeSummarizedExperiment</code> container.</p>
<p>List the <a href="https://microbiome.github.io/mia/reference/index.html">available
datasets</a> in
the <code>mia</code> package:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="example-data.html#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb67-2"><a href="example-data.html#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span></code></pre></div>
<p>Load the <code>GlobalPatterns</code> data from the <code>mia</code> package:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="example-data.html#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb68-2"><a href="example-data.html#cb68-2" aria-hidden="true" tabindex="-1"></a>GlobalPatterns</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Check the documentation for this data set:</p>
<pre><code>## No documentation for &#39;GlobalPatterns&#39; in specified packages and libraries:
## you could try &#39;??GlobalPatterns&#39;</code></pre>
</div>
<div id="experimenthub-data" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> ExperimentHub data</h2>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/ExperimentHub/inst/doc/ExperimentHub.html">ExperimentHub</a>
provides a variety of data resources, including the
<a href="https://bioconductor.org/packages/devel/data/experiment/html/microbiomeDataSets.html">microbiomeDataSets</a>
package.</p>
<p>A table of the available data sets is available through the <code>availableDataSets</code>
function.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="example-data.html#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb71-2"><a href="example-data.html#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">availableDataSets</span>()</span></code></pre></div>
<pre><code>##             Dataset
## 1  GrieneisenTSData
## 2    HintikkaXOData
## 3       LahtiMLData
## 4        LahtiMData
## 5       LahtiWAData
## 6      OKeefeDSData
## 7 SilvermanAGutData
## 8        SongQAData
## 9   SprockettTHData</code></pre>
<p>All data are downloaded from ExperimentHub and cached for local
re-use. Check the <a href="https://microbiome.github.io/microbiomeDataSets/reference/index.html">man pages of each
function</a>
for a detailed documentation of the data contents and references. Let
us retrieve a <em><a href="https://bioconductor.org/packages/3.14/MultiAssayExperiment">MultiAssayExperiment</a></em> data set:</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="example-data.html#cb73-1" aria-hidden="true" tabindex="-1"></a>mae <span class="ot">&lt;-</span> <span class="fu">HintikkaXOData</span>()</span></code></pre></div>
<p>Data is available in <em><a href="https://bioconductor.org/packages/3.14/SummarizedExperiment">SummarizedExperiment</a></em>, <code>r Biocpkg("TreeSummarizedExperiment")</code>, and <code>r Biocpkg("MultiAssayExperiment")</code> data containers; see the separate
page on <a href="https://microbiome.github.io/OMA/multitable.html">alternative
containers</a> for more
details.</p>
</div>
<div id="other-data-sources" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Other data sources</h2>
<p>The
<a href="https://waldronlab.io/curatedMetagenomicData">curatedMetagenomicData</a>
is an independent source that provides various example data sets as
<code>(Tree)SummarizedExperiment</code> objects. This resource provides curated
human microbiome data including gene families, marker abundance,
marker presence, pathway abundance, pathway coverage, and relative
abundance for samples from different body sites. See the package
homepage for more details on data availability and access.</p>
<p>As one example, let us retrieve the Vatanen (2016) <span class="citation">(<a href="#ref-Vatanen2016" role="doc-biblioref">Vatanen et al. 2016</a>)</span> data
set. This is a larger collection with a bit longer download time.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="example-data.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedMetagenomicData)</span>
<span id="cb74-2"><a href="example-data.html#cb74-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">curatedMetagenomicData</span>(<span class="st">&quot;Vatanen*&quot;</span>, <span class="at">dryrun =</span> <span class="cn">FALSE</span>, <span class="at">counts =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="session-info-1" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] microbiomeDataSets_1.1.5       MultiAssayExperiment_1.19.11  
 [3] TreeSummarizedExperiment_2.1.4 Biostrings_2.61.2             
 [5] XVector_0.33.0                 SingleCellExperiment_1.15.2   
 [7] SummarizedExperiment_1.23.4    Biobase_2.53.0                
 [9] GenomicRanges_1.45.0           GenomeInfoDb_1.29.8           
[11] IRanges_2.27.2                 S4Vectors_0.31.3              
[13] BiocGenerics_0.39.2            MatrixGenerics_1.5.4          
[15] matrixStats_0.60.1-9001        BiocStyle_2.21.3              
[17] rebook_1.3.1                  

loaded via a namespace (and not attached):
 [1] nlme_3.1-153                  bitops_1.0-7                 
 [3] bit64_4.0.5                   httr_1.4.2                   
 [5] filelock_1.0.2                tools_4.1.1                  
 [7] bslib_0.3.0                   utf8_1.2.2                   
 [9] R6_2.5.1                      DBI_1.1.1                    
[11] lazyeval_0.2.2                tidyselect_1.1.1             
[13] curl_4.3.2                    bit_4.0.4                    
[15] compiler_4.1.1                graph_1.71.2                 
[17] DelayedArray_0.19.2           bookdown_0.24                
[19] sass_0.4.0                    rappdirs_0.3.3               
[21] stringr_1.4.0                 digest_0.6.27                
[23] rmarkdown_2.10                pkgconfig_2.0.3              
[25] htmltools_0.5.2               dbplyr_2.1.1                 
[27] fastmap_1.1.0                 rlang_0.4.11                 
[29] RSQLite_2.2.8                 shiny_1.6.0                  
[31] jquerylib_0.1.4               generics_0.1.0               
[33] jsonlite_1.7.2                BiocParallel_1.27.6          
[35] dplyr_1.0.7                   RCurl_1.98-1.4               
[37] magrittr_2.0.1                GenomeInfoDbData_1.2.6       
[39] Matrix_1.3-4                  Rcpp_1.0.7                   
[41] fansi_0.5.0                   ape_5.5                      
[43] lifecycle_1.0.0               stringi_1.7.4                
[45] yaml_2.2.1                    zlibbioc_1.39.0              
[47] BiocFileCache_2.1.1           AnnotationHub_3.1.5          
[49] grid_4.1.1                    blob_1.2.2                   
[51] promises_1.2.0.1              parallel_4.1.1               
[53] ExperimentHub_2.1.4           crayon_1.4.1                 
[55] dir.expiry_1.1.0              lattice_0.20-44              
[57] KEGGREST_1.33.0               CodeDepends_0.6.5            
[59] knitr_1.33                    pillar_1.6.2                 
[61] codetools_0.2-18              XML_3.99-0.7                 
[63] glue_1.4.2                    BiocVersion_3.14.0           
[65] evaluate_0.14                 BiocManager_1.30.16          
[67] png_0.1-7                     httpuv_1.6.2                 
[69] vctrs_0.3.8                   treeio_1.17.2                
[71] purrr_0.3.4                   tidyr_1.1.3                  
[73] assertthat_0.2.1              cachem_1.0.6                 
[75] xfun_0.25                     mime_0.11                    
[77] xtable_1.8-4                  tidytree_0.3.5               
[79] later_1.3.0                   tibble_3.1.4                 
[81] AnnotationDbi_1.55.1          memoise_2.0.0                
[83] interactiveDisplayBase_1.31.2 ellipsis_0.3.2               </code></pre>
</div>

</div>
</div>



<h3>Bibliography</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Vatanen2016" class="csl-entry">
Vatanen, Tommi, Aleksandar D. Kostic, Eva d’Hennezel, Heli Siljander, Eric A. Franzosa, Moran Yassour, Raivo Kolde, et al. 2016. <span>“Variation in Microbiome LPS Immunogenicity Contributes to Autoimmunity in Humans.”</span> <em>Cell</em> 165 (May): 842–53. <a href="https://doi.org/10.1016/j.cell.2016.04.007">https://doi.org/10.1016/j.cell.2016.04.007</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="taxonomic-information.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
