<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 19 Exercises | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 19 Exercises | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 19 Exercises | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 19 Exercises | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="resources.html"/>
<link rel="next" href="extras.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#human-microbiome-compendium"><i class="fa fa-check"></i><b>3.3.4</b> Human microbiome compendium</a></li>
<li class="chapter" data-level="3.3.5" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.5</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis with igraph</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-plot"><i class="fa fa-check"></i><b>12.2.1</b> Network plot</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#centrality-measures"><i class="fa fa-check"></i><b>12.2.2</b> Centrality measures</a></li>
<li class="chapter" data-level="12.2.3" data-path="network-learning.html"><a href="network-learning.html#scale-node-sizes-by-centrality"><i class="fa fa-check"></i><b>12.2.3</b> Scale node sizes by centrality</a></li>
<li class="chapter" data-level="12.2.4" data-path="network-learning.html"><a href="network-learning.html#degree-distribution"><i class="fa fa-check"></i><b>12.2.4</b> Degree distribution</a></li>
<li class="chapter" data-level="12.2.5" data-path="network-learning.html"><a href="network-learning.html#clustered-heatmaps"><i class="fa fa-check"></i><b>12.2.5</b> Clustered heatmaps</a></li>
<li class="chapter" data-level="12.2.6" data-path="network-learning.html"><a href="network-learning.html#global-network-measures"><i class="fa fa-check"></i><b>12.2.6</b> Global network measures</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.3</b> Network analysis with NetCoMi</a></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.4</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.4.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.4.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.4.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.5</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.6</b> More about association measures</a></li>
<li class="chapter" data-level="12.7" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.7</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.7.1</b> SparCC</a></li>
<li class="chapter" data-level="12.7.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.7.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.7.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.7.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.7.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.7.4</b> Network plots</a></li>
<li class="chapter" data-level="12.7.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.7.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="exercises" class="section level1 hasAnchor" number="19">
<h1><span class="header-section-number">Chapter 19</span> Exercises<a href="exercises.html#exercises" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Here you can find assignments on different topics.</p>
<p><strong>Tips for exercises:</strong></p>
<ul>
<li>Add comments that explain what each line or lines of code do. This helps you and others understand your code and find bugs. Furthermore, it is easier for you to reuse the code, and it promotes transparency.</li>
<li>Interpret results by comparing them to literature. List main findings, so that results can easily be understood by others without advanced knowledge on data analytics.</li>
<li>Avoid hard-coding. Use variables which get values in the beginning of the pipeline. That way it is easier for you to modify parameters and reuse the code.</li>
</ul>
<div id="basics-of-rbioconductor" class="section level2 hasAnchor" number="19.1">
<h2><span class="header-section-number">19.1</span> Basics of R/Bioconductor<a href="exercises.html#basics-of-rbioconductor" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Bioconductor training material has been contributed to <a href="https://carpentries.org/">Carpentries</a>. You can check the following lessons for basic background of R and Bioconductor.</p>
<ul>
<li><p><a href="https://carpentries-incubator.github.io/bioc-intro/">Introduction to data analysis with R and Bioconductor</a></p></li>
<li><p><a href="https://carpentries-incubator.github.io/bioc-project/">Introduction to the Bioconductor project</a></p></li>
</ul>
</div>
<div id="workflows" class="section level2 hasAnchor" number="19.2">
<h2><span class="header-section-number">19.2</span> Workflows<a href="exercises.html#workflows" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="reproducible-reporting-with-quarto" class="section level3 hasAnchor" number="19.2.1">
<h3><span class="header-section-number">19.2.1</span> Reproducible reporting with Quarto<a href="exercises.html#reproducible-reporting-with-quarto" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The following batch of exercises walks you through typical use cases of Quarto
in RStudio. Before heading to the exercises, it is recommended to read the
<a href="https://quarto.org/docs/tools/rstudio.html">Quarto guidelines for RStudio</a></p>
<div id="new-document" class="section level4 hasAnchor" number="19.2.1.1">
<h4><span class="header-section-number">19.2.1.1</span> New document<a href="exercises.html#new-document" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This exercise gets you started with creating a Quarto document and adding
text to it with typing conventions borrowed from the
<a href="https://quarto.org/docs/authoring/markdown-basics.html">markdown syntax</a>.
Feel free to render the document with the <strong>Render</strong> button after each step to
see the changes in the final report.</p>
<ol style="list-style-type: decimal">
<li>Open RStudio and create a new Quarto file named <code>My first Quarto</code>.</li>
<li>Add the subtitle <code>My first section</code> and write some text of your choice
underneath. You can choose the level of headings by the number of preceding
hashes (<code>#</code>).</li>
<li>Add a subsection named <code>List of items</code> and list three items underneath, both
ordered and unordered. You can initialize items with numbers (<code>1.</code>, <code>2.</code>,
<code>3.</code>, …) or stars (<code>*</code>) for the ordered and unordered case, respectively.</li>
<li>Add another subsection named <code>Link to web</code> and add a clickable link to the
<a href="https://microbiome.github.io/OMA/">OMA book</a>, using the pattern <code>[text](url)</code>.</li>
<li>Render the document and check its appearance</li>
</ol>
<p>Nice start! You are now able to create a Quarto document, understand its syntax
and can render it into a reproducible report. If you got stuck, you can look up
the docs on <a href="https://quarto.org/docs/tools/rstudio.html#creating-documents">creating</a>
and <a href="https://quarto.org/docs/tools/rstudio.html#render-and-preview">rendering</a>
Quarto documents.</p>
</div>
<div id="code-chunks" class="section level4 hasAnchor" number="19.2.1.2">
<h4><span class="header-section-number">19.2.1.2</span> Code chunks<a href="exercises.html#code-chunks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>While customizable text is nothing new by itself, the advantage of Quarto (and
previously Rmakdown) is to combine text with code in R or other programming
languages, so that both the analytical pipeline and verbal description can be
put in one place. In this exercise, we learn how to write and run code in Quarto.</p>
<ol style="list-style-type: decimal">
<li>Open RStudio and create a new Quarto file.</li>
<li>Initialize a code chunk by pressing <code>alt</code> + <code>cmd</code> + <code>i</code> and define the variables
<code>A &lt;- "my name"</code> and <code>B &lt;- 0</code> in it.</li>
<li>Write the text <code>Below is my first code chunk</code> just above the code chunk.</li>
<li>Initialize one more code chunk and add 100 to the variable <code>B</code> in it.</li>
<li>Write the text <code>Below I change variable B</code> just above the second chunk.</li>
<li><strong>Extra</strong>: Write the following line of text: <code>my name is A and I am B years old</code>,
where A and B are variables defined in the code chunks upstream and change if
those variables are modified. Be aware that inline code can be added as
<code>&gt; r my_inline_command</code> (without <code>&gt;</code>).</li>
</ol>
<p>Good job. You are now able to combine text and code in a Quarto document. If you
got stuck, you can refer to the Quarto docs on
<a href="https://quarto.org/docs/visual-editor/technical.html#code-chunks">using code chunks</a>.</p>
</div>
<div id="knitr-options" class="section level4 hasAnchor" number="19.2.1.3">
<h4><span class="header-section-number">19.2.1.3</span> Knitr options<a href="exercises.html#knitr-options" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Code chunks can be greatly customized in terms of visibility and execution,
output size and location and much more. This is possible with the knitr chunk
options, which usually appear at the beginning of the target chunk with the
syntax <code>#| option-name: value</code>, also described
<a href="https://quarto.org/docs/computations/r.html#chunk-options">here</a>.
In this exercise, we explore part of the knitr potential.</p>
<ol style="list-style-type: decimal">
<li>Open RStudio and create a new Quarto file.</li>
<li>Initialize three code chunks and label them as <code>setup</code>, <code>fig-box</code> and
<code>tbl-coldata</code>, respectively. Remember that the name of a chunk can be
specified with the <code>label</code> option.</li>
<li>Write the following code in the corresponding chunk and render the document.</li>
</ol>
<div class="sourceCode" id="cb503"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb503-1"><a href="exercises.html#cb503-1" tabindex="-1"></a><span class="co"># setup</span></span>
<span id="cb503-2"><a href="exercises.html#cb503-2" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb503-3"><a href="exercises.html#cb503-3" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb503-4"><a href="exercises.html#cb503-4" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb503-5"><a href="exercises.html#cb503-5" tabindex="-1"></a></span>
<span id="cb503-6"><a href="exercises.html#cb503-6" tabindex="-1"></a><span class="co"># this line sets some options for all the chunks (global chunk options)</span></span>
<span id="cb503-7"><a href="exercises.html#cb503-7" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">message =</span> <span class="cn">FALSE</span>, <span class="at">warning =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb504"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb504-1"><a href="exercises.html#cb504-1" tabindex="-1"></a><span class="co"># fig-box</span></span>
<span id="cb504-2"><a href="exercises.html#cb504-2" tabindex="-1"></a><span class="fu">boxplot</span>(<span class="fu">colSums</span>(<span class="fu">assay</span>(tse)) <span class="sc">~</span> tse<span class="sc">$</span>SampleType)</span></code></pre></div>
<div class="sourceCode" id="cb505"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb505-1"><a href="exercises.html#cb505-1" tabindex="-1"></a><span class="co"># tbl-coldata</span></span>
<span id="cb505-2"><a href="exercises.html#cb505-2" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(<span class="fu">colData</span>(tse)))</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Set <code>include: false</code> in the <code>setup</code> chunk, <code>fig-width: 10</code> in the <code>fig-box</code>
chunk and <code>echo: false</code> in the <code>tbl-coldata</code> chunk. Render the document again
and find the differences from before.</li>
<li>Add the options <code>fig-cap</code> and <code>tab-cap</code> to the <code>fig-box</code> and <code>tbl-coldata</code>
chunks, respectively. They require some text input, which makes for the
caption of the figures or tables.</li>
<li><strong>Extra</strong>: Create a cross-reference to <code>fig-box</code> and <code>tbl-coldata</code> in the
text above the respective code chunk. You can do that with the syntax
<code>@chunk-name</code>.</li>
<li><strong>Extra</strong>: Define a custom folder for storing figures with <code>fig-path</code>. Insert
it in <code>knitr::opts_chunk$set</code>, so that it applies globally to all the figures
generated in the document.</li>
</ol>
<p>Congratulations! You are now familiar with the great potential and flexibility
of knitr chunk options. An exhaustive list of available options can be found
in the <a href="https://yihui.org/knitr/options/">knitr documentation</a>.</p>
</div>
<div id="yaml-instructions" class="section level4 hasAnchor" number="19.2.1.4">
<h4><span class="header-section-number">19.2.1.4</span> YAML instructions<a href="exercises.html#yaml-instructions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The box at the beginning of every Quarto document contains yaml options that let
you define the metadata of the document. They will affect the appearance of the
document when it is rendered. By default, the box includes yaml options for the
title, format and editor to be used, but much more information on layout, code
execution and figures can be specified. A comprehensive list of yaml options is
available <a href="https://quarto.org/docs/reference/formats/html.html">here</a>. In this
exercise, we will get a tiny taste of such functionality.</p>
<ol style="list-style-type: decimal">
<li>Open RStudio and create a new Quarto file.</li>
<li>In the yaml box at the beginning of the document, change the title from
<code>Untitled</code> to <code>My first Quarto</code>.</li>
<li>In the same box, add the two lines <code>author</code> and <code>date</code> followed by your name
and today’s date, respectively.</li>
<li>Render the document and check its appearance.</li>
<li><strong>Extra</strong>: Set <code>toc: true</code> to produce a table of contents. This line should
follow <code>format</code> and <code>html</code> at the second level of indentation.</li>
</ol>
<p>Well done! Now you are able to specify yaml options and understand how they
affect your Quarto document. If you got stuck, you can check
<a href="https://quarto.org/docs/tools/rstudio.html#yaml-intelligence">this section</a> of
the Quarto documentation.</p>
</div>
<div id="quarto-parameters" class="section level4 hasAnchor" number="19.2.1.5">
<h4><span class="header-section-number">19.2.1.5</span> Quarto parameters<a href="exercises.html#quarto-parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>An advanced feature of Quarto consists of execution parameters, which are
externally pre-defined variables that are also accessible in the Quarto document.
They can be specified in the yaml box as <code>params</code>. Here we learn how to use them.</p>
<ol style="list-style-type: decimal">
<li>Open RStudio and create a new Quarto file.</li>
<li>In the yaml box at the beginning of the document, add a line named <code>params</code>
followed by an indented line with <code>gamma: 10</code></li>
<li>Initialize a code chunk and type <code>str(params$gamma)</code> in it.</li>
<li>Render the document and check what happened.</li>
<li>Define one more parameter <code>beta: 3</code> and multiply <code>gamma</code> by <code>beta</code> in a code
chunk below.</li>
<li>Render the document again and check what happened.</li>
</ol>
<p>Well done! You can now use an advanced feature of Quarto such as parameters.
If you got stuck, <a href="https://quarto.org/docs/computations/parameters.html#knitr">here</a>
you can find more information about parameter definition and usage.</p>
</div>
</div>
</div>
<div id="data-containers-treese" class="section level2 hasAnchor" number="19.3">
<h2><span class="header-section-number">19.3</span> Data containers: TreeSE<a href="exercises.html#data-containers-treese" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>TreeSE containers represent the working unit of the mia package. In the
following exercises we learn how to construct, explore and work with them.
A few demo datasets can be imported with mia and can be accessed as explained
in chapter <a href="containers.html#example-data">3.3</a>.</p>
<div id="construct-TreeSE" class="section level3 hasAnchor" number="19.3.1">
<h3><span class="header-section-number">19.3.1</span> Constructing a data object<a href="exercises.html#construct-TreeSE" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we cover how to construct a TreeSE from CSV files, using the components
of OKeefeDSData from the microbiomeDataSets package as an example dataset.</p>
<ol style="list-style-type: decimal">
<li>Fetch or download the files in <a href="https://github.com/microbiome/data/tree/main/OKeefeDSData">this directory</a>.</li>
<li>Read in the csv files with <code>read.csv</code> and store them into the variables
<code>assays</code>, <code>rowdata</code> and <code>coldata</code>, respectively.</li>
<li>Create a TreeSE from the individual components with
<code>TreeSummarizedExperiment</code>. Note that the function requires three arguments:
assays, rowData and colData, to which you can give the appropriate item.</li>
<li>Check that importing is done correctly. E.g., choose random samples and
features, and check that their values equal between raw files and TreeSE.</li>
</ol>
<p>Usefuls functions: DataFrame, TreeSummarizedExperiment, matrix, rownames, colnames, SimpleList</p>
</div>
<div id="importing-data" class="section level3 hasAnchor" number="19.3.2">
<h3><span class="header-section-number">19.3.2</span> Importing data<a href="exercises.html#importing-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Raw data of different types can be imported as a TreeSE with a number of
functions explained in chapter <a href="containers.html#import-from-file">3.4.2</a>. You can also check the
<a href="https://microbiome.github.io/mia/reference/index.html">function reference in the mia package</a>.</p>
<ol style="list-style-type: decimal">
<li>Get familiar with the
<a href="https://github.com/microbiome/data">microbiome data repository</a>
and read the instructions in its
<a href="https://github.com/microbiome/data#training-microbiome-datasets">README</a>
to import and construct datasets from there.</li>
<li>Import data from another format (functions: loadFromMetaphlan | loadFromMothur | loadFromQIIME2 | makeTreeSummarizedExperimentFromBiom | makeTreeSummarizedExperimentFromDADA2 …)</li>
<li>Try out conversions between TreeSE and phyloseq data containers (makeTreeSummarizedExperimentFromPhyloseq; makephyloseqFromTreeSummarizedExperiment)</li>
</ol>
</div>
<div id="preliminary-exploration" class="section level3 hasAnchor" number="19.3.3">
<h3><span class="header-section-number">19.3.3</span> Preliminary exploration<a href="exercises.html#preliminary-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Get a summary about the TreeSE with <code>summary</code>. What is the mean count across
samples? How many features recur only once (singletons)?</li>
<li>Check the dimensions of the TreeSE with <code>dim</code> or alternatively with <code>nrow</code>
and <code>ncol</code>. How many samples and features are present?</li>
<li>List sample and features names with <code>rownames</code> and <code>colnames</code>.</li>
<li>Check what information about samples and features is contained by the
<code>colData</code> and <code>rowData</code> of the TreeSE with <code>names</code>.</li>
<li><strong>Extra</strong>: Calculate the number of unique taxa for each taxonomic rank. You
can use <code>apply</code> to count unique elements for each column of rowData.</li>
</ol>
</div>
<div id="assay-retrieval" class="section level3 hasAnchor" number="19.3.4">
<h3><span class="header-section-number">19.3.4</span> Assay retrieval<a href="exercises.html#assay-retrieval" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>List the names of all available assays with <code>assayNames</code>.</li>
<li>Fetch the list of assays with <code>assays</code>.</li>
<li>Retrieve the first assay of the TreeSE with <code>assay</code>, where the second
argument can be either the name or the index of the desired assay.</li>
</ol>
<p>Well done! You can now locate and retrieve individual assays of a TreeSE. If you
got stuck, you can refer to chapter <a href="containers.html#assay-slot">3.2.1</a> of this book.</p>
</div>
<div id="sample-information" class="section level3 hasAnchor" number="19.3.5">
<h3><span class="header-section-number">19.3.5</span> Sample information<a href="exercises.html#sample-information" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Check the names of the samples with <code>colnames</code>.</li>
<li>List the information on samples available in colData with <code>names</code>.</li>
<li>Visualize the colData with <code>View</code> and briefly look at the information stored
in the different columns.</li>
<li>Get the abundances of all features for a specific sample, such as <code>ID34</code>, for an <em>assay</em> of your choice.</li>
</ol>
</div>
<div id="feature-information" class="section level3 hasAnchor" number="19.3.6">
<h3><span class="header-section-number">19.3.6</span> Feature information<a href="exercises.html#feature-information" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Check the names of the features with <code>rownames</code>.</li>
<li>List the information on features available in rowData with <code>names</code>.</li>
<li>Visualize the rowData with <code>View</code> and briefly look at the information stored
in the different columns.</li>
<li>Get the abundances for a specific feature, such as <code>OTU1810</code>, in all the
samples. You can access feature-specific abundances for an <em>assay</em> of your choice.</li>
<li><strong>Extra</strong>: Create a taxonomy tree based on the taxonomy mappings with
<code>addTaxonomyTree</code> and display its content with <code>taxonomyTree</code> and <code>ggtree</code>.</li>
</ol>
<p>If you got stuck, you can look up chapters @fref{datamanipulation}
and <a href="taxonomic-information.html#fly-tree">6.2.1</a> on how to pick specific abundances and generate
row trees, respectively.</p>
</div>
<div id="other-elements" class="section level3 hasAnchor" number="19.3.7">
<h3><span class="header-section-number">19.3.7</span> Other elements<a href="exercises.html#other-elements" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Try to extract some of the other TreeSE elements listed in chapter <a href="containers.html#containers">3</a>.
However, such data are not always included.</p>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Fetch the metadata of the TreeSE. Is there any iformation available?</li>
<li>Access the phylogenetic tree with <code>rowTree</code>. How big is it in terms of tips
and nodes. If you like you can visualize it with <code>ggtree</code>.</li>
<li>Check if a sample tree is available with <code>colTree</code>, which is suitable for
hierarchical or nested study designs.</li>
<li>If present, obtain the information on feature DNA sequences from the DNA
sequence slot.</li>
</ol>
</div>
</div>
<div id="data-manipulation" class="section level2 hasAnchor" number="19.4">
<h2><span class="header-section-number">19.4</span> Data manipulation<a href="exercises.html#data-manipulation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="subsetting-1" class="section level3 hasAnchor" number="19.4.1">
<h3><span class="header-section-number">19.4.1</span> Subsetting<a href="exercises.html#subsetting-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Subset the TreeSE object to specific samples</li>
<li>Subset the TreeSE object to specific features</li>
<li>Subset the TreeSE object to specific samples and features</li>
</ol>
</div>
<div id="library-sizes" class="section level3 hasAnchor" number="19.4.2">
<h3><span class="header-section-number">19.4.2</span> Library sizes<a href="exercises.html#library-sizes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Calculate library sizes</li>
<li>Subsample / rarify the counts (see: subsampleCounts)</li>
</ol>
<p>Useful functions: nrow, ncol, dim, summary, table, quantile, unique, addPerCellQC, mergeFeaturesByRank</p>
</div>
<div id="prevalent-and-core-taxonomic-features" class="section level3 hasAnchor" number="19.4.3">
<h3><span class="header-section-number">19.4.3</span> Prevalent and core taxonomic features<a href="exercises.html#prevalent-and-core-taxonomic-features" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Estimate prevalence for your chosen feature (row, taxonomic group)</li>
<li>Identify all prevalent features and subset the data accordingly</li>
<li>Report the thresholds and the dimensionality of the data before and after subsetting</li>
<li>Visualize prevalence</li>
</ol>
<p>Useful functions: getPrevalence, getPrevalentFeatures, subsetByPrevalentFeatures</p>
</div>
<div id="data-exploration" class="section level3 hasAnchor" number="19.4.4">
<h3><span class="header-section-number">19.4.4</span> Data exploration<a href="exercises.html#data-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Summarize sample metadata variables. (How many age groups, how they are distributed? 0%, 25%, 50%, 75%, and 100% quantiles of library size?)</li>
<li>Create two histograms. Another shows the distribution of absolute counts, another shows how CLR transformed values are distributed.</li>
<li>Visualize how relative abundances are distributed between taxa in samples.</li>
</ol>
<p>Useful functions: nrow, ncol, dim, summary, table, quantile, unique, transformAssay, ggplot, wilcox.test, mergeFeaturesByRank, plotAbundance</p>
</div>
<div id="other-functions" class="section level3 hasAnchor" number="19.4.5">
<h3><span class="header-section-number">19.4.5</span> Other functions<a href="exercises.html#other-functions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Merge data objects (merge, mergeSEs)</li>
<li>Melt the data for visualization purposes (meltAssay)</li>
</ol>
</div>
<div id="assay-transformation" class="section level3 hasAnchor" number="19.4.6">
<h3><span class="header-section-number">19.4.6</span> Assay transformation<a href="exercises.html#assay-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Transform the counts assay into relative abundances with <code>transformAssay</code> and
store it into the TreeSE as an assay named <code>relabund</code> (see chapter <a href="taxonomic-information.html#assay-transform">6.4</a>).</li>
<li>Similarly, perform a clr transformation on the counts assay with a <code>pseudocount</code>
of 1 and add it to the TreeSE as a new assay.</li>
<li>List the available assays by name with <code>assays</code>.</li>
<li>Access the clr assay and select a subset of its first 100 features and 10
samples. Remember that assays are subsettable with <code>assay[row_idx, col_idx]</code>.</li>
<li>Take the same subset from the TreeSE, and check how this affects the individual
transformed assays. TreeSE can also be subsetted with <code>tse[row_idx, col_idx]</code>.</li>
<li><strong>Extra</strong>: If the data has phylogenetic tree, perform the phILR transformation.</li>
</ol>
</div>
</div>
<div id="abundance-tables" class="section level2 hasAnchor" number="19.5">
<h2><span class="header-section-number">19.5</span> Abundance tables<a href="exercises.html#abundance-tables" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="taxonomic-levels" class="section level3 hasAnchor" number="19.5.1">
<h3><span class="header-section-number">19.5.1</span> Taxonomic levels<a href="exercises.html#taxonomic-levels" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load one of the example data sets mentioned in Chapter 3.3
with <code>data</code> (you need one with taxonomic information at Phylum level)
and store it into a variable named <code>tse</code>.</li>
<li>List the available taxonomic ranks in the data with <code>taxonomyRanks</code>.</li>
<li>Agglomerate the data to Phylum level with <code>mergeFeaturesByRank</code> and the
appropriate value for <code>Rank</code>.</li>
<li>Report the dimensions of the TreeSE before and after agglomerating. You can
use <code>dim</code> for that.</li>
<li><strong>Extra</strong>: Perform CLR transformation on the data. Does this affect agglomeration?</li>
<li><strong>Extra</strong>: List full taxonomic information for a few selected taxa, such as
<code>OTU1</code> and <code>OTU1368</code>. For that you can use <code>mapTaxonomy</code> on a specific subset
of the TreeSE.</li>
</ol>
</div>
<div id="alternative-experiments" class="section level3 hasAnchor" number="19.5.2">
<h3><span class="header-section-number">19.5.2</span> Alternative experiments<a href="exercises.html#alternative-experiments" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load one of the example data sets mentioned in Chapter 3.3
with <code>data</code> (you need one with taxonomic information) and store it into
a variable named <code>tse</code>.</li>
<li>Check the taxonomic ranks of the features with <code>taxonomyRanks</code>. What is the
deepest taxonomic rank available?</li>
<li>Agglomerate the TreeSE to each taxonomic rank and store the resulting
experiments as altExps. This can be performed automatically with <code>splitByRanks</code>.</li>
<li>Check the names of the generated altExps with <code>altExpNames</code> and retrieve a
complete list with <code>altExps</code>.</li>
<li>Retrieve the data agglomerated by genus from the corresponding <code>altExp</code>.
As for assays, you can access the desired altExp by name or index.</li>
<li><strong>Extra</strong>: Split the data based on other features with <code>splitOn</code>.</li>
</ol>
</div>
</div>
<div id="community-alpha-diversity" class="section level2 hasAnchor" number="19.6">
<h2><span class="header-section-number">19.6</span> Community (alpha) diversity<a href="exercises.html#community-alpha-diversity" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="estimation-1" class="section level3 hasAnchor" number="19.6.1">
<h3><span class="header-section-number">19.6.1</span> Estimation<a href="exercises.html#estimation-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate multiple alpha diversity indices with <code>estimateDiversity</code> without
any additional arguments.</li>
<li>Check the names of colData with <code>names</code>. Can you identify which columns
contain the alpha diversity indices?</li>
<li><strong>Extra</strong>: Agglomerate the TreeSE by phylum and compare the mean Shannon
diversity of the original experiment with its agglomerated version. You can
use <code>mergeFeaturesByRank</code> to perform agglomeration and <code>mean</code> to calculate the
mean values of the respective columns in colData.</li>
</ol>
</div>
<div id="visualization-1" class="section level3 hasAnchor" number="19.6.2">
<h3><span class="header-section-number">19.6.2</span> Visualization<a href="exercises.html#visualization-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia and scater packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate Shannon diversity index and Faith’s phylogenetic diversity with
<code>estimateDiversity</code> and the appropriate arguments for <code>index</code>.</li>
<li>Make a boxplot of Shannon diversity on the y axis and sample type on the x
axis with <code>plotColData</code>.</li>
<li>Repeat the previous point with Faith’s phylogenetic diversity and compare the
sample distributions of the two alpha diversity indices. How greatly do they
differ?</li>
<li><strong>Extra</strong>: Make a scatterplot of Shannon diversity on the y axis and Faith’s
phylogenetic diversity on the x axis with <code>plotColData</code>. Colour the points
by sample type with the appropriate optional argument.</li>
</ol>
</div>
<div id="correlation" class="section level3 hasAnchor" number="19.6.3">
<h3><span class="header-section-number">19.6.3</span> Correlation<a href="exercises.html#correlation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia and scater packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate coverage and Shannon diversity index with <code>estimateDiversity</code> and
the appropriate arguments for <code>index</code>.</li>
<li>Test the correlation between the two indices with <code>cor.test</code>. Remember that
colData parameters are accessible with <code>tse$param_name</code>. Use Kendall tau
coefficients as <code>method</code> to measure correlation. Is the correlation
weak or strong, significant or not?</li>
<li>Make a scatterplot of Shannon diversity index on the y axis and coverage on
the x axis. You can do that with <code>plotColData</code>. How do the two indices
relate to one another?</li>
<li><strong>Extra</strong>: Compute the library size of the samples by applying <code>colSums</code> to
the counts assay of the TreeSE, and test the correlation of library size with
Shannon diversity or coverage. Which index is more correlated with library size?</li>
</ol>
<p>In this example, we inspected the correlation between two related variables,
also known as multicollinearity, and checked the correlation to library size,
which is part of quality control. However, the correlation between alpha diversity
and other numerical data about samples, such as participant’s age and weight, also
represent an important analysis in several studies.</p>
</div>
<div id="differences-between-groups" class="section level3 hasAnchor" number="19.6.4">
<h3><span class="header-section-number">19.6.4</span> Differences between groups<a href="exercises.html#differences-between-groups" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate the Gini-Simpson diversity with <code>estimateDiversity</code> and the
appropriate argument for <code>index</code>. Set <code>name</code> to <code>simpson</code>. You will use this
name to access the diversity index from colData.</li>
<li>Inspect the Diet column in the colData. Determine how the samples are grouped
in terms of diet. You can see the number of unique elements in a column with
<code>unique</code>.</li>
<li>Test differences in Gini-Simpson diversity between different diets with
<code>kruskal.test</code>. Remember that colData parameters are accessible with
<code>tse$param_name</code>.</li>
<li>Is diversity significantly different between vegan and mixed diet? To
visualize that, make a boxplot of Gini-Simpson diversity on the y axis and
diet on the x axis with <code>plotColData</code>.</li>
<li><strong>Extra</strong>: Repeat points 3 through 5, this time for age groups. Make sure
that you are using an appropriate statistical test for the number of groups
and distribution.</li>
</ol>
</div>
</div>
<div id="community-similarity-1" class="section level2 hasAnchor" number="19.7">
<h2><span class="header-section-number">19.7</span> Community similarity<a href="exercises.html#community-similarity-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="reduced-dimensions-retrieval" class="section level3 hasAnchor" number="19.7.1">
<h3><span class="header-section-number">19.7.1</span> Reduced dimensions retrieval<a href="exercises.html#reduced-dimensions-retrieval" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia package, load enterotype with <code>data</code> and store it into a
variable named <code>tse</code>.</li>
<li>List all available reduced dimensions with <code>reducedDims</code>. At this point, no
reducedDims are likely found, because we haven’t created any yet.</li>
<li>Perform PCA and store its output in the TreeSE by running
<code>tse &lt;- runPCA(tse, assay.type = "counts")</code>. Note that it is required to
specify the assay on which dimensionality reduction should be conducted.</li>
<li>View the names of all available reduced dimensions with <code>reducedDimNames</code>.
Has something new appeared?</li>
<li><strong>Extra</strong>: Access the <code>PCA</code> reducedDim object with <code>reducedDim</code> and explore
its content. How are the different dimensions stored? Try to extract an
array with only the values from the second dimension by indexing the object
with <code>[ , 2]</code>.</li>
</ol>
</div>
<div id="visualization-basics-with-pca" class="section level3 hasAnchor" number="19.7.2">
<h3><span class="header-section-number">19.7.2</span> Visualization basics with PCA<a href="exercises.html#visualization-basics-with-pca" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia and scater packages, load enterotype with <code>data</code> and store it
into a variable named <code>tse</code>.</li>
<li>Perform a 3-component PCA based on the <code>counts</code> assay. You can use <code>runPCA</code>
and set the optional arguments <code>ncomponents</code> and <code>assay.type</code> to the
appropriate values.</li>
<li>Plot the first two dimensions of PCA with <code>plotReducedDim</code>, to which you
should give the appropriate reducedDim name as the second argument. Note that
by default only the first two dimensions are shown.</li>
<li>Check which information is stored in the ColData of the TreeSE. What would
be worth visualizing in our coordination plot?</li>
<li>Make the same plot again, but this time colour the observations by Enterotype.
You can do that by setting <code>colour_by</code> to the appropriate colname in the
colData of the TreeSE.</li>
<li><strong>Extra</strong>: Plot all three dimensions of PCA with <code>plotReducedDim</code> and the
optional argument <code>ncomponents</code>. Colour observations by Enterotype. Which
pair of dimensions best explains the variance between Enterotypes?</li>
</ol>
</div>
<div id="principal-coordinate-analysis-pcoa" class="section level3 hasAnchor" number="19.7.3">
<h3><span class="header-section-number">19.7.3</span> Principal Coordinate Analysis (PCoA)<a href="exercises.html#principal-coordinate-analysis-pcoa" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>PCoA turns out to be particularly relevant for microbiome analysis, because
unlike PCA it can generate reduced dimensions from distances other than
Euclidean. There are several ecological distances to choose from and you can
find many of them under methods in the vignettes of <code>vegan::vegdist</code>.</p>
<ol style="list-style-type: decimal">
<li>Import the mia and scater packages, load enterotype with <code>data</code> and store it
into a variable named <code>tse</code>.</li>
<li>Transform the counts assay to relative abundances with <code>transformAssay</code>.</li>
<li>Perform a Multi-Dimensional Scaling (MDS) based on the relative abundance
assay in terms of Bray-Curtis dissimilarity. You can use <code>runMDS</code> with the
compulsory argument <code>FUN = vegan::vegdist</code>.</li>
<li>Plot the first two dimensions of PCA with <code>plotReducedDim</code>, to which you
should give the appropriate reducedDim name as the second argument. Colour
the observations by Enterotype with <code>colour_by</code>.</li>
<li><strong>Extra</strong>: Perform MDS again with <code>runMDS</code>, but this time use Jaccard
dissimilarity. The distance metric to use can be defined with the optional
argument <code>method</code>, choosing from the methods in <code>?vegan::vegdist</code>. If you
don’t want to overwrite the reducedDim object made in point 3, set <code>name</code> to
a name of your choice. Visualize and compare it to the plot from point 4.</li>
</ol>
<p>Good job! You are now able to produce and visualize reduced dimensions of a
TreeSE. <code>runMDS</code> is actually one of several algorithms for PCoA and dimensionality
reduction, which you can find in section <a href="community-similarity.html#other-ord-methods">8.1.2</a>.</p>
</div>
<div id="permanova-analysis" class="section level3 hasAnchor" number="19.7.4">
<h3><span class="header-section-number">19.7.4</span> PERMANOVA analysis<a href="exercises.html#permanova-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this exercise we focus on studying the weight of variables on the microbiome
composition. Significance of each variable on beta diversity is tested with
PERMANOVA (point 4) and the homogeneity assumption is also be controlled with a
PERMDISP2 analysis (point 5).</p>
<ol style="list-style-type: decimal">
<li>Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Transform the <code>counts</code> assay into relative abundances with <code>transformAssay</code>.</li>
<li>Extract the relative abundance assay, transpose it and save it into
a variable named <code>relabund_assay</code>.</li>
<li>Perform PERMANOVA with <code>adonis2</code> to see how much Diet can explain the relative
abundance assay (<code>formula = relabund_assay ~ Diet</code>) in terms of Bray-Curtis
dissimilarity (<code>method = "bray"</code>). Also set <code>data = colData(tse)</code>
<code>by = "margin"</code> and <code>permutations = 99</code>. What do the results tell you about
Diet with respect to beta diversity?</li>
<li><strong>Extra</strong>: Test homogeneity of distribution across Diet groups with
<code>anova(betadisper(my_mat), my_groups</code>, where <code>my_mat</code> is the Bray-Curtis
dissimilarity matrix of <code>relabund_assay</code> calculated with
<code>vegdist(relabund_assay, "bray")</code> and <code>my_groups</code> is the vector of Diet values
obtained from the colData of the TreeSE.</li>
</ol>
<p>Well done! You went through testing the effect and significance of Diet on beta
diversity. Keep in mind that the formula fed to <code>adonis2</code> can take more than one
independent variable, so that you can also (and very often should) include
covariates of your studies.</p>
</div>
<div id="redundancy-analysis-rda" class="section level3 hasAnchor" number="19.7.5">
<h3><span class="header-section-number">19.7.5</span> Redundancy analysis (RDA)<a href="exercises.html#redundancy-analysis-rda" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we apply RDA, an ordination method that provides dimensions with
the largest variation between the data based in the light of the specified
variables (point 3). The results of RDA are usually assessed with PERMANOVA
(point 5) and the homogeneity assumption should be checked as in the previous
exercise. This is a relatively complex procedure, but the way this is broken
down into steps below will hopefully make more sense.</p>
<ol style="list-style-type: decimal">
<li>Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Transform the <code>counts</code> assay into relative abundances with <code>transformAssay</code>.</li>
<li>Perform RDA with <code>calculateRDA</code> to see how much Diet can explain the relative
abundance assay (<code>formula = assay ~ Diet</code> and <code>assay.type = relabundance</code>) in
terms of Bray-Curtis dissimilarity (<code>method = "bray"</code>).</li>
<li>Extract the RDA dimensions from the appropriate reducedDim slot with
<code>attr(reducedDim(tse, "RDA"), "rda)</code> and store it into <code>rda</code>.</li>
<li>Test the effect and significance of Diet on beta diversity by PERMANOVA with
<code>anova.cca</code>. Feed this function with <code>rda</code> and set <code>by = "margin"</code> and
<code>permutations = 99</code>, respectively. What do the results tell you about Diet?</li>
<li><strong>Extra</strong>: Check what other parameters are stored in the colData of peerj13075,
add them to the formula (<code>formula = assay ~ Diet + ...</code>) of <code>calculateRDA</code>
and proceed to see how that changes the results of PERMANOVA.</li>
</ol>
<p>Well done! You went through an RDA analysis followed by significance testing
with PERMANOVA and BETADISPER2. In the next exercise we’ll go deeper quantify
the contributions to beta diversity.</p>
</div>
<div id="beta-diversity-analysis" class="section level3 hasAnchor" number="19.7.6">
<h3><span class="header-section-number">19.7.6</span> Beta diversity analysis<a href="exercises.html#beta-diversity-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This exercise prompts you to implement a workflow with distance-based RDA
(dbRDA). You can refer to chapter <a href="community-similarity.html#dbrda-workflow">8.3.1</a> for a step-by-step
walkthrough, which may be simplified in the future.</p>
<ol style="list-style-type: decimal">
<li>Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Create dbRDA with Bray-Curtis dissimilarities on relative abundances. Use PERMANOVA. Can differences between samples be explained with variables of sample meta data?</li>
<li>Analyze diets’ association on beta diversity. Calculate dbRDA and then PERMANOVA. Visualize coefficients. Which taxa’s abundances differ the most between samples?</li>
<li>Interpret your results. Is there association between community composition and location? What are those taxa that differ the most; find information from literature.</li>
</ol>
<p>Useful functions: runMDS, runRDA, anova.cca, transformAssay, mergeFeaturesByRank, ggplot, plotReducedDim, vegan::adonis2</p>
</div>
</div>
<div id="differential-abundance-1" class="section level2 hasAnchor" number="19.8">
<h2><span class="header-section-number">19.8</span> Differential abundance<a href="exercises.html#differential-abundance-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="standard-analysis-with-aldex2" class="section level3 hasAnchor" number="19.8.1">
<h3><span class="header-section-number">19.8.1</span> Standard analysis with ALDEx2<a href="exercises.html#standard-analysis-with-aldex2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia and ALDEx2 packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Agglomerate the TreeSE by genus and filter by a prevalence of 10%. You can
perform both operations in one go with <code>subsetByPrevalentFeatures</code> by specifying
the <code>rank</code> and <code>prevalence</code> arguments.</li>
<li>Model the counts assay of the TreeSE with <code>aldex.clr</code> and store it into the
variable <code>x</code>. As a second argument, provide the grouping variable <code>Diet</code>,
which is contained in a column of the colData.</li>
<li>Feed <code>x</code> to the functions <code>aldex.ttest</code> to erform t-test and to <code>aldex.effect</code>
to estimate effect sizes. Store the output into <code>x_tt</code> and <code>x_effect</code>,
respectively.</li>
<li>Create a data.frame named <code>aldex_out</code> which includes both <code>x_tt</code> and <code>x_effect</code>
and filter for the features with <code>wi.eBH &lt; 0.05</code>. Are there any significantly
differential abundance taxa?</li>
<li><strong>Extra</strong>: If these results appear boring, repeat steps 1 - 5, but use
<code>Gender</code> or <code>Age</code> as the grouping variable. Do we have any better luck with
Gender? What is the problem with Age?</li>
</ol>
</div>
<div id="control-daa-confounders" class="section level3 hasAnchor" number="19.8.2">
<h3><span class="header-section-number">19.8.2</span> Controlling for confounders<a href="exercises.html#control-daa-confounders" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Import the mia and MicrobiomeStat packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Agglomerate the TreeSE by genus and filter by a prevalence of 10%. You can
perform both operations in one go with <code>subsetByPrevalentFeatures</code> by specifying
the <code>rank</code> and <code>prevalence</code> arguments.</li>
<li>Model the counts assay of the TreeSE with <code>linda</code> and store the output into
a variable named <code>linda_out</code>. Provide the colData converted into a data.frame
(with <code>as.data.frame</code>) as the second argument, and a <code>formula</code> with the Age,
Gender and Diet as variables. For example, <code>formula = "~ A + B"</code> represents a
formula with variables A and B.</li>
<li>Extract the <code>output$AgeElderly</code> object from <code>linda_out</code> with <code>$</code> and store it
into a variable named <code>linda_res</code>.</li>
<li>Filter <code>linda_res</code> for features with <code>reject == TRUE</code>. How many differentially
abundant taxa were found? What are their names and how significant are they
in terms of log-fold change and adjusted p-value?</li>
</ol>
</div>
<div id="compare-daa-methods" class="section level3 hasAnchor" number="19.8.3">
<h3><span class="header-section-number">19.8.3</span> Comparing methods<a href="exercises.html#compare-daa-methods" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here, we conduct DAA with identical parameters as in the
<a href="exercises.html#control-daa-confounders">previous exercise</a>, but with a different method,
namely ZicoSeq. We aim to compare the results between these two methods and draw better informed conclusions from such comparative approach.</p>
<ol style="list-style-type: decimal">
<li>Import the mia and GUniFrac packages, load any of the example data sets mentioned in Chapter 3.3
with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Agglomerate the TreeSE by genus and filter by a prevalence of 10%. You can
perform both operations in one go with <code>subsetByPrevalentFeatures</code> by specifying
the <code>rank</code> and <code>prevalence</code> arguments.</li>
<li>Model the counts assay of the TreeSE with <code>ZicoSeq</code> as the <code>feature.dat</code>
argument and store the output into a variable named <code>zicoseq_out</code>. Provide
also the colData converted to a data.frame (with <code>as.data.frame</code>) as
<code>meta.dat</code>. In addition, set <code>grp.name</code> to <code>"Age"</code>, <code>adj.name</code> to
<code>c("Diet", "Gender")</code>, <code>feature.dat.type</code> to <code>"count"</code>, <code>return.feature.dat</code>
to <code>TRUE</code> and <code>perm.no</code> to <code>999</code>.</li>
<li>View the top six differentially abundant taxa and their adjusted p-values
with <code>head(sort(zicoseq_out$p.adj.fdr))</code>. Is there any significant taxon
according to ZicoSeq? Compared to the output of linda, do we see the same
taxa at the top in terms of significance? Overall, to what extent do the two
methods agree with one another?</li>
</ol>
</div>
<div id="workflow-1" class="section level3 hasAnchor" number="19.8.4">
<h3><span class="header-section-number">19.8.4</span> Workflow 1<a href="exercises.html#workflow-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Get the abundances for an individual feature (taxonomic group / row)</li>
<li>Visualize the abundances per group with boxplot / jitterplot</li>
<li>Is the difference significant (Wilcoxon test)?</li>
<li>Is the difference significant (linear model with covariates)?</li>
<li>How do transformations affect the outcome (log10, clr..)?</li>
<li>Get p-values for all features (taxa), for instance with a for loop</li>
<li>Do multiple testing correction</li>
<li>Compare the results from different tests with a scatterplot</li>
</ol>
<p>Useful functions: [], ggplot2::geom_boxplot, ggplot2::geom_jitter, wilcox.test, lm.test, transformAssay, p.adjust</p>
</div>
<div id="workflow-2" class="section level3 hasAnchor" number="19.8.5">
<h3><span class="header-section-number">19.8.5</span> Workflow 2<a href="exercises.html#workflow-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>install the latest development version of mia from GitHub.</li>
<li>Load experimental dataset from mia.</li>
<li>Compare abundances of each taxa between groups. First, use Wilcoxon or Kruskall-Wallis test. Then use some other method dedicated to microbiome data.</li>
<li>Summarize findings by plotting a taxa vs samples heatmap. Add column annotation that tells the group of each sample, and row annotation that tells whether the difference of certain taxa was statistically significant.</li>
<li>Choose statistically significant taxa and visualize their abundances with boxplot &amp; jitterplot.</li>
</ol>
<p>Useful functions: wilcox.test, kruskal.test, ggplot, pheatmap, ComplexHeatMap::Heatmap, ancombc, aldex2, maaslin2, mergeFeaturesByRank, transformAssay, subsetByPrevalentFeatures</p>
</div>
</div>
<div id="visualization-2" class="section level2 hasAnchor" number="19.9">
<h2><span class="header-section-number">19.9</span> Visualization<a href="exercises.html#visualization-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="multivariate-ordination" class="section level3 hasAnchor" number="19.9.1">
<h3><span class="header-section-number">19.9.1</span> Multivariate ordination<a href="exercises.html#multivariate-ordination" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Load experimental dataset from mia.</li>
<li>Create PCoA with Bray-Curtis dissimilarities</li>
<li>Create PCA with Aitchison dissimilarities</li>
<li>Visualize and compare both</li>
<li>Test other transformations, dissimilarities, and ordination methods</li>
</ol>
<p>Useful functions: runMDS, runNMDS, transformAssay, ggplot, plotReducedDim</p>
</div>
<div id="heatmap-visualization" class="section level3 hasAnchor" number="19.9.2">
<h3><span class="header-section-number">19.9.2</span> Heatmap visualization<a href="exercises.html#heatmap-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Load experimental dataset from mia.</li>
<li>Visualize abundances with heatmap</li>
<li>Visualize abundances with heatmap after CLR + Z transformation</li>
</ol>
<p>See the OMA book for examples.</p>
</div>
</div>
<div id="multiomics" class="section level2 hasAnchor" number="19.10">
<h2><span class="header-section-number">19.10</span> Multiomics<a href="exercises.html#multiomics" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="basic-exploration" class="section level3 hasAnchor" number="19.10.1">
<h3><span class="header-section-number">19.10.1</span> Basic exploration<a href="exercises.html#basic-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we learn how to conduct preliminary exploration on a MAE, using
HintikkaXOData as an example dataset.</p>
<ol style="list-style-type: decimal">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a
variable named <code>mae</code>.</li>
<li>Which experiments make up the MAE? How many samples and features are contained
by each experiment? You can get a summary for all experiments with <code>experiments</code>,
and check for each individual experiment with <code>dim</code>, <code>nrow</code> and <code>ncol</code>.</li>
<li>What are the names of the features and samples of the different experiments?
You can see that with <code>rownames</code> and <code>colnames</code>, respectively.</li>
<li>What information is known about the samples? Remember that information about
samples is stored in the <code>colData</code> of the MAE.</li>
<li><strong>Extra</strong>: How do the samples of the individual experiments map to the
columns of the MAE? You can find the sample mapping in the <code>sampleMap</code>
of the MAE.</li>
</ol>
<p>So far so good. You explored a MAE and its experiments, getting a taste of how
information is organized in its internal structure.</p>
</div>
<div id="experiment-agglomeration" class="section level3 hasAnchor" number="19.10.2">
<h3><span class="header-section-number">19.10.2</span> Experiment agglomeration<a href="exercises.html#experiment-agglomeration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we learn how to manipulate an experiment contained by a MAE and save the
new modified version of the experiment in a suitable place (the altExp slot).</p>
<ol style="list-style-type: decimal">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a
variable named <code>mae</code>.</li>
<li>Agglomerate the microbiota experiment by Genus and store the output into the
<code>altExp</code> slot of the microbiota experiment, with the custom name
<code>microbiota_genus</code>.</li>
<li>How many features remain after agglomerating? What are their names?</li>
<li><strong>Extra</strong>: create one more alternative experiment named
<code>prevalent_microbiota_family</code>, which contains the microbiota experiment
agglomerated by Family with a prevalence threshold of 10%. You can
agglomerate and in parallel select by prevalence with
<code>mergeFeaturesByPrevalence</code>.</li>
</ol>
<p>Good job! You agglomerated one of the experiments in the MAE and stored it as
an alternative experiment.</p>
</div>
<div id="experiment-transformation" class="section level3 hasAnchor" number="19.10.3">
<h3><span class="header-section-number">19.10.3</span> Experiment transformation<a href="exercises.html#experiment-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We proceed with an exercise on a different type of data manipulation, that is,
transformation of assays of individual experiments in the MAE.</p>
<ol style="list-style-type: decimal">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a
variable named <code>mae</code>.</li>
<li>What assays are contained by each individual experiment? You can check their
names with <code>assays</code>.</li>
<li>Apply a log10 transformation to the assay of the metabolite experiment.
For that you can use <code>transformAssay</code> and don’t forget to specify the assay
to be transformed with the argument <code>assay.type</code>.</li>
<li>Apply a CLR transformation to the counts assay of the microbiota experiment.
To ensure non-null values in the assay, set <code>pseudocount</code> equal to 1.</li>
</ol>
<p>You made it! You learnt how to apply different transformations to the assays
of individual experiments in a MAE with <code>transformAssay</code>, specifying optional
arguments based on the used method.</p>
</div>
<div id="assay-extraction" class="section level3 hasAnchor" number="19.10.4">
<h3><span class="header-section-number">19.10.4</span> Assay extraction<a href="exercises.html#assay-extraction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The following exercise walks you through disassembling a MAE object in
order to retrieve a specific assay, or to store its components as
multiple separate csv files.</p>
<ol style="list-style-type: decimal">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a
variable named <code>mae</code>.</li>
<li>Extract the individual metabolite experiment from the MAE into a distinct
TreeSE object named <code>metabolites</code>.</li>
<li>Which and how many assays are contained by <code>metabolites</code>? You can check that
with <code>assays</code> or <code>assayNames</code>.</li>
<li>Write a csv file for the nmr assay with <code>write.csv</code>. You can access an
individual assay of a TreeSE with <code>assay</code> by specifying the name of the
desired assay.</li>
<li><strong>Extra</strong>: Repeat step 1 thorugh 4 also for the microbiota and biomarkers
experiments, so that a completely disassembled version of the MAE is available.</li>
<li><strong>Extra</strong>: Besides experiments, MAEs also include a sampleData and a sampleMap,
which are accessible with <code>colData(mae)</code> and <code>sampleMap(mae)</code>, respectively.
Save also each of these two elements into a csv file.</li>
</ol>
<p>Well done! You just splitted a MAE into its components and stored them as csv files.
<a href="https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl/blob/main/src/assets/XO_preprocess.R">This script</a>
shows a possible approach.</p>
</div>
<div id="mae-reconstruction" class="section level3 hasAnchor" number="19.10.5">
<h3><span class="header-section-number">19.10.5</span> MAE reconstruction<a href="exercises.html#mae-reconstruction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Next, we will try to reconstruct the same MAE from the files you created.
Make sure you know their names and location! Alternatively, you can fetch or
download the CSV files in
<a href="https://github.com/microbiome/data/tree/main/HintikkaXOData">this directory</a>
with the readily disassembled components of HintikkaXOData.</p>
<ol style="list-style-type: decimal">
<li>Read in the csv files containing assays with <code>read.csv</code> and save each
of them into a variable named <code>&lt;assay name&gt;_assays</code>.</li>
<li>Create one TreeSE from each assays object with the <code>TreeSummarizedExperiment</code>
function, as explained in <a href="exercises.html#construct-TreeSE">this exercise</a>.</li>
<li>Read in the sampleData and the sampleMap and store them into the
variables <code>sample_data</code> and <code>sample_map</code>, respectively.</li>
<li>Combine the components with <code>MultiAssayExperiment</code>, where the first argument
is an <code>ExperimentList</code> (for now include only the microbiota and metabolites
TreeSEs), the second is colData and the third is sampleMap.</li>
<li>Make sure that the MAE experiments are identical to the original TreeSEs. You
can do that qualitatively by checking their <code>head</code> and quantitatively by
looking at their <code>dim</code>.</li>
<li><strong>Extra</strong>: Add the biomarkers TreeSE as a new experiment to the MAE.
Note that new experiments can be added to a MAE through simple concatenation
with <code>c(mae, experiment)</code>.</li>
</ol>
<p>Good job! Now you are aware of how MAEs are built and we can proceed to some
analytical exercises.</p>
</div>
<div id="cross-correlation-analysis" class="section level3 hasAnchor" number="19.10.6">
<h3><span class="header-section-number">19.10.6</span> Cross-correlation analysis<a href="exercises.html#cross-correlation-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now we will perform a cross-correlation analysis between two of the
experiments in the MAE.</p>
<ol style="list-style-type: decimal">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a
variable named <code>mae</code>.</li>
<li>Analyze correlations between the microbiota and the biomarkers experiments
with <code>getExperimentCrossAssociation</code>. Don’t forget to specify the experiments
you want to compare with the arguments <code>experiment1</code> and <code>experiment2</code>, and
which of their assays with <code>assay.type1</code> and <code>assay.type2</code>.</li>
<li>What does the output look like? By default, correlation is measured in terms
of Kendall tau coefficients. Repeat point 2, but this time change <code>method</code>
to Spearman coefficients.</li>
<li>Are you able to infer significance from the output? In order to also obtain
p-values from the cross-correlation analysis, repeat point 2 with the
additional argument <code>test_significance = TRUE</code>.</li>
<li>Visualize results with a heatmap similarly to the example in section
<a href="multi-assay-analyses.html#cross-correlation">15.1</a>. Do you see any significant correlations?
Interpret your results.</li>
<li><strong>Extra</strong>: Perform cross-correlation analysis between the remaining
experiments (microbiota vs metabolites and metabolites vs biomarkers) and
visualize results with heatmaps.</li>
</ol>
<p>Great job! You performed a cross-correlation analysis between two experiments of
a MAE and visualized the results with a heatmap. You are also able to customise
the correlation method and significance testing used for the analysis.</p>

</div>
</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="resources.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="extras.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/98_exercises.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/98_exercises.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/98_exercises.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
