<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 18 Extra material | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 18 Extra material | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 18 Extra material | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 18 Extra material | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="extras.html"/>
<link rel="next" href="interactive.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.1/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.2.2</b> colData</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.2.3</b> rowData</a></li>
<li class="chapter" data-level="2.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.2.4</b> rowTree</a></li>
<li class="chapter" data-level="2.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>2.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>2.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>2.3.1</b> 16S workflow</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.2</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.3" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.3</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#some-available-packages"><i class="fa fa-check"></i><b>3.2</b> Some available packages</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="packages.html"><a href="packages.html#mia-family-of-methods"><i class="fa fa-check"></i><b>3.2.1</b> mia family of methods</a></li>
<li class="chapter" data-level="3.2.2" data-path="packages.html"><a href="packages.html#tree-based-methods-sub-tree-methods"><i class="fa fa-check"></i><b>3.2.2</b> Tree-based methods {sub-tree-methods}</a></li>
<li class="chapter" data-level="3.2.3" data-path="packages.html"><a href="packages.html#differential-abundance-sub-diff-abund"><i class="fa fa-check"></i><b>3.2.3</b> Differential abundance {sub-diff-abund}</a></li>
<li class="chapter" data-level="3.2.4" data-path="packages.html"><a href="packages.html#manipulation-sub-manipulation"><i class="fa fa-check"></i><b>3.2.4</b> Manipulation {sub-manipulation}</a></li>
<li class="chapter" data-level="3.2.5" data-path="packages.html"><a href="packages.html#data"><i class="fa fa-check"></i><b>3.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.2</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.2.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
<li class="chapter" data-level="" data-path="community-diversity.html"><a href="community-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#community-comparisons-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.2</b> Community comparisons by beta diversity analysis</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#pcoa-genus"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="community-similarity.html"><a href="community-similarity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.4.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.4.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.4.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="community-similarity.html"><a href="community-similarity.html#further-reading"><i class="fa fa-check"></i><b>8.5</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="community-similarity.html"><a href="community-similarity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community typing (clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#hiearchical-clustering"><i class="fa fa-check"></i><b>10.1</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>10.2</b> K-means clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.3.1</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#community-detection"><i class="fa fa-check"></i><b>10.4</b> Community Detection</a></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.5</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.5.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.5.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>10.5.3</b> Taxa vs taxa</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.6</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>11.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.1.6</b> LinDA</a></li>
<li class="chapter" data-level="11.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.7</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.8" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.8</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>12</b> Machine learning</a>
<ul>
<li class="chapter" data-level="12.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>12.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="12.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>12.2</b> Unsupervised machine learning</a></li>
<li class="chapter" data-level="" data-path="machine_learning.html"><a href="machine_learning.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#multi-omics-factor-analysis"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
<li class="chapter" data-level="" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>14</b> Visualization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>14.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>14.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="14.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>14.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>14.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>14.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="14.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>14.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>14.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>14.3.1</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="viz-chapter.html"><a href="viz-chapter.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="15" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>15</b> Training</a>
<ul>
<li class="chapter" data-level="15.1" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>15.1</b> Code of Conduct</a></li>
<li class="chapter" data-level="15.2" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>15.2</b> Checklist</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="training.html"><a href="training.html#setting-up-your-own-computer"><i class="fa fa-check"></i><b>15.2.1</b> Setting up Your own computer</a></li>
<li class="chapter" data-level="15.2.2" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>15.2.2</b> Support and resources</a></li>
<li class="chapter" data-level="15.2.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>15.2.3</b> Study material</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="packages.html"><a href="packages.html#packages"><i class="fa fa-check"></i><b>15.3</b> Installing and loading the required R packages</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>16</b> Resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>16.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>16.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="16.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>16.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="16.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>16.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="16.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>16.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>16.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>16.2.1</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="resources.html"><a href="resources.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>16.3</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>17</b> Extra material</a>
<ul>
<li class="chapter" data-level="17.1" data-path="extras.html"><a href="extras.html#permanova-comparison"><i class="fa fa-check"></i><b>17.1</b> PERMANOVA comparison</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="extra_bayes.html"><a href="extra_bayes.html"><i class="fa fa-check"></i><b>18</b> Extra material</a>
<ul>
<li class="chapter" data-level="18.1" data-path="extra_bayes.html"><a href="extra_bayes.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>18.1</b> Bayesian Multinomial Logistic-Normal Models</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="interactive.html"><a href="interactive.html"><i class="fa fa-check"></i><b>19</b> Interactive plots</a>
<ul>
<li class="chapter" data-level="19.1" data-path="interactive.html"><a href="interactive.html#interactive-3d-plots"><i class="fa fa-check"></i><b>19.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>20</b> Acknowledgments</a></li>
<li class="chapter" data-level="21" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>21</b> Exercises</a>
<ul>
<li class="chapter" data-level="21.1" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>21.1</b> Workflows</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting"><i class="fa fa-check"></i><b>21.1.1</b> Reproducible reporting</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-basic-components"><i class="fa fa-check"></i><b>21.2</b> TreeSummarizedExperiment: basic components</a>
<ul>
<li class="chapter" data-level="21.2.1" data-path="exercises.html"><a href="exercises.html#basic-summaries"><i class="fa fa-check"></i><b>21.2.1</b> Basic summaries</a></li>
<li class="chapter" data-level="21.2.2" data-path="exercises.html"><a href="exercises.html#taxonomic-abundance-table-assay"><i class="fa fa-check"></i><b>21.2.2</b> Taxonomic abundance table (assay)</a></li>
<li class="chapter" data-level="21.2.3" data-path="exercises.html"><a href="exercises.html#sample-side-information"><i class="fa fa-check"></i><b>21.2.3</b> Sample side information</a></li>
<li class="chapter" data-level="21.2.4" data-path="exercises.html"><a href="exercises.html#feature-side-information"><i class="fa fa-check"></i><b>21.2.4</b> Feature side information</a></li>
<li class="chapter" data-level="21.2.5" data-path="exercises.html"><a href="exercises.html#other-components"><i class="fa fa-check"></i><b>21.2.5</b> Other components</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-data-import"><i class="fa fa-check"></i><b>21.3</b> TreeSummarizedExperiment: data import</a></li>
<li class="chapter" data-level="21.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>21.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="21.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>21.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="21.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>21.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="21.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>21.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="21.4.4" data-path="exercises.html"><a href="exercises.html#explore-the-data"><i class="fa fa-check"></i><b>21.4.4</b> Explore the data</a></li>
<li class="chapter" data-level="21.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>21.4.5</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="21.5" data-path="exercises.html"><a href="exercises.html#transformations"><i class="fa fa-check"></i><b>21.5</b> Transformations</a>
<ul>
<li class="chapter" data-level="21.5.1" data-path="exercises.html"><a href="exercises.html#transformations-1"><i class="fa fa-check"></i><b>21.5.1</b> Transformations</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>21.6</b> Taxonomic levels</a>
<ul>
<li class="chapter" data-level="21.6.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels-1"><i class="fa fa-check"></i><b>21.6.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="21.6.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments-altexp"><i class="fa fa-check"></i><b>21.6.2</b> Alternative experiments (altExp)</a></li>
</ul></li>
<li class="chapter" data-level="21.7" data-path="exercises.html"><a href="exercises.html#alpha-diversity"><i class="fa fa-check"></i><b>21.7</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="21.7.1" data-path="exercises.html"><a href="exercises.html#alpha-diversity-basics"><i class="fa fa-check"></i><b>21.7.1</b> Alpha diversity basics</a></li>
<li class="chapter" data-level="21.7.2" data-path="exercises.html"><a href="exercises.html#alpha-diversity-extra"><i class="fa fa-check"></i><b>21.7.2</b> Alpha diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="21.8" data-path="exercises.html"><a href="exercises.html#community-composition"><i class="fa fa-check"></i><b>21.8</b> Community composition</a>
<ul>
<li class="chapter" data-level="21.8.1" data-path="exercises.html"><a href="exercises.html#beta-diversity-basics"><i class="fa fa-check"></i><b>21.8.1</b> Beta diversity basics</a></li>
<li class="chapter" data-level="21.8.2" data-path="exercises.html"><a href="exercises.html#beta-diversity-extra"><i class="fa fa-check"></i><b>21.8.2</b> Beta diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="21.9" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>21.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="21.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>21.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="21.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>21.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="21.10" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>21.10</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="21.10.1" data-path="exercises.html"><a href="exercises.html#univariate-analyses"><i class="fa fa-check"></i><b>21.10.1</b> Univariate analyses</a></li>
<li class="chapter" data-level="21.10.2" data-path="exercises.html"><a href="exercises.html#differential-abundance-analysis-1"><i class="fa fa-check"></i><b>21.10.2</b> Differential abundance analysis</a></li>
</ul></li>
<li class="chapter" data-level="21.11" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>21.11</b> Multiomics</a>
<ul>
<li class="chapter" data-level="21.11.1" data-path="exercises.html"><a href="exercises.html#introduction-to-multiassayexperiment-mae"><i class="fa fa-check"></i><b>21.11.1</b> Introduction to MultiAssayExperiment (MAE)</a></li>
<li class="chapter" data-level="21.11.2" data-path="exercises.html"><a href="exercises.html#introduction-to-multiomics"><i class="fa fa-check"></i><b>21.11.2</b> Introduction to multiomics</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="extra_bayes" class="section level1 hasAnchor" number="18">
<h1><span class="header-section-number">Chapter 18</span> Extra material<a href="extra_bayes.html#extra_bayes" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="bayesian-multinomial-logistic-normal-models" class="section level2 hasAnchor" number="18.1">
<h2><span class="header-section-number">18.1</span> Bayesian Multinomial Logistic-Normal Models<a href="extra_bayes.html#bayesian-multinomial-logistic-normal-models" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Analysis using such model could be performed with the function
<code>pibble</code> from the <code>fido</code> package, wihch is in form of a Multinomial
Logistic-Normal Linear Regression model; see
<a href="https://jsilve24.github.io/fido/articles/introduction-to-fido.html">vignette</a>
of package.</p>
<p>The following presents such an exemplary analysis based on the
data of <span class="citation">Sprockett et al. (<a href="#ref-Sprockett2020" role="doc-biblioref">2020</a>)</span> available
through <code>microbiomeDataSets</code> package.</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="extra_bayes.html#cb414-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(fido)){</span>
<span id="cb414-2"><a href="extra_bayes.html#cb414-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># installing the fido package</span></span>
<span id="cb414-3"><a href="extra_bayes.html#cb414-3" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;jsilve24/fido&quot;</span>)</span>
<span id="cb414-4"><a href="extra_bayes.html#cb414-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Loading the libraries and importing data:</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="extra_bayes.html#cb415-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fido)</span></code></pre></div>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="extra_bayes.html#cb416-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb416-2"><a href="extra_bayes.html#cb416-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">SprockettTHData</span>()</span></code></pre></div>
<p>We pick three covariates (“Sex”,“Age_Years”,“Delivery_Mode”) during this
analysis as an example, and beforehand we check for missing data:</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb417-1"><a href="extra_bayes.html#cb417-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb417-2"><a href="extra_bayes.html#cb417-2" aria-hidden="true" tabindex="-1"></a>cov_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Sex&quot;</span>,<span class="st">&quot;Age_Years&quot;</span>,<span class="st">&quot;Delivery_Mode&quot;</span>)</span>
<span id="cb417-3"><a href="extra_bayes.html#cb417-3" aria-hidden="true" tabindex="-1"></a>na_counts <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">is.na</span>(<span class="fu">colData</span>(tse)[,cov_names]), <span class="dv">2</span>, sum)</span>
<span id="cb417-4"><a href="extra_bayes.html#cb417-4" aria-hidden="true" tabindex="-1"></a>na_summary<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(na_counts,<span class="at">row.names=</span>cov_names)</span></code></pre></div>
<p>We drop missing values of the covariates:</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb418-1"><a href="extra_bayes.html#cb418-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> tse[ , <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>Delivery_Mode) ]</span>
<span id="cb418-2"><a href="extra_bayes.html#cb418-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> tse[ , <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>Age_Years) ]</span></code></pre></div>
<p>We agglomerate microbiome data to Phylum:</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="extra_bayes.html#cb419-1" aria-hidden="true" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(tse, <span class="st">&quot;Phylum&quot;</span>)</span></code></pre></div>
<p>We extract the counts assay and covariate data to build the model
matrix:</p>
<div class="sourceCode" id="cb420"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb420-1"><a href="extra_bayes.html#cb420-1" aria-hidden="true" tabindex="-1"></a>Y <span class="ot">&lt;-</span> <span class="fu">assays</span>(tse_phylum)<span class="sc">$</span>counts</span>
<span id="cb420-2"><a href="extra_bayes.html#cb420-2" aria-hidden="true" tabindex="-1"></a><span class="co"># design matrix</span></span>
<span id="cb420-3"><a href="extra_bayes.html#cb420-3" aria-hidden="true" tabindex="-1"></a><span class="co"># taking 3 covariates</span></span>
<span id="cb420-4"><a href="extra_bayes.html#cb420-4" aria-hidden="true" tabindex="-1"></a>sample_data<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse_phylum)[,cov_names])</span>
<span id="cb420-5"><a href="extra_bayes.html#cb420-5" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">model.matrix</span>(<span class="sc">~</span>Sex<span class="sc">+</span>Age_Years<span class="sc">+</span>Delivery_Mode,<span class="at">data=</span>sample_data))</span></code></pre></div>
<p>Building the parameters for the <code>pibble</code> call to build the model; see more at <a href="https://jsilve24.github.io/fido/articles/introduction-to-fido.html">vignette</a>:</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="extra_bayes.html#cb421-1" aria-hidden="true" tabindex="-1"></a>n_taxa<span class="ot">&lt;-</span><span class="fu">nrow</span>(Y)</span>
<span id="cb421-2"><a href="extra_bayes.html#cb421-2" aria-hidden="true" tabindex="-1"></a>upsilon <span class="ot">&lt;-</span> n_taxa<span class="sc">+</span><span class="dv">3</span></span>
<span id="cb421-3"><a href="extra_bayes.html#cb421-3" aria-hidden="true" tabindex="-1"></a>Omega <span class="ot">&lt;-</span> <span class="fu">diag</span>(n_taxa)</span>
<span id="cb421-4"><a href="extra_bayes.html#cb421-4" aria-hidden="true" tabindex="-1"></a>G <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">diag</span>(n_taxa<span class="dv">-1</span>), <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb421-5"><a href="extra_bayes.html#cb421-5" aria-hidden="true" tabindex="-1"></a>Xi <span class="ot">&lt;-</span> (upsilon<span class="sc">-</span>n_taxa)<span class="sc">*</span>G<span class="sc">%*%</span>Omega<span class="sc">%*%</span><span class="fu">t</span>(G)</span>
<span id="cb421-6"><a href="extra_bayes.html#cb421-6" aria-hidden="true" tabindex="-1"></a>Theta <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, n_taxa<span class="dv">-1</span>, <span class="fu">nrow</span>(X))</span>
<span id="cb421-7"><a href="extra_bayes.html#cb421-7" aria-hidden="true" tabindex="-1"></a>Gamma <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">nrow</span>(X))</span></code></pre></div>
<p>Automatically initializing the priors and visualizing their distributions:</p>
<div class="sourceCode" id="cb422"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb422-1"><a href="extra_bayes.html#cb422-1" aria-hidden="true" tabindex="-1"></a>priors <span class="ot">&lt;-</span> <span class="fu">pibble</span>(<span class="cn">NULL</span>, X, upsilon, Theta, Gamma, Xi)</span>
<span id="cb422-2"><a href="extra_bayes.html#cb422-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names_covariates</span>(priors) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X)</span>
<span id="cb422-3"><a href="extra_bayes.html#cb422-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(priors, <span class="at">pars=</span><span class="st">&quot;Lambda&quot;</span>) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>))</span></code></pre></div>
<p><img src="97_extra_materials_bayes_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Estimating the posterior by including our response data <code>Y</code>.
Note: Some computational failures could occur (see <a href="https://github-wiki-see.page/m/jsilve24/fido/wiki/Frequently-Asked-Questions">discussion</a>)
the arguments <code>multDirichletBoot</code> <code>calcGradHess</code> could be passed in such case.</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb423-1"><a href="extra_bayes.html#cb423-1" aria-hidden="true" tabindex="-1"></a>priors<span class="sc">$</span>Y <span class="ot">&lt;-</span> Y </span>
<span id="cb423-2"><a href="extra_bayes.html#cb423-2" aria-hidden="true" tabindex="-1"></a>posterior <span class="ot">&lt;-</span> <span class="fu">refit</span>(priors, <span class="at">optim_method=</span><span class="st">&quot;adam&quot;</span>, <span class="at">multDirichletBoot=</span><span class="fl">0.5</span>) <span class="co">#calcGradHess=FALSE</span></span></code></pre></div>
<p>Printing a summary about the posterior:</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb424-1"><a href="extra_bayes.html#cb424-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ppc_summary</span>(posterior)</span></code></pre></div>
<pre><code>## Proportions of Observations within 95% Credible Interval: 0.9982402</code></pre>
<p>Plotting the summary of the posterior distributions of the regression parameters:</p>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb426-1"><a href="extra_bayes.html#cb426-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names_categories</span>(posterior) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(Y)</span>
<span id="cb426-2"><a href="extra_bayes.html#cb426-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(posterior,<span class="at">par=</span><span class="st">&quot;Lambda&quot;</span>,<span class="at">focus.cov=</span><span class="fu">rownames</span>(X)[<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>])</span></code></pre></div>
<p><img src="97_extra_materials_bayes_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Taking a closer look at “Sex” and “Delivery_Mode”:</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb427-1"><a href="extra_bayes.html#cb427-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(posterior, <span class="at">par=</span><span class="st">&quot;Lambda&quot;</span>, <span class="at">focus.cov =</span> <span class="fu">rownames</span>(X)[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)])</span></code></pre></div>
<p><img src="97_extra_materials_bayes_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>

</div>
</div>
<h3>Bibliography<a href="bibliography.html#bibliography" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Sprockett2020" class="csl-entry">
Sprockett, Daniel D., Melanie Martin, Elizabeth K. Costello, Adam R. Burns, Susan P. Holmes, Michael D. Gurven, and David A. Relman. 2020. <span>“Microbiota Assembly, Structure, and Dynamics Among <span>Tsimane</span> Horticulturalists of the <span>Bolivian Amazon</span>.”</span> <em>Nat Commun</em> 11 (1): 3772. <a href="https://doi.org/10.1038/s41467-020-17541-6">https://doi.org/10.1038/s41467-020-17541-6</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="extras.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="interactive.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
