<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Introduction to Bioconductor Classes | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Introduction to Bioconductor Classes | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Introduction to Bioconductor Classes | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to Bioconductor Classes | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="introduction.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="" data-path="introduction-to-bioconductor-classes.html"><a href="introduction-to-bioconductor-classes.html"><i class="fa fa-check"></i>Introduction to Bioconductor Classes</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#community"><i class="fa fa-check"></i><b>1.1</b> Community</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#developers-and-contributors"><i class="fa fa-check"></i><b>1.2</b> Developers and contributors</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.3</b> Acknowledgements</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#package-ecosystem"><i class="fa fa-check"></i><b>1.4</b> Package ecosystem</a></li>
</ul></li>
<li class="part"><span><b>II Introduction</b></span></li>
<li class="chapter" data-level="2" data-path="data-introduction.html"><a href="data-introduction.html"><i class="fa fa-check"></i><b>2</b> Data Infrastructure</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-introduction.html"><a href="data-introduction.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="data-introduction.html"><a href="data-introduction.html#packages"><i class="fa fa-check"></i><b>2.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="data-introduction.html"><a href="data-introduction.html#background"><i class="fa fa-check"></i><b>2.2</b> Background</a></li>
<li class="chapter" data-level="2.3" data-path="data-introduction.html"><a href="data-introduction.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="data-introduction.html"><a href="data-introduction.html#importing-data-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Importing data from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="data-introduction.html"><a href="data-introduction.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-introduction.html"><a href="data-introduction.html#metadata"><i class="fa fa-check"></i><b>2.4</b> Metadata</a></li>
<li class="chapter" data-level="2.5" data-path="data-introduction.html"><a href="data-introduction.html#microbiome-and-tree-data-specific-aspects"><i class="fa fa-check"></i><b>2.5</b> Microbiome and tree data specific aspects</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="data-introduction.html"><a href="data-introduction.html#assays"><i class="fa fa-check"></i><b>2.5.1</b> Assays</a></li>
<li class="chapter" data-level="2.5.2" data-path="data-introduction.html"><a href="data-introduction.html#coldata"><i class="fa fa-check"></i><b>2.5.2</b> colData</a></li>
<li class="chapter" data-level="2.5.3" data-path="data-introduction.html"><a href="data-introduction.html#rowdata"><i class="fa fa-check"></i><b>2.5.3</b> rowData</a></li>
<li class="chapter" data-level="2.5.4" data-path="data-introduction.html"><a href="data-introduction.html#rowtree"><i class="fa fa-check"></i><b>2.5.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="data-introduction.html"><a href="data-introduction.html#data-conversion"><i class="fa fa-check"></i><b>2.6</b> Data conversion</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="data-introduction.html"><a href="data-introduction.html#tidy-data"><i class="fa fa-check"></i><b>2.6.1</b> Tidy data</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="data-introduction.html"><a href="data-introduction.html#conclusion"><i class="fa fa-check"></i><b>2.7</b> Conclusion</a></li>
<li class="chapter" data-level="2.8" data-path="data-introduction.html"><a href="data-introduction.html#additional-reading"><i class="fa fa-check"></i><b>2.8</b> Additional Reading</a>
<ul>
<li class="chapter" data-level="2.8.1" data-path="data-introduction.html"><a href="data-introduction.html#lecture-slides"><i class="fa fa-check"></i><b>2.8.1</b> Lecture slides</a></li>
<li class="chapter" data-level="2.8.2" data-path="data-introduction.html"><a href="data-introduction.html#r-programming-resources"><i class="fa fa-check"></i><b>2.8.2</b> R programming resources</a></li>
<li class="chapter" data-level="2.8.3" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>2.8.3</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="2.8.4" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>2.8.4</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="2.8.5" data-path="data-introduction.html"><a href="data-introduction.html#further-reading"><i class="fa fa-check"></i><b>2.8.5</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-introduction.html"><a href="data-introduction.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>3</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>3.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>3.1.1</b> dada2</a></li>
<li class="chapter" data-level="3.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>3.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>3.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>3.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>3.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="3.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>3.4</b> Data transformation</a></li>
<li class="chapter" data-level="3.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>3.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>3.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="3.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>3.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>4</b> Quality Control</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quality-control.html"><a href="quality-control.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>4.1</b> Get top taxa and taxonomy</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quality-control.html"><a href="quality-control.html#features"><i class="fa fa-check"></i><b>4.1.1</b> Features</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>4.2</b> Library size</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html"><i class="fa fa-check"></i><b>5</b> Microbiome Exploration</a>
<ul>
<li class="chapter" data-level="5.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalence"><i class="fa fa-check"></i><b>5.1</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>5.1.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="5.1.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.1.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#abundance"><i class="fa fa-check"></i><b>5.2</b> Abundance</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>6</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="6.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#estimating-alpha-diversity"><i class="fa fa-check"></i><b>6.1</b> Estimating alpha diversity</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#richness"><i class="fa fa-check"></i><b>6.1.1</b> Richness</a></li>
<li class="chapter" data-level="6.1.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#diversity"><i class="fa fa-check"></i><b>6.1.2</b> Diversity</a></li>
<li class="chapter" data-level="6.1.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#evenness"><i class="fa fa-check"></i><b>6.1.3</b> Evenness</a></li>
<li class="chapter" data-level="6.1.4" data-path="alpha-diversity.html"><a href="alpha-diversity.html#dominance"><i class="fa fa-check"></i><b>6.1.4</b> Dominance</a></li>
<li class="chapter" data-level="6.1.5" data-path="alpha-diversity.html"><a href="alpha-diversity.html#rarity"><i class="fa fa-check"></i><b>6.1.5</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualize-alpha-diversities"><i class="fa fa-check"></i><b>6.2</b> Visualize alpha diversities</a></li>
<li class="chapter" data-level="" data-path="alpha-diversity.html"><a href="alpha-diversity.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>7</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="beta-diversity.html"><a href="beta-diversity.html#explained-variance"><i class="fa fa-check"></i><b>7.1</b> Explained variance</a></li>
<li class="chapter" data-level="7.2" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-beta-diversity"><i class="fa fa-check"></i><b>7.2</b> Estimating beta diversity</a></li>
<li class="chapter" data-level="7.3" data-path="beta-diversity.html"><a href="beta-diversity.html#other-ordination-methods"><i class="fa fa-check"></i><b>7.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="7.4" data-path="beta-diversity.html"><a href="beta-diversity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>7.4</b> Visualizing the most dominant genus on PCoA</a></li>
<li class="chapter" data-level="7.5" data-path="beta-diversity.html"><a href="beta-diversity.html#community-comparisons-todo-combine-with-the-material-above-for-simplicity"><i class="fa fa-check"></i><b>7.5</b> Community comparisons [TODO combine with the material above for simplicity?]</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="beta-diversity.html"><a href="beta-diversity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>7.5.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="7.5.2" data-path="beta-diversity.html"><a href="beta-diversity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>7.5.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="beta-diversity.html"><a href="beta-diversity.html#further-reading-1"><i class="fa fa-check"></i><b>7.6</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="beta-diversity.html"><a href="beta-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>8</b> Microbiome Community</a>
<ul>
<li class="chapter" data-level="8.1" data-path="microbiome-community.html"><a href="microbiome-community.html#community-composition"><i class="fa fa-check"></i><b>8.1</b> Community composition</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>8.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="8.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>8.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>9</b> Community typing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="community-typing.html"><a href="community-typing.html#community-typing-1"><i class="fa fa-check"></i><b>9.1</b> Community typing</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>9.1.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="9.1.2" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>9.1.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>9.2</b> Community Detection</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>10</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="10.1" data-path="differential-abundance.html"><a href="differential-abundance.html#introduction-to-differential-abundance-analysis"><i class="fa fa-check"></i><b>10.1</b> Introduction to differential abundance analysis</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>10.1.1</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="10.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>10.1.2</b> ALDEx2</a></li>
<li class="chapter" data-level="10.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>10.1.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="10.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>10.1.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="10.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>10.1.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="10.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>10.1.6</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>10.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>10.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multitable.html"><a href="multitable.html"><i class="fa fa-check"></i><b>11</b> Representation of Multiple Data Tables</a>
<ul>
<li class="chapter" data-level="11.1" data-path="multitable.html"><a href="multitable.html#assay-data"><i class="fa fa-check"></i><b>11.1</b> Assay data</a></li>
<li class="chapter" data-level="11.2" data-path="multitable.html"><a href="multitable.html#alternative-experiments"><i class="fa fa-check"></i><b>11.2</b> Alternative experiments</a></li>
<li class="chapter" data-level="11.3" data-path="multitable.html"><a href="multitable.html#multiassayexperiments"><i class="fa fa-check"></i><b>11.3</b> MultiAssayExperiments</a></li>
<li class="chapter" data-level="" data-path="multitable.html"><a href="multitable.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="12" data-path="example-data.html"><a href="example-data.html"><i class="fa fa-check"></i><b>12</b> Example Data</a>
<ul>
<li class="chapter" data-level="12.1" data-path="example-data.html"><a href="example-data.html#package-data"><i class="fa fa-check"></i><b>12.1</b> Package data</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="example-data.html"><a href="example-data.html#globalpatterns"><i class="fa fa-check"></i><b>12.1.1</b> GlobalPatterns</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="example-data.html"><a href="example-data.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="13" data-path="function-ref.html"><a href="function-ref.html"><i class="fa fa-check"></i><b>13</b> Function Reference</a>
<ul>
<li class="chapter" data-level="13.1" data-path="function-ref.html"><a href="function-ref.html#objects"><i class="fa fa-check"></i><b>13.1</b> Objects</a></li>
<li class="chapter" data-level="13.2" data-path="function-ref.html"><a href="function-ref.html#reduced-dimension-pca-t-sne-umap-mds-etc"><i class="fa fa-check"></i><b>13.2</b> Reduced dimension (PCA, t-SNE, UMAP, MDS, etc)</a></li>
<li class="chapter" data-level="13.3" data-path="function-ref.html"><a href="function-ref.html#plotting"><i class="fa fa-check"></i><b>13.3</b> Plotting</a></li>
</ul></li>
<li class="part"><span><b>V Appendix</b></span></li>
<li class="chapter" data-level="14" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>14</b> Extras</a>
<ul>
<li class="chapter" data-level="14.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>14.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-bioconductor-classes" class="section level1 unnumbered">
<h1>Introduction to Bioconductor Classes</h1>
<p><strong>S4 system</strong></p>
<p>S4 class system has brought several useful features to the object-oriented programming paradigm within R.
It has been constantly deployed in R packages, including the <a href="https://bioconductor.org/">Bioconductor</a> package.</p>
<div class="line-block">  Books:</div>
<ul>
<li>John M. Chambers. Software for Data Analysis: Programming with R. Springer, New York, 2008. ISBN-13 978-0387759357.</li>
<li>I Robert Gentleman. R Programming for Bioinformatics. Chapman &amp; Hall/CRC, New York, 2008. ISBN-13 978-1420063677</li>
</ul>
<div class="line-block">  Online Document:</div>
<ul>
<li>Hervé Pagès, <a href="https://bioconductor.org/packages/release/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>.</li>
<li>Laurent Gatto, <a href="https://bioconductor.org/help/course-materials/2013/CSAMA2013/friday/afternoon/S4-tutorial.pdf">A practical tutorial on S4 programming</a></li>
<li>John M. Chambers. <a href="http://developer.r-project.org/howMethodsWork.pdf">How S4 Methods Work</a></li>
</ul>
<p><strong>Additional useful classes include:</strong></p>
<ul>
<li><a href="https://rdrr.io/bioc/S4Vectors/man/DataFrame-class.html">DataFrame</a> which behaves similarly to <code>data.frame</code>, yet efficient and fast when used with large datasets.</li>
<li><a href="https://rdrr.io/bioc/Biostrings/man/DNAString-class.html">DNAString</a> along with <code>DNAStringSet</code>,<code>RNAString</code> and <code>RNAStringSet</code> efficient storage and handling of long biological sequences are offered within the <a href="https://rdrr.io/bioc/Biostrings/">Biostrings</a> package.</li>
<li><a href="https://bioconductor.org/packages/3.13/bioc/html/GenomicRanges.html">GenomicRanges</a> offers an efficient representation and manipulation of genomic annotations and alignments, see e.g. <code>GRanges</code> and <code>GRangesList</code> at <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">An Introduction to the GenomicRangesPackage</a>.</li>
</ul>
<p><a href="http://girke.bioinformatics.ucr.edu/GEN242/tutorials/rsequences/rsequences/">NGS Analysis Basics</a> is an example walk-through of the above mentioned features, with detailed examples provided.</p>

</div>



            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="introduction.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
