<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Microbiome Exploration | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 7 Microbiome Exploration | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Microbiome Exploration | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Microbiome Exploration | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="quality-control.html"/>
<link rel="next" href="alpha-diversity.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#getting-started"><i class="fa fa-check"></i><b>1.1</b> Getting started</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#community-support"><i class="fa fa-check"></i><b>1.2</b> Community support</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#alternative-experiments"><i class="fa fa-check"></i><b>2.2.2</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#multiassayexperiments"><i class="fa fa-check"></i><b>2.2.3</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration Data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="containers.html"><a href="containers.html#microbiome-data-specific-aspects"><i class="fa fa-check"></i><b>2.5</b> Microbiome data specific aspects</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="containers.html"><a href="containers.html#assays"><i class="fa fa-check"></i><b>2.5.1</b> Assays</a></li>
<li class="chapter" data-level="2.5.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.5.2</b> colData</a></li>
<li class="chapter" data-level="2.5.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.5.3</b> rowData</a></li>
<li class="chapter" data-level="2.5.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.5.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#available-packages"><i class="fa fa-check"></i><b>3.2</b> Available packages</a></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="5" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>5</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>5.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>5.1.1</b> dada2</a></li>
<li class="chapter" data-level="5.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>5.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>5.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>5.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>5.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="5.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>5.4</b> Data transformation</a></li>
<li class="chapter" data-level="5.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>5.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>5.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="5.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>5.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>6</b> Quality Control</a>
<ul>
<li class="chapter" data-level="6.1" data-path="quality-control.html"><a href="quality-control.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>6.1</b> Get top taxa and taxonomy</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="quality-control.html"><a href="quality-control.html#features"><i class="fa fa-check"></i><b>6.1.1</b> Features</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>6.2</b> Library size</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html"><i class="fa fa-check"></i><b>7</b> Microbiome Exploration</a>
<ul>
<li class="chapter" data-level="7.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalence"><i class="fa fa-check"></i><b>7.1</b> Prevalence</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>7.1.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="7.1.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#plotting-prevalence"><i class="fa fa-check"></i><b>7.1.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#abundance"><i class="fa fa-check"></i><b>7.2</b> Abundance</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>8</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#estimating-alpha-diversity"><i class="fa fa-check"></i><b>8.1</b> Estimating alpha diversity</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#richness"><i class="fa fa-check"></i><b>8.1.1</b> Richness</a></li>
<li class="chapter" data-level="8.1.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#diversity"><i class="fa fa-check"></i><b>8.1.2</b> Diversity</a></li>
<li class="chapter" data-level="8.1.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#evenness"><i class="fa fa-check"></i><b>8.1.3</b> Evenness</a></li>
<li class="chapter" data-level="8.1.4" data-path="alpha-diversity.html"><a href="alpha-diversity.html#dominance"><i class="fa fa-check"></i><b>8.1.4</b> Dominance</a></li>
<li class="chapter" data-level="8.1.5" data-path="alpha-diversity.html"><a href="alpha-diversity.html#rarity"><i class="fa fa-check"></i><b>8.1.5</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualize-alpha-diversities"><i class="fa fa-check"></i><b>8.2</b> Visualize alpha diversities</a></li>
<li class="chapter" data-level="" data-path="alpha-diversity.html"><a href="alpha-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>9</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="9.1" data-path="beta-diversity.html"><a href="beta-diversity.html#explained-variance"><i class="fa fa-check"></i><b>9.1</b> Explained variance</a></li>
<li class="chapter" data-level="9.2" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-beta-diversity"><i class="fa fa-check"></i><b>9.2</b> Estimating beta diversity</a></li>
<li class="chapter" data-level="9.3" data-path="beta-diversity.html"><a href="beta-diversity.html#other-ordination-methods"><i class="fa fa-check"></i><b>9.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="9.4" data-path="beta-diversity.html"><a href="beta-diversity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>9.4</b> Visualizing the most dominant genus on PCoA</a></li>
<li class="chapter" data-level="9.5" data-path="beta-diversity.html"><a href="beta-diversity.html#community-comparisons-todo-combine-with-the-material-above-for-simplicity"><i class="fa fa-check"></i><b>9.5</b> Community comparisons [TODO combine with the material above for simplicity?]</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="beta-diversity.html"><a href="beta-diversity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>9.5.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="9.5.2" data-path="beta-diversity.html"><a href="beta-diversity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>9.5.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="beta-diversity.html"><a href="beta-diversity.html#further-reading"><i class="fa fa-check"></i><b>9.6</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="beta-diversity.html"><a href="beta-diversity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>10</b> Microbiome Community</a>
<ul>
<li class="chapter" data-level="10.1" data-path="microbiome-community.html"><a href="microbiome-community.html#community-composition"><i class="fa fa-check"></i><b>10.1</b> Community composition</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>10.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="10.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>10.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>11</b> Community typing</a>
<ul>
<li class="chapter" data-level="11.1" data-path="community-typing.html"><a href="community-typing.html#community-typing-1"><i class="fa fa-check"></i><b>11.1</b> Community typing</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>11.1.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="11.1.2" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>11.1.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>11.2</b> Community Detection</a></li>
<li class="chapter" data-level="11.3" data-path="community-typing.html"><a href="community-typing.html#additional-community-typing"><i class="fa fa-check"></i><b>11.3</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>12</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="12.1" data-path="differential-abundance.html"><a href="differential-abundance.html#introduction-to-differential-abundance-analysis"><i class="fa fa-check"></i><b>12.1</b> Introduction to differential abundance analysis</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>12.1.1</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="12.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>12.1.2</b> ALDEx2</a></li>
<li class="chapter" data-level="12.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>12.1.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="12.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>12.1.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="12.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>12.1.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="12.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>12.1.6</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>12.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>12.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="13" data-path="appendix.html"><a href="appendix.html"><i class="fa fa-check"></i><b>13</b> Appendix</a>
<ul>
<li class="chapter" data-level="13.1" data-path="appendix.html"><a href="appendix.html#acknowledgments"><i class="fa fa-check"></i><b>13.1</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>14</b> Resources</a>
<ul>
<li class="chapter" data-level="14.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>14.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>14.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="14.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>14.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="14.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>14.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="14.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>14.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>14.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>14.2.1</b> Bioconductor Classes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="function-ref.html"><a href="function-ref.html"><i class="fa fa-check"></i><b>15</b> Function Reference</a>
<ul>
<li class="chapter" data-level="15.1" data-path="function-ref.html"><a href="function-ref.html#objects"><i class="fa fa-check"></i><b>15.1</b> Objects</a></li>
<li class="chapter" data-level="15.2" data-path="function-ref.html"><a href="function-ref.html#reduced-dimension-pca-t-sne-umap-mds-etc"><i class="fa fa-check"></i><b>15.2</b> Reduced dimension (PCA, t-SNE, UMAP, MDS, etc)</a></li>
<li class="chapter" data-level="15.3" data-path="function-ref.html"><a href="function-ref.html#plotting"><i class="fa fa-check"></i><b>15.3</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>16</b> Extras</a>
<ul>
<li class="chapter" data-level="16.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>16.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="microbiome-exploration" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> Microbiome Exploration</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>This chapter focuses on the exploration of microbiome data and establishes
commonly used descriptors of a microbiome. The main difference to quality
control is that the exploration assumes that technical aspects of the dataset
have been investigated to your satisfaction. Generally speaking, at this point
you should be quite certain that the dataset doesn’t suffer from severe
technical biases, or you should at least be aware of potential problems.</p>
<p>In reality you might need to go back and forth between QC and exploration,
since you discover through exploration of your dataset technical aspects you
need to check.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="microbiome-exploration.html#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb148-2"><a href="microbiome-exploration.html#cb148-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package=</span><span class="st">&quot;mia&quot;</span>)</span>
<span id="cb148-3"><a href="microbiome-exploration.html#cb148-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> GlobalPatterns </span></code></pre></div>
<div id="prevalence" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Prevalence</h2>
<p>Prevalence is a measurement, which describes in how many samples certain
microbes were detected.</p>
<p>Investigating the prevalence of microbes allows you either to focus on changes,
which pertain to most of the samples, or to focus on less often found microbes,
which are nonetheless abundantly found in a small number of samples.</p>
<p>On the raw data, the population prevalence (frequency) at a 1% relative
abundance threshold (<code>detection = 1/100</code> and <code>as_relative = TRUE</code>), can look
like this. The low prevalence in this example can be explained by rather
different sample types as well as the in-depth nature of the data.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="microbiome-exploration.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(se, <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<pre><code>## 331820 158660  98605 326977 145149 114821 
## 0.2308 0.2308 0.1923 0.1923 0.1538 0.1538</code></pre>
<p>The <code>detection</code> and <code>as_relative</code> can also be used to access, how many samples
do pass a threshold for raw counts. Here the population prevalence (frequency)
at the absolute abundance threshold (<code>as_relative = FALSE</code>) at read count 1
(<code>detection = 1</code>) is accessed.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="microbiome-exploration.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(se, <span class="at">detection =</span> <span class="dv">1</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>, <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb151-2"><a href="microbiome-exploration.html#cb151-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">as_relative =</span> <span class="cn">FALSE</span>))</span></code></pre></div>
<pre><code>## 145149 114821 108747 526804  98605 180658 
##      1      1      1      1      1      1</code></pre>
<p>Note that, if the output should be used for subsetting or storing the data in
the <code>rowData</code>, set <code>sort = FALSE</code>.</p>
<div id="prevalent-microbiota-analysis" class="section level3" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Prevalent microbiota analysis</h3>
<p>To investigate the microbiome data at a selected taxonomic level, two
approaches are available.</p>
<p>First the data can be agglomerated to the taxonomic level and <code>getPrevalence</code>
be used on the result.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="microbiome-exploration.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(se, <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb153-2"><a href="microbiome-exploration.html#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(<span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>), <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb153-3"><a href="microbiome-exploration.html#cb153-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<pre><code>##   Phylum:Bacteroidetes  Phylum:Proteobacteria  Phylum:Actinobacteria 
##                 1.0000                 0.9231                 0.8462 
##   Phylum:Cyanobacteria      Phylum:Firmicutes Phylum:Verrucomicrobia 
##                 0.6154                 0.5769                 0.4615</code></pre>
<p>Alternatively, the <code>rank</code> argument can be set to perform the agglomeration on
the fly.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="microbiome-exploration.html#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(se, <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb155-2"><a href="microbiome-exploration.html#cb155-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(se, <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb155-3"><a href="microbiome-exploration.html#cb155-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<pre><code>##   Bacteroidetes  Proteobacteria  Actinobacteria   Cyanobacteria      Firmicutes 
##          1.0000          0.9231          0.8462          0.6154          0.5769 
## Verrucomicrobia 
##          0.4615</code></pre>
<p>The difference in the naming scheme is that, by default, <code>na.rm = TRUE</code> is used
for agglomeration in <code>getPrevalence</code>, whereas the default for
<code>agglomerateByRank</code> is <code>FALSE</code> to prevent accidental data loss.</p>
<p>If you only need the names of the prevalent taxa, <code>getPrevalentTaxa</code> is
available. This returns the taxa that exceed the given prevalence and detection
thresholds.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="microbiome-exploration.html#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getPrevalentTaxa</span>(se, <span class="at">detection =</span> <span class="dv">0</span>, <span class="at">prevalence =</span> <span class="dv">50</span><span class="sc">/</span><span class="dv">100</span>)</span>
<span id="cb157-2"><a href="microbiome-exploration.html#cb157-2" aria-hidden="true" tabindex="-1"></a>prev <span class="ot">&lt;-</span> <span class="fu">getPrevalentTaxa</span>(se, <span class="at">detection =</span> <span class="dv">0</span>, <span class="at">prevalence =</span> <span class="dv">50</span><span class="sc">/</span><span class="dv">100</span>,</span>
<span id="cb157-3"><a href="microbiome-exploration.html#cb157-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>)</span>
<span id="cb157-4"><a href="microbiome-exploration.html#cb157-4" aria-hidden="true" tabindex="-1"></a>prev</span></code></pre></div>
<p>Note that the <code>detection</code> and <code>prevalence</code> thresholds are not the same, since
<code>detection</code> can be applied to relative counts or absolute counts depending on
whether <code>as_relative</code> is set <code>TRUE</code> or <code>FALSE</code></p>
<p>TODO
See also related functions for the analysis of rare and variable taxa
(rareMembers; rareAbundance; lowAbundance).</p>
</div>
<div id="plotting-prevalence" class="section level3" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Plotting prevalence</h3>
<p>To plot the prevalence, the data is first added to the <code>rowData</code>.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="microbiome-exploration.html#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(<span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>))<span class="sc">$</span>prevalence <span class="ot">&lt;-</span> </span>
<span id="cb158-2"><a href="microbiome-exploration.html#cb158-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getPrevalence</span>(<span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>), <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">FALSE</span>,</span>
<span id="cb158-3"><a href="microbiome-exploration.html#cb158-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Then it can be plotted via the plotting functions from the <code>scater</code> package.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="microbiome-exploration.html#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb159-2"><a href="microbiome-exploration.html#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRowData</span>(<span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>), <span class="st">&quot;prevalence&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;Phylum&quot;</span>)</span></code></pre></div>
<p><img src="13_microbiome_exploration_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Additionally, the prevalence can be plotted on the taxonomic tree using the
<code>miaViz</code> package.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="microbiome-exploration.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(se) <span class="ot">&lt;-</span> <span class="fu">splitByRanks</span>(se)</span>
<span id="cb160-2"><a href="microbiome-exploration.html#cb160-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(se) <span class="ot">&lt;-</span></span>
<span id="cb160-3"><a href="microbiome-exploration.html#cb160-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">lapply</span>(<span class="fu">altExps</span>(se),</span>
<span id="cb160-4"><a href="microbiome-exploration.html#cb160-4" aria-hidden="true" tabindex="-1"></a>          <span class="cf">function</span>(y){</span>
<span id="cb160-5"><a href="microbiome-exploration.html#cb160-5" aria-hidden="true" tabindex="-1"></a>              <span class="fu">rowData</span>(y)<span class="sc">$</span>prevalence <span class="ot">&lt;-</span> </span>
<span id="cb160-6"><a href="microbiome-exploration.html#cb160-6" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">getPrevalence</span>(y, <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">FALSE</span>,</span>
<span id="cb160-7"><a href="microbiome-exploration.html#cb160-7" aria-hidden="true" tabindex="-1"></a>                                <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>)</span>
<span id="cb160-8"><a href="microbiome-exploration.html#cb160-8" aria-hidden="true" tabindex="-1"></a>              y</span>
<span id="cb160-9"><a href="microbiome-exploration.html#cb160-9" aria-hidden="true" tabindex="-1"></a>          })</span>
<span id="cb160-10"><a href="microbiome-exploration.html#cb160-10" aria-hidden="true" tabindex="-1"></a>top_phyla <span class="ot">&lt;-</span> <span class="fu">getTopTaxa</span>(<span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>),</span>
<span id="cb160-11"><a href="microbiome-exploration.html#cb160-11" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method=</span><span class="st">&quot;prevalence&quot;</span>,</span>
<span id="cb160-12"><a href="microbiome-exploration.html#cb160-12" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top=</span>10L,</span>
<span id="cb160-13"><a href="microbiome-exploration.html#cb160-13" aria-hidden="true" tabindex="-1"></a>                        <span class="at">abund_values=</span><span class="st">&quot;counts&quot;</span>)</span>
<span id="cb160-14"><a href="microbiome-exploration.html#cb160-14" aria-hidden="true" tabindex="-1"></a>top_phyla_mean <span class="ot">&lt;-</span> <span class="fu">getTopTaxa</span>(<span class="fu">altExp</span>(se,<span class="st">&quot;Phylum&quot;</span>),</span>
<span id="cb160-15"><a href="microbiome-exploration.html#cb160-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">method=</span><span class="st">&quot;mean&quot;</span>,</span>
<span id="cb160-16"><a href="microbiome-exploration.html#cb160-16" aria-hidden="true" tabindex="-1"></a>                             <span class="at">top=</span>10L,</span>
<span id="cb160-17"><a href="microbiome-exploration.html#cb160-17" aria-hidden="true" tabindex="-1"></a>                             <span class="at">abund_values=</span><span class="st">&quot;counts&quot;</span>)</span>
<span id="cb160-18"><a href="microbiome-exploration.html#cb160-18" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">unsplitByRanks</span>(se, <span class="at">ranks =</span> <span class="fu">taxonomyRanks</span>(se)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb160-19"><a href="microbiome-exploration.html#cb160-19" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">addTaxonomyTree</span>(x)</span></code></pre></div>
<p>After some preparation the data is assembled and can be plotted via
<code>plotRowTree</code>.</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="microbiome-exploration.html#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(miaViz)</span>
<span id="cb161-2"><a href="microbiome-exploration.html#cb161-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRowTree</span>(x[<span class="fu">rowData</span>(x)<span class="sc">$</span>Phylum <span class="sc">%in%</span> top_phyla,],</span>
<span id="cb161-3"><a href="microbiome-exploration.html#cb161-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">edge_colour_by =</span> <span class="st">&quot;Phylum&quot;</span>,</span>
<span id="cb161-4"><a href="microbiome-exploration.html#cb161-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">tip_colour_by =</span> <span class="st">&quot;prevalence&quot;</span>,</span>
<span id="cb161-5"><a href="microbiome-exploration.html#cb161-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_colour_by =</span> <span class="st">&quot;prevalence&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-prev-prev"></span>
<img src="13_microbiome_exploration_files/figure-html/plot-prev-prev-1.png" alt="Prevalence of top phyla as judged by prevalence" width="672" />
<p class="caption">
Figure 7.1: Prevalence of top phyla as judged by prevalence
</p>
</div>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="microbiome-exploration.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRowTree</span>(x[<span class="fu">rowData</span>(x)<span class="sc">$</span>Phylum <span class="sc">%in%</span> top_phyla_mean,],</span>
<span id="cb162-2"><a href="microbiome-exploration.html#cb162-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">edge_colour_by =</span> <span class="st">&quot;Phylum&quot;</span>,</span>
<span id="cb162-3"><a href="microbiome-exploration.html#cb162-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">tip_colour_by =</span> <span class="st">&quot;prevalence&quot;</span>,</span>
<span id="cb162-4"><a href="microbiome-exploration.html#cb162-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_colour_by =</span> <span class="st">&quot;prevalence&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plot-prev-mean"></span>
<img src="13_microbiome_exploration_files/figure-html/plot-prev-mean-1.png" alt="Prevalence of top phyla as judged by mean abundance" width="672" />
<p class="caption">
Figure 7.2: Prevalence of top phyla as judged by mean abundance
</p>
</div>
</div>
</div>
<div id="abundance" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Abundance</h2>
<p>Another approach for exploring data is by visualizing abundance. <code>miaViz</code> offers
the function <code>plotAbundanceDensity</code> where most abundant taxa can be plotted
including several options.</p>
<p>In the following few demonstrations are shown, using the
<span class="citation">(<a href="#ref-Lahti2014" role="doc-biblioref">Lahti et al. 2014</a>)</span> dataset.</p>
<p>A Jitter plot based on relative abundance data, similar to the one presented at
<span class="citation">(<a href="#ref-Salosensaari2021" role="doc-biblioref">Salosensaari et al. 2021</a>)</span> supplementary figure 1, could be visualized as follows:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="microbiome-exploration.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loading data</span></span>
<span id="cb163-2"><a href="microbiome-exploration.html#cb163-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb163-3"><a href="microbiome-exploration.html#cb163-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">atlas1006</span>()</span>
<span id="cb163-4"><a href="microbiome-exploration.html#cb163-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-5"><a href="microbiome-exploration.html#cb163-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Counts relative abundances</span></span>
<span id="cb163-6"><a href="microbiome-exploration.html#cb163-6" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformSamples</span>(tse, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb163-7"><a href="microbiome-exploration.html#cb163-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-8"><a href="microbiome-exploration.html#cb163-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAbundanceDensity</span>(tse, <span class="at">layout =</span> <span class="st">&quot;jitter&quot;</span>, <span class="at">abund_values =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb163-9"><a href="microbiome-exploration.html#cb163-9" aria-hidden="true" tabindex="-1"></a>                     <span class="at">n =</span> <span class="dv">40</span>, <span class="at">point_size=</span><span class="dv">1</span>, <span class="at">point_shape=</span><span class="dv">19</span>, <span class="at">point_alpha=</span><span class="fl">0.1</span>) <span class="sc">+</span> </span>
<span id="cb163-10"><a href="microbiome-exploration.html#cb163-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>(<span class="at">label=</span>scales<span class="sc">::</span>percent)</span></code></pre></div>
<p><img src="13_microbiome_exploration_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>For instance, relative abundance values for the top 5 taxa can be
visualized as a density plot over a log scaled axis, using
“nationality” as an overlaying information:</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="microbiome-exploration.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAbundanceDensity</span>(tse, <span class="at">layout =</span> <span class="st">&quot;density&quot;</span>, <span class="at">abund_values =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb164-2"><a href="microbiome-exploration.html#cb164-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">n =</span> <span class="dv">5</span>, <span class="at">colour_by=</span><span class="st">&quot;nationality&quot;</span>, <span class="at">point_alpha=</span><span class="dv">1</span><span class="sc">/</span><span class="dv">10</span> ) <span class="sc">+</span></span>
<span id="cb164-3"><a href="microbiome-exploration.html#cb164-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>()</span></code></pre></div>
<p><img src="13_microbiome_exploration_files/figure-html/unnamed-chunk-8-1.png" width="672" />
## Session Info {-}</p>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] microbiomeDataSets_1.1.5       MultiAssayExperiment_1.19.16  
 [3] miaViz_1.1.4                   ggraph_2.0.5                  
 [5] scater_1.21.8                  ggplot2_3.3.5                 
 [7] scuttle_1.3.1                  mia_1.1.19                    
 [9] TreeSummarizedExperiment_2.1.4 Biostrings_2.61.2             
[11] XVector_0.33.0                 SingleCellExperiment_1.15.2   
[13] SummarizedExperiment_1.23.5    Biobase_2.53.0                
[15] GenomicRanges_1.45.0           GenomeInfoDb_1.29.8           
[17] IRanges_2.27.2                 S4Vectors_0.31.5              
[19] BiocGenerics_0.39.2            MatrixGenerics_1.5.4          
[21] matrixStats_0.61.0-9001        BiocStyle_2.21.4              
[23] rebook_1.3.1                  

loaded via a namespace (and not attached):
  [1] AnnotationHub_3.1.6           BiocFileCache_2.1.1          
  [3] plyr_1.8.6                    igraph_1.2.6                 
  [5] lazyeval_0.2.2                splines_4.1.1                
  [7] BiocParallel_1.27.17          digest_0.6.28                
  [9] yulab.utils_0.0.4             htmltools_0.5.2              
 [11] viridis_0.6.2                 fansi_0.5.0                  
 [13] magrittr_2.0.1                memoise_2.0.0                
 [15] ScaledMatrix_1.1.0            cluster_2.1.2                
 [17] DECIPHER_2.21.0               graphlayouts_0.7.1           
 [19] colorspace_2.0-2              rappdirs_0.3.3               
 [21] blob_1.2.2                    ggrepel_0.9.1                
 [23] xfun_0.26                     dplyr_1.0.7                  
 [25] crayon_1.4.1                  RCurl_1.98-1.5               
 [27] jsonlite_1.7.2                graph_1.71.2                 
 [29] ape_5.5                       glue_1.4.2                   
 [31] polyclip_1.10-0               gtable_0.3.0                 
 [33] zlibbioc_1.39.0               DelayedArray_0.19.4          
 [35] BiocSingular_1.9.1            scales_1.1.1                 
 [37] DBI_1.1.1                     Rcpp_1.0.7                   
 [39] xtable_1.8-4                  viridisLite_0.4.0            
 [41] decontam_1.13.0               gridGraphics_0.5-1           
 [43] tidytree_0.3.5                bit_4.0.4                    
 [45] rsvd_1.0.5                    httr_1.4.2                   
 [47] dir.expiry_1.1.0              ellipsis_0.3.2               
 [49] pkgconfig_2.0.3               XML_3.99-0.8                 
 [51] farver_2.1.0                  dbplyr_2.1.1                 
 [53] CodeDepends_0.6.5             sass_0.4.0                   
 [55] utf8_1.2.2                    AnnotationDbi_1.55.1         
 [57] later_1.3.0                   ggplotify_0.1.0              
 [59] tidyselect_1.1.1              labeling_0.4.2               
 [61] rlang_0.4.11                  reshape2_1.4.4               
 [63] BiocVersion_3.14.0            munsell_0.5.0                
 [65] tools_4.1.1                   cachem_1.0.6                 
 [67] ExperimentHub_2.1.4           DirichletMultinomial_1.35.0  
 [69] generics_0.1.0                RSQLite_2.2.8                
 [71] evaluate_0.14                 stringr_1.4.0                
 [73] fastmap_1.1.0                 yaml_2.2.1                   
 [75] ggtree_3.1.5                  knitr_1.36                   
 [77] bit64_4.0.5                   tidygraph_1.2.0              
 [79] purrr_0.3.4                   KEGGREST_1.33.0              
 [81] nlme_3.1-153                  sparseMatrixStats_1.5.3      
 [83] mime_0.12                     aplot_0.1.1                  
 [85] compiler_4.1.1                png_0.1-7                    
 [87] interactiveDisplayBase_1.31.2 curl_4.3.2                   
 [89] beeswarm_0.4.0                filelock_1.0.2               
 [91] treeio_1.17.2                 tibble_3.1.5                 
 [93] tweenr_1.0.2                  bslib_0.3.1                  
 [95] stringi_1.7.5                 highr_0.9                    
 [97] lattice_0.20-45               Matrix_1.3-4                 
 [99] vegan_2.5-7                   permute_0.9-5                
[101] vctrs_0.3.8                   pillar_1.6.3                 
[103] lifecycle_1.0.1               BiocManager_1.30.16          
[105] jquerylib_0.1.4               BiocNeighbors_1.11.0         
[107] cowplot_1.1.1                 bitops_1.0-7                 
[109] irlba_2.3.3                   httpuv_1.6.3                 
[111] patchwork_1.1.1               R6_2.5.1                     
[113] promises_1.2.0.1              bookdown_0.24                
[115] gridExtra_2.3                 vipor_0.4.5                  
[117] codetools_0.2-18              MASS_7.3-54                  
[119] assertthat_0.2.1              withr_2.4.2                  
[121] GenomeInfoDbData_1.2.7        mgcv_1.8-38                  
[123] parallel_4.1.1                grid_4.1.1                   
[125] ggfun_0.0.4                   beachmat_2.9.1               
[127] tidyr_1.1.4                   rmarkdown_2.11               
[129] DelayedMatrixStats_1.15.4     ggnewscale_0.4.5             
[131] ggforce_0.3.3                 shiny_1.7.1                  
[133] ggbeeswarm_0.6.0             </code></pre>
</div>

</div>
</div>
<h3>Bibliography</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Lahti2014" class="csl-entry">
Lahti, Leo, Jarkko Salojärvi, Anne Salonen, Marten Scheffer, and Willem M. de Vos. 2014. <span>“Tipping Elements in the Human Intestinal Ecosystem.”</span> <em>Nature Communications</em> 2014: 1–10. https://doi.org/<a href="https://doi.org/10.1038/ncomms5344">https://doi.org/10.1038/ncomms5344</a>.
</div>
<div id="ref-Salosensaari2021" class="csl-entry">
Salosensaari, Aaro, Ville Laitinen, Aki Havulinna, Guillaume Méric, Susan Cheng, Markus Perola, Liisa Valsta, et al. 2021. <span>“Taxonomic Signatures of Cause-Specific Mortality Risk in Human Gut Microbiome.”</span> <em>Nature Communications</em> 12: 1–8. <a href="https://www.nature.com/articles/s41467-021-22962-y">https://www.nature.com/articles/s41467-021-22962-y</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="quality-control.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="alpha-diversity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
