<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Microbiome Diversity | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 6 Microbiome Diversity | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Microbiome Diversity | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Microbiome Diversity | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="microbiome-exploration.html"/>
<link rel="next" href="microbiome-community.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="data-introduction.html"><a href="data-introduction.html"><i class="fa fa-check"></i><b>1</b> Data Infrastructure</a>
<ul>
<li class="chapter" data-level="1.1" data-path="data-introduction.html"><a href="data-introduction.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="data-introduction.html"><a href="data-introduction.html#packages"><i class="fa fa-check"></i><b>1.1.1</b> Packages</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="data-introduction.html"><a href="data-introduction.html#background"><i class="fa fa-check"></i><b>1.2</b> Background</a></li>
<li class="chapter" data-level="1.3" data-path="data-introduction.html"><a href="data-introduction.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>1.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="data-introduction.html"><a href="data-introduction.html#importing-data-from-external-files"><i class="fa fa-check"></i><b>1.3.1</b> Importing data from external files</a></li>
<li class="chapter" data-level="1.3.2" data-path="data-introduction.html"><a href="data-introduction.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>1.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="data-introduction.html"><a href="data-introduction.html#metadata"><i class="fa fa-check"></i><b>1.4</b> Metadata</a></li>
<li class="chapter" data-level="1.5" data-path="data-introduction.html"><a href="data-introduction.html#microbiome-and-tree-data-specific-aspects"><i class="fa fa-check"></i><b>1.5</b> Microbiome and tree data specific aspects</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="data-introduction.html"><a href="data-introduction.html#assays"><i class="fa fa-check"></i><b>1.5.1</b> Assays</a></li>
<li class="chapter" data-level="1.5.2" data-path="data-introduction.html"><a href="data-introduction.html#coldata"><i class="fa fa-check"></i><b>1.5.2</b> colData</a></li>
<li class="chapter" data-level="1.5.3" data-path="data-introduction.html"><a href="data-introduction.html#rowdata"><i class="fa fa-check"></i><b>1.5.3</b> rowData</a></li>
<li class="chapter" data-level="1.5.4" data-path="data-introduction.html"><a href="data-introduction.html#rowtree"><i class="fa fa-check"></i><b>1.5.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="data-introduction.html"><a href="data-introduction.html#data-conversion"><i class="fa fa-check"></i><b>1.6</b> Data conversion</a>
<ul>
<li class="chapter" data-level="1.6.1" data-path="data-introduction.html"><a href="data-introduction.html#tidy-data"><i class="fa fa-check"></i><b>1.6.1</b> Tidy data</a></li>
</ul></li>
<li class="chapter" data-level="1.7" data-path="data-introduction.html"><a href="data-introduction.html#conclusion"><i class="fa fa-check"></i><b>1.7</b> Conclusion</a></li>
<li class="chapter" data-level="1.8" data-path="data-introduction.html"><a href="data-introduction.html#additional-reading"><i class="fa fa-check"></i><b>1.8</b> Additional Reading</a>
<ul>
<li class="chapter" data-level="1.8.1" data-path="data-introduction.html"><a href="data-introduction.html#lecture-slides"><i class="fa fa-check"></i><b>1.8.1</b> Lecture slides</a></li>
<li class="chapter" data-level="1.8.2" data-path="data-introduction.html"><a href="data-introduction.html#r-programming-resources"><i class="fa fa-check"></i><b>1.8.2</b> R programming resources</a></li>
<li class="chapter" data-level="1.8.3" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>1.8.3</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="1.8.4" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>1.8.4</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="1.8.5" data-path="data-introduction.html"><a href="data-introduction.html#further-reading"><i class="fa fa-check"></i><b>1.8.5</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-introduction.html"><a href="data-introduction.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="2" data-path="overview.html"><a href="overview.html"><i class="fa fa-check"></i><b>2</b> Overview</a>
<ul>
<li class="chapter" data-level="2.1" data-path="overview.html"><a href="overview.html#introduction"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="overview.html"><a href="overview.html#microbiome-experiment-types"><i class="fa fa-check"></i><b>2.2</b> Microbiome experiment types</a></li>
<li class="chapter" data-level="2.3" data-path="overview.html"><a href="overview.html#processing-steps"><i class="fa fa-check"></i><b>2.3</b> Processing steps</a></li>
<li class="chapter" data-level="2.4" data-path="overview.html"><a href="overview.html#quick-start"><i class="fa fa-check"></i><b>2.4</b> Quick Start</a></li>
<li class="chapter" data-level="" data-path="overview.html"><a href="overview.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>3</b> Taxonomic information</a>
<ul>
<li class="chapter" data-level="3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>3.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>3.1.1</b> dada2</a></li>
<li class="chapter" data-level="3.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>3.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>3.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>3.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>3.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="3.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>3.4</b> Pick specific</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>3.4.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="3.4.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>3.4.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>4</b> Quality Control</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quality-control.html"><a href="quality-control.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>4.1</b> Get top taxa and taxonomy</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quality-control.html"><a href="quality-control.html#features"><i class="fa fa-check"></i><b>4.1.1</b> Features</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>4.2</b> Library size</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html"><i class="fa fa-check"></i><b>5</b> Microbiome Exploration</a>
<ul>
<li class="chapter" data-level="5.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalence"><i class="fa fa-check"></i><b>5.1</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>5.1.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="5.1.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.1.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html"><i class="fa fa-check"></i><b>6</b> Microbiome Diversity</a>
<ul>
<li class="chapter" data-level="6.1" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#alpha-diversity"><i class="fa fa-check"></i><b>6.1</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#estimating-alpha-diversity"><i class="fa fa-check"></i><b>6.1.1</b> Estimating alpha diversity</a></li>
<li class="chapter" data-level="6.1.2" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#visualize-alpha-diversities"><i class="fa fa-check"></i><b>6.1.2</b> Visualize alpha diversities</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#beta-diversity"><i class="fa fa-check"></i><b>6.2</b> Beta diversity</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#explained-variance"><i class="fa fa-check"></i><b>6.2.1</b> Explained variance</a></li>
<li class="chapter" data-level="6.2.2" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#estimating-beta-diversity"><i class="fa fa-check"></i><b>6.2.2</b> Estimating beta diversity</a></li>
<li class="chapter" data-level="6.2.3" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#other-ordination-methods"><i class="fa fa-check"></i><b>6.2.3</b> Other ordination methods</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#community-comparisons-todo-combine-with-the-material-above-for-simplicity"><i class="fa fa-check"></i><b>6.3</b> Community comparisons [TODO combine with the material above for simplicity?]</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>6.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="6.3.2" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>6.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#further-reading-1"><i class="fa fa-check"></i><b>6.4</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="microbiome-diversity.html"><a href="microbiome-diversity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>7</b> Microbiome Community</a>
<ul>
<li class="chapter" data-level="7.1" data-path="microbiome-community.html"><a href="microbiome-community.html#community-composition"><i class="fa fa-check"></i><b>7.1</b> Community composition</a></li>
<li class="chapter" data-level="7.2" data-path="microbiome-community.html"><a href="microbiome-community.html#community-typing"><i class="fa fa-check"></i><b>7.2</b> Community typing</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="microbiome-community.html"><a href="microbiome-community.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>7.2.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="7.2.2" data-path="microbiome-community.html"><a href="microbiome-community.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>7.2.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-timeseries.html"><a href="microbiome-timeseries.html"><i class="fa fa-check"></i><b>9</b> Microbiome time series</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-timeseries.html"><a href="microbiome-timeseries.html#stability"><i class="fa fa-check"></i><b>9.1</b> Stability</a></li>
<li class="chapter" data-level="9.2" data-path="microbiome-timeseries.html"><a href="microbiome-timeseries.html#tipping-elements"><i class="fa fa-check"></i><b>9.2</b> Tipping elements</a></li>
<li class="chapter" data-level="" data-path="microbiome-timeseries.html"><a href="microbiome-timeseries.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="multitable.html"><a href="multitable.html"><i class="fa fa-check"></i><b>10</b> Representation of multiple data tables</a>
<ul>
<li class="chapter" data-level="10.1" data-path="multitable.html"><a href="multitable.html#assay-data"><i class="fa fa-check"></i><b>10.1</b> Assay data</a></li>
<li class="chapter" data-level="10.2" data-path="multitable.html"><a href="multitable.html#alternative-experiments"><i class="fa fa-check"></i><b>10.2</b> Alternative experiments</a></li>
<li class="chapter" data-level="10.3" data-path="multitable.html"><a href="multitable.html#multiassayexperiments"><i class="fa fa-check"></i><b>10.3</b> MultiAssayExperiments</a></li>
<li class="chapter" data-level="" data-path="multitable.html"><a href="multitable.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="11" data-path="example-data.html"><a href="example-data.html"><i class="fa fa-check"></i><b>11</b> Example data</a>
<ul>
<li class="chapter" data-level="11.1" data-path="example-data.html"><a href="example-data.html#package-data"><i class="fa fa-check"></i><b>11.1</b> Package data</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="example-data.html"><a href="example-data.html#globalpatterns"><i class="fa fa-check"></i><b>11.1.1</b> GlobalPatterns</a></li>
<li class="chapter" data-level="11.1.2" data-path="example-data.html"><a href="example-data.html#enterotype"><i class="fa fa-check"></i><b>11.1.2</b> Enterotype</a></li>
<li class="chapter" data-level="11.1.3" data-path="example-data.html"><a href="example-data.html#esophagus"><i class="fa fa-check"></i><b>11.1.3</b> Esophagus</a></li>
<li class="chapter" data-level="11.1.4" data-path="example-data.html"><a href="example-data.html#soilrep"><i class="fa fa-check"></i><b>11.1.4</b> Soilrep</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="example-data.html"><a href="example-data.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="12" data-path="function-ref.html"><a href="function-ref.html"><i class="fa fa-check"></i><b>12</b> Function reference</a>
<ul>
<li class="chapter" data-level="12.1" data-path="function-ref.html"><a href="function-ref.html#objects"><i class="fa fa-check"></i><b>12.1</b> Objects</a></li>
<li class="chapter" data-level="12.2" data-path="function-ref.html"><a href="function-ref.html#reduced-dimension-pca-t-sne-umap-mds-etc"><i class="fa fa-check"></i><b>12.2</b> Reduced dimension (PCA, t-SNE, UMAP, MDS, etc)</a></li>
<li class="chapter" data-level="12.3" data-path="function-ref.html"><a href="function-ref.html#plotting"><i class="fa fa-check"></i><b>12.3</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="microbiome-diversity" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> Microbiome Diversity</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Diversity estimates are a central topic in microbiome data analysis.</p>
<p>There are three commonly employed levels of diversity measurements,
which are trying to put a number on different aspects of the questions
associated with diversity <span class="citation">(<a href="#ref-Whittaker1960" role="doc-biblioref">Whittaker 1960</a>)</span>.</p>
<p>Many different ways for estimating such diversity measurements have been
described in the literature. Which measurement is best or applicable for your
samples, is not the aim of the following sections.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="microbiome-diversity.html#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb111-2"><a href="microbiome-diversity.html#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>)</span>
<span id="cb111-3"><a href="microbiome-diversity.html#cb111-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> GlobalPatterns</span></code></pre></div>
<div id="alpha-diversity" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Alpha diversity</h2>
<p><strong><em>Alpha diversity</em></strong>, also sometimes interchangeably used with the
term <strong><em>species diversity</em></strong>, summarizes the distribution of species
abundances in a given sample into a single number that depends on
species richness and evenness. Diversity indices measure the overall
community heterogeneity. A number of ecological diversity measures are
available. The Hill coefficient combines many standard indices into a
single equation that provides observed richness, inverse Simpson, and
Shannon diversity, and generalized diversity as special cases. In
general, diversity increases together with increasing richness and
evenness. Sometimes richness, evenness, and dominance are considered
to be variants of alpha diversity.</p>
<p><strong>Richness</strong> refers to the total number of species in a community
(sample). The simplest richness index is the number of observed
species (observed richness). Assuming limited sampling from the
community, however, this may underestimate the true species
richness. Several estimators are available, including for instance ACE
<span class="citation">(<a href="#ref-Chao1992" role="doc-biblioref">A and SM 1992</a>)</span> and Chao1 <span class="citation">(<a href="#ref-Chao1984" role="doc-biblioref">A 1984</a>)</span>. Richness estimates are unaffected
by species abundances.</p>
<p><strong>Phylogenetic diversity</strong> was first proposed by <span class="citation">(<a href="#ref-Faith1992" role="doc-biblioref">Faith 1992</a>)</span>, unlike the
diversity measures mentioned above, Phylogenetic diversity (PD)
measure incorporates information from phylogenetic relationships
stored in <code>phylo</code> tree between species in a community (sample). The
Faith’s PD is calculated as the sum of branch length of all species in
a community (sample).</p>
<p><strong>Evenness</strong> focuses on species abundances, and can thus complement
the number of species. A typical evenness index is the Pielou’s
evenness, which is Shannon diversity normalized by the observed
richness.</p>
<p><strong>Dominance</strong> indices are in general negatively correlated with
diversity, and sometimes used in ecological literature. High
dominance is obtained when one or few species have a high share of
the total species abundance in the community.</p>
<div id="estimating-alpha-diversity" class="section level3" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Estimating alpha diversity</h3>
<p>Alpha diversity can be estimated with wrapper functions that interact
with other packages implementing the calculation, such as <em><code>vegan</code></em>
<span class="citation">(<a href="#ref-R-vegan" role="doc-biblioref">Oksanen et al. 2020</a>)</span>.</p>
<div id="richness" class="section level4" number="6.1.1.1">
<h4><span class="header-section-number">6.1.1.1</span> Richness</h4>
<p>Richness gives the number of features present within a community and can be calculated with <code>estimateRichness</code>. Each of the estimate diversity/richness/evenness/dominance functions adds the calculated measure(s) to the <code>colData</code> of the <code>SummarizedExperiment</code> under the given column <code>name</code>. Here, we calculate <code>observed</code> features as a measure of richness.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="microbiome-diversity.html#cb112-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateRichness</span>(se, </span>
<span id="cb112-2"><a href="microbiome-diversity.html#cb112-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb112-3"><a href="microbiome-diversity.html#cb112-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">index =</span> <span class="st">&quot;observed&quot;</span>, </span>
<span id="cb112-4"><a href="microbiome-diversity.html#cb112-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">name=</span><span class="st">&quot;observed&quot;</span>)</span>
<span id="cb112-5"><a href="microbiome-diversity.html#cb112-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-6"><a href="microbiome-diversity.html#cb112-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se)<span class="sc">$</span>observed)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
##    6964    7679    5729    2667    2574    3214</code></pre>
<p>This allows access to the values to be analyzed directly from the <code>colData</code>, for example
by plotting them using <code>plotColData</code> from the <em><code>scater</code></em> package <span class="citation">(<a href="#ref-R-scater" role="doc-biblioref">McCarthy et al. 2020</a>)</span>.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="microbiome-diversity.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb114-2"><a href="microbiome-diversity.html#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(se, </span>
<span id="cb114-3"><a href="microbiome-diversity.html#cb114-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;observed&quot;</span>, </span>
<span id="cb114-4"><a href="microbiome-diversity.html#cb114-4" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;SampleType&quot;</span>, </span>
<span id="cb114-5"><a href="microbiome-diversity.html#cb114-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>) <span class="sc">+</span></span>
<span id="cb114-6"><a href="microbiome-diversity.html#cb114-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>,<span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb114-7"><a href="microbiome-diversity.html#cb114-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(Richness[Observed]))</span></code></pre></div>
<div class="figure"><span id="fig:plot-div-shannon"></span>
<img src="14-microbiome-diversity_files/figure-html/plot-div-shannon-1.png" alt="Shannon diversity estimates plotted grouped by sample type." width="672" />
<p class="caption">
Figure 6.1: Shannon diversity estimates plotted grouped by sample type.
</p>
</div>
</div>
<div id="diversity" class="section level4" number="6.1.1.2">
<h4><span class="header-section-number">6.1.1.2</span> Diversity</h4>
<p><strong>Non-Phylogenetic measures</strong><br />
The main function, <code>estimateDiversity</code>, calculates the selected
diversity index based on the selected assay data.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="microbiome-diversity.html#cb115-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateDiversity</span>(se, </span>
<span id="cb115-2"><a href="microbiome-diversity.html#cb115-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb115-3"><a href="microbiome-diversity.html#cb115-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">index =</span> <span class="st">&quot;shannon&quot;</span>, </span>
<span id="cb115-4"><a href="microbiome-diversity.html#cb115-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">name =</span> <span class="st">&quot;shannon&quot;</span>)</span>
<span id="cb115-5"><a href="microbiome-diversity.html#cb115-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se)<span class="sc">$</span>shannon)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
##   6.577   6.777   6.498   3.828   3.288   4.289</code></pre>
<p><strong>Phylogenetic diversity</strong></p>
<p>The phylogenetic diversity is calculated by <code>mia::estimateDiversity</code>. This is a faster re-implementation of<br />
the widely function in <em><code>picante</code></em> <span class="citation"><a href="#ref-Kembel2010" role="doc-biblioref">W et al.</a> (<a href="#ref-Kembel2010" role="doc-biblioref">2010</a>)</span>.</p>
<p>Load <code>picante</code> R package and get the <code>phylo</code> stored in <code>rowTree</code>.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="microbiome-diversity.html#cb117-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateDiversity</span>(se, </span>
<span id="cb117-2"><a href="microbiome-diversity.html#cb117-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb117-3"><a href="microbiome-diversity.html#cb117-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">index =</span> <span class="st">&quot;faith&quot;</span>, </span>
<span id="cb117-4"><a href="microbiome-diversity.html#cb117-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">name =</span> <span class="st">&quot;faith&quot;</span>)</span>
<span id="cb117-5"><a href="microbiome-diversity.html#cb117-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se)<span class="sc">$</span>faith)</span></code></pre></div>
<pre><code>## [1] 250.5 262.3 208.5 117.9 119.8 135.8</code></pre>
</div>
<div id="evenness" class="section level4" number="6.1.1.3">
<h4><span class="header-section-number">6.1.1.3</span> Evenness</h4>
<p>Evenness can be calculated with <code>estimateEvenness</code>.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="microbiome-diversity.html#cb119-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">estimateEvenness</span>(se, </span>
<span id="cb119-2"><a href="microbiome-diversity.html#cb119-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb119-3"><a href="microbiome-diversity.html#cb119-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">index=</span><span class="st">&quot;simpson&quot;</span>)</span>
<span id="cb119-4"><a href="microbiome-diversity.html#cb119-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se)<span class="sc">$</span>simpson)</span></code></pre></div>
<pre><code>## [1] 0.026871 0.027197 0.047049 0.005179 0.004304 0.005011</code></pre>
</div>
<div id="dominance" class="section level4" number="6.1.1.4">
<h4><span class="header-section-number">6.1.1.4</span> Dominance</h4>
<p>Dominance can be calculated with <code>estimateDominance</code>. Here, the <code>Relative index</code> is calculated which is the relative abundance of the most dominant species in the sample.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="microbiome-diversity.html#cb121-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">estimateDominance</span>(se, </span>
<span id="cb121-2"><a href="microbiome-diversity.html#cb121-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">abund_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb121-3"><a href="microbiome-diversity.html#cb121-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">index=</span><span class="st">&quot;relative&quot;</span>)</span>
<span id="cb121-4"><a href="microbiome-diversity.html#cb121-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-5"><a href="microbiome-diversity.html#cb121-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(se)<span class="sc">$</span>relative)</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
## 0.03910 0.03226 0.01690 0.22981 0.21778 0.22329</code></pre>
</div>
<div id="rarity" class="section level4" number="6.1.1.5">
<h4><span class="header-section-number">6.1.1.5</span> Rarity</h4>
<p>TODO…</p>
</div>
</div>
<div id="visualize-alpha-diversities" class="section level3" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Visualize alpha diversities</h3>
<p>A plot comparing all the diversity measures calculated above and stored in <code>colData</code> can then be constructed directly.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="microbiome-diversity.html#cb123-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;observed&quot;</span>, <span class="st">&quot;shannon&quot;</span>,<span class="st">&quot;simpson&quot;</span>, <span class="st">&quot;relative&quot;</span>, <span class="st">&quot;faith&quot;</span>),</span>
<span id="cb123-2"><a href="microbiome-diversity.html#cb123-2" aria-hidden="true" tabindex="-1"></a>                plotColData,</span>
<span id="cb123-3"><a href="microbiome-diversity.html#cb123-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">object =</span> se,</span>
<span id="cb123-4"><a href="microbiome-diversity.html#cb123-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">x =</span> <span class="st">&quot;SampleType&quot;</span>,</span>
<span id="cb123-5"><a href="microbiome-diversity.html#cb123-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span>
<span id="cb123-6"><a href="microbiome-diversity.html#cb123-6" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plots,<span class="st">&quot;+&quot;</span>, <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>,<span class="at">hjust=</span><span class="dv">1</span>)))</span>
<span id="cb123-7"><a href="microbiome-diversity.html#cb123-7" aria-hidden="true" tabindex="-1"></a>ggpubr<span class="sc">::</span><span class="fu">ggarrange</span>(<span class="at">plotlist =</span> plots, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend =</span> <span class="st">&quot;right&quot;</span>)</span></code></pre></div>
<p><img src="14-microbiome-diversity_files/figure-html/plot-all-diversities-1.png" width="672" /></p>
</div>
</div>
<div id="beta-diversity" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Beta diversity</h2>
<p>Where alpha diversity focuses on community variation within a
community (sample), beta diversity quantifies (dis-)similarites
between communities (samples). Some of the most popular beta diversity
measures in microbiome research include Bray-Curtis index (for
compositional data), Jaccard index (for presence / absence data,
ignoring abundance information), Aitchison distance (Euclidean
distance for clr transformed abundances, aiming to avoid the
compositionality bias), and the Unifrac distances (that take into
account the phylogenetic tree information). Only some of the commonly
used beta diversity measures are actual <em>distances</em>; this is a
mathematically well-defined concept and many ecological beta diversity
measures, such as Bray-Curtis index, are not proper distances.
Therefore, the term dissimilarity or beta diversity is commonly used.</p>
<p>Technically, beta diversities are usually represented as <code>dist</code>
objects, which contain triangular data describing the distance between
each pair of samples. These distances can be further subjected to
ordination. Ordination is a common concept in ecology that aims to
reduce the dimensionsionality of the data for further evaluation or
visualization. Ordination techniques aim to capture as much of
essential information in the data as possible in a lower dimensional
representation. Dimension reduction is bound to loose information but
the common ordination techniques aim to preserve relevant information
of sample similarities in an optimal way, which is defined in
different way by different methods. [TODO add references and/or link
to ordination chapter instead?]</p>
<p>Some of the most common ordination methods in microbiome research
include Principal Component Analysis (PCA), metric and non-metric
multi-dimensional scaling (MDS, NMDS), The MDS methods is also known
as Principal Coordinates Analysis (PCoA). Other recently popular
techniques include t-SNE and UMAP.</p>
<div id="explained-variance" class="section level3" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Explained variance</h3>
<p>The percentage of explained variance is typically shown for PCA
ordination plots. This quantifies the proportion of overall variance
in the data that is captured by the PCA axes, or how well the
ordination axes reflect the original distances.</p>
<p>Sometimes a similar measure is shown for MDS/PCoA. The interpretation
is generally different, however, and hence we do not recommend using
it. PCA is a special case of PCoA with Euclidean distances. With
non-Euclidean dissimilarities PCoA uses a trick where the pointwise
dissimilarities are first cast into similarities a Euclidean space
(with some information loss i.e. stress) and then projected to the
maximal variance axes. In this case, the maximal variance axes do not
directly reflect the correspondence of the projected distances and
original distances, as they do for PCA.</p>
<p>In typical use cases, we would like to know how well the ordination
reflects the original similarity structures; then the quantity of
interest is the so-called “stress” function, which measures the
difference in pairwise similarities between the data points in the
original (high-dimensional) vs. projected (low-dimensional) space.</p>
<p>Hence, we propose that for PCoA and other ordination methods, users
would report relative stress (varies in the unit interval; the smaller
the better). This can be calculated as shown below. For further
examples, check the <a href="https://www.huber.embl.de/users/klaus/Teaching/statisticalMethods-lab.pdf">note from Huber
lab</a>.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="microbiome-diversity.html#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example data</span></span>
<span id="cb124-2"><a href="microbiome-diversity.html#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns)</span>
<span id="cb124-3"><a href="microbiome-diversity.html#cb124-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-4"><a href="microbiome-diversity.html#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Data matrix (features x samples)</span></span>
<span id="cb124-5"><a href="microbiome-diversity.html#cb124-5" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb124-6"><a href="microbiome-diversity.html#cb124-6" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(x, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb124-7"><a href="microbiome-diversity.html#cb124-7" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">assay</span>(x, <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb124-8"><a href="microbiome-diversity.html#cb124-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-9"><a href="microbiome-diversity.html#cb124-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantify dissimilarities in the original feature space</span></span>
<span id="cb124-10"><a href="microbiome-diversity.html#cb124-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb124-11"><a href="microbiome-diversity.html#cb124-11" aria-hidden="true" tabindex="-1"></a>d0 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">vegdist</span>(<span class="fu">t</span>(x), <span class="st">&quot;bray&quot;</span>))</span>
<span id="cb124-12"><a href="microbiome-diversity.html#cb124-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-13"><a href="microbiome-diversity.html#cb124-13" aria-hidden="true" tabindex="-1"></a><span class="co"># PCoA Ordination</span></span>
<span id="cb124-14"><a href="microbiome-diversity.html#cb124-14" aria-hidden="true" tabindex="-1"></a>pcoa <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cmdscale</span>(d0, <span class="at">k =</span> <span class="dv">2</span>))</span>
<span id="cb124-15"><a href="microbiome-diversity.html#cb124-15" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pcoa) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PCoA1&quot;</span>, <span class="st">&quot;PCoA2&quot;</span>)</span>
<span id="cb124-16"><a href="microbiome-diversity.html#cb124-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-17"><a href="microbiome-diversity.html#cb124-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantify dissimilarities in the ordination space</span></span>
<span id="cb124-18"><a href="microbiome-diversity.html#cb124-18" aria-hidden="true" tabindex="-1"></a>dp <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">dist</span>(pcoa))</span>
<span id="cb124-19"><a href="microbiome-diversity.html#cb124-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-20"><a href="microbiome-diversity.html#cb124-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate stress i.e. relative difference in the original and</span></span>
<span id="cb124-21"><a href="microbiome-diversity.html#cb124-21" aria-hidden="true" tabindex="-1"></a><span class="co"># projected dissimilarities</span></span>
<span id="cb124-22"><a href="microbiome-diversity.html#cb124-22" aria-hidden="true" tabindex="-1"></a>stress <span class="ot">&lt;-</span> <span class="fu">sum</span>((dp <span class="sc">-</span> d0)<span class="sc">^</span><span class="dv">2</span>)<span class="sc">/</span><span class="fu">sum</span>(d0<span class="sc">^</span><span class="dv">2</span>)</span></code></pre></div>
<p>Shepard plot visualizes the original versus projected (ordination)
dissimilarities between the data points:</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="microbiome-diversity.html#cb125-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">as.vector</span>(d0))</span>
<span id="cb125-2"><a href="microbiome-diversity.html#cb125-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">d0 =</span> <span class="fu">as.vector</span>(d0)[ord],</span>
<span id="cb125-3"><a href="microbiome-diversity.html#cb125-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dmds =</span> <span class="fu">as.vector</span>(dp)[ord])</span>
<span id="cb125-4"><a href="microbiome-diversity.html#cb125-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb125-5"><a href="microbiome-diversity.html#cb125-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb125-6"><a href="microbiome-diversity.html#cb125-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> d0, <span class="at">y =</span> dmds), <span class="at">data=</span>df) <span class="sc">+</span> </span>
<span id="cb125-7"><a href="microbiome-diversity.html#cb125-7" aria-hidden="true" tabindex="-1"></a>       <span class="fu">geom_smooth</span>() <span class="sc">+</span></span>
<span id="cb125-8"><a href="microbiome-diversity.html#cb125-8" aria-hidden="true" tabindex="-1"></a>       <span class="fu">geom_point</span>() <span class="sc">+</span>       </span>
<span id="cb125-9"><a href="microbiome-diversity.html#cb125-9" aria-hidden="true" tabindex="-1"></a>       <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Shepard plot&quot;</span>,</span>
<span id="cb125-10"><a href="microbiome-diversity.html#cb125-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;Original distance&quot;</span>,</span>
<span id="cb125-11"><a href="microbiome-diversity.html#cb125-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;MDS distance&quot;</span>,       </span>
<span id="cb125-12"><a href="microbiome-diversity.html#cb125-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">subtitle =</span> <span class="fu">paste</span>(<span class="st">&quot;Stress:&quot;</span>, <span class="fu">round</span>(stress, <span class="dv">2</span>))) <span class="sc">+</span></span>
<span id="cb125-13"><a href="microbiome-diversity.html#cb125-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="14-microbiome-diversity_files/figure-html/shepard-1.png" width="672" /></p>
</div>
<div id="estimating-beta-diversity" class="section level3" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Estimating beta diversity</h3>
<p>In the following examples dissimilarities are calculated by
functions supplied to the <code>FUN</code> argument. This function can defined by
the user. It must return a <code>dist</code> function, which can then be used to
calculate reduced dimension either via ordination methods (such as MDS
or NMDS), and the results can be stored in the <code>reducedDim</code>.</p>
<p>This entire process is wrapped in the <code>runMDS</code> and <code>runNMDS</code>
functions.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="microbiome-diversity.html#cb126-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(se, <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist, <span class="at">name =</span> <span class="st">&quot;MDS_BC&quot;</span>, <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span></code></pre></div>
<p>Sample similarities can be visualized on a lower-dimensional display
(typically 2D) using the <code>plotReducedDim</code> function in the
<code>scater</code>package. This provides also further tools to incorporate
additional information using variations in colour, shape or size.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="microbiome-diversity.html#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create ggplot object</span></span>
<span id="cb127-2"><a href="microbiome-diversity.html#cb127-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plotReducedDim</span>(se, <span class="st">&quot;MDS_BC&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span>
<span id="cb127-3"><a href="microbiome-diversity.html#cb127-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-4"><a href="microbiome-diversity.html#cb127-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add explained variance for each axis</span></span>
<span id="cb127-5"><a href="microbiome-diversity.html#cb127-5" aria-hidden="true" tabindex="-1"></a>e <span class="ot">&lt;-</span> <span class="fu">attr</span>(<span class="fu">reducedDim</span>(se, <span class="st">&quot;MDS_BC&quot;</span>), <span class="st">&quot;eig&quot;</span>);</span>
<span id="cb127-6"><a href="microbiome-diversity.html#cb127-6" aria-hidden="true" tabindex="-1"></a>rel_eig <span class="ot">&lt;-</span> e<span class="sc">/</span><span class="fu">sum</span>(e[e<span class="sc">&gt;</span><span class="dv">0</span>])          </span>
<span id="cb127-7"><a href="microbiome-diversity.html#cb127-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste</span>(<span class="st">&quot;PCoA 1 (&quot;</span>, <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> rel_eig[[<span class="dv">1</span>]],<span class="dv">1</span>), <span class="st">&quot;%&quot;</span>, <span class="st">&quot;)&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb127-8"><a href="microbiome-diversity.html#cb127-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">y =</span> <span class="fu">paste</span>(<span class="st">&quot;PCoA 2 (&quot;</span>, <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> rel_eig[[<span class="dv">2</span>]],<span class="dv">1</span>), <span class="st">&quot;%&quot;</span>, <span class="st">&quot;)&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span>
<span id="cb127-9"><a href="microbiome-diversity.html#cb127-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-10"><a href="microbiome-diversity.html#cb127-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p)</span></code></pre></div>
<div class="figure"><span id="fig:plot-mds-bray-curtis"></span>
<img src="14-microbiome-diversity_files/figure-html/plot-mds-bray-curtis-1.png" alt="MDS plot based on the Bray-Curtis distances on the GlobalPattern dataset." width="672" />
<p class="caption">
Figure 6.2: MDS plot based on the Bray-Curtis distances on the GlobalPattern dataset.
</p>
</div>
<p>With additional tools from the <code>ggplot2</code> universe, comparisons can be
performed informing on the applicability to visualize sample similarities in a
meaningful way.</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="microbiome-diversity.html#cb128-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(se, <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist, <span class="at">name =</span> <span class="st">&quot;MDS_euclidean&quot;</span>,</span>
<span id="cb128-2"><a href="microbiome-diversity.html#cb128-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb128-3"><a href="microbiome-diversity.html#cb128-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runNMDS</span>(se, <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist, <span class="at">name =</span> <span class="st">&quot;NMDS_BC&quot;</span>)</span></code></pre></div>
<pre><code>## initial  value 47.733208 
## iter   5 value 33.853364
## iter  10 value 32.891200
## final  value 32.823570 
## converged</code></pre>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="microbiome-diversity.html#cb130-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runNMDS</span>(se, <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist, <span class="at">name =</span> <span class="st">&quot;NMDS_euclidean&quot;</span>,</span>
<span id="cb130-2"><a href="microbiome-diversity.html#cb130-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>)</span></code></pre></div>
<pre><code>## initial  value 31.882673 
## final  value 31.882673 
## converged</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="microbiome-diversity.html#cb132-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;MDS_BC&quot;</span>,<span class="st">&quot;MDS_euclidean&quot;</span>,<span class="st">&quot;NMDS_BC&quot;</span>,<span class="st">&quot;NMDS_euclidean&quot;</span>),</span>
<span id="cb132-2"><a href="microbiome-diversity.html#cb132-2" aria-hidden="true" tabindex="-1"></a>                plotReducedDim, <span class="at">object =</span> se, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span>
<span id="cb132-3"><a href="microbiome-diversity.html#cb132-3" aria-hidden="true" tabindex="-1"></a>ggpubr<span class="sc">::</span><span class="fu">ggarrange</span>(<span class="at">plotlist =</span> plots, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">common.legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb132-4"><a href="microbiome-diversity.html#cb132-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">legend =</span> <span class="st">&quot;right&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:plot-mds-nmds-comparison"></span>
<img src="14-microbiome-diversity_files/figure-html/plot-mds-nmds-comparison-1.png" alt="Comparison of MDS and NMDS plots based on the Bray-Curtis or euclidean distances on the GlobalPattern dataset." width="672" />
<p class="caption">
Figure 6.3: Comparison of MDS and NMDS plots based on the Bray-Curtis or euclidean distances on the GlobalPattern dataset.
</p>
</div>
<p>The <em>UniFrac</em> method is a special case, as it requires data on the
relationship of features in form on a <code>phylo</code> tree. <code>calculateUniFrac</code>
performs the calculation to return a <code>dist</code> object, which can again be
used within <code>runMDS</code>.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="microbiome-diversity.html#cb133-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(se, <span class="at">FUN =</span> calculateUniFrac, <span class="at">name =</span> <span class="st">&quot;UniFrac&quot;</span>,</span>
<span id="cb133-2"><a href="microbiome-diversity.html#cb133-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">tree =</span> <span class="fu">rowTree</span>(se),</span>
<span id="cb133-3"><a href="microbiome-diversity.html#cb133-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">ntop =</span> <span class="fu">nrow</span>(se),</span>
<span id="cb133-4"><a href="microbiome-diversity.html#cb133-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="microbiome-diversity.html#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(se, <span class="st">&quot;UniFrac&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:plot-unifrac"></span>
<img src="14-microbiome-diversity_files/figure-html/plot-unifrac-1.png" alt="UniFrac distances scaled by MDS of the GlobalPattern dataset." width="672" />
<p class="caption">
Figure 6.4: UniFrac distances scaled by MDS of the GlobalPattern dataset.
</p>
</div>
</div>
<div id="other-ordination-methods" class="section level3" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> Other ordination methods</h3>
<p>Other dimension reduction methods, such as <code>PCA</code>, <code>t-SNE</code> and <code>UMAP</code> are
inherited directly from the <code>scater</code> package.</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="microbiome-diversity.html#cb135-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(se, <span class="at">name =</span> <span class="st">&quot;PCA&quot;</span>, <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">ncomponents =</span> <span class="dv">10</span>)</span></code></pre></div>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="microbiome-diversity.html#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(se, <span class="st">&quot;PCA&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:plot-pca"></span>
<img src="14-microbiome-diversity_files/figure-html/plot-pca-1.png" alt="PCA plot on the GlobalPatterns data set containing sample from different sources." width="672" />
<p class="caption">
Figure 6.5: PCA plot on the GlobalPatterns data set containing sample from different sources.
</p>
</div>
<p>As mentioned before, applicability of the different methods depends on your
sample set.</p>
<p>FIXME: let us switch to UMAP for the examples?</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="microbiome-diversity.html#cb137-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">runTSNE</span>(se, <span class="at">name =</span> <span class="st">&quot;TSNE&quot;</span>, <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">ncomponents =</span> <span class="dv">3</span>)</span></code></pre></div>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="microbiome-diversity.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(se, <span class="st">&quot;TSNE&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="at">ncomponents =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>))</span></code></pre></div>
<div class="figure"><span id="fig:plot-tsne"></span>
<img src="14-microbiome-diversity_files/figure-html/plot-tsne-1.png" alt="t-SNE plot on the GlobalPatterns data set containing sample from different sources." width="672" />
<p class="caption">
Figure 6.6: t-SNE plot on the GlobalPatterns data set containing sample from different sources.
</p>
</div>
</div>
</div>
<div id="community-comparisons-todo-combine-with-the-material-above-for-simplicity" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Community comparisons [TODO combine with the material above for simplicity?]</h2>
<p>A typical comparison of community composition starts with a visual
comparison of the groups on a 2D ordination.</p>
<p>Let us load an example data set:</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="microbiome-diversity.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microbiomeDataSets)</span>
<span id="cb139-2"><a href="microbiome-diversity.html#cb139-2" aria-hidden="true" tabindex="-1"></a>se.lahti <span class="ot">&lt;-</span> <span class="fu">LahtiMData</span>()</span></code></pre></div>
<p>Then we estimate relative abundances and MDS ordination based on
Bray-Curtis (BC) dissimilarity between the groups, and visualize the
results.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="microbiome-diversity.html#cb140-1" aria-hidden="true" tabindex="-1"></a>se.lahti <span class="ot">&lt;-</span> <span class="fu">relAbundanceCounts</span>(se.lahti)</span>
<span id="cb140-2"><a href="microbiome-diversity.html#cb140-2" aria-hidden="true" tabindex="-1"></a>se.lahti <span class="ot">&lt;-</span> <span class="fu">runNMDS</span>(se.lahti, <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist, <span class="at">name =</span> <span class="st">&quot;BC&quot;</span>, <span class="at">nmdsFUN =</span> <span class="st">&quot;monoMDS&quot;</span>,</span>
<span id="cb140-3"><a href="microbiome-diversity.html#cb140-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">exprs_values =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb140-4"><a href="microbiome-diversity.html#cb140-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keep_dist =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="microbiome-diversity.html#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(se.lahti, <span class="st">&quot;BC&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;group&quot;</span>)</span></code></pre></div>
<p><img src="14-microbiome-diversity_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>No clear difference between the groups can be visually observed.</p>
<div id="testing-differences-in-community-composition-between-sample-groups" class="section level3" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Testing differences in community composition between sample groups</h3>
<p>The permutational analysis of variance (PERMANOVA) <span class="citation">(<a href="#ref-Anderson2001" role="doc-biblioref">Anderson 2001</a>)</span> is
a widely used non-parametric multivariate method that can be used to
estimate the actual statistical significance of differences in the
observed community composition between two groups of
samples.</p>
<p>PERMANOVA evaluates the hypothesis that the centroids and dispersion
of the community are equivalent between the compared groups. A small
p-value indicates that the compared groups have, on average, a
different community composition.</p>
<p>This method is implemented in the <code>vegan</code> package (function <code>adonis</code>).</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="microbiome-diversity.html#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb142-2"><a href="microbiome-diversity.html#cb142-2" aria-hidden="true" tabindex="-1"></a>permanova <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis</span>(<span class="fu">t</span>(<span class="fu">assay</span>(se.lahti,<span class="st">&quot;relabundance&quot;</span>)) <span class="sc">~</span> group,</span>
<span id="cb142-3"><a href="microbiome-diversity.html#cb142-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">data =</span> <span class="fu">colData</span>(se.lahti),</span>
<span id="cb142-4"><a href="microbiome-diversity.html#cb142-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">permutations =</span> <span class="dv">9999</span>)</span>
<span id="cb142-5"><a href="microbiome-diversity.html#cb142-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-6"><a href="microbiome-diversity.html#cb142-6" aria-hidden="true" tabindex="-1"></a><span class="co"># P-value</span></span>
<span id="cb142-7"><a href="microbiome-diversity.html#cb142-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">as.data.frame</span>(permanova<span class="sc">$</span>aov.tab)[<span class="st">&quot;group&quot;</span>, <span class="st">&quot;Pr(&gt;F)&quot;</span>])</span></code></pre></div>
<pre><code>## [1] 0.2772</code></pre>
<p>In this case, the community composition is not significantly different
between the groups.</p>
<p>Let us visualize the model coefficients for species that exhibit the
largest differences between the groups. This gives some insights into
how the groups tend to differ from each other in terms of community
composition.</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="microbiome-diversity.html#cb144-1" aria-hidden="true" tabindex="-1"></a>coef <span class="ot">&lt;-</span> <span class="fu">coefficients</span>(permanova)[<span class="st">&quot;group1&quot;</span>,]</span>
<span id="cb144-2"><a href="microbiome-diversity.html#cb144-2" aria-hidden="true" tabindex="-1"></a>top.coef <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">head</span>(coef[<span class="fu">rev</span>(<span class="fu">order</span>(<span class="fu">abs</span>(coef)))],<span class="dv">20</span>))</span></code></pre></div>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="microbiome-diversity.html#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(<span class="at">x =</span> top.coef,</span>
<span id="cb145-2"><a href="microbiome-diversity.html#cb145-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">y =</span> <span class="fu">factor</span>(<span class="fu">names</span>(top.coef),</span>
<span id="cb145-3"><a href="microbiome-diversity.html#cb145-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="fu">unique</span>(<span class="fu">names</span>(top.coef)))),</span>
<span id="cb145-4"><a href="microbiome-diversity.html#cb145-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb145-5"><a href="microbiome-diversity.html#cb145-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb145-6"><a href="microbiome-diversity.html#cb145-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;&quot;</span>,<span class="at">y=</span><span class="st">&quot;&quot;</span>,<span class="at">title=</span><span class="st">&quot;Top Taxa&quot;</span>) <span class="sc">+</span></span>
<span id="cb145-7"><a href="microbiome-diversity.html#cb145-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="14-microbiome-diversity_files/figure-html/plot-top-coef-anova-1.png" width="672" /></p>
<p>In the above example, the largest differences between the two groups
can be attributed to <em>Bacteroides intestinalis</em> (elevated in the first
group) and <em>Faecalibacterium prausnitzii</em> (elevated in the second
group), and many other co-varying species.</p>
</div>
<div id="checking-the-homogeneity-condition" class="section level3" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Checking the homogeneity condition</h3>
<p>It is important to note that the application of PERMANOVA assumes
homogeneous group dispersions (variances). This can be tested with the
PERMDISP2 method <span class="citation">(<a href="#ref-Anderson2006" role="doc-biblioref">Anderson 2006</a>)</span>.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="microbiome-diversity.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(vegan<span class="sc">::</span><span class="fu">betadisper</span>(<span class="fu">attr</span>(<span class="fu">reducedDim</span>(se.lahti,<span class="st">&quot;BC&quot;</span>),<span class="st">&quot;dist&quot;</span>),</span>
<span id="cb146-2"><a href="microbiome-diversity.html#cb146-2" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">colData</span>(se.lahti)<span class="sc">$</span>group))</span></code></pre></div>
<pre><code>## Analysis of Variance Table
## 
## Response: Distances
##           Df Sum Sq Mean Sq F value Pr(&gt;F)
## Groups     1  0.000 0.00002       0   0.95
## Residuals 42  0.158 0.00376</code></pre>
<p>In our example, the groups have similar dispersion, and PERMANOVA is
an appropriate choice for comparing community compositions.</p>
</div>
</div>
<div id="further-reading-1" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Further reading</h2>
<p>In certain settings, beta diversities might be used to group samples without
prior knowledge. For this we want to point to excellent resources on
<a href="http://bioconductor.org/books/release/OSCA/clustering.html">how to extract information from the clusters</a>.</p>
<p>See also <a href="15-microbiome-community.md">community typing</a>.</p>
</div>
<div id="session-info-5" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] microbiomeDataSets_1.1.0       MultiAssayExperiment_1.19.1   
 [3] vegan_2.5-7                    lattice_0.20-44               
 [5] permute_0.9-5                  scater_1.21.2                 
 [7] ggplot2_3.3.5                  scuttle_1.3.0                 
 [9] mia_1.1.7                      TreeSummarizedExperiment_2.1.3
[11] Biostrings_2.61.1              XVector_0.33.0                
[13] SingleCellExperiment_1.15.1    SummarizedExperiment_1.23.1   
[15] Biobase_2.53.0                 GenomicRanges_1.45.0          
[17] GenomeInfoDb_1.29.3            IRanges_2.27.0                
[19] S4Vectors_0.31.0               BiocGenerics_0.39.1           
[21] MatrixGenerics_1.5.1           matrixStats_0.59.0            
[23] BiocStyle_2.21.3               rebook_1.3.0                  

loaded via a namespace (and not attached):
  [1] readxl_1.3.1                  backports_1.2.1              
  [3] AnnotationHub_3.1.3           BiocFileCache_2.1.1          
  [5] plyr_1.8.6                    lazyeval_0.2.2               
  [7] splines_4.1.0                 BiocParallel_1.27.2          
  [9] digest_0.6.27                 htmltools_0.5.1.1            
 [11] viridis_0.6.1                 fansi_0.5.0                  
 [13] magrittr_2.0.1                memoise_2.0.0                
 [15] ScaledMatrix_1.1.0            cluster_2.1.2                
 [17] DECIPHER_2.21.0               openxlsx_4.2.4               
 [19] colorspace_2.0-2              rappdirs_0.3.3               
 [21] blob_1.2.1                    haven_2.4.1                  
 [23] xfun_0.24                     dplyr_1.0.7                  
 [25] crayon_1.4.1                  RCurl_1.98-1.3               
 [27] jsonlite_1.7.2                graph_1.71.2                 
 [29] ape_5.5                       glue_1.4.2                   
 [31] gtable_0.3.0                  zlibbioc_1.39.0              
 [33] DelayedArray_0.19.1           car_3.0-11                   
 [35] BiocSingular_1.9.1            abind_1.4-5                  
 [37] scales_1.1.1                  DBI_1.1.1                    
 [39] rstatix_0.7.0                 Rcpp_1.0.7                   
 [41] xtable_1.8-4                  viridisLite_0.4.0            
 [43] decontam_1.13.0               tidytree_0.3.4               
 [45] foreign_0.8-81                bit_4.0.4                    
 [47] rsvd_1.0.5                    httr_1.4.2                   
 [49] dir.expiry_1.1.0              ellipsis_0.3.2               
 [51] pkgconfig_2.0.3               XML_3.99-0.6                 
 [53] farver_2.1.0                  dbplyr_2.1.1                 
 [55] CodeDepends_0.6.5             sass_0.4.0                   
 [57] utf8_1.2.1                    AnnotationDbi_1.55.1         
 [59] later_1.2.0                   tidyselect_1.1.1             
 [61] labeling_0.4.2                rlang_0.4.11                 
 [63] reshape2_1.4.4                BiocVersion_3.14.0           
 [65] cellranger_1.1.0              munsell_0.5.0                
 [67] tools_4.1.0                   cachem_1.0.5                 
 [69] ExperimentHub_2.1.1           DirichletMultinomial_1.35.0  
 [71] generics_0.1.0                RSQLite_2.2.7                
 [73] broom_0.7.8                   evaluate_0.14                
 [75] stringr_1.4.0                 fastmap_1.1.0                
 [77] yaml_2.2.1                    knitr_1.33                   
 [79] bit64_4.0.5                   zip_2.2.0                    
 [81] purrr_0.3.4                   KEGGREST_1.33.0              
 [83] nlme_3.1-152                  sparseMatrixStats_1.5.0      
 [85] mime_0.11                     compiler_4.1.0               
 [87] png_0.1-7                     interactiveDisplayBase_1.31.0
 [89] beeswarm_0.4.0                filelock_1.0.2               
 [91] curl_4.3.2                    ggsignif_0.6.2               
 [93] treeio_1.17.2                 tibble_3.1.2                 
 [95] bslib_0.2.5.1                 stringi_1.7.2                
 [97] highr_0.9                     forcats_0.5.1                
 [99] Matrix_1.3-4                  vctrs_0.3.8                  
[101] pillar_1.6.1                  lifecycle_1.0.0              
[103] BiocManager_1.30.16           jquerylib_0.1.4              
[105] BiocNeighbors_1.11.0          data.table_1.14.0            
[107] cowplot_1.1.1                 bitops_1.0-7                 
[109] irlba_2.3.3                   httpuv_1.6.1                 
[111] R6_2.5.0                      promises_1.2.0.1             
[113] bookdown_0.22                 gridExtra_2.3                
[115] rio_0.5.27                    vipor_0.4.5                  
[117] codetools_0.2-18              MASS_7.3-54                  
[119] assertthat_0.2.1              withr_2.4.2                  
[121] GenomeInfoDbData_1.2.6        mgcv_1.8-36                  
[123] parallel_4.1.0                hms_1.1.0                    
[125] grid_4.1.0                    beachmat_2.9.0               
[127] tidyr_1.1.3                   rmarkdown_2.9                
[129] DelayedMatrixStats_1.15.0     carData_3.0-4                
[131] Rtsne_0.15                    ggpubr_0.4.0                 
[133] shiny_1.6.0                   ggbeeswarm_0.6.0             </code></pre>
</div>

</div>
</div>
<h3>Bibliography</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Chao1984" class="csl-entry">
A, Chao. 1984. <span>“Non-Parametric Estimation of the Number of Classes in a Population.”</span> <em>Scandinavian Journal of Statistics</em> 11 (4): 265–70. <a href="https://www.jstor.org/stable/4615964">https://www.jstor.org/stable/4615964</a>.
</div>
<div id="ref-Chao1992" class="csl-entry">
A, Chao, and Lee SM. 1992. <span>“Estimating the Number of Classes via Sample Coverage.”</span> <em>Journal of the American Statistical Association</em> 87 (417): 210–17. https://doi.org/<a href="https://doi.org/10.1080/01621459.1992.10475194">https://doi.org/10.1080/01621459.1992.10475194</a>.
</div>
<div id="ref-Anderson2001" class="csl-entry">
Anderson, Marti J. 2001. <span>“A New Method for Non-Parametric Multivariate Analysis of Variance.”</span> <em>Austral Ecology</em> 26 (1): 32–46. <a href="https://doi.org/10.1111/j.1442-9993.2001.01070.pp.x">https://doi.org/10.1111/j.1442-9993.2001.01070.pp.x</a>.
</div>
<div id="ref-Anderson2006" class="csl-entry">
———. 2006. <span>“Distance-Based Tests for Homogeneity of Multivariate Dispersions.”</span> <em>Biometrics</em> 62: 245–53. <a href="https://doi.org/10.1111/j.1541-0420.2005.00440.x">https://doi.org/10.1111/j.1541-0420.2005.00440.x</a>.
</div>
<div id="ref-Faith1992" class="csl-entry">
Faith, D. P. 1992. <span>“Conservation Evaluation and Phylogenetic Diversity.”</span> <em>Biological Conservation</em> 61 (1): 10. https://doi.org/<a href="https://doi.org/10.1016/0006-3207(92)91201-3">https://doi.org/10.1016/0006-3207(92)91201-3</a>.
</div>
<div id="ref-R-scater" class="csl-entry">
McCarthy, Davis, Kieran Campbell, Aaron Lun, and Quin Wills. 2020. <em>Scater: Single-Cell Analysis Toolkit for Gene Expression Data in r</em>. <a href="http://bioconductor.org/packages/scater/">http://bioconductor.org/packages/scater/</a>.
</div>
<div id="ref-R-vegan" class="csl-entry">
Oksanen, Jari, F. Guillaume Blanchet, Michael Friendly, Roeland Kindt, Pierre Legendre, Dan McGlinn, Peter R. Minchin, et al. 2020. <em>Vegan: Community Ecology Package</em>. <a href="https://CRAN.R-project.org/package=vegan">https://CRAN.R-project.org/package=vegan</a>.
</div>
<div id="ref-Kembel2010" class="csl-entry">
W, Kembel Steven, Cowan Peter D, Helmus Matthew R, Cornwell William K, Morlon Helene, Ackerly David D, Blomberg Simon P, and Webb Campbell O. 2010. <span>“Picante: R Tools for Integrating Phylogenies and Ecology.”</span> <em>Bioinformatics</em> 26 (11): 1463–64. https://doi.org/<a href="https://doi.org/10.1093/bioinformatics/btq166">https://doi.org/10.1093/bioinformatics/btq166</a>.
</div>
<div id="ref-Whittaker1960" class="csl-entry">
Whittaker, R. H. 1960. <span>“Vegetation of the Siskiyou Mountains, Oregon and California.”</span> <em>Ecological Monographs</em> 30 (3): 279–338. https://doi.org/<a href="https://doi.org/10.2307/1943563">https://doi.org/10.2307/1943563</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="microbiome-exploration.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="microbiome-community.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
