<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 12 Representation of Multiple Data Tables | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 12 Representation of Multiple Data Tables | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 12 Representation of Multiple Data Tables | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/microbiome/OMA" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 12 Representation of Multiple Data Tables | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differential-abundance.html"/>
<link rel="next" href="function-ref.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="" data-path="introduction-to-bioconductor-classes.html"><a href="introduction-to-bioconductor-classes.html"><i class="fa fa-check"></i>Introduction to Bioconductor Classes</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#community"><i class="fa fa-check"></i><b>1.1</b> Community</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#developers-and-contributors"><i class="fa fa-check"></i><b>1.2</b> Developers and contributors</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#acknowledgements"><i class="fa fa-check"></i><b>1.3</b> Acknowledgements</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#package-ecosystem"><i class="fa fa-check"></i><b>1.4</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="introduction.html"><a href="introduction.html#data-container"><i class="fa fa-check"></i><b>1.4.1</b> Data container:</a></li>
<li class="chapter" data-level="1.4.2" data-path="introduction.html"><a href="introduction.html#packages"><i class="fa fa-check"></i><b>1.4.2</b> Packages</a></li>
<li class="chapter" data-level="1.4.3" data-path="introduction.html"><a href="introduction.html#demonstration-data"><i class="fa fa-check"></i><b>1.4.3</b> Demonstration data</a></li>
<li class="chapter" data-level="1.4.4" data-path="introduction.html"><a href="introduction.html#online-tutorials"><i class="fa fa-check"></i><b>1.4.4</b> Online tutorials</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-introduction.html"><a href="data-introduction.html"><i class="fa fa-check"></i><b>2</b> Data Infrastructure</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-introduction.html"><a href="data-introduction.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a></li>
<li class="chapter" data-level="2.2" data-path="data-introduction.html"><a href="data-introduction.html#background"><i class="fa fa-check"></i><b>2.2</b> Background</a></li>
<li class="chapter" data-level="2.3" data-path="data-introduction.html"><a href="data-introduction.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="data-introduction.html"><a href="data-introduction.html#importing-data-from-external-files"><i class="fa fa-check"></i><b>2.3.1</b> Importing data from external files</a></li>
<li class="chapter" data-level="2.3.2" data-path="data-introduction.html"><a href="data-introduction.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.2</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-introduction.html"><a href="data-introduction.html#metadata"><i class="fa fa-check"></i><b>2.4</b> Metadata</a></li>
<li class="chapter" data-level="2.5" data-path="data-introduction.html"><a href="data-introduction.html#microbiome-and-tree-data-specific-aspects"><i class="fa fa-check"></i><b>2.5</b> Microbiome and tree data specific aspects</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="data-introduction.html"><a href="data-introduction.html#assays"><i class="fa fa-check"></i><b>2.5.1</b> Assays</a></li>
<li class="chapter" data-level="2.5.2" data-path="data-introduction.html"><a href="data-introduction.html#coldata"><i class="fa fa-check"></i><b>2.5.2</b> colData</a></li>
<li class="chapter" data-level="2.5.3" data-path="data-introduction.html"><a href="data-introduction.html#rowdata"><i class="fa fa-check"></i><b>2.5.3</b> rowData</a></li>
<li class="chapter" data-level="2.5.4" data-path="data-introduction.html"><a href="data-introduction.html#rowtree"><i class="fa fa-check"></i><b>2.5.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="data-introduction.html"><a href="data-introduction.html#data-manipulation"><i class="fa fa-check"></i><b>2.6</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="data-introduction.html"><a href="data-introduction.html#tidy-data"><i class="fa fa-check"></i><b>2.6.1</b> Tidy data</a></li>
<li class="chapter" data-level="2.6.2" data-path="data-introduction.html"><a href="data-introduction.html#subsetting"><i class="fa fa-check"></i><b>2.6.2</b> Subsetting</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="data-introduction.html"><a href="data-introduction.html#additional-reading"><i class="fa fa-check"></i><b>2.7</b> Additional Reading</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="data-introduction.html"><a href="data-introduction.html#r-programming-resources"><i class="fa fa-check"></i><b>2.7.1</b> R programming resources</a></li>
<li class="chapter" data-level="2.7.2" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>2.7.2</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="2.7.3" data-path="data-introduction.html"><a href="data-introduction.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>2.7.3</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="2.7.4" data-path="data-introduction.html"><a href="data-introduction.html#further-reading"><i class="fa fa-check"></i><b>2.7.4</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-introduction.html"><a href="data-introduction.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Appendix</b></span></li>
<li class="chapter" data-level="3" data-path="example-data.html"><a href="example-data.html"><i class="fa fa-check"></i><b>3</b> Demonstration Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="example-data.html"><a href="example-data.html#package-data"><i class="fa fa-check"></i><b>3.1</b> Package data</a></li>
<li class="chapter" data-level="3.2" data-path="example-data.html"><a href="example-data.html#experimenthub-data"><i class="fa fa-check"></i><b>3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3" data-path="example-data.html"><a href="example-data.html#other-data-sources"><i class="fa fa-check"></i><b>3.3</b> Other data sources</a></li>
<li class="chapter" data-level="" data-path="example-data.html"><a href="example-data.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>4</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="4.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>4.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>4.1.1</b> dada2</a></li>
<li class="chapter" data-level="4.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>4.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>4.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>4.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>4.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="4.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>4.4</b> Data transformation</a></li>
<li class="chapter" data-level="4.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>4.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>4.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="4.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>4.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#get-top-taxa-and-taxonomy"><i class="fa fa-check"></i><b>5.1</b> Get top taxa and taxonomy</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="quality-control.html"><a href="quality-control.html#features"><i class="fa fa-check"></i><b>5.1.1</b> Features</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.2</b> Library size</a></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html"><i class="fa fa-check"></i><b>6</b> Microbiome Exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalence"><i class="fa fa-check"></i><b>6.1</b> Prevalence</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#prevalent-microbiota-analysis"><i class="fa fa-check"></i><b>6.1.1</b> Prevalent microbiota analysis</a></li>
<li class="chapter" data-level="6.1.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#plotting-prevalence"><i class="fa fa-check"></i><b>6.1.2</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="microbiome-exploration.html"><a href="microbiome-exploration.html#abundance"><i class="fa fa-check"></i><b>6.2</b> Abundance</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>7</b> Alpha Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#estimating-alpha-diversity"><i class="fa fa-check"></i><b>7.1</b> Estimating alpha diversity</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.3</b> Evenness</a></li>
<li class="chapter" data-level="7.1.4" data-path="alpha-diversity.html"><a href="alpha-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.4</b> Dominance</a></li>
<li class="chapter" data-level="7.1.5" data-path="alpha-diversity.html"><a href="alpha-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.5</b> Rarity</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualize-alpha-diversities"><i class="fa fa-check"></i><b>7.2</b> Visualize alpha diversities</a></li>
<li class="chapter" data-level="" data-path="alpha-diversity.html"><a href="alpha-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>8</b> Beta Diversity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="beta-diversity.html"><a href="beta-diversity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-beta-diversity"><i class="fa fa-check"></i><b>8.2</b> Estimating beta diversity</a></li>
<li class="chapter" data-level="8.3" data-path="beta-diversity.html"><a href="beta-diversity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="beta-diversity.html"><a href="beta-diversity.html#visualizing-the-most-dominant-genus-on-pcoa"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a></li>
<li class="chapter" data-level="8.5" data-path="beta-diversity.html"><a href="beta-diversity.html#community-comparisons-todo-combine-with-the-material-above-for-simplicity"><i class="fa fa-check"></i><b>8.5</b> Community comparisons [TODO combine with the material above for simplicity?]</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="beta-diversity.html"><a href="beta-diversity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.5.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.5.2" data-path="beta-diversity.html"><a href="beta-diversity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.5.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="beta-diversity.html"><a href="beta-diversity.html#further-reading-1"><i class="fa fa-check"></i><b>8.6</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="beta-diversity.html"><a href="beta-diversity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Microbiome Community</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#community-composition"><i class="fa fa-check"></i><b>9.1</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="community-typing.html"><a href="community-typing.html"><i class="fa fa-check"></i><b>10</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="community-typing.html"><a href="community-typing.html#community-typing-1"><i class="fa fa-check"></i><b>10.1</b> Community typing</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="community-typing.html"><a href="community-typing.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.1.1</b> Dirichlet Multinomial Mixtures (DMM)</a></li>
<li class="chapter" data-level="10.1.2" data-path="community-typing.html"><a href="community-typing.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.1.2</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="community-typing.html"><a href="community-typing.html#community-detection"><i class="fa fa-check"></i><b>10.2</b> Community Detection</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#introduction-to-differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Introduction to differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.1</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.6</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances-with-fido"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances with fido</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multitable.html"><a href="multitable.html"><i class="fa fa-check"></i><b>12</b> Representation of Multiple Data Tables</a>
<ul>
<li class="chapter" data-level="12.1" data-path="multitable.html"><a href="multitable.html#assay-data"><i class="fa fa-check"></i><b>12.1</b> Assay data</a></li>
<li class="chapter" data-level="12.2" data-path="multitable.html"><a href="multitable.html#alternative-experiments"><i class="fa fa-check"></i><b>12.2</b> Alternative experiments</a></li>
<li class="chapter" data-level="12.3" data-path="multitable.html"><a href="multitable.html#multiassayexperiments"><i class="fa fa-check"></i><b>12.3</b> MultiAssayExperiments</a></li>
<li class="chapter" data-level="" data-path="multitable.html"><a href="multitable.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="function-ref.html"><a href="function-ref.html"><i class="fa fa-check"></i><b>13</b> Function Reference</a>
<ul>
<li class="chapter" data-level="13.1" data-path="function-ref.html"><a href="function-ref.html#objects"><i class="fa fa-check"></i><b>13.1</b> Objects</a></li>
<li class="chapter" data-level="13.2" data-path="function-ref.html"><a href="function-ref.html#reduced-dimension-pca-t-sne-umap-mds-etc"><i class="fa fa-check"></i><b>13.2</b> Reduced dimension (PCA, t-SNE, UMAP, MDS, etc)</a></li>
<li class="chapter" data-level="13.3" data-path="function-ref.html"><a href="function-ref.html#plotting"><i class="fa fa-check"></i><b>13.3</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>14</b> Extras</a>
<ul>
<li class="chapter" data-level="14.1" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>14.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multitable" class="section level1" number="12">
<h1><span class="header-section-number">Chapter 12</span> Representation of Multiple Data Tables</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Microbiome data can be part of multiomics experiments and analysis strategies
and we want to outline the understanding in which we think the packages
explained and used in this book relate to these experiment layouts
using the <code>TreeSummarizedExperiment</code> and classes beyond.</p>
<p>Many microbiome experiments include multiple versions and types of
data generated independently or derived from each other through transformation
or agglomeration. We start by providing recommendations on how to represent
different varieties of multi-table data within the
<code>TreeSummarizedExperiment</code> class.</p>
<p>The options and recommendations are summarized in Table <a href="multitable.html#tab:options">12.1</a>.</p>
<div id="assay-data" class="section level2" number="12.1">
<h2><span class="header-section-number">12.1</span> Assay data</h2>
<p>The original count-based taxonomic abundance tables may have different
transformations, such as logarithmic, Centered Log-Ratio (CLR), or relative
abundance. These are typically stored in <em><strong>assays</strong></em>.</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="multitable.html#cb270-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb270-2"><a href="multitable.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns)</span>
<span id="cb270-3"><a href="multitable.html#cb270-3" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb270-4"><a href="multitable.html#cb270-4" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code></pre></div>
<pre><code>## List of length 1
## names(1): counts</code></pre>
<p>As an example, the relative abundance is calculated.</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="multitable.html#cb272-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">relAbundanceCounts</span>(se)</span>
<span id="cb272-2"><a href="multitable.html#cb272-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code></pre></div>
<pre><code>## List of length 2
## names(2): counts relabundance</code></pre>
<p>Here the dimension of the count data remains unchanged. This is
actually a requirement for any <code>SummarizedExperiment</code> object.</p>
</div>
<div id="alternative-experiments" class="section level2" number="12.2">
<h2><span class="header-section-number">12.2</span> Alternative experiments</h2>
<p><em><strong>Alternative experiments</strong></em> differ from transformations as they can
contain complementary data, which is no longer tied to the same
dimensions as the assay data. However, the number of samples (columns)
must be the same.</p>
<p>This can come into play for instance when one has taxonomic abundance
profiles quantified with different measurement technologies, such as
phylogenetic microarrays, amplicon sequencing, or metagenomic
sequencing. Such alternative experiments that concern the same samples
can be stored as</p>
<ol style="list-style-type: decimal">
<li>Separate <em>assays</em> assuming that the taxonomic information can be mapped
between feature directly 1:1; or</li>
<li>data in the <em>altExp</em> slot of the <code>TreeSummarizedExperiment</code>, if the feature
dimensions differ. Each element of the <em>altExp</em> slot is a <code>SummarizedExperiment</code>
or an object from a derived class with independent feature data.</li>
</ol>
<p>As an example, we show how to store taxonomic abundance tables
agglomerated at different taxonomic levels. However, the data could as
well originate from entirely different measurement sources as long as
the samples are matched.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="multitable.html#cb274-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate the data to Phylym level</span></span>
<span id="cb274-2"><a href="multitable.html#cb274-2" aria-hidden="true" tabindex="-1"></a>se.phylum <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(se, <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb274-3"><a href="multitable.html#cb274-3" aria-hidden="true" tabindex="-1"></a><span class="co"># both have the same number of columns (samples)</span></span>
<span id="cb274-4"><a href="multitable.html#cb274-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se)</span></code></pre></div>
<pre><code>## [1] 19216    26</code></pre>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="multitable.html#cb276-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(se.phylum)</span></code></pre></div>
<pre><code>## [1] 67 26</code></pre>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="multitable.html#cb278-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the new table as an alternative experiment</span></span>
<span id="cb278-2"><a href="multitable.html#cb278-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(se, <span class="st">&quot;Phylum&quot;</span>) <span class="ot">&lt;-</span> se.phylum</span>
<span id="cb278-3"><a href="multitable.html#cb278-3" aria-hidden="true" tabindex="-1"></a><span class="fu">altExpNames</span>(se)</span></code></pre></div>
<pre><code>## [1] &quot;Phylum&quot;</code></pre>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="multitable.html#cb280-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Pick a sample subset: this acts on both altExp and assay data</span></span>
<span id="cb280-2"><a href="multitable.html#cb280-2" aria-hidden="true" tabindex="-1"></a>se[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 10 
## metadata(0):
## assays(2): counts relabundance
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(10): CL3 CC1 ... M31Tong M11Tong
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(1): Phylum
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="multitable.html#cb282-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">altExp</span>(se[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>],<span class="st">&quot;Phylum&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 67 10</code></pre>
<p>For more details of altExp have a look at the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">Intro vignette</a> of the
<code>SingleCellExperiment</code> package <span class="citation">(<a href="#ref-R-SingleCellExperiment" role="doc-biblioref">Lun and Risso 2020</a>)</span>.</p>
</div>
<div id="multiassayexperiments" class="section level2" number="12.3">
<h2><span class="header-section-number">12.3</span> MultiAssayExperiments</h2>
<p><em><strong>Multiple experiments</strong></em> relate to complementary measurement types,
such as transcriptomic or metabolomic profiling of the microbiome or
the host. Multiple experiments can be represented using the same
options as alternative experiments, or by using the
<code>MultiAssayExperiment</code> class <span class="citation">(<a href="#ref-R-MultiAssayExperiment" role="doc-biblioref">Ramos and Waldron 2020</a>)</span>. Depending on how the
datasets relate to each other the data can be stored as:</p>
<ol style="list-style-type: decimal">
<li>Separate <em>altExp</em> if the samples can be matched directly 1:1; or</li>
<li>As <code>MultiAssayExperiment</code> objects, in which the connections between
samples are defined through a <code>sampleMap</code>. Each element on the
<code>experimentsList</code> of an <code>MultiAssayExperiment</code> is <code>matrix</code> or
<code>matrix</code>-like object including <code>SummarizedExperiment</code> objects, and the
number of samples can differ between the elements.</li>
</ol>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="multitable.html#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="co">#TODO: Find the right dataset to explain a non 1:1 sample relationship</span></span></code></pre></div>
<p>For information have a look at the <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html">intro vignette</a> of the <code>MultiAssayExperiment</code> package.</p>
<table>
<caption><span id="tab:options">Table 12.1: </span> <strong>Recommended options for storing multiple data tables in microbiome studies</strong> The <em>assays</em> are best suited for data transformations (one-to-one match between samples and columns across the assays). The <em>alternative experiments</em> are particularly suitable for alternative versions of the data that are of same type but may have a different number of features (e.g. taxonomic groups); this is for instance the case with taxonomic abundance tables agglomerated at different levels (e.g. genus vs. phyla) or alternative profiling technologies (e.g. amplicon sequencing vs. shallow shotgun metagenomics). For alternative experiments one-to-one match between samples (cols) is required but the alternative experiment tables can have different numbers of features (rows). Finally, elements of the <em>MultiAssayExperiment</em> provide the most flexible way to incorporate multi-omic data tables with flexible numbers of samples and features. We recommend these conventions as the basis for methods development and application in microbiome studies.</caption>
<thead>
<tr class="header">
<th align="right">Option</th>
<th>Rows (features)</th>
<th align="right">Cols (samples)</th>
<th align="right">Recommended</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">assays</td>
<td>match</td>
<td align="right">match</td>
<td align="right">Data transformations</td>
</tr>
<tr class="even">
<td align="right">altExp</td>
<td>free</td>
<td align="right">match</td>
<td align="right">Alternative experiments</td>
</tr>
<tr class="odd">
<td align="right">MultiAssay</td>
<td>free</td>
<td align="right">free (mapping)</td>
<td align="right">Multi-omic experiments</td>
</tr>
</tbody>
</table>
</div>
<div id="session-info-8" class="section level2 unnumbered">
<h2>Session Info</h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] mia_1.1.14                     TreeSummarizedExperiment_2.1.4
 [3] Biostrings_2.61.2              XVector_0.33.0                
 [5] SingleCellExperiment_1.15.2    SummarizedExperiment_1.23.4   
 [7] Biobase_2.53.0                 GenomicRanges_1.45.0          
 [9] GenomeInfoDb_1.29.8            IRanges_2.27.2                
[11] S4Vectors_0.31.3               BiocGenerics_0.39.2           
[13] MatrixGenerics_1.5.4           matrixStats_0.60.1-9001       
[15] BiocStyle_2.21.3               rebook_1.3.1                  

loaded via a namespace (and not attached):
 [1] ggbeeswarm_0.6.0            colorspace_2.0-2           
 [3] ellipsis_0.3.2              scuttle_1.3.1              
 [5] BiocNeighbors_1.11.0        ggrepel_0.9.1              
 [7] bit64_4.0.5                 fansi_0.5.0                
 [9] decontam_1.13.0             splines_4.1.1              
[11] codetools_0.2-18            sparseMatrixStats_1.5.3    
[13] cachem_1.0.6                knitr_1.33                 
[15] scater_1.21.3               jsonlite_1.7.2             
[17] cluster_2.1.2               graph_1.71.2               
[19] BiocManager_1.30.16         compiler_4.1.1             
[21] assertthat_0.2.1            Matrix_1.3-4               
[23] fastmap_1.1.0               lazyeval_0.2.2             
[25] BiocSingular_1.9.1          htmltools_0.5.2            
[27] tools_4.1.1                 rsvd_1.0.5                 
[29] gtable_0.3.0                glue_1.4.2                 
[31] GenomeInfoDbData_1.2.6      reshape2_1.4.4             
[33] dplyr_1.0.7                 Rcpp_1.0.7                 
[35] jquerylib_0.1.4             vctrs_0.3.8                
[37] ape_5.5                     nlme_3.1-153               
[39] DECIPHER_2.21.0             DelayedMatrixStats_1.15.4  
[41] xfun_0.25                   stringr_1.4.0              
[43] beachmat_2.9.1              lifecycle_1.0.0            
[45] irlba_2.3.3                 XML_3.99-0.7               
[47] zlibbioc_1.39.0             MASS_7.3-54                
[49] scales_1.1.1                parallel_4.1.1             
[51] yaml_2.2.1                  memoise_2.0.0              
[53] gridExtra_2.3               ggplot2_3.3.5              
[55] sass_0.4.0                  stringi_1.7.4              
[57] RSQLite_2.2.8               ScaledMatrix_1.1.0         
[59] tidytree_0.3.5              permute_0.9-5              
[61] filelock_1.0.2              BiocParallel_1.27.6        
[63] rlang_0.4.11                pkgconfig_2.0.3            
[65] bitops_1.0-7                evaluate_0.14              
[67] lattice_0.20-44             purrr_0.3.4                
[69] treeio_1.17.2               CodeDepends_0.6.5          
[71] bit_4.0.4                   tidyselect_1.1.1           
[73] plyr_1.8.6                  magrittr_2.0.1             
[75] bookdown_0.24               R6_2.5.1                   
[77] generics_0.1.0              DelayedArray_0.19.2        
[79] DBI_1.1.1                   mgcv_1.8-36                
[81] pillar_1.6.2                RCurl_1.98-1.4             
[83] tibble_3.1.4                dir.expiry_1.1.0           
[85] crayon_1.4.1                utf8_1.2.2                 
[87] rmarkdown_2.10              viridis_0.6.1              
[89] grid_4.1.1                  blob_1.2.2                 
[91] vegan_2.5-7                 digest_0.6.27              
[93] tidyr_1.1.3                 munsell_0.5.0              
[95] DirichletMultinomial_1.35.0 beeswarm_0.4.0             
[97] viridisLite_0.4.0           vipor_0.4.5                
[99] bslib_0.3.0                </code></pre>
</div>

</div>
</div>
<h3>Bibliography</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R-SingleCellExperiment" class="csl-entry">
Lun, Aaron, and Davide Risso. 2020. <em>SingleCellExperiment: S4 Classes for Single Cell Data</em>.
</div>
<div id="ref-R-MultiAssayExperiment" class="csl-entry">
Ramos, Marcel, and Levi Waldron. 2020. <em>MultiAssayExperiment: Software for the Integration of Multi-Omics Experiments in Bioconductor</em>. <a href="http://waldronlab.io/MultiAssayExperiment/">http://waldronlab.io/MultiAssayExperiment/</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differential-abundance.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="function-ref.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
