<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="machine_learning.html"/>
<link rel="next" href="viz-chapter.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.1/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.11.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>2</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>2.1</b> Data science framework</a></li>
<li class="chapter" data-level="2.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>2.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="containers.html"><a href="containers.html#assay-data"><i class="fa fa-check"></i><b>2.2.1</b> Assay data</a></li>
<li class="chapter" data-level="2.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>2.2.2</b> colData</a></li>
<li class="chapter" data-level="2.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>2.2.3</b> rowData</a></li>
<li class="chapter" data-level="2.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>2.2.4</b> rowTree</a></li>
<li class="chapter" data-level="2.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>2.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="2.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>2.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>2.3</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>2.3.1</b> 16S workflow</a></li>
<li class="chapter" data-level="2.3.2" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>2.3.2</b> Import from external files</a></li>
<li class="chapter" data-level="2.3.3" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>2.3.3</b> Conversions between data formats in R</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>2.4</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>2.4.1</b> Package data</a></li>
<li class="chapter" data-level="2.4.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>2.4.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="2.4.3" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>2.4.3</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="containers.html"><a href="containers.html#session-info"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>3</b> Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>3.1</b> Package installation</a></li>
<li class="chapter" data-level="3.2" data-path="packages.html"><a href="packages.html#some-available-packages"><i class="fa fa-check"></i><b>3.2</b> Some available packages</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="packages.html"><a href="packages.html#mia-family-of-methods"><i class="fa fa-check"></i><b>3.2.1</b> mia family of methods</a></li>
<li class="chapter" data-level="3.2.2" data-path="packages.html"><a href="packages.html#tree-based-methods-sub-tree-methods"><i class="fa fa-check"></i><b>3.2.2</b> Tree-based methods {sub-tree-methods}</a></li>
<li class="chapter" data-level="3.2.3" data-path="packages.html"><a href="packages.html#differential-abundance-sub-diff-abund"><i class="fa fa-check"></i><b>3.2.3</b> Differential abundance {sub-diff-abund}</a></li>
<li class="chapter" data-level="3.2.4" data-path="packages.html"><a href="packages.html#manipulation-sub-manipulation"><i class="fa fa-check"></i><b>3.2.4</b> Manipulation {sub-manipulation}</a></li>
<li class="chapter" data-level="3.2.5" data-path="packages.html"><a href="packages.html#data"><i class="fa fa-check"></i><b>3.2.5</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="packages.html"><a href="packages.html#session-info-1"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.2</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.2.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="quality-control.html"><a href="quality-control.html#session-info-2"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#generate-a-taxonomic-tree-on-the-fly"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-transformation"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-information.html"><a href="taxonomic-information.html#session-info-3"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
<li class="chapter" data-level="" data-path="community-diversity.html"><a href="community-diversity.html#session-info-4"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1</b> Explained variance</a></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#community-comparisons-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.2</b> Community comparisons by beta diversity analysis</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#other-ordination-methods"><i class="fa fa-check"></i><b>8.3</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#pcoa-genus"><i class="fa fa-check"></i><b>8.4</b> Visualizing the most dominant genus on PCoA</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="community-similarity.html"><a href="community-similarity.html#testing-differences-in-community-composition-between-sample-groups"><i class="fa fa-check"></i><b>8.4.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.4.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.4.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="community-similarity.html"><a href="community-similarity.html#further-reading"><i class="fa fa-check"></i><b>8.5</b> Further reading</a></li>
<li class="chapter" data-level="" data-path="community-similarity.html"><a href="community-similarity.html#session-info-5"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="microbiome-community.html"><a href="microbiome-community.html#session-info-6"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community typing (clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#hiearchical-clustering"><i class="fa fa-check"></i><b>10.1</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>10.2</b> K-means clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.3.1</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#community-detection"><i class="fa fa-check"></i><b>10.4</b> Community Detection</a></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.5</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.5.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.5.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.5.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.5.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="10.5.3" data-path="clustering.html"><a href="clustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>10.5.3</b> Taxa vs taxa</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.6</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>11.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.1.6</b> LinDA</a></li>
<li class="chapter" data-level="11.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.1.7</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.1.8" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.1.8</b> Confounding variables</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.2</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances"><i class="fa fa-check"></i><b>11.2.1</b> Group-wise associations testing based on balances</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="differential-abundance.html"><a href="differential-abundance.html#session-info-7"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>12</b> Machine learning</a>
<ul>
<li class="chapter" data-level="12.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>12.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="12.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>12.2</b> Unsupervised machine learning</a></li>
<li class="chapter" data-level="" data-path="machine_learning.html"><a href="machine_learning.html#session-info-8"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#multi-omics-factor-analysis"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
<li class="chapter" data-level="" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#session-info-9"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>14</b> Visualization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>14.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>14.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="14.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>14.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>14.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>14.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="14.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>14.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>14.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>14.3.1</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="viz-chapter.html"><a href="viz-chapter.html#session-info-10"><i class="fa fa-check"></i>Session Info</a></li>
</ul></li>
<li class="part"><span><b>III Appendix</b></span></li>
<li class="chapter" data-level="15" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>15</b> Training</a>
<ul>
<li class="chapter" data-level="15.1" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>15.1</b> Code of Conduct</a></li>
<li class="chapter" data-level="15.2" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>15.2</b> Checklist</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="training.html"><a href="training.html#setting-up-your-own-computer"><i class="fa fa-check"></i><b>15.2.1</b> Setting up Your own computer</a></li>
<li class="chapter" data-level="15.2.2" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>15.2.2</b> Support and resources</a></li>
<li class="chapter" data-level="15.2.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>15.2.3</b> Study material</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="packages.html"><a href="packages.html#packages"><i class="fa fa-check"></i><b>15.3</b> Installing and loading the required R packages</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>16</b> Resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>16.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="resources.html"><a href="resources.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>16.1.1</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="16.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>16.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="16.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>16.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="16.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>16.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>16.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>16.2.1</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="resources.html"><a href="resources.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>16.3</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>17</b> Extra material</a>
<ul>
<li class="chapter" data-level="17.1" data-path="extras.html"><a href="extras.html#permanova-comparison"><i class="fa fa-check"></i><b>17.1</b> PERMANOVA comparison</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="extra_bayes.html"><a href="extra_bayes.html"><i class="fa fa-check"></i><b>18</b> Extra material</a>
<ul>
<li class="chapter" data-level="18.1" data-path="extra_bayes.html"><a href="extra_bayes.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>18.1</b> Bayesian Multinomial Logistic-Normal Models</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="interactive.html"><a href="interactive.html"><i class="fa fa-check"></i><b>19</b> Interactive plots</a>
<ul>
<li class="chapter" data-level="19.1" data-path="interactive.html"><a href="interactive.html#interactive-3d-plots"><i class="fa fa-check"></i><b>19.1</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i><b>20</b> Acknowledgments</a></li>
<li class="chapter" data-level="21" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>21</b> Exercises</a>
<ul>
<li class="chapter" data-level="21.1" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>21.1</b> Workflows</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting"><i class="fa fa-check"></i><b>21.1.1</b> Reproducible reporting</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-basic-components"><i class="fa fa-check"></i><b>21.2</b> TreeSummarizedExperiment: basic components</a>
<ul>
<li class="chapter" data-level="21.2.1" data-path="exercises.html"><a href="exercises.html#basic-summaries"><i class="fa fa-check"></i><b>21.2.1</b> Basic summaries</a></li>
<li class="chapter" data-level="21.2.2" data-path="exercises.html"><a href="exercises.html#taxonomic-abundance-table-assay"><i class="fa fa-check"></i><b>21.2.2</b> Taxonomic abundance table (assay)</a></li>
<li class="chapter" data-level="21.2.3" data-path="exercises.html"><a href="exercises.html#sample-side-information"><i class="fa fa-check"></i><b>21.2.3</b> Sample side information</a></li>
<li class="chapter" data-level="21.2.4" data-path="exercises.html"><a href="exercises.html#feature-side-information"><i class="fa fa-check"></i><b>21.2.4</b> Feature side information</a></li>
<li class="chapter" data-level="21.2.5" data-path="exercises.html"><a href="exercises.html#other-components"><i class="fa fa-check"></i><b>21.2.5</b> Other components</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="exercises.html"><a href="exercises.html#treesummarizedexperiment-data-import"><i class="fa fa-check"></i><b>21.3</b> TreeSummarizedExperiment: data import</a></li>
<li class="chapter" data-level="21.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>21.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="21.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>21.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="21.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>21.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="21.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>21.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="21.4.4" data-path="exercises.html"><a href="exercises.html#explore-the-data"><i class="fa fa-check"></i><b>21.4.4</b> Explore the data</a></li>
<li class="chapter" data-level="21.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>21.4.5</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="21.5" data-path="exercises.html"><a href="exercises.html#transformations"><i class="fa fa-check"></i><b>21.5</b> Transformations</a>
<ul>
<li class="chapter" data-level="21.5.1" data-path="exercises.html"><a href="exercises.html#transformations-1"><i class="fa fa-check"></i><b>21.5.1</b> Transformations</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>21.6</b> Taxonomic levels</a>
<ul>
<li class="chapter" data-level="21.6.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels-1"><i class="fa fa-check"></i><b>21.6.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="21.6.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments-altexp"><i class="fa fa-check"></i><b>21.6.2</b> Alternative experiments (altExp)</a></li>
</ul></li>
<li class="chapter" data-level="21.7" data-path="exercises.html"><a href="exercises.html#alpha-diversity"><i class="fa fa-check"></i><b>21.7</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="21.7.1" data-path="exercises.html"><a href="exercises.html#alpha-diversity-basics"><i class="fa fa-check"></i><b>21.7.1</b> Alpha diversity basics</a></li>
<li class="chapter" data-level="21.7.2" data-path="exercises.html"><a href="exercises.html#alpha-diversity-extra"><i class="fa fa-check"></i><b>21.7.2</b> Alpha diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="21.8" data-path="exercises.html"><a href="exercises.html#community-composition"><i class="fa fa-check"></i><b>21.8</b> Community composition</a>
<ul>
<li class="chapter" data-level="21.8.1" data-path="exercises.html"><a href="exercises.html#beta-diversity-basics"><i class="fa fa-check"></i><b>21.8.1</b> Beta diversity basics</a></li>
<li class="chapter" data-level="21.8.2" data-path="exercises.html"><a href="exercises.html#beta-diversity-extra"><i class="fa fa-check"></i><b>21.8.2</b> Beta diversity extra</a></li>
</ul></li>
<li class="chapter" data-level="21.9" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>21.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="21.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>21.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="21.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>21.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="21.10" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>21.10</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="21.10.1" data-path="exercises.html"><a href="exercises.html#univariate-analyses"><i class="fa fa-check"></i><b>21.10.1</b> Univariate analyses</a></li>
<li class="chapter" data-level="21.10.2" data-path="exercises.html"><a href="exercises.html#differential-abundance-analysis-1"><i class="fa fa-check"></i><b>21.10.2</b> Differential abundance analysis</a></li>
</ul></li>
<li class="chapter" data-level="21.11" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>21.11</b> Multiomics</a>
<ul>
<li class="chapter" data-level="21.11.1" data-path="exercises.html"><a href="exercises.html#introduction-to-multiassayexperiment-mae"><i class="fa fa-check"></i><b>21.11.1</b> Introduction to MultiAssayExperiment (MAE)</a></li>
<li class="chapter" data-level="21.11.2" data-path="exercises.html"><a href="exercises.html#introduction-to-multiomics"><i class="fa fa-check"></i><b>21.11.2</b> Introduction to multiomics</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multi-assay-analyses" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">Chapter 13</span> Multi-assay analyses<a href="multi-assay-analyses.html#multi-assay-analyses" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb353-1"><a href="multi-assay-analyses.html#cb353-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span></code></pre></div>
<p>Multi-omics means that we integrate data from multiple sources. For
example, we can integrate microbial abundances in the gut with
biomolecular profiling data from blood samples. This kind of
integrative multi-omic approaches can support the analysis of
microbiome dysbiosis and facilitate the discovery of novel biomarkers
for health and disease.</p>
<p>With cross-correlation analysis, we can analyze how strongly and how
differently variables are associated between each other. For instance,
we can analyze if higher presence of a specific taxon equals to higher
levels of a biomolecule.</p>
<p>The data containers that the <em>miaverse</em> utilizes are scalable and they
can contain different types of data in a same container. Because of
that, the <em>miaverse</em> is well-suitable for multi-assay microbiome data
which incorporates different types of complementary data sources in a
single reproducible workflow.</p>
<p>Another experiment can be stored in altExp slot of SE data container
or both experiments can be stored side-by-side in MAE data container
(see the sections <a href="containers.html#alt-exp">2.2.5</a> and <a href="containers.html#mae">2.2.6</a> to learn more about
altExp and MAE objects, respectively).</p>
<p>Different experiments are first imported into SE or TreeSE data
container similarly to the case when only one experiment is
present. After that different experiments are combined into the same
data container. Result is one TreeSE object with alternative
experiment in altExp slot, or MAE object with multiple experiment in
its experiment slot.</p>
<p>As an example data, we use data from following publication: Hintikka L
<em>et al.</em> (2021) Xylo-oligosaccharides in prevention of hepatic
steatosis and adipose tissue inflammation: associating taxonomic and
metabolomic patterns in fecal microbiotas with biclustering
<span class="citation">(<a href="#ref-Hintikka2021" role="doc-biblioref">Hintikka et al. 2021</a>)</span>.</p>
<p>In this article, mice were fed with high-fat and low-fat diets with or
without prebiotics. The purpose of this was to study if prebiotics
would reduce the negative impacts of high-fat diet.</p>
<p>This example data can be loaded from microbiomeDataSets. The data is
already in MAE format. It includes three different experiments:
microbial abundance data, metabolite concentrations, and data about
different biomarkers. Help for importing data into SE object you can
find from
<a href="https://microbiome.github.io/OMA/containers.html#loading-experimental-microbiome-data">here</a>.</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="multi-assay-analyses.html#cb354-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb354-2"><a href="multi-assay-analyses.html#cb354-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(HintikkaXOData, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb354-3"><a href="multi-assay-analyses.html#cb354-3" aria-hidden="true" tabindex="-1"></a>mae <span class="ot">&lt;-</span> HintikkaXOData</span>
<span id="cb354-4"><a href="multi-assay-analyses.html#cb354-4" aria-hidden="true" tabindex="-1"></a>mae</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] microbiota: TreeSummarizedExperiment with 12706 rows and 40 columns
##  [2] metabolites: TreeSummarizedExperiment with 38 rows and 40 columns
##  [3] biomarkers: TreeSummarizedExperiment with 39 rows and 40 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files</code></pre>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb356-1"><a href="multi-assay-analyses.html#cb356-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(stringr)){</span>
<span id="cb356-2"><a href="multi-assay-analyses.html#cb356-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;stringr&quot;</span>)</span>
<span id="cb356-3"><a href="multi-assay-analyses.html#cb356-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(stringr)</span>
<span id="cb356-4"><a href="multi-assay-analyses.html#cb356-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb356-5"><a href="multi-assay-analyses.html#cb356-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop off those bacteria that do not include information in Phylum or lower levels</span></span>
<span id="cb356-6"><a href="multi-assay-analyses.html#cb356-6" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>Phylum), ]</span>
<span id="cb356-7"><a href="multi-assay-analyses.html#cb356-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Clean taxonomy data, so that names do not include additional characters</span></span>
<span id="cb356-8"><a href="multi-assay-analyses.html#cb356-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="fu">apply</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]]), <span class="dv">2</span>, </span>
<span id="cb356-9"><a href="multi-assay-analyses.html#cb356-9" aria-hidden="true" tabindex="-1"></a>                                     str_remove, <span class="at">pattern =</span> <span class="st">&quot;._[0-9]__&quot;</span>))</span>
<span id="cb356-10"><a href="multi-assay-analyses.html#cb356-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Microbiome data</span></span>
<span id="cb356-11"><a href="multi-assay-analyses.html#cb356-11" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 12613 40 
## metadata(0):
## assays(1): counts
## rownames(12613): GAYR01026362.62.2014 CVJT01000011.50.2173 ...
##   JRJTB:03787:02429 JRJTB:03787:02478
## rowData names(7): Phylum Class ... Species OTU
## colnames(40): C1 C2 ... C39 C40
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb358-1"><a href="multi-assay-analyses.html#cb358-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Metabolite data</span></span>
<span id="cb358-2"><a href="multi-assay-analyses.html#cb358-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">2</span>]]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 38 40 
## metadata(0):
## assays(1): nmr
## rownames(38): Butyrate Acetate ... Malonate 1,3-dihydroxyacetone
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="multi-assay-analyses.html#cb360-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Biomarker data</span></span>
<span id="cb360-2"><a href="multi-assay-analyses.html#cb360-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">3</span>]]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 39 40 
## metadata(0):
## assays(1): signals
## rownames(39): Triglycerides_liver CLSs_epi ... NPY_serum Glycogen_liver
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div id="cross-correlation-analysis" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Cross-correlation Analysis<a href="multi-assay-analyses.html#cross-correlation-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Next we can do the cross-correlation analysis. Here we analyse if
individual bacteria genera correlates with concentrations of
individual metabolites. This helps as to answer the question: If this
bacteria is present, is this metabolites concentration then low or
high?</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="multi-assay-analyses.html#cb362-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate microbiome data at family level</span></span>
<span id="cb362-2"><a href="multi-assay-analyses.html#cb362-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">agglomerateByPrevalence</span>(mae[[<span class="dv">1</span>]], <span class="at">rank =</span> <span class="st">&quot;Family&quot;</span>)</span>
<span id="cb362-3"><a href="multi-assay-analyses.html#cb362-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Does log10 transform for microbiome data</span></span>
<span id="cb362-4"><a href="multi-assay-analyses.html#cb362-4" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb362-5"><a href="multi-assay-analyses.html#cb362-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb362-6"><a href="multi-assay-analyses.html#cb362-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Give unique names so that we do not have problems when we are creating a plot</span></span>
<span id="cb362-7"><a href="multi-assay-analyses.html#cb362-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">getTaxonomyLabels</span>(mae[[<span class="dv">1</span>]])</span>
<span id="cb362-8"><a href="multi-assay-analyses.html#cb362-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb362-9"><a href="multi-assay-analyses.html#cb362-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross correlates data sets</span></span>
<span id="cb362-10"><a href="multi-assay-analyses.html#cb362-10" aria-hidden="true" tabindex="-1"></a>correlations <span class="ot">&lt;-</span> <span class="fu">testExperimentCrossCorrelation</span>(mae, </span>
<span id="cb362-11"><a href="multi-assay-analyses.html#cb362-11" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">experiment1 =</span> <span class="dv">1</span>,</span>
<span id="cb362-12"><a href="multi-assay-analyses.html#cb362-12" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">experiment2 =</span> <span class="dv">2</span>,</span>
<span id="cb362-13"><a href="multi-assay-analyses.html#cb362-13" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">assay_name1 =</span> <span class="st">&quot;log10&quot;</span>, </span>
<span id="cb362-14"><a href="multi-assay-analyses.html#cb362-14" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">assay_name2 =</span> <span class="st">&quot;nmr&quot;</span>,</span>
<span id="cb362-15"><a href="multi-assay-analyses.html#cb362-15" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">method =</span> <span class="st">&quot;spearman&quot;</span>, </span>
<span id="cb362-16"><a href="multi-assay-analyses.html#cb362-16" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">p_adj_threshold =</span> <span class="cn">NULL</span>,</span>
<span id="cb362-17"><a href="multi-assay-analyses.html#cb362-17" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">cor_threshold =</span> <span class="cn">NULL</span>,</span>
<span id="cb362-18"><a href="multi-assay-analyses.html#cb362-18" aria-hidden="true" tabindex="-1"></a>                                               <span class="co"># Remove when mia is fixed</span></span>
<span id="cb362-19"><a href="multi-assay-analyses.html#cb362-19" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>,</span>
<span id="cb362-20"><a href="multi-assay-analyses.html#cb362-20" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">sort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb362-21"><a href="multi-assay-analyses.html#cb362-21" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">show_warnings =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Creates the heatmap</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="multi-assay-analyses.html#cb363-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( <span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;ComplexHeatmap&quot;</span>) ){</span>
<span id="cb363-2"><a href="multi-assay-analyses.html#cb363-2" aria-hidden="true" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;ComplexHeatmap&quot;</span>)</span>
<span id="cb363-3"><a href="multi-assay-analyses.html#cb363-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(<span class="st">&quot;ComplexHeatmap&quot;</span>)</span>
<span id="cb363-4"><a href="multi-assay-analyses.html#cb363-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb363-5"><a href="multi-assay-analyses.html#cb363-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb363-6"><a href="multi-assay-analyses.html#cb363-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap and store it</span></span>
<span id="cb363-7"><a href="multi-assay-analyses.html#cb363-7" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(correlations<span class="sc">$</span>cor,</span>
<span id="cb363-8"><a href="multi-assay-analyses.html#cb363-8" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Print values to cells</span></span>
<span id="cb363-9"><a href="multi-assay-analyses.html#cb363-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">cell_fun =</span> <span class="cf">function</span>(j, i, x, y, width, height, fill) {</span>
<span id="cb363-10"><a href="multi-assay-analyses.html#cb363-10" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># If the p-value is under threshold</span></span>
<span id="cb363-11"><a href="multi-assay-analyses.html#cb363-11" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span>( <span class="sc">!</span><span class="fu">is.na</span>(correlations<span class="sc">$</span>p_adj[i, j]) <span class="sc">&amp;</span> correlations<span class="sc">$</span>p_adj[i, j] <span class="sc">&lt;</span> <span class="fl">0.05</span> ){</span>
<span id="cb363-12"><a href="multi-assay-analyses.html#cb363-12" aria-hidden="true" tabindex="-1"></a>                        <span class="co"># Print &quot;X&quot;</span></span>
<span id="cb363-13"><a href="multi-assay-analyses.html#cb363-13" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">grid.text</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%s&quot;</span>, <span class="st">&quot;X&quot;</span>), x, y, <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="st">&quot;black&quot;</span>))</span>
<span id="cb363-14"><a href="multi-assay-analyses.html#cb363-14" aria-hidden="true" tabindex="-1"></a>                        }</span>
<span id="cb363-15"><a href="multi-assay-analyses.html#cb363-15" aria-hidden="true" tabindex="-1"></a>                    },</span>
<span id="cb363-16"><a href="multi-assay-analyses.html#cb363-16" aria-hidden="true" tabindex="-1"></a>                <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="st">&quot;&quot;</span>, <span class="at">legend_height =</span> <span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">&quot;cm&quot;</span>))</span>
<span id="cb363-17"><a href="multi-assay-analyses.html#cb363-17" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb363-18"><a href="multi-assay-analyses.html#cb363-18" aria-hidden="true" tabindex="-1"></a>plot</span></code></pre></div>
<p><img src="23_multi-assay_analyses_files/figure-html/cross-correlation6-1.png" width="960" /></p>
</div>
<div id="multi-omics-factor-analysis" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Multi-Omics Factor Analysis<a href="multi-assay-analyses.html#multi-omics-factor-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Multi-Omics Factor Analysis <span class="citation">(<a href="#ref-Argelaguet2018" role="doc-biblioref">Argelaguet 2018</a>)</span> (MOFA) is an
unsupervised method for integrating multi-omic data sets in a
downstream analysis. It could be seen as a generalization of
principal component analysis. Yet, with the ability to infer a latent
(low-dimensional) representation, shared among the mutliple (-omics)
data sets in hand.</p>
<p>We use the R <a href="https://biofam.github.io/MOFA2/index.html">MOFA2</a>
package for the analysis, and
<a href="https://biofam.github.io/MOFA2/installation.html">install</a> the
corresponding dependencies.</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="multi-assay-analyses.html#cb364-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(MOFA2)){</span>
<span id="cb364-2"><a href="multi-assay-analyses.html#cb364-2" aria-hidden="true" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;MOFA2&quot;</span>)</span>
<span id="cb364-3"><a href="multi-assay-analyses.html#cb364-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb364-4"><a href="multi-assay-analyses.html#cb364-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb364-5"><a href="multi-assay-analyses.html#cb364-5" aria-hidden="true" tabindex="-1"></a><span class="co"># For inter-operability between Python and R, and setting Python dependencies,</span></span>
<span id="cb364-6"><a href="multi-assay-analyses.html#cb364-6" aria-hidden="true" tabindex="-1"></a><span class="co"># reticulate package is needed</span></span>
<span id="cb364-7"><a href="multi-assay-analyses.html#cb364-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(reticulate)){</span>
<span id="cb364-8"><a href="multi-assay-analyses.html#cb364-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;reticulate&quot;</span>)</span>
<span id="cb364-9"><a href="multi-assay-analyses.html#cb364-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb364-10"><a href="multi-assay-analyses.html#cb364-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb364-11"><a href="multi-assay-analyses.html#cb364-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Let us assume that these have been installed already.</span></span>
<span id="cb364-12"><a href="multi-assay-analyses.html#cb364-12" aria-hidden="true" tabindex="-1"></a><span class="co">#reticulate::install_miniconda(force = TRUE)</span></span>
<span id="cb364-13"><a href="multi-assay-analyses.html#cb364-13" aria-hidden="true" tabindex="-1"></a><span class="co">#reticulate::use_miniconda(condaenv = &quot;env1&quot;, required = FALSE)</span></span>
<span id="cb364-14"><a href="multi-assay-analyses.html#cb364-14" aria-hidden="true" tabindex="-1"></a><span class="co">#reticulate::py_install(packages = c(&quot;mofapy2&quot;), pip = TRUE, python_version=3.6)</span></span></code></pre></div>
<p>The <code>mae</code> object could be used straight to create the MOFA model. Yet,
we transform our assays since the model assumes normality per
default. Other distributions that can be used, include Poisson or
Bernoulli.</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="multi-assay-analyses.html#cb365-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MOFA2)</span>
<span id="cb365-2"><a href="multi-assay-analyses.html#cb365-2" aria-hidden="true" tabindex="-1"></a><span class="co"># For simplicity, classify all high-fat diets as high-fat, and all the low-fat </span></span>
<span id="cb365-3"><a href="multi-assay-analyses.html#cb365-3" aria-hidden="true" tabindex="-1"></a><span class="co"># diets as low-fat diets</span></span>
<span id="cb365-4"><a href="multi-assay-analyses.html#cb365-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(mae)<span class="sc">$</span>Diet <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">colData</span>(mae)<span class="sc">$</span>Diet <span class="sc">==</span> <span class="st">&quot;High-fat&quot;</span> <span class="sc">|</span> </span>
<span id="cb365-5"><a href="multi-assay-analyses.html#cb365-5" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">colData</span>(mae)<span class="sc">$</span>Diet <span class="sc">==</span> <span class="st">&quot;High-fat + XOS&quot;</span>, </span>
<span id="cb365-6"><a href="multi-assay-analyses.html#cb365-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">&quot;High-fat&quot;</span>, <span class="st">&quot;Low-fat&quot;</span>)</span>
<span id="cb365-7"><a href="multi-assay-analyses.html#cb365-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb365-8"><a href="multi-assay-analyses.html#cb365-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Removing duplicates at the microbiome data</span></span>
<span id="cb365-9"><a href="multi-assay-analyses.html#cb365-9" aria-hidden="true" tabindex="-1"></a><span class="co"># which are also in form e.g. &quot;Ambiguous&quot; and &quot;uncultured&quot; taxa</span></span>
<span id="cb365-10"><a href="multi-assay-analyses.html#cb365-10" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="sc">!</span><span class="fu">duplicated</span>(<span class="fu">rownames</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]]))), ]</span>
<span id="cb365-11"><a href="multi-assay-analyses.html#cb365-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb365-12"><a href="multi-assay-analyses.html#cb365-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Transforming microbiome data with rclr</span></span>
<span id="cb365-13"><a href="multi-assay-analyses.html#cb365-13" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb365-14"><a href="multi-assay-analyses.html#cb365-14" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">1</span>]], <span class="at">assay_name =</span> <span class="st">&quot;relabundance&quot;</span>, <span class="at">method =</span> <span class="st">&quot;rclr&quot;</span>)</span>
<span id="cb365-15"><a href="multi-assay-analyses.html#cb365-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb365-16"><a href="multi-assay-analyses.html#cb365-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Transforming metabolomic data with log10</span></span>
<span id="cb365-17"><a href="multi-assay-analyses.html#cb365-17" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">2</span>]], <span class="at">assay_name =</span> <span class="st">&quot;nmr&quot;</span>,</span>
<span id="cb365-18"><a href="multi-assay-analyses.html#cb365-18" aria-hidden="true" tabindex="-1"></a>                            <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>,</span>
<span id="cb365-19"><a href="multi-assay-analyses.html#cb365-19" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>)</span>
<span id="cb365-20"><a href="multi-assay-analyses.html#cb365-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb365-21"><a href="multi-assay-analyses.html#cb365-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Transforming biomarker data with z-transform</span></span>
<span id="cb365-22"><a href="multi-assay-analyses.html#cb365-22" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">3</span>]], <span class="at">assay_name =</span> <span class="st">&quot;signals&quot;</span>,</span>
<span id="cb365-23"><a href="multi-assay-analyses.html#cb365-23" aria-hidden="true" tabindex="-1"></a>                            <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>,</span>
<span id="cb365-24"><a href="multi-assay-analyses.html#cb365-24" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb365-25"><a href="multi-assay-analyses.html#cb365-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb365-26"><a href="multi-assay-analyses.html#cb365-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Removing assays no longer needed</span></span>
<span id="cb365-27"><a href="multi-assay-analyses.html#cb365-27" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;counts&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb365-28"><a href="multi-assay-analyses.html#cb365-28" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;log10&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb365-29"><a href="multi-assay-analyses.html#cb365-29" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">2</span>]], <span class="st">&quot;nmr&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb365-30"><a href="multi-assay-analyses.html#cb365-30" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">3</span>]], <span class="st">&quot;signals&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb365-31"><a href="multi-assay-analyses.html#cb365-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb365-32"><a href="multi-assay-analyses.html#cb365-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Building our mofa model</span></span>
<span id="cb365-33"><a href="multi-assay-analyses.html#cb365-33" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">create_mofa_from_MultiAssayExperiment</span>(mae,</span>
<span id="cb365-34"><a href="multi-assay-analyses.html#cb365-34" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">groups =</span> <span class="st">&quot;Diet&quot;</span>, </span>
<span id="cb365-35"><a href="multi-assay-analyses.html#cb365-35" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">extract_metadata =</span> <span class="cn">TRUE</span>)</span>
<span id="cb365-36"><a href="multi-assay-analyses.html#cb365-36" aria-hidden="true" tabindex="-1"></a>model</span></code></pre></div>
<pre><code>## Untrained MOFA model with the following characteristics: 
##  Number of views: 3 
##  Views names: microbiota metabolites biomarkers 
##  Number of features (per view): 38 38 39 
##  Number of groups: 2 
##  Groups names: High-fat Low-fat 
##  Number of samples (per group): 20 20 
## </code></pre>
<p>Model options could be defined as follows:</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="multi-assay-analyses.html#cb367-1" aria-hidden="true" tabindex="-1"></a>model_opts <span class="ot">&lt;-</span> <span class="fu">get_default_model_options</span>(model)</span>
<span id="cb367-2"><a href="multi-assay-analyses.html#cb367-2" aria-hidden="true" tabindex="-1"></a>model_opts<span class="sc">$</span>num_factors <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb367-3"><a href="multi-assay-analyses.html#cb367-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(model_opts)</span></code></pre></div>
<pre><code>## $likelihoods
##  microbiota metabolites  biomarkers 
##  &quot;gaussian&quot;  &quot;gaussian&quot;  &quot;gaussian&quot; 
## 
## $num_factors
## [1] 5
## 
## $spikeslab_factors
## [1] FALSE
## 
## $spikeslab_weights
## [1] TRUE
## 
## $ard_factors
## [1] TRUE
## 
## $ard_weights
## [1] TRUE</code></pre>
<p>Models training options are defined with the following:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="multi-assay-analyses.html#cb369-1" aria-hidden="true" tabindex="-1"></a>train_opts <span class="ot">&lt;-</span> <span class="fu">get_default_training_options</span>(model)</span>
<span id="cb369-2"><a href="multi-assay-analyses.html#cb369-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(train_opts)</span></code></pre></div>
<pre><code>## $maxiter
## [1] 1000
## 
## $convergence_mode
## [1] &quot;fast&quot;
## 
## $drop_factor_threshold
## [1] -1
## 
## $verbose
## [1] FALSE
## 
## $startELBO
## [1] 1
## 
## $freqELBO
## [1] 5</code></pre>
<p>Preparing and training the model:</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="multi-assay-analyses.html#cb371-1" aria-hidden="true" tabindex="-1"></a>model.prepared <span class="ot">&lt;-</span> <span class="fu">prepare_mofa</span>(</span>
<span id="cb371-2"><a href="multi-assay-analyses.html#cb371-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> model,</span>
<span id="cb371-3"><a href="multi-assay-analyses.html#cb371-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model_options =</span> model_opts</span>
<span id="cb371-4"><a href="multi-assay-analyses.html#cb371-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb371-5"><a href="multi-assay-analyses.html#cb371-5" aria-hidden="true" tabindex="-1"></a>model.trained <span class="ot">&lt;-</span> <span class="fu">run_mofa</span>(model.prepared)</span></code></pre></div>
<p>Visualizing the variance explained:</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="multi-assay-analyses.html#cb372-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb372-2"><a href="multi-assay-analyses.html#cb372-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb372-3"><a href="multi-assay-analyses.html#cb372-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb372-4"><a href="multi-assay-analyses.html#cb372-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(model.trained, <span class="at">x=</span><span class="st">&quot;view&quot;</span>, <span class="at">y=</span><span class="st">&quot;factor&quot;</span>, <span class="at">plot_total =</span> T),</span>
<span id="cb372-5"><a href="multi-assay-analyses.html#cb372-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">nrow =</span> <span class="dv">2</span></span>
<span id="cb372-6"><a href="multi-assay-analyses.html#cb372-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">&quot;Variance Explained per factor and assay&quot;</span>,</span>
<span id="cb372-7"><a href="multi-assay-analyses.html#cb372-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theme =</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)))</span></code></pre></div>
<p><img src="23_multi-assay_analyses_files/figure-html/unnamed-chunk-5-1.png" width="960" /></p>
<p>The top weights for each assay using all 5 factors:</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="multi-assay-analyses.html#cb373-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;microbiota&quot;</span>, <span class="st">&quot;metabolites&quot;</span>,<span class="st">&quot;biomarkers&quot;</span>), <span class="cf">function</span>(name) {</span>
<span id="cb373-2"><a href="multi-assay-analyses.html#cb373-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_top_weights</span>(model.trained,</span>
<span id="cb373-3"><a href="multi-assay-analyses.html#cb373-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">view =</span> name,</span>
<span id="cb373-4"><a href="multi-assay-analyses.html#cb373-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">factors =</span> <span class="st">&quot;all&quot;</span>,</span>
<span id="cb373-5"><a href="multi-assay-analyses.html#cb373-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">nfeatures =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb373-6"><a href="multi-assay-analyses.html#cb373-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;Top weights of the &quot;</span>, name,<span class="st">&quot; assay&quot;</span>))</span>
<span id="cb373-7"><a href="multi-assay-analyses.html#cb373-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb373-8"><a href="multi-assay-analyses.html#cb373-8" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">nrow =</span> <span class="dv">3</span>) <span class="sc">&amp;</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>))</span></code></pre></div>
<p><img src="23_multi-assay_analyses_files/figure-html/unnamed-chunk-6-1.png" width="960" /></p>
<p>More tutorials and examples of using the package are found at: <a href="https://biofam.github.io/MOFA2/tutorials.html">link</a></p>
</div>
<div id="session-info-9" class="section level2 unnumbered hasAnchor">
<h2>Session Info<a href="multi-assay-analyses.html#session-info-9" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ggplot2_3.4.1                  patchwork_1.1.2               
 [3] reticulate_1.28                MOFA2_1.6.0                   
 [5] ComplexHeatmap_2.12.1          stringr_1.5.0                 
 [7] mia_1.7.9                      MultiAssayExperiment_1.24.0   
 [9] TreeSummarizedExperiment_2.1.4 Biostrings_2.66.0             
[11] XVector_0.38.0                 SingleCellExperiment_1.20.0   
[13] SummarizedExperiment_1.28.0    Biobase_2.58.0                
[15] GenomicRanges_1.50.2           GenomeInfoDb_1.34.9           
[17] IRanges_2.32.0                 S4Vectors_0.36.1              
[19] BiocGenerics_0.44.0            MatrixGenerics_1.10.0         
[21] matrixStats_0.63.0-9003        BiocStyle_2.24.0              
[23] rebook_1.6.0                  

loaded via a namespace (and not attached):
  [1] circlize_0.4.15             corrplot_0.92              
  [3] BiocBaseUtils_1.0.0         plyr_1.8.8                 
  [5] lazyeval_0.2.2              splines_4.2.1              
  [7] BiocParallel_1.32.5         scater_1.26.1              
  [9] digest_0.6.31               foreach_1.5.2              
 [11] yulab.utils_0.0.6           htmltools_0.5.4            
 [13] viridis_0.6.2               fansi_1.0.4                
 [15] magrittr_2.0.3              memoise_2.0.1              
 [17] ScaledMatrix_1.6.0          cluster_2.1.4              
 [19] doParallel_1.0.17           DECIPHER_2.26.0            
 [21] colorspace_2.1-0            rappdirs_0.3.3             
 [23] blob_1.2.3                  ggrepel_0.9.3              
 [25] xfun_0.37                   dplyr_1.1.0                
 [27] crayon_1.5.2                RCurl_1.98-1.10            
 [29] jsonlite_1.8.4              graph_1.74.0               
 [31] iterators_1.0.14            ape_5.7                    
 [33] glue_1.6.2                  gtable_0.3.1               
 [35] zlibbioc_1.44.0             GetoptLong_1.0.5           
 [37] DelayedArray_0.24.0         BiocSingular_1.14.0        
 [39] Rhdf5lib_1.18.2             shape_1.4.6                
 [41] HDF5Array_1.24.2            scales_1.2.1               
 [43] pheatmap_1.0.12             DBI_1.1.3                  
 [45] Rcpp_1.0.10                 viridisLite_0.4.1          
 [47] decontam_1.18.0             clue_0.3-64                
 [49] tidytree_0.4.2              bit_4.0.5                  
 [51] rsvd_1.0.5                  dir.expiry_1.4.0           
 [53] RColorBrewer_1.1-3          farver_2.1.1               
 [55] pkgconfig_2.0.3             XML_3.99-0.13              
 [57] scuttle_1.8.4               uwot_0.1.14                
 [59] CodeDepends_0.6.5           sass_0.4.5                 
 [61] here_1.0.1                  utf8_1.2.3                 
 [63] labeling_0.4.2              tidyselect_1.2.0           
 [65] rlang_1.0.6                 reshape2_1.4.4             
 [67] munsell_0.5.0               tools_4.2.1                
 [69] cachem_1.0.7                cli_3.6.0                  
 [71] DirichletMultinomial_1.40.0 generics_0.1.3             
 [73] RSQLite_2.3.0               evaluate_0.20              
 [75] fastmap_1.1.1               yaml_2.3.7                 
 [77] knitr_1.42                  bit64_4.0.5                
 [79] purrr_1.0.1                 nlme_3.1-162               
 [81] sparseMatrixStats_1.10.0    compiler_4.2.1             
 [83] beeswarm_0.4.0              filelock_1.0.2             
 [85] png_0.1-8                   treeio_1.22.0              
 [87] tibble_3.2.0                bslib_0.4.2                
 [89] stringi_1.7.12              highr_0.10                 
 [91] basilisk.utils_1.8.0        forcats_1.0.0              
 [93] lattice_0.20-45             Matrix_1.5-3               
 [95] vegan_2.6-4                 permute_0.9-7              
 [97] vctrs_0.5.2                 rhdf5filters_1.8.0         
 [99] pillar_1.8.1                lifecycle_1.0.3            
[101] BiocManager_1.30.20         jquerylib_0.1.4            
[103] GlobalOptions_0.1.2         BiocNeighbors_1.16.0       
[105] cowplot_1.1.1               bitops_1.0-7               
[107] irlba_2.3.5.1               R6_2.5.1                   
[109] bookdown_0.33               gridExtra_2.3              
[111] vipor_0.4.5                 codetools_0.2-19           
[113] MASS_7.3-58.3               rhdf5_2.40.0               
[115] rprojroot_2.0.3             rjson_0.2.21               
[117] withr_2.5.0                 GenomeInfoDbData_1.2.9     
[119] mgcv_1.8-42                 parallel_4.2.1             
[121] beachmat_2.14.0             tidyr_1.3.0                
[123] basilisk_1.8.1              rmarkdown_2.20             
[125] DelayedMatrixStats_1.20.0   Rtsne_0.16                 
[127] Cairo_1.6-0                 ggbeeswarm_0.7.1           </code></pre>
</div>

</div>
</div>
<h3>Bibliography<a href="bibliography.html#bibliography" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Argelaguet2018" class="csl-entry">
Argelaguet, Ricard et al. 2018. <span>Multi-Omics Factor Analysisa Framework for Unsupervised Integration of Multi-Omics Data Sets.</span> <em>Molecular Systems Biology</em> 14 (6): e8124. <a href="https://doi.org/10.15252/msb.20178124">https://doi.org/10.15252/msb.20178124</a>.
</div>
<div id="ref-Hintikka2021" class="csl-entry">
Hintikka, Jukka, Sanna Lensu, Elina Mkinen, Sira Karvinen, Marjaana Honkanen, Jere Lindn, Tim Garrels, Satu Pekkala, and Leo Lahti. 2021. <span>Xylo-Oligosaccharides in Prevention of Hepatic Steatosis and Adipose Tissue Inflammation: Associating Taxonomic and Metabolomic Patterns in Fecal Microbiomes with Biclustering.</span> <em>International Journal of Environmental Research and Public Health</em> 18 (8): 4049.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="machine_learning.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="viz-chapter.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
