<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis</title>
  <meta name="description" content="Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Multi-assay analyses | Orchestrating Microbiome Analysis" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="machine_learning.html"/>
<link rel="next" href="viz-chapter.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-family-of-methods"><i class="fa fa-check"></i><b>2.2.1</b> mia family of methods</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-tree-methods"><i class="fa fa-check"></i><b>2.2.2</b> Tree-based methods</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.3</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#sub-manipulation"><i class="fa fa-check"></i><b>2.2.4</b> Manipulation</a></li>
<li class="chapter" data-level="2.2.5" data-path="packages.html"><a href="packages.html#further-options"><i class="fa fa-check"></i><b>2.2.5</b> Further options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.4</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-external-files"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
<li class="chapter" data-level="6.5" data-path="taxonomic-information.html"><a href="taxonomic-information.html#pick-specific"><i class="fa fa-check"></i><b>6.5</b> Pick specific</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>6.5.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="6.5.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>6.5.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervized-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervized ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervized-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervized ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#introduction-to-bluster-package"><i class="fa fa-check"></i><b>10.1</b> Introduction to bluster package</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="clustering.html"><a href="clustering.html#sample-clustering"><i class="fa fa-check"></i><b>10.1.1</b> Sample clustering</a></li>
<li class="chapter" data-level="10.1.2" data-path="clustering.html"><a href="clustering.html#taxa-clustering"><i class="fa fa-check"></i><b>10.1.2</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>10.3</b> K-means clustering</a></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.4</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#pcoa-for-asv-level-data-with-bray-curtis-with-dmm-clusters-shown-with-colors"><i class="fa fa-check"></i><b>10.4.1</b> PCoA for ASV-level data with Bray-Curtis; with DMM clusters shown with colors</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#community-detection"><i class="fa fa-check"></i><b>10.5</b> Community Detection</a></li>
<li class="chapter" data-level="10.6" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.6</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.6.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.6.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.6.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.6.2</b> Taxa vs biomolecules</a></li>
<li class="chapter" data-level="10.6.3" data-path="clustering.html"><a href="clustering.html#taxa-vs-taxa"><i class="fa fa-check"></i><b>10.6.3</b> Taxa vs taxa</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.7</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#differential-abundance-analysis"><i class="fa fa-check"></i><b>11.1</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="differential-abundance.html"><a href="differential-abundance.html#examples-and-tools"><i class="fa fa-check"></i><b>11.1.1</b> Examples and tools</a></li>
<li class="chapter" data-level="11.1.2" data-path="differential-abundance.html"><a href="differential-abundance.html#prevalence-filtering"><i class="fa fa-check"></i><b>11.1.2</b> Prevalence Filtering</a></li>
<li class="chapter" data-level="11.1.3" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.1.3</b> ALDEx2</a></li>
<li class="chapter" data-level="11.1.4" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.1.4</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.1.5" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.1.5</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.1.6" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.1.6</b> LinDA</a></li>
<li class="chapter" data-level="11.1.7" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.1.7</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.1.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.1.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#confounding-variables"><i class="fa fa-check"></i><b>11.2</b> Confounding variables</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#ancombc"><i class="fa fa-check"></i><b>11.2.1</b> ANCOMBC</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.2.2</b> LinDA</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.2.3</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#tree-based-methods"><i class="fa fa-check"></i><b>11.3</b> Tree-based methods</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#group-wise-associations-testing-based-on-balances"><i class="fa fa-check"></i><b>11.3.1</b> Group-wise associations testing based on balances</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>12</b> Machine learning</a>
<ul>
<li class="chapter" data-level="12.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>12.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="12.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>12.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-assay analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>14</b> Visualization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>14.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>14.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="14.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>14.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>14.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>14.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="14.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>14.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>14.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>14.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="15" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>15</b> Training</a>
<ul>
<li class="chapter" data-level="15.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>15.1</b> Checklist</a></li>
<li class="chapter" data-level="15.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>15.2</b> Recommended software</a></li>
<li class="chapter" data-level="15.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>15.3</b> Study material</a></li>
<li class="chapter" data-level="15.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>15.4</b> Support and resources</a></li>
<li class="chapter" data-level="15.5" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>15.5</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>16</b> Resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>16.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>16.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="16.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>16.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="16.1.3" data-path="resources.html"><a href="resources.html#alternative-containers-for-microbiome-data"><i class="fa fa-check"></i><b>16.1.3</b> Alternative containers for microbiome data</a></li>
<li class="chapter" data-level="16.1.4" data-path="resources.html"><a href="resources.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>16.1.4</b> Resources for phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>16.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>16.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="16.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>16.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>16.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>16.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="16.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>16.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>17</b> Exercises</a>
<ul>
<li class="chapter" data-level="17.1" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>17.1</b> Workflows</a>
<ul>
<li class="chapter" data-level="17.1.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>17.1.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="17.2" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>17.2</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>17.2.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="17.2.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>17.2.2</b> Importing data</a></li>
<li class="chapter" data-level="17.2.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>17.2.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="17.2.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>17.2.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="17.2.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>17.2.5</b> Sample information</a></li>
<li class="chapter" data-level="17.2.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>17.2.6</b> Feature information</a></li>
<li class="chapter" data-level="17.2.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>17.2.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>17.3</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>17.3.1</b> Subsetting</a></li>
<li class="chapter" data-level="17.3.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>17.3.2</b> Library sizes</a></li>
<li class="chapter" data-level="17.3.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>17.3.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="17.3.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>17.3.4</b> Data exploration</a></li>
<li class="chapter" data-level="17.3.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>17.3.5</b> Other functions</a></li>
<li class="chapter" data-level="17.3.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>17.3.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>17.4</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="17.4.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>17.4.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="17.4.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>17.4.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>17.5</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="17.5.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>17.5.1</b> Estimation</a></li>
<li class="chapter" data-level="17.5.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>17.5.2</b> Visualization</a></li>
<li class="chapter" data-level="17.5.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>17.5.3</b> Correlation</a></li>
<li class="chapter" data-level="17.5.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>17.5.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="17.6" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>17.6</b> Community similarity</a>
<ul>
<li class="chapter" data-level="17.6.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>17.6.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="17.6.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>17.6.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="17.6.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>17.6.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="17.6.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>17.6.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="17.6.5" data-path="exercises.html"><a href="exercises.html#redundance-analysis-rda"><i class="fa fa-check"></i><b>17.6.5</b> Redundance analysis (RDA)</a></li>
<li class="chapter" data-level="17.6.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>17.6.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="17.7" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>17.7</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="17.7.1" data-path="exercises.html"><a href="exercises.html#univariate-analyses"><i class="fa fa-check"></i><b>17.7.1</b> Univariate analyses</a></li>
<li class="chapter" data-level="17.7.2" data-path="exercises.html"><a href="exercises.html#differential-abundance-analysis-1"><i class="fa fa-check"></i><b>17.7.2</b> Differential abundance analysis</a></li>
</ul></li>
<li class="chapter" data-level="17.8" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>17.8</b> Visualization</a>
<ul>
<li class="chapter" data-level="17.8.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>17.8.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="17.8.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>17.8.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="17.9" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>17.9</b> Multiomics</a>
<ul>
<li class="chapter" data-level="17.9.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>17.9.1</b> Basic exploration</a></li>
<li class="chapter" data-level="17.9.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>17.9.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="17.9.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>17.9.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="17.9.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>17.9.4</b> Assay extraction</a></li>
<li class="chapter" data-level="17.9.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>17.9.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="17.9.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>17.9.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="18" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>18</b> Extra material</a>
<ul>
<li class="chapter" data-level="18.1" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>18.1</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="18.2" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>18.2</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="18.3" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>18.3</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html"><i class="fa fa-check"></i>Authors and contributors</a>
<ul>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#leo-lahti-dsc"><i class="fa fa-check"></i><em>Leo Lahti, DSc</em></a></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#tuomas-borman"><i class="fa fa-check"></i><em>Tuomas Borman</em></a></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#giulio-benedetti"><i class="fa fa-check"></i><em>Giulio Benedetti</em></a></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#felix-ernst-phd"><i class="fa fa-check"></i><em>Felix Ernst, PhD</em></a></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="" data-path="authors-and-contributors.html"><a href="authors-and-contributors.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multi-assay-analyses" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">Chapter 13</span> Multi-assay analyses<a href="multi-assay-analyses.html#multi-assay-analyses" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="multi-assay-analyses.html#cb373-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span></code></pre></div>
<p>Multi-omics approaches integrate data from multiple sources. For
example, we can integrate taxonomic abundance profiles with
metabolomic or other biomolecular profiling data to observe
associations, make predictions, or aim at causal
inferences. Integrating evidence across multiple sources can lead to
enhanced predictions, more holistic understanding, or facilitate the
discovery of novel biomarkers. In this section we demonstrate common
multi-assay data integration tasks.</p>
<p>Cross-correlation analysis is a straightforward approach that can
reveal strength and type of assocations between data sets. For instance,
we can analyze if higher presence of a specific taxon relates to higher
levels of a biomolecule.</p>
<p>The analyses can be facilitated by the multi-assay data containers,
<em>TreeSummarizedExperiment</em> and <em>MultiAssayExperiment</em>. These are
scalable and contain different types of data in a single container,
making this framework particularly suited for multi-assay microbiome
data incorporating different types of complementary data sources in a
single reproducible workflow. Solutions to a number of data
integration problems are discussed in more detail in Section
<a href="containers.html#containers">3</a>. Another experiment can be stored in <em>altExp</em> slot
of SE data container. Alternatively, both experiments can be stored
side-by-side in a MAE data container (see sections <a href="containers.html#alt-exp">3.2.5</a> and <a href="containers.html#mae">3.2.6</a>
to learn more about altExp and MAE objects, respectively). Different
experiments are first imported as single-assay data containers similarly to
the case when only one experiment is present. After that, the
different experiments can be combined into one multi-assay data
container. The result is a MAE object with multiple experiments in its
experiment slot, or a TreeSE object with alternative experiments in
the altExp slot.</p>
<p>As an example, we use a dataset from the following publication:
<span class="citation">(2021)</span> Xylo-oligosaccharides in prevention of hepatic
steatosis and adipose tissue inflammation: associating taxonomic and
metabolomic patterns in fecal microbiota with biclustering.
In this study, mice were fed either with a high-fat or a low-fat diet,
and with or without prebiotics, for the purpose studying whether prebiotics
attenuate the negative impact of a high-fat diet on health.</p>
<p>This example data can be loaded from microbiomeDataSets. The data is
already in MAE format. It includes three different experiments:
microbial abundance data, metabolite concentrations, and data about
different biomarkers. If you like to construct the same data object from the
original files instead, <a href="https://microbiome.github.io/OMA/containers.html#loading-experimental-microbiome-data">here</a>
you can find help for importing data into an SE object.</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="multi-assay-analyses.html#cb374-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb374-2"><a href="multi-assay-analyses.html#cb374-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(HintikkaXOData, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb374-3"><a href="multi-assay-analyses.html#cb374-3" aria-hidden="true" tabindex="-1"></a>mae <span class="ot">&lt;-</span> HintikkaXOData</span>
<span id="cb374-4"><a href="multi-assay-analyses.html#cb374-4" aria-hidden="true" tabindex="-1"></a>mae</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] microbiota: TreeSummarizedExperiment with 12706 rows and 40 columns
##  [2] metabolites: TreeSummarizedExperiment with 38 rows and 40 columns
##  [3] biomarkers: TreeSummarizedExperiment with 39 rows and 40 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files</code></pre>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="multi-assay-analyses.html#cb376-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb376-2"><a href="multi-assay-analyses.html#cb376-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop off those bacteria that do not include information in Phylum or lower levels</span></span>
<span id="cb376-3"><a href="multi-assay-analyses.html#cb376-3" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]])<span class="sc">$</span>Phylum), ]</span>
<span id="cb376-4"><a href="multi-assay-analyses.html#cb376-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Clean taxonomy data, so that names do not include additional characters</span></span>
<span id="cb376-5"><a href="multi-assay-analyses.html#cb376-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="fu">apply</span>(<span class="fu">rowData</span>(mae[[<span class="dv">1</span>]]), <span class="dv">2</span>, </span>
<span id="cb376-6"><a href="multi-assay-analyses.html#cb376-6" aria-hidden="true" tabindex="-1"></a>                                     str_remove, <span class="at">pattern =</span> <span class="st">&quot;._[0-9]__&quot;</span>))</span>
<span id="cb376-7"><a href="multi-assay-analyses.html#cb376-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Microbiome data</span></span>
<span id="cb376-8"><a href="multi-assay-analyses.html#cb376-8" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 12613 40 
## metadata(0):
## assays(1): counts
## rownames(12613): GAYR01026362.62.2014 CVJT01000011.50.2173 ...
##   JRJTB:03787:02429 JRJTB:03787:02478
## rowData names(7): Phylum Class ... Species OTU
## colnames(40): C1 C2 ... C39 C40
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="multi-assay-analyses.html#cb378-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Metabolite data</span></span>
<span id="cb378-2"><a href="multi-assay-analyses.html#cb378-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">2</span>]]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 38 40 
## metadata(0):
## assays(1): nmr
## rownames(38): Butyrate Acetate ... Malonate 1,3-dihydroxyacetone
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="multi-assay-analyses.html#cb380-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Biomarker data</span></span>
<span id="cb380-2"><a href="multi-assay-analyses.html#cb380-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">3</span>]]</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 39 40 
## metadata(0):
## assays(1): signals
## rownames(39): Triglycerides_liver CLSs_epi ... NPY_serum Glycogen_liver
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div id="cross-correlation" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Cross-correlation Analysis<a href="multi-assay-analyses.html#cross-correlation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Next we can perform a cross-correlation analysis. Let us analyze if
individual bacteria genera are correlated with concentrations of
individual metabolites. This helps to answer the following question: If
bacterium X is present, is the concentration of metabolite Y lower or higher?</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="multi-assay-analyses.html#cb382-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate microbiome data at family level</span></span>
<span id="cb382-2"><a href="multi-assay-analyses.html#cb382-2" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">agglomerateByPrevalence</span>(mae[[<span class="dv">1</span>]], <span class="at">rank =</span> <span class="st">&quot;Family&quot;</span>)</span>
<span id="cb382-3"><a href="multi-assay-analyses.html#cb382-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Does log10 transform for microbiome data</span></span>
<span id="cb382-4"><a href="multi-assay-analyses.html#cb382-4" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb382-5"><a href="multi-assay-analyses.html#cb382-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb382-6"><a href="multi-assay-analyses.html#cb382-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Give unique names so that we do not have problems when we are creating a plot</span></span>
<span id="cb382-7"><a href="multi-assay-analyses.html#cb382-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mae[[<span class="dv">1</span>]]) <span class="ot">&lt;-</span> <span class="fu">getTaxonomyLabels</span>(mae[[<span class="dv">1</span>]])</span>
<span id="cb382-8"><a href="multi-assay-analyses.html#cb382-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb382-9"><a href="multi-assay-analyses.html#cb382-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross correlates data sets</span></span>
<span id="cb382-10"><a href="multi-assay-analyses.html#cb382-10" aria-hidden="true" tabindex="-1"></a>correlations <span class="ot">&lt;-</span> <span class="fu">testExperimentCrossCorrelation</span>(mae, </span>
<span id="cb382-11"><a href="multi-assay-analyses.html#cb382-11" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">experiment1 =</span> <span class="dv">1</span>,</span>
<span id="cb382-12"><a href="multi-assay-analyses.html#cb382-12" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">experiment2 =</span> <span class="dv">2</span>,</span>
<span id="cb382-13"><a href="multi-assay-analyses.html#cb382-13" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">assay.type1 =</span> <span class="st">&quot;log10&quot;</span>, </span>
<span id="cb382-14"><a href="multi-assay-analyses.html#cb382-14" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">assay.type2 =</span> <span class="st">&quot;nmr&quot;</span>,</span>
<span id="cb382-15"><a href="multi-assay-analyses.html#cb382-15" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">method =</span> <span class="st">&quot;spearman&quot;</span>, </span>
<span id="cb382-16"><a href="multi-assay-analyses.html#cb382-16" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">p_adj_threshold =</span> <span class="cn">NULL</span>,</span>
<span id="cb382-17"><a href="multi-assay-analyses.html#cb382-17" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">cor_threshold =</span> <span class="cn">NULL</span>,</span>
<span id="cb382-18"><a href="multi-assay-analyses.html#cb382-18" aria-hidden="true" tabindex="-1"></a>                                               <span class="co"># Remove when mia is fixed</span></span>
<span id="cb382-19"><a href="multi-assay-analyses.html#cb382-19" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">mode =</span> <span class="st">&quot;matrix&quot;</span>,</span>
<span id="cb382-20"><a href="multi-assay-analyses.html#cb382-20" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">sort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb382-21"><a href="multi-assay-analyses.html#cb382-21" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">show_warnings =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Next, we create a heatmap depicting all cross-correlations between bacterial
genera and metabolite concentrations.</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="multi-assay-analyses.html#cb383-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap) </span>
<span id="cb383-2"><a href="multi-assay-analyses.html#cb383-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb383-3"><a href="multi-assay-analyses.html#cb383-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap and store it</span></span>
<span id="cb383-4"><a href="multi-assay-analyses.html#cb383-4" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(correlations<span class="sc">$</span>cor,</span>
<span id="cb383-5"><a href="multi-assay-analyses.html#cb383-5" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Print values to cells</span></span>
<span id="cb383-6"><a href="multi-assay-analyses.html#cb383-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">cell_fun =</span> <span class="cf">function</span>(j, i, x, y, width, height, fill) {</span>
<span id="cb383-7"><a href="multi-assay-analyses.html#cb383-7" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># If the p-value is under threshold</span></span>
<span id="cb383-8"><a href="multi-assay-analyses.html#cb383-8" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span>( <span class="sc">!</span><span class="fu">is.na</span>(correlations<span class="sc">$</span>p_adj[i, j]) <span class="sc">&amp;</span> correlations<span class="sc">$</span>p_adj[i, j] <span class="sc">&lt;</span> <span class="fl">0.05</span> ){</span>
<span id="cb383-9"><a href="multi-assay-analyses.html#cb383-9" aria-hidden="true" tabindex="-1"></a>                        <span class="co"># Print &quot;X&quot;</span></span>
<span id="cb383-10"><a href="multi-assay-analyses.html#cb383-10" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">grid.text</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%s&quot;</span>, <span class="st">&quot;X&quot;</span>), x, y, <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="st">&quot;black&quot;</span>))</span>
<span id="cb383-11"><a href="multi-assay-analyses.html#cb383-11" aria-hidden="true" tabindex="-1"></a>                        }</span>
<span id="cb383-12"><a href="multi-assay-analyses.html#cb383-12" aria-hidden="true" tabindex="-1"></a>                    },</span>
<span id="cb383-13"><a href="multi-assay-analyses.html#cb383-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="st">&quot;&quot;</span>, <span class="at">legend_height =</span> <span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">&quot;cm&quot;</span>))</span>
<span id="cb383-14"><a href="multi-assay-analyses.html#cb383-14" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb383-15"><a href="multi-assay-analyses.html#cb383-15" aria-hidden="true" tabindex="-1"></a>plot</span></code></pre></div>
<p><img src="23_multi-assay_analyses_files/figure-html/cross-correlation6-1.png" width="960" /></p>
</div>
<div id="mofa" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Multi-Omics Factor Analysis<a href="multi-assay-analyses.html#mofa" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Multi-Omics Factor Analysis (MOFA) is an unsupervised method for integrating multi-omic data sets in a downstream analysis <span class="citation">(Argelaguet 2018)</span>. It could be
seen as a generalization of principal component analysis. Yet, with the ability
to infer a latent (low-dimensional) representation, shared among the multiple
(-omics) data sets in hand.</p>
<p>We use the R <a href="https://biofam.github.io/MOFA2/index.html">MOFA2</a>
package for the analysis, and
<a href="https://biofam.github.io/MOFA2/installation.html">install</a> the
corresponding dependencies.</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="multi-assay-analyses.html#cb384-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MOFA2)</span>
<span id="cb384-2"><a href="multi-assay-analyses.html#cb384-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-3"><a href="multi-assay-analyses.html#cb384-3" aria-hidden="true" tabindex="-1"></a><span class="co"># For inter-operability between Python and R, and setting Python dependencies,</span></span>
<span id="cb384-4"><a href="multi-assay-analyses.html#cb384-4" aria-hidden="true" tabindex="-1"></a><span class="co"># reticulate package is needed</span></span>
<span id="cb384-5"><a href="multi-assay-analyses.html#cb384-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb384-6"><a href="multi-assay-analyses.html#cb384-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Let us assume that these have been installed already.</span></span>
<span id="cb384-7"><a href="multi-assay-analyses.html#cb384-7" aria-hidden="true" tabindex="-1"></a><span class="co">#reticulate::install_miniconda(force = TRUE)</span></span>
<span id="cb384-8"><a href="multi-assay-analyses.html#cb384-8" aria-hidden="true" tabindex="-1"></a><span class="co">#reticulate::use_miniconda(condaenv = &quot;env1&quot;, required = FALSE)</span></span>
<span id="cb384-9"><a href="multi-assay-analyses.html#cb384-9" aria-hidden="true" tabindex="-1"></a><span class="co">#reticulate::py_install(packages = c(&quot;mofapy2&quot;), pip = TRUE, python_version=3.6)</span></span></code></pre></div>
<p>The <code>mae</code> object could be used straight to create the MOFA model. Yet,
we transform our assays since the model assumes normality per
default. We can also use Poisson or Bernoulli distributions among others.</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="multi-assay-analyses.html#cb385-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MOFA2)</span>
<span id="cb385-2"><a href="multi-assay-analyses.html#cb385-2" aria-hidden="true" tabindex="-1"></a><span class="co"># For simplicity, classify all high-fat diets as high-fat, and all the low-fat </span></span>
<span id="cb385-3"><a href="multi-assay-analyses.html#cb385-3" aria-hidden="true" tabindex="-1"></a><span class="co"># diets as low-fat diets</span></span>
<span id="cb385-4"><a href="multi-assay-analyses.html#cb385-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(mae)<span class="sc">$</span>Diet <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">colData</span>(mae)<span class="sc">$</span>Diet <span class="sc">==</span> <span class="st">&quot;High-fat&quot;</span> <span class="sc">|</span> </span>
<span id="cb385-5"><a href="multi-assay-analyses.html#cb385-5" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">colData</span>(mae)<span class="sc">$</span>Diet <span class="sc">==</span> <span class="st">&quot;High-fat + XOS&quot;</span>, </span>
<span id="cb385-6"><a href="multi-assay-analyses.html#cb385-6" aria-hidden="true" tabindex="-1"></a>                            <span class="st">&quot;High-fat&quot;</span>, <span class="st">&quot;Low-fat&quot;</span>)</span>
<span id="cb385-7"><a href="multi-assay-analyses.html#cb385-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-8"><a href="multi-assay-analyses.html#cb385-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Removing duplicates at the microbiome data</span></span>
<span id="cb385-9"><a href="multi-assay-analyses.html#cb385-9" aria-hidden="true" tabindex="-1"></a><span class="co"># which are also in form e.g. &quot;Ambiguous&quot; and &quot;uncultured&quot; taxa</span></span>
<span id="cb385-10"><a href="multi-assay-analyses.html#cb385-10" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> mae[[<span class="dv">1</span>]][<span class="sc">!</span><span class="fu">duplicated</span>(<span class="fu">rownames</span>(<span class="fu">assay</span>(mae[[<span class="dv">1</span>]]))), ]</span>
<span id="cb385-11"><a href="multi-assay-analyses.html#cb385-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-12"><a href="multi-assay-analyses.html#cb385-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Transforming microbiome data with rclr</span></span>
<span id="cb385-13"><a href="multi-assay-analyses.html#cb385-13" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">1</span>]], <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb385-14"><a href="multi-assay-analyses.html#cb385-14" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">1</span>]], <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>, <span class="at">method =</span> <span class="st">&quot;rclr&quot;</span>)</span>
<span id="cb385-15"><a href="multi-assay-analyses.html#cb385-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-16"><a href="multi-assay-analyses.html#cb385-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Transforming metabolomic data with log10</span></span>
<span id="cb385-17"><a href="multi-assay-analyses.html#cb385-17" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">2</span>]], <span class="at">assay.type =</span> <span class="st">&quot;nmr&quot;</span>,</span>
<span id="cb385-18"><a href="multi-assay-analyses.html#cb385-18" aria-hidden="true" tabindex="-1"></a>                            <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>,</span>
<span id="cb385-19"><a href="multi-assay-analyses.html#cb385-19" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>)</span>
<span id="cb385-20"><a href="multi-assay-analyses.html#cb385-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-21"><a href="multi-assay-analyses.html#cb385-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Transforming biomarker data with z-transform</span></span>
<span id="cb385-22"><a href="multi-assay-analyses.html#cb385-22" aria-hidden="true" tabindex="-1"></a>mae[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(mae[[<span class="dv">3</span>]], <span class="at">assay.type =</span> <span class="st">&quot;signals&quot;</span>,</span>
<span id="cb385-23"><a href="multi-assay-analyses.html#cb385-23" aria-hidden="true" tabindex="-1"></a>                            <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>,</span>
<span id="cb385-24"><a href="multi-assay-analyses.html#cb385-24" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb385-25"><a href="multi-assay-analyses.html#cb385-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-26"><a href="multi-assay-analyses.html#cb385-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Removing assays no longer needed</span></span>
<span id="cb385-27"><a href="multi-assay-analyses.html#cb385-27" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;counts&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb385-28"><a href="multi-assay-analyses.html#cb385-28" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">1</span>]], <span class="st">&quot;log10&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb385-29"><a href="multi-assay-analyses.html#cb385-29" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">2</span>]], <span class="st">&quot;nmr&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb385-30"><a href="multi-assay-analyses.html#cb385-30" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(mae[[<span class="dv">3</span>]], <span class="st">&quot;signals&quot;</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb385-31"><a href="multi-assay-analyses.html#cb385-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-32"><a href="multi-assay-analyses.html#cb385-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Building our mofa model</span></span>
<span id="cb385-33"><a href="multi-assay-analyses.html#cb385-33" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">create_mofa_from_MultiAssayExperiment</span>(mae,</span>
<span id="cb385-34"><a href="multi-assay-analyses.html#cb385-34" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">groups =</span> <span class="st">&quot;Diet&quot;</span>, </span>
<span id="cb385-35"><a href="multi-assay-analyses.html#cb385-35" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">extract_metadata =</span> <span class="cn">TRUE</span>)</span>
<span id="cb385-36"><a href="multi-assay-analyses.html#cb385-36" aria-hidden="true" tabindex="-1"></a>model</span></code></pre></div>
<pre><code>## Untrained MOFA model with the following characteristics: 
##  Number of views: 3 
##  Views names: microbiota metabolites biomarkers 
##  Number of features (per view): 38 38 39 
##  Number of groups: 2 
##  Groups names: High-fat Low-fat 
##  Number of samples (per group): 20 20 
## </code></pre>
<p>Model options can be defined as follows:</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="multi-assay-analyses.html#cb387-1" aria-hidden="true" tabindex="-1"></a>model_opts <span class="ot">&lt;-</span> <span class="fu">get_default_model_options</span>(model)</span>
<span id="cb387-2"><a href="multi-assay-analyses.html#cb387-2" aria-hidden="true" tabindex="-1"></a>model_opts<span class="sc">$</span>num_factors <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb387-3"><a href="multi-assay-analyses.html#cb387-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(model_opts)</span></code></pre></div>
<pre><code>## $likelihoods
##  microbiota metabolites  biomarkers 
##  &quot;gaussian&quot;  &quot;gaussian&quot;  &quot;gaussian&quot; 
## 
## $num_factors
## [1] 5
## 
## $spikeslab_factors
## [1] FALSE
## 
## $spikeslab_weights
## [1] FALSE
## 
## $ard_factors
## [1] TRUE
## 
## $ard_weights
## [1] TRUE</code></pre>
<p>Training options for the model are defined in the following way:</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="multi-assay-analyses.html#cb389-1" aria-hidden="true" tabindex="-1"></a>train_opts <span class="ot">&lt;-</span> <span class="fu">get_default_training_options</span>(model)</span>
<span id="cb389-2"><a href="multi-assay-analyses.html#cb389-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(train_opts)</span></code></pre></div>
<pre><code>## $maxiter
## [1] 1000
## 
## $convergence_mode
## [1] &quot;fast&quot;
## 
## $drop_factor_threshold
## [1] -1
## 
## $verbose
## [1] FALSE
## 
## $startELBO
## [1] 1
## 
## $freqELBO
## [1] 5</code></pre>
<p>The model is then prepared with <code>prepare_mofa</code> and trained with <code>run_mofa</code>:</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="multi-assay-analyses.html#cb391-1" aria-hidden="true" tabindex="-1"></a>model.prepared <span class="ot">&lt;-</span> <span class="fu">prepare_mofa</span>(</span>
<span id="cb391-2"><a href="multi-assay-analyses.html#cb391-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> model,</span>
<span id="cb391-3"><a href="multi-assay-analyses.html#cb391-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model_options =</span> model_opts</span>
<span id="cb391-4"><a href="multi-assay-analyses.html#cb391-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb391-5"><a href="multi-assay-analyses.html#cb391-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb391-6"><a href="multi-assay-analyses.html#cb391-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Some systems may require the specification `use_basilisk = TRUE`</span></span>
<span id="cb391-7"><a href="multi-assay-analyses.html#cb391-7" aria-hidden="true" tabindex="-1"></a><span class="co"># so it has been added to the following code</span></span>
<span id="cb391-8"><a href="multi-assay-analyses.html#cb391-8" aria-hidden="true" tabindex="-1"></a>model.trained <span class="ot">&lt;-</span> <span class="fu">run_mofa</span>(model.prepared, <span class="at">use_basilisk =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>The explained variance is visualized with the <code>plot_variance_explained</code> function:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="multi-assay-analyses.html#cb392-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb392-2"><a href="multi-assay-analyses.html#cb392-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb392-3"><a href="multi-assay-analyses.html#cb392-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb392-4"><a href="multi-assay-analyses.html#cb392-4" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">plot_variance_explained</span>(model.trained,</span>
<span id="cb392-5"><a href="multi-assay-analyses.html#cb392-5" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">x =</span> <span class="st">&quot;view&quot;</span>, <span class="at">y =</span> <span class="st">&quot;factor&quot;</span>,</span>
<span id="cb392-6"><a href="multi-assay-analyses.html#cb392-6" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">plot_total =</span> T)</span>
<span id="cb392-7"><a href="multi-assay-analyses.html#cb392-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb392-8"><a href="multi-assay-analyses.html#cb392-8" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plot_list, <span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb392-9"><a href="multi-assay-analyses.html#cb392-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">&quot;Variance Explained per factor and assay&quot;</span>,</span>
<span id="cb392-10"><a href="multi-assay-analyses.html#cb392-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">theme =</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)))</span></code></pre></div>
<p><img src="23_multi-assay_analyses_files/figure-html/unnamed-chunk-5-1.png" width="960" /></p>
<p>The top weights for each assay using all five factors:</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="multi-assay-analyses.html#cb393-1" aria-hidden="true" tabindex="-1"></a>custom_plotter <span class="ot">&lt;-</span> <span class="cf">function</span>(name) {</span>
<span id="cb393-2"><a href="multi-assay-analyses.html#cb393-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb393-3"><a href="multi-assay-analyses.html#cb393-3" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">plot_top_weights</span>(model.trained,</span>
<span id="cb393-4"><a href="multi-assay-analyses.html#cb393-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">view =</span> name,</span>
<span id="cb393-5"><a href="multi-assay-analyses.html#cb393-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">factors =</span> <span class="st">&quot;all&quot;</span>,</span>
<span id="cb393-6"><a href="multi-assay-analyses.html#cb393-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">nfeatures =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb393-7"><a href="multi-assay-analyses.html#cb393-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;Top weights of the &quot;</span>, name, <span class="st">&quot; assay&quot;</span>))</span>
<span id="cb393-8"><a href="multi-assay-analyses.html#cb393-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb393-9"><a href="multi-assay-analyses.html#cb393-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb393-10"><a href="multi-assay-analyses.html#cb393-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-11"><a href="multi-assay-analyses.html#cb393-11" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;microbiota&quot;</span>, <span class="st">&quot;metabolites&quot;</span>, <span class="st">&quot;biomarkers&quot;</span>), custom_plotter)</span>
<span id="cb393-12"><a href="multi-assay-analyses.html#cb393-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-13"><a href="multi-assay-analyses.html#cb393-13" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plot_list, <span class="at">nrow =</span> <span class="dv">3</span>) <span class="sc">&amp;</span> <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>))</span></code></pre></div>
<p><img src="23_multi-assay_analyses_files/figure-html/unnamed-chunk-6-1.png" width="960" /></p>
<p>More tutorials and examples of using the package are found at <a href="https://biofam.github.io/MOFA2/tutorials.html">link</a></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="machine_learning.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="viz-chapter.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/23_multi-assay_analyses.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/23_multi-assay_analyses.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/23_multi-assay_analyses.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
