<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Network comparison | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 13 Network comparison | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Network comparison | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Network comparison | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="network-learning.html"/>
<link rel="next" href="machine_learning.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.4</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis with igraph</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-plot"><i class="fa fa-check"></i><b>12.2.1</b> Network plot</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#centrality-measures"><i class="fa fa-check"></i><b>12.2.2</b> Centrality measures</a></li>
<li class="chapter" data-level="12.2.3" data-path="network-learning.html"><a href="network-learning.html#scale-node-sizes-by-centrality"><i class="fa fa-check"></i><b>12.2.3</b> Scale node sizes by centrality</a></li>
<li class="chapter" data-level="12.2.4" data-path="network-learning.html"><a href="network-learning.html#degree-distribution"><i class="fa fa-check"></i><b>12.2.4</b> Degree distribution</a></li>
<li class="chapter" data-level="12.2.5" data-path="network-learning.html"><a href="network-learning.html#clustered-heatmaps"><i class="fa fa-check"></i><b>12.2.5</b> Clustered heatmaps</a></li>
<li class="chapter" data-level="12.2.6" data-path="network-learning.html"><a href="network-learning.html#global-network-measures"><i class="fa fa-check"></i><b>12.2.6</b> Global network measures</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.3</b> Network analysis with NetCoMi</a></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.4</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.4.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.4.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.4.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.5</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.6</b> More about association measures</a></li>
<li class="chapter" data-level="12.7" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.7</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.7.1</b> SparCC</a></li>
<li class="chapter" data-level="12.7.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.7.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.7.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.7.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.7.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.7.4</b> Network plots</a></li>
<li class="chapter" data-level="12.7.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.7.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="network-comparison" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">Chapter 13</span> Network comparison<a href="network-comparison.html#network-comparison" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>This chapter assumes that you are already familiar with how to construct and analyze a single microbial network, which is explained in Chapter <a href="network-learning.html#network-learning">12</a>.</p>
<p>Since microbial interactions are likely to change between conditions, such as between patients and healthy individuals or between different environmental states, identifying network differences between groups is often an integral secondary analysis step. Differences can be detected visually by plotting the networks side by side, or quantitatively using differential network analysis tools.</p>
<p>Two approaches for comparing networks between two conditions are considered in this chapter:</p>
<ul>
<li><strong>Differential network analysis</strong>, which analyzes differences in network metrics and network structure.</li>
<li><strong>Differential association analysis</strong>, which focuses on differences in the strength of individual associations.</li>
</ul>
<p>See Section <a href="network-comparison.html#netcomp-methods">13.5</a> for further details on these two approaches and the methods used in this chapter.</p>
<p>Here we use <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> <span class="citation">(Peschel et al. 2021)</span> for network comparison, which includes several <strong>differential network analysis approaches</strong> as well as functionality for *differential association analysis**, i.e., generating a differential network. How to install NetCoMi from GitHub is explained in chapter <a href="network-learning.html#network-learning">12</a>.</p>
<p>The <strong>PeerJ data set</strong> <span class="citation">(Potbhare et al. 2022)</span> containing skin microbial profiles for 58 subjects is again used in this chapter. The dataset also includes information on the subjects geographic location, gender, age and diet. Whether the <strong>skin microbiome differs</strong> between people with <strong>different diets</strong> is an interesting question that will be explored in this chapter.</p>
<div id="data-preparation-1" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Data preparation<a href="network-comparison.html#data-preparation-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We perform the same data preprocessing steps as in chapter <a href="network-learning.html#network-learning">12</a>.</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb418-1"><a href="network-comparison.html#cb418-1" tabindex="-1"></a><span class="fu">library</span>(NetCoMi)</span>
<span id="cb418-2"><a href="network-comparison.html#cb418-2" tabindex="-1"></a><span class="fu">library</span>(mia)</span></code></pre></div>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="network-comparison.html#cb419-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;peerj13075&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb419-2"><a href="network-comparison.html#cb419-2" tabindex="-1"></a>tse0 <span class="ot">&lt;-</span> peerj13075</span></code></pre></div>
<div class="sourceCode" id="cb420"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb420-1"><a href="network-comparison.html#cb420-1" tabindex="-1"></a><span class="co"># Agglomerate to genus level</span></span>
<span id="cb420-2"><a href="network-comparison.html#cb420-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(tse0, <span class="at">rank =</span> <span class="st">&quot;genus&quot;</span>) </span>
<span id="cb420-3"><a href="network-comparison.html#cb420-3" tabindex="-1"></a></span>
<span id="cb420-4"><a href="network-comparison.html#cb420-4" tabindex="-1"></a><span class="co"># Add relative abundances</span></span>
<span id="cb420-5"><a href="network-comparison.html#cb420-5" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, </span>
<span id="cb420-6"><a href="network-comparison.html#cb420-6" tabindex="-1"></a>                      <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb420-7"><a href="network-comparison.html#cb420-7" tabindex="-1"></a>                      <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb420-8"><a href="network-comparison.html#cb420-8" tabindex="-1"></a>                      <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>) </span>
<span id="cb420-9"><a href="network-comparison.html#cb420-9" tabindex="-1"></a></span>
<span id="cb420-10"><a href="network-comparison.html#cb420-10" tabindex="-1"></a><span class="co"># Filter by prevalence</span></span>
<span id="cb420-11"><a href="network-comparison.html#cb420-11" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">subsetByPrevalentFeatures</span>(tse,</span>
<span id="cb420-12"><a href="network-comparison.html#cb420-12" tabindex="-1"></a>                                 <span class="at">prevalence =</span> <span class="fl">0.2</span>,</span>
<span id="cb420-13"><a href="network-comparison.html#cb420-13" tabindex="-1"></a>                                 <span class="at">detection =</span> <span class="dv">0</span>,</span>
<span id="cb420-14"><a href="network-comparison.html#cb420-14" tabindex="-1"></a>                                 <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb420-15"><a href="network-comparison.html#cb420-15" tabindex="-1"></a></span>
<span id="cb420-16"><a href="network-comparison.html#cb420-16" tabindex="-1"></a><span class="co"># Add log10-transformed abundances</span></span>
<span id="cb420-17"><a href="network-comparison.html#cb420-17" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb420-18"><a href="network-comparison.html#cb420-18" tabindex="-1"></a></span>
<span id="cb420-19"><a href="network-comparison.html#cb420-19" tabindex="-1"></a><span class="co"># Add clr-transformed abundances</span></span>
<span id="cb420-20"><a href="network-comparison.html#cb420-20" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>Based on Diet, the <code>tse</code> object is then split into two groups: One with mixed diet subjects, and one with vegetarian subjects. Both subsets have nearly the same sample size and are therefore comparable.</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="network-comparison.html#cb421-1" tabindex="-1"></a><span class="fu">table</span>(tse<span class="sc">$</span>Diet)</span></code></pre></div>
<pre><code>## 
## Mixed   Veg 
##    28    30</code></pre>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb423-1"><a href="network-comparison.html#cb423-1" tabindex="-1"></a>tse_list <span class="ot">&lt;-</span> <span class="fu">splitOn</span>(tse, <span class="at">f =</span> <span class="st">&quot;Diet&quot;</span>, <span class="at">use_names =</span> <span class="cn">TRUE</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="network-learning-and-analysis" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Network learning and analysis<a href="network-comparison.html#network-learning-and-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The approach starts again with network construction and analysis, but this time we pass the two data sets to <code>netConstruct</code> to perform a network comparison.</p>
<p>The <code>rep.num</code> argument is set to 10 to perform only 10 repetitions in the model selection approach. This speeds up the permutation tests performed later, and has a negligible effect for this data set.</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb424-1"><a href="network-comparison.html#cb424-1" tabindex="-1"></a>spring_net_diet <span class="ot">&lt;-</span> <span class="fu">netConstruct</span>(<span class="at">data =</span> tse_list<span class="sc">$</span>Mixed,</span>
<span id="cb424-2"><a href="network-comparison.html#cb424-2" tabindex="-1"></a>                                <span class="at">data2 =</span> tse_list<span class="sc">$</span>Veg,</span>
<span id="cb424-3"><a href="network-comparison.html#cb424-3" tabindex="-1"></a>                                <span class="at">taxRank =</span> <span class="st">&quot;genus&quot;</span>,</span>
<span id="cb424-4"><a href="network-comparison.html#cb424-4" tabindex="-1"></a>                                <span class="at">filtTax =</span> <span class="st">&quot;highestFreq&quot;</span>,</span>
<span id="cb424-5"><a href="network-comparison.html#cb424-5" tabindex="-1"></a>                                <span class="at">filtTaxPar =</span> <span class="fu">list</span>(<span class="at">highestFreq  =</span> <span class="dv">100</span>),</span>
<span id="cb424-6"><a href="network-comparison.html#cb424-6" tabindex="-1"></a>                                <span class="at">measure =</span> <span class="st">&quot;spring&quot;</span>,</span>
<span id="cb424-7"><a href="network-comparison.html#cb424-7" tabindex="-1"></a>                                <span class="at">measurePar =</span> <span class="fu">list</span>(<span class="at">nlambda =</span> <span class="dv">20</span>, </span>
<span id="cb424-8"><a href="network-comparison.html#cb424-8" tabindex="-1"></a>                                                  <span class="at">rep.num =</span> <span class="dv">10</span>,</span>
<span id="cb424-9"><a href="network-comparison.html#cb424-9" tabindex="-1"></a>                                                  <span class="at">thresh =</span> <span class="fl">0.05</span>,</span>
<span id="cb424-10"><a href="network-comparison.html#cb424-10" tabindex="-1"></a>                                                  <span class="at">Rmethod =</span> <span class="st">&quot;approx&quot;</span>),</span>
<span id="cb424-11"><a href="network-comparison.html#cb424-11" tabindex="-1"></a>                                <span class="at">sparsMethod =</span> <span class="st">&quot;none&quot;</span>, </span>
<span id="cb424-12"><a href="network-comparison.html#cb424-12" tabindex="-1"></a>                                <span class="at">dissFunc =</span> <span class="st">&quot;signed&quot;</span>,</span>
<span id="cb424-13"><a href="network-comparison.html#cb424-13" tabindex="-1"></a>                                <span class="at">verbose =</span> <span class="dv">3</span>,</span>
<span id="cb424-14"><a href="network-comparison.html#cb424-14" tabindex="-1"></a>                                <span class="at">seed =</span> <span class="dv">13075</span>)</span></code></pre></div>
<p>All network measures are now computed for both networks. Also, both GCMs are plotted together with a third matrix containing the differences between the GCMs and significance codes that express if the differences are significantly different from zero.</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb425-1"><a href="network-comparison.html#cb425-1" tabindex="-1"></a>spring_netprops_diet <span class="ot">&lt;-</span> <span class="fu">netAnalyze</span>(spring_net_diet, </span>
<span id="cb425-2"><a href="network-comparison.html#cb425-2" tabindex="-1"></a>                                   <span class="at">clustMethod =</span> <span class="st">&quot;cluster_fast_greedy&quot;</span>,</span>
<span id="cb425-3"><a href="network-comparison.html#cb425-3" tabindex="-1"></a>                                   <span class="at">hubPar =</span> <span class="st">&quot;eigenvector&quot;</span>,</span>
<span id="cb425-4"><a href="network-comparison.html#cb425-4" tabindex="-1"></a>                                   <span class="at">normDeg =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="61_network_comparison_files/figure-html/netAnalyze_diet-1.png" width="960" /></p>
<p>In both of the networks, some graphlet correlations are significantly different from zero. However, none of the correlations are significantly different between the groups.</p>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb426-1"><a href="network-comparison.html#cb426-1" tabindex="-1"></a><span class="fu">summary</span>(spring_netprops_diet, <span class="at">groupNames =</span> <span class="fu">c</span>(<span class="st">&quot;Mixed diet&quot;</span>, <span class="st">&quot;Vegetarian&quot;</span>))</span></code></pre></div>
<pre><code>## 
## Component sizes
## ```````````````
## Mixed diet:                   
## size: 18 8 4 3 2  1
##    #:  1 1 1 1 3 43
## Vegetarian:                 
## size: 21 5 4 2  1
##    #:  1 1 1 3 46
## ______________________________
## Global network properties
## `````````````````````````
## Largest connected component (LCC):
##                          Mixed diet Vegetarian
## Relative LCC size           0.21951    0.25610
## Clustering coefficient      0.10393    0.22433
## Modularity                  0.51500    0.47707
## Positive edge percentage   80.00000  100.00000
## Edge density                0.13072    0.12381
## Natural connectivity        0.07312    0.06253
## Vertex connectivity         1.00000    1.00000
## Edge connectivity           1.00000    1.00000
## Average dissimilarity*      0.96009    0.96154
## Average path length**       2.22768    2.24622
## 
## Whole network:
##                          Mixed diet Vegetarian
## Number of components       50.00000   52.00000
## Clustering coefficient      0.07506    0.16530
## Modularity                  0.75184    0.66821
## Positive edge percentage   88.57143  100.00000
## Edge density                0.01054    0.01084
## Natural connectivity        0.01342    0.01347
## -----
## *: Dissimilarity = 1 - edge weight
## **: Path length = Units with average dissimilarity
## 
## ______________________________
## Clusters
## - In the whole network
## - Algorithm: cluster_fast_greedy
## ```````````````````````````````` 
## Mixed diet:                          
## name:  0 1 2 3 4 5 6 7 8 9
##    #: 43 8 6 6 6 4 3 2 2 2
## 
## Vegetarian:                        
## name:  0 1 2 3 4 5 6 7 8
##    #: 46 9 8 4 5 4 2 2 2
## 
## ______________________________
## Hubs
## - In alphabetical/numerical order
## - Based on empirical quantiles of centralities
## ```````````````````````````````````````````````
##   Mixed diet           Vegetarian
##  Citrobacter            Aeromonas
##      Erwinia              Erwinia
##  Escherichia Escherichia/Shigella
##     Serratia           Salmonella
##   Shewanella          Siccibacter
## 
## ______________________________
## Centrality measures
## - In decreasing order
## - Centrality of disconnected components is zero
## ````````````````````````````````````````````````
## Degree (unnormalized):
##                       Mixed diet Vegetarian
##               Erwinia          6          6
##           Citrobacter          4          0
##              Serratia          4          3
##         Streptococcus          3          0
##               Pantoea          2          1
##                           ______     ______
##               Erwinia          6          6
##             Aeromonas          1          5
##  Escherichia/Shigella          2          5
##           Siccibacter          0          5
##            Salmonella          2          3
## 
## Betweenness centrality (normalized):
##                       Mixed diet Vegetarian
##               Erwinia    0.60294    0.54737
##              Serratia    0.41912        0.1
##         Streptococcus    0.27941          0
##          Enterobacter    0.24265          0
##           Citrobacter    0.24265          0
##                           ______     ______
##               Erwinia    0.60294    0.54737
##  Escherichia/Shigella    0.11765    0.53158
##             Aeromonas          0    0.36842
##           Siccibacter          0        0.3
##         Thiolamprovum          0    0.20526
## 
## Closeness centrality (normalized):
##                       Mixed diet Vegetarian
##               Erwinia    0.85251    0.80605
##              Serratia    0.78452    0.56308
##           Citrobacter    0.71265          0
##          Enterobacter    0.64822    0.49217
##         Streptococcus    0.61535          0
##                           ______     ______
##  Escherichia/Shigella    0.57246    0.80747
##               Erwinia    0.85251    0.80605
##             Aeromonas     0.4768    0.74282
##           Siccibacter          0    0.72705
##            Salmonella    0.57364    0.66777
## 
## Eigenvector centrality (normalized):
##                       Mixed diet Vegetarian
##              Serratia          1    0.41577
##               Erwinia    0.97703          1
##           Citrobacter    0.80991          0
##           Escherichia    0.60982          0
##            Shewanella    0.41174     0.1238
##                           ______     ______
##               Erwinia    0.97703          1
##  Escherichia/Shigella    0.38659    0.96856
##           Siccibacter          0      0.888
##             Aeromonas    0.26033    0.68516
##            Salmonella     0.3153    0.64405</code></pre>
<p>For each centrality measure, the five nodes with the highest centrality in each group are plotted by default.</p>
<p>We notice some differences in the network properties. The differential network analysis performed in the next section will show if the differences are significant.</p>
</div>
<div id="differential-network-analysis" class="section level2 hasAnchor" number="13.3">
<h2><span class="header-section-number">13.3</span> Differential network analysis<a href="network-comparison.html#differential-network-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="visual-comparison" class="section level3 hasAnchor" number="13.3.1">
<h3><span class="header-section-number">13.3.1</span> Visual comparison<a href="network-comparison.html#visual-comparison" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We start with a visual comparison of the two networks using NetCoMis plot function. The same configuration as in chapter <a href="network-learning.html#network-learning">12</a> is used.</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb428-1"><a href="network-comparison.html#cb428-1" tabindex="-1"></a><span class="fu">plot</span>(spring_netprops_diet,</span>
<span id="cb428-2"><a href="network-comparison.html#cb428-2" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.97</span>,</span>
<span id="cb428-3"><a href="network-comparison.html#cb428-3" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb428-4"><a href="network-comparison.html#cb428-4" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb428-5"><a href="network-comparison.html#cb428-5" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">&quot;eigenvector&quot;</span>,</span>
<span id="cb428-6"><a href="network-comparison.html#cb428-6" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">2</span>,</span>
<span id="cb428-7"><a href="network-comparison.html#cb428-7" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">&quot;cluster&quot;</span>, </span>
<span id="cb428-8"><a href="network-comparison.html#cb428-8" tabindex="-1"></a>     <span class="at">sameColThresh =</span> <span class="dv">2</span>,</span>
<span id="cb428-9"><a href="network-comparison.html#cb428-9" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">&quot;darkgray&quot;</span>,</span>
<span id="cb428-10"><a href="network-comparison.html#cb428-10" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="dv">2</span>,</span>
<span id="cb428-11"><a href="network-comparison.html#cb428-11" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb428-12"><a href="network-comparison.html#cb428-12" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">&quot;Mixed diet&quot;</span>, </span>
<span id="cb428-13"><a href="network-comparison.html#cb428-13" tabindex="-1"></a>     <span class="at">title2 =</span> <span class="st">&quot;Vegetarian&quot;</span>,</span>
<span id="cb428-14"><a href="network-comparison.html#cb428-14" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb428-15"><a href="network-comparison.html#cb428-15" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="dv">2</span>,</span>
<span id="cb428-16"><a href="network-comparison.html#cb428-16" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb428-17"><a href="network-comparison.html#cb428-17" tabindex="-1"></a></span>
<span id="cb428-18"><a href="network-comparison.html#cb428-18" tabindex="-1"></a><span class="co"># Overlay a transparent plot on which the legend is plotted</span></span>
<span id="cb428-19"><a href="network-comparison.html#cb428-19" tabindex="-1"></a><span class="fu">par</span>(<span class="at">fig=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="at">oma=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">new=</span><span class="cn">TRUE</span>)</span>
<span id="cb428-20"><a href="network-comparison.html#cb428-20" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="at">type=</span><span class="st">&#39;n&#39;</span>, <span class="at">bty=</span><span class="st">&#39;n&#39;</span>, <span class="at">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="at">yaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb428-21"><a href="network-comparison.html#cb428-21" tabindex="-1"></a></span>
<span id="cb428-22"><a href="network-comparison.html#cb428-22" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">0.2</span>, <span class="sc">-</span><span class="fl">0.9</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">title =</span> <span class="st">&quot;estimated correlation:&quot;</span>,</span>
<span id="cb428-23"><a href="network-comparison.html#cb428-23" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;#009900&quot;</span>,<span class="st">&quot;red&quot;</span>), </span>
<span id="cb428-24"><a href="network-comparison.html#cb428-24" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="61_network_comparison_files/figure-html/network_plot_diet_difflay-1.png" width="1632" /></p>
<p>The layout is computed separately for each network, making it difficult to visually compare certain associations. It is therefore recommended to use the same layout for both groups (argument <code>sameLayout</code>). Instead of simply copying one layout to the other network, we set <code>layoutGroup</code> to union. This ensures that the nodes are placed as optimally as possible for both networks.</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="network-comparison.html#cb429-1" tabindex="-1"></a><span class="fu">plot</span>(spring_netprops_diet,</span>
<span id="cb429-2"><a href="network-comparison.html#cb429-2" tabindex="-1"></a>     <span class="at">sameLayout =</span> <span class="cn">TRUE</span>,</span>
<span id="cb429-3"><a href="network-comparison.html#cb429-3" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb429-4"><a href="network-comparison.html#cb429-4" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="st">&quot;inboth&quot;</span>,</span>
<span id="cb429-5"><a href="network-comparison.html#cb429-5" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb429-6"><a href="network-comparison.html#cb429-6" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">&quot;eigenvector&quot;</span>,</span>
<span id="cb429-7"><a href="network-comparison.html#cb429-7" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">2</span>,</span>
<span id="cb429-8"><a href="network-comparison.html#cb429-8" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">&quot;cluster&quot;</span>, </span>
<span id="cb429-9"><a href="network-comparison.html#cb429-9" tabindex="-1"></a>     <span class="at">sameColThresh =</span> <span class="dv">2</span>,</span>
<span id="cb429-10"><a href="network-comparison.html#cb429-10" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">&quot;darkgray&quot;</span>,</span>
<span id="cb429-11"><a href="network-comparison.html#cb429-11" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="dv">2</span>,</span>
<span id="cb429-12"><a href="network-comparison.html#cb429-12" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb429-13"><a href="network-comparison.html#cb429-13" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">&quot;Mixed diet&quot;</span>, </span>
<span id="cb429-14"><a href="network-comparison.html#cb429-14" tabindex="-1"></a>     <span class="at">title2 =</span> <span class="st">&quot;Vegetarian&quot;</span>,</span>
<span id="cb429-15"><a href="network-comparison.html#cb429-15" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb429-16"><a href="network-comparison.html#cb429-16" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="dv">2</span>,</span>
<span id="cb429-17"><a href="network-comparison.html#cb429-17" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb429-18"><a href="network-comparison.html#cb429-18" tabindex="-1"></a></span>
<span id="cb429-19"><a href="network-comparison.html#cb429-19" tabindex="-1"></a><span class="co"># Add legend</span></span>
<span id="cb429-20"><a href="network-comparison.html#cb429-20" tabindex="-1"></a><span class="fu">par</span>(<span class="at">fig=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="at">oma=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">new=</span><span class="cn">TRUE</span>)</span>
<span id="cb429-21"><a href="network-comparison.html#cb429-21" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="at">type=</span><span class="st">&#39;n&#39;</span>, <span class="at">bty=</span><span class="st">&#39;n&#39;</span>, <span class="at">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="at">yaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb429-22"><a href="network-comparison.html#cb429-22" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">0.2</span>, <span class="sc">-</span><span class="fl">0.8</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">&quot;estimated correlation:&quot;</span>,</span>
<span id="cb429-23"><a href="network-comparison.html#cb429-23" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;#009900&quot;</span>,<span class="st">&quot;red&quot;</span>), </span>
<span id="cb429-24"><a href="network-comparison.html#cb429-24" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="61_network_comparison_files/figure-html/network_plot_diet_samelay-1.png" width="1632" /></p>
<p><strong>A few notes:</strong></p>
<ul>
<li>Differences in the edge weights can now be seen at first glance. For example, Serratia and Citrobacter are strongly associated in the mixed diet group, but not at all in the vegetarian group.</li>
<li>Clusters must share at least two nodes (<code>sameColThresh</code> argument) to be colored equally in both networks, which is why the color of some clusters differs between the groups.</li>
<li>The clustering generally differs markedly. In particular, the cluster assignment of many of the nodes in the largest connected component differs between the two groups.</li>
</ul>
<p>As in Chapter <a href="network-learning.html#network-learning">12</a>, we also generate a network plot using phylum names to color the nodes and mclr-transformed abundances to scale node sizes.</p>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb430-1"><a href="network-comparison.html#cb430-1" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span></code></pre></div>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb431-1"><a href="network-comparison.html#cb431-1" tabindex="-1"></a><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span id="cb431-2"><a href="network-comparison.html#cb431-2" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>phylum)</span>
<span id="cb431-3"><a href="network-comparison.html#cb431-3" tabindex="-1"></a><span class="fu">names</span>(phyla) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(tse)<span class="sc">$</span>genus</span>
<span id="cb431-4"><a href="network-comparison.html#cb431-4" tabindex="-1"></a></span>
<span id="cb431-5"><a href="network-comparison.html#cb431-5" tabindex="-1"></a><span class="co"># Create color vector</span></span>
<span id="cb431-6"><a href="network-comparison.html#cb431-6" tabindex="-1"></a>colvec <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="fu">length</span>(<span class="fu">levels</span>(phyla)), <span class="st">&quot;Set3&quot;</span>)</span>
<span id="cb431-7"><a href="network-comparison.html#cb431-7" tabindex="-1"></a></span>
<span id="cb431-8"><a href="network-comparison.html#cb431-8" tabindex="-1"></a>p_diet <span class="ot">&lt;-</span> <span class="fu">plot</span>(spring_netprops_diet,</span>
<span id="cb431-9"><a href="network-comparison.html#cb431-9" tabindex="-1"></a>               <span class="at">sameLayout =</span> <span class="cn">TRUE</span>,</span>
<span id="cb431-10"><a href="network-comparison.html#cb431-10" tabindex="-1"></a>               <span class="at">repulsion =</span> <span class="fl">0.95</span>,</span>
<span id="cb431-11"><a href="network-comparison.html#cb431-11" tabindex="-1"></a>               <span class="at">rmSingles =</span> <span class="st">&quot;inboth&quot;</span>,</span>
<span id="cb431-12"><a href="network-comparison.html#cb431-12" tabindex="-1"></a>               <span class="at">labelScale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb431-13"><a href="network-comparison.html#cb431-13" tabindex="-1"></a>               <span class="at">nodeSize =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb431-14"><a href="network-comparison.html#cb431-14" tabindex="-1"></a>               <span class="at">nodeColor =</span> <span class="st">&quot;feature&quot;</span>, </span>
<span id="cb431-15"><a href="network-comparison.html#cb431-15" tabindex="-1"></a>               <span class="at">featVecCol =</span> phyla, </span>
<span id="cb431-16"><a href="network-comparison.html#cb431-16" tabindex="-1"></a>               <span class="at">colorVec =</span>  colvec,</span>
<span id="cb431-17"><a href="network-comparison.html#cb431-17" tabindex="-1"></a>               <span class="at">nodeTransp =</span> <span class="dv">20</span>,</span>
<span id="cb431-18"><a href="network-comparison.html#cb431-18" tabindex="-1"></a>               <span class="at">sameColThresh =</span> <span class="dv">2</span>,</span>
<span id="cb431-19"><a href="network-comparison.html#cb431-19" tabindex="-1"></a>               <span class="at">highlightHubs =</span> <span class="cn">FALSE</span>,</span>
<span id="cb431-20"><a href="network-comparison.html#cb431-20" tabindex="-1"></a>               <span class="at">cexNodes =</span> <span class="dv">2</span>,</span>
<span id="cb431-21"><a href="network-comparison.html#cb431-21" tabindex="-1"></a>               <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb431-22"><a href="network-comparison.html#cb431-22" tabindex="-1"></a>               <span class="at">title1 =</span> <span class="st">&quot;Mixed diet&quot;</span>, </span>
<span id="cb431-23"><a href="network-comparison.html#cb431-23" tabindex="-1"></a>               <span class="at">title2 =</span> <span class="st">&quot;Vegetarian&quot;</span>,</span>
<span id="cb431-24"><a href="network-comparison.html#cb431-24" tabindex="-1"></a>               <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb431-25"><a href="network-comparison.html#cb431-25" tabindex="-1"></a>               <span class="at">cexTitle =</span> <span class="dv">2</span>,</span>
<span id="cb431-26"><a href="network-comparison.html#cb431-26" tabindex="-1"></a>               <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">4</span>))</span>
<span id="cb431-27"><a href="network-comparison.html#cb431-27" tabindex="-1"></a></span>
<span id="cb431-28"><a href="network-comparison.html#cb431-28" tabindex="-1"></a><span class="co"># Add legends</span></span>
<span id="cb431-29"><a href="network-comparison.html#cb431-29" tabindex="-1"></a><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span id="cb431-30"><a href="network-comparison.html#cb431-30" tabindex="-1"></a>col_transp <span class="ot">&lt;-</span> <span class="fu">colToTransp</span>(colvec, <span class="dv">20</span>)</span>
<span id="cb431-31"><a href="network-comparison.html#cb431-31" tabindex="-1"></a></span>
<span id="cb431-32"><a href="network-comparison.html#cb431-32" tabindex="-1"></a><span class="fu">par</span>(<span class="at">fig=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="at">oma=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">new=</span><span class="cn">TRUE</span>)</span>
<span id="cb431-33"><a href="network-comparison.html#cb431-33" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="at">type=</span><span class="st">&#39;n&#39;</span>, <span class="at">bty=</span><span class="st">&#39;n&#39;</span>, <span class="at">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="at">yaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb431-34"><a href="network-comparison.html#cb431-34" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">0.15</span>, <span class="sc">-</span><span class="fl">0.8</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">&quot;estimated correlation:&quot;</span>,</span>
<span id="cb431-35"><a href="network-comparison.html#cb431-35" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;#009900&quot;</span>,<span class="st">&quot;red&quot;</span>), </span>
<span id="cb431-36"><a href="network-comparison.html#cb431-36" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span>
<span id="cb431-37"><a href="network-comparison.html#cb431-37" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">0.15</span>, <span class="fl">1.3</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">pt.cex =</span> <span class="fl">2.5</span>, <span class="at">title =</span> <span class="st">&quot;Phylum:&quot;</span>, </span>
<span id="cb431-38"><a href="network-comparison.html#cb431-38" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">levels</span>(phyla), <span class="at">col =</span> col_transp, <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">pch =</span> <span class="dv">16</span>) </span></code></pre></div>
<p><img src="61_network_comparison_files/figure-html/network_plot_diet_phyla-1.png" width="1632" /></p>
</div>
<div id="quantitative-comparison" class="section level3 hasAnchor" number="13.3.2">
<h3><span class="header-section-number">13.3.2</span> Quantitative comparison<a href="network-comparison.html#quantitative-comparison" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>netCompare()</code> enables a quantitative network comparison using comparative measures such as Jaccards Index, Adjusted Rand Index, and permutation tests.</p>
<p>To test for statistical significance of differences in network properties, we perform permutation tests with 1000 permutations. Multiple CPU cores are used to save run time. The association matrices estimated for all permutations are stored in an external file. We will reuse them later when performing differential association analysis. They could also be used to rerun <code>netCompare()</code> with different parameter settings.</p>
<p>Note that unless running on a cluster with considerably more CPU cores, a network comparison with permutation tests may take several hours. You should test the code below with a small number of permutations to make sure it works before applying it to your data.</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="network-comparison.html#cb432-1" tabindex="-1"></a>spring_netcomp_diet <span class="ot">&lt;-</span> <span class="fu">netCompare</span>(spring_netprops_diet, </span>
<span id="cb432-2"><a href="network-comparison.html#cb432-2" tabindex="-1"></a>                                  <span class="at">permTest =</span> <span class="cn">TRUE</span>,</span>
<span id="cb432-3"><a href="network-comparison.html#cb432-3" tabindex="-1"></a>                                  <span class="at">nPerm =</span> <span class="dv">1000</span>,</span>
<span id="cb432-4"><a href="network-comparison.html#cb432-4" tabindex="-1"></a>                                  <span class="at">cores =</span> <span class="dv">6</span>,</span>
<span id="cb432-5"><a href="network-comparison.html#cb432-5" tabindex="-1"></a>                                  <span class="at">seed =</span> <span class="dv">13075</span>,</span>
<span id="cb432-6"><a href="network-comparison.html#cb432-6" tabindex="-1"></a>                                  <span class="at">storeAssoPerm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb432-7"><a href="network-comparison.html#cb432-7" tabindex="-1"></a>                                  <span class="at">fileStoreAssoPerm =</span> <span class="st">&quot;general/network_data/spring_assoPerm&quot;</span>,</span>
<span id="cb432-8"><a href="network-comparison.html#cb432-8" tabindex="-1"></a>                                  <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb433"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb433-1"><a href="network-comparison.html#cb433-1" tabindex="-1"></a><span class="fu">summary</span>(spring_netcomp_diet, </span>
<span id="cb433-2"><a href="network-comparison.html#cb433-2" tabindex="-1"></a>        <span class="at">groupNames =</span> <span class="fu">c</span>(<span class="st">&quot;Mixed diet&quot;</span>, <span class="st">&quot;Vegetarian&quot;</span>),</span>
<span id="cb433-3"><a href="network-comparison.html#cb433-3" tabindex="-1"></a>        <span class="at">numbNodes =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## 
## Comparison of Network Properties
## ----------------------------------
## CALL: 
## netCompare(x = spring_netprops_diet, permTest = TRUE, verbose = TRUE, 
##     nPerm = 1000, cores = 19, libPathsClust = &quot;/dss/dsshome1/07/di93fen/R&quot;, 
##     seed = 13075, storeAssoPerm = TRUE, fileStoreAssoPerm = &quot;general/network_data/spring_assoPerm&quot;)
## 
## ______________________________
## Global network properties
## `````````````````````````
## Largest connected component (LCC):
##                          Mixed diet   Vegetarian    abs.diff.     p-value    
## Relative LCC size             0.220        0.256        0.037    0.838162    
## Clustering coefficient        0.104        0.224        0.120    0.450549    
## Modularity                    0.515        0.477        0.038    0.801199    
## Positive edge percentage     80.000      100.000       20.000    0.000999 ***
## Edge density                  0.131        0.124        0.007    0.910090    
## Natural connectivity          0.073        0.063        0.011    0.812188    
## Vertex connectivity           1.000        1.000        0.000    1.000000    
## Edge connectivity             1.000        1.000        0.000    1.000000    
## Average dissimilarity*        0.960        0.962        0.001    0.945055    
## Average path length**         2.228        2.246        0.019    0.984016    
## 
## Whole network:
##                          Mixed diet   Vegetarian    abs.diff.     p-value    
## Number of components         50.000       52.000        2.000     0.93706    
## Clustering coefficient        0.075        0.165        0.090     0.46254    
## Modularity                    0.752        0.668        0.084     0.24975    
## Positive edge percentage     88.571      100.000       11.429     0.02398 *  
## Edge density                  0.011        0.011        0.000     0.97303    
## Natural connectivity          0.013        0.013        0.000     0.89011    
## -----
## p-values: one-tailed test with null hypothesis diff=0
##  *: Dissimilarity = 1 - edge weight
## **: Path length = Units with average dissimilarity
## 
## ______________________________
## Jaccard index (similarity betw. sets of most central nodes)
## ```````````````````````````````````````````````````````````
##                     Jacc   P(&lt;=Jacc)     P(&gt;=Jacc)    
## degree             0.560      0.9944       0.01637 *  
## betweenness centr. 0.294      0.4777       0.71860    
## closeness centr.   0.560      0.9944       0.01637 *  
## eigenvec. centr.   0.560      0.9944       0.01637 *  
## hub taxa           0.111      0.1431       0.97399    
## -----
## Jaccard index in [0,1] (1 indicates perfect agreement)
## 
## ______________________________
## Adjusted Rand index (similarity betw. clusterings)
## ``````````````````````````````````````````````````
##         wholeNet       LCC
## ARI        0.367     0.035
## p-value    0.000     0.563
## -----
## ARI in [-1,1] with ARI=1: perfect agreement betw. clusterings
##                    ARI=0: expected for two random clusterings
## p-value: permutation test (n=1000) with null hypothesis ARI=0
## 
## ______________________________
## Graphlet Correlation Distance
## `````````````````````````````
##         wholeNet         LCC  
## GCD       1.5980      2.1770  
## p-value   0.4226      0.7143  
## -----
## GCD &gt;= 0 (GCD=0 indicates perfect agreement between GCMs)
## p-value: permutation test with null hypothesis GCD=0
## 
## ______________________________
## Centrality measures
## - In decreasing order
## - Centrality of disconnected components is zero
## ````````````````````````````````````````````````
## Degree (unnormalized):
##                      Mixed diet Vegetarian abs.diff. adj.p-value  
## Siccibacter                   0          5         5           1  
## Aeromonas                     1          5         4           1  
## Citrobacter                   4          0         4           1  
## Escherichia/Shigella          2          5         3           1  
## Streptococcus                 3          0         3           1  
## 
## Betweenness centrality (normalized):
##                      Mixed diet Vegetarian abs.diff. adj.p-value  
## Escherichia/Shigella      0.118      0.532     0.414           1  
## Aeromonas                 0.000      0.368     0.368           1  
## Serratia                  0.419      0.100     0.319           1  
## Siccibacter               0.000      0.300     0.300           1  
## Streptococcus             0.279      0.000     0.279           1  
## 
## Closeness centrality (normalized):
##               Mixed diet Vegetarian abs.diff. adj.p-value  
## Siccibacter        0.000      0.727     0.727      0.2676  
## Citrobacter        0.713      0.000     0.713      0.9799  
## Streptococcus      0.615      0.000     0.615      0.2676  
## Escherichia        0.587      0.000     0.587      0.9799  
## Xenorhabdus        0.000      0.527     0.527      0.2676  
## 
## Eigenvector centrality (normalized):
##                      Mixed diet Vegetarian abs.diff. adj.p-value  
## Siccibacter               0.000      0.888     0.888      0.6553  
## Citrobacter               0.810      0.000     0.810      1.0000  
## Escherichia               0.610      0.000     0.610      1.0000  
## Serratia                  1.000      0.416     0.584      1.0000  
## Escherichia/Shigella      0.387      0.969     0.582      1.0000  
## 
## _________________________________________________________
## Significance codes: ***: 0.001, **: 0.01, *: 0.05, .: 0.1</code></pre>
<p><strong>Interpreting some results:</strong></p>
<ul>
<li>Almost all global network properties are significantly different between the groups (for <span class="math inline">\(\alpha=0.1\)</span>), thus reflecting the different overall network structure we already have seen in the network plots.</li>
<li>For the Jaccard index of degree, closeness, and eigenvector centrality, the probability P(&gt;=Jacc) is significant, meaning that the sets of the most central nodes are quite similar for these three measures. The Jaccard index for the hub nodes, on the other hand, is low because the two networks share only one hub node (Erwinia).</li>
<li>As indicated by some similarities in the clusterings, the adjusted Rand index (ARI) of the whole network is significantly different from zero and thus from random clustering. The ARI of the largest connected component (LCC), however, is close to zero due to the different clusterings in the LCC.</li>
<li>The two GCD values are significantly different from zero, indicating substantial differences in the overall network structures.</li>
<li>All nodes are also tested for having significantly different centrality (only the five nodes with the highest absolute difference are shown in the summary). For <span class="math inline">\(\alpha=0.05\)</span>, some nodes have a significantly different closeness centrality, and for <span class="math inline">\(\alpha=0.1\)</span> also a significantly different eigenvector centrality. Most of these nodes have a high centrality in the one group, but are not connected in the other group.</li>
</ul>
</div>
</div>
<div id="differential-association-analysis" class="section level2 hasAnchor" number="13.4">
<h2><span class="header-section-number">13.4</span> Differential association analysis<a href="network-comparison.html#differential-association-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The <code>diffnet()</code> function provides statistical tests to assess whether the associations themselves are significantly different between the two groups. <code>NetCoMi</code> also provides a plot function to generate a differential network, where two nodes are connected if they are differentially associated between the groups.</p>
<p>Since we have already computed the permutation association matrices before, we can reuse them here (argument <code>fileLoadAssoPerm</code>).</p>
<p>The local false discovery rate is controlled at level 0.2 to account for multiplicity.</p>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb435-1"><a href="network-comparison.html#cb435-1" tabindex="-1"></a>spring_diffnet <span class="ot">&lt;-</span> <span class="fu">diffnet</span>(spring_net_diet,</span>
<span id="cb435-2"><a href="network-comparison.html#cb435-2" tabindex="-1"></a>                          <span class="at">diffMethod =</span> <span class="st">&quot;perm&quot;</span>,</span>
<span id="cb435-3"><a href="network-comparison.html#cb435-3" tabindex="-1"></a>                          <span class="at">fileLoadAssoPerm =</span> <span class="st">&quot;general/network_data/spring_assoPerm&quot;</span>,</span>
<span id="cb435-4"><a href="network-comparison.html#cb435-4" tabindex="-1"></a>                          <span class="at">adjust =</span> <span class="st">&quot;lfdr&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb436-1"><a href="network-comparison.html#cb436-1" tabindex="-1"></a><span class="fu">sum</span>(spring_diffnet<span class="sc">$</span>pAdjustVec <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode" id="cb438"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb438-1"><a href="network-comparison.html#cb438-1" tabindex="-1"></a><span class="fu">sum</span>(spring_diffnet<span class="sc">$</span>pvalsVec <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## [1] 12</code></pre>
<p>Some of the unadjusted p-values are below the usual 5% significance level. However, none of the differences remain significant after adjusting for multiple testing so that the differential network would be empty.</p>
<p>To demonstrate the interpretation of a differential network, we set <code>adjust</code> to none, which is actually statistically incorrect.</p>
<div class="sourceCode" id="cb440"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb440-1"><a href="network-comparison.html#cb440-1" tabindex="-1"></a>spring_diffnet_unadj <span class="ot">&lt;-</span> <span class="fu">diffnet</span>(spring_net_diet,</span>
<span id="cb440-2"><a href="network-comparison.html#cb440-2" tabindex="-1"></a>                                <span class="at">pvalsVec =</span> spring_diffnet<span class="sc">$</span>pvalsVec,</span>
<span id="cb440-3"><a href="network-comparison.html#cb440-3" tabindex="-1"></a>                                <span class="at">diffMethod =</span> <span class="st">&quot;perm&quot;</span>,</span>
<span id="cb440-4"><a href="network-comparison.html#cb440-4" tabindex="-1"></a>                                <span class="at">alpha =</span> <span class="fl">0.05</span>,</span>
<span id="cb440-5"><a href="network-comparison.html#cb440-5" tabindex="-1"></a>                                <span class="at">adjust =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p>The <code>diffnet</code> object it now plotted using NetCoMis plot function.</p>
<div class="sourceCode" id="cb441"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb441-1"><a href="network-comparison.html#cb441-1" tabindex="-1"></a><span class="fu">plot</span>(spring_diffnet_unadj, </span>
<span id="cb441-2"><a href="network-comparison.html#cb441-2" tabindex="-1"></a>     <span class="at">cexLabels =</span> <span class="dv">2</span>,</span>
<span id="cb441-3"><a href="network-comparison.html#cb441-3" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">0.7</span>, </span>
<span id="cb441-4"><a href="network-comparison.html#cb441-4" tabindex="-1"></a>     <span class="at">cexLegend =</span> <span class="fl">2.5</span>,</span>
<span id="cb441-5"><a href="network-comparison.html#cb441-5" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="dv">3</span>,</span>
<span id="cb441-6"><a href="network-comparison.html#cb441-6" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">15</span>),</span>
<span id="cb441-7"><a href="network-comparison.html#cb441-7" tabindex="-1"></a>     <span class="at">legendGroupnames =</span> <span class="fu">c</span>(<span class="st">&quot;Mixed diet&quot;</span>, <span class="st">&quot;Vegetarian&quot;</span>),</span>
<span id="cb441-8"><a href="network-comparison.html#cb441-8" tabindex="-1"></a>     <span class="at">legendPos =</span> <span class="fu">c</span>(<span class="fl">1.2</span>,<span class="fl">1.5</span>),</span>
<span id="cb441-9"><a href="network-comparison.html#cb441-9" tabindex="-1"></a>     <span class="at">legendArgs =</span> <span class="fu">list</span>(<span class="at">lwd =</span> <span class="dv">4</span>),</span>
<span id="cb441-10"><a href="network-comparison.html#cb441-10" tabindex="-1"></a>     <span class="at">fade =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="61_network_comparison_files/figure-html/diffnet_plot-1.png" width="1920" /></p>
<p>Edge colors represent the direction of the associations in the two groups. For example, if two OTUs are positively correlated in the mixed diet group and uncorrelated in the vegetarian group (such as Serratia and Citrobacter), the edge color is dark green.</p>
</div>
<div id="netcomp-methods" class="section level2 hasAnchor" number="13.5">
<h2><span class="header-section-number">13.5</span> Network comparison methods<a href="network-comparison.html#netcomp-methods" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>While many approaches exist for the detection of <strong>differential correlations</strong>, e.g. <span class="citation">(Yu et al. 2019; McKenzie et al. 2016; Siska and Kechris 2017)</span>, the literature on the more general case of <strong>differential association</strong> detection is scarce. <span class="citation">Bhuva et al. (2019)</span> compare various methods in a simulation study, which again includes many differential correlation approaches, but also more general methods such as latent differential graphical models. <span class="citation">Gill, Datta, and Datta (2010)</span> introduce an approach to analyze whether the connectivity of individual nodes is different between two groups using permutation tests, which is applicable to any kind of association. <span class="citation">He et al. (2019)</span> propose a test to infer the differential network structure for two conditional dependence networks.</p>
<p>Performing <strong>differential network analysis</strong> is challenging because network measures do not follow classical statistical distributions. <span class="citation">Shojaie (2021)</span> provide an overview of differential network analysis methods, but focus only on changes in edge sets. <span class="citation">Lichtblau et al. (2017)</span> compare differential network analysis methods that incorporate multiple local and global network measures. <span class="citation">Jardim et al. (2019)</span> present a tool BioNetStat for differential analysis of biological networks, which is able to compare certain network measures between groups.</p>
<p>The <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> package used for network comparison in this chapter includes the following <strong>differential network analysis approaches:</strong>:</p>
<ul>
<li><strong>Permutation approach</strong> to test global network measures (e.g., transitivity, connectivity, or average path length) as well as centrality measures for group differences.</li>
<li><strong>Jaccard</strong> index to assess the similarity between sets of most central nodes</li>
<li><strong>Adjusted Rand index</strong> to assess the similarity between clusterings</li>
<li><strong>Graphlet Correlation Distance</strong> (GCM)</li>
</ul>
<p>See <span class="citation">(Peschel et al. 2021)</span> for an explanation of the first three approaches. The GCM was proposed by <span class="citation">Yaverolu et al. (2014)</span>.</p>
<p>Two methods (Fishers z-test <span class="citation">(Fisher 1970)</span> and the Discordant method <span class="citation">(Siska, Bowler, and Kechris 2016)</span>) are available for identifying differential correlations, and <strong>permutation tests</strong> for the more general case of identifying <strong>differential associations</strong>. See <span class="citation">(Peschel et al. 2021)</span> for details. NetCoMi offers also a function for plotting a differential network.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="network-learning.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="machine_learning.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/61_network_comparison.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/61_network_comparison.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/61_network_comparison.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
