<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 12 Network learning and analysis | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 12 Network learning and analysis | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 12 Network learning and analysis | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 12 Network learning and analysis | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differential-abundance.html"/>
<link rel="next" href="network-comparison.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.4</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-igraph"><i class="fa fa-check"></i><b>12.2.1</b> Network analysis with igraph</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.2.2</b> Network analysis with NetCoMi</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.3</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.3.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.3.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.3.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.4</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.5</b> More about association measures</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.6</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.6.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.6.1</b> SparCC</a></li>
<li class="chapter" data-level="12.6.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.6.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.6.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.6.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.6.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.6.4</b> Network plots</a></li>
<li class="chapter" data-level="12.6.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.6.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="network-learning" class="section level1 hasAnchor" number="12">
<h1><span class="header-section-number">Chapter 12</span> Network learning and analysis<a href="network-learning.html#network-learning" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Learning and analyzing microbial association networks is another common exploratory data analysis approach aimed at understanding the complex interplay of microbial communities in their natural habitat. Microbial networks consist of nodes, representing microbial species or taxa, and edges, expressing their association. The mathematical representation of a network is the adjacency matrix, which has a non-zero entry for each edge in the network.</p>
<p>A typical workflow for estimating a microbial network involves several steps, including data preprocessing, estimating microbial associations, and transforming them into edge weights. The resulting network can be analyzed using local network properties, such as centrality measures, or global measures that describe the overall structure of the network. Network plots provide further insight into the microbial community structure and allow for exploratory analysis of microbial relationships.</p>
<p>In this chapter, we go through the complete workflow of constructing and analyzing a single microbial network, step by step. A potential next analysis task is to compare networks between groups, such as patients and controls, different environmental conditions, or different time points. How to compare networks between two groups is explained in chapter <a href="network-comparison.html#network-comparison">13</a></p>
<div id="network-learning-1" class="section level2 hasAnchor" number="12.1">
<h2><span class="header-section-number">12.1</span> Network learning<a href="network-learning.html#network-learning-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="network-learning-workflow" class="section level3 hasAnchor" number="12.1.1">
<h3><span class="header-section-number">12.1.1</span> Typical workflow<a href="network-learning.html#network-learning-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Figure <a href="network-learning.html#fig:network-workflow">12.1</a> shows the workflow for learning/constructing a microbial association network as proposed by <span class="citation">Peschel et al. (2021)</span>. The respective steps are explained below.</p>
<div class="figure"><span style="display:block;" id="fig:network-workflow"></span>
<img src="general/figures/FigureNetworkLearning.png" alt="The typical input is a $p$ x $n$ dimensional count matrix coming from a sequencing process, where $n$ is the number of samples and $p$ the number of features / ASVs / OTUs. Steps 1 through 6 are explained below. Each matrix resulting from steps 4, 5, and 6 plays a specific role in the final network: The adjacency matrix is used for edge colors, dissimilarity for layout, and similarity for edge weights. In weighted networks, the similarity matrix equals the adjacency matrix." width="100%" />
<p class="caption">
Figure 12.1: The typical input is a <span class="math inline">\(p\)</span> x <span class="math inline">\(n\)</span> dimensional count matrix coming from a sequencing process, where <span class="math inline">\(n\)</span> is the number of samples and <span class="math inline">\(p\)</span> the number of features / ASVs / OTUs. Steps 1 through 6 are explained below. Each matrix resulting from steps 4, 5, and 6 plays a specific role in the final network: The adjacency matrix is used for edge colors, dissimilarity for layout, and similarity for edge weights. In weighted networks, the similarity matrix equals the adjacency matrix.
</p>
</div>
<ol style="list-style-type: decimal">
<li><p><strong>Zero replacement:</strong> Since the following steps usually require non-zero entries in the read count matrix, zero counts must be replaced. A simple solution is to add a pseudo count to the data. Other possible approaches are implemented in the R package <a href="https://cran.r-project.org/web/packages/zCompositions/index.html">zCompositions</a></p></li>
<li><p><strong>Normalization:</strong> To avoid compositional effects, the data are normalized using a compositionality aware transformation. A common approach is the centered log-ratio (clr) transformation, which moves the data from a <span class="math inline">\(p\)</span>-dimensional simplex to Euclidean space so that standard statistical analysis methods are valid. A variance stabilizing transformation (vst) is also a suitable approach for normalizing microbial count data <span class="citation">(Badri et al. 2020)</span>.</p></li>
<li><p><strong>Association estimation:</strong> This is the crucial step in network learning to obtain statistical relations between the taxa. Common association measures include correlation, conditional dependence (which we will equate to partial correlation), and proportionality. Further information on these three types of association and their application can be found in Section <a href="network-learning.html#more-about-association">12.5</a>. The following list gives a selection of compositionality aware approaches:</p>
<ul>
<li><strong>Compositionality aware correlation estimation methods:</strong>
<ul>
<li>Pearsons correlation coefficient (+ normalization)</li>
<li>Spearmans rank correlation coefficient (+ normalization)</li>
<li>Covariance shrinkage (<a href="https://strimmerlab.github.io/software/corpcor/">corpcor</a> package) (+ normalization)</li>
<li>SparCC (implemented in <a href="https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html">SpiecEasi</a>); applied in Section <a href="network-learning.html#sparcc-correlation">12.6.1</a></li>
<li>CCREPE (<a href="https://bioconductor.org/packages/release/bioc/html/ccrepe.html">ccrepe</a> package)</li>
<li>CCLasso (<a href="https://github.com/huayingfang/CCLasso">R code on GitHub</a>)</li>
</ul></li>
<li><strong>Compositionality aware measures of conditional dependence / partial correlation:</strong>
<ul>
<li><a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a> with Meinshausen and Bhlmann (MB) neighborhood selection; applied in Section <a href="network-learning.html#spieceasi-mb">12.6.3</a></li>
<li><a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a> with the graphical lasso (glasso)</li>
<li>gCoda (<a href="https://github.com/huayingfang/gCoda">R code on GitHub</a>)</li>
<li><a href="https://github.com/GraceYoon/SPRING">SPRING</a>; applied in Section <a href="network-learning.html#netcomi-spring">12.1.5</a></li>
</ul></li>
<li><strong>Proportionality measures (proportionality aware by definition):</strong>
<ul>
<li><a href="https://github.com/tpq/propr">propr</a></li>
<li><a href="https://github.com/MichelleBadri/NormCorr-manuscript/blob/master/code/helpers/norm_functions.R">Shrinkage proportionality estimator</a>; applied in Section <a href="network-learning.html#shrinkage-prop">12.6.2</a></li>
</ul></li>
</ul></li>
<li><p><strong>Sparsification:</strong> Transforming the estimated associations directly into adjacencies would lead to a dense network where all nodes are connected and only weighted network measures are meaningful. Therefore, the association matrix is usually sparsified to select edges of interest. A common sparsification approach for correlations is thresholding, where correlations with a magnitude below the threshold are set to zero. Another possibility is a statistical test (Students t-test or permutation test) with the null hypothesis that the correlation is equal to zero. SpiecEasi uses the StARS stability selection approach <span class="citation">(H. Liu, Roeder, and Wasserman 2010)</span> to decide on an appropriate sparsification level of the inferred conditional dependence graph.</p></li>
<li><p><strong>Transformation into dissimilarity:</strong> A common next step is to simply use the absolute values of the sparsified associations as edge weights. In this way, correlations of high magnitude (both positive and negative) will have a high edge weight. From a biological point of view, it would also make sense to assign a low edge weight to taxa that are strongly negatively associated, which would correspond to a high dissimilarity value. Here we follow <span class="citation">Dongen and Enright (2012)</span> to directly transform the sparse associations <span class="math inline">\(r_{ij}^*\)</span> into dissimilarities, which can later be used for shortest path network measures. Depending on the desired handling of negative associations, one of the two proposed transformations should be chosen:</p>
<p>5a: <strong>signed:</strong> <span class="math inline">\(d_{ij} = \sqrt{0.5(1-r^*_{ij})}\)</span>, where strongly negatively associated taxa have the largest distance and are placed further apart in the network.</p>
<p>5b: <strong>unsigned:</strong> <span class="math inline">\(d_{ij} = \sqrt{1-{r_{ij}^*}^2}\)</span>, resulting in a small distance between strongly associated taxa (regardless of the sign).</p></li>
<li><p><strong>Transformation into similarity / edge weight:</strong> Finally, the dissimilarities are transformed into similarities by <span class="math inline">\(s_{ij} = 1 - d_{ij}\)</span>, which are used as edge weights. Thus, the similarity matrix is equal to the adjacency matrix in a weighted network.</p></li>
</ol>
<p>The main association measure used in this chapter is the <a href="https://github.com/GraceYoon/SPRING">SPRING</a> (Semi-Parametric Rank-based approach for INference in Graphical model) method proposed by <span class="citation">Yoon, Gaynanova, and Mller (2019)</span>. <code>SPRING</code> learns conditional dependency graphs for compositional data and follows the neighborhood selection method introduced by <span class="citation">Meinshausen and Bhlmann (2006)</span> (MB). We will show how to apply the method directly, as well as how to use it in conjunction with the R package <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a>, which is specifically designed for the construction and analysis of networks for microbiome data.</p>
<p>See Section <a href="network-learning.html#more-about-association">12.5</a> for a comparison of all three association types (correlation, partial correlation, and proportionality) with more information on each measure and applications.</p>
<p>We demonstrate the workflow using the the <strong>PeerJ data set</strong> <span class="citation">(Potbhare et al. 2022)</span>. It contains skin microbial profiles of 58 subjects.</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="network-learning.html#cb354-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span></code></pre></div>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="network-learning.html#cb355-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;peerj13075&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb355-2"><a href="network-learning.html#cb355-2" tabindex="-1"></a>tse0 <span class="ot">&lt;-</span> peerj13075</span>
<span id="cb355-3"><a href="network-learning.html#cb355-3" tabindex="-1"></a><span class="fu">dim</span>(tse0)</span></code></pre></div>
<pre><code>## [1] 674  58</code></pre>
</div>
<div id="install-packages" class="section level3 hasAnchor" number="12.1.2">
<h3><span class="header-section-number">12.1.2</span> Install packages<a href="network-learning.html#install-packages" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Three packages used in this chapter are available on GitHub only: <code>SpiecEasi</code>, <code>SPRING</code>, and <code>NetCoMi</code>. We recommend that you install these packages before proceeding.</p>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb357-1"><a href="network-learning.html#cb357-1" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(SpiecEasi)){</span>
<span id="cb357-2"><a href="network-learning.html#cb357-2" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;zdk123/SpiecEasi&quot;</span>)</span>
<span id="cb357-3"><a href="network-learning.html#cb357-3" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## pulsar (NA -&gt; 0.3.11) [CRAN]
## huge   (NA -&gt; 1.3.5 ) [CRAN]
##  R CMD build 
## * checking for file /tmp/RtmpHCeGz7/remotes383975225676/zdk123-SpiecEasi-5f396da/DESCRIPTION ... OK
## * preparing SpiecEasi:
## * checking DESCRIPTION meta-information ... OK
## * cleaning src
## * checking for LF line-endings in source and make files and shell scripts
## * checking for empty or unneeded directories
## Removed empty directory SpiecEasi/inst
## * looking to see if a data/datalist file should be added
## * building SpiecEasi_1.1.3.tar.gz</code></pre>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb359-1"><a href="network-learning.html#cb359-1" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(SPRING)){</span>
<span id="cb359-2"><a href="network-learning.html#cb359-2" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;GraceYoon/SPRING&quot;</span>)</span>
<span id="cb359-3"><a href="network-learning.html#cb359-3" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## mixedCCA     (NA -&gt; 4c2b63f75...) [GitHub]
## egg          (NA -&gt; 0.4.5       ) [CRAN]
## linprog      (NA -&gt; 0.9-4       ) [CRAN]
## lpSolve      (NA -&gt; 5.6.19      ) [CRAN]
## magic        (NA -&gt; 1.6-1       ) [CRAN]
## sn           (NA -&gt; 2.1.1       ) [CRAN]
## cubature     (NA -&gt; 2.1.0       ) [CRAN]
## microbenc... (NA -&gt; 1.4.10      ) [CRAN]
## doFuture     (NA -&gt; 1.0.0       ) [CRAN]
## geometry     (NA -&gt; 0.4.7       ) [CRAN]
## heatmaply    (NA -&gt; 1.5.0       ) [CRAN]
## mnormt       (NA -&gt; 2.1.1       ) [CRAN]
## fMultivar    (NA -&gt; 4031.84     ) [CRAN]
## latentcor    (NA -&gt; 2.0.1       ) [CRAN]
## linprog      (NA -&gt; 0.9-4  ) [CRAN]
## lpSolve      (NA -&gt; 5.6.19 ) [CRAN]
## magic        (NA -&gt; 1.6-1  ) [CRAN]
## egg          (NA -&gt; 0.4.5  ) [CRAN]
## mnormt       (NA -&gt; 2.1.1  ) [CRAN]
## sn           (NA -&gt; 2.1.1  ) [CRAN]
## cubature     (NA -&gt; 2.1.0  ) [CRAN]
## microbenc... (NA -&gt; 1.4.10 ) [CRAN]
## doFuture     (NA -&gt; 1.0.0  ) [CRAN]
## geometry     (NA -&gt; 0.4.7  ) [CRAN]
## heatmaply    (NA -&gt; 1.5.0  ) [CRAN]
## fMultivar    (NA -&gt; 4031.84) [CRAN]
## latentcor    (NA -&gt; 2.0.1  ) [CRAN]
##  R CMD build 
## * checking for file /tmp/RtmpHCeGz7/remotes383954e68e9/irinagain-mixedCCA-4c2b63f/DESCRIPTION ... OK
## * preparing mixedCCA:
## * checking DESCRIPTION meta-information ... OK
## * cleaning src
## * checking for LF line-endings in source and make files and shell scripts
## * checking for empty or unneeded directories
## Omitted LazyData from DESCRIPTION
## * building mixedCCA_1.6.2.tar.gz
## 
##  R CMD build 
## * checking for file /tmp/RtmpHCeGz7/remotes38391a5cd2b4/GraceYoon-SPRING-3d641a4/DESCRIPTION ... OK
## * preparing SPRING:
## * checking DESCRIPTION meta-information ... OK
## * checking for LF line-endings in source and make files and shell scripts
## * checking for empty or unneeded directories
## * building SPRING_1.0.4.tar.gz</code></pre>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="network-learning.html#cb361-1" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(NetCoMi)){</span>
<span id="cb361-2"><a href="network-learning.html#cb361-2" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;stefpeschel/NetCoMi&quot;</span>, <span class="at">force =</span> <span class="cn">TRUE</span>, <span class="at">ref =</span> <span class="st">&quot;TSE&quot;</span>,</span>
<span id="cb361-3"><a href="network-learning.html#cb361-3" tabindex="-1"></a>                           <span class="at">dependencies =</span> <span class="fu">c</span>(<span class="st">&quot;Depends&quot;</span>, <span class="st">&quot;Imports&quot;</span>, <span class="st">&quot;LinkingTo&quot;</span>),</span>
<span id="cb361-4"><a href="network-learning.html#cb361-4" tabindex="-1"></a>                           <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">&quot;https://cloud.r-project.org/&quot;</span>,</span>
<span id="cb361-5"><a href="network-learning.html#cb361-5" tabindex="-1"></a>                                     BiocManager<span class="sc">::</span><span class="fu">repositories</span>()))</span>
<span id="cb361-6"><a href="network-learning.html#cb361-6" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## Biobase      (2.60.0 -&gt; 8201fbbe5...) [Bioc]
## phyloseq     (1.44.0 -&gt; 73201334b...) [Bioc]
## stringi      (1.8.1  -&gt; 1.8.2       ) [CRAN]
## bslib        (0.6.0  -&gt; 0.6.1       ) [CRAN]
## scales       (1.2.1  -&gt; 1.3.0       ) [CRAN]
## pbivnorm     (NA     -&gt; 0.6.0       ) [CRAN]
## sp           (2.1-1  -&gt; 2.1-2       ) [CRAN]
## crosstalk    (1.2.0  -&gt; 1.2.1       ) [CRAN]
## seriation    (1.5.1  -&gt; 1.5.3       ) [CRAN]
## mvtnorm      (1.2-3  -&gt; 1.2-4       ) [CRAN]
## Cairo        (1.6-1  -&gt; 1.6-2       ) [CRAN]
## GO.db        (NA     -&gt; 3.17.0      ) [CRAN]
## preproces... (NA     -&gt; 1.62.1      ) [CRAN]
## impute       (NA     -&gt; 1.74.1      ) [CRAN]
## fastcluster  (NA     -&gt; 1.2.3       ) [CRAN]
## dynamicTr... (NA     -&gt; 1.63-1      ) [CRAN]
## fdrtool      (NA     -&gt; 1.2.17      ) [CRAN]
## glasso       (NA     -&gt; 1.11        ) [CRAN]
## corpcor      (NA     -&gt; 1.6.10      ) [CRAN]
## jpeg         (NA     -&gt; 0.1-10      ) [CRAN]
## lavaan       (NA     -&gt; 0.6-16      ) [CRAN]
## psych        (NA     -&gt; 2.3.9       ) [CRAN]
## WGCNA        (NA     -&gt; 1.72-1      ) [CRAN]
## qgraph       (NA     -&gt; 1.9.8       ) [CRAN]
## orca         (NA     -&gt; 1.1-1       ) [CRAN]
## filematrix   (NA     -&gt; 1.3         ) [CRAN]
## doSNOW       (NA     -&gt; 1.0.20      ) [CRAN]
## 
##  R CMD build 
## * checking for file /tmp/RtmpHCeGz7/file38392ce56ce0/DESCRIPTION ... OK
## * preparing Biobase:
## * checking DESCRIPTION meta-information ... OK
## * cleaning src
## * checking for LF line-endings in source and make files and shell scripts
## * checking for empty or unneeded directories
## * looking to see if a data/datalist file should be added
## * building Biobase_2.62.0.tar.gz
## 
## 
##  R CMD build 
## * checking for file /tmp/RtmpHCeGz7/file3839201fc447/DESCRIPTION ... OK
## * preparing phyloseq:
## * checking DESCRIPTION meta-information ... OK
## * checking for LF line-endings in source and make files and shell scripts
## * checking for empty or unneeded directories
## * looking to see if a data/datalist file should be added
## * building phyloseq_1.46.0.tar.gz
## 
##  R CMD build 
## * checking for file /tmp/RtmpHCeGz7/remotes38392e14953e/stefpeschel-NetCoMi-03abaf1/DESCRIPTION ... OK
## * preparing NetCoMi:
## * checking DESCRIPTION meta-information ... OK
## * installing the package to process help pages
## Loading required namespace: NetCoMi
## Registered S3 methods overwritten by &#39;treeio&#39;:
##   method              from    
##   MRCA.phylo          tidytree
##   MRCA.treedata       tidytree
##   Nnode.treedata      tidytree
##   Ntip.treedata       tidytree
##   ancestor.phylo      tidytree
##   ancestor.treedata   tidytree
##   child.phylo         tidytree
##   child.treedata      tidytree
##   full_join.phylo     tidytree
##   full_join.treedata  tidytree
##   groupClade.phylo    tidytree
##   groupClade.treedata tidytree
##   groupOTU.phylo      tidytree
##   groupOTU.treedata   tidytree
##   inner_join.phylo    tidytree
##   inner_join.treedata tidytree
##   is.rooted.treedata  tidytree
##   nodeid.phylo        tidytree
##   nodeid.treedata     tidytree
##   nodelab.phylo       tidytree
##   nodelab.treedata    tidytree
##   offspring.phylo     tidytree
##   offspring.treedata  tidytree
##   parent.phylo        tidytree
##   parent.treedata     tidytree
##   root.treedata       tidytree
##   rootnode.phylo      tidytree
##   sibling.phylo       tidytree
## 
## * saving partial Rd database
## * checking for LF line-endings in source and make files and shell scripts
## * checking for empty or unneeded directories
## * building NetCoMi_1.1.0.9000.tar.gz</code></pre>
</div>
<div id="data-preparation" class="section level3 hasAnchor" number="12.1.3">
<h3><span class="header-section-number">12.1.3</span> Data preparation<a href="network-learning.html#data-preparation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before applying the network learning methods, we perform some data preparation steps:</p>
<ul>
<li>Aggregation to genus level</li>
<li>Add relative abundance assay</li>
<li>Prevalence filtering (keep genera with prevalence &gt; 20%)</li>
<li>Add assay with log10 transformed abundances</li>
<li>Add assay with clr transformed abundances</li>
</ul>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="network-learning.html#cb363-1" tabindex="-1"></a><span class="co"># Agglomerate to genus level</span></span>
<span id="cb363-2"><a href="network-learning.html#cb363-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(tse0, <span class="at">rank =</span> <span class="st">&quot;genus&quot;</span>) </span>
<span id="cb363-3"><a href="network-learning.html#cb363-3" tabindex="-1"></a></span>
<span id="cb363-4"><a href="network-learning.html#cb363-4" tabindex="-1"></a><span class="co"># Add relative abundances</span></span>
<span id="cb363-5"><a href="network-learning.html#cb363-5" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, </span>
<span id="cb363-6"><a href="network-learning.html#cb363-6" tabindex="-1"></a>                      <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb363-7"><a href="network-learning.html#cb363-7" tabindex="-1"></a>                      <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb363-8"><a href="network-learning.html#cb363-8" tabindex="-1"></a>                      <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>) </span>
<span id="cb363-9"><a href="network-learning.html#cb363-9" tabindex="-1"></a></span>
<span id="cb363-10"><a href="network-learning.html#cb363-10" tabindex="-1"></a><span class="co"># Filter by prevalence</span></span>
<span id="cb363-11"><a href="network-learning.html#cb363-11" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">subsetByPrevalentFeatures</span>(tse,</span>
<span id="cb363-12"><a href="network-learning.html#cb363-12" tabindex="-1"></a>                                 <span class="at">prevalence =</span> <span class="fl">0.2</span>,</span>
<span id="cb363-13"><a href="network-learning.html#cb363-13" tabindex="-1"></a>                                 <span class="at">detection =</span> <span class="dv">0</span>,</span>
<span id="cb363-14"><a href="network-learning.html#cb363-14" tabindex="-1"></a>                                 <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb363-15"><a href="network-learning.html#cb363-15" tabindex="-1"></a></span>
<span id="cb363-16"><a href="network-learning.html#cb363-16" tabindex="-1"></a><span class="co"># Add log10-transformed abundances</span></span>
<span id="cb363-17"><a href="network-learning.html#cb363-17" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;log10&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb363-18"><a href="network-learning.html#cb363-18" tabindex="-1"></a></span>
<span id="cb363-19"><a href="network-learning.html#cb363-19" tabindex="-1"></a><span class="co"># Add clr-transformed abundances</span></span>
<span id="cb363-20"><a href="network-learning.html#cb363-20" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb363-21"><a href="network-learning.html#cb363-21" tabindex="-1"></a></span>
<span id="cb363-22"><a href="network-learning.html#cb363-22" tabindex="-1"></a><span class="fu">dim</span>(tse)</span></code></pre></div>
<pre><code>## [1] 147  58</code></pre>
</div>
<div id="spring-network" class="section level3 hasAnchor" number="12.1.4">
<h3><span class="header-section-number">12.1.4</span> SPRING network<a href="network-learning.html#spring-network" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As explained in Section <a href="network-learning.html#network-learning-workflow">12.1.1</a>, we use <a href="https://github.com/GraceYoon/SPRING">SPRING</a> (Semi-Parametric Rank-based approach for INference in Graphical model) as association measure. We first use the <code>SPRING</code> function directly to construct a conditional dependency graph.</p>
<p>Neither zero replacement nor normalization (steps 1 and 2 in our workflow) are required because SPRING uses a modified clr (mclr) transformation that can handle zero counts, and the correlation estimation method itself can also deal with zeros in the data. mclr is similar to the clr transformation except that mclr considers only the non-zero values. More precisely, the geometric mean is derived from positive values only, and zero counts remain zero after the transformation. This approach is similar to the robust clr (rclr) transformation included in the <code>vegan</code> package, except that mclr applies a positive shift to all non-zero values to make them strictly positive. See <span class="citation">(Yoon, Gaynanova, and Mller 2019)</span> for details.</p>
<p>The <code>Rmethod</code> argument is set to approx to estimate the correlations using a hybrid multi-linear interpolation approach proposed by <span class="citation">Yoon, Mller, and Gaynanova (2021)</span>. This method considerably reduces the runtime while controlling the approximation error.</p>
<p><code>SPRING</code> uses the StARS (Stability Approach to Regularization Selection) method <span class="citation">(H. Liu, Roeder, and Wasserman 2010)</span> to obtain a sparse association matrix. Thus, also step 4 of our workflow is already included. We set the StARS threshold to 0.05 to get a sparser graph.</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="network-learning.html#cb365-1" tabindex="-1"></a><span class="fu">library</span>(SPRING)</span></code></pre></div>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="network-learning.html#cb366-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13075</span>)</span>
<span id="cb366-2"><a href="network-learning.html#cb366-2" tabindex="-1"></a>spring_est <span class="ot">&lt;-</span> <span class="fu">SPRING</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>)), </span>
<span id="cb366-3"><a href="network-learning.html#cb366-3" tabindex="-1"></a>                     <span class="at">Rmethod =</span> <span class="st">&quot;approx&quot;</span>, </span>
<span id="cb366-4"><a href="network-learning.html#cb366-4" tabindex="-1"></a>                     <span class="at">thresh =</span> <span class="fl">0.05</span>,</span>
<span id="cb366-5"><a href="network-learning.html#cb366-5" tabindex="-1"></a>                     <span class="at">lambdaseq =</span> <span class="st">&quot;data-specific&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="network-learning.html#cb367-1" tabindex="-1"></a><span class="co"># Get index of the optimal lambda selected by StARS</span></span>
<span id="cb367-2"><a href="network-learning.html#cb367-2" tabindex="-1"></a>opt.K <span class="ot">&lt;-</span> spring_est<span class="sc">$</span>output<span class="sc">$</span>stars<span class="sc">$</span>opt.index</span>
<span id="cb367-3"><a href="network-learning.html#cb367-3" tabindex="-1"></a></span>
<span id="cb367-4"><a href="network-learning.html#cb367-4" tabindex="-1"></a><span class="co"># Store partial correlation matrix belonging to the optimal lambda as matrix</span></span>
<span id="cb367-5"><a href="network-learning.html#cb367-5" tabindex="-1"></a>spring_cor <span class="ot">&lt;-</span> SpiecEasi<span class="sc">::</span><span class="fu">symBeta</span>(<span class="fu">as.matrix</span>(spring_est<span class="sc">$</span>output<span class="sc">$</span>est<span class="sc">$</span>beta[[opt.K]]))</span>
<span id="cb367-6"><a href="network-learning.html#cb367-6" tabindex="-1"></a>spring_cor <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(spring_cor)</span>
<span id="cb367-7"><a href="network-learning.html#cb367-7" tabindex="-1"></a><span class="fu">rownames</span>(spring_cor) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(spring_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(tse)</span>
<span id="cb367-8"><a href="network-learning.html#cb367-8" tabindex="-1"></a><span class="fu">diag</span>(spring_cor) <span class="ot">&lt;-</span> <span class="dv">1</span></span></code></pre></div>
<p>As explained in Section <a href="network-learning.html#network-learning-workflow">12.1.1</a>, the estimated associations are sparsified, transformed into dissimilarities, and finally transformed into similarities, which are the adjacency values. We write a function for these steps, which will be reused later.</p>
<p>Since <code>SPRING</code> already includes a sparsification approach, the <code>thresh</code> argument is not needed here, but will be needed in Section <a href="network-learning.html#more-about-association">12.5</a> for other association measures.</p>
<p>To be consistent with the workflow, we provide two dissimilarity transformations: signed and unsigned (see Section <a href="network-learning.html#network-learning-workflow">12.1.1</a> for an explanation). These transformations were introduced by <span class="citation">Dongen and Enright (2012)</span>. We use the signed transformation in our examples so that strongly negatively associated genera have low edge weights.</p>
<p>The output of the function is an <code>igraph</code> object, which can be plotted and analyzed using functions from the <code>igraph</code> package.</p>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="network-learning.html#cb368-1" tabindex="-1"></a><span class="co"># Arguments:</span></span>
<span id="cb368-2"><a href="network-learning.html#cb368-2" tabindex="-1"></a><span class="co"># - assoMat: association matrix</span></span>
<span id="cb368-3"><a href="network-learning.html#cb368-3" tabindex="-1"></a><span class="co"># - threshold: associations below the threshold are set to zero</span></span>
<span id="cb368-4"><a href="network-learning.html#cb368-4" tabindex="-1"></a><span class="co"># - dissTrans: dissimilarity transformation (&quot;signed&quot; or &quot;unsigned&quot;)</span></span>
<span id="cb368-5"><a href="network-learning.html#cb368-5" tabindex="-1"></a></span>
<span id="cb368-6"><a href="network-learning.html#cb368-6" tabindex="-1"></a>transform_asso <span class="ot">&lt;-</span> <span class="cf">function</span>(assoMat, <span class="at">thresh =</span> <span class="cn">NULL</span>, <span class="at">dissTrans =</span> <span class="st">&quot;signed&quot;</span>) {</span>
<span id="cb368-7"><a href="network-learning.html#cb368-7" tabindex="-1"></a>  <span class="co"># Sparsification</span></span>
<span id="cb368-8"><a href="network-learning.html#cb368-8" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(thresh)) {</span>
<span id="cb368-9"><a href="network-learning.html#cb368-9" tabindex="-1"></a>    assoMat[<span class="fu">abs</span>(assoMat) <span class="sc">&lt;</span> thresh] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb368-10"><a href="network-learning.html#cb368-10" tabindex="-1"></a>  }</span>
<span id="cb368-11"><a href="network-learning.html#cb368-11" tabindex="-1"></a>  </span>
<span id="cb368-12"><a href="network-learning.html#cb368-12" tabindex="-1"></a>  <span class="co"># Compute dissimilarity matrix</span></span>
<span id="cb368-13"><a href="network-learning.html#cb368-13" tabindex="-1"></a>  <span class="cf">if</span> (dissTrans <span class="sc">==</span> <span class="st">&quot;signed&quot;</span>) {</span>
<span id="cb368-14"><a href="network-learning.html#cb368-14" tabindex="-1"></a>    dissMat <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fl">0.5</span> <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> assoMat))</span>
<span id="cb368-15"><a href="network-learning.html#cb368-15" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb368-16"><a href="network-learning.html#cb368-16" tabindex="-1"></a>    dissMat <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="dv">1</span> <span class="sc">-</span> assoMat<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb368-17"><a href="network-learning.html#cb368-17" tabindex="-1"></a>  }</span>
<span id="cb368-18"><a href="network-learning.html#cb368-18" tabindex="-1"></a>  </span>
<span id="cb368-19"><a href="network-learning.html#cb368-19" tabindex="-1"></a>  <span class="co"># Dissimilarity between nodes with zero correlation is set to 1</span></span>
<span id="cb368-20"><a href="network-learning.html#cb368-20" tabindex="-1"></a>  <span class="co"># (these nodes are unconnected and thus should have maximum dissimilarity)</span></span>
<span id="cb368-21"><a href="network-learning.html#cb368-21" tabindex="-1"></a>  dissMat[assoMat <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb368-22"><a href="network-learning.html#cb368-22" tabindex="-1"></a>  </span>
<span id="cb368-23"><a href="network-learning.html#cb368-23" tabindex="-1"></a>  <span class="co"># Compute similarity matrix</span></span>
<span id="cb368-24"><a href="network-learning.html#cb368-24" tabindex="-1"></a>  simMat <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> dissMat</span>
<span id="cb368-25"><a href="network-learning.html#cb368-25" tabindex="-1"></a>  </span>
<span id="cb368-26"><a href="network-learning.html#cb368-26" tabindex="-1"></a>  <span class="co"># Turn into igraph object</span></span>
<span id="cb368-27"><a href="network-learning.html#cb368-27" tabindex="-1"></a>  graphObj <span class="ot">&lt;-</span> SpiecEasi<span class="sc">::</span><span class="fu">adj2igraph</span>(simMat)</span>
<span id="cb368-28"><a href="network-learning.html#cb368-28" tabindex="-1"></a>  </span>
<span id="cb368-29"><a href="network-learning.html#cb368-29" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">graph =</span> graphObj, <span class="at">adja =</span> simMat, <span class="at">asso =</span> assoMat, <span class="at">diss =</span> dissMat))</span>
<span id="cb368-30"><a href="network-learning.html#cb368-30" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="network-learning.html#cb369-1" tabindex="-1"></a><span class="co"># Create graph object</span></span>
<span id="cb369-2"><a href="network-learning.html#cb369-2" tabindex="-1"></a>spring_graph <span class="ot">&lt;-</span> <span class="fu">transform_asso</span>(spring_cor)<span class="sc">$</span>graph</span></code></pre></div>
<p>The graph object can now be plotted using the <a href="https://r.igraph.org/">igraph</a> package, which is a state-of-the-art package for network analysis and visualization. Other packages that could be used for network plotting are the <a href="https://rdrr.io/cran/qgraph/">qgraph</a> package or the <a href="https://briatte.github.io/ggnet/">ggnet2</a> package. Since we will use igraph for network analysis later on, we are using its plotting function here as well.</p>
<p>We use the Fruchterman-Reingold layout (a force-directed layout) for node placement. By placing strongly connected nodes close together and those with low edge weight far apart, this layout results in an easy-to-read network plot.</p>
<p>The node size is proportional to a taxons log10-transformed abundance, which we previously added to the <code>tse</code> object, averaged across all samples. The values are rescaled to be visually distinguishable.</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="network-learning.html#cb370-1" tabindex="-1"></a><span class="fu">library</span>(igraph)</span></code></pre></div>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="network-learning.html#cb371-1" tabindex="-1"></a><span class="co"># Node sizes</span></span>
<span id="cb371-2"><a href="network-learning.html#cb371-2" tabindex="-1"></a>vsize <span class="ot">&lt;-</span> (<span class="fu">colMeans</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;log10&quot;</span>))) <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">*</span> <span class="dv">3</span></span>
<span id="cb371-3"><a href="network-learning.html#cb371-3" tabindex="-1"></a></span>
<span id="cb371-4"><a href="network-learning.html#cb371-4" tabindex="-1"></a><span class="co"># Fruchterman-Reingold layout from igraph package</span></span>
<span id="cb371-5"><a href="network-learning.html#cb371-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13075</span>)</span>
<span id="cb371-6"><a href="network-learning.html#cb371-6" tabindex="-1"></a>lay_fr <span class="ot">&lt;-</span> <span class="fu">layout_with_fr</span>(spring_graph)</span>
<span id="cb371-7"><a href="network-learning.html#cb371-7" tabindex="-1"></a></span>
<span id="cb371-8"><a href="network-learning.html#cb371-8" tabindex="-1"></a><span class="fu">plot</span>(spring_graph, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, <span class="at">vertex.label =</span> <span class="cn">NA</span>, </span>
<span id="cb371-9"><a href="network-learning.html#cb371-9" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">&quot;Conditional dependence graph</span><span class="sc">\n</span><span class="st">(generated with SPRING)&quot;</span>)</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/nets_sparcc_prop_se-1.png" width="480" /></p>
</div>
<div id="netcomi-spring" class="section level3 hasAnchor" number="12.1.5">
<h3><span class="header-section-number">12.1.5</span> NetCoMi network<a href="network-learning.html#netcomi-spring" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> <span class="citation">(Peschel et al. 2021)</span> package is specifically designed to construct, analyze, and compare networks for microbiome data and implements the complete workflow described in Section <a href="network-learning.html#network-learning-workflow">12.1.1</a>. Instead of using several functions for each of the steps, <code>NetCoMi</code> provides a single function for network construction (<code>netConstruct()</code>), so the package streamlines the workflow considerably. The user can choose from a variety of methods for data preprocessing, association estimation, sparsification, and transformation. The returned <code>microNet</code> object can then be passed to <code>netAnalyze()</code> (the network analysis function) so that all necessary information is available for the network analysis workflow.</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="network-learning.html#cb372-1" tabindex="-1"></a><span class="fu">library</span>(NetCoMi)</span></code></pre></div>
<p>We again use <code>SPRING</code> as one of the association measures available in NetCoMi to construct a conditional dependency graph.</p>
<p>To demonstrate how taxa are filtered with <code>netConstruct()</code>, we will use the unfiltered <code>tse</code> object this time. The filtering is the same as before: Taxa occurring in less than 20% of the samples are removed.</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="network-learning.html#cb373-1" tabindex="-1"></a>netcomi_net <span class="ot">&lt;-</span> <span class="fu">netConstruct</span>(tse,</span>
<span id="cb373-2"><a href="network-learning.html#cb373-2" tabindex="-1"></a>                            <span class="at">taxRank =</span> <span class="st">&quot;genus&quot;</span>,</span>
<span id="cb373-3"><a href="network-learning.html#cb373-3" tabindex="-1"></a>                            <span class="at">filtTax =</span> <span class="st">&quot;numbSamp&quot;</span>,</span>
<span id="cb373-4"><a href="network-learning.html#cb373-4" tabindex="-1"></a>                            <span class="at">filtTaxPar =</span> <span class="fu">list</span>(<span class="at">numbSamp =</span> <span class="fl">0.2</span>),</span>
<span id="cb373-5"><a href="network-learning.html#cb373-5" tabindex="-1"></a>                            <span class="at">measure =</span> <span class="st">&quot;spring&quot;</span>,</span>
<span id="cb373-6"><a href="network-learning.html#cb373-6" tabindex="-1"></a>                            <span class="at">measurePar =</span> <span class="fu">list</span>(<span class="at">thresh =</span> <span class="fl">0.05</span>,</span>
<span id="cb373-7"><a href="network-learning.html#cb373-7" tabindex="-1"></a>                                              <span class="at">Rmethod =</span> <span class="st">&quot;approx&quot;</span>),</span>
<span id="cb373-8"><a href="network-learning.html#cb373-8" tabindex="-1"></a>                            <span class="at">sparsMethod =</span> <span class="st">&quot;none&quot;</span>, </span>
<span id="cb373-9"><a href="network-learning.html#cb373-9" tabindex="-1"></a>                            <span class="at">dissFunc =</span> <span class="st">&quot;signed&quot;</span>,</span>
<span id="cb373-10"><a href="network-learning.html#cb373-10" tabindex="-1"></a>                            <span class="at">seed =</span> <span class="dv">13075</span>)</span></code></pre></div>
<p><code>netConstruct()</code> returns an object of the class <code>microNet</code>, which contains all matrices generated during network construction.</p>
<p>The object also contains an edge list, giving each edges estimated association, dissimilarity, and adjacency. Lets take a quick look at the edges with the highest and lowest edge weights:</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="network-learning.html#cb374-1" tabindex="-1"></a>edgelist <span class="ot">&lt;-</span> netcomi_net<span class="sc">$</span>edgelist1[<span class="fu">order</span>(netcomi_net<span class="sc">$</span>edgelist1<span class="sc">$</span>adja, </span>
<span id="cb374-2"><a href="network-learning.html#cb374-2" tabindex="-1"></a>                                        <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb374-3"><a href="network-learning.html#cb374-3" tabindex="-1"></a><span class="fu">head</span>(edgelist)</span></code></pre></div>
<pre><code>##                 v1                    v2   asso   diss   adja
## 73     Citrobacter           Escherichia 0.3682 0.5621 0.4379
## 63    Buttiauxella              Serratia 0.2426 0.6154 0.3846
## 69   Chitinivibrio Pseudogracilibacillus 0.2203 0.6244 0.3756
## 19        Algicola           Siccibacter 0.2193 0.6248 0.3752
## 111     Haliangium          Marinobacter 0.2148 0.6266 0.3734
## 143 Planomicrobium         Virgibacillus 0.2006 0.6322 0.3678</code></pre>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="network-learning.html#cb376-1" tabindex="-1"></a><span class="fu">tail</span>(edgelist)</span></code></pre></div>
<pre><code>##                v1             v2       asso   diss   adja
## 132 Mycobacterium Salinibacillus  0.0017483 0.7065 0.2935
## 24      Amphritea    Providencia  0.0014085 0.7066 0.2934
## 102       Erwinia    Siccibacter  0.0013114 0.7066 0.2934
## 116     Holophaga   Methylarcula  0.0007828 0.7068 0.2932
## 17       Algicola     Lysobacter  0.0005191 0.7069 0.2931
## 95   Enterobacter     Janibacter -0.0013921 0.7076 0.2924</code></pre>
<p>As before, the adjacency matrix is converted into an <code>igraph</code> object. Further steps like sparsification and transformation are not necessary because they are done internally by <code>netConstruct()</code>.</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="network-learning.html#cb378-1" tabindex="-1"></a>netcomi_graph <span class="ot">&lt;-</span> SpiecEasi<span class="sc">::</span><span class="fu">adj2igraph</span>(<span class="fu">abs</span>(netcomi_net<span class="sc">$</span>adjaMat1))</span></code></pre></div>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="network-learning.html#cb379-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb379-2"><a href="network-learning.html#cb379-2" tabindex="-1"></a><span class="fu">plot</span>(spring_graph, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, </span>
<span id="cb379-3"><a href="network-learning.html#cb379-3" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="st">&quot;SPRING network&quot;</span>)</span>
<span id="cb379-4"><a href="network-learning.html#cb379-4" tabindex="-1"></a><span class="fu">plot</span>(netcomi_graph, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, </span>
<span id="cb379-5"><a href="network-learning.html#cb379-5" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="st">&quot;NetCoMi network</span><span class="sc">\n</span><span class="st">(with SPRING associations)&quot;</span>)</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/nets_sparcc_se_spring-1.png" width="768" /></p>
<p>The left network constructed directly with the <code>SPRING()</code> function and the right network constructed with <code>NetCoMi</code> are exactly the same.</p>
</div>
</div>
<div id="network-analysis" class="section level2 hasAnchor" number="12.2">
<h2><span class="header-section-number">12.2</span> Network analysis<a href="network-learning.html#network-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The computed network is now analyzed using appropriate methods. We will first use the <code>igraph</code> package to analyze the <code>SPRING</code> network. NetCoMis <code>netAnalyze()</code> function will be used later to analyze the constructed <code>microNet</code> object.</p>
<div id="network-analysis-with-igraph" class="section level3 hasAnchor" number="12.2.1">
<h3><span class="header-section-number">12.2.1</span> Network analysis with igraph<a href="network-learning.html#network-analysis-with-igraph" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="centrality-measures" class="section level4 hasAnchor" number="12.2.1.1">
<h4><span class="header-section-number">12.2.1.1</span> Centrality measures<a href="network-learning.html#centrality-measures" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Centrality measures express the importance of nodes within the network. Common measures are the degree, betweenness, closeness, and eigenvector centrality. The <code>igraph</code> package provides functions to compute these measures.</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="network-learning.html#cb380-1" tabindex="-1"></a><span class="co"># We access igraph directly with &quot;::&quot; because there are more packages loaded in </span></span>
<span id="cb380-2"><a href="network-learning.html#cb380-2" tabindex="-1"></a><span class="co"># this chapter that contain a degree() function.</span></span>
<span id="cb380-3"><a href="network-learning.html#cb380-3" tabindex="-1"></a>centr_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Degree =</span> igraph<span class="sc">::</span><span class="fu">degree</span>(spring_graph))</span>
<span id="cb380-4"><a href="network-learning.html#cb380-4" tabindex="-1"></a>centr_df<span class="sc">$</span>Betweenness <span class="ot">&lt;-</span> <span class="fu">betweenness</span>(spring_graph)</span>
<span id="cb380-5"><a href="network-learning.html#cb380-5" tabindex="-1"></a>centr_df<span class="sc">$</span>Closeness <span class="ot">&lt;-</span> <span class="fu">closeness</span>(spring_graph)</span>
<span id="cb380-6"><a href="network-learning.html#cb380-6" tabindex="-1"></a>centr_df<span class="sc">$</span>Eigenvector <span class="ot">&lt;-</span> <span class="fu">eigen_centrality</span>(spring_graph)<span class="sc">$</span>vector</span>
<span id="cb380-7"><a href="network-learning.html#cb380-7" tabindex="-1"></a></span>
<span id="cb380-8"><a href="network-learning.html#cb380-8" tabindex="-1"></a><span class="fu">rownames</span>(centr_df) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spring_cor)</span>
<span id="cb380-9"><a href="network-learning.html#cb380-9" tabindex="-1"></a><span class="fu">head</span>(centr_df, <span class="dv">15</span>)</span></code></pre></div>
<pre><code>##                  Degree Betweenness Closeness Eigenvector
## Abyssicoccus          0           0       NaN   2.198e-17
## Acidaminococcus       2           0  0.006338   1.862e-01
## Acinetobacter         3           5  0.005436   1.257e-01
## Actinomyces           0           0       NaN   2.198e-17
## Actinoplanes          2          99  0.005004   1.099e-02
## Aerococcus            0           0       NaN   2.198e-17
## Aeromonas             4         351  0.007331   2.023e-01
## Agromyces             6         435  0.007581   5.293e-01
## Algicola              4         264  0.007064   1.500e-01
## Alicyclobacillus      0           0       NaN   2.198e-17
## Alteribacillus        0           0       NaN   2.198e-17
## Ammoniibacillus       1           0  0.004965   5.419e-02
## Amphritea             5         382  0.006133   1.335e-02
## Amycolatopsis         1           0  0.005997   1.338e-01
## Anaerococcus          2         392  0.003985   1.288e-03</code></pre>
<p>The closeness centrality is NaN for some genera. These are unconnected nodes, as can be seen by the zero degree and betweenness centrality.</p>
</div>
<div id="scale-node-sizes-by-degree" class="section level4 hasAnchor" number="12.2.1.2">
<h4><span class="header-section-number">12.2.1.2</span> Scale node sizes by degree<a href="network-learning.html#scale-node-sizes-by-degree" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Centrality measures can be visualized in the network plot by scaling the node sizes according to one of these measures. We plot the <code>Spring</code> graph using the same layout as before and with the node sizes scaled according to all four centrality measures.</p>
<p>Of the four centrality measures, only the degree has a range suitable to be used as node size. The other centrality measures must be rescaled because their range is either too small or too large. The following scaling is a suggestion that works for this example. The values might be adapted for other data sets.</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="network-learning.html#cb382-1" tabindex="-1"></a>vsize_df <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(centr_df)</span>
<span id="cb382-2"><a href="network-learning.html#cb382-2" tabindex="-1"></a>vsize_df[, <span class="st">&quot;Betweenness&quot;</span>] <span class="ot">&lt;-</span> <span class="fu">log</span>(vsize_df[, <span class="st">&quot;Betweenness&quot;</span>])</span>
<span id="cb382-3"><a href="network-learning.html#cb382-3" tabindex="-1"></a>vsize_df[, <span class="st">&quot;Closeness&quot;</span>] <span class="ot">&lt;-</span> <span class="fu">log</span>(vsize_df[, <span class="st">&quot;Closeness&quot;</span>] <span class="sc">*</span> <span class="dv">1000</span>)</span>
<span id="cb382-4"><a href="network-learning.html#cb382-4" tabindex="-1"></a>vsize_df[, <span class="st">&quot;Eigenvector&quot;</span>] <span class="ot">&lt;-</span> vsize_df[, <span class="st">&quot;Eigenvector&quot;</span>] <span class="sc">*</span> <span class="dv">10</span></span>
<span id="cb382-5"><a href="network-learning.html#cb382-5" tabindex="-1"></a></span>
<span id="cb382-6"><a href="network-learning.html#cb382-6" tabindex="-1"></a>vsize_df[<span class="fu">is.infinite</span>(vsize_df) <span class="sc">|</span> <span class="fu">is.na</span>(vsize_df)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb382-7"><a href="network-learning.html#cb382-7" tabindex="-1"></a></span>
<span id="cb382-8"><a href="network-learning.html#cb382-8" tabindex="-1"></a><span class="fu">head</span>(vsize_df)</span></code></pre></div>
<pre><code>##                 Degree Betweenness Closeness Eigenvector
## Abyssicoccus         0       0.000     0.000   2.198e-16
## Acidaminococcus      2       0.000     1.846   1.862e+00
## Acinetobacter        3       1.609     1.693   1.257e+00
## Actinomyces          0       0.000     0.000   2.198e-16
## Actinoplanes         2       4.595     1.610   1.099e-01
## Aerococcus           0       0.000     0.000   2.198e-16</code></pre>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="network-learning.html#cb384-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb384-2"><a href="network-learning.html#cb384-2" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(centr_df)) {</span>
<span id="cb384-3"><a href="network-learning.html#cb384-3" tabindex="-1"></a>  <span class="fu">plot</span>(spring_graph, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize_df[, i], </span>
<span id="cb384-4"><a href="network-learning.html#cb384-4" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="fu">colnames</span>(centr_df)[i])</span>
<span id="cb384-5"><a href="network-learning.html#cb384-5" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/unnamed-chunk-12-1.png" width="960" /></p>
</div>
<div id="degree-distribution" class="section level4 hasAnchor" number="12.2.1.3">
<h4><span class="header-section-number">12.2.1.3</span> Degree distribution<a href="network-learning.html#degree-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The degree distribution is another popular measure that expresses the probability distribution of degrees over the entire network. It thus provides insight into the overall network structure. We plot the degree distribution for all four association estimation methods to compare the network structure.</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="network-learning.html#cb385-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="network-learning.html#cb386-1" tabindex="-1"></a><span class="co"># Compute degree distribution</span></span>
<span id="cb386-2"><a href="network-learning.html#cb386-2" tabindex="-1"></a>ddist<span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">degree.distribution</span>(spring_graph)</span>
<span id="cb386-3"><a href="network-learning.html#cb386-3" tabindex="-1"></a></span>
<span id="cb386-4"><a href="network-learning.html#cb386-4" tabindex="-1"></a><span class="co"># Data frame needed for ggplot2</span></span>
<span id="cb386-5"><a href="network-learning.html#cb386-5" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Degree =</span> <span class="fu">as.factor</span>((<span class="fu">seq_along</span>(ddist)) <span class="sc">-</span> <span class="dv">1</span>),</span>
<span id="cb386-6"><a href="network-learning.html#cb386-6" tabindex="-1"></a>                 <span class="at">Fraction =</span> ddist)</span>
<span id="cb386-7"><a href="network-learning.html#cb386-7" tabindex="-1"></a></span>
<span id="cb386-8"><a href="network-learning.html#cb386-8" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> df, <span class="fu">aes</span>(<span class="at">x =</span> Degree, <span class="at">y =</span> Fraction, <span class="at">group =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb386-9"><a href="network-learning.html#cb386-9" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb386-10"><a href="network-learning.html#cb386-10" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb386-11"><a href="network-learning.html#cb386-11" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>The network has a large number of singletons and sparsely connected nodes, and only a small number of nodes with a higher degree of 7 or more.</p>
</div>
<div id="clustered-heatmaps" class="section level4 hasAnchor" number="12.2.1.4">
<h4><span class="header-section-number">12.2.1.4</span> Clustered heatmaps<a href="network-learning.html#clustered-heatmaps" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Using the <code>ComplexHeatmap</code> package, we plot a heatmap of the association matrix estimated with <code>SPRING</code>. Rows and columns are sorted according to the clusters identified via hierarchical clustering.</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="network-learning.html#cb387-1" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb387-2"><a href="network-learning.html#cb387-2" tabindex="-1"></a><span class="fu">library</span>(circlize)</span></code></pre></div>
<p>We select the 50 nodes with the highest sum of edge weights to get a smaller heatmap.</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="network-learning.html#cb388-1" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">rowSums</span>(spring_cor), <span class="at">decreasing =</span> <span class="cn">TRUE</span>))[<span class="fu">seq_len</span>(<span class="dv">50</span>)]</span>
<span id="cb388-2"><a href="network-learning.html#cb388-2" tabindex="-1"></a>adja_sel <span class="ot">&lt;-</span> spring_cor[sel, sel]</span></code></pre></div>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="network-learning.html#cb389-1" tabindex="-1"></a><span class="co"># Color vector</span></span>
<span id="cb389-2"><a href="network-learning.html#cb389-2" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), </span>
<span id="cb389-3"><a href="network-learning.html#cb389-3" tabindex="-1"></a>                  <span class="fu">c</span>(<span class="st">&quot;royalblue4&quot;</span>, <span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;orange&quot;</span>, <span class="st">&quot;firebrick3&quot;</span>))</span>
<span id="cb389-4"><a href="network-learning.html#cb389-4" tabindex="-1"></a></span>
<span id="cb389-5"><a href="network-learning.html#cb389-5" tabindex="-1"></a><span class="fu">Heatmap</span>(adja_sel, </span>
<span id="cb389-6"><a href="network-learning.html#cb389-6" tabindex="-1"></a>        <span class="at">col =</span> col, </span>
<span id="cb389-7"><a href="network-learning.html#cb389-7" tabindex="-1"></a>        <span class="at">rect_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">&quot;gray&quot;</span>, <span class="at">lwd =</span> <span class="dv">1</span>),</span>
<span id="cb389-8"><a href="network-learning.html#cb389-8" tabindex="-1"></a>        <span class="at">show_row_names =</span> <span class="cn">FALSE</span>, </span>
<span id="cb389-9"><a href="network-learning.html#cb389-9" tabindex="-1"></a>        <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb389-10"><a href="network-learning.html#cb389-10" tabindex="-1"></a>        <span class="at">name =</span> <span class="st">&quot;Association&quot;</span>)</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/clustered_heatmap-1.png" width="556.8" /></p>
<p>The associations are generally quite low, and there are no prominent clusters detected by hierarchical clustering.</p>
</div>
<div id="global-network-measures" class="section level4 hasAnchor" number="12.2.1.5">
<h4><span class="header-section-number">12.2.1.5</span> Global network measures<a href="network-learning.html#global-network-measures" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Global measures describe the overall network structure. We take a look at three common measures: density, transitivity, and average path length. The values are again computed with <code>igraph</code> functions.</p>
<div id="density" class="section level5 hasAnchor" number="12.2.1.5.1">
<h5><span class="header-section-number">12.2.1.5.1</span> Density<a href="network-learning.html#density" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Definition: Proportion of present edges from all possible edges.</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="network-learning.html#cb390-1" tabindex="-1"></a><span class="fu">edge_density</span>(spring_graph)</span></code></pre></div>
<pre><code>## [1] 0.01444</code></pre>
</div>
<div id="transitivity-clustering-coefficient" class="section level5 hasAnchor" number="12.2.1.5.2">
<h5><span class="header-section-number">12.2.1.5.2</span> Transitivity (clustering coefficient)<a href="network-learning.html#transitivity-clustering-coefficient" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Here, we consider only the global clustering coefficient, which is defined as the ratio of triangles to connected triples.</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="network-learning.html#cb392-1" tabindex="-1"></a><span class="fu">transitivity</span>(spring_graph)</span></code></pre></div>
<pre><code>## [1] 0.1456</code></pre>
</div>
<div id="average-path-length" class="section level5 hasAnchor" number="12.2.1.5.3">
<h5><span class="header-section-number">12.2.1.5.3</span> Average path length<a href="network-learning.html#average-path-length" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Definition: Mean of the shortest distance between each pair of nodes.</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="network-learning.html#cb394-1" tabindex="-1"></a><span class="fu">average.path.length</span>(spring_graph)</span></code></pre></div>
<pre><code>## [1] 1.699</code></pre>
</div>
</div>
</div>
<div id="network-analysis-with-netcomi" class="section level3 hasAnchor" number="12.2.2">
<h3><span class="header-section-number">12.2.2</span> Network analysis with NetCoMi<a href="network-learning.html#network-analysis-with-netcomi" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <code>netcomi_net</code> object of class <code>microNet</code> created before is now passed to <code>netAnalyze()</code> to perform network analysis with <code>NetCoMi</code>.</p>
<p>The function computes several common network characteristics such as centrality measures, cluster assignment, the graphlet correlation matrix, as well as global network measures.</p>
<p>The user has several options to choose from, such as a clustering method, how to define hubs, and whether or not to normalize centrality values. See the help page <code>?netAnalyze</code> for a description of the arguments.</p>
<p>By default, a heatmap of the Graphlet Correlation Matrix (GCM) is returned (with graphlet correlations in the upper triangle and significance codes resulting from Students t-test in the lower triangle). See <code>?calcGCM</code> and <code>?testGCM</code> for details.</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="network-learning.html#cb396-1" tabindex="-1"></a>netcomi_netprops <span class="ot">&lt;-</span> <span class="fu">netAnalyze</span>(netcomi_net, </span>
<span id="cb396-2"><a href="network-learning.html#cb396-2" tabindex="-1"></a>                               <span class="at">clustMethod =</span> <span class="st">&quot;cluster_fast_greedy&quot;</span>,</span>
<span id="cb396-3"><a href="network-learning.html#cb396-3" tabindex="-1"></a>                               <span class="at">hubPar =</span> <span class="st">&quot;eigenvector&quot;</span>,</span>
<span id="cb396-4"><a href="network-learning.html#cb396-4" tabindex="-1"></a>                               <span class="at">normDeg =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/netAnalyze_single-1.png" width="75%" /></p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb397-1"><a href="network-learning.html#cb397-1" tabindex="-1"></a><span class="fu">summary</span>(netcomi_netprops, <span class="at">numbNodes =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## 
## Component sizes
## ```````````````              
## size: 103 2  1
##    #:   1 1 42
## ______________________________
## Global network properties
## `````````````````````````
## Largest connected component (LCC):
##                                  
## Relative LCC size         0.70068
## Clustering coefficient    0.21253
## Modularity                0.70495
## Positive edge percentage 99.35065
## Edge density              0.02932
## Natural connectivity      0.01173
## Vertex connectivity       1.00000
## Edge connectivity         1.00000
## Average dissimilarity*    0.99084
## Average path length**     3.81101
## 
## Whole network:
##                                  
## Number of components     44.00000
## Clustering coefficient    0.21253
## Modularity                0.70787
## Positive edge percentage 99.35484
## Edge density              0.01444
## Natural connectivity      0.00782
## -----
## *: Dissimilarity = 1 - edge weight
## **: Path length = Units with average dissimilarity
## 
## ______________________________
## Clusters
## - In the whole network
## - Algorithm: cluster_fast_greedy
## ```````````````````````````````` 
##                                   
## name:  0  1  2  3  4 5  6 7 8 9 10
##    #: 42 16 11 14 20 9 17 8 5 3  2
## 
## ______________________________
## Hubs
## - In alphabetical/numerical order
## - Based on empirical quantiles of centralities
## ```````````````````````````````````````````````                      
##  Agromyces            
##  Aneurinibacillus     
##  Anoxybacillus        
##  Chitinivibrio        
##  Erwinia              
##  Geobacillus          
##  Janibacter           
##  Pseudogracilibacillus
## 
## ______________________________
## Centrality measures
## - In decreasing order
## - Centrality of disconnected components is zero
## ````````````````````````````````````````````````
## Degree (unnormalized):
##                        
## Anoxybacillus        10
## Erwinia              10
## Chitinivibrio         9
## Janibacter            8
## Escherichia/Shigella  7
## 
## Betweenness centrality (normalized):
##                     
## Janibacter    0.3941
## Chitinivibrio 0.3137
## Enterobacter  0.2780
## Buttiauxella  0.2454
## Erwinia       0.2095
## 
## Closeness centrality (normalized):
##                             
## Chitinivibrio         0.5247
## Janibacter            0.5067
## Anoxybacillus         0.4841
## Pseudogracilibacillus 0.4659
## Erwinia               0.4653
## 
## Eigenvector centrality (normalized):
##                             
## Anoxybacillus         1.0000
## Chitinivibrio         0.9831
## Pseudogracilibacillus 0.8657
## Janibacter            0.8361
## Aneurinibacillus      0.6890</code></pre>
<p><strong>Interpretation of some findings:</strong></p>
<ul>
<li>The largest connected component (LCC) has 103 nodes and the network contains 42 singletons.</li>
<li>10 clusters have been identified, containing 2 to 20 nodes.</li>
<li>There are 8 hub nodes detected, which by definition are the nodes with the highest eigenvector centrality.</li>
<li>The average path length in the LCC is 3.811. This means that on average it takes 3.811 steps (step length is the average dissimilarity) to get from one node to another. Note that the average path length in <code>NetCoMi</code> is defined differently than in the <code>igraph</code> package, which is why the values differ.</li>
<li>Low values of edge density and the connectivity measures indicate that the network is rather sparse and not robust to perturbations (i.e., removal of nodes or edges).</li>
</ul>
</div>
</div>
<div id="network-visualization" class="section level2 hasAnchor" number="12.3">
<h2><span class="header-section-number">12.3</span> Network visualization<a href="network-learning.html#network-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Further insight into the network structure can be gained by visualizing the network. We have already seen examples of how to plot a network using the <code>igraph</code> package. Here we will use NetCoMis plot function. It takes as input the <code>microNetProps</code> object returned by <code>netAnalyze()</code>, which contains all computed network properties. This has the advantage that the user can choose which properties to plot by simply changing some arguments. The plot function is based on <a href="https://rdrr.io/cran/qgraph/">qgraph</a>, which is another state-of-the-art R package for network visualization. The help page can be accessed via <code>?plot.microNetProps</code>.</p>
<div id="highlight-node-properties" class="section level3 hasAnchor" number="12.3.1">
<h3><span class="header-section-number">12.3.1</span> Highlight node properties<a href="network-learning.html#highlight-node-properties" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the first plot, node colors represent the detected clusters and node sizes are scaled by eigenvector centrality. Hub nodes are highlighted by default. Singletons are not included in the plot. To improve the readability, NetCoMis intelligent label shortening approach is used.</p>
<p>Note that nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion argument until you find a value where the labels are legible, but also the clusters are still well recognizable.</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="network-learning.html#cb399-1" tabindex="-1"></a><span class="fu">plot</span>(netcomi_netprops,</span>
<span id="cb399-2"><a href="network-learning.html#cb399-2" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.98</span>,</span>
<span id="cb399-3"><a href="network-learning.html#cb399-3" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb399-4"><a href="network-learning.html#cb399-4" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">&quot;intelligent&quot;</span>,</span>
<span id="cb399-5"><a href="network-learning.html#cb399-5" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb399-6"><a href="network-learning.html#cb399-6" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">&quot;eigenvector&quot;</span>,</span>
<span id="cb399-7"><a href="network-learning.html#cb399-7" tabindex="-1"></a>     <span class="at">nodeSizeSpread =</span> <span class="dv">3</span>,</span>
<span id="cb399-8"><a href="network-learning.html#cb399-8" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">&quot;cluster&quot;</span>, </span>
<span id="cb399-9"><a href="network-learning.html#cb399-9" tabindex="-1"></a>     <span class="at">hubBorderCol =</span> <span class="st">&quot;gray40&quot;</span>,</span>
<span id="cb399-10"><a href="network-learning.html#cb399-10" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.8</span>,</span>
<span id="cb399-11"><a href="network-learning.html#cb399-11" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb399-12"><a href="network-learning.html#cb399-12" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">&quot;Network properties highlighted&quot;</span>, </span>
<span id="cb399-13"><a href="network-learning.html#cb399-13" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb399-14"><a href="network-learning.html#cb399-14" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb399-15"><a href="network-learning.html#cb399-15" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">8</span>))</span>
<span id="cb399-16"><a href="network-learning.html#cb399-16" tabindex="-1"></a></span>
<span id="cb399-17"><a href="network-learning.html#cb399-17" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">&quot;estimated correlation:&quot;</span>,</span>
<span id="cb399-18"><a href="network-learning.html#cb399-18" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;#009900&quot;</span>,<span class="st">&quot;red&quot;</span>), </span>
<span id="cb399-19"><a href="network-learning.html#cb399-19" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/network_plot_single_cluster-1.png" width="1248" /></p>
</div>
<div id="highlight-data-features" class="section level3 hasAnchor" number="12.3.2">
<h3><span class="header-section-number">12.3.2</span> Highlight data features<a href="network-learning.html#highlight-data-features" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We now color nodes according to their phylum. The node sizes are proportional to a taxons sum of mclr-transformed abundances. As already mentioned in Section <a href="network-learning.html#spring-network">12.1.4</a>, this is the normalization method used by <code>SPRING</code>. A color palette from <code>RColorBrewer</code> is used here.</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb400-1"><a href="network-learning.html#cb400-1" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span></code></pre></div>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="network-learning.html#cb401-1" tabindex="-1"></a><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span id="cb401-2"><a href="network-learning.html#cb401-2" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>phylum)</span>
<span id="cb401-3"><a href="network-learning.html#cb401-3" tabindex="-1"></a><span class="fu">names</span>(phyla) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(tse)<span class="sc">$</span>genus</span>
<span id="cb401-4"><a href="network-learning.html#cb401-4" tabindex="-1"></a></span>
<span id="cb401-5"><a href="network-learning.html#cb401-5" tabindex="-1"></a><span class="co"># Create color vector</span></span>
<span id="cb401-6"><a href="network-learning.html#cb401-6" tabindex="-1"></a>colvec <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="fu">length</span>(<span class="fu">levels</span>(phyla)), <span class="st">&quot;Set3&quot;</span>)</span>
<span id="cb401-7"><a href="network-learning.html#cb401-7" tabindex="-1"></a></span>
<span id="cb401-8"><a href="network-learning.html#cb401-8" tabindex="-1"></a><span class="fu">plot</span>(netcomi_netprops,</span>
<span id="cb401-9"><a href="network-learning.html#cb401-9" tabindex="-1"></a>     <span class="at">repulsion =</span> <span class="fl">0.98</span>,</span>
<span id="cb401-10"><a href="network-learning.html#cb401-10" tabindex="-1"></a>     <span class="at">rmSingles =</span> <span class="cn">TRUE</span>,</span>
<span id="cb401-11"><a href="network-learning.html#cb401-11" tabindex="-1"></a>     <span class="at">shortenLabels =</span> <span class="st">&quot;intelligent&quot;</span>,</span>
<span id="cb401-12"><a href="network-learning.html#cb401-12" tabindex="-1"></a>     <span class="at">labelScale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb401-13"><a href="network-learning.html#cb401-13" tabindex="-1"></a>     <span class="at">nodeSize =</span> <span class="st">&quot;mclr&quot;</span>,</span>
<span id="cb401-14"><a href="network-learning.html#cb401-14" tabindex="-1"></a>     <span class="at">nodeColor =</span> <span class="st">&quot;feature&quot;</span>, </span>
<span id="cb401-15"><a href="network-learning.html#cb401-15" tabindex="-1"></a>     <span class="at">featVecCol =</span> phyla, </span>
<span id="cb401-16"><a href="network-learning.html#cb401-16" tabindex="-1"></a>     <span class="at">colorVec =</span>  colvec,</span>
<span id="cb401-17"><a href="network-learning.html#cb401-17" tabindex="-1"></a>     <span class="at">nodeTransp =</span> <span class="dv">20</span>,</span>
<span id="cb401-18"><a href="network-learning.html#cb401-18" tabindex="-1"></a>     <span class="at">highlightHubs =</span> <span class="cn">FALSE</span>,</span>
<span id="cb401-19"><a href="network-learning.html#cb401-19" tabindex="-1"></a>     <span class="at">cexNodes =</span> <span class="fl">1.8</span>,</span>
<span id="cb401-20"><a href="network-learning.html#cb401-20" tabindex="-1"></a>     <span class="at">edgeTranspHigh =</span> <span class="dv">20</span>,</span>
<span id="cb401-21"><a href="network-learning.html#cb401-21" tabindex="-1"></a>     <span class="at">title1 =</span> <span class="st">&quot;Data features highlighted&quot;</span>, </span>
<span id="cb401-22"><a href="network-learning.html#cb401-22" tabindex="-1"></a>     <span class="at">showTitle =</span> <span class="cn">TRUE</span>,</span>
<span id="cb401-23"><a href="network-learning.html#cb401-23" tabindex="-1"></a>     <span class="at">cexTitle =</span> <span class="fl">2.3</span>,</span>
<span id="cb401-24"><a href="network-learning.html#cb401-24" tabindex="-1"></a>     <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">4</span>, <span class="dv">6</span>))</span>
<span id="cb401-25"><a href="network-learning.html#cb401-25" tabindex="-1"></a></span>
<span id="cb401-26"><a href="network-learning.html#cb401-26" tabindex="-1"></a><span class="co"># Add legends</span></span>
<span id="cb401-27"><a href="network-learning.html#cb401-27" tabindex="-1"></a><span class="fu">legend</span>(<span class="fl">0.7</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">title =</span> <span class="st">&quot;estimated correlation:&quot;</span>,</span>
<span id="cb401-28"><a href="network-learning.html#cb401-28" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>), <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;#009900&quot;</span>,<span class="st">&quot;red&quot;</span>), </span>
<span id="cb401-29"><a href="network-learning.html#cb401-29" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">horiz =</span> <span class="cn">TRUE</span>)</span>
<span id="cb401-30"><a href="network-learning.html#cb401-30" tabindex="-1"></a></span>
<span id="cb401-31"><a href="network-learning.html#cb401-31" tabindex="-1"></a><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span id="cb401-32"><a href="network-learning.html#cb401-32" tabindex="-1"></a>col_transp <span class="ot">&lt;-</span> <span class="fu">colToTransp</span>(colvec, <span class="dv">20</span>)</span>
<span id="cb401-33"><a href="network-learning.html#cb401-33" tabindex="-1"></a></span>
<span id="cb401-34"><a href="network-learning.html#cb401-34" tabindex="-1"></a><span class="fu">legend</span>(<span class="sc">-</span><span class="fl">1.8</span>, <span class="fl">1.1</span>, <span class="at">cex =</span> <span class="fl">1.7</span>, <span class="at">pt.cex =</span> <span class="fl">2.5</span>, <span class="at">title =</span> <span class="st">&quot;Phylum:&quot;</span>, </span>
<span id="cb401-35"><a href="network-learning.html#cb401-35" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">levels</span>(phyla), <span class="at">col =</span> col_transp, <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">pch =</span> <span class="dv">16</span>) </span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/network_plot_single_phylum-1.png" width="1440" /></p>
<p><strong>A few things to observe:</strong></p>
<ul>
<li>Genera belonging to the same phylum tend to cluster together, though not perfectly.</li>
<li>Genera with a low total count play a rather unimportant role in the network, i.e., they have a low centrality.</li>
<li>There is only one negative edge in the network. This edge is between two clusters, as expected when using the signed transformation.</li>
</ul>
</div>
</div>
<div id="which-methods-to-choose" class="section level2 hasAnchor" number="12.4">
<h2><span class="header-section-number">12.4</span> Which method(s) to choose?<a href="network-learning.html#which-methods-to-choose" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Throughout all the steps from primary data to potentially significant network features, there is a variety of methods and parameters to choose from. However, there is no general consensus in the community on the right way to estimate and analyze microbial networks. In the absence of a best method for inferring and analyzing microbial networks, researchers may be tempted to try different methods and report only the optimal results or those that fit some prior knowledge. This carries the risk of overfitting the analysis to the existing data so that the results are not replicable for new data <span class="citation">(Ullmann et al. 2023)</span>.</p>
<p>Therefore, the selection of the workflow building blocks should be set up once and independently of any hypothesis about the data, thus avoiding the fallacy of starting to fish for results that best fit a previously formulated hypothesis. For example, one should ask prior to the analysis whether correlation or conditional dependence as a measure of association better fits the research question and choose the method accordingly. Another example is the choice of transformation from estimated association to dissimilarity (i.e., signed or unsigned), which completely changes the interpretation and characteristics of the network. This choice should be made based on the research question before starting the analysis.</p>
</div>
<div id="more-about-association" class="section level2 hasAnchor" number="12.5">
<h2><span class="header-section-number">12.5</span> More about association measures<a href="network-learning.html#more-about-association" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>As mentioned in the introduction of this chapter, there are three types of association measures that are commonly used to express relationships between taxa: correlation, conditional dependence, and proportionality. Below, we provide a brief explanation of each of these measures, along with lists of available compositionality-aware approaches.</p>
<ul>
<li><strong>Correlation:</strong> Two popular measures of ecological association are Pearsons correlation coefficient and Spearmans rank correlation coefficient, both of which can be inferred from empirical (sample) covariances. However, in the <span class="math inline">\(p\gg n\)</span> setting, which most microbiome datasets are in, sample covariances and correlations are unreliable because the parameters being estimated are typically underdetermined. One way to improve sample covariance estimates is to assume that the underlying covariance matrix is sparse and use a regularized covariance estimator to implement this structural assumption. The Schfer-Strimmer shrinkage estimator <span class="citation">(Schfer and Strimmer 2005)</span> is one possible method for estimating a sparse correlation matrix. Other popular methods, especially designed to estimate correlations for compositional data, are SparCC (Sparse Correlations for Compositional data) by <span class="citation">J. Friedman and Alm (2012)</span>, CCREPE (Compositionality Corrected by REnormalization and PErmutation) by <span class="citation">Faust et al. (2012)</span>, and CCLasso (Correlation inference for Compositional data through Lasso) by <span class="citation">Huaying Fang et al. (2015)</span>. The latter three methods already include a compositionality aware normalization, and SparCC also includes a zero replacement approach.</li>
<li><strong>Conditional dependence:</strong> Since standard correlations include both direct and indirect dependencies, conditional dependence or partial correlation is often preferred for measuring association. Unlike (marginal) correlation, it expresses the relationship between two features conditioned on all other features in the data set. The approach and R package <a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a> (Sparse InversE Covariance estimation for Ecological Association and Statistical Inference) by <span class="citation">Kurtz et al. (2015)</span> is specifically designed for inferring ecological networks from microbiome data and includes two approaches for estimating conditional dependence structures between taxa: Neighborhood Selection; short MB <span class="citation">(Meinshausen and Bhlmann 2006)</span> and (inverse) covariance selection <span class="citation">(Jerome Friedman, Hastie, and Tibshirani 2008)</span>, which is based on a penalized maximum likelihood approach and is also known as graphical lasso. Another approach and R package for inferring partial correlations from microbiome data is <a href="https://github.com/GraceYoon/SPRING">SPRING</a> (Semi-Parametric Rank-based approach for INference in Graphical model) by <span class="citation">Yoon, Gaynanova, and Mller (2019)</span>. They also use the MB neighborhood selection method, but introduce a novel semi-parametric rank-based approach for sparse partial correlation estimation that can naturally handle the excess of zeros in the data. gCoda <span class="citation">(H. Fang et al. 2017)</span> is another conditional dependence measure based on penalized maximum likelihood estimation. All of the aforementioned conditional dependence measures address the high dimensionality of microbiome data.</li>
<li><strong>Proportionality:</strong> <span class="citation">Lovell et al. (2015)</span> introduce proportionality as an alternative measure of pairwise association for compositional data. The idea is that if the relative abundances between two taxa <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are proportional, then their corresponding absolute abundances are also proportional: <span class="math inline">\(\frac{\omega_i}{m} \propto \frac{\omega_j}{m} \Rightarrow \omega_i \propto \omega_j\)</span>, where <span class="math inline">\(m\)</span> is the sum of counts in the sample. It follows that proportionality is identical for the observed (relative) read counts and the true unobserved counts. The proportionality measure proposed by <span class="citation">Lovell et al. (2015)</span> is based on log-ratio variance <span class="math inline">\(var(log \frac{x_i}{x_j})\)</span>, which is zero when <span class="math inline">\(\omega_i\)</span> and <span class="math inline">\(\omega_j\)</span> are perfectly proportional. Proportionality is implemented in the R package <a href="https://github.com/tpq/propr">propr</a>. <span class="citation">Badri et al. (2020)</span> extend the proportionality measure to a so-called shrinkage proportionality estimator. It combines proportionality with the covariance shrinkage approach to obtain consistent association estimates even with small sample sizes.</li>
</ul>
</div>
<div id="comparison-of-association-measures" class="section level2 hasAnchor" number="12.6">
<h2><span class="header-section-number">12.6</span> Comparison of association measures<a href="network-learning.html#comparison-of-association-measures" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we provide three additional examples for constructing a network using each of the three types of association:</p>
<ul>
<li>Correlation using <code>SparCC</code></li>
<li>Partial correlation using <code>SpiecEasi</code></li>
<li>Proportionality using the shrinkage proportionality measure</li>
</ul>
<div id="sparcc-correlation" class="section level3 hasAnchor" number="12.6.1">
<h3><span class="header-section-number">12.6.1</span> SparCC<a href="network-learning.html#sparcc-correlation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The first association measure we look at is SparCC (Sparse Correlations for Compositional data), introduced by <span class="citation">J. Friedman and Alm (2012)</span>. It estimates Pearson correlations while taking into account the compositional structure of the data. The <a href="https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html">SpiecEasi</a> package provides an implementation of this method.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="network-learning.html#cb402-1" tabindex="-1"></a><span class="co"># Set seed for reproducibility</span></span>
<span id="cb402-2"><a href="network-learning.html#cb402-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13075</span>)</span>
<span id="cb402-3"><a href="network-learning.html#cb402-3" tabindex="-1"></a><span class="co"># Compute correlation matrix</span></span>
<span id="cb402-4"><a href="network-learning.html#cb402-4" tabindex="-1"></a>sparcc_cor <span class="ot">&lt;-</span> SpiecEasi<span class="sc">::</span><span class="fu">sparcc</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>)))<span class="sc">$</span>Cor</span>
<span id="cb402-5"><a href="network-learning.html#cb402-5" tabindex="-1"></a><span class="fu">rownames</span>(sparcc_cor) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sparcc_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(tse)</span></code></pre></div>
<p>We reuse the <code>transform_asso()</code> function created in Section <a href="network-learning.html#spring-network">12.1.4</a>, which sparsifies the association matrix, transforms it into a similarity matrix, and finally returns an <code>igraph</code> object.</p>
<p>Two threshold values are used to see the effect of sparsification later in the network plot.</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="network-learning.html#cb403-1" tabindex="-1"></a>sparcc_trans03 <span class="ot">&lt;-</span> <span class="fu">transform_asso</span>(sparcc_cor, <span class="at">thresh =</span> <span class="fl">0.3</span>)</span>
<span id="cb403-2"><a href="network-learning.html#cb403-2" tabindex="-1"></a>sparcc_trans04 <span class="ot">&lt;-</span> <span class="fu">transform_asso</span>(sparcc_cor, <span class="at">thresh =</span> <span class="fl">0.4</span>)</span>
<span id="cb403-3"><a href="network-learning.html#cb403-3" tabindex="-1"></a></span>
<span id="cb403-4"><a href="network-learning.html#cb403-4" tabindex="-1"></a>sparcc_graph03 <span class="ot">&lt;-</span> sparcc_trans03<span class="sc">$</span>graph</span>
<span id="cb403-5"><a href="network-learning.html#cb403-5" tabindex="-1"></a>sparcc_graph04 <span class="ot">&lt;-</span> sparcc_trans04<span class="sc">$</span>graph</span></code></pre></div>
</div>
<div id="shrinkage-prop" class="section level3 hasAnchor" number="12.6.2">
<h3><span class="header-section-number">12.6.2</span> Shrinkage proportionality<a href="network-learning.html#shrinkage-prop" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the second example, microbial associations are measured by proportionality, originally introduced by <span class="citation">Lovell et al. (2015)</span>. We use the shrinkage proportionality estimator proposed by <span class="citation">Badri et al. (2020)</span>, which gives consistent results even for small sample sizes. Since there is no R package implementing this estimator, we use the <code>rho_shrink_est()</code> function provided in the <a href="https://github.com/MichelleBadri/NormCorr-manuscript/blob/master/code/helpers/norm_functions.R">GitHub repository</a> associated with the paper. The function is slightly modified to take normalized counts as input.</p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb404-1"><a href="network-learning.html#cb404-1" tabindex="-1"></a><span class="fu">library</span>(corpcor)</span></code></pre></div>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="network-learning.html#cb405-1" tabindex="-1"></a><span class="co"># norm_counts: clr-transformed count matrix with samples in rows</span></span>
<span id="cb405-2"><a href="network-learning.html#cb405-2" tabindex="-1"></a>rho_shrink_est <span class="ot">&lt;-</span> <span class="cf">function</span>(norm_counts, ...) {</span>
<span id="cb405-3"><a href="network-learning.html#cb405-3" tabindex="-1"></a>  shrunk_cov <span class="ot">&lt;-</span> <span class="fu">cov.shrink</span>(norm_counts, ...)</span>
<span id="cb405-4"><a href="network-learning.html#cb405-4" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ncol</span>(norm_counts)</span>
<span id="cb405-5"><a href="network-learning.html#cb405-5" tabindex="-1"></a>  J <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rep</span>(<span class="fu">diag</span>(shrunk_cov), p), p)</span>
<span id="cb405-6"><a href="network-learning.html#cb405-6" tabindex="-1"></a>  rho <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> shrunk_cov <span class="sc">/</span> (J <span class="sc">+</span> <span class="fu">t</span>(J))</span>
<span id="cb405-7"><a href="network-learning.html#cb405-7" tabindex="-1"></a>  (rho <span class="sc">+</span> <span class="fu">t</span>(rho)) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb405-8"><a href="network-learning.html#cb405-8" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb406-1"><a href="network-learning.html#cb406-1" tabindex="-1"></a><span class="co"># Apply the shrinkage proportionality estimator to the clr-transformed counts</span></span>
<span id="cb406-2"><a href="network-learning.html#cb406-2" tabindex="-1"></a>prop_est <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">rho_shrink_est</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;clr&quot;</span>))), <span class="st">&quot;matrix&quot;</span>)</span></code></pre></div>
<pre><code>## Estimating optimal shrinkage intensity lambda.var (variance vector): 0.0857 
## 
## Estimating optimal shrinkage intensity lambda (correlation matrix): 0.3634</code></pre>
<p>Again, we use our transformation function to convert the association matrix into a graph object.</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb408-1"><a href="network-learning.html#cb408-1" tabindex="-1"></a>prop_trans <span class="ot">&lt;-</span> <span class="fu">transform_asso</span>(prop_est, <span class="at">thresh =</span> <span class="fl">0.4</span>)</span>
<span id="cb408-2"><a href="network-learning.html#cb408-2" tabindex="-1"></a>prop_graph <span class="ot">&lt;-</span> prop_trans<span class="sc">$</span>graph</span></code></pre></div>
</div>
<div id="spieceasi-mb" class="section level3 hasAnchor" number="12.6.3">
<h3><span class="header-section-number">12.6.3</span> SpiecEasi - MB<a href="network-learning.html#spieceasi-mb" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As third example, we use the SpiecEasi (Sparse InversE Covariance estimation for Ecological Association and Statistical Inference) approach proposed by <span class="citation">Kurtz et al. (2015)</span> to estimate a sparse conditional dependency graph. The neighborhood selection method (MB) introduced by <span class="citation">Meinshausen and Bhlmann (2006)</span> is used for network learning. The approach is implemented in the R package <a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a>.</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="network-learning.html#cb409-1" tabindex="-1"></a><span class="fu">library</span>(SpiecEasi)</span></code></pre></div>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="network-learning.html#cb410-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13075</span>)</span>
<span id="cb410-2"><a href="network-learning.html#cb410-2" tabindex="-1"></a>se_mb_est <span class="ot">&lt;-</span> <span class="fu">spiec.easi</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;counts&quot;</span>)), </span>
<span id="cb410-3"><a href="network-learning.html#cb410-3" tabindex="-1"></a>                        <span class="at">method =</span> <span class="st">&#39;mb&#39;</span>, <span class="at">nlambda =</span> <span class="dv">20</span>, </span>
<span id="cb410-4"><a href="network-learning.html#cb410-4" tabindex="-1"></a>                        <span class="at">pulsar.params =</span> <span class="fu">list</span>(<span class="at">rep.num =</span> <span class="dv">20</span>))</span></code></pre></div>
<p>Since <code>SpiecEasi</code> uses the StARS (Stability Approach to Regularization Selection) method <span class="citation">(H. Liu, Roeder, and Wasserman 2010)</span> to obtain a sparse association matrix, we dont need to set a threshold here. We store the partial correlations corresponding to the StARS-optimal lambda and convert them into an <code>igraph</code> object.</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb411-1"><a href="network-learning.html#cb411-1" tabindex="-1"></a><span class="co"># Get optimal matrix with partial correlations</span></span>
<span id="cb411-2"><a href="network-learning.html#cb411-2" tabindex="-1"></a>se_mb_cor <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">getOptBeta</span>(se_mb_est))</span>
<span id="cb411-3"><a href="network-learning.html#cb411-3" tabindex="-1"></a>se_mb_cor <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">symBeta</span>(se_mb_cor))</span>
<span id="cb411-4"><a href="network-learning.html#cb411-4" tabindex="-1"></a><span class="fu">rownames</span>(se_mb_cor) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(se_mb_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(tse)</span>
<span id="cb411-5"><a href="network-learning.html#cb411-5" tabindex="-1"></a><span class="fu">diag</span>(se_mb_cor) <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb411-6"><a href="network-learning.html#cb411-6" tabindex="-1"></a></span>
<span id="cb411-7"><a href="network-learning.html#cb411-7" tabindex="-1"></a><span class="co"># Create graph object</span></span>
<span id="cb411-8"><a href="network-learning.html#cb411-8" tabindex="-1"></a>se_mb_graph <span class="ot">&lt;-</span> <span class="fu">transform_asso</span>(se_mb_cor)<span class="sc">$</span>graph</span></code></pre></div>
</div>
<div id="network-plots" class="section level3 hasAnchor" number="12.6.4">
<h3><span class="header-section-number">12.6.4</span> Network plots<a href="network-learning.html#network-plots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The graph objects can now be plotted using the <a href="https://r.igraph.org/">igraph</a> package. The same layout is used in all four plots so that the networks are comparable.</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="network-learning.html#cb412-1" tabindex="-1"></a><span class="fu">library</span>(igraph)</span></code></pre></div>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb413-1"><a href="network-learning.html#cb413-1" tabindex="-1"></a><span class="co"># Node sizes</span></span>
<span id="cb413-2"><a href="network-learning.html#cb413-2" tabindex="-1"></a>vsize <span class="ot">&lt;-</span> (<span class="fu">colMeans</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;log10&quot;</span>))) <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">*</span> <span class="dv">3</span></span>
<span id="cb413-3"><a href="network-learning.html#cb413-3" tabindex="-1"></a></span>
<span id="cb413-4"><a href="network-learning.html#cb413-4" tabindex="-1"></a><span class="co"># Use Fruchterman-Reingold (force-directed) layout</span></span>
<span id="cb413-5"><a href="network-learning.html#cb413-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13075</span>)</span>
<span id="cb413-6"><a href="network-learning.html#cb413-6" tabindex="-1"></a>lay_fr <span class="ot">&lt;-</span> <span class="fu">layout_with_fr</span>(se_mb_graph)</span>
<span id="cb413-7"><a href="network-learning.html#cb413-7" tabindex="-1"></a></span>
<span id="cb413-8"><a href="network-learning.html#cb413-8" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb413-9"><a href="network-learning.html#cb413-9" tabindex="-1"></a><span class="fu">plot</span>(sparcc_graph03, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, </span>
<span id="cb413-10"><a href="network-learning.html#cb413-10" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="st">&quot;SparCC (thresh 0.3)&quot;</span>)</span>
<span id="cb413-11"><a href="network-learning.html#cb413-11" tabindex="-1"></a><span class="fu">plot</span>(sparcc_graph04, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, </span>
<span id="cb413-12"><a href="network-learning.html#cb413-12" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="st">&quot;SparCC (thresh 0.4)&quot;</span>)</span>
<span id="cb413-13"><a href="network-learning.html#cb413-13" tabindex="-1"></a><span class="fu">plot</span>(prop_graph, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, </span>
<span id="cb413-14"><a href="network-learning.html#cb413-14" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="st">&quot;Shrinkage proportionality</span><span class="sc">\n</span><span class="st">(thresh 0.4)&quot;</span>)</span>
<span id="cb413-15"><a href="network-learning.html#cb413-15" tabindex="-1"></a><span class="fu">plot</span>(se_mb_graph, <span class="at">layout =</span> lay_fr, <span class="at">vertex.size =</span> vsize, </span>
<span id="cb413-16"><a href="network-learning.html#cb413-16" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>, <span class="at">main =</span> <span class="st">&quot;SpiecEasi (MB)&quot;</span>)</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/netplots-1.png" width="960" /></p>
<p><strong>A view observations:</strong><br />
The density of SparCC (threshold 0.4), proportionality and SpiecEasi is comparable, while the SparCC correlation network with threshold 0.3 is much denser. However, there are edges in the proportionality and SpiecEasi networks that are not present in the two SparCC networks. The SpiecEasi network has less highly connected nodes than the other three networks, but more nodes with one or two connections.</p>
<p>We will look at the degree distribution in the next section to quantify these observations.</p>
</div>
<div id="network-analysis-1" class="section level3 hasAnchor" number="12.6.5">
<h3><span class="header-section-number">12.6.5</span> Network analysis<a href="network-learning.html#network-analysis-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here we repeat some of the network analysis approaches explained in Section <a href="network-learning.html#network-analysis">12.2</a>. The analyses are performed simultaneously for the three association measures as well as the SPRING network constructed in Section <a href="network-learning.html#spring-network">12.1.4</a>. Therefore, we start by creating a list of all the graph objects we need for the analyses.</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="network-learning.html#cb414-1" tabindex="-1"></a>graphlist <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">SparCC =</span> sparcc_graph04, </span>
<span id="cb414-2"><a href="network-learning.html#cb414-2" tabindex="-1"></a>                  <span class="at">Proportionality =</span> prop_graph,</span>
<span id="cb414-3"><a href="network-learning.html#cb414-3" tabindex="-1"></a>                  <span class="at">SpiecEasi =</span> se_mb_graph,</span>
<span id="cb414-4"><a href="network-learning.html#cb414-4" tabindex="-1"></a>                  <span class="at">SPRING =</span> spring_graph)</span></code></pre></div>
<div id="degree-distribution-1" class="section level4 hasAnchor" number="12.6.5.1">
<h4><span class="header-section-number">12.6.5.1</span> Degree distribution<a href="network-learning.html#degree-distribution-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The degree distribution is plotted for all four measures to compare the overall network structure.</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="network-learning.html#cb415-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="network-learning.html#cb416-1" tabindex="-1"></a><span class="co"># Compute degree distributions</span></span>
<span id="cb416-2"><a href="network-learning.html#cb416-2" tabindex="-1"></a>ddlist <span class="ot">&lt;-</span> <span class="fu">lapply</span>(graphlist, igraph<span class="sc">::</span>degree.distribution)</span>
<span id="cb416-3"><a href="network-learning.html#cb416-3" tabindex="-1"></a></span>
<span id="cb416-4"><a href="network-learning.html#cb416-4" tabindex="-1"></a><span class="co"># Maximum degree</span></span>
<span id="cb416-5"><a href="network-learning.html#cb416-5" tabindex="-1"></a>maxdeg <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">lengths</span>(ddlist))</span>
<span id="cb416-6"><a href="network-learning.html#cb416-6" tabindex="-1"></a></span>
<span id="cb416-7"><a href="network-learning.html#cb416-7" tabindex="-1"></a><span class="co"># Make list elements the same length</span></span>
<span id="cb416-8"><a href="network-learning.html#cb416-8" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(graphlist)) {</span>
<span id="cb416-9"><a href="network-learning.html#cb416-9" tabindex="-1"></a>  <span class="fu">length</span>(ddlist[[i]]) <span class="ot">&lt;-</span> maxdeg</span>
<span id="cb416-10"><a href="network-learning.html#cb416-10" tabindex="-1"></a>}</span>
<span id="cb416-11"><a href="network-learning.html#cb416-11" tabindex="-1"></a></span>
<span id="cb416-12"><a href="network-learning.html#cb416-12" tabindex="-1"></a><span class="co"># Data frame needed for ggplot2</span></span>
<span id="cb416-13"><a href="network-learning.html#cb416-13" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Degree =</span> <span class="fu">rep</span>(<span class="fu">seq_len</span>(maxdeg), <span class="fu">length</span>(graphlist)), </span>
<span id="cb416-14"><a href="network-learning.html#cb416-14" tabindex="-1"></a>                 <span class="at">Fraction =</span> <span class="fu">unlist</span>(ddlist), </span>
<span id="cb416-15"><a href="network-learning.html#cb416-15" tabindex="-1"></a>                 <span class="at">Method =</span> <span class="fu">rep</span>(<span class="fu">names</span>(graphlist), <span class="at">each =</span> maxdeg))</span>
<span id="cb416-16"><a href="network-learning.html#cb416-16" tabindex="-1"></a></span>
<span id="cb416-17"><a href="network-learning.html#cb416-17" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> Degree, <span class="at">y =</span> Fraction, <span class="at">group =</span> Method)) <span class="sc">+</span></span>
<span id="cb416-18"><a href="network-learning.html#cb416-18" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">color =</span> Method)) <span class="sc">+</span></span>
<span id="cb416-19"><a href="network-learning.html#cb416-19" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> Method)) <span class="sc">+</span></span>
<span id="cb416-20"><a href="network-learning.html#cb416-20" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
<p>The SparCC and shrinkage proportionality networks have a considerably higher proportion of singletons (zero-degree nodes) than the two conditional dependency graphs, but a lower proportion of nodes with degrees between one and five. The SpiecEasi and SPRING graphs, on the other hand, have a higher proportion of low degree nodes, but no highly connected nodes with a degree greater than eleven.</p>
</div>
<div id="clustered-heatmaps-1" class="section level4 hasAnchor" number="12.6.5.2">
<h4><span class="header-section-number">12.6.5.2</span> Clustered heatmaps<a href="network-learning.html#clustered-heatmaps-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Using the <code>ComplexHeatmap</code> package, we plot heatmaps of the association matrices for the four considered association measures. Rows and columns are sorted according to the clusters identified via hierarchical clustering.</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb417-1"><a href="network-learning.html#cb417-1" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb417-2"><a href="network-learning.html#cb417-2" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb417-3"><a href="network-learning.html#cb417-3" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code></pre></div>
<p>For each association measure, we select the 50 nodes with the highest sum of edge weights.</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb418-1"><a href="network-learning.html#cb418-1" tabindex="-1"></a><span class="co"># Function for selecting taxa with highest sum of edge weights</span></span>
<span id="cb418-2"><a href="network-learning.html#cb418-2" tabindex="-1"></a>select_taxa <span class="ot">&lt;-</span> <span class="cf">function</span>(adja, <span class="at">ntaxa =</span> <span class="dv">50</span>) {</span>
<span id="cb418-3"><a href="network-learning.html#cb418-3" tabindex="-1"></a>  sel <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">rowSums</span>(adja), <span class="at">decreasing =</span> <span class="cn">TRUE</span>))[<span class="fu">seq_len</span>(ntaxa)]</span>
<span id="cb418-4"><a href="network-learning.html#cb418-4" tabindex="-1"></a>  adja[sel, sel]</span>
<span id="cb418-5"><a href="network-learning.html#cb418-5" tabindex="-1"></a>}</span>
<span id="cb418-6"><a href="network-learning.html#cb418-6" tabindex="-1"></a></span>
<span id="cb418-7"><a href="network-learning.html#cb418-7" tabindex="-1"></a>assolist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb418-8"><a href="network-learning.html#cb418-8" tabindex="-1"></a>assolist<span class="sc">$</span>SparCC <span class="ot">&lt;-</span> <span class="fu">select_taxa</span>(sparcc_trans04<span class="sc">$</span>adja)</span>
<span id="cb418-9"><a href="network-learning.html#cb418-9" tabindex="-1"></a>assolist<span class="sc">$</span>Proportionality <span class="ot">&lt;-</span> <span class="fu">select_taxa</span>(prop_trans<span class="sc">$</span>adja)</span>
<span id="cb418-10"><a href="network-learning.html#cb418-10" tabindex="-1"></a>assolist<span class="sc">$</span>SpiecEasi <span class="ot">&lt;-</span> <span class="fu">select_taxa</span>(se_mb_cor)</span>
<span id="cb418-11"><a href="network-learning.html#cb418-11" tabindex="-1"></a>assolist<span class="sc">$</span>SPRING <span class="ot">&lt;-</span> <span class="fu">select_taxa</span>(spring_cor)</span></code></pre></div>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="network-learning.html#cb419-1" tabindex="-1"></a><span class="co"># Color vector for the legend</span></span>
<span id="cb419-2"><a href="network-learning.html#cb419-2" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), </span>
<span id="cb419-3"><a href="network-learning.html#cb419-3" tabindex="-1"></a>                  <span class="fu">c</span>(<span class="st">&quot;royalblue4&quot;</span>, <span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;orange&quot;</span>, <span class="st">&quot;firebrick3&quot;</span>))</span>
<span id="cb419-4"><a href="network-learning.html#cb419-4" tabindex="-1"></a></span>
<span id="cb419-5"><a href="network-learning.html#cb419-5" tabindex="-1"></a>hm_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb419-6"><a href="network-learning.html#cb419-6" tabindex="-1"></a></span>
<span id="cb419-7"><a href="network-learning.html#cb419-7" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(assolist)) {</span>
<span id="cb419-8"><a href="network-learning.html#cb419-8" tabindex="-1"></a>  <span class="cf">if</span> (i <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>)) {</span>
<span id="cb419-9"><a href="network-learning.html#cb419-9" tabindex="-1"></a>    showlegend <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb419-10"><a href="network-learning.html#cb419-10" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb419-11"><a href="network-learning.html#cb419-11" tabindex="-1"></a>    showlegend <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb419-12"><a href="network-learning.html#cb419-12" tabindex="-1"></a>  }</span>
<span id="cb419-13"><a href="network-learning.html#cb419-13" tabindex="-1"></a>  </span>
<span id="cb419-14"><a href="network-learning.html#cb419-14" tabindex="-1"></a>  hm_list[[i]] <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(assolist[[i]], </span>
<span id="cb419-15"><a href="network-learning.html#cb419-15" tabindex="-1"></a>                          <span class="at">col =</span> col, </span>
<span id="cb419-16"><a href="network-learning.html#cb419-16" tabindex="-1"></a>                          <span class="at">rect_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">&quot;gray&quot;</span>, <span class="at">lwd =</span> <span class="dv">1</span>),</span>
<span id="cb419-17"><a href="network-learning.html#cb419-17" tabindex="-1"></a>                          <span class="at">show_row_names =</span> <span class="cn">FALSE</span>, </span>
<span id="cb419-18"><a href="network-learning.html#cb419-18" tabindex="-1"></a>                          <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb419-19"><a href="network-learning.html#cb419-19" tabindex="-1"></a>                          <span class="at">column_title =</span> <span class="fu">names</span>(assolist)[i], </span>
<span id="cb419-20"><a href="network-learning.html#cb419-20" tabindex="-1"></a>                          <span class="at">name =</span> <span class="st">&quot;Association&quot;</span>,</span>
<span id="cb419-21"><a href="network-learning.html#cb419-21" tabindex="-1"></a>                          <span class="at">show_heatmap_legend =</span> showlegend) <span class="sc">%&gt;%</span> </span>
<span id="cb419-22"><a href="network-learning.html#cb419-22" tabindex="-1"></a>    <span class="fu">draw</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb419-23"><a href="network-learning.html#cb419-23" tabindex="-1"></a>    <span class="fu">grid.grabExpr</span>()</span>
<span id="cb419-24"><a href="network-learning.html#cb419-24" tabindex="-1"></a>}</span>
<span id="cb419-25"><a href="network-learning.html#cb419-25" tabindex="-1"></a></span>
<span id="cb419-26"><a href="network-learning.html#cb419-26" tabindex="-1"></a><span class="co"># Plot with wrap_plots() function from patchwork package</span></span>
<span id="cb419-27"><a href="network-learning.html#cb419-27" tabindex="-1"></a><span class="fu">wrap_plots</span>(hm_list, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">8</span>, <span class="dv">10</span>))</span></code></pre></div>
<p><img src="60_network_learning_files/figure-html/clustered_heatmap_all-1.png" width="960" /></p>
<p>The SparCC and the proportionality network show a block structure, where each block corresponds to a cluster. The clusters are less pronounced in the conditional dependence networks. The latter also generally have lower edge weights.</p>
</div>
<div id="global-network-measures-1" class="section level4 hasAnchor" number="12.6.5.3">
<h4><span class="header-section-number">12.6.5.3</span> Global network measures<a href="network-learning.html#global-network-measures-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For each association measure, the three global network measures density, transitivity, and average path length are computed and stored in a data frame for comparison.</p>
<div class="sourceCode" id="cb420"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb420-1"><a href="network-learning.html#cb420-1" tabindex="-1"></a><span class="co"># Compute density and store in a data frame</span></span>
<span id="cb420-2"><a href="network-learning.html#cb420-2" tabindex="-1"></a>glob <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Density =</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(graphlist, edge_density)))</span>
<span id="cb420-3"><a href="network-learning.html#cb420-3" tabindex="-1"></a></span>
<span id="cb420-4"><a href="network-learning.html#cb420-4" tabindex="-1"></a><span class="co"># Transitivity</span></span>
<span id="cb420-5"><a href="network-learning.html#cb420-5" tabindex="-1"></a>glob<span class="sc">$</span>Transitivity <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(graphlist, transitivity))</span>
<span id="cb420-6"><a href="network-learning.html#cb420-6" tabindex="-1"></a></span>
<span id="cb420-7"><a href="network-learning.html#cb420-7" tabindex="-1"></a><span class="co"># Average path length</span></span>
<span id="cb420-8"><a href="network-learning.html#cb420-8" tabindex="-1"></a>glob<span class="sc">$</span>Av.path <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(graphlist, average.path.length))</span>
<span id="cb420-9"><a href="network-learning.html#cb420-9" tabindex="-1"></a></span>
<span id="cb420-10"><a href="network-learning.html#cb420-10" tabindex="-1"></a>glob</span></code></pre></div>
<pre><code>##                 Density Transitivity Av.path
## SparCC          0.01510       0.5458  0.8932
## Proportionality 0.01295       0.5402  1.1847
## SpiecEasi       0.02162       0.2145  1.6322
## SPRING          0.01444       0.1456  1.6993</code></pre>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differential-abundance.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="network-comparison.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/60_network_learning.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/60_network_learning.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/60_network_learning.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
