% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{book}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{booktabs}
% Solution from:
% https://github.com/ulyngs/oxforddown/commit/ee17c2fff6d5df37df972936dd9df25866c088bb

\usepackage{fvextra}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage[]{natbib}
\bibliographystyle{apalike}
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Orchestrating Microbiome Analysis with Bioconductor},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Orchestrating Microbiome Analysis with Bioconductor}
\author{}
\date{\vspace{-2.5em}\textbf{Authors:} Leo Lahti {[}aut{]}, Tuomas Borman {[}aut, cre{]}, Felix GM Ernst {[}aut{]}, and others (see the full list of contributors) {[}ctb{]} \textbf{Version:} 0.98.16 \textbf{Modified:} 2023-07-29 \textbf{Compiled:} 2023-12-03 \textbf{Environment:} R version 4.3.1 (2023-06-16), Bioconductor 3.17 \textbf{License:} CC BY-NC-SA 3.0 US \textbf{Copyright:} \textbf{Source:} \url{https://github.com/microbiome/OMA}}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{welcome}{%
\chapter*{Welcome}\label{welcome}}
\addcontentsline{toc}{chapter}{Welcome}

You are reading the online book, \href{https://microbiome.github.io/OMA/}{\textbf{Orchestrating Microbiome Analysis
with Bioconductor}} \citep{OMA}, where we
walk through common strategies and workflows in microbiome data
science.

The book shows through concrete examples how you can take advantage of
the latest developments in R/Bioconductor for the manipulation,
analysis, and reproducible reporting of hierarchical and heterogeneous
microbiome profiling data sets. The book was borne out of necessity,
while updating microbiome analysis tools to work with Bioconductor
classes that provide support for multi-modal data collections. Many of
these techniques are generic and widely applicable in other contexts
as well.

This work has been heavily influenced by other similar resources, in
particular the Orchestrating Single-Cell Analysis with Bioconductor
\citep{Amezquita2020}, \href{http://joey711.github.io/phyloseq/tutorials-index}{phyloseq
tutorials}
\citep{Callahan2016} and \href{https://microbiome.github.io/tutorials/}{microbiome
tutorials} \citep{Shetty2019}.
This book extends these resources to teach the grammar of Bioconductor
workflows in the context of microbiome data science. As such, it
supports the adoption of general skills in the analysis of large,
hierarchical, and multi-modal data collections. We focus on microbiome
analysis tools, including entirely new, partially updated as well as
previously established methods.

This online resource and its associated ecosystem of microbiome data
science tools are a result of a community-driven development process,
and welcoming new contributors. Several individuals have
\href{https://github.com/microbiome/OMA/graphs/contributors}{contributed}
methods, workflows and improvements as acknowledged in the
Introduction. You can find more information on how to find us online
and join the developer community through the project homepage at
\href{https://microbiome.github.io}{microbiome.github.io}. This online
resource has been written in RMarkdown with the bookdown R
package. The material is \textbf{free to use} with the \href{https://creativecommons.org/licenses/by-nc/3.0/us/}{Creative Commons
Attribution-NonCommercial
3.0} License.

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

\hypertarget{part-introduction}{%
\part{Introduction}\label{part-introduction}}

\hypertarget{intro}{%
\chapter{Introduction}\label{intro}}

This work - \href{https://microbiome.github.io/OMA/}{\textbf{Orchestrating Microbiome Analysis with Bioconductor}} \citep{OMA} - contributes novel
methods and educational resources for microbiome data science. It
aims to teach the grammar of Bioconductor workflows in the context of
microbiome data science. We show through concrete examples how to use
the latest developments and data analytical strategies in
R/Bioconductor for the manipulation, analysis, and reproducible
reporting of hierarchical, heterogeneous, and multi-modal microbiome
profiling data. The data science methodology is tightly integrated
with the broader R/Bioconductor ecosystem that focuses on the
development of high-quality open research software for life
sciences (\citet{Gentleman2004}, \citet{Huber2015}). The support for modularity and
interoperability is a key to efficient resource sharing and
collaborative development both within and across research fields. The
central data infrastructure, the \texttt{SummarizedExperiment} data container
and its derivatives, have already been widely adopted in microbiome
research, single cell sequencing, and in other fields, allowing a
rapid adoption and extensions of emerging data science techniques
across application domains.

We assume that the reader is already familiar with R programming. For
references and tips on introductory material for R and Bioconductor,
see Chapter \ref{resources}. This online resource and its associated
ecosystem of microbiome data science tools are a result of a
community-driven development process, and welcoming new users and
contributors. You can find more information on how to find us online
and join the developer community through the project homepage at
\href{https://microbiome.github.io}{microbiome.github.io}.

The book is organized into three parts. We start by introducing the
material and link to further resources for learning R and
Bioconductor. We describe the key data infrastructure, the
\texttt{TreeSummarizedExperiment} class that provides a container for
microbiome data, and how to get started by loading microbiome data set
in the context of this new framework. The second section, \emph{Focus
Topics}, covers the common steps in microbiome data analysis,
beginning with the most common steps and progressing to more
specialized methods in subsequent sections. Third, \emph{Workflows},
provides case studies for the various datasets used throughout the
book. Finally, \emph{Appendix}, links to further resources and
acknowledgments.

\hypertarget{packages}{%
\chapter{Packages}\label{packages}}

The Bioconductor microbiome data science framework consists of:

\begin{itemize}
\tightlist
\item
  \textbf{data containers}, designed to organize multi-assay microbiome data
\item
  \textbf{R/Bioconductor packages} that provide dedicated methods
\item
  \textbf{community} of users and developers
\end{itemize}

This section provides an overview of the package ecosystem. Section
\ref{example-data} links to various open microbiome data resources
that support this framework.

\hypertarget{package-installation}{%
\section{Package installation}\label{package-installation}}

You can install all packages that are required to run every example in this book via the following command:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{source}\NormalTok{(}\StringTok{"https://raw.githubusercontent.com/microbiome/OMA/master/install\_packages.R"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{packages_specific}{%
\subsection{Installing specific packages}\label{packages_specific}}

You can install R packages of your choice with the following command
line procedure.

\textbf{Bioconductor development version} requires the installation of the
latest R beta version. This is primarily recommended for those who
already have experience with R/Bioconductor and need access to the
latest updates.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{BiocManager}\SpecialCharTok{::}\FunctionTok{install}\NormalTok{(}\StringTok{"microbiome/mia"}\NormalTok{, }\AttributeTok{version=}\StringTok{"devel"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbf{Github development version} provides access to the latest but
potentially unstable features. This is useful when you want access to
all available tools.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{devtools}\SpecialCharTok{::}\FunctionTok{install\_github}\NormalTok{(}\StringTok{"microbiome/mia"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{ecosystem}{%
\section{Package ecosystem}\label{ecosystem}}

Methods for \texttt{(Tree)SummarizedExperiment} and \texttt{MultiAssayExperiment}
data containers are provided by multiple independent developers
through R/Bioconductor packages. Some of these are listed below (tips
on new packages are \href{https://microbiome.github.io}{welcome}).

\hypertarget{mia-package-family}{%
\subsection{mia package family}\label{mia-package-family}}

The mia package family provides general methods for microbiome data wrangling, analysis and visualization.

\begin{itemize}
\tightlist
\item
  \href{https://microbiome.github.io/mia/}{mia}: Microbiome analysis tools \citep{R_mia}
\item
  \href{https://microbiome.github.io/miaViz/}{miaViz}: Microbiome analysis specific visualization \citep{Ernst2022}
\item
  \href{https://microbiome.github.io/miaSim/}{miaSim}: Microbiome data simulations \citep{Simsek2021}
\item
  \href{https://microbiome.github.io/miaTime/}{miaTime}: Microbiome time series analysis \citep{Lahti2021}
\end{itemize}

\hypertarget{sub-diff-abund}{%
\subsection{Differential abundance}\label{sub-diff-abund}}

The following DA methods support \texttt{(Tree)SummarizedExperiment}.

\begin{itemize}
\tightlist
\item
  \href{https://bioconductor.org/packages/devel/bioc/html/ANCOMBC.html}{ANCOMBC} for differential abundance analysis
\item
  \href{https://bioconductor.org/packages/release/bioc/vignettes/benchdamic/inst/doc/intro.html}{benchdamic} for benchmarking differential abundance methods
\item
  \href{https://www.bioconductor.org/packages/release/bioc/html/ALDEx2.html}{ALDEx2} for differential abundance analysis
\end{itemize}

\hypertarget{other-packages}{%
\subsection{Other packages}\label{other-packages}}

\begin{itemize}
\tightlist
\item
  \href{http://bioconductor.org/packages/devel/bioc/html/philr.html}{philr} (\citet{Silverman2017}) phylogeny-aware phILR transformation
\item
  \href{https://bioconductor.org/packages/release/bioc/html/MicrobiotaProcess.html}{MicrobiotaProcess} for ``tidy'' analysis of microbiome and other ecological data
\item
  \href{https://microsud.github.io/Tools-Microbiome-Analysis/}{Tools for Microbiome
  Analysis}
  site listed over 130 R packages for microbiome data science in

  \begin{enumerate}
  \def\labelenumi{\arabic{enumi}.}
  \setcounter{enumi}{2022}
  \tightlist
  \item
    Many of these are not in Bioconductor, or do not directly
    support the data containers used in this book but can be often used
    with minor modifications. The phyloseq-based tools can be used by
    converting the TreeSE data into phyloseq with
    \texttt{makePhyloseqFromTreeSummarizedExperiment}.
  \end{enumerate}
\end{itemize}

\hypertarget{open-microbiome-data}{%
\subsection{Open microbiome data}\label{open-microbiome-data}}

Hundreds of published microbiome data sets are readily available in
these data containers (see \ref{example-data}).

\hypertarget{containers}{%
\chapter{Microbiome Data}\label{containers}}

\hypertarget{data-science-framework}{%
\section{Data science framework}\label{data-science-framework}}

The building blocks of the framework are \textbf{data container}
(SummarizedExperiment and its derivatives), \textbf{packages} from various
developers using the TreeSE container, open \textbf{demonstration data
sets}, in a separate chapter \ref{example-data}, and \textbf{online
tutorials} including this online book as well as the various package
vignettes and other materials.

\includegraphics[width=18.67in]{general/figures/FigureOverviewV2_mod}

\hypertarget{data-containers}{%
\section{Data containers}\label{data-containers}}

\texttt{SummarizedExperiment} (\texttt{SE}) \citep{R_SummarizedExperiment} is a generic and highly optimized container for complex data
structures. It has become a common choice for analysing various types
of biomedical profiling data, such as RNAseq, ChIp-Seq, microarrays,
flow cytometry, proteomics, and single-cell
sequencing.

{[}\texttt{TreeSummarizedExperiment}{]} (\texttt{TreeSE}) \citep{R_TreeSummarizedExperiment} was developed as an extension to incorporate hierarchical
information (such as phylogenetic trees and sample hierarchies) and
reference sequences.

{[}\texttt{MultiAssayExperiment}{]} (\texttt{MAE}) \citep{Ramos2017} provides an organized
way to bind several different data containers together in a single
object. For example, we can bind microbiome data (in \texttt{TreeSE}
container) with metabolomic profiling data (in \texttt{SE}) container, with
(partially) shared sample metadata. This is convenient and robust for
instance in subsetting and other data manipulation tasks. Microbiome
data can be part of multiomics experiments and analysis strategies. We
highlight how the methods used througout in this book relate to this
data framework by using the \texttt{TreeSummarizedExperiment},
\texttt{MultiAssayExperiment}, and classes beyond.

This section provides an introductions to these data containers. In
microbiome data science, these containers link taxonomic abundance
tables with rich side information on the features and
samples. Taxonomic abundance data can be obtained by 16S rRNA amplicon
or metagenomic sequencing, phylogenetic microarrays, or by other
means. Many microbiome experiments include multiple versions and types
of data generated independently or derived from each other through
transformation or agglomeration. We start by providing recommendations
on how to represent different varieties of multi-table data within the
\texttt{TreeSummarizedExperiment} class.

The options and recommendations are summarized in Table \ref{tab:options}.

\hypertarget{assay-slot}{%
\subsection{Assay data}\label{assay-slot}}

The original count-based taxonomic abundance tables may have different
transformations, such as logarithmic, Centered Log-Ratio (CLR), or relative
abundance. These are typically stored in \emph{\textbf{assays}}.

Let us load example data and rename it as tse.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"hitchip1006"}\NormalTok{, }\AttributeTok{package =} \StringTok{"miaTime"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ hitchip1006}
\end{Highlighting}
\end{Shaded}

The \texttt{assays} slot contains the experimental data as multiple count matrices. The result of \texttt{assays} is a list of matrices.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assays}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of length 1
## names(1): counts
\end{verbatim}

Individual assays can be accessed via \texttt{assay}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{,}\DecValTok{1}\SpecialCharTok{:}\DecValTok{7}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                              Sample-1 Sample-2 Sample-3 Sample-4 Sample-5
## Actinomycetaceae                    0        0        0        0        0
## Aerococcus                          0        0        0        0        0
## Aeromonas                           0        0        0        0        0
## Akkermansia                        21       36      475       61       34
## Alcaligenes faecalis et rel.        1        1        1        2        1
##                              Sample-6 Sample-7
## Actinomycetaceae                    0        0
## Aerococcus                          0        0
## Aeromonas                           0        0
## Akkermansia                        14       27
## Alcaligenes faecalis et rel.        1        1
\end{verbatim}

To illustrate the use of multiple assays, the relative abundance data can be
calculated and stored along the original count data using \texttt{transformAssay}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\FunctionTok{assays}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of length 2
## names(2): counts relabundance
\end{verbatim}

Now there are two assays available in the \texttt{tse} object, \texttt{counts} and
\texttt{relabundance}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"relabundance"}\NormalTok{)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{,}\DecValTok{1}\SpecialCharTok{:}\DecValTok{7}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                               Sample-1  Sample-2  Sample-3  Sample-4  Sample-5
## Actinomycetaceae             0.0000000 0.000e+00 0.0000000 0.0000000 0.000e+00
## Aerococcus                   0.0000000 0.000e+00 0.0000000 0.0000000 0.000e+00
## Aeromonas                    0.0000000 0.000e+00 0.0000000 0.0000000 0.000e+00
## Akkermansia                  0.0027657 3.547e-03 0.0666106 0.0056195 2.833e-03
## Alcaligenes faecalis et rel. 0.0001317 9.854e-05 0.0001402 0.0001842 8.333e-05
##                               Sample-6  Sample-7
## Actinomycetaceae             0.0000000 0.0000000
## Aerococcus                   0.0000000 0.0000000
## Aeromonas                    0.0000000 0.0000000
## Akkermansia                  0.0017690 0.0045570
## Alcaligenes faecalis et rel. 0.0001264 0.0001688
\end{verbatim}

Here the dimension of the count data remains unchanged in
transformation. This is in fact, a requirement for the assays.

\hypertarget{coldata}{%
\subsection{colData}\label{coldata}}

\texttt{colData} contains data on the samples.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{colData}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 1151 rows and 10 columns
##                   age      sex nationality DNA_extraction_method  project
##             <integer> <factor>    <factor>              <factor> <factor>
## Sample-1           28   male            US                    NA        1
## Sample-2           24   female          US                    NA        1
## Sample-3           52   male            US                    NA        1
## Sample-4           22   female          US                    NA        1
## Sample-5           25   female          US                    NA        1
## ...               ...      ...         ...                   ...      ...
## Sample-1168        50   female Scandinavia                     r       40
## Sample-1169        31   female Scandinavia                     r       40
## Sample-1170        31   female Scandinavia                     r       40
## Sample-1171        52   male   Scandinavia                     r       40
## Sample-1172        52   male   Scandinavia                     r       40
##             diversity   bmi_group  subject      time      sample
##             <numeric>    <factor> <factor> <numeric> <character>
## Sample-1         5.76 severeobese        1         0    Sample-1
## Sample-2         6.06 obese              2         0    Sample-2
## Sample-3         5.50 lean               3         0    Sample-3
## Sample-4         5.87 underweight        4         0    Sample-4
## Sample-5         5.89 lean               5         0    Sample-5
## ...               ...         ...      ...       ...         ...
## Sample-1168      5.87 severeobese      244       8.1 Sample-1168
## Sample-1169      5.87 overweight       245       2.3 Sample-1169
## Sample-1170      5.92 overweight       245       8.2 Sample-1170
## Sample-1171      6.04 overweight       246       2.1 Sample-1171
## Sample-1172      5.74 overweight       246       7.9 Sample-1172
\end{verbatim}

\hypertarget{rowdata}{%
\subsection{rowData}\label{rowdata}}

\texttt{rowData} contains data on the features of the analyzed samples. Of particular
interest to the microbiome field, this is used to store taxonomic information.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rowData}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 130 rows and 3 columns
##                                       Phylum          Family
##                                  <character>     <character>
## Actinomycetaceae              Actinobacteria  Actinobacteria
## Aerococcus                        Firmicutes         Bacilli
## Aeromonas                     Proteobacteria  Proteobacteria
## Akkermansia                  Verrucomicrobia Verrucomicrobia
## Alcaligenes faecalis et rel.  Proteobacteria  Proteobacteria
## ...                                      ...             ...
## Vibrio                        Proteobacteria  Proteobacteria
## Weissella et rel.                 Firmicutes         Bacilli
## Wissella et rel.                  Firmicutes         Bacilli
## Xanthomonadaceae              Proteobacteria  Proteobacteria
## Yersinia et rel.              Proteobacteria  Proteobacteria
##                                               Genus
##                                         <character>
## Actinomycetaceae                   Actinomycetaceae
## Aerococcus                               Aerococcus
## Aeromonas                                 Aeromonas
## Akkermansia                             Akkermansia
## Alcaligenes faecalis et rel. Alcaligenes faecalis..
## ...                                             ...
## Vibrio                                       Vibrio
## Weissella et rel.                 Weissella et rel.
## Wissella et rel.                   Wissella et rel.
## Xanthomonadaceae                   Xanthomonadaceae
## Yersinia et rel.                   Yersinia et rel.
\end{verbatim}

\hypertarget{rowtree}{%
\subsection{rowTree}\label{rowtree}}

Phylogenetic trees also play an important role in the microbiome field. The
\texttt{TreeSummarizedExperiment} class can keep track of features and node
relations via two functions, \texttt{rowTree} and \texttt{rowLinks}.

A tree can be accessed via \texttt{rowTree} as \texttt{phylo} object.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rowTree}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## NULL
\end{verbatim}

The links to the individual features are available through \texttt{rowLinks}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rowLinks}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## NULL
\end{verbatim}

Please note that there can be a 1:1 relationship between tree nodes and
features, but this is not a must-have. This means there can be features, which
are not linked to nodes, and nodes, which are not linked to features. To change
the links in an existing object, the \texttt{changeTree} function is available.

\hypertarget{alt-exp}{%
\subsection{Alternative experiments}\label{alt-exp}}

\emph{\textbf{Alternative experiments}} complement \emph{assays}. They can contain
complementary data, which is no longer tied to the same dimensions as
the assay data. However, the number of samples (columns) must be the
same.

This can come into play, for instance, when one has taxonomic
abundance profiles quantified with different measurement technologies,
such as phylogenetic microarrays, amplicon sequencing, or metagenomic
sequencing. Another common use case is including abundance tables for
different taxonomic ranks. Such alternative experiments concerning the
same set of samples can be stored as

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Separate \emph{assays} assuming that the taxonomic information can be mapped
  between features directly 1:1; or
\item
  Data in the \emph{altExp} slot of the \texttt{TreeSummarizedExperiment}, if the feature
  dimensions differ. Each element of the \emph{altExp} slot is a \texttt{SummarizedExperiment}
  or an object from a derived class with independent feature data.
\end{enumerate}

The following shows how to store taxonomic abundance tables
agglomerated at different taxonomic levels. However, the data could as
well originate from entirely different measurement sources as long as
the samples match.

Let us first agglomerate the data to Phylum level. This yields a new
TreeSE data object.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\StringTok{"Phylum"}\NormalTok{, }\AttributeTok{na.rm=}\ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# Both have the same number of columns (samples)}
\FunctionTok{dim}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  130 1151
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(tse\_phylum)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]    8 1151
\end{verbatim}

Then we can add the new phylum-level data object as an alternative experiment in the original data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Add the new data object to the original data object as an alternative experiment with the name "Phylum"}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Phylum"}\NormalTok{) }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum}

\CommentTok{\# Check the alternative experiment names available in the data}
\FunctionTok{altExpNames}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Phylum"
\end{verbatim}

We can now subset the data, for instance, and this acts on both altExp and assay data.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse[,}\DecValTok{1}\SpecialCharTok{:}\DecValTok{10}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 130 10 
## metadata(0):
## assays(2): counts relabundance
## rownames(130): Actinomycetaceae Aerococcus ... Xanthomonadaceae
##   Yersinia et rel.
## rowData names(3): Phylum Family Genus
## colnames(10): Sample-1 Sample-2 ... Sample-9 Sample-10
## colData names(10): age sex ... time sample
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(1): Phylum
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{dim}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse[,}\DecValTok{1}\SpecialCharTok{:}\DecValTok{10}\NormalTok{],}\StringTok{"Phylum"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  8 10
\end{verbatim}

For more details on \emph{altExp}, you can check the \href{https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html}{introduction} to the \texttt{SingleCellExperiment} package \citep{R_SingleCellExperiment}.

\hypertarget{mae}{%
\subsection{MultiAssayExperiments}\label{mae}}

\emph{\textbf{Multiple experiments}} relate to complementary measurement types,
such as transcriptomic or metabolomic profiling of the microbiome or
the host. Multiple experiments can be represented using the same
options as alternative experiments, or by using the
\texttt{MultiAssayExperiment} class \citep{Ramos2017}. Depending on how the
datasets relate to each other the data can be stored as:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Separate \emph{altExp} if the samples can be matched directly 1:1; or
\item
  As \texttt{MultiAssayExperiment} objects, in which the connections between
  samples are defined through a \texttt{sampleMap}. Each element on the
  \texttt{experimentsList} of an \texttt{MultiAssayExperiment} is \texttt{matrix} or
  \texttt{matrix}-like objects, including \texttt{SummarizedExperiment} objects, and
  the number of samples can differ between the elements.
\end{enumerate}

For information have a look at the \href{https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html}{intro vignette} of the \texttt{MultiAssayExperiment} package.

\begin{longtable}[]{@{}rlrr@{}}
\caption{\label{tab:options} \textbf{Recommended options for storing multiple data tables in microbiome studies} The \emph{assays} are best suited for data transformations (one-to-one match between samples and columns across the assays). The \emph{alternative experiments} are particularly suitable for alternative versions of the data that are of same type but may have a different number of features (e.g.~taxonomic groups); this is for instance the case with taxonomic abundance tables agglomerated at different levels (e.g.~genus vs.~phyla) or alternative profiling technologies (e.g.~amplicon sequencing vs.~shallow shotgun metagenomics). For alternative experiments one-to-one match between samples (cols) is libraryd but the alternative experiment tables can have different numbers of features (rows). Finally, elements of the \emph{MultiAssayExperiment} provide the most flexible way to incorporate multi-omic data tables with flexible numbers of samples and features. We recommend these conventions as the basis for methods development and application in microbiome studies.}\tabularnewline
\toprule\noalign{}
Option & Rows (features) & Cols (samples) & Recommended \\
\midrule\noalign{}
\endfirsthead
\toprule\noalign{}
Option & Rows (features) & Cols (samples) & Recommended \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
assays & match & match & Data transformations \\
altExp & free & match & Alternative experiments \\
MultiAssay & free & free (mapping) & Multi-omic experiments \\
\end{longtable}

\hypertarget{example-data}{%
\section{Demonstration data}\label{example-data}}

Open demonstration data for testing and benchmarking purposes is
available from multiple locations. This chapter introduces some
options. The other chapters of this book provide ample examples about
the use of the data.

\hypertarget{package-data}{%
\subsection{Package data}\label{package-data}}

The \texttt{mia} R package contains example datasets that are direct
conversions from the alternative \texttt{phyloseq} container to the
\texttt{TreeSummarizedExperiment} container.

List the \href{https://microbiome.github.io/mia/reference/mia-datasets.html}{available
datasets} in
the \texttt{mia} package:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Load the \texttt{GlobalPatterns} data from the \texttt{mia} package:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{GlobalPatterns}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\hypertarget{tengeler-desc}{%
\subsubsection{Tengeler2020}\label{tengeler-desc}}

Tengeler2020 is derived from a randomised blinded study on the effects of gut
microbiome on attention-deficit/hyperactivity disorder (ADHD) in humanised mice
\citep{Tengeler2020}. The dataset is briefly presented in
\href{https://microbiome.github.io/outreach/tengeler2020_presentation.html}{these slides}.

\hypertarget{hintikka-desc}{%
\subsubsection{HintikkaXOData}\label{hintikka-desc}}

\href{https://microbiome.github.io/microbiomeDataSets/reference/HintikkaXOData.html}{HintikkaXOData}
is derived from a study about the effects of fat diet and prebiotics on the
microbiome of rat models \citep{Hintikka2021}. It is available in the MAE data
container for R. The dataset is briefly summarized in
\href{https://microbiome.github.io/outreach/hintikkaxo_presentation.html}{these slides}.

\hypertarget{experimenthub-data}{%
\subsection{ExperimentHub data}\label{experimenthub-data}}

\href{https://bioconductor.org/packages/release/bioc/vignettes/ExperimentHub/inst/doc/ExperimentHub.html}{ExperimentHub}
provides a variety of data resources, including the
\href{https://bioconductor.org/packages/release/data/experiment/html/microbiomeDataSets.html}{microbiomeDataSets}
package \citep{Morgan2021, microlahti2021}.

A table of the available datasets is available through the
\texttt{availableDataSets} function.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(microbiomeDataSets)}
\FunctionTok{availableDataSets}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##             Dataset
## 1  GrieneisenTSData
## 2    HintikkaXOData
## 3       LahtiMLData
## 4        LahtiMData
## 5       LahtiWAData
## 6      OKeefeDSData
## 7 SilvermanAGutData
## 8        SongQAData
## 9   SprockettTHData
\end{verbatim}

All data are downloaded from ExperimentHub and cached for local
re-use. Check the \href{https://microbiome.github.io/microbiomeDataSets/reference/index.html}{man pages of each
function}
for a detailed documentation of the data contents and references. Let
us retrieve a \emph{\href{https://bioconductor.org/packages/3.17/MultiAssayExperiment}{MultiAssayExperiment}} dataset:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# mae \textless{}{-} HintikkaXOData()}
\CommentTok{\# Since HintikkaXOData is now added to mia, we can load it directly from there}
\CommentTok{\# We suggest to check other datasets from microbiomeDataSets}
\FunctionTok{data}\NormalTok{(HintikkaXOData, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{mae }\OtherTok{\textless{}{-}}\NormalTok{ HintikkaXOData}
\end{Highlighting}
\end{Shaded}

Data is available in \emph{\href{https://bioconductor.org/packages/3.17/SummarizedExperiment}{SummarizedExperiment}}, \texttt{r\ Biocpkg("TreeSummarizedExperiment")} and \texttt{r\ Biocpkg("MultiAssayExperiment")} data containers; see the separate
page on \href{https://microbiome.github.io/OMA/multitable.html}{alternative
containers} for more
details.

\hypertarget{curated-metagenomic-data}{%
\subsection{Curated metagenomic data}\label{curated-metagenomic-data}}

\href{https://bioconductor.org/packages/release/data/experiment/html/curatedMetagenomicData.html}{curatedMetagenomicData}
is a large collection of curated human microbiome datasets, provided as
\texttt{(Tree)SummarizedExperiment} objects \citep{Pasolli2017}. The resource
provides curated human microbiome data including gene families, marker
abundance, marker presence, pathway abundance, pathway coverage, and
relative abundance for samples from different body sites. See the
package homepage for more details on data availability and access.

As one example, let us retrieve the Vatanen (2016) \citep{Vatanen2016} data
set. This is a larger collection with a bit longer download time.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(curatedMetagenomicData)}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{curatedMetagenomicData}\NormalTok{(}\StringTok{"Vatanen*"}\NormalTok{, }\AttributeTok{dryrun =} \ConstantTok{FALSE}\NormalTok{, }\AttributeTok{counts =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{human-microbiome-compendium}{%
\subsection{Human microbiome compendium}\label{human-microbiome-compendium}}

\href{https://seandavi.github.io/MicroBioMap/}{MicroBioMap} dataset includes
over 170k samples of publicly available 16S rRNA amplicon sequencing data,
all processed using the same pipeline and reference database\citep{Abdill2023}.
After installing the MicroBioMap package (see the \href{https://github.com/seandavi/MicroBioMap\#microbiome-compendium}{original website} for instructions), you can load the compendium with

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(MicroBioMap)}
\NormalTok{cpd }\OtherTok{\textless{}{-}} \FunctionTok{getCompendium}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

This returns a TreeSummarizedExperiment object. Currently,
the ``tree'' part of the TreeSummarizedExperiment is not populated, but
that is on the roadmap\citep{compendiumpackage}.

After loading the compendium, you will have immediate access to nearly 170,000 microbiome samples of publicly available 16S rRNA amplicon sequencing data, all processed using the same pipeline and reference database. For more use examples in R/Bioconductor, see the \href{https://seandavi.github.io/MicroBioMap/articles/overview.html}{MicroBioMap vignette}.

\hypertarget{other-data-sources}{%
\subsection{Other data sources}\label{other-data-sources}}

The current collections provide access to vast microbiome data
resources. The output has to be converted into TreeSE/MAE separately.

\begin{itemize}
\tightlist
\item
  \href{https://github.com/beadyallen/MGnifyR}{MGnifyR} provides access to \href{https://www.ebi.ac.uk/metagenomics/}{EBI/MGnify}
\item
  \href{https://github.com/cran/qiitr}{qiitr} provides access to \href{https://qiita.com/about}{QIITA}
\end{itemize}

\hypertarget{loading-experimental-microbiome-data}{%
\section{Loading experimental microbiome data}\label{loading-experimental-microbiome-data}}

\hypertarget{s-workflow}{%
\subsection{16S workflow}\label{s-workflow}}

Result of amplicon sequencing is a large number of files that include all the sequences
that were read from samples. Those sequences need to be matched with taxa. Additionally,
we need to know how many times each taxa were found from each sample.

There are several algorithms to do that, and DADA2 is one of the most common.
You can find DADA2 pipeline tutorial, for example,
\href{https://benjjneb.github.io/dada2/tutorial.html}{here}.
After the DADA2 portion of the tutorial is completed, the data is stored into \emph{phyloseq} object
(Bonus: Handoff to phyloseq). To store the data to \emph{TreeSummarizedExperiment},
follow the example below.

You can find full workflow script without further explanations and comments from
\href{https://github.com/microbiome/OMA/blob/master/dada2_workflow.Rmd}{here}

Load required packages.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{library}\NormalTok{(ggplot2)}
\FunctionTok{library}\NormalTok{(BiocManager)}
\FunctionTok{library}\NormalTok{(Biostrings)}
\end{Highlighting}
\end{Shaded}

Create arbitrary example sample metadata like it was done in the tutorial. Usually,
sample metadata is imported as a file.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{samples.out }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(seqtab.nochim)}
\NormalTok{subject }\OtherTok{\textless{}{-}} \FunctionTok{sapply}\NormalTok{(}\FunctionTok{strsplit}\NormalTok{(samples.out, }\StringTok{"D"}\NormalTok{), }\StringTok{\textasciigrave{}}\AttributeTok{[}\StringTok{\textasciigrave{}}\NormalTok{, }\DecValTok{1}\NormalTok{)}
\NormalTok{gender }\OtherTok{\textless{}{-}} \FunctionTok{substr}\NormalTok{(subject,}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{)}
\NormalTok{subject }\OtherTok{\textless{}{-}} \FunctionTok{substr}\NormalTok{(subject,}\DecValTok{2}\NormalTok{,}\DecValTok{999}\NormalTok{)}
\NormalTok{day }\OtherTok{\textless{}{-}} \FunctionTok{as.integer}\NormalTok{(}\FunctionTok{sapply}\NormalTok{(}\FunctionTok{strsplit}\NormalTok{(samples.out, }\StringTok{"D"}\NormalTok{), }\StringTok{\textasciigrave{}}\AttributeTok{[}\StringTok{\textasciigrave{}}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\NormalTok{samdf }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{Subject=}\NormalTok{subject, }\AttributeTok{Gender=}\NormalTok{gender, }\AttributeTok{Day=}\NormalTok{day)}
\NormalTok{samdf}\SpecialCharTok{$}\NormalTok{When }\OtherTok{\textless{}{-}} \StringTok{"Early"}
\NormalTok{samdf}\SpecialCharTok{$}\NormalTok{When[samdf}\SpecialCharTok{$}\NormalTok{Day}\SpecialCharTok{\textgreater{}}\DecValTok{100}\NormalTok{] }\OtherTok{\textless{}{-}} \StringTok{"Late"}
\FunctionTok{rownames}\NormalTok{(samdf) }\OtherTok{\textless{}{-}}\NormalTok{ samples.out}
\end{Highlighting}
\end{Shaded}

Convert data into right format and create a \emph{TreeSE} object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create a list that contains assays}
\NormalTok{counts }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(seqtab.nochim)}
\NormalTok{counts }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(counts)}
\NormalTok{assays }\OtherTok{\textless{}{-}} \FunctionTok{SimpleList}\NormalTok{(}\AttributeTok{counts =}\NormalTok{ counts)}

\CommentTok{\# Convert colData and rowData into DataFrame}
\NormalTok{samdf }\OtherTok{\textless{}{-}} \FunctionTok{DataFrame}\NormalTok{(samdf)}
\NormalTok{taxa }\OtherTok{\textless{}{-}} \FunctionTok{DataFrame}\NormalTok{(taxa)}

\CommentTok{\# Create TreeSE}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{TreeSummarizedExperiment}\NormalTok{(}\AttributeTok{assays =}\NormalTok{ assays,}
                                \AttributeTok{colData =}\NormalTok{ samdf,}
                                \AttributeTok{rowData =}\NormalTok{ taxa}
\NormalTok{                                )}

\CommentTok{\# Remove mock sample like it is also done in DADA2 pipeline tutorial}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ tse[ , }\FunctionTok{colnames}\NormalTok{(tse) }\SpecialCharTok{!=} \StringTok{"mock"}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

Add sequences into \emph{referenceSeq} slot and convert rownames into simpler format.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Convert sequences into right format}
\NormalTok{dna }\OtherTok{\textless{}{-}}\NormalTok{ Biostrings}\SpecialCharTok{::}\FunctionTok{DNAStringSet}\NormalTok{( }\FunctionTok{rownames}\NormalTok{(tse) )}
\CommentTok{\# Add sequences into referenceSeq slot}
\FunctionTok{referenceSeq}\NormalTok{(tse) }\OtherTok{\textless{}{-}}\NormalTok{ dna}
\CommentTok{\# Convert rownames into ASV\_number format}
\FunctionTok{rownames}\NormalTok{(tse) }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"ASV"}\NormalTok{, }\FunctionTok{seq}\NormalTok{( }\FunctionTok{nrow}\NormalTok{(tse) ))}
\NormalTok{tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 232 20 
## metadata(0):
## assays(1): counts
## rownames(232): ASV1 ASV2 ... ASV231 ASV232
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(20): F3D0 F3D1 ... F3D9 Mock
## colData names(4): Subject Gender Day When
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
## referenceSeq: a DNAStringSet (232 sequences)
\end{verbatim}

\hypertarget{import-from-file}{%
\subsection{Import from external files}\label{import-from-file}}

Microbiome (taxonomic) profiling data is commonly distributed in
various file formats. You can import such external data files as a
(Tree)SummarizedExperiment object, but the details depend on the file
format. Here, we provide examples for common formats. Some datasets and raw
files to learn how to import raw data and construct TreeSE/MAE containers are
available in the
\href{https://github.com/microbiome/data}{microbiome data repository}.

\hypertarget{csv-import}{%
\subsubsection{CSV import}\label{csv-import}}

\textbf{CSV data tables} can be imported with the standard R functions,
then converted to the desired format. For detailed examples, you can
check the \href{https://bioconductor.org/help/course-materials/2019/BSS2019/04_Practical_CoreApproachesInBioconductor.html}{Bioconductor course
material}
by Martin Morgan. You can also check the \href{https://github.com/microbiome/OMA/tree/master/data}{example
files} and
construct your own CSV files accordingly.

Recommendations for the CSV files are the following. File names are
arbitrary; we refer here to the same names as in the examples:

\begin{itemize}
\item
  Abundance table (\texttt{assay\_taxa.csv}): data matrix (features x
  samples); first column provides feature IDs, the first row provides
  sample IDs; other values should be numeric (abundances).
\item
  Row data (\texttt{rowdata\_taxa.csv}): data table (features x info); first
  column provides feature IDs, the first row provides column headers;
  this file usually contains the taxonomic mapping between different
  taxonomic levels. Ideally, the feature IDs (row names) match one-to-one with
  the abundance table row names.
\item
  Column data (\texttt{coldata.csv}): data table (samples x info); first
  column provides sample IDs, the first row provides column headers;
  this file usually contains the sample metadata/phenodata (such as
  subject age, health etc). Ideally, the sample IDs match one-to-one with
  the abundance table column names.
\end{itemize}

After you have set up the CSV files, you can read them in R:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{count\_file  }\OtherTok{\textless{}{-}} \StringTok{"data/assay\_taxa.csv"}
\NormalTok{tax\_file    }\OtherTok{\textless{}{-}} \StringTok{"data/rowdata\_taxa.csv"}
\NormalTok{sample\_file }\OtherTok{\textless{}{-}} \StringTok{"data/coldata.csv"}

\CommentTok{\# Load files}
\NormalTok{counts  }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(count\_file, }\AttributeTok{row.names=}\DecValTok{1}\NormalTok{)   }\CommentTok{\# Abundance table (e.g. ASV data; to assay data)}
\NormalTok{tax     }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(tax\_file, }\AttributeTok{row.names=}\DecValTok{1}\NormalTok{)     }\CommentTok{\# Taxonomy table (to rowData)}
\NormalTok{samples }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(sample\_file, }\AttributeTok{row.names=}\DecValTok{1}\NormalTok{)  }\CommentTok{\# Sample data (to colData)}
\end{Highlighting}
\end{Shaded}

After reading the data in R, ensure the following:

\begin{itemize}
\item
  abundance table (\texttt{counts}): numeric \texttt{matrix}, with feature IDs as
  rownames and sample IDs as column names
\item
  rowdata (\texttt{tax}): \texttt{DataFrame}, with feature IDs as rownames. If this
  is a \texttt{data.frame} you can use the function \texttt{DataFrame()} to change
  the format. Column names are free but in microbiome analysis they
  usually they refer to taxonomic ranks. The rownames in rowdata
  should match with rownames in abundance table.
\item
  coldata (\texttt{samples}): \texttt{DataFrame}, with sample IDs as rownames. If
  this is a \texttt{data.frame} you can use the function \texttt{DataFrame()} to
  change the format. Column names are free. The rownames in coldata
  should match with colnames in abundance table.
\end{itemize}

\textbf{Always ensure that the tables have rownames!} The \emph{TreeSE} constructor compares
rownames and ensures that, for example, right samples are linked with right patient.

Also ensure that the row and column names match one-to-one between
abundance table, rowdata, and coldata:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Match rows and columns}
\NormalTok{counts }\OtherTok{\textless{}{-}}\NormalTok{ counts[}\FunctionTok{rownames}\NormalTok{(tax), }\FunctionTok{rownames}\NormalTok{(samples)]}

\CommentTok{\# Let us ensure that the data is in correct (numeric matrix) format:}
\NormalTok{counts }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(counts)}
\end{Highlighting}
\end{Shaded}

If you hesitate about the format of the data, you can compare to one
of the available demonstration datasets, and make sure that your data
components have the same format.

There are many different source files and many different ways to read
data in R. One can do data manipulation in R as well. Investigate the
entries as follows.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# coldata rownames match assay colnames}
\FunctionTok{all}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(samples) }\SpecialCharTok{==} \FunctionTok{colnames}\NormalTok{(counts)) }\CommentTok{\# our dataset}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{class}\NormalTok{(samples) }\CommentTok{\# should be data.frame or DataFrame}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "data.frame"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# rowdata rownames match assay rownames}
\FunctionTok{all}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(tax) }\SpecialCharTok{==} \FunctionTok{rownames}\NormalTok{(counts)) }\CommentTok{\# our dataset}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{class}\NormalTok{(tax) }\CommentTok{\# should be data.frame or DataFrame}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "data.frame"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Counts }
\FunctionTok{class}\NormalTok{(counts) }\CommentTok{\# should be a numeric matrix}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "matrix" "array"
\end{verbatim}

\hypertarget{constructing-treesummarizedexperiment}{%
\subsection{Constructing TreeSummarizedExperiment}\label{constructing-treesummarizedexperiment}}

Now let us create the TreeSE object from the input data tables. Here
we also convert the data objects in their preferred formats:

\begin{itemize}
\tightlist
\item
  counts --\textgreater{} numeric matrix
\item
  rowData --\textgreater{} DataFrame
\item
  colData --\textgreater{} DataFrame
\end{itemize}

The \texttt{SimpleList} could be used to include multiple alternative assays, if
necessary.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create a TreeSE}
\NormalTok{tse\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{TreeSummarizedExperiment}\NormalTok{(}\AttributeTok{assays =}  \FunctionTok{SimpleList}\NormalTok{(}\AttributeTok{counts =}\NormalTok{ counts),}
                                     \AttributeTok{colData =} \FunctionTok{DataFrame}\NormalTok{(samples),}
                                     \AttributeTok{rowData =} \FunctionTok{DataFrame}\NormalTok{(tax))}

\NormalTok{tse\_taxa}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 12706 40 
## metadata(0):
## assays(1): counts
## rownames(12706): GAYR01026362.62.2014 CVJT01000011.50.2173 ...
##   JRJTB:03787:02429 JRJTB:03787:02478
## rowData names(7): Phylum Class ... Species OTU
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

Now you should have a ready-made TreeSE data object that can be used in downstream analyses.

\hypertarget{constructing-multiassayexperiment}{%
\subsection{Constructing MultiAssayExperiment}\label{constructing-multiassayexperiment}}

To construct a \emph{MultiAssayExperiment} object, just combine multiple \emph{TreeSE} data containers.
Here we import metabolite data from the same study.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{count\_file }\OtherTok{\textless{}{-}} \StringTok{"data/assay\_metabolites.csv"}
\NormalTok{sample\_file }\OtherTok{\textless{}{-}} \StringTok{"data/coldata.csv"}

\CommentTok{\# Load files}
\NormalTok{counts  }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(count\_file, }\AttributeTok{row.names=}\DecValTok{1}\NormalTok{)  }
\NormalTok{samples }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(sample\_file, }\AttributeTok{row.names=}\DecValTok{1}\NormalTok{)}

\CommentTok{\# Create a TreeSE for the metabolite data}
\NormalTok{tse\_metabolite }\OtherTok{\textless{}{-}} \FunctionTok{TreeSummarizedExperiment}\NormalTok{(}\AttributeTok{assays =} \FunctionTok{SimpleList}\NormalTok{(}\AttributeTok{concs =} \FunctionTok{as.matrix}\NormalTok{(counts)),}
                                           \AttributeTok{colData =} \FunctionTok{DataFrame}\NormalTok{(samples))}

\NormalTok{tse\_metabolite}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 38 40 
## metadata(0):
## assays(1): concs
## rownames(38): Butyrate Acetate ... Malonate 1,3-dihydroxyacetone
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

Now we can combine these two experiments into \emph{MAE}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create an ExperimentList that includes experiments}
\NormalTok{experiments }\OtherTok{\textless{}{-}} \FunctionTok{ExperimentList}\NormalTok{(}\AttributeTok{microbiome =}\NormalTok{ tse\_taxa, }
                              \AttributeTok{metabolite =}\NormalTok{ tse\_metabolite)}

\CommentTok{\# Create a MAE}
\NormalTok{mae }\OtherTok{\textless{}{-}} \FunctionTok{MultiAssayExperiment}\NormalTok{(}\AttributeTok{experiments =}\NormalTok{ experiments)}

\NormalTok{mae}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 2:
##  [1] microbiome: TreeSummarizedExperiment with 12706 rows and 40 columns
##  [2] metabolite: TreeSummarizedExperiment with 38 rows and 40 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files
\end{verbatim}

\hypertarget{import-functions-for-standard-formats}{%
\subsection{Import functions for standard formats}\label{import-functions-for-standard-formats}}

Specific import functions are provided for:

\begin{itemize}
\tightlist
\item
  Biom files (see \texttt{help(mia::loadFromBiom)})
\item
  QIIME2 files (see \texttt{help(mia::loadFromQIIME2)})
\item
  Mothur files (see \texttt{help(mia::loadFromMothur)})
\end{itemize}

\hypertarget{biom-import}{%
\subsubsection{Biom import}\label{biom-import}}

Here we show how \href{https://biom-format.org/}{Biom files} are imported into
a TreeSE object using as an example Tengeler2020, which is further described in section \ref{tengeler-desc}. This dataset consists of 3 files, which can be
fetched or downloaded from
\href{https://github.com/microbiome/data/tree/main/Tengeler2020}{this repository}:

\begin{itemize}
\tightlist
\item
  biom file: abundance table and taxonomy information
\item
  csv file: sample metadata
\item
  tree file: phylogenetic tree
\end{itemize}

To begin with, we store the data in a local directory within the working
directory, such as \emph{data/}, and define the source file paths.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{biom\_file\_path }\OtherTok{\textless{}{-}} \StringTok{"data/Aggregated\_humanization2.biom"}
\NormalTok{sample\_meta\_file\_path }\OtherTok{\textless{}{-}} \StringTok{"data/Mapping\_file\_ADHD\_aggregated.csv"}
\NormalTok{tree\_file\_path }\OtherTok{\textless{}{-}} \StringTok{"data/Data\_humanization\_phylo\_aggregation.tre"}
\end{Highlighting}
\end{Shaded}

Now we can read in the biom file and convert it into a TreeSE object. In addition, we retrieve the rank names from the prefixes of the feature names and then remove them with the \texttt{rankFromPrefix} and \texttt{removeTaxaPrefixes} optional arguments.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}

\CommentTok{\# read biom and convert it to TreeSE}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{loadFromBiom}\NormalTok{(biom\_file\_path,}
                    \AttributeTok{rankFromPrefix =} \ConstantTok{TRUE}\NormalTok{,}
                    \AttributeTok{removeTaxaPrefixes =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Check}
\NormalTok{tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): taxonomy1 Phylum ... Family Genus
## colnames(27): A110 A111 ... A38 A39
## colData names(0):
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

The \texttt{assays} slot includes a list of abundance tables. The imported
abundance table is named as ``counts''. Let us inspect only the first
cols and rows.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{3}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\DecValTok{3}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##           A110  A111  A12
## 1726470  17722 11630    0
## 1726471  12052     0 2679
## 17264731     0   970    0
\end{verbatim}

The \texttt{rowdata} includes taxonomic information from the biom file. The \texttt{head()} command shows just the beginning of the data table for an overview.

\texttt{knitr::kable()} helps print the information more nicely.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 6 rows and 6 columns
##             taxonomy1          Phylum            Class              Order
##           <character>     <character>      <character>        <character>
## 1726470  "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726471  "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264731 "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264726 "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726472  "k__Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales
## 17264724 "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
##                       Family            Genus
##                  <character>      <character>
## 1726470       Bacteroidaceae     Bacteroides"
## 1726471       Bacteroidaceae     Bacteroides"
## 17264731  Porphyromonadaceae Parabacteroides"
## 17264726      Bacteroidaceae     Bacteroides"
## 1726472  Verrucomicrobiaceae     Akkermansia"
## 17264724      Bacteroidaceae     Bacteroides"
\end{verbatim}

We further polish the feature names by removing unnecessary characters and then replace the original rowData with its updated version.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Genus level has additional \textquotesingle{}\textbackslash{}"\textquotesingle{}, so let\textquotesingle{}s delete that also}
\NormalTok{rowdata\_modified }\OtherTok{\textless{}{-}}\NormalTok{ BiocParallel}\SpecialCharTok{::}\FunctionTok{bplapply}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse), }
                                           \AttributeTok{FUN =}\NormalTok{ stringr}\SpecialCharTok{::}\NormalTok{str\_remove, }
                                           \AttributeTok{pattern =} \StringTok{\textquotesingle{}}\SpecialCharTok{\textbackslash{}"}\StringTok{\textquotesingle{}}\NormalTok{)}

\CommentTok{\# rowdata\_modified is a list, so convert this back to DataFrame format. }
\CommentTok{\# and assign the cleaned data back to the TSE rowData}
\FunctionTok{rowData}\NormalTok{(tse) }\OtherTok{\textless{}{-}} \FunctionTok{DataFrame}\NormalTok{(rowdata\_modified)}

\CommentTok{\# Now we have a nicer table}
\FunctionTok{head}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 6 rows and 6 columns
##            taxonomy1          Phylum            Class              Order
##          <character>     <character>      <character>        <character>
## 1726470  k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726471  k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264731 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 17264726 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
## 1726472  k__Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales
## 17264724 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales
##                       Family           Genus
##                  <character>     <character>
## 1726470       Bacteroidaceae     Bacteroides
## 1726471       Bacteroidaceae     Bacteroides
## 17264731  Porphyromonadaceae Parabacteroides
## 17264726      Bacteroidaceae     Bacteroides
## 1726472  Verrucomicrobiaceae     Akkermansia
## 17264724      Bacteroidaceae     Bacteroides
\end{verbatim}

We notice that the imported biom file did not contain any colData yet,
so only an empty dataframe appears in this slot.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 6 rows and 0 columns
\end{verbatim}

Let us add colData from the sample metadata, which is stored in a CSV file.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# CSV file with colnames in the first row and rownames in the first column}
\NormalTok{sample\_meta }\OtherTok{\textless{}{-}} \FunctionTok{read.csv}\NormalTok{(sample\_meta\_file\_path,}
                        \AttributeTok{sep =} \StringTok{","}\NormalTok{, }\AttributeTok{row.names =} \DecValTok{1}\NormalTok{)}

\CommentTok{\# Add this sample data to colData of the taxonomic data object}
\CommentTok{\# Note that the data must be given in a DataFrame format (required for our purposes)}
\FunctionTok{colData}\NormalTok{(tse) }\OtherTok{\textless{}{-}} \FunctionTok{DataFrame}\NormalTok{(sample\_meta)}
\end{Highlighting}
\end{Shaded}

Now the colData includes the sample metadata.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 6 rows and 4 columns
##        Treatment      Cohort TreatmentxCohort Description
##      <character> <character>      <character> <character>
## A110        ADHD    Cohort_1    ADHD_Cohort_1        A110
## A12         ADHD    Cohort_1    ADHD_Cohort_1         A12
## A15         ADHD    Cohort_1    ADHD_Cohort_1         A15
## A19         ADHD    Cohort_1    ADHD_Cohort_1         A19
## A21         ADHD    Cohort_2    ADHD_Cohort_2         A21
## A23         ADHD    Cohort_2    ADHD_Cohort_2         A23
\end{verbatim}

Finally, we add a phylogenetic tree to the rowData slot. Such feature is available only in TreeSE objects. Similarly, Trees specifying the sample hierarchy can be stored in the colTree slot.

Here, we read in the file containing the phylogenetic tree and insert it in corresponding slot of the TreeSE object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Reads the tree file}
\NormalTok{tree }\OtherTok{\textless{}{-}}\NormalTok{ ape}\SpecialCharTok{::}\FunctionTok{read.tree}\NormalTok{(tree\_file\_path)}

\CommentTok{\# Add tree to rowTree}
\FunctionTok{rowTree}\NormalTok{(tse) }\OtherTok{\textless{}{-}}\NormalTok{ tree}

\CommentTok{\# Check}
\NormalTok{tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 151 27 
## metadata(0):
## assays(1): counts
## rownames(151): 1726470 1726471 ... 17264756 17264757
## rowData names(6): taxonomy1 Phylum ... Family Genus
## colnames(27): A110 A12 ... A35 A38
## colData names(4): Treatment Cohort TreatmentxCohort Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (151 rows)
## rowTree: 1 phylo tree(s) (151 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

Now the rowTree slot contains the phylogenetic tree:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{rowTree}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\hypertarget{conversions-between-data-formats-in-r}{%
\subsection{Conversions between data formats in R}\label{conversions-between-data-formats-in-r}}

If the data has already been imported in R in another format, it
can be readily converted into \texttt{TreeSummarizedExperiment}, as shown in our next
example. Note that similar conversion functions to
\texttt{TreeSummarizedExperiment} are available for multiple data formats via
the \texttt{mia} package (see makeTreeSummarizedExperimentFrom* for phyloseq,
Biom, and DADA2).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}

\CommentTok{\# phyloseq example data}
\FunctionTok{data}\NormalTok{(GlobalPatterns, }\AttributeTok{package=}\StringTok{"phyloseq"}\NormalTok{) }
\NormalTok{GlobalPatterns\_phyloseq }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}
\NormalTok{GlobalPatterns\_phyloseq}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# convert phyloseq to TSE}
\NormalTok{GlobalPatterns\_TSE }\OtherTok{\textless{}{-}} \FunctionTok{makeTreeSummarizedExperimentFromPhyloseq}\NormalTok{(GlobalPatterns\_phyloseq) }
\NormalTok{GlobalPatterns\_TSE}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

We can also convert \texttt{TreeSummarizedExperiment} objects into \texttt{phyloseq}
with respect to the shared components that are supported by both
formats (i.e.~taxonomic abundance table, sample metadata, taxonomic
table, phylogenetic tree, sequence information). This is useful for
instance when additional methods are available for \texttt{phyloseq}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# convert TSE to phyloseq}
\NormalTok{GlobalPatterns\_phyloseq2 }\OtherTok{\textless{}{-}} \FunctionTok{makePhyloseqFromTreeSummarizedExperiment}\NormalTok{(GlobalPatterns\_TSE) }
\NormalTok{GlobalPatterns\_phyloseq2}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]
\end{verbatim}

Conversion is possible between other data formats. Interested readers can refer to the following functions:
* \href{https://microbiome.github.io/mia/reference/makeTreeSummarizedExperimentFromDADA2.html}{makeTreeSummarizedExperimentFromDADA2}
* \href{https://microbiome.github.io/mia/reference/makeSummarizedExperimentFromBiom.html}{makeSummarizedExperimentFromBiom}
* \href{https://microbiome.github.io/mia/reference/loadFromMetaphlan.html}{loadFromMetaphlan}
* \href{https://microbiome.github.io/mia/reference/loadFromQIIME2.html}{readQZA}

\hypertarget{part-focus-topics}{%
\part{Focus Topics}\label{part-focus-topics}}

\hypertarget{datamanipulation}{%
\chapter{Data Manipulation}\label{datamanipulation}}

\hypertarget{tidying-and-subsetting}{%
\section{Tidying and subsetting}\label{tidying-and-subsetting}}

\hypertarget{tidy-data}{%
\subsection{Tidy data}\label{tidy-data}}

For several custom analysis and visualization packages, such as those from
\texttt{tidyverse}, the \texttt{SE} data can be converted to a long data.frame format with
\texttt{meltAssay}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(GlobalPatterns, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{, }\AttributeTok{method=}\StringTok{"relabundance"}\NormalTok{)}
\NormalTok{molten\_tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{meltAssay}\NormalTok{(tse,}
                        \AttributeTok{add\_row\_data =} \ConstantTok{TRUE}\NormalTok{,}
                        \AttributeTok{add\_col\_data =} \ConstantTok{TRUE}\NormalTok{,}
                        \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}
\NormalTok{molten\_tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 499,616 x 17
##    FeatureID SampleID relabundance Kingdom Phylum       Class Order Family Genus
##    <fct>     <fct>           <dbl> <chr>   <chr>        <chr> <chr> <chr>  <chr>
##  1 549322    CL3                 0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  2 549322    CC1                 0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  3 549322    SV1                 0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  4 549322    M31Fcsw             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  5 549322    M11Fcsw             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  6 549322    M31Plmr             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  7 549322    M11Plmr             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  8 549322    F21Plmr             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
##  9 549322    M31Tong             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
## 10 549322    M11Tong             0 Archaea Crenarchaeo~ Ther~ <NA>  <NA>   <NA> 
## # i 499,606 more rows
## # i 8 more variables: Species <chr>, X.SampleID <fct>, Primer <fct>,
## #   Final_Barcode <fct>, Barcode_truncated_plus_T <fct>,
## #   Barcode_full_length <fct>, SampleType <fct>, Description <fct>
\end{verbatim}

\hypertarget{subsetting}{%
\subsection{Subsetting}\label{subsetting}}

\textbf{Subsetting} data helps to draw the focus of analysis on particular
sets of samples and / or features. When dealing with large datasets,
the subset of interest can be extracted and investigated
separately. This might improve performance and reduce the
computational load.

Load:

\begin{itemize}
\tightlist
\item
  mia
\item
  dplyr
\item
  knitr
\item
  data \texttt{GlobalPatterns}
\end{itemize}

Let us store \texttt{GlobalPatterns} into \texttt{tse} and check its original number of features (rows) and samples (columns). \textbf{Note}: when subsetting by sample, expect the number of columns to decrease; when subsetting by feature, expect the number of rows to decrease.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Store data into se and check dimensions}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}
\CommentTok{\# Show dimensions (features x samples)}
\FunctionTok{dim}\NormalTok{(tse) }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 19216    26
\end{verbatim}

\hypertarget{subset-by-sample-column-wise}{%
\subsubsection{Subset by sample (column-wise)}\label{subset-by-sample-column-wise}}

For the sake of demonstration, here we will extract a subset containing only the samples of human origin (feces, skin or tongue), stored as \texttt{SampleType} within \texttt{colData(tse)} and also in \texttt{tse}.

First, we would like to see all the possible values that \texttt{SampleType} can take on and how frequent those are:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Inspect possible values for SampleType}
\FunctionTok{unique}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{SampleType)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] Soil               Feces              Skin               Tongue            
## [5] Freshwater         Freshwater (creek) Ocean              Sediment (estuary)
## [9] Mock              
## 9 Levels: Feces Freshwater Freshwater (creek) Mock ... Tongue
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Show the frequency of each value}
\NormalTok{tse}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{table}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\linewidth}{!}{
\begin{tabular}{l|r}
\hline
. & Freq\\
\hline
Feces & 4\\
\hline
Freshwater & 2\\
\hline
Freshwater (creek) & 3\\
\hline
Mock & 3\\
\hline
Ocean & 3\\
\hline
Sediment (estuary) & 3\\
\hline
Skin & 3\\
\hline
Soil & 3\\
\hline
Tongue & 2\\
\hline
\end{tabular}}
\end{table}

\textbf{Note}: after subsetting, expect the number of columns to equal the
sum of the frequencies of the samples that you are interested
in. For instance, \texttt{ncols\ =\ Feces\ +\ Skin\ +\ Tongue\ =\ 4\ +\ 3\ +\ 2\ =\ 9}.

Next, we \emph{logical index} across the columns of \texttt{tse} (make sure to
leave the first index empty to select all rows) and filter for the
samples of human origin. For this, we use the information on the
samples from the meta data \texttt{colData(tse)}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Subset by sample}
\NormalTok{tse\_subset\_by\_sample }\OtherTok{\textless{}{-}}\NormalTok{ tse[ , tse}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Feces"}\NormalTok{, }\StringTok{"Skin"}\NormalTok{, }\StringTok{"Tongue"}\NormalTok{)]}

\CommentTok{\# Show dimensions}
\FunctionTok{dim}\NormalTok{(tse\_subset\_by\_sample)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 19216     9
\end{verbatim}

As a sanity check, the new object \texttt{tse\_subset\_by\_sample} should have
the original number of features (rows) and a number of samples
(columns) equal to the sum of the samples of interest (in this case
9).

Several characteristics can be used to subset by sample:

\begin{itemize}
\tightlist
\item
  origin
\item
  sampling time
\item
  sequencing method
\item
  DNA / RNA barcode
\item
  cohort
\end{itemize}

\hypertarget{subset-by-feature-row-wise}{%
\subsubsection{Subset by feature (row-wise)}\label{subset-by-feature-row-wise}}

Similarly, here we will extract a subset containing only the features
that belong to the phyla Actinobacteria and Chlamydiae, stored as
\texttt{Phylum} within \texttt{rowData(tse)}. However, subsetting by feature implies
a few more obstacles, such as the presence of \texttt{NA} elements and the
possible need for agglomeration.

As previously, we would first like to see all the possible values that
\texttt{Phylum} can take on and how frequent those are:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Inspect possible values for phylum}
\FunctionTok{unique}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  [1] "Crenarchaeota"    "Euryarchaeota"    "Actinobacteria"   "Spirochaetes"    
##  [5] "MVP-15"           "Proteobacteria"   "SBR1093"          "Fusobacteria"    
##  [9] "Tenericutes"      "ZB3"              "Cyanobacteria"    "GOUTA4"          
## [13] "TG3"              "Chlorobi"         "Bacteroidetes"    "Caldithrix"      
## [17] "KSB1"             "SAR406"           "LCP-89"           "Thermi"          
## [21] "Gemmatimonadetes" "Fibrobacteres"    "GN06"             "AC1"             
## [25] "TM6"              "OP8"              "Elusimicrobia"    "NC10"            
## [29] "SPAM"             NA                 "Acidobacteria"    "CCM11b"          
## [33] "Nitrospirae"      "NKB19"            "BRC1"             "Hyd24-12"        
## [37] "WS3"              "PAUC34f"          "GN04"             "GN12"            
## [41] "Verrucomicrobia"  "Lentisphaerae"    "LD1"              "Chlamydiae"      
## [45] "OP3"              "Planctomycetes"   "Firmicutes"       "OP9"             
## [49] "WPS-2"            "Armatimonadetes"  "SC3"              "TM7"             
## [53] "GN02"             "SM2F11"           "ABY1_OD1"         "ZB2"             
## [57] "OP11"             "Chloroflexi"      "SC4"              "WS1"             
## [61] "GAL15"            "AD3"              "WS2"              "Caldiserica"     
## [65] "Thermotogae"      "Synergistetes"    "SR1"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Show the frequency of each value}
\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{table}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\linewidth}{!}{
\begin{tabular}{l|r}
\hline
. & Freq\\
\hline
ABY1\_OD1 & 7\\
\hline
AC1 & 1\\
\hline
Acidobacteria & 1021\\
\hline
Actinobacteria & 1631\\
\hline
AD3 & 9\\
\hline
Armatimonadetes & 61\\
\hline
Bacteroidetes & 2382\\
\hline
BRC1 & 13\\
\hline
Caldiserica & 3\\
\hline
Caldithrix & 10\\
\hline
CCM11b & 2\\
\hline
Chlamydiae & 21\\
\hline
Chlorobi & 64\\
\hline
Chloroflexi & 437\\
\hline
Crenarchaeota & 106\\
\hline
Cyanobacteria & 393\\
\hline
Elusimicrobia & 31\\
\hline
Euryarchaeota & 102\\
\hline
Fibrobacteres & 7\\
\hline
Firmicutes & 4356\\
\hline
Fusobacteria & 37\\
\hline
GAL15 & 2\\
\hline
Gemmatimonadetes & 191\\
\hline
GN02 & 8\\
\hline
GN04 & 7\\
\hline
GN06 & 2\\
\hline
GN12 & 1\\
\hline
GOUTA4 & 11\\
\hline
Hyd24-12 & 4\\
\hline
KSB1 & 6\\
\hline
LCP-89 & 2\\
\hline
LD1 & 2\\
\hline
Lentisphaerae & 21\\
\hline
MVP-15 & 5\\
\hline
NC10 & 9\\
\hline
Nitrospirae & 74\\
\hline
NKB19 & 16\\
\hline
OP11 & 6\\
\hline
OP3 & 30\\
\hline
OP8 & 27\\
\hline
OP9 & 4\\
\hline
PAUC34f & 3\\
\hline
Planctomycetes & 638\\
\hline
Proteobacteria & 6416\\
\hline
SAR406 & 21\\
\hline
SBR1093 & 9\\
\hline
SC3 & 8\\
\hline
SC4 & 8\\
\hline
SM2F11 & 5\\
\hline
SPAM & 22\\
\hline
Spirochaetes & 124\\
\hline
SR1 & 5\\
\hline
Synergistetes & 7\\
\hline
Tenericutes & 143\\
\hline
TG3 & 5\\
\hline
Thermi & 46\\
\hline
Thermotogae & 1\\
\hline
TM6 & 27\\
\hline
TM7 & 32\\
\hline
Verrucomicrobia & 470\\
\hline
WPS-2 & 20\\
\hline
WS1 & 5\\
\hline
WS2 & 2\\
\hline
WS3 & 70\\
\hline
ZB2 & 2\\
\hline
ZB3 & 2\\
\hline
\end{tabular}}
\end{table}

\textbf{Note}: after subsetting, expect the number of columns to equal the
sum of the frequencies of the feature(s) that you are interested
in. For instance, \texttt{nrows\ =\ Actinobacteria\ +\ Chlamydiae\ =\ 1631\ +\ 21\ =\ \ \ 1652}.

Depending on your research question, you might or might not need to
agglomerate the data in the first place: if you want to find the
abundance of each and every feature that belongs to Actinobacteria and
Chlamydiae, agglomeration is not needed; if you want to find the total
abundance of all features that belong to Actinobacteria or
Chlamydiae, agglomeration is recommended.

\hypertarget{non-agglomerated-data}{%
\paragraph{Non-agglomerated data}\label{non-agglomerated-data}}

Next, we \emph{logical index} across the rows of \texttt{tse} (make sure to leave
the second index empty to select all columns) and filter for the
features that fall in either Actinobacteria or Chlamydiae group. For this,
we use the information on the samples from the metadata
\texttt{rowData(tse)}.

The first term with the \texttt{\%in\%} operator includes all the features
of interest, whereas the second term after the AND operator \texttt{\&}
filters out all features that have an \texttt{NA} in place of the phylum variable.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Subset by feature}
\NormalTok{tse\_subset\_by\_feature }\OtherTok{\textless{}{-}}\NormalTok{ tse[}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Actinobacteria"}\NormalTok{, }\StringTok{"Chlamydiae"}\NormalTok{) }\SpecialCharTok{\&} \SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum), ]}

\CommentTok{\# Show dimensions}
\FunctionTok{dim}\NormalTok{(tse\_subset\_by\_feature)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1652   26
\end{verbatim}

As a sanity check, the new object, \texttt{tse\_subset\_by\_feature}, should have the original number of samples (columns) and a number of features (rows) equal to the sum of the features of interest (in this case, 1652).

\hypertarget{agglomerated-data}{%
\paragraph{Agglomerated data}\label{agglomerated-data}}

When total abundances of certain phyla are of relevance, the data is initially agglomerated by Phylum. Then, similar steps as in the case of non-agglomerated data are followed.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate by phylum}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}}\NormalTok{ tse }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{mergeFeaturesByRank}\NormalTok{(}\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{)}

\CommentTok{\# Subset by feature and remove NAs}
\NormalTok{tse\_phylum\_subset\_by\_feature }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum[}\FunctionTok{rowData}\NormalTok{(tse\_phylum)}\SpecialCharTok{$}\NormalTok{Phylum }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Actinobacteria"}\NormalTok{, }\StringTok{"Chlamydiae"}\NormalTok{) }\SpecialCharTok{\&} \SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse\_phylum)}\SpecialCharTok{$}\NormalTok{Phylum), ]}

\CommentTok{\# Show dimensions}
\FunctionTok{dim}\NormalTok{(tse\_phylum\_subset\_by\_feature)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  2 26
\end{verbatim}

\textbf{Note}: as data was agglomerated, the number of rows should equal the
number of phyla used to index (in this case, just 2).

Alternatively:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Store features of interest into phyla}
\NormalTok{phyla }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Phylum:Actinobacteria"}\NormalTok{, }\StringTok{"Phylum:Chlamydiae"}\NormalTok{)}
\CommentTok{\# subset by feature}
\NormalTok{tse\_phylum\_subset\_by\_feature }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum[phyla, ]}
\CommentTok{\# Show dimensions}
\FunctionTok{dim}\NormalTok{(tse\_subset\_by\_feature)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1652   26
\end{verbatim}

The code above returns the non-agglomerated version of the data.

Fewer characteristics can be used to subset by feature:

\begin{itemize}
\tightlist
\item
  Taxonomic rank
\item
  Meta-taxonomic group
\end{itemize}

For subsetting by kingdom, agglomeration does not apply, whereas for
the other ranks it can be applied if necessary.

\hypertarget{subset-by-sample-and-feature}{%
\subsubsection{Subset by sample and feature}\label{subset-by-sample-and-feature}}

Finally, we can subset data by sample and feature at once. The
resulting subset contains all the samples of human origin and all the
features of phyla Actinobacteria or Chlamydiae.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Subset by sample and feature and remove NAs}
\NormalTok{tse\_subset\_by\_sample\_feature }\OtherTok{\textless{}{-}}\NormalTok{ tse[}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Actinobacteria"}\NormalTok{, }\StringTok{"Chlamydiae"}\NormalTok{) }\SpecialCharTok{\&} \SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum), tse}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Feces"}\NormalTok{, }\StringTok{"Skin"}\NormalTok{, }\StringTok{"Tongue"}\NormalTok{)]}

\CommentTok{\# Show dimensions}
\FunctionTok{dim}\NormalTok{(tse\_subset\_by\_sample\_feature)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1652    9
\end{verbatim}

\textbf{Note}: the dimensions of \texttt{tse\_subset\_by\_sample\_feature} agree with
those of the previous subsets (9 columns filtered by sample and 1652
rows filtered by feature).

If a study was to consider and quantify the presence of Actinobacteria
as well as Chlamydiae in different sites of the human body,
\texttt{tse\_subset\_by\_sample\_feature} might be a suitable subset to start
with.

\hypertarget{remove-empty-columns-and-rows}{%
\subsubsection{Remove empty columns and rows}\label{remove-empty-columns-and-rows}}

Sometimes data might contain, e.g., features that are not present in any of the samples.
This can occur, for example, after the data subsetting. In certain analyses, we might want to
remove those instances.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate data at Genus level }
\NormalTok{tse\_genus }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{)}
\CommentTok{\# List bacteria that we want to include}
\NormalTok{genera }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Class:Thermoprotei"}\NormalTok{, }\StringTok{"Genus:Sulfolobus"}\NormalTok{, }\StringTok{"Genus:Sediminicola"}\NormalTok{)}
\CommentTok{\# Subset data}
\NormalTok{tse\_genus\_sub }\OtherTok{\textless{}{-}}\NormalTok{ tse\_genus[genera, ]}

\NormalTok{tse\_genus\_sub}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 3 26 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(3): Class:Thermoprotei Genus:Sulfolobus Genus:Sediminicola
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (3 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# List total counts of each sample}
\FunctionTok{colSums}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse\_genus\_sub, }\StringTok{"counts"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      CL3      CC1      SV1  M31Fcsw  M11Fcsw  M31Plmr  M11Plmr  F21Plmr 
##        1        0        0        1        1        0        4        1 
##  M31Tong  M11Tong LMEpi24M SLEpi20M   AQC1cm   AQC4cm   AQC7cm      NP2 
##        7        3        0        2       64      105      136      222 
##      NP3      NP5  TRRsed1  TRRsed2  TRRsed3     TS28     TS29    Even1 
##     6433     1154        2        2        2        0        0        0 
##    Even2    Even3 
##        2        0
\end{verbatim}

Now we can see that certain samples do not include any bacteria. We can remove those.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Remove samples that do not contain any bacteria}
\NormalTok{tse\_genus\_sub }\OtherTok{\textless{}{-}}\NormalTok{ tse\_genus\_sub[ , }\FunctionTok{colSums}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse\_genus\_sub, }\StringTok{"counts"}\NormalTok{)) }\SpecialCharTok{!=} \DecValTok{0}\NormalTok{ ]}
\NormalTok{tse\_genus\_sub}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 3 18 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(3): Class:Thermoprotei Genus:Sulfolobus Genus:Sediminicola
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(18): CL3 M31Fcsw ... TRRsed3 Even2
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (3 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

The same action can also be applied to the features.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Take only those samples that are collected from feces, skin, or tongue}
\NormalTok{tse\_genus\_sub }\OtherTok{\textless{}{-}}\NormalTok{ tse\_genus[ , tse\_genus}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Feces"}\NormalTok{, }\StringTok{"Skin"}\NormalTok{, }\StringTok{"Tongue"}\NormalTok{)]}

\NormalTok{tse\_genus\_sub}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 1516 9 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(1516): Class:Thermoprotei Genus:Sulfolobus ...
##   Genus:Coprothermobacter Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(9): M31Fcsw M11Fcsw ... TS28 TS29
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (1516 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# What is the number of bacteria that are not present?}
\FunctionTok{sum}\NormalTok{(}\FunctionTok{rowSums}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse\_genus\_sub, }\StringTok{"counts"}\NormalTok{)) }\SpecialCharTok{==} \DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 435
\end{verbatim}

We can see that there are bacteria that are not present in these samples we chose.
We can remove those bacteria from the data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Take only those bacteria that are present}
\NormalTok{tse\_genus\_sub }\OtherTok{\textless{}{-}}\NormalTok{ tse\_genus\_sub[}\FunctionTok{rowSums}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse\_genus\_sub, }\StringTok{"counts"}\NormalTok{)) }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{, ]}

\NormalTok{tse\_genus\_sub}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 1081 9 
## metadata(1): agglomerated_by_rank
## assays(1): counts
## rownames(1081): Genus:Sulfolobus Order:NRP-J ...
##   Genus:Coprothermobacter Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(9): M31Fcsw M11Fcsw ... TS28 TS29
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (1081 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\hypertarget{splitting}{%
\subsection{Splitting}\label{splitting}}

You can split the data based on variables by using the functions \texttt{splitByRanks}
and \texttt{splitOn}.

\texttt{splitByRanks} splits the data based on taxonomic ranks. Since the elements of the output list
share columns, they can be stored into \texttt{altExp}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{altExps}\NormalTok{(tse) }\OtherTok{\textless{}{-}} \FunctionTok{splitByRanks}\NormalTok{(tse)}
\FunctionTok{altExps}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of length 7
## names(7): Kingdom Phylum Class Order Family Genus Species
\end{verbatim}

If you want to split the data based on another variable than taxonomic rank, use
\texttt{splitOn}. It works for row-wise and column-wise splitting.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{splitOn}\NormalTok{(tse, }\StringTok{"SampleType"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of length 9
## names(9): Soil Feces Skin Tongue ... Ocean Sediment (estuary) Mock
\end{verbatim}

\hypertarget{add-or-modify-data}{%
\section{Add or modify data}\label{add-or-modify-data}}

The information contained by the \texttt{colData} of a \texttt{TreeSE} can be modified by
accessing the desired variables.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# modify the Description entries}
\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Description }\OtherTok{\textless{}{-}} \FunctionTok{paste}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Description, }\StringTok{"modified description"}\NormalTok{)}

\CommentTok{\# view modified variable}
\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{Description)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Calhoun South Carolina Pine soil, pH 4.9 modified description"  
## [2] "Cedar Creek Minnesota, grassland, pH 6.1 modified description"  
## [3] "Sevilleta new Mexico, desert scrub, pH 8.3 modified description"
## [4] "M3, Day 1, fecal swab, whole body study modified description"   
## [5] "M1, Day 1, fecal swab, whole body study  modified description"  
## [6] "M3, Day 1, right palm, whole body study modified description"
\end{verbatim}

New information can also be added to the experiment by creating a new variable.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# simulate new data}
\NormalTok{new\_data }\OtherTok{\textless{}{-}} \FunctionTok{runif}\NormalTok{(}\FunctionTok{ncol}\NormalTok{(tse))}

\CommentTok{\# store new data as new variable in colData}
\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{NewVariable }\OtherTok{\textless{}{-}}\NormalTok{ new\_data}

\CommentTok{\# view new variable}
\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{NewVariable)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.12155 0.35614 0.58483 0.20994 0.30377 0.01013
\end{verbatim}

\hypertarget{merge-data}{%
\section{Merge data}\label{merge-data}}

\texttt{mia} package has \texttt{mergeSEs} function that merges multiple \texttt{SummarizedExperiment}
objects. For example, it is possible to combine multiple \texttt{TreeSE} objects which each
includes one sample.

\texttt{mergeSEs} works like \texttt{dplyr} joining functions. In fact, there are available
\texttt{dplyr-like} aliases of \texttt{mergeSEs}, such as \texttt{full\_join}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Take subsets for demonstration purposes}
\NormalTok{tse1 }\OtherTok{\textless{}{-}}\NormalTok{ tse[, }\DecValTok{1}\NormalTok{]}
\NormalTok{tse2 }\OtherTok{\textless{}{-}}\NormalTok{ tse[, }\DecValTok{2}\NormalTok{]}
\NormalTok{tse3 }\OtherTok{\textless{}{-}}\NormalTok{ tse[, }\DecValTok{3}\NormalTok{]}
\NormalTok{tse4 }\OtherTok{\textless{}{-}}\NormalTok{ tse[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{100}\NormalTok{, }\DecValTok{4}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# With inner join, we want to include all shared rows. When using mergeSEs function}
\CommentTok{\# all samples are always preserved.}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{mergeSEs}\NormalTok{(}\FunctionTok{list}\NormalTok{(tse1, tse2, tse3, tse4), }\AttributeTok{join =} \StringTok{"inner"}\NormalTok{)}
\NormalTok{tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 100 4 
## metadata(0):
## assays(1): counts
## rownames(100): 239672 243675 ... 549322 951
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(4): CC1 CL3 M31Fcsw SV1
## colData names(8): X.SampleID Primer ... Description NewVariable
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (100 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Left join preserves all rows of the 1st object}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{left\_join}\NormalTok{(tse1, tse4, }\AttributeTok{missing\_values =} \DecValTok{0}\NormalTok{)}
\NormalTok{tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 19216 2 
## metadata(0):
## assays(1): counts
## rownames(19216): 239672 243675 ... 239967 254851
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(2): CL3 M31Fcsw
## colData names(8): X.SampleID Primer ... Description NewVariable
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (19216 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\hypertarget{additional-functions}{%
\subsection{Additional functions}\label{additional-functions}}

\begin{itemize}
\tightlist
\item
  \href{https://microbiome.github.io/mia/reference/taxonomy-methods.html}{mapTaxonomy}
\item
  \href{https://microbiome.github.io/mia/reference/merge-methods.html}{mergeFeatures/mergeSamples}
\end{itemize}

\hypertarget{quality-control}{%
\chapter{Exploration and Quality Control}\label{quality-control}}

This chapter focuses on the quality control and exploration of
microbiome data and establishes commonly used descriptive
summaries. Familiarizing with the peculiarities of a given dataset is
the essential basis for any data analysis and model building.

The dataset should not suffer from severe technical biases, and you
should at least be aware of potential challenges, such as outliers,
biases, unexpected patterns and so forth. Standard summaries and
visualizations can help, and the rest comes with experience. The
exploration and quality control can be iterative processes.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\end{Highlighting}
\end{Shaded}

\hypertarget{abundance}{%
\section{Abundance}\label{abundance}}

Abundance visualization is an important data exploration
approach. \texttt{miaViz} offers the function \texttt{plotAbundanceDensity} to plot
the most abundant taxa with several options.

Next, a few demonstrations are shown, using the \citep{Lahti2014}
dataset. A Jitter plot based on relative abundance data, similar to
the one presented at \citep{Salosensaari2021} supplementary figure 1, could
be visualized as follows:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load example data}
\FunctionTok{library}\NormalTok{(miaTime)}
\FunctionTok{library}\NormalTok{(miaViz)}
\FunctionTok{data}\NormalTok{(hitchip1006)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ hitchip1006}

\CommentTok{\# Add relative abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Use argument names}
\CommentTok{\# assay.type / assay.type / assay.type}
\CommentTok{\# depending on the mia package version}
\FunctionTok{plotAbundanceDensity}\NormalTok{(tse, }\AttributeTok{layout =} \StringTok{"jitter"}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
                     \AttributeTok{n =} \DecValTok{40}\NormalTok{, }\AttributeTok{point\_size=}\DecValTok{1}\NormalTok{, }\AttributeTok{point\_shape=}\DecValTok{19}\NormalTok{, }\AttributeTok{point\_alpha=}\FloatTok{0.1}\NormalTok{) }\SpecialCharTok{+} 
                     \FunctionTok{scale\_x\_log10}\NormalTok{(}\AttributeTok{label=}\NormalTok{scales}\SpecialCharTok{::}\NormalTok{percent)}
\end{Highlighting}
\end{Shaded}

\includegraphics{12_quality_control_files/figure-latex/unnamed-chunk-2-1.pdf}

The relative abundance values for the top-5 taxonomic features can be
visualized as a density plot over a log scaled axis, with
``nationality'' indicated by colors:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plotAbundanceDensity}\NormalTok{(tse, }\AttributeTok{layout =} \StringTok{"density"}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
                     \AttributeTok{n =} \DecValTok{5}\NormalTok{, }\AttributeTok{colour\_by=}\StringTok{"nationality"}\NormalTok{, }\AttributeTok{point\_alpha=}\DecValTok{1}\SpecialCharTok{/}\DecValTok{10}\NormalTok{) }\SpecialCharTok{+}
    \FunctionTok{scale\_x\_log10}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\includegraphics{12_quality_control_files/figure-latex/unnamed-chunk-3-1.pdf}

\hypertarget{prevalence}{%
\section{Prevalence}\label{prevalence}}

Prevalence quantifies the frequency of samples where certain microbes
were detected (above a given detection threshold). The prevalence can
be given as sample size (N) or percentage (unit interval).

Investigating prevalence allows you either to focus on changes which
pertain to the majority of the samples, or identify rare microbes,
which may be \emph{conditionally abundant} in a small number of samples.

The population prevalence (frequency) at a 1\% relative abundance
threshold (\texttt{detection\ =\ 1/100} and \texttt{as\_relative\ =\ TRUE}), can look
like this.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getPrevalence}\NormalTok{(tse, }\AttributeTok{detection =} \DecValTok{1}\SpecialCharTok{/}\DecValTok{100}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{as\_relative =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Faecalibacterium prausnitzii et rel.           Ruminococcus obeum et rel. 
##                               0.9522                               0.9140 
##   Oscillospira guillermondii et rel.        Clostridium symbiosum et rel. 
##                               0.8801                               0.8714 
##     Subdoligranulum variable at rel.     Clostridium orbiscindens et rel. 
##                               0.8358                               0.8315
\end{verbatim}

The function arguments \texttt{detection} and \texttt{as\_relative} can also be used
to access, how many samples do pass a threshold for raw counts. Here,
the population prevalence (frequency) at the absolute abundance
threshold (\texttt{as\_relative\ =\ FALSE}) at read count 1 (\texttt{detection\ =\ 1}) is
accessed.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getPrevalence}\NormalTok{(tse, }\AttributeTok{detection =} \DecValTok{1}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                   \AttributeTok{as\_relative =} \ConstantTok{FALSE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##            Uncultured Mollicutes      Uncultured Clostridiales II 
##                                1                                1 
##       Uncultured Clostridiales I               Tannerella et rel. 
##                                1                                1 
##   Sutterella wadsworthia et rel. Subdoligranulum variable at rel. 
##                                1                                1
\end{verbatim}

If the output should be used for subsetting or storing the data in the
\texttt{rowData}, set \texttt{sort\ =\ FALSE}.

\hypertarget{prevalence-analysis}{%
\subsection{Prevalence analysis}\label{prevalence-analysis}}

To investigate microbiome prevalence at a selected taxonomic level, two
approaches are available.

First the data can be agglomerated to the taxonomic level and \texttt{getPrevalence}
applied on the resulting object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate taxa abundances to Phylum level, and add the new table}
\CommentTok{\# to the altExp slot}
\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\StringTok{"Phylum"}\NormalTok{)}
\CommentTok{\# Check prevalence for the Phylum abundance table from the altExp slot}
\FunctionTok{head}\NormalTok{(}\FunctionTok{getPrevalence}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{), }\AttributeTok{detection =} \DecValTok{1}\SpecialCharTok{/}\DecValTok{100}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{TRUE}\NormalTok{,}
                   \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{as\_relative =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      Firmicutes   Bacteroidetes  Actinobacteria  Proteobacteria Verrucomicrobia 
##       1.0000000       0.9852302       0.4821894       0.2988705       0.1277150 
##   Cyanobacteria 
##       0.0008688
\end{verbatim}

Alternatively, the \texttt{rank} argument could be set to perform the
agglomeration on the fly.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getPrevalence}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{, }\AttributeTok{detection =} \DecValTok{1}\SpecialCharTok{/}\DecValTok{100}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{TRUE}\NormalTok{,}
                   \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{as\_relative =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      Firmicutes   Bacteroidetes  Actinobacteria  Proteobacteria Verrucomicrobia 
##       1.0000000       0.9852302       0.4821894       0.2988705       0.1277150 
##   Cyanobacteria 
##       0.0008688
\end{verbatim}

Note that, by default, \texttt{na.rm\ =\ TRUE} is used for agglomeration in
\texttt{getPrevalence}, whereas the default for \texttt{mergeFeaturesByRank} is
\texttt{FALSE} to prevent accidental data loss.

If you only need the names of the prevalent taxa, \texttt{getPrevalentFeatures}
is available. This returns the taxa that exceed the given prevalence
and detection thresholds.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{getPrevalentFeatures}\NormalTok{(tse, }\AttributeTok{detection =} \DecValTok{0}\NormalTok{, }\AttributeTok{prevalence =} \DecValTok{50}\SpecialCharTok{/}\DecValTok{100}\NormalTok{)}
\NormalTok{prev }\OtherTok{\textless{}{-}} \FunctionTok{getPrevalentFeatures}\NormalTok{(tse, }\AttributeTok{detection =} \DecValTok{0}\NormalTok{, }\AttributeTok{prevalence =} \DecValTok{50}\SpecialCharTok{/}\DecValTok{100}\NormalTok{,}
                         \AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{prev}
\end{Highlighting}
\end{Shaded}

Note that the \texttt{detection} and \texttt{prevalence} thresholds are not the same, since
\texttt{detection} can be applied to relative counts or absolute counts depending on
whether \texttt{as\_relative} is set \texttt{TRUE} or \texttt{FALSE}

The function `getPrevalentAbundance' can be used to check the total
relative abundance of the prevalent taxa (between 0 and 1).

\hypertarget{rare-taxa}{%
\subsection{Rare taxa}\label{rare-taxa}}

Related functions are available for the analysis of rare taxa
(\texttt{rareMembers}; \texttt{rareAbundance}; \texttt{lowAbundance}, \texttt{getRareFeatures},
\texttt{subsetByRareFeatures}).

\hypertarget{plotting-prevalence}{%
\subsection{Plotting prevalence}\label{plotting-prevalence}}

To plot the prevalence, add the prevalence of each taxon to
\texttt{rowData}. Here, we are analysing the Phylum level abundances, which
are stored in the \texttt{altExp} slot.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rowData}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{prevalence }\OtherTok{\textless{}{-}} 
    \FunctionTok{getPrevalence}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{), }\AttributeTok{detection =} \DecValTok{1}\SpecialCharTok{/}\DecValTok{100}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{FALSE}\NormalTok{,}
                  \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{as\_relative =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The prevalences can then be plotted using the plotting functions from
the \texttt{scater} package.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(scater)}
\FunctionTok{plotRowData}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{), }\StringTok{"prevalence"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"Phylum"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{12_quality_control_files/figure-latex/unnamed-chunk-7-1.pdf}

The prevalence can also be visualized on the taxonomic tree with the
\texttt{miaViz} package.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{altExps}\NormalTok{(tse) }\OtherTok{\textless{}{-}} \FunctionTok{splitByRanks}\NormalTok{(tse)}
\FunctionTok{altExps}\NormalTok{(tse) }\OtherTok{\textless{}{-}}
   \FunctionTok{lapply}\NormalTok{(}\FunctionTok{altExps}\NormalTok{(tse),}
          \ControlFlowTok{function}\NormalTok{(y)\{}
              \FunctionTok{rowData}\NormalTok{(y)}\SpecialCharTok{$}\NormalTok{prevalence }\OtherTok{\textless{}{-}} 
                  \FunctionTok{getPrevalence}\NormalTok{(y, }\AttributeTok{detection =} \DecValTok{1}\SpecialCharTok{/}\DecValTok{100}\NormalTok{, }\AttributeTok{sort =} \ConstantTok{FALSE}\NormalTok{,}
                                \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{as\_relative =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{              y}
\NormalTok{          \})}
\NormalTok{top\_phyla }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{),}
                        \AttributeTok{method=}\StringTok{"prevalence"}\NormalTok{,}
                        \AttributeTok{top=}\NormalTok{5L,}
                        \AttributeTok{assay.type=}\StringTok{"counts"}\NormalTok{)}
\NormalTok{top\_phyla\_mean }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse,}\StringTok{"Phylum"}\NormalTok{),}
                             \AttributeTok{method=}\StringTok{"mean"}\NormalTok{,}
                             \AttributeTok{top=}\NormalTok{5L,}
                             \AttributeTok{assay.type=}\StringTok{"counts"}\NormalTok{)}
\NormalTok{x }\OtherTok{\textless{}{-}} \FunctionTok{unsplitByRanks}\NormalTok{(tse, }\AttributeTok{ranks =} \FunctionTok{taxonomyRanks}\NormalTok{(tse)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{6}\NormalTok{])}
\NormalTok{x }\OtherTok{\textless{}{-}} \FunctionTok{addTaxonomyTree}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

After some preparation, the data is assembled and can be plotted with
\texttt{plotRowTree}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(miaViz)}
\FunctionTok{plotRowTree}\NormalTok{(x[}\FunctionTok{rowData}\NormalTok{(x)}\SpecialCharTok{$}\NormalTok{Phylum }\SpecialCharTok{\%in\%}\NormalTok{ top\_phyla,],}
            \AttributeTok{edge\_colour\_by =} \StringTok{"Phylum"}\NormalTok{,}
            \AttributeTok{tip\_colour\_by =} \StringTok{"prevalence"}\NormalTok{,}
            \AttributeTok{node\_colour\_by =} \StringTok{"prevalence"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{12_quality_control_files/figure-latex/plot-prev-prev-1.pdf}
\caption{\label{fig:plot-prev-prev}Prevalence of top phyla as judged by prevalence}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plotRowTree}\NormalTok{(x[}\FunctionTok{rowData}\NormalTok{(x)}\SpecialCharTok{$}\NormalTok{Phylum }\SpecialCharTok{\%in\%}\NormalTok{ top\_phyla\_mean,],}
            \AttributeTok{edge\_colour\_by =} \StringTok{"Phylum"}\NormalTok{,}
            \AttributeTok{tip\_colour\_by =} \StringTok{"prevalence"}\NormalTok{,}
            \AttributeTok{node\_colour\_by =} \StringTok{"prevalence"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{12_quality_control_files/figure-latex/plot-prev-mean-1.pdf}
\caption{\label{fig:plot-prev-mean}Prevalence of top phyla as judged by mean abundance}
\end{figure}

\hypertarget{qc}{%
\section{Quality control}\label{qc}}

Next, let us load the \texttt{GlobalPatterns} dataset to illustrate standard
microbiome data summaries.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns }
\end{Highlighting}
\end{Shaded}

\hypertarget{top-taxa}{%
\subsection{Top taxa}\label{top-taxa}}

The \texttt{getTopFeatures} identifies top taxa in the data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Pick the top taxa}
\NormalTok{top\_features }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse, }\AttributeTok{method=}\StringTok{"median"}\NormalTok{, }\AttributeTok{top=}\DecValTok{10}\NormalTok{)}

\CommentTok{\# Check the information for these}
\FunctionTok{rowData}\NormalTok{(tse)[top\_features, }\FunctionTok{taxonomyRanks}\NormalTok{(tse)]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 10 rows and 7 columns
##            Kingdom         Phylum               Class             Order
##        <character>    <character>         <character>       <character>
## 549656    Bacteria  Cyanobacteria         Chloroplast     Stramenopiles
## 331820    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
## 317182    Bacteria  Cyanobacteria         Chloroplast     Stramenopiles
## 94166     Bacteria Proteobacteria Gammaproteobacteria    Pasteurellales
## 279599    Bacteria  Cyanobacteria    Nostocophycideae        Nostocales
## 158660    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
## 329744    Bacteria Actinobacteria      Actinobacteria   Actinomycetales
## 326977    Bacteria Actinobacteria      Actinobacteria Bifidobacteriales
## 248140    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
## 550960    Bacteria Proteobacteria Gammaproteobacteria Enterobacteriales
##                    Family           Genus                Species
##               <character>     <character>            <character>
## 549656                 NA              NA                     NA
## 331820     Bacteroidaceae     Bacteroides                     NA
## 317182                 NA              NA                     NA
## 94166     Pasteurellaceae     Haemophilus Haemophilusparainflu..
## 279599        Nostocaceae  Dolichospermum                     NA
## 158660     Bacteroidaceae     Bacteroides                     NA
## 329744             ACK-M1              NA                     NA
## 326977 Bifidobacteriaceae Bifidobacterium Bifidobacteriumadole..
## 248140     Bacteroidaceae     Bacteroides      Bacteroidescaccae
## 550960 Enterobacteriaceae     Providencia                     NA
\end{verbatim}

\hypertarget{library-size-read-count}{%
\subsection{Library size / read count}\label{library-size-read-count}}

The total counts/sample can be calculated using \texttt{perCellQCMetrics}/\texttt{addPerCellQC} from the \texttt{scater} package. The former one
just calculates the values, whereas the latter one directly adds them to
\texttt{colData}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(scater)}
\FunctionTok{perCellQCMetrics}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 26 rows and 3 columns
##               sum  detected     total
##         <numeric> <numeric> <numeric>
## CL3        864077      6964    864077
## CC1       1135457      7679   1135457
## SV1        697509      5729    697509
## M31Fcsw   1543451      2667   1543451
## M11Fcsw   2076476      2574   2076476
## ...           ...       ...       ...
## TS28       937466      2679    937466
## TS29      1211071      2629   1211071
## Even1     1216137      4213   1216137
## Even2      971073      3130    971073
## Even3     1078241      2776   1078241
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{addPerCellQC}\NormalTok{(tse)}
\FunctionTok{colData}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 26 rows and 10 columns
##         X.SampleID   Primer Final_Barcode Barcode_truncated_plus_T
##           <factor> <factor>      <factor>                 <factor>
## CL3        CL3      ILBC_01        AACGCA                   TGCGTT
## CC1        CC1      ILBC_02        AACTCG                   CGAGTT
## SV1        SV1      ILBC_03        AACTGT                   ACAGTT
## M31Fcsw    M31Fcsw  ILBC_04        AAGAGA                   TCTCTT
## M11Fcsw    M11Fcsw  ILBC_05        AAGCTG                   CAGCTT
## ...            ...      ...           ...                      ...
## TS28         TS28   ILBC_25        ACCAGA                   TCTGGT
## TS29         TS29   ILBC_26        ACCAGC                   GCTGGT
## Even1        Even1  ILBC_27        ACCGCA                   TGCGGT
## Even2        Even2  ILBC_28        ACCTCG                   CGAGGT
## Even3        Even3  ILBC_29        ACCTGT                   ACAGGT
##         Barcode_full_length SampleType
##                    <factor>   <factor>
## CL3             CTAGCGTGCGT      Soil 
## CC1             CATCGACGAGT      Soil 
## SV1             GTACGCACAGT      Soil 
## M31Fcsw         TCGACATCTCT      Feces
## M11Fcsw         CGACTGCAGCT      Feces
## ...                     ...        ...
## TS28            GCATCGTCTGG      Feces
## TS29            CTAGTCGCTGG      Feces
## Even1           TGACTCTGCGG      Mock 
## Even2           TCTGATCGAGG      Mock 
## Even3           AGAGAGACAGG      Mock 
##                                        Description       sum  detected
##                                           <factor> <numeric> <numeric>
## CL3     Calhoun South Carolina Pine soil, pH 4.9      864077      6964
## CC1     Cedar Creek Minnesota, grassland, pH 6.1     1135457      7679
## SV1     Sevilleta new Mexico, desert scrub, pH 8.3    697509      5729
## M31Fcsw M3, Day 1, fecal swab, whole body study      1543451      2667
## M11Fcsw M1, Day 1, fecal swab, whole body study      2076476      2574
## ...                                            ...       ...       ...
## TS28                                       Twin #1    937466      2679
## TS29                                       Twin #2   1211071      2629
## Even1                                      Even1     1216137      4213
## Even2                                      Even2      971073      3130
## Even3                                      Even3     1078241      2776
##             total
##         <numeric>
## CL3        864077
## CC1       1135457
## SV1        697509
## M31Fcsw   1543451
## M11Fcsw   2076476
## ...           ...
## TS28       937466
## TS29      1211071
## Even1     1216137
## Even2      971073
## Even3     1078241
\end{verbatim}

The distribution of calculated library sizes can be visualized as a
histogram (left), or by sorting the samples by library size (right).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}

\NormalTok{p1 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse))) }\SpecialCharTok{+}
        \FunctionTok{geom\_histogram}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ sum), }\AttributeTok{color =} \StringTok{"black"}\NormalTok{, }\AttributeTok{fill =} \StringTok{"gray"}\NormalTok{, }\AttributeTok{bins =} \DecValTok{30}\NormalTok{) }\SpecialCharTok{+}
        \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Library size"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Frequency (n)"}\NormalTok{) }\SpecialCharTok{+} 
        \CommentTok{\# scale\_x\_log10(breaks = scales::trans\_breaks("log10", function(x) 10\^{}x), }
        \CommentTok{\# labels = scales::trans\_format("log10", scales::math\_format(10\^{}.x))) +}
        \FunctionTok{theme\_bw}\NormalTok{() }\SpecialCharTok{+}
        \FunctionTok{theme}\NormalTok{(}\AttributeTok{panel.grid.major =} \FunctionTok{element\_blank}\NormalTok{(), }\CommentTok{\# Removes the grid}
          \AttributeTok{panel.grid.minor =} \FunctionTok{element\_blank}\NormalTok{(),}
          \AttributeTok{panel.border =} \FunctionTok{element\_blank}\NormalTok{(),}
          \AttributeTok{panel.background =} \FunctionTok{element\_blank}\NormalTok{(),}
          \AttributeTok{axis.line =} \FunctionTok{element\_line}\NormalTok{(}\AttributeTok{colour =} \StringTok{"black"}\NormalTok{)) }\CommentTok{\# Adds y{-}axis}

\FunctionTok{library}\NormalTok{(dplyr)}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)) }\SpecialCharTok{\%\textgreater{}\%}
        \FunctionTok{arrange}\NormalTok{(sum) }\SpecialCharTok{\%\textgreater{}\%}
        \FunctionTok{mutate}\NormalTok{(}\AttributeTok{index =} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{n}\NormalTok{())}
\NormalTok{p2 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{y =}\NormalTok{ index, }\AttributeTok{x =}\NormalTok{ sum}\SpecialCharTok{/}\FloatTok{1e6}\NormalTok{)) }\SpecialCharTok{+}
        \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+}  
        \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Library size (million reads)"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Sample index"}\NormalTok{) }\SpecialCharTok{+}  
        \FunctionTok{theme\_bw}\NormalTok{() }\SpecialCharTok{+}
        \FunctionTok{theme}\NormalTok{(}\AttributeTok{panel.grid.major =} \FunctionTok{element\_blank}\NormalTok{(), }\CommentTok{\# Removes the grid}
          \AttributeTok{panel.grid.minor =} \FunctionTok{element\_blank}\NormalTok{(),}
          \AttributeTok{panel.border =} \FunctionTok{element\_blank}\NormalTok{(),}
          \AttributeTok{panel.background =} \FunctionTok{element\_blank}\NormalTok{(),}
          \AttributeTok{axis.line =} \FunctionTok{element\_line}\NormalTok{(}\AttributeTok{colour =} \StringTok{"black"}\NormalTok{)) }\CommentTok{\# Adds y{-}axis}

\FunctionTok{library}\NormalTok{(patchwork)}
\NormalTok{p1 }\SpecialCharTok{+}\NormalTok{ p2}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{12_quality_control_files/figure-latex/plot-viz-lib-size-1-1.pdf}
\caption{\label{fig:plot-viz-lib-size-1}Library size distribution.}
\end{figure}

Library sizes other variables from \texttt{colData} can be
visualized by using specified function called \texttt{plotColData}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}
\CommentTok{\# Sort samples by read count, order the factor levels, and store back to tse as DataFrame}
\CommentTok{\# }\AlertTok{TODO}\CommentTok{: plotColData could include an option for sorting samples based on colData variables}
\FunctionTok{colData}\NormalTok{(tse) }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)) }\SpecialCharTok{\%\textgreater{}\%}
                 \FunctionTok{arrange}\NormalTok{(X.SampleID) }\SpecialCharTok{\%\textgreater{}\%}
             \FunctionTok{mutate}\NormalTok{(}\AttributeTok{X.SampleID =} \FunctionTok{factor}\NormalTok{(X.SampleID, }\AttributeTok{levels=}\NormalTok{X.SampleID)) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{         DataFrame}
\FunctionTok{plotColData}\NormalTok{(tse,}\StringTok{"sum"}\NormalTok{,}\StringTok{"X.SampleID"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"SampleType"}\NormalTok{) }\SpecialCharTok{+} 
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle =} \DecValTok{45}\NormalTok{, }\AttributeTok{hjust=}\DecValTok{1}\NormalTok{)) }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{y =} \StringTok{"Library size (N)"}\NormalTok{, }\AttributeTok{x =} \StringTok{"Sample ID"}\NormalTok{)       }
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{12_quality_control_files/figure-latex/plot-viz-lib-size-2-1.pdf}
\caption{\label{fig:plot-viz-lib-size-2}Library sizes per sample.}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plotColData}\NormalTok{(tse,}\StringTok{"sum"}\NormalTok{,}\StringTok{"SampleType"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"SampleType"}\NormalTok{) }\SpecialCharTok{+} 
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle =} \DecValTok{45}\NormalTok{, }\AttributeTok{hjust=}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{12_quality_control_files/figure-latex/plot-viz-lib-size-3-1.pdf}
\caption{\label{fig:plot-viz-lib-size-3}Library sizes per sample type.}
\end{figure}

In addition, data can be rarefied with
\href{https://microbiome.github.io/mia/reference/subsampleCounts.html}{subsampleCounts},
which normalises the samples to an equal number of reads. However,
this practice has been discouraged for the analysis of differentially
abundant microorganisms (see \citep{mcmurdie2014waste}).

\hypertarget{contaminant-sequences}{%
\subsection{Contaminant sequences}\label{contaminant-sequences}}

Samples might be contaminated with exogenous sequences. The impact of
each contaminant can be estimated based on their frequencies and
concentrations across the samples.

The following \href{https://microbiome.github.io/mia/reference/isContaminant.html}{decontam
functions}
are based on the \citep{davis2018simple} and support such functionality:

\begin{itemize}
\tightlist
\item
  \texttt{isContaminant}, \texttt{isNotContaminant}
\item
  \texttt{addContaminantQC}, \texttt{addNotContaminantQC}
\end{itemize}

\hypertarget{taxonomic-information}{%
\chapter{Taxonomic Information}\label{taxonomic-information}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}
\end{Highlighting}
\end{Shaded}

Taxonomic information is a key part of analyzing microbiome data and without
it, any type of data analysis probably will not make much sense. However,
the degree of detail of taxonomic information differs depending on the dataset
and annotation data used.

Therefore, the mia package expects a loose assembly of taxonomic information
and assumes certain key aspects:

\begin{itemize}
\tightlist
\item
  Taxonomic information is given as character vectors or factors in the
  \texttt{rowData} of a \texttt{SummarizedExperiment} object.
\item
  The columns containing the taxonomic information must be named \texttt{domain},
  \texttt{kingdom}, \texttt{phylum}, \texttt{class}, \texttt{order}, \texttt{family}, \texttt{genus}, \texttt{species} or with
  a capital first letter.
\item
  the columns must be given in the order shown above
\item
  column can be omited, but the order must remain
\end{itemize}

In this chapter, we will refer to co-abundant groups as CAGs, which are
clusters of taxa that co-vary across samples.

\hypertarget{assigning-taxonomic-information.}{%
\section{Assigning taxonomic information.}\label{assigning-taxonomic-information.}}

There are a number of methods to assign taxonomic information. We like to give
a short introduction about the methods available without ranking one over the
other. This has to be your choice based on the result for the individual
dataset.

\hypertarget{dada2}{%
\subsection{dada2}\label{dada2}}

The dada2 package \citep{Callahan2016dada2} implements the \texttt{assignTaxonomy}
function, which takes as input the ASV sequences associated with each
row of data and a training dataset. For more information visit the
\href{https://benjjneb.github.io/dada2/assign.html}{dada2 homepage}.

\hypertarget{decipher}{%
\subsection{DECIPHER}\label{decipher}}

The DECIPHER package \citep{R_DECIPHER} implements the \texttt{IDTAXA} algorithm to assign
either taxonomic information or function information. For \texttt{mia}
only the first option is of interest for now and more information can be
found on the \href{http://www2.decipher.codes/Classification.html}{DECIPHER website}.

\hypertarget{functions-to-access-taxonomic-information}{%
\section{Functions to access taxonomic information}\label{functions-to-access-taxonomic-information}}

\texttt{checkTaxonomy} checks whether the taxonomic information is usable for \texttt{mia}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{checkTaxonomy}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

Since the \texttt{rowData} can contain other data, \texttt{taxonomyRanks} will return the
columns \texttt{mia} assumes to contain the taxonomic information.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{taxonomyRanks}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"
\end{verbatim}

This can then be used to subset the \texttt{rowData} to columns needed.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{rowData}\NormalTok{(tse)[, }\FunctionTok{taxonomyRanks}\NormalTok{(tse)]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 19216 rows and 7 columns
##            Kingdom        Phylum        Class        Order        Family
##        <character>   <character>  <character>  <character>   <character>
## 549322     Archaea Crenarchaeota Thermoprotei           NA            NA
## 522457     Archaea Crenarchaeota Thermoprotei           NA            NA
## 951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 244423     Archaea Crenarchaeota        Sd-NA           NA            NA
## 586076     Archaea Crenarchaeota        Sd-NA           NA            NA
## ...            ...           ...          ...          ...           ...
## 278222    Bacteria           SR1           NA           NA            NA
## 463590    Bacteria           SR1           NA           NA            NA
## 535321    Bacteria           SR1           NA           NA            NA
## 200359    Bacteria           SR1           NA           NA            NA
## 271582    Bacteria           SR1           NA           NA            NA
##              Genus                Species
##        <character>            <character>
## 549322          NA                     NA
## 522457          NA                     NA
## 951     Sulfolobus Sulfolobusacidocalda..
## 244423          NA                     NA
## 586076          NA                     NA
## ...            ...                    ...
## 278222          NA                     NA
## 463590          NA                     NA
## 535321          NA                     NA
## 200359          NA                     NA
## 271582          NA                     NA
\end{verbatim}

\texttt{taxonomyRankEmpty} checks for empty values in the given \texttt{rank} and returns a
logical vector of \texttt{length(x)}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{all}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{taxonomyRankEmpty}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Kingdom"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{table}\NormalTok{(}\FunctionTok{taxonomyRankEmpty}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## FALSE  TRUE 
##  8008 11208
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{table}\NormalTok{(}\FunctionTok{taxonomyRankEmpty}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Species"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## FALSE  TRUE 
##  1413 17803
\end{verbatim}

\texttt{getTaxonomyLabels} is a multi-purpose function, which turns taxonomic
information into a character vector of \texttt{length(x)}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getTaxonomyLabels}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Class:Thermoprotei"               "Class:Thermoprotei_1"            
## [3] "Species:Sulfolobusacidocaldarius" "Class:Sd-NA"                     
## [5] "Class:Sd-NA_1"                    "Class:Sd-NA_2"
\end{verbatim}

By default, this will use the lowest non-empty information to construct a
string with the following scheme \texttt{level:value}. If all levels are the same,
this part is omitted, but can be added by setting \texttt{with\_rank\ =\ TRUE}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{phylum }\OtherTok{\textless{}{-}} \SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum) }\SpecialCharTok{\&}
    \FunctionTok{vapply}\NormalTok{(}\FunctionTok{data.frame}\NormalTok{(}\FunctionTok{apply}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)[, }\FunctionTok{taxonomyRanks}\NormalTok{(tse)[}\DecValTok{3}\SpecialCharTok{:}\DecValTok{7}\NormalTok{]], 1L, is.na)), all, }\FunctionTok{logical}\NormalTok{(}\DecValTok{1}\NormalTok{))}
\FunctionTok{head}\NormalTok{(}\FunctionTok{getTaxonomyLabels}\NormalTok{(tse[phylum,]))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Crenarchaeota"    "Crenarchaeota_1"  "Crenarchaeota_2"  "Actinobacteria"  
## [5] "Actinobacteria_1" "Spirochaetes"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getTaxonomyLabels}\NormalTok{(tse[phylum,], }\AttributeTok{with\_rank =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Phylum:Crenarchaeota"    "Phylum:Crenarchaeota_1" 
## [3] "Phylum:Crenarchaeota_2"  "Phylum:Actinobacteria"  
## [5] "Phylum:Actinobacteria_1" "Phylum:Spirochaetes"
\end{verbatim}

By default the return value of \texttt{getTaxonomyLabels} contains only
unique elements by passing it through \texttt{make.unique}. This step can be
omitted by setting \texttt{make\_unique\ =\ FALSE}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getTaxonomyLabels}\NormalTok{(tse[phylum,], }\AttributeTok{with\_rank =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{make\_unique =} \ConstantTok{FALSE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Phylum:Crenarchaeota"  "Phylum:Crenarchaeota"  "Phylum:Crenarchaeota" 
## [4] "Phylum:Actinobacteria" "Phylum:Actinobacteria" "Phylum:Spirochaetes"
\end{verbatim}

To apply the loop resolving function \texttt{resolveLoop} from the
\texttt{TreeSummarizedExperiment} package \citep{R_TreeSummarizedExperiment} within
\texttt{getTaxonomyLabels}, set \texttt{resolve\_loops\ =\ TRUE}.

The function \texttt{getUniqueFeatures} gives a list of unique taxa for the
specified taxonomic rank.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{getUniqueFeatures}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Crenarchaeota"  "Euryarchaeota"  "Actinobacteria" "Spirochaetes"  
## [5] "MVP-15"         "Proteobacteria"
\end{verbatim}

\hypertarget{fly-tree}{%
\subsection{Generate a taxonomic tree on the fly}\label{fly-tree}}

To create a taxonomic tree, \texttt{taxonomyTree} used the information and returns a
\texttt{phylo} object. Duplicate information from the \texttt{rowData} is removed.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{taxonomyTree}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Phylogenetic tree with 1645 tips and 1089 internal nodes.
## 
## Tip labels:
##   Species:Cenarchaeumsymbiosum, Species:pIVWA5, Species:CandidatusNitrososphaeragargensis, Species:SCA1145, Species:SCA1170, Species:Sulfolobusacidocaldarius, ...
## Node labels:
##   root:ALL, Kingdom:Archaea, Phylum:Crenarchaeota, Class:C2, Class:Sd-NA, Class:Thaumarchaeota, ...
## 
## Rooted; includes branch lengths.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{addTaxonomyTree}\NormalTok{(tse)}
\NormalTok{tse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): Class:Thermoprotei Class:Thermoprotei ... Phylum:SR1
##   Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (1645 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

The implementation is based on the \texttt{toTree} function from the
\texttt{TreeSummarizedExperiment} package \citep{R_TreeSummarizedExperiment}.

\hypertarget{data-agglomeration}{%
\section{Data agglomeration}\label{data-agglomeration}}

One of the main applications of taxonomic information in regards to count data
is to agglomerate count data on taxonomic levels and track the influence of
changing conditions through these levels. For this \texttt{mia} contains the
\texttt{mergeFeaturesByRank} function. The ideal location to store the agglomerated data
is as an alternative experiment.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Family"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Family"}\NormalTok{,}
                                           \AttributeTok{agglomerateTree =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Family"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 603 26 
## metadata(1): agglomerated_by_rank
## assays(2): counts relabundance
## rownames(603): Class:Thermoprotei Family:Sulfolobaceae ...
##   Family:Thermodesulfobiaceae Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (603 rows)
## rowTree: 1 phylo tree(s) (496 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

If multiple assays (counts and relabundance) exist, both will be agglomerated.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assayNames}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "counts"       "relabundance"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assayNames}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Family"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "counts"       "relabundance"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assay}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Family"}\NormalTok{), }\StringTok{"relabundance"}\NormalTok{)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\DecValTok{7}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                            CL3       CC1 SV1 M31Fcsw M11Fcsw M31Plmr   M11Plmr
## Class:Thermoprotei   0.0000000 0.000e+00   0       0       0       0 0.000e+00
## Family:Sulfolobaceae 0.0000000 0.000e+00   0       0       0       0 2.305e-06
## Class:Sd-NA          0.0000000 0.000e+00   0       0       0       0 0.000e+00
## Order:NRP-J          0.0001991 2.070e-04   0       0       0       0 6.914e-06
## Family:SAGMA-X       0.0000000 6.165e-06   0       0       0       0 0.000e+00
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assay}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Family"}\NormalTok{), }\StringTok{"counts"}\NormalTok{)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{5}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\DecValTok{7}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                      CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## Class:Thermoprotei     0   0   0       0       0       0       0
## Family:Sulfolobaceae   0   0   0       0       0       0       1
## Class:Sd-NA            0   0   0       0       0       0       0
## Order:NRP-J          172 235   0       0       0       0       3
## Family:SAGMA-X         0   7   0       0       0       0       0
\end{verbatim}

\texttt{altExpNames} now consists of \texttt{Family} level data. This can be extended to use
any taxonomic level listed in \texttt{mia::taxonomyRanks(tse)}.

Rare taxa can also be aggregated into a single group ``Other'' instead of
filtering them out. A suitable function for this is \texttt{mergeFeaturesByPrevalence}.
The number of rare taxa is higher on the species level, which causes the need
for data agglomeration by prevalence.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Species\_byPrevalence"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByPrevalence}\NormalTok{(tse, }
                                                               \AttributeTok{rank =} \StringTok{"Species"}\NormalTok{, }
                                                               \AttributeTok{other\_label =} \StringTok{"Other"}\NormalTok{, }
                                                               \AttributeTok{prevalence =} \DecValTok{5} \SpecialCharTok{/} \DecValTok{100}\NormalTok{, }
                                                               \AttributeTok{detection =} \DecValTok{1} \SpecialCharTok{/} \DecValTok{100}\NormalTok{, }
                                                               \AttributeTok{as\_relative =}\NormalTok{ T)}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Species\_byPrevalence"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 113 26 
## metadata(2): agglomerated_by_rank agglomerated_by_rank
## assays(2): counts relabundance
## rownames(113): Family:MarinegroupII Order:Actinomycetales ...
##   Species:Veillonellaparvula Other
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{assay}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"Species\_byPrevalence"}\NormalTok{), }\StringTok{"relabundance"}\NormalTok{)[}\DecValTok{88}\SpecialCharTok{:}\DecValTok{92}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\DecValTok{7}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                           CL3       CC1       SV1   M31Fcsw   M11Fcsw   M31Plmr
## Genus:Luteolibacter 9.837e-05 8.076e-04 0.0001233 3.239e-06 5.297e-06 4.034e-05
## Genus:MC18          5.664e-02 2.110e-01 0.0289530 4.471e-05 2.649e-05 1.071e-04
## Class:Phycisphaerae 6.958e-03 1.816e-02 0.0585985 5.831e-06 4.816e-06 4.729e-05
## Order:envOPS12      2.315e-05 7.046e-06 0.0000000 1.296e-06 4.816e-07 1.391e-06
## Order:Clostridiales 1.017e-03 1.629e-04 0.0003857 1.473e-02 2.005e-02 1.647e-03
##                       M11Plmr
## Genus:Luteolibacter 1.291e-04
## Genus:MC18          7.882e-04
## Class:Phycisphaerae 4.033e-04
## Order:envOPS12      4.609e-06
## Order:Clostridiales 1.719e-03
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Saving the tse for later}
\NormalTok{tseGlobalPatterns }\OtherTok{\textless{}{-}}\NormalTok{ tse}
\end{Highlighting}
\end{Shaded}

\hypertarget{taxa-clustering}{%
\subsection{Taxa clustering}\label{taxa-clustering}}

Another way to agglomerate the data is to cluster the taxa. To do so,
we usually start by doing a compositionality aware transformation such as CLR,
followed by the application of a standard clustering method.

Here is an example that does a CLR transformation followed by the hierarchical
clustering algorithm.

First, we import the library \texttt{bluster} that simplifies the clustering.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(bluster)}
\end{Highlighting}
\end{Shaded}

Then we do the CLR transform followed by the clustering. We will cluster with
two different distances: the euclidean distance and the kendall distance.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get the data}
\FunctionTok{data}\NormalTok{(}\StringTok{"peerj13075"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ peerj13075}

\CommentTok{\# The result of the CLR transform is stored in the assay clr}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{method =} \StringTok{"z"}\NormalTok{, }
                      \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{)}

\CommentTok{\# Cluster (with euclidean distance) on the features of the z assay}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{cluster}\NormalTok{(tse,}
               \AttributeTok{assay.type =} \StringTok{"z"}\NormalTok{,}
               \AttributeTok{clust.col =} \StringTok{"hclustEuclidean"}\NormalTok{,}
           \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{,}
               \FunctionTok{HclustParam}\NormalTok{(}\AttributeTok{dist.fun =}\NormalTok{ stats}\SpecialCharTok{::}\NormalTok{dist, }\AttributeTok{method =} \StringTok{"ward.D2"}\NormalTok{))}

\CommentTok{\# Declare the Kendall dissimilarity computation function}
\NormalTok{kendall\_dissimilarity }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(x) \{}
    \FunctionTok{as.dist}\NormalTok{(}\DecValTok{1} \SpecialCharTok{{-}} \FunctionTok{cor}\NormalTok{(}\FunctionTok{t}\NormalTok{(x), }\AttributeTok{method =} \StringTok{"kendall"}\NormalTok{))}
\NormalTok{\}}

\CommentTok{\# Cluster (with Kendall dissimilarity) on the features of the z assay}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{cluster}\NormalTok{(tse,}
               \AttributeTok{assay.type =} \StringTok{"z"}\NormalTok{,}
               \AttributeTok{clust.col =} \StringTok{"hclustKendall"}\NormalTok{,}
               \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{,            }
               \FunctionTok{HclustParam}\NormalTok{(}\AttributeTok{dist.fun =}\NormalTok{ kendall\_dissimilarity, }\AttributeTok{method =} \StringTok{"ward.D2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Let us store the resulting cluster indices in the \texttt{rowData} column specified
with the \texttt{clust.col} parameter.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Checking the clusters}
\NormalTok{clusters\_euclidean }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{hclustEuclidean}
\FunctionTok{head}\NormalTok{(clusters\_euclidean, }\DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 
##     1     2     1     1     1     1     3     4     3     2 
## Levels: 1 2 3 4 5
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{clusters\_kendall }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{hclustKendall}
\FunctionTok{head}\NormalTok{(clusters\_kendall, }\DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 
##     1     2     1     3     3     1     3     1     1     3 
## Levels: 1 2 3 4
\end{verbatim}

To better visualize the results and the distribution of the clusters, we can
plot the histogram of the clusters.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}
\FunctionTok{library}\NormalTok{(patchwork) }\CommentTok{\# TO arrange several plots as a grid}
\NormalTok{plot1 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)), }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ clusters\_euclidean)) }\SpecialCharTok{+}
    \FunctionTok{geom\_bar}\NormalTok{() }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"CAG size distribution (Euclidean distance)"}\NormalTok{,}
         \AttributeTok{x =} \StringTok{"Clusters"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Feature count (n)"}\NormalTok{)}
\NormalTok{plot2 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)), }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ clusters\_kendall)) }\SpecialCharTok{+}
    \FunctionTok{geom\_bar}\NormalTok{() }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"CAG size distribution (1 {-} tau)"}\NormalTok{,}
         \AttributeTok{x =} \StringTok{"Clusters"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Feature count (n)"}\NormalTok{)}
\NormalTok{plot1 }\SpecialCharTok{+}\NormalTok{ plot2 }\SpecialCharTok{+} \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{ncol =} \DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{11_taxonomic_information_files/figure-latex/taxa_clustering_histogram-1.pdf}

It's also possible to merge the rows by cluster.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Aggregate clusters as a sum of each cluster values}
\NormalTok{tse\_merged }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeatures}\NormalTok{(tse, clusters\_euclidean)}
\NormalTok{tse\_merged}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 5 58 
## metadata(0):
## assays(3): counts clr z
## rownames(5): 1 2 3 4 5
## rowData names(8): kingdom phylum ... hclustEuclidean hclustKendall
## colnames(58): ID1 ID2 ... ID57 ID58
## colData names(5): Sample Geographical_location Gender Age Diet
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

We can note that it worked as planned since there were 5 clusters and there are
now 5 rows.

\hypertarget{assay-transform}{%
\section{Data transformation}\label{assay-transform}}

Data transformations are common in microbiome analysis. Examples
include the logarithmic transformation, calculation of relative
abundances (percentages), and compositionality-aware transformations
such as the centered log-ratio transformation (clr).

In mia package, transformations are applied to abundance data. The transformed
abundance table is stored back to `assays'. mia includes transformation
function (`transformAssay()') which applies sample-wise or column-wise
transformation when MARGIN = `samples', feature-wise or row-wise transformation
when MARGIN = `features'.

For a complete list of available transformations and parameters, see function
\href{https://microbiome.github.io/mia/reference/transformAssay.html}{help}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ tseGlobalPatterns}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(}\AttributeTok{x =}\NormalTok{ tse, }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{, }
                      \AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{, }\AttributeTok{name =} \StringTok{"clr"}\NormalTok{)}

\FunctionTok{head}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"clr"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                                         CL3        CC1        SV1    M31Fcsw
## Class:Thermoprotei               -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Thermoprotei               -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Species:Sulfolobusacidocaldarius -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
##                                     M11Fcsw    M31Plmr    M11Plmr    F21Plmr
## Class:Thermoprotei               -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Class:Thermoprotei               -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Species:Sulfolobusacidocaldarius -4.947e-05 -4.931e-05 -4.658e-05 -4.671e-05
## Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
##                                     M31Tong    M11Tong   LMEpi24M   SLEpi20M
## Class:Thermoprotei               -4.846e-05 -4.257e-05 -4.756e-05 -4.837e-05
## Class:Thermoprotei               -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Species:Sulfolobusacidocaldarius -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
##                                      AQC1cm     AQC4cm     AQC7cm        NP2
## Class:Thermoprotei               -2.385e-05 -4.438e-06  2.787e-05 -4.731e-05
## Class:Thermoprotei               -4.660e-05 -4.568e-05 -4.428e-05 -4.915e-05
## Species:Sulfolobusacidocaldarius -4.660e-05 -4.652e-05 -4.777e-05 -4.915e-05
## Class:Sd-NA                      -4.660e-05 -3.726e-05 -3.090e-05 -4.915e-05
## Class:Sd-NA                      -4.660e-05 -4.568e-05 -4.719e-05 -4.915e-05
## Class:Sd-NA                      -4.660e-05 -4.610e-05 -4.603e-05 -4.915e-05
##                                         NP3        NP5    TRRsed1    TRRsed2
## Class:Thermoprotei               -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Thermoprotei               -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Species:Sulfolobusacidocaldarius -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
##                                     TRRsed3       TS28       TS29      Even1
## Class:Thermoprotei               -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Thermoprotei               -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Species:Sulfolobusacidocaldarius -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
##                                       Even2      Even3
## Class:Thermoprotei               -5.017e-05 -5.034e-05
## Class:Thermoprotei               -5.017e-05 -5.034e-05
## Species:Sulfolobusacidocaldarius -5.017e-05 -5.034e-05
## Class:Sd-NA                      -5.017e-05 -5.034e-05
## Class:Sd-NA                      -5.017e-05 -5.034e-05
## Class:Sd-NA                      -5.017e-05 -5.034e-05
\end{verbatim}

\begin{itemize}
\tightlist
\item
  In `pa' transformation, abundance table is converted to present/absent table.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"pa"}\NormalTok{)}

\FunctionTok{head}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"pa"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                                  CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## Class:Thermoprotei                 0   0   0       0       0       0       0
## Class:Thermoprotei                 0   0   0       0       0       0       0
## Species:Sulfolobusacidocaldarius   0   0   0       0       0       0       1
## Class:Sd-NA                        0   0   0       0       0       0       0
## Class:Sd-NA                        0   0   0       0       0       0       0
## Class:Sd-NA                        0   0   0       0       0       0       0
##                                  F21Plmr M31Tong M11Tong LMEpi24M SLEpi20M
## Class:Thermoprotei                     0       0       0        0        1
## Class:Thermoprotei                     0       0       0        0        0
## Species:Sulfolobusacidocaldarius       0       0       0        0        0
## Class:Sd-NA                            0       0       0        0        0
## Class:Sd-NA                            0       0       0        0        0
## Class:Sd-NA                            0       0       0        0        0
##                                  AQC1cm AQC4cm AQC7cm NP2 NP3 NP5 TRRsed1
## Class:Thermoprotei                    1      1      1   1   0   0       0
## Class:Thermoprotei                    0      1      1   0   0   0       0
## Species:Sulfolobusacidocaldarius      0      0      0   0   0   0       0
## Class:Sd-NA                           0      1      1   0   0   0       0
## Class:Sd-NA                           0      1      1   0   0   0       0
## Class:Sd-NA                           0      1      1   0   0   0       0
##                                  TRRsed2 TRRsed3 TS28 TS29 Even1 Even2 Even3
## Class:Thermoprotei                     0       0    0    0     0     0     0
## Class:Thermoprotei                     0       0    0    0     0     0     0
## Species:Sulfolobusacidocaldarius       0       0    0    0     0     0     0
## Class:Sd-NA                            0       0    0    0     0     0     0
## Class:Sd-NA                            0       0    0    0     0     0     0
## Class:Sd-NA                            0       0    0    0     0     0     0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# list of abundance tables that assays slot contains}
\FunctionTok{assays}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of length 4
## names(4): counts relabundance clr pa
\end{verbatim}

\hypertarget{community-diversity}{%
\chapter{Community Diversity}\label{community-diversity}}

Diversity estimates are a central topic in microbiome data analysis.

There are three commonly employed levels of diversity measurements,
which are trying to put a number on different aspects of the questions
associated with diversity \citep{Whittaker1960}.

Many different ways for estimating such diversity measurements have been
described in the literature. Which measurement is best or applicable for your
samples, is not the aim of the following sections.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}
\end{Highlighting}
\end{Shaded}

\textbf{\emph{Alpha diversity}}, also sometimes interchangeably used with the
term \textbf{\emph{species diversity}}, summarizes the distribution of species
abundances in a given sample into a single number that depends on
species richness and evenness. Diversity indices measure the overall
community heterogeneity. A number of ecological diversity measures are
available. The Hill coefficient combines many standard indices into a
single equation that provides observed richness, inverse Simpson, and
Shannon diversity, and generalized diversity as special cases. In
general, diversity increases together with increasing richness and
evenness. Sometimes richness, phylogenetic diversity, evenness, dominance,
and rarity are considered to be variants of alpha diversity.

\textbf{Richness} refers to the total number of species in a community
(sample). The simplest richness index is the number of observed
species (observed richness). Assuming limited sampling from the
community, however, this may underestimate the true species
richness. Several estimators are available, including for instance ACE
\citep{Chao1992} and Chao1 \citep{Chao1984}. Richness estimates are unaffected
by species abundances.

\textbf{Phylogenetic diversity} was first proposed by \citep{Faith1992}. Unlike the
diversity measures mentioned above, Phylogenetic diversity (PD)
measure incorporates information from phylogenetic relationships
stored in \texttt{phylo} tree between species in a community (sample). The
Faith's PD is calculated as the sum of branch length of all species in
a community (sample).

\textbf{Evenness} focuses on species abundances, and can thus complement
the number of species. A typical evenness index is the Pielou's
evenness, which is Shannon diversity normalized by the observed
richness.

\textbf{Dominance} indices are in general negatively correlated with
diversity, and sometimes used in ecological literature. High
dominance is obtained when one or few species have a high share of
the total species abundance in the community.

\textbf{Rarity} indices characterize the concentration of taxa at low abundance.
Prevalence and detection thresholds determine rare taxa whose total concentration
is represented as a rarity index.

\hypertarget{estimation}{%
\section{Estimation}\label{estimation}}

Alpha diversity can be estimated with wrapper functions that interact
with other packages implementing the calculation, such as \emph{\texttt{vegan}}
\citep{R_vegan}.

\hypertarget{richness}{%
\subsection{Richness}\label{richness}}

Richness gives the number of features present within a community and can be calculated with \texttt{estimateRichness}. Each of the estimate diversity/richness/evenness/dominance functions adds the calculated measure(s) to the \texttt{colData} of the \texttt{SummarizedExperiment} under the given column \texttt{name}. Here, we calculate \texttt{observed} features as a measure of richness.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateRichness}\NormalTok{(tse, }
                             \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }
                             \AttributeTok{index =} \StringTok{"observed"}\NormalTok{, }
                             \AttributeTok{name=}\StringTok{"observed"}\NormalTok{)}

\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{observed)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
##    6964    7679    5729    2667    2574    3214
\end{verbatim}

This allows access to the values to be analyzed directly from the \texttt{colData}, for example
by plotting them using \texttt{plotColData} from the \emph{\texttt{scater}} package \citep{R_scater}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(scater)}
\FunctionTok{plotColData}\NormalTok{(tse, }
            \StringTok{"observed"}\NormalTok{, }
            \StringTok{"SampleType"}\NormalTok{, }
            \AttributeTok{colour\_by =} \StringTok{"Final\_Barcode"}\NormalTok{) }\SpecialCharTok{+}
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle=}\DecValTok{45}\NormalTok{,}\AttributeTok{hjust=}\DecValTok{1}\NormalTok{)) }\SpecialCharTok{+} 
  \FunctionTok{ylab}\NormalTok{(}\FunctionTok{expression}\NormalTok{(Richness[Observed]))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{14_alpha_diversity_files/figure-latex/plot-div-shannon-1.pdf}
\caption{\label{fig:plot-div-shannon}Shannon diversity estimates plotted grouped by sample type with colour-labeled barcode.}
\end{figure}

\hypertarget{estimate-diversity}{%
\subsection{Diversity}\label{estimate-diversity}}

The main function, \texttt{estimateDiversity}, calculates the selected
diversity index based on the selected assay data.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateDiversity}\NormalTok{(tse, }
                              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                              \AttributeTok{index =} \StringTok{"shannon"}\NormalTok{, }
                              \AttributeTok{name =} \StringTok{"shannon"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{shannon)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
##   6.577   6.777   6.498   3.828   3.288   4.289
\end{verbatim}

Alpha diversities can be visualized with boxplot. Here, Shannon index is compared
between different sample type groups. Individual data points are visualized by
plotting them as points with \texttt{geom\_jitter}.

\texttt{geom\_signif} is used to test whether these differences are statistically significant.
It adds p-values to plot.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggsignif)}
\FunctionTok{library}\NormalTok{(ggplot2)}
\FunctionTok{library}\NormalTok{(patchwork)}
\FunctionTok{library}\NormalTok{(ggsignif)}

\CommentTok{\# Subsets the data. Takes only those samples that are from feces, skin, or tongue,}
\CommentTok{\# and creates data frame from the collected data}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)[tse}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%in\%} 
                 \FunctionTok{c}\NormalTok{(}\StringTok{"Feces"}\NormalTok{, }\StringTok{"Skin"}\NormalTok{, }\StringTok{"Tongue"}\NormalTok{), ])}

\CommentTok{\# Changes old levels with new levels}
\NormalTok{df}\SpecialCharTok{$}\NormalTok{SampleType }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(df}\SpecialCharTok{$}\NormalTok{SampleType)}

\CommentTok{\# For significance testing, all different combinations are determined}
\NormalTok{comb }\OtherTok{\textless{}{-}} \FunctionTok{split}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{combn}\NormalTok{(}\FunctionTok{levels}\NormalTok{(df}\SpecialCharTok{$}\NormalTok{SampleType), }\DecValTok{2}\NormalTok{)), }
           \FunctionTok{seq}\NormalTok{(}\FunctionTok{nrow}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{combn}\NormalTok{(}\FunctionTok{levels}\NormalTok{(df}\SpecialCharTok{$}\NormalTok{SampleType), }\DecValTok{2}\NormalTok{)))))}

\FunctionTok{ggplot}\NormalTok{(df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ SampleType, }\AttributeTok{y =}\NormalTok{ shannon)) }\SpecialCharTok{+}
  \CommentTok{\# Outliers are removed, because otherwise each data point would be plotted twice; }
  \CommentTok{\# as an outlier of boxplot and as a point of dotplot.}
  \FunctionTok{geom\_boxplot}\NormalTok{(}\AttributeTok{outlier.shape =} \ConstantTok{NA}\NormalTok{) }\SpecialCharTok{+} 
  \FunctionTok{geom\_jitter}\NormalTok{(}\AttributeTok{width =} \FloatTok{0.2}\NormalTok{) }\SpecialCharTok{+} 
  \FunctionTok{geom\_signif}\NormalTok{(}\AttributeTok{comparisons =}\NormalTok{ comb, }\AttributeTok{map\_signif\_level =} \ConstantTok{FALSE}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{text =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{10}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\includegraphics{14_alpha_diversity_files/figure-latex/visualize-shannon-1.pdf}

\hypertarget{faith-diversity}{%
\subsection{Faith phylogenetic diversity}\label{faith-diversity}}

The Faith index is returned by the function \texttt{estimateFaith}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateFaith}\NormalTok{(tse,}
                          \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{faith)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0 0 0 0 0 0
\end{verbatim}

\textbf{Note}: because \texttt{tse} is a \texttt{TreeSummarizedExperiment} object, its phylogenetic tree is used by default. However, the optional argument \texttt{tree} must be provided if \texttt{tse} does not contain one.

Below a visual comparison between shannon and faith indices is shown with a violin plot.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"shannon"}\NormalTok{, }\StringTok{"faith"}\NormalTok{),}
\NormalTok{                plotColData,}
                \AttributeTok{object =}\NormalTok{ tse, }\AttributeTok{colour\_by =} \StringTok{"SampleType"}\NormalTok{)}
\NormalTok{plots[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{+}\NormalTok{ plots[[}\DecValTok{2}\NormalTok{]] }\SpecialCharTok{+}
  \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{14_alpha_diversity_files/figure-latex/phylo-div-2-1.pdf}

Alternatively, the phylogenetic diversity can be calculated by \texttt{mia::estimateDiversity}. This is a faster re-implementation of\\
the widely used function in \emph{\texttt{picante}} \citep[\citet{Kembel2010}]{R_picante}.

Load \texttt{picante} R package and get the \texttt{phylo} stored in \texttt{rowTree}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateDiversity}\NormalTok{(tse, }
                              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                              \AttributeTok{index =} \StringTok{"faith"}\NormalTok{, }
                              \AttributeTok{name =} \StringTok{"faith"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{evenness}{%
\subsection{Evenness}\label{evenness}}

Evenness can be calculated with \texttt{estimateEvenness}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{estimateEvenness}\NormalTok{(tse, }
                        \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }
                        \AttributeTok{index=}\StringTok{"simpson"}\NormalTok{)}
\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{simpson)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.026871 0.027197 0.047049 0.005179 0.004304 0.005011
\end{verbatim}

\hypertarget{dominance}{%
\subsection{Dominance}\label{dominance}}

Dominance can be calculated with \texttt{estimateDominance}. Here, the \texttt{Relative\ index} is calculated which is the relative abundance of the most dominant species in the sample.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{estimateDominance}\NormalTok{(tse, }
                         \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }
                         \AttributeTok{index=}\StringTok{"relative"}\NormalTok{)}

\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{relative)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr 
## 0.03910 0.03226 0.01690 0.22981 0.21778 0.22329
\end{verbatim}

\hypertarget{rarity}{%
\subsection{Rarity}\label{rarity}}

\texttt{mia} package provides one rarity index called log-modulo skewness. It can be
calculated with \texttt{estimateDiversity}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateDiversity}\NormalTok{(tse, }
                              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                              \AttributeTok{index =} \StringTok{"log\_modulo\_skewness"}\NormalTok{)}

\FunctionTok{head}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{log\_modulo\_skewness)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 2.061 2.061 2.061 2.061 2.061 2.061
\end{verbatim}

\hypertarget{divergence}{%
\subsection{Divergence}\label{divergence}}

Divergence can be evaluated with \texttt{estimateDivergence}. Reference and algorithm for the calculation of divergence can be specified as \texttt{reference} and \texttt{FUN}, respectively.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateDivergence}\NormalTok{(tse,}
                               \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                               \AttributeTok{reference =} \StringTok{"median"}\NormalTok{,}
                               \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist)}
\end{Highlighting}
\end{Shaded}

\hypertarget{visualization}{%
\section{Visualization}\label{visualization}}

A plot comparing all the diversity measures calculated above and stored in \texttt{colData} can then be constructed directly.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"observed"}\NormalTok{, }\StringTok{"shannon"}\NormalTok{, }\StringTok{"simpson"}\NormalTok{, }\StringTok{"relative"}\NormalTok{, }\StringTok{"faith"}\NormalTok{, }\StringTok{"log\_modulo\_skewness"}\NormalTok{),}
\NormalTok{                plotColData,}
                \AttributeTok{object =}\NormalTok{ tse,}
                \AttributeTok{x =} \StringTok{"SampleType"}\NormalTok{,}
                \AttributeTok{colour\_by =} \StringTok{"SampleType"}\NormalTok{)}

\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(plots, }\StringTok{"+"}\NormalTok{, }
                \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_blank}\NormalTok{(),}
                      \AttributeTok{axis.title.x =} \FunctionTok{element\_blank}\NormalTok{(),}
                      \AttributeTok{axis.ticks.x =} \FunctionTok{element\_blank}\NormalTok{()))}

\NormalTok{((plots[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{|}\NormalTok{ plots[[}\DecValTok{2}\NormalTok{]] }\SpecialCharTok{|}\NormalTok{ plots[[}\DecValTok{3}\NormalTok{]]) }\SpecialCharTok{/} 
\NormalTok{(plots[[}\DecValTok{4}\NormalTok{]] }\SpecialCharTok{|}\NormalTok{ plots[[}\DecValTok{5}\NormalTok{]] }\SpecialCharTok{|}\NormalTok{ plots[[}\DecValTok{6}\NormalTok{]])) }\SpecialCharTok{+}
  \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{14_alpha_diversity_files/figure-latex/plot-all-diversities-1.pdf}

\hypertarget{community-similarity}{%
\chapter{Community Similarity}\label{community-similarity}}

Beta diversity quantifies the dissimilarity between communities (multiple
samples), as opposed to alpha diversity which focuses on variation within a
community (one sample). In microbiome research, commonly used metrics of beta
diversity include the Bray-Curtis index (for compositional data), Jaccard index
(for presence/absence data, ignoring abundance information), Aitchison distance
(Euclidean distance for clr transformed abundances, aiming to avoid the
compositionality bias), and the Unifrac distance (that takes into account the
phylogenetic tree information). Notably, only some of these measures are actual
\emph{distances}, as this is a mathematical concept whose definition is not satisfied
by certain ecological measure, such as the Bray-Curtis index. Therefore, the terms
dissimilarity and beta diversity are preferred.

\begin{longtable}[]{@{}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3973}}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.2877}}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3151}}@{}}
\toprule\noalign{}
\begin{minipage}[b]{\linewidth}\centering
Method description
\end{minipage} & \begin{minipage}[b]{\linewidth}\centering
Assay type
\end{minipage} & \begin{minipage}[b]{\linewidth}\centering
Beta diversity metric
\end{minipage} \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Quantitative profiling & Absolute counts & Bray-Curtis \\
Relative profiling & Relative abundances & Bray-Curtis \\
Aitchison distance & Absolute counts & Aitchison \\
Aitchison distance & clr & Euclidean \\
Robust Aitchison distance & rclr & Euclidean \\
Presence/Absence similarity & Relative abundances & Jaccard \\
Presence/Absence similarity & Absolute counts & Jaccard \\
Phylogenetic distance & Rarefied counts & Unifrac \\
\end{longtable}

In practice, beta diversity is usually represented as a \texttt{dist} object, a
triangular matrix where the distance between each pair of samples is encoded by
a specific cell. This distance matrix can then undergo ordination, which is an
important ecological tool to reduce the dimensionality of data for a more
efficient analysis and visualization. Ordination techniques aim to capture as
much essential information from the data as possible and turn it into a lower
dimensional representation. Dimension reduction is bound to lose information but
commonly used ordination techniques can preserve relevant information of sample
similarities in an optimal way, which is defined in different ways by different
methods.

Based on the type of algorithm, ordination methods in microbiome research can
be generally divided in two categories: unsupervised and supervised ordination.
The former includes Principal Coordinate Analysis (PCoA), Principal Component
Analysis (PCA) and Uniform Manifold Approximation and Projection for Dimension
Reduction (UMAP), whereas the latter is mainly represented by distance-based
Redundancy Analysis (dbRDA). We will first discuss unsupervised ordination
methods and then proceed to supervised ones.

\hypertarget{unsupervised-ordination}{%
\section{Unsupervised ordination}\label{unsupervised-ordination}}

Unsupervised ordination methods variation in the data without additional
information on covariates or other supervision of the model. Among the different
approaches, Multi-Dimensional Scaling (MDS) and non-metric MDS (NMDS) can be
regarded as the standard. They are jointly referred to as PCoA. For this
demonstration we will analyse beta diversity in GlobalPatterns, and observe the
variation between stool samples and those with a different origin.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load mia and import sample dataset}
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}

\CommentTok{\# Beta diversity metrics like Bray{-}Curtis are often applied to relabundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse,}
                      \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                      \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Other metrics like Aitchison to clr{-}transformed data}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse,}
                      \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
                      \AttributeTok{method =} \StringTok{"clr"}\NormalTok{,}
                      \AttributeTok{pseudocount =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Add group information Feces yes/no}
\NormalTok{tse}\SpecialCharTok{$}\NormalTok{Group }\OtherTok{\textless{}{-}}\NormalTok{ tse}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{==} \StringTok{"Feces"}
\end{Highlighting}
\end{Shaded}

\hypertarget{comparing-communities-by-beta-diversity-analysis}{%
\subsection{Comparing communities by beta diversity analysis}\label{comparing-communities-by-beta-diversity-analysis}}

A typical comparison of community compositions starts with a visual
representation of the groups by a 2D ordination. Then we estimate relative
abundances and MDS ordination based on Bray-Curtis index between the groups,
and visualize the results.

In the following examples dissimilarity is calculated with the function supplied
to the \texttt{FUN} argument. Several metrics of beta diversity are defined by the \texttt{vegdist}
function of the vegan package, which is often used in this context. However, such
custom functions created by the user also work, as long as they return a \texttt{dist}
object. In either case, this function is then applied to calculate reduced
dimensions via an ordination method, the results of which can be stored in the
\texttt{reducedDim} slot of the TreeSE. This entire process is contained by the \texttt{runMDS}
and \texttt{runNMDS} functions.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package to plot reducedDim}
\FunctionTok{library}\NormalTok{(scater)}

\CommentTok{\# Run PCoA on relabundance assay with Bray{-}Curtis distances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(tse,}
              \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
              \AttributeTok{method =} \StringTok{"bray"}\NormalTok{,}
              \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
              \AttributeTok{name =} \StringTok{"MDS\_bray"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Sample dissimilarity can be visualized on a lower-dimensional display (typically
2D) using the \texttt{plotReducedDim} function from the scater package. This also
provides tools to incorporate additional information encoded by color, shape,
size and other aesthetics. Can you find any difference between the groups?

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create ggplot object}
\NormalTok{p }\OtherTok{\textless{}{-}} \FunctionTok{plotReducedDim}\NormalTok{(tse, }\StringTok{"MDS\_bray"}\NormalTok{,}
                    \AttributeTok{colour\_by =} \StringTok{"Group"}\NormalTok{)}

\CommentTok{\# Calculate explained variance}
\NormalTok{e }\OtherTok{\textless{}{-}} \FunctionTok{attr}\NormalTok{(}\FunctionTok{reducedDim}\NormalTok{(tse, }\StringTok{"MDS\_bray"}\NormalTok{), }\StringTok{"eig"}\NormalTok{)}
\NormalTok{rel\_eig }\OtherTok{\textless{}{-}}\NormalTok{ e }\SpecialCharTok{/} \FunctionTok{sum}\NormalTok{(e[e }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{])}

\CommentTok{\# Add explained variance for each axis}
\NormalTok{p }\OtherTok{\textless{}{-}}\NormalTok{ p }\SpecialCharTok{+} \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PCoA 1 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(}\DecValTok{100} \SpecialCharTok{*}\NormalTok{ rel\_eig[[}\DecValTok{1}\NormalTok{]], }\DecValTok{1}\NormalTok{), }\StringTok{"\%"}\NormalTok{, }\StringTok{")"}\NormalTok{, }\AttributeTok{sep =} \StringTok{""}\NormalTok{),}
              \AttributeTok{y =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PCoA 2 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(}\DecValTok{100} \SpecialCharTok{*}\NormalTok{ rel\_eig[[}\DecValTok{2}\NormalTok{]], }\DecValTok{1}\NormalTok{), }\StringTok{"\%"}\NormalTok{, }\StringTok{")"}\NormalTok{, }\AttributeTok{sep =} \StringTok{""}\NormalTok{))}

\NormalTok{p}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{20_beta_diversity_files/figure-latex/plot-mds-bray-curtis-1.png}
\caption{\label{fig:plot-mds-bray-curtis}MDS plot based on the Bray-Curtis distances on the GlobalPattern dataset.}
\end{figure}

A few combinations of beta diversity metrics and assay types are typically
used. For instance, Bray-Curtis dissimilarity and Euclidean distance are often
applied to the relative abundance and the clr assays, respectively. Besides
\textbf{beta diversity metric} and \textbf{assay type}, the \textbf{PCoA algorithm} is also a
variable that should be considered. Below, we show how the choice of these three
factors can affect the resulting lower-dimensional data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Run NMDS on relabundance assay with Bray{-}Curtis distances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runNMDS}\NormalTok{(tse,}
               \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
               \AttributeTok{method =} \StringTok{"bray"}\NormalTok{,}
               \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
               \AttributeTok{name =} \StringTok{"NMDS\_bray"}\NormalTok{)}

\CommentTok{\# Run MDS on clr assay with Aitchison distances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(tse,}
              \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
              \AttributeTok{method =} \StringTok{"euclidean"}\NormalTok{,}
              \AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{,}
              \AttributeTok{name =} \StringTok{"MDS\_aitchison"}\NormalTok{)}

\CommentTok{\# Run NMDS on clr assay with Euclidean distances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runNMDS}\NormalTok{(tse,}
               \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
               \AttributeTok{method =} \StringTok{"euclidean"}\NormalTok{,}
               \AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{,}
               \AttributeTok{name =} \StringTok{"NMDS\_aitchison"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Multiple ordination plots are combined into a multi-panel plot with the
patchwork package, so that different methods can be compared to find similarities
between them or select the most suitable one to visualize beta diversity in the
light of the research question.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package for multi{-}panel plotting}
\FunctionTok{library}\NormalTok{(patchwork)}

\CommentTok{\# Generate plots for all 4 reducedDims}
\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"MDS\_bray"}\NormalTok{, }\StringTok{"MDS\_aitchison"}\NormalTok{,}
                  \StringTok{"NMDS\_bray"}\NormalTok{, }\StringTok{"NMDS\_aitchison"}\NormalTok{),}
\NormalTok{                plotReducedDim,}
                \AttributeTok{object =}\NormalTok{ tse,}
                \AttributeTok{colour\_by =} \StringTok{"Group"}\NormalTok{)}

\CommentTok{\# Generate multi{-}panel plot}
\FunctionTok{wrap\_plots}\NormalTok{(plots) }\SpecialCharTok{+}
  \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{20_beta_diversity_files/figure-latex/unnamed-chunk-2-1.png}
\caption{\label{fig:unnamed-chunk-2}Comparison of MDS and NMDS plots based on the Bray-Curtis or Aitchison distances on the GlobalPattern dataset.}
\end{figure}

The \emph{Unifrac} method is a special case, as it requires data on the
relationship of features in form on a \texttt{phylo} tree. \texttt{calculateUnifrac}
performs the calculation to return a \texttt{dist} object, which can again be
used within \texttt{runMDS}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(tse,}
              \AttributeTok{FUN =}\NormalTok{ mia}\SpecialCharTok{::}\NormalTok{calculateUnifrac,}
              \AttributeTok{name =} \StringTok{"Unifrac"}\NormalTok{,}
              \AttributeTok{tree =} \FunctionTok{rowTree}\NormalTok{(tse),}
              \AttributeTok{ntop =} \FunctionTok{nrow}\NormalTok{(tse),}
              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{)}

\FunctionTok{plotReducedDim}\NormalTok{(tse, }\StringTok{"Unifrac"}\NormalTok{,}
               \AttributeTok{colour\_by =} \StringTok{"Group"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{20_beta_diversity_files/figure-latex/plot-unifrac-1.png}
\caption{\label{fig:plot-unifrac}Unifrac distances scaled by MDS of the GlobalPattern dataset.}
\end{figure}

\hypertarget{other-ord-methods}{%
\subsection{Other ordination methods}\label{other-ord-methods}}

Other dimension reduction methods, such as PCA and UMAP, are inherited from the
scater package.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runPCA}\NormalTok{(tse,}
              \AttributeTok{name =} \StringTok{"PCA"}\NormalTok{,}
              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
              \AttributeTok{ncomponents =} \DecValTok{10}\NormalTok{)}

\FunctionTok{plotReducedDim}\NormalTok{(tse, }\StringTok{"PCA"}\NormalTok{,}
               \AttributeTok{colour\_by =} \StringTok{"Group"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{20_beta_diversity_files/figure-latex/plot-pca-1.png}
\caption{\label{fig:plot-pca}PCA plot on the GlobalPatterns data set containing sample from different sources.}
\end{figure}

As mentioned before, applicability of the different methods depends on your
sample set and research question.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runUMAP}\NormalTok{(tse,}
               \AttributeTok{name =} \StringTok{"UMAP"}\NormalTok{,}
               \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
               \AttributeTok{ncomponents =} \DecValTok{3}\NormalTok{)}

\FunctionTok{plotReducedDim}\NormalTok{(tse, }\StringTok{"UMAP"}\NormalTok{,}
               \AttributeTok{colour\_by =} \StringTok{"Group"}\NormalTok{,}
               \AttributeTok{ncomponents =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\SpecialCharTok{:}\DecValTok{3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{20_beta_diversity_files/figure-latex/plot-umap-1.png}
\caption{\label{fig:plot-umap}UMAP plot on the GlobalPatterns data set containing sample from different sources.}
\end{figure}

\hypertarget{explained-variance}{%
\subsection{Explained variance}\label{explained-variance}}

The percentage of explained variance is typically shown for PCA
ordination plots. This quantifies the proportion of overall variance
in the data that is captured by the PCA axes, or how well the
ordination axes reflect the original distances.

Sometimes a similar measure is shown for MDS/PCoA. The interpretation
is generally different, however, and hence we do not recommend using
it. PCA is a special case of PCoA with Euclidean distances. With
non-Euclidean dissimilarities PCoA uses a trick where the pointwise
dissimilarities are first cast into similarities in a Euclidean space
(with some information loss i.e.~stress) and then projected to the
maximal variance axes. In this case, the maximal variance axes do not
directly reflect the correspondence of the projected distances and
original distances, as they do for PCA.

In typical use cases, we would like to know how well the ordination
reflects the original similarity structures; then the quantity of
interest is the so-called ``stress'' function, which measures the
difference in pairwise similarities between the data points in the
original (high-dimensional) vs.~projected (low-dimensional) space.

Hence, we propose that for PCoA and other ordination methods, users
would report relative stress, which varies in the unit interval and is better
if smaller. This can be calculated as shown below.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load vegan package}
\FunctionTok{library}\NormalTok{(vegan)}

\CommentTok{\# Quantify dissimilarities in the original feature space}
\NormalTok{x }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse, }\StringTok{"relabundance"}\NormalTok{) }\CommentTok{\# Pick relabunance assay separately}
\NormalTok{d0 }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{vegdist}\NormalTok{(}\FunctionTok{t}\NormalTok{(x), }\StringTok{"bray"}\NormalTok{))}

\CommentTok{\# PCoA Ordination}
\NormalTok{pcoa }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{cmdscale}\NormalTok{(d0, }\AttributeTok{k =} \DecValTok{2}\NormalTok{))}
\FunctionTok{names}\NormalTok{(pcoa) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"PCoA1"}\NormalTok{, }\StringTok{"PCoA2"}\NormalTok{)}

\CommentTok{\# Quantify dissimilarities in the ordination space}
\NormalTok{dp }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{dist}\NormalTok{(pcoa))}

\CommentTok{\# Calculate stress i.e. relative difference in the original and}
\CommentTok{\# projected dissimilarities}
\NormalTok{stress }\OtherTok{\textless{}{-}} \FunctionTok{sum}\NormalTok{((dp }\SpecialCharTok{{-}}\NormalTok{ d0)}\SpecialCharTok{\^{}}\DecValTok{2}\NormalTok{) }\SpecialCharTok{/} \FunctionTok{sum}\NormalTok{(d0}\SpecialCharTok{\^{}}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

A Shepard plot visualizes the original versus the ordinated dissimilarity
between the observations.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ord }\OtherTok{\textless{}{-}} \FunctionTok{order}\NormalTok{(}\FunctionTok{as.vector}\NormalTok{(d0))}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{d0 =} \FunctionTok{as.vector}\NormalTok{(d0)[ord],}
                 \AttributeTok{dmds =} \FunctionTok{as.vector}\NormalTok{(dp)[ord])}

\FunctionTok{ggplot}\NormalTok{(df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ d0, }\AttributeTok{y =}\NormalTok{ dmds)) }\SpecialCharTok{+}
  \FunctionTok{geom\_smooth}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+}    
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \StringTok{"Shepard plot"}\NormalTok{,}
       \AttributeTok{x =} \StringTok{"Original distance"}\NormalTok{,}
       \AttributeTok{y =} \StringTok{"MDS distance"}\NormalTok{,   }
       \AttributeTok{subtitle =} \FunctionTok{paste}\NormalTok{(}\StringTok{"Stress:"}\NormalTok{, }\FunctionTok{round}\NormalTok{(stress, }\DecValTok{2}\NormalTok{))) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\includegraphics{20_beta_diversity_files/figure-latex/shepard-1.png}

\hypertarget{supervised-ordination}{%
\section{Supervised ordination}\label{supervised-ordination}}

dbRDA is a supervised counterpart of PCoA, that is, it takes into account the
covariates specified by the user to maximize the variance with respect to the
them. The result shows how much each covariate affects beta diversity. The table
below illustrates the relation between supervised and unsupervised ordination
methods.

\begin{longtable}[]{@{}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3421}}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3158}}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3421}}@{}}
\toprule\noalign{}
\begin{minipage}[b]{\linewidth}\centering
\end{minipage} & \begin{minipage}[b]{\linewidth}\centering
supervised ordination
\end{minipage} & \begin{minipage}[b]{\linewidth}\centering
unsupervised ordination
\end{minipage} \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Euclidean distance & RDA & PCA \\
non-Euclidean distance & dbRDA & PCoA/MDS, NMDS and UMAP \\
\end{longtable}

We demonstrate the usage of dbRDA with the enterotype dataset, where samples
correspond to patients. The colData contains the clinical status of each patient
and a few covariates such as their gender and age.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load data}
\FunctionTok{data}\NormalTok{(}\StringTok{"enterotype"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse2 }\OtherTok{\textless{}{-}}\NormalTok{ enterotype}

\CommentTok{\# Apply relative transform}
\NormalTok{tse2 }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse2,}
                       \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

dbRDA can be perfomed with the \texttt{runRDA} function. In addition to the arguments
previously defined for \protect\hyperlink{unsupervised-ordination}{unsupervised ordination}, this
function takes a formula to control for variables and an action to treat missing
values. Along with clinical status, which is the main outcome, we control for
gender and age, and exclude observations where one of these variables is missing.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Perform RDA}
\NormalTok{tse2 }\OtherTok{\textless{}{-}} \FunctionTok{runRDA}\NormalTok{(tse2,}
               \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
               \AttributeTok{formula =}\NormalTok{ assay }\SpecialCharTok{\textasciitilde{}}\NormalTok{ ClinicalStatus }\SpecialCharTok{+}\NormalTok{ Gender }\SpecialCharTok{+}\NormalTok{ Age,}
               \AttributeTok{distance =} \StringTok{"bray"}\NormalTok{,}
               \AttributeTok{na.action =}\NormalTok{ na.exclude)}

\CommentTok{\# Store results of PERMANOVA test}
\NormalTok{rda\_info }\OtherTok{\textless{}{-}} \FunctionTok{attr}\NormalTok{(}\FunctionTok{reducedDim}\NormalTok{(tse2, }\StringTok{"RDA"}\NormalTok{), }\StringTok{"significance"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The importance of each variable on the similarity between samples can be
assessed from the results of PERMANOVA, automatically provided by the \texttt{runRDA}
function. We see that both clinical status and age explain more than 10\% of the
variance, but only age shows statistical significance.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{rda\_info}\SpecialCharTok{$}\NormalTok{permanova }\SpecialCharTok{|\textgreater{}}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r|r|r|r|r}
\hline
  & Df & SumOfSqs & F & Pr(>F) & Total variance & Explained variance\\
\hline
Model & 6 & 1.1157 & 1.940 & 0.033 & 3.991 & 0.2795\\
\hline
ClinicalStatus & 4 & 0.5837 & 1.522 & 0.147 & 3.991 & 0.1463\\
\hline
Gender & 1 & 0.1679 & 1.751 & 0.107 & 3.991 & 0.0421\\
\hline
Age & 1 & 0.5245 & 5.471 & 0.001 & 3.991 & 0.1314\\
\hline
Residual & 30 & 2.8757 & NA & NA & 3.991 & 0.7205\\
\hline
\end{tabular}

To ensure that the homogeneity assumption holds, we retrieve the corresponding
information from the results of RDA. In this case, none of the p-values is lower
than the significance threshold, and thus homogeneity is observed.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{rda\_info}\SpecialCharTok{$}\NormalTok{homogeneity }\SpecialCharTok{|\textgreater{}}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r|r|r|r|r|r|r}
\hline
  & Df & Sum Sq & Mean Sq & F & N.Perm & Pr(>F) & Total variance & Explained variance\\
\hline
ClinicalStatus & 4 & 0.2511 & 0.0628 & 2.7440 & 999 & 0.119 & 1.0288 & 0.2440\\
\hline
Gender & 1 & 0.0103 & 0.0103 & 0.4158 & 999 & 0.537 & 0.9283 & 0.0111\\
\hline
Age & 29 & 0.3272 & 0.0113 & 17.0255 & 999 & 0.437 & 0.3319 & 0.9860\\
\hline
\end{tabular}

Next, we proceed to visualize the weight and significance of each variable on
the similarity between samples with an RDA plot, which can be generated with
the \texttt{plotRDA} function from the miaViz package.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load packages for plotting function}
\FunctionTok{library}\NormalTok{(miaViz)}

\CommentTok{\# Generate RDA plot coloured by clinical status}
\FunctionTok{plotRDA}\NormalTok{(tse2, }\StringTok{"RDA"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"ClinicalStatus"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{20_beta_diversity_files/figure-latex/plot-rda-1.png}

From the plot above, we can see that only age significantly describes
differences between the microbial profiles of different samples. Such visual
approach complements the previous results of PERMANOVA.

\hypertarget{case-studies}{%
\section{Case studies}\label{case-studies}}

\hypertarget{pcoa-genus}{%
\subsubsection{Visualizing the most dominant genus on PCoA}\label{pcoa-genus}}

In this section, we visualize the most dominant genus on PCoA. A similar
visualization was proposed by \citeyearpar{Salosensaari2021}. First, we agglomerate the
data at the Genus level and get the dominant taxa per sample.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate to genus level}
\NormalTok{tse\_genus }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse,}
                                 \AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{)}

\CommentTok{\# Convert to relative abundances}
\NormalTok{tse\_genus }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse,}
                            \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{,}
                            \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{)}

\CommentTok{\# Add info on dominant genus per sample}
\NormalTok{tse\_genus }\OtherTok{\textless{}{-}} \FunctionTok{addPerSampleDominantFeatures}\NormalTok{(tse\_genus,}
                                          \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
                                          \AttributeTok{name =} \StringTok{"dominant\_taxa"}\NormalTok{)}
\CommentTok{\# Overview}
\FunctionTok{countDominantFeatures}\NormalTok{(tse\_genus, }\AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{, }\AttributeTok{digits =} \DecValTok{3}\NormalTok{, }\AttributeTok{name =} \StringTok{"dominant\_taxa"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 17 x 3
##    dominant_taxa                  n rel_freq
##    <chr>                      <int>    <dbl>
##  1 Genus:Bacteroides              5    0.192
##  2 Order:Stramenopiles            4    0.154
##  3 Family:Desulfobulbaceae        2    0.077
##  4 Genus:Streptococcus            2    0.077
##  5 Class:Chloracidobacteria       1    0.038
##  6 Family:ACK-M1                  1    0.038
##  7 Family:Flavobacteriaceae       1    0.038
##  8 Family:Moraxellaceae           1    0.038
##  9 Family:Ruminococcaceae         1    0.038
## 10 Genus:CandidatusSolibacter     1    0.038
## 11 Genus:Dolichospermum           1    0.038
## 12 Genus:Faecalibacterium         1    0.038
## 13 Genus:MC18                     1    0.038
## 14 Genus:Neisseria                1    0.038
## 15 Genus:Prochlorococcus          1    0.038
## 16 Genus:Veillonella              1    0.038
## 17 Order:Chromatiales             1    0.038
\end{verbatim}

Next, we perform PCoA with Bray-Curtis dissimilarity.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse\_genus }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(tse\_genus,}
                    \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
                    \AttributeTok{name =} \StringTok{"PCoA\_BC"}\NormalTok{,}
                    \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Finally, we get the top taxa and and visualize their abundances on PCoA. Note
that A 3D interactive version of the plot below can be found in \ref{extras}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Getting the top taxa}
\NormalTok{top\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse\_genus,}
                           \AttributeTok{top =} \DecValTok{6}\NormalTok{,}
                           \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Naming all the rest of non top{-}taxa as "Other"}
\NormalTok{most\_abundant }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse\_genus)}\SpecialCharTok{$}\NormalTok{dominant\_taxa,}
                        \ControlFlowTok{function}\NormalTok{(x) \{}\ControlFlowTok{if}\NormalTok{ (x }\SpecialCharTok{\%in\%}\NormalTok{ top\_taxa) \{x\} }\ControlFlowTok{else}\NormalTok{ \{}\StringTok{"Other"}\NormalTok{\}\})}

\CommentTok{\# Storing the previous results as a new column within colData}
\FunctionTok{colData}\NormalTok{(tse\_genus)}\SpecialCharTok{$}\NormalTok{most\_abundant }\OtherTok{\textless{}{-}} \FunctionTok{as.character}\NormalTok{(most\_abundant)}

\CommentTok{\# Calculating percentage of the most abundant}
\NormalTok{most\_abundant\_freq }\OtherTok{\textless{}{-}} \FunctionTok{table}\NormalTok{(}\FunctionTok{as.character}\NormalTok{(most\_abundant))}
\NormalTok{most\_abundant\_percent }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(most\_abundant\_freq }\SpecialCharTok{/} \FunctionTok{sum}\NormalTok{(most\_abundant\_freq) }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{1}\NormalTok{)}

\CommentTok{\# Retrieving the explained variance}
\NormalTok{e }\OtherTok{\textless{}{-}} \FunctionTok{attr}\NormalTok{(}\FunctionTok{reducedDim}\NormalTok{(tse\_genus, }\StringTok{"PCoA\_BC"}\NormalTok{), }\StringTok{"eig"}\NormalTok{)}
\NormalTok{var\_explained }\OtherTok{\textless{}{-}}\NormalTok{ e }\SpecialCharTok{/} \FunctionTok{sum}\NormalTok{(e[e }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{]) }\SpecialCharTok{*} \DecValTok{100}

\CommentTok{\# Define colors for visualization}
\NormalTok{my\_colors }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"black"}\NormalTok{, }\StringTok{"blue"}\NormalTok{, }\StringTok{"lightblue"}\NormalTok{, }\StringTok{"darkgray"}\NormalTok{, }\StringTok{"magenta"}\NormalTok{, }\StringTok{"darkgreen"}\NormalTok{, }\StringTok{"red"}\NormalTok{)}

\CommentTok{\# Visualization}
\NormalTok{plot }\OtherTok{\textless{}{-}}\FunctionTok{plotReducedDim}\NormalTok{(tse\_genus, }\StringTok{"PCoA\_BC"}\NormalTok{,}
                      \AttributeTok{colour\_by =} \StringTok{"most\_abundant"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_colour\_manual}\NormalTok{(}\AttributeTok{values =}\NormalTok{ my\_colors,}
                      \AttributeTok{labels =} \FunctionTok{paste0}\NormalTok{(}\FunctionTok{names}\NormalTok{(most\_abundant\_percent), }\StringTok{"("}\NormalTok{, most\_abundant\_percent, }\StringTok{"\%)"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PC 1 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(var\_explained[}\DecValTok{1}\NormalTok{], }\DecValTok{1}\NormalTok{), }\StringTok{"\%)"}\NormalTok{),}
       \AttributeTok{y =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PC 2 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(var\_explained[}\DecValTok{2}\NormalTok{], }\DecValTok{1}\NormalTok{), }\StringTok{"\%)"}\NormalTok{),}
       \AttributeTok{color =} \StringTok{""}\NormalTok{)}

\NormalTok{plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{20_beta_diversity_files/figure-latex/unnamed-chunk-5-1.png}

Similarly, we visualize and compare the sub-population.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Calculating the frequencies and percentages for both categories}
\NormalTok{freq\_TRUE }\OtherTok{\textless{}{-}} \FunctionTok{table}\NormalTok{(}\FunctionTok{as.character}\NormalTok{(most\_abundant[}\FunctionTok{colData}\NormalTok{(tse\_genus)}\SpecialCharTok{$}\NormalTok{Group }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{]))}
\NormalTok{freq\_FALSE }\OtherTok{\textless{}{-}} \FunctionTok{table}\NormalTok{(}\FunctionTok{as.character}\NormalTok{(most\_abundant[}\FunctionTok{colData}\NormalTok{(tse\_genus)}\SpecialCharTok{$}\NormalTok{Group }\SpecialCharTok{==} \ConstantTok{FALSE}\NormalTok{]))}
\NormalTok{percent\_TRUE }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(freq\_TRUE }\SpecialCharTok{/} \FunctionTok{sum}\NormalTok{(freq\_TRUE) }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{1}\NormalTok{)}
\NormalTok{percent\_FALSE }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(freq\_FALSE }\SpecialCharTok{/} \FunctionTok{sum}\NormalTok{(freq\_FALSE) }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{1}\NormalTok{)}

\CommentTok{\# Visualization}
\FunctionTok{plotReducedDim}\NormalTok{(tse\_genus[ , }\FunctionTok{colData}\NormalTok{(tse\_genus)}\SpecialCharTok{$}\NormalTok{Group }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{], }\StringTok{"PCoA\_BC"}\NormalTok{,}
               \AttributeTok{colour\_by =} \StringTok{"most\_abundant"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_colour\_manual}\NormalTok{(}\AttributeTok{values =}\NormalTok{ my\_colors,}
                      \AttributeTok{labels =} \FunctionTok{paste0}\NormalTok{(}\FunctionTok{names}\NormalTok{(percent\_TRUE), }\StringTok{"("}\NormalTok{, percent\_TRUE, }\StringTok{"\%)"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PC 1 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(var\_explained[}\DecValTok{1}\NormalTok{], }\DecValTok{1}\NormalTok{), }\StringTok{"\%)"}\NormalTok{),}
       \AttributeTok{y =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PC 2 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(var\_explained[}\DecValTok{2}\NormalTok{], }\DecValTok{1}\NormalTok{), }\StringTok{"\%)"}\NormalTok{),}
       \AttributeTok{title =} \StringTok{"Group = TRUE"}\NormalTok{, }\AttributeTok{color =} \StringTok{""}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{20_beta_diversity_files/figure-latex/unnamed-chunk-6-1.png}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plotReducedDim}\NormalTok{(tse\_genus[ , }\FunctionTok{colData}\NormalTok{(tse\_genus)}\SpecialCharTok{$}\NormalTok{Group }\SpecialCharTok{==} \ConstantTok{FALSE}\NormalTok{], }\StringTok{"PCoA\_BC"}\NormalTok{,}
               \AttributeTok{colour\_by =} \StringTok{"most\_abundant"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_colour\_manual}\NormalTok{(}\AttributeTok{values =}\NormalTok{ my\_colors,}
                      \AttributeTok{labels =} \FunctionTok{paste0}\NormalTok{(}\FunctionTok{names}\NormalTok{(percent\_FALSE), }\StringTok{"("}\NormalTok{, percent\_FALSE, }\StringTok{"\%)"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PC 1 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(var\_explained[}\DecValTok{1}\NormalTok{], }\DecValTok{1}\NormalTok{), }\StringTok{"\%)"}\NormalTok{),}
       \AttributeTok{y =} \FunctionTok{paste}\NormalTok{(}\StringTok{"PC 2 ("}\NormalTok{, }\FunctionTok{round}\NormalTok{(var\_explained[}\DecValTok{2}\NormalTok{], }\DecValTok{1}\NormalTok{), }\StringTok{"\%)"}\NormalTok{),}
       \AttributeTok{title =} \StringTok{"Group = FALSE"}\NormalTok{, }\AttributeTok{color =} \StringTok{""}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{20_beta_diversity_files/figure-latex/unnamed-chunk-6-2.png}

\hypertarget{dbrda-workflow}{%
\subsection{Testing differences in community composition between sample groups}\label{dbrda-workflow}}

Permutational Analysis of Variance (PERMANOVA; \citeyearpar{Anderson2001}) is
a widely used non-parametric multivariate method that aims to
estimate the actual statistical significance of differences in the
observed community composition between two groups of samples.

PERMANOVA tests the hypothesis that the centroids and dispersion
of the community are equivalent between the compared groups. A p-value
smaller than the significance threshold indicates that the groups have a
different community composition. This method is implemented with the
\href{https://www.rdocumentation.org/packages/vegan/versions/2.4-2/topics/adonis}{\texttt{adonis2}}
function from the vegan package.

By default, the argument \texttt{by} is set to \texttt{"terms"}, in which the order of variables
in the formula matters. In this case, each variable is analyzed sequentially, and
the result is different when more than 1 variable is introduced and their order differs.
Therefore, it is recommended to set \texttt{by\ =\ "margin"}, which specifies that the
marginal effect of each variable is analyzed individually. You can view a
comparison between the two designs in chapter \ref{compare-permanova}.

We can perform PERMANOVA either with \texttt{adonis2} function or by first performing
dbRDA and then applying permutational test its results. An advantage of the latter
approach is that by doing so we can get coefficients: how much each taxa affects
the variation between communities.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate data to Species level}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse,}
                           \AttributeTok{rank =} \StringTok{"Species"}\NormalTok{)}

\CommentTok{\# Set seed for reproducibility}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{1576}\NormalTok{)}
\CommentTok{\# We choose 99 random permutations. Consider applying more (999 or 9999) in your}
\CommentTok{\# analysis. }
\NormalTok{permanova }\OtherTok{\textless{}{-}} \FunctionTok{adonis2}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"relabundance"}\NormalTok{)) }\SpecialCharTok{\textasciitilde{}}\NormalTok{ Group,}
                     \AttributeTok{by =} \StringTok{"margin"}\NormalTok{, }\CommentTok{\# each term (here only \textquotesingle{}Group\textquotesingle{}) analyzed individually}
                     \AttributeTok{data =} \FunctionTok{colData}\NormalTok{(tse),}
                     \AttributeTok{method =} \StringTok{"euclidean"}\NormalTok{,}
                     \AttributeTok{permutations =} \DecValTok{99}\NormalTok{)}

\CommentTok{\# Set seed for reproducibility}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{1576}\NormalTok{)}
\CommentTok{\# Perform dbRDA}
\NormalTok{dbrda }\OtherTok{\textless{}{-}} \FunctionTok{dbrda}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse,}\StringTok{"relabundance"}\NormalTok{)) }\SpecialCharTok{\textasciitilde{}}\NormalTok{ Group, }
               \AttributeTok{data =} \FunctionTok{colData}\NormalTok{(tse))}
\CommentTok{\# Perform permutational analysis}
\NormalTok{permanova2 }\OtherTok{\textless{}{-}} \FunctionTok{anova.cca}\NormalTok{(dbrda,}
                        \AttributeTok{by =} \StringTok{"margin"}\NormalTok{, }\CommentTok{\# each term (here only \textquotesingle{}Group\textquotesingle{}) analyzed individually}
                        \AttributeTok{method =} \StringTok{"euclidean"}\NormalTok{,}
                        \AttributeTok{permutations =} \DecValTok{99}\NormalTok{)}

\CommentTok{\# Get p{-}values}
\NormalTok{p\_values }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(permanova[}\StringTok{"Group"}\NormalTok{, }\StringTok{"Pr(\textgreater{}F)"}\NormalTok{], permanova2[}\StringTok{"Group"}\NormalTok{, }\StringTok{"Pr(\textgreater{}F)"}\NormalTok{])}
\NormalTok{p\_values }\OtherTok{\textless{}{-}}\FunctionTok{as.data.frame}\NormalTok{(p\_values)}
\FunctionTok{rownames}\NormalTok{(p\_values) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"adonis2"}\NormalTok{, }\StringTok{"dbRDA+anova.cca"}\NormalTok{)}
\NormalTok{p\_values}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                 p_values
## adonis2             0.02
## dbRDA+anova.cca     0.02
\end{verbatim}

As we can see, the community composition is significantly different
between the groups (p \textless{} 0.05), and these two methods give equal p-values.

Let us visualize the model coefficients for species that exhibit the
largest differences between the groups. This gives some insights into
how the groups tend to differ from each other in terms of community
composition.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Add taxa info}
\FunctionTok{sppscores}\NormalTok{(dbrda) }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"relabundance"}\NormalTok{))}
\CommentTok{\# Get coefficients}
\NormalTok{coef }\OtherTok{\textless{}{-}}\NormalTok{ dbrda}\SpecialCharTok{$}\NormalTok{CCA}\SpecialCharTok{$}\NormalTok{v}
\CommentTok{\# Get the taxa with biggest weights}
\NormalTok{top.coef }\OtherTok{\textless{}{-}} \FunctionTok{head}\NormalTok{(coef[}\FunctionTok{rev}\NormalTok{(}\FunctionTok{order}\NormalTok{(}\FunctionTok{abs}\NormalTok{(coef))), , }\AttributeTok{drop =} \ConstantTok{FALSE}\NormalTok{], }\DecValTok{20}\NormalTok{)}
\CommentTok{\# Sort weights in increasing order}
\NormalTok{top.coef }\OtherTok{\textless{}{-}}\NormalTok{ top.coef[}\FunctionTok{order}\NormalTok{(top.coef), ]}
\CommentTok{\# Get top names}
\NormalTok{top\_names }\OtherTok{\textless{}{-}} \FunctionTok{names}\NormalTok{(top.coef)[}\FunctionTok{order}\NormalTok{(}\FunctionTok{abs}\NormalTok{(top.coef), }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{)]}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{x =}\NormalTok{ top.coef,}
                 \AttributeTok{y =} \FunctionTok{factor}\NormalTok{(}\FunctionTok{names}\NormalTok{(top.coef), }\FunctionTok{unique}\NormalTok{(}\FunctionTok{names}\NormalTok{(top.coef))))}

\FunctionTok{ggplot}\NormalTok{(df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ x, }\AttributeTok{y =}\NormalTok{ y)) }\SpecialCharTok{+}
  \FunctionTok{geom\_bar}\NormalTok{(}\AttributeTok{stat =} \StringTok{"identity"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{""}\NormalTok{, }\AttributeTok{y=} \StringTok{""}\NormalTok{, }\AttributeTok{title =} \StringTok{"Top Taxa"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\includegraphics{20_beta_diversity_files/figure-latex/plot-top-coef-anova-1.png}

In the example above, the largest differences between the two groups
can be attributed to \emph{Genus:Bacteroides} (elevated in the first
group) and \emph{Family:Ruminococcaceae} (elevated in the second
group), and many other co-varying species.

\hypertarget{checking-the-homogeneity-condition}{%
\subsection{Checking the homogeneity condition}\label{checking-the-homogeneity-condition}}

It is important to note that the application of PERMANOVA assumes
homogeneous group dispersions (variances). This can be tested with the
PERMDISP2 method \citep{Anderson2006} by using the same assay and distance
method than in PERMANOVA.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{anova}\NormalTok{(}\FunctionTok{betadisper}\NormalTok{(}\FunctionTok{vegdist}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{))), }\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Group))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Analysis of Variance Table
## 
## Response: Distances
##           Df Sum Sq Mean Sq F value  Pr(>F)    
## Groups     1 0.2385  0.2385     103 3.6e-10 ***
## Residuals 24 0.0554  0.0023                    
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
\end{verbatim}

If the groups have similar dispersion, PERMANOVA can be seen as an
appropriate choice for comparing community compositions.

\hypertarget{summary}{%
\section{Summary}\label{summary}}

As a final note, we provide a comprehensive list of functions for the evaluation
of dissimilarity indices available in the mia and scater packages. The
\texttt{calculate} methods return a reducedDim object as an output, whereas the \texttt{run}
methods store the reducedDim object into the specified TreeSE.

\begin{itemize}
\tightlist
\item
  Canonical Correspondence Analysis (CCA): \texttt{calculateCCA} and \texttt{runCCA}
\item
  dbRDA: \texttt{calculateRDA} and \texttt{runRDA}
\item
  Double Principal Coordinate Analysis (DPCoA): \texttt{calculateDPCoA} and \texttt{runDPCoA}
\item
  Jensen-Shannon Divergence (JSD): \texttt{calculateJSD} and \texttt{runJSD}
\item
  MDS: \texttt{calculateMDS} and \texttt{runMDS}
\item
  NMDS: \texttt{calculateNMDS} and \texttt{runNMDS}
\item
  Overlap: \texttt{calculateOverlap} and \texttt{runOverlap}
\item
  t-distributed Stochastic Neighbor Embedding (t-SNE): \texttt{calculateTSNE} and \texttt{runTSNE}
\item
  UMAP: \texttt{calculateUMAP} and \texttt{runUMAP}
\end{itemize}

For more information on clustering samples by beta diversity, you can refer to:

\begin{itemize}
\tightlist
\item
  \href{http://bioconductor.org/books/release/OSCA/clustering.html}{How to extract information from clusters}
\item
  Chapter \ref{clustering} on community typing
\end{itemize}

\hypertarget{microbiome-community}{%
\chapter{Community Composition}\label{microbiome-community}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}
\end{Highlighting}
\end{Shaded}

\hypertarget{visual-composition}{%
\section{Visualizing taxonomic composition}\label{visual-composition}}

\hypertarget{composition-barplot}{%
\subsection{Composition barplot}\label{composition-barplot}}

A typical way to visualize microbiome composition is by using
composition barplot. In the following, relative abundance is
calculated and top taxa are retrieved for the Phylum rank. Thereafter,
the barplot is visualized ordering rank by abundance values and
samples by ``Bacteroidetes'':

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(miaViz)}
\CommentTok{\# Computing relative abundance}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Getting top taxa on a Phylum level}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\AttributeTok{rank =}\StringTok{"Phylum"}\NormalTok{, }\AttributeTok{onRankOnly=}\ConstantTok{TRUE}\NormalTok{)}
\NormalTok{top\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse\_phylum,}\AttributeTok{top =} \DecValTok{5}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Renaming the "Phylum" rank to keep only top taxa and the rest to "Other"}
\NormalTok{phylum\_renamed }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum,}
                   \ControlFlowTok{function}\NormalTok{(x)\{}\ControlFlowTok{if}\NormalTok{ (x }\SpecialCharTok{\%in\%}\NormalTok{ top\_taxa) \{x\} }\ControlFlowTok{else}\NormalTok{ \{}\StringTok{"Other"}\NormalTok{\}\})}
\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Phylum }\OtherTok{\textless{}{-}} \FunctionTok{as.character}\NormalTok{(phylum\_renamed)}

\CommentTok{\# Visualizing the composition barplot, with samples order by "Bacteroidetes"}
\FunctionTok{plotAbundance}\NormalTok{(tse, }\AttributeTok{assay.type=}\StringTok{"relabundance"}\NormalTok{, }\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{,}
              \AttributeTok{order\_rank\_by=}\StringTok{"abund"}\NormalTok{, }
              \AttributeTok{order\_sample\_by =} \StringTok{"Bacteroidetes"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{21_microbiome_community_files/figure-latex/unnamed-chunk-1-1.pdf}

\hypertarget{composition-heatmap}{%
\subsection{Composition heatmap}\label{composition-heatmap}}

Community composition can be visualized with heatmap, where the
horizontal axis represents samples and the vertical axis the
taxa. Color of each intersection point represents abundance of a taxon
in a specific sample.

Here, abundances are first CLR (centered log-ratio) transformed to
remove compositionality bias. Then Z transformation is applied to
CLR-transformed data. This shifts all taxa to zero mean and unit
variance, allowing visual comparison between taxa that have different
absolute abundance levels. After these rough visual exploration
techniques, we can visualize the abundances at Phylum level.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}

\CommentTok{\# Add clr{-}transformation on samples}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                              \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum, }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
                              \AttributeTok{method =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\CommentTok{\# Add z{-}transformation on features (taxa)}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum, }\AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{, }
                              \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{,}
                              \AttributeTok{method =} \StringTok{"z"}\NormalTok{, }\AttributeTok{name =} \StringTok{"clr\_z"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Visualize as heatmap.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Melt the assay for plotting purposes}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{meltAssay}\NormalTok{(tse\_phylum, }\AttributeTok{assay.type =} \StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Determines the scaling of colours}
\NormalTok{maxval }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(df}\SpecialCharTok{$}\NormalTok{clr\_z)))}
\NormalTok{limits }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{maxval, maxval)}
\NormalTok{breaks }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\AttributeTok{from =} \FunctionTok{min}\NormalTok{(limits), }\AttributeTok{to =} \FunctionTok{max}\NormalTok{(limits), }\AttributeTok{by =} \FloatTok{0.5}\NormalTok{)}
\NormalTok{colours }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"darkblue"}\NormalTok{, }\StringTok{"blue"}\NormalTok{, }\StringTok{"white"}\NormalTok{, }\StringTok{"red"}\NormalTok{, }\StringTok{"darkred"}\NormalTok{)}

\CommentTok{\# Creates a ggplot object}
\FunctionTok{ggplot}\NormalTok{(df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ SampleID, }\AttributeTok{y =}\NormalTok{ FeatureID, }\AttributeTok{fill =}\NormalTok{ clr\_z)) }\SpecialCharTok{+}
  \FunctionTok{geom\_tile}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_gradientn}\NormalTok{(}\AttributeTok{name =} \StringTok{"CLR + Z transform"}\NormalTok{, }
                       \AttributeTok{breaks =}\NormalTok{ breaks, }\AttributeTok{limits =}\NormalTok{ limits, }\AttributeTok{colours =}\NormalTok{ colours) }\SpecialCharTok{+} 
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{text =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size=}\DecValTok{10}\NormalTok{),}
        \AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle=}\DecValTok{45}\NormalTok{, }\AttributeTok{hjust=}\DecValTok{1}\NormalTok{),}
        \AttributeTok{legend.key.size =} \FunctionTok{unit}\NormalTok{(}\DecValTok{1}\NormalTok{, }\StringTok{"cm"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Samples"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Taxa"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{21_microbiome_community_files/figure-latex/heatmapvisu-1.pdf}

\emph{pheatmap} is a package that provides methods to plot clustered heatmaps.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(pheatmap)}

\CommentTok{\# Takes subset: only samples from feces, skin, or tongue}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum[ , tse\_phylum}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Feces"}\NormalTok{, }\StringTok{"Skin"}\NormalTok{, }\StringTok{"Tongue"}\NormalTok{) ]}

\CommentTok{\# Add clr{-}transformation}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum\_subset,}
                         \AttributeTok{method =} \StringTok{"clr"}\NormalTok{,}
                 \AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum\_subset, }\AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{,}
                                     \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{, }
                                     \AttributeTok{method =} \StringTok{"z"}\NormalTok{, }\AttributeTok{name =} \StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Get n most abundant taxa, and subsets the data by them}
\NormalTok{top\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse\_phylum\_subset, }\AttributeTok{top =} \DecValTok{20}\NormalTok{)}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum\_subset[top\_taxa, ]}

\CommentTok{\# Gets the assay table}
\NormalTok{mat }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse\_phylum\_subset, }\StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Creates the heatmap}
\FunctionTok{pheatmap}\NormalTok{(mat)}
\end{Highlighting}
\end{Shaded}

\includegraphics{21_microbiome_community_files/figure-latex/pheatmap1-1.pdf}

We can create clusters by hierarchical clustering and add them to the plot.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ape)}

\CommentTok{\# Hierarchical clustering}
\NormalTok{taxa\_hclust }\OtherTok{\textless{}{-}} \FunctionTok{hclust}\NormalTok{(}\FunctionTok{dist}\NormalTok{(mat), }\AttributeTok{method =} \StringTok{"complete"}\NormalTok{)}

\CommentTok{\# Creates a phylogenetic tree}
\NormalTok{taxa\_tree }\OtherTok{\textless{}{-}} \FunctionTok{as.phylo}\NormalTok{(taxa\_hclust)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggtree)}

\CommentTok{\# Plot taxa tree}
\NormalTok{taxa\_tree }\OtherTok{\textless{}{-}} \FunctionTok{ggtree}\NormalTok{(taxa\_tree) }\SpecialCharTok{+} 
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{)) }\CommentTok{\# removes margins}

\CommentTok{\# Get order of taxa in plot}
\NormalTok{taxa\_ordered }\OtherTok{\textless{}{-}} \FunctionTok{get\_taxa\_name}\NormalTok{(taxa\_tree)}

\NormalTok{taxa\_tree}
\end{Highlighting}
\end{Shaded}

\includegraphics{21_microbiome_community_files/figure-latex/pheatmap3-1.pdf}

Based on phylo tree, we decide to create three clusters.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Creates clusters}
\NormalTok{taxa\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{cutree}\NormalTok{(}\AttributeTok{tree =}\NormalTok{ taxa\_hclust, }\AttributeTok{k =} \DecValTok{3}\NormalTok{)}

\CommentTok{\# Converts into data frame}
\NormalTok{taxa\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{clusters =}\NormalTok{ taxa\_clusters)}
\NormalTok{taxa\_clusters}\SpecialCharTok{$}\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(taxa\_clusters}\SpecialCharTok{$}\NormalTok{clusters)}

\CommentTok{\# Order data so that it\textquotesingle{}s same as in phylo tree}
\NormalTok{taxa\_clusters }\OtherTok{\textless{}{-}}\NormalTok{ taxa\_clusters[taxa\_ordered, , drop }\OtherTok{=} \ConstantTok{FALSE}\NormalTok{] }

\CommentTok{\# Prints taxa and their clusters}
\NormalTok{taxa\_clusters}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                  clusters
## Chloroflexi             3
## Actinobacteria          3
## Crenarchaeota           3
## Planctomycetes          3
## Gemmatimonadetes        3
## Thermi                  3
## Acidobacteria           3
## Spirochaetes            2
## Fusobacteria            2
## SR1                     2
## Cyanobacteria           2
## Proteobacteria          2
## Synergistetes           2
## Lentisphaerae           1
## Bacteroidetes           1
## Verrucomicrobia         1
## Tenericutes             1
## Firmicutes              1
## Euryarchaeota           1
## SAR406                  1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Adds information to rowData}
\FunctionTok{rowData}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{clusters }\OtherTok{\textless{}{-}}\NormalTok{ taxa\_clusters[}\FunctionTok{order}\NormalTok{(}\FunctionTok{match}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(taxa\_clusters), }\FunctionTok{rownames}\NormalTok{(tse\_phylum\_subset))), ]}

\CommentTok{\# Prints taxa and their clusters}
\FunctionTok{rowData}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{clusters}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  [1] 1 1 2 3 2 2 1 1 1 1 3 2 3 3 3 2 2 3 3 1
## Levels: 1 2 3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Hierarchical clustering}
\NormalTok{sample\_hclust }\OtherTok{\textless{}{-}} \FunctionTok{hclust}\NormalTok{(}\FunctionTok{dist}\NormalTok{(}\FunctionTok{t}\NormalTok{(mat)), }\AttributeTok{method =} \StringTok{"complete"}\NormalTok{)}

\CommentTok{\# Creates a phylogenetic tree}
\NormalTok{sample\_tree }\OtherTok{\textless{}{-}} \FunctionTok{as.phylo}\NormalTok{(sample\_hclust)}

\CommentTok{\# Plot sample tree}
\NormalTok{sample\_tree }\OtherTok{\textless{}{-}} \FunctionTok{ggtree}\NormalTok{(sample\_tree) }\SpecialCharTok{+} \FunctionTok{layout\_dendrogram}\NormalTok{() }\SpecialCharTok{+} 
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{)) }\CommentTok{\# removes margins}

\CommentTok{\# Get order of samples in plot}
\NormalTok{samples\_ordered }\OtherTok{\textless{}{-}} \FunctionTok{rev}\NormalTok{(}\FunctionTok{get\_taxa\_name}\NormalTok{(sample\_tree))}

\NormalTok{sample\_tree}
\end{Highlighting}
\end{Shaded}

\includegraphics{21_microbiome_community_files/figure-latex/pheatmap6-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Creates clusters}
\NormalTok{sample\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(}\FunctionTok{cutree}\NormalTok{(}\AttributeTok{tree =}\NormalTok{ sample\_hclust, }\AttributeTok{k =} \DecValTok{3}\NormalTok{))}

\CommentTok{\# Converts into data frame}
\NormalTok{sample\_data }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{clusters =}\NormalTok{ sample\_clusters)}

\CommentTok{\# Order data so that it\textquotesingle{}s same as in phylo tree}
\NormalTok{sample\_data }\OtherTok{\textless{}{-}}\NormalTok{ sample\_data[samples\_ordered, , drop }\OtherTok{=} \ConstantTok{FALSE}\NormalTok{] }

\CommentTok{\# Order data based on }
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum\_subset[ , }\FunctionTok{rownames}\NormalTok{(sample\_data)]}

\CommentTok{\# Add sample type data}
\NormalTok{sample\_data}\SpecialCharTok{$}\NormalTok{sample\_types }\OtherTok{\textless{}{-}} \FunctionTok{unfactor}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{SampleType)}

\NormalTok{sample\_data}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##         clusters sample_types
## M11Plmr        2         Skin
## M31Plmr        2         Skin
## F21Plmr        2         Skin
## M31Fcsw        1        Feces
## M11Fcsw        1        Feces
## TS28           3        Feces
## TS29           3        Feces
## M31Tong        3       Tongue
## M11Tong        3       Tongue
\end{verbatim}

Now we can create heatmap with additional annotations.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Determines the scaling of colorss}
\CommentTok{\# Scale colors}
\NormalTok{breaks }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\SpecialCharTok{{-}}\FunctionTok{ceiling}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))), }\FunctionTok{ceiling}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))), }
              \AttributeTok{length.out =} \FunctionTok{ifelse}\NormalTok{( }\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))}\SpecialCharTok{\textgreater{}}\DecValTok{5}\NormalTok{, }\DecValTok{2}\SpecialCharTok{*}\FunctionTok{ceiling}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))), }\DecValTok{10}\NormalTok{ ) )}
\NormalTok{colors }\OtherTok{\textless{}{-}} \FunctionTok{colorRampPalette}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"darkblue"}\NormalTok{, }\StringTok{"blue"}\NormalTok{, }\StringTok{"white"}\NormalTok{, }\StringTok{"red"}\NormalTok{, }\StringTok{"darkred"}\NormalTok{))(}\FunctionTok{length}\NormalTok{(breaks)}\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{)}

\FunctionTok{pheatmap}\NormalTok{(mat, }\AttributeTok{annotation\_row =}\NormalTok{ taxa\_clusters, }
         \AttributeTok{annotation\_col =}\NormalTok{ sample\_data,}
         \AttributeTok{breaks =}\NormalTok{ breaks,}
         \AttributeTok{color =}\NormalTok{ colors)}
\end{Highlighting}
\end{Shaded}

\includegraphics{21_microbiome_community_files/figure-latex/pheatmap8-1.pdf}

In addition, there are also other packages that provide functions for
more complex heatmaps, such as
\href{https://docs.ropensci.org/iheatmapr/articles/full_vignettes/iheatmapr.html}{\emph{iheatmapr}}
and ComplexHeatmap \citep{ComplexHeatmap}.
\href{http://www.bioconductor.org/packages/release/bioc/vignettes/sechm/inst/doc/sechm.html}{sechm}
package provides wrapper for \emph{ComplexHeatmap} and its usage is
explained in chapter \ref{viz-chapter} along with the \texttt{pheatmap}
package for clustered heatmaps.

\hypertarget{clustering}{%
\chapter{Community Typing (Clustering)}\label{clustering}}

Clustering techniques are unsupervised machine learning that aims to find groups, called clusters, that share a pattern in the data. In the microbiome context, clustering techniques are included in microbiome community typing methods. For example, clustering allow samples to be distinguished from each other based on their microbiome community composition. There are multiple clustering algorithms available.

The data can be clustered either based on features or samples. Hence, depending on the analysis goal the data might require transformation. The examples below are focused on sample clustering. To learn about feature clustering, check out chapter \ref{taxa-clustering}.

\hypertarget{custom-tools}{%
\section{Custom tools}\label{custom-tools}}

\emph{bluster} is a Bioconductor package providing tools for clustering data in
in the \texttt{SummarizedExperiment} container. It offers multiple algorithms such
as hierarchical clustering, DBSCAN, and K-means.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load dependencies}
\FunctionTok{library}\NormalTok{(bluster)}
\FunctionTok{library}\NormalTok{(kableExtra)}
\end{Highlighting}
\end{Shaded}

In the first examples of microbiome community typing we use \texttt{enterotype} data.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"enterotype"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ enterotype}

\CommentTok{\# Apply transformation}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The main focus in this example is to show how to use mia's \texttt{cluster} function to
cluster enterotype data. \texttt{cluster} function allows to choose a clustering algorithm and offers multiple parameters to shape the result.

In this example, HclustParam() parameter is chosen for hierarchical clustering.
HclustParam() parameter itself has parameters on its own
\href{https://rdrr.io/github/LTLA/bluster/man/HclustParam-class.html}{HclustParam documentation}.
A parameter is \texttt{MARGIN} defines whether to cluster features or samples .

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Simple use of the hierarchical clustering. Here, the default parameters}
\CommentTok{\# Set the cut height to half of the dendrogram height}

\CommentTok{\# Save as an alternative experiment that contains clustering information}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{cluster}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{, }
               \AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{, }\FunctionTok{HclustParam}\NormalTok{())}

\CommentTok{\# The result can be found in \textquotesingle{}clusters\textquotesingle{} column of colData}

\CommentTok{\# The number of samples included in each cluster}
\FunctionTok{summary}\NormalTok{(}\FunctionTok{colData}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{clusters)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  1  2  3  4  5  6  7  8  9 
## 92 25 21  2 55 41 20 23  1
\end{verbatim}

Once the clustering on the samples is done, we can also plot the clusters.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(scater)}

\CommentTok{\# Add the MDS dimensions for plotting}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust"}\NormalTok{), }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{, }
              \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist, }\AttributeTok{method =} \StringTok{"bray"}\NormalTok{)}

\CommentTok{\# Plot the clusters}
\FunctionTok{plotReducedDim}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust"}\NormalTok{), }\StringTok{"MDS"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"clusters"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/bluster_sample_plot-1.pdf}

\hypertarget{hierarchical-clustering}{%
\section{Hierarchical clustering}\label{hierarchical-clustering}}

The hierarchical clustering algorithm aims to find hierarchy between
samples/features. There are to approaches: agglomerative (``bottom-up'')
and divisive (``top-down''). In agglomerative approach, each observation is first in a unique cluster. The algorithm continues by agglomerating similar clusters. The divisive approach, instead, starts with one cluster that contains all observations. Clusters are split recursively into clusters that differ the most. The clustering ends when each cluster contains only one observation. In this algorithm, the similarity of two clusters is based on the distance
between them.

Hierarchical clustering can be visualized with a dendrogram tree. In each
splitting point, the tree is divided into two clusters leading to the
hierarchy.

Hierarchical clustering requires two steps.
1. Computation of the dissimilarities with a given distance.\\
2. Clustering based on dissimilarities.

Additionally, since sequencing data is compositional, we apply relative
transformation (as seen in the previous example).

In this example, we want to add information on the clustering. To do so,
we use the \texttt{full} parameter. We also compute the dissimilarities with the
bray distance. Finally, the \texttt{clust.col} parameter allows us to choose the name
of the column in the colData (default name is \texttt{clusters}).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(vegan)}

\CommentTok{\# Save another alternative experiment that contains full clustering information}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust\_full"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{cluster}\NormalTok{(tse,}
               \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
               \AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{,}
               \FunctionTok{HclustParam}\NormalTok{(}\AttributeTok{method =} \StringTok{"complete"}\NormalTok{,}
                           \AttributeTok{dist.fun =}\NormalTok{ vegdist,}
                           \AttributeTok{metric =} \StringTok{"bray"}\NormalTok{),}
               \AttributeTok{full =} \ConstantTok{TRUE}\NormalTok{,}
               \AttributeTok{clust.col =} \StringTok{"Hclust"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We plot the dendrogram, which is possible since we got the
additional information from the clustering.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(dendextend)}

\CommentTok{\# Get hclust data from metadata}
\NormalTok{hclust\_data }\OtherTok{\textless{}{-}} \FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust\_full"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{clusters}\SpecialCharTok{$}\NormalTok{hclust}

\CommentTok{\# Get the dendrogram object}
\NormalTok{dendro }\OtherTok{\textless{}{-}} \FunctionTok{as.dendrogram}\NormalTok{(hclust\_data)}

\CommentTok{\# Plot dendrogram}
\NormalTok{dendro }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{set}\NormalTok{(}\StringTok{"labels"}\NormalTok{, }\ConstantTok{NULL}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{plot}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/hclust3-1.pdf}

In our case, we cut the dendrogram in half by default. To know how many clusters
we have, we can check the colData.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get the clusters}
\FunctionTok{summary}\NormalTok{(}\FunctionTok{colData}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust\_full"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{Hclust)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 
##  3 65 35 12  9  2 23 23 27 15  4 11  6 11  2  1  9  7  4  4  1  2  1  1  1  1
\end{verbatim}

We can see that there are 26 clusters, but that probably is not optimal since the
the number of clusters has been chosen arbitrarily. To determine the number of
clusters, we can use the dendrogram. Usually the tree is split where the branch
length is the largest. However, as we can see from the dendrogram, clusters are
not clear. There are algorithms to identify the optimal number of clusters.

The \texttt{NbClust} library is useful to that end as it offers multiple methods to
determine the optimal number of clusters. Here we will use the silhouette
analysis to determine the optimal number of clusters. For each data point, this
analysis measures the distance to other data points in the same cluster
(cohesion), and the distance to the other clusters (separation), establishing a
score. That score is then combined across the data points. \texttt{NbClust} does
this for multiple number of clusters and the best score corresponds to the
optimal number of clusters.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(NbClust)}
\NormalTok{diss }\OtherTok{\textless{}{-}} \FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"hclust\_full"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{clusters}\SpecialCharTok{$}\NormalTok{dist}

\CommentTok{\# Apply the silhouette analysis on the distance matrix}
\NormalTok{res }\OtherTok{\textless{}{-}} \FunctionTok{NbClust}\NormalTok{(}\AttributeTok{diss =}\NormalTok{ diss, }\AttributeTok{distance =} \ConstantTok{NULL}\NormalTok{, }\AttributeTok{method =} \StringTok{"ward.D2"}\NormalTok{,}
               \AttributeTok{index =} \StringTok{"silhouette"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##  Only frey, mcclain, cindex, sihouette and dunn can be computed. To compute the other indices, data matrix is needed
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{res}\SpecialCharTok{$}\NormalTok{Best.nc}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Number_clusters     Value_Index 
##          2.0000          0.4783
\end{verbatim}

Based on the result, let's divide observations into 2 clusters.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(dendextend)}

\CommentTok{\# Get optimal number of clusters}
\NormalTok{k }\OtherTok{\textless{}{-}}\NormalTok{ res}\SpecialCharTok{$}\NormalTok{Best.nc[}\DecValTok{1}\NormalTok{]}

\CommentTok{\# Making colors for 2 clusters}
\NormalTok{col\_val\_map }\OtherTok{\textless{}{-}}\NormalTok{ randomcoloR}\SpecialCharTok{::}\FunctionTok{distinctColorPalette}\NormalTok{(k) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{as.list}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{setNames}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(}\StringTok{"clust\_"}\NormalTok{, }\FunctionTok{seq}\NormalTok{(k)))}

\NormalTok{dend }\OtherTok{\textless{}{-}} \FunctionTok{color\_branches}\NormalTok{(dendro, }\AttributeTok{k =}\NormalTok{ k, }\AttributeTok{col =} \FunctionTok{unlist}\NormalTok{(col\_val\_map))}
\FunctionTok{labels}\NormalTok{(dend) }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\FunctionTok{plot}\NormalTok{(dend)}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/hclust6-1.pdf}

\hypertarget{dirichlet-multinomial-mixtures-dmm}{%
\section{Dirichlet Multinomial Mixtures (DMM)}\label{dirichlet-multinomial-mixtures-dmm}}

This section focus on \href{https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0030126}{Dirichlet-Multinomial Mixture
Model} analysis. It is a probabilistic technique that allows to search for sample patterns that
reflect sample similarity in the data. DMM has a property of determining an optimal number of clusters (k) to obtain the best model.
The minimum value of Laplace approximation to the negative log model evidence for DMM models as a function of k, determines an optimal k. The optimal k suggests to fit a model with k mixtures of Dirichlet distributions. For the best model, k probabilities for each sample to belong to each cluster are obtained.

In this example, we cluster the data with DMM clustering. Since the data set is large, the
algorithm requires a lot of computational capacity. Therefore, we use only a subset
of the data that is agglomerated by Phylum as a rank.

In the example of DMM we use \texttt{GlobalPatterns} data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get the data}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}

\CommentTok{\# Agglomerate by rank}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{, }\AttributeTok{agglomerateTree =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

In the example below, we calculate model fit using Laplace approximation. The cluster information is added in the metadata with an optional \texttt{name}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Run the model and calculates the most likely number of clusters from 1 to 7}

\CommentTok{\# Save as an alternative experiment that contains clustering information}
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{cluster}\NormalTok{(tse, }\AttributeTok{name =} \StringTok{"DMM"}\NormalTok{, }\FunctionTok{DmmParam}\NormalTok{(}\AttributeTok{k =} \DecValTok{1}\SpecialCharTok{:}\DecValTok{7}\NormalTok{, }\AttributeTok{type =} \StringTok{"laplace"}\NormalTok{), }
                   \AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{, }\AttributeTok{full =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# The dmm information is stored in the metadata under the \textquotesingle{}DMM\textquotesingle{} column that includes information about all seven models }
\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 67 26 
## metadata(2): agglomerated_by_rank DMM
## assays(1): counts
## rownames(67): Phylum:Crenarchaeota Phylum:Euryarchaeota ...
##   Phylum:Synergistetes Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(8): X.SampleID Primer ... Description clusters
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (67 rows)
## rowTree: 1 phylo tree(s) (66 leaves)
## colLinks: NULL
## colTree: NULL
\end{verbatim}

The plot below represents the Laplace approximation to the model evidence for each of the k models. We can see that the best number of clusters is two.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(miaViz)}
\FunctionTok{plotDMNFit}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{), }\AttributeTok{type =} \StringTok{"laplace"}\NormalTok{, }\AttributeTok{name =} \StringTok{"DMM"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/dmm4-1.pdf}

The best model can be confirmed with the following operation.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get the the best model}
\NormalTok{bestFit }\OtherTok{\textless{}{-}} \FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{DMM}\SpecialCharTok{$}\NormalTok{dmm[[}\FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{DMM}\SpecialCharTok{$}\NormalTok{best]]}
\NormalTok{bestFit}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: DMN 
## k: 2 
## samples x taxa: 26 x 67 
## Laplace: 7673 BIC: 7927 AIC: 7842
\end{verbatim}

The clusters for the best model are saved in the colData under `cluster' column.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{colData}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{clusters, }\DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong 
##       1       1       1       2       2       2       2       2       2       2 
## Levels: 1 2
\end{verbatim}

More detailed information about the clusters can be accessed in the metadata. The metadata contains samples-cluster assignment probabilities that tell us the likelihood for each sample to belong to each cluster.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(}\FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{DMM}\SpecialCharTok{$}\NormalTok{prob, }\DecValTok{10}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                 1         2
## CL3     1.000e+00 5.080e-17
## CC1     1.000e+00 3.936e-22
## SV1     1.000e+00 1.963e-12
## M31Fcsw 7.914e-26 1.000e+00
## M11Fcsw 1.130e-16 1.000e+00
## M31Plmr 1.121e-13 1.000e+00
## M11Plmr 3.441e-06 1.000e+00
## F21Plmr 4.360e-11 1.000e+00
## M31Tong 1.449e-08 1.000e+00
## M11Tong 2.223e-06 1.000e+00
\end{verbatim}

Once the optimal model have been confirmed, we can find out which samples are grouped with each other. The table below shows one sample of each sample type clustered in either of the groups. We can notice that DMM can distinguish environmental samples into one group, and mock and human samples into another. For clarity, in this example, the probabilities for each sample to belong in each cluster have been rounded.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(dplyr)}

\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}\FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{DMM}\SpecialCharTok{$}\NormalTok{prob, }\DecValTok{1}\NormalTok{)}
\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{cbind}\NormalTok{(clusters, }\FunctionTok{levels}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{)}\SpecialCharTok{$}\NormalTok{SampleType)[}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{)}\SpecialCharTok{$}\NormalTok{SampleType])) }\CommentTok{\# add sample type information}
\FunctionTok{colnames}\NormalTok{(clusters) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Group1"}\NormalTok{, }\StringTok{"Group2"}\NormalTok{, }\StringTok{"SampleType"}\NormalTok{)}

\NormalTok{clusters }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{group\_by}\NormalTok{(SampleType) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(Group1) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\FunctionTok{row\_number}\NormalTok{()}\SpecialCharTok{==}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 9 x 3
## # Groups:   SampleType [9]
##   Group1 Group2 SampleType        
##   <chr>  <chr>  <chr>             
## 1 0      1      Feces             
## 2 0      1      Skin              
## 3 0      1      Tongue            
## 4 0      1      Mock              
## 5 1      0      Soil              
## 6 1      0      Freshwater        
## 7 1      0      Freshwater (creek)
## 8 1      0      Ocean             
## 9 1      0      Sediment (estuary)
\end{verbatim}

We can also plot the driver Phyla in each group. In this case, it reflects the differences between environmental and human samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get the estimates on how much each phyla contributes on each cluster}
\NormalTok{best\_model }\OtherTok{\textless{}{-}} \FunctionTok{metadata}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{DMM}\SpecialCharTok{$}\NormalTok{dmm[}\DecValTok{2}\NormalTok{]}
\NormalTok{drivers }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(best\_model[[}\DecValTok{1}\NormalTok{]]}\SpecialCharTok{@}\NormalTok{fit}\SpecialCharTok{$}\NormalTok{Estimate)}

\NormalTok{drivers}\SpecialCharTok{$}\NormalTok{phyla }\OtherTok{\textless{}{-}} \FunctionTok{gsub}\NormalTok{(}\StringTok{"Phylum:"}\NormalTok{, }\StringTok{""}\NormalTok{, }\FunctionTok{rownames}\NormalTok{(drivers)) }\CommentTok{\# Clean phylum names}

\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\DecValTok{2}\NormalTok{) \{}
\NormalTok{  drivers }\OtherTok{\textless{}{-}}\NormalTok{ drivers[}\FunctionTok{order}\NormalTok{(drivers[[i]], }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{),]}
\NormalTok{  p }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\FunctionTok{head}\NormalTok{(drivers, }\DecValTok{10}\NormalTok{), }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =} \FunctionTok{reorder}\NormalTok{(}\FunctionTok{head}\NormalTok{(phyla, }\DecValTok{10}\NormalTok{), }\SpecialCharTok{+} \FunctionTok{head}\NormalTok{(drivers[[i]], }\DecValTok{10}\NormalTok{)), }\AttributeTok{y =} \FunctionTok{head}\NormalTok{(drivers[[i]], }\DecValTok{10}\NormalTok{))) }\SpecialCharTok{+}
          \FunctionTok{geom\_bar}\NormalTok{(}\AttributeTok{stat =} \StringTok{"identity"}\NormalTok{, }\AttributeTok{fill =} \StringTok{"deeppink4"}\NormalTok{, }\AttributeTok{alpha =} \FloatTok{0.5}\NormalTok{) }\SpecialCharTok{+}
          \FunctionTok{coord\_flip}\NormalTok{() }\SpecialCharTok{+} \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \FunctionTok{paste}\NormalTok{(}\StringTok{"Top phyla in group"}\NormalTok{, i)) }\SpecialCharTok{+}
          \FunctionTok{theme\_light}\NormalTok{(}\AttributeTok{base\_size =} \DecValTok{15}\NormalTok{) }\SpecialCharTok{+} \FunctionTok{labs}\NormalTok{(}\AttributeTok{x=}\StringTok{""}\NormalTok{, }\AttributeTok{y=}\StringTok{""}\NormalTok{) }\SpecialCharTok{+} \FunctionTok{scale\_y\_continuous}\NormalTok{(}\AttributeTok{limits=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{7}\NormalTok{))}

  \FunctionTok{print}\NormalTok{(p)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/dmm9-1.pdf} \includegraphics{24_clustering_files/figure-latex/dmm9-2.pdf}

We use \texttt{calculateDMNgroup} function to have an overview of the best model. The function groups samples by \texttt{SampleType} column from colData and returns DMNGroup object that contains a summary.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{dmm\_group }\OtherTok{\textless{}{-}} \FunctionTok{calculateDMNgroup}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{), }\AttributeTok{variable =} \StringTok{"SampleType"}\NormalTok{, }
                               \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{k =} \DecValTok{2}\NormalTok{, }
                               \AttributeTok{seed =}\NormalTok{ .Machine}\SpecialCharTok{$}\NormalTok{integer.max)}

\NormalTok{dmm\_group}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: DMNGroup 
## summary:
##                    k samples taxa    NLE  LogDet Laplace    BIC  AIC
## Feces              2       4   67 1078.3 -106.26   901.1 1171.9 1213
## Freshwater         2       2   67  889.6  -97.20   716.9  936.4 1025
## Freshwater (creek) 2       3   67 1600.3  862.19  1907.3 1674.5 1735
## Mock               2       3   67 1008.4  -55.40   856.6 1082.5 1143
## Ocean              2       3   67 1096.7  -56.66   944.3 1170.9 1232
## Sediment (estuary) 2       3   67 1195.5   18.63  1080.8 1269.7 1331
## Skin               2       3   67  992.6  -85.05   826.1 1066.8 1128
## Soil               2       3   67 1380.3   11.20  1261.8 1454.5 1515
## Tongue             2       2   67  783.0 -107.79   605.0  829.8  918
\end{verbatim}

Mixture weights can be used for having a rough approximation of the cluster size.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{DirichletMultinomial}\SpecialCharTok{::}\FunctionTok{mixturewt}\NormalTok{(bestFit)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##       pi theta
## 1 0.5385 20.58
## 2 0.4615 15.29
\end{verbatim}

\hypertarget{pcoa-with-dmm-clusters}{%
\subsection{PCoA with DMM clusters}\label{pcoa-with-dmm-clusters}}

In this section we show how to calculate principal coordinates for clr transformed abundance data. To calculate PCoA, we use Aitchison distance as a distance metrics that calculates Euclidean distances for clr transformed compositions.

In the visualization section, we project the sample distances on two dimensional space of first two principal coordinates. We colour the samples based on their DMM clusters. The visualization demonstrates that the DMM clusters can be distinguished on a PCoA plot, although the clusters are not coherent. This means that two-dimensional representation of the data created by PCoA preserves similar information that drives the DMM cluster division.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# add pseudocount, because data contains zeros}
\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"pseudo"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{) }\SpecialCharTok{+} \DecValTok{1}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"pseudo"}\NormalTok{, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# clr transformation}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\StringTok{"relabundance"}\NormalTok{, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{)}

\CommentTok{\# principal coordinate analysis}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{calculateMDS}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{method =} \StringTok{"euclidean"}\NormalTok{)}

\CommentTok{\# Create a data frame from principal coordinates}
\NormalTok{euclidean\_pcoa\_df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{pcoa1 =}\NormalTok{ df[, }\DecValTok{1}\NormalTok{], }\AttributeTok{pcoa2 =}\NormalTok{ df[, }\DecValTok{2}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create a data frame that contains principal coordinates and DMM information}
\NormalTok{euclidean\_dmm\_pcoa\_df }\OtherTok{\textless{}{-}} \FunctionTok{cbind}\NormalTok{(euclidean\_pcoa\_df,}
                               \AttributeTok{dmm\_component =} \FunctionTok{colData}\NormalTok{(}\FunctionTok{altExp}\NormalTok{(tse, }\StringTok{"dmm"}\NormalTok{))}\SpecialCharTok{$}\NormalTok{clusters)}

\CommentTok{\# Create a plot}
\NormalTok{euclidean\_dmm\_plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ euclidean\_dmm\_pcoa\_df,}
                             \FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ pcoa1, }\AttributeTok{y =}\NormalTok{ pcoa2, }\AttributeTok{color =}\NormalTok{ dmm\_component)) }\SpecialCharTok{+}
    \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Coordinate 1"}\NormalTok{,}\AttributeTok{y =} \StringTok{"Coordinate 2"}\NormalTok{, }
         \AttributeTok{title =} \StringTok{"PCoA with Aitchison distances"}\NormalTok{) }\SpecialCharTok{+}
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{plot.title =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{12}\NormalTok{, }\CommentTok{\# makes titles smaller}
                                    \AttributeTok{hjust =} \FloatTok{0.5}\NormalTok{)) }

\NormalTok{euclidean\_dmm\_plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/dmm12-1.pdf}

\hypertarget{biclustering}{%
\section{Biclustering}\label{biclustering}}

Biclustering methods cluster rows and columns simultaneously in order
to find subsets of correlated features/samples.

Here, we use following packages:

\begin{itemize}
\tightlist
\item
  \href{https://cran.r-project.org/web/packages/biclust/index.html}{biclust}
\item
  \href{https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13582}{cobiclust}
\end{itemize}

\emph{cobiclust} is especially developed for microbiome data whereas \emph{biclust} is more
general method. In this section, we show two different cases and example
solutions to apply biclustering to them.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Taxa vs samples
\item
  Taxa vs biomolecule/biomarker
\end{enumerate}

Biclusters can be visualized using heatmap or boxplot, for
instance. For checking purposes, also scatter plot might be valid
choice.

Check more ideas for heatmaps from chapters \ref{viz-chapter} and
\ref{microbiome-community}.

\hypertarget{taxa-vs-samples}{%
\subsection{Taxa vs samples}\label{taxa-vs-samples}}

When you have microbial abundance matrices, we suggest to use
\emph{cobiclust} which is designed for microbial data.

Load example data

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(cobiclust)}
\FunctionTok{data}\NormalTok{(}\StringTok{"HintikkaXOData"}\NormalTok{)}
\NormalTok{mae }\OtherTok{\textless{}{-}}\NormalTok{ HintikkaXOData}
\end{Highlighting}
\end{Shaded}

Only the most prevalent taxa are included in analysis.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Subset data in the first experiment}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{subsetByPrevalentFeatures}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{, }
                                      \AttributeTok{prevalence =} \FloatTok{0.2}\NormalTok{, }
                                      \AttributeTok{detection =} \FloatTok{0.001}\NormalTok{)}

\CommentTok{\# rclr{-}transform in the first experiment}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{method =} \StringTok{"rclr"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\emph{cobiclust} takes counts table as an input and gives \emph{cobiclust} object as an output.
It includes clusters for taxa and samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Do clustering using counts table}
\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{cobiclust}\NormalTok{(}\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\StringTok{"counts"}\NormalTok{))}

\CommentTok{\# Get clusters}
\NormalTok{row\_clusters }\OtherTok{\textless{}{-}}\NormalTok{ clusters}\SpecialCharTok{$}\NormalTok{classification}\SpecialCharTok{$}\NormalTok{rowclass}
\NormalTok{col\_clusters }\OtherTok{\textless{}{-}}\NormalTok{ clusters}\SpecialCharTok{$}\NormalTok{classification}\SpecialCharTok{$}\NormalTok{colclass}

\CommentTok{\# Add clusters to rowdata and coldata}
\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(row\_clusters)}
\FunctionTok{colData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(col\_clusters)}

\CommentTok{\# Order data based on clusters}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ mae[[}\DecValTok{1}\NormalTok{]][}\FunctionTok{order}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{clusters),}
                     \FunctionTok{order}\NormalTok{(}\FunctionTok{colData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{clusters)]}

\CommentTok{\# Print clusters}
\NormalTok{clusters}\SpecialCharTok{$}\NormalTok{classification}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $rowclass
##   [1] 1 1 1 1 1 1 1 1 1 2 2 1 2 1 1 1 2 1 2 1 1 1 2 2 1 1 2 2 1 2 2 1 1 2 1 1 2
##  [38] 1 2 2 2 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 2 1 1 2 1 2 2 1 1 1 1 1 1 1 1 1 1 1
##  [75] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 2 2 2 2 1 1 1 1 1 1 1 2 2 1 1 1 1
## [112] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $colclass
##  C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 C20 
##   1   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 
## C21 C22 C23 C24 C25 C26 C27 C28 C29 C30 C31 C32 C33 C34 C35 C36 C37 C38 C39 C40 
##   2   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   1
\end{verbatim}

Next we can plot clusters. Annotated heatmap is a common choice.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(pheatmap)}
\CommentTok{\# z{-}transform for heatmap}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{assay.type =} \StringTok{"rclr"}\NormalTok{,}
                            \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{, }\AttributeTok{method =} \StringTok{"z"}\NormalTok{, }\AttributeTok{name =} \StringTok{"rclr\_z"}\NormalTok{)}

\CommentTok{\# Create annotations. When column names are equal, they should share levels.}
\CommentTok{\# Here samples include 3 clusters, and taxa 2. That is why we have to make}
\CommentTok{\# column names unique.}
\NormalTok{annotation\_col }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])[, }\StringTok{"clusters"}\NormalTok{, }\AttributeTok{drop =}\NormalTok{ F])}
\FunctionTok{colnames}\NormalTok{(annotation\_col) }\OtherTok{\textless{}{-}} \StringTok{"col\_clusters"}

\NormalTok{annotation\_row }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])[, }\StringTok{"clusters"}\NormalTok{, }\AttributeTok{drop =}\NormalTok{ F])}
\FunctionTok{colnames}\NormalTok{(annotation\_row) }\OtherTok{\textless{}{-}} \StringTok{"row\_clusters"}
\end{Highlighting}
\end{Shaded}

Plot the heatmap.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{pheatmap}\NormalTok{(}\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\StringTok{"rclr\_z"}\NormalTok{), }\AttributeTok{cluster\_rows =}\NormalTok{ F, }\AttributeTok{cluster\_cols =}\NormalTok{ F,}
         \AttributeTok{annotation\_col =}\NormalTok{ annotation\_col, }\AttributeTok{annotation\_row =}\NormalTok{ annotation\_row)}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/cobiclust_3b-1.pdf}

Boxplot is commonly used to summarize the results:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}
\FunctionTok{library}\NormalTok{(patchwork)}

\CommentTok{\# ggplot requires data in melted format}
\NormalTok{melt\_assay }\OtherTok{\textless{}{-}} \FunctionTok{meltAssay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{assay.type =} \StringTok{"rclr"}\NormalTok{, }
                        \AttributeTok{add\_col\_data =}\NormalTok{ T, }\AttributeTok{add\_row\_data =}\NormalTok{ T)}

\CommentTok{\# patchwork two plots side{-}by{-}side}
\NormalTok{p1 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(melt\_assay) }\SpecialCharTok{+}
    \FunctionTok{geom\_boxplot}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ clusters.x, }\AttributeTok{y =}\NormalTok{ rclr)) }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Taxa clusters"}\NormalTok{)}

\NormalTok{p2 }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(melt\_assay) }\SpecialCharTok{+}
    \FunctionTok{geom\_boxplot}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ clusters.y, }\AttributeTok{y =}\NormalTok{ rclr)) }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Sample clusters"}\NormalTok{)}

\NormalTok{p1 }\SpecialCharTok{+}\NormalTok{ p2}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/cobiclust_4-1.pdf}

\hypertarget{taxa-vs-biomolecules}{%
\subsection{Taxa vs biomolecules}\label{taxa-vs-biomolecules}}

Here, we analyze cross-correlation between taxa and metabolites. This
is a case, where we use \emph{biclust} method which is suitable for numeric
matrices in general. First we pre-process the data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Samples must be in equal order}
\CommentTok{\# (Only 1st experiment was ordered in cobiclust step leading to unequal order)}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ mae[[}\DecValTok{1}\NormalTok{]][, }\FunctionTok{colnames}\NormalTok{(mae[[}\DecValTok{2}\NormalTok{]])]}

\CommentTok{\# Make rownames unique since it is required by other steps}
\FunctionTok{rownames}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]) }\OtherTok{\textless{}{-}} \FunctionTok{make.unique}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]))}

\CommentTok{\# Transform the metabolites to be in log basis}
\NormalTok{mae[[}\DecValTok{2}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{2}\NormalTok{]], }\AttributeTok{assay.type =} \StringTok{"nmr"}\NormalTok{, }\AttributeTok{method =} \StringTok{"log10"}\NormalTok{)}

\CommentTok{\# Add missing data to the metabolites}
\NormalTok{replace\_na }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(row) \{}
\NormalTok{    na\_indices }\OtherTok{\textless{}{-}} \FunctionTok{which}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(row))}
\NormalTok{    non\_na\_values }\OtherTok{\textless{}{-}}\NormalTok{ row[}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(row)]}
\NormalTok{    row[na\_indices] }\OtherTok{\textless{}{-}} \FunctionTok{sample}\NormalTok{(non\_na\_values, }\FunctionTok{length}\NormalTok{(na\_indices), }\AttributeTok{replace =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{    row}
\NormalTok{\}}
\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{2}\NormalTok{]], }\StringTok{"log10"}\NormalTok{) }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(}\FunctionTok{apply}\NormalTok{(}\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{2}\NormalTok{]], }\StringTok{"log10"}\NormalTok{), }\DecValTok{1}\NormalTok{, replace\_na))}
\end{Highlighting}
\end{Shaded}

Next, we compute the spearman correlation matrix.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Calculate correlations}
\NormalTok{corr }\OtherTok{\textless{}{-}} \FunctionTok{getExperimentCrossCorrelation}\NormalTok{(mae, }\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{, }\AttributeTok{assay.type1 =} \StringTok{"rclr"}\NormalTok{,}
                                      \AttributeTok{assay.type2 =} \StringTok{"log10"}\NormalTok{, }\AttributeTok{mode =} \StringTok{"matrix"}\NormalTok{,}
                                      \AttributeTok{correlation =} \StringTok{"spearman"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\emph{biclust} takes a matrix as an input and returns a \emph{biclust} object.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(biclust)}
\CommentTok{\# Set seed for reproducibility}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{3973}\NormalTok{)}

\CommentTok{\# Find biclusters}
\NormalTok{bc }\OtherTok{\textless{}{-}} \FunctionTok{biclust}\NormalTok{(corr, }\AttributeTok{method =} \FunctionTok{BCPlaid}\NormalTok{(), }\AttributeTok{verbose =} \ConstantTok{FALSE}\NormalTok{)}

\NormalTok{bc}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## An object of class Biclust 
## 
## call:
##  biclust(x = corr, method = BCPlaid(), verbose = FALSE)
## 
## Number of Clusters found:  4 
## 
## First  4  Cluster sizes:
##                    BC 1 BC 2 BC 3 BC 4
## Number of Rows:      21   20    7    2
## Number of Columns:   13   10   10    9
\end{verbatim}

The object includes cluster information. However compared to
\emph{cobiclust}, \emph{biclust} object includes only information about clusters
that were found, not general cluster.

Meaning that if one cluster size of 5 features was found out of 20 features,
those 15 features do not belong to any cluster. That is why we have to create an
additional cluster for features/samples that are not assigned into any cluster.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Functions for obtaining biclust information}

\CommentTok{\# Get clusters for rows and columns}
\NormalTok{.get\_biclusters\_from\_biclust }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(bc, assay) \{}
    \CommentTok{\# Get cluster information for columns and rows}
\NormalTok{    bc\_columns }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(bc}\SpecialCharTok{@}\NormalTok{NumberxCol)}
\NormalTok{    bc\_columns }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(bc\_columns)}
\NormalTok{    bc\_rows }\OtherTok{\textless{}{-}}\NormalTok{ bc}\SpecialCharTok{@}\NormalTok{RowxNumber}
\NormalTok{    bc\_rows }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(bc\_rows)}

    \CommentTok{\# Get data into right format}
\NormalTok{    bc\_columns }\OtherTok{\textless{}{-}} \FunctionTok{.manipulate\_bc\_data}\NormalTok{(bc\_columns, assay, }\StringTok{"col"}\NormalTok{)}
\NormalTok{    bc\_rows }\OtherTok{\textless{}{-}} \FunctionTok{.manipulate\_bc\_data}\NormalTok{(bc\_rows, assay, }\StringTok{"row"}\NormalTok{)}
    
    \FunctionTok{return}\NormalTok{(}\FunctionTok{list}\NormalTok{(}\AttributeTok{bc\_columns =}\NormalTok{ bc\_columns, }\AttributeTok{bc\_rows =}\NormalTok{ bc\_rows))}
\NormalTok{\}}

\CommentTok{\# Input clusters, and how many observations there should be, i.e.,}
\CommentTok{\# the number of samples or features}
\NormalTok{.manipulate\_bc\_data }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(bc\_clusters, assay, row\_col) \{}
    \CommentTok{\# Get right dimension}
\NormalTok{    dim }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{(row\_col }\SpecialCharTok{==} \StringTok{"col"}\NormalTok{, }\FunctionTok{ncol}\NormalTok{(assay), }\FunctionTok{nrow}\NormalTok{(assay))}
    \CommentTok{\# Get column/row names}
    \ControlFlowTok{if}\NormalTok{ (row\_col }\SpecialCharTok{==} \StringTok{"col"}\NormalTok{) \{}
\NormalTok{        names }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(assay)}
\NormalTok{    \} }\ControlFlowTok{else}\NormalTok{ \{}
\NormalTok{        names }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(assay)}
\NormalTok{    \}}

    \CommentTok{\# If no clusters were found, create one. Otherwise create additional}
    \CommentTok{\# cluster which}
    \CommentTok{\# contain those samples that are not included in clusters that were found.}
    \ControlFlowTok{if}\NormalTok{ (}\FunctionTok{nrow}\NormalTok{(bc\_clusters) }\SpecialCharTok{!=}\NormalTok{ dim) \{}
\NormalTok{        bc\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{cluster =} \FunctionTok{rep}\NormalTok{(}\ConstantTok{TRUE}\NormalTok{, dim))}
\NormalTok{    \} }\ControlFlowTok{else}\NormalTok{ \{}
        \CommentTok{\# Create additional cluster that includes those samples/features that}
        \CommentTok{\# are not included in other clusters.}
\NormalTok{        vec }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{rowSums}\NormalTok{(bc\_clusters) }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{, }\ConstantTok{FALSE}\NormalTok{, }\ConstantTok{TRUE}\NormalTok{)}

        \CommentTok{\# If additional cluster contains samples, then add it}
        \ControlFlowTok{if}\NormalTok{ (}\FunctionTok{any}\NormalTok{(vec)) \{}
\NormalTok{            bc\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{cbind}\NormalTok{(bc\_clusters, vec)}
\NormalTok{        \}}
\NormalTok{    \}}
    
    \CommentTok{\# Adjust row and column names}
    \FunctionTok{rownames}\NormalTok{(bc\_clusters) }\OtherTok{\textless{}{-}}\NormalTok{ names}
    \FunctionTok{colnames}\NormalTok{(bc\_clusters) }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"cluster\_"}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\FunctionTok{ncol}\NormalTok{(bc\_clusters))}
    \FunctionTok{return}\NormalTok{(bc\_clusters)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get biclusters}
\NormalTok{bcs }\OtherTok{\textless{}{-}} \FunctionTok{.get\_biclusters\_from\_biclust}\NormalTok{(bc, corr)}

\NormalTok{bicluster\_rows }\OtherTok{\textless{}{-}}\NormalTok{ bcs}\SpecialCharTok{$}\NormalTok{bc\_rows}
\NormalTok{bicluster\_columns }\OtherTok{\textless{}{-}}\NormalTok{ bcs}\SpecialCharTok{$}\NormalTok{bc\_columns}

\CommentTok{\# Print biclusters for rows}
\FunctionTok{head}\NormalTok{(bicluster\_rows)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                                                                                                                       cluster_1
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_2
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                           TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_3
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                           TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_4
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE
##                                                                                                                       cluster_5
## D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                    TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella                TRUE
## D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                 TRUE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella      TRUE
## D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE
## D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                   TRUE
\end{verbatim}

Let's collect information for the scatter plot.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Function for obtaining sample{-}wise sum, mean, median, and mean variance}
\CommentTok{\# for each cluster}

\NormalTok{.sum\_mean\_median\_var }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(tse1, tse2, assay.type1, assay.type2, clusters1, clusters2) \{}
\NormalTok{    list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
    \CommentTok{\# Create a data frame that includes all the information}
    \ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{ncol}\NormalTok{(clusters1)) \{}
        \CommentTok{\# Subset data based on cluster}
\NormalTok{        tse\_subset1 }\OtherTok{\textless{}{-}}\NormalTok{ tse1[clusters1[, i], ]}
\NormalTok{        tse\_subset2 }\OtherTok{\textless{}{-}}\NormalTok{ tse2[clusters2[, i], ]}
        \CommentTok{\# Get assay}
\NormalTok{        assay1 }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse\_subset1, assay.type1)}
\NormalTok{        assay2 }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse\_subset2, assay.type2)}
        \CommentTok{\# Calculate sum, mean, median, and mean variance}
\NormalTok{        sum1 }\OtherTok{\textless{}{-}} \FunctionTok{colSums2}\NormalTok{(assay1, }\AttributeTok{na.rm =}\NormalTok{ T)}
\NormalTok{        mean1 }\OtherTok{\textless{}{-}} \FunctionTok{colMeans2}\NormalTok{(assay1, }\AttributeTok{na.rm =}\NormalTok{ T)}
\NormalTok{        median1 }\OtherTok{\textless{}{-}} \FunctionTok{colMedians}\NormalTok{(assay1, }\AttributeTok{na.rm =}\NormalTok{ T)}
\NormalTok{        var1 }\OtherTok{\textless{}{-}} \FunctionTok{colVars}\NormalTok{(assay1, }\AttributeTok{na.rm =}\NormalTok{ T)}
        
\NormalTok{        sum2 }\OtherTok{\textless{}{-}} \FunctionTok{colSums2}\NormalTok{(assay2, }\AttributeTok{na.rm =}\NormalTok{ T)}
\NormalTok{        mean2 }\OtherTok{\textless{}{-}} \FunctionTok{colMeans2}\NormalTok{(assay2, }\AttributeTok{na.rm =}\NormalTok{ T)}
\NormalTok{        median2 }\OtherTok{\textless{}{-}} \FunctionTok{colMedians}\NormalTok{(assay2, }\AttributeTok{na.rm =}\NormalTok{ T)}
\NormalTok{        var2 }\OtherTok{\textless{}{-}} \FunctionTok{colVars}\NormalTok{(assay2, }\AttributeTok{na.rm =}\NormalTok{ T)}
        
\NormalTok{        list[[i]] }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{sample =} \FunctionTok{colnames}\NormalTok{(tse1), sum1, sum2, mean1, }
\NormalTok{                                 mean2, median1, median2, var1, var2)}
\NormalTok{    \}}
    \FunctionTok{return}\NormalTok{(list)}
\NormalTok{\}}

\CommentTok{\# Calculate info}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{.sum\_mean\_median\_var}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], mae[[}\DecValTok{2}\NormalTok{]], }\StringTok{"rclr"}\NormalTok{, }\StringTok{"log10"}\NormalTok{, bicluster\_rows, bicluster\_columns)}
\end{Highlighting}
\end{Shaded}

Now we can create a scatter plot. X-axis includes median clr abundance
of microbiome and y-axis median absolute concentration of each
metabolite. Each data point represents a single sample.

From the plots, we can see that there is low negative correlation in
both cluster 1 and 3. This means that when abundance of bacteria
belonging to cluster 1 or 3 is higher, the concentration of
metabolites of cluster 1 or 3 is lower, and vice versa.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pics }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \FunctionTok{seq\_along}\NormalTok{(df)) \{}
\NormalTok{    pics[[i]] }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(df[[i]]) }\SpecialCharTok{+}
        \FunctionTok{geom\_point}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ median1, }\AttributeTok{y =}\NormalTok{ median2)) }\SpecialCharTok{+}
        \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"Cluster "}\NormalTok{, i), }\AttributeTok{x =} \StringTok{"Taxa (rclr median)"}\NormalTok{,}
             \AttributeTok{y =} \StringTok{"Metabolites (abs. median)"}\NormalTok{)}
    \FunctionTok{print}\NormalTok{(pics[[i]])}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\includegraphics[width=0.33\linewidth]{24_clustering_files/figure-latex/biclust_7-1}
\includegraphics[width=0.33\linewidth]{24_clustering_files/figure-latex/biclust_7-2}
\includegraphics[width=0.33\linewidth]{24_clustering_files/figure-latex/biclust_7-3}
\includegraphics[width=0.33\linewidth]{24_clustering_files/figure-latex/biclust_7-4}
\includegraphics[width=0.33\linewidth]{24_clustering_files/figure-latex/biclust_7-5}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pics[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{+}\NormalTok{ pics[[}\DecValTok{2}\NormalTok{]] }\SpecialCharTok{+}\NormalTok{ pics[[}\DecValTok{3}\NormalTok{]]}
\end{Highlighting}
\end{Shaded}

\includegraphics[width=0.33\linewidth]{24_clustering_files/figure-latex/biclust_7-6}

\emph{pheatmap} does not allow boolean values, so they must be converted into factors.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{bicluster\_columns }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\FunctionTok{apply}\NormalTok{(bicluster\_columns, }\DecValTok{2}\NormalTok{, as.factor))}
\NormalTok{bicluster\_rows }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\FunctionTok{apply}\NormalTok{(bicluster\_rows, }\DecValTok{2}\NormalTok{, as.factor))}
\end{Highlighting}
\end{Shaded}

Again, we can plot clusters with heatmap.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Adjust colors for all clusters}
\ControlFlowTok{if}\NormalTok{ (}\FunctionTok{ncol}\NormalTok{(bicluster\_rows) }\SpecialCharTok{\textgreater{}} \FunctionTok{ncol}\NormalTok{(bicluster\_columns)) \{}
\NormalTok{    cluster\_names }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(bicluster\_rows)}
\NormalTok{\} }\ControlFlowTok{else}\NormalTok{ \{}
\NormalTok{    cluster\_names }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(bicluster\_columns)}
\NormalTok{\}}
\NormalTok{annotation\_colors }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{ (name }\ControlFlowTok{in}\NormalTok{ cluster\_names) \{}
\NormalTok{    annotation\_colors[[name]] }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"TRUE"} \OtherTok{=} \StringTok{"red"}\NormalTok{, }\StringTok{"FALSE"} \OtherTok{=} \StringTok{"white"}\NormalTok{)}
\NormalTok{\}}

\CommentTok{\# Create a heatmap}
\FunctionTok{pheatmap}\NormalTok{(corr, }\AttributeTok{cluster\_cols =}\NormalTok{ F, }\AttributeTok{cluster\_rows =}\NormalTok{ F,}
         \AttributeTok{annotation\_col =}\NormalTok{ bicluster\_columns, }\AttributeTok{annotation\_row =}\NormalTok{ bicluster\_rows,}
         \AttributeTok{annotation\_colors =}\NormalTok{ annotation\_colors)}
\end{Highlighting}
\end{Shaded}

\includegraphics{24_clustering_files/figure-latex/biclust_9-1.pdf}

\hypertarget{additional-community-typing}{%
\section{Additional Community Typing}\label{additional-community-typing}}

For more community typing techniques applied to the `SprockettTHData'
data set, see the attached .Rmd file.

Link:

\begin{itemize}
\tightlist
\item
  \href{add-comm-typing.Rmd}{Rmd}
\end{itemize}

\hypertarget{differential-abundance}{%
\chapter{Differential Abundance}\label{differential-abundance}}

This section provides an introduction to Differential Abundance Analysis (DAA),
which is used to identify differences in the abundances of individual taxa (at
any taxonomic level) between two or more groups, such as treatment and control.
Here, we demonstrate its implementation on Tengeler2020, described in chapter
\ref{tengeler-desc}.

The goal of DAA is to identify biomarkers of a certain phenotype or condition,
and gain understanding of a complex system by looking at its isolated components.
For example, the identification of a bacterial taxon that is differentially
abundant between healthy patients and diseased patients can lead to important
insights into the pathophysiology of the disease. In other words, differentially
abundant taxa can be involved in the dynamics of the disease, which in turn helps
understand the system as a whole. Despite its relevance in current research,
the DAA approach has also been subject to debate \citep{Quinn2021}.

\hypertarget{statistical-challenges-of-microbiome-data}{%
\section{Statistical challenges of microbiome data}\label{statistical-challenges-of-microbiome-data}}

Microbiome data display unique properties that are exclusively addressed by DAA
tools developed for microbiome analysis. Specifically, microbiome data are
characterized by high variability, zero-inflation and compositionality.
\textbf{High variability} expresses that abundance of taxa often varies by several
orders of magnitude from sample to sample. \textbf{Zero-inflation} means that
typically more than 70\% of the values are zeros, which could be due to either
physical absence (structural zeros) or insufficient sampling effort (sampling
zeros). \textbf{Compositionality} implies that a change in the absolute abundance of
one taxon will lead to apparent variations in the relative abundances of other
taxa in the same sample. If neglected, such properties may cause significant
bias in the results of DAA. Therefore, several approaches have been developed to
address the unique properties of microbiome data and provide statistically
useful results.

The first approach to target zero-inflated data consists of specialized models,
such as \textbf{over-dispersed count models} and \textbf{zero-inflated mixture models}.
DESeq2, edgeR and corncorb are based on over-dispersed count models, whereas
metagenomeSeq, RAIDA, ZIBB and Omnibus implement zero-inflated mixture models to
address zero-inflation. Typically, these models assume a negative binomial,
beta-binomial or normal/log-normal distribution. Alternatively, \textbf{zero imputation}
also represents a valid approach to deal with zero-inflated data. ALDEx2 and
eBay apply a Bayesian model to impute the zeros when working with proportion
data, accounting for sampling variability and sequencing depth variation. Other
methods, such as MaAsLin2 and ANCOMBC impute the zeros with a pseudo-count strategy.

Regarding the compositionality of microbiome data, several approaches have been
developed to perform \textbf{robust normalization} with methods specifically designed
to reduce the bias found in compositional data. Some examples include trimmed
mean of M-values (TMM) normalization used by edgeR, relative log expression (RLE)
normalization used by DESeq2, cumulative sum scaling (CSS) normalization used
by metagenomeSeq, centered log-ratio transformation (CLR) normalization used
by ALDEx2 and geometric mean of pairwise ratios (GMPR) normalization used by
Omnibus and Wrench normalization \citep{Kumar2018}, which corrects the compositional
bias by an empirical Bayes approach. Other methods to deal with compositional
data entail reference taxa approach used by DACOMP and RAIDA, analyzing the
pattern of pairwise log ratios as done by ANCOM and bias-correction applied by
ANCOMBC.

We recommend to have a look at \citet{Nearing2022}. In this study, multiple DAA methods
were applied to 38 different datasets and their results were compared to one
another. Because each method follows a slightly different approach in terms of
assumptions and normalization techniques, it was shown that results on the same
dataset may differ substantially depending on the method. Recently, \citet{Yang2022}
comprehensively evaluated DAA methods via a semi-parametric framework and 106
real datasets. This study also concluded that different methods can produce
contradictory results, creating the risk of cherry-picking the most favorable
options for one's own hypothesis. Therefore, it is highly recommended to perform
DAA with multiple methods to determine whether the findings can be reproduced by
different approaches. Built on the findings of \citet{Calgaro2020}, the \emph{benchdamic} \citep{Calgaro2022} package could offer a valuable support in this regard. Through a comprehensive evaluation process it serves both practitioners by comparing DA methods from existing literature, and method developers by providing an impartial tool to evaluate their new approaches in comparison to what is already available. For details, check its extensive \href{https://www.bioconductor.org/packages/release/bioc/vignettes/benchdamic/inst/doc/intro.html}{vignette}.

\hypertarget{using-the-tools}{%
\section{Using the tools}\label{using-the-tools}}

In this section we demonstrate the use of four methods that can be
recommended based on recent literature (ANCOM-BC \citep{ancombc2020},
\emph{ALDEx2} \citep{Gloor2016}, \emph{Maaslin2} \citep{Mallick2020}, \emph{LinDA} \citep{Zhou2022}
and \emph{ZicoSeq} \citep{Yang2022}).

The purpose of this section is to show how to perform DAA in R, not
how to correctly do causal inference. Depending on your experimental
setup and your theory, you must determine how to specify any model
exactly. E.g., there might be confounding factors that might drive
(the absence of) differences between the shown groups that we ignore
here for simplicity. Or your dataset is repeated sampling design,
matched-pair design or the general longitudianl design. We will
demonstrate how to include covariates in those models. We picked a
dataset that merely has microbial abundances in a TSE object as well
as a grouping variable in the sample data. We simplify the examples by
only including two of the three groups.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{library}\NormalTok{(tidyverse)}

\CommentTok{\# Import dataset}
\FunctionTok{data}\NormalTok{(}\StringTok{"Tengeler2020"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ Tengeler2020}

\CommentTok{\# Show patient status by cohort}
\FunctionTok{table}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{patient\_status, tse}\SpecialCharTok{$}\NormalTok{cohort) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r|r}
\hline
  & Cohort\_1 & Cohort\_2 & Cohort\_3\\
\hline
ADHD & 4 & 5 & 4\\
\hline
Control & 6 & 5 & 3\\
\hline
\end{tabular}

\hypertarget{preparing-the-data-for-daa}{%
\subsection{Preparing the data for DAA}\label{preparing-the-data-for-daa}}

Before starting the analysis, it is recommended to reduce the size and
complexity of the data to make the results more reproducible. For this
purpose, we agglomerate the features by genus and filter them by a prevalence
threshold of 10\%.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate by genus and subset by prevalence}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{subsetByPrevalentFeatures}\NormalTok{(tse,}
                             \AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{,}
                             \AttributeTok{prevalence =} \DecValTok{10} \SpecialCharTok{/} \DecValTok{100}\NormalTok{)}

\CommentTok{\# Transform count assay to relative abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse,}
                      \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                      \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

While some DAA tools provide optional arguments for prevalence filtering, here
we filtered the tse object directly. This way, we ensure that the input data
remains the same when multiple tools are used.

\hypertarget{aldex2}{%
\subsection{ALDEx2}\label{aldex2}}

In this section, we will show how to perform DAA with ALDEx2, which can be
regarded as the method of choice for its consistency, as it normally identifies features that are also found by complementary methods \citep{Nearing2022}. A more
extensive introduction to its functionality is available in the
\href{https://bioconductor.org/packages/release/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.html}{ALDEx2 vignette}.

ALDEx2 estimates technical variation within each sample per taxon by utilizing
the Dirichlet distribution. It furthermore applies the CLR transformation (or
closely related log-ratio transforms). Depending on the experimental setup, it
will perform a two sample Welch's t-test and Wilcoxon test or a one-way ANOVA
and Kruskal-Wallis test. For more complex study designs, there is a possibility
to utilize the \texttt{glm} functionality within ALDEx2. The Benjamini-Hochberg
procedure is applied by default to correct for multiple testing.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package}
\FunctionTok{library}\NormalTok{(ALDEx2)}

\CommentTok{\# Generate Monte Carlo samples of the Dirichlet distribution for each sample.}
\CommentTok{\# Convert each instance using the centered log{-}ratio transform.}
\CommentTok{\# This is the input for all further analyses.}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{123}\NormalTok{)}
\NormalTok{x }\OtherTok{\textless{}{-}} \FunctionTok{aldex.clr}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse), tse}\SpecialCharTok{$}\NormalTok{patient\_status)     }
\end{Highlighting}
\end{Shaded}

The t-test:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# calculates expected values of the Welch\textquotesingle{}s t{-}test and Wilcoxon rank}
\CommentTok{\# test on the data returned by aldex.clr}
\NormalTok{x\_tt }\OtherTok{\textless{}{-}} \FunctionTok{aldex.ttest}\NormalTok{(x, }\AttributeTok{paired.test =} \ConstantTok{FALSE}\NormalTok{, }\AttributeTok{verbose =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Effect sizes:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Determines the median clr abundance of the feature in all samples and in}
\CommentTok{\# groups, the median difference between the two groups, the median variation}
\CommentTok{\# within each group and the effect size, which is the median of the ratio}
\CommentTok{\# of the between group difference and the larger of the variance within groups}
\NormalTok{x\_effect }\OtherTok{\textless{}{-}} \FunctionTok{aldex.effect}\NormalTok{(x, }\AttributeTok{CI =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{verbose =} \ConstantTok{FALSE}\NormalTok{)}

\CommentTok{\# combine all outputs }
\NormalTok{aldex\_out }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(x\_tt, x\_effect)}
\end{Highlighting}
\end{Shaded}

Now, we can create a so called Bland-Altman or MA plot (left). It shows the
association between the relative abundance and the magnitude of the difference
per sample. Next to that, we can also create a plot that shows the dispersion
on the x-axis instead of log-ratio abundance. Red dots represent genera that are
differentially abundant (\(q \leq 0.1\)) between the 2 groups. Black points are
rare taxa and grey ones are abundant taxa. The dashed line represent an effect
size of 1. \citet{Gloor2016} provides more information on these plots.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{))}

\FunctionTok{aldex.plot}\NormalTok{(aldex\_out,}
           \AttributeTok{type =} \StringTok{"MA"}\NormalTok{,}
           \AttributeTok{test =} \StringTok{"welch"}\NormalTok{,}
           \AttributeTok{xlab =} \StringTok{"Log{-}ratio abundance"}\NormalTok{,}
           \AttributeTok{ylab =} \StringTok{"Difference"}\NormalTok{,}
           \AttributeTok{cutoff =} \FloatTok{0.05}\NormalTok{)}

\FunctionTok{aldex.plot}\NormalTok{(aldex\_out,}
           \AttributeTok{type =} \StringTok{"MW"}\NormalTok{,}
           \AttributeTok{test =} \StringTok{"welch"}\NormalTok{,}
           \AttributeTok{xlab =} \StringTok{"Dispersion"}\NormalTok{,}
           \AttributeTok{ylab =} \StringTok{"Difference"}\NormalTok{,}
           \AttributeTok{cutoff =} \FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{30_differential_abundance_files/figure-latex/plot-aldex2-1.pdf}

The evaluation as differential abundant in above plots is based on the
corrected p-value. According to the ALDEx2 developers, the safest
approach is to identify those features where the 95\% CI of the effect
size does not cross 0. As we can see in below table, this is not the
case for any of the identified genera (see overlap column, which
indicates the proportion of overlap). Also, the authors recommend to
focus on effect sizes and CIs rather than interpreting the p-value. To
keep the comparison simple, we will here use the p-value as decision
criterion. But please be aware that the effect size together with the
CI is a better answer to the question we are typically interested in.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{aldex\_out }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{rownames\_to\_column}\NormalTok{(}\AttributeTok{var =} \StringTok{"Genus"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \CommentTok{\# here we choose the wilcoxon output rather than t{-}test output}
  \FunctionTok{filter}\NormalTok{(wi.eBH }\SpecialCharTok{\textless{}=} \FloatTok{0.05}\NormalTok{)  }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{select}\NormalTok{(Genus, we.eBH, wi.eBH, effect, overlap) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r|r|r}
\hline
Genus & we.eBH & wi.eBH & effect & overlap\\
\hline
Genus:[Ruminococcus]\_gauvreauii\_group & 0.073 & 0.0379 & 0.8151 & 0.1357\\
\hline
\end{tabular}

\hypertarget{ancom-bc}{%
\subsection{ANCOM-BC}\label{ancom-bc}}

The analysis of composition of microbiomes with bias correction
(ANCOM-BC) \citep{ancombc2020} is a recently developed method for differential
abundance testing. It is based on an earlier published approach
\citep{Mandal2015}. The previous version of ANCOM was among the methods
that produced the most consistent results and is probably a
conservative approach \citep{Nearing2022}. However, the new ANCOM-BC
method operates quite differently compared to the former ANCOM method.

As the only method, ANCOM-BC incorporates the so called \emph{sampling
fraction} into the model. The latter term could be empirically
estimated by the ratio of the library size to the microbial
load. According to the authors, ignoring variations
in this sampling fraction would bias DAA results. Furthermore,
this method provides p-values and confidence intervals for each
taxon. It also controls the FDR and it is computationally simple to
implement.

Note that the original method was implemented in the \texttt{ancombc()} function (see
\href{https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html}{extended tutorial}).
The method has since then been updated and new features have been added to enable
multi-group comparisons and repeated measurements among other improvements.
We do not cover the more advanced features of ANCOMBC in this tutorial
as these features are documented in detail in this
\href{https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html}{tutorial}.

We now proceed with a simple example. First, we specify a formula. In this
formula, other covariates could potentially be included to adjust for
confounding. We show this further below. Again, please make sure to check the
\href{https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html}{function documentation}
as well as the linked tutorials to learn about the additional arguments
that we specify.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package}
\FunctionTok{library}\NormalTok{(ANCOMBC)}

\CommentTok{\# Run ANCOM{-}BC at the genus level and only including the prevalent genera}
\NormalTok{ancombc2\_out }\OtherTok{\textless{}{-}} \FunctionTok{ancombc2}\NormalTok{(}\AttributeTok{data =}\NormalTok{ tse,}
                         \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                         \AttributeTok{fix\_formula =} \StringTok{"patient\_status"}\NormalTok{,}
                         \AttributeTok{p\_adj\_method =} \StringTok{"fdr"}\NormalTok{,}
                         \AttributeTok{prv\_cut =} \DecValTok{0}\NormalTok{,}
                         \AttributeTok{group =} \StringTok{"patient\_status"}\NormalTok{,}
                         \AttributeTok{struc\_zero =} \ConstantTok{TRUE}\NormalTok{,}
                         \AttributeTok{neg\_lb =} \ConstantTok{TRUE}\NormalTok{,}
                         \CommentTok{\# multi group comparison is deactivated automatically}
                         \AttributeTok{global =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The object \texttt{out} contains all model output. Again, see the
\href{https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html}{documentation of the
function}
under \textbf{Value} for details. Our question whether taxa are
differentially abundant can be answered by looking at the \texttt{res}
object, which contains dataframes with the coefficients, standard
errors, p-values and q-values. Below we show the first entries of this
dataframe.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# store the FDR adjusted results }
\NormalTok{ancombc2\_out}\SpecialCharTok{$}\NormalTok{res }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{select}\NormalTok{(taxon, lfc\_patient\_statusControl, q\_patient\_statusControl) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(q\_patient\_statusControl }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(q\_patient\_statusControl) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{head}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r}
\hline
taxon & lfc\_patient\_statusControl & q\_patient\_statusControl\\
\hline
Genus:Subdoligranulum & 1.817 & 0.0011\\
\hline
Genus:Ruminococcus\_1 & 2.819 & 0.0011\\
\hline
Genus:[Ruminococcus]\_gauvreauii\_group & 1.442 & 0.0016\\
\hline
Genus:[Clostridium]\_innocuum\_group & 1.388 & 0.0053\\
\hline
Genus:[Eubacterium]\_rectale\_group & 1.259 & 0.0053\\
\hline
Genus:Coprobacter & -1.185 & 0.0053\\
\hline
\end{tabular}

\hypertarget{maaslin2}{%
\subsection{MaAsLin2}\label{maaslin2}}

Let us next illustrate MaAsLin2 \citep{Mallick2020}. This method is based on
generalized linear models and flexible for different study designs
and covariate structures. For details, check their
\href{https://github.com/biobakery/biobakery/wiki/maaslin2}{Biobakery tutorial}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package}
\FunctionTok{library}\NormalTok{(Maaslin2)}

\CommentTok{\# maaslin expects features as columns and samples as rows }
\CommentTok{\# for both the abundance table as well as metadata }

\CommentTok{\# We can specify different GLMs/normalizations/transforms.}
\CommentTok{\# specifying a ref is especially important if you have more than 2 levels}
\NormalTok{maaslin2\_out }\OtherTok{\textless{}{-}} \FunctionTok{Maaslin2}\NormalTok{(}\AttributeTok{input\_data =} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse))),}
                         \AttributeTok{input\_metadata =} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)),}
                         \AttributeTok{output =} \StringTok{"DAA example"}\NormalTok{,}
                         \AttributeTok{transform =} \StringTok{"AST"}\NormalTok{,}
                         \AttributeTok{fixed\_effects =} \StringTok{"patient\_status"}\NormalTok{,}
                         \CommentTok{\# you can also fit MLM by specifying random effects}
                         \CommentTok{\# random\_effects = c(...),}
                         \AttributeTok{reference =} \StringTok{"patient\_status,Control"}\NormalTok{,}
                         \AttributeTok{normalization =} \StringTok{"TSS"}\NormalTok{,}
                         \AttributeTok{standardize =} \ConstantTok{FALSE}\NormalTok{,}
                         \CommentTok{\# filtering was previously performed}
                         \AttributeTok{min\_prevalence =} \DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Which genera are identified as differentially abundant? (leave out ``head'' to see all).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{maaslin2\_out}\SpecialCharTok{$}\NormalTok{results }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(qval }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|l|l|r|r|r|l|r|r|r}
\hline
feature & metadata & value & coef & stderr & pval & name & qval & N & N.not.zero\\
\hline
Genus..Ruminococcus.\_gauvreauii\_group & patient\_status & ADHD & -0.0662 & 0.0131 & 0.0000 & patient\_statusADHD & 0.0017 & 27 & 21\\
\hline
Genus.Faecalibacterium & patient\_status & ADHD & 0.1195 & 0.0363 & 0.0030 & patient\_statusADHD & 0.0391 & 27 & 11\\
\hline
Genus..Clostridium.\_innocuum\_group & patient\_status & ADHD & -0.0678 & 0.0209 & 0.0033 & patient\_statusADHD & 0.0391 & 27 & 25\\
\hline
Order.Bacteroidales & patient\_status & ADHD & 0.0447 & 0.0139 & 0.0035 & patient\_statusADHD & 0.0391 & 27 & 6\\
\hline
Genus.Catabacter & patient\_status & ADHD & 0.0288 & 0.0090 & 0.0036 & patient\_statusADHD & 0.0391 & 27 & 9\\
\hline
\end{tabular}

This will create a folder that is called like in the output specified
above. It contains also figures to visualize difference between
significant taxa.

\hypertarget{linda}{%
\subsection{LinDA}\label{linda}}

Lastly, we cover linear models for differential abundance analysis of
microbiome compositional data (\citet{Zhou2022}). This is very similar to
ANCOMBC with few differences: 1) LinDA corrects for the compositional
bias differently using the mode of all regression coefficients. 2) it
is faster (100x-1000x than ANCOMBC and according to the authors); 3)
it supports hierarchical models. The latest ANCOMBC versions are also
supporting hierarchical models. Nevertheless, LinDA seems a promising
tool that achieves a very good power/fdr trade-off together with
ANCOMBC according to the review. The speed improvements might make it
critical especially for datasets that have higher sample or feature
set sizes.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package}
\FunctionTok{library}\NormalTok{(MicrobiomeStat)}

\CommentTok{\# Run LinDA}
\NormalTok{linda\_out }\OtherTok{\textless{}{-}} \FunctionTok{linda}\NormalTok{(}\AttributeTok{feature.dat =} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse)),}
                   \AttributeTok{meta.dat =} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)),}
                   \AttributeTok{formula =} \StringTok{"\textasciitilde{} patient\_status"}\NormalTok{,}
                   \AttributeTok{alpha =} \FloatTok{0.05}\NormalTok{,}
                   \AttributeTok{prev.filter =} \DecValTok{0}\NormalTok{,}
                   \AttributeTok{mean.abund.filter =} \DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 0  features are filtered!
## The filtered data has  27  samples and  54  features will be tested!
## Pseudo-count approach is used.
## Fit linear models ...
## Completed.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# List genera for which H0 could be rejected:}
\NormalTok{linda\_out}\SpecialCharTok{$}\NormalTok{output}\SpecialCharTok{$}\NormalTok{patient\_statusControl }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(reject) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{select}\NormalTok{(stat, padj) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{rownames\_to\_column}\NormalTok{(}\AttributeTok{var =} \StringTok{"feature"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r}
\hline
feature & stat & padj\\
\hline
Genus:Faecalibacterium & -4.194 & 0.0101\\
\hline
Genus:Erysipelatoclostridium & 3.031 & 0.0474\\
\hline
Genus:[Ruminococcus]\_gauvreauii\_group & 4.108 & 0.0101\\
\hline
Genus:Barnesiella & -3.091 & 0.0474\\
\hline
Order:Bacteroidales & -3.606 & 0.0243\\
\hline
Genus:Ruminococcaceae\_UCG-014 & -2.993 & 0.0474\\
\hline
Genus:Catabacter & -3.387 & 0.0316\\
\hline
\end{tabular}

\hypertarget{zicoseq}{%
\subsection{ZicoSeq}\label{zicoseq}}

Subsequently, we demonstrate DAA with ZicoSeq, a method based on linear models
and permutation. Further details can be found in
\href{https://cran.r-project.org/web/packages/GUniFrac/vignettes/ZicoSeq.html}{this tutorial}.
This approach has been assessed to exhibit high power and a low false discovery
rate, which has the following components:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Winsorization to decrease the influence of outliers;
\item
  Posterior sampling based on a beta mixture prior to address
  sampling variability and zero inflation;
\item
  Reference-based multiple-stage normalization to address
  compositional effects;
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package}
\FunctionTok{library}\NormalTok{(GUniFrac)}

\FunctionTok{set.seed}\NormalTok{(}\DecValTok{123}\NormalTok{)}
\NormalTok{zicoseq\_out }\OtherTok{\textless{}{-}} \FunctionTok{ZicoSeq}\NormalTok{(}\AttributeTok{feature.dat =} \FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse)),}
                       \AttributeTok{meta.dat =} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)),}
                       \AttributeTok{grp.name =} \StringTok{"patient\_status"}\NormalTok{,}
                       \AttributeTok{feature.dat.type =} \StringTok{"count"}\NormalTok{,}
                       \AttributeTok{return.feature.dat =} \ConstantTok{TRUE}\NormalTok{,}
                       \AttributeTok{prev.filter =} \DecValTok{0}\NormalTok{,}
                       \AttributeTok{mean.abund.filter =} \DecValTok{0}\NormalTok{,}
                       \AttributeTok{max.abund.filter =} \DecValTok{0}\NormalTok{,}
                       \AttributeTok{perm.no =} \DecValTok{999}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## For sample size less than 40, posterior sampling will not be used!
## 0  features are filtered!
## The data has  27  samples and  54  features will be tested!
## On average,  1  outlier counts will be replaced for each feature!
## Finding the references ...
## Permutation testing ...
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## Completed!
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zicoseq\_res }\OtherTok{\textless{}{-}} \FunctionTok{cbind.data.frame}\NormalTok{(}\AttributeTok{p.raw =}\NormalTok{ zicoseq\_out}\SpecialCharTok{$}\NormalTok{p.raw,}
                                \AttributeTok{p.adj.fdr =}\NormalTok{ zicoseq\_out}\SpecialCharTok{$}\NormalTok{p.adj.fdr)}

\NormalTok{zicoseq\_res }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(p.adj.fdr }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(p.adj.fdr) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r}
\hline
  & p.raw & p.adj.fdr\\
\hline
Genus:[Ruminococcus]\_gauvreauii\_group & 0.001 & 0.0040\\
\hline
Genus:[Clostridium]\_innocuum\_group & 0.003 & 0.0210\\
\hline
Genus:Faecalibacterium & 0.001 & 0.0230\\
\hline
Order:Bacteroidales & 0.011 & 0.0358\\
\hline
Genus:Catabacter & 0.005 & 0.0366\\
\hline
\end{tabular}

\begin{Shaded}
\begin{Highlighting}[]
\DocumentationTok{\#\# x{-}axis is the effect size: R2 * direction of coefficient}
\FunctionTok{ZicoSeq.plot}\NormalTok{(}\AttributeTok{ZicoSeq.obj =}\NormalTok{ zicoseq\_out,}
             \AttributeTok{pvalue.type =} \StringTok{\textquotesingle{}p.adj.fdr\textquotesingle{}}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{30_differential_abundance_files/figure-latex/plot-zicoseq-1.pdf}

\hypertarget{philr}{%
\subsection{PhILR}\label{philr}}

PhILR is a tree-based method that tests group-wise associations based on
balances. A detailed introduction to this method is available in
\href{https://www.bioconductor.org/packages/devel/bioc/vignettes/philr/inst/doc/philr-intro.html}{this Bioconductor tutorial}.

\hypertarget{comparison-of-methods}{%
\subsection{Comparison of methods}\label{comparison-of-methods}}

Although the methods described above yield unidentical results, they are
expected to agree on a few differentially abundant taxa. To draw more informed
conclusions, it is good practice to compare the outcomes of different methods in terms of found features, their effect sizes and significances, as well as other method-specific aspects. Such comparative approach is outlined in
\protect\hyperlink{compare-daa-methods}{this exercise}.

\hypertarget{daa-with-confounding}{%
\section{DAA with confounding}\label{daa-with-confounding}}

Confounders can be defined as variables that are related to and affect the
apparent dynamics between the response and the main independent variable.
They are common in experimental studies. Generally, they can be classified
into 3 groups:

\begin{itemize}
\item
  Biological confounders, such as age and sex
\item
  Technical confounders produced during sample collection, processing and analysis
\item
  Confounders resulting from experimental models, such as batch effects and sample history
\end{itemize}

Controlling for confounders is an important practice to reach an unbiased
conclusion. To perform causal inference, it is crucial that the method
is able to include confounders in the model. This is not possible with
statistical tests of general use, such as the Wilcoxon test. In contrast,
methods that target DAA, such as those described in this chapter, allow
controlling for confounders. In the following examples, we will perform DAA
with a main independent variable and a few confounders.

\hypertarget{selecting-confounders}{%
\subsection{Selecting confounders}\label{selecting-confounders}}

In addition to patient status, we will now control for two confounders: cohort
and library size. The former is a categorical variable with three factors,
whereas the latter is a discrete numerical variable. Remarkably, most DAA
methods accept these two and several other data types.

For demonstration, library size is treated as a confounder and included
in the formulas of the DAA methods. Although this is a satisfactory approach to
control for uneven sequencing efforts across samples, rarefaction generally represents a better solution \citep{Schloss2023}. With that said, library size can be readily computed and added to the colData.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Compute and store library size in colData}
\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{library\_size }\OtherTok{\textless{}{-}} \FunctionTok{colSums}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{ancom-bc-1}{%
\subsection{ANCOM-BC}\label{ancom-bc-1}}

Here, confounders can be added to the formula along with patient status, the
main outcome variable. This way, the model evaluates whether differentially
abundant taxa are associated with one of the variables when the other two
are kept constant.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# perform the analysis }
\NormalTok{ancombc2\_out }\OtherTok{\textless{}{-}} \FunctionTok{ancombc2}\NormalTok{(tse,}
                         \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                         \AttributeTok{fix\_formula =} \StringTok{"patient\_status + cohort + library\_size"}\NormalTok{,}
                         \AttributeTok{p\_adj\_method =} \StringTok{"fdr"}\NormalTok{,}
                         \AttributeTok{lib\_cut =} \DecValTok{0}\NormalTok{,}
                         \AttributeTok{group =} \StringTok{"patient\_status"}\NormalTok{, }
                         \AttributeTok{struc\_zero =} \ConstantTok{TRUE}\NormalTok{, }
                         \AttributeTok{neg\_lb =} \ConstantTok{TRUE}\NormalTok{,}
                         \AttributeTok{alpha =} \FloatTok{0.05}\NormalTok{,}
                         \CommentTok{\# multi{-}group comparison is deactivated automatically}
                         \AttributeTok{global =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

In the output, each taxon is assigned with several effect sizes (lfc, which
stands for log-fold change) and adjusted p-values (q). For categorical variables such as patient status and cohort, the statistics indicate whether the abundance
of a given taxon is significantly different between the specified group (column
name) and the reference group (the group that does not appear in the column
names), whereas for numerical variables such as library size, they indicate
whether the abundance of a given taxon varies with that variable.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ancombc2\_out}\SpecialCharTok{$}\NormalTok{res }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{select}\NormalTok{(}\FunctionTok{starts\_with}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"taxon"}\NormalTok{, }\StringTok{"lfc"}\NormalTok{, }\StringTok{"q"}\NormalTok{))) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{arrange}\NormalTok{(q\_patient\_statusControl) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{head}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r|r|r|r|r|r|r|r|r}
\hline
taxon & lfc\_(Intercept) & lfc\_patient\_statusControl & lfc\_cohortCohort\_2 & lfc\_cohortCohort\_3 & lfc\_library\_size & q\_(Intercept) & q\_patient\_statusControl & q\_cohortCohort\_2 & q\_cohortCohort\_3 & q\_library\_size\\
\hline
Genus:Hungatella & -0.2343 & -0.7177 & -0.1473 & -0.0996 & 0e+00 & 0 & 0 & 0 & 0 & 0.0000\\
\hline
Genus:Ruminococcaceae\_UCG-013 & -1.0393 & -0.7369 & 0.6178 & -0.0217 & 1e-04 & 0 & 0 & 0 & 0 & 0.0000\\
\hline
Family:Lachnospiraceae & -0.9455 & -0.6247 & 0.5917 & 0.5393 & 0e+00 & 0 & 0 & 0 & 0 & 0.0000\\
\hline
Genus:Alistipes & -0.0252 & -0.4359 & -0.4571 & 0.0100 & 0e+00 & 0 & 0 & 0 & 0 & 0.0035\\
\hline
Genus:Bacteroides & -0.3896 & -1.1519 & 0.0942 & 0.7381 & 0e+00 & 0 & 0 & 0 & 0 & 0.0001\\
\hline
Genus:Escherichia-Shigella & -1.3544 & -0.9021 & 1.3086 & 0.2621 & 1e-04 & 0 & 0 & 0 & 0 & 0.0000\\
\hline
\end{tabular}

\hypertarget{linda-1}{%
\subsection{LinDA}\label{linda-1}}

As in the previous method, confounders can be included in the formula with the
main outcome variable.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{linda\_out }\OtherTok{\textless{}{-}} \FunctionTok{linda}\NormalTok{(}\FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{)),}
                   \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)),}
                   \AttributeTok{formula =} \StringTok{"\textasciitilde{} patient\_status + cohort + library\_size"}\NormalTok{,}
                   \AttributeTok{alpha =} \FloatTok{0.05}\NormalTok{,}
                   \AttributeTok{prev.filter =} \DecValTok{0}\NormalTok{,}
                   \AttributeTok{mean.abund.filter =} \DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 0  features are filtered!
## The filtered data has  27  samples and  54  features will be tested!
## Imputation approach is used.
## Fit linear models ...
## Completed.
\end{verbatim}

The model returns an output for every variable included in the formula. Normally,
only the results on the main outcome variable are relevant and can be retrieved
as shown below. However, the statistics on the confounders can be similarly
obtained by accessing the corresponding items from the output object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Select results for the patient status}
\NormalTok{linda\_res }\OtherTok{\textless{}{-}}\NormalTok{ linda\_out}\SpecialCharTok{$}\NormalTok{output}\SpecialCharTok{$}\NormalTok{patient\_statusControl}

\NormalTok{linda\_res }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(reject) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{select}\NormalTok{(log2FoldChange, stat, padj) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{rownames\_to\_column}\NormalTok{(}\AttributeTok{var =} \StringTok{"feature"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{head}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r|r}
\hline
feature & log2FoldChange & stat & padj\\
\hline
Genus:Faecalibacterium & -5.843 & -4.473 & 0.0051\\
\hline
Genus:Erysipelatoclostridium & 3.827 & 3.048 & 0.0398\\
\hline
Genus:[Ruminococcus]\_gauvreauii\_group & 4.303 & 4.494 & 0.0051\\
\hline
Genus:Barnesiella & -3.802 & -3.067 & 0.0398\\
\hline
Order:Bacteroidales & -4.098 & -3.938 & 0.0126\\
\hline
Genus:Ruminococcaceae\_UCG-014 & -2.980 & -3.592 & 0.0175\\
\hline
\end{tabular}

The output shows effect sizes in terms of log-fold changes and a derived
statistic (stat) as well as the corresponding adjusted p-values for differences
in abundance of each taxon between the control and treated group.

\hypertarget{zicoseq-1}{%
\subsection{ZicoSeq}\label{zicoseq-1}}

For this method, confounders can be added as a list to the \texttt{adj.name} argument.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{123}\NormalTok{)}
\NormalTok{zicoseq\_out }\OtherTok{\textless{}{-}} \FunctionTok{ZicoSeq}\NormalTok{(}\AttributeTok{feature.dat =} \FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse)),}
                       \AttributeTok{meta.dat =} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)),}
                       \AttributeTok{grp.name =} \StringTok{"patient\_status"}\NormalTok{,}
                       \AttributeTok{adj.name =} \FunctionTok{c}\NormalTok{(}\StringTok{"cohort"}\NormalTok{, }\StringTok{"library\_size"}\NormalTok{), }
                       \AttributeTok{feature.dat.type =} \StringTok{"count"}\NormalTok{,}
                       \AttributeTok{return.feature.dat =} \ConstantTok{TRUE}\NormalTok{,}
                       \AttributeTok{prev.filter =} \DecValTok{0}\NormalTok{,}
                       \AttributeTok{mean.abund.filter =} \DecValTok{0}\NormalTok{,}
                       \AttributeTok{max.abund.filter =} \DecValTok{0}\NormalTok{,}
                       \AttributeTok{perm.no =} \DecValTok{999}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## For sample size less than 40, posterior sampling will not be used!
## 0  features are filtered!
## The data has  27  samples and  54  features will be tested!
## On average,  1  outlier counts will be replaced for each feature!
## Finding the references ...
## Permutation testing ...
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## ...................................................................................................
## Completed!
\end{verbatim}

The output shows the raw and adjusted p-values for clinical status.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{zicoseq\_res }\OtherTok{\textless{}{-}} \FunctionTok{cbind.data.frame}\NormalTok{(}\AttributeTok{p.raw =}\NormalTok{ zicoseq\_out}\SpecialCharTok{$}\NormalTok{p.raw,}
                                \AttributeTok{p.adj.fdr =}\NormalTok{ zicoseq\_out}\SpecialCharTok{$}\NormalTok{p.adj.fdr)}

\NormalTok{zicoseq\_res }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(p.adj.fdr }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{head}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{l|r|r}
\hline
  & p.raw & p.adj.fdr\\
\hline
Genus:Faecalibacterium & 0.002 & 0.0167\\
\hline
Genus:[Clostridium]\_innocuum\_group & 0.003 & 0.0206\\
\hline
Genus:[Ruminococcus]\_gauvreauii\_group & 0.001 & 0.0005\\
\hline
Order:Bacteroidales & 0.004 & 0.0167\\
\hline
Genus:Ruminococcus\_2 & 0.003 & 0.0264\\
\hline
Genus:Ruminococcaceae\_UCG-014 & 0.004 & 0.0316\\
\hline
\end{tabular}

\hypertarget{additional-resources}{%
\section{Additional resources}\label{additional-resources}}

DAA can be performed by several means. Although most of them provide similar
functionality, some may be more suitable than others given a certain study
design or data type. Commonly used DAA tools include:

\begin{itemize}
\tightlist
\item
  ALDEx2 \citep{Gloor2016}
\item
  ANCOM \citep{Mandal2015}
\item
  ANCOMBC \citep{ancombc2020}
\item
  corncob \citep{Martin2021}
\item
  DACOMP \citep{Brill2019}
\item
  DESeq2 \citep{Love2014}
\item
  eBay \citep{Liu2020}
\item
  edgeR \citep{Chen2016}
\item
  fastANCOM \citep{fastANCOM2022}
\item
  LDM \citep{Hu2020}
\item
  lefser \citep{Khlebrodova2021}
\item
  limma \citep{Ritchie2015}
\item
  LinDA \citep{Zhou2022}
\item
  MaAsLin2 \citep{Mallick2020}
\item
  metagenomeSeq \citep{Paulson2017}
\item
  Omnibus \citep{Omnibus2018}
\item
  RAIDA \citep{Sohn2015}
\item
  t-test
\item
  Wilcoxon test
\item
  ZicoSeq \citep{Yang2022}
\item
  ZINQ \citep{Ling2021}
\end{itemize}

\hypertarget{network-learning}{%
\chapter{Network learning and analysis}\label{network-learning}}

Learning and analyzing microbial association networks is another common exploratory data analysis approach aimed at understanding the complex interplay of microbial communities in their natural habitat. Microbial networks consist of nodes, representing microbial species or taxa, and edges, expressing their association. The mathematical representation of a network is the adjacency matrix, which has a non-zero entry for each edge in the network.

A typical workflow for estimating a microbial network involves several steps, including data preprocessing, estimating microbial associations, and transforming them into edge weights. The resulting network can be analyzed using local network properties, such as centrality measures, or global measures that describe the overall structure of the network. Network plots provide further insight into the microbial community structure and allow for exploratory analysis of microbial relationships.

In this chapter, we go through the complete workflow of constructing and analyzing a single microbial network, step by step. A potential next analysis task is to compare networks between groups, such as patients and controls, different environmental conditions, or different time points. How to compare networks between two groups is explained in chapter \ref{network-comparison}

\hypertarget{network-learning-1}{%
\section{Network learning}\label{network-learning-1}}

\hypertarget{network-learning-workflow}{%
\subsection{Typical workflow}\label{network-learning-workflow}}

Figure \ref{fig:network-workflow} shows the workflow for learning/constructing a microbial association network as proposed by \citet{peschel2021netcomi}. The respective steps are explained below.

\begin{figure}
\includegraphics[width=1\linewidth]{general/figures/FigureNetworkLearning} \caption{The typical input is a $p$ x $n$ dimensional count matrix coming from a sequencing process, where $n$ is the number of samples and $p$ the number of features / ASVs / OTUs. Steps 1 through 6 are explained below. Each matrix resulting from steps 4, 5, and 6 plays a specific role in the final network: The adjacency matrix is used for edge colors, dissimilarity for layout, and similarity for edge weights. In weighted networks, the similarity matrix equals the adjacency matrix.}\label{fig:network-workflow}
\end{figure}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  \textbf{Zero replacement:} Since the following steps usually require non-zero entries in the read count matrix, zero counts must be replaced. A simple solution is to add a pseudo count to the data. Other possible approaches are implemented in the R package \href{https://cran.r-project.org/web/packages/zCompositions/index.html}{zCompositions}
\item
  \textbf{Normalization:} To avoid compositional effects, the data are normalized using a compositionality aware transformation. A common approach is the centered log-ratio (clr) transformation, which moves the data from a \(p\)-dimensional simplex to Euclidean space so that standard statistical analysis methods are valid. A variance stabilizing transformation (vst) is also a suitable approach for normalizing microbial count data \citep{badri2020shrinkage}.
\item
  \textbf{Association estimation:} This is the crucial step in network learning to obtain statistical relations between the taxa. Common association measures include correlation, conditional dependence (which we will equate to partial correlation), and proportionality. Further information on these three types of association and their application can be found in Section \ref{more-about-association}. The following list gives a selection of compositionality aware approaches:

  \begin{itemize}
  \tightlist
  \item
    \textbf{Compositionality aware correlation estimation methods:}

    \begin{itemize}
    \tightlist
    \item
      Pearson's correlation coefficient (+ normalization)
    \item
      Spearman's rank correlation coefficient (+ normalization)
    \item
      Covariance shrinkage (\href{https://strimmerlab.github.io/software/corpcor/}{corpcor} package) (+ normalization)
    \item
      SparCC (implemented in \href{https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html}{SpiecEasi}); applied in Section \ref{sparcc-correlation}
    \item
      CCREPE (\href{https://bioconductor.org/packages/release/bioc/html/ccrepe.html}{ccrepe} package)
    \item
      CCLasso (\href{https://github.com/huayingfang/CCLasso}{R code on GitHub})
    \end{itemize}
  \item
    \textbf{Compositionality aware measures of conditional dependence / partial correlation:}

    \begin{itemize}
    \tightlist
    \item
      \href{https://github.com/zdk123/SpiecEasi}{SpiecEasi} with Meinshausen and Bühlmann (MB) neighborhood selection; applied in Section \ref{spieceasi-mb}
    \item
      \href{https://github.com/zdk123/SpiecEasi}{SpiecEasi} with the graphical lasso (glasso)
    \item
      gCoda (\href{https://github.com/huayingfang/gCoda}{R code on GitHub})
    \item
      \href{https://github.com/GraceYoon/SPRING}{SPRING}; applied in Section \ref{netcomi-spring}
    \end{itemize}
  \item
    \textbf{Proportionality measures (proportionality aware by definition):}

    \begin{itemize}
    \tightlist
    \item
      \href{https://github.com/tpq/propr}{propr}
    \item
      \href{https://github.com/MichelleBadri/NormCorr-manuscript/blob/master/code/helpers/norm_functions.R}{Shrinkage proportionality estimator}; applied in Section \ref{shrinkage-prop}
    \end{itemize}
  \end{itemize}
\item
  \textbf{Sparsification:} Transforming the estimated associations directly into adjacencies would lead to a dense network where all nodes are connected and only weighted network measures are meaningful. Therefore, the association matrix is usually sparsified to select edges of interest. A common sparsification approach for correlations is thresholding, where correlations with a magnitude below the threshold are set to zero. Another possibility is a statistical test (Student's t-test or permutation test) with the null hypothesis that the correlation is equal to zero. SpiecEasi uses the StARS stability selection approach \citep{liu2010stability} to decide on an appropriate sparsification level of the inferred conditional dependence graph.
\item
  \textbf{Transformation into dissimilarity:} A common next step is to simply use the absolute values of the sparsified associations as edge weights. In this way, correlations of high magnitude (both positive and negative) will have a high edge weight. From a biological point of view, it would also make sense to assign a low edge weight to taxa that are strongly negatively associated, which would correspond to a high dissimilarity value. Here we follow \citet{vanDongen2012metric} to directly transform the sparse associations \(r_{ij}^*\) into dissimilarities, which can later be used for shortest path network measures. Depending on the desired handling of negative associations, one of the two proposed transformations should be chosen:

  5a: \textbf{``signed'':} \(d_{ij} = \sqrt{0.5(1-r^*_{ij})}\), where strongly negatively associated taxa have the largest distance and are placed further apart in the network.

  5b: \textbf{``unsigned'':} \(d_{ij} = \sqrt{1-{r_{ij}^*}^2}\), resulting in a small distance between strongly associated taxa (regardless of the sign).
\item
  \textbf{Transformation into similarity / edge weight:} Finally, the dissimilarities are transformed into similarities by \(s_{ij} = 1 - d_{ij}\), which are used as edge weights. Thus, the similarity matrix is equal to the adjacency matrix in a weighted network.
\end{enumerate}

The main association measure used in this chapter is the \href{https://github.com/GraceYoon/SPRING}{SPRING} (``Semi-Parametric Rank-based approach for INference in Graphical model'') method proposed by \citet{yoon2019microbial}. \texttt{SPRING} learns conditional dependency graphs for compositional data and follows the neighborhood selection method introduced by \citet{meinshausen2006high} (``MB''). We will show how to apply the method directly, as well as how to use it in conjunction with the R package \href{https://github.com/stefpeschel/NetCoMi}{NetCoMi}, which is specifically designed for the construction and analysis of networks for microbiome data.

See Section \ref{more-about-association} for a comparison of all three association types (correlation, partial correlation, and proportionality) with more information on each measure and applications.

We demonstrate the workflow using the the \textbf{PeerJ data set} \citep{potbhare2022skin}. It contains skin microbial profiles of 58 subjects.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{data}\NormalTok{(}\StringTok{"peerj13075"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse0 }\OtherTok{\textless{}{-}}\NormalTok{ peerj13075}
\FunctionTok{dim}\NormalTok{(tse0)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 674  58
\end{verbatim}

\hypertarget{install-packages}{%
\subsection{Install packages}\label{install-packages}}

Three packages used in this chapter are available on GitHub only: \texttt{SpiecEasi}, \texttt{SPRING}, and \texttt{NetCoMi}. We recommend that you install these packages before proceeding.

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{if}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{require}\NormalTok{(SpiecEasi))\{}
\NormalTok{  devtools}\SpecialCharTok{::}\FunctionTok{install\_github}\NormalTok{(}\StringTok{"zdk123/SpiecEasi"}\NormalTok{)}
\NormalTok{\}}

\ControlFlowTok{if}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{require}\NormalTok{(SPRING))\{}
\NormalTok{  devtools}\SpecialCharTok{::}\FunctionTok{install\_github}\NormalTok{(}\StringTok{"GraceYoon/SPRING"}\NormalTok{)}
\NormalTok{\}}

\ControlFlowTok{if}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{require}\NormalTok{(NetCoMi))\{}
\NormalTok{  devtools}\SpecialCharTok{::}\FunctionTok{install\_github}\NormalTok{(}\StringTok{"stefpeschel/NetCoMi"}\NormalTok{, }\AttributeTok{force =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{ref =} \StringTok{"TSE"}\NormalTok{,}
                           \AttributeTok{dependencies =} \FunctionTok{c}\NormalTok{(}\StringTok{"Depends"}\NormalTok{, }\StringTok{"Imports"}\NormalTok{, }\StringTok{"LinkingTo"}\NormalTok{),}
                           \AttributeTok{repos =} \FunctionTok{c}\NormalTok{(}\StringTok{"https://cloud.r{-}project.org/"}\NormalTok{,}
\NormalTok{                                     BiocManager}\SpecialCharTok{::}\FunctionTok{repositories}\NormalTok{()))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\hypertarget{data-preparation}{%
\subsection{Data preparation}\label{data-preparation}}

Before applying the network learning methods, we perform some data preparation steps:

\begin{itemize}
\tightlist
\item
  Aggregation to genus level
\item
  Add relative abundance assay
\item
  Prevalence filtering (keep genera with prevalence \textgreater{} 20\%)
\item
  Add assay with log10 transformed abundances
\item
  Add assay with clr transformed abundances
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate to genus level}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{agglomerateByRank}\NormalTok{(tse0, }\AttributeTok{rank =} \StringTok{"genus"}\NormalTok{) }

\CommentTok{\# Add relative abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }
                      \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }
                      \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{,}
                      \AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{) }

\CommentTok{\# Filter by prevalence}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{subsetByPrevalentFeatures}\NormalTok{(tse,}
                                 \AttributeTok{prevalence =} \FloatTok{0.2}\NormalTok{,}
                                 \AttributeTok{detection =} \DecValTok{0}\NormalTok{,}
                                 \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Add log10{-}transformed abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"log10"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\CommentTok{\# Add clr{-}transformed abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\FunctionTok{dim}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 147  58
\end{verbatim}

\hypertarget{spring-network}{%
\subsection{SPRING network}\label{spring-network}}

As explained in Section \ref{network-learning-workflow}, we use \href{https://github.com/GraceYoon/SPRING}{SPRING} (``Semi-Parametric Rank-based approach for INference in Graphical model'') as association measure. We first use the \texttt{SPRING} function directly to construct a conditional dependency graph.

Neither zero replacement nor normalization (steps 1 and 2 in our workflow) are required because SPRING uses a modified clr (mclr) transformation that can handle zero counts, and the correlation estimation method itself can also deal with zeros in the data. mclr is similar to the clr transformation except that mclr considers only the non-zero values. More precisely, the geometric mean is derived from positive values only, and zero counts remain zero after the transformation. This approach is similar to the ``robust clr'' (rclr) transformation included in the \texttt{vegan} package, except that mclr applies a positive shift to all non-zero values to make them strictly positive. See \citep{yoon2019microbial} for details.

The \texttt{Rmethod} argument is set to ``approx'' to estimate the correlations using a hybrid multi-linear interpolation approach proposed by \citet{yoon2021fast}. This method considerably reduces the runtime while controlling the approximation error.

\texttt{SPRING} uses the StARS (``Stability Approach to Regularization Selection'') method \citep{liu2010stability} to obtain a sparse association matrix. Thus, also step 4 of our workflow is already included. We set the StARS threshold to 0.05 to get a sparser graph.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(SPRING)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{13075}\NormalTok{)}
\NormalTok{spring\_est }\OtherTok{\textless{}{-}} \FunctionTok{SPRING}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{)), }
                     \AttributeTok{Rmethod =} \StringTok{"approx"}\NormalTok{, }
                     \AttributeTok{thresh =} \FloatTok{0.05}\NormalTok{,}
                     \AttributeTok{lambdaseq =} \StringTok{"data{-}specific"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get index of the optimal lambda selected by StARS}
\NormalTok{opt.K }\OtherTok{\textless{}{-}}\NormalTok{ spring\_est}\SpecialCharTok{$}\NormalTok{output}\SpecialCharTok{$}\NormalTok{stars}\SpecialCharTok{$}\NormalTok{opt.index}

\CommentTok{\# Store partial correlation matrix belonging to the optimal lambda as matrix}
\NormalTok{spring\_cor }\OtherTok{\textless{}{-}}\NormalTok{ SpiecEasi}\SpecialCharTok{::}\FunctionTok{symBeta}\NormalTok{(}\FunctionTok{as.matrix}\NormalTok{(spring\_est}\SpecialCharTok{$}\NormalTok{output}\SpecialCharTok{$}\NormalTok{est}\SpecialCharTok{$}\NormalTok{beta[[opt.K]]))}
\NormalTok{spring\_cor }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(spring\_cor)}
\FunctionTok{rownames}\NormalTok{(spring\_cor) }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(spring\_cor) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(tse)}
\FunctionTok{diag}\NormalTok{(spring\_cor) }\OtherTok{\textless{}{-}} \DecValTok{1}
\end{Highlighting}
\end{Shaded}

As explained in Section \ref{network-learning-workflow}, the estimated associations are sparsified, transformed into dissimilarities, and finally transformed into similarities, which are the adjacency values. We write a function for these steps, which will be reused later.

Since \texttt{SPRING} already includes a sparsification approach, the \texttt{thresh} argument is not needed here, but will be needed in Section \ref{more-about-association} for other association measures.

To be consistent with the workflow, we provide two dissimilarity transformations: ``signed'' and ``unsigned'' (see Section \ref{network-learning-workflow} for an explanation). These transformations were introduced by \citet{vanDongen2012metric}. We use the ``signed'' transformation in our examples so that strongly negatively associated genera have low edge weights.

The output of the function is an \texttt{igraph} object, which can be plotted and analyzed using functions from the \texttt{igraph} package.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Arguments:}
\CommentTok{\# {-} assoMat: association matrix}
\CommentTok{\# {-} threshold: associations below the threshold are set to zero}
\CommentTok{\# {-} dissTrans: dissimilarity transformation ("signed" or "unsigned")}

\NormalTok{transform\_asso }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(assoMat, }\AttributeTok{thresh =} \ConstantTok{NULL}\NormalTok{, }\AttributeTok{dissTrans =} \StringTok{"signed"}\NormalTok{) \{}
  \CommentTok{\# Sparsification}
  \ControlFlowTok{if}\NormalTok{ (}\SpecialCharTok{!}\FunctionTok{is.null}\NormalTok{(thresh)) \{}
\NormalTok{    assoMat[}\FunctionTok{abs}\NormalTok{(assoMat) }\SpecialCharTok{\textless{}}\NormalTok{ thresh] }\OtherTok{\textless{}{-}} \DecValTok{0}
\NormalTok{  \}}
  
  \CommentTok{\# Compute dissimilarity matrix}
  \ControlFlowTok{if}\NormalTok{ (dissTrans }\SpecialCharTok{==} \StringTok{"signed"}\NormalTok{) \{}
\NormalTok{    dissMat }\OtherTok{\textless{}{-}} \FunctionTok{sqrt}\NormalTok{(}\FloatTok{0.5} \SpecialCharTok{*}\NormalTok{ (}\DecValTok{1} \SpecialCharTok{{-}}\NormalTok{ assoMat))}
\NormalTok{  \} }\ControlFlowTok{else}\NormalTok{ \{}
\NormalTok{    dissMat }\OtherTok{\textless{}{-}} \FunctionTok{sqrt}\NormalTok{(}\DecValTok{1} \SpecialCharTok{{-}}\NormalTok{ assoMat}\SpecialCharTok{\^{}}\DecValTok{2}\NormalTok{)}
\NormalTok{  \}}
  
  \CommentTok{\# Dissimilarity between nodes with zero correlation is set to 1}
  \CommentTok{\# (these nodes are unconnected and thus should have maximum dissimilarity)}
\NormalTok{  dissMat[assoMat }\SpecialCharTok{==} \DecValTok{0}\NormalTok{] }\OtherTok{\textless{}{-}} \DecValTok{1}
  
  \CommentTok{\# Compute similarity matrix}
\NormalTok{  simMat }\OtherTok{\textless{}{-}} \DecValTok{1} \SpecialCharTok{{-}}\NormalTok{ dissMat}
  
  \CommentTok{\# Turn into igraph object}
\NormalTok{  graphObj }\OtherTok{\textless{}{-}}\NormalTok{ SpiecEasi}\SpecialCharTok{::}\FunctionTok{adj2igraph}\NormalTok{(simMat)}
  
  \FunctionTok{return}\NormalTok{(}\FunctionTok{list}\NormalTok{(}\AttributeTok{graph =}\NormalTok{ graphObj, }\AttributeTok{adja =}\NormalTok{ simMat, }\AttributeTok{asso =}\NormalTok{ assoMat, }\AttributeTok{diss =}\NormalTok{ dissMat))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create graph object}
\NormalTok{spring\_graph }\OtherTok{\textless{}{-}} \FunctionTok{transform\_asso}\NormalTok{(spring\_cor)}\SpecialCharTok{$}\NormalTok{graph}
\end{Highlighting}
\end{Shaded}

\hypertarget{netcomi-spring}{%
\subsection{NetCoMi network}\label{netcomi-spring}}

The \href{https://github.com/stefpeschel/NetCoMi}{NetCoMi} \citep{peschel2021netcomi} package is specifically designed to construct, analyze, and compare networks for microbiome data and implements the complete workflow described in Section \ref{network-learning-workflow}. Instead of using several functions for each of the steps, \texttt{NetCoMi} provides a single function for network construction (\texttt{netConstruct()}), so the package streamlines the workflow considerably. The user can choose from a variety of methods for data preprocessing, association estimation, sparsification, and transformation. The returned \texttt{microNet} object can then be passed to \texttt{netAnalyze()} (the network analysis function) so that all necessary information is available for the network analysis workflow.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(NetCoMi)}
\end{Highlighting}
\end{Shaded}

We again use \texttt{SPRING} as one of the association measures available in NetCoMi to construct a conditional dependency graph.

To demonstrate how taxa are filtered with \texttt{netConstruct()}, we will use the unfiltered \texttt{tse} object this time. The filtering is the same as before: Taxa occurring in less than 20\% of the samples are removed.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{netcomi\_net }\OtherTok{\textless{}{-}} \FunctionTok{netConstruct}\NormalTok{(tse,}
                            \AttributeTok{taxRank =} \StringTok{"genus"}\NormalTok{,}
                            \AttributeTok{filtTax =} \StringTok{"numbSamp"}\NormalTok{,}
                            \AttributeTok{filtTaxPar =} \FunctionTok{list}\NormalTok{(}\AttributeTok{numbSamp =} \FloatTok{0.2}\NormalTok{),}
                            \AttributeTok{measure =} \StringTok{"spring"}\NormalTok{,}
                            \AttributeTok{measurePar =} \FunctionTok{list}\NormalTok{(}\AttributeTok{thresh =} \FloatTok{0.05}\NormalTok{,}
                                              \AttributeTok{Rmethod =} \StringTok{"approx"}\NormalTok{),}
                            \AttributeTok{sparsMethod =} \StringTok{"none"}\NormalTok{, }
                            \AttributeTok{dissFunc =} \StringTok{"signed"}\NormalTok{,}
                            \AttributeTok{seed =} \DecValTok{13075}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\texttt{netConstruct()} returns an object of the class \texttt{microNet}, which contains all matrices generated during network construction.

The object also contains an edge list, giving each edge's estimated association, dissimilarity, and adjacency. Let's take a quick look at the edges with the highest and lowest edge weights:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{edgelist }\OtherTok{\textless{}{-}}\NormalTok{ netcomi\_net}\SpecialCharTok{$}\NormalTok{edgelist1[}\FunctionTok{order}\NormalTok{(netcomi\_net}\SpecialCharTok{$}\NormalTok{edgelist1}\SpecialCharTok{$}\NormalTok{adja, }
                                        \AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{), ]}
\FunctionTok{head}\NormalTok{(edgelist)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                 v1                    v2   asso   diss   adja
## 73     Citrobacter           Escherichia 0.3682 0.5621 0.4379
## 63    Buttiauxella              Serratia 0.2426 0.6154 0.3846
## 69   Chitinivibrio Pseudogracilibacillus 0.2203 0.6244 0.3756
## 19        Algicola           Siccibacter 0.2193 0.6248 0.3752
## 111     Haliangium          Marinobacter 0.2148 0.6266 0.3734
## 143 Planomicrobium         Virgibacillus 0.2006 0.6322 0.3678
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tail}\NormalTok{(edgelist)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                v1             v2       asso   diss   adja
## 132 Mycobacterium Salinibacillus  0.0017483 0.7065 0.2935
## 24      Amphritea    Providencia  0.0014085 0.7066 0.2934
## 102       Erwinia    Siccibacter  0.0013114 0.7066 0.2934
## 116     Holophaga   Methylarcula  0.0007828 0.7068 0.2932
## 17       Algicola     Lysobacter  0.0005191 0.7069 0.2931
## 95   Enterobacter     Janibacter -0.0013921 0.7076 0.2924
\end{verbatim}

As before, the adjacency matrix is converted into an \texttt{igraph} object. Further steps like sparsification and transformation are not necessary because they are done internally by \texttt{netConstruct()}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{netcomi\_graph }\OtherTok{\textless{}{-}}\NormalTok{ SpiecEasi}\SpecialCharTok{::}\FunctionTok{adj2igraph}\NormalTok{(}\FunctionTok{abs}\NormalTok{(netcomi\_net}\SpecialCharTok{$}\NormalTok{adjaMat1))}
\end{Highlighting}
\end{Shaded}

\hypertarget{network-analysis}{%
\section{Network analysis with igraph}\label{network-analysis}}

The computed network is now analyzed using appropriate methods. We will first use the \texttt{igraph} package to analyze the \texttt{SPRING} network. NetCoMi's \texttt{netAnalyze()} function will be used later to analyze the constructed \texttt{microNet} object.

\hypertarget{network-plot}{%
\subsection{Network plot}\label{network-plot}}

To get an overview of the network structure, a first common analysis method is to plot the network. We here use the \href{https://r.igraph.org/}{igraph} package, which is a state-of-the-art package for network analysis and visualization. Other packages that could be used for network plotting are the \href{https://rdrr.io/cran/qgraph/}{qgraph} package or the \href{https://briatte.github.io/ggnet/}{ggnet2} package. Since we will use igraph for network analysis later on, we are using its plotting function here as well.

We use the Fruchterman-Reingold layout (a force-directed layout) for node placement. By placing strongly connected nodes close together and those with low edge weight far apart, this layout results in an easy-to-read network plot.

The node sizes are proportional to a taxon's log10-transformed abundance, which we previously added to the \texttt{tse} object, averaged across all samples. The values are rescaled to be visually distinguishable.

Since we created two graph objects, one with SPRING and one with NetCoMi, we plot them side by side. The two plots should be identical.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(igraph)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Node sizes}
\NormalTok{vsize }\OtherTok{\textless{}{-}}\NormalTok{ (}\FunctionTok{colMeans}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"log10"}\NormalTok{))) }\SpecialCharTok{+} \DecValTok{1}\NormalTok{) }\SpecialCharTok{*} \DecValTok{3}

\CommentTok{\# Fruchterman{-}Reingold layout from igraph package}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{13075}\NormalTok{)}
\NormalTok{lay\_fr }\OtherTok{\textless{}{-}} \FunctionTok{layout\_with\_fr}\NormalTok{(spring\_graph)}

\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(spring\_graph, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize, }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \StringTok{"SPRING network"}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(netcomi\_graph, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize, }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \StringTok{"NetCoMi network}\SpecialCharTok{\textbackslash{}n}\StringTok{(with SPRING associations)"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/plots_spring_netcomi-1.pdf}

\hypertarget{centrality-measures}{%
\subsection{Centrality measures}\label{centrality-measures}}

Centrality measures express the importance of nodes within the network. Common measures are the degree, betweenness, closeness, and eigenvector centrality. The \texttt{igraph} package provides functions to compute these measures. We wrap a function around the code to reuse it later.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{get\_centr }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(graph\_obj) \{}
  \CommentTok{\# We access igraph directly with "::" because there are more packages loaded in }
  \CommentTok{\# this chapter that contain a degree() function.}
\NormalTok{  df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{Degree =}\NormalTok{ igraph}\SpecialCharTok{::}\FunctionTok{degree}\NormalTok{(graph\_obj))}
\NormalTok{  df}\SpecialCharTok{$}\NormalTok{Betweenness }\OtherTok{\textless{}{-}} \FunctionTok{betweenness}\NormalTok{(graph\_obj)}
\NormalTok{  df}\SpecialCharTok{$}\NormalTok{Closeness }\OtherTok{\textless{}{-}} \FunctionTok{closeness}\NormalTok{(graph\_obj, }\AttributeTok{normalized =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{  df}\SpecialCharTok{$}\NormalTok{Eigenvector }\OtherTok{\textless{}{-}} \FunctionTok{eigen\_centrality}\NormalTok{(graph\_obj)}\SpecialCharTok{$}\NormalTok{vector}
  \FunctionTok{return}\NormalTok{(df)}
\NormalTok{\}}

\NormalTok{centr\_df }\OtherTok{\textless{}{-}} \FunctionTok{get\_centr}\NormalTok{(spring\_graph)}
\FunctionTok{rownames}\NormalTok{(centr\_df) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(spring\_cor)}
\FunctionTok{head}\NormalTok{(centr\_df, }\DecValTok{15}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                  Degree Betweenness Closeness Eigenvector
## Abyssicoccus          0           0       NaN   9.278e-18
## Acidaminococcus       2           0    0.6464   1.862e-01
## Acinetobacter         3           5    0.5544   1.257e-01
## Actinomyces           0           0       NaN   9.278e-18
## Actinoplanes          2          99    0.5104   1.099e-02
## Aerococcus            0           0       NaN   9.278e-18
## Aeromonas             4         351    0.7477   2.023e-01
## Agromyces             6         435    0.7733   5.293e-01
## Algicola              4         264    0.7206   1.500e-01
## Alicyclobacillus      0           0       NaN   9.278e-18
## Alteribacillus        0           0       NaN   9.278e-18
## Ammoniibacillus       1           0    0.5064   5.419e-02
## Amphritea             5         382    0.6256   1.335e-02
## Amycolatopsis         1           0    0.6117   1.338e-01
## Anaerococcus          2         392    0.4064   1.288e-03
\end{verbatim}

The closeness centrality is ``NaN'' for some genera. These are unconnected nodes, as can be seen by the zero degree and betweenness centrality.

\hypertarget{scale-node-sizes-by-centrality}{%
\subsection{Scale node sizes by centrality}\label{scale-node-sizes-by-centrality}}

Centrality measures can be visualized in the network plot by scaling the node sizes according to one of these measures. We plot the \texttt{Spring} graph using the same layout as before and with the node sizes scaled according to all four centrality measures.

Of the four centrality measures, only the degree has a range suitable to be used as node size. The other centrality measures must be rescaled because their range is either too small or too large. The following scaling is a suggestion that works for this example. The values might be adapted for other data sets.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{get\_vsizes }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(centr\_df) \{}
\NormalTok{  df }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(centr\_df)}
\NormalTok{  df[, }\StringTok{"Betweenness"}\NormalTok{] }\OtherTok{\textless{}{-}} \FunctionTok{log}\NormalTok{(df[, }\StringTok{"Betweenness"}\NormalTok{])}
\NormalTok{  df[, }\StringTok{"Closeness"}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ df[, }\StringTok{"Closeness"}\NormalTok{] }\SpecialCharTok{*} \DecValTok{10}
\NormalTok{  df[, }\StringTok{"Eigenvector"}\NormalTok{] }\OtherTok{\textless{}{-}}\NormalTok{ df[, }\StringTok{"Eigenvector"}\NormalTok{] }\SpecialCharTok{*} \DecValTok{10}
\NormalTok{  df[}\FunctionTok{is.infinite}\NormalTok{(df) }\SpecialCharTok{|} \FunctionTok{is.na}\NormalTok{(df)] }\OtherTok{\textless{}{-}} \DecValTok{0}
  \FunctionTok{return}\NormalTok{(df)}
\NormalTok{\}}

\NormalTok{vsize\_df }\OtherTok{\textless{}{-}} \FunctionTok{get\_vsizes}\NormalTok{(centr\_df )}
\FunctionTok{head}\NormalTok{(vsize\_df)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                 Degree Betweenness Closeness Eigenvector
## Abyssicoccus         0       0.000     0.000   9.278e-17
## Acidaminococcus      2       0.000     6.464   1.862e+00
## Acinetobacter        3       1.609     5.544   1.257e+00
## Actinomyces          0       0.000     0.000   9.278e-17
## Actinoplanes         2       4.595     5.104   1.099e-01
## Aerococcus           0       0.000     0.000   9.278e-17
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \FunctionTok{seq\_along}\NormalTok{(centr\_df)) \{}
  \FunctionTok{plot}\NormalTok{(spring\_graph, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize\_df[, i], }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \FunctionTok{colnames}\NormalTok{(centr\_df)[i])}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/plots_centrality-1.pdf}

We observe that the two-node component at the bottom has a much higher closeness centrality than the nodes belonging to the main component of the network. Obviously, closeness centrality as commonly defined is misleading when the network consists of disconnected components. Nodes belonging to smaller components are seen as closer to others than in larger components. To overcome this problem, centrality values, and especially closeness centrality, are often calculated only for the largest connected component (LCC), which we will do below.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Extract the LCC}
\NormalTok{dg\_net }\OtherTok{\textless{}{-}}\NormalTok{ igraph}\SpecialCharTok{::}\FunctionTok{decompose.graph}\NormalTok{(spring\_graph)}
\NormalTok{idx\_lcc }\OtherTok{\textless{}{-}} \FunctionTok{which.max}\NormalTok{(}\FunctionTok{unlist}\NormalTok{(}\FunctionTok{lapply}\NormalTok{(dg\_net, }\ControlFlowTok{function}\NormalTok{(x) }\FunctionTok{length}\NormalTok{(igraph}\SpecialCharTok{::}\FunctionTok{V}\NormalTok{(x)))))}
\NormalTok{lcc }\OtherTok{\textless{}{-}}\NormalTok{ dg\_net[[idx\_lcc]]}

\CommentTok{\# Compute centrality values for the LCC}
\NormalTok{centr\_df\_lcc }\OtherTok{\textless{}{-}} \FunctionTok{get\_centr}\NormalTok{(lcc)}

\CommentTok{\# Replace centrality values by those for LCC and set all others to zero}
\NormalTok{lcc\_nodes }\OtherTok{\textless{}{-}} \FunctionTok{as.numeric}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(centr\_df\_lcc))}
\NormalTok{centr\_df[lcc\_nodes, ] }\OtherTok{\textless{}{-}}\NormalTok{ centr\_df\_lcc}
\NormalTok{centr\_df[}\SpecialCharTok{{-}}\NormalTok{lcc\_nodes, ] }\OtherTok{\textless{}{-}} \DecValTok{0}

\CommentTok{\# Node/vertex sizes}
\NormalTok{vsize\_df }\OtherTok{\textless{}{-}} \FunctionTok{get\_vsizes}\NormalTok{(centr\_df)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \FunctionTok{seq\_along}\NormalTok{(centr\_df)) \{}
  \FunctionTok{plot}\NormalTok{(spring\_graph, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize\_df[, i], }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \FunctionTok{colnames}\NormalTok{(centr\_df)[i])}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/plots_centrality_lcc-1.pdf}

Note that \texttt{NetCoMi} follows a different approach to overcome this problem. \texttt{NetCoMi} uses the definition of closeness centrality proposed by \href{https://toreopsahl.com/2010/03/20/closeness-centrality-in-networks-with-disconnected-components/}{Tore Opsahl}, which is well defined even for disconnected networks and assigns higher closeness centrality values to nodes in larger components. This is more intuitive because nodes in a larger component are connected to a larger number of other nodes than in small components.

\hypertarget{degree-distribution}{%
\subsection{Degree distribution}\label{degree-distribution}}

The degree distribution is another popular measure that expresses the probability distribution of degrees over the entire network. It thus provides insight into the overall network structure. We plot the degree distribution for all four association estimation methods to compare the network structure.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Compute degree distribution}
\NormalTok{ddist}\OtherTok{\textless{}{-}}\NormalTok{ igraph}\SpecialCharTok{::}\FunctionTok{degree.distribution}\NormalTok{(spring\_graph)}

\CommentTok{\# Data frame needed for ggplot2}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{Degree =} \FunctionTok{as.factor}\NormalTok{((}\FunctionTok{seq\_along}\NormalTok{(ddist)) }\SpecialCharTok{{-}} \DecValTok{1}\NormalTok{),}
                 \AttributeTok{Fraction =}\NormalTok{ ddist)}

\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ Degree, }\AttributeTok{y =}\NormalTok{ Fraction, }\AttributeTok{group =} \DecValTok{1}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{geom\_line}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/degree_dist_spring-1.pdf}

The network has a large number of singletons and sparsely connected nodes, and only a small number of nodes with a higher degree of 7 or more.

\hypertarget{clustered-heatmaps}{%
\subsection{Clustered heatmaps}\label{clustered-heatmaps}}

Using the \texttt{ComplexHeatmap} package, we plot a heatmap of the association matrix estimated with \texttt{SPRING}. Rows and columns are sorted according to the clusters identified via hierarchical clustering.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ComplexHeatmap)}
\FunctionTok{library}\NormalTok{(circlize)}
\end{Highlighting}
\end{Shaded}

We select the 50 nodes with the highest sum of edge weights to get a smaller heatmap.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sel }\OtherTok{\textless{}{-}} \FunctionTok{names}\NormalTok{(}\FunctionTok{sort}\NormalTok{(}\FunctionTok{rowSums}\NormalTok{(spring\_cor), }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))[}\FunctionTok{seq\_len}\NormalTok{(}\DecValTok{50}\NormalTok{)]}
\NormalTok{adja\_sel }\OtherTok{\textless{}{-}}\NormalTok{ spring\_cor[sel, sel]}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Color vector}
\NormalTok{col }\OtherTok{\textless{}{-}} \FunctionTok{colorRamp2}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{, }\SpecialCharTok{{-}}\FloatTok{0.5}\NormalTok{, }\DecValTok{0}\NormalTok{, }\FloatTok{0.5}\NormalTok{, }\DecValTok{1}\NormalTok{), }
                  \FunctionTok{c}\NormalTok{(}\StringTok{"royalblue4"}\NormalTok{, }\StringTok{"lightblue"}\NormalTok{, }\StringTok{"white"}\NormalTok{, }\StringTok{"orange"}\NormalTok{, }\StringTok{"firebrick3"}\NormalTok{))}

\FunctionTok{Heatmap}\NormalTok{(adja\_sel, }
        \AttributeTok{col =}\NormalTok{ col, }
        \AttributeTok{rect\_gp =} \FunctionTok{gpar}\NormalTok{(}\AttributeTok{col =} \StringTok{"gray"}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{1}\NormalTok{),}
        \AttributeTok{show\_row\_names =} \ConstantTok{FALSE}\NormalTok{, }
        \AttributeTok{show\_column\_names =} \ConstantTok{FALSE}\NormalTok{,}
        \AttributeTok{name =} \StringTok{"Association"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/clustered_heatmap-1.pdf}

The associations are generally quite low, and there are no prominent clusters detected by hierarchical clustering.

\hypertarget{global-network-measures}{%
\subsection{Global network measures}\label{global-network-measures}}

Global measures describe the overall network structure. We take a look at three common measures: density, transitivity, and average path length. The values are again computed with \texttt{igraph} functions.

\hypertarget{density}{%
\subsubsection{Density}\label{density}}

Definition: Proportion of present edges from all possible edges.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{edge\_density}\NormalTok{(spring\_graph)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.01444
\end{verbatim}

\hypertarget{transitivity-clustering-coefficient}{%
\subsubsection{Transitivity (clustering coefficient)}\label{transitivity-clustering-coefficient}}

Here, we consider only the global clustering coefficient, which is defined as the ratio of triangles to connected triples.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{transitivity}\NormalTok{(spring\_graph)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.1456
\end{verbatim}

\hypertarget{average-path-length}{%
\subsubsection{Average path length}\label{average-path-length}}

Definition: Mean of the shortest distance between each pair of nodes.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{average.path.length}\NormalTok{(spring\_graph)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1.699
\end{verbatim}

\hypertarget{network-analysis-with-netcomi}{%
\section{Network analysis with NetCoMi}\label{network-analysis-with-netcomi}}

The \texttt{netcomi\_net} object of class \texttt{microNet} created before is now passed to \texttt{netAnalyze()} to perform network analysis with \texttt{NetCoMi}.

The function computes several common network characteristics such as centrality measures, cluster assignment, the graphlet correlation matrix, as well as global network measures.

The user has several options to choose from, such as a clustering method, how to define hubs, and whether or not to normalize centrality values. See the help page \texttt{?netAnalyze} for a description of the arguments.

By default, a heatmap of the Graphlet Correlation Matrix (GCM) is returned (with graphlet correlations in the upper triangle and significance codes resulting from Student's t-test in the lower triangle). See \texttt{?calcGCM} and \texttt{?testGCM} for details.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{netcomi\_netprops }\OtherTok{\textless{}{-}} \FunctionTok{netAnalyze}\NormalTok{(netcomi\_net, }
                               \AttributeTok{clustMethod =} \StringTok{"cluster\_fast\_greedy"}\NormalTok{,}
                               \AttributeTok{hubPar =} \StringTok{"eigenvector"}\NormalTok{,}
                               \AttributeTok{normDeg =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics[width=0.75\linewidth]{60_network_learning_files/figure-latex/netAnalyze_single-1}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(netcomi\_netprops, }\AttributeTok{numbNodes =} \DecValTok{5}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Component sizes
## ```````````````              
## size: 103 2  1
##    #:   1 1 42
## ______________________________
## Global network properties
## `````````````````````````
## Largest connected component (LCC):
##                                  
## Relative LCC size         0.70068
## Clustering coefficient    0.21253
## Modularity                0.70495
## Positive edge percentage 99.35065
## Edge density              0.02932
## Natural connectivity      0.01173
## Vertex connectivity       1.00000
## Edge connectivity         1.00000
## Average dissimilarity*    0.99084
## Average path length**     3.81101
## 
## Whole network:
##                                  
## Number of components     44.00000
## Clustering coefficient    0.21253
## Modularity                0.70787
## Positive edge percentage 99.35484
## Edge density              0.01444
## Natural connectivity      0.00782
## -----
## *: Dissimilarity = 1 - edge weight
## **: Path length = Units with average dissimilarity
## 
## ______________________________
## Clusters
## - In the whole network
## - Algorithm: cluster_fast_greedy
## ```````````````````````````````` 
##                                   
## name:  0  1  2  3  4 5  6 7 8 9 10
##    #: 42 16 11 14 20 9 17 8 5 3  2
## 
## ______________________________
## Hubs
## - In alphabetical/numerical order
## - Based on empirical quantiles of centralities
## ```````````````````````````````````````````````                      
##  Agromyces            
##  Aneurinibacillus     
##  Anoxybacillus        
##  Chitinivibrio        
##  Erwinia              
##  Geobacillus          
##  Janibacter           
##  Pseudogracilibacillus
## 
## ______________________________
## Centrality measures
## - In decreasing order
## - Centrality of disconnected components is zero
## ````````````````````````````````````````````````
## Degree (unnormalized):
##                        
## Anoxybacillus        10
## Erwinia              10
## Chitinivibrio         9
## Janibacter            8
## Escherichia/Shigella  7
## 
## Betweenness centrality (normalized):
##                     
## Janibacter    0.3941
## Chitinivibrio 0.3137
## Enterobacter  0.2780
## Buttiauxella  0.2454
## Erwinia       0.2095
## 
## Closeness centrality (normalized):
##                             
## Chitinivibrio         0.5247
## Janibacter            0.5067
## Anoxybacillus         0.4841
## Pseudogracilibacillus 0.4659
## Erwinia               0.4653
## 
## Eigenvector centrality (normalized):
##                             
## Anoxybacillus         1.0000
## Chitinivibrio         0.9831
## Pseudogracilibacillus 0.8657
## Janibacter            0.8361
## Aneurinibacillus      0.6890
\end{verbatim}

\textbf{Interpretation of some findings:}

\begin{itemize}
\tightlist
\item
  The largest connected component (LCC) has 103 nodes and the network contains 42 singletons.
\item
  10 clusters have been identified, containing 2 to 20 nodes.
\item
  There are 8 hub nodes detected, which by definition are the nodes with the highest eigenvector centrality.
\item
  The average path length in the LCC is 3.811. This means that on average it takes 3.811 steps (step length is the average dissimilarity) to get from one node to another. Note that the average path length in \texttt{NetCoMi} is defined differently than in the \texttt{igraph} package, which is why the values differ.
\item
  Low values of edge density and the connectivity measures indicate that the network is rather sparse and not robust to perturbations (i.e., removal of nodes or edges).
\end{itemize}

\hypertarget{network-visualization}{%
\section{Network visualization}\label{network-visualization}}

Further insight into the network structure can be gained by visualizing the network. We have already seen examples of how to plot a network using the \texttt{igraph} package. Here we will use NetCoMi's plot function. It takes as input the \texttt{microNetProps} object returned by \texttt{netAnalyze()}, which contains all computed network properties. This has the advantage that the user can choose which properties to plot by simply changing some arguments. The plot function is based on \href{https://rdrr.io/cran/qgraph/}{qgraph}, which is another state-of-the-art R package for network visualization. The help page can be accessed via \texttt{?plot.microNetProps}.

\hypertarget{highlight-node-properties}{%
\subsection{Highlight node properties}\label{highlight-node-properties}}

In the first plot, node colors represent the detected clusters and node sizes are scaled by eigenvector centrality. Hub nodes are highlighted by default. Singletons are not included in the plot. To improve the readability, NetCoMi's ``intelligent'' label shortening approach is used.

Note that nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion argument until you find a value where the labels are legible, but also the clusters are still well recognizable.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(netcomi\_netprops,}
     \AttributeTok{repulsion =} \FloatTok{0.98}\NormalTok{,}
     \AttributeTok{rmSingles =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{shortenLabels =} \StringTok{"intelligent"}\NormalTok{,}
     \AttributeTok{labelScale =} \ConstantTok{FALSE}\NormalTok{,}
     \AttributeTok{nodeSize =} \StringTok{"eigenvector"}\NormalTok{,}
     \AttributeTok{nodeSizeSpread =} \DecValTok{3}\NormalTok{,}
     \AttributeTok{nodeColor =} \StringTok{"cluster"}\NormalTok{, }
     \AttributeTok{hubBorderCol =} \StringTok{"gray40"}\NormalTok{,}
     \AttributeTok{cexNodes =} \FloatTok{1.8}\NormalTok{,}
     \AttributeTok{edgeTranspHigh =} \DecValTok{20}\NormalTok{,}
     \AttributeTok{title1 =} \StringTok{"Network properties highlighted"}\NormalTok{, }
     \AttributeTok{showTitle =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{cexTitle =} \FloatTok{2.3}\NormalTok{,}
     \AttributeTok{mar =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{8}\NormalTok{))}

\FunctionTok{legend}\NormalTok{(}\FloatTok{0.7}\NormalTok{, }\FloatTok{1.1}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.7}\NormalTok{, }\AttributeTok{title =} \StringTok{"estimated correlation:"}\NormalTok{,}
       \AttributeTok{legend =} \FunctionTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{,}\StringTok{"{-}"}\NormalTok{), }\AttributeTok{lty =} \DecValTok{1}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{3}\NormalTok{, }\AttributeTok{col =} \FunctionTok{c}\NormalTok{(}\StringTok{"\#009900"}\NormalTok{,}\StringTok{"red"}\NormalTok{), }
       \AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{horiz =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/network_plot_single_cluster-1.pdf}

\hypertarget{highlight-data-features}{%
\subsection{Highlight data features}\label{highlight-data-features}}

We now color nodes according to their phylum. The node sizes are proportional to a taxon's sum of mclr-transformed abundances. As already mentioned in Section \ref{spring-network}, this is the normalization method used by \texttt{SPRING}. A color palette from \texttt{RColorBrewer} is used here.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(RColorBrewer)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Generate vector with phylum names for node coloring}
\NormalTok{phyla }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{phylum)}
\FunctionTok{names}\NormalTok{(phyla) }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{genus}

\CommentTok{\# Create color vector}
\NormalTok{colvec }\OtherTok{\textless{}{-}}\NormalTok{ RColorBrewer}\SpecialCharTok{::}\FunctionTok{brewer.pal}\NormalTok{(}\FunctionTok{length}\NormalTok{(}\FunctionTok{levels}\NormalTok{(phyla)), }\StringTok{"Set3"}\NormalTok{)}

\FunctionTok{plot}\NormalTok{(netcomi\_netprops,}
     \AttributeTok{repulsion =} \FloatTok{0.98}\NormalTok{,}
     \AttributeTok{rmSingles =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{shortenLabels =} \StringTok{"intelligent"}\NormalTok{,}
     \AttributeTok{labelScale =} \ConstantTok{FALSE}\NormalTok{,}
     \AttributeTok{nodeSize =} \StringTok{"mclr"}\NormalTok{,}
     \AttributeTok{nodeColor =} \StringTok{"feature"}\NormalTok{, }
     \AttributeTok{featVecCol =}\NormalTok{ phyla, }
     \AttributeTok{colorVec =}\NormalTok{  colvec,}
     \AttributeTok{nodeTransp =} \DecValTok{20}\NormalTok{,}
     \AttributeTok{highlightHubs =} \ConstantTok{FALSE}\NormalTok{,}
     \AttributeTok{cexNodes =} \FloatTok{1.8}\NormalTok{,}
     \AttributeTok{edgeTranspHigh =} \DecValTok{20}\NormalTok{,}
     \AttributeTok{title1 =} \StringTok{"Data features highlighted"}\NormalTok{, }
     \AttributeTok{showTitle =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{cexTitle =} \FloatTok{2.3}\NormalTok{,}
     \AttributeTok{mar =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{10}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{6}\NormalTok{))}

\CommentTok{\# Add legends}
\FunctionTok{legend}\NormalTok{(}\FloatTok{0.7}\NormalTok{, }\FloatTok{1.1}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.7}\NormalTok{, }\AttributeTok{title =} \StringTok{"estimated correlation:"}\NormalTok{,}
       \AttributeTok{legend =} \FunctionTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{,}\StringTok{"{-}"}\NormalTok{), }\AttributeTok{lty =} \DecValTok{1}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{3}\NormalTok{, }\AttributeTok{col =} \FunctionTok{c}\NormalTok{(}\StringTok{"\#009900"}\NormalTok{,}\StringTok{"red"}\NormalTok{), }
       \AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{horiz =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Colors used in the legend should be equally transparent as in the plot}
\NormalTok{col\_transp }\OtherTok{\textless{}{-}} \FunctionTok{colToTransp}\NormalTok{(colvec, }\DecValTok{20}\NormalTok{)}

\FunctionTok{legend}\NormalTok{(}\SpecialCharTok{{-}}\FloatTok{1.8}\NormalTok{, }\FloatTok{1.1}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.7}\NormalTok{, }\AttributeTok{pt.cex =} \FloatTok{2.5}\NormalTok{, }\AttributeTok{title =} \StringTok{"Phylum:"}\NormalTok{, }
       \AttributeTok{legend=}\FunctionTok{levels}\NormalTok{(phyla), }\AttributeTok{col =}\NormalTok{ col\_transp, }\AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{pch =} \DecValTok{16}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/network_plot_single_phylum-1.pdf}

\textbf{A few things to observe:}

\begin{itemize}
\tightlist
\item
  Genera belonging to the same phylum tend to cluster together, though not perfectly.
\item
  Genera with a low total count play a rather unimportant role in the network, i.e., they have a low centrality.
\item
  There is only one negative edge in the network. This edge is between two clusters, as expected when using the ``signed'' transformation.
\end{itemize}

\hypertarget{which-methods-to-choose}{%
\section{Which method(s) to choose?}\label{which-methods-to-choose}}

Throughout all the steps from primary data to potentially significant network features, there is a variety of methods and parameters to choose from. However, there is no general consensus in the community on the ``right'' way to estimate and analyze microbial networks. In the absence of a ``best method'' for inferring and analyzing microbial networks, researchers may be tempted to try different methods and report only the optimal results or those that fit some prior knowledge. This carries the risk of ``overfitting'' the analysis to the existing data so that the results are not replicable for new data \citep{ullmann2023over}.

Therefore, the selection of the workflow building blocks should be set up once and independently of any hypothesis about the data, thus avoiding the fallacy of starting to ``fish'' for results that best fit a previously formulated hypothesis. For example, one should ask prior to the analysis whether correlation or conditional dependence as a measure of association better fits the research question and choose the method accordingly. Another example is the choice of transformation from estimated association to dissimilarity (i.e., ``signed'' or ``unsigned''), which completely changes the interpretation and characteristics of the network. This choice should be made based on the research question before starting the analysis.

\hypertarget{more-about-association}{%
\section{More about association measures}\label{more-about-association}}

As mentioned in the introduction of this chapter, there are three types of association measures that are commonly used to express relationships between taxa: correlation, conditional dependence, and proportionality. Below, we provide a brief explanation of each of these measures, along with lists of available compositionality-aware approaches.

\begin{itemize}
\tightlist
\item
  \textbf{Correlation:} Two popular measures of ecological association are Pearson's correlation coefficient and Spearman's rank correlation coefficient, both of which can be inferred from empirical (sample) covariances. However, in the \(p\gg n\) setting, which most microbiome datasets are in, sample covariances and correlations are unreliable because the parameters being estimated are typically underdetermined. One way to improve sample covariance estimates is to assume that the underlying covariance matrix is sparse and use a regularized covariance estimator to implement this structural assumption. The Schäfer-Strimmer shrinkage estimator \citep{schafer2005shrinkage} is one possible method for estimating a sparse correlation matrix. Other popular methods, especially designed to estimate correlations for compositional data, are SparCC (``Sparse Correlations for Compositional data'') by \citet{Friedman2012}, CCREPE (``Compositionality Corrected by REnormalization and PErmutation'') by \citet{faust2012microbial}, and CCLasso (``Correlation inference for Compositional data through Lasso'') by \citet{fang2015cclasso}. The latter three methods already include a compositionality aware normalization, and SparCC also includes a zero replacement approach.
\item
  \textbf{Conditional dependence:} Since standard correlations include both direct and indirect dependencies, conditional dependence or partial correlation is often preferred for measuring association. Unlike (marginal) correlation, it expresses the relationship between two features conditioned on all other features in the data set. The approach and R package \href{https://github.com/zdk123/SpiecEasi}{SpiecEasi} (``Sparse InversE Covariance estimation for Ecological Association and Statistical Inference'') by \citet{Kurtz2015} is specifically designed for inferring ecological networks from microbiome data and includes two approaches for estimating conditional dependence structures between taxa: Neighborhood Selection; short ``MB'' \citep{meinshausen2006high} and (inverse) covariance selection \citep{friedman2008sparse}, which is based on a penalized maximum likelihood approach and is also known as ``graphical lasso''. Another approach and R package for inferring partial correlations from microbiome data is \href{https://github.com/GraceYoon/SPRING}{SPRING} (``Semi-Parametric Rank-based approach for INference in Graphical model'') by \citet{yoon2019microbial}. They also use the MB neighborhood selection method, but introduce a novel semi-parametric rank-based approach for sparse partial correlation estimation that can naturally handle the excess of zeros in the data. gCoda \citep{fang2017gcoda} is another conditional dependence measure based on penalized maximum likelihood estimation. All of the aforementioned conditional dependence measures address the high dimensionality of microbiome data.
\item
  \textbf{Proportionality:} \citet{lovell2015proportionality} introduce proportionality as an alternative measure of pairwise association for compositional data. The idea is that if the relative abundances between two taxa \(i\) and \(j\) are proportional, then their corresponding absolute abundances are also proportional: \(\frac{\omega_i}{m} \propto \frac{\omega_j}{m} \Rightarrow \omega_i \propto \omega_j\), where \(m\) is the sum of counts in the sample. It follows that proportionality is identical for the observed (relative) read counts and the true unobserved counts. The proportionality measure proposed by \citet{lovell2015proportionality} is based on log-ratio variance \(var(log \frac{x_i}{x_j})\), which is zero when \(\omega_i\) and \(\omega_j\) are perfectly proportional. Proportionality is implemented in the R package \href{https://github.com/tpq/propr}{propr}. \citet{badri2020shrinkage} extend the proportionality measure to a so-called ``shrinkage proportionality estimator''. It combines proportionality with the covariance shrinkage approach to obtain consistent association estimates even with small sample sizes.
\end{itemize}

\hypertarget{comparison-of-association-measures}{%
\section{Comparison of association measures}\label{comparison-of-association-measures}}

In this section, we provide three additional examples for constructing a network using each of the three types of association:

\begin{itemize}
\tightlist
\item
  Correlation using \texttt{SparCC}
\item
  Partial correlation using \texttt{SpiecEasi}
\item
  Proportionality using the shrinkage proportionality measure
\end{itemize}

\hypertarget{sparcc-correlation}{%
\subsection{SparCC}\label{sparcc-correlation}}

The first association measure we look at is SparCC (``Sparse Correlations for Compositional data''), introduced by \citet{Friedman2012}. It estimates Pearson correlations while taking into account the compositional structure of the data. The \href{https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html}{SpiecEasi} package provides an implementation of this method.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Set seed for reproducibility}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{13075}\NormalTok{)}
\CommentTok{\# Compute correlation matrix}
\NormalTok{sparcc\_cor }\OtherTok{\textless{}{-}}\NormalTok{ SpiecEasi}\SpecialCharTok{::}\FunctionTok{sparcc}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{)))}\SpecialCharTok{$}\NormalTok{Cor}
\FunctionTok{rownames}\NormalTok{(sparcc\_cor) }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(sparcc\_cor) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(tse)}
\end{Highlighting}
\end{Shaded}

We reuse the \texttt{transform\_asso()} function created in Section \ref{spring-network}, which sparsifies the association matrix, transforms it into a similarity matrix, and finally returns an \texttt{igraph} object.

Two threshold values are used to see the effect of sparsification later in the network plot.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sparcc\_trans03 }\OtherTok{\textless{}{-}} \FunctionTok{transform\_asso}\NormalTok{(sparcc\_cor, }\AttributeTok{thresh =} \FloatTok{0.3}\NormalTok{)}
\NormalTok{sparcc\_trans04 }\OtherTok{\textless{}{-}} \FunctionTok{transform\_asso}\NormalTok{(sparcc\_cor, }\AttributeTok{thresh =} \FloatTok{0.4}\NormalTok{)}

\NormalTok{sparcc\_graph03 }\OtherTok{\textless{}{-}}\NormalTok{ sparcc\_trans03}\SpecialCharTok{$}\NormalTok{graph}
\NormalTok{sparcc\_graph04 }\OtherTok{\textless{}{-}}\NormalTok{ sparcc\_trans04}\SpecialCharTok{$}\NormalTok{graph}
\end{Highlighting}
\end{Shaded}

\hypertarget{shrinkage-prop}{%
\subsection{Shrinkage proportionality}\label{shrinkage-prop}}

In the second example, microbial associations are measured by proportionality, originally introduced by \citet{lovell2015proportionality}. We use the shrinkage proportionality estimator proposed by \citet{badri2020shrinkage}, which gives consistent results even for small sample sizes. Since there is no R package implementing this estimator, we use the \texttt{rho\_shrink\_est()} function provided in the \href{https://github.com/MichelleBadri/NormCorr-manuscript/blob/master/code/helpers/norm_functions.R}{GitHub repository} associated with the paper. The function is slightly modified to take normalized counts as input.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(corpcor)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# norm\_counts: clr{-}transformed count matrix with samples in rows}
\NormalTok{rho\_shrink\_est }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(norm\_counts, ...) \{}
\NormalTok{  shrunk\_cov }\OtherTok{\textless{}{-}} \FunctionTok{cov.shrink}\NormalTok{(norm\_counts, ...)}
\NormalTok{  p }\OtherTok{\textless{}{-}} \FunctionTok{ncol}\NormalTok{(norm\_counts)}
\NormalTok{  J }\OtherTok{\textless{}{-}} \FunctionTok{matrix}\NormalTok{(}\FunctionTok{rep}\NormalTok{(}\FunctionTok{diag}\NormalTok{(shrunk\_cov), p), p)}
\NormalTok{  rho }\OtherTok{\textless{}{-}} \DecValTok{2} \SpecialCharTok{*}\NormalTok{ shrunk\_cov }\SpecialCharTok{/}\NormalTok{ (J }\SpecialCharTok{+} \FunctionTok{t}\NormalTok{(J))}
\NormalTok{  (rho }\SpecialCharTok{+} \FunctionTok{t}\NormalTok{(rho)) }\SpecialCharTok{/} \DecValTok{2}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Apply the shrinkage proportionality estimator to the clr{-}transformed counts}
\NormalTok{prop\_est }\OtherTok{\textless{}{-}} \FunctionTok{as}\NormalTok{(}\FunctionTok{rho\_shrink\_est}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"clr"}\NormalTok{))), }\StringTok{"matrix"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Estimating optimal shrinkage intensity lambda.var (variance vector): 0.0857 
## 
## Estimating optimal shrinkage intensity lambda (correlation matrix): 0.3634
\end{verbatim}

Again, we use our transformation function to convert the association matrix into a graph object.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{prop\_trans }\OtherTok{\textless{}{-}} \FunctionTok{transform\_asso}\NormalTok{(prop\_est, }\AttributeTok{thresh =} \FloatTok{0.4}\NormalTok{)}
\NormalTok{prop\_graph }\OtherTok{\textless{}{-}}\NormalTok{ prop\_trans}\SpecialCharTok{$}\NormalTok{graph}
\end{Highlighting}
\end{Shaded}

\hypertarget{spieceasi-mb}{%
\subsection{SpiecEasi - MB}\label{spieceasi-mb}}

As third example, we use the SpiecEasi (``Sparse InversE Covariance estimation for Ecological Association and Statistical Inference'') approach proposed by \citet{Kurtz2015} to estimate a sparse conditional dependency graph. The neighborhood selection method (``MB'') introduced by \citet{meinshausen2006high} is used for network learning. The approach is implemented in the R package \href{https://github.com/zdk123/SpiecEasi}{SpiecEasi}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(SpiecEasi)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{13075}\NormalTok{)}
\NormalTok{se\_mb\_est }\OtherTok{\textless{}{-}} \FunctionTok{spiec.easi}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"counts"}\NormalTok{)), }
                        \AttributeTok{method =} \StringTok{\textquotesingle{}mb\textquotesingle{}}\NormalTok{, }\AttributeTok{nlambda =} \DecValTok{20}\NormalTok{, }
                        \AttributeTok{pulsar.params =} \FunctionTok{list}\NormalTok{(}\AttributeTok{rep.num =} \DecValTok{20}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Since \texttt{SpiecEasi} uses the StARS (``Stability Approach to Regularization Selection'') method \citep{liu2010stability} to obtain a sparse association matrix, we don't need to set a threshold here. We store the partial correlations corresponding to the StARS-optimal lambda and convert them into an \texttt{igraph} object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get optimal matrix with partial correlations}
\NormalTok{se\_mb\_cor }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{getOptBeta}\NormalTok{(se\_mb\_est))}
\NormalTok{se\_mb\_cor }\OtherTok{\textless{}{-}} \FunctionTok{as.matrix}\NormalTok{(}\FunctionTok{symBeta}\NormalTok{(se\_mb\_cor))}
\FunctionTok{rownames}\NormalTok{(se\_mb\_cor) }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(se\_mb\_cor) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(tse)}
\FunctionTok{diag}\NormalTok{(se\_mb\_cor) }\OtherTok{\textless{}{-}} \DecValTok{1}

\CommentTok{\# Create graph object}
\NormalTok{se\_mb\_graph }\OtherTok{\textless{}{-}} \FunctionTok{transform\_asso}\NormalTok{(se\_mb\_cor)}\SpecialCharTok{$}\NormalTok{graph}
\end{Highlighting}
\end{Shaded}

\hypertarget{network-plots}{%
\subsection{Network plots}\label{network-plots}}

The graph objects can now be plotted using the \href{https://r.igraph.org/}{igraph} package. The same layout is used in all four plots so that the networks are comparable.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(igraph)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Node sizes}
\NormalTok{vsize }\OtherTok{\textless{}{-}}\NormalTok{ (}\FunctionTok{colMeans}\NormalTok{(}\FunctionTok{t}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse, }\StringTok{"log10"}\NormalTok{))) }\SpecialCharTok{+} \DecValTok{1}\NormalTok{) }\SpecialCharTok{*} \DecValTok{3}

\CommentTok{\# Use Fruchterman{-}Reingold (force{-}directed) layout}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{13075}\NormalTok{)}
\NormalTok{lay\_fr }\OtherTok{\textless{}{-}} \FunctionTok{layout\_with\_fr}\NormalTok{(se\_mb\_graph)}

\FunctionTok{par}\NormalTok{(}\AttributeTok{mfrow =} \FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\FunctionTok{plot}\NormalTok{(sparcc\_graph03, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize, }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \StringTok{"SparCC (thresh 0.3)"}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(sparcc\_graph04, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize, }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \StringTok{"SparCC (thresh 0.4)"}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(prop\_graph, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize, }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \StringTok{"Shrinkage proportionality}\SpecialCharTok{\textbackslash{}n}\StringTok{(thresh 0.4)"}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(se\_mb\_graph, }\AttributeTok{layout =}\NormalTok{ lay\_fr, }\AttributeTok{vertex.size =}\NormalTok{ vsize, }
     \AttributeTok{vertex.label =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{main =} \StringTok{"SpiecEasi (MB)"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/netplots-1.pdf}

\textbf{A few observations:}\\
The density of SparCC (threshold 0.4), proportionality and SpiecEasi is comparable, while the SparCC correlation network with threshold 0.3 is much denser. However, there are edges in the proportionality and SpiecEasi networks that are not present in the two SparCC networks. The SpiecEasi network has less highly connected nodes than the other three networks, but more nodes with one or two connections.

We will look at the degree distribution in the next section to quantify these observations.

\hypertarget{network-analysis-1}{%
\subsection{Network analysis}\label{network-analysis-1}}

Here we repeat some of the network analysis approaches explained in Section \ref{network-analysis}. The analyses are performed simultaneously for the three association measures as well as the SPRING network constructed in Section \ref{spring-network}. Therefore, we start by creating a list of all the graph objects we need for the analyses.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{graphlist }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{(}\AttributeTok{SparCC =}\NormalTok{ sparcc\_graph04, }
                  \AttributeTok{Proportionality =}\NormalTok{ prop\_graph,}
                  \AttributeTok{SpiecEasi =}\NormalTok{ se\_mb\_graph,}
                  \AttributeTok{SPRING =}\NormalTok{ spring\_graph)}
\end{Highlighting}
\end{Shaded}

\hypertarget{degree-distribution-1}{%
\subsubsection{Degree distribution}\label{degree-distribution-1}}

The degree distribution is plotted for all four measures to compare the overall network structure.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ggplot2)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Compute degree distributions}
\NormalTok{ddlist }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(graphlist, igraph}\SpecialCharTok{::}\NormalTok{degree.distribution)}

\CommentTok{\# Maximum degree}
\NormalTok{maxdeg }\OtherTok{\textless{}{-}} \FunctionTok{max}\NormalTok{(}\FunctionTok{lengths}\NormalTok{(ddlist))}

\CommentTok{\# Make list elements the same length}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \FunctionTok{seq\_along}\NormalTok{(graphlist)) \{}
  \FunctionTok{length}\NormalTok{(ddlist[[i]]) }\OtherTok{\textless{}{-}}\NormalTok{ maxdeg}
\NormalTok{\}}

\CommentTok{\# Data frame needed for ggplot2}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{Degree =} \FunctionTok{rep}\NormalTok{(}\FunctionTok{seq\_len}\NormalTok{(maxdeg), }\FunctionTok{length}\NormalTok{(graphlist)), }
                 \AttributeTok{Fraction =} \FunctionTok{unlist}\NormalTok{(ddlist), }
                 \AttributeTok{Method =} \FunctionTok{rep}\NormalTok{(}\FunctionTok{names}\NormalTok{(graphlist), }\AttributeTok{each =}\NormalTok{ maxdeg))}

\FunctionTok{ggplot}\NormalTok{(df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ Degree, }\AttributeTok{y =}\NormalTok{ Fraction, }\AttributeTok{group =}\NormalTok{ Method)) }\SpecialCharTok{+}
  \FunctionTok{geom\_line}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{color =}\NormalTok{ Method)) }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{color =}\NormalTok{ Method)) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/degree_dist-1.pdf}

The SparCC and shrinkage proportionality networks have a considerably higher proportion of singletons (zero-degree nodes) than the two conditional dependency graphs, but a lower proportion of nodes with degrees between one and five. The SpiecEasi and SPRING graphs, on the other hand, have a higher proportion of low degree nodes, but no highly connected nodes with a degree greater than eleven.

\hypertarget{clustered-heatmaps-1}{%
\subsubsection{Clustered heatmaps}\label{clustered-heatmaps-1}}

Using the \texttt{ComplexHeatmap} package, we plot heatmaps of the association matrices for the four considered association measures. Rows and columns are sorted according to the clusters identified via hierarchical clustering.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ComplexHeatmap)}
\FunctionTok{library}\NormalTok{(circlize)}
\FunctionTok{library}\NormalTok{(patchwork)}
\end{Highlighting}
\end{Shaded}

For each association measure, we select the 50 nodes with the highest sum of edge weights.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Function for selecting taxa with highest sum of edge weights}
\NormalTok{select\_taxa }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(adja, }\AttributeTok{ntaxa =} \DecValTok{50}\NormalTok{) \{}
\NormalTok{  sel }\OtherTok{\textless{}{-}} \FunctionTok{names}\NormalTok{(}\FunctionTok{sort}\NormalTok{(}\FunctionTok{rowSums}\NormalTok{(adja), }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{))[}\FunctionTok{seq\_len}\NormalTok{(ntaxa)]}
\NormalTok{  adja[sel, sel]}
\NormalTok{\}}

\NormalTok{assolist }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\NormalTok{assolist}\SpecialCharTok{$}\NormalTok{SparCC }\OtherTok{\textless{}{-}} \FunctionTok{select\_taxa}\NormalTok{(sparcc\_trans04}\SpecialCharTok{$}\NormalTok{adja)}
\NormalTok{assolist}\SpecialCharTok{$}\NormalTok{Proportionality }\OtherTok{\textless{}{-}} \FunctionTok{select\_taxa}\NormalTok{(prop\_trans}\SpecialCharTok{$}\NormalTok{adja)}
\NormalTok{assolist}\SpecialCharTok{$}\NormalTok{SpiecEasi }\OtherTok{\textless{}{-}} \FunctionTok{select\_taxa}\NormalTok{(se\_mb\_cor)}
\NormalTok{assolist}\SpecialCharTok{$}\NormalTok{SPRING }\OtherTok{\textless{}{-}} \FunctionTok{select\_taxa}\NormalTok{(spring\_cor)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Color vector for the legend}
\NormalTok{col }\OtherTok{\textless{}{-}} \FunctionTok{colorRamp2}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{, }\SpecialCharTok{{-}}\FloatTok{0.5}\NormalTok{, }\DecValTok{0}\NormalTok{, }\FloatTok{0.5}\NormalTok{, }\DecValTok{1}\NormalTok{), }
                  \FunctionTok{c}\NormalTok{(}\StringTok{"royalblue4"}\NormalTok{, }\StringTok{"lightblue"}\NormalTok{, }\StringTok{"white"}\NormalTok{, }\StringTok{"orange"}\NormalTok{, }\StringTok{"firebrick3"}\NormalTok{))}

\NormalTok{hm\_list }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}

\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \FunctionTok{seq\_along}\NormalTok{(assolist)) \{}
  \ControlFlowTok{if}\NormalTok{ (i }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{, }\DecValTok{4}\NormalTok{)) \{}
\NormalTok{    showlegend }\OtherTok{\textless{}{-}} \ConstantTok{TRUE}
\NormalTok{  \} }\ControlFlowTok{else}\NormalTok{ \{}
\NormalTok{    showlegend }\OtherTok{\textless{}{-}} \ConstantTok{FALSE}
\NormalTok{  \}}
  
\NormalTok{  hm\_list[[i]] }\OtherTok{\textless{}{-}} \FunctionTok{Heatmap}\NormalTok{(assolist[[i]], }
                          \AttributeTok{col =}\NormalTok{ col, }
                          \AttributeTok{rect\_gp =} \FunctionTok{gpar}\NormalTok{(}\AttributeTok{col =} \StringTok{"gray"}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{1}\NormalTok{),}
                          \AttributeTok{show\_row\_names =} \ConstantTok{FALSE}\NormalTok{, }
                          \AttributeTok{show\_column\_names =} \ConstantTok{FALSE}\NormalTok{,}
                          \AttributeTok{column\_title =} \FunctionTok{names}\NormalTok{(assolist)[i], }
                          \AttributeTok{name =} \StringTok{"Association"}\NormalTok{,}
                          \AttributeTok{show\_heatmap\_legend =}\NormalTok{ showlegend) }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{draw}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{grid.grabExpr}\NormalTok{()}
\NormalTok{\}}

\CommentTok{\# Plot with wrap\_plots() function from patchwork package}
\FunctionTok{wrap\_plots}\NormalTok{(hm\_list, }\AttributeTok{ncol =} \DecValTok{2}\NormalTok{, }\AttributeTok{widths =} \FunctionTok{c}\NormalTok{(}\DecValTok{8}\NormalTok{, }\DecValTok{10}\NormalTok{, }\DecValTok{8}\NormalTok{, }\DecValTok{10}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\includegraphics{60_network_learning_files/figure-latex/clustered_heatmap_all-1.pdf}

The SparCC and the proportionality network show a block structure, where each block corresponds to a cluster. The clusters are less pronounced in the conditional dependence networks. The latter also generally have lower edge weights.

\hypertarget{global-network-measures-1}{%
\subsubsection{Global network measures}\label{global-network-measures-1}}

For each association measure, the three global network measures density, transitivity, and average path length are computed and stored in a data frame for comparison.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Compute density and store in a data frame}
\NormalTok{glob }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{Density =} \FunctionTok{unlist}\NormalTok{(}\FunctionTok{lapply}\NormalTok{(graphlist, edge\_density)))}

\CommentTok{\# Transitivity}
\NormalTok{glob}\SpecialCharTok{$}\NormalTok{Transitivity }\OtherTok{\textless{}{-}} \FunctionTok{unlist}\NormalTok{(}\FunctionTok{lapply}\NormalTok{(graphlist, transitivity))}

\CommentTok{\# Average path length}
\NormalTok{glob}\SpecialCharTok{$}\NormalTok{Av.path }\OtherTok{\textless{}{-}} \FunctionTok{unlist}\NormalTok{(}\FunctionTok{lapply}\NormalTok{(graphlist, average.path.length))}

\NormalTok{glob}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                 Density Transitivity Av.path
## SparCC          0.01510       0.5458  0.8932
## Proportionality 0.01295       0.5402  1.1847
## SpiecEasi       0.02162       0.2145  1.6322
## SPRING          0.01444       0.1456  1.6993
\end{verbatim}

\hypertarget{network-comparison}{%
\chapter{Network comparison}\label{network-comparison}}

This chapter assumes that you are already familiar with how to construct and analyze a single microbial network, which is explained in Chapter \ref{network-learning}.

Since microbial interactions are likely to change between conditions, such as between patients and healthy individuals or between different environmental states, identifying network differences between groups is often an integral secondary analysis step. Differences can be detected visually by plotting the networks side by side, or quantitatively using differential network analysis tools.

Two approaches for comparing networks between two conditions are considered in this chapter:

\begin{itemize}
\tightlist
\item
  \textbf{Differential network analysis}, which analyzes differences in network metrics and network structure.
\item
  \textbf{Differential association analysis}, which focuses on differences in the strength of individual associations.
\end{itemize}

See Section \ref{netcomp-methods} for further details on these two approaches and the methods used in this chapter.

Here we use \href{https://github.com/stefpeschel/NetCoMi}{NetCoMi} \citep{peschel2021netcomi} for network comparison, which includes several \textbf{differential network analysis approaches} as well as functionality for *differential association analysis**, i.e., generating a differential network. How to install NetCoMi from GitHub is explained in chapter \ref{network-learning}.

The \textbf{PeerJ data set} \citep{potbhare2022skin} containing skin microbial profiles for 58 subjects is again used in this chapter. The dataset also includes information on the subjects' geographic location, gender, age and diet. Whether the \textbf{skin microbiome differs} between people with \textbf{different diets} is an interesting question that will be explored in this chapter.

\hypertarget{data-preparation-1}{%
\section{Data preparation}\label{data-preparation-1}}

We perform the same data preprocessing steps as in chapter \ref{network-learning}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(NetCoMi)}
\FunctionTok{library}\NormalTok{(mia)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{data}\NormalTok{(}\StringTok{"peerj13075"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse0 }\OtherTok{\textless{}{-}}\NormalTok{ peerj13075}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate to genus level}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{agglomerateByRank}\NormalTok{(tse0, }\AttributeTok{rank =} \StringTok{"genus"}\NormalTok{) }

\CommentTok{\# Add relative abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }
                      \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }
                      \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{,}
                      \AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{) }

\CommentTok{\# Filter by prevalence}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{subsetByPrevalentFeatures}\NormalTok{(tse,}
                                 \AttributeTok{prevalence =} \FloatTok{0.2}\NormalTok{,}
                                 \AttributeTok{detection =} \DecValTok{0}\NormalTok{,}
                                 \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Add log10{-}transformed abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"log10"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\CommentTok{\# Add clr{-}transformed abundances}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Based on ``Diet'', the \texttt{tse} object is then split into two groups: One with mixed diet subjects, and one with vegetarian subjects. Both subsets have nearly the same sample size and are therefore comparable.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{table}\NormalTok{(tse}\SpecialCharTok{$}\NormalTok{Diet)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Mixed   Veg 
##    28    30
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse\_list }\OtherTok{\textless{}{-}} \FunctionTok{splitOn}\NormalTok{(tse, }\AttributeTok{f =} \StringTok{"Diet"}\NormalTok{, }\AttributeTok{use\_names =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{MARGIN =} \DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{network-learning-and-analysis}{%
\section{Network learning and analysis}\label{network-learning-and-analysis}}

The approach starts again with network construction and analysis, but this time we pass the two data sets to \texttt{netConstruct} to perform a network comparison.

The \texttt{rep.num} argument is set to 10 to perform only 10 repetitions in the model selection approach. This speeds up the permutation tests performed later, and has a negligible effect for this data set.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{spring\_net\_diet }\OtherTok{\textless{}{-}} \FunctionTok{netConstruct}\NormalTok{(}\AttributeTok{data =}\NormalTok{ tse\_list}\SpecialCharTok{$}\NormalTok{Mixed,}
                                \AttributeTok{data2 =}\NormalTok{ tse\_list}\SpecialCharTok{$}\NormalTok{Veg,}
                                \AttributeTok{taxRank =} \StringTok{"genus"}\NormalTok{,}
                                \AttributeTok{filtTax =} \StringTok{"highestFreq"}\NormalTok{,}
                                \AttributeTok{filtTaxPar =} \FunctionTok{list}\NormalTok{(}\AttributeTok{highestFreq  =} \DecValTok{100}\NormalTok{),}
                                \AttributeTok{measure =} \StringTok{"spring"}\NormalTok{,}
                                \AttributeTok{measurePar =} \FunctionTok{list}\NormalTok{(}\AttributeTok{nlambda =} \DecValTok{20}\NormalTok{, }
                                                  \AttributeTok{rep.num =} \DecValTok{10}\NormalTok{,}
                                                  \AttributeTok{thresh =} \FloatTok{0.05}\NormalTok{,}
                                                  \AttributeTok{Rmethod =} \StringTok{"approx"}\NormalTok{),}
                                \AttributeTok{sparsMethod =} \StringTok{"none"}\NormalTok{, }
                                \AttributeTok{dissFunc =} \StringTok{"signed"}\NormalTok{,}
                                \AttributeTok{verbose =} \DecValTok{3}\NormalTok{,}
                                \AttributeTok{seed =} \DecValTok{13075}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

All network measures are now computed for both networks. Also, both GCMs are plotted together with a third matrix containing the differences between the GCMs and significance codes that express if the differences are significantly different from zero.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{spring\_netprops\_diet }\OtherTok{\textless{}{-}} \FunctionTok{netAnalyze}\NormalTok{(spring\_net\_diet, }
                                   \AttributeTok{clustMethod =} \StringTok{"cluster\_fast\_greedy"}\NormalTok{,}
                                   \AttributeTok{hubPar =} \StringTok{"eigenvector"}\NormalTok{,}
                                   \AttributeTok{normDeg =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{61_network_comparison_files/figure-latex/netAnalyze_diet-1.pdf}

In both of the networks, some graphlet correlations are significantly different from zero. However, none of the correlations are significantly different between the groups.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(spring\_netprops\_diet, }\AttributeTok{groupNames =} \FunctionTok{c}\NormalTok{(}\StringTok{"Mixed diet"}\NormalTok{, }\StringTok{"Vegetarian"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Component sizes
## ```````````````
## Mixed diet:                   
## size: 18 8 4 3 2  1
##    #:  1 1 1 1 3 43
## Vegetarian:                 
## size: 21 5 4 2  1
##    #:  1 1 1 3 46
## ______________________________
## Global network properties
## `````````````````````````
## Largest connected component (LCC):
##                          Mixed diet Vegetarian
## Relative LCC size           0.21951    0.25610
## Clustering coefficient      0.10393    0.22433
## Modularity                  0.51500    0.47707
## Positive edge percentage   80.00000  100.00000
## Edge density                0.13072    0.12381
## Natural connectivity        0.07312    0.06253
## Vertex connectivity         1.00000    1.00000
## Edge connectivity           1.00000    1.00000
## Average dissimilarity*      0.96009    0.96154
## Average path length**       2.22768    2.24622
## 
## Whole network:
##                          Mixed diet Vegetarian
## Number of components       50.00000   52.00000
## Clustering coefficient      0.07506    0.16530
## Modularity                  0.75184    0.66821
## Positive edge percentage   88.57143  100.00000
## Edge density                0.01054    0.01084
## Natural connectivity        0.01342    0.01347
## -----
## *: Dissimilarity = 1 - edge weight
## **: Path length = Units with average dissimilarity
## 
## ______________________________
## Clusters
## - In the whole network
## - Algorithm: cluster_fast_greedy
## ```````````````````````````````` 
## Mixed diet:                          
## name:  0 1 2 3 4 5 6 7 8 9
##    #: 43 8 6 6 6 4 3 2 2 2
## 
## Vegetarian:                        
## name:  0 1 2 3 4 5 6 7 8
##    #: 46 9 8 4 5 4 2 2 2
## 
## ______________________________
## Hubs
## - In alphabetical/numerical order
## - Based on empirical quantiles of centralities
## ```````````````````````````````````````````````
##   Mixed diet           Vegetarian
##  Citrobacter            Aeromonas
##      Erwinia              Erwinia
##  Escherichia Escherichia/Shigella
##     Serratia           Salmonella
##   Shewanella          Siccibacter
## 
## ______________________________
## Centrality measures
## - In decreasing order
## - Centrality of disconnected components is zero
## ````````````````````````````````````````````````
## Degree (unnormalized):
##                       Mixed diet Vegetarian
##               Erwinia          6          6
##           Citrobacter          4          0
##              Serratia          4          3
##         Streptococcus          3          0
##               Pantoea          2          1
##                           ______     ______
##               Erwinia          6          6
##             Aeromonas          1          5
##  Escherichia/Shigella          2          5
##           Siccibacter          0          5
##            Salmonella          2          3
## 
## Betweenness centrality (normalized):
##                       Mixed diet Vegetarian
##               Erwinia    0.60294    0.54737
##              Serratia    0.41912        0.1
##         Streptococcus    0.27941          0
##          Enterobacter    0.24265          0
##           Citrobacter    0.24265          0
##                           ______     ______
##               Erwinia    0.60294    0.54737
##  Escherichia/Shigella    0.11765    0.53158
##             Aeromonas          0    0.36842
##           Siccibacter          0        0.3
##         Thiolamprovum          0    0.20526
## 
## Closeness centrality (normalized):
##                       Mixed diet Vegetarian
##               Erwinia    0.85251    0.80605
##              Serratia    0.78452    0.56308
##           Citrobacter    0.71265          0
##          Enterobacter    0.64822    0.49217
##         Streptococcus    0.61535          0
##                           ______     ______
##  Escherichia/Shigella    0.57246    0.80747
##               Erwinia    0.85251    0.80605
##             Aeromonas     0.4768    0.74282
##           Siccibacter          0    0.72705
##            Salmonella    0.57364    0.66777
## 
## Eigenvector centrality (normalized):
##                       Mixed diet Vegetarian
##              Serratia          1    0.41577
##               Erwinia    0.97703          1
##           Citrobacter    0.80991          0
##           Escherichia    0.60982          0
##            Shewanella    0.41174     0.1238
##                           ______     ______
##               Erwinia    0.97703          1
##  Escherichia/Shigella    0.38659    0.96856
##           Siccibacter          0      0.888
##             Aeromonas    0.26033    0.68516
##            Salmonella     0.3153    0.64405
\end{verbatim}

For each centrality measure, the five nodes with the highest centrality in each group are plotted by default.

We notice some differences in the network properties. The differential network analysis performed in the next section will show if the differences are significant.

\hypertarget{differential-network-analysis}{%
\section{Differential network analysis}\label{differential-network-analysis}}

\hypertarget{visual-comparison}{%
\subsection{Visual comparison}\label{visual-comparison}}

We start with a visual comparison of the two networks using NetCoMi's plot function. The same configuration as in chapter \ref{network-learning} is used.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(spring\_netprops\_diet,}
     \AttributeTok{repulsion =} \FloatTok{0.97}\NormalTok{,}
     \AttributeTok{rmSingles =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{labelScale =} \ConstantTok{FALSE}\NormalTok{,}
     \AttributeTok{nodeSize =} \StringTok{"eigenvector"}\NormalTok{,}
     \AttributeTok{nodeSizeSpread =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{nodeColor =} \StringTok{"cluster"}\NormalTok{, }
     \AttributeTok{sameColThresh =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{hubBorderCol =} \StringTok{"darkgray"}\NormalTok{,}
     \AttributeTok{cexNodes =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{edgeTranspHigh =} \DecValTok{20}\NormalTok{,}
     \AttributeTok{title1 =} \StringTok{"Mixed diet"}\NormalTok{, }
     \AttributeTok{title2 =} \StringTok{"Vegetarian"}\NormalTok{,}
     \AttributeTok{showTitle =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{cexTitle =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{mar =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{))}

\CommentTok{\# Overlay a transparent plot on which the legend is plotted}
\FunctionTok{par}\NormalTok{(}\AttributeTok{fig=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{), }\AttributeTok{oma=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{), }\AttributeTok{mar=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{), }\AttributeTok{new=}\ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\AttributeTok{type=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{bty=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{xaxt=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{yaxt=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{)}

\FunctionTok{legend}\NormalTok{(}\SpecialCharTok{{-}}\FloatTok{0.2}\NormalTok{, }\SpecialCharTok{{-}}\FloatTok{0.9}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.5}\NormalTok{, }\AttributeTok{title =} \StringTok{"estimated correlation:"}\NormalTok{,}
       \AttributeTok{legend =} \FunctionTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{,}\StringTok{"{-}"}\NormalTok{), }\AttributeTok{lty =} \DecValTok{1}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{3}\NormalTok{, }\AttributeTok{col =} \FunctionTok{c}\NormalTok{(}\StringTok{"\#009900"}\NormalTok{,}\StringTok{"red"}\NormalTok{), }
       \AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{horiz =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{61_network_comparison_files/figure-latex/network_plot_diet_difflay-1.pdf}

The layout is computed separately for each network, making it difficult to visually compare certain associations. It is therefore recommended to use the same layout for both groups (argument \texttt{sameLayout}). Instead of simply copying one layout to the other network, we set \texttt{layoutGroup} to ``union''. This ensures that the nodes are placed as optimally as possible for both networks.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(spring\_netprops\_diet,}
     \AttributeTok{sameLayout =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{repulsion =} \FloatTok{0.95}\NormalTok{,}
     \AttributeTok{rmSingles =} \StringTok{"inboth"}\NormalTok{,}
     \AttributeTok{labelScale =} \ConstantTok{FALSE}\NormalTok{,}
     \AttributeTok{nodeSize =} \StringTok{"eigenvector"}\NormalTok{,}
     \AttributeTok{nodeSizeSpread =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{nodeColor =} \StringTok{"cluster"}\NormalTok{, }
     \AttributeTok{sameColThresh =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{hubBorderCol =} \StringTok{"darkgray"}\NormalTok{,}
     \AttributeTok{cexNodes =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{edgeTranspHigh =} \DecValTok{20}\NormalTok{,}
     \AttributeTok{title1 =} \StringTok{"Mixed diet"}\NormalTok{, }
     \AttributeTok{title2 =} \StringTok{"Vegetarian"}\NormalTok{,}
     \AttributeTok{showTitle =} \ConstantTok{TRUE}\NormalTok{,}
     \AttributeTok{cexTitle =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{mar =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{))}

\CommentTok{\# Add legend}
\FunctionTok{par}\NormalTok{(}\AttributeTok{fig=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{), }\AttributeTok{oma=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{), }\AttributeTok{mar=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{), }\AttributeTok{new=}\ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\AttributeTok{type=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{bty=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{xaxt=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{yaxt=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{)}
\FunctionTok{legend}\NormalTok{(}\SpecialCharTok{{-}}\FloatTok{0.2}\NormalTok{, }\SpecialCharTok{{-}}\FloatTok{0.8}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.7}\NormalTok{, }\AttributeTok{title =} \StringTok{"estimated correlation:"}\NormalTok{,}
       \AttributeTok{legend =} \FunctionTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{,}\StringTok{"{-}"}\NormalTok{), }\AttributeTok{lty =} \DecValTok{1}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{3}\NormalTok{, }\AttributeTok{col =} \FunctionTok{c}\NormalTok{(}\StringTok{"\#009900"}\NormalTok{,}\StringTok{"red"}\NormalTok{), }
       \AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{horiz =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{61_network_comparison_files/figure-latex/network_plot_diet_samelay-1.pdf}

\textbf{A few notes:}

\begin{itemize}
\tightlist
\item
  Differences in the edge weights can now be seen at first glance. For example, Serratia and Citrobacter are strongly associated in the mixed diet group, but not at all in the vegetarian group.
\item
  Clusters must share at least two nodes (\texttt{sameColThresh} argument) to be colored equally in both networks, which is why the color of some clusters differs between the groups.
\item
  The clustering generally differs markedly. In particular, the cluster assignment of many of the nodes in the largest connected component differs between the two groups.
\end{itemize}

As in Chapter \ref{network-learning}, we also generate a network plot using phylum names to color the nodes and mclr-transformed abundances to scale node sizes.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(RColorBrewer)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Generate vector with phylum names for node coloring}
\NormalTok{phyla }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{phylum)}
\FunctionTok{names}\NormalTok{(phyla) }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{genus}

\CommentTok{\# Create color vector}
\NormalTok{colvec }\OtherTok{\textless{}{-}}\NormalTok{ RColorBrewer}\SpecialCharTok{::}\FunctionTok{brewer.pal}\NormalTok{(}\FunctionTok{length}\NormalTok{(}\FunctionTok{levels}\NormalTok{(phyla)), }\StringTok{"Set3"}\NormalTok{)}

\NormalTok{p\_diet }\OtherTok{\textless{}{-}} \FunctionTok{plot}\NormalTok{(spring\_netprops\_diet,}
               \AttributeTok{sameLayout =} \ConstantTok{TRUE}\NormalTok{,}
               \AttributeTok{repulsion =} \FloatTok{0.95}\NormalTok{,}
               \AttributeTok{rmSingles =} \StringTok{"inboth"}\NormalTok{,}
               \AttributeTok{labelScale =} \ConstantTok{FALSE}\NormalTok{,}
               \AttributeTok{nodeSize =} \StringTok{"clr"}\NormalTok{,}
               \AttributeTok{nodeColor =} \StringTok{"feature"}\NormalTok{, }
               \AttributeTok{featVecCol =}\NormalTok{ phyla, }
               \AttributeTok{colorVec =}\NormalTok{  colvec,}
               \AttributeTok{nodeTransp =} \DecValTok{20}\NormalTok{,}
               \AttributeTok{sameColThresh =} \DecValTok{2}\NormalTok{,}
               \AttributeTok{highlightHubs =} \ConstantTok{FALSE}\NormalTok{,}
               \AttributeTok{cexNodes =} \DecValTok{2}\NormalTok{,}
               \AttributeTok{edgeTranspHigh =} \DecValTok{20}\NormalTok{,}
               \AttributeTok{title1 =} \StringTok{"Mixed diet"}\NormalTok{, }
               \AttributeTok{title2 =} \StringTok{"Vegetarian"}\NormalTok{,}
               \AttributeTok{showTitle =} \ConstantTok{TRUE}\NormalTok{,}
               \AttributeTok{cexTitle =} \DecValTok{2}\NormalTok{,}
               \AttributeTok{mar =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{))}

\CommentTok{\# Add legends}
\CommentTok{\# Colors used in the legend should be equally transparent as in the plot}
\NormalTok{col\_transp }\OtherTok{\textless{}{-}} \FunctionTok{colToTransp}\NormalTok{(colvec, }\DecValTok{20}\NormalTok{)}

\FunctionTok{par}\NormalTok{(}\AttributeTok{fig=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{), }\AttributeTok{oma=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{), }\AttributeTok{mar=}\FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{), }\AttributeTok{new=}\ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{plot}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{0}\NormalTok{, }\AttributeTok{type=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{bty=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{xaxt=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{, }\AttributeTok{yaxt=}\StringTok{\textquotesingle{}n\textquotesingle{}}\NormalTok{)}
\FunctionTok{legend}\NormalTok{(}\SpecialCharTok{{-}}\FloatTok{0.15}\NormalTok{, }\SpecialCharTok{{-}}\FloatTok{0.8}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.7}\NormalTok{, }\AttributeTok{title =} \StringTok{"estimated correlation:"}\NormalTok{,}
       \AttributeTok{legend =} \FunctionTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{,}\StringTok{"{-}"}\NormalTok{), }\AttributeTok{lty =} \DecValTok{1}\NormalTok{, }\AttributeTok{lwd =} \DecValTok{3}\NormalTok{, }\AttributeTok{col =} \FunctionTok{c}\NormalTok{(}\StringTok{"\#009900"}\NormalTok{,}\StringTok{"red"}\NormalTok{), }
       \AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{horiz =} \ConstantTok{TRUE}\NormalTok{)}
\FunctionTok{legend}\NormalTok{(}\SpecialCharTok{{-}}\FloatTok{0.15}\NormalTok{, }\FloatTok{1.3}\NormalTok{, }\AttributeTok{cex =} \FloatTok{1.7}\NormalTok{, }\AttributeTok{pt.cex =} \FloatTok{2.5}\NormalTok{, }\AttributeTok{title =} \StringTok{"Phylum:"}\NormalTok{, }
       \AttributeTok{legend=}\FunctionTok{levels}\NormalTok{(phyla), }\AttributeTok{col =}\NormalTok{ col\_transp, }\AttributeTok{bty =} \StringTok{"n"}\NormalTok{, }\AttributeTok{pch =} \DecValTok{16}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

\includegraphics{61_network_comparison_files/figure-latex/network_plot_diet_phyla-1.pdf}

\hypertarget{quantitative-comparison}{%
\subsection{Quantitative comparison}\label{quantitative-comparison}}

\texttt{netCompare()} enables a quantitative network comparison using comparative measures such as Jaccard's Index, Adjusted Rand Index, and permutation tests.

To test for statistical significance of differences in network properties, we perform permutation tests with 1000 permutations. Multiple CPU cores are used to save run time. The association matrices estimated for all permutations are stored in an external file. We will reuse them later when performing differential association analysis. They could also be used to rerun \texttt{netCompare()} with different parameter settings.

Note that unless running on a cluster with considerably more CPU cores, a network comparison with permutation tests may take several hours. You should test the code below with a small number of permutations to make sure it works before applying it to your data.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{spring\_netcomp\_diet }\OtherTok{\textless{}{-}} \FunctionTok{netCompare}\NormalTok{(spring\_netprops\_diet, }
                                  \AttributeTok{permTest =} \ConstantTok{TRUE}\NormalTok{,}
                                  \AttributeTok{nPerm =} \DecValTok{1000}\NormalTok{,}
                                  \AttributeTok{cores =} \DecValTok{6}\NormalTok{,}
                                  \AttributeTok{seed =} \DecValTok{13075}\NormalTok{,}
                                  \AttributeTok{storeAssoPerm =} \ConstantTok{TRUE}\NormalTok{,}
                                  \AttributeTok{fileStoreAssoPerm =} \StringTok{"general/network\_data/spring\_assoPerm"}\NormalTok{,}
                                  \AttributeTok{verbose =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{summary}\NormalTok{(spring\_netcomp\_diet, }
        \AttributeTok{groupNames =} \FunctionTok{c}\NormalTok{(}\StringTok{"Mixed diet"}\NormalTok{, }\StringTok{"Vegetarian"}\NormalTok{),}
        \AttributeTok{numbNodes =} \DecValTok{5}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Comparison of Network Properties
## ----------------------------------
## CALL: 
## netCompare(x = spring_netprops_diet, permTest = TRUE, verbose = TRUE, 
##     nPerm = 1000, cores = 19, libPathsClust = "/dss/dsshome1/07/di93fen/R", 
##     seed = 13075, storeAssoPerm = TRUE, fileStoreAssoPerm = "general/network_data/spring_assoPerm")
## 
## ______________________________
## Global network properties
## `````````````````````````
## Largest connected component (LCC):
##                          Mixed diet   Vegetarian    abs.diff.     p-value    
## Relative LCC size             0.220        0.256        0.037    0.838162    
## Clustering coefficient        0.104        0.224        0.120    0.450549    
## Modularity                    0.515        0.477        0.038    0.801199    
## Positive edge percentage     80.000      100.000       20.000    0.000999 ***
## Edge density                  0.131        0.124        0.007    0.910090    
## Natural connectivity          0.073        0.063        0.011    0.812188    
## Vertex connectivity           1.000        1.000        0.000    1.000000    
## Edge connectivity             1.000        1.000        0.000    1.000000    
## Average dissimilarity*        0.960        0.962        0.001    0.945055    
## Average path length**         2.228        2.246        0.019    0.984016    
## 
## Whole network:
##                          Mixed diet   Vegetarian    abs.diff.     p-value    
## Number of components         50.000       52.000        2.000     0.93706    
## Clustering coefficient        0.075        0.165        0.090     0.46254    
## Modularity                    0.752        0.668        0.084     0.24975    
## Positive edge percentage     88.571      100.000       11.429     0.02398 *  
## Edge density                  0.011        0.011        0.000     0.97303    
## Natural connectivity          0.013        0.013        0.000     0.89011    
## -----
## p-values: one-tailed test with null hypothesis diff=0
##  *: Dissimilarity = 1 - edge weight
## **: Path length = Units with average dissimilarity
## 
## ______________________________
## Jaccard index (similarity betw. sets of most central nodes)
## ```````````````````````````````````````````````````````````
##                     Jacc   P(<=Jacc)     P(>=Jacc)    
## degree             0.560      0.9944       0.01637 *  
## betweenness centr. 0.294      0.4777       0.71860    
## closeness centr.   0.560      0.9944       0.01637 *  
## eigenvec. centr.   0.560      0.9944       0.01637 *  
## hub taxa           0.111      0.1431       0.97399    
## -----
## Jaccard index in [0,1] (1 indicates perfect agreement)
## 
## ______________________________
## Adjusted Rand index (similarity betw. clusterings)
## ``````````````````````````````````````````````````
##         wholeNet       LCC
## ARI        0.367     0.035
## p-value    0.000     0.563
## -----
## ARI in [-1,1] with ARI=1: perfect agreement betw. clusterings
##                    ARI=0: expected for two random clusterings
## p-value: permutation test (n=1000) with null hypothesis ARI=0
## 
## ______________________________
## Graphlet Correlation Distance
## `````````````````````````````
##         wholeNet         LCC  
## GCD       1.5980      2.1770  
## p-value   0.4226      0.7143  
## -----
## GCD >= 0 (GCD=0 indicates perfect agreement between GCMs)
## p-value: permutation test with null hypothesis GCD=0
## 
## ______________________________
## Centrality measures
## - In decreasing order
## - Centrality of disconnected components is zero
## ````````````````````````````````````````````````
## Degree (unnormalized):
##                      Mixed diet Vegetarian abs.diff. adj.p-value  
## Siccibacter                   0          5         5           1  
## Aeromonas                     1          5         4           1  
## Citrobacter                   4          0         4           1  
## Escherichia/Shigella          2          5         3           1  
## Streptococcus                 3          0         3           1  
## 
## Betweenness centrality (normalized):
##                      Mixed diet Vegetarian abs.diff. adj.p-value  
## Escherichia/Shigella      0.118      0.532     0.414           1  
## Aeromonas                 0.000      0.368     0.368           1  
## Serratia                  0.419      0.100     0.319           1  
## Siccibacter               0.000      0.300     0.300           1  
## Streptococcus             0.279      0.000     0.279           1  
## 
## Closeness centrality (normalized):
##               Mixed diet Vegetarian abs.diff. adj.p-value  
## Siccibacter        0.000      0.727     0.727      0.2676  
## Citrobacter        0.713      0.000     0.713      0.9799  
## Streptococcus      0.615      0.000     0.615      0.2676  
## Escherichia        0.587      0.000     0.587      0.9799  
## Xenorhabdus        0.000      0.527     0.527      0.2676  
## 
## Eigenvector centrality (normalized):
##                      Mixed diet Vegetarian abs.diff. adj.p-value  
## Siccibacter               0.000      0.888     0.888      0.6553  
## Citrobacter               0.810      0.000     0.810      1.0000  
## Escherichia               0.610      0.000     0.610      1.0000  
## Serratia                  1.000      0.416     0.584      1.0000  
## Escherichia/Shigella      0.387      0.969     0.582      1.0000  
## 
## _________________________________________________________
## Significance codes: ***: 0.001, **: 0.01, *: 0.05, .: 0.1
\end{verbatim}

\textbf{Interpreting some results:}

\begin{itemize}
\tightlist
\item
  Almost all global network properties are significantly different between the groups (for \(\alpha=0.1\)), thus reflecting the different overall network structure we already have seen in the network plots.
\item
  For the Jaccard index of degree, closeness, and eigenvector centrality, the probability P(\textgreater=Jacc) is significant, meaning that the sets of the most central nodes are quite similar for these three measures. The Jaccard index for the hub nodes, on the other hand, is low because the two networks share only one hub node (``Erwinia'').
\item
  As indicated by some similarities in the clusterings, the adjusted Rand index (ARI) of the whole network is significantly different from zero and thus from random clustering. The ARI of the largest connected component (LCC), however, is close to zero due to the different clusterings in the LCC.
\item
  The two GCD values are significantly different from zero, indicating substantial differences in the overall network structures.
\item
  All nodes are also tested for having significantly different centrality (only the five nodes with the highest absolute difference are shown in the summary). For \(\alpha=0.05\), some nodes have a significantly different closeness centrality, and for \(\alpha=0.1\) also a significantly different eigenvector centrality. Most of these nodes have a high centrality in the one group, but are not connected in the other group.
\end{itemize}

\hypertarget{differential-association-analysis}{%
\section{Differential association analysis}\label{differential-association-analysis}}

The \texttt{diffnet()} function provides statistical tests to assess whether the associations themselves are significantly different between the two groups. \texttt{NetCoMi} also provides a plot function to generate a differential network, where two nodes are connected if they are differentially associated between the groups.

Since we have already computed the permutation association matrices before, we can reuse them here (argument \texttt{fileLoadAssoPerm}).

The local false discovery rate is controlled at level 0.2 to account for multiplicity.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{spring\_diffnet }\OtherTok{\textless{}{-}} \FunctionTok{diffnet}\NormalTok{(spring\_net\_diet,}
                          \AttributeTok{diffMethod =} \StringTok{"perm"}\NormalTok{,}
                          \AttributeTok{fileLoadAssoPerm =} \StringTok{"general/network\_data/spring\_assoPerm"}\NormalTok{,}
                          \AttributeTok{adjust =} \StringTok{"lfdr"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(spring\_diffnet}\SpecialCharTok{$}\NormalTok{pAdjustVec }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sum}\NormalTok{(spring\_diffnet}\SpecialCharTok{$}\NormalTok{pvalsVec }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 12
\end{verbatim}

Some of the unadjusted p-values are below the usual 5\% significance level. However, none of the differences remain significant after adjusting for multiple testing so that the differential network would be empty.

To demonstrate the interpretation of a differential network, we set \texttt{adjust} to ``none'', which is actually statistically incorrect.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{spring\_diffnet\_unadj }\OtherTok{\textless{}{-}} \FunctionTok{diffnet}\NormalTok{(spring\_net\_diet,}
                                \AttributeTok{pvalsVec =}\NormalTok{ spring\_diffnet}\SpecialCharTok{$}\NormalTok{pvalsVec,}
                                \AttributeTok{diffMethod =} \StringTok{"perm"}\NormalTok{,}
                                \AttributeTok{alpha =} \FloatTok{0.05}\NormalTok{,}
                                \AttributeTok{adjust =} \StringTok{"none"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The \texttt{diffnet} object it now plotted using NetCoMi's plot function.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(spring\_diffnet\_unadj, }
     \AttributeTok{cexLabels =} \DecValTok{2}\NormalTok{,}
     \AttributeTok{cexNodes =} \FloatTok{0.7}\NormalTok{, }
     \AttributeTok{cexLegend =} \FloatTok{2.5}\NormalTok{,}
     \AttributeTok{cexTitle =} \DecValTok{3}\NormalTok{,}
     \AttributeTok{mar =} \FunctionTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{5}\NormalTok{,}\DecValTok{15}\NormalTok{),}
     \AttributeTok{legendGroupnames =} \FunctionTok{c}\NormalTok{(}\StringTok{"Mixed diet"}\NormalTok{, }\StringTok{"Vegetarian"}\NormalTok{),}
     \AttributeTok{legendPos =} \FunctionTok{c}\NormalTok{(}\FloatTok{1.2}\NormalTok{,}\FloatTok{1.5}\NormalTok{),}
     \AttributeTok{legendArgs =} \FunctionTok{list}\NormalTok{(}\AttributeTok{lwd =} \DecValTok{4}\NormalTok{),}
     \AttributeTok{fade =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{61_network_comparison_files/figure-latex/diffnet_plot-1.pdf}

Edge colors represent the direction of the associations in the two groups. For example, if two OTUs are positively correlated in the mixed diet group and uncorrelated in the vegetarian group (such as Serratia and Citrobacter), the edge color is dark green.

\hypertarget{netcomp-methods}{%
\section{Network comparison methods}\label{netcomp-methods}}

While many approaches exist for the detection of \textbf{differential correlations}, e.g. \citep{yu2019new, mckenzie2016dgca, siska2017differential}, the literature on the more general case of \textbf{differential association} detection is scarce. \citet{bhuva2019differential} compare various methods in a simulation study, which again includes many differential correlation approaches, but also more general methods such as latent differential graphical models. \citet{gill2010statistical} introduce an approach to analyze whether the connectivity of individual nodes is different between two groups using permutation tests, which is applicable to any kind of association. \citet{he2019statistical} propose a test to infer the differential network structure for two conditional dependence networks.

Performing \textbf{differential network analysis} is challenging because network measures do not follow classical statistical distributions. \citet{shojaie2021differential} provide an overview of differential network analysis methods, but focus only on changes in edge sets. \citet{lichtblau2017comparative} compare differential network analysis methods that incorporate multiple local and global network measures. \citet{jardim2019bionetstat} present a tool ``BioNetStat'' for differential analysis of biological networks, which is able to compare certain network measures between groups.

The \href{https://github.com/stefpeschel/NetCoMi}{NetCoMi} package used for network comparison in this chapter includes the following \textbf{differential network analysis approaches:}:

\begin{itemize}
\tightlist
\item
  \textbf{Permutation approach} to test global network measures (e.g., transitivity, connectivity, or average path length) as well as centrality measures for group differences.
\item
  \textbf{Jaccard} index to assess the similarity between sets of most central nodes
\item
  \textbf{Adjusted Rand index} to assess the similarity between clusterings
\item
  \textbf{Graphlet Correlation Distance} (GCM)
\end{itemize}

See \citep{peschel2021netcomi} for an explanation of the first three approaches. The GCM was proposed by \citet{yaverouglu2014revealing}.

Two methods (Fisher's z-test \citep{fisher1970statistical} and the Discordant method \citep{siska2016discordant}) are available for identifying differential correlations, and \textbf{permutation tests} for the more general case of identifying \textbf{differential associations}. See \citep{peschel2021netcomi} for details. NetCoMi offers also a function for plotting a differential network.

\hypertarget{machine_learning}{%
\chapter{Machine Learning}\label{machine_learning}}

Machine learning (ML) is a part of artificial intelligence. There are multiple
definitions, but ``machine'' refers to computation and ``learning'' to improving
performance based on the data by finding patterns from it. Machine learning
includes wide variety of methods from simple statistical methods to more
complex methods such as neural-networks.

Machine learning can be divided into supervised and unsupervised machine learning.
Supervised ML is used to predict outcome based on the data. Unsupervised ML is used,
for example, to reduce dimensionality (e.g.~PCA) and to find clusters from the
data (e.g., k-means clustering).

\hypertarget{supervised-machine-learning}{%
\section{Supervised machine learning}\label{supervised-machine-learning}}

``Supervised'' means that the training data is introduced before. The training data
contains labels (e.g., patient status), and the model is fitted based on the
training data. After fitting, the model is utilized to predict labels of data whose
labels are not known.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}

\CommentTok{\# Load experimental data}
\FunctionTok{data}\NormalTok{(peerj13075, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ peerj13075}
\end{Highlighting}
\end{Shaded}

Let's first preprocess the data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate data}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\AttributeTok{rank =} \StringTok{"order"}\NormalTok{)}

\CommentTok{\# Apply CLR transform}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{,}
                       \AttributeTok{MARGIN=}\StringTok{"samples"}\NormalTok{, }\AttributeTok{pseudocount=}\DecValTok{1}\NormalTok{)}

\CommentTok{\# Get assay}
\NormalTok{assay }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse, }\StringTok{"clr"}\NormalTok{)}
\CommentTok{\# Transpose assay}
\NormalTok{assay }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(assay)}

\CommentTok{\# Convert into data.frame}
\NormalTok{df }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(assay)}

\CommentTok{\# Add labels to assay}
\NormalTok{labels }\OtherTok{\textless{}{-}} \FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Diet}
\NormalTok{labels }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(labels)}
\NormalTok{df}\SpecialCharTok{$}\NormalTok{diet }\OtherTok{\textless{}{-}}\NormalTok{ labels }

\NormalTok{df[}\DecValTok{5}\NormalTok{, }\DecValTok{5}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] -0.4612
\end{verbatim}

In the example below, we use \href{https://journals.asm.org/doi/10.1128/mBio.00434-20}{mikropml}
package. We try to predict the diet type based on the data.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mikropml)}

\CommentTok{\# Run random forest }
\NormalTok{results }\OtherTok{\textless{}{-}} \FunctionTok{run\_ml}\NormalTok{(df, }\StringTok{"rf"}\NormalTok{, }\AttributeTok{outcome\_colname =} \StringTok{"diet"}\NormalTok{, }
                  \AttributeTok{kfold =} \DecValTok{2}\NormalTok{, }\AttributeTok{cv\_times =} \DecValTok{5}\NormalTok{, }\AttributeTok{training\_frac =} \FloatTok{0.8}\NormalTok{)}

\CommentTok{\# Print result}
\FunctionTok{confusionMatrix}\NormalTok{(}\AttributeTok{data =}\NormalTok{ results}\SpecialCharTok{$}\NormalTok{trained\_model}\SpecialCharTok{$}\NormalTok{finalModel}\SpecialCharTok{$}\NormalTok{predicted, }
                \AttributeTok{reference =}\NormalTok{ results}\SpecialCharTok{$}\NormalTok{trained\_model}\SpecialCharTok{$}\NormalTok{finalModel}\SpecialCharTok{$}\NormalTok{y)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Mixed Veg
##      Mixed    12   9
##      Veg      11  15
##                                         
##                Accuracy : 0.574         
##                  95% CI : (0.422, 0.717)
##     No Information Rate : 0.511         
##     P-Value [Acc > NIR] : 0.233         
##                                         
##                   Kappa : 0.147         
##                                         
##  Mcnemar's Test P-Value : 0.823         
##                                         
##             Sensitivity : 0.522         
##             Specificity : 0.625         
##          Pos Pred Value : 0.571         
##          Neg Pred Value : 0.577         
##              Prevalence : 0.489         
##          Detection Rate : 0.255         
##    Detection Prevalence : 0.447         
##       Balanced Accuracy : 0.573         
##                                         
##        'Positive' Class : Mixed         
## 
\end{verbatim}

mikropml offers easier interface to \href{https://cran.r-project.org/web/packages/caret/index.html}{caret}
package. However, we can also use it directly.

Let's use xgboost model which is another commonly used algorithm in bioinformatics.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Set seed for reproducibility}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{6358}\NormalTok{)}

\CommentTok{\# Specify train control}
\NormalTok{train\_control }\OtherTok{\textless{}{-}} \FunctionTok{trainControl}\NormalTok{(}\AttributeTok{method =} \StringTok{"cv"}\NormalTok{, }\AttributeTok{number =} \DecValTok{5}\NormalTok{,}
                              \AttributeTok{classProbs =} \ConstantTok{TRUE}\NormalTok{, }
                              \AttributeTok{savePredictions =} \StringTok{"final"}\NormalTok{,}
                              \AttributeTok{allowParallel =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Specify hyperparameter tuning grid}
\NormalTok{tune\_grid }\OtherTok{\textless{}{-}} \FunctionTok{expand.grid}\NormalTok{(}\AttributeTok{nrounds =} \FunctionTok{c}\NormalTok{(}\DecValTok{50}\NormalTok{, }\DecValTok{100}\NormalTok{, }\DecValTok{200}\NormalTok{),}
                         \AttributeTok{max\_depth =} \FunctionTok{c}\NormalTok{(}\DecValTok{6}\NormalTok{, }\DecValTok{8}\NormalTok{, }\DecValTok{10}\NormalTok{),}
                         \AttributeTok{colsample\_bytree =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.6}\NormalTok{, }\FloatTok{0.8}\NormalTok{, }\DecValTok{1}\NormalTok{),}
                         \AttributeTok{eta =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.1}\NormalTok{, }\FloatTok{0.3}\NormalTok{),}
                         \AttributeTok{gamma =} \DecValTok{0}\NormalTok{,}
                         \AttributeTok{min\_child\_weight =} \FunctionTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{5}\NormalTok{),}
                         \AttributeTok{subsample =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.6}\NormalTok{, }\FloatTok{0.8}\NormalTok{)}
\NormalTok{                         )}

\CommentTok{\# Train the model, use LOOCV to evaluate performance}
\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{train}\NormalTok{(}\AttributeTok{x =}\NormalTok{ assay, }
               \AttributeTok{y =}\NormalTok{ labels, }
               \AttributeTok{method =} \StringTok{"xgbTree"}\NormalTok{,}
               \AttributeTok{objective =} \StringTok{"binary:logistic"}\NormalTok{,}
               \AttributeTok{trControl =}\NormalTok{ train\_control,}
               \AttributeTok{tuneGrid =}\NormalTok{ tune\_grid,}
               \AttributeTok{metric =} \StringTok{"AUC"}\NormalTok{,}
               \AttributeTok{verbosity =} \DecValTok{0}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Let's create ROC curve which is a commonly used method in binary classification.
For unbalanced data, you might want to plot precision-recall curve.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(MLeval)}

\CommentTok{\# Calculate different evaluation metrics}
\NormalTok{res }\OtherTok{\textless{}{-}} \FunctionTok{evalm}\NormalTok{(model, }\AttributeTok{showplots =} \ConstantTok{FALSE}\NormalTok{)}

\CommentTok{\# Use patchwork to plot ROC and precision{-}recall curve side{-}by{-}side}
\FunctionTok{library}\NormalTok{(patchwork)}
\NormalTok{res}\SpecialCharTok{$}\NormalTok{roc }\SpecialCharTok{+}\NormalTok{ res}\SpecialCharTok{$}\NormalTok{proc }\SpecialCharTok{+} 
    \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{) }\SpecialCharTok{\&} \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position =} \StringTok{\textquotesingle{}bottom\textquotesingle{}}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{40_machine_learning_files/figure-latex/super5-1.pdf}

\hypertarget{unsupervised-machine-learning}{%
\section{Unsupervised machine learning}\label{unsupervised-machine-learning}}

``Unsupervised'' means that the labels (e.g., patient status is not known),
and patterns are learned based only the abundance table, for instance.
Unsupervised ML is also known as a data mining where patterns are extracted
from big datasets.

For unsupervised machine learning, please refer to chapters that are listed below:

\begin{itemize}
\tightlist
\item
  Chapter \ref{clustering}
\item
  Chapter \ref{community-similarity}
\end{itemize}

\hypertarget{multi-assay-analyses}{%
\chapter{Multi-Assay Analyses}\label{multi-assay-analyses}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\end{Highlighting}
\end{Shaded}

Multi-omics approaches integrate data from multiple sources. For
example, we can integrate taxonomic abundance profiles with
metabolomic or other biomolecular profiling data to observe
associations, make predictions, or aim at causal
inferences. Integrating evidence across multiple sources can lead to
enhanced predictions, more holistic understanding, or facilitate the
discovery of novel biomarkers. In this section we demonstrate common
multi-assay data integration tasks.

Cross-correlation analysis is a straightforward approach that can
reveal strength and type of assocations between data sets. For instance,
we can analyze if higher presence of a specific taxon relates to higher
levels of a biomolecule.

The analyses can be facilitated by the multi-assay data containers,
\emph{TreeSummarizedExperiment} and \emph{MultiAssayExperiment}. These are
scalable and contain different types of data in a single container,
making this framework particularly suited for multi-assay microbiome
data incorporating different types of complementary data sources in a
single reproducible workflow. Solutions to a number of data
integration problems are discussed in more detail in Section
\ref{containers}. Another experiment can be stored in \emph{altExp} slot
of SE data container. Alternatively, both experiments can be stored
side-by-side in a MAE data container (see sections \ref{alt-exp} and \ref{mae}
to learn more about altExp and MAE objects, respectively). Different
experiments are first imported as single-assay data containers similarly to
the case when only one experiment is present. After that, the
different experiments can be combined into one multi-assay data
container. The result is a MAE object with multiple experiments in its
experiment slot, or a TreeSE object with alternative experiments in
the altExp slot.

As an example, we use a dataset from the following publication:
\citeyearpar{Hintikka2021} Xylo-oligosaccharides in prevention of hepatic
steatosis and adipose tissue inflammation: associating taxonomic and
metabolomic patterns in fecal microbiota with biclustering.
In this study, mice were fed either with a high-fat or a low-fat diet,
and with or without prebiotics, for the purpose studying whether prebiotics
attenuate the negative impact of a high-fat diet on health.

This example data can be loaded from microbiomeDataSets. The data is
already in MAE format. It includes three different experiments:
microbial abundance data, metabolite concentrations, and data about
different biomarkers. If you like to construct the same data object from the
original files instead, \href{https://microbiome.github.io/OMA/containers.html\#loading-experimental-microbiome-data}{here}
you can find help for importing data into an SE object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load the data}
\FunctionTok{data}\NormalTok{(HintikkaXOData, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{mae }\OtherTok{\textless{}{-}}\NormalTok{ HintikkaXOData}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(stringr)}
\CommentTok{\# Drop off those bacteria that do not include information in Phylum or lower levels}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ mae[[}\DecValTok{1}\NormalTok{]][}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{Phylum), ]}
\CommentTok{\# Clean taxonomy data, so that names do not include additional characters}
\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]) }\OtherTok{\textless{}{-}} \FunctionTok{DataFrame}\NormalTok{(}\FunctionTok{apply}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]), }\DecValTok{2}\NormalTok{, }
\NormalTok{                                     str\_remove, }\AttributeTok{pattern =} \StringTok{".\_[0{-}9]\_\_"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Available alternative experiments}
\FunctionTok{experiments}\NormalTok{(mae)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ExperimentList class object of length 3:
##  [1] microbiota: TreeSummarizedExperiment with 12613 rows and 40 columns
##  [2] metabolites: TreeSummarizedExperiment with 38 rows and 40 columns
##  [3] biomarkers: TreeSummarizedExperiment with 39 rows and 40 columns
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Microbiome data}
\FunctionTok{getWithColData}\NormalTok{(mae, }\StringTok{"microbiota"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 12613 40 
## metadata(0):
## assays(1): counts
## rownames(12613): GAYR01026362.62.2014 CVJT01000011.50.2173 ...
##   JRJTB:03787:02429 JRJTB:03787:02478
## rowData names(7): Phylum Class ... Species OTU
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Metabolite data}
\FunctionTok{getWithColData}\NormalTok{(mae, }\StringTok{"metabolites"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 38 40 
## metadata(0):
## assays(1): nmr
## rownames(38): Butyrate Acetate ... Malonate 1,3-dihydroxyacetone
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Biomarker data}
\FunctionTok{getWithColData}\NormalTok{(mae, }\StringTok{"biomarkers"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: TreeSummarizedExperiment 
## dim: 39 40 
## metadata(0):
## assays(1): signals
## rownames(39): Triglycerides_liver CLSs_epi ... NPY_serum Glycogen_liver
## rowData names(0):
## colnames(40): C1 C2 ... C39 C40
## colData names(6): Sample Rat ... Fat XOS
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL
\end{verbatim}

\hypertarget{cross-correlation}{%
\section{Cross-correlation Analysis}\label{cross-correlation}}

Next we can perform a cross-correlation analysis. Let us analyze if
individual bacteria genera are correlated with concentrations of
individual metabolites. This helps to answer the following question: ``If
bacterium X is present, is the concentration of metabolite Y lower or higher''?

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate microbiome data at family level}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByPrevalence}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{rank =} \StringTok{"Family"}\NormalTok{)}
\CommentTok{\# Does log10 transform for microbiome data}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{method =} \StringTok{"log10"}\NormalTok{, }\AttributeTok{pseudocount =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Give unique names so that we do not have problems when we are creating a plot}
\FunctionTok{rownames}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]) }\OtherTok{\textless{}{-}} \FunctionTok{getTaxonomyLabels}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}

\CommentTok{\# Cross correlates data sets}
\NormalTok{correlations }\OtherTok{\textless{}{-}} \FunctionTok{testExperimentCrossCorrelation}\NormalTok{(mae, }
                                               \AttributeTok{experiment1 =} \DecValTok{1}\NormalTok{,}
                                               \AttributeTok{experiment2 =} \DecValTok{2}\NormalTok{,}
                                               \AttributeTok{assay.type1 =} \StringTok{"log10"}\NormalTok{, }
                                               \AttributeTok{assay.type2 =} \StringTok{"nmr"}\NormalTok{,}
                                               \AttributeTok{method =} \StringTok{"spearman"}\NormalTok{, }
                                               \AttributeTok{p\_adj\_threshold =} \ConstantTok{NULL}\NormalTok{,}
                                               \AttributeTok{cor\_threshold =} \ConstantTok{NULL}\NormalTok{,}
                                               \CommentTok{\# Remove when mia is fixed}
                                               \AttributeTok{mode =} \StringTok{"matrix"}\NormalTok{,}
                                               \AttributeTok{sort =} \ConstantTok{TRUE}\NormalTok{,}
                                               \AttributeTok{show\_warnings =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Next, we create a heatmap depicting all cross-correlations between bacterial
genera and metabolite concentrations.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(ComplexHeatmap) }

\CommentTok{\# Create a heatmap and store it}
\NormalTok{plot }\OtherTok{\textless{}{-}} \FunctionTok{Heatmap}\NormalTok{(correlations}\SpecialCharTok{$}\NormalTok{cor,}
                \CommentTok{\# Print values to cells}
                \AttributeTok{cell\_fun =} \ControlFlowTok{function}\NormalTok{(j, i, x, y, width, height, fill) \{}
                    \CommentTok{\# If the p{-}value is under threshold}
                    \ControlFlowTok{if}\NormalTok{( }\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(correlations}\SpecialCharTok{$}\NormalTok{p\_adj[i, j]) }\SpecialCharTok{\&}\NormalTok{ correlations}\SpecialCharTok{$}\NormalTok{p\_adj[i, j] }\SpecialCharTok{\textless{}} \FloatTok{0.05}\NormalTok{ )\{}
                        \CommentTok{\# Print "X"}
                        \FunctionTok{grid.text}\NormalTok{(}\FunctionTok{sprintf}\NormalTok{(}\StringTok{"\%s"}\NormalTok{, }\StringTok{"X"}\NormalTok{), x, y, }\AttributeTok{gp =} \FunctionTok{gpar}\NormalTok{(}\AttributeTok{fontsize =} \DecValTok{10}\NormalTok{, }\AttributeTok{col =} \StringTok{"\#1dff00"}\NormalTok{))}
\NormalTok{                        \}}
\NormalTok{                    \},}
                \AttributeTok{heatmap\_legend\_param =} \FunctionTok{list}\NormalTok{(}\AttributeTok{title =} \StringTok{""}\NormalTok{, }\AttributeTok{legend\_height =} \FunctionTok{unit}\NormalTok{(}\DecValTok{5}\NormalTok{, }\StringTok{"cm"}\NormalTok{))}
\NormalTok{                )}
\NormalTok{plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{23_multi-assay_analyses_files/figure-latex/cross-correlation6-1.pdf}

\hypertarget{mofa}{%
\section{Multi-Omics Factor Analysis}\label{mofa}}

Multi-Omics Factor Analysis (MOFA) is an unsupervised method for integrating multi-omic data sets in a downstream analysis \citep{Argelaguet2018}. It could be
seen as a generalization of principal component analysis. Yet, with the ability
to infer a latent (low-dimensional) representation, shared among the multiple
(-omics) data sets in hand.

We use the R \href{https://biofam.github.io/MOFA2/index.html}{MOFA2}
package for the analysis, and
\href{https://biofam.github.io/MOFA2/installation.html}{install} the
corresponding dependencies.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(MOFA2)}

\CommentTok{\# For inter{-}operability between Python and R, and setting Python dependencies,}
\CommentTok{\# reticulate package is needed}
\FunctionTok{library}\NormalTok{(reticulate)}
\CommentTok{\# Let us assume that these have been installed already.}
\CommentTok{\#reticulate::install\_miniconda(force = TRUE)}
\CommentTok{\#reticulate::use\_miniconda(condaenv = "env1", required = FALSE)}
\CommentTok{\#reticulate::py\_install(packages = c("mofapy2"), pip = TRUE, python\_version=3.6)}
\end{Highlighting}
\end{Shaded}

The \texttt{mae} object could be used straight to create the MOFA model. Yet,
we transform our assays since the model assumes normality per
default. We can also use Poisson or Bernoulli distributions among others.

Note that duplicates, such as ``uncultured'', might appear when aggregating the microbiome data by a taxonomic rank. To check for duplicates, run \texttt{any(duplicated(rownames(mae{[}{[}1{]}{]})))}. If it returns \texttt{TRUE}, then the
duplicates are present. We can add \texttt{rownames(mae{[}{[}1{]}{]})\ \textless{}-\ getTaxonomyLabels(mae{[}{[}1{]}{]},\ make\_unique=TRUE)} to remove them.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(MOFA2)}
\CommentTok{\# For simplicity, classify all high{-}fat diets as high{-}fat, and all the low{-}fat }
\CommentTok{\# diets as low{-}fat diets}
\FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet }\SpecialCharTok{==} \StringTok{"High{-}fat"} \SpecialCharTok{|} 
                              \FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet }\SpecialCharTok{==} \StringTok{"High{-}fat + XOS"}\NormalTok{, }
                            \StringTok{"High{-}fat"}\NormalTok{, }\StringTok{"Low{-}fat"}\NormalTok{)}

\CommentTok{\# Transforming microbiome data with rclr}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{, }\AttributeTok{method =} \StringTok{"rclr"}\NormalTok{)}

\CommentTok{\# Transforming metabolomic data with log10}
\NormalTok{mae[[}\DecValTok{2}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{2}\NormalTok{]], }\AttributeTok{assay.type =} \StringTok{"nmr"}\NormalTok{,}
                            \AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{,}
                            \AttributeTok{method =} \StringTok{"log10"}\NormalTok{)}

\CommentTok{\# Transforming biomarker data with z{-}transform}
\NormalTok{mae[[}\DecValTok{3}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(mae[[}\DecValTok{3}\NormalTok{]], }\AttributeTok{assay.type =} \StringTok{"signals"}\NormalTok{,}
                           \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{,}
                           \AttributeTok{method =} \StringTok{"z"}\NormalTok{, }\AttributeTok{pseudocount =} \DecValTok{1}\NormalTok{)}

\CommentTok{\# Removing assays no longer needed}
\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\StringTok{"counts"}\NormalTok{) }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\StringTok{"log10"}\NormalTok{) }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{2}\NormalTok{]], }\StringTok{"nmr"}\NormalTok{) }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\FunctionTok{assay}\NormalTok{(mae[[}\DecValTok{3}\NormalTok{]], }\StringTok{"signals"}\NormalTok{) }\OtherTok{\textless{}{-}} \ConstantTok{NULL}

\CommentTok{\# Building our mofa model}
\NormalTok{model }\OtherTok{\textless{}{-}} \FunctionTok{create\_mofa\_from\_MultiAssayExperiment}\NormalTok{(mae,}
                                               \AttributeTok{groups =} \StringTok{"Diet"}\NormalTok{, }
                                               \AttributeTok{extract\_metadata =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{model}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Untrained MOFA model with the following characteristics: 
##  Number of views: 3 
##  Views names: microbiota metabolites biomarkers 
##  Number of features (per view): 39 38 39 
##  Number of groups: 2 
##  Groups names: High-fat Low-fat 
##  Number of samples (per group): 20 20 
## 
\end{verbatim}

Model options can be defined as follows:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{model\_opts }\OtherTok{\textless{}{-}} \FunctionTok{get\_default\_model\_options}\NormalTok{(model)}
\NormalTok{model\_opts}\SpecialCharTok{$}\NormalTok{num\_factors }\OtherTok{\textless{}{-}} \DecValTok{5}
\FunctionTok{head}\NormalTok{(model\_opts)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $likelihoods
##  microbiota metabolites  biomarkers 
##  "gaussian"  "gaussian"  "gaussian" 
## 
## $num_factors
## [1] 5
## 
## $spikeslab_factors
## [1] FALSE
## 
## $spikeslab_weights
## [1] FALSE
## 
## $ard_factors
## [1] TRUE
## 
## $ard_weights
## [1] TRUE
\end{verbatim}

Training options for the model are defined in the following way:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{train\_opts }\OtherTok{\textless{}{-}} \FunctionTok{get\_default\_training\_options}\NormalTok{(model)}
\FunctionTok{head}\NormalTok{(train\_opts)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $maxiter
## [1] 1000
## 
## $convergence_mode
## [1] "fast"
## 
## $drop_factor_threshold
## [1] -1
## 
## $verbose
## [1] FALSE
## 
## $startELBO
## [1] 1
## 
## $freqELBO
## [1] 5
\end{verbatim}

The model is then prepared with \texttt{prepare\_mofa} and trained with \texttt{run\_mofa}:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{model.prepared }\OtherTok{\textless{}{-}} \FunctionTok{prepare\_mofa}\NormalTok{(}
  \AttributeTok{object =}\NormalTok{ model,}
  \AttributeTok{model\_options =}\NormalTok{ model\_opts}
\NormalTok{)}

\CommentTok{\# Some systems may require the specification \textasciigrave{}use\_basilisk = TRUE\textasciigrave{}}
\CommentTok{\# so it has been added to the following code}
\NormalTok{model.trained }\OtherTok{\textless{}{-}} \FunctionTok{run\_mofa}\NormalTok{(model.prepared, }\AttributeTok{use\_basilisk =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##         #########################################################
##         ###           __  __  ____  ______                    ### 
##         ###          |  \/  |/ __ \|  ____/\    _             ### 
##         ###          | \  / | |  | | |__ /  \ _| |_           ### 
##         ###          | |\/| | |  | |  __/ /\ \_   _|          ###
##         ###          | |  | | |__| | | / ____ \|_|            ###
##         ###          |_|  |_|\____/|_|/_/    \_\              ###
##         ###                                                   ### 
##         ######################################################### 
##        
##  
##         
## use_float32 set to True: replacing float64 arrays by float32 arrays to speed up computations...
## 
## Successfully loaded view='microbiota' group='High-fat' with N=20 samples and D=39 features...
## Successfully loaded view='microbiota' group='Low-fat' with N=20 samples and D=39 features...
## Successfully loaded view='metabolites' group='High-fat' with N=20 samples and D=38 features...
## Successfully loaded view='metabolites' group='Low-fat' with N=20 samples and D=38 features...
## Successfully loaded view='biomarkers' group='High-fat' with N=20 samples and D=39 features...
## Successfully loaded view='biomarkers' group='Low-fat' with N=20 samples and D=39 features...
## 
## 
## Model options:
## - Automatic Relevance Determination prior on the factors: True
## - Automatic Relevance Determination prior on the weights: True
## - Spike-and-slab prior on the factors: False
## - Spike-and-slab prior on the weights: False
## Likelihoods:
## - View 0 (microbiota): gaussian
## - View 1 (metabolites): gaussian
## - View 2 (biomarkers): gaussian
## 
## 
## 
## 
## ######################################
## ## Training the model with seed 42 ##
## ######################################
## 
## 
## ELBO before training: -22197.07 
## 
## Iteration 1: time=0.00, ELBO=-4038.83, deltaELBO=18158.238 (81.80467202%), Factors=5
## Iteration 2: time=0.00, Factors=5
## Iteration 3: time=0.00, Factors=5
## Iteration 4: time=0.00, Factors=5
## Iteration 5: time=0.00, Factors=5
## Iteration 6: time=0.00, ELBO=939.45, deltaELBO=4978.275 (22.42762338%), Factors=5
## Iteration 7: time=0.00, Factors=5
## Iteration 8: time=0.00, Factors=5
## Iteration 9: time=0.00, Factors=5
## Iteration 10: time=0.00, Factors=5
## Iteration 11: time=0.00, ELBO=974.43, deltaELBO=34.980 (0.15758673%), Factors=5
## Iteration 12: time=0.00, Factors=5
## Iteration 13: time=0.00, Factors=5
## Iteration 14: time=0.00, Factors=5
## Iteration 15: time=0.00, Factors=5
## Iteration 16: time=0.00, ELBO=980.98, deltaELBO=6.558 (0.02954301%), Factors=5
## Iteration 17: time=0.00, Factors=5
## Iteration 18: time=0.00, Factors=5
## Iteration 19: time=0.00, Factors=5
## Iteration 20: time=0.00, Factors=5
## Iteration 21: time=0.00, ELBO=984.05, deltaELBO=3.071 (0.01383618%), Factors=5
## Iteration 22: time=0.00, Factors=5
## Iteration 23: time=0.00, Factors=5
## Iteration 24: time=0.00, Factors=5
## Iteration 25: time=0.00, Factors=5
## Iteration 26: time=0.00, ELBO=985.89, deltaELBO=1.837 (0.00827635%), Factors=5
## Iteration 27: time=0.00, Factors=5
## Iteration 28: time=0.00, Factors=5
## Iteration 29: time=0.00, Factors=5
## Iteration 30: time=0.00, Factors=5
## Iteration 31: time=0.00, ELBO=987.13, deltaELBO=1.240 (0.00558588%), Factors=5
## Iteration 32: time=0.00, Factors=5
## Iteration 33: time=0.00, Factors=5
## Iteration 34: time=0.00, Factors=5
## Iteration 35: time=0.00, Factors=5
## Iteration 36: time=0.00, ELBO=988.02, deltaELBO=0.893 (0.00402099%), Factors=5
## Iteration 37: time=0.00, Factors=5
## Iteration 38: time=0.00, Factors=5
## Iteration 39: time=0.00, Factors=5
## Iteration 40: time=0.00, Factors=5
## Iteration 41: time=0.00, ELBO=988.70, deltaELBO=0.676 (0.00304483%), Factors=5
## Iteration 42: time=0.00, Factors=5
## Iteration 43: time=0.00, Factors=5
## Iteration 44: time=0.00, Factors=5
## Iteration 45: time=0.00, Factors=5
## Iteration 46: time=0.00, ELBO=989.23, deltaELBO=0.531 (0.00239446%), Factors=5
## Iteration 47: time=0.00, Factors=5
## Iteration 48: time=0.00, Factors=5
## Iteration 49: time=0.00, Factors=5
## Iteration 50: time=0.00, Factors=5
## Iteration 51: time=0.00, ELBO=989.66, deltaELBO=0.428 (0.00192983%), Factors=5
## Iteration 52: time=0.00, Factors=5
## Iteration 53: time=0.00, Factors=5
## Iteration 54: time=0.00, Factors=5
## Iteration 55: time=0.00, Factors=5
## Iteration 56: time=0.00, ELBO=990.01, deltaELBO=0.355 (0.00159733%), Factors=5
## Iteration 57: time=0.00, Factors=5
## Iteration 58: time=0.00, Factors=5
## Iteration 59: time=0.00, Factors=5
## Iteration 60: time=0.00, Factors=5
## Iteration 61: time=0.00, ELBO=990.31, deltaELBO=0.299 (0.00134727%), Factors=5
## Iteration 62: time=0.00, Factors=5
## Iteration 63: time=0.00, Factors=5
## Iteration 64: time=0.00, Factors=5
## Iteration 65: time=0.00, Factors=5
## Iteration 66: time=0.00, ELBO=990.57, deltaELBO=0.260 (0.00117140%), Factors=5
## Iteration 67: time=0.00, Factors=5
## Iteration 68: time=0.00, Factors=5
## Iteration 69: time=0.00, Factors=5
## Iteration 70: time=0.00, Factors=5
## Iteration 71: time=0.00, ELBO=990.80, deltaELBO=0.224 (0.00100732%), Factors=5
## Iteration 72: time=0.00, Factors=5
## Iteration 73: time=0.00, Factors=5
## Iteration 74: time=0.00, Factors=5
## Iteration 75: time=0.00, Factors=5
## Iteration 76: time=0.00, ELBO=991.00, deltaELBO=0.199 (0.00089717%), Factors=5
## Iteration 77: time=0.00, Factors=5
## Iteration 78: time=0.00, Factors=5
## Iteration 79: time=0.00, Factors=5
## Iteration 80: time=0.00, Factors=5
## Iteration 81: time=0.00, ELBO=991.17, deltaELBO=0.176 (0.00079169%), Factors=5
## Iteration 82: time=0.00, Factors=5
## Iteration 83: time=0.00, Factors=5
## Iteration 84: time=0.00, Factors=5
## Iteration 85: time=0.00, Factors=5
## Iteration 86: time=0.00, ELBO=991.33, deltaELBO=0.159 (0.00071840%), Factors=5
## Iteration 87: time=0.00, Factors=5
## Iteration 88: time=0.00, Factors=5
## Iteration 89: time=0.00, Factors=5
## Iteration 90: time=0.00, Factors=5
## Iteration 91: time=0.00, ELBO=991.48, deltaELBO=0.144 (0.00064963%), Factors=5
## Iteration 92: time=0.00, Factors=5
## Iteration 93: time=0.00, Factors=5
## Iteration 94: time=0.00, Factors=5
## Iteration 95: time=0.00, Factors=5
## Iteration 96: time=0.00, ELBO=991.61, deltaELBO=0.133 (0.00059834%), Factors=5
## Iteration 97: time=0.00, Factors=5
## Iteration 98: time=0.00, Factors=5
## Iteration 99: time=0.00, Factors=5
## Iteration 100: time=0.00, Factors=5
## Iteration 101: time=0.00, ELBO=991.73, deltaELBO=0.122 (0.00054741%), Factors=5
## Iteration 102: time=0.00, Factors=5
## Iteration 103: time=0.00, Factors=5
## Iteration 104: time=0.00, Factors=5
## Iteration 105: time=0.00, Factors=5
## Iteration 106: time=0.00, ELBO=991.84, deltaELBO=0.111 (0.00050136%), Factors=5
## Iteration 107: time=0.00, Factors=5
## Iteration 108: time=0.00, Factors=5
## Iteration 109: time=0.00, Factors=5
## Iteration 110: time=0.00, Factors=5
## Iteration 111: time=0.00, ELBO=991.94, deltaELBO=0.103 (0.00046562%), Factors=5
## Iteration 112: time=0.00, Factors=5
## Iteration 113: time=0.00, Factors=5
## Iteration 114: time=0.00, Factors=5
## Iteration 115: time=0.00, Factors=5
## Iteration 116: time=0.00, ELBO=992.04, deltaELBO=0.096 (0.00043145%), Factors=5
## 
## Converged!
## 
## 
## 
## #######################
## ## Training finished ##
## #######################
## 
## 
## Saving model in /tmp/Rtmp44jyLj/mofa_20231203-164529.hdf5...
\end{verbatim}

The explained variance is visualized with the \texttt{plot\_variance\_explained} function:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(patchwork)}
\FunctionTok{library}\NormalTok{(ggplot2)}

\NormalTok{plot\_list }\OtherTok{\textless{}{-}} \FunctionTok{plot\_variance\_explained}\NormalTok{(model.trained,}
                                     \AttributeTok{x =} \StringTok{"view"}\NormalTok{, }\AttributeTok{y =} \StringTok{"factor"}\NormalTok{,}
                                     \AttributeTok{plot\_total =}\NormalTok{ T)}

\FunctionTok{wrap\_plots}\NormalTok{(plot\_list, }\AttributeTok{nrow =} \DecValTok{2}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{plot\_annotation}\NormalTok{(}\AttributeTok{title =} \StringTok{"Variance Explained per factor and assay"}\NormalTok{,}
                  \AttributeTok{theme =} \FunctionTok{theme}\NormalTok{(}\AttributeTok{plot.title =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{hjust =} \FloatTok{0.5}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\includegraphics{23_multi-assay_analyses_files/figure-latex/unnamed-chunk-7-1.pdf}

The top weights for each assay using all five factors:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{custom\_plotter }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(name) \{}
  
\NormalTok{  p }\OtherTok{\textless{}{-}} \FunctionTok{plot\_top\_weights}\NormalTok{(model.trained,}
                        \AttributeTok{view =}\NormalTok{ name,}
                        \AttributeTok{factors =} \StringTok{"all"}\NormalTok{,}
                        \AttributeTok{nfeatures =} \DecValTok{10}\NormalTok{) }\SpecialCharTok{+}
    \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"Top weights of the "}\NormalTok{, name, }\StringTok{" assay"}\NormalTok{))}
  
\NormalTok{\}}

\NormalTok{plot\_list }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"microbiota"}\NormalTok{, }\StringTok{"metabolites"}\NormalTok{, }\StringTok{"biomarkers"}\NormalTok{), custom\_plotter)}

\FunctionTok{wrap\_plots}\NormalTok{(plot\_list, }\AttributeTok{nrow =} \DecValTok{3}\NormalTok{) }\SpecialCharTok{\&} \FunctionTok{theme}\NormalTok{(}\AttributeTok{text =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{8}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\includegraphics{23_multi-assay_analyses_files/figure-latex/unnamed-chunk-8-1.pdf}

More tutorials and examples of using the package are found at \href{https://biofam.github.io/MOFA2/tutorials.html}{link}

\hypertarget{viz-chapter}{%
\chapter{Visualization}\label{viz-chapter}}

Data visualization will inevitably shape interpretation and motivate
the next steps of the analysis. A variety of visualization methods are
available for microbiome analysis but the application requires careful
attention to details. Knowledge on the available tools and their
limitations plays an important role in selecting the most suitable
methods to address a given question.

This chapter introduces the reader to a number of visualization
techniques found in this book, such as:

\begin{itemize}
\tightlist
\item
  barplots
\item
  boxplots
\item
  heatmaps
\item
  ordination charts
\item
  regression charts
\item
  trees
\end{itemize}

The toolkit which provides the essential plotting functionality
includes the following packages:

\begin{itemize}
\tightlist
\item
  \emph{patchwork}, \emph{cowplot}, \emph{ggpubr} and \emph{gridExtra}: plot layout and multi-panel plotting
\item
  \emph{miaViz}: specific visualization tools for \texttt{TreeSummaizedExperiment} objects
\item
  \emph{scater}: specific visualization tools for \texttt{SingleCellExperiment} objects
\item
  \emph{ggplot2}, \emph{pheatmap}, \emph{ggtree}, \emph{sechm}: composition heatmaps
\item
  \emph{ANCOMBC}, \emph{ALDEx2} and \emph{Maaslin2}: visual differential abundance
\item
  \emph{fido}: tree-based methods for differential abundance
\item
  \emph{plotly}: animated and 3D plotting
\end{itemize}

For systematic and extensive tutorials on the visual tools available
in \emph{mia}, readers can refer to the following material:

\begin{itemize}
\tightlist
\item
  \href{https://microbiome.github.io/tutorials/}{microbiome tutorials}
\end{itemize}

\hypertarget{pre-analysis-exploration}{%
\section{Pre-analysis exploration}\label{pre-analysis-exploration}}

\hypertarget{accessing-row-and-column-data}{%
\subsection{Accessing row and column data}\label{accessing-row-and-column-data}}

\texttt{SCE} and \texttt{TreeSE} objects contain multiple layers of information in the
form of rows, columns and meta data. The \emph{scater} package supports in
accessing, modifying and graphing the meta data related to features as
well as samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# list row meta data}
\FunctionTok{names}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# list column meta data}
\FunctionTok{names}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "X.SampleID"               "Primer"                  
## [3] "Final_Barcode"            "Barcode_truncated_plus_T"
## [5] "Barcode_full_length"      "SampleType"              
## [7] "Description"
\end{verbatim}

Such meta data can be directly plotted with the functions
\texttt{plotRowData} and \texttt{plotColData}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# obtain QC data}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{addPerCellQC}\NormalTok{(tse)}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{addPerFeatureQC}\NormalTok{(tse)}
\CommentTok{\# plot QC Mean against Species}
\FunctionTok{plotRowData}\NormalTok{(tse, }\StringTok{"mean"}\NormalTok{, }\StringTok{"Species"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_blank}\NormalTok{()) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Species"}\NormalTok{, }\AttributeTok{y =} \StringTok{"QC Mean"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-3-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# plot QC Sum against Sample ID, colour{-}labeled by Sample Type}
\FunctionTok{plotColData}\NormalTok{(tse, }\StringTok{"sum"}\NormalTok{, }\StringTok{"X.SampleID"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"SampleType"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle =} \DecValTok{45}\NormalTok{, }\AttributeTok{hjust =} \DecValTok{1}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Sample ID"}\NormalTok{, }\AttributeTok{y =} \StringTok{"QC Sum"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-3-2.pdf}

Alternatively, they can be converted to a \texttt{data.frame} object and
passed to \texttt{ggplot}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# store colData into a data frame}
\NormalTok{coldata }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse))}
\CommentTok{\# plot Number of Samples against Sampling Site}
\FunctionTok{ggplot}\NormalTok{(coldata, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ SampleType)) }\SpecialCharTok{+}
  \FunctionTok{geom\_bar}\NormalTok{(}\AttributeTok{width =} \FloatTok{0.5}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle =} \DecValTok{45}\NormalTok{, }\AttributeTok{hjust =} \DecValTok{1}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Sampling Site"}\NormalTok{,}
       \AttributeTok{y =} \StringTok{"Number of Samples"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-4-1.pdf}

Further methods of application can be found in the chapters \ref{qc}
and \ref{richness} and in a few \href{https://github.com/davismcc/scater_tutorials_open_data}{external
tutorials}
with open data. Additionally, \texttt{rowData} and \texttt{colData} allow
manipulation and subsetting of large data sets into smaller units, as
explained in chapter \ref{datamanipulation}.

\hypertarget{viewing-abundance-and-prevalence-patterns}{%
\subsection{Viewing abundance and prevalence patterns}\label{viewing-abundance-and-prevalence-patterns}}

Prior-to-analysis exploration may involve questions such as how microorganisms
are distributed across samples (abundance) and what microorganisms are present
in most of the samples (prevalence). The information on abundance and prevalence
can be summarized into a \textbf{jitter} or \textbf{density plot} and a \textbf{tree},
respectively, with the \emph{miaViz} package.

Specifically, the functions \texttt{plotAbundance}, \texttt{plotAbundanceDensity}
and \texttt{plotRowTree} are used, and examples on their usage are discussed
throughout chapter \ref{quality-control}.

\hypertarget{diversity-estimation}{%
\section{Diversity estimation}\label{diversity-estimation}}

Alpha diversity is commonly measured as one of the diversity indices
explained in chapter \ref{community-diversity}. Because the focus
lies on each sample separately, one-dimensional plots, such as
\textbf{scatter}, \textbf{violin} and \textbf{box plots}, are suitable.

Beta diversity is generally evaluated as one of the dissimilarity
indices reported in chapter \ref{community-similarity}. Unlike alpha
diversity, samples are compared collectively to estimate the
heterogeneity across them, therefore multidimensional plots, such as
\textbf{Shepard} and \textbf{ordination plots} are suitable.

\begin{longtable}[]{@{}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3125}}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3500}}
  >{\centering\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3375}}@{}}
\toprule\noalign{}
\begin{minipage}[b]{\linewidth}\centering
\end{minipage} & \begin{minipage}[b]{\linewidth}\centering
alpha diversity
\end{minipage} & \begin{minipage}[b]{\linewidth}\centering
beta diversity
\end{minipage} \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
used metrics & diversity indices & dissimilarity indices \\
metric dimensionality & one-dimensional & multidimensional \\
suitable visualization & scatter, violin, box plots & Shepard, ordination plots \\
\end{longtable}

In conclusion, visualization techniques for alpha and beta diversity
significantly differ from one another.

\hypertarget{alpha-diversity-with-scatter-violin-and-box-plots}{%
\subsection{Alpha diversity with scatter, violin and box plots}\label{alpha-diversity-with-scatter-violin-and-box-plots}}

The basic method to visualize the diversity values assigned to the
different samples in a \texttt{TSE} object includes the following, where each
data point represents one sample:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# estimate shannon diversity index}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateDiversity}\NormalTok{(tse, }
                              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                              \AttributeTok{index =} \StringTok{"shannon"}\NormalTok{, }
                              \AttributeTok{name =} \StringTok{"shannon"}\NormalTok{)}
\CommentTok{\# plot shannon diversity index, colour{-}labeled by Sample Type}
\FunctionTok{plotColData}\NormalTok{(tse, }\StringTok{"shannon"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"SampleType"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-5-1.pdf}

The several indices available for the evaluation of alpha diversity
often return slightly divergent results, which can be visually
compared with a multiple violin or box plot. For this purpose,
\texttt{plotColData} (for violin plots) or \texttt{ggplot} (for box plots) are
recursively applied to a number of diversity indices with the function
\texttt{lapply} and the multi-panel plotting functionality of the \emph{patchwork}
package is then exploited.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# estimate faith diversity index}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ mia}\SpecialCharTok{::}\FunctionTok{estimateFaith}\NormalTok{(tse,}
                          \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{)}
\CommentTok{\# store colData into a data frame}
\NormalTok{coldata }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse))}
\CommentTok{\# generate plots for shannon and faith indices}
\CommentTok{\# and store them into a list}
\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"shannon"}\NormalTok{, }\StringTok{"faith"}\NormalTok{),}
                \ControlFlowTok{function}\NormalTok{(i) }\FunctionTok{ggplot}\NormalTok{(coldata, }\FunctionTok{aes\_string}\NormalTok{(}\AttributeTok{y =}\NormalTok{ i)) }\SpecialCharTok{+}
                  \FunctionTok{geom\_boxplot}\NormalTok{() }\SpecialCharTok{+}
                  \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.text.x =} \FunctionTok{element\_blank}\NormalTok{(),}
                        \AttributeTok{axis.ticks.x =} \FunctionTok{element\_blank}\NormalTok{()))}
\CommentTok{\# combine plots with patchwork}
\NormalTok{plots[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{+}\NormalTok{ plots[[}\DecValTok{2}\NormalTok{]]}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-6-1.pdf}

The analogous output in the form of a violin plot is obtained in
chapter \ref{faith-diversity}. In addition, box plots that group
samples according to certain information, such as origin, sex, age and
health condition, can be labeled with p-values for significant
differences with the package \emph{ggsignif} package, as shown in chapter
\ref{estimate-diversity}.

\hypertarget{beta-diversity-with-shepard-and-coordination-plots}{%
\subsection{Beta diversity with Shepard and coordination plots}\label{beta-diversity-with-shepard-and-coordination-plots}}

The \emph{scater} package offers the general function \texttt{plotReducedDim}. In
its basic form, it takes a \texttt{TSE} object and the results on sample
similarity stored in the same object, which can be evaluated with the
following coordination methods:

\begin{itemize}
\tightlist
\item
  \texttt{runMDS}
\item
  \texttt{runNMDS}
\item
  \texttt{runPCA}
\item
  \texttt{runTSNE}
\item
  \texttt{runUMAP}
\end{itemize}

Since these clustering techniques allow for multiple coordinates or
components, \textbf{coordination plots} can also span multiple dimensions,
which is explained in chapter \ref{extras}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# perform NMDS coordination method}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runNMDS}\NormalTok{(tse,}
               \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
               \AttributeTok{name =} \StringTok{"NMDS"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## initial  value 47.733208 
## iter   5 value 33.853364
## iter  10 value 32.891200
## final  value 32.823570 
## converged
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# plot results of a 2{-}component NMDS on tse,}
\CommentTok{\# coloured{-}scaled by shannon diversity index}
\FunctionTok{plotReducedDim}\NormalTok{(tse, }\StringTok{"NMDS"}\NormalTok{, }\AttributeTok{colour\_by =} \StringTok{"shannon"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-7-1.pdf}

Multiple combinations of coordinates or dimensions can also be integrated into a multi-panel arrangement.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# perform MDS coordination method}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(tse,}
              \AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist,}
              \AttributeTok{method =} \StringTok{"bray"}\NormalTok{,}
              \AttributeTok{name =} \StringTok{"MDS"}\NormalTok{,}
              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
              \AttributeTok{ncomponents =} \DecValTok{3}\NormalTok{)}
\CommentTok{\# plot results of a 3{-}component MDS on tse,}
\CommentTok{\# coloured{-}scaled by faith diversity index}
\FunctionTok{plotReducedDim}\NormalTok{(tse, }\StringTok{"MDS"}\NormalTok{, }\AttributeTok{ncomponents =} \FunctionTok{c}\NormalTok{(}\DecValTok{1}\SpecialCharTok{:}\DecValTok{3}\NormalTok{), }\AttributeTok{colour\_by =} \StringTok{"faith"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-8-1.pdf}

Similarly to iterating \texttt{plotColData} over indices of alpha diversity,
\texttt{lapply} can be used in combination with \emph{patchwork} to recursively
apply \texttt{plotReducedDim} and visually compare results among various
coordination methods.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# generate plots for MDS and NMDS methods}
\CommentTok{\# and store them into a list}
\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"MDS"}\NormalTok{, }\StringTok{"NMDS"}\NormalTok{),}
\NormalTok{                plotReducedDim,}
                \AttributeTok{object =}\NormalTok{ tse,}
                \AttributeTok{colour\_by =} \StringTok{"shannon"}\NormalTok{)}
\CommentTok{\# combine plots with patchwork}
\NormalTok{plots[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{+}\NormalTok{ plots[[}\DecValTok{2}\NormalTok{]] }\SpecialCharTok{+}
  \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/unnamed-chunk-9-1.pdf}

For similar examples, readers are referred to chapter
\ref{community-similarity}. Further material on the graphic
capabilities of \emph{patchwork} is available in its \href{https://patchwork.data-imaginist.com/articles/patchwork.html}{official package
tutorial}.

\hypertarget{statistical-analysis}{%
\section{Statistical analysis}\label{statistical-analysis}}

\hypertarget{heatmaps}{%
\subsection{Heatmaps}\label{heatmaps}}

As described in chapter \ref{visual-composition}, bar plots and
heatmaps can offer a useful insight into the composition of a
community. Simple methods involve the functions \texttt{plotAbundance} and
\texttt{geom\_tile} in combination with \texttt{scale\_fill\_gradientn} from the
packages \emph{miaViz} and \emph{ggplot2}, respectively.

For instance, below the composition of multiple samples (x axis) is
reported in terms of relative abundances (y axis) for the top 10 taxa
at the Order rank. Bar plots and heatmaps with analogous information
at the Phylum level are available in the aforementioned chapter.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# agglomerate tse by Order}
\NormalTok{tse\_order }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse,}
                                \AttributeTok{rank =} \StringTok{"Order"}\NormalTok{,}
                                \AttributeTok{onRankOnly =} \ConstantTok{TRUE}\NormalTok{)}
\CommentTok{\# transform counts into relative abundance}
\NormalTok{tse\_order }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_order,}
                              \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{,}
                              \AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\CommentTok{\# get top orders}
\NormalTok{top\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse\_order,}
                       \AttributeTok{top =} \DecValTok{10}\NormalTok{,}
                       \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{)}
\CommentTok{\# leave only names for top 10 orders and label the rest with "Other"}
\NormalTok{order\_renamed }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(tse\_order)}\SpecialCharTok{$}\NormalTok{Order,}
                   \ControlFlowTok{function}\NormalTok{(x)\{}\ControlFlowTok{if}\NormalTok{ (x }\SpecialCharTok{\%in\%}\NormalTok{ top\_taxa) \{x\} }\ControlFlowTok{else}\NormalTok{ \{}\StringTok{"Other"}\NormalTok{\}\})}
\FunctionTok{rowData}\NormalTok{(tse\_order)}\SpecialCharTok{$}\NormalTok{Order }\OtherTok{\textless{}{-}} \FunctionTok{as.character}\NormalTok{(order\_renamed)}
\CommentTok{\# plot composition as a bar plot}
\FunctionTok{plotAbundance}\NormalTok{(tse\_order,}
              \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
              \AttributeTok{rank =} \StringTok{"Order"}\NormalTok{,}
              \AttributeTok{order\_rank\_by =} \StringTok{"abund"}\NormalTok{,}
              \AttributeTok{order\_sample\_by =} \StringTok{"Clostridiales"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/plotAbundance1-1.pdf}

To add a sample annotation, you can combine plots that you get from the output
of \emph{plotAbundance}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create plots}
\NormalTok{plots }\OtherTok{\textless{}{-}} \FunctionTok{plotAbundance}\NormalTok{(tse\_order,}
            \AttributeTok{assay.type =} \StringTok{"relabundance"}\NormalTok{,}
        \AttributeTok{rank =} \StringTok{"Order"}\NormalTok{,}
            \AttributeTok{order\_rank\_by =} \StringTok{"abund"}\NormalTok{,}
        \AttributeTok{order\_sample\_by =} \StringTok{"Clostridiales"}\NormalTok{,}
            \AttributeTok{features =} \StringTok{"SampleType"}\NormalTok{)}

\CommentTok{\# Modify the legend of the first plot to be smaller }
\NormalTok{plots[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ plots[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{+}
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.key.size =} \FunctionTok{unit}\NormalTok{(}\FloatTok{0.3}\NormalTok{, }\StringTok{\textquotesingle{}cm\textquotesingle{}}\NormalTok{),}
          \AttributeTok{legend.text =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{6}\NormalTok{),}
          \AttributeTok{legend.title =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{8}\NormalTok{))}

\CommentTok{\# Modify the legend of the second plot to be smaller }
\NormalTok{plots[[}\DecValTok{2}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ plots[[}\DecValTok{2}\NormalTok{]] }\SpecialCharTok{+}
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.key.height =} \FunctionTok{unit}\NormalTok{(}\FloatTok{0.3}\NormalTok{, }\StringTok{\textquotesingle{}cm\textquotesingle{}}\NormalTok{),}
          \AttributeTok{legend.key.width =} \FunctionTok{unit}\NormalTok{(}\FloatTok{0.3}\NormalTok{, }\StringTok{\textquotesingle{}cm\textquotesingle{}}\NormalTok{),}
          \AttributeTok{legend.text =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{6}\NormalTok{),}
          \AttributeTok{legend.title =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{size =} \DecValTok{8}\NormalTok{),}
          \AttributeTok{legend.direction =} \StringTok{"vertical"}\NormalTok{)}

\CommentTok{\# Load required packages}
\FunctionTok{library}\NormalTok{(ggpubr)}
\FunctionTok{library}\NormalTok{(patchwork) }
\CommentTok{\# Combine legends}
\NormalTok{legend }\OtherTok{\textless{}{-}} \FunctionTok{wrap\_plots}\NormalTok{(}\FunctionTok{as\_ggplot}\NormalTok{(}\FunctionTok{get\_legend}\NormalTok{(plots[[}\DecValTok{1}\NormalTok{]])), }\FunctionTok{as\_ggplot}\NormalTok{(}\FunctionTok{get\_legend}\NormalTok{(plots[[}\DecValTok{2}\NormalTok{]])), }\AttributeTok{ncol =} \DecValTok{1}\NormalTok{) }

\CommentTok{\# Remove legends from the plots}
\NormalTok{plots[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ plots[[}\DecValTok{1}\NormalTok{]] }\SpecialCharTok{+} \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position =} \StringTok{"none"}\NormalTok{)}
\NormalTok{plots[[}\DecValTok{2}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ plots[[}\DecValTok{2}\NormalTok{]] }\SpecialCharTok{+} \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.position =} \StringTok{"none"}\NormalTok{, }\AttributeTok{axis.title.x=}\FunctionTok{element\_blank}\NormalTok{()) }

\CommentTok{\# Combine plots}
\NormalTok{plot }\OtherTok{\textless{}{-}} \FunctionTok{wrap\_plots}\NormalTok{(plots[[}\DecValTok{2}\NormalTok{]], plots[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{ncol =} \DecValTok{1}\NormalTok{, }\AttributeTok{heights =} \FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{, }\DecValTok{10}\NormalTok{))}
\CommentTok{\# Combine the plot with the legend}
\FunctionTok{wrap\_plots}\NormalTok{(plot, legend, }\AttributeTok{nrow =} \DecValTok{1}\NormalTok{, }\AttributeTok{widths =} \FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{, }\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/plotAbundance2-1.pdf}

For more sophisticated visualizations than those produced with \texttt{plotAbundance}
and \emph{ggplot2}, the packages \emph{pheatmap} and \emph{sechm} provide methods to include
feature and sample clusters in a heatmap, along with further functionality.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Agglomerate tse by phylum}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse,}
                                \AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{,}
                                \AttributeTok{onRankOnly =} \ConstantTok{TRUE}\NormalTok{)}

\CommentTok{\# Add clr{-}transformation on samples}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum, }\AttributeTok{MARGIN =} \StringTok{"samples"}\NormalTok{, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{pseudocount=}\DecValTok{1}\NormalTok{)}

\CommentTok{\# Add z{-}transformation on features (taxa)}
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum, }\AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{,}
                              \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{, }
                              \AttributeTok{method =} \StringTok{"z"}\NormalTok{, }\AttributeTok{name =} \StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Take subset: only samples from feces, skin, or tongue}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum[ , tse\_phylum}\SpecialCharTok{$}\NormalTok{SampleType }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Feces"}\NormalTok{, }\StringTok{"Skin"}\NormalTok{, }\StringTok{"Tongue"}\NormalTok{) ]}

\CommentTok{\# Add clr{-}transformation}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum\_subset, }\AttributeTok{method =} \StringTok{"clr"}\NormalTok{,}
                                     \AttributeTok{MARGIN=}\StringTok{"samples"}\NormalTok{,}
                                     \AttributeTok{assay.type =} \StringTok{"counts"}\NormalTok{, }\AttributeTok{pseudocount=}\DecValTok{1}\NormalTok{)}
\CommentTok{\# Does z{-}transformation}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(tse\_phylum\_subset, }\AttributeTok{assay.type =} \StringTok{"clr"}\NormalTok{,}
                                     \AttributeTok{MARGIN =} \StringTok{"features"}\NormalTok{, }
                                     \AttributeTok{method =} \StringTok{"z"}\NormalTok{, }\AttributeTok{name =} \StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Get n most abundant taxa, and subsets the data by them}
\NormalTok{top\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse\_phylum\_subset, }\AttributeTok{top =} \DecValTok{20}\NormalTok{)}
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum\_subset[top\_taxa, ]}

\CommentTok{\# Gets the assay table}
\NormalTok{mat }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(tse\_phylum\_subset, }\StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Creates the heatmap}
\FunctionTok{pheatmap}\NormalTok{(mat)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/pheatmap1-1.pdf}

We can cluster both samples and features hierarchically and add them to the
x and y axes of the heatmap, respectively.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Hierarchical clustering}
\NormalTok{taxa\_hclust }\OtherTok{\textless{}{-}} \FunctionTok{hclust}\NormalTok{(}\FunctionTok{dist}\NormalTok{(mat), }\AttributeTok{method =} \StringTok{"complete"}\NormalTok{)}

\CommentTok{\# Creates a phylogenetic tree}
\NormalTok{taxa\_tree }\OtherTok{\textless{}{-}} \FunctionTok{as.phylo}\NormalTok{(taxa\_hclust)}

\CommentTok{\# Plot taxa tree}
\NormalTok{taxa\_tree }\OtherTok{\textless{}{-}} \FunctionTok{ggtree}\NormalTok{(taxa\_tree) }\SpecialCharTok{+} 
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{)) }\CommentTok{\# removes margins}

\CommentTok{\# Get order of taxa in plot}
\NormalTok{taxa\_ordered }\OtherTok{\textless{}{-}} \FunctionTok{get\_taxa\_name}\NormalTok{(taxa\_tree)}

\CommentTok{\# to view the tree, run}
\CommentTok{\# taxa\_tree}
\end{Highlighting}
\end{Shaded}

Based on phylo tree, we decide to create three clusters.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Creates clusters}
\NormalTok{taxa\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{cutree}\NormalTok{(}\AttributeTok{tree =}\NormalTok{ taxa\_hclust, }\AttributeTok{k =} \DecValTok{3}\NormalTok{)}

\CommentTok{\# Converts into data frame}
\NormalTok{taxa\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{clusters =}\NormalTok{ taxa\_clusters)}
\NormalTok{taxa\_clusters}\SpecialCharTok{$}\NormalTok{clusters }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(taxa\_clusters}\SpecialCharTok{$}\NormalTok{clusters)}

\CommentTok{\# Order data so that it\textquotesingle{}s same as in phylo tree}
\NormalTok{taxa\_clusters }\OtherTok{\textless{}{-}}\NormalTok{ taxa\_clusters[taxa\_ordered, , drop }\OtherTok{=} \ConstantTok{FALSE}\NormalTok{] }

\CommentTok{\# Prints taxa and their clusters}
\NormalTok{taxa\_clusters}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                  clusters
## Chloroflexi             3
## Actinobacteria          3
## Crenarchaeota           3
## Planctomycetes          3
## Gemmatimonadetes        3
## Thermi                  3
## Acidobacteria           3
## Spirochaetes            2
## Fusobacteria            2
## SR1                     2
## Cyanobacteria           2
## Proteobacteria          2
## Synergistetes           2
## Lentisphaerae           1
## Bacteroidetes           1
## Verrucomicrobia         1
## Tenericutes             1
## Firmicutes              1
## Euryarchaeota           1
## SAR406                  1
\end{verbatim}

The information on the clusters is then added to the feature meta data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Adds information to rowData}
\FunctionTok{rowData}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{clusters }\OtherTok{\textless{}{-}}\NormalTok{ taxa\_clusters[}\FunctionTok{order}\NormalTok{(}\FunctionTok{match}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(taxa\_clusters), }\FunctionTok{rownames}\NormalTok{(tse\_phylum\_subset))), ]}

\CommentTok{\# Prints taxa and their clusters}
\FunctionTok{rowData}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{clusters}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  [1] 1 1 2 3 2 2 1 1 1 1 3 2 3 3 3 2 2 3 3 1
## Levels: 1 2 3
\end{verbatim}

Similarly, samples are hierarchically grouped into clusters, the most suitable
number of clusters for the plot is selected and the new information is stored
into the sample meta data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Hierarchical clustering}
\NormalTok{sample\_hclust }\OtherTok{\textless{}{-}} \FunctionTok{hclust}\NormalTok{(}\FunctionTok{dist}\NormalTok{(}\FunctionTok{t}\NormalTok{(mat)), }\AttributeTok{method =} \StringTok{"complete"}\NormalTok{)}

\CommentTok{\# Creates a phylogenetic tree}
\NormalTok{sample\_tree }\OtherTok{\textless{}{-}} \FunctionTok{as.phylo}\NormalTok{(sample\_hclust)}

\CommentTok{\# Plot sample tree}
\NormalTok{sample\_tree }\OtherTok{\textless{}{-}} \FunctionTok{ggtree}\NormalTok{(sample\_tree) }\SpecialCharTok{+} \FunctionTok{layout\_dendrogram}\NormalTok{() }\SpecialCharTok{+} 
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{)) }\CommentTok{\# removes margins}

\CommentTok{\# Get order of samples in plot}
\NormalTok{samples\_ordered }\OtherTok{\textless{}{-}} \FunctionTok{rev}\NormalTok{(}\FunctionTok{get\_taxa\_name}\NormalTok{(sample\_tree))}

\CommentTok{\# to view the tree, run}
\CommentTok{\# sample\_tree}

\CommentTok{\# Creates clusters}
\NormalTok{sample\_clusters }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(}\FunctionTok{cutree}\NormalTok{(}\AttributeTok{tree =}\NormalTok{ sample\_hclust, }\AttributeTok{k =} \DecValTok{3}\NormalTok{))}

\CommentTok{\# Converts into data frame}
\NormalTok{sample\_data }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{clusters =}\NormalTok{ sample\_clusters)}

\CommentTok{\# Order data so that it\textquotesingle{}s same as in phylo tree}
\NormalTok{sample\_data }\OtherTok{\textless{}{-}}\NormalTok{ sample\_data[samples\_ordered, , drop }\OtherTok{=} \ConstantTok{FALSE}\NormalTok{] }

\CommentTok{\# Order data based on }
\NormalTok{tse\_phylum\_subset }\OtherTok{\textless{}{-}}\NormalTok{ tse\_phylum\_subset[ , }\FunctionTok{rownames}\NormalTok{(sample\_data)]}

\CommentTok{\# Add sample type data}
\NormalTok{sample\_data}\SpecialCharTok{$}\NormalTok{sample\_types }\OtherTok{\textless{}{-}} \FunctionTok{unfactor}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{SampleType)}

\NormalTok{sample\_data}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##         clusters sample_types
## M11Plmr        2         Skin
## M31Plmr        2         Skin
## F21Plmr        2         Skin
## M31Fcsw        1        Feces
## M11Fcsw        1        Feces
## TS28           3        Feces
## TS29           3        Feces
## M31Tong        3       Tongue
## M11Tong        3       Tongue
\end{verbatim}

Now we can create heatmap with additional annotations.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Determines the scaling of colorss}
\CommentTok{\# Scale colors}
\NormalTok{breaks }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\SpecialCharTok{{-}}\FunctionTok{ceiling}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))), }\FunctionTok{ceiling}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))), }
              \AttributeTok{length.out =} \FunctionTok{ifelse}\NormalTok{( }\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))}\SpecialCharTok{\textgreater{}}\DecValTok{5}\NormalTok{, }\DecValTok{2}\SpecialCharTok{*}\FunctionTok{ceiling}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(mat))), }\DecValTok{10}\NormalTok{ ) )}
\NormalTok{colors }\OtherTok{\textless{}{-}} \FunctionTok{colorRampPalette}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\StringTok{"darkblue"}\NormalTok{, }\StringTok{"blue"}\NormalTok{, }\StringTok{"white"}\NormalTok{, }\StringTok{"red"}\NormalTok{, }\StringTok{"darkred"}\NormalTok{))(}\FunctionTok{length}\NormalTok{(breaks)}\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{)}

\FunctionTok{pheatmap}\NormalTok{(mat, }\AttributeTok{annotation\_row =}\NormalTok{ taxa\_clusters, }
         \AttributeTok{annotation\_col =}\NormalTok{ sample\_data,}
         \AttributeTok{breaks =}\NormalTok{ breaks,}
         \AttributeTok{color =}\NormalTok{ colors)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/pheatmap6-1.pdf}

The package \emph{sechm} allows for further visual capabilities and flexibility.
In this case, the clustering step is automatically performed by the plotting
function and does not need to be executed in advance.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Stores annotation colros to metadata}
\FunctionTok{metadata}\NormalTok{(tse\_phylum\_subset)}\SpecialCharTok{$}\NormalTok{anno\_colors}\SpecialCharTok{$}\NormalTok{SampleType }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\AttributeTok{Feces =} \StringTok{"blue"}\NormalTok{, }
                                                        \AttributeTok{Skin =} \StringTok{"red"}\NormalTok{, }
                                                        \AttributeTok{Tongue =} \StringTok{"gray"}\NormalTok{)}

\CommentTok{\# Create a plot}
\FunctionTok{sechm}\NormalTok{(tse\_phylum\_subset, }
      \AttributeTok{features =} \FunctionTok{rownames}\NormalTok{(tse\_phylum\_subset), }
      \AttributeTok{assayName =} \StringTok{"clr"}\NormalTok{, }
      \AttributeTok{do.scale =} \ConstantTok{TRUE}\NormalTok{, }
      \AttributeTok{top\_annotation =} \FunctionTok{c}\NormalTok{(}\StringTok{"SampleType"}\NormalTok{), }
      \AttributeTok{gaps\_at =} \StringTok{"SampleType"}\NormalTok{,}
      \AttributeTok{cluster\_cols =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{cluster\_rows =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/sechm-1.pdf}

It is also possible to create an analogous heatmap by just using the
\emph{ggplot2} package. However, a relatively long code is required to
generate an identical output.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Add feature names to column as a factor}
\NormalTok{taxa\_clusters}\SpecialCharTok{$}\NormalTok{Feature }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(taxa\_clusters)}
\NormalTok{taxa\_clusters}\SpecialCharTok{$}\NormalTok{Feature }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(taxa\_clusters}\SpecialCharTok{$}\NormalTok{Feature, }\AttributeTok{levels =}\NormalTok{ taxa\_clusters}\SpecialCharTok{$}\NormalTok{Feature)}

\CommentTok{\# Create annotation plot}
\NormalTok{row\_annotation }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(taxa\_clusters) }\SpecialCharTok{+} 
  \FunctionTok{geom\_tile}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{y =}\NormalTok{ Feature, }\AttributeTok{fill =}\NormalTok{ clusters)) }\SpecialCharTok{+}
  \FunctionTok{coord\_equal}\NormalTok{(}\AttributeTok{ratio =} \DecValTok{1}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}
        \AttributeTok{axis.text.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.text.y=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.ticks.y=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.title.y=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.title.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle =} \DecValTok{90}\NormalTok{, }\AttributeTok{vjust =} \FloatTok{0.5}\NormalTok{, }\AttributeTok{hjust=}\DecValTok{1}\NormalTok{),}
        \AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}
\NormalTok{        ) }\SpecialCharTok{+}
      \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"Clusters"}\NormalTok{, }\AttributeTok{x =} \StringTok{"Clusters"}\NormalTok{)}

\CommentTok{\# to view the notation, run}
\CommentTok{\# row\_annotation}

\CommentTok{\# Add sample names to one of the columns}
\NormalTok{sample\_data}\SpecialCharTok{$}\NormalTok{sample }\OtherTok{\textless{}{-}} \FunctionTok{factor}\NormalTok{(}\FunctionTok{rownames}\NormalTok{(sample\_data), }\AttributeTok{levels =} \FunctionTok{rownames}\NormalTok{(sample\_data))}

\CommentTok{\# Create annotation plot}
\NormalTok{sample\_types\_annotation }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(sample\_data) }\SpecialCharTok{+}
  \FunctionTok{scale\_y\_discrete}\NormalTok{(}\AttributeTok{position =} \StringTok{"right"}\NormalTok{, }\AttributeTok{expand =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{geom\_tile}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{y =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{x =}\NormalTok{ sample, }\AttributeTok{fill =}\NormalTok{ sample\_types)) }\SpecialCharTok{+}
  \FunctionTok{coord\_equal}\NormalTok{(}\AttributeTok{ratio =} \DecValTok{1}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}
        \AttributeTok{axis.text.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.text.y=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.title.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.ticks.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}
        \AttributeTok{axis.title.y.right =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle=}\DecValTok{0}\NormalTok{, }\AttributeTok{vjust =} \FloatTok{0.5}\NormalTok{)}
\NormalTok{        ) }\SpecialCharTok{+}
      \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"Sample types"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Sample types"}\NormalTok{)}
\CommentTok{\# to view the notation, run}
\CommentTok{\# sample\_types\_annotation}

\CommentTok{\# Create annotation plot}
\NormalTok{sample\_clusters\_annotation }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(sample\_data) }\SpecialCharTok{+}
  \FunctionTok{scale\_y\_discrete}\NormalTok{(}\AttributeTok{position =} \StringTok{"right"}\NormalTok{, }\AttributeTok{expand =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{geom\_tile}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{y =} \ConstantTok{NA}\NormalTok{, }\AttributeTok{x =}\NormalTok{ sample, }\AttributeTok{fill =}\NormalTok{ clusters)) }\SpecialCharTok{+}
  \FunctionTok{coord\_equal}\NormalTok{(}\AttributeTok{ratio =} \DecValTok{1}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}
        \AttributeTok{axis.text.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.text.y=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.title.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.ticks.x=}\FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}
        \AttributeTok{axis.title.y.right =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle=}\DecValTok{0}\NormalTok{, }\AttributeTok{vjust =} \FloatTok{0.5}\NormalTok{)}
\NormalTok{        ) }\SpecialCharTok{+}
      \FunctionTok{labs}\NormalTok{(}\AttributeTok{fill =} \StringTok{"Clusters"}\NormalTok{, }\AttributeTok{y =} \StringTok{"Clusters"}\NormalTok{)}
\CommentTok{\# to view the notation, run}
\CommentTok{\# sample\_clusters\_annotation}

\CommentTok{\# Order data based on clusters and sample types}
\NormalTok{mat }\OtherTok{\textless{}{-}}\NormalTok{ mat[}\FunctionTok{unfactor}\NormalTok{(taxa\_clusters}\SpecialCharTok{$}\NormalTok{Feature), }\FunctionTok{unfactor}\NormalTok{(sample\_data}\SpecialCharTok{$}\NormalTok{sample)]}

\CommentTok{\# ggplot requires data in melted format}
\NormalTok{melted\_mat }\OtherTok{\textless{}{-}} \FunctionTok{melt}\NormalTok{(mat)}
\FunctionTok{colnames}\NormalTok{(melted\_mat) }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Taxa"}\NormalTok{, }\StringTok{"Sample"}\NormalTok{, }\StringTok{"clr\_z"}\NormalTok{)}

\CommentTok{\# Determines the scaling of colorss}
\NormalTok{maxval }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(}\FunctionTok{max}\NormalTok{(}\FunctionTok{abs}\NormalTok{(melted\_mat}\SpecialCharTok{$}\NormalTok{clr\_z)))}
\NormalTok{limits }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{maxval, maxval)}
\NormalTok{breaks }\OtherTok{\textless{}{-}} \FunctionTok{seq}\NormalTok{(}\AttributeTok{from =} \FunctionTok{min}\NormalTok{(limits), }\AttributeTok{to =} \FunctionTok{max}\NormalTok{(limits), }\AttributeTok{by =} \FloatTok{0.5}\NormalTok{)}
\NormalTok{colours }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"darkblue"}\NormalTok{, }\StringTok{"blue"}\NormalTok{, }\StringTok{"white"}\NormalTok{, }\StringTok{"red"}\NormalTok{, }\StringTok{"darkred"}\NormalTok{)}

\NormalTok{heatmap }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(melted\_mat) }\SpecialCharTok{+} 
  \FunctionTok{geom\_tile}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ Sample, }\AttributeTok{y =}\NormalTok{ Taxa, }\AttributeTok{fill =}\NormalTok{ clr\_z)) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}
    \AttributeTok{axis.title.y=}\FunctionTok{element\_blank}\NormalTok{(),}
    \AttributeTok{axis.title.x=}\FunctionTok{element\_blank}\NormalTok{(),}
    \AttributeTok{axis.ticks.y=}\FunctionTok{element\_blank}\NormalTok{(),}
    \AttributeTok{axis.text.x =} \FunctionTok{element\_text}\NormalTok{(}\AttributeTok{angle =} \DecValTok{90}\NormalTok{, }\AttributeTok{vjust =} \FloatTok{0.5}\NormalTok{, }\AttributeTok{hjust=}\DecValTok{1}\NormalTok{),}
    
    \AttributeTok{plot.margin=}\FunctionTok{margin}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{), }\CommentTok{\# removes margins}
    \AttributeTok{legend.key.height=} \FunctionTok{unit}\NormalTok{(}\DecValTok{1}\NormalTok{, }\StringTok{\textquotesingle{}cm\textquotesingle{}}\NormalTok{)}
\NormalTok{    ) }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_gradientn}\NormalTok{(}\AttributeTok{name =} \StringTok{"CLR + Z transform"}\NormalTok{, }
                       \AttributeTok{breaks =}\NormalTok{ breaks, }
                       \AttributeTok{limits =}\NormalTok{ limits, }
                       \AttributeTok{colours =}\NormalTok{ colours) }\SpecialCharTok{+} 
  \FunctionTok{scale\_y\_discrete}\NormalTok{(}\AttributeTok{position =} \StringTok{"right"}\NormalTok{)}

\NormalTok{heatmap}
\end{Highlighting}
\end{Shaded}

\includegraphics{19_visualization_techniques_files/figure-latex/more_complex_heatmap-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(patchwork)}

\CommentTok{\# Create layout}
\NormalTok{design }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{3}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{1}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{3}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{2}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{3}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{3}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{3}\NormalTok{)}
\NormalTok{)}
\CommentTok{\# to view the design, run}
\CommentTok{\# plot(design)}

\CommentTok{\# Combine plots}
\NormalTok{plot }\OtherTok{\textless{}{-}}\NormalTok{ row\_annotation }\SpecialCharTok{+}\NormalTok{ sample\_clusters\_annotation }\SpecialCharTok{+}
\NormalTok{                         sample\_types\_annotation }\SpecialCharTok{+}
\NormalTok{             heatmap  }\SpecialCharTok{+}
    \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{design =}\NormalTok{ design, }\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{,}
                \CommentTok{\# Specify layout, collect legends}
                
                \CommentTok{\# Adjust widths and heights to align plots.}
                \CommentTok{\# When annotation plot is larger, it might not fit into}
        \CommentTok{\# its column/row.}
                \CommentTok{\# Then you need to make column/row larger.}
                
                \CommentTok{\# Relative widths and heights of each column and row:}
                \CommentTok{\# Currently, the width of the first column is 15 \% and the height of}
                \CommentTok{\# first two rows are 30 \% the size of others}
                
                \CommentTok{\# To get this work most of the times, you can adjust all sizes to be 1, i.e. equal, }
                \CommentTok{\# but then the gaps between plots are larger.}
                \AttributeTok{widths =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.15}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{),}
                \AttributeTok{heights =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.3}\NormalTok{, }\FloatTok{0.3}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{))}

\CommentTok{\# plot}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Create layout}
\NormalTok{design }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{4}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{5}\NormalTok{, }\DecValTok{1}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{4}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{5}\NormalTok{, }\DecValTok{2}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{4}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{2}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{4}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{3}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{4}\NormalTok{),}
\NormalTok{  patchwork}\SpecialCharTok{::}\FunctionTok{area}\NormalTok{(}\DecValTok{4}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{5}\NormalTok{, }\DecValTok{4}\NormalTok{)}
\NormalTok{)}

\CommentTok{\# to view the design, run}
\CommentTok{\# plot(design)}

\CommentTok{\# Combine plots}
\NormalTok{plot }\OtherTok{\textless{}{-}}\NormalTok{ taxa\_tree }\SpecialCharTok{+} 
\NormalTok{  row\_annotation }\SpecialCharTok{+}
\NormalTok{  sample\_tree }\SpecialCharTok{+} 
\NormalTok{  sample\_clusters\_annotation }\SpecialCharTok{+}
\NormalTok{  sample\_types\_annotation }\SpecialCharTok{+}
\NormalTok{  heatmap }\SpecialCharTok{+}
    \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{design =}\NormalTok{ design, }\AttributeTok{guides =} \StringTok{"collect"}\NormalTok{, }\CommentTok{\# Specify layout, collect legends}
                \AttributeTok{widths =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.2}\NormalTok{, }\FloatTok{0.15}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{),}
                \AttributeTok{heights =} \FunctionTok{c}\NormalTok{(}\FloatTok{0.1}\NormalTok{, }\FloatTok{0.15}\NormalTok{, }\FloatTok{0.15}\NormalTok{, }\FloatTok{0.25}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{))}

\NormalTok{plot}
\end{Highlighting}
\end{Shaded}

Heatmaps find several other applications in biclustering and
multi-assay analyses. These are discussed further in chapters
\ref{clustering} and \ref{multi-assay-analyses}.

\hypertarget{part-training}{%
\part{Training}\label{part-training}}

\hypertarget{training}{%
\chapter{Training}\label{training}}

The page provides practical information to support training and self-study.

\hypertarget{checklist}{%
\section{Checklist}\label{checklist}}

Brief checklist to prepare for training (see below for links).

\begin{itemize}
\tightlist
\item
  Install the recommended software
\item
  If the time allows, watch the short online videos and familiarize with the other available material
\item
  Join Gitter online chat for support
\end{itemize}

\hypertarget{software}{%
\section{Recommended software}\label{software}}

We recommend to install and set up the relevant software packages on
your own computer as this will support later use. The essential
components to install include:

\begin{itemize}
\item
  \href{https://www.r-project.org/}{R (the latest official release)}
\item
  \href{https://www.rstudio.com/products/rstudio/download/}{RStudio};
  choose ``Rstudio Desktop'' to download the latest version. Check the
  \href{https://www.rstudio.com/}{Rstudio home page} for more
  information. RStudio is optional.
\item
  Install key R packages (Section \ref{packages} provides an installation script)
\item
  After a successful installation you can consider trying out examples
  from Section \ref{exercises} already before training. \textbf{You can run
  the workflows by simply copy-pasting examples.} You can then test
  further examples from this tutorial, modifying and applying these
  techniques to your own data. Plain source code for the individual chapters of this book are available via \href{https://github.com/microbiome/OMA/tree/master/R}{Github}
\end{itemize}

\hypertarget{material}{%
\section{Study material}\label{material}}

We encourage to familiarize with the material and test examples in advance but this is optional:

\begin{itemize}
\item
  \href{https://carpentries-incubator.github.io/bioc-intro/}{Introduction to data analysis with R and Bioconductor} (for beginners with R)
\item
  \href{https://www.youtube.com/playlist?list=PLjiXAZO27elAJEptP59BN3whVJ61XIkST}{Short online videos} on microbiome data science with R/Bioconductor
\item
  \href{https://microbiome.github.io/outreach/index.html}{Quarto presentations}
\item
  \href{https://microbiome.github.io/OMA/}{Orchestrating Microbiome Analysis with Bioconductor (OMA)} (this book)
\item
  \href{https://github.com/microbiome/outreach}{Other outreach material}
\item
  \protect\hyperlink{exercises}{Exercises} for self-study
\item
  \protect\hyperlink{resources}{Resources} and links to complementary external material
\end{itemize}

\hypertarget{support-and-resources}{%
\section{Support and resources}\label{support-and-resources}}

For online support on installation and other matters, join us at
\href{https://gitter.im/microbiome/miaverse?utm_source=badge\&utm_medium=badge\&utm_campaign=pr-badge\&utm_content=badge}{Gitter}.

You are also welcome to connect through various channels with our
broader \href{https://microbiome.github.io}{developer and user community}.

\hypertarget{further-reading}{%
\section{Further reading}\label{further-reading}}

The following online books provide good general data science background:

\begin{itemize}
\tightlist
\item
  (Data science basics in R{]}(\url{https://r4ds.had.co.nz})
\item
  (Modern Statistics for Modern Biology){[}\url{https://www.huber.embl.de/msmb/}{]} open access book (Holmes S, Huber W)
\item
  \href{https://carpentries-incubator.github.io/bioc-project/}{The Bioconductor
  project}
  (background on the Bioconductor project; Carpentries workshop)
\end{itemize}

\hypertarget{coc}{%
\section{Code of Conduct}\label{coc}}

We support the \href{https://bioconductor.github.io/bioc_coc_multilingual/}{Bioconductor Code of Conduct}. The community values an open approach to science that promotes

\begin{itemize}
\tightlist
\item
  sharing of ideas, code, software and expertise
\item
  a kind and welcoming environment, diversity and inclusivity
\item
  community contributions and collaboration
\end{itemize}

\hypertarget{resources}{%
\chapter{Resources}\label{resources}}

\hypertarget{data-containers-1}{%
\section{Data containers}\label{data-containers-1}}

\hypertarget{data-container-documentation}{%
\subsection{Data container documentation}\label{data-container-documentation}}

\begin{itemize}
\tightlist
\item
  SingleCellExperiment \citep{R_SingleCellExperiment}

  \begin{itemize}
  \tightlist
  \item
    \href{https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html}{Online tutorial}
  \item
    \href{https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html}{Project page}
  \item
    \href{https://doi.org/10.1038/s41592-019-0654-x}{Publication}
  \end{itemize}
\item
  SummarizedExperiment \citep{R_SummarizedExperiment}

  \begin{itemize}
  \tightlist
  \item
    \href{https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html}{Online tutorial}
  \item
    \href{https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html}{Project page}
  \end{itemize}
\item
  TreeSummarizedExperiment \citep{R_TreeSummarizedExperiment}

  \begin{itemize}
  \tightlist
  \item
    \href{https://bioconductor.org/packages/release/bioc/vignettes/TreeSummarizedExperiment/inst/doc/Introduction_to_treeSummarizedExperiment.html}{Online tutorial}
  \item
    \href{https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html}{Project page}
  \item
    \href{https://doi.org/10.12688/f1000research.26669.2}{Publication}
  \end{itemize}
\item
  MultiAssayExperiment \citep{Ramos2017}

  \begin{itemize}
  \tightlist
  \item
    \href{https://www.bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html}{Online tutorial}
  \item
    \href{https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html}{Project page}\\
  \item
    \href{https://doi.org/10.1158/0008-5472.CAN-17-0344}{Publication}
  \end{itemize}
\end{itemize}

\hypertarget{other-relevant-containers}{%
\subsection{Other relevant containers}\label{other-relevant-containers}}

\begin{itemize}
\tightlist
\item
  \href{https://rdrr.io/bioc/S4Vectors/man/DataFrame-class.html}{DataFrame} which behaves similarly to \texttt{data.frame}, yet efficient and fast when used with large datasets.
\item
  \href{https://rdrr.io/bioc/Biostrings/man/DNAString-class.html}{DNAString} along with \texttt{DNAStringSet},\texttt{RNAString} and \texttt{RNAStringSet} efficient storage and handling of long biological sequences are offered within the Biostrings package \citep{R_Biostrings}.
\item
  GenomicRanges (\citep{GenomicRanges2013}) offers an efficient representation and manipulation of genomic annotations and alignments, see e.g.~\texttt{GRanges} and \texttt{GRangesList} at \href{https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html}{An Introduction to the GenomicRangesPackage}.
\end{itemize}

\href{http://girke.bioinformatics.ucr.edu/GEN242/tutorials/rsequences/rsequences/}{NGS Analysis Basics} provides a walk-through of the above-mentioned features with detailed examples.

\hypertarget{phyloseq-an-alternative-container-for-microbiome-data}{%
\subsection{phyloseq: an alternative container for microbiome data}\label{phyloseq-an-alternative-container-for-microbiome-data}}

The \texttt{phyloseq} package and class became the first widely used data
container for microbiome data science in R. Many methods for taxonomic
profiling data are readily available for this class. We provide here a
short description how \texttt{phyloseq} and \texttt{*Experiment} classes relate to
each other.

\texttt{assays} : This slot is similar to \texttt{otu\_table} in \texttt{phyloseq}. In a
\texttt{SummarizedExperiment} object multiple assays, raw
counts, transformed counts can be stored. See also
\citeyearpar{Ramos2017}
for storing data from multiple \texttt{experiments} such as
RNASeq, Proteomics, etc. \texttt{rowData} : This slot is
similar to \texttt{tax\_table} in \texttt{phyloseq} to store taxonomic
information. \texttt{colData} : This slot is similar to
\texttt{sample\_data} in \texttt{phyloseq} to store information
related to samples. \texttt{rowTree} : This slot is similar
to \texttt{phy\_tree} in \texttt{phyloseq} to store phylogenetic tree.

In this book, you will encounter terms such as \texttt{FeatureIDs} and
\texttt{SampleIDs}. \texttt{FeatureIDs} : These are basically OTU/ASV ids which are
row names in \texttt{assays} and \texttt{rowData}. \texttt{SampleIDs} : As the name
suggests, these are sample ids which are column names in \texttt{assays} and
row names in \texttt{colData}. \texttt{FeatureIDs} and \texttt{SampleIDs} are used but the
technical terms \texttt{rownames} and \texttt{colnames} are encouraged to be used, since
they relate to actual objects we work with.

\hypertarget{benchmarking-treese-with-phyloseq}{%
\subsubsection{Benchmarking TreeSE with phyloseq}\label{benchmarking-treese-with-phyloseq}}

TreeSE objects can be converted into phyloseq objects and vice versa, therefore
it is possible to compare the two containers in terms of computational efficiency.
Remarkably, TreeSE and phyloseq were benchmarked against one another in mia v1.2.3
and phyloseq v1.38.0, respectively. 5 standard microbiome analysis operationswere
applied to 4 datasets of varying size with both containers. In a nutshell, TreeSE
and phyloseq showed a similar performance for datasets of small and medium size
for most of the operations. However, TreeSE performed more efficiently as the size
of the datasets increased. Further details on such results can be found in the
\href{https://github.com/microbiome/benchmarking}{benchmarking repository}.

\hypertarget{resources-on-phyloseq}{%
\subsubsection{Resources on phyloseq}\label{resources-on-phyloseq}}

The phyloseq container provides analogous methods to TreeSE. The following
material can be used to familiarize with such alternative methods:

\begin{itemize}
\tightlist
\item
  \href{https://microsud.github.io/Tools-Microbiome-Analysis/}{List of R tools for microbiome analysis}
\item
  phyloseq \citep{McMurdie2013}
\item
  \href{http://microbiome.github.io/tutorials/}{microbiome tutorial}
\item
  \href{https://microsud.github.io/microbiomeutilities/}{microbiomeutilities}
\item
  Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses \citep{Callahan2016}.
\end{itemize}

\hypertarget{r-programming-resources}{%
\section{R programming resources}\label{r-programming-resources}}

\hypertarget{base-r-and-rstudio}{%
\subsection{Base R and RStudio}\label{base-r-and-rstudio}}

If you are new to R, you could try \href{https://swirlstats.com/}{swirl}
for a kickstart to R programming. Further support resources are
available through the Bioconductor
project \citep{Huber2015}.

\begin{itemize}
\tightlist
\item
  \href{https://geomoer.github.io/moer-base-r/cheatsheet.html}{Base R and RStudio cheatsheets}
\item
  \href{https://www.rstudio.com/resources/cheatsheets/}{Package-specific cheatsheets}
\item
  \href{https://ggplot2.tidyverse.org/}{Visualization with ggplot2}
\item
  \href{http://www.cookbook-r.com/Graphs/}{R graphics cookbook}
\end{itemize}

\hypertarget{bioc_intro}{%
\subsection{Bioconductor Classes}\label{bioc_intro}}

\textbf{S4 system}

S4 class system has brought several useful features to the
object-oriented programming paradigm within R, and it is constantly
deployed in R/Bioconductor packages \citep{Huber2015}.

~~Online Document:

\begin{itemize}
\tightlist
\item
  Hervé Pagès, \href{https://bioconductor.org/packages/release/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf}{A quick overview of the S4 class system}.
\item
  Laurent Gatto, \href{https://bioconductor.org/help/course-materials/2013/CSAMA2013/friday/afternoon/S4-tutorial.pdf}{A practical tutorial on S4 programming}
\item
  How S4 Methods Work \citep{Chambers2006}
\end{itemize}

~~Books:

\begin{itemize}
\tightlist
\item
  John M. Chambers. Software for Data Analysis: Programming with R. Springer, New York, 2008. ISBN-13 978-0387759357 \citep{Chambers2008}
\item
  I Robert Gentleman. R Programming for Bioinformatics. Chapman \& Hall/CRC, New York, 2008. ISBN-13 978-1420063677 \citep{gentleman2008r}
\end{itemize}

\hypertarget{quarto}{%
\section{Reproducible reporting with Quarto}\label{quarto}}

\hypertarget{learn-quarto}{%
\subsection{Learn Quarto}\label{learn-quarto}}

Reproducible reporting is the starting point for robust interactive
data science. Perform the following tasks:

\begin{itemize}
\item
  If you are entirely new to Quarto, take
  \href{https://quarto.org/docs/get-started/hello/rstudio.html}{this short tutorial}
  to get introduced to the most important functions within Quarto.
  Then experiment with different options from the
  \href{https://res.cloudinary.com/dyd911kmh/image/upload/v1676540721/Marketing/Blog/Quarto_Cheat_Sheet.pdf}{Quarto cheatsheet}.
\item
  Create a Quarto template in RStudio, and render it into a
  document (markdown, PDF, docx or other format). In case you are new
  to Quarto, its documentation provides guidelines to use Quarto with the
  R language (\href{https://quarto.org/docs/computations/r.html}{here})
  and the RStudio IDE (\href{https://quarto.org/docs/tools/rstudio.html}{here}).
\item
  Further examples are tips for Quarto are available in \href{https://appsilon.com/r-quarto-tutorial/}{this online tutorial}
  to interactive reproducible reporting.
\end{itemize}

\hypertarget{additional-material-on-rmarkdown}{%
\subsection{Additional material on Rmarkdown}\label{additional-material-on-rmarkdown}}

Being able to use Quarto in R partly relies on your previous knowledge of Rmarkdown. The following resources can help you get familiar with Rmarkdown:

\begin{itemize}
\tightlist
\item
  \href{https://www.markdowntutorial.com/}{Online tutorial}
\item
  \href{https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf}{Cheatsheet}
\item
  \href{https://rmarkdown.rstudio.com/lesson-1.html}{Documentation}
\item
  \href{https://rpubs.com/marschmi/RMarkdown}{Dr.~C Titus Brown's tutorial}
\end{itemize}

\textbf{Figure sources:}

\textbf{Original article}
- Huang R \emph{et al}. (2021) TreeSummarizedExperiment: a S4 class
for data with hierarchical structure. F1000Research 9:1246. \citep{Huang2021}

\textbf{Reference Sequence slot extension}
- Lahti L \emph{et al}. (2020) \href{https://doi.org/10.7490/\%20f1000research.1118447.1}{Upgrading the R/Bioconductor ecosystem for microbiome
research} F1000Research 9:1464 (slides).

\hypertarget{exercises}{%
\chapter{Exercises}\label{exercises}}

Here you can find assignments on different topics.

\textbf{Tips for exercises:}

\begin{itemize}
\tightlist
\item
  Add comments that explain what each line or lines of code do. This helps you and others understand your code and find bugs. Furthermore, it is easier for you to reuse the code, and it promotes transparency.
\item
  Interpret results by comparing them to literature. List main findings, so that results can easily be understood by others without advanced knowledge on data analytics.
\item
  Avoid hard-coding. Use variables which get values in the beginning of the pipeline. That way it is easier for you to modify parameters and reuse the code.
\end{itemize}

\hypertarget{basics-of-rbioconductor}{%
\section{Basics of R/Bioconductor}\label{basics-of-rbioconductor}}

Bioconductor training material has been contributed to \href{https://carpentries.org/}{Carpentries}. You can check the following lessons for basic background of R and Bioconductor.

\begin{itemize}
\item
  \href{https://carpentries-incubator.github.io/bioc-intro/}{Introduction to data analysis with R and Bioconductor}
\item
  \href{https://carpentries-incubator.github.io/bioc-project/}{Introduction to the Bioconductor project}
\end{itemize}

\hypertarget{workflows}{%
\section{Workflows}\label{workflows}}

\hypertarget{reproducible-reporting-with-quarto}{%
\subsection{Reproducible reporting with Quarto}\label{reproducible-reporting-with-quarto}}

The following batch of exercises walks you through typical use cases of Quarto
in RStudio. Before heading to the exercises, it is recommended to read the
\href{https://quarto.org/docs/tools/rstudio.html}{Quarto guidelines for RStudio}

\hypertarget{new-document}{%
\subsubsection{New document}\label{new-document}}

This exercise gets you started with creating a Quarto document and adding
text to it with typing conventions borrowed from the
\href{https://quarto.org/docs/authoring/markdown-basics.html}{markdown syntax}.
Feel free to render the document with the \textbf{Render} button after each step to
see the changes in the final report.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Open RStudio and create a new Quarto file named \texttt{My\ first\ Quarto}.
\item
  Add the subtitle \texttt{My\ first\ section} and write some text of your choice
  underneath. You can choose the level of headings by the number of preceding
  hashes (\texttt{\#}).
\item
  Add a subsection named \texttt{List\ of\ items} and list three items underneath, both
  ordered and unordered. You can initialize items with numbers (\texttt{1.}, \texttt{2.},
  \texttt{3.}, \ldots) or stars (\texttt{*}) for the ordered and unordered case, respectively.
\item
  Add another subsection named \texttt{Link\ to\ web} and add a clickable link to the
  \href{https://microbiome.github.io/OMA/}{OMA book}, using the pattern \texttt{{[}text{]}(url)}.
\item
  Render the document and check its appearance
\end{enumerate}

Nice start! You are now able to create a Quarto document, understand its syntax
and can render it into a reproducible report. If you got stuck, you can look up
the docs on \href{https://quarto.org/docs/tools/rstudio.html\#creating-documents}{creating}
and \href{https://quarto.org/docs/tools/rstudio.html\#render-and-preview}{rendering}
Quarto documents.

\hypertarget{code-chunks}{%
\subsubsection{Code chunks}\label{code-chunks}}

While customizable text is nothing new by itself, the advantage of Quarto (and
previously Rmakdown) is to combine text with code in R or other programming
languages, so that both the analytical pipeline and verbal description can be
put in one place. In this exercise, we learn how to write and run code in Quarto.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Open RStudio and create a new Quarto file.
\item
  Initialize a code chunk by pressing \texttt{alt} + \texttt{cmd} + \texttt{i} and define the variables
  \texttt{A\ \textless{}-\ "my\ name"} and \texttt{B\ \textless{}-\ 0} in it.
\item
  Write the text \texttt{Below\ is\ my\ first\ code\ chunk} just above the code chunk.
\item
  Initialize one more code chunk and add 100 to the variable \texttt{B} in it.
\item
  Write the text \texttt{Below\ I\ change\ variable\ B} just above the second chunk.
\item
  \textbf{Extra}: Write the following line of text: \texttt{my\ name\ is\ A\ and\ I\ am\ B\ years\ old},
  where A and B are variables defined in the code chunks upstream and change if
  those variables are modified. Be aware that inline code can be added as
  \texttt{\textgreater{}\ r\ my\_inline\_command} (without \texttt{\textgreater{}}).
\end{enumerate}

Good job. You are now able to combine text and code in a Quarto document. If you
got stuck, you can refer to the Quarto docs on
\href{https://quarto.org/docs/visual-editor/technical.html\#code-chunks}{using code chunks}.

\hypertarget{knitr-options}{%
\subsubsection{Knitr options}\label{knitr-options}}

Code chunks can be greatly customized in terms of visibility and execution,
output size and location and much more. This is possible with the knitr chunk
options, which usually appear at the beginning of the target chunk with the
syntax \texttt{\#\textbar{}\ option-name:\ value}, also described
\href{https://quarto.org/docs/computations/r.html\#chunk-options}{here}.
In this exercise, we explore part of the knitr potential.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Open RStudio and create a new Quarto file.
\item
  Initialize three code chunks and label them as \texttt{setup}, \texttt{fig-box} and
  \texttt{tbl-coldata}, respectively. Remember that the name of a chunk can be
  specified with the \texttt{label} option.
\item
  Write the following code in the corresponding chunk and render the document.
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# setup}
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"GlobalPatterns"}\NormalTok{, }\AttributeTok{package =} \StringTok{"mia"}\NormalTok{)}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ GlobalPatterns}

\CommentTok{\# this line sets some options for all the chunks (global chunk options)}
\NormalTok{knitr}\SpecialCharTok{::}\NormalTok{opts\_chunk}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{message =} \ConstantTok{FALSE}\NormalTok{, }\AttributeTok{warning =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# fig{-}box}
\FunctionTok{boxplot}\NormalTok{(}\FunctionTok{colSums}\NormalTok{(}\FunctionTok{assay}\NormalTok{(tse)) }\SpecialCharTok{\textasciitilde{}}\NormalTok{ tse}\SpecialCharTok{$}\NormalTok{SampleType)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# tbl{-}coldata}
\NormalTok{knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{(}\FunctionTok{head}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)))}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\tightlist
\item
  Set \texttt{include:\ false} in the \texttt{setup} chunk, \texttt{fig-width:\ 10} in the \texttt{fig-box}
  chunk and \texttt{echo:\ false} in the \texttt{tbl-coldata} chunk. Render the document again
  and find the differences from before.
\item
  Add the options \texttt{fig-cap} and \texttt{tab-cap} to the \texttt{fig-box} and \texttt{tbl-coldata}
  chunks, respectively. They require some text input, which makes for the
  caption of the figures or tables.
\item
  \textbf{Extra}: Create a cross-reference to \texttt{fig-box} and \texttt{tbl-coldata} in the
  text above the respective code chunk. You can do that with the syntax
  \texttt{@chunk-name}.
\item
  \textbf{Extra}: Define a custom folder for storing figures with \texttt{fig-path}. Insert
  it in \texttt{knitr::opts\_chunk\$set}, so that it applies globally to all the figures
  generated in the document.
\end{enumerate}

Congratulations! You are now familiar with the great potential and flexibility
of knitr chunk options. An exhaustive list of available options can be found
in the \href{https://yihui.org/knitr/options/}{knitr documentation}.

\hypertarget{yaml-instructions}{%
\subsubsection{YAML instructions}\label{yaml-instructions}}

The box at the beginning of every Quarto document contains yaml options that let
you define the metadata of the document. They will affect the appearance of the
document when it is rendered. By default, the box includes yaml options for the
title, format and editor to be used, but much more information on layout, code
execution and figures can be specified. A comprehensive list of yaml options is
available \href{https://quarto.org/docs/reference/formats/html.html}{here}. In this
exercise, we will get a tiny taste of such functionality.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Open RStudio and create a new Quarto file.
\item
  In the yaml box at the beginning of the document, change the title from
  \texttt{Untitled} to \texttt{My\ first\ Quarto}.
\item
  In the same box, add the two lines \texttt{author} and \texttt{date} followed by your name
  and today's date, respectively.
\item
  Render the document and check its appearance.
\item
  \textbf{Extra}: Set \texttt{toc:\ true} to produce a table of contents. This line should
  follow \texttt{format} and \texttt{html} at the second level of indentation.
\end{enumerate}

Well done! Now you are able to specify yaml options and understand how they
affect your Quarto document. If you got stuck, you can check
\href{https://quarto.org/docs/tools/rstudio.html\#yaml-intelligence}{this section} of
the Quarto documentation.

\hypertarget{quarto-parameters}{%
\subsubsection{Quarto parameters}\label{quarto-parameters}}

An advanced feature of Quarto consists of execution parameters, which are
externally pre-defined variables that are also accessible in the Quarto document.
They can be specified in the yaml box as \texttt{params}. Here we learn how to use them.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Open RStudio and create a new Quarto file.
\item
  In the yaml box at the beginning of the document, add a line named \texttt{params}
  followed by an indented line with \texttt{gamma:\ 10}
\item
  Initialize a code chunk and type \texttt{str(params\$gamma)} in it.
\item
  Render the document and check what happened.
\item
  Define one more parameter \texttt{beta:\ 3} and multiply \texttt{gamma} by \texttt{beta} in a code
  chunk below.
\item
  Render the document again and check what happened.
\end{enumerate}

Well done! You can now use an advanced feature of Quarto such as parameters.
If you got stuck, \href{https://quarto.org/docs/computations/parameters.html\#knitr}{here}
you can find more information about parameter definition and usage.

\hypertarget{data-containers-treese}{%
\section{Data containers: TreeSE}\label{data-containers-treese}}

TreeSE containers represent the working unit of the mia package. In the
following exercises we learn how to construct, explore and work with them.
A few demo datasets can be imported with mia and can be accessed as explained
in chapter \ref{example-data}.

\hypertarget{construct-TreeSE}{%
\subsection{Constructing a data object}\label{construct-TreeSE}}

Here we cover how to construct a TreeSE from CSV files, using the components
of OKeefeDSData from the microbiomeDataSets package as an example dataset.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Fetch or download the files in \href{https://github.com/microbiome/data/tree/main/OKeefeDSData}{this directory}.
\item
  Read in the csv files with \texttt{read.csv} and store them into the variables
  \texttt{assays}, \texttt{rowdata} and \texttt{coldata}, respectively.
\item
  Create a TreeSE from the individual components with
  \texttt{TreeSummarizedExperiment}. Note that the function requires three arguments:
  assays, rowData and colData, to which you can give the appropriate item.
\item
  Check that importing is done correctly. E.g., choose random samples and
  features, and check that their values equal between raw files and TreeSE.
\end{enumerate}

Usefuls functions: DataFrame, TreeSummarizedExperiment, matrix, rownames, colnames, SimpleList

\hypertarget{importing-data}{%
\subsection{Importing data}\label{importing-data}}

Raw data of different types can be imported as a TreeSE with a number of
functions explained in chapter \ref{import-from-file}. You can also check the
\href{https://microbiome.github.io/mia/reference/index.html}{function reference in the mia package}.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Get familiar with the
  \href{https://github.com/microbiome/data}{microbiome data repository}
  and read the instructions in its
  \href{https://github.com/microbiome/data\#training-microbiome-datasets}{README}
  to import and construct datasets from there.
\item
  Import data from another format (functions: loadFromMetaphlan \textbar{} loadFromMothur \textbar{} loadFromQIIME2 \textbar{} makeTreeSummarizedExperimentFromBiom \textbar{} makeTreeSummarizedExperimentFromDADA2 \ldots)
\item
  Try out conversions between TreeSE and phyloseq data containers (makeTreeSummarizedExperimentFromPhyloseq; makephyloseqFromTreeSummarizedExperiment)
\end{enumerate}

\hypertarget{preliminary-exploration}{%
\subsection{Preliminary exploration}\label{preliminary-exploration}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Get a summary about the TreeSE with \texttt{summary}. What is the mean count across
  samples? How many features recur only once (singletons)?
\item
  Check the dimensions of the TreeSE with \texttt{dim} or alternatively with \texttt{nrow}
  and \texttt{ncol}. How many samples and features are present?
\item
  List sample and features names with \texttt{rownames} and \texttt{colnames}.
\item
  Check what information about samples and features is contained by the
  \texttt{colData} and \texttt{rowData} of the TreeSE with \texttt{names}.
\item
  \textbf{Extra}: Calculate the number of unique taxa for each taxonomic rank. You
  can use \texttt{apply} to count unique elements for each column of rowData.
\end{enumerate}

\hypertarget{assay-retrieval}{%
\subsection{Assay retrieval}\label{assay-retrieval}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  List the names of all available assays with \texttt{assayNames}.
\item
  Fetch the list of assays with \texttt{assays}.
\item
  Retrieve the first assay of the TreeSE with \texttt{assay}, where the second
  argument can be either the name or the index of the desired assay.
\end{enumerate}

Well done! You can now locate and retrieve individual assays of a TreeSE. If you
got stuck, you can refer to chapter \ref{assay-slot} of this book.

\hypertarget{sample-information}{%
\subsection{Sample information}\label{sample-information}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Check the names of the samples with \texttt{colnames}.
\item
  List the information on samples available in colData with \texttt{names}.
\item
  Visualize the colData with \texttt{View} and briefly look at the information stored
  in the different columns.
\item
  Get the abundances of all features for a specific sample, such as \texttt{ID34}, for an \emph{assay} of your choice.
\end{enumerate}

\hypertarget{feature-information}{%
\subsection{Feature information}\label{feature-information}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Check the names of the features with \texttt{rownames}.
\item
  List the information on features available in rowData with \texttt{names}.
\item
  Visualize the rowData with \texttt{View} and briefly look at the information stored
  in the different columns.
\item
  Get the abundances for a specific feature, such as \texttt{OTU1810}, in all the
  samples. You can access feature-specific abundances for an \emph{assay} of your choice.
\item
  \textbf{Extra}: Create a taxonomy tree based on the taxonomy mappings with
  \texttt{addTaxonomyTree} and display its content with \texttt{taxonomyTree} and \texttt{ggtree}.
\end{enumerate}

If you got stuck, you can look up chapters @fref\{datamanipulation\}
and \ref{fly-tree} on how to pick specific abundances and generate
row trees, respectively.

\hypertarget{other-elements}{%
\subsection{Other elements}\label{other-elements}}

Try to extract some of the other TreeSE elements listed in chapter \ref{containers}.
However, such data are not always included.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Fetch the metadata of the TreeSE. Is there any iformation available?
\item
  Access the phylogenetic tree with \texttt{rowTree}. How big is it in terms of tips
  and nodes. If you like you can visualize it with \texttt{ggtree}.
\item
  Check if a sample tree is available with \texttt{colTree}, which is suitable for
  hierarchical or nested study designs.
\item
  If present, obtain the information on feature DNA sequences from the DNA
  sequence slot.
\end{enumerate}

\hypertarget{data-manipulation}{%
\section{Data manipulation}\label{data-manipulation}}

\hypertarget{subsetting-1}{%
\subsection{Subsetting}\label{subsetting-1}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Subset the TreeSE object to specific samples
\item
  Subset the TreeSE object to specific features
\item
  Subset the TreeSE object to specific samples and features
\end{enumerate}

\hypertarget{library-sizes}{%
\subsection{Library sizes}\label{library-sizes}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Calculate library sizes
\item
  Subsample / rarify the counts (see: subsampleCounts)
\end{enumerate}

Useful functions: nrow, ncol, dim, summary, table, quantile, unique, addPerCellQC, mergeFeaturesByRank

\hypertarget{prevalent-and-core-taxonomic-features}{%
\subsection{Prevalent and core taxonomic features}\label{prevalent-and-core-taxonomic-features}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Estimate prevalence for your chosen feature (row, taxonomic group)
\item
  Identify all prevalent features and subset the data accordingly
\item
  Report the thresholds and the dimensionality of the data before and after subsetting
\item
  Visualize prevalence
\end{enumerate}

Useful functions: getPrevalence, getPrevalentFeatures, subsetByPrevalentFeatures

\hypertarget{data-exploration}{%
\subsection{Data exploration}\label{data-exploration}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Summarize sample metadata variables. (How many age groups, how they are distributed? 0\%, 25\%, 50\%, 75\%, and 100\% quantiles of library size?)
\item
  Create two histograms. Another shows the distribution of absolute counts, another shows how CLR transformed values are distributed.
\item
  Visualize how relative abundances are distributed between taxa in samples.
\end{enumerate}

Useful functions: nrow, ncol, dim, summary, table, quantile, unique, transformAssay, ggplot, wilcox.test, mergeFeaturesByRank, plotAbundance

\hypertarget{other-functions}{%
\subsection{Other functions}\label{other-functions}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Merge data objects (merge, mergeSEs)
\item
  Melt the data for visualization purposes (meltAssay)
\end{enumerate}

\hypertarget{assay-transformation}{%
\subsection{Assay transformation}\label{assay-transformation}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Transform the counts assay into relative abundances with \texttt{transformAssay} and
  store it into the TreeSE as an assay named \texttt{relabund} (see chapter \ref{assay-transform}).
\item
  Similarly, perform a clr transformation on the counts assay with a \texttt{pseudocount}
  of 1 and add it to the TreeSE as a new assay.
\item
  List the available assays by name with \texttt{assays}.
\item
  Access the clr assay and select a subset of its first 100 features and 10
  samples. Remember that assays are subsettable with \texttt{assay{[}row\_idx,\ col\_idx{]}}.
\item
  Take the same subset from the TreeSE, and check how this affects the individual
  transformed assays. TreeSE can also be subsetted with \texttt{tse{[}row\_idx,\ col\_idx{]}}.
\item
  \textbf{Extra}: If the data has phylogenetic tree, perform the phILR transformation.
\end{enumerate}

\hypertarget{abundance-tables}{%
\section{Abundance tables}\label{abundance-tables}}

\hypertarget{taxonomic-levels}{%
\subsection{Taxonomic levels}\label{taxonomic-levels}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load one of the example data sets mentioned in Chapter 3.3
  with \texttt{data} (you need one with taxonomic information at Phylum level)
  and store it into a variable named \texttt{tse}.
\item
  List the available taxonomic ranks in the data with \texttt{taxonomyRanks}.
\item
  Agglomerate the data to Phylum level with \texttt{mergeFeaturesByRank} and the
  appropriate value for \texttt{Rank}.
\item
  Report the dimensions of the TreeSE before and after agglomerating. You can
  use \texttt{dim} for that.
\item
  \textbf{Extra}: Perform CLR transformation on the data. Does this affect agglomeration?
\item
  \textbf{Extra}: List full taxonomic information for a few selected taxa, such as
  \texttt{OTU1} and \texttt{OTU1368}. For that you can use \texttt{mapTaxonomy} on a specific subset
  of the TreeSE.
\end{enumerate}

\hypertarget{alternative-experiments}{%
\subsection{Alternative experiments}\label{alternative-experiments}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load one of the example data sets mentioned in Chapter 3.3
  with \texttt{data} (you need one with taxonomic information) and store it into
  a variable named \texttt{tse}.
\item
  Check the taxonomic ranks of the features with \texttt{taxonomyRanks}. What is the
  deepest taxonomic rank available?
\item
  Agglomerate the TreeSE to each taxonomic rank and store the resulting
  experiments as altExps. This can be performed automatically with \texttt{splitByRanks}.
\item
  Check the names of the generated altExps with \texttt{altExpNames} and retrieve a
  complete list with \texttt{altExps}.
\item
  Retrieve the data agglomerated by genus from the corresponding \texttt{altExp}.
  As for assays, you can access the desired altExp by name or index.
\item
  \textbf{Extra}: Split the data based on other features with \texttt{splitOn}.
\end{enumerate}

\hypertarget{community-alpha-diversity}{%
\section{Community (alpha) diversity}\label{community-alpha-diversity}}

\hypertarget{estimation-1}{%
\subsection{Estimation}\label{estimation-1}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Calculate multiple alpha diversity indices with \texttt{estimateDiversity} without
  any additional arguments.
\item
  Check the names of colData with \texttt{names}. Can you identify which columns
  contain the alpha diversity indices?
\item
  \textbf{Extra}: Agglomerate the TreeSE by phylum and compare the mean Shannon
  diversity of the original experiment with its agglomerated version. You can
  use \texttt{mergeFeaturesByRank} to perform agglomeration and \texttt{mean} to calculate the
  mean values of the respective columns in colData.
\end{enumerate}

\hypertarget{visualization-1}{%
\subsection{Visualization}\label{visualization-1}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and scater packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Calculate Shannon diversity index and Faith's phylogenetic diversity with
  \texttt{estimateDiversity} and the appropriate arguments for \texttt{index}.
\item
  Make a boxplot of Shannon diversity on the y axis and sample type on the x
  axis with \texttt{plotColData}.
\item
  Repeat the previous point with Faith's phylogenetic diversity and compare the
  sample distributions of the two alpha diversity indices. How greatly do they
  differ?
\item
  \textbf{Extra}: Make a scatterplot of Shannon diversity on the y axis and Faith's
  phylogenetic diversity on the x axis with \texttt{plotColData}. Colour the points
  by sample type with the appropriate optional argument.
\end{enumerate}

\hypertarget{correlation}{%
\subsection{Correlation}\label{correlation}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and scater packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Calculate coverage and Shannon diversity index with \texttt{estimateDiversity} and
  the appropriate arguments for \texttt{index}.
\item
  Test the correlation between the two indices with \texttt{cor.test}. Remember that
  colData parameters are accessible with \texttt{tse\$param\_name}. Use Kendall tau
  coefficients as \texttt{method} to measure correlation. Is the correlation
  weak or strong, significant or not?
\item
  Make a scatterplot of Shannon diversity index on the y axis and coverage on
  the x axis. You can do that with \texttt{plotColData}. How do the two indices
  relate to one another?
\item
  \textbf{Extra}: Compute the library size of the samples by applying \texttt{colSums} to
  the counts assay of the TreeSE, and test the correlation of library size with
  Shannon diversity or coverage. Which index is more correlated with library size?
\end{enumerate}

In this example, we inspected the correlation between two related variables,
also known as multicollinearity, and checked the correlation to library size,
which is part of quality control. However, the correlation between alpha diversity
and other numerical data about samples, such as participant's age and weight, also
represent an important analysis in several studies.

\hypertarget{differences-between-groups}{%
\subsection{Differences between groups}\label{differences-between-groups}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Calculate the Gini-Simpson diversity with \texttt{estimateDiversity} and the
  appropriate argument for \texttt{index}. Set \texttt{name} to \texttt{simpson}. You will use this
  name to access the diversity index from colData.
\item
  Inspect the Diet column in the colData. Determine how the samples are grouped
  in terms of diet. You can see the number of unique elements in a column with
  \texttt{unique}.
\item
  Test differences in Gini-Simpson diversity between different diets with
  \texttt{kruskal.test}. Remember that colData parameters are accessible with
  \texttt{tse\$param\_name}.
\item
  Is diversity significantly different between vegan and mixed diet? To
  visualize that, make a boxplot of Gini-Simpson diversity on the y axis and
  diet on the x axis with \texttt{plotColData}.
\item
  \textbf{Extra}: Repeat points 3 through 5, this time for age groups. Make sure
  that you are using an appropriate statistical test for the number of groups
  and distribution.
\end{enumerate}

\hypertarget{community-similarity-1}{%
\section{Community similarity}\label{community-similarity-1}}

\hypertarget{reduced-dimensions-retrieval}{%
\subsection{Reduced dimensions retrieval}\label{reduced-dimensions-retrieval}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load enterotype with \texttt{data} and store it into a
  variable named \texttt{tse}.
\item
  List all available reduced dimensions with \texttt{reducedDims}. At this point, no
  reducedDims are likely found, because we haven't created any yet.
\item
  Perform PCA and store its output in the TreeSE by running
  \texttt{tse\ \textless{}-\ runPCA(tse,\ assay.type\ =\ "counts")}. Note that it is required to
  specify the assay on which dimensionality reduction should be conducted.
\item
  View the names of all available reduced dimensions with \texttt{reducedDimNames}.
  Has something new appeared?
\item
  \textbf{Extra}: Access the \texttt{PCA} reducedDim object with \texttt{reducedDim} and explore
  its content. How are the different dimensions stored? Try to extract an
  array with only the values from the second dimension by indexing the object
  with \texttt{{[}\ ,\ 2{]}}.
\end{enumerate}

\hypertarget{visualization-basics-with-pca}{%
\subsection{Visualization basics with PCA}\label{visualization-basics-with-pca}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and scater packages, load enterotype with \texttt{data} and store it
  into a variable named \texttt{tse}.
\item
  Perform a 3-component PCA based on the \texttt{counts} assay. You can use \texttt{runPCA}
  and set the optional arguments \texttt{ncomponents} and \texttt{assay.type} to the
  appropriate values.
\item
  Plot the first two dimensions of PCA with \texttt{plotReducedDim}, to which you
  should give the appropriate reducedDim name as the second argument. Note that
  by default only the first two dimensions are shown.
\item
  Check which information is stored in the ColData of the TreeSE. What would
  be worth visualizing in our coordination plot?
\item
  Make the same plot again, but this time colour the observations by Enterotype.
  You can do that by setting \texttt{colour\_by} to the appropriate colname in the
  colData of the TreeSE.
\item
  \textbf{Extra}: Plot all three dimensions of PCA with \texttt{plotReducedDim} and the
  optional argument \texttt{ncomponents}. Colour observations by Enterotype. Which
  pair of dimensions best explains the variance between Enterotypes?
\end{enumerate}

\hypertarget{principal-coordinate-analysis-pcoa}{%
\subsection{Principal Coordinate Analysis (PCoA)}\label{principal-coordinate-analysis-pcoa}}

PCoA turns out to be particularly relevant for microbiome analysis, because
unlike PCA it can generate reduced dimensions from distances other than
Euclidean. There are several ecological distances to choose from and you can
find many of them under methods in the vignettes of \texttt{vegan::vegdist}.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and scater packages, load enterotype with \texttt{data} and store it
  into a variable named \texttt{tse}.
\item
  Transform the counts assay to relative abundances with \texttt{transformAssay}.
\item
  Perform a Multi-Dimensional Scaling (MDS) based on the relative abundance
  assay in terms of Bray-Curtis dissimilarity. You can use \texttt{runMDS} with the
  compulsory argument \texttt{FUN\ =\ vegan::vegdist}.
\item
  Plot the first two dimensions of PCA with \texttt{plotReducedDim}, to which you
  should give the appropriate reducedDim name as the second argument. Colour
  the observations by Enterotype with \texttt{colour\_by}.
\item
  \textbf{Extra}: Perform MDS again with \texttt{runMDS}, but this time use Jaccard
  dissimilarity. The distance metric to use can be defined with the optional
  argument \texttt{method}, choosing from the methods in \texttt{?vegan::vegdist}. If you
  don't want to overwrite the reducedDim object made in point 3, set \texttt{name} to
  a name of your choice. Visualize and compare it to the plot from point 4.
\end{enumerate}

Good job! You are now able to produce and visualize reduced dimensions of a
TreeSE. \texttt{runMDS} is actually one of several algorithms for PCoA and dimensionality
reduction, which you can find in section \ref{other-ord-methods}.

\hypertarget{permanova-analysis}{%
\subsection{PERMANOVA analysis}\label{permanova-analysis}}

In this exercise we focus on studying the weight of variables on the microbiome
composition. Significance of each variable on beta diversity is tested with
PERMANOVA (point 4) and the homogeneity assumption is also be controlled with a
PERMDISP2 analysis (point 5).

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Transform the \texttt{counts} assay into relative abundances with \texttt{transformAssay}.
\item
  Extract the relative abundance assay, transpose it and save it into
  a variable named \texttt{relabund\_assay}.
\item
  Perform PERMANOVA with \texttt{adonis2} to see how much Diet can explain the relative
  abundance assay (\texttt{formula\ =\ relabund\_assay\ \textasciitilde{}\ Diet}) in terms of Bray-Curtis
  dissimilarity (\texttt{method\ =\ "bray"}). Also set \texttt{data\ =\ colData(tse)}
  \texttt{by\ =\ "margin"} and \texttt{permutations\ =\ 99}. What do the results tell you about
  Diet with respect to beta diversity?
\item
  \textbf{Extra}: Test homogeneity of distribution across Diet groups with
  \texttt{anova(betadisper(my\_mat),\ my\_groups}, where \texttt{my\_mat} is the Bray-Curtis
  dissimilarity matrix of \texttt{relabund\_assay} calculated with
  \texttt{vegdist(relabund\_assay,\ "bray")} and \texttt{my\_groups} is the vector of Diet values
  obtained from the colData of the TreeSE.
\end{enumerate}

Well done! You went through testing the effect and significance of Diet on beta
diversity. Keep in mind that the formula fed to \texttt{adonis2} can take more than one
independent variable, so that you can also (and very often should) include
covariates of your studies.

\hypertarget{redundancy-analysis-rda}{%
\subsection{Redundancy analysis (RDA)}\label{redundancy-analysis-rda}}

Here we apply RDA, an ordination method that provides dimensions with
the largest variation between the data based in the light of the specified
variables (point 3). The results of RDA are usually assessed with PERMANOVA
(point 5) and the homogeneity assumption should be checked as in the previous
exercise. This is a relatively complex procedure, but the way this is broken
down into steps below will hopefully make more sense.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Transform the \texttt{counts} assay into relative abundances with \texttt{transformAssay}.
\item
  Perform RDA with \texttt{calculateRDA} to see how much Diet can explain the relative
  abundance assay (\texttt{formula\ =\ assay\ \textasciitilde{}\ Diet} and \texttt{assay.type\ =\ relabundance}) in
  terms of Bray-Curtis dissimilarity (\texttt{method\ =\ "bray"}).
\item
  Extract the RDA dimensions from the appropriate reducedDim slot with
  \texttt{attr(reducedDim(tse,\ "RDA"),\ "rda)} and store it into \texttt{rda}.
\item
  Test the effect and significance of Diet on beta diversity by PERMANOVA with
  \texttt{anova.cca}. Feed this function with \texttt{rda} and set \texttt{by\ =\ "margin"} and
  \texttt{permutations\ =\ 99}, respectively. What do the results tell you about Diet?
\item
  \textbf{Extra}: Check what other parameters are stored in the colData of peerj13075,
  add them to the formula (\texttt{formula\ =\ assay\ \textasciitilde{}\ Diet\ +\ ...}) of \texttt{calculateRDA}
  and proceed to see how that changes the results of PERMANOVA.
\end{enumerate}

Well done! You went through an RDA analysis followed by significance testing
with PERMANOVA and BETADISPER2. In the next exercise we'll go deeper quantify
the contributions to beta diversity.

\hypertarget{beta-diversity-analysis}{%
\subsection{Beta diversity analysis}\label{beta-diversity-analysis}}

This exercise prompts you to implement a workflow with distance-based RDA
(dbRDA). You can refer to chapter \ref{dbrda-workflow} for a step-by-step
walkthrough, which may be simplified in the future.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Create dbRDA with Bray-Curtis dissimilarities on relative abundances. Use PERMANOVA. Can differences between samples be explained with variables of sample meta data?
\item
  Analyze diets' association on beta diversity. Calculate dbRDA and then PERMANOVA. Visualize coefficients. Which taxa's abundances differ the most between samples?
\item
  Interpret your results. Is there association between community composition and location? What are those taxa that differ the most; find information from literature.
\end{enumerate}

Useful functions: runMDS, runRDA, anova.cca, transformAssay, mergeFeaturesByRank, ggplot, plotReducedDim, vegan::adonis2

\hypertarget{differential-abundance-1}{%
\section{Differential abundance}\label{differential-abundance-1}}

\hypertarget{standard-analysis-with-aldex2}{%
\subsection{Standard analysis with ALDEx2}\label{standard-analysis-with-aldex2}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and ALDEx2 packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Agglomerate the TreeSE by genus and filter by a prevalence of 10\%. You can
  perform both operations in one go with \texttt{subsetByPrevalentFeatures} by specifying
  the \texttt{rank} and \texttt{prevalence} arguments.
\item
  Model the counts assay of the TreeSE with \texttt{aldex.clr} and store it into the
  variable \texttt{x}. As a second argument, provide the grouping variable \texttt{Diet},
  which is contained in a column of the colData.
\item
  Feed \texttt{x} to the functions \texttt{aldex.ttest} to erform t-test and to \texttt{aldex.effect}
  to estimate effect sizes. Store the output into \texttt{x\_tt} and \texttt{x\_effect},
  respectively.
\item
  Create a data.frame named \texttt{aldex\_out} which includes both \texttt{x\_tt} and \texttt{x\_effect}
  and filter for the features with \texttt{wi.eBH\ \textless{}\ 0.05}. Are there any significantly
  differential abundance taxa?
\item
  \textbf{Extra}: If these results appear boring, repeat steps 1 - 5, but use
  \texttt{Gender} or \texttt{Age} as the grouping variable. Do we have any better luck with
  Gender? What is the problem with Age?
\end{enumerate}

\hypertarget{control-daa-confounders}{%
\subsection{Controlling for confounders}\label{control-daa-confounders}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and MicrobiomeStat packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Agglomerate the TreeSE by genus and filter by a prevalence of 10\%. You can
  perform both operations in one go with \texttt{subsetByPrevalentFeatures} by specifying
  the \texttt{rank} and \texttt{prevalence} arguments.
\item
  Model the counts assay of the TreeSE with \texttt{linda} and store the output into
  a variable named \texttt{linda\_out}. Provide the colData converted into a data.frame
  (with \texttt{as.data.frame}) as the second argument, and a \texttt{formula} with the Age,
  Gender and Diet as variables. For example, \texttt{formula\ =\ "\textasciitilde{}\ A\ +\ B"} represents a
  formula with variables A and B.
\item
  Extract the \texttt{output\$AgeElderly} object from \texttt{linda\_out} with \texttt{\$} and store it
  into a variable named \texttt{linda\_res}.
\item
  Filter \texttt{linda\_res} for features with \texttt{reject\ ==\ TRUE}. How many differentially
  abundant taxa were found? What are their names and how significant are they
  in terms of log-fold change and adjusted p-value?
\end{enumerate}

\hypertarget{compare-daa-methods}{%
\subsection{Comparing methods}\label{compare-daa-methods}}

Here, we conduct DAA with identical parameters as in the
\protect\hyperlink{control-daa-confounders}{previous exercise}, but with a different method,
namely ZicoSeq. We aim to compare the results between these two methods and draw better informed conclusions from such comparative approach.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia and GUniFrac packages, load any of the example data sets mentioned in Chapter 3.3
  with \texttt{data} and store it into a variable named \texttt{tse}.
\item
  Agglomerate the TreeSE by genus and filter by a prevalence of 10\%. You can
  perform both operations in one go with \texttt{subsetByPrevalentFeatures} by specifying
  the \texttt{rank} and \texttt{prevalence} arguments.
\item
  Model the counts assay of the TreeSE with \texttt{ZicoSeq} as the \texttt{feature.dat}
  argument and store the output into a variable named \texttt{zicoseq\_out}. Provide
  also the colData converted to a data.frame (with \texttt{as.data.frame}) as
  \texttt{meta.dat}. In addition, set \texttt{grp.name} to \texttt{"Age"}, \texttt{adj.name} to
  \texttt{c("Diet",\ "Gender")}, \texttt{feature.dat.type} to \texttt{"count"}, \texttt{return.feature.dat}
  to \texttt{TRUE} and \texttt{perm.no} to \texttt{999}.
\item
  View the top six differentially abundant taxa and their adjusted p-values
  with \texttt{head(sort(zicoseq\_out\$p.adj.fdr))}. Is there any significant taxon
  according to ZicoSeq? Compared to the output of linda, do we see the same
  taxa at the top in terms of significance? Overall, to what extent do the two
  methods agree with one another?
\end{enumerate}

\hypertarget{workflow-1}{%
\subsection{Workflow 1}\label{workflow-1}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Get the abundances for an individual feature (taxonomic group / row)
\item
  Visualize the abundances per group with boxplot / jitterplot
\item
  Is the difference significant (Wilcoxon test)?
\item
  Is the difference significant (linear model with covariates)?
\item
  How do transformations affect the outcome (log10, clr..)?
\item
  Get p-values for all features (taxa), for instance with a for loop
\item
  Do multiple testing correction
\item
  Compare the results from different tests with a scatterplot
\end{enumerate}

Useful functions: {[}{]}, ggplot2::geom\_boxplot, ggplot2::geom\_jitter, wilcox.test, lm.test, transformAssay, p.adjust

\hypertarget{workflow-2}{%
\subsection{Workflow 2}\label{workflow-2}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  install the latest development version of mia from GitHub.
\item
  Load experimental dataset from mia.
\item
  Compare abundances of each taxa between groups. First, use Wilcoxon or Kruskall-Wallis test. Then use some other method dedicated to microbiome data.
\item
  Summarize findings by plotting a taxa vs samples heatmap. Add column annotation that tells the group of each sample, and row annotation that tells whether the difference of certain taxa was statistically significant.
\item
  Choose statistically significant taxa and visualize their abundances with boxplot \& jitterplot.
\end{enumerate}

Useful functions: wilcox.test, kruskal.test, ggplot, pheatmap, ComplexHeatMap::Heatmap, ancombc, aldex2, maaslin2, mergeFeaturesByRank, transformAssay, subsetByPrevalentFeatures

\hypertarget{visualization-2}{%
\section{Visualization}\label{visualization-2}}

\hypertarget{multivariate-ordination}{%
\subsection{Multivariate ordination}\label{multivariate-ordination}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Load experimental dataset from mia.
\item
  Create PCoA with Bray-Curtis dissimilarities
\item
  Create PCA with Aitchison dissimilarities
\item
  Visualize and compare both
\item
  Test other transformations, dissimilarities, and ordination methods
\end{enumerate}

Useful functions: runMDS, runNMDS, transformAssay, ggplot, plotReducedDim

\hypertarget{heatmap-visualization}{%
\subsection{Heatmap visualization}\label{heatmap-visualization}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Load experimental dataset from mia.
\item
  Visualize abundances with heatmap
\item
  Visualize abundances with heatmap after CLR + Z transformation
\end{enumerate}

See the OMA book for examples.

\hypertarget{multiomics}{%
\section{Multiomics}\label{multiomics}}

\hypertarget{basic-exploration}{%
\subsection{Basic exploration}\label{basic-exploration}}

Here we learn how to conduct preliminary exploration on a MAE, using
HintikkaXOData as an example dataset.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load HintikkaXOData with \texttt{data} and store it into a
  variable named \texttt{mae}.
\item
  Which experiments make up the MAE? How many samples and features are contained
  by each experiment? You can get a summary for all experiments with \texttt{experiments},
  and check for each individual experiment with \texttt{dim}, \texttt{nrow} and \texttt{ncol}.
\item
  What are the names of the features and samples of the different experiments?
  You can see that with \texttt{rownames} and \texttt{colnames}, respectively.
\item
  What information is known about the samples? Remember that information about
  samples is stored in the \texttt{colData} of the MAE.
\item
  \textbf{Extra}: How do the samples of the individual experiments map to the
  columns of the MAE? You can find the sample mapping in the \texttt{sampleMap}
  of the MAE.
\end{enumerate}

So far so good. You explored a MAE and its experiments, getting a taste of how
information is organized in its internal structure.

\hypertarget{experiment-agglomeration}{%
\subsection{Experiment agglomeration}\label{experiment-agglomeration}}

Here we learn how to manipulate an experiment contained by a MAE and save the
new modified version of the experiment in a suitable place (the altExp slot).

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load HintikkaXOData with \texttt{data} and store it into a
  variable named \texttt{mae}.
\item
  Agglomerate the microbiota experiment by Genus and store the output into the
  \texttt{altExp} slot of the microbiota experiment, with the custom name
  \texttt{microbiota\_genus}.
\item
  How many features remain after agglomerating? What are their names?
\item
  \textbf{Extra}: create one more alternative experiment named
  \texttt{prevalent\_microbiota\_family}, which contains the microbiota experiment
  agglomerated by Family with a prevalence threshold of 10\%. You can
  agglomerate and in parallel select by prevalence with
  \texttt{mergeFeaturesByPrevalence}.
\end{enumerate}

Good job! You agglomerated one of the experiments in the MAE and stored it as
an alternative experiment.

\hypertarget{experiment-transformation}{%
\subsection{Experiment transformation}\label{experiment-transformation}}

We proceed with an exercise on a different type of data manipulation, that is,
transformation of assays of individual experiments in the MAE.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load HintikkaXOData with \texttt{data} and store it into a
  variable named \texttt{mae}.
\item
  What assays are contained by each individual experiment? You can check their
  names with \texttt{assays}.
\item
  Apply a log10 transformation to the assay of the metabolite experiment.
  For that you can use \texttt{transformAssay} and don't forget to specify the assay
  to be transformed with the argument \texttt{assay.type}.
\item
  Apply a CLR transformation to the counts assay of the microbiota experiment.
  To ensure non-null values in the assay, set \texttt{pseudocount} equal to 1.
\end{enumerate}

You made it! You learnt how to apply different transformations to the assays
of individual experiments in a MAE with \texttt{transformAssay}, specifying optional
arguments based on the used method.

\hypertarget{assay-extraction}{%
\subsection{Assay extraction}\label{assay-extraction}}

The following exercise walks you through disassembling a MAE object in
order to retrieve a specific assay, or to store its components as
multiple separate csv files.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load HintikkaXOData with \texttt{data} and store it into a
  variable named \texttt{mae}.
\item
  Extract the individual metabolite experiment from the MAE into a distinct
  TreeSE object named \texttt{metabolites}.
\item
  Which and how many assays are contained by \texttt{metabolites}? You can check that
  with \texttt{assays} or \texttt{assayNames}.
\item
  Write a csv file for the nmr assay with \texttt{write.csv}. You can access an
  individual assay of a TreeSE with \texttt{assay} by specifying the name of the
  desired assay.
\item
  \textbf{Extra}: Repeat step 1 thorugh 4 also for the microbiota and biomarkers
  experiments, so that a completely disassembled version of the MAE is available.
\item
  \textbf{Extra}: Besides experiments, MAEs also include a sampleData and a sampleMap,
  which are accessible with \texttt{colData(mae)} and \texttt{sampleMap(mae)}, respectively.
  Save also each of these two elements into a csv file.
\end{enumerate}

Well done! You just splitted a MAE into its components and stored them as csv files.
\href{https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl/blob/main/src/assets/XO_preprocess.R}{This script}
shows a possible approach.

\hypertarget{mae-reconstruction}{%
\subsection{MAE reconstruction}\label{mae-reconstruction}}

Next, we will try to reconstruct the same MAE from the files you created.
Make sure you know their names and location! Alternatively, you can fetch or
download the CSV files in
\href{https://github.com/microbiome/data/tree/main/HintikkaXOData}{this directory}
with the readily disassembled components of HintikkaXOData.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Read in the csv files containing assays with \texttt{read.csv} and save each
  of them into a variable named \texttt{\textless{}assay\ name\textgreater{}\_assays}.
\item
  Create one TreeSE from each assays object with the \texttt{TreeSummarizedExperiment}
  function, as explained in \protect\hyperlink{construct-TreeSE}{this exercise}.
\item
  Read in the sampleData and the sampleMap and store them into the
  variables \texttt{sample\_data} and \texttt{sample\_map}, respectively.
\item
  Combine the components with \texttt{MultiAssayExperiment}, where the first argument
  is an \texttt{ExperimentList} (for now include only the microbiota and metabolites
  TreeSEs), the second is colData and the third is sampleMap.
\item
  Make sure that the MAE experiments are identical to the original TreeSEs. You
  can do that qualitatively by checking their \texttt{head} and quantitatively by
  looking at their \texttt{dim}.
\item
  \textbf{Extra}: Add the biomarkers TreeSE as a new experiment to the MAE.
  Note that new experiments can be added to a MAE through simple concatenation
  with \texttt{c(mae,\ experiment)}.
\end{enumerate}

Good job! Now you are aware of how MAEs are built and we can proceed to some
analytical exercises.

\hypertarget{cross-correlation-analysis}{%
\subsection{Cross-correlation analysis}\label{cross-correlation-analysis}}

Now we will perform a cross-correlation analysis between two of the
experiments in the MAE.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Import the mia package, load HintikkaXOData with \texttt{data} and store it into a
  variable named \texttt{mae}.
\item
  Analyze correlations between the microbiota and the biomarkers experiments
  with \texttt{getExperimentCrossAssociation}. Don't forget to specify the experiments
  you want to compare with the arguments \texttt{experiment1} and \texttt{experiment2}, and
  which of their assays with \texttt{assay.type1} and \texttt{assay.type2}.
\item
  What does the output look like? By default, correlation is measured in terms
  of Kendall tau coefficients. Repeat point 2, but this time change \texttt{method}
  to Spearman coefficients.
\item
  Are you able to infer significance from the output? In order to also obtain
  p-values from the cross-correlation analysis, repeat point 2 with the
  additional argument \texttt{test\_significance\ =\ TRUE}.
\item
  Visualize results with a heatmap similarly to the example in section
  \ref{cross-correlation}. Do you see any significant correlations?
  Interpret your results.
\item
  \textbf{Extra}: Perform cross-correlation analysis between the remaining
  experiments (microbiota vs metabolites and metabolites vs biomarkers) and
  visualize results with heatmaps.
\end{enumerate}

Great job! You performed a cross-correlation analysis between two experiments of
a MAE and visualized the results with a heatmap. You are also able to customise
the correlation method and significance testing used for the analysis.

\hypertarget{part-appendix}{%
\part{Appendix}\label{part-appendix}}

\hypertarget{extras}{%
\chapter{Extra material}\label{extras}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\SpecialCharTok{::}\NormalTok{opts\_chunk}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{eval=}\ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{slides}{%
\section{Slides}\label{slides}}

\href{https://microbiome.github.io/outreach/}{Outreach material} includes
slide sets for training events.

\hypertarget{compare-permanova}{%
\section{PERMANOVA comparison}\label{compare-permanova}}

Here we present two possible uses of the \texttt{adonis2} function which performs PERMANOVA. The
optional argument \texttt{by} has an effect on the statistical outcome, so its two options are
compared here.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# import necessary packages}
\FunctionTok{library}\NormalTok{(gtools)}
\FunctionTok{library}\NormalTok{(purrr)}
\FunctionTok{library}\NormalTok{(vegan)}
\FunctionTok{library}\NormalTok{(gtools)}
\FunctionTok{library}\NormalTok{(purrr)}
\end{Highlighting}
\end{Shaded}

Let us load the \emph{enterotype} TSE object and run PERMANOVA for
different orders of three variables with two different approaches:
\texttt{by\ =\ "margin"} or \texttt{by\ =\ "terms"}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# load and prepare data}
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{data}\NormalTok{(}\StringTok{"enterotype"}\NormalTok{, }\AttributeTok{package=}\StringTok{"mia"}\NormalTok{)}
\NormalTok{enterotype }\OtherTok{\textless{}{-}} \FunctionTok{transformAssay}\NormalTok{(enterotype, }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\CommentTok{\# drop samples missing meta data}
\NormalTok{enterotype }\OtherTok{\textless{}{-}}\NormalTok{ enterotype[ , }\SpecialCharTok{!}\FunctionTok{rowSums}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{colData}\NormalTok{(enterotype)[, }\FunctionTok{c}\NormalTok{(}\StringTok{"Nationality"}\NormalTok{, }\StringTok{"Gender"}\NormalTok{, }\StringTok{"ClinicalStatus"}\NormalTok{)]) }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{)]}
\CommentTok{\# define variables and list all possible combinations}
\NormalTok{vars }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Nationality"}\NormalTok{, }\StringTok{"Gender"}\NormalTok{, }\StringTok{"ClinicalStatus"}\NormalTok{)}
\NormalTok{var\_perm }\OtherTok{\textless{}{-}} \FunctionTok{permutations}\NormalTok{(}\AttributeTok{n =} \DecValTok{3}\NormalTok{, }\AttributeTok{r =} \DecValTok{3}\NormalTok{, vars)}
\NormalTok{formulas }\OtherTok{\textless{}{-}} \FunctionTok{apply}\NormalTok{(var\_perm, }\DecValTok{1}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(row) purrr}\SpecialCharTok{::}\FunctionTok{reduce}\NormalTok{(row, }\ControlFlowTok{function}\NormalTok{(x, y) }\FunctionTok{paste}\NormalTok{(x, }\StringTok{"+"}\NormalTok{, y)))}
\CommentTok{\# create empty data.frames for further storing p{-}values}
\NormalTok{terms\_df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\StringTok{"Formula"} \OtherTok{=}\NormalTok{ formulas,}
                       \StringTok{"ClinicalStatus"} \OtherTok{=} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{6}\NormalTok{),}
                       \StringTok{"Gender"} \OtherTok{=} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{6}\NormalTok{),}
                       \StringTok{"Nationality"} \OtherTok{=} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{6}\NormalTok{))}
\NormalTok{margin\_df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\StringTok{"Formula"} \OtherTok{=}\NormalTok{ formulas,}
                        \StringTok{"ClinicalStatus"} \OtherTok{=} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{6}\NormalTok{),}
                        \StringTok{"Gender"} \OtherTok{=} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{6}\NormalTok{),}
                        \StringTok{"Nationality"} \OtherTok{=} \FunctionTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{, }\DecValTok{6}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{for}\NormalTok{ (row\_idx }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{nrow}\NormalTok{(var\_perm)) \{}
  
  \CommentTok{\# generate temporary formula (i.e. "assay \textasciitilde{} ClinicalStatus + Nationality + Gender")}
\NormalTok{  tmp\_formula }\OtherTok{\textless{}{-}}\NormalTok{ purrr}\SpecialCharTok{::}\FunctionTok{reduce}\NormalTok{(var\_perm[row\_idx, ], }\ControlFlowTok{function}\NormalTok{(x, y) }\FunctionTok{paste}\NormalTok{(x, }\StringTok{"+"}\NormalTok{, y))}
\NormalTok{  tmp\_formula }\OtherTok{\textless{}{-}} \FunctionTok{as.formula}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(}\StringTok{\textquotesingle{}t(assay(enterotype, "relabundance")) \textasciitilde{} \textquotesingle{}}\NormalTok{,}
\NormalTok{                            tmp\_formula))}

  \CommentTok{\# multiple variables, default: by = "terms"}
  \FunctionTok{set.seed}\NormalTok{(}\DecValTok{75}\NormalTok{)}
\NormalTok{  with\_terms }\OtherTok{\textless{}{-}} \FunctionTok{adonis2}\NormalTok{(tmp\_formula, }
                \AttributeTok{by =} \StringTok{"terms"}\NormalTok{,}
                \AttributeTok{data =} \FunctionTok{colData}\NormalTok{(enterotype),}
                \AttributeTok{permutations =} \DecValTok{99}\NormalTok{)}
  
  \CommentTok{\# multiple variables, by = "margin"}
  \FunctionTok{set.seed}\NormalTok{(}\DecValTok{75}\NormalTok{)}
\NormalTok{  with\_margin }\OtherTok{\textless{}{-}} \FunctionTok{adonis2}\NormalTok{(tmp\_formula, }
                 \AttributeTok{by =} \StringTok{"margin"}\NormalTok{,}
                 \AttributeTok{data =} \FunctionTok{colData}\NormalTok{(enterotype),}
                 \AttributeTok{permutations =} \DecValTok{99}\NormalTok{)}

  \CommentTok{\# extract p{-}values}
\NormalTok{  terms\_p }\OtherTok{\textless{}{-}}\NormalTok{ with\_terms[[}\StringTok{"Pr(\textgreater{}F)"}\NormalTok{]]}
\NormalTok{  terms\_p }\OtherTok{\textless{}{-}}\NormalTok{ terms\_p[}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(terms\_p)]}
\NormalTok{  margin\_p }\OtherTok{\textless{}{-}}\NormalTok{ with\_margin[[}\StringTok{"Pr(\textgreater{}F)"}\NormalTok{]]}
\NormalTok{  margin\_p }\OtherTok{\textless{}{-}}\NormalTok{ margin\_p[}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(margin\_p)]}
  
  \CommentTok{\# store p{-}values into data.frames}
  \ControlFlowTok{for}\NormalTok{ (col\_idx }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{ncol}\NormalTok{(var\_perm)) \{}
    
\NormalTok{    terms\_df[var\_perm[row\_idx, col\_idx]][row\_idx, ] }\OtherTok{\textless{}{-}}\NormalTok{ terms\_p[col\_idx]}
\NormalTok{    margin\_df[var\_perm[row\_idx, col\_idx]][row\_idx, ] }\OtherTok{\textless{}{-}}\NormalTok{ margin\_p[col\_idx]}
    
\NormalTok{  \}}
  
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

The following table displays the p-values for the three variables
ClinicalStatus, Gender and Nationality obtained by PERMANOVA with
\texttt{adonis2}. Note that the p-values remain identical when \texttt{by\ =\ "margin"}, but change with the order of the variables in the
formula when \texttt{by\ =\ "terms"} (default).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{df }\OtherTok{\textless{}{-}}\NormalTok{ terms\_df }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{inner\_join}\NormalTok{(margin\_df, }\AttributeTok{by =} \StringTok{"Formula"}\NormalTok{, }\AttributeTok{suffix =} \FunctionTok{c}\NormalTok{(}\StringTok{" (terms)"}\NormalTok{, }\StringTok{" (margin)"}\NormalTok{))}

\NormalTok{knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{(df)}
\end{Highlighting}
\end{Shaded}

\hypertarget{bayesian-multinomial-logistic-normal-models}{%
\section{Bayesian Multinomial Logistic-Normal Models}\label{bayesian-multinomial-logistic-normal-models}}

Analysis using such model could be performed with the function
\texttt{pibble} from the \texttt{fido} package, wihch is in form of a Multinomial
Logistic-Normal Linear Regression model; see
\href{https://jsilve24.github.io/fido/articles/introduction-to-fido.html}{vignette}
of package.

The following presents such an exemplary analysis based on the
data of \citet{Sprockett2020} available
through \texttt{microbiomeDataSets} package.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(fido)}
\end{Highlighting}
\end{Shaded}

Loading the libraries and importing data:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(fido)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(microbiomeDataSets)}
\NormalTok{tse }\OtherTok{\textless{}{-}} \FunctionTok{SprockettTHData}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

We pick three covariates (``Sex'',``Age\_Years'',``Delivery\_Mode'') during this
analysis as an example, and beforehand we check for missing data:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(mia)}
\NormalTok{cov\_names }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Sex"}\NormalTok{,}\StringTok{"Age\_Years"}\NormalTok{,}\StringTok{"Delivery\_Mode"}\NormalTok{)}
\NormalTok{na\_counts }\OtherTok{\textless{}{-}} \FunctionTok{apply}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)[,cov\_names]), }\DecValTok{2}\NormalTok{, sum)}
\NormalTok{na\_summary}\OtherTok{\textless{}{-}}\FunctionTok{as.data.frame}\NormalTok{(na\_counts,}\AttributeTok{row.names=}\NormalTok{cov\_names)}
\end{Highlighting}
\end{Shaded}

We drop missing values of the covariates:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ tse[ , }\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Delivery\_Mode) ]}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ tse[ , }\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse)}\SpecialCharTok{$}\NormalTok{Age\_Years) ]}
\end{Highlighting}
\end{Shaded}

We agglomerate microbiome data to Phylum:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tse\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\StringTok{"Phylum"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We extract the counts assay and covariate data to build the model
matrix:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{Y }\OtherTok{\textless{}{-}} \FunctionTok{assays}\NormalTok{(tse\_phylum)}\SpecialCharTok{$}\NormalTok{counts}
\CommentTok{\# design matrix}
\CommentTok{\# taking 3 covariates}
\NormalTok{sample\_data}\OtherTok{\textless{}{-}}\FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse\_phylum)[,cov\_names])}
\NormalTok{X }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(}\FunctionTok{model.matrix}\NormalTok{(}\SpecialCharTok{\textasciitilde{}}\NormalTok{Sex}\SpecialCharTok{+}\NormalTok{Age\_Years}\SpecialCharTok{+}\NormalTok{Delivery\_Mode,}\AttributeTok{data=}\NormalTok{sample\_data))}
\end{Highlighting}
\end{Shaded}

Building the parameters for the \texttt{pibble} call to build the model; see more at \href{https://jsilve24.github.io/fido/articles/introduction-to-fido.html}{vignette}:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{n\_taxa}\OtherTok{\textless{}{-}}\FunctionTok{nrow}\NormalTok{(Y)}
\NormalTok{upsilon }\OtherTok{\textless{}{-}}\NormalTok{ n\_taxa}\SpecialCharTok{+}\DecValTok{3}
\NormalTok{Omega }\OtherTok{\textless{}{-}} \FunctionTok{diag}\NormalTok{(n\_taxa)}
\NormalTok{G }\OtherTok{\textless{}{-}} \FunctionTok{cbind}\NormalTok{(}\FunctionTok{diag}\NormalTok{(n\_taxa}\DecValTok{{-}1}\NormalTok{), }\SpecialCharTok{{-}}\DecValTok{1}\NormalTok{)}
\NormalTok{Xi }\OtherTok{\textless{}{-}}\NormalTok{ (upsilon}\SpecialCharTok{{-}}\NormalTok{n\_taxa)}\SpecialCharTok{*}\NormalTok{G}\SpecialCharTok{\%*\%}\NormalTok{Omega}\SpecialCharTok{\%*\%}\FunctionTok{t}\NormalTok{(G)}
\NormalTok{Theta }\OtherTok{\textless{}{-}} \FunctionTok{matrix}\NormalTok{(}\DecValTok{0}\NormalTok{, n\_taxa}\DecValTok{{-}1}\NormalTok{, }\FunctionTok{nrow}\NormalTok{(X))}
\NormalTok{Gamma }\OtherTok{\textless{}{-}} \FunctionTok{diag}\NormalTok{(}\FunctionTok{nrow}\NormalTok{(X))}
\end{Highlighting}
\end{Shaded}

Automatically initializing the priors and visualizing their distributions:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{priors }\OtherTok{\textless{}{-}} \FunctionTok{pibble}\NormalTok{(}\ConstantTok{NULL}\NormalTok{, X, upsilon, Theta, Gamma, Xi)}
\FunctionTok{names\_covariates}\NormalTok{(priors) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(X)}
\FunctionTok{plot}\NormalTok{(priors, }\AttributeTok{pars=}\StringTok{"Lambda"}\NormalTok{) }\SpecialCharTok{+}\NormalTok{ ggplot2}\SpecialCharTok{::}\FunctionTok{xlim}\NormalTok{(}\FunctionTok{c}\NormalTok{(}\SpecialCharTok{{-}}\DecValTok{5}\NormalTok{, }\DecValTok{5}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Estimating the posterior by including our response data \texttt{Y}.
Note: Some computational failures could occur (see \href{https://github-wiki-see.page/m/jsilve24/fido/wiki/Frequently-Asked-Questions}{discussion})
the arguments \texttt{multDirichletBoot} \texttt{calcGradHess} could be passed in such case.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{priors}\SpecialCharTok{$}\NormalTok{Y }\OtherTok{\textless{}{-}}\NormalTok{ Y }
\NormalTok{posterior }\OtherTok{\textless{}{-}} \FunctionTok{refit}\NormalTok{(priors, }\AttributeTok{optim\_method=}\StringTok{"adam"}\NormalTok{, }\AttributeTok{multDirichletBoot=}\FloatTok{0.5}\NormalTok{) }\CommentTok{\#calcGradHess=FALSE}
\end{Highlighting}
\end{Shaded}

Printing a summary about the posterior:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ppc\_summary}\NormalTok{(posterior)}
\end{Highlighting}
\end{Shaded}

Plotting the summary of the posterior distributions of the regression parameters:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{names\_categories}\NormalTok{(posterior) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(Y)}
\FunctionTok{plot}\NormalTok{(posterior,}\AttributeTok{par=}\StringTok{"Lambda"}\NormalTok{,}\AttributeTok{focus.cov=}\FunctionTok{rownames}\NormalTok{(X)[}\DecValTok{2}\SpecialCharTok{:}\DecValTok{4}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

Taking a closer look at ``Sex'' and ``Delivery\_Mode'':

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(posterior, }\AttributeTok{par=}\StringTok{"Lambda"}\NormalTok{, }\AttributeTok{focus.cov =} \FunctionTok{rownames}\NormalTok{(X)[}\FunctionTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{4}\NormalTok{)])}
\end{Highlighting}
\end{Shaded}

\hypertarget{interactive-3d-plots}{%
\section{Interactive 3D Plots}\label{interactive-3d-plots}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load libraries}
\FunctionTok{library}\NormalTok{(rgl)}
\FunctionTok{library}\NormalTok{(plotly)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(knitr)}
\NormalTok{knitr}\SpecialCharTok{::}\NormalTok{knit\_hooks}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{webgl =}\NormalTok{ hook\_webgl)}
\end{Highlighting}
\end{Shaded}

In this section we make a 3D version of the earlier Visualizing the most dominant genus on PCoA (see \ref{quality-control}), with the help of the plotly \citep{Sievert2020}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Importing necessary libraries}
\FunctionTok{library}\NormalTok{(curatedMetagenomicData)}
\FunctionTok{library}\NormalTok{(dplyr)}
\FunctionTok{library}\NormalTok{(DT)}
\FunctionTok{library}\NormalTok{(mia)}
\FunctionTok{library}\NormalTok{(scater)}

\CommentTok{\# Querying the data}
\NormalTok{tse }\OtherTok{\textless{}{-}}\NormalTok{ sampleMetadata }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{filter}\NormalTok{(age }\SpecialCharTok{\textgreater{}=} \DecValTok{18}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} \CommentTok{\# taking only data of age 18 or above}
    \FunctionTok{filter}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(alcohol)) }\SpecialCharTok{\%\textgreater{}\%} \CommentTok{\# excluding missing values}
    \FunctionTok{returnSamples}\NormalTok{(}\StringTok{"relative\_abundance"}\NormalTok{)}

\NormalTok{tse\_Genus }\OtherTok{\textless{}{-}} \FunctionTok{mergeFeaturesByRank}\NormalTok{(tse, }\AttributeTok{rank=}\StringTok{"genus"}\NormalTok{)}
\NormalTok{tse\_Genus }\OtherTok{\textless{}{-}} \FunctionTok{addPerSampleDominantFeatures}\NormalTok{(tse\_Genus,}\AttributeTok{assay.type=}\StringTok{"relative\_abundance"}\NormalTok{, }\AttributeTok{name =} \StringTok{"dominant\_taxa"}\NormalTok{)}

\CommentTok{\# Performing PCoA with Bray{-}Curtis dissimilarity.}
\NormalTok{tse\_Genus }\OtherTok{\textless{}{-}} \FunctionTok{runMDS}\NormalTok{(tse\_Genus, }\AttributeTok{FUN =}\NormalTok{ vegan}\SpecialCharTok{::}\NormalTok{vegdist, }\AttributeTok{ncomponents =} \DecValTok{3}\NormalTok{,}
              \AttributeTok{name =} \StringTok{"PCoA\_BC"}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"relative\_abundance"}\NormalTok{)}

\CommentTok{\# Getting the 6 top taxa}
\NormalTok{top\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{getTopFeatures}\NormalTok{(tse\_Genus,}\AttributeTok{top =} \DecValTok{6}\NormalTok{, }\AttributeTok{assay.type =} \StringTok{"relative\_abundance"}\NormalTok{)}

\CommentTok{\# Naming all the rest of non top{-}taxa as "Other"}
\NormalTok{most\_abundant }\OtherTok{\textless{}{-}} \FunctionTok{lapply}\NormalTok{(}\FunctionTok{colData}\NormalTok{(tse\_Genus)}\SpecialCharTok{$}\NormalTok{dominant\_taxa,}
                   \ControlFlowTok{function}\NormalTok{(x)\{}\ControlFlowTok{if}\NormalTok{ (x }\SpecialCharTok{\%in\%}\NormalTok{ top\_taxa) \{x\} }\ControlFlowTok{else}\NormalTok{ \{}\StringTok{"Other"}\NormalTok{\}\})}

\CommentTok{\# Storing the previous results as a new column within colData}
\FunctionTok{colData}\NormalTok{(tse\_Genus)}\SpecialCharTok{$}\NormalTok{most\_abundant }\OtherTok{\textless{}{-}} \FunctionTok{as.character}\NormalTok{(most\_abundant)}

\CommentTok{\# Calculating percentage of the most abundant}
\NormalTok{most\_abundant\_freq }\OtherTok{\textless{}{-}} \FunctionTok{table}\NormalTok{(}\FunctionTok{as.character}\NormalTok{(most\_abundant))}
\NormalTok{most\_abundant\_percent }\OtherTok{\textless{}{-}} \FunctionTok{round}\NormalTok{(most\_abundant\_freq}\SpecialCharTok{/}\FunctionTok{sum}\NormalTok{(most\_abundant\_freq)}\SpecialCharTok{*}\DecValTok{100}\NormalTok{, }\DecValTok{1}\NormalTok{)}

\CommentTok{\# Retrieving the explained variance}
\NormalTok{e }\OtherTok{\textless{}{-}} \FunctionTok{attr}\NormalTok{(}\FunctionTok{reducedDim}\NormalTok{(tse\_Genus, }\StringTok{"PCoA\_BC"}\NormalTok{), }\StringTok{"eig"}\NormalTok{);}
\NormalTok{var\_explained }\OtherTok{\textless{}{-}}\NormalTok{ e}\SpecialCharTok{/}\FunctionTok{sum}\NormalTok{(e[e}\SpecialCharTok{\textgreater{}}\DecValTok{0}\NormalTok{])}\SpecialCharTok{*}\DecValTok{100}
\end{Highlighting}
\end{Shaded}

\hypertarget{developers}{%
\chapter*{Developers}\label{developers}}
\addcontentsline{toc}{chapter}{Developers}

\hypertarget{core-team}{%
\subsection*{Core team}\label{core-team}}
\addcontentsline{toc}{subsection}{Core team}

Contributions to this Gitbook from the various developers are
coordinated by:

\begin{itemize}
\item
  \emph{Leo Lahti, DSc}, professor in Data Science at the \href{https://datascience.utu.fi/}{Department of
  Computing, University of Turku,
  Finland}, with a focus on
  computational microbiome analysis. Lahti obtained doctoral degree
  (DSc) from Aalto University in Finland (2010), developing
  probabilistic machine learning with applications to high-throughput
  life science data integration. Since then he has focused on
  microbiome research and developed, for instance, the
  \emph{phyloseq}-based \href{https://bioconductor.org/packages/release/bioc/html/microbiome.html}{microbiome R
  package}
  before starting to develop the \emph{TreeSummarizedExperiment} /
  \emph{MultiAssayExperiment} framework and the mia family of Bioconductor
  packages for microbiome data science introduced in this
  gitbook. Lahti led the development of \href{https://avointiede.fi/en/policies-materials/policies-open-science-and-research-finland/policy-open-research-data-and-methods}{national policy on open
  access to research methods in
  Finland}.
  He is current member in the \href{https://bioconductor.org/about/community-advisory-board/}{Bioconductor Community Advisory
  Board}
  and runs regular training workshops in microbiome data science.
\item
  \emph{Tuomas Borman}, PhD researcher and the lead developer of OMA/mia at
  the Department of Computing, University of Turku.
\end{itemize}

\hypertarget{contributors}{%
\subsection*{Contributors}\label{contributors}}
\addcontentsline{toc}{subsection}{Contributors}

This work is a remarkably collaborative effort. The full list of
contributors is available via
\href{https://github.com/microbiome/OMA/graphs/contributors}{Github}. Some
key authors/contributors include:

\begin{itemize}
\item
  \emph{Felix Ernst, PhD}, among the first developers of R/Bioc methods for
  microbiome research based on the \emph{SummarizedExperiment} class and
  its derivatives.
\item
  \emph{Giulio Benedetti}, scientific programmer at the Department of
  Computing, University of Turku. His research interest is mostly
  related to Data Science. He has also helped to expand the
  SummarizedExperiment-based microbiome analysis framework to the
  Julia language, implementing
  \href{https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl}{MicrobiomeAnalysis.jl}.
\item
  \emph{Sudarshan Shetty, PhD} has supported the establishment of the
  framework and associated tools. He also maintains a list of
  \href{https://microsud.github.io/Tools-Microbiome-Analysis/}{microbiome R
  packages}.
\item
  \emph{Henrik Eckermann}, in particular to the development of
  the differential abundance analyses
\item
  \emph{Chouaib Benchraka} provided various contributions to the package ecosystem and the OMA book
\item
  \emph{Yağmur Şimşek} converted the miaSim R package to support the Bioconductor framework
\item
  \emph{Basil Courbayre} provided various contributions to the package
  ecosystem and the OMA book, in particular on unsupervised machine
  learning
\item
  \emph{Matti Ruuskanen, PhD}, added machine learning techniques for
  microbiome analysis
\item
  \emph{Stefanie Peschel} has contributed chapters on the construction, analysis, and comparison of microbial association networks.
\item
  \emph{Christian L. Müller}, group leader at the \href{https://www.helmholtz-munich.de/en/computational-health-center}{Computational Health Center, Helmholtz Zentrum München, Germany} and a Professor for Biomedical Statistics and Data Science at \href{https://www.en.statistik.uni-muenchen.de/index.html}{LMU Munich}. He assisted in writing the chapters on network learning and comparison.
\item
  \emph{Shigdel Rajesh, PhD}
\item
  \emph{Artur Sannikov}
\item
  \emph{Jeba Akewak}
\item
  \emph{Himmi Lindgren}
\item
  \emph{Lu Yang}
\end{itemize}

\hypertarget{acknowledgments}{%
\subsection*{Acknowledgments}\label{acknowledgments}}
\addcontentsline{toc}{subsection}{Acknowledgments}

This work would not have been possible without the countless
contributions and interactions with other researchers, developers, and
users. We express our gratitude to the entire Bioconductor community
for developing this high-quality open research software repository for
life science analytics, continuously pushing the limits in emerging
fields \citep{Gentleman2004}, \citep{Huber2015}.

The presented framework for microbiome data science is based on the
\emph{TreeSummarizedExperiment} data container created by Ruizhu Huang and
others \citep{R_TreeSummarizedExperiment}, \citep{Ernst2020}, and on the
\emph{MultiAssayExperiment} by Marcel Ramos et al. \citep{Ramos2017}. The idea
of using these containers as a basis for microbiome data science was
initially advanced by the groundwork of Domenick Braccia, Héctor
Corrada Bravo and others and brought together with other microbiome
data science developers \citep{Shetty2019}. Setting up the base ecosystem
of packages and tutorials was then subsequently led by Tuomas Borman,
Felix Ernst, and \href{http://www.iki.fi/Leo.Lahti}{Leo Lahti}. We would
specifically like to thank everyone who contributed to the work
supporting the \emph{TreeSummarizedExperiment} ecosystem for microbiome
research, including but not limited to the R packages mia, miaViz,
miaTime, miaSim, philr, ANCOMBC, curatedMetagenomicData, scater,
scuttle, and other packages, some of which are listed in Section
\ref{ecosystem}. A number of other contributors have advanced the
ecosystem further, and will be acknowledged in the individual
packages, \href{https://github.com/microbiome/OMA/graphs/contributors}{pull
requests},
\href{https://github.com/microbiome/OMA/issues}{issues}, and other work.

Ample demonstration data resources supporting this framework have been
made available through the
\href{https://waldronlab.io/curatedMetagenomicData/}{curatedMetagenomicData}
project by Edoardo Pasolli, Lucas Schiffer, Levi Waldron and others
\citep{Pasolli2017}.

The work has drawn initial inspiration from many sources, most notably
from the work on \emph{phyloseq} by Paul McMurdie and Susan Holmes
\citep{McMurdie2013} who pioneered the work on rigorous and reproducible
microbiome data science ecosystems in R/Bioconductor. The phyloseq
framework continues to provide a vast array of complementary packages
and methods for microbiome studies. The Orchestrating Single-Cell
Analysis with Bioconductor, or
\href{https://bioconductor.org/books/release/OSCA/}{\emph{OSCA}} book by Robert
Amezquita, Aaron Lun, Stephanie Hicks, and Raphael Gottardo
\citep{Amezquita2020natmeth} has implemented closely related work on the
\emph{SummarizedExperiment} data container and its derivatives in the field
of single cell sequencing studies that have inspired this work.

In the background, the open source books by Susan Holmes and Wolfgang
Huber, Modern Statistics for Modern Biology \citep{Holmes2019} and by
Garret Grolemund and Hadley Wickham, the R for Data Science
\citep{Grolemund2017}, and Richard McElreath's Statistical Rethinking and
the associated online resources by Solomon Kurz \citep{McElreath2020} are
key references that have advanced reproducible data science training
and dissemination.

\hypertarget{how-to-contribute}{%
\subsection{How to contribute}\label{how-to-contribute}}

To contribute reports, follow the Git flow procedure (you can see instructions
to \href{https://docs.github.com/en/get-started}{getting started with Github}):

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Fork the project
\item
  Clone your fork
\item
  Modify the material
\item
  Check locally that the changes render successfully (see above)
\item
  Add and commit the changes to your fork
\item
  Create a pull request (PR) from your fork back to the original repo
\item
  Fix and discuss issues in the review process
\end{enumerate}

More detailed instructions for contributing can be found on OMA \href{https://github.com/microbiome/OMA/blob/master/README.md}{README}.

\hypertarget{support}{%
\subsection*{Support}\label{support}}
\addcontentsline{toc}{subsection}{Support}

This work has been supported by:

\begin{itemize}
\item
  \href{https://www.aka.fi/}{Research Council of Finland}
\item
  \href{https://www.findingpheno.eu/}{FindingPheno} European Union's Horizon 2020 research and innovation programme under grant agreement No 952914
\item
  COST Action network on Statistical and Machine Learning Techniques for Human Microbiome Studies
  (\href{https://www.ml4microbiome.eu/}{ML4microbiome}) \citep{MorenoIndias2021}.
\item
  Computational Life Science Research Program, \href{https://biocityturku.fi/}{Biocity Turku}
\item
  \href{https://www.yliopistosaatio.fi/en/}{Turku University Foundation}
\end{itemize}

\hypertarget{sessioninfo}{%
\chapter*{Sessioninfo}\label{sessioninfo}}
\addcontentsline{toc}{chapter}{Sessioninfo}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sessionInfo}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## R version 4.3.1 (2023-06-16)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 22.04.3 LTS
## 
## Matrix products: default
## 
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: UTC
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] BiocStyle_2.28.1 rebook_1.10.1   
## 
## loaded via a namespace (and not attached):
## [1] knitr_1.45      shiny_1.8.0     htmltools_0.5.7 rmarkdown_2.25 
## [5] bookdown_0.36   miniUI_0.1.1.1  tools_4.3.1
\end{verbatim}

  \bibliography{book.bib}

\end{document}
