<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Taxonomic Information | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 6 Taxonomic Information | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Taxonomic Information | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Taxonomic Information | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="quality-control.html"/>
<link rel="next" href="community-diversity.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#human-microbiome-compendium"><i class="fa fa-check"></i><b>3.3.4</b> Human microbiome compendium</a></li>
<li class="chapter" data-level="3.3.5" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.5</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="network-learning.html"><a href="network-learning.html"><i class="fa fa-check"></i><b>12</b> Network learning and analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-1"><i class="fa fa-check"></i><b>12.1</b> Network learning</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="network-learning.html"><a href="network-learning.html#network-learning-workflow"><i class="fa fa-check"></i><b>12.1.1</b> Typical workflow</a></li>
<li class="chapter" data-level="12.1.2" data-path="network-learning.html"><a href="network-learning.html#install-packages"><i class="fa fa-check"></i><b>12.1.2</b> Install packages</a></li>
<li class="chapter" data-level="12.1.3" data-path="network-learning.html"><a href="network-learning.html#data-preparation"><i class="fa fa-check"></i><b>12.1.3</b> Data preparation</a></li>
<li class="chapter" data-level="12.1.4" data-path="network-learning.html"><a href="network-learning.html#spring-network"><i class="fa fa-check"></i><b>12.1.4</b> SPRING network</a></li>
<li class="chapter" data-level="12.1.5" data-path="network-learning.html"><a href="network-learning.html#netcomi-spring"><i class="fa fa-check"></i><b>12.1.5</b> NetCoMi network</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="network-learning.html"><a href="network-learning.html#network-analysis"><i class="fa fa-check"></i><b>12.2</b> Network analysis with igraph</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="network-learning.html"><a href="network-learning.html#network-plot"><i class="fa fa-check"></i><b>12.2.1</b> Network plot</a></li>
<li class="chapter" data-level="12.2.2" data-path="network-learning.html"><a href="network-learning.html#centrality-measures"><i class="fa fa-check"></i><b>12.2.2</b> Centrality measures</a></li>
<li class="chapter" data-level="12.2.3" data-path="network-learning.html"><a href="network-learning.html#scale-node-sizes-by-centrality"><i class="fa fa-check"></i><b>12.2.3</b> Scale node sizes by centrality</a></li>
<li class="chapter" data-level="12.2.4" data-path="network-learning.html"><a href="network-learning.html#degree-distribution"><i class="fa fa-check"></i><b>12.2.4</b> Degree distribution</a></li>
<li class="chapter" data-level="12.2.5" data-path="network-learning.html"><a href="network-learning.html#clustered-heatmaps"><i class="fa fa-check"></i><b>12.2.5</b> Clustered heatmaps</a></li>
<li class="chapter" data-level="12.2.6" data-path="network-learning.html"><a href="network-learning.html#global-network-measures"><i class="fa fa-check"></i><b>12.2.6</b> Global network measures</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="network-learning.html"><a href="network-learning.html#network-analysis-with-netcomi"><i class="fa fa-check"></i><b>12.3</b> Network analysis with NetCoMi</a></li>
<li class="chapter" data-level="12.4" data-path="network-learning.html"><a href="network-learning.html#network-visualization"><i class="fa fa-check"></i><b>12.4</b> Network visualization</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="network-learning.html"><a href="network-learning.html#highlight-node-properties"><i class="fa fa-check"></i><b>12.4.1</b> Highlight node properties</a></li>
<li class="chapter" data-level="12.4.2" data-path="network-learning.html"><a href="network-learning.html#highlight-data-features"><i class="fa fa-check"></i><b>12.4.2</b> Highlight data features</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="network-learning.html"><a href="network-learning.html#which-methods-to-choose"><i class="fa fa-check"></i><b>12.5</b> Which method(s) to choose?</a></li>
<li class="chapter" data-level="12.6" data-path="network-learning.html"><a href="network-learning.html#more-about-association"><i class="fa fa-check"></i><b>12.6</b> More about association measures</a></li>
<li class="chapter" data-level="12.7" data-path="network-learning.html"><a href="network-learning.html#comparison-of-association-measures"><i class="fa fa-check"></i><b>12.7</b> Comparison of association measures</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="network-learning.html"><a href="network-learning.html#sparcc-correlation"><i class="fa fa-check"></i><b>12.7.1</b> SparCC</a></li>
<li class="chapter" data-level="12.7.2" data-path="network-learning.html"><a href="network-learning.html#shrinkage-prop"><i class="fa fa-check"></i><b>12.7.2</b> Shrinkage proportionality</a></li>
<li class="chapter" data-level="12.7.3" data-path="network-learning.html"><a href="network-learning.html#spieceasi-mb"><i class="fa fa-check"></i><b>12.7.3</b> SpiecEasi - MB</a></li>
<li class="chapter" data-level="12.7.4" data-path="network-learning.html"><a href="network-learning.html#network-plots"><i class="fa fa-check"></i><b>12.7.4</b> Network plots</a></li>
<li class="chapter" data-level="12.7.5" data-path="network-learning.html"><a href="network-learning.html#network-analysis-1"><i class="fa fa-check"></i><b>12.7.5</b> Network analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="network-comparison.html"><a href="network-comparison.html"><i class="fa fa-check"></i><b>13</b> Network comparison</a>
<ul>
<li class="chapter" data-level="13.1" data-path="network-comparison.html"><a href="network-comparison.html#data-preparation-1"><i class="fa fa-check"></i><b>13.1</b> Data preparation</a></li>
<li class="chapter" data-level="13.2" data-path="network-comparison.html"><a href="network-comparison.html#network-learning-and-analysis"><i class="fa fa-check"></i><b>13.2</b> Network learning and analysis</a></li>
<li class="chapter" data-level="13.3" data-path="network-comparison.html"><a href="network-comparison.html#differential-network-analysis"><i class="fa fa-check"></i><b>13.3</b> Differential network analysis</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="network-comparison.html"><a href="network-comparison.html#visual-comparison"><i class="fa fa-check"></i><b>13.3.1</b> Visual comparison</a></li>
<li class="chapter" data-level="13.3.2" data-path="network-comparison.html"><a href="network-comparison.html#quantitative-comparison"><i class="fa fa-check"></i><b>13.3.2</b> Quantitative comparison</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="network-comparison.html"><a href="network-comparison.html#differential-association-analysis"><i class="fa fa-check"></i><b>13.4</b> Differential association analysis</a></li>
<li class="chapter" data-level="13.5" data-path="network-comparison.html"><a href="network-comparison.html#netcomp-methods"><i class="fa fa-check"></i><b>13.5</b> Network comparison methods</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>14</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="14.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>14.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="14.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>14.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>15</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="15.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>15.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="15.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>15.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>16</b> Visualization</a>
<ul>
<li class="chapter" data-level="16.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>16.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>16.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="16.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>16.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>16.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>16.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="16.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>16.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>16.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>16.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="17" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>17</b> Training</a>
<ul>
<li class="chapter" data-level="17.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>17.1</b> Checklist</a></li>
<li class="chapter" data-level="17.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>17.2</b> Recommended software</a></li>
<li class="chapter" data-level="17.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>17.3</b> Study material</a></li>
<li class="chapter" data-level="17.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>17.4</b> Support and resources</a></li>
<li class="chapter" data-level="17.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>17.5</b> Further reading</a></li>
<li class="chapter" data-level="17.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>17.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>18</b> Resources</a>
<ul>
<li class="chapter" data-level="18.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>18.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>18.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="18.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>18.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="18.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>18.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>18.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>18.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="18.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>18.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>18.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>18.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="18.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>18.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>19</b> Exercises</a>
<ul>
<li class="chapter" data-level="19.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>19.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="19.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>19.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>19.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>19.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>19.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="19.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>19.3.2</b> Importing data</a></li>
<li class="chapter" data-level="19.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>19.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="19.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>19.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="19.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>19.3.5</b> Sample information</a></li>
<li class="chapter" data-level="19.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>19.3.6</b> Feature information</a></li>
<li class="chapter" data-level="19.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>19.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>19.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>19.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="19.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>19.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="19.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>19.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="19.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>19.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="19.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>19.4.5</b> Other functions</a></li>
<li class="chapter" data-level="19.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>19.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>19.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="19.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>19.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="19.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>19.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>19.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="19.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>19.6.1</b> Estimation</a></li>
<li class="chapter" data-level="19.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>19.6.2</b> Visualization</a></li>
<li class="chapter" data-level="19.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>19.6.3</b> Correlation</a></li>
<li class="chapter" data-level="19.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>19.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>19.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="19.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>19.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="19.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>19.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="19.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>19.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="19.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>19.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="19.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>19.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="19.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>19.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="19.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>19.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="19.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>19.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="19.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>19.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="19.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>19.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="19.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>19.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="19.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>19.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="19.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>19.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="19.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>19.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="19.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>19.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="19.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>19.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="19.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>19.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="19.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>19.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="19.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>19.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="19.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>19.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="19.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>19.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="19.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>19.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="20" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>20</b> Extra material</a>
<ul>
<li class="chapter" data-level="20.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>20.1</b> Slides</a></li>
<li class="chapter" data-level="20.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>20.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="20.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>20.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="20.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>20.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="20.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>20.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="taxonomic-information" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Taxonomic Information<a href="taxonomic-information.html#taxonomic-information" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="taxonomic-information.html#cb157-1" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb157-2"><a href="taxonomic-information.html#cb157-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb157-3"><a href="taxonomic-information.html#cb157-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span></code></pre></div>
<p>Taxonomic information is a key part of analyzing microbiome data and without
it, any type of data analysis probably will not make much sense. However,
the degree of detail of taxonomic information differs depending on the dataset
and annotation data used.</p>
<p>Therefore, the mia package expects a loose assembly of taxonomic information
and assumes certain key aspects:</p>
<ul>
<li>Taxonomic information is given as character vectors or factors in the
<code>rowData</code> of a <code>SummarizedExperiment</code> object.</li>
<li>The columns containing the taxonomic information must be named <code>domain</code>,
<code>kingdom</code>, <code>phylum</code>, <code>class</code>, <code>order</code>, <code>family</code>, <code>genus</code>, <code>species</code> or with
a capital first letter.</li>
<li>the columns must be given in the order shown above</li>
<li>column can be omited, but the order must remain</li>
</ul>
<p>In this chapter, we will refer to co-abundant groups as CAGs, which are
clusters of taxa that co-vary across samples.</p>
<div id="assigning-taxonomic-information." class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Assigning taxonomic information.<a href="taxonomic-information.html#assigning-taxonomic-information." class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There are a number of methods to assign taxonomic information. We like to give
a short introduction about the methods available without ranking one over the
other. This has to be your choice based on the result for the individual
dataset.</p>
<div id="dada2" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> dada2<a href="taxonomic-information.html#dada2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The dada2 package <span class="citation">(Benjamin J. Callahan et al. 2016)</span> implements the <code>assignTaxonomy</code>
function, which takes as input the ASV sequences associated with each
row of data and a training dataset. For more information visit the
<a href="https://benjjneb.github.io/dada2/assign.html">dada2 homepage</a>.</p>
</div>
<div id="decipher" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> DECIPHER<a href="taxonomic-information.html#decipher" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The DECIPHER package <span class="citation">(Wright 2020)</span> implements the <code>IDTAXA</code> algorithm to assign
either taxonomic information or function information. For <code>mia</code>
only the first option is of interest for now and more information can be
found on the <a href="http://www2.decipher.codes/Classification.html">DECIPHER website</a>.</p>
</div>
</div>
<div id="functions-to-access-taxonomic-information" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Functions to access taxonomic information<a href="taxonomic-information.html#functions-to-access-taxonomic-information" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>checkTaxonomy</code> checks whether the taxonomic information is usable for <code>mia</code></p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="taxonomic-information.html#cb158-1" tabindex="-1"></a><span class="fu">checkTaxonomy</span>(tse)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Since the <code>rowData</code> can contain other data, <code>taxonomyRanks</code> will return the
columns <code>mia</code> assumes to contain the taxonomic information.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="taxonomic-information.html#cb160-1" tabindex="-1"></a><span class="fu">taxonomyRanks</span>(tse)</span></code></pre></div>
<pre><code>## [1] &quot;Kingdom&quot; &quot;Phylum&quot;  &quot;Class&quot;   &quot;Order&quot;   &quot;Family&quot;  &quot;Genus&quot;   &quot;Species&quot;</code></pre>
<p>This can then be used to subset the <code>rowData</code> to columns needed.</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="taxonomic-information.html#cb162-1" tabindex="-1"></a><span class="fu">rowData</span>(tse)[, <span class="fu">taxonomyRanks</span>(tse)]</span></code></pre></div>
<pre><code>## DataFrame with 19216 rows and 7 columns
##            Kingdom        Phylum        Class        Order        Family
##        &lt;character&gt;   &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;
## 549322     Archaea Crenarchaeota Thermoprotei           NA            NA
## 522457     Archaea Crenarchaeota Thermoprotei           NA            NA
## 951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 244423     Archaea Crenarchaeota        Sd-NA           NA            NA
## 586076     Archaea Crenarchaeota        Sd-NA           NA            NA
## ...            ...           ...          ...          ...           ...
## 278222    Bacteria           SR1           NA           NA            NA
## 463590    Bacteria           SR1           NA           NA            NA
## 535321    Bacteria           SR1           NA           NA            NA
## 200359    Bacteria           SR1           NA           NA            NA
## 271582    Bacteria           SR1           NA           NA            NA
##              Genus                Species
##        &lt;character&gt;            &lt;character&gt;
## 549322          NA                     NA
## 522457          NA                     NA
## 951     Sulfolobus Sulfolobusacidocalda..
## 244423          NA                     NA
## 586076          NA                     NA
## ...            ...                    ...
## 278222          NA                     NA
## 463590          NA                     NA
## 535321          NA                     NA
## 200359          NA                     NA
## 271582          NA                     NA</code></pre>
<p><code>taxonomyRankEmpty</code> checks for empty values in the given <code>rank</code> and returns a
logical vector of <code>length(x)</code>.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="taxonomic-information.html#cb164-1" tabindex="-1"></a><span class="fu">all</span>(<span class="sc">!</span><span class="fu">taxonomyRankEmpty</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Kingdom&quot;</span>))</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="taxonomic-information.html#cb166-1" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">taxonomyRankEmpty</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>))</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##  8008 11208</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="taxonomic-information.html#cb168-1" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">taxonomyRankEmpty</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Species&quot;</span>))</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##  1413 17803</code></pre>
<p><code>getTaxonomyLabels</code> is a multi-purpose function, which turns taxonomic
information into a character vector of <code>length(x)</code></p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="taxonomic-information.html#cb170-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse))</span></code></pre></div>
<pre><code>## [1] &quot;Class:Thermoprotei&quot;               &quot;Class:Thermoprotei_1&quot;            
## [3] &quot;Species:Sulfolobusacidocaldarius&quot; &quot;Class:Sd-NA&quot;                     
## [5] &quot;Class:Sd-NA_1&quot;                    &quot;Class:Sd-NA_2&quot;</code></pre>
<p>By default, this will use the lowest non-empty information to construct a
string with the following scheme <code>level:value</code>. If all levels are the same,
this part is omitted, but can be added by setting <code>with_rank = TRUE</code>.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="taxonomic-information.html#cb172-1" tabindex="-1"></a>phylum <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum) <span class="sc">&amp;</span></span>
<span id="cb172-2"><a href="taxonomic-information.html#cb172-2" tabindex="-1"></a>    <span class="fu">vapply</span>(<span class="fu">data.frame</span>(<span class="fu">apply</span>(<span class="fu">rowData</span>(tse)[, <span class="fu">taxonomyRanks</span>(tse)[<span class="dv">3</span><span class="sc">:</span><span class="dv">7</span>]], 1L, is.na)), all, <span class="fu">logical</span>(<span class="dv">1</span>))</span>
<span id="cb172-3"><a href="taxonomic-information.html#cb172-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse[phylum,]))</span></code></pre></div>
<pre><code>## [1] &quot;Crenarchaeota&quot;    &quot;Crenarchaeota_1&quot;  &quot;Crenarchaeota_2&quot;  &quot;Actinobacteria&quot;  
## [5] &quot;Actinobacteria_1&quot; &quot;Spirochaetes&quot;</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="taxonomic-information.html#cb174-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse[phylum,], <span class="at">with_rank =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Phylum:Crenarchaeota&quot;    &quot;Phylum:Crenarchaeota_1&quot; 
## [3] &quot;Phylum:Crenarchaeota_2&quot;  &quot;Phylum:Actinobacteria&quot;  
## [5] &quot;Phylum:Actinobacteria_1&quot; &quot;Phylum:Spirochaetes&quot;</code></pre>
<p>By default the return value of <code>getTaxonomyLabels</code> contains only
unique elements by passing it through <code>make.unique</code>. This step can be
omitted by setting <code>make_unique = FALSE</code>.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="taxonomic-information.html#cb176-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse[phylum,], <span class="at">with_rank =</span> <span class="cn">TRUE</span>, <span class="at">make_unique =</span> <span class="cn">FALSE</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Phylum:Crenarchaeota&quot;  &quot;Phylum:Crenarchaeota&quot;  &quot;Phylum:Crenarchaeota&quot; 
## [4] &quot;Phylum:Actinobacteria&quot; &quot;Phylum:Actinobacteria&quot; &quot;Phylum:Spirochaetes&quot;</code></pre>
<p>To apply the loop resolving function <code>resolveLoop</code> from the
<code>TreeSummarizedExperiment</code> package <span class="citation">(Huang 2020)</span> within
<code>getTaxonomyLabels</code>, set <code>resolve_loops = TRUE</code>.</p>
<p>The function <code>getUniqueFeatures</code> gives a list of unique taxa for the
specified taxonomic rank.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="taxonomic-information.html#cb178-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getUniqueFeatures</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Crenarchaeota&quot;  &quot;Euryarchaeota&quot;  &quot;Actinobacteria&quot; &quot;Spirochaetes&quot;  
## [5] &quot;MVP-15&quot;         &quot;Proteobacteria&quot;</code></pre>
<div id="fly-tree" class="section level3 hasAnchor" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Generate a taxonomic tree on the fly<a href="taxonomic-information.html#fly-tree" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To create a taxonomic tree, <code>taxonomyTree</code> used the information and returns a
<code>phylo</code> object. Duplicate information from the <code>rowData</code> is removed.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="taxonomic-information.html#cb180-1" tabindex="-1"></a><span class="fu">taxonomyTree</span>(tse)</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 1645 tips and 1089 internal nodes.
## 
## Tip labels:
##   Species:Cenarchaeumsymbiosum, Species:pIVWA5, Species:CandidatusNitrososphaeragargensis, Species:SCA1145, Species:SCA1170, Species:Sulfolobusacidocaldarius, ...
## Node labels:
##   root:ALL, Kingdom:Archaea, Phylum:Crenarchaeota, Class:C2, Class:Sd-NA, Class:Thaumarchaeota, ...
## 
## Rooted; includes branch lengths.</code></pre>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="taxonomic-information.html#cb182-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">addTaxonomyTree</span>(tse)</span>
<span id="cb182-2"><a href="taxonomic-information.html#cb182-2" tabindex="-1"></a>tse</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(1): counts
## rownames(19216): Class:Thermoprotei Class:Thermoprotei ... Phylum:SR1
##   Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: 1 phylo tree(s) (1645 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>The implementation is based on the <code>toTree</code> function from the
<code>TreeSummarizedExperiment</code> package <span class="citation">(Huang 2020)</span>.</p>
</div>
</div>
<div id="data-agglomeration" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> Data agglomeration<a href="taxonomic-information.html#data-agglomeration" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the main applications of taxonomic information in regards to count data
is to agglomerate count data on taxonomic levels and track the influence of
changing conditions through these levels. For this <code>mia</code> contains the
<code>mergeFeaturesByRank</code> function. The ideal location to store the agglomerated data
is as an alternative experiment.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="taxonomic-information.html#cb184-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb184-2"><a href="taxonomic-information.html#cb184-2" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;Family&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb184-3"><a href="taxonomic-information.html#cb184-3" tabindex="-1"></a>                                           <span class="at">agglomerateTree =</span> <span class="cn">TRUE</span>)</span>
<span id="cb184-4"><a href="taxonomic-information.html#cb184-4" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;Family&quot;</span>)</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 603 26 
## metadata(1): agglomerated_by_rank
## assays(2): counts relabundance
## rownames(603): Class:Thermoprotei Family:Sulfolobaceae ...
##   Family:Thermodesulfobiaceae Phylum:SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: a LinkDataFrame (603 rows)
## rowTree: 1 phylo tree(s) (496 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>If multiple assays (counts and relabundance) exist, both will be agglomerated.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="taxonomic-information.html#cb186-1" tabindex="-1"></a><span class="fu">assayNames</span>(tse)</span></code></pre></div>
<pre><code>## [1] &quot;counts&quot;       &quot;relabundance&quot;</code></pre>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="taxonomic-information.html#cb188-1" tabindex="-1"></a><span class="fu">assayNames</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;Family&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;counts&quot;       &quot;relabundance&quot;</code></pre>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="taxonomic-information.html#cb190-1" tabindex="-1"></a><span class="fu">assay</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;Family&quot;</span>), <span class="st">&quot;relabundance&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                            CL3       CC1 SV1 M31Fcsw M11Fcsw M31Plmr   M11Plmr
## Class:Thermoprotei   0.0000000 0.000e+00   0       0       0       0 0.000e+00
## Family:Sulfolobaceae 0.0000000 0.000e+00   0       0       0       0 2.305e-06
## Class:Sd-NA          0.0000000 0.000e+00   0       0       0       0 0.000e+00
## Order:NRP-J          0.0001991 2.070e-04   0       0       0       0 6.914e-06
## Family:SAGMA-X       0.0000000 6.165e-06   0       0       0       0 0.000e+00</code></pre>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="taxonomic-information.html#cb192-1" tabindex="-1"></a><span class="fu">assay</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;Family&quot;</span>), <span class="st">&quot;counts&quot;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                      CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## Class:Thermoprotei     0   0   0       0       0       0       0
## Family:Sulfolobaceae   0   0   0       0       0       0       1
## Class:Sd-NA            0   0   0       0       0       0       0
## Order:NRP-J          172 235   0       0       0       0       3
## Family:SAGMA-X         0   7   0       0       0       0       0</code></pre>
<p><code>altExpNames</code> now consists of <code>Family</code> level data. This can be extended to use
any taxonomic level listed in <code>mia::taxonomyRanks(tse)</code>.</p>
<p>Rare taxa can also be aggregated into a single group “Other” instead of
filtering them out. A suitable function for this is <code>mergeFeaturesByPrevalence</code>.
The number of rare taxa is higher on the species level, which causes the need
for data agglomeration by prevalence.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="taxonomic-information.html#cb194-1" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;Species_byPrevalence&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByPrevalence</span>(tse, </span>
<span id="cb194-2"><a href="taxonomic-information.html#cb194-2" tabindex="-1"></a>                                                               <span class="at">rank =</span> <span class="st">&quot;Species&quot;</span>, </span>
<span id="cb194-3"><a href="taxonomic-information.html#cb194-3" tabindex="-1"></a>                                                               <span class="at">other_label =</span> <span class="st">&quot;Other&quot;</span>, </span>
<span id="cb194-4"><a href="taxonomic-information.html#cb194-4" tabindex="-1"></a>                                                               <span class="at">prevalence =</span> <span class="dv">5</span> <span class="sc">/</span> <span class="dv">100</span>, </span>
<span id="cb194-5"><a href="taxonomic-information.html#cb194-5" tabindex="-1"></a>                                                               <span class="at">detection =</span> <span class="dv">1</span> <span class="sc">/</span> <span class="dv">100</span>, </span>
<span id="cb194-6"><a href="taxonomic-information.html#cb194-6" tabindex="-1"></a>                                                               <span class="at">as_relative =</span> T)</span>
<span id="cb194-7"><a href="taxonomic-information.html#cb194-7" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">&quot;Species_byPrevalence&quot;</span>)</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 113 26 
## metadata(2): agglomerated_by_rank agglomerated_by_rank
## assays(2): counts relabundance
## rownames(113): Family:MarinegroupII Order:Actinomycetales ...
##   Species:Veillonellaparvula Other
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="taxonomic-information.html#cb196-1" tabindex="-1"></a><span class="fu">assay</span>(<span class="fu">altExp</span>(tse, <span class="st">&quot;Species_byPrevalence&quot;</span>), <span class="st">&quot;relabundance&quot;</span>)[<span class="dv">88</span><span class="sc">:</span><span class="dv">92</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                           CL3       CC1       SV1   M31Fcsw   M11Fcsw   M31Plmr
## Genus:Luteolibacter 9.837e-05 8.076e-04 0.0001233 3.239e-06 5.297e-06 4.034e-05
## Genus:MC18          5.664e-02 2.110e-01 0.0289530 4.471e-05 2.649e-05 1.071e-04
## Class:Phycisphaerae 6.958e-03 1.816e-02 0.0585985 5.831e-06 4.816e-06 4.729e-05
## Order:envOPS12      2.315e-05 7.046e-06 0.0000000 1.296e-06 4.816e-07 1.391e-06
## Order:Clostridiales 1.017e-03 1.629e-04 0.0003857 1.473e-02 2.005e-02 1.647e-03
##                       M11Plmr
## Genus:Luteolibacter 1.291e-04
## Genus:MC18          7.882e-04
## Class:Phycisphaerae 4.033e-04
## Order:envOPS12      4.609e-06
## Order:Clostridiales 1.719e-03</code></pre>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="taxonomic-information.html#cb198-1" tabindex="-1"></a><span class="co"># Saving the tse for later</span></span>
<span id="cb198-2"><a href="taxonomic-information.html#cb198-2" tabindex="-1"></a>tseGlobalPatterns <span class="ot">&lt;-</span> tse</span></code></pre></div>
<div id="taxa-clustering" class="section level3 hasAnchor" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Taxa clustering<a href="taxonomic-information.html#taxa-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Another way to agglomerate the data is to cluster the taxa. To do so,
we usually start by doing a compositionality aware transformation such as CLR,
followed by the application of a standard clustering method.</p>
<p>Here is an example that does a CLR transformation followed by the hierarchical
clustering algorithm.</p>
<p>First, we import the library <code>bluster</code> that simplifies the clustering.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="taxonomic-information.html#cb199-1" tabindex="-1"></a><span class="fu">library</span>(bluster)</span></code></pre></div>
<p>Then we do the CLR transform followed by the clustering. We will cluster with
two different distances: the euclidean distance and the kendall distance.</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="taxonomic-information.html#cb200-1" tabindex="-1"></a><span class="co"># Get the data</span></span>
<span id="cb200-2"><a href="taxonomic-information.html#cb200-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;peerj13075&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mia&quot;</span>)</span>
<span id="cb200-3"><a href="taxonomic-information.html#cb200-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> peerj13075</span>
<span id="cb200-4"><a href="taxonomic-information.html#cb200-4" tabindex="-1"></a></span>
<span id="cb200-5"><a href="taxonomic-information.html#cb200-5" tabindex="-1"></a><span class="co"># The result of the CLR transform is stored in the assay clr</span></span>
<span id="cb200-6"><a href="taxonomic-information.html#cb200-6" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb200-7"><a href="taxonomic-information.html#cb200-7" tabindex="-1"></a></span>
<span id="cb200-8"><a href="taxonomic-information.html#cb200-8" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, </span>
<span id="cb200-9"><a href="taxonomic-information.html#cb200-9" tabindex="-1"></a>                      <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>)</span>
<span id="cb200-10"><a href="taxonomic-information.html#cb200-10" tabindex="-1"></a></span>
<span id="cb200-11"><a href="taxonomic-information.html#cb200-11" tabindex="-1"></a><span class="co"># Cluster (with euclidean distance) on the features of the z assay</span></span>
<span id="cb200-12"><a href="taxonomic-information.html#cb200-12" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse,</span>
<span id="cb200-13"><a href="taxonomic-information.html#cb200-13" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;z&quot;</span>,</span>
<span id="cb200-14"><a href="taxonomic-information.html#cb200-14" tabindex="-1"></a>               <span class="at">clust.col =</span> <span class="st">&quot;hclustEuclidean&quot;</span>,</span>
<span id="cb200-15"><a href="taxonomic-information.html#cb200-15" tabindex="-1"></a>           <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>,</span>
<span id="cb200-16"><a href="taxonomic-information.html#cb200-16" tabindex="-1"></a>               <span class="fu">HclustParam</span>(<span class="at">dist.fun =</span> stats<span class="sc">::</span>dist, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>))</span>
<span id="cb200-17"><a href="taxonomic-information.html#cb200-17" tabindex="-1"></a></span>
<span id="cb200-18"><a href="taxonomic-information.html#cb200-18" tabindex="-1"></a><span class="co"># Declare the Kendall dissimilarity computation function</span></span>
<span id="cb200-19"><a href="taxonomic-information.html#cb200-19" tabindex="-1"></a>kendall_dissimilarity <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb200-20"><a href="taxonomic-information.html#cb200-20" tabindex="-1"></a>    <span class="fu">as.dist</span>(<span class="dv">1</span> <span class="sc">-</span> <span class="fu">cor</span>(<span class="fu">t</span>(x), <span class="at">method =</span> <span class="st">&quot;kendall&quot;</span>))</span>
<span id="cb200-21"><a href="taxonomic-information.html#cb200-21" tabindex="-1"></a>}</span>
<span id="cb200-22"><a href="taxonomic-information.html#cb200-22" tabindex="-1"></a></span>
<span id="cb200-23"><a href="taxonomic-information.html#cb200-23" tabindex="-1"></a><span class="co"># Cluster (with Kendall dissimilarity) on the features of the z assay</span></span>
<span id="cb200-24"><a href="taxonomic-information.html#cb200-24" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse,</span>
<span id="cb200-25"><a href="taxonomic-information.html#cb200-25" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">&quot;z&quot;</span>,</span>
<span id="cb200-26"><a href="taxonomic-information.html#cb200-26" tabindex="-1"></a>               <span class="at">clust.col =</span> <span class="st">&quot;hclustKendall&quot;</span>,</span>
<span id="cb200-27"><a href="taxonomic-information.html#cb200-27" tabindex="-1"></a>               <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>,            </span>
<span id="cb200-28"><a href="taxonomic-information.html#cb200-28" tabindex="-1"></a>               <span class="fu">HclustParam</span>(<span class="at">dist.fun =</span> kendall_dissimilarity, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>))</span></code></pre></div>
<p>Let us store the resulting cluster indices in the <code>rowData</code> column specified
with the <code>clust.col</code> parameter.</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="taxonomic-information.html#cb201-1" tabindex="-1"></a><span class="co"># Checking the clusters</span></span>
<span id="cb201-2"><a href="taxonomic-information.html#cb201-2" tabindex="-1"></a>clusters_euclidean <span class="ot">&lt;-</span> <span class="fu">rowData</span>(tse)<span class="sc">$</span>hclustEuclidean</span>
<span id="cb201-3"><a href="taxonomic-information.html#cb201-3" tabindex="-1"></a><span class="fu">head</span>(clusters_euclidean, <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##  OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 
##     1     2     1     1     1     1     3     4     3     2 
## Levels: 1 2 3 4 5</code></pre>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="taxonomic-information.html#cb203-1" tabindex="-1"></a>clusters_kendall <span class="ot">&lt;-</span> <span class="fu">rowData</span>(tse)<span class="sc">$</span>hclustKendall</span>
<span id="cb203-2"><a href="taxonomic-information.html#cb203-2" tabindex="-1"></a><span class="fu">head</span>(clusters_kendall, <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##  OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 
##     1     2     1     3     3     1     3     1     1     3 
## Levels: 1 2 3 4</code></pre>
<p>To better visualize the results and the distribution of the clusters, we can
plot the histogram of the clusters.</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="taxonomic-information.html#cb205-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb205-2"><a href="taxonomic-information.html#cb205-2" tabindex="-1"></a><span class="fu">library</span>(patchwork) <span class="co"># TO arrange several plots as a grid</span></span>
<span id="cb205-3"><a href="taxonomic-information.html#cb205-3" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">rowData</span>(tse)), <span class="fu">aes</span>(<span class="at">x =</span> clusters_euclidean)) <span class="sc">+</span></span>
<span id="cb205-4"><a href="taxonomic-information.html#cb205-4" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb205-5"><a href="taxonomic-information.html#cb205-5" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;CAG size distribution (Euclidean distance)&quot;</span>,</span>
<span id="cb205-6"><a href="taxonomic-information.html#cb205-6" tabindex="-1"></a>         <span class="at">x =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Feature count (n)&quot;</span>)</span>
<span id="cb205-7"><a href="taxonomic-information.html#cb205-7" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">rowData</span>(tse)), <span class="fu">aes</span>(<span class="at">x =</span> clusters_kendall)) <span class="sc">+</span></span>
<span id="cb205-8"><a href="taxonomic-information.html#cb205-8" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb205-9"><a href="taxonomic-information.html#cb205-9" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;CAG size distribution (1 - tau)&quot;</span>,</span>
<span id="cb205-10"><a href="taxonomic-information.html#cb205-10" tabindex="-1"></a>         <span class="at">x =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Feature count (n)&quot;</span>)</span>
<span id="cb205-11"><a href="taxonomic-information.html#cb205-11" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="11_taxonomic_information_files/figure-html/taxa_clustering_histogram-1.png" width="672" /></p>
<p>It’s also possible to merge the rows by cluster.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="taxonomic-information.html#cb206-1" tabindex="-1"></a><span class="co"># Aggregate clusters as a sum of each cluster values</span></span>
<span id="cb206-2"><a href="taxonomic-information.html#cb206-2" tabindex="-1"></a>tse_merged <span class="ot">&lt;-</span> <span class="fu">mergeFeatures</span>(tse, clusters_euclidean)</span>
<span id="cb206-3"><a href="taxonomic-information.html#cb206-3" tabindex="-1"></a>tse_merged</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 5 58 
## metadata(0):
## assays(3): counts clr z
## rownames(5): 1 2 3 4 5
## rowData names(8): kingdom phylum ... hclustEuclidean hclustKendall
## colnames(58): ID1 ID2 ... ID57 ID58
## colData names(5): Sample Geographical_location Gender Age Diet
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## rowLinks: NULL
## rowTree: NULL
## colLinks: NULL
## colTree: NULL</code></pre>
<p>We can note that it worked as planned since there were 5 clusters and there are
now 5 rows.</p>
</div>
</div>
<div id="assay-transform" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> Data transformation<a href="taxonomic-information.html#assay-transform" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Data transformations are common in microbiome analysis. Examples
include the logarithmic transformation, calculation of relative
abundances (percentages), and compositionality-aware transformations
such as the centered log-ratio transformation (clr).</p>
<p>In mia package, transformations are applied to abundance data. The transformed
abundance table is stored back to ‘assays’. mia includes transformation
function (‘transformAssay()’) which applies sample-wise or column-wise
transformation when MARGIN = ‘samples’, feature-wise or row-wise transformation
when MARGIN = ‘features’.</p>
<p>For a complete list of available transformations and parameters, see function
<a href="https://microbiome.github.io/mia/reference/transformAssay.html">help</a>.</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="taxonomic-information.html#cb208-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> tseGlobalPatterns</span>
<span id="cb208-2"><a href="taxonomic-information.html#cb208-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb208-3"><a href="taxonomic-information.html#cb208-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(<span class="at">x =</span> tse, <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>, </span>
<span id="cb208-4"><a href="taxonomic-information.html#cb208-4" tabindex="-1"></a>                      <span class="at">pseudocount =</span> <span class="dv">1</span>, <span class="at">name =</span> <span class="st">&quot;clr&quot;</span>)</span>
<span id="cb208-5"><a href="taxonomic-information.html#cb208-5" tabindex="-1"></a></span>
<span id="cb208-6"><a href="taxonomic-information.html#cb208-6" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;clr&quot;</span>))</span></code></pre></div>
<pre><code>##                                         CL3        CC1        SV1    M31Fcsw
## Class:Thermoprotei               -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Thermoprotei               -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Species:Sulfolobusacidocaldarius -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
## Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
##                                     M11Fcsw    M31Plmr    M11Plmr    F21Plmr
## Class:Thermoprotei               -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Class:Thermoprotei               -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Species:Sulfolobusacidocaldarius -4.947e-05 -4.931e-05 -4.658e-05 -4.671e-05
## Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
## Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
##                                     M31Tong    M11Tong   LMEpi24M   SLEpi20M
## Class:Thermoprotei               -4.846e-05 -4.257e-05 -4.756e-05 -4.837e-05
## Class:Thermoprotei               -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Species:Sulfolobusacidocaldarius -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
## Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
##                                      AQC1cm     AQC4cm     AQC7cm        NP2
## Class:Thermoprotei               -2.385e-05 -4.438e-06  2.787e-05 -4.731e-05
## Class:Thermoprotei               -4.660e-05 -4.568e-05 -4.428e-05 -4.915e-05
## Species:Sulfolobusacidocaldarius -4.660e-05 -4.652e-05 -4.777e-05 -4.915e-05
## Class:Sd-NA                      -4.660e-05 -3.726e-05 -3.090e-05 -4.915e-05
## Class:Sd-NA                      -4.660e-05 -4.568e-05 -4.719e-05 -4.915e-05
## Class:Sd-NA                      -4.660e-05 -4.610e-05 -4.603e-05 -4.915e-05
##                                         NP3        NP5    TRRsed1    TRRsed2
## Class:Thermoprotei               -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Thermoprotei               -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Species:Sulfolobusacidocaldarius -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
## Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
##                                     TRRsed3       TS28       TS29      Even1
## Class:Thermoprotei               -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Thermoprotei               -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Species:Sulfolobusacidocaldarius -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
## Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
##                                       Even2      Even3
## Class:Thermoprotei               -5.017e-05 -5.034e-05
## Class:Thermoprotei               -5.017e-05 -5.034e-05
## Species:Sulfolobusacidocaldarius -5.017e-05 -5.034e-05
## Class:Sd-NA                      -5.017e-05 -5.034e-05
## Class:Sd-NA                      -5.017e-05 -5.034e-05
## Class:Sd-NA                      -5.017e-05 -5.034e-05</code></pre>
<ul>
<li>In ‘pa’ transformation, abundance table is converted to present/absent table.</li>
</ul>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="taxonomic-information.html#cb210-1" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">&quot;pa&quot;</span>)</span>
<span id="cb210-2"><a href="taxonomic-information.html#cb210-2" tabindex="-1"></a></span>
<span id="cb210-3"><a href="taxonomic-information.html#cb210-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(tse, <span class="st">&quot;pa&quot;</span>))</span></code></pre></div>
<pre><code>##                                  CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## Class:Thermoprotei                 0   0   0       0       0       0       0
## Class:Thermoprotei                 0   0   0       0       0       0       0
## Species:Sulfolobusacidocaldarius   0   0   0       0       0       0       1
## Class:Sd-NA                        0   0   0       0       0       0       0
## Class:Sd-NA                        0   0   0       0       0       0       0
## Class:Sd-NA                        0   0   0       0       0       0       0
##                                  F21Plmr M31Tong M11Tong LMEpi24M SLEpi20M
## Class:Thermoprotei                     0       0       0        0        1
## Class:Thermoprotei                     0       0       0        0        0
## Species:Sulfolobusacidocaldarius       0       0       0        0        0
## Class:Sd-NA                            0       0       0        0        0
## Class:Sd-NA                            0       0       0        0        0
## Class:Sd-NA                            0       0       0        0        0
##                                  AQC1cm AQC4cm AQC7cm NP2 NP3 NP5 TRRsed1
## Class:Thermoprotei                    1      1      1   1   0   0       0
## Class:Thermoprotei                    0      1      1   0   0   0       0
## Species:Sulfolobusacidocaldarius      0      0      0   0   0   0       0
## Class:Sd-NA                           0      1      1   0   0   0       0
## Class:Sd-NA                           0      1      1   0   0   0       0
## Class:Sd-NA                           0      1      1   0   0   0       0
##                                  TRRsed2 TRRsed3 TS28 TS29 Even1 Even2 Even3
## Class:Thermoprotei                     0       0    0    0     0     0     0
## Class:Thermoprotei                     0       0    0    0     0     0     0
## Species:Sulfolobusacidocaldarius       0       0    0    0     0     0     0
## Class:Sd-NA                            0       0    0    0     0     0     0
## Class:Sd-NA                            0       0    0    0     0     0     0
## Class:Sd-NA                            0       0    0    0     0     0     0</code></pre>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="taxonomic-information.html#cb212-1" tabindex="-1"></a><span class="co"># list of abundance tables that assays slot contains</span></span>
<span id="cb212-2"><a href="taxonomic-information.html#cb212-2" tabindex="-1"></a><span class="fu">assays</span>(tse)</span></code></pre></div>
<pre><code>## List of length 4
## names(4): counts relabundance clr pa</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="quality-control.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="community-diversity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/11_taxonomic_information.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/11_taxonomic_information.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/11_taxonomic_information.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
