# Taxonomic Information {#sec-taxonomic-information}

```{r setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble()
```

```{r, message=FALSE}
library(mia)
data("GlobalPatterns", package = "mia")
tse <- GlobalPatterns
```

Taxonomic information is a key part of analyzing microbiome data and without
it, any type of data analysis probably will not make much sense. However,
the degree of detail of taxonomic information differs depending on the dataset
and annotation data used.

Therefore, the mia package expects a loose assembly of taxonomic information
and assumes certain key aspects:

* Taxonomic information is given as character vectors or factors in the 
`rowData` of a `SummarizedExperiment` object.
* The columns containing the taxonomic information must be named `domain`,
`kingdom`, `phylum`, `class`, `order`, `family`, `genus`, `species` or with
a capital first letter.
* the columns must be given in the order shown above
* column can be omited, but the order must remain

In this chapter, we will refer to co-abundant groups as CAGs, which are
clusters of taxa that co-vary across samples.

## Assigning taxonomic information.

There are a number of methods to assign taxonomic information. We like to give
a short introduction about the methods available without ranking one over the 
other. This has to be your choice based on the result for the individual 
dataset.

### dada2

The dada2 package [@Callahan2016dada2] implements the `assignTaxonomy`
function, which takes as input the ASV sequences associated with each
row of data and a training dataset. For more information visit the
[dada2 homepage](https://benjjneb.github.io/dada2/assign.html).

### DECIPHER

The DECIPHER package [@R_DECIPHER] implements the `IDTAXA` algorithm to assign
either taxonomic information or function information. For `mia`
only the first option is of interest for now and more information can be
found on the [DECIPHER website](http://www2.decipher.codes/Classification.html).

## Functions to access taxonomic information

`checkTaxonomy` checks whether the taxonomic information is usable for `mia`

```{r}
checkTaxonomy(tse)
```

Since the `rowData` can contain other data, `taxonomyRanks` will return the 
columns `mia` assumes to contain the taxonomic information.

```{r}
taxonomyRanks(tse)
```

This can then be used to subset the `rowData` to columns needed.

```{r}
rowData(tse)[, taxonomyRanks(tse)]
```

`taxonomyRankEmpty` checks for empty values in the given `rank` and returns a 
logical vector of `length(x)`.

```{r}
all(!taxonomyRankEmpty(tse, rank = "Kingdom"))
table(taxonomyRankEmpty(tse, rank = "Genus"))
table(taxonomyRankEmpty(tse, rank = "Species"))
```

`getTaxonomyLabels` is a multi-purpose function, which turns taxonomic
information into a character vector of `length(x)`

```{r}
head(getTaxonomyLabels(tse))
```

By default, this will use the lowest non-empty information to construct a
string with the following scheme `level:value`. If all levels are the same,
this part is omitted, but can be added by setting `with_rank = TRUE`.

```{r}
phylum <- !is.na(rowData(tse)$Phylum) &
    vapply(data.frame(apply(rowData(tse)[, taxonomyRanks(tse)[3:7]], 1L, is.na)), all, logical(1))
head(getTaxonomyLabels(tse[phylum,]))
head(getTaxonomyLabels(tse[phylum,], with_rank = TRUE))
```

By default the return value of `getTaxonomyLabels` contains only
unique elements by passing it through `make.unique`. This step can be
omitted by setting `make_unique = FALSE`.

```{r}
head(getTaxonomyLabels(tse[phylum,], with_rank = TRUE, make_unique = FALSE))
```

To apply the loop resolving function `resolveLoop` from the
`TreeSummarizedExperiment` package [@R_TreeSummarizedExperiment] within
`getTaxonomyLabels`, set `resolve_loops = TRUE`.

The function `getUnique` gives a list of unique taxa for the
specified taxonomic rank.

```{r}
head(getUnique(tse, rank = "Phylum"))
```

### Generate a hierarchy tree on the fly {#sec-fly-tree}

A hierarchy tree is not an actual phylogenetic tree.
A hierarchy tree shows mapping between the taxonomic levels in taxonomic rank table (included in rowData), rather than the detailed phylogenetic relations included in a phylogenetic tree

For instance, the TreeSummarizedExperiment dataset Tengeler2020 contains an actual phylogenetic tree.

```{r}
data(Tengeler2020)
tse2 <- Tengeler2020
phylo_tree <- rowTree(tse2)
phylo_tree
```

To create a hierarchy tree, `getHierarchyTree` used the information and returns a
`phylo` object. Duplicate information from the `rowData` is removed.

```{r}
getHierarchyTree(tse)
```

```{r}
tse <- addHierarchyTree(tse)
tse
```

The implementation is based on the `toTree` function from the
`TreeSummarizedExperiment` package [@R_TreeSummarizedExperiment].

## Data agglomeration {#sec-data-agglomeration}

One of the main applications of taxonomic information in regards to count data
is to agglomerate count data on taxonomic levels and track the influence of 
changing conditions through these levels. For this `mia` contains the
`mergeFeaturesByRank` function. The ideal location to store the agglomerated data
is as an alternative experiment.

```{r}
tse <- transformAssay(tse, assay.type = "counts", method = "relabundance")
altExp(tse, "Family") <- mergeFeaturesByRank(tse, rank = "Family",
                                           agglomerateTree = TRUE)
altExp(tse, "Family")
```

If multiple assays (counts and relabundance) exist, both will be agglomerated.

```{r}
assayNames(tse)
assayNames(altExp(tse, "Family"))
```

```{r}
assay(altExp(tse, "Family"), "relabundance")[1:5, 1:7]
```
  
```{r taxinfo_altexp_example}
assay(altExp(tse, "Family"), "counts")[1:5, 1:7]
```

`altExpNames` now consists of `Family` level data. This can be extended to use 
any taxonomic level listed in `mia::taxonomyRanks(tse)`.

Rare taxa can also be aggregated into a single group "Other" instead of 
filtering them out. A suitable function for this is `mergeFeaturesByPrevalence`.
The number of rare taxa is higher on the species level, which causes the need 
for data agglomeration by prevalence.

```{r}
altExp(tse, "Species_byPrevalence") <- mergeFeaturesByPrevalence(tse, 
                                                               rank = "Species", 
                                                               other_label = "Other", 
                                                               prevalence = 5 / 100, 
                                                               detection = 1 / 100, 
                                                               as_relative = T)
altExp(tse, "Species_byPrevalence")

assay(altExp(tse, "Species_byPrevalence"), "relabundance")[1:5, 1:3]
```

```{r}
# Saving the tse for later
tseGlobalPatterns <- tse
```

### Taxa clustering {#sec-taxa-clustering}
Another way to agglomerate the data is to cluster the taxa. To do so,
we usually start by doing a compositionality aware transformation such as CLR, 
followed by the application of a standard clustering method.

Here is an example that does a CLR transformation followed by the hierarchical
clustering algorithm. 

First, we import the library `bluster` that simplifies the clustering.

```{r bluster_dependence}
library(bluster)
```

Then we do the CLR transform followed by the clustering. We will cluster with
two different distances: the euclidean distance and the kendall distance.

```{r taxa_clustering}
# Get the data
data("peerj13075", package = "mia")
tse <- peerj13075

# The result of the CLR transform is stored in the assay clr
tse <- transformAssay(tse, method = "clr", pseudocount = 1)

tse <- transformAssay(tse, assay.type = "clr", method = "z", 
                      MARGIN = "features")

# Cluster (with euclidean distance) on the features of the z assay
tse <- addCluster(tse,
               assay.type = "z",
               clust.col = "hclustEuclidean",
	       MARGIN = "features",
               HclustParam(dist.fun = stats::dist, method = "ward.D2"))

# Declare the Kendall dissimilarity computation function
kendall_dissimilarity <- function(x) {
    as.dist(1 - cor(t(x), method = "kendall"))
}

# Cluster (with Kendall dissimilarity) on the features of the z assay
tse <- addCluster(tse,
               assay.type = "z",
               clust.col = "hclustKendall",
       	       MARGIN = "features", 	       
               HclustParam(dist.fun = kendall_dissimilarity, method = "ward.D2"))
```

Let us store the resulting cluster indices in the `rowData` column specified 
with the `clust.col` parameter.

```{r taxa_clustering_result}
# Checking the clusters
clusters_euclidean <- rowData(tse)$hclustEuclidean
head(clusters_euclidean, 10)

clusters_kendall <- rowData(tse)$hclustKendall
head(clusters_kendall, 10)
```

To better visualize the results and the distribution of the clusters, we can 
plot the histogram of the clusters.

```{r taxa_clustering_histogram}
library(ggplot2)
library(patchwork) # TO arrange several plots as a grid
plot1 <- ggplot(rowData(tse), aes(x = clusters_euclidean)) +
    geom_bar() +
    labs(title = "CAG size distribution (Euclidean distance)",
         x = "Clusters", y = "Feature count (n)")
plot2 <- ggplot(rowData(tse), aes(x = clusters_kendall)) +
    geom_bar() +
    labs(title = "CAG size distribution (1 - tau)",
         x = "Clusters", y = "Feature count (n)")
plot1 + plot2 + plot_layout(ncol = 2)
```

It's also possible to merge the rows by cluster.

```{r taxa_clustering_row_merge}
# Aggregate clusters as a sum of each cluster values
tse_merged <- mergeFeatures(tse, clusters_euclidean)
tse_merged
```
We can note that it worked as planned since there were 5 clusters and there are
now 5 rows.


## Data transformation {#sec-assay-transform}

Data transformations are common in (microbial) ecology [@Legendre2001]
and used to improve compatibility with assumptions related to specific
statistical methods, mitigate biases, enhance the comparability of
samples or features, or to obtain more interpretable values.

Examples include the logarithmic transformation, calculation of
relative abundances (percentages), and compositionality-aware
transformations such as the centered log-ratio transformation (clr).

Let us summarize some commonly used transformations in microbiome data
science; further details and benchmarkings available in the
references.

 * 'relabundance' relative transformation; also known as total sum
   scaling (TSS) and compositional transformation. This converts
   counts into percentages (at the scale [0, 1]) that sum up to
   1. Much of the currently available taxonomic abundance data from
   high-throughput assays (16S, metagenomic sequencing) is
   compositional by nature, even if the data is provided as counts
   [@Gloor2017].
 
 * 'clr' Centered log ratio transformation [@Aitchison1986] is used to
   reduce data skewness and compositionality bias in relative
   abundances, while bringing the data to the logarithmic scale. This
   transformation is frequently applied in microbial ecology
   [@Gloor2017]. However, this transformation only applies to positive
   values. Usual solution is to add pseudocount, which adds another
   type of bias in the data. The robust clr transformation ('rclr')
   aims to circumvent the need to add a pseudocount. While the
   resulting values from these transformations are difficult interpret
   directly, this transformation may enhance comparability of relative
   differences between samples. It is part of a broader Aitchison
   family of transformations; the additive log ratio transformation
   (`alr') is also available. The robust clr ("rclr") is similar to
   regular clr (see above) but allows data with zeroes and avoids the
   need to add pseudocount [@Keshavan2010, @Martino2019].

  * 'pa' presence/absence transformation ignores abundances and only
   indicates whether the given feature is detected above the given
   threshold (default: 0). This simple transformation is relatively
   widely used in ecological research. It has shown good performance
   in microbiome-based classification performance [@Giliberti2022,
   Karwowska2024].

 * 'z' Z transformation scales data to zero mean and unit variance;
   this us used to bring features (or samples) to more comparable
   levels in terms of mean and scale of the values. This can enhance
   visualization and interpretation of the data

* 'log', 'log2', 'log10' Logarithmic transformations; used e.g. to
   reduce data skewness; with compositional data the `clr` (or `rclr`)
   transformation is often preferred.
   
 * 'hellinger' Hellinger transformation equals to the square root of
   relative abundances. This ecological transformation can be useful
   if we are interested in changes in relative abundances.

 * 'rank' Rank transformation replaces each value by its rank. Also
   see 'rrank' (relative rank transformation). This has use for
   instance in non-parametric statistics.

 * Other available transformations include Chi square ('chi.square'),
   Frequency transformation ('frequency'), and Make margin sum of
   squares equal to one ('normalize')


### Transforming the data in practice

Transformations on abundance assays can be performed with
`mia::transformAssay()`, keeping both the original and the transformed
assay(s). The transformed abundance assay is then stored back to the
'assays' slot in the data object. The function applies sample-wise or
column-wise transformation when MARGIN = 'samples', feature-wise or
row-wise transformation when MARGIN = 'features'.

A complete list of available transformations and parameters, is
available in the function
[help](https://microbiome.github.io/mia/reference/transformAssay.html).


```{r}
# Load example data
library(mia)
data("GlobalPatterns", package = "mia")
tse <- GlobalPatterns

# Transform "counts" assay to relative abundances ("relabundance"), with pseudocount 1 
tse <- transformAssay(tse, assay.type = "counts", method = "relabundance", pseudocount = 1)

# Transform relative abundance assay ("relabundance") to "clr", using pseudocount if necessary;
# name the resulting assay to "clr" 
tse <- transformAssay(x = tse, assay.type = "relabundance", method = "clr", 
                      pseudocount = TRUE, name = "clr")

```


Get the values in the resulting assay, and view some of the first
entries of it with the `head` command.

```{r}
head(assay(tse, "clr"))
```


In 'pa' transformation, abundance table is converted to
presence/absence table that ignores abundances and only indicates
whether the given feature is detected. This simple transformation is
relatively widely used in ecological research. It has shown good
performance in microbiome-based classification performance
[@Giliberti2022, Karwowska2024].


```{r}
# Here, `assay.type` is not explicitly specified.
# Then The function uses the "counts" assay for the transformation.
tse <- transformAssay(tse, method = "pa")
head(assay(tse, "pa"))
```

You can now view the entire list of abundance assays in your data object with:

```{r}
assays(tse)
```


