<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 2&nbsp; Microbiome Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/10_manipulation.html" rel="next">
<link href="../pages/06_packages.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../pages/images/mia_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Orchestrating Microbiome Analysis</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-bi-github">
<li>
    <a class="dropdown-item" href="https://github.com/microbiome/OMA/">
 <span class="dropdown-text">Source Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://microbiome.github.io/OMA/docs/devel/">
 <span class="dropdown-text">Browse version <code>devel</code></span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/06_packages.html">OMA ecosystem</a></li><li class="breadcrumb-item"><a href="../pages/04_containers.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#data-science-framework" id="toc-data-science-framework" class="nav-link active" data-scroll-target="#data-science-framework"><span class="header-section-number">2.1</span> Data science framework</a></li>
  <li>
<a href="#data-containers" id="toc-data-containers" class="nav-link" data-scroll-target="#data-containers"><span class="header-section-number">2.2</span> Data containers</a>
  <ul class="collapse">
<li><a href="#sec-assay-slot" id="toc-sec-assay-slot" class="nav-link" data-scroll-target="#sec-assay-slot"><span class="header-section-number">2.2.1</span> Assay data</a></li>
  <li><a href="#coldata" id="toc-coldata" class="nav-link" data-scroll-target="#coldata"><span class="header-section-number">2.2.2</span> colData</a></li>
  <li><a href="#rowdata" id="toc-rowdata" class="nav-link" data-scroll-target="#rowdata"><span class="header-section-number">2.2.3</span> rowData</a></li>
  <li><a href="#rowtree" id="toc-rowtree" class="nav-link" data-scroll-target="#rowtree"><span class="header-section-number">2.2.4</span> rowTree</a></li>
  <li><a href="#sec-alt-exp" id="toc-sec-alt-exp" class="nav-link" data-scroll-target="#sec-alt-exp"><span class="header-section-number">2.2.5</span> Alternative experiments</a></li>
  <li><a href="#sec-mae" id="toc-sec-mae" class="nav-link" data-scroll-target="#sec-mae"><span class="header-section-number">2.2.6</span> MAEs</a></li>
  </ul>
</li>
  <li>
<a href="#sec-example-data" id="toc-sec-example-data" class="nav-link" data-scroll-target="#sec-example-data"><span class="header-section-number">2.3</span> Demonstration data</a>
  <ul class="collapse">
<li><a href="#sec-package-data" id="toc-sec-package-data" class="nav-link" data-scroll-target="#sec-package-data"><span class="header-section-number">2.3.1</span> Package data</a></li>
  <li><a href="#experimenthub-data" id="toc-experimenthub-data" class="nav-link" data-scroll-target="#experimenthub-data"><span class="header-section-number">2.3.2</span> ExperimentHub data</a></li>
  <li><a href="#curated-metagenomic-data" id="toc-curated-metagenomic-data" class="nav-link" data-scroll-target="#curated-metagenomic-data"><span class="header-section-number">2.3.3</span> Curated metagenomic data</a></li>
  <li><a href="#human-microbiome-compendium" id="toc-human-microbiome-compendium" class="nav-link" data-scroll-target="#human-microbiome-compendium"><span class="header-section-number">2.3.4</span> Human microbiome compendium</a></li>
  <li><a href="#other-data-sources" id="toc-other-data-sources" class="nav-link" data-scroll-target="#other-data-sources"><span class="header-section-number">2.3.5</span> Other data sources</a></li>
  </ul>
</li>
  <li>
<a href="#sec-loading-experimental-microbiome-data" id="toc-sec-loading-experimental-microbiome-data" class="nav-link" data-scroll-target="#sec-loading-experimental-microbiome-data"><span class="header-section-number">2.4</span> Loading experimental microbiome data</a>
  <ul class="collapse">
<li><a href="#s-workflow" id="toc-s-workflow" class="nav-link" data-scroll-target="#s-workflow"><span class="header-section-number">2.4.1</span> 16S workflow</a></li>
  <li><a href="#sec-import-from-file" id="toc-sec-import-from-file" class="nav-link" data-scroll-target="#sec-import-from-file"><span class="header-section-number">2.4.2</span> Import from external files</a></li>
  <li><a href="#constructing-treese" id="toc-constructing-treese" class="nav-link" data-scroll-target="#constructing-treese"><span class="header-section-number">2.4.3</span> Constructing TreeSE</a></li>
  <li><a href="#constructing-mae" id="toc-constructing-mae" class="nav-link" data-scroll-target="#constructing-mae"><span class="header-section-number">2.4.4</span> Constructing MAE</a></li>
  <li><a href="#import-functions-for-standard-formats" id="toc-import-functions-for-standard-formats" class="nav-link" data-scroll-target="#import-functions-for-standard-formats"><span class="header-section-number">2.4.5</span> Import functions for standard formats</a></li>
  <li><a href="#conversions-between-data-formats-in-r" id="toc-conversions-between-data-formats-in-r" class="nav-link" data-scroll-target="#conversions-between-data-formats-in-r"><span class="header-section-number">2.4.6</span> Conversions between data formats in R</a></li>
  </ul>
</li>
  <li>
<a href="#sec-exporting-data-container" id="toc-sec-exporting-data-container" class="nav-link" data-scroll-target="#sec-exporting-data-container"><span class="header-section-number">2.5</span> Exporting data container</a>
  <ul class="collapse">
<li><a href="#export-treesummarizedexperiment" id="toc-export-treesummarizedexperiment" class="nav-link" data-scroll-target="#export-treesummarizedexperiment"><span class="header-section-number">2.5.1</span> Export TreeSummarizedExperiment</a></li>
  <li><a href="#export-multiassayexperiment" id="toc-export-multiassayexperiment" class="nav-link" data-scroll-target="#export-multiassayexperiment"><span class="header-section-number">2.5.2</span> Export MultiAssayExperiment</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/06_packages.html">OMA ecosystem</a></li><li class="breadcrumb-item"><a href="../pages/04_containers.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-containers" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<section id="data-science-framework" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="data-science-framework">
<span class="header-section-number">2.1</span> Data science framework</h2>
<p>The building blocks of the framework are <strong>data container</strong> (SummarizedExperiment and its derivatives), <strong>packages</strong> from various developers using the <code>TreeSE</code> container, open <strong>demonstration datasets</strong>, in a separate chapter <a href="#sec-example-data" class="quarto-xref"><span>Section 2.3</span></a>, and <strong>online tutorials</strong> including this online book as well as the various package vignettes and other materials.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/FigureOverviewV2_mod.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="data-containers" class="level2 page-columns page-full" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="data-containers">
<span class="header-section-number">2.2</span> Data containers</h2>
<p><code>SummarizedExperiment</code> (<code>SE</code>) <span class="citation" data-cites="R_SummarizedExperiment">(<a href="#ref-R_SummarizedExperiment" role="doc-biblioref">Morgan et al. 2020</a>)</span> is a generic and highly optimized container for complex data structures. It has become a common choice for analysing various types of biomedical profiling data, such as RNAseq, ChIp-Seq, microarrays, flow cytometry, proteomics, and single-cell sequencing.</p>
<div class="no-row-height column-margin column-container"><div id="ref-R_SummarizedExperiment" class="csl-entry" role="listitem">
Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020. <em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://bioconductor.org/packages/SummarizedExperiment">https://bioconductor.org/packages/SummarizedExperiment</a>.
</div><div id="ref-R_TreeSummarizedExperiment" class="csl-entry" role="listitem">
Huang, Ruizhu. 2020. <em>TreeSummarizedExperiment: A S4 Class for Data with Tree Structures</em>.
</div></div><p>[<code>TreeSummarizedExperiment</code>] (<code>TreeSE</code>) <span class="citation" data-cites="R_TreeSummarizedExperiment">(<a href="#ref-R_TreeSummarizedExperiment" role="doc-biblioref">Huang 2020</a>)</span> was developed as an extension to incorporate hierarchical information (such as phylogenetic trees and sample hierarchies) and reference sequences.</p>
<p>[<code>MultiAssayExperiment</code>] (<code>MAE</code>) <span class="citation" data-cites="Ramos2017">(<a href="#ref-Ramos2017" role="doc-biblioref">Ramos et al. 2017</a>)</span> provides an organized way to bind several different data containers together in a single object. For example, we can bind microbiome data (in <code>TreeSE</code> container) with metabolomic profiling data (in <code>SE</code>) container, with (partially) shared sample metadata. This is convenient and robust for instance in subsetting and other data manipulation tasks. Microbiome data can be part of multiomics experiments and analysis strategies. We highlight how the methods used througout in this book relate to this data framework by using the <code>TreeSE</code>, <code>MAE</code>, and classes beyond.</p>
<div class="no-row-height column-margin column-container"></div><p>This section provides an introduction to these data containers. In microbiome data science, these containers link taxonomic abundance tables with rich side information on the features and samples. Taxonomic abundance data can be obtained by 16S rRNA amplicon or metagenomic sequencing, phylogenetic microarrays, or by other means. Many microbiome experiments include multiple versions and types of data generated independently or derived from each other through transformation or agglomeration. We start by providing recommendations on how to represent different varieties of multi-table data within the <code>TreeSE</code> class.</p>
<p>The options and recommendations are summarized in <a href="#tbl-options" class="quarto-xref">Table&nbsp;<span>2.1</span></a>.</p>
<section id="sec-assay-slot" class="level3" data-number="2.2.1"><h3 data-number="2.2.1" class="anchored" data-anchor-id="sec-assay-slot">
<span class="header-section-number">2.2.1</span> Assay data</h3>
<p>The microbiome is the collection of all microbes (such as bacteria, viruses, fungi, etc.) in the body. When studying these microbes, data is needed, and that’s where assays come in.</p>
<p>An assay is a way of measuring the presence and abundance of different types of microbes in a sample. For example, if you want to know how many bacteria of a certain type are in your gut, you can use an assay to measure this. When storing assays, the original data is count based. However, the original count-based taxonomic abundance tables may have different transformations, such as logarithmic, Centered Log-Ratio (CLR), or relative abundance. These are typically stored in <em><strong>assays</strong></em>.</p>
<p>Let us load example data and rename it as tse.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>assays</code> slot contains the experimental data as multiple count matrices. The result of <code>assays</code> is a list of matrices.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  List of length 1</span></span>
<span><span class="co">##  names(1): counts</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Individual assays can be accessed via <code>assay</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="co">##         CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr</span></span>
<span><span class="co">##  549322   0   0   0       0       0       0       0</span></span>
<span><span class="co">##  522457   0   0   0       0       0       0       0</span></span>
<span><span class="co">##  951      0   0   0       0       0       0       1</span></span>
<span><span class="co">##  244423   0   0   0       0       0       0       0</span></span>
<span><span class="co">##  586076   0   0   0       0       0       0       0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So, in summary, in the world of microbiome analysis, an assay is essentially a way to quantify and understand the composition of microbes in a given sample, which is super important for all kinds of research, from human health to the environment.</p>
<p>Furthermore, to illustrate the use of multiple assays, the relative abundance data can be calculated and stored along the original count data using <code>transformAssay</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  List of length 2</span></span>
<span><span class="co">##  names(2): counts relabundance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now there are two assays available in the <code>tse</code> object, <code>counts</code> and <code>relabundance</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"relabundance"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="co">##         CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr   M11Plmr</span></span>
<span><span class="co">##  549322   0   0   0       0       0       0 0.000e+00</span></span>
<span><span class="co">##  522457   0   0   0       0       0       0 0.000e+00</span></span>
<span><span class="co">##  951      0   0   0       0       0       0 2.305e-06</span></span>
<span><span class="co">##  244423   0   0   0       0       0       0 0.000e+00</span></span>
<span><span class="co">##  586076   0   0   0       0       0       0 0.000e+00</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here the dimension of the count data remains unchanged in transformation. This is in fact, a requirement for the assays.</p>
</section><section id="coldata" class="level3" data-number="2.2.2"><h3 data-number="2.2.2" class="anchored" data-anchor-id="coldata">
<span class="header-section-number">2.2.2</span> colData</h3>
<p><code>colData</code> contains information about the samples used in the study. This information can include details such as the sample ID, the primers used in the analysis, the barcodes associated with the sample (truncated or complete), the type of sample (e.g.&nbsp;soil, faecal, mock) and a description of the sample.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 26 rows and 7 columns</span></span>
<span><span class="co">##          X.SampleID   Primer Final_Barcode Barcode_truncated_plus_T</span></span>
<span><span class="co">##            &lt;factor&gt; &lt;factor&gt;      &lt;factor&gt;                 &lt;factor&gt;</span></span>
<span><span class="co">##  CL3        CL3      ILBC_01        AACGCA                   TGCGTT</span></span>
<span><span class="co">##  CC1        CC1      ILBC_02        AACTCG                   CGAGTT</span></span>
<span><span class="co">##  SV1        SV1      ILBC_03        AACTGT                   ACAGTT</span></span>
<span><span class="co">##  M31Fcsw    M31Fcsw  ILBC_04        AAGAGA                   TCTCTT</span></span>
<span><span class="co">##  M11Fcsw    M11Fcsw  ILBC_05        AAGCTG                   CAGCTT</span></span>
<span><span class="co">##  ...            ...      ...           ...                      ...</span></span>
<span><span class="co">##  TS28         TS28   ILBC_25        ACCAGA                   TCTGGT</span></span>
<span><span class="co">##  TS29         TS29   ILBC_26        ACCAGC                   GCTGGT</span></span>
<span><span class="co">##  Even1        Even1  ILBC_27        ACCGCA                   TGCGGT</span></span>
<span><span class="co">##  Even2        Even2  ILBC_28        ACCTCG                   CGAGGT</span></span>
<span><span class="co">##  Even3        Even3  ILBC_29        ACCTGT                   ACAGGT</span></span>
<span><span class="co">##          Barcode_full_length SampleType</span></span>
<span><span class="co">##                     &lt;factor&gt;   &lt;factor&gt;</span></span>
<span><span class="co">##  CL3             CTAGCGTGCGT      Soil </span></span>
<span><span class="co">##  CC1             CATCGACGAGT      Soil </span></span>
<span><span class="co">##  SV1             GTACGCACAGT      Soil </span></span>
<span><span class="co">##  M31Fcsw         TCGACATCTCT      Feces</span></span>
<span><span class="co">##  M11Fcsw         CGACTGCAGCT      Feces</span></span>
<span><span class="co">##  ...                     ...        ...</span></span>
<span><span class="co">##  TS28            GCATCGTCTGG      Feces</span></span>
<span><span class="co">##  TS29            CTAGTCGCTGG      Feces</span></span>
<span><span class="co">##  Even1           TGACTCTGCGG      Mock </span></span>
<span><span class="co">##  Even2           TCTGATCGAGG      Mock </span></span>
<span><span class="co">##  Even3           AGAGAGACAGG      Mock </span></span>
<span><span class="co">##                                         Description</span></span>
<span><span class="co">##                                            &lt;factor&gt;</span></span>
<span><span class="co">##  CL3     Calhoun South Carolina Pine soil, pH 4.9  </span></span>
<span><span class="co">##  CC1     Cedar Creek Minnesota, grassland, pH 6.1  </span></span>
<span><span class="co">##  SV1     Sevilleta new Mexico, desert scrub, pH 8.3</span></span>
<span><span class="co">##  M31Fcsw M3, Day 1, fecal swab, whole body study   </span></span>
<span><span class="co">##  M11Fcsw M1, Day 1, fecal swab, whole body study   </span></span>
<span><span class="co">##  ...                                            ...</span></span>
<span><span class="co">##  TS28                                       Twin #1</span></span>
<span><span class="co">##  TS29                                       Twin #2</span></span>
<span><span class="co">##  Even1                                      Even1  </span></span>
<span><span class="co">##  Even2                                      Even2  </span></span>
<span><span class="co">##  Even3                                      Even3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To illustrate, <strong><em>X.SampleID</em></strong> gives the sample identifier, <strong><em>SampleType</em></strong> indicates the sample type (e.g.&nbsp;soil, fecal matter, control) and <strong><em>Description</em></strong> provides an additional description of the sample.</p>
</section><section id="rowdata" class="level3" data-number="2.2.3"><h3 data-number="2.2.3" class="anchored" data-anchor-id="rowdata">
<span class="header-section-number">2.2.3</span> rowData</h3>
<p><code>rowData</code> contains data on the features of the analyzed samples. Of particular interest to the microbiome field, this is used to store taxonomic information. This taxonomic information is extremely important for understanding the composition and diversity of the microbiome in each sample analyzed. It enables you to identify the different types of micro-organisms present in samples. It also allows you to explore the relationships between microbiome composition and various environmental or health factors.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 19216 rows and 7 columns</span></span>
<span><span class="co">##             Kingdom        Phylum        Class        Order        Family</span></span>
<span><span class="co">##         &lt;character&gt;   &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">##  549322     Archaea Crenarchaeota Thermoprotei           NA            NA</span></span>
<span><span class="co">##  522457     Archaea Crenarchaeota Thermoprotei           NA            NA</span></span>
<span><span class="co">##  951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae</span></span>
<span><span class="co">##  244423     Archaea Crenarchaeota        Sd-NA           NA            NA</span></span>
<span><span class="co">##  586076     Archaea Crenarchaeota        Sd-NA           NA            NA</span></span>
<span><span class="co">##  ...            ...           ...          ...          ...           ...</span></span>
<span><span class="co">##  278222    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  463590    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  535321    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  200359    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  271582    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##               Genus                Species</span></span>
<span><span class="co">##         &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">##  549322          NA                     NA</span></span>
<span><span class="co">##  522457          NA                     NA</span></span>
<span><span class="co">##  951     Sulfolobus Sulfolobusacidocalda..</span></span>
<span><span class="co">##  244423          NA                     NA</span></span>
<span><span class="co">##  586076          NA                     NA</span></span>
<span><span class="co">##  ...            ...                    ...</span></span>
<span><span class="co">##  278222          NA                     NA</span></span>
<span><span class="co">##  463590          NA                     NA</span></span>
<span><span class="co">##  535321          NA                     NA</span></span>
<span><span class="co">##  200359          NA                     NA</span></span>
<span><span class="co">##  271582          NA                     NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="rowtree" class="level3" data-number="2.2.4"><h3 data-number="2.2.4" class="anchored" data-anchor-id="rowtree">
<span class="header-section-number">2.2.4</span> rowTree</h3>
<p>Phylogenetic trees also play an important role in the microbiome field. The <code>TreeSE</code> class can keep track of features and node relations via two functions, <code>rowTree</code> and <code>rowLinks</code>.</p>
<p>A tree can be accessed via <code>rowTree</code> as <code>phylo</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Phylogenetic tree with 19216 tips and 19215 internal nodes.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Tip labels:</span></span>
<span><span class="co">##    549322, 522457, 951, 244423, 586076, 246140, ...</span></span>
<span><span class="co">##  Node labels:</span></span>
<span><span class="co">##    , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Rooted; includes branch lengths.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The links to the individual features are available through <code>rowLinks</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowLinks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  LinkDataFrame with 19216 rows and 5 columns</span></span>
<span><span class="co">##            nodeLab   nodeNum nodeLab_alias    isLeaf   whichTree</span></span>
<span><span class="co">##        &lt;character&gt; &lt;integer&gt;   &lt;character&gt; &lt;logical&gt; &lt;character&gt;</span></span>
<span><span class="co">##  1          549322         1       alias_1      TRUE       phylo</span></span>
<span><span class="co">##  2          522457         2       alias_2      TRUE       phylo</span></span>
<span><span class="co">##  3             951         3       alias_3      TRUE       phylo</span></span>
<span><span class="co">##  4          244423         4       alias_4      TRUE       phylo</span></span>
<span><span class="co">##  5          586076         5       alias_5      TRUE       phylo</span></span>
<span><span class="co">##  ...           ...       ...           ...       ...         ...</span></span>
<span><span class="co">##  19212      278222     19212   alias_19212      TRUE       phylo</span></span>
<span><span class="co">##  19213      463590     19213   alias_19213      TRUE       phylo</span></span>
<span><span class="co">##  19214      535321     19214   alias_19214      TRUE       phylo</span></span>
<span><span class="co">##  19215      200359     19215   alias_19215      TRUE       phylo</span></span>
<span><span class="co">##  19216      271582     19216   alias_19216      TRUE       phylo</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Please note that there can be a 1:1 relationship between tree nodes and features, but this is not a must-have. This means there can be features, which are not linked to nodes, and nodes, which are not linked to features. To change the links in an existing object, the <code>changeTree</code> function is available.</p>
</section><section id="sec-alt-exp" class="level3 page-columns page-full" data-number="2.2.5"><h3 data-number="2.2.5" class="anchored" data-anchor-id="sec-alt-exp">
<span class="header-section-number">2.2.5</span> Alternative experiments</h3>
<p><em><strong>Alternative experiments</strong></em> complement <em>assays</em>. They can contain complementary data, which is no longer tied to the same dimensions as the assay data. However, the number of samples (columns) must be the same.</p>
<p>This can come into play, for instance, when one has taxonomic abundance profiles quantified with different measurement technologies, such as phylogenetic microarrays, amplicon sequencing, or metagenomic sequencing. Another common use case is including abundance tables for different taxonomic ranks. Such alternative experiments concerning the same set of samples can be stored as</p>
<ol type="1">
<li>Separate <em>assays</em> assuming that the taxonomic information can be mapped between features directly 1:1; or</li>
<li>Data in the <em>altExp</em> slot of the <code>TreeSE</code>, if the feature dimensions differ. Each element of the <em>altExp</em> slot is a <code>SE</code> or an object from a derived class with independent feature data.</li>
</ol>
<p>The following shows how to store taxonomic abundance tables agglomerated at different taxonomic levels. However, the data could as well originate from entirely different measurement sources as long as the samples match.</p>
<p>Let us first agglomerate the data to Phylum level. This yields a new <code>TreeSE</code> data object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse_phylum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/deprecate.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Phylum"</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Both have the same number of columns (samples)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] 19216    26</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">tse_phylum</span><span class="op">)</span></span>
<span><span class="co">##  [1] 66 26</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we can add the new phylum-level data object as an alternative experiment in the original data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add the new data object to the original data object as an alternative experiment with the name "Phylum"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Phylum"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tse_phylum</span></span>
<span></span>
<span><span class="co"># Check the alternative experiment names available in the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExpNames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Phylum"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now subset the data, for instance, and this acts on both altExp and assay data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 19216 10 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(2): counts relabundance</span></span>
<span><span class="co">##  rownames(19216): 549322 522457 ... 200359 271582</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(10): CL3 CC1 ... M31Tong M11Tong</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(1): Phylum</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (19216 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (19216 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,<span class="st">"Phylum"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 66 10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For more details on <em>altExp</em>, you can check the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">introduction</a> to the <code>SingleCellExperiment</code> package <span class="citation" data-cites="R_SingleCellExperiment">(<a href="#ref-R_SingleCellExperiment" role="doc-biblioref">Lun and Risso 2020</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-R_SingleCellExperiment" class="csl-entry" role="listitem">
Lun, Aaron, and Davide Risso. 2020. <em>SingleCellExperiment: S4 Classes for Single Cell Data</em>.
</div></div></section><section id="sec-mae" class="level3 page-columns page-full" data-number="2.2.6"><h3 data-number="2.2.6" class="anchored" data-anchor-id="sec-mae">
<span class="header-section-number">2.2.6</span> MAEs</h3>
<p><em><strong>Multiple experiments</strong></em> relate to complementary measurement types, such as transcriptomic or metabolomic profiling of the microbiome or the host. Multiple experiments can be represented using the same options as alternative experiments, or by using the <code>MAE</code> class <span class="citation" data-cites="Ramos2017">(<a href="#ref-Ramos2017" role="doc-biblioref">Ramos et al. 2017</a>)</span>. Depending on how the datasets relate to each other the data can be stored as:</p>
<div class="no-row-height column-margin column-container"><div id="ref-Ramos2017" class="csl-entry" role="listitem">
Ramos, Marcel, Lucas Schiffer, Angela Re, Rimsha Azhar, Azfar Basunia, Carmen Rodriguez Cabrera, Tiffany Chan, et al. 2017. <span>“Software for the Integration of Multiomics Experiments in Bioconductor.”</span> <em>Cancer Research</em>. <a href="https://doi.org/10.1158/0008-5472.CAN-17-0344">https://doi.org/10.1158/0008-5472.CAN-17-0344</a>.
</div></div><ol type="1">
<li>Separate <em>altExp</em> if the samples can be matched directly 1:1; or</li>
<li>As <code>MAE</code> objects, in which the connections between samples are defined through a <code>sampleMap</code>. Each element on the <code>experimentsList</code> of an <code>MAE</code> is <code>matrix</code> or <code>matrix</code>-like objects, including <code>SE</code> objects, and the number of samples can differ between the elements.</li>
</ol>
<!--

::: {.cell layout-align="center"}

```{.r .cell-code}
#TODO: Find the right dataset to explain a non 1:1 sample relationship
```
:::

--><p>For information have a look at the <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html">intro vignette</a> of the <code>MultiAssayExperiment</code> package.</p>
<div id="tbl-options" class="quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-options-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;2.1: <strong>Recommended options for storing multiple data tables in microbiome studies</strong> The <em>assays</em> are best suited for data transformations (one-to-one match between samples and columns across the assays). The <em>alternative experiments</em> are particularly suitable for alternative versions of the data that are of same type but may have a different number of features (e.g.&nbsp;taxonomic groups); this is for instance the case with taxonomic abundance tables agglomerated at different levels (e.g.&nbsp;genus vs.&nbsp;phyla) or alternative profiling technologies (e.g.&nbsp;amplicon sequencing vs.&nbsp;shallow shotgun metagenomics). For alternative experiments one-to-one match between samples (cols) is libraryd but the alternative experiment tables can have different numbers of features (rows). Finally, elements of the <em>MAE</em> provide the most flexible way to incorporate multi-omic data tables with flexible numbers of samples and features. We recommend these conventions as the basis for methods development and application in microbiome studies.
</figcaption><div aria-describedby="tbl-options-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead><tr class="header">
<th style="text-align: right;">Option</th>
<th>Rows (features)</th>
<th style="text-align: right;">Cols (samples)</th>
<th style="text-align: right;">Recommended</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">assays</td>
<td>match</td>
<td style="text-align: right;">match</td>
<td style="text-align: right;">Data transformations</td>
</tr>
<tr class="even">
<td style="text-align: right;">altExp</td>
<td>free</td>
<td style="text-align: right;">match</td>
<td style="text-align: right;">Alternative experiments</td>
</tr>
<tr class="odd">
<td style="text-align: right;">MultiAssay</td>
<td>free</td>
<td style="text-align: right;">free (mapping)</td>
<td style="text-align: right;">Multi-omic experiments</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
</section></section><section id="sec-example-data" class="level2 page-columns page-full" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="sec-example-data">
<span class="header-section-number">2.3</span> Demonstration data</h2>
<p>Open demonstration data for testing and benchmarking purposes is available from multiple locations. This chapter introduces some options. The other chapters of this book provide ample examples about the use of the data.</p>
<section id="sec-package-data" class="level3" data-number="2.3.1"><h3 data-number="2.3.1" class="anchored" data-anchor-id="sec-package-data">
<span class="header-section-number">2.3.1</span> Package data</h3>
<p>The <code>mia</code> R package contains example datasets that are direct conversions from the alternative <code>phyloseq</code> container to the <code>TreeSE</code> container.</p>
<p>List the <a href="https://microbiome.github.io/mia/reference/mia-datasets.html">available datasets</a> in the <code>mia</code> package:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load the <code>GlobalPatterns</code> data from the <code>mia</code> package:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">GlobalPatterns</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 19216 26 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(19216): 549322 522457 ... 200359 271582</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (19216 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (19216 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>R packages contain additional demonstration data sets (see the Datasets section of the reference page):</p>
<ul>
<li><a href="https://microbiome.github.io/mia/reference/index.html">mia reference</a></li>
<li><a href="https://microbiome.github.io/miaViz/reference/index.html">miaViz reference</a></li>
<li><a href="https://microbiome.github.io/miaTime/reference/index.html">miaTime reference</a></li>
</ul></section><section id="experimenthub-data" class="level3 page-columns page-full" data-number="2.3.2"><h3 data-number="2.3.2" class="anchored" data-anchor-id="experimenthub-data">
<span class="header-section-number">2.3.2</span> ExperimentHub data</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/ExperimentHub/inst/doc/ExperimentHub.html">ExperimentHub</a> provides a variety of data resources, including the <a href="https://bioconductor.org/packages/release/data/experiment/html/microbiomeDataSets.html">microbiomeDataSets</a> package <span class="citation" data-cites="Morgan2021 microlahti2021">(<a href="#ref-Morgan2021" role="doc-biblioref">Morgan and Shepherd 2021</a>; <a href="#ref-microlahti2021" role="doc-biblioref">Lahti, Ernst, and Shetty 2021</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Morgan2021" class="csl-entry" role="listitem">
Morgan, Martin, and Lori Shepherd. 2021. <em>ExperimentHub: Client to Access ExperimentHub Resources</em>.
</div><div id="ref-microlahti2021" class="csl-entry" role="listitem">
Lahti, Leo, Felix G. M. Ernst, and Sudarshan Shetty. 2021. <em>microbiomeDataSets: Experiment Hub Based Microbiome Datasets</em>.
</div></div><p>A table of the available datasets is available through the <code>availableDataSets</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">microbiomeDataSets</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/microbiomeDataSets/man/availableDataSets.html">availableDataSets</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##              Dataset</span></span>
<span><span class="co">##  1  GrieneisenTSData</span></span>
<span><span class="co">##  2    HintikkaXOData</span></span>
<span><span class="co">##  3       LahtiMLData</span></span>
<span><span class="co">##  4        LahtiMData</span></span>
<span><span class="co">##  5       LahtiWAData</span></span>
<span><span class="co">##  6      OKeefeDSData</span></span>
<span><span class="co">##  7 SilvermanAGutData</span></span>
<span><span class="co">##  8        SongQAData</span></span>
<span><span class="co">##  9   SprockettTHData</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All data are downloaded from ExperimentHub and cached for local re-use. Check the <a href="https://microbiome.github.io/microbiomeDataSets/reference/index.html">man pages of each function</a> for a detailed documentation of the data contents and references. Let us retrieve a <em><a href="https://bioconductor.org/packages/3.20/MAE">MAE</a></em> dataset:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># mae &lt;- HintikkaXOData()</span></span>
<span><span class="co"># Since HintikkaXOData is now added to mia, we can load it directly from there</span></span>
<span><span class="co"># We suggest to check other datasets from microbiomeDataSets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">HintikkaXOData</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">mae</span> <span class="op">&lt;-</span> <span class="va">HintikkaXOData</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Data is available in <em><a href="https://bioconductor.org/packages/3.20/SE">SE</a></em>, <code>r Biocpkg("TreeSE")</code> and <code>r Biocpkg("MAE")</code> data containers; see the separate page on <a href="../pages/23_multi-assay_analyses.html">alternative containers</a> for more details.</p>
</section><section id="curated-metagenomic-data" class="level3 page-columns page-full" data-number="2.3.3"><h3 data-number="2.3.3" class="anchored" data-anchor-id="curated-metagenomic-data">
<span class="header-section-number">2.3.3</span> Curated metagenomic data</h3>
<p><a href="https://bioconductor.org/packages/release/data/experiment/html/curatedMetagenomicData.html">curatedMetagenomicData</a> is a large collection of curated human microbiome datasets, provided as <code>TreeSE</code> objects <span class="citation" data-cites="Pasolli2017">(<a href="#ref-Pasolli2017" role="doc-biblioref">E 2017</a>)</span>. The resource provides curated human microbiome data including gene families, marker abundance, marker presence, pathway abundance, pathway coverage, and relative abundance for samples from different body sites. See the package homepage for more details on data availability and access.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Pasolli2017" class="csl-entry" role="listitem">
E, Pasolli. 2017. <span>“Accessible, Curated Metagenomic Data Through ExperimentHub.”</span> <em>Nature Methods</em> 14 (11): 1023–24. <a href="https://www.nature.com/articles/nmeth.4468">https://www.nature.com/articles/nmeth.4468</a>.
</div><div id="ref-Vatanen2016" class="csl-entry" role="listitem">
Vatanen, Tommi, Aleksandar D. Kostic, Eva d’Hennezel, Heli Siljander, Eric A. Franzosa, Moran Yassour, Raivo Kolde, et al. 2016. <span>“Variation in Microbiome LPS Immunogenicity Contributes to Autoimmunity in Humans.”</span> <em>Cell</em> 165 (May): 842–53. <a href="https://doi.org/10.1016/j.cell.2016.04.007">https://doi.org/10.1016/j.cell.2016.04.007</a>.
</div></div><p>As one example, let us retrieve the Vatanen (2016) <span class="citation" data-cites="Vatanen2016">(<a href="#ref-Vatanen2016" role="doc-biblioref">Vatanen et al. 2016</a>)</span> data set. This is a larger collection with a bit longer download time.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/curatedMetagenomicData">curatedMetagenomicData</a></span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/curatedMetagenomicData.html">curatedMetagenomicData</a></span><span class="op">(</span><span class="st">"Vatanen*"</span>, dryrun <span class="op">=</span> <span class="cn">FALSE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="human-microbiome-compendium" class="level3" data-number="2.3.4"><h3 data-number="2.3.4" class="anchored" data-anchor-id="human-microbiome-compendium">
<span class="header-section-number">2.3.4</span> Human microbiome compendium</h3>
<p><a href="https://seandavi.github.io/MicroBioMap/">MicroBioMap</a> dataset includes over 170k samples of publicly available 16S rRNA amplicon sequencing data, all processed using the same pipeline and reference database<span class="citation" data-cites="Abdill2023">(<a href="#ref-Abdill2023" role="doc-biblioref"><strong>Abdill2023?</strong></a>)</span>. After installing the MicroBioMap package (see the <a href="https://github.com/seandavi/MicroBioMap#microbiome-compendium">original website</a> for instructions), you can load the compendium with</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MicroBioMap</span><span class="op">)</span></span>
<span><span class="va">cpd</span> <span class="op">&lt;-</span> <span class="fu">getCompendium</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This returns a <code>TreeSE</code> object. Currently, the “tree” part of the <code>TreeSE</code> is not populated, but that is on the roadmap<span class="citation" data-cites="compendiumpackage">(<a href="#ref-compendiumpackage" role="doc-biblioref"><strong>compendiumpackage?</strong></a>)</span>.</p>
<p>After loading the compendium, you will have immediate access to nearly 170,000 microbiome samples of publicly available 16S rRNA amplicon sequencing data, all processed using the same pipeline and reference database. For more use examples in R/Bioconductor, see the <a href="https://seandavi.github.io/MicroBioMap/articles/overview.html">MicroBioMap vignette</a>.</p>
</section><section id="other-data-sources" class="level3" data-number="2.3.5"><h3 data-number="2.3.5" class="anchored" data-anchor-id="other-data-sources">
<span class="header-section-number">2.3.5</span> Other data sources</h3>
<p>The current collections provide access to vast microbiome data resources. The output has to be converted into <code>TreeSE/MAE</code> separately.</p>
<ul>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/MGnifyR.html">MGnifyR</a> provides access to <a href="https://www.ebi.ac.uk/metagenomics/">EBI/MGnify</a>
</li>
<li>
<a href="https://github.com/cran/qiitr">qiitr</a> provides access to <a href="https://qiita.com/about">QIITA</a>
</li>
<li>
<a href="https://github.com/microbiome/qiime2R">qiime2R</a> provides access to <a href="https://docs.qiime2.org/2024.2/">QIIME2</a>
</li>
</ul></section></section><section id="sec-loading-experimental-microbiome-data" class="level2" data-number="2.4"><h2 data-number="2.4" class="anchored" data-anchor-id="sec-loading-experimental-microbiome-data">
<span class="header-section-number">2.4</span> Loading experimental microbiome data</h2>
<section id="s-workflow" class="level3" data-number="2.4.1"><h3 data-number="2.4.1" class="anchored" data-anchor-id="s-workflow">
<span class="header-section-number">2.4.1</span> 16S workflow</h3>
<p>Result of amplicon sequencing is a large number of files that include all the sequences that were read from samples. Those sequences need to be matched with taxa. Additionally, we need to know how many times each taxa were found from each sample.</p>
<p>There are several algorithms to do that, and DADA2 is one of the most common. You can find DADA2 pipeline tutorial, for example, <a href="https://benjjneb.github.io/dada2/tutorial.html">here</a>. After the DADA2 portion of the tutorial is completed, the data is stored into <em>phyloseq</em> object (Bonus: Handoff to phyloseq). To store the data to <em>TreeSE</em>, follow the example below.</p>
<p>You can find full workflow script without further explanations and comments from <a href="https://github.com/microbiome/OMA/blob/master/dada2_workflow.Rmd">here</a></p>
<p>Load required packages.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/">BiocManager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Create arbitrary example sample metadata like it was done in the tutorial. Usually, sample metadata is imported as a file.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">samples.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">seqtab.nochim</span><span class="op">)</span></span>
<span><span class="va">subject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">samples.out</span>, <span class="st">"D"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gender</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">subject</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">subject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">subject</span>,<span class="fl">2</span>,<span class="fl">999</span><span class="op">)</span></span>
<span><span class="va">day</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">samples.out</span>, <span class="st">"D"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">samdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Subject<span class="op">=</span><span class="va">subject</span>, Gender<span class="op">=</span><span class="va">gender</span>, Day<span class="op">=</span><span class="va">day</span><span class="op">)</span></span>
<span><span class="va">samdf</span><span class="op">$</span><span class="va">When</span> <span class="op">&lt;-</span> <span class="st">"Early"</span></span>
<span><span class="va">samdf</span><span class="op">$</span><span class="va">When</span><span class="op">[</span><span class="va">samdf</span><span class="op">$</span><span class="va">Day</span><span class="op">&gt;</span><span class="fl">100</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Late"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">samdf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples.out</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Convert data into right format and create a <em>TreeSE</em> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a list that contains assays</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html">t</a></span><span class="op">(</span><span class="va">seqtab.nochim</span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">assays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/SimpleList-class.html">SimpleList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert colData and rowData into DataFrame</span></span>
<span><span class="va">samdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">samdf</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create TreeSE</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-constructor.html">TreeSummarizedExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="va">assays</span>,</span>
<span>                                colData <span class="op">=</span> <span class="va">samdf</span>,</span>
<span>                                rowData <span class="op">=</span> <span class="va">taxa</span></span>
<span>                                <span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove mock sample like it is also done in DADA2 pipeline tutorial</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span> , <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"mock"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Add sequences into <em>referenceSeq</em> slot and convert rownames into simpler format.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Convert sequences into right format</span></span>
<span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="co"># Add sequences into referenceSeq slot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">referenceSeq</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dna</span></span>
<span><span class="co"># Convert rownames into ASV_number format</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"ASV"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html">nrow</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 232 20 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(232): ASV1 ASV2 ... ASV231 ASV232</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(20): F3D0 F3D1 ... F3D9 Mock</span></span>
<span><span class="co">##  colData names(4): Subject Gender Day When</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span>
<span><span class="co">##  referenceSeq: a DNAStringSet (232 sequences)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-import-from-file" class="level3" data-number="2.4.2"><h3 data-number="2.4.2" class="anchored" data-anchor-id="sec-import-from-file">
<span class="header-section-number">2.4.2</span> Import from external files</h3>
<p>Microbiome (taxonomic) profiling data is commonly distributed in various file formats. You can import such external data files as a <code>TreeSE</code> object, but the details depend on the file format. Here, we provide examples for common formats. Some datasets and raw files to learn how to import raw data and construct <code>TreeSE/MAE</code> containers are available in the <a href="https://github.com/microbiome/data">microbiome data repository</a>.</p>
<section id="csv-import" class="level4" data-number="2.4.2.1"><h4 data-number="2.4.2.1" class="anchored" data-anchor-id="csv-import">
<span class="header-section-number">2.4.2.1</span> CSV import</h4>
<p><strong>CSV data tables</strong> can be imported with the standard R functions, then converted to the desired format. For detailed examples, you can check the <a href="https://bioconductor.org/help/course-materials/2019/BSS2019/04_Practical_CoreApproachesInBioconductor.html">Bioconductor course material</a> by Martin Morgan. You can also check the <a href="https://github.com/microbiome/OMA/tree/master/data">example files</a> and construct your own CSV files accordingly.</p>
<p>Recommendations for the CSV files are the following. File names are arbitrary; we refer here to the same names as in the examples:</p>
<ul>
<li><p>Abundance table (<code>assay_taxa.csv</code>): data matrix (features x samples); first column provides feature IDs, the first row provides sample IDs; other values should be numeric (abundances).</p></li>
<li><p>Row data (<code>rowdata_taxa.csv</code>): data table (features x info); first column provides feature IDs, the first row provides column headers; this file usually contains the taxonomic mapping between different taxonomic levels. Ideally, the feature IDs (row names) match one-to-one with the abundance table row names.</p></li>
<li><p>Column data (<code>coldata.csv</code>): data table (samples x info); first column provides sample IDs, the first row provides column headers; this file usually contains the sample metadata/phenodata (such as subject age, health etc). Ideally, the sample IDs match one-to-one with the abundance table column names.</p></li>
</ul>
<p>After you have set up the CSV files, you can read them in R:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">count_file</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"assay_taxa.csv"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span><span class="va">tax_file</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rowdata_taxa.csv"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span><span class="va">sample_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"coldata.csv"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load files</span></span>
<span><span class="va">counts</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">count_file</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>   <span class="co"># Abundance table (e.g. ASV data; to assay data)</span></span>
<span><span class="va">tax</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">tax_file</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>     <span class="co"># Taxonomy table (to rowData)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">sample_file</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>  <span class="co"># Sample data (to colData)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After reading the data in R, ensure the following:</p>
<ul>
<li><p>abundance table (<code>counts</code>): numeric <code>matrix</code>, with feature IDs as rownames and sample IDs as column names.</p></li>
<li><p>rowdata (<code>tax</code>): <code>DataFrame</code>, with feature IDs as rownames. If this is a <code>data.frame</code> you can use the function <code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame()</a></code> to change the format. Column names are free but in microbiome analysis they usually they refer to taxonomic ranks. The rownames in rowdata should match with rownames in abundance table.</p></li>
<li><p>coldata (<code>samples</code>): <code>DataFrame</code>, with sample IDs as rownames. If this is a <code>data.frame</code> you can use the function <code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame()</a></code> to change the format. Column names are free. The rownames in <code>coldata</code> should match with colnames in abundance table.</p></li>
</ul>
<p><strong>Always ensure that the tables have rownames!</strong> The <em>TreeSE</em> constructor compares rownames and ensures that, for example, right samples are linked with right patient.</p>
<p>Also ensure that the row and column names match one-to-one between abundance table, <code>rowdata</code>, and <code>coldata</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Match rows and columns</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">tax</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Let's ensure that the data is in correct (numeric matrix) format:</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you hesitate about the format of the data, you can compare to one of the available demonstration datasets, and make sure that your data components have the same format.</p>
<p>There are many different source files and many different ways to read data in R. One can do data manipulation in R as well. Investigate the entries as follows.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># coldata rownames match assay colnames</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="co"># our dataset</span></span>
<span><span class="co">##  [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span> <span class="co"># should be data.frame or DataFrame</span></span>
<span><span class="co">##  [1] "data.frame"</span></span>
<span></span>
<span><span class="co"># rowdata rownames match assay rownames</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">tax</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="co"># our dataset</span></span>
<span><span class="co">##  [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">tax</span><span class="op">)</span> <span class="co"># should be data.frame or DataFrame</span></span>
<span><span class="co">##  [1] "data.frame"</span></span>
<span></span>
<span><span class="co"># Counts </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="co"># should be a numeric matrix</span></span>
<span><span class="co">##  [1] "matrix" "array"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="constructing-treese" class="level3" data-number="2.4.3"><h3 data-number="2.4.3" class="anchored" data-anchor-id="constructing-treese">
<span class="header-section-number">2.4.3</span> Constructing TreeSE</h3>
<p>Now, let’s create the <code>TreeSE</code> object from the input data tables. Here we also convert the data objects in their preferred formats:</p>
<ul>
<li>counts –&gt; numeric matrix</li>
<li>rowData –&gt; DataFrame</li>
<li>colData –&gt; DataFrame</li>
</ul>
<p>The <code>SimpleList</code> could be used to include multiple alternative assays, if necessary.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a TreeSE</span></span>
<span><span class="va">tse_taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-constructor.html">TreeSummarizedExperiment</a></span><span class="op">(</span>assays <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/SimpleList-class.html">SimpleList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span><span class="op">)</span>,</span>
<span>                                     colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>,</span>
<span>                                     rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">tax</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tse_taxa</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 12706 40 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(12706): GAYR01026362.62.2014 CVJT01000011.50.2173 ...</span></span>
<span><span class="co">##    JRJTB:03787:02429 JRJTB:03787:02478</span></span>
<span><span class="co">##  rowData names(7): Phylum Class ... Species OTU</span></span>
<span><span class="co">##  colnames(40): C1 C2 ... C39 C40</span></span>
<span><span class="co">##  colData names(6): Sample Rat ... Fat XOS</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now you should have a ready-made <code>TreeSE</code> data object that can be used in downstream analyses.</p>
</section><section id="constructing-mae" class="level3" data-number="2.4.4"><h3 data-number="2.4.4" class="anchored" data-anchor-id="constructing-mae">
<span class="header-section-number">2.4.4</span> Constructing MAE</h3>
<p>To construct a <em>MAE</em> object, just combine multiple <em>TreeSE</em> data containers. Here we import metabolite data from the same study.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">count_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"assay_metabolites.csv"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span><span class="va">sample_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"coldata.csv"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load files</span></span>
<span><span class="va">counts</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">count_file</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span>  </span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">sample_file</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a TreeSE for the metabolite data</span></span>
<span><span class="va">tse_metabolite</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-constructor.html">TreeSummarizedExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/SimpleList-class.html">SimpleList</a></span><span class="op">(</span>concs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                           colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tse_metabolite</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 38 40 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): concs</span></span>
<span><span class="co">##  rownames(38): Butyrate Acetate ... Malonate 1,3-dihydroxyacetone</span></span>
<span><span class="co">##  rowData names(0):</span></span>
<span><span class="co">##  colnames(40): C1 C2 ... C39 C40</span></span>
<span><span class="co">##  colData names(6): Sample Rat ... Fat XOS</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can combine these two experiments into <em>MAE</em>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create an ExperimentList that includes experiments</span></span>
<span><span class="va">experiments</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/ExperimentList.html">ExperimentList</a></span><span class="op">(</span>microbiome <span class="op">=</span> <span class="va">tse_taxa</span>, </span>
<span>                              metabolite <span class="op">=</span> <span class="va">tse_metabolite</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a MAE</span></span>
<span><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span><span class="op">(</span>experiments <span class="op">=</span> <span class="va">experiments</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mae</span></span>
<span><span class="co">##  A MultiAssayExperiment object of 2 listed</span></span>
<span><span class="co">##   experiments with user-defined names and respective classes.</span></span>
<span><span class="co">##   Containing an ExperimentList class object of length 2:</span></span>
<span><span class="co">##   [1] microbiome: TreeSummarizedExperiment with 12706 rows and 40 columns</span></span>
<span><span class="co">##   [2] metabolite: TreeSummarizedExperiment with 38 rows and 40 columns</span></span>
<span><span class="co">##  Functionality:</span></span>
<span><span class="co">##   experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">##   colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">##   sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">##   `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">##   *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">##   assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">##   exportClass() - save data to flat files</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="import-functions-for-standard-formats" class="level3" data-number="2.4.5"><h3 data-number="2.4.5" class="anchored" data-anchor-id="import-functions-for-standard-formats">
<span class="header-section-number">2.4.5</span> Import functions for standard formats</h3>
<p>Specific import functions are provided for:</p>
<ul>
<li>Biom files (see <code>help(mia::importBIOM)</code>)</li>
<li>QIIME2 files (see <code>help(mia::importQIIME2)</code>)</li>
<li>Mothur files (see <code>help(mia::importMothur)</code>)</li>
</ul>
<section id="biom-import" class="level4" data-number="2.4.5.1"><h4 data-number="2.4.5.1" class="anchored" data-anchor-id="biom-import">
<span class="header-section-number">2.4.5.1</span> Biom import</h4>
<p>Here we show how <a href="https://biom-format.org/">Biom files</a> are imported into a <code>TreeSE</code> object using as an example Tengeler2020, which is further described <a href="https://microbiome.github.io/mia/reference/Tengeler2020.html">here</a>. This dataset consists of 3 files, which can be fetched or downloaded from <a href="https://github.com/microbiome/data/tree/main/Tengeler2020">this repository</a>:</p>
<ul>
<li>biom file: abundance table and taxonomy information</li>
<li>csv file: sample metadata</li>
<li>tree file: phylogenetic tree</li>
</ul>
<p>To begin with, we store the data in a local directory within the working directory, such as <em>data/</em>, and define the source file paths.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">biom_file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Aggregated_humanization2.biom"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span><span class="va">sample_meta_file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Mapping_file_ADHD_aggregated.csv"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span>
<span><span class="va">tree_file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Data_humanization_phylo_aggregation.tre"</span>, package <span class="op">=</span> <span class="st">"OMA"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can read in the biom file and convert it into a <code>TreeSE</code> object. In addition, we retrieve the rank names from the prefixes of the feature names and then remove them with the <code>rankFromPrefix</code> and <code>removeTaxaPrefixes</code> optional arguments.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read biom and convert it to TreeSE</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/makeTreeSEFromBiom.html">importBIOM</a></span><span class="op">(</span><span class="va">biom_file_path</span>,</span>
<span>                    rankFromPrefix <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    removeTaxaPrefixes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 151 27 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(151): 1726470 1726471 ... 17264756 17264757</span></span>
<span><span class="co">##  rowData names(6): taxonomy1 Phylum ... Family Genus</span></span>
<span><span class="co">##  colnames(27): A110 A111 ... A38 A39</span></span>
<span><span class="co">##  colData names(0):</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>assays</code> slot includes a list of abundance tables. The imported abundance table is named as “counts”. Let us inspect only the first cols and rows.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">##            A110  A111  A12</span></span>
<span><span class="co">##  1726470  17722 11630    0</span></span>
<span><span class="co">##  1726471  12052     0 2679</span></span>
<span><span class="co">##  17264731     0   970    0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>rowdata</code> includes taxonomic information from the biom file. The <code><a href="https://rdrr.io/r/utils/head.html">head()</a></code> command shows just the beginning of the data table for an overview.</p>
<p><code><a href="https://rdrr.io/pkg/knitr/man/kable.html">knitr::kable()</a></code> helps print the information more nicely.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">##              taxonomy1          Phylum            Class              Order</span></span>
<span><span class="co">##            &lt;character&gt;     &lt;character&gt;      &lt;character&gt;        &lt;character&gt;</span></span>
<span><span class="co">##  1726470  "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  1726471  "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  17264731 "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  17264726 "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  1726472  "k__Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales</span></span>
<span><span class="co">##  17264724 "k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##                        Family            Genus</span></span>
<span><span class="co">##                   &lt;character&gt;      &lt;character&gt;</span></span>
<span><span class="co">##  1726470       Bacteroidaceae     Bacteroides"</span></span>
<span><span class="co">##  1726471       Bacteroidaceae     Bacteroides"</span></span>
<span><span class="co">##  17264731  Porphyromonadaceae Parabacteroides"</span></span>
<span><span class="co">##  17264726      Bacteroidaceae     Bacteroides"</span></span>
<span><span class="co">##  1726472  Verrucomicrobiaceae     Akkermansia"</span></span>
<span><span class="co">##  17264724      Bacteroidaceae     Bacteroides"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We further polish the feature names by removing unnecessary characters and then replace the original rowData with its updated version.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Genus level has additional '\"', so let's delete that also</span></span>
<span><span class="va">rowdata_modified</span> <span class="op">&lt;-</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span>, </span>
<span>                                           FUN <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="va"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span>, </span>
<span>                                           pattern <span class="op">=</span> <span class="st">'\"'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rowdata_modified is a list, so convert this back to DataFrame format. </span></span>
<span><span class="co"># and assign the cleaned data back to the TSE rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">rowdata_modified</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now we have a nicer table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">##             taxonomy1          Phylum            Class              Order</span></span>
<span><span class="co">##           &lt;character&gt;     &lt;character&gt;      &lt;character&gt;        &lt;character&gt;</span></span>
<span><span class="co">##  1726470  k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  1726471  k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  17264731 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  17264726 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##  1726472  k__Bacteria Verrucomicrobia Verrucomicrobiae Verrucomicrobiales</span></span>
<span><span class="co">##  17264724 k__Bacteria   Bacteroidetes      Bacteroidia      Bacteroidales</span></span>
<span><span class="co">##                        Family           Genus</span></span>
<span><span class="co">##                   &lt;character&gt;     &lt;character&gt;</span></span>
<span><span class="co">##  1726470       Bacteroidaceae     Bacteroides</span></span>
<span><span class="co">##  1726471       Bacteroidaceae     Bacteroides</span></span>
<span><span class="co">##  17264731  Porphyromonadaceae Parabacteroides</span></span>
<span><span class="co">##  17264726      Bacteroidaceae     Bacteroides</span></span>
<span><span class="co">##  1726472  Verrucomicrobiaceae     Akkermansia</span></span>
<span><span class="co">##  17264724      Bacteroidaceae     Bacteroides</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We notice that the imported biom file did not contain any <code>colData</code> yet, so only an empty dataframe appears in this slot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 6 rows and 0 columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let us add <code>colData</code> from the sample metadata, which is stored in a CSV file.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># CSV file with colnames in the first row and rownames in the first column</span></span>
<span><span class="va">sample_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">sample_meta_file_path</span>,</span>
<span>                        sep <span class="op">=</span> <span class="st">","</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add this sample data to colData of the taxonomic data object</span></span>
<span><span class="co"># Note that the data must be given in a DataFrame format (required for our purposes)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span><span class="va">sample_meta</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the <code>colData</code> includes the sample metadata.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##         Treatment      Cohort TreatmentxCohort Description</span></span>
<span><span class="co">##       &lt;character&gt; &lt;character&gt;      &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##  A110        ADHD    Cohort_1    ADHD_Cohort_1        A110</span></span>
<span><span class="co">##  A12         ADHD    Cohort_1    ADHD_Cohort_1         A12</span></span>
<span><span class="co">##  A15         ADHD    Cohort_1    ADHD_Cohort_1         A15</span></span>
<span><span class="co">##  A19         ADHD    Cohort_1    ADHD_Cohort_1         A19</span></span>
<span><span class="co">##  A21         ADHD    Cohort_2    ADHD_Cohort_2         A21</span></span>
<span><span class="co">##  A23         ADHD    Cohort_2    ADHD_Cohort_2         A23</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we add a phylogenetic tree to the rowData slot. Such feature is available only in <code>TreeSE</code> objects. Similarly, Trees specifying the sample hierarchy can be stored in the colTree slot.</p>
<p>Here, we read in the file containing the phylogenetic tree and insert it in corresponding slot of the <code>TreeSE</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Reads the tree file</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="va">tree_file_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add tree to rowTree</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tree</span></span>
<span></span>
<span><span class="co"># Check</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 151 27 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(151): 1726470 1726471 ... 17264756 17264757</span></span>
<span><span class="co">##  rowData names(6): taxonomy1 Phylum ... Family Genus</span></span>
<span><span class="co">##  colnames(27): A110 A12 ... A35 A38</span></span>
<span><span class="co">##  colData names(4): Treatment Cohort TreatmentxCohort Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (151 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (151 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the rowTree slot contains the phylogenetic tree:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="conversions-between-data-formats-in-r" class="level3" data-number="2.4.6"><h3 data-number="2.4.6" class="anchored" data-anchor-id="conversions-between-data-formats-in-r">
<span class="header-section-number">2.4.6</span> Conversions between data formats in R</h3>
<p>If the data has already been imported in R in another format, it can be readily converted into <code>TreeSE</code>, as shown in our next example. Note that similar conversion functions to <code>TreeSE</code> are available for multiple data formats via the <code>mia</code> package (see makeTreeSummarizedExperimentFrom* for phyloseq, Biom, and DADA2).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># phyloseq example data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package<span class="op">=</span><span class="st">"phyloseq"</span><span class="op">)</span> </span>
<span><span class="va">GlobalPatterns_phyloseq</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span>
<span><span class="va">GlobalPatterns_phyloseq</span></span>
<span><span class="co">##  phyloseq-class experiment-level object</span></span>
<span><span class="co">##  otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]</span></span>
<span><span class="co">##  sample_data() Sample Data:       [ 26 samples by 7 sample variables ]</span></span>
<span><span class="co">##  tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]</span></span>
<span><span class="co">##  phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># convert phyloseq to TSE</span></span>
<span><span class="va">GlobalPatterns_TSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/makeTreeSEFromPhyloseq.html">makeTreeSummarizedExperimentFromPhyloseq</a></span><span class="op">(</span><span class="va">GlobalPatterns_phyloseq</span><span class="op">)</span> </span>
<span><span class="va">GlobalPatterns_TSE</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 19216 26 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(19216): 549322 522457 ... 200359 271582</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (19216 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (19216 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also convert <code>TreeSE</code> objects into <code>phyloseq</code> with respect to the shared components that are supported by both formats (i.e.&nbsp;taxonomic abundance table, sample metadata, taxonomic table, phylogenetic tree, sequence information). This is useful for instance when additional methods are available for <code>phyloseq</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># convert TSE to phyloseq</span></span>
<span><span class="va">GlobalPatterns_phyloseq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/makePhyloseqFromTreeSE.html">makePhyloseqFromTreeSummarizedExperiment</a></span><span class="op">(</span><span class="va">GlobalPatterns_TSE</span><span class="op">)</span> </span>
<span><span class="va">GlobalPatterns_phyloseq2</span></span>
<span><span class="co">##  phyloseq-class experiment-level object</span></span>
<span><span class="co">##  otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]</span></span>
<span><span class="co">##  sample_data() Sample Data:       [ 26 samples by 7 sample variables ]</span></span>
<span><span class="co">##  tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]</span></span>
<span><span class="co">##  phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Conversion is possible between other data formats. Interested readers can refer to the following functions:</p>
<ul>
<li>
<a href="https://microbiome.github.io/mia/reference/makeTreeSEFromDADA2.html">makeTreeSEFromDADA2</a><br>
</li>
<li>
<a href="https://microbiome.github.io/mia/reference/makeTreeSEFromBiom.html">makeTreeSEFromBiom</a><br>
</li>
<li>
<a href="https://microbiome.github.io/mia/reference/importMetaPhlAn.html">importMetaPhlAn</a><br>
</li>
<li><a href="https://microbiome.github.io/mia/reference/importQIIME2.html">importQZA</a></li>
</ul></section></section><section id="sec-exporting-data-container" class="level2" data-number="2.5"><h2 data-number="2.5" class="anchored" data-anchor-id="sec-exporting-data-container">
<span class="header-section-number">2.5</span> Exporting data container</h2>
<section id="export-treesummarizedexperiment" class="level3" data-number="2.5.1"><h3 data-number="2.5.1" class="anchored" data-anchor-id="export-treesummarizedexperiment">
<span class="header-section-number">2.5.1</span> Export TreeSummarizedExperiment</h3>
<p>Exporting a <code>TreeSE</code> data container can be done using <em>feather</em> package. TSE object has to be converted into a dataframe (<code>data.frame</code> and not <code>DataFrame</code>) . The output file is a .feather file, which can be imported in other languages such as Julia or Python. For information, have a look at the <a href="https://juliapackages.com/p/featherfiles">FeatherFile</a> Julia package and <a href="https://pypi.org/project/feather-format/">feather-format</a> Python library.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Import TSE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span>
<span><span class="co"># Convert into a data.frame </span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, MARGIN <span class="op">=</span> <span class="st">"samples"</span>, method<span class="op">=</span><span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="va">molten_tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/deprecate.html">meltAssay</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                        add_row_data <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        add_col_data <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        assay.type <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="co"># Export as a feather file</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wesm/feather">feather</a></span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"path/to/tse.feather"</span></span>
<span><span class="fu">write_feather</span><span class="op">(</span><span class="va">molten_tse</span>,<span class="va">path</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Another way could be using a CSV file. This works the same as for a feather file, make sure you have converted your <code>TreeSE</code> data container as a dataframe. Here note that you can decide whether you want to write the row names or not.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span><span class="op">(</span><span class="va">molten_tse</span>, <span class="st">"path/to/tse.csv"</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="export-multiassayexperiment" class="level3" data-number="2.5.2"><h3 data-number="2.5.2" class="anchored" data-anchor-id="export-multiassayexperiment">
<span class="header-section-number">2.5.2</span> Export MultiAssayExperiment</h3>
<p>Exporting a <code>MultiAssayExperiment</code> data container can also be done using <em>feather</em> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Convert into a data.frame</span></span>
<span><span class="va">molten_mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html">longFormat</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span></span>
<span><span class="va">molten_mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">molten_mae</span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"path/to/mae.feather"</span></span>
<span><span class="fu">write_feather</span><span class="op">(</span><span class="va">molten_mae</span>,<span class="va">path</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And as a CSV file.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span><span class="op">(</span><span class="va">molten_mae</span>, <span class="st">"path/to/mae.csv"</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/06_packages.html" class="pagination-link" aria-label="Packages">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/10_manipulation.html" class="pagination-link" aria-label="Data Manipulation">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>