<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 5&nbsp; Taxonomic Information</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/14_alpha_diversity.html" rel="next">
<link href="../pages/12_quality_control.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/10_manipulation.html">Focus topics</a></li><li class="breadcrumb-item"><a href="../pages/11_taxonomic_information.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Microbiome Analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/microbiome/OMA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://microbiome.github.io/OMA/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#assigning-taxonomic-information." id="toc-assigning-taxonomic-information." class="nav-link active" data-scroll-target="#assigning-taxonomic-information."><span class="header-section-number">5.1</span> Assigning taxonomic information.</a>
  <ul class="collapse">
<li><a href="#dada2" id="toc-dada2" class="nav-link" data-scroll-target="#dada2"><span class="header-section-number">5.1.1</span> dada2</a></li>
  <li><a href="#decipher" id="toc-decipher" class="nav-link" data-scroll-target="#decipher"><span class="header-section-number">5.1.2</span> DECIPHER</a></li>
  </ul>
</li>
  <li>
<a href="#functions-to-access-taxonomic-information" id="toc-functions-to-access-taxonomic-information" class="nav-link" data-scroll-target="#functions-to-access-taxonomic-information"><span class="header-section-number">5.2</span> Functions to access taxonomic information</a>
  <ul class="collapse">
<li><a href="#sec-fly-tree" id="toc-sec-fly-tree" class="nav-link" data-scroll-target="#sec-fly-tree"><span class="header-section-number">5.2.1</span> Generate a hierarchy tree on the fly</a></li>
  </ul>
</li>
  <li>
<a href="#sec-data-agglomeration" id="toc-sec-data-agglomeration" class="nav-link" data-scroll-target="#sec-data-agglomeration"><span class="header-section-number">5.3</span> Data agglomeration</a>
  <ul class="collapse">
<li><a href="#sec-taxa-clustering" id="toc-sec-taxa-clustering" class="nav-link" data-scroll-target="#sec-taxa-clustering"><span class="header-section-number">5.3.1</span> Taxa clustering</a></li>
  </ul>
</li>
  <li>
<a href="#sec-assay-transform" id="toc-sec-assay-transform" class="nav-link" data-scroll-target="#sec-assay-transform"><span class="header-section-number">5.4</span> Data transformation</a>
  <ul class="collapse">
<li><a href="#transforming-the-data-in-practice" id="toc-transforming-the-data-in-practice" class="nav-link" data-scroll-target="#transforming-the-data-in-practice"><span class="header-section-number">5.4.1</span> Transforming the data in practice</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/10_manipulation.html">Focus topics</a></li><li class="breadcrumb-item"><a href="../pages/11_taxonomic_information.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-taxonomic-information" class="quarto-section-identifier"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Taxonomic information is a key part of analyzing microbiome data and without it, any type of data analysis probably will not make much sense. However, the degree of detail of taxonomic information differs depending on the dataset and annotation data used.</p>
<p>Therefore, the mia package expects a loose assembly of taxonomic information and assumes certain key aspects:</p>
<ul>
<li>Taxonomic information is given as character vectors or factors in the <code>rowData</code> of a <code>SummarizedExperiment</code> object.</li>
<li>The columns containing the taxonomic information must be named <code>domain</code>, <code>kingdom</code>, <code>phylum</code>, <code>class</code>, <code>order</code>, <code>family</code>, <code>genus</code>, <code>species</code> or with a capital first letter.</li>
<li>the columns must be given in the order shown above</li>
<li>column can be omited, but the order must remain</li>
</ul>
<p>In this chapter, we will refer to co-abundant groups as CAGs, which are clusters of taxa that co-vary across samples.</p>
<section id="assigning-taxonomic-information." class="level2 page-columns page-full" data-number="5.1"><h2 data-number="5.1" class="anchored" data-anchor-id="assigning-taxonomic-information.">
<span class="header-section-number">5.1</span> Assigning taxonomic information.</h2>
<p>There are a number of methods to assign taxonomic information. We like to give a short introduction about the methods available without ranking one over the other. This has to be your choice based on the result for the individual dataset.</p>
<section id="dada2" class="level3 page-columns page-full" data-number="5.1.1"><h3 data-number="5.1.1" class="anchored" data-anchor-id="dada2">
<span class="header-section-number">5.1.1</span> dada2</h3>
<p>The dada2 package <span class="citation" data-cites="Callahan2016dada2">(<a href="#ref-Callahan2016dada2" role="doc-biblioref">Callahan et al. 2016</a>)</span> implements the <code>assignTaxonomy</code> function, which takes as input the ASV sequences associated with each row of data and a training dataset. For more information visit the <a href="https://benjjneb.github.io/dada2/assign.html">dada2 homepage</a>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Callahan2016dada2" class="csl-entry" role="listitem">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han, Amy Jo A Johnson, and Susan P Holmes. 2016. <span>“DADA2: High-Resolution Sample Inference from Illumina Amplicon Data.”</span> <em>Nature Methods</em> 13: 581–83. <a href="https://doi.org/10.1038/nmeth.3869">https://doi.org/10.1038/nmeth.3869</a>.
</div></div></section><section id="decipher" class="level3 page-columns page-full" data-number="5.1.2"><h3 data-number="5.1.2" class="anchored" data-anchor-id="decipher">
<span class="header-section-number">5.1.2</span> DECIPHER</h3>
<p>The DECIPHER package <span class="citation" data-cites="R_DECIPHER">(<a href="#ref-R_DECIPHER" role="doc-biblioref">Wright 2020</a>)</span> implements the <code>IDTAXA</code> algorithm to assign either taxonomic information or function information. For <code>mia</code> only the first option is of interest for now and more information can be found on the <a href="http://www2.decipher.codes/Classification.html">DECIPHER website</a>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-R_DECIPHER" class="csl-entry" role="listitem">
Wright, Erik. 2020. <em>DECIPHER: Tools for Curating, Analyzing, and Manipulating Biological Sequences</em>.
</div></div></section></section><section id="functions-to-access-taxonomic-information" class="level2 page-columns page-full" data-number="5.2"><h2 data-number="5.2" class="anchored" data-anchor-id="functions-to-access-taxonomic-information">
<span class="header-section-number">5.2</span> Functions to access taxonomic information</h2>
<p><code>checkTaxonomy</code> checks whether the taxonomic information is usable for <code>mia</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">checkTaxonomy</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since the <code>rowData</code> can contain other data, <code>taxonomyRanks</code> will return the columns <code>mia</code> assumes to contain the taxonomic information.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This can then be used to subset the <code>rowData</code> to columns needed.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##  DataFrame with 19216 rows and 7 columns</span></span>
<span><span class="co">##             Kingdom        Phylum        Class        Order        Family</span></span>
<span><span class="co">##         &lt;character&gt;   &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">##  549322     Archaea Crenarchaeota Thermoprotei           NA            NA</span></span>
<span><span class="co">##  522457     Archaea Crenarchaeota Thermoprotei           NA            NA</span></span>
<span><span class="co">##  951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae</span></span>
<span><span class="co">##  244423     Archaea Crenarchaeota        Sd-NA           NA            NA</span></span>
<span><span class="co">##  586076     Archaea Crenarchaeota        Sd-NA           NA            NA</span></span>
<span><span class="co">##  ...            ...           ...          ...          ...           ...</span></span>
<span><span class="co">##  278222    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  463590    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  535321    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  200359    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##  271582    Bacteria           SR1           NA           NA            NA</span></span>
<span><span class="co">##               Genus                Species</span></span>
<span><span class="co">##         &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">##  549322          NA                     NA</span></span>
<span><span class="co">##  522457          NA                     NA</span></span>
<span><span class="co">##  951     Sulfolobus Sulfolobusacidocalda..</span></span>
<span><span class="co">##  244423          NA                     NA</span></span>
<span><span class="co">##  586076          NA                     NA</span></span>
<span><span class="co">##  ...            ...                    ...</span></span>
<span><span class="co">##  278222          NA                     NA</span></span>
<span><span class="co">##  463590          NA                     NA</span></span>
<span><span class="co">##  535321          NA                     NA</span></span>
<span><span class="co">##  200359          NA                     NA</span></span>
<span><span class="co">##  271582          NA                     NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>taxonomyRankEmpty</code> checks for empty values in the given <code>rank</code> and returns a logical vector of <code>length(x)</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRankEmpty</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Kingdom"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRankEmpty</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  FALSE  TRUE </span></span>
<span><span class="co">##   8008 11208</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRankEmpty</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Species"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  FALSE  TRUE </span></span>
<span><span class="co">##   1413 17803</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>getTaxonomyLabels</code> is a multi-purpose function, which turns taxonomic information into a character vector of <code>length(x)</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Class:Thermoprotei"               "Class:Thermoprotei_1"            </span></span>
<span><span class="co">##  [3] "Species:Sulfolobusacidocaldarius" "Class:Sd-NA"                     </span></span>
<span><span class="co">##  [5] "Class:Sd-NA_1"                    "Class:Sd-NA_2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, this will use the lowest non-empty information to construct a string with the following scheme <code>level:value</code>. If all levels are the same, this part is omitted, but can be added by setting <code>with_rank = TRUE</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phylum</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">Phylum</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1L</span>, <span class="va">is.na</span><span class="op">)</span><span class="op">)</span>, <span class="va">all</span>, <span class="fu"><a href="https://rdrr.io/r/base/logical.html">logical</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="va">tse</span><span class="op">[</span><span class="va">phylum</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Crenarchaeota"    "Crenarchaeota_1"  "Crenarchaeota_2" </span></span>
<span><span class="co">##  [4] "Actinobacteria"   "Actinobacteria_1" "Spirochaetes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="va">tse</span><span class="op">[</span><span class="va">phylum</span>,<span class="op">]</span>, with_rank <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Phylum:Crenarchaeota"    "Phylum:Crenarchaeota_1" </span></span>
<span><span class="co">##  [3] "Phylum:Crenarchaeota_2"  "Phylum:Actinobacteria"  </span></span>
<span><span class="co">##  [5] "Phylum:Actinobacteria_1" "Phylum:Spirochaetes"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default the return value of <code>getTaxonomyLabels</code> contains only unique elements by passing it through <code>make.unique</code>. This step can be omitted by setting <code>make_unique = FALSE</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">getTaxonomyLabels</a></span><span class="op">(</span><span class="va">tse</span><span class="op">[</span><span class="va">phylum</span>,<span class="op">]</span>, with_rank <span class="op">=</span> <span class="cn">TRUE</span>, make_unique <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Phylum:Crenarchaeota"  "Phylum:Crenarchaeota"  "Phylum:Crenarchaeota" </span></span>
<span><span class="co">##  [4] "Phylum:Actinobacteria" "Phylum:Actinobacteria" "Phylum:Spirochaetes"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To apply the loop resolving function <code>resolveLoop</code> from the <code>TreeSummarizedExperiment</code> package <span class="citation" data-cites="R_TreeSummarizedExperiment">(<a href="#ref-R_TreeSummarizedExperiment" role="doc-biblioref">Huang 2020</a>)</span> within <code>getTaxonomyLabels</code>, set <code>resolve_loops = TRUE</code>.</p>
<div class="no-row-height column-margin column-container"></div><p>The function <code>getUnique</code> gives a list of unique taxa for the specified taxonomic rank.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/summaries.html">getUnique</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Crenarchaeota"  "Euryarchaeota"  "Actinobacteria" "Spirochaetes"  </span></span>
<span><span class="co">##  [5] "MVP-15"         "Proteobacteria"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sec-fly-tree" class="level3 page-columns page-full" data-number="5.2.1"><h3 data-number="5.2.1" class="anchored" data-anchor-id="sec-fly-tree">
<span class="header-section-number">5.2.1</span> Generate a hierarchy tree on the fly</h3>
<p>A hierarchy tree is not an actual phylogenetic tree. A hierarchy tree shows mapping between the taxonomic levels in taxonomic rank table (included in rowData), rather than the detailed phylogenetic relations included in a phylogenetic tree</p>
<p>For instance, the TreeSummarizedExperiment dataset Tengeler2020 contains an actual phylogenetic tree.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Tengeler2020</span><span class="op">)</span></span>
<span><span class="va">tse2</span> <span class="op">&lt;-</span> <span class="va">Tengeler2020</span></span>
<span><span class="va">phylo_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse2</span><span class="op">)</span></span>
<span><span class="va">phylo_tree</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Phylogenetic tree with 151 tips and 149 internal nodes.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Tip labels:</span></span>
<span><span class="co">##    [Eubacterium]_coprostanoligenes_group, [Clostridium]_innocuum_group, [Clostridium]_innocuum_group_1, Dielma, Unidentified_Lachnospiraceae_12, Epulopiscium, ...</span></span>
<span><span class="co">##  Node labels:</span></span>
<span><span class="co">##    node_1, node_2, node_3, node_4, node_5, node_6, ...</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Unrooted; includes branch lengths.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To create a hierarchy tree, <code>getHierarchyTree</code> used the information and returns a <code>phylo</code> object. Duplicate information from the <code>rowData</code> is removed.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/hierarchy-tree.html">getHierarchyTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Phylogenetic tree with 1645 tips and 1089 internal nodes.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Tip labels:</span></span>
<span><span class="co">##    Species:Cenarchaeumsymbiosum, Species:pIVWA5, Species:CandidatusNitrososphaeragargensis, Species:SCA1145, Species:SCA1170, Species:Sulfolobusacidocaldarius, ...</span></span>
<span><span class="co">##  Node labels:</span></span>
<span><span class="co">##    root:ALL, Kingdom:Archaea, Phylum:Crenarchaeota, Class:C2, Class:Sd-NA, Class:Thaumarchaeota, ...</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Rooted; includes branch lengths.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/hierarchy-tree.html">addHierarchyTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 19216 26 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(19216): 549322 522457 ... 200359 271582</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (19216 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (1645 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The implementation is based on the <code>toTree</code> function from the <code>TreeSummarizedExperiment</code> package <span class="citation" data-cites="R_TreeSummarizedExperiment">(<a href="#ref-R_TreeSummarizedExperiment" role="doc-biblioref">Huang 2020</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-R_TreeSummarizedExperiment" class="csl-entry" role="listitem">
Huang, Ruizhu. 2020. <em>TreeSummarizedExperiment: A S4 Class for Data with Tree Structures</em>.
</div></div></section></section><section id="sec-data-agglomeration" class="level2" data-number="5.3"><h2 data-number="5.3" class="anchored" data-anchor-id="sec-data-agglomeration">
<span class="header-section-number">5.3</span> Data agglomeration</h2>
<p>One of the main applications of taxonomic information in regards to count data is to agglomerate count data on taxonomic levels and track the influence of changing conditions through these levels. For this <code>mia</code> contains the <code>mergeFeaturesByRank</code> function. The ideal location to store the agglomerated data is as an alternative experiment.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/deprecate.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Family"</span>,</span>
<span>                                           agglomerateTree <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 334 26 </span></span>
<span><span class="co">##  metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">##  assays(2): counts relabundance</span></span>
<span><span class="co">##  rownames(334): 125ds10 211ds20 ... vadinHA31 wb1_P06</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (334 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (334 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If multiple assays (counts and relabundance) exist, both will be agglomerated.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assayNames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] "counts"       "relabundance"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assayNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "counts"       "relabundance"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span>, <span class="st">"relabundance"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="co">##                CL3       CC1       SV1   M31Fcsw   M11Fcsw   M31Plmr</span></span>
<span><span class="co">##  125ds10 5.324e-05 0.000e+00 0.000e+00 0.000e+00 0.000e+00 0.000e+00</span></span>
<span><span class="co">##  211ds20 1.389e-05 3.699e-05 4.731e-04 0.000e+00 0.000e+00 5.564e-06</span></span>
<span><span class="co">##  5B-12   1.180e-04 2.272e-04 2.638e-04 0.000e+00 0.000e+00 0.000e+00</span></span>
<span><span class="co">##  A714017 3.472e-06 3.523e-06 1.434e-06 0.000e+00 2.890e-06 1.391e-06</span></span>
<span><span class="co">##  ACK-M1  5.902e-05 5.636e-05 2.151e-04 4.147e-05 4.671e-05 4.312e-05</span></span>
<span><span class="co">##            M11Plmr</span></span>
<span><span class="co">##  125ds10 0.000e+00</span></span>
<span><span class="co">##  211ds20 8.297e-05</span></span>
<span><span class="co">##  5B-12   0.000e+00</span></span>
<span><span class="co">##  A714017 9.219e-06</span></span>
<span><span class="co">##  ACK-M1  3.342e-04</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="co">##          CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr</span></span>
<span><span class="co">##  125ds10  46   0   0       0       0       0       0</span></span>
<span><span class="co">##  211ds20  12  42 330       0       0       4      36</span></span>
<span><span class="co">##  5B-12   102 258 184       0       0       0       0</span></span>
<span><span class="co">##  A714017   3   4   1       0       6       1       4</span></span>
<span><span class="co">##  ACK-M1   51  64 150      64      97      31     145</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>altExpNames</code> now consists of <code>Family</code> level data. This can be extended to use any taxonomic level listed in <code>mia::taxonomyRanks(tse)</code>.</p>
<p>Rare taxa can also be aggregated into a single group “Other” instead of filtering them out. A suitable function for this is <code>mergeFeaturesByPrevalence</code>. The number of rare taxa is higher on the species level, which causes the need for data agglomeration by prevalence.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Species_byPrevalence"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/deprecate.html">mergeFeaturesByPrevalence</a></span><span class="op">(</span><span class="va">tse</span>, </span>
<span>                                                               rank <span class="op">=</span> <span class="st">"Species"</span>, </span>
<span>                                                               other_label <span class="op">=</span> <span class="st">"Other"</span>, </span>
<span>                                                               prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, </span>
<span>                                                               detection <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">100</span>, </span>
<span>                                                               as_relative <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Species_byPrevalence"</span><span class="op">)</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 27 26 </span></span>
<span><span class="co">##  metadata(2): agglomerated_by_rank agglomerated_by_rank</span></span>
<span><span class="co">##  assays(2): counts relabundance</span></span>
<span><span class="co">##  rownames(27): Class:Thermoprotei Species:Actinobacillusporcinus ...</span></span>
<span><span class="co">##    Species:Veillonellaparvula Other</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Species_byPrevalence"</span><span class="op">)</span>, <span class="st">"relabundance"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">##                                         CL3       CC1       SV1</span></span>
<span><span class="co">##  Class:Thermoprotei               9.722e-01 9.757e-01 8.938e-01</span></span>
<span><span class="co">##  Species:Actinobacillusporcinus   1.389e-05 2.466e-05 1.147e-05</span></span>
<span><span class="co">##  Species:Alistipesputredinis      4.166e-05 3.963e-05 5.161e-05</span></span>
<span><span class="co">##  Species:Anaerococcushydrogenalis 7.407e-05 8.807e-05 6.595e-05</span></span>
<span><span class="co">##  Species:Bacteroidescaccae        1.007e-04 7.926e-05 1.104e-04</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Saving the tse for later</span></span>
<span><span class="va">tseGlobalPatterns</span> <span class="op">&lt;-</span> <span class="va">tse</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sec-taxa-clustering" class="level3" data-number="5.3.1"><h3 data-number="5.3.1" class="anchored" data-anchor-id="sec-taxa-clustering">
<span class="header-section-number">5.3.1</span> Taxa clustering</h3>
<p>Another way to agglomerate the data is to cluster the taxa. To do so, we usually start by doing a compositionality aware transformation such as CLR, followed by the application of a standard clustering method.</p>
<p>Here is an example that does a CLR transformation followed by the hierarchical clustering algorithm.</p>
<p>First, we import the library <code>bluster</code> that simplifies the clustering.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we do the CLR transform followed by the clustering. We will cluster with two different distances: the euclidean distance and the kendall distance.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"peerj13075"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">peerj13075</span></span>
<span></span>
<span><span class="co"># The result of the CLR transform is stored in the assay clr</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"clr"</span>, method <span class="op">=</span> <span class="st">"z"</span>, </span>
<span>                      MARGIN <span class="op">=</span> <span class="st">"features"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cluster (with euclidean distance) on the features of the z assay</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addCluster.html">addCluster</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>               assay.type <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>               clust.col <span class="op">=</span> <span class="st">"hclustEuclidean"</span>,</span>
<span>           MARGIN <span class="op">=</span> <span class="st">"features"</span>,</span>
<span>               <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/HclustParam-class.html">HclustParam</a></span><span class="op">(</span>dist.fun <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Declare the Kendall dissimilarity computation function</span></span>
<span><span class="va">kendall_dissimilarity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"kendall"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Cluster (with Kendall dissimilarity) on the features of the z assay</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addCluster.html">addCluster</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>               assay.type <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>               clust.col <span class="op">=</span> <span class="st">"hclustKendall"</span>,</span>
<span>               MARGIN <span class="op">=</span> <span class="st">"features"</span>,            </span>
<span>               <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/HclustParam-class.html">HclustParam</a></span><span class="op">(</span>dist.fun <span class="op">=</span> <span class="va">kendall_dissimilarity</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let us store the resulting cluster indices in the <code>rowData</code> column specified with the <code>clust.col</code> parameter.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Checking the clusters</span></span>
<span><span class="va">clusters_euclidean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">hclustEuclidean</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">clusters_euclidean</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">##   OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 </span></span>
<span><span class="co">##      1     2     1     1     1     1     3     4     3     2 </span></span>
<span><span class="co">##  Levels: 1 2 3 4 5</span></span>
<span></span>
<span><span class="va">clusters_kendall</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">hclustKendall</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">clusters_kendall</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">##   OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 </span></span>
<span><span class="co">##      1     2     1     3     3     1     3     1     1     3 </span></span>
<span><span class="co">##  Levels: 1 2 3 4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To better visualize the results and the distribution of the clusters, we can plot the histogram of the clusters.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span> <span class="co"># TO arrange several plots as a grid</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">clusters_euclidean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CAG size distribution (Euclidean distance)"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Clusters"</span>, y <span class="op">=</span> <span class="st">"Feature count (n)"</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">clusters_kendall</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CAG size distribution (1 - tau)"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Clusters"</span>, y <span class="op">=</span> <span class="st">"Feature count (n)"</span><span class="op">)</span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="11_taxonomic_information_files/figure-html/taxa_clustering_histogram-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It’s also possible to merge the rows by cluster.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Aggregate clusters as a sum of each cluster values</span></span>
<span><span class="va">tse_merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/deprecate.html">mergeFeatures</a></span><span class="op">(</span><span class="va">tse</span>, <span class="va">clusters_euclidean</span><span class="op">)</span></span>
<span><span class="va">tse_merged</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 5 58 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(3): counts clr z</span></span>
<span><span class="co">##  rownames(5): 1 2 3 4 5</span></span>
<span><span class="co">##  rowData names(8): kingdom phylum ... hclustEuclidean hclustKendall</span></span>
<span><span class="co">##  colnames(58): ID1 ID2 ... ID57 ID58</span></span>
<span><span class="co">##  colData names(5): Sample Geographical_location Gender Age Diet</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can note that it worked as planned since there were 5 clusters and there are now 5 rows.</p>
</section></section><section id="sec-assay-transform" class="level2 page-columns page-full" data-number="5.4"><h2 data-number="5.4" class="anchored" data-anchor-id="sec-assay-transform">
<span class="header-section-number">5.4</span> Data transformation</h2>
<p>Data transformations are common in (microbial) ecology <span class="citation" data-cites="Legendre2001">(<a href="#ref-Legendre2001" role="doc-biblioref">Legendre and Gallagher 2001</a>)</span> and used to improve compatibility with assumptions related to specific statistical methods, mitigate biases, enhance the comparability of samples or features, or to obtain more interpretable values.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Legendre2001" class="csl-entry" role="listitem">
Legendre, Pierre, and Eugene D. Gallagher. 2001. <span>“Ecologically Meaningful Transformations for Ordination of Species Data.”</span> <em>Oecologia</em> 129 (2): 271–80. <a href="https://doi.org/10.1007/s004420100716">https://doi.org/10.1007/s004420100716</a>.
</div></div><p>Examples include the logarithmic transformation, calculation of relative abundances (percentages), and compositionality-aware transformations such as the centered log-ratio transformation (clr).</p>
<p>Let us summarize some commonly used transformations in microbiome data science; further details and benchmarkings available in the references.</p>
<ul>
<li>
<p>‘relabundance’ relative transformation; also known as total sum scaling (TSS) and compositional transformation. This converts counts into percentages (at the scale [0, 1]) that sum up to</p>
<ol type="1">
<li>Much of the currently available taxonomic abundance data from high-throughput assays (16S, metagenomic sequencing) is compositional by nature, even if the data is provided as counts <span class="citation" data-cites="Gloor2017">(<a href="#ref-Gloor2017" role="doc-biblioref">Gloor et al. 2017</a>)</span>.</li>
</ol>
</li>
<li><p>‘clr’ Centered log ratio transformation <span class="citation" data-cites="Aitchison1986">(<a href="#ref-Aitchison1986" role="doc-biblioref">Aitchison 1986</a>)</span> is used to reduce data skewness and compositionality bias in relative abundances, while bringing the data to the logarithmic scale. This transformation is frequently applied in microbial ecology <span class="citation" data-cites="Gloor2017">(<a href="#ref-Gloor2017" role="doc-biblioref">Gloor et al. 2017</a>)</span>. However, this transformation only applies to positive values. Usual solution is to add pseudocount, which adds another type of bias in the data. The robust clr transformation (‘rclr’) aims to circumvent the need to add a pseudocount. While the resulting values from these transformations are difficult interpret directly, this transformation may enhance comparability of relative differences between samples. It is part of a broader Aitchison family of transformations; the additive log ratio transformation (`alr’) is also available. The robust clr (“rclr”) is similar to regular clr (see above) but allows data with zeroes and avoids the need to add pseudocount <span class="citation" data-cites="Keshavan2010">Martino et al. (<a href="#ref-Martino2019" role="doc-biblioref">2019</a>)</span>.</p></li>
<li><p>‘pa’ presence/absence transformation ignores abundances and only indicates whether the given feature is detected above the given threshold (default: 0). This simple transformation is relatively widely used in ecological research. It has shown good performance in microbiome-based classification performance <span class="citation" data-cites="Giliberti2022">(<a href="#ref-Giliberti2022" role="doc-biblioref">Giliberti et al. 2022, Karwowska2024</a>)</span>.</p></li>
<li><p>‘z’ Z transformation scales data to zero mean and unit variance; this us used to bring features (or samples) to more comparable levels in terms of mean and scale of the values. This can enhance visualization and interpretation of the data</p></li>
<li><p>‘log’, ‘log2’, ‘log10’ Logarithmic transformations; used e.g.&nbsp;to reduce data skewness; with compositional data the <code>clr</code> (or <code>rclr</code>) transformation is often preferred.</p></li>
<li><p>‘hellinger’ Hellinger transformation equals to the square root of relative abundances. This ecological transformation can be useful if we are interested in changes in relative abundances.</p></li>
<li><p>‘rank’ Rank transformation replaces each value by its rank. Also see ‘rrank’ (relative rank transformation). This has use for instance in non-parametric statistics.</p></li>
<li><p>Other available transformations include Chi square (‘chi.square’), Frequency transformation (‘frequency’), and Make margin sum of squares equal to one (‘normalize’)</p></li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-Gloor2017" class="csl-entry" role="listitem">
Gloor, GB, JM Macklaim, V Pawlowsky-Glahn, and JJ Egozcue. 2017. <span>“Microbiome Datasets Are Compositional: And This Is Not Optional.”</span> <em>Frontiers in Microbiology</em> 8. <a href="https://doi.org/10.3389/fmicb.2017.02224">https://doi.org/10.3389/fmicb.2017.02224</a>.
</div><div id="ref-Aitchison1986" class="csl-entry" role="listitem">
Aitchison, J. 1986. <em>The Statistical Analysis of Compositional Data</em>. London, UK: Chapman &amp; Hall.
</div><div id="ref-Martino2019" class="csl-entry" role="listitem">
Martino, C, J. T. Morton, C. A. Marotz, L. R. Thompson, A Tripathi, R Knight, and K Zengler. 2019. <span>“<a href="">A Novel Sparse Compositional Technique Reveals Microbial Perturbations</a>.”</span> <em>mSystems</em> 4.
</div></div><section id="transforming-the-data-in-practice" class="level3 page-columns page-full" data-number="5.4.1"><h3 data-number="5.4.1" class="anchored" data-anchor-id="transforming-the-data-in-practice">
<span class="header-section-number">5.4.1</span> Transforming the data in practice</h3>
<p>Transformations on abundance assays can be performed with <code><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">mia::transformAssay()</a></code>, keeping both the original and the transformed assay(s). The transformed abundance assay is then stored back to the ‘assays’ slot in the data object. The function applies sample-wise or column-wise transformation when MARGIN = ‘samples’, feature-wise or row-wise transformation when MARGIN = ‘features’.</p>
<p>A complete list of available transformations and parameters, is available in the function <a href="https://microbiome.github.io/mia/reference/transformAssay.html">help</a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load example data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span>
<span></span>
<span><span class="co"># Transform "counts" assay to relative abundances ("relabundance"), with pseudocount 1 </span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"relabundance"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transform relative abundance assay ("relabundance") to "clr", using pseudocount if necessary;</span></span>
<span><span class="co"># name the resulting assay to "clr" </span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"relabundance"</span>, method <span class="op">=</span> <span class="st">"clr"</span>, </span>
<span>                      pseudocount <span class="op">=</span> <span class="cn">TRUE</span>, name <span class="op">=</span> <span class="st">"clr"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get the values in the resulting assay, and view some of the first entries of it with the <code>head</code> command.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"clr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##             CL3     CC1    SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr</span></span>
<span><span class="co">##  549322 -0.8644 -0.9978 -0.683 -0.2439 -0.2096  -0.323 -0.4427 -0.2559</span></span>
<span><span class="co">##  522457 -0.8644 -0.9978 -0.683 -0.2439 -0.2096  -0.323 -0.4427 -0.2559</span></span>
<span><span class="co">##  951    -0.8644 -0.9978 -0.683 -0.2439 -0.2096  -0.323  0.1670 -0.2559</span></span>
<span><span class="co">##  244423 -0.8644 -0.9978 -0.683 -0.2439 -0.2096  -0.323 -0.4427 -0.2559</span></span>
<span><span class="co">##  586076 -0.8644 -0.9978 -0.683 -0.2439 -0.2096  -0.323 -0.4427 -0.2559</span></span>
<span><span class="co">##  246140 -0.8644 -0.9978 -0.683 -0.2439 -0.2096  -0.323 -0.4427 -0.2559</span></span>
<span><span class="co">##         M31Tong M11Tong LMEpi24M SLEpi20M  AQC1cm   AQC4cm  AQC7cm     NP2</span></span>
<span><span class="co">##  549322 -0.1956 -0.1537  -0.2802   0.2318  2.3342  3.27787  3.6714  0.3713</span></span>
<span><span class="co">##  522457 -0.1956 -0.1537  -0.2802  -0.2737 -0.6105  0.03919  0.8349 -0.2243</span></span>
<span><span class="co">##  951    -0.1956 -0.1537  -0.2802  -0.2737 -0.6105 -0.65395 -0.6651 -0.2243</span></span>
<span><span class="co">##  244423 -0.1956 -0.1537  -0.2802  -0.2737 -0.6105  1.83095  2.2157 -0.2243</span></span>
<span><span class="co">##  586076 -0.1956 -0.1537  -0.2802  -0.2737 -0.6105  0.03919 -0.2075 -0.2243</span></span>
<span><span class="co">##  246140 -0.1956 -0.1537  -0.2802  -0.2737 -0.6105 -0.24849  0.3432 -0.2243</span></span>
<span><span class="co">##             NP3     NP5 TRRsed1 TRRsed2 TRRsed3    TS28    TS29  Even1</span></span>
<span><span class="co">##  549322 -0.3632 -0.2974 -0.2641 -0.4479 -0.4185 -0.2371 -0.2189 -0.289</span></span>
<span><span class="co">##  522457 -0.3632 -0.2974 -0.2641 -0.4479 -0.4185 -0.2371 -0.2189 -0.289</span></span>
<span><span class="co">##  951    -0.3632 -0.2974 -0.2641 -0.4479 -0.4185 -0.2371 -0.2189 -0.289</span></span>
<span><span class="co">##  244423 -0.3632 -0.2974 -0.2641 -0.4479 -0.4185 -0.2371 -0.2189 -0.289</span></span>
<span><span class="co">##  586076 -0.3632 -0.2974 -0.2641 -0.4479 -0.4185 -0.2371 -0.2189 -0.289</span></span>
<span><span class="co">##  246140 -0.3632 -0.2974 -0.2641 -0.4479 -0.4185 -0.2371 -0.2189 -0.289</span></span>
<span><span class="co">##           Even2   Even3</span></span>
<span><span class="co">##  549322 -0.2174 -0.2031</span></span>
<span><span class="co">##  522457 -0.2174 -0.2031</span></span>
<span><span class="co">##  951    -0.2174 -0.2031</span></span>
<span><span class="co">##  244423 -0.2174 -0.2031</span></span>
<span><span class="co">##  586076 -0.2174 -0.2031</span></span>
<span><span class="co">##  246140 -0.2174 -0.2031</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In ‘pa’ transformation, abundance table is converted to presence/absence table that ignores abundances and only indicates whether the given feature is detected. This simple transformation is relatively widely used in ecological research. It has shown good performance in microbiome-based classification performance <span class="citation" data-cites="Giliberti2022">(<a href="#ref-Giliberti2022" role="doc-biblioref">Giliberti et al. 2022, Karwowska2024</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Giliberti2022" class="csl-entry" role="listitem">
Giliberti, R, S Cavaliere, IE Mauriello, D Ercolini, and E Pasolli. 2022. <span>“Host Phenotype Classification from Human Microbiome Data Is Mainly Driven by the Presence of Microbial Taxa.”</span> <em>PLoS Comput Biol.</em> 18 (4): e1010066. <a href="https://doi.org/10.1371/journal.pcbi.1010066">https://doi.org/10.1371/journal.pcbi.1010066</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Here, `assay.type` is not explicitly specified.</span></span>
<span><span class="co"># Then The function uses the "counts" assay for the transformation.</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"pa"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pa"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##         CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong</span></span>
<span><span class="co">##  549322   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  522457   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  951      0   0   0       0       0       0       1       0       0       0</span></span>
<span><span class="co">##  244423   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  586076   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  246140   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##         LMEpi24M SLEpi20M AQC1cm AQC4cm AQC7cm NP2 NP3 NP5 TRRsed1 TRRsed2</span></span>
<span><span class="co">##  549322        0        1      1      1      1   1   0   0       0       0</span></span>
<span><span class="co">##  522457        0        0      0      1      1   0   0   0       0       0</span></span>
<span><span class="co">##  951           0        0      0      0      0   0   0   0       0       0</span></span>
<span><span class="co">##  244423        0        0      0      1      1   0   0   0       0       0</span></span>
<span><span class="co">##  586076        0        0      0      1      1   0   0   0       0       0</span></span>
<span><span class="co">##  246140        0        0      0      1      1   0   0   0       0       0</span></span>
<span><span class="co">##         TRRsed3 TS28 TS29 Even1 Even2 Even3</span></span>
<span><span class="co">##  549322       0    0    0     0     0     0</span></span>
<span><span class="co">##  522457       0    0    0     0     0     0</span></span>
<span><span class="co">##  951          0    0    0     0     0     0</span></span>
<span><span class="co">##  244423       0    0    0     0     0     0</span></span>
<span><span class="co">##  586076       0    0    0     0     0     0</span></span>
<span><span class="co">##  246140       0    0    0     0     0     0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can now view the entire list of abundance assays in your data object with:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  List of length 4</span></span>
<span><span class="co">##  names(4): counts relabundance clr pa</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/12_quality_control.html" class="pagination-link" aria-label="Exploration and Quality Control">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/14_alpha_diversity.html" class="pagination-link" aria-label="Community Diversity">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>