<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 4&nbsp; Exploration and Quality Control</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/11_taxonomic_information.html" rel="next">
<link href="../pages/10_manipulation.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/10_manipulation.html">Focus topics</a></li><li class="breadcrumb-item"><a href="../pages/12_quality_control.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Microbiome Analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/microbiome/OMA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://microbiome.github.io/OMA/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#abundance" id="toc-abundance" class="nav-link active" data-scroll-target="#abundance"><span class="header-section-number">4.1</span> Abundance</a></li>
  <li>
<a href="#prevalence" id="toc-prevalence" class="nav-link" data-scroll-target="#prevalence"><span class="header-section-number">4.2</span> Prevalence</a>
  <ul class="collapse">
<li><a href="#prevalence-analysis" id="toc-prevalence-analysis" class="nav-link" data-scroll-target="#prevalence-analysis"><span class="header-section-number">4.2.1</span> Prevalence analysis</a></li>
  <li><a href="#rare-taxa" id="toc-rare-taxa" class="nav-link" data-scroll-target="#rare-taxa"><span class="header-section-number">4.2.2</span> Rare taxa</a></li>
  <li><a href="#plotting-prevalence" id="toc-plotting-prevalence" class="nav-link" data-scroll-target="#plotting-prevalence"><span class="header-section-number">4.2.3</span> Plotting prevalence</a></li>
  </ul>
</li>
  <li>
<a href="#sec-qc" id="toc-sec-qc" class="nav-link" data-scroll-target="#sec-qc"><span class="header-section-number">4.3</span> Quality control</a>
  <ul class="collapse">
<li><a href="#top-taxa" id="toc-top-taxa" class="nav-link" data-scroll-target="#top-taxa"><span class="header-section-number">4.3.1</span> Top taxa</a></li>
  <li><a href="#library-size-read-count" id="toc-library-size-read-count" class="nav-link" data-scroll-target="#library-size-read-count"><span class="header-section-number">4.3.2</span> Library size / read count</a></li>
  <li><a href="#contaminant-sequences" id="toc-contaminant-sequences" class="nav-link" data-scroll-target="#contaminant-sequences"><span class="header-section-number">4.3.3</span> Contaminant sequences</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-quality-control" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>This chapter focuses on the quality control and exploration of microbiome data and establishes commonly used descriptive summaries. Familiarizing with the peculiarities of a given dataset is the essential basis for any data analysis and model building.</p>
<p>The dataset should not suffer from severe technical biases, and you should at least be aware of potential challenges, such as outliers, biases, unexpected patterns and so forth. Standard summaries and visualizations can help, and the rest comes with experience. The exploration and quality control can be iterative processes.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-1_adc6b6318c94101e17c6bf9755e13616">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="abundance" class="level2 page-columns page-full" data-number="4.1"><h2 data-number="4.1" class="anchored" data-anchor-id="abundance">
<span class="header-section-number">4.1</span> Abundance</h2>
<p>Abundance visualization is an important data exploration approach. <code>miaViz</code> offers the function <code>plotAbundanceDensity</code> to plot the most abundant taxa with several options.</p>
<p>Next, a few demonstrations are shown, using the <span class="citation" data-cites="Lahti2014">(<a href="#ref-Lahti2014" role="doc-biblioref">Lahti et al. 2014</a>)</span> dataset. A Jitter plot based on relative abundance data, similar to the one presented at <span class="citation" data-cites="Salosensaari2021">(<a href="#ref-Salosensaari2021" role="doc-biblioref">Salosensaari et al. 2021</a>)</span> supplementary figure 1, could be visualized as follows:</p>
<div class="no-row-height column-margin column-container"><div id="ref-Lahti2014" class="csl-entry" role="listitem">
Lahti, L, JSalojärvi, A Salonen, M Scheffer, and WM de Vos. 2014. <span>“Tipping Elements in the Human Intestinal Ecosystem.”</span> <em>Nature Communications</em> 2014: 1–10. https://doi.org/<a href="https://doi.org/10.1038/ncomms5344">https://doi.org/10.1038/ncomms5344</a>.
</div><div id="ref-Salosensaari2021" class="csl-entry" role="listitem">
Salosensaari, Aaro, Ville Laitinen, Aki Havulinna, Guillaume Méric, Susan Cheng, Markus Perola, Liisa Valsta, et al. 2021. <span>“Taxonomic Signatures of Cause-Specific Mortality Risk in Human Gut Microbiome.”</span> <em>Nature Communications</em> 12: 1–8. <a href="https://www.nature.com/articles/s41467-021-22962-y">https://www.nature.com/articles/s41467-021-22962-y</a>.
</div></div><div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-2_d9f68f90aa0ecc5991aa0f7d5088cc92">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load example data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/miaTime">miaTime</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">miaViz</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">hitchip1006</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">hitchip1006</span></span>
<span></span>
<span><span class="co"># Add relative abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, MARGIN <span class="op">=</span> <span class="st">"samples"</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use argument names</span></span>
<span><span class="co"># assay.type / assay.type / assay.type</span></span>
<span><span class="co"># depending on the mia package version</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotAbundanceDensity.html">plotAbundanceDensity</a></span><span class="op">(</span><span class="va">tse</span>, layout <span class="op">=</span> <span class="st">"jitter"</span>, assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>                     n <span class="op">=</span> <span class="fl">40</span>, point_size<span class="op">=</span><span class="fl">1</span>, point_shape<span class="op">=</span><span class="fl">19</span>, point_alpha<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span>label<span class="op">=</span><span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html">percent</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The relative abundance values for the top-5 taxonomic features can be visualized as a density plot over a log scaled axis, with “nationality” indicated by colors:</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-3_d10de44955758a21248d49a3fc1d7110">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotAbundanceDensity.html">plotAbundanceDensity</a></span><span class="op">(</span><span class="va">tse</span>, layout <span class="op">=</span> <span class="st">"density"</span>, assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>                     n <span class="op">=</span> <span class="fl">5</span>, colour_by<span class="op">=</span><span class="st">"nationality"</span>, point_alpha<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="prevalence" class="level2" data-number="4.2"><h2 data-number="4.2" class="anchored" data-anchor-id="prevalence">
<span class="header-section-number">4.2</span> Prevalence</h2>
<p>Prevalence quantifies the frequency of samples where certain microbes were detected (above a given detection threshold). The prevalence can be given as sample size (N) or percentage (unit interval).</p>
<p>Investigating prevalence allows you either to focus on changes which pertain to the majority of the samples, or identify rare microbes, which may be <em>conditionally abundant</em> in a small number of samples.</p>
<p>The population prevalence (frequency) at a 1% relative abundance threshold (<code>detection = 1/100</code> and <code>as_relative = TRUE</code>), can look like this.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/exploration-prevalence_18d776119772febd38d220e9568a0dd1">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="va">tse</span>, detection <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">100</span>, sort <span class="op">=</span> <span class="cn">TRUE</span>, as_relative <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  Faecalibacterium prausnitzii et rel.           Ruminococcus obeum et rel. </span></span>
<span><span class="co">##                                0.9522                               0.9140 </span></span>
<span><span class="co">##    Oscillospira guillermondii et rel.        Clostridium symbiosum et rel. </span></span>
<span><span class="co">##                                0.8801                               0.8714 </span></span>
<span><span class="co">##      Subdoligranulum variable at rel.     Clostridium orbiscindens et rel. </span></span>
<span><span class="co">##                                0.8358                               0.8315</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function arguments <code>detection</code> and <code>as_relative</code> can also be used to access, how many samples do pass a threshold for raw counts. Here, the population prevalence (frequency) at the absolute abundance threshold (<code>as_relative = FALSE</code>) at read count 1 (<code>detection = 1</code>) is accessed.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/concepts_prevalence2_9c48f0601798c5f7a1c0f28160532daa">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="va">tse</span>, detection <span class="op">=</span> <span class="fl">1</span>, sort <span class="op">=</span> <span class="cn">TRUE</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                   as_relative <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##             Uncultured Mollicutes      Uncultured Clostridiales II </span></span>
<span><span class="co">##                                 1                                1 </span></span>
<span><span class="co">##        Uncultured Clostridiales I               Tannerella et rel. </span></span>
<span><span class="co">##                                 1                                1 </span></span>
<span><span class="co">##    Sutterella wadsworthia et rel. Subdoligranulum variable at rel. </span></span>
<span><span class="co">##                                 1                                1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If the output should be used for subsetting or storing the data in the <code>rowData</code>, set <code>sort = FALSE</code>.</p>
<section id="prevalence-analysis" class="level3" data-number="4.2.1"><h3 data-number="4.2.1" class="anchored" data-anchor-id="prevalence-analysis">
<span class="header-section-number">4.2.1</span> Prevalence analysis</h3>
<p>To investigate microbiome prevalence at a selected taxonomic level, two approaches are available.</p>
<p>First the data can be agglomerated to the taxonomic level and <code>getPrevalence</code> applied on the resulting object.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-4_f428f1edcce0054a23d1096c743bdd6f">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Agglomerate taxa abundances to Phylum level, and add the new table</span></span>
<span><span class="co"># to the altExp slot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Phylum"</span><span class="op">)</span></span>
<span><span class="co"># Check prevalence for the Phylum abundance table from the altExp slot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span>, detection <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">100</span>, sort <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   assay.type <span class="op">=</span> <span class="st">"counts"</span>, as_relative <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##       Firmicutes   Bacteroidetes  Actinobacteria  Proteobacteria </span></span>
<span><span class="co">##        1.0000000       0.9852302       0.4821894       0.2988705 </span></span>
<span><span class="co">##  Verrucomicrobia   Cyanobacteria </span></span>
<span><span class="co">##        0.1277150       0.0008688</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, the <code>rank</code> argument could be set to perform the agglomeration on the fly.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-5_af7d8b21f0595935d29a2713ce079de0">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span>, detection <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">100</span>, sort <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   assay.type <span class="op">=</span> <span class="st">"counts"</span>, as_relative <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##       Firmicutes   Bacteroidetes  Actinobacteria  Proteobacteria </span></span>
<span><span class="co">##        1.0000000       0.9852302       0.4821894       0.2988705 </span></span>
<span><span class="co">##  Verrucomicrobia   Cyanobacteria </span></span>
<span><span class="co">##        0.1277150       0.0008688</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that, by default, <code>na.rm = TRUE</code> is used for agglomeration in <code>getPrevalence</code>, whereas the default for <code>mergeFeaturesByRank</code> is <code>FALSE</code> to prevent accidental data loss.</p>
<p>If you only need the names of the prevalent taxa, <code>getPrevalentFeatures</code> is available. This returns the taxa that exceed the given prevalence and detection thresholds.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/core-members_fbae80dbbb91472562f999064eed80c4">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalentFeatures</a></span><span class="op">(</span><span class="va">tse</span>, detection <span class="op">=</span> <span class="fl">0</span>, prevalence <span class="op">=</span> <span class="fl">50</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">prev</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalentFeatures</a></span><span class="op">(</span><span class="va">tse</span>, detection <span class="op">=</span> <span class="fl">0</span>, prevalence <span class="op">=</span> <span class="fl">50</span><span class="op">/</span><span class="fl">100</span>,</span>
<span>                         rank <span class="op">=</span> <span class="st">"Phylum"</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">prev</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that the <code>detection</code> and <code>prevalence</code> thresholds are not the same, since <code>detection</code> can be applied to relative counts or absolute counts depending on whether <code>as_relative</code> is set <code>TRUE</code> or <code>FALSE</code></p>
<p>The function ‘getPrevalentAbundance’ can be used to check the total relative abundance of the prevalent taxa (between 0 and 1).</p>
</section><section id="rare-taxa" class="level3" data-number="4.2.2"><h3 data-number="4.2.2" class="anchored" data-anchor-id="rare-taxa">
<span class="header-section-number">4.2.2</span> Rare taxa</h3>
<p>Related functions are available for the analysis of rare taxa (<code>rareMembers</code>; <code>rareAbundance</code>; <code>lowAbundance</code>, <code>getRareFeatures</code>, <code>subsetByRareFeatures</code>).</p>
</section><section id="plotting-prevalence" class="level3" data-number="4.2.3"><h3 data-number="4.2.3" class="anchored" data-anchor-id="plotting-prevalence">
<span class="header-section-number">4.2.3</span> Plotting prevalence</h3>
<p>To plot the prevalence, add the prevalence of each taxon to <code>rowData</code>. Here, we are analysing the Phylum level abundances, which are stored in the <code>altExp</code> slot.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-6_f7955725543217ee2d3b6a09bbe8dac4">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">prevalence</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span>, detection <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">100</span>, sort <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                  assay.type <span class="op">=</span> <span class="st">"counts"</span>, as_relative <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The prevalences can then be plotted using the plotting functions from the <code>scater</code> package.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-7_8dde1f75f6550bf61e17cc018b5e400a">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html">plotRowData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span>, <span class="st">"prevalence"</span>, colour_by <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The prevalence can also be visualized on the taxonomic tree with the <code>miaViz</code> package.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-8_667a834918f7c843390eadd1c89789cf">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExps</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/splitByRanks.html">splitByRanks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExps</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExps</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span>,</span>
<span>          <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">prevalence</span> <span class="op">&lt;-</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">getPrevalence</a></span><span class="op">(</span><span class="va">y</span>, detection <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">100</span>, sort <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                assay.type <span class="op">=</span> <span class="st">"counts"</span>, as_relative <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>              <span class="va">y</span></span>
<span>          <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">top_phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/summaries.html">getTopFeatures</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span>,</span>
<span>                        method<span class="op">=</span><span class="st">"prevalence"</span>,</span>
<span>                        top<span class="op">=</span><span class="fl">5L</span>,</span>
<span>                        assay.type<span class="op">=</span><span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">top_phyla_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/summaries.html">getTopFeatures</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"Phylum"</span><span class="op">)</span>,</span>
<span>                             method<span class="op">=</span><span class="st">"mean"</span>,</span>
<span>                             top<span class="op">=</span><span class="fl">5L</span>,</span>
<span>                             assay.type<span class="op">=</span><span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/splitByRanks.html">unsplitByRanks</a></span><span class="op">(</span><span class="va">tse</span>, ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">addTaxonomyTree</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After some preparation, the data is assembled and can be plotted with <code>plotRowTree</code>.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/plot-prev-prev_f23d98d8bbe9db7b0f693af69292013a">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">miaViz</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotTree.html">plotRowTree</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">Phylum</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">top_phyla</span>,<span class="op">]</span>,</span>
<span>            edge_colour_by <span class="op">=</span> <span class="st">"Phylum"</span>,</span>
<span>            tip_colour_by <span class="op">=</span> <span class="st">"prevalence"</span>,</span>
<span>            node_colour_by <span class="op">=</span> <span class="st">"prevalence"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/plot-prev-prev-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Prevalence of top phyla as judged by prevalence</figcaption></figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/plot-prev-mean_f0f8664e35b7a04cf3e76a8a664df79c">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotTree.html">plotRowTree</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">Phylum</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">top_phyla_mean</span>,<span class="op">]</span>,</span>
<span>            edge_colour_by <span class="op">=</span> <span class="st">"Phylum"</span>,</span>
<span>            tip_colour_by <span class="op">=</span> <span class="st">"prevalence"</span>,</span>
<span>            node_colour_by <span class="op">=</span> <span class="st">"prevalence"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/plot-prev-mean-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Prevalence of top phyla as judged by mean abundance</figcaption></figure>
</div>
</div>
</div>
</section></section><section id="sec-qc" class="level2 page-columns page-full" data-number="4.3"><h2 data-number="4.3" class="anchored" data-anchor-id="sec-qc">
<span class="header-section-number">4.3</span> Quality control</h2>
<p>Next, let us load the <code>GlobalPatterns</code> dataset to illustrate standard microbiome data summaries.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/unnamed-chunk-9_32bc654252d9ef19aac51cb8e495bf0d">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="top-taxa" class="level3" data-number="4.3.1"><h3 data-number="4.3.1" class="anchored" data-anchor-id="top-taxa">
<span class="header-section-number">4.3.1</span> Top taxa</h3>
<p>The <code>getTopFeatures</code> identifies top taxa in the data.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/top-feature-taxo_e1e1f4e88ff213d9396ce887eb96c3ce">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Pick the top taxa</span></span>
<span><span class="va">top_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/summaries.html">getTopFeatures</a></span><span class="op">(</span><span class="va">tse</span>, method<span class="op">=</span><span class="st">"median"</span>, top<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the information for these</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span><span class="va">top_features</span>, <span class="fu"><a href="https://rdrr.io/pkg/mia/man/taxonomy-methods.html">taxonomyRanks</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##  DataFrame with 10 rows and 7 columns</span></span>
<span><span class="co">##             Kingdom         Phylum               Class             Order</span></span>
<span><span class="co">##         &lt;character&gt;    &lt;character&gt;         &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">##  549656    Bacteria  Cyanobacteria         Chloroplast     Stramenopiles</span></span>
<span><span class="co">##  331820    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales</span></span>
<span><span class="co">##  317182    Bacteria  Cyanobacteria         Chloroplast     Stramenopiles</span></span>
<span><span class="co">##  94166     Bacteria Proteobacteria Gammaproteobacteria    Pasteurellales</span></span>
<span><span class="co">##  279599    Bacteria  Cyanobacteria    Nostocophycideae        Nostocales</span></span>
<span><span class="co">##  158660    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales</span></span>
<span><span class="co">##  329744    Bacteria Actinobacteria      Actinobacteria   Actinomycetales</span></span>
<span><span class="co">##  326977    Bacteria Actinobacteria      Actinobacteria Bifidobacteriales</span></span>
<span><span class="co">##  248140    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales</span></span>
<span><span class="co">##  550960    Bacteria Proteobacteria Gammaproteobacteria Enterobacteriales</span></span>
<span><span class="co">##                     Family           Genus                Species</span></span>
<span><span class="co">##                &lt;character&gt;     &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">##  549656                 NA              NA                     NA</span></span>
<span><span class="co">##  331820     Bacteroidaceae     Bacteroides                     NA</span></span>
<span><span class="co">##  317182                 NA              NA                     NA</span></span>
<span><span class="co">##  94166     Pasteurellaceae     Haemophilus Haemophilusparainflu..</span></span>
<span><span class="co">##  279599        Nostocaceae  Dolichospermum                     NA</span></span>
<span><span class="co">##  158660     Bacteroidaceae     Bacteroides                     NA</span></span>
<span><span class="co">##  329744             ACK-M1              NA                     NA</span></span>
<span><span class="co">##  326977 Bifidobacteriaceae Bifidobacterium Bifidobacteriumadole..</span></span>
<span><span class="co">##  248140     Bacteroidaceae     Bacteroides      Bacteroidescaccae</span></span>
<span><span class="co">##  550960 Enterobacteriaceae     Providencia                     NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="library-size-read-count" class="level3 page-columns page-full" data-number="4.3.2"><h3 data-number="4.3.2" class="anchored" data-anchor-id="library-size-read-count">
<span class="header-section-number">4.3.2</span> Library size / read count</h3>
<p>The total counts/sample can be calculated using <code>perCellQCMetrics</code>/<code>addPerCellQC</code> from the <code>scater</code> package. The former one just calculates the values, whereas the latter one directly adds them to <code>colData</code>.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/lib-size_f6d8398ec7c0be8c35a853776463d926">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 26 rows and 3 columns</span></span>
<span><span class="co">##                sum  detected     total</span></span>
<span><span class="co">##          &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##  CL3        864077      6964    864077</span></span>
<span><span class="co">##  CC1       1135457      7679   1135457</span></span>
<span><span class="co">##  SV1        697509      5729    697509</span></span>
<span><span class="co">##  M31Fcsw   1543451      2667   1543451</span></span>
<span><span class="co">##  M11Fcsw   2076476      2574   2076476</span></span>
<span><span class="co">##  ...           ...       ...       ...</span></span>
<span><span class="co">##  TS28       937466      2679    937466</span></span>
<span><span class="co">##  TS29      1211071      2629   1211071</span></span>
<span><span class="co">##  Even1     1216137      4213   1216137</span></span>
<span><span class="co">##  Even2      971073      3130    971073</span></span>
<span><span class="co">##  Even3     1078241      2776   1078241</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html">addPerCellQC</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 26 rows and 10 columns</span></span>
<span><span class="co">##          X.SampleID   Primer Final_Barcode Barcode_truncated_plus_T</span></span>
<span><span class="co">##            &lt;factor&gt; &lt;factor&gt;      &lt;factor&gt;                 &lt;factor&gt;</span></span>
<span><span class="co">##  CL3        CL3      ILBC_01        AACGCA                   TGCGTT</span></span>
<span><span class="co">##  CC1        CC1      ILBC_02        AACTCG                   CGAGTT</span></span>
<span><span class="co">##  SV1        SV1      ILBC_03        AACTGT                   ACAGTT</span></span>
<span><span class="co">##  M31Fcsw    M31Fcsw  ILBC_04        AAGAGA                   TCTCTT</span></span>
<span><span class="co">##  M11Fcsw    M11Fcsw  ILBC_05        AAGCTG                   CAGCTT</span></span>
<span><span class="co">##  ...            ...      ...           ...                      ...</span></span>
<span><span class="co">##  TS28         TS28   ILBC_25        ACCAGA                   TCTGGT</span></span>
<span><span class="co">##  TS29         TS29   ILBC_26        ACCAGC                   GCTGGT</span></span>
<span><span class="co">##  Even1        Even1  ILBC_27        ACCGCA                   TGCGGT</span></span>
<span><span class="co">##  Even2        Even2  ILBC_28        ACCTCG                   CGAGGT</span></span>
<span><span class="co">##  Even3        Even3  ILBC_29        ACCTGT                   ACAGGT</span></span>
<span><span class="co">##          Barcode_full_length SampleType</span></span>
<span><span class="co">##                     &lt;factor&gt;   &lt;factor&gt;</span></span>
<span><span class="co">##  CL3             CTAGCGTGCGT      Soil </span></span>
<span><span class="co">##  CC1             CATCGACGAGT      Soil </span></span>
<span><span class="co">##  SV1             GTACGCACAGT      Soil </span></span>
<span><span class="co">##  M31Fcsw         TCGACATCTCT      Feces</span></span>
<span><span class="co">##  M11Fcsw         CGACTGCAGCT      Feces</span></span>
<span><span class="co">##  ...                     ...        ...</span></span>
<span><span class="co">##  TS28            GCATCGTCTGG      Feces</span></span>
<span><span class="co">##  TS29            CTAGTCGCTGG      Feces</span></span>
<span><span class="co">##  Even1           TGACTCTGCGG      Mock </span></span>
<span><span class="co">##  Even2           TCTGATCGAGG      Mock </span></span>
<span><span class="co">##  Even3           AGAGAGACAGG      Mock </span></span>
<span><span class="co">##                                         Description       sum  detected</span></span>
<span><span class="co">##                                            &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##  CL3     Calhoun South Carolina Pine soil, pH 4.9      864077      6964</span></span>
<span><span class="co">##  CC1     Cedar Creek Minnesota, grassland, pH 6.1     1135457      7679</span></span>
<span><span class="co">##  SV1     Sevilleta new Mexico, desert scrub, pH 8.3    697509      5729</span></span>
<span><span class="co">##  M31Fcsw M3, Day 1, fecal swab, whole body study      1543451      2667</span></span>
<span><span class="co">##  M11Fcsw M1, Day 1, fecal swab, whole body study      2076476      2574</span></span>
<span><span class="co">##  ...                                            ...       ...       ...</span></span>
<span><span class="co">##  TS28                                       Twin #1    937466      2679</span></span>
<span><span class="co">##  TS29                                       Twin #2   1211071      2629</span></span>
<span><span class="co">##  Even1                                      Even1     1216137      4213</span></span>
<span><span class="co">##  Even2                                      Even2      971073      3130</span></span>
<span><span class="co">##  Even3                                      Even3     1078241      2776</span></span>
<span><span class="co">##              total</span></span>
<span><span class="co">##          &lt;numeric&gt;</span></span>
<span><span class="co">##  CL3        864077</span></span>
<span><span class="co">##  CC1       1135457</span></span>
<span><span class="co">##  SV1        697509</span></span>
<span><span class="co">##  M31Fcsw   1543451</span></span>
<span><span class="co">##  M11Fcsw   2076476</span></span>
<span><span class="co">##  ...           ...</span></span>
<span><span class="co">##  TS28       937466</span></span>
<span><span class="co">##  TS29      1211071</span></span>
<span><span class="co">##  Even1     1216137</span></span>
<span><span class="co">##  Even2      971073</span></span>
<span><span class="co">##  Even3     1078241</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The distribution of calculated library sizes can be visualized as a histogram (left), or by sorting the samples by library size (right).</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/plot-viz-lib-size-1_39d3d03b0802d5d2cf1314b3a995a990">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sum</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"gray"</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Library size"</span>, y <span class="op">=</span> <span class="st">"Frequency (n)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="co"># scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x), </span></span>
<span>        <span class="co"># labels = scales::trans_format("log10", scales::math_format(10^.x))) +</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, <span class="co"># Removes the grid</span></span>
<span>          panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="co"># Adds y-axis</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">sum</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">index</span>, x <span class="op">=</span> <span class="va">sum</span><span class="op">/</span><span class="fl">1e6</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Library size (million reads)"</span>, y <span class="op">=</span> <span class="st">"Sample index"</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, <span class="co"># Removes the grid</span></span>
<span>          panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="co"># Adds y-axis</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/plot-viz-lib-size-1-1.png" class="img-fluid figure-img" width="768"></p>
<figcaption class="figure-caption">Library size distribution.</figcaption></figure>
</div>
</div>
</div>
<p>Library sizes other variables from <code>colData</code> can be visualized by using specified function called <code>plotColData</code>.</p>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/plot-viz-lib-size-2_42e8e3320a0278c17beddf05c9bacb15">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co"># Sort samples by read count, order the factor levels, and store back to tse as DataFrame</span></span>
<span><span class="co"># TODO: plotColData could include an option for sorting samples based on colData variables</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>                 <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">X.SampleID</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>             <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>X.SampleID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">X.SampleID</span>, levels<span class="op">=</span><span class="va">X.SampleID</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>         <span class="va">DataFrame</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"sum"</span>,<span class="st">"X.SampleID"</span>, colour_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Library size (N)"</span>, x <span class="op">=</span> <span class="st">"Sample ID"</span><span class="op">)</span>       </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/plot-viz-lib-size-2-1.png" class="img-fluid figure-img" width="768"></p>
<figcaption class="figure-caption">Library sizes per sample.</figcaption></figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="12_quality_control_cache/html/plot-viz-lib-size-3_3c79fc9855dc75db93ef5917337868f7">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span><span class="op">(</span><span class="va">tse</span>,<span class="st">"sum"</span>,<span class="st">"SampleType"</span>, colour_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="12_quality_control_files/figure-html/plot-viz-lib-size-3-1.png" class="img-fluid figure-img" width="768"></p>
<figcaption class="figure-caption">Library sizes per sample type.</figcaption></figure>
</div>
</div>
</div>
<p>In addition, data can be rarefied with <a href="https://microbiome.github.io/mia/reference/subsampleCounts.html">subsampleCounts</a>, which normalises the samples to an equal number of reads. However, this practice has been discouraged for the analysis of differentially abundant microorganisms (see <span class="citation" data-cites="mcmurdie2014waste">(<a href="#ref-mcmurdie2014waste" role="doc-biblioref">McMurdie and Holmes 2014</a>)</span>).</p>
<div class="no-row-height column-margin column-container"><div id="ref-mcmurdie2014waste" class="csl-entry" role="listitem">
McMurdie, Paul J, and Susan Holmes. 2014. <span>“Waste Not, Want Not: Why Rarefying Microbiome Data Is Inadmissible.”</span> <em>PLoS Computational Biology</em> 10 (4): e1003531.
</div></div></section><section id="contaminant-sequences" class="level3 page-columns page-full" data-number="4.3.3"><h3 data-number="4.3.3" class="anchored" data-anchor-id="contaminant-sequences">
<span class="header-section-number">4.3.3</span> Contaminant sequences</h3>
<p>Samples might be contaminated with exogenous sequences. The impact of each contaminant can be estimated based on their frequencies and concentrations across the samples.</p>
<p>The following <a href="https://microbiome.github.io/mia/reference/isContaminant.html">decontam functions</a> are based on the <span class="citation" data-cites="davis2018simple">(<a href="#ref-davis2018simple" role="doc-biblioref">Davis et al. 2018</a>)</span> and support such functionality:</p>
<div class="no-row-height column-margin column-container"><div id="ref-davis2018simple" class="csl-entry" role="listitem">
Davis, Nicole M, Diana M Proctor, Susan P Holmes, David A Relman, and Benjamin J Callahan. 2018. <span>“Simple Statistical Identification and Removal of Contaminant Sequences in Marker-Gene and Metagenomics Data.”</span> <em>Microbiome</em> 6 (1): 1–14.
</div></div><ul>
<li>
<code>isContaminant</code>, <code>isNotContaminant</code><br>
</li>
<li>
<code>addContaminantQC</code>, <code>addNotContaminantQC</code>
</li>
</ul>



</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/10_manipulation.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/11_taxonomic_information.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>