<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 9&nbsp; Community Typing (Clustering)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/30_differential_abundance.html" rel="next">
<link href="../pages/21_microbiome_community.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/10_manipulation.html">Focus topics</a></li><li class="breadcrumb-item"><a href="../pages/24_clustering.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Microbiome Analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/microbiome/OMA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://microbiome.github.io/OMA/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#custom-tools" id="toc-custom-tools" class="nav-link active" data-scroll-target="#custom-tools"><span class="header-section-number">9.1</span> Custom tools</a></li>
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering"><span class="header-section-number">9.2</span> Hierarchical clustering</a></li>
  <li>
<a href="#dirichlet-multinomial-mixtures-dmm" id="toc-dirichlet-multinomial-mixtures-dmm" class="nav-link" data-scroll-target="#dirichlet-multinomial-mixtures-dmm"><span class="header-section-number">9.3</span> Dirichlet Multinomial Mixtures (DMM)</a>
  <ul class="collapse">
<li><a href="#pcoa-with-dmm-clusters" id="toc-pcoa-with-dmm-clusters" class="nav-link" data-scroll-target="#pcoa-with-dmm-clusters"><span class="header-section-number">9.3.1</span> PCoA with DMM clusters</a></li>
  </ul>
</li>
  <li>
<a href="#biclustering" id="toc-biclustering" class="nav-link" data-scroll-target="#biclustering"><span class="header-section-number">9.4</span> Biclustering</a>
  <ul class="collapse">
<li><a href="#taxa-vs-samples" id="toc-taxa-vs-samples" class="nav-link" data-scroll-target="#taxa-vs-samples"><span class="header-section-number">9.4.1</span> Taxa vs samples</a></li>
  <li><a href="#taxa-vs-biomolecules" id="toc-taxa-vs-biomolecules" class="nav-link" data-scroll-target="#taxa-vs-biomolecules"><span class="header-section-number">9.4.2</span> Taxa vs biomolecules</a></li>
  </ul>
</li>
  <li><a href="#additional-community-typing" id="toc-additional-community-typing" class="nav-link" data-scroll-target="#additional-community-typing"><span class="header-section-number">9.5</span> Additional Community Typing</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-clustering" class="quarto-section-identifier"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>Clustering techniques are unsupervised machine learning that aims to find groups, called clusters, that share a pattern in the data. In the microbiome context, clustering techniques are included in microbiome community typing methods. For example, clustering allow samples to be distinguished from each other based on their microbiome community composition. There are multiple clustering algorithms available.</p>
<p>The data can be clustered either based on features or samples. Hence, depending on the analysis goal the data might require transformation. The examples below are focused on sample clustering. To learn about feature clustering, check out chapter <a href="11_taxonomic_information.html#sec-taxa-clustering"><span>Section&nbsp;5.3.1</span></a>.</p>
<section id="custom-tools" class="level2" data-number="9.1"><h2 data-number="9.1" class="anchored" data-anchor-id="custom-tools">
<span class="header-section-number">9.1</span> Custom tools</h2>
<p><em>bluster</em> is a Bioconductor package providing tools for clustering data in in the <code>SummarizedExperiment</code> container. It offers multiple algorithms such as hierarchical clustering, DBSCAN, and K-means.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/load_bluster_a45fa0c4b7f7498d77f602bcac85a9ed">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load dependencies</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/">kableExtra</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the first examples of microbiome community typing we use <code>enterotype</code> data.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/load-pkg-data1_c8de2ca962cbb8201074da67fc1ba48d">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"enterotype"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">enterotype</span></span>
<span></span>
<span><span class="co"># Apply transformation</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The main focus in this example is to show how to use mia’s <code>addCluster</code> function to cluster enterotype data. <code>addCluster</code> function allows to choose a clustering algorithm and offers multiple parameters to shape the result.</p>
<p>In this example, HclustParam() parameter is chosen for hierarchical clustering. HclustParam() parameter itself has parameters on its own <a href="https://rdrr.io/github/LTLA/bluster/man/HclustParam-class.html">HclustParam documentation</a>. A parameter is <code>MARGIN</code> defines whether to cluster features or samples .</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/bluster_sample_hclust1_7aee0bff1a8842d1670bd6be81681fc2">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Simple use of the hierarchical clustering. Here, the default parameters</span></span>
<span><span class="co"># Set the cut height to half of the dendrogram height</span></span>
<span></span>
<span><span class="co"># Save as an alternative experiment that contains clustering information</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addCluster.html">addCluster</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"relabundance"</span>, </span>
<span>               MARGIN <span class="op">=</span> <span class="st">"samples"</span>, <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/HclustParam-class.html">HclustParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The result can be found in 'clusters' column of colData</span></span>
<span></span>
<span><span class="co"># The number of samples included in each cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="co">##   1  2  3  4  5  6  7  8  9 </span></span>
<span><span class="co">##  92 25 21  2 55 41 20 23  1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once the clustering on the samples is done, we can also plot the clusters.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/bluster_sample_plot_c7126ea1803f0e2a56d4a6723d1417ef">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the MDS dimensions for plotting</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runMDS.html">runMDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust"</span><span class="op">)</span>, assay.type <span class="op">=</span> <span class="st">"relabundance"</span>, </span>
<span>              FUN <span class="op">=</span> <span class="fu">vegan</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span>, method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the clusters</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust"</span><span class="op">)</span>, <span class="st">"MDS"</span>, colour_by <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/bluster_sample_plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="hierarchical-clustering" class="level2" data-number="9.2"><h2 data-number="9.2" class="anchored" data-anchor-id="hierarchical-clustering">
<span class="header-section-number">9.2</span> Hierarchical clustering</h2>
<p>The hierarchical clustering algorithm aims to find hierarchy between samples/features. There are two approaches: agglomerative (“bottom-up”) and divisive (“top-down”). In agglomerative approach, each observation is first in a unique cluster. The algorithm continues by agglomerating similar clusters. The divisive approach, instead, starts with one cluster that contains all observations. Clusters are split recursively into clusters that differ the most. The clustering ends when each cluster contains only one observation. In this algorithm, the similarity of two clusters is based on the distance between them.</p>
<p>Hierarchical clustering can be visualized with a dendrogram tree. In each splitting point, the tree is divided into two clusters leading to the hierarchy.</p>
<p>Hierarchical clustering requires two steps. 1. Computation of the dissimilarities with a given distance.<br>
2. Clustering based on dissimilarities.</p>
<p>Additionally, since sequencing data is compositional, we apply relative transformation (as seen in the previous example).</p>
<p>In this example, we want to add information on the clustering. To do so, we use the <code>full</code> parameter. We also compute the dissimilarities with the bray distance. Finally, the <code>clust.col</code> parameter allows us to choose the name of the column in the colData (default name is <code>clusters</code>).</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/hclust2_f80b051b58b8cc3aa111580649d7b14a">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vegandevs/vegan">vegan</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save another alternative experiment that contains full clustering information</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust_full"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addCluster.html">addCluster</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>               assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>               MARGIN <span class="op">=</span> <span class="st">"samples"</span>,</span>
<span>               <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/HclustParam-class.html">HclustParam</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>                           dist.fun <span class="op">=</span> <span class="va">vegdist</span>,</span>
<span>                           metric <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span>,</span>
<span>               full <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               clust.col <span class="op">=</span> <span class="st">"Hclust"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We plot the dendrogram, which is possible since we got the additional information from the clustering.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/hclust3_daa8410b2ecdfe69046bdb3f2f5bd147">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://talgalili.github.io/dendextend/">dendextend</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get hclust data from metadata</span></span>
<span><span class="va">hclust_data</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust_full"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span><span class="op">$</span><span class="va">hclust</span></span>
<span></span>
<span><span class="co"># Get the dendrogram object</span></span>
<span><span class="va">dendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hclust_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot dendrogram</span></span>
<span><span class="va">dendro</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="http://talgalili.github.io/dendextend/reference/set.html">set</a></span><span class="op">(</span><span class="st">"labels"</span>, <span class="cn">NULL</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/hclust3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In our case, we cut the dendrogram in half by default. To know how many clusters we have, we can check the colData.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/hclust4_68c273ac5e55acc370934f2a6fce3fed">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get the clusters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust_full"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">Hclust</span><span class="op">)</span></span>
<span><span class="co">##   1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 </span></span>
<span><span class="co">##   3 65 35 12  9  2 23 23 27 15  4 11  6 11  2  1  9  7  4  4  1  2  1  1  1 </span></span>
<span><span class="co">##  26 </span></span>
<span><span class="co">##   1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that there are 26 clusters, but that probably is not optimal since the the number of clusters has been chosen arbitrarily. To determine the number of clusters, we can use the dendrogram. Usually the tree is split where the branch length is the largest. However, as we can see from the dendrogram, clusters are not clear. There are algorithms to identify the optimal number of clusters.</p>
<p>The <code>NbClust</code> library is useful to that end as it offers multiple methods to determine the optimal number of clusters. Here we will use the silhouette analysis to determine the optimal number of clusters. For each data point, this analysis measures the distance to other data points in the same cluster (cohesion), and the distance to the other clusters (separation), establishing a score. That score is then combined across the data points. <code>NbClust</code> does this for multiple number of clusters and the best score corresponds to the optimal number of clusters.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/hclust5_5de097c3ce034a6d807711afdb3d5b69">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sites.google.com/site/malikacharrad/research/nbclust-package">NbClust</a></span><span class="op">)</span></span>
<span><span class="va">diss</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"hclust_full"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span><span class="op">$</span><span class="va">dist</span></span>
<span></span>
<span><span class="co"># Apply the silhouette analysis on the distance matrix</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NbClust/man/NbClust.html">NbClust</a></span><span class="op">(</span>diss <span class="op">=</span> <span class="va">diss</span>, distance <span class="op">=</span> <span class="cn">NULL</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span>,</span>
<span>               index <span class="op">=</span> <span class="st">"silhouette"</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##   Only frey, mcclain, cindex, sihouette and dunn can be computed. To compute the other indices, data matrix is needed</span></span>
<span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">Best.nc</span></span>
<span><span class="co">##  Number_clusters     Value_Index </span></span>
<span><span class="co">##           2.0000          0.4783</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Based on the result, let’s divide observations into 2 clusters.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/hclust6_c157ebf58e50cd1ad74f7c677efdf9d8">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://talgalili.github.io/dendextend/">dendextend</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get optimal number of clusters</span></span>
<span><span class="va">k</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">Best.nc</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Making colors for 2 clusters</span></span>
<span><span class="va">col_val_map</span> <span class="op">&lt;-</span> <span class="fu">randomcoloR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/randomcoloR/man/distinctColorPalette.html">distinctColorPalette</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"clust_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dend</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://talgalili.github.io/dendextend/reference/color_branches.html">color_branches</a></span><span class="op">(</span><span class="va">dendro</span>, k <span class="op">=</span> <span class="va">k</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">col_val_map</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/labels.html">labels</a></span><span class="op">(</span><span class="va">dend</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">dend</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/hclust6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="dirichlet-multinomial-mixtures-dmm" class="level2" data-number="9.3"><h2 data-number="9.3" class="anchored" data-anchor-id="dirichlet-multinomial-mixtures-dmm">
<span class="header-section-number">9.3</span> Dirichlet Multinomial Mixtures (DMM)</h2>
<p>This section focus on <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0030126">Dirichlet-Multinomial Mixture Model</a> analysis. It is a probabilistic technique that allows to search for sample patterns that reflect sample similarity in the data. DMM has a property of determining an optimal number of clusters (k) to obtain the best model. The minimum value of Laplace approximation to the negative log model evidence for DMM models as a function of k, determines an optimal k. The optimal k suggests to fit a model with k mixtures of Dirichlet distributions. For the best model, k probabilities for each sample to belong to each cluster are obtained.</p>
<p>In this example, we cluster the data with DMM clustering. Since the data set is large, the algorithm requires a lot of computational capacity. Therefore, we use only a subset of the data that is agglomerated by Phylum as a rank.</p>
<p>In the example of DMM we use <code>GlobalPatterns</code> data.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm1_a42c1e69eedcac90b672f1a976c106a3">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span>
<span></span>
<span><span class="co"># Agglomerate by rank</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span>, agglomerateTree <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the example below, we calculate model fit using Laplace approximation. The cluster information is added in the metadata with an optional <code>name</code>.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm2_6fdb9d9fd031787ad7ff2179a6e3b550">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run the model and calculates the most likely number of clusters from 1 to 7</span></span>
<span></span>
<span><span class="co"># Save as an alternative experiment that contains clustering information</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/addCluster.html">addCluster</a></span><span class="op">(</span><span class="va">tse</span>, name <span class="op">=</span> <span class="st">"DMM"</span>, <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/DmmParam-class.html">DmmParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, type <span class="op">=</span> <span class="st">"laplace"</span><span class="op">)</span>, </span>
<span>                   MARGIN <span class="op">=</span> <span class="st">"samples"</span>, full <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm3_62d5dfbc0a4dc8494f6c5a23521e487b">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># The dmm information is stored in the metadata under the 'DMM' column that includes information about all seven models </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 67 26 </span></span>
<span><span class="co">##  metadata(2): agglomerated_by_rank DMM</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(67): Phylum:Crenarchaeota Phylum:Euryarchaeota ...</span></span>
<span><span class="co">##    Phylum:Synergistetes Phylum:SR1</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(8): X.SampleID Primer ... Description clusters</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (67 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (19216 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The plot below represents the Laplace approximation to the model evidence for each of the k models. We can see that the best number of clusters is two.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm4_e8d2db17bb5bcf012fb20cfc8135e590">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">miaViz</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/miaViz/man/plotDMN.html">plotDMNFit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"laplace"</span>, name <span class="op">=</span> <span class="st">"DMM"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/dmm4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The best model can be confirmed with the following operation.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm5_552cf714b28762700c7e069cc5f17cae">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get the the best model</span></span>
<span><span class="va">bestFit</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">DMM</span><span class="op">$</span><span class="va">dmm</span><span class="op">[[</span><span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">DMM</span><span class="op">$</span><span class="va">best</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">bestFit</span></span>
<span><span class="co">##  class: DMN </span></span>
<span><span class="co">##  k: 2 </span></span>
<span><span class="co">##  samples x taxa: 26 x 67 </span></span>
<span><span class="co">##  Laplace: 7673 BIC: 7927 AIC: 7842</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The clusters for the best model are saved in the colData under ‘clusters’ column.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm6_36034f8bb7788807ba49890ddc1edffd">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">##      CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong </span></span>
<span><span class="co">##        1       1       1       2       2       2       2       2       2 </span></span>
<span><span class="co">##  M11Tong </span></span>
<span><span class="co">##        2 </span></span>
<span><span class="co">##  Levels: 1 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>More detailed information about the clusters can be accessed in the metadata. The metadata contains samples-cluster assignment probabilities that tell us the likelihood for each sample to belong to each cluster.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm7_265200bb5887de4edad2ba2468efdd04">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">DMM</span><span class="op">$</span><span class="va">prob</span>, <span class="fl">10</span><span class="op">)</span> </span>
<span><span class="co">##                  1         2</span></span>
<span><span class="co">##  CL3     1.000e+00 4.995e-17</span></span>
<span><span class="co">##  CC1     1.000e+00 3.835e-22</span></span>
<span><span class="co">##  SV1     1.000e+00 1.948e-12</span></span>
<span><span class="co">##  M31Fcsw 7.900e-26 1.000e+00</span></span>
<span><span class="co">##  M11Fcsw 1.133e-16 1.000e+00</span></span>
<span><span class="co">##  M31Plmr 1.119e-13 1.000e+00</span></span>
<span><span class="co">##  M11Plmr 3.397e-06 1.000e+00</span></span>
<span><span class="co">##  F21Plmr 4.306e-11 1.000e+00</span></span>
<span><span class="co">##  M31Tong 1.441e-08 1.000e+00</span></span>
<span><span class="co">##  M11Tong 2.223e-06 1.000e+00</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once the optimal model have been confirmed, we can find out which samples are grouped with each other. The table below shows one sample of each sample type clustered in either of the groups. We can notice that DMM can distinguish environmental samples into one group, and mock and human samples into another. For clarity, in this example, the probabilities for each sample to belong in each cluster have been rounded.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm8_7a2eb63749a5bf0b3c102a76f7c6211b">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">DMM</span><span class="op">$</span><span class="va">prob</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">clusters</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">$</span><span class="va">SampleType</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="co"># add sample type information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Group1"</span>, <span class="st">"Group2"</span>, <span class="st">"SampleType"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">SampleType</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">Group1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">##  # A tibble: 9 × 3</span></span>
<span><span class="co">##  # Groups:   SampleType [9]</span></span>
<span><span class="co">##    Group1 Group2 SampleType</span></span>
<span><span class="co">##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;     </span></span>
<span><span class="co">##  1 0      1      Feces     </span></span>
<span><span class="co">##  2 0      1      Skin      </span></span>
<span><span class="co">##  3 0      1      Tongue    </span></span>
<span><span class="co">##  4 0      1      Mock      </span></span>
<span><span class="co">##  5 1      0      Soil      </span></span>
<span><span class="co">##  6 1      0      Freshwater</span></span>
<span><span class="co">##  # ℹ 3 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also plot the driver Phyla in each group. In this case, it reflects the differences between environmental and human samples.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm9_210eebf353cafa063209a3b6b2b12125">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get the estimates on how much each phyla contributes on each cluster</span></span>
<span><span class="va">best_model</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">DMM</span><span class="op">$</span><span class="va">dmm</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">drivers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">best_model</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">fit</span><span class="op">$</span><span class="va">Estimate</span><span class="op">)</span></span>
<span></span>
<span><span class="va">drivers</span><span class="op">$</span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"Phylum:"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">drivers</span><span class="op">)</span><span class="op">)</span> <span class="co"># Clean phylum names</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">drivers</span> <span class="op">&lt;-</span> <span class="va">drivers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">drivers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">drivers</span>, <span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">phyla</span>, <span class="fl">10</span><span class="op">)</span>, <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">drivers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">drivers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, fill <span class="op">=</span> <span class="st">"deeppink4"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Top phyla in group"</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_light</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">""</span>, y<span class="op">=</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/dmm9-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/dmm9-2.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>We use <code>calculateDMNgroup</code> function to have an overview of the best model. The function groups samples by <code>SampleType</code> column from colData and returns DMNGroup object that contains a summary.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm_group_71d2432d75f1202a65f6dcad6e70cf19">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dmm_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/calculateDMN.html">calculateDMNgroup</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span>, variable <span class="op">=</span> <span class="st">"SampleType"</span>, </span>
<span>                               assay.type <span class="op">=</span> <span class="st">"counts"</span>, k <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                               seed <span class="op">=</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">integer.max</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dmm_group</span></span>
<span><span class="co">##  class: DMNGroup </span></span>
<span><span class="co">##  summary:</span></span>
<span><span class="co">##                     k samples taxa    NLE  LogDet Laplace    BIC  AIC</span></span>
<span><span class="co">##  Feces              2       4   67 1078.3 -106.26   901.1 1171.9 1213</span></span>
<span><span class="co">##  Freshwater         2       2   67  889.6  -97.20   716.9  936.4 1025</span></span>
<span><span class="co">##  Freshwater (creek) 2       3   67 1600.3  862.19  1907.3 1674.5 1735</span></span>
<span><span class="co">##  Mock               2       3   67 1008.4  -55.40   856.6 1082.5 1143</span></span>
<span><span class="co">##  Ocean              2       3   67 1096.7  -56.66   944.3 1170.9 1232</span></span>
<span><span class="co">##  Sediment (estuary) 2       3   67 1195.5   18.63  1080.8 1269.7 1331</span></span>
<span><span class="co">##  Skin               2       3   67  992.6  -85.05   826.1 1066.8 1128</span></span>
<span><span class="co">##  Soil               2       3   67 1380.3   11.20  1261.8 1454.5 1515</span></span>
<span><span class="co">##  Tongue             2       2   67  783.0 -107.79   605.0  829.8  918</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Mixture weights can be used for having a rough approximation of the cluster size.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm10_3a25f3835311ecf5dc14294809f0ddb3">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">DirichletMultinomial</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DirichletMultinomial/man/fitted.html">mixturewt</a></span><span class="op">(</span><span class="va">bestFit</span><span class="op">)</span></span>
<span><span class="co">##        pi theta</span></span>
<span><span class="co">##  1 0.5385 20.58</span></span>
<span><span class="co">##  2 0.4615 15.28</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="pcoa-with-dmm-clusters" class="level3" data-number="9.3.1"><h3 data-number="9.3.1" class="anchored" data-anchor-id="pcoa-with-dmm-clusters">
<span class="header-section-number">9.3.1</span> PCoA with DMM clusters</h3>
<p>In this section we show how to calculate principal coordinates for clr transformed abundance data. To calculate PCoA, we use Aitchison distance as a distance metrics that calculates Euclidean distances for clr transformed compositions.</p>
<p>In the visualization section, we project the sample distances on two dimensional space of first two principal coordinates. We colour the samples based on their DMM clusters. The visualization demonstrates that the DMM clusters can be distinguished on a PCoA plot, although the clusters are not coherent. This means that two-dimensional representation of the data created by PCoA preserves similar information that drives the DMM cluster division.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm11_9fe43fb4e3fd0e32488c38246fa9629e">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add pseudocount, because data contains zeros</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pseudo"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"pseudo"</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># clr transformation</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"relabundance"</span>, method <span class="op">=</span> <span class="st">"clr"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># principal coordinate analysis</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runMDS.html">calculateMDS</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"clr"</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a data frame from principal coordinates</span></span>
<span><span class="va">euclidean_pcoa_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>pcoa1 <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, pcoa2 <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/dmm12_609f7a1c225e708b55ce809480655af9">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a data frame that contains principal coordinates and DMM information</span></span>
<span><span class="va">euclidean_dmm_pcoa_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">euclidean_pcoa_df</span>,</span>
<span>                               dmm_component <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"dmm"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a plot</span></span>
<span><span class="va">euclidean_dmm_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">euclidean_dmm_pcoa_df</span>,</span>
<span>                             <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pcoa1</span>, y <span class="op">=</span> <span class="va">pcoa2</span>, color <span class="op">=</span> <span class="va">dmm_component</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Coordinate 1"</span>,y <span class="op">=</span> <span class="st">"Coordinate 2"</span>, </span>
<span>         title <span class="op">=</span> <span class="st">"PCoA with Aitchison distances"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, <span class="co"># makes titles smaller</span></span>
<span>                                    hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">euclidean_dmm_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/dmm12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="biclustering" class="level2" data-number="9.4"><h2 data-number="9.4" class="anchored" data-anchor-id="biclustering">
<span class="header-section-number">9.4</span> Biclustering</h2>
<p>Biclustering methods cluster rows and columns simultaneously in order to find subsets of correlated features/samples.</p>
<p>Here, we use following packages:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/biclust/index.html">biclust</a></li>
<li><a href="https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13582">cobiclust</a></li>
</ul>
<p><em>cobiclust</em> is especially developed for microbiome data whereas <em>biclust</em> is more general method. In this section, we show two different cases and example solutions to apply biclustering to them.</p>
<ol type="1">
<li>Taxa vs samples</li>
<li>Taxa vs biomolecule/biomarker</li>
</ol>
<p>Biclusters can be visualized using heatmap or boxplot, for instance. For checking purposes, also scatter plot might be valid choice.</p>
<p>Check more ideas for heatmaps from chapters <a href="19_visualization_techniques.html"><span>Chapter&nbsp;18</span></a> and <a href="21_microbiome_community.html"><span>Chapter&nbsp;8</span></a>.</p>
<section id="taxa-vs-samples" class="level3" data-number="9.4.1"><h3 data-number="9.4.1" class="anchored" data-anchor-id="taxa-vs-samples">
<span class="header-section-number">9.4.1</span> Taxa vs samples</h3>
<p>When you have microbial abundance matrices, we suggest to use <em>cobiclust</em> which is designed for microbial data.</p>
<p>Load example data</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/load-pkg-data2_c5943d170a820af88c04a6fcd820aaa8">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/julieaubert/cobiclust">cobiclust</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"HintikkaXOData"</span><span class="op">)</span></span>
<span><span class="va">mae</span> <span class="op">&lt;-</span> <span class="va">HintikkaXOData</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Only the most prevalent taxa are included in analysis.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/cobiclust_1_34cd5fd4df91203e92d4d30a6c59e4e0">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subset data in the first experiment</span></span>
<span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalentFeatures</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, </span>
<span>                                      prevalence <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>                                      detection <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rclr-transform in the first experiment</span></span>
<span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"rclr"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>cobiclust</em> takes counts table as an input and gives <em>cobiclust</em> object as an output. It includes clusters for taxa and samples.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/cobiclust_2_dea345594522d042c75d9003e981a119">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Do clustering using counts table</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cobiclust/man/cobiclust.html">cobiclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get clusters</span></span>
<span><span class="va">row_clusters</span> <span class="op">&lt;-</span> <span class="va">clusters</span><span class="op">$</span><span class="va">classification</span><span class="op">$</span><span class="va">rowclass</span></span>
<span><span class="va">col_clusters</span> <span class="op">&lt;-</span> <span class="va">clusters</span><span class="op">$</span><span class="va">classification</span><span class="op">$</span><span class="va">colclass</span></span>
<span></span>
<span><span class="co"># Add clusters to rowdata and coldata</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">row_clusters</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">col_clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Order data based on clusters</span></span>
<span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span>,</span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Print clusters</span></span>
<span><span class="va">clusters</span><span class="op">$</span><span class="va">classification</span></span>
<span><span class="co">##  $rowclass</span></span>
<span><span class="co">##    [1] 1 1 1 1 1 1 1 1 1 2 2 1 2 1 1 1 2 1 2 1 1 1 2 2 1 1 2 2 1 2 2 1 1 2 1 1</span></span>
<span><span class="co">##   [37] 2 1 2 2 2 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 2 1 1 2 1 2 2 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">##   [73] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 2 2 2 2 1 1 1 1 1 1 1 2 2 1</span></span>
<span><span class="co">##  [109] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $colclass</span></span>
<span><span class="co">##   C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 </span></span>
<span><span class="co">##    1   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 </span></span>
<span><span class="co">##  C20 C21 C22 C23 C24 C25 C26 C27 C28 C29 C30 C31 C32 C33 C34 C35 C36 C37 C38 </span></span>
<span><span class="co">##    2   2   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3 </span></span>
<span><span class="co">##  C39 C40 </span></span>
<span><span class="co">##    3   1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next we can plot clusters. Annotated heatmap is a common choice.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/cobiclust_3a_77aea09add1225bf75091242d9dfdb82">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="co"># z-transform for heatmap</span></span>
<span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, assay.type <span class="op">=</span> <span class="st">"rclr"</span>,</span>
<span>                            MARGIN <span class="op">=</span> <span class="st">"features"</span>, method <span class="op">=</span> <span class="st">"z"</span>, name <span class="op">=</span> <span class="st">"rclr_z"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create annotations. When column names are equal, they should share levels.</span></span>
<span><span class="co"># Here samples include 3 clusters, and taxa 2. That is why we have to make</span></span>
<span><span class="co"># column names unique.</span></span>
<span><span class="va">annotation_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>, <span class="st">"clusters"</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">annotation_col</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"col_clusters"</span></span>
<span></span>
<span><span class="va">annotation_row</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>, <span class="st">"clusters"</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">annotation_row</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"row_clusters"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot the heatmap.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/cobiclust_3b_d5940a851d0df9487951836ae8a73ed7">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"rclr_z"</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">F</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>,</span>
<span>         annotation_col <span class="op">=</span> <span class="va">annotation_col</span>, annotation_row <span class="op">=</span> <span class="va">annotation_row</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/cobiclust_3b-1.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
</div>
<p>Boxplot is commonly used to summarize the results:</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/cobiclust_4_59cbb968a35a6ffe5c771ed1abefc82d">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ggplot requires data in melted format</span></span>
<span><span class="va">melt_assay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/meltAssay.html">meltAssay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, assay.type <span class="op">=</span> <span class="st">"rclr"</span>, </span>
<span>                        add_col_data <span class="op">=</span> <span class="cn">T</span>, add_row_data <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># patchwork two plots side-by-side</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">melt_assay</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">clusters.x</span>, y <span class="op">=</span> <span class="va">rclr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Taxa clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">melt_assay</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">clusters.y</span>, y <span class="op">=</span> <span class="va">rclr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Sample clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/cobiclust_4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="taxa-vs-biomolecules" class="level3" data-number="9.4.2"><h3 data-number="9.4.2" class="anchored" data-anchor-id="taxa-vs-biomolecules">
<span class="header-section-number">9.4.2</span> Taxa vs biomolecules</h3>
<p>Here, we analyze cross-correlation between taxa and metabolites. This is a case, where we use <em>biclust</em> method which is suitable for numeric matrices in general. First we pre-process the data.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_1_7bbc97756a586d155e11eb55a4b3c605">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Samples must be in equal order</span></span>
<span><span class="co"># (Only 1st experiment was ordered in cobiclust step leading to unequal order)</span></span>
<span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Make rownames unique since it is required by other steps</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html">make.unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transform the metabolites to be in log basis</span></span>
<span><span class="va">mae</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, assay.type <span class="op">=</span> <span class="st">"nmr"</span>, method <span class="op">=</span> <span class="st">"log10"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add missing data to the metabolites</span></span>
<span><span class="va">replace_na</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">row</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">na_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">row</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">non_na_values</span> <span class="op">&lt;-</span> <span class="va">row</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">row</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="va">row</span><span class="op">[</span><span class="va">na_indices</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">non_na_values</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">na_indices</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">row</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"log10"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"log10"</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">replace_na</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we compute the spearman correlation matrix.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_2_dd002d0bc2f026bd8247fb0578c5a6c5">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Calculate correlations</span></span>
<span><span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getExperimentCrossAssociation.html">getExperimentCrossCorrelation</a></span><span class="op">(</span><span class="va">mae</span>, <span class="fl">1</span>, <span class="fl">2</span>, assay.type1 <span class="op">=</span> <span class="st">"rclr"</span>,</span>
<span>                                      assay.type2 <span class="op">=</span> <span class="st">"log10"</span>, mode <span class="op">=</span> <span class="st">"matrix"</span>,</span>
<span>                                      correlation <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>biclust</em> takes a matrix as an input and returns a <em>biclust</em> object.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_3_7760bf71fc5250a4915c801a218d0060">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">biclust</span><span class="op">)</span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3973</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find biclusters</span></span>
<span><span class="va">bc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biclust/man/biclust-method.html">biclust</a></span><span class="op">(</span><span class="va">corr</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/biclust/man/BCPlaid.html">BCPlaid</a></span><span class="op">(</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bc</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  An object of class Biclust </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  call:</span></span>
<span><span class="co">##      biclust(x = corr, method = BCPlaid(), verbose = FALSE)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Number of Clusters found:  6 </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  First  5  Cluster sizes:</span></span>
<span><span class="co">##                     BC 1 BC 2 BC 3 BC 4 BC 5</span></span>
<span><span class="co">##  Number of Rows:      19   19    7    3    3</span></span>
<span><span class="co">##  Number of Columns:   14   10   12    8   11</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The object includes cluster information. However compared to <em>cobiclust</em>, <em>biclust</em> object includes only information about clusters that were found, not general cluster.</p>
<p>Meaning that if one cluster size of 5 features was found out of 20 features, those 15 features do not belong to any cluster. That is why we have to create an additional cluster for features/samples that are not assigned into any cluster.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_4_d9337f73a74187f82c2f3a203d138482">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Functions for obtaining biclust information</span></span>
<span></span>
<span><span class="co"># Get clusters for rows and columns</span></span>
<span><span class="va">.get_biclusters_from_biclust</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">bc</span>, <span class="va">assay</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Get cluster information for columns and rows</span></span>
<span>    <span class="va">bc_columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">bc</span><span class="op">@</span><span class="va">NumberxCol</span><span class="op">)</span></span>
<span>    <span class="va">bc_columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">bc_columns</span><span class="op">)</span></span>
<span>    <span class="va">bc_rows</span> <span class="op">&lt;-</span> <span class="va">bc</span><span class="op">@</span><span class="va">RowxNumber</span></span>
<span>    <span class="va">bc_rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">bc_rows</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># Get data into right format</span></span>
<span>    <span class="va">bc_columns</span> <span class="op">&lt;-</span> <span class="fu">.manipulate_bc_data</span><span class="op">(</span><span class="va">bc_columns</span>, <span class="va">assay</span>, <span class="st">"col"</span><span class="op">)</span></span>
<span>    <span class="va">bc_rows</span> <span class="op">&lt;-</span> <span class="fu">.manipulate_bc_data</span><span class="op">(</span><span class="va">bc_rows</span>, <span class="va">assay</span>, <span class="st">"row"</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>bc_columns <span class="op">=</span> <span class="va">bc_columns</span>, bc_rows <span class="op">=</span> <span class="va">bc_rows</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Input clusters, and how many observations there should be, i.e.,</span></span>
<span><span class="co"># the number of samples or features</span></span>
<span><span class="va">.manipulate_bc_data</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">bc_clusters</span>, <span class="va">assay</span>, <span class="va">row_col</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Get right dimension</span></span>
<span>    <span class="va">dim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">row_col</span> <span class="op">==</span> <span class="st">"col"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">assay</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">assay</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># Get column/row names</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">row_col</span> <span class="op">==</span> <span class="st">"col"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">assay</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">assay</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co"># If no clusters were found, create one. Otherwise create additional</span></span>
<span>    <span class="co"># cluster which</span></span>
<span>    <span class="co"># contain those samples that are not included in clusters that were found.</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">bc_clusters</span><span class="op">)</span> <span class="op">!=</span> <span class="va">dim</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">bc_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="va">dim</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="co"># Create additional cluster that includes those samples/features that</span></span>
<span>        <span class="co"># are not included in other clusters.</span></span>
<span>        <span class="va">vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">bc_clusters</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co"># If additional cluster contains samples, then add it</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="va">vec</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">bc_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">bc_clusters</span>, <span class="va">vec</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co"># Adjust row and column names</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bc_clusters</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bc_clusters</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cluster_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">bc_clusters</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">bc_clusters</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_5_36a31235a084c512f64f64c5b5187cf0">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get biclusters</span></span>
<span><span class="va">bcs</span> <span class="op">&lt;-</span> <span class="fu">.get_biclusters_from_biclust</span><span class="op">(</span><span class="va">bc</span>, <span class="va">corr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bicluster_rows</span> <span class="op">&lt;-</span> <span class="va">bcs</span><span class="op">$</span><span class="va">bc_rows</span></span>
<span><span class="va">bicluster_columns</span> <span class="op">&lt;-</span> <span class="va">bcs</span><span class="op">$</span><span class="va">bc_columns</span></span>
<span></span>
<span><span class="co"># Print biclusters for rows</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">bicluster_rows</span><span class="op">)</span></span>
<span><span class="co">##                                                                                                                        cluster_1</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE</span></span>
<span><span class="co">##                                                                                                                        cluster_2</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                           TRUE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE</span></span>
<span><span class="co">##                                                                                                                        cluster_3</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                           TRUE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE</span></span>
<span><span class="co">##                                                                                                                        cluster_4</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE</span></span>
<span><span class="co">##                                                                                                                        cluster_5</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE</span></span>
<span><span class="co">##                                                                                                                        cluster_6</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                   FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella               FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella     FALSE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                  FALSE</span></span>
<span><span class="co">##                                                                                                                        cluster_7</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Bacillales_D_4__Staphylococcaceae_D_5__Staphylococcus                                    TRUE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Klebsiella                TRUE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Bacilli_D_3__Lactobacillales_D_4__Streptococcaceae_D_5__Streptococcus                                 TRUE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Enterobacteriales_D_4__Enterobacteriaceae_D_5__Escherichia-Shigella      TRUE</span></span>
<span><span class="co">##  D_1__Firmicutes_D_2__Clostridia_D_3__Clostridiales_D_4__Ruminococcaceae_D_5__Ruminiclostridium 5                          FALSE</span></span>
<span><span class="co">##  D_1__Proteobacteria_D_2__Gammaproteobacteria_D_3__Pseudomonadales_D_4__Pseudomonadaceae_D_5__Pseudomonas                   TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s collect information for the scatter plot.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_6_31e4ffabf443fef222e1714bfc41af76">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Function for obtaining sample-wise sum, mean, median, and mean variance</span></span>
<span><span class="co"># for each cluster</span></span>
<span></span>
<span><span class="va">.sum_mean_median_var</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">tse1</span>, <span class="va">tse2</span>, <span class="va">assay.type1</span>, <span class="va">assay.type2</span>, <span class="va">clusters1</span>, <span class="va">clusters2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="co"># Create a data frame that includes all the information</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">clusters1</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># Subset data based on cluster</span></span>
<span>        <span class="va">tse_subset1</span> <span class="op">&lt;-</span> <span class="va">tse1</span><span class="op">[</span><span class="va">clusters1</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>, <span class="op">]</span></span>
<span>        <span class="va">tse_subset2</span> <span class="op">&lt;-</span> <span class="va">tse2</span><span class="op">[</span><span class="va">clusters2</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>, <span class="op">]</span></span>
<span>        <span class="co"># Get assay</span></span>
<span>        <span class="va">assay1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse_subset1</span>, <span class="va">assay.type1</span><span class="op">)</span></span>
<span>        <span class="va">assay2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse_subset2</span>, <span class="va">assay.type2</span><span class="op">)</span></span>
<span>        <span class="co"># Calculate sum, mean, median, and mean variance</span></span>
<span>        <span class="va">sum1</span> <span class="op">&lt;-</span> <span class="fu">colSums2</span><span class="op">(</span><span class="va">assay1</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        <span class="va">mean1</span> <span class="op">&lt;-</span> <span class="fu">colMeans2</span><span class="op">(</span><span class="va">assay1</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        <span class="va">median1</span> <span class="op">&lt;-</span> <span class="fu">colMedians</span><span class="op">(</span><span class="va">assay1</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        <span class="va">var1</span> <span class="op">&lt;-</span> <span class="fu">colVars</span><span class="op">(</span><span class="va">assay1</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="va">sum2</span> <span class="op">&lt;-</span> <span class="fu">colSums2</span><span class="op">(</span><span class="va">assay2</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        <span class="va">mean2</span> <span class="op">&lt;-</span> <span class="fu">colMeans2</span><span class="op">(</span><span class="va">assay2</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        <span class="va">median2</span> <span class="op">&lt;-</span> <span class="fu">colMedians</span><span class="op">(</span><span class="va">assay2</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        <span class="va">var2</span> <span class="op">&lt;-</span> <span class="fu">colVars</span><span class="op">(</span><span class="va">assay2</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="va">list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tse1</span><span class="op">)</span>, <span class="va">sum1</span>, <span class="va">sum2</span>, <span class="va">mean1</span>, </span>
<span>                                 <span class="va">mean2</span>, <span class="va">median1</span>, <span class="va">median2</span>, <span class="va">var1</span>, <span class="va">var2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">list</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Calculate info</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">.sum_mean_median_var</span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">mae</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"rclr"</span>, <span class="st">"log10"</span>, <span class="va">bicluster_rows</span>, <span class="va">bicluster_columns</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can create a scatter plot. X-axis includes median clr abundance of microbiome and y-axis median absolute concentration of each metabolite. Each data point represents a single sample.</p>
<p>From the plots, we can see that there is low negative correlation in both cluster 1 and 3. This means that when abundance of bacteria belonging to cluster 1 or 3 is higher, the concentration of metabolites of cluster 1 or 3 is lower, and vice versa.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_7_9e8b1b39f2a89bfd2e007b882a1df184">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">pics</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">median1</span>, y <span class="op">=</span> <span class="va">median2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster "</span>, <span class="va">i</span><span class="op">)</span>, x <span class="op">=</span> <span class="st">"Taxa (rclr median)"</span>,</span>
<span>             y <span class="op">=</span> <span class="st">"Metabolites (abs. median)"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pics</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-1.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-2.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-3.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-4.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-5.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-6.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-7.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pics</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">pics</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">pics</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_7-8.png" class="img-fluid figure-img" style="width:33.0%"></p>
</figure>
</div>
</div>
</div>
<p><em>pheatmap</em> does not allow boolean values, so they must be converted into factors.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_8_84fbc089d5461c279813b7b72f5c3131">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">bicluster_columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">bicluster_columns</span>, <span class="fl">2</span>, <span class="va">as.factor</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bicluster_rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">bicluster_rows</span>, <span class="fl">2</span>, <span class="va">as.factor</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again, we can plot clusters with heatmap.</p>
<div class="cell" data-layout-align="center" data-hash="24_clustering_cache/html/biclust_9_a35e1cb2daa864a070c6b454b8d7f30f">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Adjust colors for all clusters</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">bicluster_rows</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">bicluster_columns</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cluster_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bicluster_rows</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">cluster_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bicluster_columns</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">annotation_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">name</span> <span class="kw">in</span> <span class="va">cluster_names</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">annotation_colors</span><span class="op">[[</span><span class="va">name</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Create a heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">corr</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>, cluster_rows <span class="op">=</span> <span class="cn">F</span>,</span>
<span>         annotation_col <span class="op">=</span> <span class="va">bicluster_columns</span>, annotation_row <span class="op">=</span> <span class="va">bicluster_rows</span>,</span>
<span>         annotation_colors <span class="op">=</span> <span class="va">annotation_colors</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="24_clustering_files/figure-html/biclust_9-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="additional-community-typing" class="level2" data-number="9.5"><h2 data-number="9.5" class="anchored" data-anchor-id="additional-community-typing">
<span class="header-section-number">9.5</span> Additional Community Typing</h2>
<p>For more community typing techniques applied to the ‘SprockettTHData’ data set, see the attached .Rmd file.</p>
<p>Link:</p>
<ul>
<li><a href="add-comm-typing.Rmd">Rmd</a></li>
</ul>


</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/21_microbiome_community.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/30_differential_abundance.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>