<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 10&nbsp; Differential Abundance</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/40_machine_learning.html" rel="next">
<link href="../pages/24_clustering.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar docked nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../docs/pages/images/mia_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Orchestrating Microbiome Analysis</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">TEST</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-bi-github">
<li>
    <a class="dropdown-item" href="https://github.com/microbiome/OMA/">
 <span class="dropdown-text">Source Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://microbiome.github.io/OMA/docs/devel/">
 <span class="dropdown-text">Browse version <code>devel</code></span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/10_manipulation.html">Focus topics</a></li><li class="breadcrumb-item"><a href="../pages/30_differential_abundance.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#statistical-challenges-of-microbiome-data" id="toc-statistical-challenges-of-microbiome-data" class="nav-link active" data-scroll-target="#statistical-challenges-of-microbiome-data"><span class="header-section-number">10.1</span> Statistical challenges of microbiome data</a></li>
  <li>
<a href="#using-the-tools" id="toc-using-the-tools" class="nav-link" data-scroll-target="#using-the-tools"><span class="header-section-number">10.2</span> Using the tools</a>
  <ul class="collapse">
<li><a href="#preparing-the-data-for-daa" id="toc-preparing-the-data-for-daa" class="nav-link" data-scroll-target="#preparing-the-data-for-daa"><span class="header-section-number">10.2.1</span> Preparing the data for DAA</a></li>
  <li><a href="#aldex2" id="toc-aldex2" class="nav-link" data-scroll-target="#aldex2"><span class="header-section-number">10.2.2</span> ALDEx2</a></li>
  <li><a href="#ancom-bc" id="toc-ancom-bc" class="nav-link" data-scroll-target="#ancom-bc"><span class="header-section-number">10.2.3</span> ANCOM-BC</a></li>
  <li><a href="#maaslin2" id="toc-maaslin2" class="nav-link" data-scroll-target="#maaslin2"><span class="header-section-number">10.2.4</span> MaAsLin2</a></li>
  <li><a href="#linda" id="toc-linda" class="nav-link" data-scroll-target="#linda"><span class="header-section-number">10.2.5</span> LinDA</a></li>
  <li><a href="#zicoseq" id="toc-zicoseq" class="nav-link" data-scroll-target="#zicoseq"><span class="header-section-number">10.2.6</span> ZicoSeq</a></li>
  <li><a href="#philr" id="toc-philr" class="nav-link" data-scroll-target="#philr"><span class="header-section-number">10.2.7</span> PhILR</a></li>
  <li><a href="#comparison-of-methods" id="toc-comparison-of-methods" class="nav-link" data-scroll-target="#comparison-of-methods"><span class="header-section-number">10.2.8</span> Comparison of methods</a></li>
  </ul>
</li>
  <li>
<a href="#daa-with-confounding" id="toc-daa-with-confounding" class="nav-link" data-scroll-target="#daa-with-confounding"><span class="header-section-number">10.3</span> DAA with confounding</a>
  <ul class="collapse">
<li><a href="#selecting-confounders" id="toc-selecting-confounders" class="nav-link" data-scroll-target="#selecting-confounders"><span class="header-section-number">10.3.1</span> Selecting confounders</a></li>
  <li><a href="#ancom-bc-1" id="toc-ancom-bc-1" class="nav-link" data-scroll-target="#ancom-bc-1"><span class="header-section-number">10.3.2</span> ANCOM-BC</a></li>
  <li><a href="#linda-1" id="toc-linda-1" class="nav-link" data-scroll-target="#linda-1"><span class="header-section-number">10.3.3</span> LinDA</a></li>
  <li><a href="#zicoseq-1" id="toc-zicoseq-1" class="nav-link" data-scroll-target="#zicoseq-1"><span class="header-section-number">10.3.4</span> ZicoSeq</a></li>
  </ul>
</li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><span class="header-section-number">10.4</span> Additional resources</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/10_manipulation.html">Focus topics</a></li><li class="breadcrumb-item"><a href="../pages/30_differential_abundance.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-differential-abundance" class="quarto-section-identifier"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>This section provides an introduction to Differential Abundance Analysis (DAA), which is used to identify differences in the abundances of individual taxa (at any taxonomic level) between two or more groups, such as treatment and control. Here, we demonstrate its implementation on <a href="https://microbiome.github.io/mia/reference/Tengeler2020.html">Tengeler2020</a> data set.</p>
<p>The goal of DAA is to identify biomarkers of a certain phenotype or condition, and gain understanding of a complex system by looking at its isolated components. For example, the identification of a bacterial taxon that is differentially abundant between healthy patients and diseased patients can lead to important insights into the pathophysiology of the disease. In other words, differentially abundant taxa can be involved in the dynamics of the disease, which in turn helps understand the system as a whole. Despite its relevance in current research, the DAA approach has also been subject to debate <span class="citation" data-cites="Quinn2021">(<a href="#ref-Quinn2021" role="doc-biblioref">Quinn, Gordon-Rodriguez, and Erb 2021</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Quinn2021" class="csl-entry" role="listitem">
Quinn, Thomas P., Elliott Gordon-Rodriguez, and Ionas Erb. 2021. <span>“A <span>Critique</span> of <span>Differential Abundance Analysis</span>, and <span>Advocacy</span> for an <span>Alternative</span>.”</span> <em>arXiv:2104.07266 [q-Bio, Stat]</em>, April. <a href="https://arxiv.org/abs/2104.07266">https://arxiv.org/abs/2104.07266</a>.
</div></div><section id="statistical-challenges-of-microbiome-data" class="level2 page-columns page-full" data-number="10.1"><h2 data-number="10.1" class="anchored" data-anchor-id="statistical-challenges-of-microbiome-data">
<span class="header-section-number">10.1</span> Statistical challenges of microbiome data</h2>
<p>Microbiome data display unique properties that are exclusively addressed by DAA tools developed for microbiome analysis. Specifically, microbiome data are characterized by high variability, zero-inflation and compositionality. <strong>High variability</strong> expresses that abundance of taxa often varies by several orders of magnitude from sample to sample. <strong>Zero-inflation</strong> means that typically more than 70% of the values are zeros, which could be due to either physical absence (structural zeros) or insufficient sampling effort (sampling zeros). <strong>Compositionality</strong> implies that a change in the absolute abundance of one taxon will lead to apparent variations in the relative abundances of other taxa in the same sample. If neglected, such properties may cause significant bias in the results of DAA. Therefore, several approaches have been developed to address the unique properties of microbiome data and provide statistically useful results.</p>
<p>The first approach to target zero-inflated data consists of specialized models, such as <strong>over-dispersed count models</strong> and <strong>zero-inflated mixture models</strong>. DESeq2, edgeR and corncorb are based on over-dispersed count models, whereas metagenomeSeq, RAIDA, ZIBB and Omnibus implement zero-inflated mixture models to address zero-inflation. Typically, these models assume a negative binomial, beta-binomial or normal/log-normal distribution. Alternatively, <strong>zero imputation</strong> also represents a valid approach to deal with zero-inflated data. ALDEx2 and eBay apply a Bayesian model to impute the zeros when working with proportion data, accounting for sampling variability and sequencing depth variation. Other methods, such as MaAsLin2 and ANCOMBC impute the zeros with a pseudo-count strategy.</p>
<p>Regarding the compositionality of microbiome data, several approaches have been developed to perform <strong>robust normalization</strong> with methods specifically designed to reduce the bias found in compositional data. Some examples include trimmed mean of M-values (TMM) normalization used by edgeR, relative log expression (RLE) normalization used by DESeq2, cumulative sum scaling (CSS) normalization used by metagenomeSeq, centered log-ratio transformation (CLR) normalization used by ALDEx2 and geometric mean of pairwise ratios (GMPR) normalization used by Omnibus and Wrench normalization <span class="citation" data-cites="Kumar2018">(<a href="#ref-Kumar2018" role="doc-biblioref">Kumar et al. 2018</a>)</span>, which corrects the compositional bias by an empirical Bayes approach. Other methods to deal with compositional data entail reference taxa approach used by DACOMP and RAIDA, analyzing the pattern of pairwise log ratios as done by ANCOM and bias-correction applied by ANCOMBC.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Kumar2018" class="csl-entry" role="listitem">
Kumar, M. Senthil, V. Eric Slud, Kwame Okrah, C. Stephanie Hicks, Sridhar Hannenhalli, and Corrada Héctor Bravo. 2018. <span>“Analysis and Correction of Compositional Bias in Sparse Sequencing Count Data.”</span> <em>BMC Genomics</em> 19 (November). <a href="https://doi.org/10.1186/s12864-018-5160-5">https://doi.org/10.1186/s12864-018-5160-5</a>.
</div><div id="ref-Calgaro2020" class="csl-entry" role="listitem">
Calgaro, Matteo, Chiara Romualdi, Levi Waldron, Davide Risso, and Nicola Vitulo. 2020. <span>“Assessment of Statistical Methods from Single Cell, Bulk RNA-Seq, and Metagenomics Applied to Microbiome Data.”</span> <em>Genome Biology</em> 21 (1): 191. <a href="https://doi.org/10.1186/s13059-020-02104-1">https://doi.org/10.1186/s13059-020-02104-1</a>.
</div><div id="ref-Calgaro2022" class="csl-entry" role="listitem">
Calgaro, Matteo, Chiara Romualdi, Davide Risso, and Nicola Vitulo. 2022. <span>“Benchdamic: Benchmarking of Differential Abundance Methods for Microbiome Data.”</span> <em>Bioinformatics</em> 39 (1). <a href="https://doi.org/10.1093/bioinformatics/btac778">https://doi.org/10.1093/bioinformatics/btac778</a>.
</div></div><p>We recommend to have a look at <span class="citation" data-cites="Nearing2022">Nearing et al. (<a href="#ref-Nearing2022" role="doc-biblioref">2022</a>)</span>. In this study, multiple DAA methods were applied to 38 different datasets and their results were compared to one another. Because each method follows a slightly different approach in terms of assumptions and normalization techniques, it was shown that results on the same dataset may differ substantially depending on the method. Recently, <span class="citation" data-cites="Yang2022">Yang and Chen (<a href="#ref-Yang2022" role="doc-biblioref">2022</a>)</span> comprehensively evaluated DAA methods via a semi-parametric framework and 106 real datasets. This study also concluded that different methods can produce contradictory results, creating the risk of cherry-picking the most favorable options for one’s own hypothesis. Therefore, it is highly recommended to perform DAA with multiple methods to determine whether the findings can be reproduced by different approaches. Built on the findings of <span class="citation" data-cites="Calgaro2020">Calgaro et al. (<a href="#ref-Calgaro2020" role="doc-biblioref">2020</a>)</span>, the <em>benchdamic</em> <span class="citation" data-cites="Calgaro2022">(<a href="#ref-Calgaro2022" role="doc-biblioref">Calgaro et al. 2022</a>)</span> package could offer a valuable support in this regard. Through a comprehensive evaluation process it serves both practitioners by comparing DA methods from existing literature, and method developers by providing an impartial tool to evaluate their new approaches in comparison to what is already available. For details, check its extensive <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/benchdamic/inst/doc/intro.html">vignette</a>.</p>
</section><section id="using-the-tools" class="level2 page-columns page-full" data-number="10.2"><h2 data-number="10.2" class="anchored" data-anchor-id="using-the-tools">
<span class="header-section-number">10.2</span> Using the tools</h2>
<p>In this section we demonstrate the use of four methods that can be recommended based on recent literature (ANCOM-BC <span class="citation" data-cites="ancombc2020">(<a href="#ref-ancombc2020" role="doc-biblioref">Lin and Peddada 2020</a>)</span>, <em>ALDEx2</em> <span class="citation" data-cites="Gloor2016">(<a href="#ref-Gloor2016" role="doc-biblioref">Gloor, Macklaim, and Fernandes 2016</a>)</span>, <em>Maaslin2</em> <span class="citation" data-cites="Mallick2020">(<a href="#ref-Mallick2020" role="doc-biblioref">Mallick, Rahnavard, and McIver 2020</a>)</span>, <em>LinDA</em> <span class="citation" data-cites="Zhou2022">(<a href="#ref-Zhou2022" role="doc-biblioref">H. Zhou et al. 2022</a>)</span> and <em>ZicoSeq</em> <span class="citation" data-cites="Yang2022">(<a href="#ref-Yang2022" role="doc-biblioref">Yang and Chen 2022</a>)</span>).</p>
<div class="no-row-height column-margin column-container"></div><p>The purpose of this section is to show how to perform DAA in R, not how to correctly do causal inference. Depending on your experimental setup and your theory, you must determine how to specify any model exactly. E.g., there might be confounding factors that might drive (the absence of) differences between the shown groups that we ignore here for simplicity. Or your dataset is repeated sampling design, matched-pair design or the general longitudianl design. We will demonstrate how to include covariates in those models. We picked a dataset that merely has microbial abundances in a TSE object as well as a grouping variable in the sample data. We simplify the examples by only including two of the three groups.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"Tengeler2020"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">Tengeler2020</span></span>
<span></span>
<span><span class="co"># Show patient status by cohort</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">patient_status</span>, <span class="va">tse</span><span class="op">$</span><span class="va">cohort</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Cohort_1</th>
<th style="text-align: right;">Cohort_2</th>
<th style="text-align: right;">Cohort_3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Control</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">3</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="preparing-the-data-for-daa" class="level3" data-number="10.2.1"><h3 data-number="10.2.1" class="anchored" data-anchor-id="preparing-the-data-for-daa">
<span class="header-section-number">10.2.1</span> Preparing the data for DAA</h3>
<p>Before starting the analysis, it is recommended to reduce the size and complexity of the data to make the results more reproducible. For this purpose, we agglomerate the features by genus and filter them by a prevalence threshold of 10%.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Agglomerate by genus and subset by prevalence</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalent</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                            rank <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span>                            prevalence <span class="op">=</span> <span class="fl">10</span> <span class="op">/</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transform count assay to relative abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                      assay.type <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>While some DAA tools provide optional arguments for prevalence filtering, here we filtered the tse object directly. This way, we ensure that the input data remains the same when multiple tools are used.</p>
</section><section id="aldex2" class="level3 page-columns page-full" data-number="10.2.2"><h3 data-number="10.2.2" class="anchored" data-anchor-id="aldex2">
<span class="header-section-number">10.2.2</span> ALDEx2</h3>
<p>In this section, we will show how to perform DAA with ALDEx2, which can be regarded as the method of choice for its consistency, as it normally identifies features that are also found by complementary methods <span class="citation" data-cites="Nearing2022">(<a href="#ref-Nearing2022" role="doc-biblioref">Nearing et al. 2022</a>)</span>. A more extensive introduction to its functionality is available in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.html">ALDEx2 vignette</a>.</p>
<div class="no-row-height column-margin column-container"></div><p>ALDEx2 estimates technical variation within each sample per taxon by utilizing the Dirichlet distribution. It furthermore applies the CLR transformation (or closely related log-ratio transforms). Depending on the experimental setup, it will perform a two sample Welch’s t-test and Wilcoxon test or a one-way ANOVA and Kruskal-Wallis test. For more complex study designs, there is a possibility to utilize the <code>glm</code> functionality within ALDEx2. The Benjamini-Hochberg procedure is applied by default to correct for multiple testing.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ggloor/ALDEx_bioc">ALDEx2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate Monte Carlo samples of the Dirichlet distribution for each sample.</span></span>
<span><span class="co"># Convert each instance using the centered log-ratio transform.</span></span>
<span><span class="co"># This is the input for all further analyses.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.clr.function.html">aldex.clr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span>, <span class="va">tse</span><span class="op">$</span><span class="va">patient_status</span><span class="op">)</span>     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The t-test:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># calculates expected values of the Welch's t-test and Wilcoxon rank</span></span>
<span><span class="co"># test on the data returned by aldex.clr</span></span>
<span><span class="va">x_tt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.ttest.html">aldex.ttest</a></span><span class="op">(</span><span class="va">x</span>, paired.test <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Effect sizes:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determines the median clr abundance of the feature in all samples and in</span></span>
<span><span class="co"># groups, the median difference between the two groups, the median variation</span></span>
<span><span class="co"># within each group and the effect size, which is the median of the ratio</span></span>
<span><span class="co"># of the between group difference and the larger of the variance within groups</span></span>
<span><span class="va">x_effect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.effect.html">aldex.effect</a></span><span class="op">(</span><span class="va">x</span>, CI <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine all outputs </span></span>
<span><span class="va">aldex_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">x_tt</span>, <span class="va">x_effect</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can create a so called Bland-Altman or MA plot (left). It shows the association between the relative abundance and the magnitude of the difference per sample. Next to that, we can also create a plot that shows the dispersion on the x-axis instead of log-ratio abundance. Red dots represent genera that are differentially abundant (<span class="math inline">\(q \leq 0.1\)</span>) between the 2 groups. Black points are rare taxa and grey ones are abundant taxa. The dashed line represent an effect size of 1. <span class="citation" data-cites="Gloor2016">Gloor, Macklaim, and Fernandes (<a href="#ref-Gloor2016" role="doc-biblioref">2016</a>)</span> provides more information on these plots.</p>
<div class="no-row-height column-margin column-container"></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.plot.html">aldex.plot</a></span><span class="op">(</span><span class="va">aldex_out</span>,</span>
<span>           type <span class="op">=</span> <span class="st">"MA"</span>,</span>
<span>           test <span class="op">=</span> <span class="st">"welch"</span>,</span>
<span>           xlab <span class="op">=</span> <span class="st">"Log-ratio abundance"</span>,</span>
<span>           ylab <span class="op">=</span> <span class="st">"Difference"</span>,</span>
<span>           cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.plot.html">aldex.plot</a></span><span class="op">(</span><span class="va">aldex_out</span>,</span>
<span>           type <span class="op">=</span> <span class="st">"MW"</span>,</span>
<span>           test <span class="op">=</span> <span class="st">"welch"</span>,</span>
<span>           xlab <span class="op">=</span> <span class="st">"Dispersion"</span>,</span>
<span>           ylab <span class="op">=</span> <span class="st">"Difference"</span>,</span>
<span>           cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30_differential_abundance_files/figure-html/plot-aldex2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The evaluation as differential abundant in above plots is based on the corrected p-value. According to the ALDEx2 developers, the safest approach is to identify those features where the 95% CI of the effect size does not cross 0. As we can see in below table, this is not the case for any of the identified genera (see overlap column, which indicates the proportion of overlap). Also, the authors recommend to focus on effect sizes and CIs rather than interpreting the p-value. To keep the comparison simple, we will here use the p-value as decision criterion. But please be aware that the effect size together with the CI is a better answer to the question we are typically interested in.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">aldex_out</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># here we choose the wilcoxon output rather than t-test output</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">wi.eBH</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Genus</span>, <span class="va">we.eBH</span>, <span class="va">wi.eBH</span>, <span class="va">effect</span>, <span class="va">overlap</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 56%">
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 11%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Genus</th>
<th style="text-align: right;">we.eBH</th>
<th style="text-align: right;">wi.eBH</th>
<th style="text-align: right;">effect</th>
<th style="text-align: right;">overlap</th>
</tr></thead>
<tbody><tr class="odd">
<td style="text-align: left;">Genus:[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">0.0942</td>
<td style="text-align: right;">0.0235</td>
<td style="text-align: right;">0.8516</td>
<td style="text-align: right;">0.1224</td>
</tr></tbody>
</table>
</div>
</div>
</section><section id="ancom-bc" class="level3 page-columns page-full" data-number="10.2.3"><h3 data-number="10.2.3" class="anchored" data-anchor-id="ancom-bc">
<span class="header-section-number">10.2.3</span> ANCOM-BC</h3>
<p>The analysis of composition of microbiomes with bias correction (ANCOM-BC) <span class="citation" data-cites="ancombc2020">(<a href="#ref-ancombc2020" role="doc-biblioref">Lin and Peddada 2020</a>)</span> is a recently developed method for differential abundance testing. It is based on an earlier published approach <span class="citation" data-cites="Mandal2015">(<a href="#ref-Mandal2015" role="doc-biblioref">Mandal et al. 2015</a>)</span>. The previous version of ANCOM was among the methods that produced the most consistent results and is probably a conservative approach <span class="citation" data-cites="Nearing2022">(<a href="#ref-Nearing2022" role="doc-biblioref">Nearing et al. 2022</a>)</span>. However, the new ANCOM-BC method operates quite differently compared to the former ANCOM method.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Nearing2022" class="csl-entry" role="listitem">
Nearing, Jacob T., Gavin M. Douglas, Molly G. Hayes, Jocelyn MacDonald, Dhwani K. Desai, Nicole Allward, Casey M. A. Jones, et al. 2022. <span>“Microbiome Differential Abundance Methods Produce Different Results Across 38 Datasets.”</span> <em>Nature Communications</em> 13 (1): 342. <a href="https://doi.org/10.1038/s41467-022-28034-z">https://doi.org/10.1038/s41467-022-28034-z</a>.
</div></div><p>As the only method, ANCOM-BC incorporates the so called <em>sampling fraction</em> into the model. The latter term could be empirically estimated by the ratio of the library size to the microbial load. According to the authors, ignoring variations in this sampling fraction would bias DAA results. Furthermore, this method provides p-values and confidence intervals for each taxon. It also controls the FDR and it is computationally simple to implement.</p>
<p>Note that the original method was implemented in the <code><a href="https://rdrr.io/pkg/ANCOMBC/man/ancombc.html">ancombc()</a></code> function (see <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html">extended tutorial</a>). The method has since then been updated and new features have been added to enable multi-group comparisons and repeated measurements among other improvements. We do not cover the more advanced features of ANCOMBC in this tutorial as these features are documented in detail in this <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html">tutorial</a>.</p>
<p>We now proceed with a simple example. First, we specify a formula. In this formula, other covariates could potentially be included to adjust for confounding. We show this further below. Again, please make sure to check the <a href="https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html">function documentation</a> as well as the linked tutorials to learn about the additional arguments that we specify.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FrederickHuangLin/ANCOMBC">ANCOMBC</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run ANCOM-BC at the genus level and only including the prevalent genera</span></span>
<span><span class="va">ancombc2_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ANCOMBC/man/ancombc2.html">ancombc2</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">tse</span>,</span>
<span>                         assay.type <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                         fix_formula <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>                         p_adj_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>                         prv_cut <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                         group <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>                         struc_zero <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         neg_lb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         <span class="co"># multi group comparison is deactivated automatically</span></span>
<span>                         global <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The object <code>out</code> contains all model output. Again, see the <a href="https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html">documentation of the function</a> under <strong>Value</strong> for details. Our question whether taxa are differentially abundant can be answered by looking at the <code>res</code> object, which contains dataframes with the coefficients, standard errors, p-values and q-values. Below we show the first entries of this dataframe.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># store the FDR adjusted results </span></span>
<span><span class="va">ancombc2_out</span><span class="op">$</span><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">taxon</span>, <span class="va">lfc_patient_statusControl</span>, <span class="va">q_patient_statusControl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">q_patient_statusControl</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">q_patient_statusControl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 35%">
<col style="width: 33%">
<col style="width: 31%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">taxon</th>
<th style="text-align: right;">lfc_patient_statusControl</th>
<th style="text-align: right;">q_patient_statusControl</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Genus:Coprobacter</td>
<td style="text-align: right;">-1.736</td>
<td style="text-align: right;">0.0048</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Ruminococcus_1</td>
<td style="text-align: right;">2.286</td>
<td style="text-align: right;">0.0101</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:uncultured_bacterium</td>
<td style="text-align: right;">-1.375</td>
<td style="text-align: right;">0.0344</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Subdoligranulum</td>
<td style="text-align: right;">1.278</td>
<td style="text-align: right;">0.0365</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section id="maaslin2" class="level3 page-columns page-full" data-number="10.2.4"><h3 data-number="10.2.4" class="anchored" data-anchor-id="maaslin2">
<span class="header-section-number">10.2.4</span> MaAsLin2</h3>
<p>Let us next illustrate MaAsLin2 <span class="citation" data-cites="Mallick2020">(<a href="#ref-Mallick2020" role="doc-biblioref">Mallick, Rahnavard, and McIver 2020</a>)</span>. This method is based on generalized linear models and flexible for different study designs and covariate structures. For details, check their <a href="https://github.com/biobakery/biobakery/wiki/maaslin2">Biobakery tutorial</a>.</p>
<div class="no-row-height column-margin column-container"></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://huttenhower.sph.harvard.edu/maaslin2">Maaslin2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># maaslin expects features as columns and samples as rows </span></span>
<span><span class="co"># for both the abundance table as well as metadata </span></span>
<span></span>
<span><span class="co"># We can specify different GLMs/normalizations/transforms.</span></span>
<span><span class="co"># specifying a ref is especially important if you have more than 2 levels</span></span>
<span><span class="va">maaslin2_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Maaslin2/man/Maaslin2.html">Maaslin2</a></span><span class="op">(</span>input_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                         input_metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                         output <span class="op">=</span> <span class="st">"DAA example"</span>,</span>
<span>                         transform <span class="op">=</span> <span class="st">"AST"</span>,</span>
<span>                         fixed_effects <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>                         <span class="co"># you can also fit MLM by specifying random effects</span></span>
<span>                         <span class="co"># random_effects = c(...),</span></span>
<span>                         reference <span class="op">=</span> <span class="st">"patient_status,Control"</span>,</span>
<span>                         normalization <span class="op">=</span> <span class="st">"TSS"</span>,</span>
<span>                         standardize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                         <span class="co"># filtering was previously performed</span></span>
<span>                         min_prevalence <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Which genera are identified as differentially abundant? (leave out “head” to see all).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">maaslin2_out</span><span class="op">$</span><span class="va">results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 31%">
<col style="width: 12%">
<col style="width: 4%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 15%">
<col style="width: 5%">
<col style="width: 2%">
<col style="width: 9%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">feature</th>
<th style="text-align: left;">metadata</th>
<th style="text-align: left;">value</th>
<th style="text-align: right;">coef</th>
<th style="text-align: right;">stderr</th>
<th style="text-align: right;">pval</th>
<th style="text-align: left;">name</th>
<th style="text-align: right;">qval</th>
<th style="text-align: right;">N</th>
<th style="text-align: right;">N.not.zero</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Genus..Ruminococcus._gauvreauii_group</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">-0.0662</td>
<td style="text-align: right;">0.0131</td>
<td style="text-align: right;">0.0000</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0015</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus.Catabacter</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">0.0288</td>
<td style="text-align: right;">0.0090</td>
<td style="text-align: right;">0.0036</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0347</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus.Faecalibacterium</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">0.1195</td>
<td style="text-align: right;">0.0363</td>
<td style="text-align: right;">0.0030</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0347</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus..Clostridium._innocuum_group</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">-0.0678</td>
<td style="text-align: right;">0.0209</td>
<td style="text-align: right;">0.0033</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0347</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">25</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus.uncultured</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">-0.0450</td>
<td style="text-align: right;">0.0130</td>
<td style="text-align: right;">0.0019</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0347</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">27</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This will create a folder that is called like in the output specified above. It contains also figures to visualize difference between significant taxa.</p>
</section><section id="linda" class="level3 page-columns page-full" data-number="10.2.5"><h3 data-number="10.2.5" class="anchored" data-anchor-id="linda">
<span class="header-section-number">10.2.5</span> LinDA</h3>
<p>Lastly, we cover linear models for differential abundance analysis of microbiome compositional data (<span class="citation" data-cites="Zhou2022">H. Zhou et al. (<a href="#ref-Zhou2022" role="doc-biblioref">2022</a>)</span>). This is very similar to ANCOMBC with few differences: 1) LinDA corrects for the compositional bias differently using the mode of all regression coefficients. 2) it is faster (100x-1000x than ANCOMBC and according to the authors); 3) it supports hierarchical models. The latest ANCOMBC versions are also supporting hierarchical models. Nevertheless, LinDA seems a promising tool that achieves a very good power/fdr trade-off together with ANCOMBC according to the review. The speed improvements might make it critical especially for datasets that have higher sample or feature set sizes.</p>
<div class="no-row-height column-margin column-container"></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MicrobiomeStat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run LinDA</span></span>
<span><span class="va">linda_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MicrobiomeStat/man/linda.html">linda</a></span><span class="op">(</span>feature.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   meta.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   formula <span class="op">=</span> <span class="st">"~ patient_status"</span>,</span>
<span>                   alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                   prev.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                   mean.abund.filter <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">##  0  features are filtered!</span></span>
<span><span class="co">##  The filtered data has  27  samples and  48  features will be tested!</span></span>
<span><span class="co">##  Pseudo-count approach is used.</span></span>
<span><span class="co">##  Fit linear models ...</span></span>
<span><span class="co">##  Completed.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># List genera for which H0 could be rejected:</span></span>
<span><span class="va">linda_out</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">patient_statusControl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">reject</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">stat</span>, <span class="va">padj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"feature"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;">feature</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">padj</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Genus:Catabacter</td>
<td style="text-align: right;">-3.303</td>
<td style="text-align: right;">0.0461</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Faecalibacterium</td>
<td style="text-align: right;">-4.192</td>
<td style="text-align: right;">0.0078</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">4.165</td>
<td style="text-align: right;">0.0078</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section id="zicoseq" class="level3" data-number="10.2.6"><h3 data-number="10.2.6" class="anchored" data-anchor-id="zicoseq">
<span class="header-section-number">10.2.6</span> ZicoSeq</h3>
<p>Subsequently, we demonstrate DAA with ZicoSeq, a method based on linear models and permutation. Further details can be found in <a href="https://cran.r-project.org/web/packages/GUniFrac/vignettes/ZicoSeq.html">this tutorial</a>. This approach has been assessed to exhibit high power and a low false discovery rate, which has the following components:</p>
<ol type="1">
<li>Winsorization to decrease the influence of outliers;</li>
<li>Posterior sampling based on a beta mixture prior to address sampling variability and zero inflation;</li>
<li>Reference-based multiple-stage normalization to address compositional effects;</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GUniFrac</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">zicoseq_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GUniFrac/man/ZicoSeq.html">ZicoSeq</a></span><span class="op">(</span>feature.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       meta.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       grp.name <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>                       feature.dat.type <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>                       return.feature.dat <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       prev.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       mean.abund.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       max.abund.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       perm.no <span class="op">=</span> <span class="fl">999</span><span class="op">)</span></span>
<span><span class="co">##  For sample size less than 40, posterior sampling will not be used!</span></span>
<span><span class="co">##  0  features are filtered!</span></span>
<span><span class="co">##  The data has  27  samples and  48  features will be tested!</span></span>
<span><span class="co">##  On average,  1  outlier counts will be replaced for each feature!</span></span>
<span><span class="co">##  Finding the references ...</span></span>
<span><span class="co">##  Permutation testing ...</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  Completed!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zicoseq_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind.data.frame</a></span><span class="op">(</span>p.raw <span class="op">=</span> <span class="va">zicoseq_out</span><span class="op">$</span><span class="va">p.raw</span>,</span>
<span>                                p.adj.fdr <span class="op">=</span> <span class="va">zicoseq_out</span><span class="op">$</span><span class="va">p.adj.fdr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">zicoseq_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">p.adj.fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.adj.fdr</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">p.raw</th>
<th style="text-align: right;">p.adj.fdr</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Genus:[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.0030</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:[Clostridium]_innocuum_group</td>
<td style="text-align: right;">0.003</td>
<td style="text-align: right;">0.0155</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Faecalibacterium</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.0240</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:uncultured</td>
<td style="text-align: right;">0.003</td>
<td style="text-align: right;">0.0240</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Catabacter</td>
<td style="text-align: right;">0.007</td>
<td style="text-align: right;">0.0316</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## x-axis is the effect size: R2 * direction of coefficient</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GUniFrac/man/ZicoSeq.plot.html">ZicoSeq.plot</a></span><span class="op">(</span>ZicoSeq.obj <span class="op">=</span> <span class="va">zicoseq_out</span>,</span>
<span>             pvalue.type <span class="op">=</span> <span class="st">'p.adj.fdr'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="30_differential_abundance_files/figure-html/plot-zicoseq-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="philr" class="level3" data-number="10.2.7"><h3 data-number="10.2.7" class="anchored" data-anchor-id="philr">
<span class="header-section-number">10.2.7</span> PhILR</h3>
<p>PhILR is a tree-based method that tests group-wise associations based on balances. A detailed introduction to this method is available in <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/philr/inst/doc/philr-intro.html">this Bioconductor tutorial</a>.</p>
</section><section id="comparison-of-methods" class="level3" data-number="10.2.8"><h3 data-number="10.2.8" class="anchored" data-anchor-id="comparison-of-methods">
<span class="header-section-number">10.2.8</span> Comparison of methods</h3>
<p>Although the methods described above yield unidentical results, they are expected to agree on a few differentially abundant taxa. To draw more informed conclusions, it is good practice to compare the outcomes of different methods in terms of found features, their effect sizes and significances, as well as other method-specific aspects. Such comparative approach is outlined in <a href="#compare-daa-methods">this exercise</a>.</p>
</section></section><section id="daa-with-confounding" class="level2 page-columns page-full" data-number="10.3"><h2 data-number="10.3" class="anchored" data-anchor-id="daa-with-confounding">
<span class="header-section-number">10.3</span> DAA with confounding</h2>
<p>Confounders can be defined as variables that are related to and affect the apparent dynamics between the response and the main independent variable. They are common in experimental studies. Generally, they can be classified into 3 groups:</p>
<ul>
<li><p>Biological confounders, such as age and sex</p></li>
<li><p>Technical confounders produced during sample collection, processing and analysis</p></li>
<li><p>Confounders resulting from experimental models, such as batch effects and sample history</p></li>
</ul>
<p>Controlling for confounders is an important practice to reach an unbiased conclusion. To perform causal inference, it is crucial that the method is able to include confounders in the model. This is not possible with statistical tests of general use, such as the Wilcoxon test. In contrast, methods that target DAA, such as those described in this chapter, allow controlling for confounders. In the following examples, we will perform DAA with a main independent variable and a few confounders.</p>
<section id="selecting-confounders" class="level3 page-columns page-full" data-number="10.3.1"><h3 data-number="10.3.1" class="anchored" data-anchor-id="selecting-confounders">
<span class="header-section-number">10.3.1</span> Selecting confounders</h3>
<p>In addition to patient status, we will now control for two confounders: cohort and library size. The former is a categorical variable with three factors, whereas the latter is a discrete numerical variable. Remarkably, most DAA methods accept these two and several other data types.</p>
<p>For demonstration, library size is treated as a confounder and included in the formulas of the DAA methods. Although this is a satisfactory approach to control for uneven sequencing efforts across samples, rarefaction generally represents a better solution <span class="citation" data-cites="Schloss2023">(<a href="#ref-Schloss2023" role="doc-biblioref">Schloss 2023</a>)</span>. With that said, library size can be readily computed and added to the colData.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Schloss2023" class="csl-entry" role="listitem">
Schloss, Patrick D. 2023. <span>“Rarefaction Is Currently the Best Approach to Control for Uneven Sequencing Effort in Amplicon Sequence Analyses.”</span> <em>bioRxiv</em>, 2023–06. <a href="https://doi.org/10.1101/2023.06.23.546313">https://doi.org/10.1101/2023.06.23.546313</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute and store library size in colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">library_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="ancom-bc-1" class="level3" data-number="10.3.2"><h3 data-number="10.3.2" class="anchored" data-anchor-id="ancom-bc-1">
<span class="header-section-number">10.3.2</span> ANCOM-BC</h3>
<p>Here, confounders can be added to the formula along with patient status, the main outcome variable. This way, the model evaluates whether differentially abundant taxa are associated with one of the variables when the other two are kept constant.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># perform the analysis </span></span>
<span><span class="va">ancombc2_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ANCOMBC/man/ancombc2.html">ancombc2</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                         assay.type <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                         fix_formula <span class="op">=</span> <span class="st">"patient_status + cohort + library_size"</span>,</span>
<span>                         p_adj_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>                         lib_cut <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                         group <span class="op">=</span> <span class="st">"patient_status"</span>, </span>
<span>                         struc_zero <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                         neg_lb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                         <span class="co"># multi-group comparison is deactivated automatically</span></span>
<span>                         global <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the output, each taxon is assigned with several effect sizes (lfc, which stands for log-fold change) and adjusted p-values (q). For categorical variables such as patient status and cohort, the statistics indicate whether the abundance of a given taxon is significantly different between the specified group (column name) and the reference group (the group that does not appear in the column names), whereas for numerical variables such as library size, they indicate whether the abundance of a given taxon varies with that variable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ancombc2_out</span><span class="op">$</span><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"taxon"</span>, <span class="st">"lfc"</span>, <span class="st">"q"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">q_patient_statusControl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 14%">
<col style="width: 7%">
<col style="width: 12%">
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">taxon</th>
<th style="text-align: right;">lfc_(Intercept)</th>
<th style="text-align: right;">lfc_patient_statusControl</th>
<th style="text-align: right;">lfc_cohortCohort_2</th>
<th style="text-align: right;">lfc_cohortCohort_3</th>
<th style="text-align: right;">lfc_library_size</th>
<th style="text-align: right;">q_(Intercept)</th>
<th style="text-align: right;">q_patient_statusControl</th>
<th style="text-align: right;">q_cohortCohort_2</th>
<th style="text-align: right;">q_cohortCohort_3</th>
<th style="text-align: right;">q_library_size</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Family:Lachnospiraceae</td>
<td style="text-align: right;">-0.8809</td>
<td style="text-align: right;">-0.8845</td>
<td style="text-align: right;">0.6863</td>
<td style="text-align: right;">0.4128</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.9989</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Hungatella</td>
<td style="text-align: right;">-0.2495</td>
<td style="text-align: right;">-0.8367</td>
<td style="text-align: right;">-0.1362</td>
<td style="text-align: right;">-0.1479</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0619</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Ruminococcaceae_UCG-013</td>
<td style="text-align: right;">-1.0545</td>
<td style="text-align: right;">-0.8559</td>
<td style="text-align: right;">0.6289</td>
<td style="text-align: right;">-0.0700</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0165</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Alistipes</td>
<td style="text-align: right;">-0.0404</td>
<td style="text-align: right;">-0.5549</td>
<td style="text-align: right;">-0.4461</td>
<td style="text-align: right;">-0.0383</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0001</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Akkermansia</td>
<td style="text-align: right;">-1.0334</td>
<td style="text-align: right;">-0.2804</td>
<td style="text-align: right;">0.2866</td>
<td style="text-align: right;">0.4300</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.8241</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Bacteroides</td>
<td style="text-align: right;">-0.4045</td>
<td style="text-align: right;">-1.2689</td>
<td style="text-align: right;">0.1047</td>
<td style="text-align: right;">0.6851</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.6514</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section id="linda-1" class="level3" data-number="10.3.3"><h3 data-number="10.3.3" class="anchored" data-anchor-id="linda-1">
<span class="header-section-number">10.3.3</span> LinDA</h3>
<p>As in the previous method, confounders can be included in the formula with the main outcome variable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">linda_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MicrobiomeStat/man/linda.html">linda</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   formula <span class="op">=</span> <span class="st">"~ patient_status + cohort + library_size"</span>,</span>
<span>                   alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                   prev.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                   mean.abund.filter <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">##  0  features are filtered!</span></span>
<span><span class="co">##  The filtered data has  27  samples and  48  features will be tested!</span></span>
<span><span class="co">##  Imputation approach is used.</span></span>
<span><span class="co">##  Fit linear models ...</span></span>
<span><span class="co">##  Completed.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The model returns an output for every variable included in the formula. Normally, only the results on the main outcome variable are relevant and can be retrieved as shown below. However, the statistics on the confounders can be similarly obtained by accessing the corresponding items from the output object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Select results for the patient status</span></span>
<span><span class="va">linda_res</span> <span class="op">&lt;-</span> <span class="va">linda_out</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">patient_statusControl</span></span>
<span></span>
<span><span class="va">linda_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">reject</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">log2FoldChange</span>, <span class="va">stat</span>, <span class="va">padj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"feature"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;">feature</th>
<th style="text-align: right;">log2FoldChange</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">padj</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Genus:Barnesiella</td>
<td style="text-align: right;">-3.832</td>
<td style="text-align: right;">-3.057</td>
<td style="text-align: right;">0.0397</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Butyricicoccus</td>
<td style="text-align: right;">-2.309</td>
<td style="text-align: right;">-3.056</td>
<td style="text-align: right;">0.0397</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Catabacter</td>
<td style="text-align: right;">-3.266</td>
<td style="text-align: right;">-3.569</td>
<td style="text-align: right;">0.0206</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Erysipelatoclostridium</td>
<td style="text-align: right;">3.797</td>
<td style="text-align: right;">3.090</td>
<td style="text-align: right;">0.0397</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Faecalibacterium</td>
<td style="text-align: right;">-5.873</td>
<td style="text-align: right;">-4.481</td>
<td style="text-align: right;">0.0045</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Ruminococcaceae_UCG-014</td>
<td style="text-align: right;">-3.009</td>
<td style="text-align: right;">-3.575</td>
<td style="text-align: right;">0.0206</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The output shows effect sizes in terms of log-fold changes and a derived statistic (stat) as well as the corresponding adjusted p-values for differences in abundance of each taxon between the control and treated group.</p>
</section><section id="zicoseq-1" class="level3" data-number="10.3.4"><h3 data-number="10.3.4" class="anchored" data-anchor-id="zicoseq-1">
<span class="header-section-number">10.3.4</span> ZicoSeq</h3>
<p>For this method, confounders can be added as a list to the <code>adj.name</code> argument.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">zicoseq_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GUniFrac/man/ZicoSeq.html">ZicoSeq</a></span><span class="op">(</span>feature.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       meta.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       grp.name <span class="op">=</span> <span class="st">"patient_status"</span>,</span>
<span>                       adj.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cohort"</span>, <span class="st">"library_size"</span><span class="op">)</span>, </span>
<span>                       feature.dat.type <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>                       return.feature.dat <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       prev.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       mean.abund.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       max.abund.filter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       perm.no <span class="op">=</span> <span class="fl">999</span><span class="op">)</span></span>
<span><span class="co">##  For sample size less than 40, posterior sampling will not be used!</span></span>
<span><span class="co">##  0  features are filtered!</span></span>
<span><span class="co">##  The data has  27  samples and  48  features will be tested!</span></span>
<span><span class="co">##  On average,  1  outlier counts will be replaced for each feature!</span></span>
<span><span class="co">##  Finding the references ...</span></span>
<span><span class="co">##  Permutation testing ...</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  ...................................................................................................</span></span>
<span><span class="co">##  Completed!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output shows the raw and adjusted p-values for clinical status.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zicoseq_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind.data.frame</a></span><span class="op">(</span>p.raw <span class="op">=</span> <span class="va">zicoseq_out</span><span class="op">$</span><span class="va">p.raw</span>,</span>
<span>                                p.adj.fdr <span class="op">=</span> <span class="va">zicoseq_out</span><span class="op">$</span><span class="va">p.adj.fdr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">zicoseq_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">p.adj.fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">p.raw</th>
<th style="text-align: right;">p.adj.fdr</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Genus:Catabacter</td>
<td style="text-align: right;">0.004</td>
<td style="text-align: right;">0.0196</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Faecalibacterium</td>
<td style="text-align: right;">0.002</td>
<td style="text-align: right;">0.0165</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:Ruminococcaceae_UCG-014</td>
<td style="text-align: right;">0.004</td>
<td style="text-align: right;">0.0312</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:Ruminococcus_2</td>
<td style="text-align: right;">0.007</td>
<td style="text-align: right;">0.0312</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Genus:[Clostridium]_innocuum_group</td>
<td style="text-align: right;">0.003</td>
<td style="text-align: right;">0.0196</td>
</tr>
<tr class="even">
<td style="text-align: left;">Genus:[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.0005</td>
</tr>
</tbody>
</table>
</div>
</div>
</section></section><section id="additional-resources" class="level2 page-columns page-full" data-number="10.4"><h2 data-number="10.4" class="anchored" data-anchor-id="additional-resources">
<span class="header-section-number">10.4</span> Additional resources</h2>
<p>DAA can be performed by several means. Although most of them provide similar functionality, some may be more suitable than others given a certain study design or data type. Commonly used DAA tools include:</p>
<ul>
<li>ALDEx2 <span class="citation" data-cites="Gloor2016">(<a href="#ref-Gloor2016" role="doc-biblioref">Gloor, Macklaim, and Fernandes 2016</a>)</span>
</li>
<li>ANCOM <span class="citation" data-cites="Mandal2015">(<a href="#ref-Mandal2015" role="doc-biblioref">Mandal et al. 2015</a>)</span>
</li>
<li>ANCOMBC <span class="citation" data-cites="ancombc2020">(<a href="#ref-ancombc2020" role="doc-biblioref">Lin and Peddada 2020</a>)</span>
</li>
<li>corncob <span class="citation" data-cites="Martin2021">(<a href="#ref-Martin2021" role="doc-biblioref">Martin, Witten, and Willis 2021</a>)</span>
</li>
<li>DACOMP <span class="citation" data-cites="Brill2019">(<a href="#ref-Brill2019" role="doc-biblioref">Brill, Amnon, and Ruth 2022</a>)</span>
</li>
<li>DESeq2 <span class="citation" data-cites="Love2014">(<a href="#ref-Love2014" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span>
</li>
<li>eBay <span class="citation" data-cites="Liu2020">(<a href="#ref-Liu2020" role="doc-biblioref">Liu, Zhao, and Wang 2020</a>)</span>
</li>
<li>edgeR <span class="citation" data-cites="Chen2016">(<a href="#ref-Chen2016" role="doc-biblioref">Y. Chen, Lun, and Smyth 2016</a>)</span>
</li>
<li>fastANCOM <span class="citation" data-cites="fastANCOM2022">(<a href="#ref-fastANCOM2022" role="doc-biblioref">C. Zhou et al. 2022</a>)</span>
</li>
<li>LDM <span class="citation" data-cites="Hu2020">(<a href="#ref-Hu2020" role="doc-biblioref">Hu and Satten 2020</a>)</span>
</li>
<li>lefser <span class="citation" data-cites="Khlebrodova2021">(<a href="#ref-Khlebrodova2021" role="doc-biblioref">Khleborodova 2021</a>)</span>
</li>
<li>limma <span class="citation" data-cites="Ritchie2015">(<a href="#ref-Ritchie2015" role="doc-biblioref">Ritchie et al. 2015</a>)</span>
</li>
<li>LinDA <span class="citation" data-cites="Zhou2022">(<a href="#ref-Zhou2022" role="doc-biblioref">H. Zhou et al. 2022</a>)</span>
</li>
<li>MaAsLin2 <span class="citation" data-cites="Mallick2020">(<a href="#ref-Mallick2020" role="doc-biblioref">Mallick, Rahnavard, and McIver 2020</a>)</span>
</li>
<li>metagenomeSeq <span class="citation" data-cites="Paulson2017">(<a href="#ref-Paulson2017" role="doc-biblioref">Paulson, Talukder, and Bravo 2017</a>)</span>
</li>
<li>Omnibus <span class="citation" data-cites="Omnibus2018">(<a href="#ref-Omnibus2018" role="doc-biblioref">J. Chen et al. 2018</a>)</span>
</li>
<li>RAIDA <span class="citation" data-cites="Sohn2015">(<a href="#ref-Sohn2015" role="doc-biblioref">Sohn, Du, and An 2015</a>)</span>
</li>
<li>t-test</li>
<li>Wilcoxon test</li>
<li>ZicoSeq <span class="citation" data-cites="Yang2022">(<a href="#ref-Yang2022" role="doc-biblioref">Yang and Chen 2022</a>)</span>
</li>
<li>ZINQ <span class="citation" data-cites="Ling2021">(<a href="#ref-Ling2021" role="doc-biblioref">Ling et al. 2021</a>)</span>
</li>
</ul>


<div class="no-row-height column-margin column-container"><div id="ref-Gloor2016" class="csl-entry" role="listitem">
Gloor, Gregory B., Jean M. Macklaim, and Andrew D. Fernandes. 2016. <span>“Displaying <span>Variation</span> in <span>Large Datasets</span>: <span>Plotting</span> a <span>Visual Summary</span> of <span>Effect Sizes</span>.”</span> <em>Journal of Computational and Graphical Statistics</em> 25 (3): 971–79. <a href="https://doi.org/10.1080/10618600.2015.1131161">https://doi.org/10.1080/10618600.2015.1131161</a>.
</div><div id="ref-Mandal2015" class="csl-entry" role="listitem">
Mandal, Siddhartha, Will Van Treuren, Richard A. White, Merete Eggesbø, Rob Knight, and Shyamal D. Peddada. 2015. <span>“Analysis of Composition of Microbiomes: A Novel Method for Studying Microbial Composition.”</span> <em>Microbial Ecology in Health &amp; Disease</em> 26 (0). <a href="https://doi.org/10.3402/mehd.v26.27663">https://doi.org/10.3402/mehd.v26.27663</a>.
</div><div id="ref-ancombc2020" class="csl-entry" role="listitem">
Lin, Huang, and Shyamal Das Peddada. 2020. <span>“Analysis of Compositions of Microbiomes with Bias Correction.”</span> <em>Nature Communications</em> 11 (1): 1–11. https://doi.org/<a href="https://doi.org/10.1038/s41467-020-17041-7">https://doi.org/10.1038/s41467-020-17041-7</a>.
</div><div id="ref-Martin2021" class="csl-entry" role="listitem">
Martin, Bryan D, Daniela Witten, and Amy D Willis. 2021. <em>Corncob: Count Regression for Correlated Observations with the Beta-Binomial</em>. <a href="https://CRAN.R-project.org/package=corncob">https://CRAN.R-project.org/package=corncob</a>.
</div><div id="ref-Brill2019" class="csl-entry" role="listitem">
Brill, Barak, Amir Amnon, and Heller Ruth. 2022. <span>“Testing for Differential Abundance in Compositional Counts Data, with Application to Microbiome Studies.”</span> <em>The Annals of Applied Statistics</em> 16 (December).
</div><div id="ref-Love2014" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div><div id="ref-Liu2020" class="csl-entry" role="listitem">
Liu, Tiantian, Hongyu Zhao, and Tao Wang. 2020. <span>“An Empirical Bayes Approach to Normalization and Differential Abundance Testing for Microbiome Data.”</span> <em>BMC Bioinformatics</em> 21 (June). <a href="https://doi.org/10.1186/s12859-020-03552-z">https://doi.org/10.1186/s12859-020-03552-z</a>.
</div><div id="ref-Chen2016" class="csl-entry" role="listitem">
Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”</span> <em>F1000Research</em> 5: 1438. <a href="https://doi.org/10.12688/f1000research.8987.2">https://doi.org/10.12688/f1000research.8987.2</a>.
</div><div id="ref-fastANCOM2022" class="csl-entry" role="listitem">
Zhou, Chao, Huimin Wang, Hongyu Zhao, and Tao Wang. 2022. <span>“fastANCOM: A Fast Method for Analysis of Compositions of Microbiomes.”</span> <em>Bioinformatics</em> 38 (March): 2039–41.
</div><div id="ref-Hu2020" class="csl-entry" role="listitem">
Hu, Yijuan, and Glen A Satten. 2020. <span>“Testing Hypotheses about the Microbiome Using the Linear Decomposition Model (LDM).”</span> <em>Bioinformatics</em> 36 (July): 4106--4115.
</div><div id="ref-Khlebrodova2021" class="csl-entry" role="listitem">
Khleborodova, Asya. 2021. <em>Lefser: R Implementation of the LEfSE Method for Microbiome Biomarker Discovery</em>. <a href="https://github.com/waldronlab/lefser">https://github.com/waldronlab/lefser</a>.
</div><div id="ref-Ritchie2015" class="csl-entry" role="listitem">
Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. <span>“<span class="nocase">limma</span> Powers Differential Expression Analyses for <span>RNA</span>-Sequencing and Microarray Studies.”</span> <em>Nucleic Acids Research</em> 43 (7): e47. <a href="https://doi.org/10.1093/nar/gkv007">https://doi.org/10.1093/nar/gkv007</a>.
</div><div id="ref-Zhou2022" class="csl-entry" role="listitem">
Zhou, Huijuan, Kejun He, Jun Chen, and Xianyang Zhang. 2022. <span>“<span>LinDA</span>: Linear Models for Differential Abundance Analysis of Microbiome Compositional Data.”</span> <em>Genome Biology</em> 23 (1): 95. <a href="https://doi.org/10.1186/s13059-022-02655-5">https://doi.org/10.1186/s13059-022-02655-5</a>.
</div><div id="ref-Mallick2020" class="csl-entry" role="listitem">
Mallick, Himel, Ali Rahnavard, and Lauren J. McIver. 2020. <em>MaAsLin 2: Multivariable Association in Population-Scale Meta-Omics Studies.</em> <a href="http://huttenhower.sph.harvard.edu/maaslin2">http://huttenhower.sph.harvard.edu/maaslin2</a>.
</div><div id="ref-Paulson2017" class="csl-entry" role="listitem">
Paulson, JN, H Talukder, and HC Bravo. 2017. <span>“Longitudinal Differential Abundance Analysis of Marker-Gene Surveys Using Smoothing Splines.”</span> <em>Biorxiv</em>. <a href="https://doi.org/10.1101/099457">https://doi.org/10.1101/099457</a>.
</div><div id="ref-Omnibus2018" class="csl-entry" role="listitem">
Chen, Jun, Emily King, Rebecca Deek, Zhi Wei, Yue Yu, Diane Grill, and Karla Ballman. 2018. <span>“An Omnibus Test for Differential Distribution Analysis of Microbiome Sequencing Data.”</span> <em>Bioinformatics</em> 34 (February).
</div><div id="ref-Sohn2015" class="csl-entry" role="listitem">
Sohn, Michael B., Ruofei Du, and Lingling An. 2015. <span>“A Robust Approach for Identifying Differentially Abundant Features in Metagenomic Samples.”</span> <em>Bioinformatics</em> 31 (July). <a href="https://doi.org/10.1093/bioinformatics/btv165">https://doi.org/10.1093/bioinformatics/btv165</a>.
</div><div id="ref-Yang2022" class="csl-entry" role="listitem">
Yang, Lu, and Jun Chen. 2022. <span>“A Comprehensive Evaluation of Microbial Differential Abundance Analysis Methods: Current Status and Potential Solutions.”</span> <em>Microbiome</em> 10 (130): 2049–2618. <a href="https://doi.org/10.1186/s40168-022-01320-0">https://doi.org/10.1186/s40168-022-01320-0</a>.
</div><div id="ref-Ling2021" class="csl-entry" role="listitem">
Ling, Wodan, Zhao Ni, M. Plantinga Anna, J. Launer Lenore, A. Fodor Anthony, A. Meyer Katie, and C. Wu Michael. 2021. <span>“Powerful and Robust Non-Parametric Association Testing for Microbiome Data via a Zero-Inflated Quantile Approach (ZINQ).”</span> <em>Microbiome</em> 181 (September). <a href="https://doi.org/10.1186/s40168-021-01129-3">https://doi.org/10.1186/s40168-021-01129-3</a>.
</div></div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/24_clustering.html" class="pagination-link" aria-label="Community Typing (Clustering)">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/40_machine_learning.html" class="pagination-link" aria-label="Machine Learning">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>