<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 13&nbsp; Network learning and analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/61_network_comparison.html" rel="next">
<link href="../pages/23_multi-assay_analyses.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar docked nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../docs/pages/images/mia_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Orchestrating Microbiome Analysis</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">TEST</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-bi-github">
<li>
    <a class="dropdown-item" href="https://github.com/microbiome/OMA/">
 <span class="dropdown-text">Source Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://microbiome.github.io/OMA/docs/devel/">
 <span class="dropdown-text">Browse version <code>devel</code></span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/60_network_learning.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#network-learning" id="toc-network-learning" class="nav-link active" data-scroll-target="#network-learning"><span class="header-section-number">13.1</span> Network learning</a>
  <ul class="collapse">
<li><a href="#sec-network-learning-workflow" id="toc-sec-network-learning-workflow" class="nav-link" data-scroll-target="#sec-network-learning-workflow"><span class="header-section-number">13.1.1</span> Typical workflow</a></li>
  <li><a href="#install-packages" id="toc-install-packages" class="nav-link" data-scroll-target="#install-packages"><span class="header-section-number">13.1.2</span> Install packages</a></li>
  <li><a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation"><span class="header-section-number">13.1.3</span> Data preparation</a></li>
  <li><a href="#sec-spring-network" id="toc-sec-spring-network" class="nav-link" data-scroll-target="#sec-spring-network"><span class="header-section-number">13.1.4</span> SPRING network</a></li>
  <li><a href="#sec-netcomi-spring" id="toc-sec-netcomi-spring" class="nav-link" data-scroll-target="#sec-netcomi-spring"><span class="header-section-number">13.1.5</span> NetCoMi network</a></li>
  </ul>
</li>
  <li>
<a href="#sec-network-analysis" id="toc-sec-network-analysis" class="nav-link" data-scroll-target="#sec-network-analysis"><span class="header-section-number">13.2</span> Network analysis with igraph</a>
  <ul class="collapse">
<li><a href="#network-plot" id="toc-network-plot" class="nav-link" data-scroll-target="#network-plot"><span class="header-section-number">13.2.1</span> Network plot</a></li>
  <li><a href="#centrality-measures" id="toc-centrality-measures" class="nav-link" data-scroll-target="#centrality-measures"><span class="header-section-number">13.2.2</span> Centrality measures</a></li>
  <li><a href="#scale-node-sizes-by-degree" id="toc-scale-node-sizes-by-degree" class="nav-link" data-scroll-target="#scale-node-sizes-by-degree"><span class="header-section-number">13.2.3</span> Scale node sizes by degree</a></li>
  <li><a href="#degree-distribution" id="toc-degree-distribution" class="nav-link" data-scroll-target="#degree-distribution"><span class="header-section-number">13.2.4</span> Degree distribution</a></li>
  <li><a href="#clustered-heatmaps" id="toc-clustered-heatmaps" class="nav-link" data-scroll-target="#clustered-heatmaps"><span class="header-section-number">13.2.5</span> Clustered heatmaps</a></li>
  <li><a href="#global-network-measures" id="toc-global-network-measures" class="nav-link" data-scroll-target="#global-network-measures"><span class="header-section-number">13.2.6</span> Global network measures</a></li>
  </ul>
</li>
  <li><a href="#network-analysis-with-netcomi" id="toc-network-analysis-with-netcomi" class="nav-link" data-scroll-target="#network-analysis-with-netcomi"><span class="header-section-number">13.3</span> Network analysis with NetCoMi</a></li>
  <li>
<a href="#network-visualization" id="toc-network-visualization" class="nav-link" data-scroll-target="#network-visualization"><span class="header-section-number">13.4</span> Network visualization</a>
  <ul class="collapse">
<li><a href="#highlight-node-properties" id="toc-highlight-node-properties" class="nav-link" data-scroll-target="#highlight-node-properties"><span class="header-section-number">13.4.1</span> Highlight node properties</a></li>
  <li><a href="#highlight-data-features" id="toc-highlight-data-features" class="nav-link" data-scroll-target="#highlight-data-features"><span class="header-section-number">13.4.2</span> Highlight data features</a></li>
  </ul>
</li>
  <li><a href="#which-methods-to-choose" id="toc-which-methods-to-choose" class="nav-link" data-scroll-target="#which-methods-to-choose"><span class="header-section-number">13.5</span> Which method(s) to choose?</a></li>
  <li><a href="#sec-more-about-association" id="toc-sec-more-about-association" class="nav-link" data-scroll-target="#sec-more-about-association"><span class="header-section-number">13.6</span> More about association measures</a></li>
  <li>
<a href="#comparison-of-association-measures" id="toc-comparison-of-association-measures" class="nav-link" data-scroll-target="#comparison-of-association-measures"><span class="header-section-number">13.7</span> Comparison of association measures</a>
  <ul class="collapse">
<li><a href="#sec-sparcc-correlation" id="toc-sec-sparcc-correlation" class="nav-link" data-scroll-target="#sec-sparcc-correlation"><span class="header-section-number">13.7.1</span> SparCC</a></li>
  <li><a href="#sec-shrinkage-prop" id="toc-sec-shrinkage-prop" class="nav-link" data-scroll-target="#sec-shrinkage-prop"><span class="header-section-number">13.7.2</span> Shrinkage proportionality</a></li>
  <li><a href="#sec-spieceasi-mb" id="toc-sec-spieceasi-mb" class="nav-link" data-scroll-target="#sec-spieceasi-mb"><span class="header-section-number">13.7.3</span> SpiecEasi - MB</a></li>
  <li><a href="#network-plots" id="toc-network-plots" class="nav-link" data-scroll-target="#network-plots"><span class="header-section-number">13.7.4</span> Network plots</a></li>
  <li><a href="#network-analysis" id="toc-network-analysis" class="nav-link" data-scroll-target="#network-analysis"><span class="header-section-number">13.7.5</span> Network analysis</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/60_network_learning.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-network-learning" class="quarto-section-identifier"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>Learning and analyzing microbial association networks is another common exploratory data analysis approach aimed at understanding the complex interplay of microbial communities in their natural habitat. Microbial networks consist of nodes, representing microbial species or taxa, and edges, expressing their association. The mathematical representation of a network is the adjacency matrix, which has a non-zero entry for each edge in the network.</p>
<p>A typical workflow for estimating a microbial network involves several steps, including data preprocessing, estimating microbial associations, and transforming them into edge weights. The resulting network can be analyzed using local network properties, such as centrality measures, or global measures that describe the overall structure of the network. Network plots provide further insight into the microbial community structure and allow for exploratory analysis of microbial relationships.</p>
<p>In this chapter, we go through the complete workflow of constructing and analyzing a single microbial network, step by step. A potential next analysis task is to compare networks between groups, such as patients and controls, different environmental conditions, or different time points. How to compare networks between two groups is explained in <a href="61_network_comparison.html" class="quarto-xref"><span>Chapter 14</span></a>.</p>
<section id="network-learning" class="level2 page-columns page-full" data-number="13.1"><h2 data-number="13.1" class="anchored" data-anchor-id="network-learning">
<span class="header-section-number">13.1</span> Network learning</h2>
<section id="sec-network-learning-workflow" class="level3 page-columns page-full" data-number="13.1.1"><h3 data-number="13.1.1" class="anchored" data-anchor-id="sec-network-learning-workflow">
<span class="header-section-number">13.1.1</span> Typical workflow</h3>
<p><a href="#fig-network-workflow" class="quarto-xref">Figure&nbsp;<span>13.1</span></a> shows the workflow for learning/constructing a microbial association network as proposed by <span class="citation" data-cites="peschel2021netcomi">Peschel et al. (<a href="#ref-peschel2021netcomi" role="doc-biblioref">2021</a>)</span>. The respective steps are explained below.</p>
<div class="no-row-height column-margin column-container"></div><div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-network-workflow" class="quarto-figure quarto-figure-center quarto-float anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-network-workflow-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/FigureNetworkLearning.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-network-workflow-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;13.1: The typical input is a <span class="math inline">\(p\)</span> x <span class="math inline">\(n\)</span> dimensional count matrix coming from a sequencing process, where <span class="math inline">\(n\)</span> is the number of samples and <span class="math inline">\(p\)</span> the number of features / ASVs / OTUs. Steps 1 through 6 are explained below. Each matrix resulting from steps 4, 5, and 6 plays a specific role in the final network: The adjacency matrix is used for edge colors, dissimilarity for layout, and similarity for edge weights. In weighted networks, the similarity matrix equals the adjacency matrix.
</figcaption></figure>
</div>
</div>
</div>
<ol type="1">
<li><p><strong>Zero replacement:</strong> Since the following steps usually require non-zero entries in the read count matrix, zero counts must be replaced. A simple solution is to add a pseudo count to the data. Other possible approaches are implemented in the R package <a href="https://cran.r-project.org/web/packages/zCompositions/index.html">zCompositions</a></p></li>
<li><p><strong>Normalization:</strong> To avoid compositional effects, the data are normalized using a compositionality aware transformation. A common approach is the centered log-ratio (clr) transformation, which moves the data from a <span class="math inline">\(p\)</span>-dimensional simplex to Euclidean space so that standard statistical analysis methods are valid. A variance stabilizing transformation (vst) is also a suitable approach for normalizing microbial count data <span class="citation" data-cites="badri2020shrinkage">(<a href="#ref-badri2020shrinkage" role="doc-biblioref">Badri et al. 2020</a>)</span>.</p></li>
<li>
<p><strong>Association estimation:</strong> This is the crucial step in network learning to obtain statistical relations between the taxa. Common association measures include correlation, conditional dependence (which we will equate to partial correlation), and proportionality. Further information on these three types of association and their application can be found in <a href="#sec-more-about-association" class="quarto-xref"><span>Section 13.6</span></a>. The following list gives a selection of compositionality aware approaches:</p>
<ul>
<li>
<strong>Compositionality aware correlation estimation methods:</strong>
<ul>
<li>Pearson’s correlation coefficient (+ normalization)</li>
<li>Spearman’s rank correlation coefficient (+ normalization)</li>
<li>Covariance shrinkage (<a href="https://strimmerlab.github.io/software/corpcor/">corpcor</a> package) (+ normalization)</li>
<li>SparCC (implemented in <a href="https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html">SpiecEasi</a>); applied in <a href="#sec-sparcc-correlation" class="quarto-xref"><span>Section 13.7.1</span></a>
</li>
<li>CCREPE (<a href="https://bioconductor.org/packages/release/bioc/html/ccrepe.html">ccrepe</a> package)</li>
<li>CCLasso (<a href="https://github.com/huayingfang/CCLasso">R code on GitHub</a>)</li>
</ul>
</li>
<li>
<strong>Compositionality aware measures of conditional dependence / partial correlation:</strong>
<ul>
<li>
<a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a> with Meinshausen and Bühlmann (MB) neighborhood selection; applied in <a href="#sec-spieceasi-mb" class="quarto-xref"><span>Section 13.7.3</span></a>
</li>
<li>
<a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a> with the graphical lasso (glasso)</li>
<li>gCoda (<a href="https://github.com/huayingfang/gCoda">R code on GitHub</a>)</li>
<li>
<a href="https://github.com/GraceYoon/SPRING">SPRING</a>; applied in <a href="#sec-spring-network" class="quarto-xref"><span>Section 13.1.4</span></a>
</li>
</ul>
</li>
<li>
<strong>Proportionality measures (proportionality aware by definition):</strong>
<ul>
<li><a href="https://github.com/tpq/propr">propr</a></li>
<li>
<a href="https://github.com/MichelleBadri/NormCorr-manuscript/blob/master/code/helpers/norm_functions.R">Shrinkage proportionality estimator</a>; applied in <a href="#sec-shrinkage-prop" class="quarto-xref"><span>Section 13.7.2</span></a>
</li>
</ul>
</li>
</ul>
</li>
<li><p><strong>Sparsification:</strong> Transforming the estimated associations directly into adjacencies would lead to a dense network where all nodes are connected and only weighted network measures are meaningful. Therefore, the association matrix is usually sparsified to select edges of interest. A common sparsification approach for correlations is thresholding, where correlations with a magnitude below the threshold are set to zero. Another possibility is a statistical test (Student’s t-test or permutation test) with the null hypothesis that the correlation is equal to zero. SpiecEasi uses the StARS stability selection approach <span class="citation" data-cites="liu2010stability">(<a href="#ref-liu2010stability" role="doc-biblioref">Liu, Roeder, and Wasserman 2010</a>)</span> to decide on an appropriate sparsification level of the inferred conditional dependence graph.</p></li>
<li>
<p><strong>Transformation into dissimilarity:</strong> A common next step is to simply use the absolute values of the sparsified associations as edge weights. In this way, correlations of high magnitude (both positive and negative) will have a high edge weight. From a biological point of view, it would also make sense to assign a low edge weight to taxa that are strongly negatively associated, which would correspond to a high dissimilarity value. Here we follow <span class="citation" data-cites="vanDongen2012metric">Dongen and Enright (<a href="#ref-vanDongen2012metric" role="doc-biblioref">2012</a>)</span> to directly transform the sparse associations <span class="math inline">\(r_{ij}^*\)</span> into dissimilarities, which can later be used for shortest path network measures. Depending on the desired handling of negative associations, one of the two proposed transformations should be chosen:</p>
<p>5a: <strong>“signed”:</strong> <span class="math inline">\(d_{ij} = \sqrt{0.5(1-r^*_{ij})}\)</span>, where strongly negatively associated taxa have the largest distance and are placed further apart in the network.</p>
<p>5b: <strong>“unsigned”:</strong> <span class="math inline">\(d_{ij} = \sqrt{1-{r_{ij}^*}^2}\)</span>, resulting in a small distance between strongly associated taxa (regardless of the sign).</p>
</li>
<li><p><strong>Transformation into similarity / edge weight:</strong> Finally, the dissimilarities are transformed into similarities by <span class="math inline">\(s_{ij} = 1 - d_{ij}\)</span>, which are used as edge weights. Thus, the similarity matrix is equal to the adjacency matrix in a weighted network.</p></li>
</ol>
<div class="no-row-height column-margin column-container"></div><p>The main association measure used in this chapter is the <a href="https://github.com/GraceYoon/SPRING">SPRING</a> (“Semi-Parametric Rank-based approach for INference in Graphical model”) method proposed by <span class="citation" data-cites="yoon2019microbial">Yoon, Gaynanova, and Müller (<a href="#ref-yoon2019microbial" role="doc-biblioref">2019</a>)</span>. <code>SPRING</code> learns conditional dependency graphs for compositional data and follows the neighborhood selection method introduced by <span class="citation" data-cites="meinshausen2006high">Meinshausen and Bühlmann (<a href="#ref-meinshausen2006high" role="doc-biblioref">2006</a>)</span> (“MB”). We will show how to apply the method directly, as well as how to use it in conjunction with the R package <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a>, which is specifically designed for the construction and analysis of networks for microbiome data.</p>
<p>See <a href="#sec-more-about-association" class="quarto-xref"><span>Section 13.6</span></a> for a comparison of all three association types (correlation, partial correlation, and proportionality) with more information on each measure and applications.</p>
<p>We demonstrate the workflow using the the <strong>PeerJ data set</strong> <span class="citation" data-cites="potbhare2022skin">(<a href="#ref-potbhare2022skin" role="doc-biblioref">Potbhare et al. 2022</a>)</span>. It contains skin microbial profiles of 58 subjects.</p>
<div class="no-row-height column-margin column-container"><div id="ref-potbhare2022skin" class="csl-entry" role="listitem">
Potbhare, Renuka, Ameeta RaviKumar, Eveliina Munukka, Leo Lahti, and Richa Ashma. 2022. <span>“Skin Microbiota Diversity Among Genetically Unrelated Individuals of Indian Origin.”</span> <em><span>PeerJ</span></em> 10: e13075. <a href="https://doi.org/10.7717/peerj.13075">https://doi.org/10.7717/peerj.13075</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"peerj13075"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse0</span> <span class="op">&lt;-</span> <span class="va">peerj13075</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">tse0</span><span class="op">)</span></span>
<span><span class="co">##  [1] 674  58</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="install-packages" class="level3" data-number="13.1.2"><h3 data-number="13.1.2" class="anchored" data-anchor-id="install-packages">
<span class="header-section-number">13.1.2</span> Install packages</h3>
<p>Three packages used in this chapter are available on GitHub only: <code>SpiecEasi</code>, <code>SPRING</code>, and <code>NetCoMi</code>. We recommend that you install these packages before proceeding.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">SpiecEasi</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"zdk123/SpiecEasi"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">SPRING</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"GraceYoon/SPRING"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">NetCoMi</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"stefpeschel/NetCoMi"</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, ref <span class="op">=</span> <span class="st">"TSE"</span>,</span>
<span>                           dependencies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"LinkingTo"</span><span class="op">)</span>,</span>
<span>                           repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"https://cloud.r-project.org/"</span>,</span>
<span>                                     <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="data-preparation" class="level3" data-number="13.1.3"><h3 data-number="13.1.3" class="anchored" data-anchor-id="data-preparation">
<span class="header-section-number">13.1.3</span> Data preparation</h3>
<p>Before applying the network learning methods, we perform some data preparation steps:</p>
<ul>
<li>Aggregation to genus level</li>
<li>Add relative abundance assay</li>
<li>Prevalence filtering (keep genera with prevalence &gt; 20%)</li>
<li>Add assay with log10 transformed abundances</li>
<li>Add assay with clr transformed abundances</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Agglomerate to genus level</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">tse0</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Add relative abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, </span>
<span>                      assay.type <span class="op">=</span> <span class="st">"counts"</span>, </span>
<span>                      method <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>                      MARGIN <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Filter by prevalence</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalent</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                            prevalence <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                            detection <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                            assay.type <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add log10-transformed abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"log10"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add clr-transformed abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] 147  58</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-spring-network" class="level3 page-columns page-full" data-number="13.1.4"><h3 data-number="13.1.4" class="anchored" data-anchor-id="sec-spring-network">
<span class="header-section-number">13.1.4</span> SPRING network</h3>
<p>As explained in <a href="#sec-network-learning-workflow" class="quarto-xref"><span>Section 13.1.1</span></a>, we use <a href="https://github.com/GraceYoon/SPRING">SPRING</a> (“Semi-Parametric Rank-based approach for INference in Graphical model”) as association measure. We first use the <code>SPRING</code> function directly to construct a conditional dependency graph.</p>
<p>Neither zero replacement nor normalization (steps 1 and 2 in our workflow) are required because SPRING uses a modified clr (mclr) transformation that can handle zero counts, and the correlation estimation method itself can also deal with zeros in the data. mclr is similar to the clr transformation except that mclr considers only the non-zero values. More precisely, the geometric mean is derived from positive values only, and zero counts remain zero after the transformation. This approach is similar to the “robust clr” (rclr) transformation included in the <code>vegan</code> package, except that mclr applies a positive shift to all non-zero values to make them strictly positive. See <span class="citation" data-cites="yoon2019microbial">(<a href="#ref-yoon2019microbial" role="doc-biblioref">Yoon, Gaynanova, and Müller 2019</a>)</span> for details.</p>
<div class="no-row-height column-margin column-container"><div id="ref-yoon2021fast" class="csl-entry" role="listitem">
Yoon, Grace, Christian L Müller, and Irina Gaynanova. 2021. <span>“Fast Computation of Latent Correlations.”</span> <em><span>Journal of Computational and Graphical Statistics</span></em> 30 (4): 1249–56.
</div></div><p>The <code>Rmethod</code> argument is set to “approx” to estimate the correlations using a hybrid multi-linear interpolation approach proposed by <span class="citation" data-cites="yoon2021fast">Yoon, Müller, and Gaynanova (<a href="#ref-yoon2021fast" role="doc-biblioref">2021</a>)</span>. This method considerably reduces the runtime while controlling the approximation error.</p>
<p><code>SPRING</code> uses the StARS (“Stability Approach to Regularization Selection”) method <span class="citation" data-cites="liu2010stability">(<a href="#ref-liu2010stability" role="doc-biblioref">Liu, Roeder, and Wasserman 2010</a>)</span> to obtain a sparse association matrix. Thus, also step 4 of our workflow is already included. We set the StARS threshold to 0.05 to get a sparser graph.</p>
<div class="no-row-height column-margin column-container"></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SPRING</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">13075</span><span class="op">)</span></span>
<span><span class="va">spring_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPRING/man/SPRING.html">SPRING</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                     Rmethod <span class="op">=</span> <span class="st">"approx"</span>, </span>
<span>                     thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                     lambdaseq <span class="op">=</span> <span class="st">"data-specific"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get index of the optimal lambda selected by StARS</span></span>
<span><span class="va">opt.K</span> <span class="op">&lt;-</span> <span class="va">spring_est</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">stars</span><span class="op">$</span><span class="va">opt.index</span></span>
<span></span>
<span><span class="co"># Store partial correlation matrix belonging to the optimal lambda as matrix</span></span>
<span><span class="va">spring_cor</span> <span class="op">&lt;-</span> <span class="fu">SpiecEasi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/symBeta.html">symBeta</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">spring_est</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">est</span><span class="op">$</span><span class="va">beta</span><span class="op">[[</span><span class="va">opt.K</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spring_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As explained in <a href="#sec-network-learning-workflow" class="quarto-xref"><span>Section 13.1.1</span></a>, the estimated associations are sparsified, transformed into dissimilarities, and finally transformed into similarities, which are the adjacency values. We write a function for these steps, which will be reused later.</p>
<p>Since <code>SPRING</code> already includes a sparsification approach, the <code>thresh</code> argument is not needed here, but will be needed in <a href="#sec-more-about-association" class="quarto-xref"><span>Section 13.6</span></a> for other association measures.</p>
<p>To be consistent with the workflow, we provide two dissimilarity transformations: “signed” and “unsigned” (see <a href="#sec-network-learning-workflow" class="quarto-xref"><span>Section 13.1.1</span></a> for an explanation). These transformations were introduced by <span class="citation" data-cites="vanDongen2012metric">Dongen and Enright (<a href="#ref-vanDongen2012metric" role="doc-biblioref">2012</a>)</span>. We use the “signed” transformation in our examples so that strongly negatively associated genera have low edge weights.</p>
<div class="no-row-height column-margin column-container"><div id="ref-vanDongen2012metric" class="csl-entry" role="listitem">
Dongen, Stijn van, and Anton J. Enright. 2012. <span>“<span class="nocase">Metric distances derived from cosine similarity and Pearson and Spearman correlations</span>.”</span> <em><span>arXiv Preprint</span></em> 2: 2–6. <a href="http://arxiv.org/abs/1208.3145">http://arxiv.org/abs/1208.3145</a>.
</div></div><p>The output of the function is an <code>igraph</code> object, which can be plotted and analyzed using functions from the <code>igraph</code> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Arguments:</span></span>
<span><span class="co"># - assoMat: association matrix</span></span>
<span><span class="co"># - threshold: associations below the threshold are set to zero</span></span>
<span><span class="co"># - dissTrans: dissimilarity transformation ("signed" or "unsigned")</span></span>
<span></span>
<span><span class="va">transform_asso</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">assoMat</span>, <span class="va">thresh</span> <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">dissTrans</span> <span class="op">=</span> <span class="st">"signed"</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Sparsification</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html">is.null</a></span><span class="op">(</span><span class="va">thresh</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">assoMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">assoMat</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">thresh</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Compute dissimilarity matrix</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">dissTrans</span> <span class="op">==</span> <span class="st">"signed"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">dissMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fl">0.5</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">assoMat</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">dissMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">assoMat</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Dissimilarity between nodes with zero correlation is set to 1</span></span>
<span>  <span class="co"># (these nodes are unconnected and thus should have maximum dissimilarity)</span></span>
<span>  <span class="va">dissMat</span><span class="op">[</span><span class="va">assoMat</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>  </span>
<span>  <span class="co"># Compute similarity matrix</span></span>
<span>  <span class="va">simMat</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">dissMat</span></span>
<span>  </span>
<span>  <span class="co"># Turn into igraph object</span></span>
<span>  <span class="va">graphObj</span> <span class="op">&lt;-</span> <span class="fu">SpiecEasi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/adj2igraph.html">adj2igraph</a></span><span class="op">(</span><span class="va">simMat</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graphObj</span>, adja <span class="op">=</span> <span class="va">simMat</span>, asso <span class="op">=</span> <span class="va">assoMat</span>, diss <span class="op">=</span> <span class="va">dissMat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create graph object</span></span>
<span><span class="va">spring_graph</span> <span class="op">&lt;-</span> <span class="fu">transform_asso</span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span><span class="op">$</span><span class="va">graph</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-netcomi-spring" class="level3 page-columns page-full" data-number="13.1.5"><h3 data-number="13.1.5" class="anchored" data-anchor-id="sec-netcomi-spring">
<span class="header-section-number">13.1.5</span> NetCoMi network</h3>
<p>The <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> <span class="citation" data-cites="peschel2021netcomi">(<a href="#ref-peschel2021netcomi" role="doc-biblioref">Peschel et al. 2021</a>)</span> package is specifically designed to construct, analyze, and compare networks for microbiome data and implements the complete workflow described in <a href="#sec-network-learning-workflow" class="quarto-xref"><span>Section 13.1.1</span></a>. Instead of using several functions for each of the steps, <code>NetCoMi</code> provides a single function for network construction (<code><a href="https://rdrr.io/pkg/NetCoMi/man/netConstruct.html">netConstruct()</a></code>), so the package streamlines the workflow considerably. The user can choose from a variety of methods for data preprocessing, association estimation, sparsification, and transformation. The returned <code>microNet</code> object can then be passed to <code><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">netAnalyze()</a></code> (the network analysis function) so that all necessary information is available for the network analysis workflow.</p>
<div class="no-row-height column-margin column-container"><div id="ref-peschel2021netcomi" class="csl-entry" role="listitem">
Peschel, Stefanie, Christian L Müller, Erika von Mutius, Anne-Laure Boulesteix, and Martin Depner. 2021. <span>“<span class="nocase">NetCoMi: network construction and comparison for microbiome data in R</span>.”</span> <em><span>Briefings in Bioinformatics</span></em> 22 (4): bbaa290. <a href="https://doi.org/10.1093/bib/bbaa290">https://doi.org/10.1093/bib/bbaa290</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">NetCoMi</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We again use <code>SPRING</code> as one of the association measures available in NetCoMi to construct a conditional dependency graph.</p>
<p>To demonstrate how taxa are filtered with <code><a href="https://rdrr.io/pkg/NetCoMi/man/netConstruct.html">netConstruct()</a></code>, we will use the unfiltered <code>tse</code> object this time. The filtering is the same as before: Taxa occurring in less than 20% of the samples are removed.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">netcomi_net</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                            taxRank <span class="op">=</span> <span class="st">"genus"</span>,</span>
<span>                            filtTax <span class="op">=</span> <span class="st">"numbSamp"</span>,</span>
<span>                            filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>numbSamp <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,</span>
<span>                            measure <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>                            measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                              Rmethod <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span>,</span>
<span>                            sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                            dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>                            seed <span class="op">=</span> <span class="fl">13075</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code><a href="https://rdrr.io/pkg/NetCoMi/man/netConstruct.html">netConstruct()</a></code> returns an object of the class <code>microNet</code>, which contains all matrices generated during network construction.</p>
<p>The object also contains an edge list, giving each edge’s estimated association, dissimilarity, and adjacency. Let’s take a quick look at the edges with the highest and lowest edge weights:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">edgelist</span> <span class="op">&lt;-</span> <span class="va">netcomi_net</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">netcomi_net</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">$</span><span class="va">adja</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">edgelist</span><span class="op">)</span></span>
<span><span class="co">##                  v1                    v2   asso   diss   adja</span></span>
<span><span class="co">##  73     Citrobacter           Escherichia 0.3682 0.5621 0.4379</span></span>
<span><span class="co">##  63    Buttiauxella              Serratia 0.2426 0.6154 0.3846</span></span>
<span><span class="co">##  69   Chitinivibrio Pseudogracilibacillus 0.2203 0.6244 0.3756</span></span>
<span><span class="co">##  19        Algicola           Siccibacter 0.2193 0.6248 0.3752</span></span>
<span><span class="co">##  111     Haliangium          Marinobacter 0.2148 0.6266 0.3734</span></span>
<span><span class="co">##  143 Planomicrobium         Virgibacillus 0.2006 0.6322 0.3678</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">edgelist</span><span class="op">)</span></span>
<span><span class="co">##                 v1             v2       asso   diss   adja</span></span>
<span><span class="co">##  132 Mycobacterium Salinibacillus  0.0017483 0.7065 0.2935</span></span>
<span><span class="co">##  24      Amphritea    Providencia  0.0014085 0.7066 0.2934</span></span>
<span><span class="co">##  102       Erwinia    Siccibacter  0.0013114 0.7066 0.2934</span></span>
<span><span class="co">##  116     Holophaga   Methylarcula  0.0007828 0.7068 0.2932</span></span>
<span><span class="co">##  17       Algicola     Lysobacter  0.0005191 0.7069 0.2931</span></span>
<span><span class="co">##  95   Enterobacter     Janibacter -0.0013921 0.7076 0.2924</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As before, the adjacency matrix is converted into an <code>igraph</code> object. Further steps like sparsification and transformation are not necessary because they are done internally by <code><a href="https://rdrr.io/pkg/NetCoMi/man/netConstruct.html">netConstruct()</a></code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">netcomi_graph</span> <span class="op">&lt;-</span> <span class="fu">SpiecEasi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/adj2igraph.html">adj2igraph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">netcomi_net</span><span class="op">$</span><span class="va">adjaMat1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="sec-network-analysis" class="level2" data-number="13.2"><h2 data-number="13.2" class="anchored" data-anchor-id="sec-network-analysis">
<span class="header-section-number">13.2</span> Network analysis with igraph</h2>
<p>The computed network is now analyzed using appropriate methods. We will first use the <code>igraph</code> package to analyze the <code>SPRING</code> network. NetCoMi’s <code><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">netAnalyze()</a></code> function will be used later to analyze the constructed <code>microNet</code> object.</p>
<section id="network-plot" class="level3" data-number="13.2.1"><h3 data-number="13.2.1" class="anchored" data-anchor-id="network-plot">
<span class="header-section-number">13.2.1</span> Network plot</h3>
<p>To get an overview of the network structure, a first common analysis method is to plot the network. We here use the <a href="https://r.igraph.org/">igraph</a> package, which is a state-of-the-art package for network analysis and visualization. Other packages that could be used for network plotting are the <a href="https://rdrr.io/cran/qgraph/">qgraph</a> package or the <a href="https://briatte.github.io/ggnet/">ggnet2</a> package. Since we will use igraph for network analysis later on, we are using its plotting function here as well.</p>
<p>We use the Fruchterman-Reingold layout (a force-directed layout) for node placement. By placing strongly connected nodes close together and those with low edge weight far apart, this layout results in an easy-to-read network plot.</p>
<p>The node sizes are proportional to a taxon’s log10-transformed abundance, which we previously added to the <code>tse</code> object, averaged across all samples. The values are rescaled to be visually distinguishable.</p>
<p>Since we created two graph objects, one with SPRING and one with NetCoMi, we plot them side by side. The two plots should be identical.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/">igraph</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Node sizes</span></span>
<span><span class="va">vsize</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"log10"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fl">3</span></span>
<span></span>
<span><span class="co"># Fruchterman-Reingold layout from igraph package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">13075</span><span class="op">)</span></span>
<span><span class="va">lay_fr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/layout_with_fr.html">layout_with_fr</a></span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_graph</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"SPRING network"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">netcomi_graph</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"NetCoMi network\n(with SPRING associations)"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/plots_spring_netcomi-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section><section id="centrality-measures" class="level3" data-number="13.2.2"><h3 data-number="13.2.2" class="anchored" data-anchor-id="centrality-measures">
<span class="header-section-number">13.2.2</span> Centrality measures</h3>
<p>Centrality measures express the importance of nodes within the network. Common measures are the degree, betweenness, closeness, and eigenvector centrality. The <code>igraph</code> package provides functions to compute these measures. We wrap a function around the code to reuse it later.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">get_centr</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">graph_obj</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># We access igraph directly with "::" because there are more packages loaded in </span></span>
<span>  <span class="co"># this chapter that contain a degree() function.</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Degree <span class="op">=</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/degree.html">degree</a></span><span class="op">(</span><span class="va">graph_obj</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">Betweenness</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/betweenness.html">betweenness</a></span><span class="op">(</span><span class="va">graph_obj</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">Closeness</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/closeness.html">closeness</a></span><span class="op">(</span><span class="va">graph_obj</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">Eigenvector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/eigen_centrality.html">eigen_centrality</a></span><span class="op">(</span><span class="va">graph_obj</span><span class="op">)</span><span class="op">$</span><span class="va">vector</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">centr_df</span> <span class="op">&lt;-</span> <span class="fu">get_centr</span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">centr_df</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">##                   Degree Betweenness Closeness Eigenvector</span></span>
<span><span class="co">##  Abyssicoccus          0           0       NaN   9.278e-18</span></span>
<span><span class="co">##  Acidaminococcus       2           0    0.6464   1.862e-01</span></span>
<span><span class="co">##  Acinetobacter         3           5    0.5544   1.257e-01</span></span>
<span><span class="co">##  Actinomyces           0           0       NaN   9.278e-18</span></span>
<span><span class="co">##  Actinoplanes          2          99    0.5104   1.099e-02</span></span>
<span><span class="co">##  Aerococcus            0           0       NaN   9.278e-18</span></span>
<span><span class="co">##  Aeromonas             4         351    0.7477   2.023e-01</span></span>
<span><span class="co">##  Agromyces             6         435    0.7733   5.293e-01</span></span>
<span><span class="co">##  Algicola              4         264    0.7206   1.500e-01</span></span>
<span><span class="co">##  Alicyclobacillus      0           0       NaN   9.278e-18</span></span>
<span><span class="co">##  Alteribacillus        0           0       NaN   9.278e-18</span></span>
<span><span class="co">##  Ammoniibacillus       1           0    0.5064   5.419e-02</span></span>
<span><span class="co">##  Amphritea             5         382    0.6256   1.335e-02</span></span>
<span><span class="co">##  Amycolatopsis         1           0    0.6117   1.338e-01</span></span>
<span><span class="co">##  Anaerococcus          2         392    0.4064   1.288e-03</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The closeness centrality is “NaN” for some genera. These are unconnected nodes, as can be seen by the zero degree and betweenness centrality.</p>
</section><section id="scale-node-sizes-by-degree" class="level3" data-number="13.2.3"><h3 data-number="13.2.3" class="anchored" data-anchor-id="scale-node-sizes-by-degree">
<span class="header-section-number">13.2.3</span> Scale node sizes by degree</h3>
<p>Centrality measures can be visualized in the network plot by scaling the node sizes according to one of these measures. We plot the <code>Spring</code> graph using the same layout as before and with the node sizes scaled according to all four centrality measures.</p>
<p>Of the four centrality measures, only the degree has a range suitable to be used as node size. The other centrality measures must be rescaled because their range is either too small or too large. The following scaling is a suggestion that works for this example. The values might be adapted for other data sets.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">get_vsizes</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">[</span>, <span class="st">"Betweenness"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>, <span class="st">"Betweenness"</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">[</span>, <span class="st">"Closeness"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span>, <span class="st">"Closeness"</span><span class="op">]</span> <span class="op">*</span> <span class="fl">10</span></span>
<span>  <span class="va">df</span><span class="op">[</span>, <span class="st">"Eigenvector"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span>, <span class="st">"Eigenvector"</span><span class="op">]</span> <span class="op">*</span> <span class="fl">10</span></span>
<span>  <span class="va">df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html">is.infinite</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">vsize_df</span> <span class="op">&lt;-</span> <span class="fu">get_vsizes</span><span class="op">(</span><span class="va">centr_df</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">vsize_df</span><span class="op">)</span></span>
<span><span class="co">##                  Degree Betweenness Closeness Eigenvector</span></span>
<span><span class="co">##  Abyssicoccus         0       0.000     0.000   9.278e-17</span></span>
<span><span class="co">##  Acidaminococcus      2       0.000     6.464   1.862e+00</span></span>
<span><span class="co">##  Acinetobacter        3       1.609     5.544   1.257e+00</span></span>
<span><span class="co">##  Actinomyces          0       0.000     0.000   9.278e-17</span></span>
<span><span class="co">##  Actinoplanes         2       4.595     5.104   1.099e-01</span></span>
<span><span class="co">##  Aerococcus           0       0.000     0.000   9.278e-17</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_graph</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize_df</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/plots_centrality-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>We observe that the two-node component at the bottom has a much higher closeness centrality than the nodes belonging to the main component of the network. Obviously, closeness centrality as commonly defined is misleading when the network consists of disconnected components. Nodes belonging to smaller components are seen as closer to others than in larger components. To overcome this problem, centrality values, and especially closeness centrality, are often calculated only for the largest connected component (LCC), which we will do below.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Extract the LCC</span></span>
<span><span class="va">dg_net</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/decompose.graph.html">decompose.graph</a></span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span><span class="va">idx_lcc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dg_net</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/V.html">V</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lcc</span> <span class="op">&lt;-</span> <span class="va">dg_net</span><span class="op">[[</span><span class="va">idx_lcc</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Compute centrality values for the LCC</span></span>
<span><span class="va">centr_df_lcc</span> <span class="op">&lt;-</span> <span class="fu">get_centr</span><span class="op">(</span><span class="va">lcc</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace centrality values by those for LCC and set all others to zero</span></span>
<span><span class="va">lcc_nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">centr_df_lcc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">centr_df</span><span class="op">[</span><span class="va">lcc_nodes</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="va">centr_df_lcc</span></span>
<span><span class="va">centr_df</span><span class="op">[</span><span class="op">-</span><span class="va">lcc_nodes</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># Node/vertex sizes</span></span>
<span><span class="va">vsize_df</span> <span class="op">&lt;-</span> <span class="fu">get_vsizes</span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_graph</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize_df</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">centr_df</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/plots_centrality_lcc-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>Note that <code>NetCoMi</code> follows a different approach to overcome this problem. <code>NetCoMi</code> uses the definition of closeness centrality proposed by <a href="https://toreopsahl.com/2010/03/20/closeness-centrality-in-networks-with-disconnected-components/">Tore Opsahl</a>, which is well defined even for disconnected networks and assigns higher closeness centrality values to nodes in larger components. This is more intuitive because nodes in a larger component are connected to a larger number of other nodes than in small components.</p>
</section><section id="degree-distribution" class="level3" data-number="13.2.4"><h3 data-number="13.2.4" class="anchored" data-anchor-id="degree-distribution">
<span class="header-section-number">13.2.4</span> Degree distribution</h3>
<p>The degree distribution is another popular measure that expresses the probability distribution of degrees over the entire network. It thus provides insight into the overall network structure. We plot the degree distribution for all four association estimation methods to compare the network structure.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute degree distribution</span></span>
<span><span class="va">ddist</span><span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/degree.distribution.html">degree.distribution</a></span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Data frame needed for ggplot2</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Degree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">ddist</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 Fraction <span class="op">=</span> <span class="va">ddist</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Degree</span>, y <span class="op">=</span> <span class="va">Fraction</span>, group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/degree_dist_spring-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The network has a large number of singletons and sparsely connected nodes, and only a small number of nodes with a higher degree of 7 or more.</p>
</section><section id="clustered-heatmaps" class="level3" data-number="13.2.5"><h3 data-number="13.2.5" class="anchored" data-anchor-id="clustered-heatmaps">
<span class="header-section-number">13.2.5</span> Clustered heatmaps</h3>
<p>Using the <code>ComplexHeatmap</code> package, we plot a heatmap of the association matrix estimated with <code>SPRING</code>. Rows and columns are sorted according to the clusters identified via hierarchical clustering.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We select the 50 nodes with the highest sum of edge weights to get a smaller heatmap.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">adja_sel</span> <span class="op">&lt;-</span> <span class="va">spring_cor</span><span class="op">[</span><span class="va">sel</span>, <span class="va">sel</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Color vector</span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"royalblue4"</span>, <span class="st">"lightblue"</span>, <span class="st">"white"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">adja_sel</span>, </span>
<span>        col <span class="op">=</span> <span class="va">col</span>, </span>
<span>        rect_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"gray"</span>, lwd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>        show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        name <span class="op">=</span> <span class="st">"Association"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/clustered_heatmap-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="556"></p>
</figure>
</div>
</div>
</div>
<p>The associations are generally quite low, and there are no prominent clusters detected by hierarchical clustering.</p>
</section><section id="global-network-measures" class="level3" data-number="13.2.6"><h3 data-number="13.2.6" class="anchored" data-anchor-id="global-network-measures">
<span class="header-section-number">13.2.6</span> Global network measures</h3>
<p>Global measures describe the overall network structure. We take a look at three common measures: density, transitivity, and average path length. The values are again computed with <code>igraph</code> functions.</p>
<section id="density" class="level4" data-number="13.2.6.1"><h4 data-number="13.2.6.1" class="anchored" data-anchor-id="density">
<span class="header-section-number">13.2.6.1</span> Density</h4>
<p>Definition: Proportion of present edges from all possible edges.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://r.igraph.org/reference/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0.01444</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="transitivity-clustering-coefficient" class="level4" data-number="13.2.6.2"><h4 data-number="13.2.6.2" class="anchored" data-anchor-id="transitivity-clustering-coefficient">
<span class="header-section-number">13.2.6.2</span> Transitivity (clustering coefficient)</h4>
<p>Here, we consider only the global clustering coefficient, which is defined as the ratio of triangles to connected triples.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://r.igraph.org/reference/transitivity.html">transitivity</a></span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0.1456</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="average-path-length" class="level4" data-number="13.2.6.3"><h4 data-number="13.2.6.3" class="anchored" data-anchor-id="average-path-length">
<span class="header-section-number">13.2.6.3</span> Average path length</h4>
<p>Definition: Mean of the shortest distance between each pair of nodes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://r.igraph.org/reference/average.path.length.html">average.path.length</a></span><span class="op">(</span><span class="va">spring_graph</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1.699</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section></section><section id="network-analysis-with-netcomi" class="level2" data-number="13.3"><h2 data-number="13.3" class="anchored" data-anchor-id="network-analysis-with-netcomi">
<span class="header-section-number">13.3</span> Network analysis with NetCoMi</h2>
<p>The <code>netcomi_net</code> object of class <code>microNet</code> created before is now passed to <code><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">netAnalyze()</a></code> to perform network analysis with <code>NetCoMi</code>.</p>
<p>The function computes several common network characteristics such as centrality measures, cluster assignment, the graphlet correlation matrix, as well as global network measures.</p>
<p>The user has several options to choose from, such as a clustering method, how to define hubs, and whether or not to normalize centrality values. See the help page <code><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">?netAnalyze</a></code> for a description of the arguments.</p>
<p>By default, a heatmap of the Graphlet Correlation Matrix (GCM) is returned (with graphlet correlations in the upper triangle and significance codes resulting from Student’s t-test in the lower triangle). See <code><a href="https://rdrr.io/pkg/NetCoMi/man/calcGCM.html">?calcGCM</a></code> and <code><a href="https://rdrr.io/pkg/NetCoMi/man/testGCM.html">?testGCM</a></code> for details.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">netcomi_netprops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">netcomi_net</span>, </span>
<span>                               clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                               hubPar <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>                               normDeg <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/netAnalyze_single-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">netcomi_netprops</span>, numbNodes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>##  
##  Component sizes
##  ```````````````              
##  size: 103 2  1
##     #:   1 1 42
##  ______________________________
##  Global network properties
##  `````````````````````````
##  Largest connected component (LCC):
##                                   
##  Relative LCC size         0.70068
##  Clustering coefficient    0.21253
##  Modularity                0.70495
##  Positive edge percentage 99.35065
##  Edge density              0.02932
##  Natural connectivity      0.01173
##  Vertex connectivity       1.00000
##  Edge connectivity         1.00000
##  Average dissimilarity*    0.99084
##  Average path length**     3.81101
##  
##  Whole network:
##                                   
##  Number of components     44.00000
##  Clustering coefficient    0.21253
##  Modularity                0.70787
##  Positive edge percentage 99.35484
##  Edge density              0.01444
##  Natural connectivity      0.00782
##  -----
##  *: Dissimilarity = 1 - edge weight
##  **: Path length = Units with average dissimilarity
##  
##  ______________________________
##  Clusters
##  - In the whole network
##  - Algorithm: cluster_fast_greedy
##  ```````````````````````````````` 
##                                    
##  name:  0  1  2  3  4 5  6 7 8 9 10
##     #: 42 16 11 14 20 9 17 8 5 3  2
##  
##  ______________________________
##  Hubs
##  - In alphabetical/numerical order
##  - Based on empirical quantiles of centralities
##  ```````````````````````````````````````````````                      
##   Agromyces            
##   Aneurinibacillus     
##   Anoxybacillus        
##   Chitinivibrio        
##   Erwinia              
##   Geobacillus          
##   Janibacter           
##   Pseudogracilibacillus
##  
##  ______________________________
##  Centrality measures
##  - In decreasing order
##  - Centrality of disconnected components is zero
##  ````````````````````````````````````````````````
##  Degree (unnormalized):
##                         
##  Anoxybacillus        10
##  Erwinia              10
##  Chitinivibrio         9
##  Janibacter            8
##  Escherichia/Shigella  7
##  
##  Betweenness centrality (normalized):
##                      
##  Janibacter    0.3941
##  Chitinivibrio 0.3137
##  Enterobacter  0.2780
##  Buttiauxella  0.2454
##  Erwinia       0.2095
##  
##  Closeness centrality (normalized):
##                              
##  Chitinivibrio         0.5247
##  Janibacter            0.5067
##  Anoxybacillus         0.4841
##  Pseudogracilibacillus 0.4659
##  Erwinia               0.4653
##  
##  Eigenvector centrality (normalized):
##                              
##  Anoxybacillus         1.0000
##  Chitinivibrio         0.9831
##  Pseudogracilibacillus 0.8657
##  Janibacter            0.8361
##  Aneurinibacillus      0.6890</code></pre>
</div>
<p><strong>Interpretation of some findings:</strong></p>
<ul>
<li>The largest connected component (LCC) has 103 nodes and the network contains 42 singletons.</li>
<li>10 clusters have been identified, containing 2 to 20 nodes.</li>
<li>There are 8 hub nodes detected, which by definition are the nodes with the highest eigenvector centrality.</li>
<li>The average path length in the LCC is 3.811. This means that on average it takes 3.811 steps (step length is the average dissimilarity) to get from one node to another. Note that the average path length in <code>NetCoMi</code> is defined differently than in the <code>igraph</code> package, which is why the values differ.</li>
<li>Low values of edge density and the connectivity measures indicate that the network is rather sparse and not robust to perturbations (i.e., removal of nodes or edges).</li>
</ul></section><section id="network-visualization" class="level2" data-number="13.4"><h2 data-number="13.4" class="anchored" data-anchor-id="network-visualization">
<span class="header-section-number">13.4</span> Network visualization</h2>
<p>Further insight into the network structure can be gained by visualizing the network. We have already seen examples of how to plot a network using the <code>igraph</code> package. Here we will use NetCoMi’s plot function. It takes as input the <code>microNetProps</code> object returned by <code><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">netAnalyze()</a></code>, which contains all computed network properties. This has the advantage that the user can choose which properties to plot by simply changing some arguments. The plot function is based on <a href="https://rdrr.io/cran/qgraph/">qgraph</a>, which is another state-of-the-art R package for network visualization. The help page can be accessed via <code><a href="https://rdrr.io/pkg/NetCoMi/man/plot.microNetProps.html">?plot.microNetProps</a></code>.</p>
<section id="highlight-node-properties" class="level3" data-number="13.4.1"><h3 data-number="13.4.1" class="anchored" data-anchor-id="highlight-node-properties">
<span class="header-section-number">13.4.1</span> Highlight node properties</h3>
<p>In the first plot, node colors represent the detected clusters and node sizes are scaled by eigenvector centrality. Hub nodes are highlighted by default. Singletons are not included in the plot. To improve the readability, NetCoMi’s “intelligent” label shortening approach is used.</p>
<p>Note that nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion argument until you find a value where the labels are legible, but also the clusters are still well recognizable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">netcomi_netprops</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.98</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"intelligent"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"gray40"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network properties highlighted"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/network_plot_single_cluster-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1248"></p>
</figure>
</div>
</div>
</div>
</section><section id="highlight-data-features" class="level3" data-number="13.4.2"><h3 data-number="13.4.2" class="anchored" data-anchor-id="highlight-data-features">
<span class="header-section-number">13.4.2</span> Highlight data features</h3>
<p>We now color nodes according to their phylum. The node sizes are proportional to a taxon’s sum of mclr-transformed abundances. As already mentioned in <a href="#sec-spring-network" class="quarto-xref"><span>Section 13.1.4</span></a>, this is the normalization method used by <code>SPRING</code>. A color palette from <code>RColorBrewer</code> is used here.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">phylum</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">genus</span></span>
<span></span>
<span><span class="co"># Create color vector</span></span>
<span><span class="va">colvec</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span><span class="op">)</span>, <span class="st">"Set3"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">netcomi_netprops</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.98</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"intelligent"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"mclr"</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"feature"</span>, </span>
<span>     featVecCol <span class="op">=</span> <span class="va">phyla</span>, </span>
<span>     colorVec <span class="op">=</span>  <span class="va">colvec</span>,</span>
<span>     nodeTransp <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     highlightHubs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Data features highlighted"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add legends</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span><span class="va">col_transp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/colToTransp.html">colToTransp</a></span><span class="op">(</span><span class="va">colvec</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.8</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, pt.cex <span class="op">=</span> <span class="fl">2.5</span>, title <span class="op">=</span> <span class="st">"Phylum:"</span>, </span>
<span>       legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col_transp</span>, bty <span class="op">=</span> <span class="st">"n"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/network_plot_single_phylum-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1440"></p>
</figure>
</div>
</div>
</div>
<p><strong>A few things to observe:</strong></p>
<ul>
<li>Genera belonging to the same phylum tend to cluster together, though not perfectly.</li>
<li>Genera with a low total count play a rather unimportant role in the network, i.e., they have a low centrality.</li>
<li>There is only one negative edge in the network. This edge is between two clusters, as expected when using the “signed” transformation.</li>
</ul></section></section><section id="which-methods-to-choose" class="level2 page-columns page-full" data-number="13.5"><h2 data-number="13.5" class="anchored" data-anchor-id="which-methods-to-choose">
<span class="header-section-number">13.5</span> Which method(s) to choose?</h2>
<p>Throughout all the steps from primary data to potentially significant network features, there is a variety of methods and parameters to choose from. However, there is no general consensus in the community on the “right” way to estimate and analyze microbial networks. In the absence of a “best method” for inferring and analyzing microbial networks, researchers may be tempted to try different methods and report only the optimal results or those that fit some prior knowledge. This carries the risk of “overfitting” the analysis to the existing data so that the results are not replicable for new data <span class="citation" data-cites="ullmann2023over">(<a href="#ref-ullmann2023over" role="doc-biblioref">Ullmann et al. 2023</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-ullmann2023over" class="csl-entry" role="listitem">
Ullmann, Theresa, Stefanie Peschel, Philipp Finger, Christian L Müller, and Anne-Laure Boulesteix. 2023. <span>“Over-Optimism in Unsupervised Microbiome Analysis: Insights from Network Learning and Clustering.”</span> <em><span>PLoS Computational Biology</span></em> 19 (1): e1010820. <a href="https://doi.org/10.1371/journal.pcbi.1010820">https://doi.org/10.1371/journal.pcbi.1010820</a>.
</div></div><p>Therefore, the selection of the workflow building blocks should be set up once and independently of any hypothesis about the data, thus avoiding the fallacy of starting to “fish” for results that best fit a previously formulated hypothesis. For example, one should ask prior to the analysis whether correlation or conditional dependence as a measure of association better fits the research question and choose the method accordingly. Another example is the choice of transformation from estimated association to dissimilarity (i.e., “signed” or “unsigned”), which completely changes the interpretation and characteristics of the network. This choice should be made based on the research question before starting the analysis.</p>
</section><section id="sec-more-about-association" class="level2 page-columns page-full" data-number="13.6"><h2 data-number="13.6" class="anchored" data-anchor-id="sec-more-about-association">
<span class="header-section-number">13.6</span> More about association measures</h2>
<p>As mentioned in the introduction of this chapter, there are three types of association measures that are commonly used to express relationships between taxa: correlation, conditional dependence, and proportionality. Below, we provide a brief explanation of each of these measures, along with lists of available compositionality-aware approaches.</p>
<ul>
<li>
<strong>Correlation:</strong> Two popular measures of ecological association are Pearson’s correlation coefficient and Spearman’s rank correlation coefficient, both of which can be inferred from empirical (sample) covariances. However, in the <span class="math inline">\(p\gg n\)</span> setting, which most microbiome datasets are in, sample covariances and correlations are unreliable because the parameters being estimated are typically underdetermined. One way to improve sample covariance estimates is to assume that the underlying covariance matrix is sparse and use a regularized covariance estimator to implement this structural assumption. The Schäfer-Strimmer shrinkage estimator <span class="citation" data-cites="schafer2005shrinkage">(<a href="#ref-schafer2005shrinkage" role="doc-biblioref">Schäfer and Strimmer 2005</a>)</span> is one possible method for estimating a sparse correlation matrix. Other popular methods, especially designed to estimate correlations for compositional data, are SparCC (“Sparse Correlations for Compositional data”) by <span class="citation" data-cites="Friedman2012">J. Friedman and Alm (<a href="#ref-Friedman2012" role="doc-biblioref">2012</a>)</span>, CCREPE (“Compositionality Corrected by REnormalization and PErmutation”) by <span class="citation" data-cites="faust2012microbial">Faust et al. (<a href="#ref-faust2012microbial" role="doc-biblioref">2012</a>)</span>, and CCLasso (“Correlation inference for Compositional data through Lasso”) by <span class="citation" data-cites="fang2015cclasso">Huaying Fang et al. (<a href="#ref-fang2015cclasso" role="doc-biblioref">2015</a>)</span>. The latter three methods already include a compositionality aware normalization, and SparCC also includes a zero replacement approach.</li>
<li>
<strong>Conditional dependence:</strong> Since standard correlations include both direct and indirect dependencies, conditional dependence or partial correlation is often preferred for measuring association. Unlike (marginal) correlation, it expresses the relationship between two features conditioned on all other features in the data set. The approach and R package <a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a> (“Sparse InversE Covariance estimation for Ecological Association and Statistical Inference”) by <span class="citation" data-cites="Kurtz2015">Kurtz et al. (<a href="#ref-Kurtz2015" role="doc-biblioref">2015</a>)</span> is specifically designed for inferring ecological networks from microbiome data and includes two approaches for estimating conditional dependence structures between taxa: Neighborhood Selection; short “MB” <span class="citation" data-cites="meinshausen2006high">(<a href="#ref-meinshausen2006high" role="doc-biblioref">Meinshausen and Bühlmann 2006</a>)</span> and (inverse) covariance selection <span class="citation" data-cites="friedman2008sparse">(<a href="#ref-friedman2008sparse" role="doc-biblioref">Jerome Friedman, Hastie, and Tibshirani 2008</a>)</span>, which is based on a penalized maximum likelihood approach and is also known as “graphical lasso”. Another approach and R package for inferring partial correlations from microbiome data is <a href="https://github.com/GraceYoon/SPRING">SPRING</a> (“Semi-Parametric Rank-based approach for INference in Graphical model”) by <span class="citation" data-cites="yoon2019microbial">Yoon, Gaynanova, and Müller (<a href="#ref-yoon2019microbial" role="doc-biblioref">2019</a>)</span>. They also use the MB neighborhood selection method, but introduce a novel semi-parametric rank-based approach for sparse partial correlation estimation that can naturally handle the excess of zeros in the data. gCoda <span class="citation" data-cites="fang2017gcoda">(<a href="#ref-fang2017gcoda" role="doc-biblioref">H. Fang et al. 2017</a>)</span> is another conditional dependence measure based on penalized maximum likelihood estimation. All of the aforementioned conditional dependence measures address the high dimensionality of microbiome data.</li>
<li>
<strong>Proportionality:</strong> <span class="citation" data-cites="lovell2015proportionality">Lovell et al. (<a href="#ref-lovell2015proportionality" role="doc-biblioref">2015</a>)</span> introduce proportionality as an alternative measure of pairwise association for compositional data. The idea is that if the relative abundances between two taxa <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are proportional, then their corresponding absolute abundances are also proportional: <span class="math inline">\(\frac{\omega_i}{m} \propto \frac{\omega_j}{m} \Rightarrow \omega_i \propto \omega_j\)</span>, where <span class="math inline">\(m\)</span> is the sum of counts in the sample. It follows that proportionality is identical for the observed (relative) read counts and the true unobserved counts. The proportionality measure proposed by <span class="citation" data-cites="lovell2015proportionality">Lovell et al. (<a href="#ref-lovell2015proportionality" role="doc-biblioref">2015</a>)</span> is based on log-ratio variance <span class="math inline">\(var(log \frac{x_i}{x_j})\)</span>, which is zero when <span class="math inline">\(\omega_i\)</span> and <span class="math inline">\(\omega_j\)</span> are perfectly proportional. Proportionality is implemented in the R package <a href="https://github.com/tpq/propr">propr</a>. <span class="citation" data-cites="badri2020shrinkage">Badri et al. (<a href="#ref-badri2020shrinkage" role="doc-biblioref">2020</a>)</span> extend the proportionality measure to a so-called “shrinkage proportionality estimator”. It combines proportionality with the covariance shrinkage approach to obtain consistent association estimates even with small sample sizes.</li>
</ul><div class="no-row-height column-margin column-container"><div id="ref-schafer2005shrinkage" class="csl-entry" role="listitem">
Schäfer, Juliane, and Korbinian Strimmer. 2005. <span>“<span class="nocase">A shrinkage approach to large-scale covariance matrix estimation and implications for functional genomics</span>.”</span> <em><span>Statistical Applications in Genetics and Molecular Biology</span></em> 4 (1): 1–30. <a href="https://doi.org/10.2202/1544-6115.1175">https://doi.org/10.2202/1544-6115.1175</a>.
</div><div id="ref-faust2012microbial" class="csl-entry" role="listitem">
Faust, Karoline, J. Fah Sathirapongsasuti, Jacques Izard, Nicola Segata, Dirk Gevers, Jeroen Raes, and Curtis Huttenhower. 2012. <span>“<span class="nocase">Microbial Co-occurrence Relationships in the Human Microbiome</span>.”</span> <em><span>PLoS Computational Biology</span></em> 8 (7): e1002606. <a href="https://doi.org/10.1371/journal.pcbi.1002606">https://doi.org/10.1371/journal.pcbi.1002606</a>.
</div><div id="ref-fang2015cclasso" class="csl-entry" role="listitem">
Fang, Huaying, Chengcheng Huang, Hongyu Zhao, and Minghua Deng. 2015. <span>“<span class="nocase">CCLasso: Correlation inference for compositional data through Lasso</span>.”</span> <em>Bioinformatics</em> 31 (19): 3172–80. <a href="https://doi.org/10.1093/bioinformatics/btv349">https://doi.org/10.1093/bioinformatics/btv349</a>.
</div><div id="ref-friedman2008sparse" class="csl-entry" role="listitem">
Friedman, Jerome, Trevor Hastie, and Robert Tibshirani. 2008. <span>“<span class="nocase">Sparse inverse covariance estimation with the graphical lasso</span>.”</span> <em><span>Biostatistics (Oxford, England)</span></em> 9 (3): 432–41. <a href="https://doi.org/10.1093/BIOSTATISTICS/KXM045">https://doi.org/10.1093/BIOSTATISTICS/KXM045</a>.
</div><div id="ref-yoon2019microbial" class="csl-entry" role="listitem">
Yoon, Grace, Irina Gaynanova, and Christian L Müller. 2019. <span>“<span class="nocase">Microbial networks in SPRING - Semi-parametric rank-based correlation and partial correlation estimation for quantitative microbiome data</span>.”</span> <em><span>Frontiers in Genetics</span></em> 10: 516. <a href="https://doi.org/10.3389/fgene.2019.00516">https://doi.org/10.3389/fgene.2019.00516</a>.
</div><div id="ref-fang2017gcoda" class="csl-entry" role="listitem">
Fang, H, C Huang, H Zhao, and M Deng. 2017. <span>“<span class="nocase">gCoda: Conditional Dependence Network Inference for Compositional Data</span>.”</span> <em><span>Journal of Computational Biology</span></em> 24 (7): 699–708. <a href="https://doi.org/10.1089/cmb.2017.0054">https://doi.org/10.1089/cmb.2017.0054</a>.
</div></div></section><section id="comparison-of-association-measures" class="level2 page-columns page-full" data-number="13.7"><h2 data-number="13.7" class="anchored" data-anchor-id="comparison-of-association-measures">
<span class="header-section-number">13.7</span> Comparison of association measures</h2>
<p>In this section, we provide three additional examples for constructing a network using each of the three types of association:</p>
<ul>
<li>Correlation using <code>SparCC</code>
</li>
<li>Partial correlation using <code>SpiecEasi</code>
</li>
<li>Proportionality using the shrinkage proportionality measure</li>
</ul>
<section id="sec-sparcc-correlation" class="level3 page-columns page-full" data-number="13.7.1"><h3 data-number="13.7.1" class="anchored" data-anchor-id="sec-sparcc-correlation">
<span class="header-section-number">13.7.1</span> SparCC</h3>
<p>The first association measure we look at is SparCC (“Sparse Correlations for Compositional data”), introduced by <span class="citation" data-cites="Friedman2012">J. Friedman and Alm (<a href="#ref-Friedman2012" role="doc-biblioref">2012</a>)</span>. It estimates Pearson correlations while taking into account the compositional structure of the data. The <a href="https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html">SpiecEasi</a> package provides an implementation of this method.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Friedman2012" class="csl-entry" role="listitem">
Friedman, J, and EJ Alm. 2012. <span>“Inferring Correlation Networks from Genomic Survey Data.”</span> <em><span>PLoS Computational Biology</span></em> 8 (9): e1002687. <a href="https://doi.org/10.1371/journal.pcbi.1002687">https://doi.org/10.1371/journal.pcbi.1002687</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">13075</span><span class="op">)</span></span>
<span><span class="co"># Compute correlation matrix</span></span>
<span><span class="va">sparcc_cor</span> <span class="op">&lt;-</span> <span class="fu">SpiecEasi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/sparcc.html">sparcc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">Cor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sparcc_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sparcc_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We reuse the <code>transform_asso()</code> function created in <a href="#sec-spring-network" class="quarto-xref"><span>Section 13.1.4</span></a>, which sparsifies the association matrix, transforms it into a similarity matrix, and finally returns an <code>igraph</code> object.</p>
<p>Two threshold values are used to see the effect of sparsification later in the network plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sparcc_trans03</span> <span class="op">&lt;-</span> <span class="fu">transform_asso</span><span class="op">(</span><span class="va">sparcc_cor</span>, thresh <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="va">sparcc_trans04</span> <span class="op">&lt;-</span> <span class="fu">transform_asso</span><span class="op">(</span><span class="va">sparcc_cor</span>, thresh <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sparcc_graph03</span> <span class="op">&lt;-</span> <span class="va">sparcc_trans03</span><span class="op">$</span><span class="va">graph</span></span>
<span><span class="va">sparcc_graph04</span> <span class="op">&lt;-</span> <span class="va">sparcc_trans04</span><span class="op">$</span><span class="va">graph</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-shrinkage-prop" class="level3 page-columns page-full" data-number="13.7.2"><h3 data-number="13.7.2" class="anchored" data-anchor-id="sec-shrinkage-prop">
<span class="header-section-number">13.7.2</span> Shrinkage proportionality</h3>
<p>In the second example, microbial associations are measured by proportionality, originally introduced by <span class="citation" data-cites="lovell2015proportionality">Lovell et al. (<a href="#ref-lovell2015proportionality" role="doc-biblioref">2015</a>)</span>. We use the shrinkage proportionality estimator proposed by <span class="citation" data-cites="badri2020shrinkage">Badri et al. (<a href="#ref-badri2020shrinkage" role="doc-biblioref">2020</a>)</span>, which gives consistent results even for small sample sizes. Since there is no R package implementing this estimator, we use the <code>rho_shrink_est()</code> function provided in the <a href="https://github.com/MichelleBadri/NormCorr-manuscript/blob/master/code/helpers/norm_functions.R">GitHub repository</a> associated with the paper. The function is slightly modified to take normalized counts as input.</p>
<div class="no-row-height column-margin column-container"><div id="ref-lovell2015proportionality" class="csl-entry" role="listitem">
Lovell, David, Vera Pawlowsky-Glahn, Juan José Egozcue, Samuel Marguerat, and Jürg Bähler. 2015. <span>“<span class="nocase">Proportionality: A Valid Alternative to Correlation for Relative Data</span>.”</span> <em><span>PLoS Computational Biology</span></em> 11 (3): 1–12. <a href="https://doi.org/10.1371/journal.pcbi.1004075">https://doi.org/10.1371/journal.pcbi.1004075</a>.
</div><div id="ref-badri2020shrinkage" class="csl-entry" role="listitem">
Badri, Michelle, Zachary D. Kurtz, Richard Bonneau, and Christian L. Müller. 2020. <span>“<span class="nocase">Shrinkage improves estimation of microbial associations under different normalization methods</span>.”</span> <em><span>NAR Genomics and Bioinformatics</span></em> 2 (4). <a href="https://doi.org/10.1093/NARGAB/LQAA100">https://doi.org/10.1093/NARGAB/LQAA100</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://strimmerlab.github.io/software/corpcor/">corpcor</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># norm_counts: clr-transformed count matrix with samples in rows</span></span>
<span><span class="va">rho_shrink_est</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">norm_counts</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">shrunk_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/corpcor/man/cov.shrink.html">cov.shrink</a></span><span class="op">(</span><span class="va">norm_counts</span>, <span class="va">...</span><span class="op">)</span></span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">norm_counts</span><span class="op">)</span></span>
<span>  <span class="va">J</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">shrunk_cov</span><span class="op">)</span>, <span class="va">p</span><span class="op">)</span>, <span class="va">p</span><span class="op">)</span></span>
<span>  <span class="va">rho</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">shrunk_cov</span> <span class="op">/</span> <span class="op">(</span><span class="va">J</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">J</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">(</span><span class="va">rho</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Apply the shrinkage proportionality estimator to the clr-transformed counts</span></span>
<span><span class="va">prop_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="fu">rho_shrink_est</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"clr"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span><span class="co">##  Estimating optimal shrinkage intensity lambda.var (variance vector): 0.0857 </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Estimating optimal shrinkage intensity lambda (correlation matrix): 0.3634</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again, we use our transformation function to convert the association matrix into a graph object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop_trans</span> <span class="op">&lt;-</span> <span class="fu">transform_asso</span><span class="op">(</span><span class="va">prop_est</span>, thresh <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="va">prop_graph</span> <span class="op">&lt;-</span> <span class="va">prop_trans</span><span class="op">$</span><span class="va">graph</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-spieceasi-mb" class="level3 page-columns page-full" data-number="13.7.3"><h3 data-number="13.7.3" class="anchored" data-anchor-id="sec-spieceasi-mb">
<span class="header-section-number">13.7.3</span> SpiecEasi - MB</h3>
<p>As third example, we use the SpiecEasi (“Sparse InversE Covariance estimation for Ecological Association and Statistical Inference”) approach proposed by <span class="citation" data-cites="Kurtz2015">Kurtz et al. (<a href="#ref-Kurtz2015" role="doc-biblioref">2015</a>)</span> to estimate a sparse conditional dependency graph. The neighborhood selection method (“MB”) introduced by <span class="citation" data-cites="meinshausen2006high">Meinshausen and Bühlmann (<a href="#ref-meinshausen2006high" role="doc-biblioref">2006</a>)</span> is used for network learning. The approach is implemented in the R package <a href="https://github.com/zdk123/SpiecEasi">SpiecEasi</a>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Kurtz2015" class="csl-entry" role="listitem">
Kurtz, ZD, CL Müller, ER Miraldi, DR Littman, MJ Blaser, and Bonneau RA. 2015. <span>“Sparse and Compositionally Robust Inference of Microbial Ecological Networks.”</span> <em><span>PLoS Computational Biology</span></em> 11 (5): e1004226. <a href="https://doi.org/10.1371/journal.pcbi.1004226">https://doi.org/10.1371/journal.pcbi.1004226</a>.
</div><div id="ref-meinshausen2006high" class="csl-entry" role="listitem">
Meinshausen, Nicolai, and Peter Bühlmann. 2006. <span>“<span class="nocase">High-dimensional graphs and variable selection with the Lasso</span>.”</span> <em><span>Annals of Statistics</span></em> 34 (3): 1436–62. <a href="https://doi.org/10.1214/009053606000000281">https://doi.org/10.1214/009053606000000281</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SpiecEasi</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">13075</span><span class="op">)</span></span>
<span><span class="va">se_mb_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/spiec.easi.html">spiec.easi</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                        method <span class="op">=</span> <span class="st">'mb'</span>, nlambda <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                        pulsar.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>rep.num <span class="op">=</span> <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since <code>SpiecEasi</code> uses the StARS (“Stability Approach to Regularization Selection”) method <span class="citation" data-cites="liu2010stability">(<a href="#ref-liu2010stability" role="doc-biblioref">Liu, Roeder, and Wasserman 2010</a>)</span> to obtain a sparse association matrix, we don’t need to set a threshold here. We store the partial correlations corresponding to the StARS-optimal lambda and convert them into an <code>igraph</code> object.</p>
<div class="no-row-height column-margin column-container"><div id="ref-liu2010stability" class="csl-entry" role="listitem">
Liu, Han, Kathryn Roeder, and Larry Wasserman. 2010. <span>“<span class="nocase">Stability Approach to Regularization Selection (StARS) for High Dimensional Graphical Models</span>.”</span> <em><span>Advances in Neural Information Processing Systems</span></em>. <a href="https://doi.org/10.48550/arxiv.1006.3316">https://doi.org/10.48550/arxiv.1006.3316</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get optimal matrix with partial correlations</span></span>
<span><span class="va">se_mb_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/getOptInd.html">getOptBeta</a></span><span class="op">(</span><span class="va">se_mb_est</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_mb_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/symBeta.html">symBeta</a></span><span class="op">(</span><span class="va">se_mb_cor</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">se_mb_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">se_mb_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">se_mb_cor</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># Create graph object</span></span>
<span><span class="va">se_mb_graph</span> <span class="op">&lt;-</span> <span class="fu">transform_asso</span><span class="op">(</span><span class="va">se_mb_cor</span><span class="op">)</span><span class="op">$</span><span class="va">graph</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="network-plots" class="level3" data-number="13.7.4"><h3 data-number="13.7.4" class="anchored" data-anchor-id="network-plots">
<span class="header-section-number">13.7.4</span> Network plots</h3>
<p>The graph objects can now be plotted using the <a href="https://r.igraph.org/">igraph</a> package. The same layout is used in all four plots so that the networks are comparable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/">igraph</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Node sizes</span></span>
<span><span class="va">vsize</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"log10"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fl">3</span></span>
<span></span>
<span><span class="co"># Use Fruchterman-Reingold (force-directed) layout</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">13075</span><span class="op">)</span></span>
<span><span class="va">lay_fr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/layout_with_fr.html">layout_with_fr</a></span><span class="op">(</span><span class="va">se_mb_graph</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sparcc_graph03</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"SparCC (thresh 0.3)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sparcc_graph04</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"SparCC (thresh 0.4)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">prop_graph</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"Shrinkage proportionality\n(thresh 0.4)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">se_mb_graph</span>, layout <span class="op">=</span> <span class="va">lay_fr</span>, vertex.size <span class="op">=</span> <span class="va">vsize</span>, </span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"SpiecEasi (MB)"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/netplots-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p><strong>A few observations:</strong><br>
The density of SparCC (threshold 0.4), proportionality and SpiecEasi is comparable, while the SparCC correlation network with threshold 0.3 is much denser. However, there are edges in the proportionality and SpiecEasi networks that are not present in the two SparCC networks. The SpiecEasi network has less highly connected nodes than the other three networks, but more nodes with one or two connections.</p>
<p>We will look at the degree distribution in the next section to quantify these observations.</p>
</section><section id="network-analysis" class="level3" data-number="13.7.5"><h3 data-number="13.7.5" class="anchored" data-anchor-id="network-analysis">
<span class="header-section-number">13.7.5</span> Network analysis</h3>
<p>Here we repeat some of the network analysis approaches explained in <a href="#sec-network-analysis" class="quarto-xref"><span>Section 13.2</span></a>. The analyses are performed simultaneously for the three association measures as well as the SPRING network constructed in <a href="#sec-spring-network" class="quarto-xref"><span>Section 13.1.4</span></a>. Therefore, we start by creating a list of all the graph objects we need for the analyses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">graphlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>SparCC <span class="op">=</span> <span class="va">sparcc_graph04</span>, </span>
<span>                  Proportionality <span class="op">=</span> <span class="va">prop_graph</span>,</span>
<span>                  SpiecEasi <span class="op">=</span> <span class="va">se_mb_graph</span>,</span>
<span>                  SPRING <span class="op">=</span> <span class="va">spring_graph</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="degree-distribution-1" class="level4" data-number="13.7.5.1"><h4 data-number="13.7.5.1" class="anchored" data-anchor-id="degree-distribution-1">
<span class="header-section-number">13.7.5.1</span> Degree distribution</h4>
<p>The degree distribution is plotted for all four measures to compare the overall network structure.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute degree distributions</span></span>
<span><span class="va">ddlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">graphlist</span>, <span class="fu">igraph</span><span class="fu">::</span><span class="va"><a href="https://r.igraph.org/reference/degree.distribution.html">degree.distribution</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Maximum degree</span></span>
<span><span class="va">maxdeg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">ddlist</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make list elements the same length</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">graphlist</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ddlist</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">maxdeg</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Data frame needed for ggplot2</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Degree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="va">maxdeg</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">graphlist</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                 Fraction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">ddlist</span><span class="op">)</span>, </span>
<span>                 Method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">graphlist</span><span class="op">)</span>, each <span class="op">=</span> <span class="va">maxdeg</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Degree</span>, y <span class="op">=</span> <span class="va">Fraction</span>, group <span class="op">=</span> <span class="va">Method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/degree_dist-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The SparCC and shrinkage proportionality networks have a considerably higher proportion of singletons (zero-degree nodes) than the two conditional dependency graphs, but a lower proportion of nodes with degrees between one and five. The SpiecEasi and SPRING graphs, on the other hand, have a higher proportion of low degree nodes, but no highly connected nodes with a degree greater than eleven.</p>
</section><section id="clustered-heatmaps-1" class="level4" data-number="13.7.5.2"><h4 data-number="13.7.5.2" class="anchored" data-anchor-id="clustered-heatmaps-1">
<span class="header-section-number">13.7.5.2</span> Clustered heatmaps</h4>
<p>Using the <code>ComplexHeatmap</code> package, we plot heatmaps of the association matrices for the four considered association measures. Rows and columns are sorted according to the clusters identified via hierarchical clustering.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For each association measure, we select the 50 nodes with the highest sum of edge weights.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Function for selecting taxa with highest sum of edge weights</span></span>
<span><span class="va">select_taxa</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">adja</span>, <span class="va">ntaxa</span> <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">adja</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="va">ntaxa</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">adja</span><span class="op">[</span><span class="va">sel</span>, <span class="va">sel</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">assolist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">assolist</span><span class="op">$</span><span class="va">SparCC</span> <span class="op">&lt;-</span> <span class="fu">select_taxa</span><span class="op">(</span><span class="va">sparcc_trans04</span><span class="op">$</span><span class="va">adja</span><span class="op">)</span></span>
<span><span class="va">assolist</span><span class="op">$</span><span class="va">Proportionality</span> <span class="op">&lt;-</span> <span class="fu">select_taxa</span><span class="op">(</span><span class="va">prop_trans</span><span class="op">$</span><span class="va">adja</span><span class="op">)</span></span>
<span><span class="va">assolist</span><span class="op">$</span><span class="va">SpiecEasi</span> <span class="op">&lt;-</span> <span class="fu">select_taxa</span><span class="op">(</span><span class="va">se_mb_cor</span><span class="op">)</span></span>
<span><span class="va">assolist</span><span class="op">$</span><span class="va">SPRING</span> <span class="op">&lt;-</span> <span class="fu">select_taxa</span><span class="op">(</span><span class="va">spring_cor</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Color vector for the legend</span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"royalblue4"</span>, <span class="st">"lightblue"</span>, <span class="st">"white"</span>, <span class="st">"orange"</span>, <span class="st">"firebrick3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hm_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">assolist</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">i</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">showlegend</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">showlegend</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">hm_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">assolist</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                          col <span class="op">=</span> <span class="va">col</span>, </span>
<span>                          rect_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"gray"</span>, lwd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                          show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                          show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          column_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">assolist</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, </span>
<span>                          name <span class="op">=</span> <span class="st">"Association"</span>,</span>
<span>                          show_heatmap_legend <span class="op">=</span> <span class="va">showlegend</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Plot with wrap_plots() function from patchwork package</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">hm_list</span>, ncol <span class="op">=</span> <span class="fl">2</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">10</span>, <span class="fl">8</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="60_network_learning_files/figure-html/clustered_heatmap_all-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>The SparCC and the proportionality network show a block structure, where each block corresponds to a cluster. The clusters are less pronounced in the conditional dependence networks. The latter also generally have lower edge weights.</p>
</section><section id="global-network-measures-1" class="level4" data-number="13.7.5.3"><h4 data-number="13.7.5.3" class="anchored" data-anchor-id="global-network-measures-1">
<span class="header-section-number">13.7.5.3</span> Global network measures</h4>
<p>For each association measure, the three global network measures density, transitivity, and average path length are computed and stored in a data frame for comparison.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute density and store in a data frame</span></span>
<span><span class="va">glob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Density <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">graphlist</span>, <span class="va">edge_density</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transitivity</span></span>
<span><span class="va">glob</span><span class="op">$</span><span class="va">Transitivity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">graphlist</span>, <span class="va">transitivity</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Average path length</span></span>
<span><span class="va">glob</span><span class="op">$</span><span class="va">Av.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">graphlist</span>, <span class="va">average.path.length</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">glob</span></span>
<span><span class="co">##                  Density Transitivity Av.path</span></span>
<span><span class="co">##  SparCC          0.01510       0.5458  0.8932</span></span>
<span><span class="co">##  Proportionality 0.01295       0.5402  1.1847</span></span>
<span><span class="co">##  SpiecEasi       0.02162       0.2145  1.6322</span></span>
<span><span class="co">##  SPRING          0.01444       0.1456  1.6993</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</section></section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/23_multi-assay_analyses.html" class="pagination-link" aria-label="Multi-Assay Analyses">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/61_network_comparison.html" class="pagination-link" aria-label="Network comparison">
        <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>