<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 14&nbsp; Network comparison</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/IntegratedLearner.html" rel="next">
<link href="../pages/60_network_learning.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar docked nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../docs/pages/images/mia_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Orchestrating Microbiome Analysis</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">TEST</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-bi-github">
<li>
    <a class="dropdown-item" href="https://github.com/microbiome/OMA/">
 <span class="dropdown-text">Source Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://microbiome.github.io/OMA/docs/devel/">
 <span class="dropdown-text">Browse version <code>devel</code></span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/61_network_comparison.html"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#data-preparation" id="toc-data-preparation" class="nav-link active" data-scroll-target="#data-preparation"><span class="header-section-number">14.1</span> Data preparation</a></li>
  <li><a href="#network-learning-and-analysis" id="toc-network-learning-and-analysis" class="nav-link" data-scroll-target="#network-learning-and-analysis"><span class="header-section-number">14.2</span> Network learning and analysis</a></li>
  <li>
<a href="#differential-network-analysis" id="toc-differential-network-analysis" class="nav-link" data-scroll-target="#differential-network-analysis"><span class="header-section-number">14.3</span> Differential network analysis</a>
  <ul class="collapse">
<li><a href="#visual-comparison" id="toc-visual-comparison" class="nav-link" data-scroll-target="#visual-comparison"><span class="header-section-number">14.3.1</span> Visual comparison</a></li>
  <li><a href="#quantitative-comparison" id="toc-quantitative-comparison" class="nav-link" data-scroll-target="#quantitative-comparison"><span class="header-section-number">14.3.2</span> Quantitative comparison</a></li>
  </ul>
</li>
  <li><a href="#differential-association-analysis" id="toc-differential-association-analysis" class="nav-link" data-scroll-target="#differential-association-analysis"><span class="header-section-number">14.4</span> Differential association analysis</a></li>
  <li><a href="#sec-netcomp-methods" id="toc-sec-netcomp-methods" class="nav-link" data-scroll-target="#sec-netcomp-methods"><span class="header-section-number">14.5</span> Network comparison methods</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/61_network_comparison.html"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-network-comparison" class="quarto-section-identifier"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>This chapter assumes that you are already familiar with how to construct and analyze a single microbial network, which is explained in <a href="60_network_learning.html" class="quarto-xref"><span>Chapter 13</span></a>.</p>
<p>Since microbial interactions are likely to change between conditions, such as between patients and healthy individuals or between different environmental states, identifying network differences between groups is often an integral secondary analysis step. Differences can be detected visually by plotting the networks side by side, or quantitatively using differential network analysis tools.</p>
<p>Two approaches for comparing networks between two conditions are considered in this chapter:</p>
<ul>
<li>
<strong>Differential network analysis</strong>, which analyzes differences in network metrics and network structure.</li>
<li>
<strong>Differential association analysis</strong>, which focuses on differences in the strength of individual associations.</li>
</ul>
<p>See <a href="#sec-netcomp-methods" class="quarto-xref"><span>Section 14.5</span></a> for further details on these two approaches and the methods used in this chapter.</p>
<p>Here we use <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> <span class="citation" data-cites="peschel2021netcomi">(<a href="#ref-peschel2021netcomi" role="doc-biblioref">Peschel et al. 2021</a>)</span> for network comparison, which includes several <strong>differential network analysis approaches</strong> as well as functionality for *differential association analysis**, i.e., generating a differential network. How to install NetCoMi from GitHub is explained in <a href="60_network_learning.html" class="quarto-xref"><span>Chapter 13</span></a>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-potbhare2022skin" class="csl-entry" role="listitem">
Potbhare, Renuka, Ameeta RaviKumar, Eveliina Munukka, Leo Lahti, and Richa Ashma. 2022. <span>“Skin Microbiota Diversity Among Genetically Unrelated Individuals of Indian Origin.”</span> <em><span>PeerJ</span></em> 10: e13075. <a href="https://doi.org/10.7717/peerj.13075">https://doi.org/10.7717/peerj.13075</a>.
</div></div><p>The <strong>PeerJ data set</strong> <span class="citation" data-cites="potbhare2022skin">(<a href="#ref-potbhare2022skin" role="doc-biblioref">Potbhare et al. 2022</a>)</span> containing skin microbial profiles for 58 subjects is again used in this chapter. The dataset also includes information on the subjects’ geographic location, gender, age and diet. Whether the <strong>skin microbiome differs</strong> between people with <strong>different diets</strong> is an interesting question that will be explored in this chapter.</p>
<section id="data-preparation" class="level2" data-number="14.1"><h2 data-number="14.1" class="anchored" data-anchor-id="data-preparation">
<span class="header-section-number">14.1</span> Data preparation</h2>
<p>We perform the same data preprocessing steps as in <a href="60_network_learning.html" class="quarto-xref"><span>Chapter 13</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">NetCoMi</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"peerj13075"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse0</span> <span class="op">&lt;-</span> <span class="va">peerj13075</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Agglomerate to genus level</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">agglomerateByRank</a></span><span class="op">(</span><span class="va">tse0</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Add relative abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, </span>
<span>                      assay.type <span class="op">=</span> <span class="st">"counts"</span>, </span>
<span>                      method <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>                      MARGIN <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Filter by prevalence</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalent</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                            prevalence <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                            detection <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                            assay.type <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add log10-transformed abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"log10"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add clr-transformed abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Based on “Diet”, the <code>tse</code> object is then split into two groups: One with mixed diet subjects, and one with vegetarian subjects. Both subsets have nearly the same sample size and are therefore comparable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">Diet</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Mixed   Veg </span></span>
<span><span class="co">##     28    30</span></span>
<span><span class="va">tse_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/splitOn.html">splitOn</a></span><span class="op">(</span><span class="va">tse</span>, f <span class="op">=</span> <span class="st">"Diet"</span>, use_names <span class="op">=</span> <span class="cn">TRUE</span>, MARGIN <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="network-learning-and-analysis" class="level2" data-number="14.2"><h2 data-number="14.2" class="anchored" data-anchor-id="network-learning-and-analysis">
<span class="header-section-number">14.2</span> Network learning and analysis</h2>
<p>The approach starts again with network construction and analysis, but this time we pass the two data sets to <code>netConstruct</code> to perform a network comparison.</p>
<p>The <code>rep.num</code> argument is set to 10 to perform only 10 repetitions in the model selection approach. This speeds up the permutation tests performed later, and has a negligible effect for this data set.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spring_net_diet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/netConstruct.html">netConstruct</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">tse_list</span><span class="op">$</span><span class="va">Mixed</span>,</span>
<span>                                data2 <span class="op">=</span> <span class="va">tse_list</span><span class="op">$</span><span class="va">Veg</span>,</span>
<span>                                taxRank <span class="op">=</span> <span class="st">"genus"</span>,</span>
<span>                                filtTax <span class="op">=</span> <span class="st">"highestFreq"</span>,</span>
<span>                                filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>highestFreq  <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>,</span>
<span>                                measure <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>                                measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>nlambda <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                                                  rep.num <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                                  thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                  Rmethod <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span>,</span>
<span>                                sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                                dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>                                verbose <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                seed <span class="op">=</span> <span class="fl">13075</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All network measures are now computed for both networks. Also, both GCMs are plotted together with a third matrix containing the differences between the GCMs and significance codes that express if the differences are significantly different from zero.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spring_netprops_diet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">spring_net_diet</span>, </span>
<span>                                   clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                                   hubPar <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>                                   normDeg <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="61_network_comparison_files/figure-html/netAnalyze_diet-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>In both of the networks, some graphlet correlations are significantly different from zero. However, none of the correlations are significantly different between the groups.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">spring_netprops_diet</span>, groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mixed diet"</span>, <span class="st">"Vegetarian"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>##  
##  Component sizes
##  ```````````````
##  Mixed diet:                   
##  size: 18 8 4 3 2  1
##     #:  1 1 1 1 3 43
##  Vegetarian:                 
##  size: 21 5 4 2  1
##     #:  1 1 1 3 46
##  ______________________________
##  Global network properties
##  `````````````````````````
##  Largest connected component (LCC):
##                           Mixed diet Vegetarian
##  Relative LCC size           0.21951    0.25610
##  Clustering coefficient      0.10393    0.22433
##  Modularity                  0.51500    0.47707
##  Positive edge percentage   80.00000  100.00000
##  Edge density                0.13072    0.12381
##  Natural connectivity        0.07312    0.06253
##  Vertex connectivity         1.00000    1.00000
##  Edge connectivity           1.00000    1.00000
##  Average dissimilarity*      0.96009    0.96154
##  Average path length**       2.22768    2.24622
##  
##  Whole network:
##                           Mixed diet Vegetarian
##  Number of components       50.00000   52.00000
##  Clustering coefficient      0.07506    0.16530
##  Modularity                  0.75184    0.66821
##  Positive edge percentage   88.57143  100.00000
##  Edge density                0.01054    0.01084
##  Natural connectivity        0.01342    0.01347
##  -----
##  *: Dissimilarity = 1 - edge weight
##  **: Path length = Units with average dissimilarity
##  
##  ______________________________
##  Clusters
##  - In the whole network
##  - Algorithm: cluster_fast_greedy
##  ```````````````````````````````` 
##  Mixed diet:                          
##  name:  0 1 2 3 4 5 6 7 8 9
##     #: 43 8 6 6 6 4 3 2 2 2
##  
##  Vegetarian:                        
##  name:  0 1 2 3 4 5 6 7 8
##     #: 46 9 8 4 5 4 2 2 2
##  
##  ______________________________
##  Hubs
##  - In alphabetical/numerical order
##  - Based on empirical quantiles of centralities
##  ```````````````````````````````````````````````
##    Mixed diet           Vegetarian
##   Citrobacter            Aeromonas
##       Erwinia              Erwinia
##   Escherichia Escherichia/Shigella
##      Serratia           Salmonella
##    Shewanella          Siccibacter
##  
##  ______________________________
##  Centrality measures
##  - In decreasing order
##  - Centrality of disconnected components is zero
##  ````````````````````````````````````````````````
##  Degree (unnormalized):
##                        Mixed diet Vegetarian
##                Erwinia          6          6
##            Citrobacter          4          0
##               Serratia          4          3
##          Streptococcus          3          0
##                Pantoea          2          1
##                            ______     ______
##                Erwinia          6          6
##              Aeromonas          1          5
##   Escherichia/Shigella          2          5
##            Siccibacter          0          5
##             Salmonella          2          3
##  
##  Betweenness centrality (normalized):
##                        Mixed diet Vegetarian
##                Erwinia    0.60294    0.54737
##               Serratia    0.41912        0.1
##          Streptococcus    0.27941          0
##           Enterobacter    0.24265          0
##            Citrobacter    0.24265          0
##                            ______     ______
##                Erwinia    0.60294    0.54737
##   Escherichia/Shigella    0.11765    0.53158
##              Aeromonas          0    0.36842
##            Siccibacter          0        0.3
##          Thiolamprovum          0    0.20526
##  
##  Closeness centrality (normalized):
##                        Mixed diet Vegetarian
##                Erwinia    0.85251    0.80605
##               Serratia    0.78452    0.56308
##            Citrobacter    0.71265          0
##           Enterobacter    0.64822    0.49217
##          Streptococcus    0.61535          0
##                            ______     ______
##   Escherichia/Shigella    0.57246    0.80747
##                Erwinia    0.85251    0.80605
##              Aeromonas     0.4768    0.74282
##            Siccibacter          0    0.72705
##             Salmonella    0.57364    0.66777
##  
##  Eigenvector centrality (normalized):
##                        Mixed diet Vegetarian
##               Serratia          1    0.41577
##                Erwinia    0.97703          1
##            Citrobacter    0.80991          0
##            Escherichia    0.60982          0
##             Shewanella    0.41174     0.1238
##                            ______     ______
##                Erwinia    0.97703          1
##   Escherichia/Shigella    0.38659    0.96856
##            Siccibacter          0      0.888
##              Aeromonas    0.26033    0.68516
##             Salmonella     0.3153    0.64405</code></pre>
</div>
<p>For each centrality measure, the five nodes with the highest centrality in each group are plotted by default.</p>
<p>We notice some differences in the network properties. The differential network analysis performed in the next section will show if the differences are significant.</p>
</section><section id="differential-network-analysis" class="level2" data-number="14.3"><h2 data-number="14.3" class="anchored" data-anchor-id="differential-network-analysis">
<span class="header-section-number">14.3</span> Differential network analysis</h2>
<section id="visual-comparison" class="level3" data-number="14.3.1"><h3 data-number="14.3.1" class="anchored" data-anchor-id="visual-comparison">
<span class="header-section-number">14.3.1</span> Visual comparison</h3>
<p>We start with a visual comparison of the two networks using NetCoMi’s plot function. The same configuration as in <a href="60_network_learning.html" class="quarto-xref"><span>Chapter 13</span></a> is used.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_netprops_diet</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.97</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     sameColThresh <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Mixed diet"</span>, </span>
<span>     title2 <span class="op">=</span> <span class="st">"Vegetarian"</span>,</span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Overlay a transparent plot on which the legend is plotted</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>fig<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, oma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, new<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, type<span class="op">=</span><span class="st">'n'</span>, bty<span class="op">=</span><span class="st">'n'</span>, xaxt<span class="op">=</span><span class="st">'n'</span>, yaxt<span class="op">=</span><span class="st">'n'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.2</span>, <span class="op">-</span><span class="fl">0.9</span>, cex <span class="op">=</span> <span class="fl">1.5</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="61_network_comparison_files/figure-html/network_plot_diet_difflay-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1632"></p>
</figure>
</div>
</div>
</div>
<p>The layout is computed separately for each network, making it difficult to visually compare certain associations. It is therefore recommended to use the same layout for both groups (argument <code>sameLayout</code>). Instead of simply copying one layout to the other network, we set <code>layoutGroup</code> to “union”. This ensures that the nodes are placed as optimally as possible for both networks.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_netprops_diet</span>,</span>
<span>     sameLayout <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="st">"inboth"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     sameColThresh <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Mixed diet"</span>, </span>
<span>     title2 <span class="op">=</span> <span class="st">"Vegetarian"</span>,</span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add legend</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>fig<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, oma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, new<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, type<span class="op">=</span><span class="st">'n'</span>, bty<span class="op">=</span><span class="st">'n'</span>, xaxt<span class="op">=</span><span class="st">'n'</span>, yaxt<span class="op">=</span><span class="st">'n'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.2</span>, <span class="op">-</span><span class="fl">0.8</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="61_network_comparison_files/figure-html/network_plot_diet_samelay-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1632"></p>
</figure>
</div>
</div>
</div>
<p><strong>A few notes:</strong></p>
<ul>
<li>Differences in the edge weights can now be seen at first glance. For example, Serratia and Citrobacter are strongly associated in the mixed diet group, but not at all in the vegetarian group.</li>
<li>Clusters must share at least two nodes (<code>sameColThresh</code> argument) to be colored equally in both networks, which is why the color of some clusters differs between the groups.</li>
<li>The clustering generally differs markedly. In particular, the cluster assignment of many of the nodes in the largest connected component differs between the two groups.</li>
</ul>
<p>As in <a href="60_network_learning.html" class="quarto-xref"><span>Chapter 13</span></a>, we also generate a network plot using phylum names to color the nodes and mclr-transformed abundances to scale node sizes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">phylum</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">genus</span></span>
<span></span>
<span><span class="co"># Create color vector</span></span>
<span><span class="va">colvec</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span><span class="op">)</span>, <span class="st">"Set3"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_diet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_netprops_diet</span>,</span>
<span>               sameLayout <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               repulsion <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>               rmSingles <span class="op">=</span> <span class="st">"inboth"</span>,</span>
<span>               labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>               nodeSize <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>               nodeColor <span class="op">=</span> <span class="st">"feature"</span>, </span>
<span>               featVecCol <span class="op">=</span> <span class="va">phyla</span>, </span>
<span>               colorVec <span class="op">=</span>  <span class="va">colvec</span>,</span>
<span>               nodeTransp <span class="op">=</span> <span class="fl">20</span>,</span>
<span>               sameColThresh <span class="op">=</span> <span class="fl">2</span>,</span>
<span>               highlightHubs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>               cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>               edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>               title1 <span class="op">=</span> <span class="st">"Mixed diet"</span>, </span>
<span>               title2 <span class="op">=</span> <span class="st">"Vegetarian"</span>,</span>
<span>               showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               cexTitle <span class="op">=</span> <span class="fl">2</span>,</span>
<span>               mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add legends</span></span>
<span><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span><span class="va">col_transp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/colToTransp.html">colToTransp</a></span><span class="op">(</span><span class="va">colvec</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>fig<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, oma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, new<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, type<span class="op">=</span><span class="st">'n'</span>, bty<span class="op">=</span><span class="st">'n'</span>, xaxt<span class="op">=</span><span class="st">'n'</span>, yaxt<span class="op">=</span><span class="st">'n'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>, <span class="op">-</span><span class="fl">0.8</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>, <span class="fl">1.3</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, pt.cex <span class="op">=</span> <span class="fl">2.5</span>, title <span class="op">=</span> <span class="st">"Phylum:"</span>, </span>
<span>       legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col_transp</span>, bty <span class="op">=</span> <span class="st">"n"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="61_network_comparison_files/figure-html/network_plot_diet_phyla-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1632"></p>
</figure>
</div>
</div>
</div>
</section><section id="quantitative-comparison" class="level3" data-number="14.3.2"><h3 data-number="14.3.2" class="anchored" data-anchor-id="quantitative-comparison">
<span class="header-section-number">14.3.2</span> Quantitative comparison</h3>
<p><code><a href="https://rdrr.io/pkg/NetCoMi/man/netCompare.html">netCompare()</a></code> enables a quantitative network comparison using comparative measures such as Jaccard’s Index, Adjusted Rand Index, and permutation tests.</p>
<p>To test for statistical significance of differences in network properties, we perform permutation tests with 1000 permutations. Multiple CPU cores are used to save run time. The association matrices estimated for all permutations are stored in an external file. We will reuse them later when performing differential association analysis. They could also be used to rerun <code><a href="https://rdrr.io/pkg/NetCoMi/man/netCompare.html">netCompare()</a></code> with different parameter settings.</p>
<p>Note that unless running on a cluster with considerably more CPU cores, a network comparison with permutation tests may take several hours. You should test the code below with a small number of permutations to make sure it works before applying it to your data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spring_netcomp_diet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/netCompare.html">netCompare</a></span><span class="op">(</span><span class="va">spring_netprops_diet</span>, </span>
<span>                                  permTest <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  nPerm <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                  cores <span class="op">=</span> <span class="fl">6</span>,</span>
<span>                                  seed <span class="op">=</span> <span class="fl">13075</span>,</span>
<span>                                  storeAssoPerm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  fileStoreAssoPerm <span class="op">=</span> <span class="st">"general/network_data/spring_assoPerm"</span>,</span>
<span>                                  verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">spring_netcomp_diet</span>, </span>
<span>        groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mixed diet"</span>, <span class="st">"Vegetarian"</span><span class="op">)</span>,</span>
<span>        numbNodes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>##  
##  Comparison of Network Properties
##  ----------------------------------
##  CALL: 
##  netCompare(x = spring_netprops_diet, permTest = TRUE, verbose = TRUE, 
##      nPerm = 1000, cores = 19, libPathsClust = "/dss/dsshome1/07/di93fen/R", 
##      seed = 13075, storeAssoPerm = TRUE, fileStoreAssoPerm = "general/network_data/spring_assoPerm")
##  
##  ______________________________
##  Global network properties
##  `````````````````````````
##  Largest connected component (LCC):
##                           Mixed diet   Vegetarian    abs.diff.     p-value
##  Relative LCC size             0.220        0.256        0.037    0.838162
##  Clustering coefficient        0.104        0.224        0.120    0.450549
##  Modularity                    0.515        0.477        0.038    0.801199
##  Positive edge percentage     80.000      100.000       20.000    0.000999
##  Edge density                  0.131        0.124        0.007    0.910090
##  Natural connectivity          0.073        0.063        0.011    0.812188
##  Vertex connectivity           1.000        1.000        0.000    1.000000
##  Edge connectivity             1.000        1.000        0.000    1.000000
##  Average dissimilarity*        0.960        0.962        0.001    0.945055
##  Average path length**         2.228        2.246        0.019    0.984016
##                              
##  Relative LCC size           
##  Clustering coefficient      
##  Modularity                  
##  Positive edge percentage ***
##  Edge density                
##  Natural connectivity        
##  Vertex connectivity         
##  Edge connectivity           
##  Average dissimilarity*      
##  Average path length**       
##  
##  Whole network:
##                           Mixed diet   Vegetarian    abs.diff.     p-value
##  Number of components         50.000       52.000        2.000     0.93706
##  Clustering coefficient        0.075        0.165        0.090     0.46254
##  Modularity                    0.752        0.668        0.084     0.24975
##  Positive edge percentage     88.571      100.000       11.429     0.02398
##  Edge density                  0.011        0.011        0.000     0.97303
##  Natural connectivity          0.013        0.013        0.000     0.89011
##                              
##  Number of components        
##  Clustering coefficient      
##  Modularity                  
##  Positive edge percentage *  
##  Edge density                
##  Natural connectivity        
##  -----
##  p-values: one-tailed test with null hypothesis diff=0
##   *: Dissimilarity = 1 - edge weight
##  **: Path length = Units with average dissimilarity
##  
##  ______________________________
##  Jaccard index (similarity betw. sets of most central nodes)
##  ```````````````````````````````````````````````````````````
##                      Jacc   P(&lt;=Jacc)     P(&gt;=Jacc)    
##  degree             0.560      0.9944       0.01637 *  
##  betweenness centr. 0.294      0.4777       0.71860    
##  closeness centr.   0.560      0.9944       0.01637 *  
##  eigenvec. centr.   0.560      0.9944       0.01637 *  
##  hub taxa           0.111      0.1431       0.97399    
##  -----
##  Jaccard index in [0,1] (1 indicates perfect agreement)
##  
##  ______________________________
##  Adjusted Rand index (similarity betw. clusterings)
##  ``````````````````````````````````````````````````
##          wholeNet       LCC
##  ARI        0.367     0.035
##  p-value    0.000     0.563
##  -----
##  ARI in [-1,1] with ARI=1: perfect agreement betw. clusterings
##                     ARI=0: expected for two random clusterings
##  p-value: permutation test (n=1000) with null hypothesis ARI=0
##  
##  ______________________________
##  Graphlet Correlation Distance
##  `````````````````````````````
##          wholeNet         LCC  
##  GCD       1.5980      2.1770  
##  p-value   0.4226      0.7143  
##  -----
##  GCD &gt;= 0 (GCD=0 indicates perfect agreement between GCMs)
##  p-value: permutation test with null hypothesis GCD=0
##  
##  ______________________________
##  Centrality measures
##  - In decreasing order
##  - Centrality of disconnected components is zero
##  ````````````````````````````````````````````````
##  Degree (unnormalized):
##                       Mixed diet Vegetarian abs.diff. adj.p-value  
##  Siccibacter                   0          5         5           1  
##  Aeromonas                     1          5         4           1  
##  Citrobacter                   4          0         4           1  
##  Escherichia/Shigella          2          5         3           1  
##  Streptococcus                 3          0         3           1  
##  
##  Betweenness centrality (normalized):
##                       Mixed diet Vegetarian abs.diff. adj.p-value  
##  Escherichia/Shigella      0.118      0.532     0.414           1  
##  Aeromonas                 0.000      0.368     0.368           1  
##  Serratia                  0.419      0.100     0.319           1  
##  Siccibacter               0.000      0.300     0.300           1  
##  Streptococcus             0.279      0.000     0.279           1  
##  
##  Closeness centrality (normalized):
##                Mixed diet Vegetarian abs.diff. adj.p-value  
##  Siccibacter        0.000      0.727     0.727      0.2676  
##  Citrobacter        0.713      0.000     0.713      0.9799  
##  Streptococcus      0.615      0.000     0.615      0.2676  
##  Escherichia        0.587      0.000     0.587      0.9799  
##  Xenorhabdus        0.000      0.527     0.527      0.2676  
##  
##  Eigenvector centrality (normalized):
##                       Mixed diet Vegetarian abs.diff. adj.p-value  
##  Siccibacter               0.000      0.888     0.888      0.6553  
##  Citrobacter               0.810      0.000     0.810      1.0000  
##  Escherichia               0.610      0.000     0.610      1.0000  
##  Serratia                  1.000      0.416     0.584      1.0000  
##  Escherichia/Shigella      0.387      0.969     0.582      1.0000  
##  
##  _________________________________________________________
##  Significance codes: ***: 0.001, **: 0.01, *: 0.05, .: 0.1</code></pre>
</div>
<p><strong>Interpreting some results:</strong></p>
<ul>
<li>Almost all global network properties are significantly different between the groups (for <span class="math inline">\(\alpha=0.1\)</span>), thus reflecting the different overall network structure we already have seen in the network plots.</li>
<li>For the Jaccard index of degree, closeness, and eigenvector centrality, the probability P(&gt;=Jacc) is significant, meaning that the sets of the most central nodes are quite similar for these three measures. The Jaccard index for the hub nodes, on the other hand, is low because the two networks share only one hub node (“Erwinia”).</li>
<li>As indicated by some similarities in the clusterings, the adjusted Rand index (ARI) of the whole network is significantly different from zero and thus from random clustering. The ARI of the largest connected component (LCC), however, is close to zero due to the different clusterings in the LCC.</li>
<li>The two GCD values are significantly different from zero, indicating substantial differences in the overall network structures.</li>
<li>All nodes are also tested for having significantly different centrality (only the five nodes with the highest absolute difference are shown in the summary). For <span class="math inline">\(\alpha=0.05\)</span>, some nodes have a significantly different closeness centrality, and for <span class="math inline">\(\alpha=0.1\)</span> also a significantly different eigenvector centrality. Most of these nodes have a high centrality in the one group, but are not connected in the other group.</li>
</ul></section></section><section id="differential-association-analysis" class="level2" data-number="14.4"><h2 data-number="14.4" class="anchored" data-anchor-id="differential-association-analysis">
<span class="header-section-number">14.4</span> Differential association analysis</h2>
<p>The <code><a href="https://rdrr.io/pkg/NetCoMi/man/diffnet.html">diffnet()</a></code> function provides statistical tests to assess whether the associations themselves are significantly different between the two groups. <code>NetCoMi</code> also provides a plot function to generate a differential network, where two nodes are connected if they are differentially associated between the groups.</p>
<p>Since we have already computed the permutation association matrices before, we can reuse them here (argument <code>fileLoadAssoPerm</code>).</p>
<p>The local false discovery rate is controlled at level 0.2 to account for multiplicity.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spring_diffnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/diffnet.html">diffnet</a></span><span class="op">(</span><span class="va">spring_net_diet</span>,</span>
<span>                          diffMethod <span class="op">=</span> <span class="st">"perm"</span>,</span>
<span>                          fileLoadAssoPerm <span class="op">=</span> <span class="st">"general/network_data/spring_assoPerm"</span>,</span>
<span>                          adjust <span class="op">=</span> <span class="st">"lfdr"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">spring_diffnet</span><span class="op">$</span><span class="va">pAdjustVec</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">spring_diffnet</span><span class="op">$</span><span class="va">pvalsVec</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">##  [1] 12</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Some of the unadjusted p-values are below the usual 5% significance level. However, none of the differences remain significant after adjusting for multiple testing so that the differential network would be empty.</p>
<p>To demonstrate the interpretation of a differential network, we set <code>adjust</code> to “none”, which is actually statistically incorrect.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spring_diffnet_unadj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NetCoMi/man/diffnet.html">diffnet</a></span><span class="op">(</span><span class="va">spring_net_diet</span>,</span>
<span>                                pvalsVec <span class="op">=</span> <span class="va">spring_diffnet</span><span class="op">$</span><span class="va">pvalsVec</span>,</span>
<span>                                diffMethod <span class="op">=</span> <span class="st">"perm"</span>,</span>
<span>                                alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                adjust <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>diffnet</code> object it now plotted using NetCoMi’s plot function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">spring_diffnet_unadj</span>, </span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span>     cexLegend <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">5</span>,<span class="fl">15</span><span class="op">)</span>,</span>
<span>     legendGroupnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mixed diet"</span>, <span class="st">"Vegetarian"</span><span class="op">)</span>,</span>
<span>     legendPos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.2</span>,<span class="fl">1.5</span><span class="op">)</span>,</span>
<span>     legendArgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>lwd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,</span>
<span>     fade <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="61_network_comparison_files/figure-html/diffnet_plot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1920"></p>
</figure>
</div>
</div>
</div>
<p>Edge colors represent the direction of the associations in the two groups. For example, if two OTUs are positively correlated in the mixed diet group and uncorrelated in the vegetarian group (such as Serratia and Citrobacter), the edge color is dark green.</p>
</section><section id="sec-netcomp-methods" class="level2 page-columns page-full" data-number="14.5"><h2 data-number="14.5" class="anchored" data-anchor-id="sec-netcomp-methods">
<span class="header-section-number">14.5</span> Network comparison methods</h2>
<p>While many approaches exist for the detection of <strong>differential correlations</strong>, e.g. <span class="citation" data-cites="yu2019new mckenzie2016dgca siska2017differential">(<a href="#ref-yu2019new" role="doc-biblioref">Yu et al. 2019</a>; <a href="#ref-mckenzie2016dgca" role="doc-biblioref">McKenzie et al. 2016</a>; <a href="#ref-siska2017differential" role="doc-biblioref">Siska and Kechris 2017</a>)</span>, the literature on the more general case of <strong>differential association</strong> detection is scarce. <span class="citation" data-cites="bhuva2019differential">Bhuva et al. (<a href="#ref-bhuva2019differential" role="doc-biblioref">2019</a>)</span> compare various methods in a simulation study, which again includes many differential correlation approaches, but also more general methods such as latent differential graphical models. <span class="citation" data-cites="gill2010statistical">Gill, Datta, and Datta (<a href="#ref-gill2010statistical" role="doc-biblioref">2010</a>)</span> introduce an approach to analyze whether the connectivity of individual nodes is different between two groups using permutation tests, which is applicable to any kind of association. <span class="citation" data-cites="he2019statistical">He et al. (<a href="#ref-he2019statistical" role="doc-biblioref">2019</a>)</span> propose a test to infer the differential network structure for two conditional dependence networks.</p>
<div class="no-row-height column-margin column-container"><div id="ref-yu2019new" class="csl-entry" role="listitem">
Yu, Danyang, Zeyu Zhang, Kimberly Glass, Jessica Su, Dawn L. DeMeo, Kelan Tantisira, Scott T. Weiss, and Weiliang Qiu. 2019. <span>“<span class="nocase">New Statistical Methods for Constructing Robust Differential Correlation Networks to characterize the interactions among microRNAs</span>.”</span> <em><span>Scientific Reports</span></em> 9 (1): 1–12. <a href="https://doi.org/10.1038/s41598-019-40167-8">https://doi.org/10.1038/s41598-019-40167-8</a>.
</div><div id="ref-mckenzie2016dgca" class="csl-entry" role="listitem">
McKenzie, Andrew T., Igor Katsyv, Won Min Song, Minghui Wang, and Bin Zhang. 2016. <span>“<span class="nocase">DGCA: A comprehensive R package for Differential Gene Correlation Analysis</span>.”</span> <em><span>BMC Systems Biology</span></em> 10 (1): 1–25. <a href="https://doi.org/10.1186/s12918-016-0349-1">https://doi.org/10.1186/s12918-016-0349-1</a>.
</div><div id="ref-siska2017differential" class="csl-entry" role="listitem">
Siska, Charlotte, and Katerina Kechris. 2017. <span>“<span class="nocase">Differential correlation for sequencing data</span>.”</span> <em><span>BMC Research Notes</span></em> 10 (1): 1–9. <a href="https://doi.org/10.1186/s13104-016-2331-9">https://doi.org/10.1186/s13104-016-2331-9</a>.
</div><div id="ref-bhuva2019differential" class="csl-entry" role="listitem">
Bhuva, Dharmesh D., Joseph Cursons, Gordon K. Smyth, and Melissa J. Davis. 2019. <span>“<span class="nocase">Differential co-expression-based detection of conditional relationships in transcriptional data: comparative analysis and application to breast cancer</span>.”</span> <em><span>Genome Biology</span></em> 20 (1). <a href="https://doi.org/10.1186/S13059-019-1851-8">https://doi.org/10.1186/S13059-019-1851-8</a>.
</div><div id="ref-gill2010statistical" class="csl-entry" role="listitem">
Gill, Ryan, Somnath Datta, and Susmita Datta. 2010. <span>“<span class="nocase">A statistical framework for differential network analysis from microarray data</span>.”</span> <em><span>BMC Bioinformatics</span></em> 11 (1): 95.
</div><div id="ref-he2019statistical" class="csl-entry" role="listitem">
He, Hao, Shaolong Cao, Ji gang Zhang, Hui Shen, Yu Ping Wang, and Hong wen Deng. 2019. <span>“<span class="nocase">A Statistical Test for Differential Network Analysis Based on Inference of Gaussian Graphical Model</span>.”</span> <em><span>Scientific Reports</span></em> 9 (1): 1–8. <a href="https://doi.org/10.1038/s41598-019-47362-7">https://doi.org/10.1038/s41598-019-47362-7</a>.
</div><div id="ref-shojaie2021differential" class="csl-entry" role="listitem">
Shojaie, Ali. 2021. <span>“<span class="nocase">Differential network analysis: A statistical perspective</span>.”</span> <em><span>Wiley Interdisciplinary Reviews: Computational Statistics</span></em> 13 (2): e1508. <a href="https://doi.org/10.1002/WICS.1508">https://doi.org/10.1002/WICS.1508</a>.
</div><div id="ref-lichtblau2017comparative" class="csl-entry" role="listitem">
Lichtblau, Yvonne, Karin Zimmermann, Berit Haldemann, Dido Lenze, Michael Hummel, and Ulf Leser. 2017. <span>“<span class="nocase">Comparative assessment of differential network analysis methods</span>.”</span> <em><span>Briefings in Bioinformatics</span></em> 18 (5): 837–50. <a href="https://doi.org/10.1093/bib/bbw061">https://doi.org/10.1093/bib/bbw061</a>.
</div><div id="ref-jardim2019bionetstat" class="csl-entry" role="listitem">
Jardim, Vinícius Carvalho, Suzana De Siqueira Santos, Andre Fujita, and Marcos Silveira Buckeridge. 2019. <span>“<span class="nocase">BioNetStat: A tool for biological networks differential analysis</span>.”</span> <em><span>Frontiers in Genetics</span></em> 10 (JUN): 1–13. <a href="https://doi.org/10.3389/fgene.2019.00594">https://doi.org/10.3389/fgene.2019.00594</a>.
</div></div><p>Performing <strong>differential network analysis</strong> is challenging because network measures do not follow classical statistical distributions. <span class="citation" data-cites="shojaie2021differential">Shojaie (<a href="#ref-shojaie2021differential" role="doc-biblioref">2021</a>)</span> provide an overview of differential network analysis methods, but focus only on changes in edge sets. <span class="citation" data-cites="lichtblau2017comparative">Lichtblau et al. (<a href="#ref-lichtblau2017comparative" role="doc-biblioref">2017</a>)</span> compare differential network analysis methods that incorporate multiple local and global network measures. <span class="citation" data-cites="jardim2019bionetstat">Jardim et al. (<a href="#ref-jardim2019bionetstat" role="doc-biblioref">2019</a>)</span> present a tool “BioNetStat” for differential analysis of biological networks, which is able to compare certain network measures between groups.</p>
<p>The <a href="https://github.com/stefpeschel/NetCoMi">NetCoMi</a> package used for network comparison in this chapter includes the following <strong>differential network analysis approaches:</strong>:</p>
<ul>
<li>
<strong>Permutation approach</strong> to test global network measures (e.g., transitivity, connectivity, or average path length) as well as centrality measures for group differences.</li>
<li>
<strong>Jaccard</strong> index to assess the similarity between sets of most central nodes</li>
<li>
<strong>Adjusted Rand index</strong> to assess the similarity between clusterings</li>
<li>
<strong>Graphlet Correlation Distance</strong> (GCM)</li>
</ul>
<p>See <span class="citation" data-cites="peschel2021netcomi">(<a href="#ref-peschel2021netcomi" role="doc-biblioref">Peschel et al. 2021</a>)</span> for an explanation of the first three approaches. The GCM was proposed by <span class="citation" data-cites="yaverouglu2014revealing">Yaveroğlu et al. (<a href="#ref-yaverouglu2014revealing" role="doc-biblioref">2014</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-peschel2021netcomi" class="csl-entry" role="listitem">
Peschel, Stefanie, Christian L Müller, Erika von Mutius, Anne-Laure Boulesteix, and Martin Depner. 2021. <span>“<span class="nocase">NetCoMi: network construction and comparison for microbiome data in R</span>.”</span> <em><span>Briefings in Bioinformatics</span></em> 22 (4): bbaa290. <a href="https://doi.org/10.1093/bib/bbaa290">https://doi.org/10.1093/bib/bbaa290</a>.
</div><div id="ref-yaverouglu2014revealing" class="csl-entry" role="listitem">
Yaveroğlu, Ömer Nebil, Noël Malod-Dognin, Darren Davis, Zoran Levnajic, Vuk Janjic, Rasa Karapandza, Aleksandar Stojmirovic, and Nataša Pržulj. 2014. <span>“<span class="nocase">Revealing the Hidden Language of Complex Networks</span>.”</span> <em><span>Scientific Reports</span></em> 4 (1): 1–9. <a href="https://doi.org/10.1038/srep04547">https://doi.org/10.1038/srep04547</a>.
</div><div id="ref-fisher1970statistical" class="csl-entry" role="listitem">
Fisher, Ronald Aylmer. 1970. <span>“Statistical Methods for Research Workers.”</span> In <em><span class="nocase">Breakthroughs in statistics: Methodology and distribution</span></em>, 66–70. Springer.
</div><div id="ref-siska2016discordant" class="csl-entry" role="listitem">
Siska, Charlotte, Russell Bowler, and Katerina Kechris. 2016. <span>“<span class="nocase">The discordant method: A novel approach for differential correlation</span>.”</span> <em>Bioinformatics</em> 32 (5): 690–96. <a href="https://doi.org/10.1093/bioinformatics/btv633">https://doi.org/10.1093/bioinformatics/btv633</a>.
</div></div><p>Two methods (Fisher’s z-test <span class="citation" data-cites="fisher1970statistical">(<a href="#ref-fisher1970statistical" role="doc-biblioref">Fisher 1970</a>)</span> and the Discordant method <span class="citation" data-cites="siska2016discordant">(<a href="#ref-siska2016discordant" role="doc-biblioref">Siska, Bowler, and Kechris 2016</a>)</span>) are available for identifying differential correlations, and <strong>permutation tests</strong> for the more general case of identifying <strong>differential associations</strong>. See <span class="citation" data-cites="peschel2021netcomi">(<a href="#ref-peschel2021netcomi" role="doc-biblioref">Peschel et al. 2021</a>)</span> for details. NetCoMi offers also a function for plotting a differential network.</p>



</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/60_network_learning.html" class="pagination-link" aria-label="Network learning and analysis">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/IntegratedLearner.html" class="pagination-link" aria-label="Multi-omics Integration">
        <span class="nav-page-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>