<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Orchestrating Microbiome Analysis - 20&nbsp; Resources</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/98_exercises.html" rel="next">
<link href="../pages/80_training.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/80_training.html">Training</a></li><li class="breadcrumb-item"><a href="../pages/95_resources.html"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Microbiome Analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/microbiome/OMA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://microbiome.github.io/OMA/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
    </div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#data-containers" id="toc-data-containers" class="nav-link active" data-scroll-target="#data-containers"><span class="header-section-number">20.1</span> Data containers</a>
  <ul class="collapse">
  <li><a href="#data-container-documentation" id="toc-data-container-documentation" class="nav-link" data-scroll-target="#data-container-documentation"><span class="header-section-number">20.1.1</span> Data container documentation</a></li>
  <li><a href="#other-relevant-containers" id="toc-other-relevant-containers" class="nav-link" data-scroll-target="#other-relevant-containers"><span class="header-section-number">20.1.2</span> Other relevant containers</a></li>
  <li><a href="#phyloseq-an-alternative-container-for-microbiome-data" id="toc-phyloseq-an-alternative-container-for-microbiome-data" class="nav-link" data-scroll-target="#phyloseq-an-alternative-container-for-microbiome-data"><span class="header-section-number">20.1.3</span> phyloseq: an alternative container for microbiome data</a></li>
  </ul></li>
  <li><a href="#r-programming-resources" id="toc-r-programming-resources" class="nav-link" data-scroll-target="#r-programming-resources"><span class="header-section-number">20.2</span> R programming resources</a>
  <ul class="collapse">
  <li><a href="#base-r-and-rstudio" id="toc-base-r-and-rstudio" class="nav-link" data-scroll-target="#base-r-and-rstudio"><span class="header-section-number">20.2.1</span> Base R and RStudio</a></li>
  <li><a href="#sec-bioc_intro" id="toc-sec-bioc_intro" class="nav-link" data-scroll-target="#sec-bioc_intro"><span class="header-section-number">20.2.2</span> Bioconductor Classes</a></li>
  </ul></li>
  <li><a href="#sec-quarto" id="toc-sec-quarto" class="nav-link" data-scroll-target="#sec-quarto"><span class="header-section-number">20.3</span> Reproducible reporting with Quarto</a>
  <ul class="collapse">
  <li><a href="#learn-quarto" id="toc-learn-quarto" class="nav-link" data-scroll-target="#learn-quarto"><span class="header-section-number">20.3.1</span> Learn Quarto</a></li>
  <li><a href="#additional-material-on-rmarkdown" id="toc-additional-material-on-rmarkdown" class="nav-link" data-scroll-target="#additional-material-on-rmarkdown"><span class="header-section-number">20.3.2</span> Additional material on Rmarkdown</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/80_training.html">Training</a></li><li class="breadcrumb-item"><a href="../pages/95_resources.html"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span id="sec-resources" class="quarto-section-identifier"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="data-containers" class="level2 page-columns page-full" data-number="20.1">
<h2 data-number="20.1" class="anchored" data-anchor-id="data-containers"><span class="header-section-number">20.1</span> Data containers</h2>
<section id="data-container-documentation" class="level3 page-columns page-full" data-number="20.1.1">
<h3 data-number="20.1.1" class="anchored" data-anchor-id="data-container-documentation"><span class="header-section-number">20.1.1</span> Data container documentation</h3>
<ul>
<li><p>SingleCellExperiment <span class="citation" data-cites="R_SingleCellExperiment">(<a href="#ref-R_SingleCellExperiment" role="doc-biblioref">Lun and Risso 2020</a>)</span></p>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">Online tutorial</a><br>
</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">Project page</a><br>
</li>
<li><a href="https://doi.org/10.1038/s41592-019-0654-x">Publication</a></li>
</ul></li>
<li><p>SummarizedExperiment <span class="citation" data-cites="R_SummarizedExperiment">(<a href="#ref-R_SummarizedExperiment" role="doc-biblioref">Morgan et al. 2020</a>)</span></p>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">Online tutorial</a><br>
</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">Project page</a></li>
</ul></li>
<li><p>TreeSummarizedExperiment <span class="citation" data-cites="R_TreeSummarizedExperiment">(<a href="#ref-R_TreeSummarizedExperiment" role="doc-biblioref">Huang 2020</a>)</span></p>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/TreeSummarizedExperiment/inst/doc/Introduction_to_treeSummarizedExperiment.html">Online tutorial</a><br>
</li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html">Project page</a><br>
</li>
<li><a href="https://doi.org/10.12688/f1000research.26669.2">Publication</a></li>
</ul></li>
<li><p>MultiAssayExperiment <span class="citation" data-cites="Ramos2017">(<a href="#ref-Ramos2017" role="doc-biblioref">Ramos et al. 2017</a>)</span></p>
<ul>
<li><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html">Online tutorial</a><br>
</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html">Project page</a><br>
</li>
<li><a href="https://doi.org/10.1158/0008-5472.CAN-17-0344">Publication</a></li>
</ul></li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-R_SingleCellExperiment" class="csl-entry" role="listitem">
Lun, Aaron, and Davide Risso. 2020. <em>SingleCellExperiment: S4 Classes for Single Cell Data</em>.
</div><div id="ref-R_SummarizedExperiment" class="csl-entry" role="listitem">
Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020. <em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://bioconductor.org/packages/SummarizedExperiment">https://bioconductor.org/packages/SummarizedExperiment</a>.
</div><div id="ref-R_TreeSummarizedExperiment" class="csl-entry" role="listitem">
Huang, Ruizhu. 2020. <em>TreeSummarizedExperiment: A S4 Class for Data with Tree Structures</em>.
</div></div></section>
<section id="other-relevant-containers" class="level3 page-columns page-full" data-number="20.1.2">
<h3 data-number="20.1.2" class="anchored" data-anchor-id="other-relevant-containers"><span class="header-section-number">20.1.2</span> Other relevant containers</h3>
<ul>
<li><a href="https://rdrr.io/bioc/S4Vectors/man/DataFrame-class.html">DataFrame</a> which behaves similarly to <code>data.frame</code>, yet efficient and fast when used with large datasets.<br>
</li>
<li><a href="https://rdrr.io/bioc/Biostrings/man/DNAString-class.html">DNAString</a> along with <code>DNAStringSet</code>,<code>RNAString</code> and <code>RNAStringSet</code> efficient storage and handling of long biological sequences are offered within the Biostrings package <span class="citation" data-cites="R_Biostrings">(<a href="#ref-R_Biostrings" role="doc-biblioref">Pagès et al. 2020</a>)</span>.<br>
</li>
<li>GenomicRanges (<span class="citation" data-cites="GenomicRanges2013">(<a href="#ref-GenomicRanges2013" role="doc-biblioref">Lawrence et al. 2013</a>)</span>) offers an efficient representation and manipulation of genomic annotations and alignments, see e.g.&nbsp;<code>GRanges</code> and <code>GRangesList</code> at <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">An Introduction to the GenomicRangesPackage</a>.</li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-R_Biostrings" class="csl-entry" role="listitem">
Pagès, H., P. Aboyoun, R. Gentleman, and S. DebRoy. 2020. <em>Biostrings: Efficient Manipulation of Biological Strings</em>. <a href="https://bioconductor.org/packages/Biostrings">https://bioconductor.org/packages/Biostrings</a>.
</div><div id="ref-GenomicRanges2013" class="csl-entry" role="listitem">
Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin Morgan, and Vincent Carey. 2013. <span>“Software for Computing and Annotating Genomic Ranges.”</span> <em><span>PLoS</span> Computational Biology</em> 9. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div></div><p><a href="http://girke.bioinformatics.ucr.edu/GEN242/tutorials/rsequences/rsequences/">NGS Analysis Basics</a> provides a walk-through of the above-mentioned features with detailed examples.</p>
</section>
<section id="phyloseq-an-alternative-container-for-microbiome-data" class="level3 page-columns page-full" data-number="20.1.3">
<h3 data-number="20.1.3" class="anchored" data-anchor-id="phyloseq-an-alternative-container-for-microbiome-data"><span class="header-section-number">20.1.3</span> phyloseq: an alternative container for microbiome data</h3>
<p>The <code>phyloseq</code> package and class became the first widely used data container for microbiome data science in R. Many methods for taxonomic profiling data are readily available for this class. We provide here a short description how <code>phyloseq</code> and <code>*Experiment</code> classes relate to each other.</p>
<p><code>assays</code> : This slot is similar to <code>otu_table</code> in <code>phyloseq</code>. In a <code>SummarizedExperiment</code> object multiple assays, raw counts, transformed counts can be stored. See also <span class="citation" data-cites="Ramos2017">(<a href="#ref-Ramos2017" role="doc-biblioref">2017</a>)</span> for storing data from multiple <code>experiments</code> such as RNASeq, Proteomics, etc. <code>rowData</code> : This slot is similar to <code>tax_table</code> in <code>phyloseq</code> to store taxonomic information. <code>colData</code> : This slot is similar to <code>sample_data</code> in <code>phyloseq</code> to store information related to samples. <code>rowTree</code> : This slot is similar to <code>phy_tree</code> in <code>phyloseq</code> to store phylogenetic tree.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Ramos2017" class="csl-entry" role="listitem">
Ramos, Marcel, Lucas Schiffer, Angela Re, Rimsha Azhar, Azfar Basunia, Carmen Rodriguez Cabrera, Tiffany Chan, et al. 2017. <span>“Software for the Integration of Multiomics Experiments in Bioconductor.”</span> <em>Cancer Research</em>. <a href="https://doi.org/10.1158/0008-5472.CAN-17-0344">https://doi.org/10.1158/0008-5472.CAN-17-0344</a>.
</div></div><p>In this book, you will encounter terms such as <code>FeatureIDs</code> and <code>SampleIDs</code>. <code>FeatureIDs</code> : These are basically OTU/ASV ids which are row names in <code>assays</code> and <code>rowData</code>. <code>SampleIDs</code> : As the name suggests, these are sample ids which are column names in <code>assays</code> and row names in <code>colData</code>. <code>FeatureIDs</code> and <code>SampleIDs</code> are used but the technical terms <code>rownames</code> and <code>colnames</code> are encouraged to be used, since they relate to actual objects we work with.</p>
<p><img src="https://raw.githubusercontent.com/FelixErnst/TreeSummarizedExperiment
/2293440c6e70ae4d6e978b6fdf2c42fdea7fb36a/vignettes/tse2.png" width="100%"></p>
<section id="benchmarking-treese-with-phyloseq" class="level4" data-number="20.1.3.1">
<h4 data-number="20.1.3.1" class="anchored" data-anchor-id="benchmarking-treese-with-phyloseq"><span class="header-section-number">20.1.3.1</span> Benchmarking TreeSE with phyloseq</h4>
<p>TreeSE objects can be converted into phyloseq objects and vice versa, therefore it is possible to compare the two containers in terms of computational efficiency. Remarkably, TreeSE and phyloseq were benchmarked against one another in mia v1.2.3 and phyloseq v1.38.0, respectively. 5 standard microbiome analysis operationswere applied to 4 datasets of varying size with both containers. In a nutshell, TreeSE and phyloseq showed a similar performance for datasets of small and medium size for most of the operations. However, TreeSE performed more efficiently as the size of the datasets increased. Further details on such results can be found in the <a href="https://github.com/microbiome/benchmarking">benchmarking repository</a>.</p>
</section>
<section id="resources-on-phyloseq" class="level4 page-columns page-full" data-number="20.1.3.2">
<h4 data-number="20.1.3.2" class="anchored" data-anchor-id="resources-on-phyloseq"><span class="header-section-number">20.1.3.2</span> Resources on phyloseq</h4>
<p>The phyloseq container provides analogous methods to TreeSE. The following material can be used to familiarize with such alternative methods:</p>
<ul>
<li><a href="https://microsud.github.io/Tools-Microbiome-Analysis/">List of R tools for microbiome analysis</a><br>
</li>
<li>phyloseq <span class="citation" data-cites="McMurdie2013">(<a href="#ref-McMurdie2013" role="doc-biblioref">McMurdie and Holmes 2013</a>)</span><br>
</li>
<li><a href="http://microbiome.github.io/tutorials/">microbiome tutorial</a><br>
</li>
<li><a href="https://microsud.github.io/microbiomeutilities/">microbiomeutilities</a></li>
<li><a href="https://microbiome.github.io/OMA/docs/devel/pages/97_extra_materials.html#cheatsheet">phyloseq/TreeSE cheatsheet</a></li>
<li>Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses <span class="citation" data-cites="Callahan2016">(<a href="#ref-Callahan2016" role="doc-biblioref">Callahan et al. 2016</a>)</span>.</li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-McMurdie2013" class="csl-entry" role="listitem">
McMurdie, PJ, and S Holmes. 2013. <span>“<span><em>Phyloseq:</em></span> An r Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.”</span> <em>PLoS ONE</em> 8: e61217. <a href="https://doi.org/10.1371/journal.pone.0061217">https://doi.org/10.1371/journal.pone.0061217</a>.
</div><div id="ref-Callahan2016" class="csl-entry" role="listitem">
Callahan, Ben J., Kris Sankaran, Julia A. Fukuyama, Paul J. McMurdie, and Susan P. Holmes. 2016. <span>“Bioconductor Workflow for Microbiome Data Analysis: From Raw Reads to Community Analyses [Version 2; Peer Review: 3 Approved].”</span> <em>F1000Research</em> 5: 1492. <a href="https://doi.org/10.12688/f1000research.8986.2">https://doi.org/10.12688/f1000research.8986.2</a>.
</div></div></section>
</section>
</section>
<section id="r-programming-resources" class="level2 page-columns page-full" data-number="20.2">
<h2 data-number="20.2" class="anchored" data-anchor-id="r-programming-resources"><span class="header-section-number">20.2</span> R programming resources</h2>
<section id="base-r-and-rstudio" class="level3 page-columns page-full" data-number="20.2.1">
<h3 data-number="20.2.1" class="anchored" data-anchor-id="base-r-and-rstudio"><span class="header-section-number">20.2.1</span> Base R and RStudio</h3>
<p>If you are new to R, you could try <a href="https://swirlstats.com/">swirl</a> for a kickstart to R programming. Further support resources are available through the Bioconductor project <span class="citation" data-cites="Huber2015">(<a href="#ref-Huber2015" role="doc-biblioref">Huber et al. 2015</a>)</span>.</p>
<div class="no-row-height column-margin column-container"></div><ul>
<li><a href="https://geomoer.github.io/moer-base-r/cheatsheet.html">Base R and RStudio cheatsheets</a><br>
</li>
<li><a href="https://www.rstudio.com/resources/cheatsheets/">Package-specific cheatsheets</a><br>
</li>
<li><a href="https://ggplot2.tidyverse.org/">Visualization with ggplot2</a><br>
</li>
<li><a href="http://www.cookbook-r.com/Graphs/">R graphics cookbook</a></li>
</ul>
</section>
<section id="sec-bioc_intro" class="level3 page-columns page-full" data-number="20.2.2">
<h3 data-number="20.2.2" class="anchored" data-anchor-id="sec-bioc_intro"><span class="header-section-number">20.2.2</span> Bioconductor Classes</h3>
<p><a href="https://carpentries-incubator.github.io/bioc-intro/">Introduction to data analysis with R and Bioconductor</a>; Carpentries introduction, including R &amp; RStudio installation instructions</p>
<p><strong>S4 system</strong></p>
<p>S4 class system has brought several useful features to the object-oriented programming paradigm within R, and it is constantly deployed in R/Bioconductor packages <span class="citation" data-cites="Huber2015">(<a href="#ref-Huber2015" role="doc-biblioref">Huber et al. 2015</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Huber2015" class="csl-entry" role="listitem">
Huber, W., V. J. Carey, R. Gentleman, S. Anders, M. Carlson, B. S. Carvalho, H. C. Bravo, et al. 2015. <span>“<span>O</span>rchestrating High-Throughput Genomic Analysis with <span>B</span>ioconductor.”</span> <em>Nature Methods</em> 12 (2): 115–21. <a href="http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html">http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html</a>.
</div></div><div class="line-block">&nbsp;&nbsp;Online Document:</div>
<ul>
<li>Hervé Pagès, <a href="https://bioconductor.org/packages/release/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>.<br>
</li>
<li>Laurent Gatto, <a href="https://bioconductor.org/help/course-materials/2013/CSAMA2013/friday/afternoon/S4-tutorial.pdf">A practical tutorial on S4 programming</a><br>
</li>
<li>How S4 Methods Work <span class="citation" data-cites="Chambers2006">(<a href="#ref-Chambers2006" role="doc-biblioref">J. Chambers 2006</a>)</span></li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-Chambers2006" class="csl-entry" role="listitem">
Chambers, JM. 2006. <span>“How S4 Methods Work.”</span> Technical report.
</div></div><div class="line-block">&nbsp;&nbsp;Books:</div>
<ul>
<li>John M. Chambers. Software for Data Analysis: Programming with R. Springer, New York, 2008. ISBN-13 978-0387759357 <span class="citation" data-cites="Chambers2008">(<a href="#ref-Chambers2008" role="doc-biblioref">J. M. Chambers 2008</a>)</span><br>
</li>
<li>I Robert Gentleman. R Programming for Bioinformatics. Chapman &amp; Hall/CRC, New York, 2008. ISBN-13 978-1420063677 <span class="citation" data-cites="gentleman2008r">(<a href="#ref-gentleman2008r" role="doc-biblioref">Gentleman 2008</a>)</span></li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-Chambers2008" class="csl-entry" role="listitem">
Chambers, John M. 2008. <em>Software for Data Analysis: Programming with r</em>. Vol. 2. Springer.
</div><div id="ref-gentleman2008r" class="csl-entry" role="listitem">
Gentleman, Robert. 2008. <em>R Programming for Bioinformatics</em>. CRC Press.
</div></div></section>
</section>
<section id="sec-quarto" class="level2 page-columns page-full" data-number="20.3">
<h2 data-number="20.3" class="anchored" data-anchor-id="sec-quarto"><span class="header-section-number">20.3</span> Reproducible reporting with Quarto</h2>
<section id="learn-quarto" class="level3" data-number="20.3.1">
<h3 data-number="20.3.1" class="anchored" data-anchor-id="learn-quarto"><span class="header-section-number">20.3.1</span> Learn Quarto</h3>
<p>Reproducible reporting is the starting point for robust interactive data science. Perform the following tasks:</p>
<ul>
<li><p>If you are entirely new to Quarto, take <a href="https://quarto.org/docs/get-started/hello/rstudio.html">this short tutorial</a> to get introduced to the most important functions within Quarto. Then experiment with different options from the <a href="https://res.cloudinary.com/dyd911kmh/image/upload/v1676540721/Marketing/Blog/Quarto_Cheat_Sheet.pdf">Quarto cheatsheet</a>.</p></li>
<li><p>Create a Quarto template in RStudio, and render it into a document (markdown, PDF, docx or other format). In case you are new to Quarto, its documentation provides guidelines to use Quarto with the R language (<a href="https://quarto.org/docs/computations/r.html">here</a>) and the RStudio IDE (<a href="https://quarto.org/docs/tools/rstudio.html">here</a>).</p></li>
<li><p>Further examples are tips for Quarto are available in <a href="https://appsilon.com/r-quarto-tutorial/">this online tutorial</a> to interactive reproducible reporting.</p></li>
</ul>
</section>
<section id="additional-material-on-rmarkdown" class="level3 page-columns page-full" data-number="20.3.2">
<h3 data-number="20.3.2" class="anchored" data-anchor-id="additional-material-on-rmarkdown"><span class="header-section-number">20.3.2</span> Additional material on Rmarkdown</h3>
<p>Being able to use Quarto in R partly relies on your previous knowledge of Rmarkdown. The following resources can help you get familiar with Rmarkdown:</p>
<ul>
<li><a href="https://www.markdowntutorial.com/">Online tutorial</a><br>
</li>
<li><a href="https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">Cheatsheet</a><br>
</li>
<li><a href="https://rmarkdown.rstudio.com/lesson-1.html">Documentation</a><br>
</li>
<li><a href="https://rpubs.com/marschmi/RMarkdown">Dr.&nbsp;C Titus Brown’s tutorial</a></li>
</ul>
<p><strong>Figure sources:</strong></p>
<p><strong>Original article</strong></p>
<ul>
<li>Huang R <em>et al</em>. (2021) TreeSummarizedExperiment: a S4 class for data with hierarchical structure. F1000Research 9:1246. <span class="citation" data-cites="Huang2021">(<a href="#ref-Huang2021" role="doc-biblioref">Huang et al. 2021</a>)</span></li>
</ul>
<div class="no-row-height column-margin column-container"><div id="ref-Huang2021" class="csl-entry" role="listitem">
Huang, Ruizhu, Charlotte Soneson, Felix G. M. Ernst, et al. 2021. <span>“TreeSummarizedExperiment: A S4 Class for Data with Hierarchical Structure [Version 2; Peer Review: 3 Approved].”</span> <em>F1000Research</em> 9: 1246. <a href="https://doi.org/10.12688/f1000research.26669.2">https://doi.org/10.12688/f1000research.26669.2</a>.
</div></div><p><strong>Reference Sequence slot extension</strong></p>
<ul>
<li>Lahti L <em>et al</em>. (2020) <a href="https://doi.org/10.7490/%20f1000research.1118447.1">Upgrading the R/Bioconductor ecosystem for microbiome research</a> F1000Research 9:1464 (slides).</li>
</ul>



</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../pages/80_training.html" class="pagination-link" aria-label="Training">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/98_exercises.html" class="pagination-link" aria-label="Exercises">
        <span class="nav-page-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>




</body></html>