<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - Appendix A — Extra material</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/90_acknowledgments.html" rel="next">
<link href="../pages/98_exercises.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/97_extra_materials.html">Appendices</a></li><li class="breadcrumb-item"><a href="../pages/97_extra_materials.html"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Microbiome Analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/microbiome/OMA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://microbiome.github.io/OMA/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#slides" id="toc-slides" class="nav-link active" data-scroll-target="#slides"><span class="header-section-number">A.1</span> Slides</a></li>
  <li><a href="#compare-permanova" id="toc-compare-permanova" class="nav-link" data-scroll-target="#compare-permanova"><span class="header-section-number">A.2</span> PERMANOVA comparison</a></li>
  <li><a href="#bayesian-multinomial-logistic-normal-models" id="toc-bayesian-multinomial-logistic-normal-models" class="nav-link" data-scroll-target="#bayesian-multinomial-logistic-normal-models"><span class="header-section-number">A.3</span> Bayesian Multinomial Logistic-Normal Models</a></li>
  <li><a href="#interactive-3d-plots" id="toc-interactive-3d-plots" class="nav-link" data-scroll-target="#interactive-3d-plots"><span class="header-section-number">A.4</span> Interactive 3D Plots</a></li>
  <li>
<a href="#phyloseq-vs-treese-cheatsheet" id="toc-phyloseq-vs-treese-cheatsheet" class="nav-link" data-scroll-target="#phyloseq-vs-treese-cheatsheet"><span class="header-section-number">A.5</span> phyloseq vs TreeSE cheatsheet</a>
  <ul class="collapse">
<li><a href="#accessing-different-types-of-data-in-phyloseq-versus-treese" id="toc-accessing-different-types-of-data-in-phyloseq-versus-treese" class="nav-link" data-scroll-target="#accessing-different-types-of-data-in-phyloseq-versus-treese"><span class="header-section-number">A.5.1</span> Accessing different types of data in phyloseq versus TreeSE</a></li>
  <li><a href="#building-phyloseq-objects-vs-treese-objects-phyloseq-treesummarizedexperiment" id="toc-building-phyloseq-objects-vs-treese-objects-phyloseq-treesummarizedexperiment" class="nav-link" data-scroll-target="#building-phyloseq-objects-vs-treese-objects-phyloseq-treesummarizedexperiment"><span class="header-section-number">A.5.2</span> Building phyloseq objects vs TreeSE objects: phyloseq = TreeSummarizedExperiment</a></li>
  <li><a href="#handling-different-otu-table-normalizations-in-phyloseq-vs-treese" id="toc-handling-different-otu-table-normalizations-in-phyloseq-vs-treese" class="nav-link" data-scroll-target="#handling-different-otu-table-normalizations-in-phyloseq-vs-treese"><span class="header-section-number">A.5.3</span> Handling different OTU table normalizations in phyloseq vs TreeSE</a></li>
  <li><a href="#subsetting-samples-and-taxa" id="toc-subsetting-samples-and-taxa" class="nav-link" data-scroll-target="#subsetting-samples-and-taxa"><span class="header-section-number">A.5.4</span> Subsetting samples and taxa</a></li>
  <li><a href="#calculating-alpha-diversity-estimate_richness-estimatediversity" id="toc-calculating-alpha-diversity-estimate_richness-estimatediversity" class="nav-link" data-scroll-target="#calculating-alpha-diversity-estimate_richness-estimatediversity"><span class="header-section-number">A.5.5</span> Calculating alpha diversity: estimate_richness = estimateDiversity</a></li>
  <li><a href="#calculating-beta-diversity-ordinate-runmds" id="toc-calculating-beta-diversity-ordinate-runmds" class="nav-link" data-scroll-target="#calculating-beta-diversity-ordinate-runmds"><span class="header-section-number">A.5.6</span> Calculating beta diversity: ordinate = runMDS</a></li>
  <li><a href="#plotting-ordinations-plot_ordination-plotreduceddim" id="toc-plotting-ordinations-plot_ordination-plotreduceddim" class="nav-link" data-scroll-target="#plotting-ordinations-plot_ordination-plotreduceddim"><span class="header-section-number">A.5.7</span> Plotting ordinations: plot_ordination = plotReducedDim</a></li>
  <li><a href="#agglomerating-taxa-tax_glommergefeaturesbyrank" id="toc-agglomerating-taxa-tax_glommergefeaturesbyrank" class="nav-link" data-scroll-target="#agglomerating-taxa-tax_glommergefeaturesbyrank"><span class="header-section-number">A.5.8</span> Agglomerating taxa: tax_glom=mergeFeaturesByRank</a></li>
  <li><a href="#cheatsheet" id="toc-cheatsheet" class="nav-link" data-scroll-target="#cheatsheet"><span class="header-section-number">A.5.9</span> Cheatsheet</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-extras" class="quarto-section-identifier">Appendix A — Extra material</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>eval<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="slides" class="level2" data-number="A.1"><h2 data-number="A.1" class="anchored" data-anchor-id="slides">
<span class="header-section-number">A.1</span> Slides</h2>
<p><a href="https://microbiome.github.io/outreach/">Outreach material</a> includes slide sets for training events.</p>
</section><section id="compare-permanova" class="level2" data-number="A.2"><h2 data-number="A.2" class="anchored" data-anchor-id="compare-permanova">
<span class="header-section-number">A.2</span> PERMANOVA comparison</h2>
<p>Here we present two possible uses of the <code>adonis2</code> function which performs PERMANOVA. The optional argument <code>by</code> has an effect on the statistical outcome, so its two options are compared here.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># import necessary packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-gregmisc/gtools">gtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vegandevs/vegan">vegan</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-gregmisc/gtools">gtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let us load the <em>enterotype</em> TSE object and run PERMANOVA for different orders of three variables with two different approaches: <code>by = "margin"</code> or <code>by = "terms"</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load and prepare data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"enterotype"</span>, package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">enterotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">enterotype</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="co"># drop samples missing meta data</span></span>
<span><span class="va">enterotype</span> <span class="op">&lt;-</span> <span class="va">enterotype</span><span class="op">[</span> , <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Nationality"</span>, <span class="st">"Gender"</span>, <span class="st">"ClinicalStatus"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># define variables and list all possible combinations</span></span>
<span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Nationality"</span>, <span class="st">"Gender"</span>, <span class="st">"ClinicalStatus"</span><span class="op">)</span></span>
<span><span class="va">var_perm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gtools/man/combinations.html">permutations</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span>, r <span class="op">=</span> <span class="fl">3</span>, <span class="va">vars</span><span class="op">)</span></span>
<span><span class="va">formulas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">var_perm</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">row</span><span class="op">)</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span><span class="op">(</span><span class="va">row</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"+"</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># create empty data.frames for further storing p-values</span></span>
<span><span class="va">terms_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"Formula"</span> <span class="op">=</span> <span class="va">formulas</span>,</span>
<span>                       <span class="st">"ClinicalStatus"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>                       <span class="st">"Gender"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>                       <span class="st">"Nationality"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">margin_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"Formula"</span> <span class="op">=</span> <span class="va">formulas</span>,</span>
<span>                        <span class="st">"ClinicalStatus"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>                        <span class="st">"Gender"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>                        <span class="st">"Nationality"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">row_idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">var_perm</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># generate temporary formula (i.e. "assay ~ ClinicalStatus + Nationality + Gender")</span></span>
<span>  <span class="va">tmp_formula</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span><span class="op">(</span><span class="va">var_perm</span><span class="op">[</span><span class="va">row_idx</span>, <span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"+"</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">tmp_formula</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'t(assay(enterotype, "relabundance")) ~ '</span>,</span>
<span>                            <span class="va">tmp_formula</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># multiple variables, default: by = "terms"</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">75</span><span class="op">)</span></span>
<span>  <span class="va">with_terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/adonis.html">adonis2</a></span><span class="op">(</span><span class="va">tmp_formula</span>, </span>
<span>                by <span class="op">=</span> <span class="st">"terms"</span>,</span>
<span>                data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span>,</span>
<span>                permutations <span class="op">=</span> <span class="fl">99</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># multiple variables, by = "margin"</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">75</span><span class="op">)</span></span>
<span>  <span class="va">with_margin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/adonis.html">adonis2</a></span><span class="op">(</span><span class="va">tmp_formula</span>, </span>
<span>                 by <span class="op">=</span> <span class="st">"margin"</span>,</span>
<span>                 data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span>,</span>
<span>                 permutations <span class="op">=</span> <span class="fl">99</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># extract p-values</span></span>
<span>  <span class="va">terms_p</span> <span class="op">&lt;-</span> <span class="va">with_terms</span><span class="op">[[</span><span class="st">"Pr(&gt;F)"</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">terms_p</span> <span class="op">&lt;-</span> <span class="va">terms_p</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">terms_p</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">margin_p</span> <span class="op">&lt;-</span> <span class="va">with_margin</span><span class="op">[[</span><span class="st">"Pr(&gt;F)"</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">margin_p</span> <span class="op">&lt;-</span> <span class="va">margin_p</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">margin_p</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># store p-values into data.frames</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">col_idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">var_perm</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    </span>
<span>    <span class="va">terms_df</span><span class="op">[</span><span class="va">var_perm</span><span class="op">[</span><span class="va">row_idx</span>, <span class="va">col_idx</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">row_idx</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="va">terms_p</span><span class="op">[</span><span class="va">col_idx</span><span class="op">]</span></span>
<span>    <span class="va">margin_df</span><span class="op">[</span><span class="va">var_perm</span><span class="op">[</span><span class="va">row_idx</span>, <span class="va">col_idx</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">row_idx</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="va">margin_p</span><span class="op">[</span><span class="va">col_idx</span><span class="op">]</span></span>
<span>    </span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following table displays the p-values for the three variables ClinicalStatus, Gender and Nationality obtained by PERMANOVA with <code>adonis2</code>. Note that the p-values remain identical when <code>by = "margin"</code>, but change with the order of the variables in the formula when <code>by = "terms"</code> (default).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">terms_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">margin_df</span>, by <span class="op">=</span> <span class="st">"Formula"</span>, suffix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">" (terms)"</span>, <span class="st">" (margin)"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 24%">
<col style="width: 14%">
<col style="width: 9%">
<col style="width: 12%">
<col style="width: 15%">
<col style="width: 10%">
<col style="width: 13%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Formula</th>
<th style="text-align: right;">ClinicalStatus (terms)</th>
<th style="text-align: right;">Gender (terms)</th>
<th style="text-align: right;">Nationality (terms)</th>
<th style="text-align: right;">ClinicalStatus (margin)</th>
<th style="text-align: right;">Gender (margin)</th>
<th style="text-align: right;">Nationality (margin)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ClinicalStatus + Gender + Nationality</td>
<td style="text-align: right;">0.20</td>
<td style="text-align: right;">0.70</td>
<td style="text-align: right;">0.04</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
</tr>
<tr class="even">
<td style="text-align: left;">ClinicalStatus + Nationality + Gender</td>
<td style="text-align: right;">0.20</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.05</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Gender + ClinicalStatus + Nationality</td>
<td style="text-align: right;">0.17</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.04</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
</tr>
<tr class="even">
<td style="text-align: left;">Gender + Nationality + ClinicalStatus</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.03</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Nationality + ClinicalStatus + Gender</td>
<td style="text-align: right;">0.61</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
</tr>
<tr class="even">
<td style="text-align: left;">Nationality + Gender + ClinicalStatus</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.39</td>
<td style="text-align: right;">0.04</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">0.04</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section id="bayesian-multinomial-logistic-normal-models" class="level2 page-columns page-full" data-number="A.3"><h2 data-number="A.3" class="anchored" data-anchor-id="bayesian-multinomial-logistic-normal-models">
<span class="header-section-number">A.3</span> Bayesian Multinomial Logistic-Normal Models</h2>
<p>Analysis using such model could be performed with the function <code>pibble</code> from the <code>fido</code> package, wihch is in form of a Multinomial Logistic-Normal Linear Regression model; see <a href="https://jsilve24.github.io/fido/articles/introduction-to-fido.html">vignette</a> of package.</p>
<p>The following presents such an exemplary analysis based on the data of <span class="citation" data-cites="Sprockett2020">Sprockett et al. (<a href="#ref-Sprockett2020" role="doc-biblioref">2020</a>)</span> available through <code>microbiomeDataSets</code> package.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Sprockett2020" class="csl-entry" role="listitem">
Sprockett, Daniel D., Melanie Martin, Elizabeth K. Costello, Adam R. Burns, Susan P. Holmes, Michael D. Gurven, and David A. Relman. 2020. <span>“Microbiota Assembly, Structure, and Dynamics Among <span>Tsimane</span> Horticulturalists of the <span>Bolivian Amazon</span>.”</span> <em>Nat Commun</em> 11 (1): 3772. <a href="https://doi.org/10.1038/s41467-020-17541-6">https://doi.org/10.1038/s41467-020-17541-6</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://jsilve24.github.io/fido/">fido</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Loading the libraries and importing data:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://jsilve24.github.io/fido/">fido</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">microbiomeDataSets</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiomeDataSets/man/SprockettTHData.html">SprockettTHData</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We pick three covariates (“Sex”,“Age_Years”,“Delivery_Mode”) during this analysis as an example, and beforehand we check for missing data:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="va">cov_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sex"</span>,<span class="st">"Age_Years"</span>,<span class="st">"Delivery_Mode"</span><span class="op">)</span></span>
<span><span class="va">na_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>,<span class="va">cov_names</span><span class="op">]</span><span class="op">)</span>, <span class="fl">2</span>, <span class="va">sum</span><span class="op">)</span></span>
<span><span class="va">na_summary</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">na_counts</span>,row.names<span class="op">=</span><span class="va">cov_names</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We drop missing values of the covariates:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span> , <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">Delivery_Mode</span><span class="op">)</span> <span class="op">]</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span> , <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">Age_Years</span><span class="op">)</span> <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We agglomerate microbiome data to Phylum:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse_phylum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Phylum"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We extract the counts assay and covariate data to build the model matrix:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse_phylum</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="co"># design matrix</span></span>
<span><span class="co"># taking 3 covariates</span></span>
<span><span class="va">sample_data</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse_phylum</span><span class="op">)</span><span class="op">[</span>,<span class="va">cov_names</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">Sex</span><span class="op">+</span><span class="va">Age_Years</span><span class="op">+</span><span class="va">Delivery_Mode</span>,data<span class="op">=</span><span class="va">sample_data</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Building the parameters for the <code>pibble</code> call to build the model; see more at <a href="https://jsilve24.github.io/fido/articles/introduction-to-fido.html">vignette</a>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_taxa</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="va">upsilon</span> <span class="op">&lt;-</span> <span class="va">n_taxa</span><span class="op">+</span><span class="fl">3</span></span>
<span><span class="va">Omega</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">n_taxa</span><span class="op">)</span></span>
<span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">n_taxa</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">Xi</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">upsilon</span><span class="op">-</span><span class="va">n_taxa</span><span class="op">)</span><span class="op">*</span><span class="va">G</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span><span class="va">Omega</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span>
<span><span class="va">Theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">n_taxa</span><span class="op">-</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Gamma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Automatically initializing the priors and visualizing their distributions:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://jsilve24.github.io/fido/reference/pibble_fit.html">pibble</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="va">X</span>, <span class="va">upsilon</span>, <span class="va">Theta</span>, <span class="va">Gamma</span>, <span class="va">Xi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://jsilve24.github.io/fido/reference/name_dims.html">names_covariates</a></span><span class="op">(</span><span class="va">priors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">priors</span>, pars<span class="op">=</span><span class="st">"Lambda"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="97_extra_materials_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Estimating the posterior by including our response data <code>Y</code>. Note: Some computational failures could occur (see <a href="https://github-wiki-see.page/m/jsilve24/fido/wiki/Frequently-Asked-Questions">discussion</a>) the arguments <code>multDirichletBoot</code> <code>calcGradHess</code> could be passed in such case.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">priors</span><span class="op">$</span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">Y</span> </span>
<span><span class="va">posterior</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://jsilve24.github.io/fido/reference/refit.html">refit</a></span><span class="op">(</span><span class="va">priors</span>, optim_method<span class="op">=</span><span class="st">"adam"</span>, multDirichletBoot<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="co">#calcGradHess=FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Printing a summary about the posterior:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://jsilve24.github.io/fido/reference/ppc_summary.html">ppc_summary</a></span><span class="op">(</span><span class="va">posterior</span><span class="op">)</span></span>
<span><span class="co">##  Proportions of Observations within 95% Credible Interval: 0.9977226</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plotting the summary of the posterior distributions of the regression parameters:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://jsilve24.github.io/fido/reference/name_dims.html">names_categories</a></span><span class="op">(</span><span class="va">posterior</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">posterior</span>,par<span class="op">=</span><span class="st">"Lambda"</span>,focus.cov<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="97_extra_materials_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Taking a closer look at “Sex” and “Delivery_Mode”:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">posterior</span>, par<span class="op">=</span><span class="st">"Lambda"</span>, focus.cov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="97_extra_materials_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="interactive-3d-plots" class="level2 page-columns page-full" data-number="A.4"><h2 data-number="A.4" class="anchored" data-anchor-id="interactive-3d-plots">
<span class="header-section-number">A.4</span> Interactive 3D Plots</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dmurdoch/rgl">rgl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/knit_hooks.html">knit_hooks</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>webgl <span class="op">=</span> <span class="va">hook_webgl</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this section we make a 3D version of the earlier Visualizing the most dominant genus on PCoA (see @ref(quality-control)), with the help of the plotly <span class="citation" data-cites="Sievert2020">(<a href="#ref-Sievert2020" role="doc-biblioref">Sievert 2020</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Sievert2020" class="csl-entry" role="listitem">
Sievert, Carson. 2020. <em>Interactive Web-Based Data Visualization with r, Plotly, and Shiny</em>. Chapman; Hall/CRC. <a href="https://plotly-r.com">https://plotly-r.com</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Importing necessary libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/curatedMetagenomicData">curatedMetagenomicData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT">DT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Querying the data</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">age</span> <span class="op">&gt;=</span> <span class="fl">18</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># taking only data of age 18 or above</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">alcohol</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># excluding missing values</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/returnSamples.html">returnSamples</a></span><span class="op">(</span><span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tse_Genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>, rank<span class="op">=</span><span class="st">"genus"</span><span class="op">)</span></span>
<span><span class="va">tse_Genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/perSampleDominantTaxa.html">addPerSampleDominantFeatures</a></span><span class="op">(</span><span class="va">tse_Genus</span>,assay.type<span class="op">=</span><span class="st">"relative_abundance"</span>, name <span class="op">=</span> <span class="st">"dominant_taxa"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Performing PCoA with Bray-Curtis dissimilarity.</span></span>
<span><span class="va">tse_Genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runMDS.html">runMDS</a></span><span class="op">(</span><span class="va">tse_Genus</span>, FUN <span class="op">=</span> <span class="fu">vegan</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span>, ncomponents <span class="op">=</span> <span class="fl">3</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"PCoA_BC"</span>, assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Getting the 6 top taxa</span></span>
<span><span class="va">top_taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/summaries.html">getTopFeatures</a></span><span class="op">(</span><span class="va">tse_Genus</span>,top <span class="op">=</span> <span class="fl">6</span>, assay.type <span class="op">=</span> <span class="st">"relative_abundance"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Naming all the rest of non top-taxa as "Other"</span></span>
<span><span class="va">most_abundant</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse_Genus</span><span class="op">)</span><span class="op">$</span><span class="va">dominant_taxa</span>,</span>
<span>                   <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="kw">if</span> <span class="op">(</span><span class="va">x</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">top_taxa</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span><span class="st">"Other"</span><span class="op">}</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Storing the previous results as a new column within colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse_Genus</span><span class="op">)</span><span class="op">$</span><span class="va">most_abundant</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">most_abundant</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculating percentage of the most abundant</span></span>
<span><span class="va">most_abundant_freq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">most_abundant</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">most_abundant_percent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">most_abundant_freq</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">most_abundant_freq</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieving the explained variance</span></span>
<span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">tse_Genus</span>, <span class="st">"PCoA_BC"</span><span class="op">)</span>, <span class="st">"eig"</span><span class="op">)</span>;</span>
<span><span class="va">var_explained</span> <span class="op">&lt;-</span> <span class="va">e</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">e</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="phyloseq-vs-treese-cheatsheet" class="level2" data-number="A.5"><h2 data-number="A.5" class="anchored" data-anchor-id="phyloseq-vs-treese-cheatsheet">
<span class="header-section-number">A.5</span> phyloseq vs TreeSE cheatsheet</h2>
<p>This section has a cheatsheet for translating common functions in phyloseq to TreeSE/mia with example code.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Download libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'phyloseq'</span></span>
<span><span class="co">##  The following object is masked from 'package:fido':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      nsamples</span></span>
<span><span class="co">##  The following object is masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      distance</span></span>
<span><span class="co">##  The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      sampleNames</span></span>
<span><span class="co">##  The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      distance</span></span>
<span><span class="co">##  The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      distance</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'dplyr'</span></span>
<span><span class="co">##  The following objects are masked from 'package:mia':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      full_join, inner_join, left_join, right_join</span></span>
<span><span class="co">##  The following objects are masked from 'package:Biostrings':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      collapse, intersect, setdiff, setequal, union</span></span>
<span><span class="co">##  The following object is masked from 'package:XVector':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      slice</span></span>
<span><span class="co">##  The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      combine</span></span>
<span><span class="co">##  The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      intersect, setdiff, union</span></span>
<span><span class="co">##  The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      intersect</span></span>
<span><span class="co">##  The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      collapse, desc, intersect, setdiff, slice, union</span></span>
<span><span class="co">##  The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      first, intersect, rename, setdiff, setequal, union</span></span>
<span><span class="co">##  The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      combine, intersect, setdiff, union</span></span>
<span><span class="co">##  The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      count</span></span>
<span><span class="co">##  The following objects are masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      filter, lag</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Start by loading data as a phyloseq object phy and as TreeSE object tse.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Loading example data</span></span>
<span><span class="co"># Using GlobalPatterns dataset</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"phyloseq"</span>, <span class="st">"GlobalPatterns"</span><span class="op">)</span> <span class="co"># phyloseq object</span></span>
<span><span class="va">phy</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span> <span class="co"># Rename</span></span>
<span><span class="va">phy</span> <span class="co"># Check the phyloseq object</span></span>
<span><span class="co">##  phyloseq-class experiment-level object</span></span>
<span><span class="co">##  otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]</span></span>
<span><span class="co">##  sample_data() Sample Data:       [ 26 samples by 7 sample variables ]</span></span>
<span><span class="co">##  tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]</span></span>
<span><span class="co">##  phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"mia"</span>, <span class="st">"GlobalPatterns"</span><span class="op">)</span> <span class="co"># TreeSE object</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span> <span class="co"># Rename</span></span>
<span><span class="va">tse</span> <span class="co"># Check the tse object</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 19216 26 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(1): counts</span></span>
<span><span class="co">##  rownames(19216): 549322 522457 ... 200359 271582</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: a LinkDataFrame (19216 rows)</span></span>
<span><span class="co">##  rowTree: 1 phylo tree(s) (19216 leaves)</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="accessing-different-types-of-data-in-phyloseq-versus-treese" class="level3" data-number="A.5.1"><h3 data-number="A.5.1" class="anchored" data-anchor-id="accessing-different-types-of-data-in-phyloseq-versus-treese">
<span class="header-section-number">A.5.1</span> Accessing different types of data in phyloseq versus TreeSE</h3>
<p>Often microbiome datasets contain three different types of tables, one which defines the microbes’ taxonomy from domain to species level, one that describes sample level information like whether the sample is from a healthy or a diseased person, and one that has the abundances of taxa from mapping, like an OTU table.</p>
<p>There are slightly different names for these tables in phyloseq and tse, but they can be retrieved from the phyloseq and tse containers in analogous ways.</p>
<p><strong>Accessing the table of taxonomic names: tax_table = rowData</strong></p>
<p>phyloseq and TreeSE objects’ taxonomy tables can be accessed with tax_table and rowData commands.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_taxtable</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># Access the phyloseq taxonomic name table</span></span>
<span>  <span class="va">data.frame</span> <span class="co"># Make into a data frame</span></span>
<span></span>
<span><span class="va">tse_taxtable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># Same for tse</span></span>
<span>  <span class="va">data.frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Accessing sample data: sample_data = colData</strong></p>
<p>Sample data can be accessed with sample_data and colData commands.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_sampledata</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span></span>
<span></span>
<span><span class="va">tse_sampledata</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Accessing operational taxonomic unit (OTU) abundance objects: otu_table = assay</strong></p>
<p>OTU tables can be accessed with otu_table and assay commands. The assay can also hold other types of information like taxa abundances from shotgun metagenomic annotation, or functional gene abundances.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_otutable</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span></span>
<span></span>
<span><span class="va">tse_otutable</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="building-phyloseq-objects-vs-treese-objects-phyloseq-treesummarizedexperiment" class="level3" data-number="A.5.2"><h3 data-number="A.5.2" class="anchored" data-anchor-id="building-phyloseq-objects-vs-treese-objects-phyloseq-treesummarizedexperiment">
<span class="header-section-number">A.5.2</span> Building phyloseq objects vs TreeSE objects: phyloseq = TreeSummarizedExperiment</h3>
<p>After learning how to access various data types from TreeSE, let’s see how creating TreeSE objects compares to creating phyloseq objects. We will use the vanilla dataframes we created from the phyloseq object to demonstrate making both types of data objects. These are identical to the equivalent tse dataframes but for demonstration we will use ones created from phy.</p>
<p>Let’s start by checking what we have.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_otutable</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">##         CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong</span></span>
<span><span class="co">##  549322   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  522457   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  951      0   0   0       0       0       0       1       0       0       0</span></span>
<span><span class="co">##  244423   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  586076   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  246140   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##         LMEpi24M SLEpi20M AQC1cm AQC4cm AQC7cm NP2 NP3 NP5 TRRsed1 TRRsed2</span></span>
<span><span class="co">##  549322        0        1     27    100    130   1   0   0       0       0</span></span>
<span><span class="co">##  522457        0        0      0      2      6   0   0   0       0       0</span></span>
<span><span class="co">##  951           0        0      0      0      0   0   0   0       0       0</span></span>
<span><span class="co">##  244423        0        0      0     22     29   0   0   0       0       0</span></span>
<span><span class="co">##  586076        0        0      0      2      1   0   0   0       0       0</span></span>
<span><span class="co">##  246140        0        0      0      1      3   0   0   0       0       0</span></span>
<span><span class="co">##         TRRsed3 TS28 TS29 Even1 Even2 Even3</span></span>
<span><span class="co">##  549322       0    0    0     0     0     0</span></span>
<span><span class="co">##  522457       0    0    0     0     0     0</span></span>
<span><span class="co">##  951          0    0    0     0     0     0</span></span>
<span><span class="co">##  244423       0    0    0     0     0     0</span></span>
<span><span class="co">##  586076       0    0    0     0     0     0</span></span>
<span><span class="co">##  246140       0    0    0     0     0     0</span></span>
<span><span class="va">phy_sampledata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">##          X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T</span></span>
<span><span class="co">##  CL3            CL3 ILBC_01        AACGCA                   TGCGTT</span></span>
<span><span class="co">##  CC1            CC1 ILBC_02        AACTCG                   CGAGTT</span></span>
<span><span class="co">##  SV1            SV1 ILBC_03        AACTGT                   ACAGTT</span></span>
<span><span class="co">##  M31Fcsw    M31Fcsw ILBC_04        AAGAGA                   TCTCTT</span></span>
<span><span class="co">##  M11Fcsw    M11Fcsw ILBC_05        AAGCTG                   CAGCTT</span></span>
<span><span class="co">##  M31Plmr    M31Plmr ILBC_07        AATCGT                   ACGATT</span></span>
<span><span class="co">##          Barcode_full_length SampleType</span></span>
<span><span class="co">##  CL3             CTAGCGTGCGT       Soil</span></span>
<span><span class="co">##  CC1             CATCGACGAGT       Soil</span></span>
<span><span class="co">##  SV1             GTACGCACAGT       Soil</span></span>
<span><span class="co">##  M31Fcsw         TCGACATCTCT      Feces</span></span>
<span><span class="co">##  M11Fcsw         CGACTGCAGCT      Feces</span></span>
<span><span class="co">##  M31Plmr         CGAGTCACGAT       Skin</span></span>
<span><span class="co">##                                         Description</span></span>
<span><span class="co">##  CL3       Calhoun South Carolina Pine soil, pH 4.9</span></span>
<span><span class="co">##  CC1       Cedar Creek Minnesota, grassland, pH 6.1</span></span>
<span><span class="co">##  SV1     Sevilleta new Mexico, desert scrub, pH 8.3</span></span>
<span><span class="co">##  M31Fcsw    M3, Day 1, fecal swab, whole body study</span></span>
<span><span class="co">##  M11Fcsw   M1, Day 1, fecal swab, whole body study </span></span>
<span><span class="co">##  M31Plmr    M3, Day 1, right palm, whole body study</span></span>
<span><span class="va">phy_taxtable</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">##         Kingdom        Phylum        Class        Order        Family</span></span>
<span><span class="co">##  549322 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;</span></span>
<span><span class="co">##  522457 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;</span></span>
<span><span class="co">##  951    Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae</span></span>
<span><span class="co">##  244423 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;</span></span>
<span><span class="co">##  586076 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;</span></span>
<span><span class="co">##  246140 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;</span></span>
<span><span class="co">##              Genus                  Species</span></span>
<span><span class="co">##  549322       &lt;NA&gt;                     &lt;NA&gt;</span></span>
<span><span class="co">##  522457       &lt;NA&gt;                     &lt;NA&gt;</span></span>
<span><span class="co">##  951    Sulfolobus Sulfolobusacidocaldarius</span></span>
<span><span class="co">##  244423       &lt;NA&gt;                     &lt;NA&gt;</span></span>
<span><span class="co">##  586076       &lt;NA&gt;                     &lt;NA&gt;</span></span>
<span><span class="co">##  246140       &lt;NA&gt;                     &lt;NA&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Ok, these are all normal data frames which could come from upstream bioinformatics, like OTU tables that come from 16S analysis, and taxonomy tables.</p>
<p>Let’s demo how to create the treeSE object, how it compares to creating phyloseq and how assay in treeSE compares to otu_table in phyloseq.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># Create phyloseq object</span></span>
<span><span class="va">OTU_phy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">phy_otutable</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">as.matrix</span>, taxa_are_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Make OTU table</span></span>
<span><span class="va">TAX_phy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">phy_taxtable</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">as.matrix</span><span class="op">)</span> <span class="co"># Make TAX table</span></span>
<span><span class="va">SAMPLE_phy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">phy_sampledata</span><span class="op">)</span> <span class="co"># Make sample data table</span></span>
<span></span>
<span><span class="va">phy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq.html">phyloseq</a></span><span class="op">(</span><span class="va">OTU_phy</span>, <span class="va">TAX_phy</span>, <span class="va">SAMPLE_phy</span><span class="op">)</span> <span class="co"># Combine into phyloseq object</span></span>
<span><span class="va">phy</span> <span class="co"># Inspect</span></span>
<span><span class="co">##  phyloseq-class experiment-level object</span></span>
<span><span class="co">##  otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]</span></span>
<span><span class="co">##  sample_data() Sample Data:       [ 26 samples by 7 sample variables ]</span></span>
<span><span class="co">##  tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s start by checking our otu table, and see if it is counts or already normalized. We will use the same data frame extracted from the phy object as before.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check if we have counts or normalized data</span></span>
<span></span>
<span><span class="va">phy_otutable</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">##         CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong</span></span>
<span><span class="co">##  549322   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  522457   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  951      0   0   0       0       0       0       1       0       0       0</span></span>
<span><span class="co">##  244423   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  586076   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##  246140   0   0   0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##         LMEpi24M SLEpi20M AQC1cm AQC4cm AQC7cm NP2 NP3 NP5 TRRsed1 TRRsed2</span></span>
<span><span class="co">##  549322        0        1     27    100    130   1   0   0       0       0</span></span>
<span><span class="co">##  522457        0        0      0      2      6   0   0   0       0       0</span></span>
<span><span class="co">##  951           0        0      0      0      0   0   0   0       0       0</span></span>
<span><span class="co">##  244423        0        0      0     22     29   0   0   0       0       0</span></span>
<span><span class="co">##  586076        0        0      0      2      1   0   0   0       0       0</span></span>
<span><span class="co">##  246140        0        0      0      1      3   0   0   0       0       0</span></span>
<span><span class="co">##         TRRsed3 TS28 TS29 Even1 Even2 Even3</span></span>
<span><span class="co">##  549322       0    0    0     0     0     0</span></span>
<span><span class="co">##  522457       0    0    0     0     0     0</span></span>
<span><span class="co">##  951          0    0    0     0     0     0</span></span>
<span><span class="co">##  244423       0    0    0     0     0     0</span></span>
<span><span class="co">##  586076       0    0    0     0     0     0</span></span>
<span><span class="co">##  246140       0    0    0     0     0     0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have counts!</p>
<p>Since TreeSEs can hold many different versions of the OTU table, most commonly either relative abundances or counts, we will need to give our assay (which corresponds to otu_table in Phyloseq) a name and list the different types of assays or transformations we have. In this example we only have one item ‘counts’ in the list.</p>
<p>Let’s convert the data frame to a matrix and make the list of assays.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create TreeSE</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">phy_otutable</span><span class="op">)</span> <span class="co"># Convert to a matrix</span></span>
<span><span class="va">assays</span> <span class="op">&lt;-</span> <span class="fu">SimpleList</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-constructor.html">TreeSummarizedExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="va">assays</span>, </span>
<span>                                colData <span class="op">=</span> <span class="va">phy_sampledata</span>,</span>
<span>                                rowData <span class="op">=</span> <span class="va">phy_taxtable</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s check the different assay names we have.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assayNames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] "counts"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="handling-different-otu-table-normalizations-in-phyloseq-vs-treese" class="level3" data-number="A.5.3"><h3 data-number="A.5.3" class="anchored" data-anchor-id="handling-different-otu-table-normalizations-in-phyloseq-vs-treese">
<span class="header-section-number">A.5.3</span> Handling different OTU table normalizations in phyloseq vs TreeSE</h3>
<p>Adding the assays as a list might seem inconvenient if you only have one type of OTU table (counts in our example), but let’s see why it is actually very convenient to be able to hold multiple assays in one data object.</p>
<p>Here we’ll show an example of how to add relative abundances and CLR normalized OTU tables to your tse assays.</p>
<p>With phyloseq you would need three different phyloseq objects, each taking up 7.7 MB of memory, whilst the tse with the three assays takes up only 18.3 MB.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add another assay that holds the relative abundance normalized OTU table</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="co"># Let's check</span></span>
<span><span class="co">##  List of length 2</span></span>
<span><span class="co">##  names(2): counts relabundance</span></span>
<span></span>
<span><span class="co"># With phyloseq you would need to have two different phyloseq objects</span></span>
<span><span class="va">phy_relab</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/transformcounts.html">transform_sample_counts</a></span><span class="op">(</span><span class="va">phy</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">x</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Let's add clr transformed data just for the fun of it :)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/mia/man/transformAssay.html">transformAssay</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                 assay.type <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                 method <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                 pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="co"># Let's check</span></span>
<span><span class="co">##  List of length 3</span></span>
<span><span class="co">##  names(3): counts relabundance clr</span></span>
<span></span>
<span><span class="co"># With phyloseq you would need to have a third phyloseq object.</span></span>
<span><span class="co"># phy_CLR &lt;- microbiome::transform(phy, 'clr') # Example, don't run</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="subsetting-samples-and-taxa" class="level3" data-number="A.5.4"><h3 data-number="A.5.4" class="anchored" data-anchor-id="subsetting-samples-and-taxa">
<span class="header-section-number">A.5.4</span> Subsetting samples and taxa</h3>
<p><strong>Subsetting samples: subset_samples = indexing columns</strong></p>
<p>Next let’s learn how to subset samples. In phyloseq we use subset_samples command, but since the sample data is stored in columns in the TreeSe, we can access it by indexing columns.</p>
<p>In this section we will remove the “Mock” samples and make new data objects.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_nomock</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html">subset_samples</a></span><span class="op">(</span><span class="va">phy</span>, <span class="op">!</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Mock"</span><span class="op">)</span> <span class="co"># Removing mock samples in phyloseq</span></span>
<span></span>
<span><span class="va">tse_nomock</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">tse</span><span class="op">[</span>,<span class="op">!</span><span class="va">tse</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Mock"</span><span class="op">]</span> <span class="co"># tse uses indexing columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s see what we have now.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_names-methods.html">sample_names</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  [1] 26</span></span>
<span><span class="va">phy_nomock</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_names-methods.html">sample_names</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  [1] 23</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  [1] 26</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tse_nomock</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  [1] 23</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have removed three samples that where SampleType “Mock”.</p>
<p><strong>Subsetting taxa: subset_taxa = indexing rows</strong></p>
<p>Taxa are stored in rows in TreeSE and the TreeSE equivalent to subset_taxa is indexing rows.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_nomock_bacteria</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html">subset_taxa</a></span><span class="op">(</span><span class="va">phy_nomock</span>, <span class="va">Kingdom</span> <span class="op">==</span> <span class="st">"Bacteria"</span><span class="op">)</span></span>
<span><span class="va">tse_nomock_bacteria</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">tse</span><span class="op">[</span><span class="va">tse</span><span class="op">$</span><span class="va">Kingdom</span> <span class="op">==</span> <span class="st">"Bacteria"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">phy_nomock_bacteria</span> <span class="co"># We have 19008 taxa (only bacteria) and before 19216</span></span>
<span><span class="co">##  phyloseq-class experiment-level object</span></span>
<span><span class="co">##  otu_table()   OTU Table:         [ 19008 taxa and 23 samples ]</span></span>
<span><span class="co">##  sample_data() Sample Data:       [ 23 samples by 7 sample variables ]</span></span>
<span><span class="co">##  tax_table()   Taxonomy Table:    [ 19008 taxa by 7 taxonomic ranks ]</span></span>
<span><span class="va">tse_nomock_bacteria</span> </span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 0 26 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(3): counts relabundance clr</span></span>
<span><span class="co">##  rownames(0):</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(7): X.SampleID Primer ... SampleType Description</span></span>
<span><span class="co">##  reducedDimNames(0):</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="calculating-alpha-diversity-estimate_richness-estimatediversity" class="level3" data-number="A.5.5"><h3 data-number="A.5.5" class="anchored" data-anchor-id="calculating-alpha-diversity-estimate_richness-estimatediversity">
<span class="header-section-number">A.5.5</span> Calculating alpha diversity: estimate_richness = estimateDiversity</h3>
<p>Now we know how data stored in TreeSE can be accessed and the TreeSE data objects created. Let’s look at how we can calculate alpha diversity using mia compared to phyloseq package.</p>
<p>The mia command estimateDiversity will return a TreeSE and the results are stored in colData, unlike the phyloseq command that outputs a data frame with just the diversity estimates.</p>
<p>In phyloseq you would need to add the alpha diversity separately to your sample data to keep it safe with the other sample level data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Alpha diversity with phyloseq</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/estimate_richness.html">estimate_richness</a></span><span class="op">(</span><span class="va">phy</span>, measures <span class="op">=</span> <span class="st">"Shannon"</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="co"># Inspect</span></span>
<span><span class="co">##           Shannon</span></span>
<span><span class="co">##  CL3     6.576517</span></span>
<span><span class="co">##  CC1     6.776603</span></span>
<span><span class="co">##  SV1     6.498494</span></span>
<span><span class="co">##  M31Fcsw 3.828368</span></span>
<span><span class="co">##  M11Fcsw 3.287666</span></span>
<span><span class="co">##  M31Plmr 4.289269</span></span>
<span></span>
<span><span class="co"># Add Shannon to the sample_data to keep results safe with other sample data</span></span>
<span><span class="va">phy_sampledata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span></span>
<span><span class="va">phy_sampledata</span><span class="op">$</span><span class="va">shannon</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">$</span><span class="va">Shannon</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">phy_sampledata</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">phy</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span>  <span class="co"># Inspect</span></span>
<span><span class="co">##          X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T</span></span>
<span><span class="co">##  CL3            CL3 ILBC_01        AACGCA                   TGCGTT</span></span>
<span><span class="co">##  CC1            CC1 ILBC_02        AACTCG                   CGAGTT</span></span>
<span><span class="co">##  SV1            SV1 ILBC_03        AACTGT                   ACAGTT</span></span>
<span><span class="co">##  M31Fcsw    M31Fcsw ILBC_04        AAGAGA                   TCTCTT</span></span>
<span><span class="co">##  M11Fcsw    M11Fcsw ILBC_05        AAGCTG                   CAGCTT</span></span>
<span><span class="co">##  M31Plmr    M31Plmr ILBC_07        AATCGT                   ACGATT</span></span>
<span><span class="co">##          Barcode_full_length SampleType</span></span>
<span><span class="co">##  CL3             CTAGCGTGCGT       Soil</span></span>
<span><span class="co">##  CC1             CATCGACGAGT       Soil</span></span>
<span><span class="co">##  SV1             GTACGCACAGT       Soil</span></span>
<span><span class="co">##  M31Fcsw         TCGACATCTCT      Feces</span></span>
<span><span class="co">##  M11Fcsw         CGACTGCAGCT      Feces</span></span>
<span><span class="co">##  M31Plmr         CGAGTCACGAT       Skin</span></span>
<span><span class="co">##                                         Description  shannon</span></span>
<span><span class="co">##  CL3       Calhoun South Carolina Pine soil, pH 4.9 6.576517</span></span>
<span><span class="co">##  CC1       Cedar Creek Minnesota, grassland, pH 6.1 6.776603</span></span>
<span><span class="co">##  SV1     Sevilleta new Mexico, desert scrub, pH 8.3 6.498494</span></span>
<span><span class="co">##  M31Fcsw    M3, Day 1, fecal swab, whole body study 3.828368</span></span>
<span><span class="co">##  M11Fcsw   M1, Day 1, fecal swab, whole body study  3.287666</span></span>
<span><span class="co">##  M31Plmr    M3, Day 1, right palm, whole body study 4.289269</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the tse we will need to specify which assay (which normalization of the OTU table) we want to use, since we have three options now with the counts, relative abundance and CLR. We can check the assay names first.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assayNames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="co"># Check the assay names</span></span>
<span><span class="co">##  [1] "counts"       "relabundance" "clr"</span></span>
<span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/estimateDiversity.html">estimateDiversity</a></span><span class="op">(</span><span class="va">tse</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>, index <span class="op">=</span> <span class="st">"shannon"</span><span class="op">)</span> <span class="co"># Let's use counts</span></span>
<span><span class="co"># Inspect the new colData with added alpha diversity estimate</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">names</span> <span class="co"># shannon has been added to the colData</span></span>
<span><span class="co">##  [1] "X.SampleID"               "Primer"                  </span></span>
<span><span class="co">##  [3] "Final_Barcode"            "Barcode_truncated_plus_T"</span></span>
<span><span class="co">##  [5] "Barcode_full_length"      "SampleType"              </span></span>
<span><span class="co">##  [7] "Description"              "shannon"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we want to extract a data frame that only has the alpha diversity it can be done easily.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Extract</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">matches</a></span><span class="op">(</span><span class="st">"shannon"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="calculating-beta-diversity-ordinate-runmds" class="level3" data-number="A.5.6"><h3 data-number="A.5.6" class="anchored" data-anchor-id="calculating-beta-diversity-ordinate-runmds">
<span class="header-section-number">A.5.6</span> Calculating beta diversity: ordinate = runMDS</h3>
<p>We can calculate PCoA with Bray-Curtis distances in phyloseq using the ordinate command. The beta diversity calculation in mia outputs a TreeSE with a new type of data, reduced dimensions or reducedDim.</p>
<p>Here we will use the scater package that runs the PCoA with runMDS. (PCoA and MDS mean the same thing)</p>
<p>In phyloseq you would again need to add the dimensions to the sample data if you want to keep them safe with other metadata.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run PCoA on the relative abundance data and store in phy_ord list</span></span>
<span><span class="va">phy_ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ordinate.html">ordinate</a></span><span class="op">(</span><span class="va">phy_relab</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>, distance <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: scuttle</span></span>
<span></span>
<span><span class="co"># Ordinate with runMDS and implement the vegan's Bray-Curtis dissimilarity distance calculation</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runMDS.html">runMDS</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>              FUN <span class="op">=</span> <span class="fu">vegan</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span>,</span>
<span>              method <span class="op">=</span> <span class="st">"bray"</span>,</span>
<span>              assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"MDS_bray"</span>,</span>
<span>              ncomponents <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># Let's also define how many dimensions</span></span>
<span><span class="va">tse</span> <span class="co"># Inspect, now we have new reducedDim "MDS_bray"</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 19216 26 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(3): counts relabundance clr</span></span>
<span><span class="co">##  rownames(19216): 549322 522457 ... 200359 271582</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(8): X.SampleID Primer ... Description shannon</span></span>
<span><span class="co">##  reducedDimNames(1): MDS_bray</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="plotting-ordinations-plot_ordination-plotreduceddim" class="level3" data-number="A.5.7"><h3 data-number="A.5.7" class="anchored" data-anchor-id="plotting-ordinations-plot_ordination-plotreduceddim">
<span class="header-section-number">A.5.7</span> Plotting ordinations: plot_ordination = plotReducedDim</h3>
<p>phyloseq has it’s own plotting fuction for ordinations.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/plot_ordination.html">plot_ordination</a></span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">phy</span>, ordination <span class="op">=</span> <span class="va">phy_ord</span>, color <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="97_extra_materials_files/figure-html/plot_ord-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It is also easy to plot the ordination stored in reducedDim in the tse using the plotReducedDim function. We can first check what the name of the Bray-Curtis MDS/PCoA was incase we forgot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check reduced dim names</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDimNames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="co">##  [1] "MDS_bray"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Ok, let’s plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"MDS_bray"</span>, color_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="97_extra_materials_files/figure-html/plot_red_dim-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># The sign is given arbitrarily. We can change it to match the plot_ordination</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"MDS_bray"</span>, color_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="97_extra_materials_files/figure-html/plot_red_dim-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="agglomerating-taxa-tax_glommergefeaturesbyrank" class="level3" data-number="A.5.8"><h3 data-number="A.5.8" class="anchored" data-anchor-id="agglomerating-taxa-tax_glommergefeaturesbyrank">
<span class="header-section-number">A.5.8</span> Agglomerating taxa: tax_glom=mergeFeaturesByRank</h3>
<p>Often you might want to study your data using different taxonomic ranks, for example check if you see differences in the abundances of higher taxonomic levels.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">phy_fam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_glom.html">tax_glom</a></span><span class="op">(</span><span class="va">phy</span>, taxrank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This family level data object can again be conveniently stored in a tse object under altExp.</p>
<p>Tax_glom removes the taxa which have not been assigned to the level given in taxrank by default (NArm = TRUE). So we will add the onRankOnly = TRUE to mergeFeaturesByRank function which is equivalent to the default behaviour of tax_glom.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/mia/man/agglomerate-methods.html">mergeFeaturesByRank</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                      rank <span class="op">=</span> <span class="st">"Family"</span>,</span>
<span>                      onRankOnly <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># Let's add the equivalent option to NArm =TRUE</span></span>
<span>                      agglomerateTree <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  Warning in .local(x, ...): 'agglomerateByRank' is deprecated.</span></span>
<span><span class="co">##  Use 'mergeFeaturesByRank' instead.</span></span>
<span><span class="co">##  See help("Deprecated") and help("Now agglomerateByRank is deprecated. Use mergeFeaturesByRank instead.-deprecated").</span></span>
<span><span class="co">##  Warning: 'clr' includes negative values.</span></span>
<span><span class="co">##  Agglomeration of it might lead to meaningless values.</span></span>
<span><span class="co">##  Check the assay, and consider doing transformation again manually with agglomerated data.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"Family"</span><span class="op">)</span></span>
<span><span class="co">##  class: TreeSummarizedExperiment </span></span>
<span><span class="co">##  dim: 334 26 </span></span>
<span><span class="co">##  metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">##  assays(3): counts relabundance clr</span></span>
<span><span class="co">##  rownames(334): Sulfolobaceae SAGMA-X ... Dethiosulfovibrionaceae</span></span>
<span><span class="co">##    Thermodesulfobiaceae</span></span>
<span><span class="co">##  rowData names(7): Kingdom Phylum ... Genus Species</span></span>
<span><span class="co">##  colnames(26): CL3 CC1 ... Even2 Even3</span></span>
<span><span class="co">##  colData names(8): X.SampleID Primer ... Description shannon</span></span>
<span><span class="co">##  reducedDimNames(1): MDS_bray</span></span>
<span><span class="co">##  mainExpName: NULL</span></span>
<span><span class="co">##  altExpNames(0):</span></span>
<span><span class="co">##  rowLinks: NULL</span></span>
<span><span class="co">##  rowTree: NULL</span></span>
<span><span class="co">##  colLinks: NULL</span></span>
<span><span class="co">##  colTree: NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="cheatsheet" class="level3" data-number="A.5.9"><h3 data-number="A.5.9" class="anchored" data-anchor-id="cheatsheet">
<span class="header-section-number">A.5.9</span> Cheatsheet</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    Functionality <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"Access sample data"</span>, <span class="co"># Row 1</span></span>
<span>      <span class="st">"Access tax table"</span>, <span class="co"># Row 2</span></span>
<span>      <span class="st">"Access OTU table"</span>, </span>
<span>      <span class="st">"Build data object"</span>,</span>
<span>      <span class="st">"Calculate alpha diversity"</span>,</span>
<span>      <span class="st">"Calculate beta diversity"</span>,</span>
<span>      <span class="st">"Plot ordination"</span>,</span>
<span>      <span class="st">"Subset taxa"</span>,</span>
<span>      <span class="st">"Subset samples"</span>,</span>
<span>      <span class="st">"Aggromerate taxa"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    phyloseq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"sample_data()"</span>,</span>
<span>      <span class="st">"tax_table()"</span>,</span>
<span>      <span class="st">"otu_table()"</span>,</span>
<span>      <span class="st">"phyloseq()"</span>,</span>
<span>      <span class="st">"estimate_richness()"</span>,</span>
<span>      <span class="st">"ordinate()"</span>,</span>
<span>      <span class="st">"plot_ordination()"</span>,</span>
<span>      <span class="st">"subset_taxa()"</span>,</span>
<span>      <span class="st">"subset_samples()"</span>,</span>
<span>      <span class="st">"tax_glom()"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    miaTreeSE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"Index columns"</span>,</span>
<span>      <span class="st">"Index rows"</span>,</span>
<span>      <span class="st">"assays()"</span>,</span>
<span>      <span class="st">"TreeSummarizedExperiment()"</span>,</span>
<span>      <span class="st">"estimateDiversity()"</span>,</span>
<span>      <span class="st">"runMDS()"</span>,</span>
<span>      <span class="st">"plotReducedDim()"</span>,</span>
<span>      <span class="st">"Index rows"</span>,</span>
<span>      <span class="st">"Index columns"</span>,</span>
<span>      <span class="st">"mergeFeaturesByRank()"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">df2</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    Data_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"OTU table"</span>, <span class="co"># Row 1</span></span>
<span>                  <span class="st">"Taxonomy table"</span>, <span class="co"># Row2</span></span>
<span>                  <span class="st">"Sample data table"</span><span class="op">)</span>, <span class="co"># Row 3</span></span>
<span>    phyloseq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"otu_table"</span>, <span class="co"># Row 1</span></span>
<span>                 <span class="st">"tax_table"</span>, <span class="co"># Row2</span></span>
<span>                 <span class="st">"sample_data"</span><span class="op">)</span>,<span class="co"># Row 3</span></span>
<span>    TreeSE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"assay"</span>, <span class="co"># Row 1</span></span>
<span>               <span class="st">"rowData"</span>, <span class="co"># Row2</span></span>
<span>               <span class="st">"colData"</span><span class="op">)</span> <span class="co"># Row 3</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 35%">
<col style="width: 27%">
<col style="width: 36%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Functionality</th>
<th style="text-align: left;">phyloseq</th>
<th style="text-align: left;">miaTreeSE</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Access sample data</td>
<td style="text-align: left;">sample_data()</td>
<td style="text-align: left;">Index columns</td>
</tr>
<tr class="even">
<td style="text-align: left;">Access tax table</td>
<td style="text-align: left;">tax_table()</td>
<td style="text-align: left;">Index rows</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Access OTU table</td>
<td style="text-align: left;">otu_table()</td>
<td style="text-align: left;">assays()</td>
</tr>
<tr class="even">
<td style="text-align: left;">Build data object</td>
<td style="text-align: left;">phyloseq()</td>
<td style="text-align: left;">TreeSummarizedExperiment()</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Calculate alpha diversity</td>
<td style="text-align: left;">estimate_richness()</td>
<td style="text-align: left;">estimateDiversity()</td>
</tr>
<tr class="even">
<td style="text-align: left;">Calculate beta diversity</td>
<td style="text-align: left;">ordinate()</td>
<td style="text-align: left;">runMDS()</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Plot ordination</td>
<td style="text-align: left;">plot_ordination()</td>
<td style="text-align: left;">plotReducedDim()</td>
</tr>
<tr class="even">
<td style="text-align: left;">Subset taxa</td>
<td style="text-align: left;">subset_taxa()</td>
<td style="text-align: left;">Index rows</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Subset samples</td>
<td style="text-align: left;">subset_samples()</td>
<td style="text-align: left;">Index columns</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aggromerate taxa</td>
<td style="text-align: left;">tax_glom()</td>
<td style="text-align: left;">mergeFeaturesByRank()</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">df2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;">Data_type</th>
<th style="text-align: left;">phyloseq</th>
<th style="text-align: left;">TreeSE</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">OTU table</td>
<td style="text-align: left;">otu_table</td>
<td style="text-align: left;">assay</td>
</tr>
<tr class="even">
<td style="text-align: left;">Taxonomy table</td>
<td style="text-align: left;">tax_table</td>
<td style="text-align: left;">rowData</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Sample data table</td>
<td style="text-align: left;">sample_data</td>
<td style="text-align: left;">colData</td>
</tr>
</tbody>
</table>
</div>
</div>



</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/98_exercises.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exercises</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/90_acknowledgments.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>