<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 21&nbsp; Exercises</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/97_extra_materials.html" rel="next">
<link href="../pages/95_resources.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../docs/pages/images/mia_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Orchestrating Microbiome Analysis</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">TEST</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-bi-github">
<li>
    <a class="dropdown-item" href="https://github.com/microbiome/OMA/">
 <span class="dropdown-text">Source Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://microbiome.github.io/OMA/docs/devel/">
 <span class="dropdown-text">Browse version <code>devel</code></span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/80_training.html">Training</a></li><li class="breadcrumb-item"><a href="../pages/98_exercises.html"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#basics-of-rbioconductor" id="toc-basics-of-rbioconductor" class="nav-link active" data-scroll-target="#basics-of-rbioconductor"><span class="header-section-number">21.1</span> Basics of R/Bioconductor</a></li>
  <li>
<a href="#workflows" id="toc-workflows" class="nav-link" data-scroll-target="#workflows"><span class="header-section-number">21.2</span> Workflows</a>
  <ul class="collapse">
<li><a href="#reproducible-reporting-with-quarto" id="toc-reproducible-reporting-with-quarto" class="nav-link" data-scroll-target="#reproducible-reporting-with-quarto"><span class="header-section-number">21.2.1</span> Reproducible reporting with Quarto</a></li>
  </ul>
</li>
  <li>
<a href="#data-containers-treese" id="toc-data-containers-treese" class="nav-link" data-scroll-target="#data-containers-treese"><span class="header-section-number">21.3</span> Data containers: TreeSE</a>
  <ul class="collapse">
<li><a href="#sec-construct-TreeSE" id="toc-sec-construct-TreeSE" class="nav-link" data-scroll-target="#sec-construct-TreeSE"><span class="header-section-number">21.3.1</span> Constructing a data object</a></li>
  <li><a href="#importing-data" id="toc-importing-data" class="nav-link" data-scroll-target="#importing-data"><span class="header-section-number">21.3.2</span> Importing data</a></li>
  <li><a href="#preliminary-exploration" id="toc-preliminary-exploration" class="nav-link" data-scroll-target="#preliminary-exploration"><span class="header-section-number">21.3.3</span> Preliminary exploration</a></li>
  <li><a href="#assay-retrieval" id="toc-assay-retrieval" class="nav-link" data-scroll-target="#assay-retrieval"><span class="header-section-number">21.3.4</span> Assay retrieval</a></li>
  <li><a href="#sample-information" id="toc-sample-information" class="nav-link" data-scroll-target="#sample-information"><span class="header-section-number">21.3.5</span> Sample information</a></li>
  <li><a href="#feature-information" id="toc-feature-information" class="nav-link" data-scroll-target="#feature-information"><span class="header-section-number">21.3.6</span> Feature information</a></li>
  <li><a href="#other-elements" id="toc-other-elements" class="nav-link" data-scroll-target="#other-elements"><span class="header-section-number">21.3.7</span> Other elements</a></li>
  </ul>
</li>
  <li>
<a href="#data-manipulation" id="toc-data-manipulation" class="nav-link" data-scroll-target="#data-manipulation"><span class="header-section-number">21.4</span> Data manipulation</a>
  <ul class="collapse">
<li><a href="#subsetting" id="toc-subsetting" class="nav-link" data-scroll-target="#subsetting"><span class="header-section-number">21.4.1</span> Subsetting</a></li>
  <li><a href="#library-sizes" id="toc-library-sizes" class="nav-link" data-scroll-target="#library-sizes"><span class="header-section-number">21.4.2</span> Library sizes</a></li>
  <li><a href="#prevalent-and-core-taxonomic-features" id="toc-prevalent-and-core-taxonomic-features" class="nav-link" data-scroll-target="#prevalent-and-core-taxonomic-features"><span class="header-section-number">21.4.3</span> Prevalent and core taxonomic features</a></li>
  <li><a href="#data-exploration" id="toc-data-exploration" class="nav-link" data-scroll-target="#data-exploration"><span class="header-section-number">21.4.4</span> Data exploration</a></li>
  <li><a href="#other-functions" id="toc-other-functions" class="nav-link" data-scroll-target="#other-functions"><span class="header-section-number">21.4.5</span> Other functions</a></li>
  <li><a href="#assay-transformation" id="toc-assay-transformation" class="nav-link" data-scroll-target="#assay-transformation"><span class="header-section-number">21.4.6</span> Assay transformation</a></li>
  </ul>
</li>
  <li>
<a href="#abundance-tables" id="toc-abundance-tables" class="nav-link" data-scroll-target="#abundance-tables"><span class="header-section-number">21.5</span> Abundance tables</a>
  <ul class="collapse">
<li><a href="#taxonomic-levels" id="toc-taxonomic-levels" class="nav-link" data-scroll-target="#taxonomic-levels"><span class="header-section-number">21.5.1</span> Taxonomic levels</a></li>
  <li><a href="#alternative-experiments" id="toc-alternative-experiments" class="nav-link" data-scroll-target="#alternative-experiments"><span class="header-section-number">21.5.2</span> Alternative experiments</a></li>
  </ul>
</li>
  <li>
<a href="#community-alpha-diversity" id="toc-community-alpha-diversity" class="nav-link" data-scroll-target="#community-alpha-diversity"><span class="header-section-number">21.6</span> Community (alpha) diversity</a>
  <ul class="collapse">
<li><a href="#estimation" id="toc-estimation" class="nav-link" data-scroll-target="#estimation"><span class="header-section-number">21.6.1</span> Estimation</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization"><span class="header-section-number">21.6.2</span> Visualization</a></li>
  <li><a href="#correlation" id="toc-correlation" class="nav-link" data-scroll-target="#correlation"><span class="header-section-number">21.6.3</span> Correlation</a></li>
  <li><a href="#differences-between-groups" id="toc-differences-between-groups" class="nav-link" data-scroll-target="#differences-between-groups"><span class="header-section-number">21.6.4</span> Differences between groups</a></li>
  </ul>
</li>
  <li>
<a href="#community-similarity" id="toc-community-similarity" class="nav-link" data-scroll-target="#community-similarity"><span class="header-section-number">21.7</span> Community similarity</a>
  <ul class="collapse">
<li><a href="#reduced-dimensions-retrieval" id="toc-reduced-dimensions-retrieval" class="nav-link" data-scroll-target="#reduced-dimensions-retrieval"><span class="header-section-number">21.7.1</span> Reduced dimensions retrieval</a></li>
  <li><a href="#visualization-basics-with-pca" id="toc-visualization-basics-with-pca" class="nav-link" data-scroll-target="#visualization-basics-with-pca"><span class="header-section-number">21.7.2</span> Visualization basics with PCA</a></li>
  <li><a href="#principal-coordinate-analysis-pcoa" id="toc-principal-coordinate-analysis-pcoa" class="nav-link" data-scroll-target="#principal-coordinate-analysis-pcoa"><span class="header-section-number">21.7.3</span> Principal Coordinate Analysis (PCoA)</a></li>
  <li><a href="#permanova-analysis" id="toc-permanova-analysis" class="nav-link" data-scroll-target="#permanova-analysis"><span class="header-section-number">21.7.4</span> PERMANOVA analysis</a></li>
  <li><a href="#redundancy-analysis-rda" id="toc-redundancy-analysis-rda" class="nav-link" data-scroll-target="#redundancy-analysis-rda"><span class="header-section-number">21.7.5</span> Redundancy analysis (RDA)</a></li>
  <li><a href="#beta-diversity-analysis" id="toc-beta-diversity-analysis" class="nav-link" data-scroll-target="#beta-diversity-analysis"><span class="header-section-number">21.7.6</span> Beta diversity analysis</a></li>
  </ul>
</li>
  <li>
<a href="#differential-abundance" id="toc-differential-abundance" class="nav-link" data-scroll-target="#differential-abundance"><span class="header-section-number">21.8</span> Differential abundance</a>
  <ul class="collapse">
<li><a href="#standard-analysis-with-aldex2" id="toc-standard-analysis-with-aldex2" class="nav-link" data-scroll-target="#standard-analysis-with-aldex2"><span class="header-section-number">21.8.1</span> Standard analysis with ALDEx2</a></li>
  <li><a href="#sec-control-daa-confounders" id="toc-sec-control-daa-confounders" class="nav-link" data-scroll-target="#sec-control-daa-confounders"><span class="header-section-number">21.8.2</span> Controlling for confounders</a></li>
  <li><a href="#sec-compare-daa-methods" id="toc-sec-compare-daa-methods" class="nav-link" data-scroll-target="#sec-compare-daa-methods"><span class="header-section-number">21.8.3</span> Comparing methods</a></li>
  <li><a href="#workflow-1" id="toc-workflow-1" class="nav-link" data-scroll-target="#workflow-1"><span class="header-section-number">21.8.4</span> Workflow 1</a></li>
  <li><a href="#workflow-2" id="toc-workflow-2" class="nav-link" data-scroll-target="#workflow-2"><span class="header-section-number">21.8.5</span> Workflow 2</a></li>
  </ul>
</li>
  <li>
<a href="#visualization-1" id="toc-visualization-1" class="nav-link" data-scroll-target="#visualization-1"><span class="header-section-number">21.9</span> Visualization</a>
  <ul class="collapse">
<li><a href="#multivariate-ordination" id="toc-multivariate-ordination" class="nav-link" data-scroll-target="#multivariate-ordination"><span class="header-section-number">21.9.1</span> Multivariate ordination</a></li>
  <li><a href="#heatmap-visualization" id="toc-heatmap-visualization" class="nav-link" data-scroll-target="#heatmap-visualization"><span class="header-section-number">21.9.2</span> Heatmap visualization</a></li>
  </ul>
</li>
  <li>
<a href="#multiomics" id="toc-multiomics" class="nav-link" data-scroll-target="#multiomics"><span class="header-section-number">21.10</span> Multiomics</a>
  <ul class="collapse">
<li><a href="#basic-exploration" id="toc-basic-exploration" class="nav-link" data-scroll-target="#basic-exploration"><span class="header-section-number">21.10.1</span> Basic exploration</a></li>
  <li><a href="#experiment-agglomeration" id="toc-experiment-agglomeration" class="nav-link" data-scroll-target="#experiment-agglomeration"><span class="header-section-number">21.10.2</span> Experiment agglomeration</a></li>
  <li><a href="#experiment-transformation" id="toc-experiment-transformation" class="nav-link" data-scroll-target="#experiment-transformation"><span class="header-section-number">21.10.3</span> Experiment transformation</a></li>
  <li><a href="#assay-extraction" id="toc-assay-extraction" class="nav-link" data-scroll-target="#assay-extraction"><span class="header-section-number">21.10.4</span> Assay extraction</a></li>
  <li><a href="#mae-reconstruction" id="toc-mae-reconstruction" class="nav-link" data-scroll-target="#mae-reconstruction"><span class="header-section-number">21.10.5</span> MAE reconstruction</a></li>
  <li><a href="#cross-correlation-analysis" id="toc-cross-correlation-analysis" class="nav-link" data-scroll-target="#cross-correlation-analysis"><span class="header-section-number">21.10.6</span> Cross-correlation analysis</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/80_training.html">Training</a></li><li class="breadcrumb-item"><a href="../pages/98_exercises.html"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-exercises" class="quarto-section-identifier"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>Here you can find assignments on different topics.</p>
<p><strong>Tips for exercises:</strong></p>
<ul>
<li>Add comments that explain what each line or lines of code do. This helps you and others understand your code and find bugs. Furthermore, it is easier for you to reuse the code, and it promotes transparency.</li>
<li>Interpret results by comparing them to literature. List main findings, so that results can easily be understood by others without advanced knowledge on data analytics.</li>
<li>Avoid hard-coding. Use variables which get values in the beginning of the pipeline. That way it is easier for you to modify parameters and reuse the code.</li>
</ul>
<section id="basics-of-rbioconductor" class="level2" data-number="21.1"><h2 data-number="21.1" class="anchored" data-anchor-id="basics-of-rbioconductor">
<span class="header-section-number">21.1</span> Basics of R/Bioconductor</h2>
<p>Bioconductor training material has been contributed to <a href="https://carpentries.org/">Carpentries</a>. You can check the following lessons for basic background of R and Bioconductor.</p>
<ul>
<li><p><a href="https://carpentries-incubator.github.io/bioc-intro/">Introduction to data analysis with R and Bioconductor</a></p></li>
<li><p><a href="https://carpentries-incubator.github.io/bioc-project/">Introduction to the Bioconductor project</a></p></li>
</ul></section><section id="workflows" class="level2" data-number="21.2"><h2 data-number="21.2" class="anchored" data-anchor-id="workflows">
<span class="header-section-number">21.2</span> Workflows</h2>
<section id="reproducible-reporting-with-quarto" class="level3" data-number="21.2.1"><h3 data-number="21.2.1" class="anchored" data-anchor-id="reproducible-reporting-with-quarto">
<span class="header-section-number">21.2.1</span> Reproducible reporting with Quarto</h3>
<p>The following batch of exercises walks you through typical use cases of Quarto in RStudio. Before heading to the exercises, it is recommended to read the <a href="https://quarto.org/docs/tools/rstudio.html">Quarto guidelines for RStudio</a></p>
<section id="new-document" class="level4" data-number="21.2.1.1"><h4 data-number="21.2.1.1" class="anchored" data-anchor-id="new-document">
<span class="header-section-number">21.2.1.1</span> New document</h4>
<p>This exercise gets you started with creating a Quarto document and adding text to it with typing conventions borrowed from the <a href="https://quarto.org/docs/authoring/markdown-basics.html">markdown syntax</a>. Feel free to render the document with the <strong>Render</strong> button after each step to see the changes in the final report.</p>
<ol type="1">
<li>Open RStudio and create a new Quarto file named <code>My first Quarto</code>.</li>
<li>Add the subtitle <code>My first section</code> and write some text of your choice underneath. You can choose the level of headings by the number of preceding hashes (<code>#</code>).</li>
<li>Add a subsection named <code>List of items</code> and list three items underneath, both ordered and unordered. You can initialize items with numbers (<code>1.</code>, <code>2.</code>, <code>3.</code>, …) or stars (<code>*</code>) for the ordered and unordered case, respectively.</li>
<li>Add another subsection named <code>Link to web</code> and add a clickable link to the <a href="https://microbiome.github.io/OMA/">OMA book</a>, using the pattern <code>[text](url)</code>.</li>
<li>Render the document and check its appearance</li>
</ol>
<p>Nice start! You are now able to create a Quarto document, understand its syntax and can render it into a reproducible report. If you got stuck, you can look up the docs on <a href="https://quarto.org/docs/tools/rstudio.html#creating-documents">creating</a> and <a href="https://quarto.org/docs/tools/rstudio.html#render-and-preview">rendering</a> Quarto documents.</p>
</section><section id="code-chunks" class="level4" data-number="21.2.1.2"><h4 data-number="21.2.1.2" class="anchored" data-anchor-id="code-chunks">
<span class="header-section-number">21.2.1.2</span> Code chunks</h4>
<p>While customizable text is nothing new by itself, the advantage of Quarto (and previously Rmakdown) is to combine text with code in R or other programming languages, so that both the analytical pipeline and verbal description can be put in one place. In this exercise, we learn how to write and run code in Quarto.</p>
<ol type="1">
<li>Open RStudio and create a new Quarto file.<br>
</li>
<li>Initialize a code chunk by pressing <code>alt</code> + <code>cmd</code> + <code>i</code> and define the variables <code>A &lt;- "my name"</code> and <code>B &lt;- 0</code> in it.<br>
</li>
<li>Write the text <code>Below is my first code chunk</code> just above the code chunk.<br>
</li>
<li>Initialize one more code chunk and add 100 to the variable <code>B</code> in it.<br>
</li>
<li>Write the text <code>Below I change variable B</code> just above the second chunk.<br>
</li>
<li>
<strong>Extra</strong>: Write the following line of text: <code>my name is A and I am B years old</code>, where A and B are variables defined in the code chunks upstream and change if those variables are modified. Be aware that inline code can be added as <code>&gt; r my_inline_command</code> (without <code>&gt;</code>).</li>
</ol>
<p>Good job. You are now able to combine text and code in a Quarto document. If you got stuck, you can refer to the Quarto docs on <a href="https://quarto.org/docs/visual-editor/technical.html#code-chunks">using code chunks</a>.</p>
</section><section id="knitr-options" class="level4" data-number="21.2.1.3"><h4 data-number="21.2.1.3" class="anchored" data-anchor-id="knitr-options">
<span class="header-section-number">21.2.1.3</span> Knitr options</h4>
<p>Code chunks can be greatly customized in terms of visibility and execution, output size and location and much more. This is possible with the knitr chunk options, which usually appear at the beginning of the target chunk with the syntax <code>#| option-name: value</code>, also described <a href="https://quarto.org/docs/computations/r.html#chunk-options">here</a>. In this exercise, we explore part of the knitr potential.</p>
<ol type="1">
<li>Open RStudio and create a new Quarto file.<br>
</li>
<li>Initialize three code chunks and label them as <code>setup</code>, <code>fig-box</code> and <code>tbl-coldata</code>, respectively. Remember that the name of a chunk can be specified with the <code>label</code> option.<br>
</li>
<li>Write the following code in the corresponding chunk and render the document.</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># setup</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span>
<span></span>
<span><span class="co"># this line sets some options for all the chunks (global chunk options)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>message <span class="op">=</span> <span class="cn">FALSE</span>, warning <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># fig-box</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span> <span class="op">~</span> <span class="va">tse</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># tbl-coldata</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="4" type="1">
<li>Set <code>include: false</code> in the <code>setup</code> chunk, <code>fig-width: 10</code> in the <code>fig-box</code> chunk and <code>echo: false</code> in the <code>tbl-coldata</code> chunk. Render the document again and find the differences from before.<br>
</li>
<li>Add the options <code>fig-cap</code> and <code>tab-cap</code> to the <code>fig-box</code> and <code>tbl-coldata</code> chunks, respectively. They require some text input, which makes for the caption of the figures or tables.<br>
</li>
<li>
<strong>Extra</strong>: Create a cross-reference to <code>fig-box</code> and <code>tbl-coldata</code> in the text above the respective code chunk. You can do that with the syntax <code>@chunk-name</code>.<br>
</li>
<li>
<strong>Extra</strong>: Define a custom folder for storing figures with <code>fig-path</code>. Insert it in <code>knitr::opts_chunk$set</code>, so that it applies globally to all the figures generated in the document.</li>
</ol>
<p>Congratulations! You are now familiar with the great potential and flexibility of knitr chunk options. An exhaustive list of available options can be found in the <a href="https://yihui.org/knitr/options/">knitr documentation</a>.</p>
</section><section id="yaml-instructions" class="level4" data-number="21.2.1.4"><h4 data-number="21.2.1.4" class="anchored" data-anchor-id="yaml-instructions">
<span class="header-section-number">21.2.1.4</span> YAML instructions</h4>
<p>The box at the beginning of every Quarto document contains yaml options that let you define the metadata of the document. They will affect the appearance of the document when it is rendered. By default, the box includes yaml options for the title, format and editor to be used, but much more information on layout, code execution and figures can be specified. A comprehensive list of yaml options is available <a href="https://quarto.org/docs/reference/formats/html.html">here</a>. In this exercise, we will get a tiny taste of such functionality.</p>
<ol type="1">
<li>Open RStudio and create a new Quarto file.<br>
</li>
<li>In the yaml box at the beginning of the document, change the title from <code>Untitled</code> to <code>My first Quarto</code>.<br>
</li>
<li>In the same box, add the two lines <code>author</code> and <code>date</code> followed by your name and today’s date, respectively.<br>
</li>
<li>Render the document and check its appearance.<br>
</li>
<li>
<strong>Extra</strong>: Set <code>toc: true</code> to produce a table of contents. This line should follow <code>format</code> and <code>html</code> at the second level of indentation.</li>
</ol>
<p>Well done! Now you are able to specify yaml options and understand how they affect your Quarto document. If you got stuck, you can check <a href="https://quarto.org/docs/tools/rstudio.html#yaml-intelligence">this section</a> of the Quarto documentation.</p>
</section><section id="quarto-parameters" class="level4" data-number="21.2.1.5"><h4 data-number="21.2.1.5" class="anchored" data-anchor-id="quarto-parameters">
<span class="header-section-number">21.2.1.5</span> Quarto parameters</h4>
<p>An advanced feature of Quarto consists of execution parameters, which are externally pre-defined variables that are also accessible in the Quarto document. They can be specified in the yaml box as <code>params</code>. Here we learn how to use them.</p>
<ol type="1">
<li>Open RStudio and create a new Quarto file.<br>
</li>
<li>In the yaml box at the beginning of the document, add a line named <code>params</code> followed by an indented line with <code>gamma: 10</code><br>
</li>
<li>Initialize a code chunk and type <code>str(params$gamma)</code> in it.<br>
</li>
<li>Render the document and check what happened.<br>
</li>
<li>Define one more parameter <code>beta: 3</code> and multiply <code>gamma</code> by <code>beta</code> in a code chunk below.<br>
</li>
<li>Render the document again and check what happened.</li>
</ol>
<p>Well done! You can now use an advanced feature of Quarto such as parameters. If you got stuck, <a href="https://quarto.org/docs/computations/parameters.html#knitr">here</a> you can find more information about parameter definition and usage.</p>
</section></section></section><section id="data-containers-treese" class="level2" data-number="21.3"><h2 data-number="21.3" class="anchored" data-anchor-id="data-containers-treese">
<span class="header-section-number">21.3</span> Data containers: TreeSE</h2>
<p>TreeSE containers represent the working unit of the mia package. In the following exercises we learn how to construct, explore and work with them. A few demo datasets can be imported with mia and can be accessed as explained in chapter <a href="04_containers.html#sec-example-data" class="quarto-xref"><span>Section 2.3</span></a>.</p>
<section id="sec-construct-TreeSE" class="level3" data-number="21.3.1"><h3 data-number="21.3.1" class="anchored" data-anchor-id="sec-construct-TreeSE">
<span class="header-section-number">21.3.1</span> Constructing a data object</h3>
<p>Here we cover how to construct a TreeSE from CSV files, using the components of OKeefeDSData from the microbiomeDataSets package as an example dataset.</p>
<ol type="1">
<li>Fetch or download the files in <a href="https://github.com/microbiome/data/tree/main/OKeefeDSData">this directory</a>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">url_to_assay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">"https://github.com/microbiome/data/raw/main/OKeefeDSData/okeefe_assay.csv"</span><span class="op">)</span></span>
<span><span class="va">url_to_coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">"https://github.com/microbiome/data/raw/main/OKeefeDSData/okeefe_coldata.csv"</span><span class="op">)</span></span>
<span><span class="va">url_to_rowdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">"https://github.com/microbiome/data/raw/main/OKeefeDSData/okeefe_rowdata.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Read in the csv files with <code>read.csv</code> and store them into the variables <code>assays</code>, <code>rowdata</code> and <code>coldata</code>, respectively.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">okeefe_assay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">url_to_assay</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">okeefe_coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">url_to_coldata</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">okeefe_rowdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">url_to_rowdata</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Create a TreeSE from the individual components with <code>TreeSummarizedExperiment</code>. Note that the function requires three arguments: assays, rowData and colData, to which you can give the appropriate item.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-constructor.html">TreeSummarizedExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">okeefe_assay</span><span class="op">)</span>,</span>
<span>                                colData <span class="op">=</span> <span class="va">okeefe_coldata</span>,</span>
<span>                                rowData <span class="op">=</span> <span class="va">okeefe_rowdata</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Check that importing is done correctly. E.g., choose random samples and features, and check that their values equal between raw files and TreeSE.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">okeefe_assay</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Usefuls functions: DataFrame, TreeSummarizedExperiment, matrix, rownames, colnames, SimpleList</p>
</section><section id="importing-data" class="level3" data-number="21.3.2"><h3 data-number="21.3.2" class="anchored" data-anchor-id="importing-data">
<span class="header-section-number">21.3.2</span> Importing data</h3>
<p>Raw data of different types can be imported as a TreeSE with a number of functions explained in chapter <a href="04_containers.html#sec-import-from-file" class="quarto-xref"><span>Section 2.4.2</span></a>. You can also check the <a href="https://microbiome.github.io/mia/reference/index.html">function reference in the mia package</a>.</p>
<ol type="1">
<li>Get familiar with the <a href="https://github.com/microbiome/data">microbiome data repository</a> and read the instructions in its <a href="https://github.com/microbiome/data#training-microbiome-datasets">README</a> to import and construct datasets from there.<br>
</li>
<li>Import data by using mia. (functions: <a href="https://microbiome.github.io/mia/reference/loadFromMetaphlan.html">importMetaPhlAn</a> | <a href="https://microbiome.github.io/mia/reference/importMothur.html">importMothur</a> | <a href="https://microbiome.github.io/mia/reference/importQIIME2.html">importQIIME2</a> | <a href="https://microbiome.github.io/mia/reference/makeTreeSEFromBiom.html">makeTreeSummarizedExperimentFromBiom</a> | <a href="https://microbiome.github.io/mia/reference/makeTreeSEFromDADA2.html">makeTreeSummarizedExperimentFromDADA2</a> …)</li>
<li>Try out conversions between TreeSE and phyloseq data containers (<a href="https://microbiome.github.io/mia/reference/makeTreeSEFromPhyloseq.html">makeTreeSummarizedExperimentFromPhyloseq</a>; <a href="https://microbiome.github.io/mia/reference/makePhyloseqFromTreeSE.html">makephyloseqFromTreeSummarizedExperiment</a>)</li>
</ol></section><section id="preliminary-exploration" class="level3" data-number="21.3.3"><h3 data-number="21.3.3" class="anchored" data-anchor-id="preliminary-exploration">
<span class="header-section-number">21.3.3</span> Preliminary exploration</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Get a summary about the TreeSE with <code>summary</code>. What is the mean count across samples? How many features recur only once (singletons)?</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Check the dimensions of the TreeSE with <code>dim</code> or alternatively with <code>nrow</code> and <code>ncol</code>. How many samples and features are present?</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>List sample and features names with <code>rownames</code> and <code>colnames</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="5" type="1">
<li>Check what information about samples and features is contained by the <code>colData</code> and <code>rowData</code> of the TreeSE with <code>names</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="6" type="1">
<li>
<strong>Extra</strong>: Calculate the number of unique taxa for each taxonomic rank. You can use <code>apply</code> to count unique elements for each column of rowData.</li>
</ol></section><section id="assay-retrieval" class="level3" data-number="21.3.4"><h3 data-number="21.3.4" class="anchored" data-anchor-id="assay-retrieval">
<span class="header-section-number">21.3.4</span> Assay retrieval</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package<span class="op">=</span><span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>List the names of all available assays with <code>assayNames</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assayNames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Fetch the list of assays with <code>assays</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Retrieve the first assay of the TreeSE with <code>assay</code>, where the second argument can be either the name or the index of the desired assay.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Well done! You can now locate and retrieve individual assays of a TreeSE. If you got stuck, you can refer to chapter <a href="04_containers.html#sec-assay-slot" class="quarto-xref"><span>Section 2.2.1</span></a> of this book.</p>
</section><section id="sample-information" class="level3" data-number="21.3.5"><h3 data-number="21.3.5" class="anchored" data-anchor-id="sample-information">
<span class="header-section-number">21.3.5</span> Sample information</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Check the names of the samples with <code>colnames</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>List the information on samples available in colData with <code>names</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Visualize the colData with <code>View</code> and briefly look at the information stored in the different columns.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="5" type="1">
<li>Get the abundances of all features for a specific sample, such as <code>ID34</code>, for an <em>assay</em> of your choice.</li>
</ol></section><section id="feature-information" class="level3" data-number="21.3.6"><h3 data-number="21.3.6" class="anchored" data-anchor-id="feature-information">
<span class="header-section-number">21.3.6</span> Feature information</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Check the names of the features with <code>rownames</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>List the information on features available in rowData with <code>names</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Visualize the rowData with <code>View</code> and briefly look at the information stored in the different columns.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="5" type="1">
<li>Get the abundances for a specific feature, such as <code>OTU1810</code>, in all the samples. You can access feature-specific abundances for an <em>assay</em> of your choice.</li>
</ol>
<p>If you got stuck, you can look up chapters <a href="10_manipulation.html" class="quarto-xref"><span>Chapter 3</span></a> and <a href="11_taxonomic_information.html#sec-fly-tree" class="quarto-xref"><span>Section 5.2.1</span></a> on how to pick specific abundances and generate row trees, respectively.</p>
</section><section id="other-elements" class="level3" data-number="21.3.7"><h3 data-number="21.3.7" class="anchored" data-anchor-id="other-elements">
<span class="header-section-number">21.3.7</span> Other elements</h3>
<p>Extract some of the other TreeSE elements listed in chapter <a href="04_containers.html" class="quarto-xref"><span>Chapter 2</span></a>. However, such data are not always included.</p>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Fetch the metadata of the TreeSE. Is there any information available?</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Access the phylogenetic tree with <code>rowTree</code>. How big is it in terms of tips and nodes. If you like you can visualize it with <code>ggtree</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">rowTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Check if a sample tree is available with <code>colTree</code>, which is suitable for hierarchical or nested study designs.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/TreeSummarizedExperiment/man/TreeSummarizedExperiment-accessor.html">colTree</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="5" type="1">
<li>If present, obtain the information on feature DNA sequences from the DNA sequence slot.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">Barcode_full_length</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section></section><section id="data-manipulation" class="level2" data-number="21.4"><h2 data-number="21.4" class="anchored" data-anchor-id="data-manipulation">
<span class="header-section-number">21.4</span> Data manipulation</h2>
<section id="subsetting" class="level3" data-number="21.4.1"><h3 data-number="21.4.1" class="anchored" data-anchor-id="subsetting">
<span class="header-section-number">21.4.1</span> Subsetting</h3>
<ol type="1">
<li>Subset the TreeSE object to specific samples.<br>
</li>
<li>Subset the TreeSE object to specific features.<br>
</li>
<li>Subset the TreeSE object to specific samples and features.</li>
</ol></section><section id="library-sizes" class="level3" data-number="21.4.2"><h3 data-number="21.4.2" class="anchored" data-anchor-id="library-sizes">
<span class="header-section-number">21.4.2</span> Library sizes</h3>
<ol type="1">
<li>Calculate library sizes<br>
</li>
<li>Subsample / rarify the counts (see: rarefyAssay)</li>
</ol>
<p>Useful functions: nrow, ncol, dim, summary, table, quantile, unique, scater::addPerCellQC, mia::mergeFeaturesByRank</p>
</section><section id="prevalent-and-core-taxonomic-features" class="level3" data-number="21.4.3"><h3 data-number="21.4.3" class="anchored" data-anchor-id="prevalent-and-core-taxonomic-features">
<span class="header-section-number">21.4.3</span> Prevalent and core taxonomic features</h3>
<ol type="1">
<li>Estimate prevalence for your chosen feature (row, taxonomic group)<br>
</li>
<li>Identify all prevalent features and subset the data accordingly<br>
</li>
<li>Report the thresholds and the dimensionality of the data before and after subsetting<br>
</li>
<li>Visualize prevalence</li>
</ol>
<p>Useful functions: getPrevalence, getPrevalent, subsetByPrevalent</p>
</section><section id="data-exploration" class="level3" data-number="21.4.4"><h3 data-number="21.4.4" class="anchored" data-anchor-id="data-exploration">
<span class="header-section-number">21.4.4</span> Data exploration</h3>
<ol type="1">
<li>Summarize sample metadata variables. (How many age groups, how they are distributed? 0%, 25%, 50%, 75%, and 100% quantiles of library size?)<br>
</li>
<li>Create two histograms. Another shows the distribution of absolute counts, another shows how CLR transformed values are distributed.<br>
</li>
<li>Visualize how relative abundances are distributed between taxa in samples.</li>
</ol>
<p>Useful functions: nrow, ncol, dim, summary, table, quantile, unique, transformAssay, ggplot, wilcox.test, mergeFeaturesByRank, miaViz::plotAbundance</p>
</section><section id="other-functions" class="level3" data-number="21.4.5"><h3 data-number="21.4.5" class="anchored" data-anchor-id="other-functions">
<span class="header-section-number">21.4.5</span> Other functions</h3>
<ol type="1">
<li>Merge data objects (merge, mergeSEs)<br>
</li>
<li>Melt the data for visualization purposes (meltSE)</li>
</ol></section><section id="assay-transformation" class="level3" data-number="21.4.6"><h3 data-number="21.4.6" class="anchored" data-anchor-id="assay-transformation">
<span class="header-section-number">21.4.6</span> Assay transformation</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.<br>
</li>
<li>Transform the counts assay into relative abundances with <code>transformAssay</code> and store it into the TreeSE as an assay named <code>relabund</code> (see chapter <a href="11_taxonomic_information.html#sec-assay-transform" class="quarto-xref"><span>Section 5.4</span></a>).<br>
</li>
<li>Similarly, perform a clr transformation on the counts assay with a <code>pseudocount</code> of 1 and add it to the TreeSE as a new assay.<br>
</li>
<li>List the available assays by name with <code>assays</code>.<br>
</li>
<li>Access the clr assay and select a subset of its first 100 features and 10 samples. Remember that assays are subsettable with <code>assay[row_idx, col_idx]</code>.<br>
</li>
<li>Take the same subset from the TreeSE, and check how this affects the individual transformed assays. TreeSE can also be subsetted with <code>tse[row_idx, col_idx]</code>.<br>
</li>
<li>
<strong>Extra</strong>: If the data has phylogenetic tree, perform the phILR transformation.</li>
</ol></section></section><section id="abundance-tables" class="level2" data-number="21.5"><h2 data-number="21.5" class="anchored" data-anchor-id="abundance-tables">
<span class="header-section-number">21.5</span> Abundance tables</h2>
<section id="taxonomic-levels" class="level3" data-number="21.5.1"><h3 data-number="21.5.1" class="anchored" data-anchor-id="taxonomic-levels">
<span class="header-section-number">21.5.1</span> Taxonomic levels</h3>
<ol type="1">
<li>Import the mia package, load one of the example data sets mentioned in Chapter 3.3 with <code>data</code> (you need one with taxonomic information at Phylum level) and store it into a variable named <code>tse</code>.<br>
</li>
<li>List the available taxonomic ranks in the data with <code>taxonomyRanks</code>.<br>
</li>
<li>Agglomerate the data to Phylum level with <code>mergeFeaturesByRank</code> and the appropriate value for <code>Rank</code>.<br>
</li>
<li>Report the dimensions of the TreeSE before and after agglomerating. You can use <code>dim</code> for that.<br>
</li>
<li>
<strong>Extra</strong>: Perform CLR transformation on the data. Does this affect agglomeration?<br>
</li>
<li>
<strong>Extra</strong>: List full taxonomic information for a few selected taxa, such as <code>OTU1</code> and <code>OTU1368</code>. For that you can use <code>mapTaxonomy</code> on a specific subset of the TreeSE.</li>
</ol></section><section id="alternative-experiments" class="level3" data-number="21.5.2"><h3 data-number="21.5.2" class="anchored" data-anchor-id="alternative-experiments">
<span class="header-section-number">21.5.2</span> Alternative experiments</h3>
<ol type="1">
<li>Import the mia package, load one of the example data sets mentioned in Chapter 3.3 with <code>data</code> (you need one with taxonomic information) and store it into a variable named <code>tse</code>. Can you find the same taxonomyRanks from rowData?</li>
<li>Check the taxonomic ranks of the features with <code>taxonomyRanks</code>. What is the deepest taxonomic rank available?<br>
</li>
<li>Agglomerate the TreeSE to each taxonomic rank and store the resulting experiments as altExps. This can be performed automatically with <code>agglomerateByRanks</code>.<br>
</li>
<li>Check the names of the generated altExps with <code>altExpNames</code> and retrieve a complete list with <code>altExps</code>.<br>
</li>
<li>Retrieve the data agglomerated by genus from the corresponding <code>altExp</code>. As for assays, you can access the desired altExp by name or index.<br>
</li>
<li>
<strong>Extra</strong>: Split the data based on other features with <code>splitOn</code>.</li>
</ol></section></section><section id="community-alpha-diversity" class="level2" data-number="21.6"><h2 data-number="21.6" class="anchored" data-anchor-id="community-alpha-diversity">
<span class="header-section-number">21.6</span> Community (alpha) diversity</h2>
<section id="estimation" class="level3" data-number="21.6.1"><h3 data-number="21.6.1" class="anchored" data-anchor-id="estimation">
<span class="header-section-number">21.6.1</span> Estimation</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate multiple alpha diversity indices with <code>estimateDiversity</code> without any additional arguments.</li>
<li>Check the names of colData with <code>names</code>. Can you identify which columns contain the alpha diversity indices?</li>
<li>
<strong>Extra</strong>: Agglomerate the TreeSE by phylum and compare the mean Shannon diversity of the original experiment with its agglomerated version. You can use <code>mergeFeaturesByRank</code> to perform agglomeration and <code>mean</code> to calculate the mean values of the respective columns in colData.</li>
</ol></section><section id="visualization" class="level3" data-number="21.6.2"><h3 data-number="21.6.2" class="anchored" data-anchor-id="visualization">
<span class="header-section-number">21.6.2</span> Visualization</h3>
<ol type="1">
<li>Import the mia and scater packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate Shannon diversity index and Faith’s phylogenetic diversity with <code>estimateDiversity</code> and the appropriate arguments for <code>index</code>.</li>
<li>Make a boxplot of Shannon diversity on the y axis and sample type on the x axis using <code><a href="https://rdrr.io/pkg/scater/man/plotColData.html">scater::plotColData</a></code>.</li>
<li>Repeat the previous point with Faith’s phylogenetic diversity and compare the sample distributions of the two alpha diversity indices. How greatly do they differ?</li>
<li>
<strong>Extra</strong>: Make a scatterplot of Shannon diversity on the y axis and Faith’s phylogenetic diversity on the x axis with <code>plotColData</code>. Colour the points by sample type with the appropriate optional argument.</li>
</ol></section><section id="correlation" class="level3" data-number="21.6.3"><h3 data-number="21.6.3" class="anchored" data-anchor-id="correlation">
<span class="header-section-number">21.6.3</span> Correlation</h3>
<ol type="1">
<li>Import the mia and scater packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate coverage and Shannon diversity index with <code>estimateDiversity</code> and the appropriate arguments for <code>index</code>.</li>
<li>Test the correlation between the two indices with <code>cor.test</code>. Remember that colData parameters are accessible with <code>tse$param_name</code>. Use Kendall tau coefficients as <code>method</code> to measure correlation. Is the correlation weak or strong, significant or not?</li>
<li>Make a scatterplot of Shannon diversity index on the y axis and coverage on the x axis. You can do that with <code>plotColData</code>. How do the two indices relate to one another?</li>
<li>
<strong>Extra</strong>: Compute the library size of the samples by applying <code>colSums</code> to the counts assay of the TreeSE, and test the correlation of library size with Shannon diversity or coverage. Which index is more correlated with library size?</li>
</ol>
<p>In this example, we inspected the correlation between two related variables, also known as multicollinearity, and checked the correlation to library size, which is part of quality control. However, the correlation between alpha diversity and other numerical data about samples, such as participant’s age and weight, also represent an important analysis in several studies.</p>
</section><section id="differences-between-groups" class="level3" data-number="21.6.4"><h3 data-number="21.6.4" class="anchored" data-anchor-id="differences-between-groups">
<span class="header-section-number">21.6.4</span> Differences between groups</h3>
<ol type="1">
<li>Import the mia package, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Calculate the Gini-Simpson diversity with <code>estimateDiversity</code> and the appropriate argument for <code>index</code>. Set <code>name</code> to <code>simpson</code>. You will use this name to access the diversity index from colData.</li>
<li>Inspect the Diet column in the colData. Determine how the samples are grouped in terms of diet. You can see the number of unique elements in a column with <code>unique</code>.</li>
<li>Test differences in Gini-Simpson diversity between different diets with <code>kruskal.test</code>. Remember that colData parameters are accessible with <code>tse$param_name</code>.</li>
<li>Is diversity significantly different between vegan and mixed diet? To visualize that, make a boxplot of Gini-Simpson diversity on the y axis and diet on the x axis with <code>plotColData</code>.</li>
<li>
<strong>Extra</strong>: Repeat points 3 through 5, this time for age groups. Make sure that you are using an appropriate statistical test for the number of groups and distribution.</li>
</ol></section></section><section id="community-similarity" class="level2" data-number="21.7"><h2 data-number="21.7" class="anchored" data-anchor-id="community-similarity">
<span class="header-section-number">21.7</span> Community similarity</h2>
<section id="reduced-dimensions-retrieval" class="level3" data-number="21.7.1"><h3 data-number="21.7.1" class="anchored" data-anchor-id="reduced-dimensions-retrieval">
<span class="header-section-number">21.7.1</span> Reduced dimensions retrieval</h3>
<ol type="1">
<li>Import the mia package, load enterotype with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>List all available reduced dimensions with <code>reducedDims</code>. At this point, no reducedDims are likely found, because we haven’t created any yet.</li>
<li>Perform PCA using the scater library and store its output in the TreeSE by running <code>tse &lt;- runPCA(tse, assay.type = "counts")</code>. Note that it is required to specify the assay on which dimensionality reduction should be conducted.</li>
<li>View the names of all available reduced dimensions with <code>reducedDimNames</code>. Has something new appeared?</li>
<li>
<strong>Extra</strong>: Access the <code>scater::PCA</code> reducedDim object with <code>reducedDim</code> and explore its content. How are the different dimensions stored? Extract an array with only the values from the second dimension by indexing the object with <code>[ , 2]</code>.</li>
</ol></section><section id="visualization-basics-with-pca" class="level3" data-number="21.7.2"><h3 data-number="21.7.2" class="anchored" data-anchor-id="visualization-basics-with-pca">
<span class="header-section-number">21.7.2</span> Visualization basics with PCA</h3>
<ol type="1">
<li>Import the mia and scater packages, load enterotype with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Perform a 3-component PCA based on the <code>counts</code> assay. You can use <code>runPCA</code> and set the optional arguments <code>ncomponents</code> and <code>assay.type</code> to the appropriate values.</li>
<li>Plot the first two dimensions of PCA with <code><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">scater::plotReducedDim</a></code>, to which you should give the appropriate reducedDim name as the second argument. Note that by default only the first two dimensions are shown.</li>
<li>Check which information is stored in the ColData of the TreeSE. What would be worth visualizing in our coordination plot?</li>
<li>Make the same plot again, but this time colour the observations by Enterotype. You can do that by setting <code>colour_by</code> to the appropriate colname in the colData of the TreeSE.</li>
<li>
<strong>Extra</strong>: Plot all three dimensions of PCA with <code><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">scater::plotReducedDim</a></code> and the optional argument <code>ncomponents</code>. Colour observations by Enterotype. Which pair of dimensions best explains the variance between Enterotypes?</li>
</ol></section><section id="principal-coordinate-analysis-pcoa" class="level3" data-number="21.7.3"><h3 data-number="21.7.3" class="anchored" data-anchor-id="principal-coordinate-analysis-pcoa">
<span class="header-section-number">21.7.3</span> Principal Coordinate Analysis (PCoA)</h3>
<p>PCoA turns out to be particularly relevant for microbiome analysis, because unlike PCA it can generate reduced dimensions from distances other than Euclidean. There are several ecological distances to choose from and you can find many of them under methods in the vignettes of <code><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegan::vegdist</a></code>.</p>
<ol type="1">
<li>Import the mia and scater packages, load enterotype with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Transform the counts assay to relative abundances with <code>transformAssay</code>.</li>
<li>Perform a Multi-Dimensional Scaling (MDS) based on the relative abundance assay in terms of Bray-Curtis dissimilarity. You can use <code><a href="https://rdrr.io/pkg/scater/man/runMDS.html">scater::runMDS</a></code> with the compulsory argument <code>FUN = vegan::vegdist</code>.</li>
<li>Plot the first two dimensions of PCA with <code>plotReducedDim</code>, to which you should give the appropriate reducedDim name as the second argument. Colour the observations by Enterotype with <code>colour_by</code>.</li>
<li>
<strong>Extra</strong>: Perform MDS again with <code><a href="https://rdrr.io/pkg/scater/man/runMDS.html">scater::runMDS</a></code>, but this time use Jaccard dissimilarity. The distance metric to use can be defined with the optional argument <code>method</code>, choosing from the methods in <code><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">?vegan::vegdist</a></code>. If you don’t want to overwrite the reducedDim object made in point 3, set <code>name</code> to a name of your choice. Visualize and compare it to the plot from point 4.</li>
</ol>
<p>Good job! You are now able to produce and visualize reduced dimensions of a TreeSE. <code>runMDS</code> is actually one of several algorithms for PCoA and dimensionality reduction, which you can find in section <a href="20_beta_diversity.html#sec-other-ord-methods" class="quarto-xref"><span>Section 7.2.2</span></a>.</p>
</section><section id="permanova-analysis" class="level3" data-number="21.7.4"><h3 data-number="21.7.4" class="anchored" data-anchor-id="permanova-analysis">
<span class="header-section-number">21.7.4</span> PERMANOVA analysis</h3>
<p>In this exercise we focus on studying the weight of variables on the microbiome composition. Significance of each variable on beta diversity is tested with PERMANOVA (point 4) and the homogeneity assumption is also be controlled with a PERMDISP2 analysis (point 5).</p>
<ol type="1">
<li>Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Transform the <code>counts</code> assay into relative abundances with <code>transformAssay</code>.</li>
<li>Extract the relative abundance assay, transpose it and save it into a variable named <code>relabund_assay</code>.</li>
<li>Perform PERMANOVA with <code><a href="https://rdrr.io/pkg/vegan/man/adonis.html">vegan::adonis2</a></code> to see how much Diet can explain the relative abundance assay (<code>formula = relabund_assay ~ Diet</code>) in terms of Bray-Curtis dissimilarity (<code>method = "bray"</code>). Also set <code>data = colData(tse)</code> <code>by = "margin"</code> and <code>permutations = 99</code>. What do the results tell you about Diet with respect to beta diversity?</li>
<li>
<strong>Extra</strong>: Test homogeneity of distribution across Diet groups with <code>anova(betadisper(my_mat), my_groups</code>, where <code>my_mat</code> is the Bray-Curtis dissimilarity matrix of <code>relabund_assay</code> calculated with <code>vegdist(relabund_assay, "bray")</code> and <code>my_groups</code> is the vector of Diet values obtained from the colData of the TreeSE.</li>
</ol>
<p>Well done! You went through testing the effect and significance of Diet on beta diversity. Keep in mind that the formula fed to <code>adonis2</code> can take more than one independent variable, so that you can also (and very often should) include covariates of your studies.</p>
</section><section id="redundancy-analysis-rda" class="level3" data-number="21.7.5"><h3 data-number="21.7.5" class="anchored" data-anchor-id="redundancy-analysis-rda">
<span class="header-section-number">21.7.5</span> Redundancy analysis (RDA)</h3>
<p>Here we apply RDA, an ordination method that provides dimensions with the largest variation between the data based in the light of the specified variables (point 3). The results of RDA are usually assessed with PERMANOVA (point 5) and the homogeneity assumption should be checked as in the previous exercise. This is a relatively complex procedure, but the way this is broken down into steps below will hopefully make more sense.</p>
<ol type="1">
<li>Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Transform the <code>counts</code> assay into relative abundances with <code>transformAssay</code>.</li>
<li>Perform RDA with <code>calculateRDA</code> to see how much Diet can explain the relative abundance assay (<code>formula = assay ~ Diet</code> and <code>assay.type = relabundance</code>) in terms of Bray-Curtis dissimilarity (<code>method = "bray"</code>).</li>
<li>Extract the RDA dimensions from the appropriate reducedDim slot with <code>attr(reducedDim(tse, "RDA"), "rda)</code> and store it into <code>rda</code>.</li>
<li>Test the effect and significance of Diet on beta diversity by PERMANOVA with <code><a href="https://rdrr.io/pkg/vegan/man/anova.cca.html">vegan::anova.cca</a></code>. Feed this function with <code>rda</code> and set <code>by = "margin"</code> and <code>permutations = 99</code>, respectively. What do the results tell you about Diet?</li>
<li>
<strong>Extra</strong>: Check what other parameters are stored in the colData of peerj13075, add them to the formula (<code>formula = assay ~ Diet + ...</code>) of <code>calculateRDA</code> and proceed to see how that changes the results of PERMANOVA.</li>
</ol>
<p>Well done! You went through an RDA analysis followed by significance testing with PERMANOVA and BETADISPER2. In the next exercise we’ll go deeper quantify the contributions to beta diversity.</p>
</section><section id="beta-diversity-analysis" class="level3" data-number="21.7.6"><h3 data-number="21.7.6" class="anchored" data-anchor-id="beta-diversity-analysis">
<span class="header-section-number">21.7.6</span> Beta diversity analysis</h3>
<p>This exercise prompts you to implement a workflow with distance-based RDA (dbRDA). You can refer to chapter <a href="20_beta_diversity.html#sec-dbrda-workflow" class="quarto-xref"><span>Section 7.4.1</span></a> for a step-by-step walkthrough, which may be simplified in the future.</p>
<ol type="1">
<li>Import the mia and vegan packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
<li>Create dbRDA with Bray-Curtis dissimilarities on relative abundances. Use PERMANOVA. Can differences between samples be explained with variables of sample meta data?</li>
<li>Analyze diets’ association on beta diversity. Calculate dbRDA and then PERMANOVA. Visualize coefficients. Which taxa’s abundances differ the most between samples?</li>
<li>Interpret your results. Is there association between community composition and location? What are those taxa that differ the most; find information from literature.</li>
</ol>
<p>Useful functions: scater::runMDS, runRDA, vegan::anova.cca, transformAssay, mergeFeaturesByRank, ggplot, scater::plotReducedDim, vegan::adonis2</p>
</section></section><section id="differential-abundance" class="level2" data-number="21.8"><h2 data-number="21.8" class="anchored" data-anchor-id="differential-abundance">
<span class="header-section-number">21.8</span> Differential abundance</h2>
<section id="standard-analysis-with-aldex2" class="level3" data-number="21.8.1"><h3 data-number="21.8.1" class="anchored" data-anchor-id="standard-analysis-with-aldex2">
<span class="header-section-number">21.8.1</span> Standard analysis with ALDEx2</h3>
<ol type="1">
<li>Import the mia and ALDEx2 packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ggloor/ALDEx_bioc">ALDEx2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"peerj13075"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">peerj13075</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Agglomerate the TreeSE by genus and filter by a prevalence of 10%. You can perform both operations in one go with <code>subsetByPrevalent</code> by specifying the <code>rank</code> and <code>prevalence</code> arguments.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalent</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                                 rank <span class="op">=</span> <span class="st">"genus"</span>,</span>
<span>                                 prevalence <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Model the counts assay of the TreeSE with <code>aldex.clr</code> and store it into the variable <code>x</code>. As a second argument, provide the grouping variable <code>Diet</code>, which is contained in a column of the colData.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.clr.function.html">aldex.clr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span>, <span class="va">tse</span><span class="op">$</span><span class="va">Gender</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Feed <code>x</code> to the functions <code>aldex.ttest</code> to perform t-test and to <code>aldex.effect</code> to estimate effect sizes. Store the output into <code>x_tt</code> and <code>x_effect</code>, respectively.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x_tt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.ttest.html">aldex.ttest</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">x_effect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ALDEx2/man/aldex.effect.html">aldex.effect</a></span><span class="op">(</span><span class="va">x</span>, CI <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="5" type="1">
<li>Create a data.frame named <code>aldex_out</code> which includes both <code>x_tt</code> and <code>x_effect</code> and filter for the features with <code>wi.eBH &lt; 0.05</code>. Are there any significantly differential abundance taxa?</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">aldex_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">x_tt</span>, <span class="va">x_effect</span><span class="op">)</span></span>
<span></span>
<span><span class="va">aldex_out</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">wi.eBH</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span>  <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Genus</span>, <span class="va">we.eBH</span>, <span class="va">wi.eBH</span>, <span class="va">effect</span>, <span class="va">overlap</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="6" type="1">
<li>
<strong>Extra</strong>: If these results appear boring, repeat steps 1 - 5, but use <code>Gender</code> or <code>Age</code> as the grouping variable. Do we have any better luck with Gender? What is the problem with Age?</li>
</ol></section><section id="sec-control-daa-confounders" class="level3" data-number="21.8.2"><h3 data-number="21.8.2" class="anchored" data-anchor-id="sec-control-daa-confounders">
<span class="header-section-number">21.8.2</span> Controlling for confounders</h3>
<ol type="1">
<li>Import the mia and MicrobiomeStat packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MicrobiomeStat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"peerj13075"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">peerj13075</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Agglomerate the TreeSE by genus and filter by a prevalence of 10%. You can perform both operations in one go with <code>subsetByPrevalent</code> by specifying the <code>rank</code> and <code>prevalence</code> arguments.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalent</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                                 rank <span class="op">=</span> <span class="st">"genus"</span>,</span>
<span>                                 prevalence <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Model the counts assay of the TreeSE with <code>linda</code> and store the output into a variable named <code>linda_out</code>. Provide the colData converted into a data.frame (with <code>as.data.frame</code>) as the second argument, and a <code>formula</code> with the Age, Gender and Diet as variables. For example, <code>formula = "~ A + B"</code> represents a formula with variables A and B.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">linda_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MicrobiomeStat/man/linda.html">linda</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   formula <span class="op">=</span> <span class="st">"~ Diet + Gender + Age"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>Extract the <code>output$AgeElderly</code> object from <code>linda_out</code> with <code>$</code> and store it into a variable named <code>linda_res</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">linda_res</span> <span class="op">&lt;-</span> <span class="va">linda_out</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">AgeElderly</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="5" type="1">
<li>Filter <code>linda_res</code> for features with <code>reject == TRUE</code>. How many differentially abundant taxa were found? What are their names and how significant are they in terms of log-fold change and adjusted p-value?</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">linda_res</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">reject</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">log2FoldChange</span>, <span class="va">stat</span>, <span class="va">padj</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"feature"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">linda_res</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="sec-compare-daa-methods" class="level3" data-number="21.8.3"><h3 data-number="21.8.3" class="anchored" data-anchor-id="sec-compare-daa-methods">
<span class="header-section-number">21.8.3</span> Comparing methods</h3>
<p>Here, we conduct DAA with identical parameters as in the <a href="#control-daa-confounders">previous exercise</a>, but with a different method, namely ZicoSeq. We aim to compare the results between these two methods and draw better informed conclusions from such comparative approach.</p>
<ol type="1">
<li>Import the mia and GUniFrac packages, load any of the example data sets mentioned in Chapter 3.3 with <code>data</code> and store it into a variable named <code>tse</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/microbiome/mia">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GUniFrac</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"peerj13075"</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">peerj13075</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Agglomerate the TreeSE by genus and filter by a prevalence of 10%. You can perform both operations in one go with <code>subsetByPrevalent</code> by specifying the <code>rank</code> and <code>prevalence</code> arguments.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mia/man/getPrevalence.html">subsetByPrevalent</a></span><span class="op">(</span><span class="va">tse</span>,</span>
<span>                            rank <span class="op">=</span> <span class="st">"genus"</span>,</span>
<span>                            prevalence <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="3" type="1">
<li>Model the counts assay of the TreeSE with <code>ZicoSeq</code> as the <code>feature.dat</code> argument and store the output into a variable named <code>zicoseq_out</code>. Provide also the colData converted to a data.frame (with <code>as.data.frame</code>) as <code>meta.dat</code>. In addition, set <code>grp.name</code> to <code>"Age"</code>, <code>adj.name</code> to <code>c("Diet", "Gender")</code>, <code>feature.dat.type</code> to <code>"count"</code>, <code>return.feature.dat</code> to <code>TRUE</code> and <code>perm.no</code> to <code>999</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zicoseq_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GUniFrac/man/ZicoSeq.html">ZicoSeq</a></span><span class="op">(</span>feature.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       meta.dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       grp.name <span class="op">=</span> <span class="st">"Age"</span>,</span>
<span>                       adj.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Diet"</span>, <span class="st">"Gender"</span><span class="op">)</span>, </span>
<span>                       feature.dat.type <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>                       return.feature.dat <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       perm.no <span class="op">=</span> <span class="fl">999</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="4" type="1">
<li>View the top six differentially abundant taxa and their adjusted p-values with <code>head(sort(zicoseq_out$p.adj.fdr))</code>. Is there any significant taxon according to ZicoSeq? Compared to the output of linda, do we see the same taxa at the top in terms of significance? Overall, to what extent do the two methods agree with one another?</li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show solution</summary><div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">zicoseq_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind.data.frame</a></span><span class="op">(</span>p.raw <span class="op">=</span> <span class="va">zicoseq_out</span><span class="op">$</span><span class="va">p.raw</span>,</span>
<span>                                p.adj.fdr <span class="op">=</span> <span class="va">zicoseq_out</span><span class="op">$</span><span class="va">p.adj.fdr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">zicoseq_res</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">p.raw</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="workflow-1" class="level3" data-number="21.8.4"><h3 data-number="21.8.4" class="anchored" data-anchor-id="workflow-1">
<span class="header-section-number">21.8.4</span> Workflow 1</h3>
<ol type="1">
<li>Get the abundances for an individual feature (taxonomic group / row)</li>
<li>Visualize the abundances per group with boxplot / jitterplot</li>
<li>Is the difference significant (Wilcoxon test)?</li>
<li>Is the difference significant (linear model with covariates)?</li>
<li>How do transformations affect the outcome (log10, clr..)?</li>
<li>Get p-values for all features (taxa), for instance with a for loop</li>
<li>Do multiple testing correction</li>
<li>Compare the results from different tests with a scatterplot</li>
</ol>
<p>Useful functions: [], ggplot2::geom_boxplot, ggplot2::geom_jitter, wilcox.test, lm.test, transformAssay, p.adjust</p>
</section><section id="workflow-2" class="level3" data-number="21.8.5"><h3 data-number="21.8.5" class="anchored" data-anchor-id="workflow-2">
<span class="header-section-number">21.8.5</span> Workflow 2</h3>
<ol type="1">
<li>install the latest development version of mia from GitHub.</li>
<li>Load experimental dataset from mia.</li>
<li>Compare abundances of each taxa between groups. First, use Wilcoxon or Kruskall-Wallis test. Then use some other method dedicated to microbiome data.</li>
<li>Summarize findings by plotting a taxa vs samples heatmap. Add column annotation that tells the group of each sample, and row annotation that tells whether the difference of certain taxa was statistically significant.</li>
<li>Choose statistically significant taxa and visualize their abundances with boxplot &amp; jitterplot.</li>
</ol>
<p>Useful functions: wilcox.test, kruskal.test, ggplot2::ggplot, pheatmap::pheatmap, ComplexHeatMap::Heatmap, ANCOMBC::ancombc2, ALDEx2::aldex, Maaslin2::Maaslin2, agglomerateByRank, transformAssay, subsetByPrevalent</p>
</section></section><section id="visualization-1" class="level2" data-number="21.9"><h2 data-number="21.9" class="anchored" data-anchor-id="visualization-1">
<span class="header-section-number">21.9</span> Visualization</h2>
<section id="multivariate-ordination" class="level3" data-number="21.9.1"><h3 data-number="21.9.1" class="anchored" data-anchor-id="multivariate-ordination">
<span class="header-section-number">21.9.1</span> Multivariate ordination</h3>
<ol type="1">
<li>Load experimental dataset from mia.</li>
<li>Create PCoA with Bray-Curtis dissimilarities</li>
<li>Create PCA with Aitchison dissimilarities</li>
<li>Visualize and compare both</li>
<li>Test other transformations, dissimilarities, and ordination methods</li>
</ol>
<p>Useful functions: scater::runMDS, runNMDS, transformAssay, ggplot, scater::plotReducedDim</p>
</section><section id="heatmap-visualization" class="level3" data-number="21.9.2"><h3 data-number="21.9.2" class="anchored" data-anchor-id="heatmap-visualization">
<span class="header-section-number">21.9.2</span> Heatmap visualization</h3>
<ol type="1">
<li>Load experimental dataset from mia.</li>
<li>Visualize abundances with heatmap.</li>
<li>Visualize abundances with heatmap after CLR + Z transformation.</li>
</ol>
<p>See the OMA book for examples.</p>
</section></section><section id="multiomics" class="level2" data-number="21.10"><h2 data-number="21.10" class="anchored" data-anchor-id="multiomics">
<span class="header-section-number">21.10</span> Multiomics</h2>
<section id="basic-exploration" class="level3" data-number="21.10.1"><h3 data-number="21.10.1" class="anchored" data-anchor-id="basic-exploration">
<span class="header-section-number">21.10.1</span> Basic exploration</h3>
<p>Here we learn how to conduct preliminary exploration on a MAE, using HintikkaXOData as an example dataset.</p>
<ol type="1">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a variable named <code>mae</code>.</li>
<li>Which experiments make up the MAE? How many samples and features are contained by each experiment? You can get a summary for all experiments with <code>experiments</code>, and check for each individual experiment with <code>dim</code>, <code>nrow</code> and <code>ncol</code>.</li>
<li>What are the names of the features and samples of the different experiments? You can see that with <code>rownames</code> and <code>colnames</code>, respectively.</li>
<li>What information is known about the samples? Remember that information about samples is stored in the <code>colData</code> of the MAE.</li>
<li>
<strong>Extra</strong>: How do the samples of the individual experiments map to the columns of the MAE? You can find the sample mapping in the <code>sampleMap</code> of the MAE.</li>
</ol>
<p>So far so good. You explored a MAE and its experiments, getting a taste of how information is organized in its internal structure.</p>
</section><section id="experiment-agglomeration" class="level3" data-number="21.10.2"><h3 data-number="21.10.2" class="anchored" data-anchor-id="experiment-agglomeration">
<span class="header-section-number">21.10.2</span> Experiment agglomeration</h3>
<p>Here we learn how to manipulate an experiment contained by a MAE and save the new modified version of the experiment in a suitable place (the altExp slot).</p>
<ol type="1">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a variable named <code>mae</code>.</li>
<li>Agglomerate the microbiota experiment by Genus and store the output into the <code>altExp</code> slot of the microbiota experiment, with the custom name <code>microbiota_genus</code>.</li>
<li>How many features remain after agglomerating? What are their names?</li>
<li>
<strong>Extra</strong>: create one more alternative experiment named <code>prevalent_microbiota_family</code>, which contains the microbiota experiment agglomerated by Family with a prevalence threshold of 10%. You can agglomerate and in parallel select by prevalence with <code>mergeFeaturesByPrevalence</code>.</li>
</ol>
<p>Good job! You agglomerated one of the experiments in the MAE and stored it as an alternative experiment.</p>
</section><section id="experiment-transformation" class="level3" data-number="21.10.3"><h3 data-number="21.10.3" class="anchored" data-anchor-id="experiment-transformation">
<span class="header-section-number">21.10.3</span> Experiment transformation</h3>
<p>We proceed with an exercise on a different type of data manipulation, that is, transformation of assays of individual experiments in the MAE.</p>
<ol type="1">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a variable named <code>mae</code>.</li>
<li>What assays are contained by each individual experiment? You can check their names with <code>assays</code>.</li>
<li>Apply a log10 transformation to the assay of the metabolite experiment. For that you can use <code>transformAssay</code> and don’t forget to specify the assay to be transformed with the argument <code>assay.type</code>.</li>
<li>Apply a CLR transformation to the counts assay of the microbiota experiment. To ensure non-null values in the assay, set <code>pseudocount</code> equal to 1.</li>
</ol>
<p>You made it! You learnt how to apply different transformations to the assays of individual experiments in a MAE with <code>transformAssay</code>, specifying optional arguments based on the used method.</p>
</section><section id="assay-extraction" class="level3" data-number="21.10.4"><h3 data-number="21.10.4" class="anchored" data-anchor-id="assay-extraction">
<span class="header-section-number">21.10.4</span> Assay extraction</h3>
<p>The following exercise walks you through disassembling a MAE object in order to retrieve a specific assay, or to store its components as multiple separate csv files.</p>
<ol type="1">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a variable named <code>mae</code>.</li>
<li>Extract the individual metabolite experiment from the MAE into a distinct TreeSE object named <code>metabolites</code>.</li>
<li>Which and how many assays are contained by <code>metabolites</code>? You can check that with <code>assays</code> or <code>assayNames</code>.</li>
<li>Write a csv file for the nmr assay with <code>write.csv</code>. You can access an individual assay of a TreeSE with <code>assay</code> by specifying the name of the desired assay.</li>
<li>
<strong>Extra</strong>: Repeat step 1 thorugh 4 also for the microbiota and biomarkers experiments, so that a completely disassembled version of the MAE is available.</li>
<li>
<strong>Extra</strong>: Besides experiments, MAEs also include a sampleData and a sampleMap, which are accessible with <code>colData(mae)</code> and <code>sampleMap(mae)</code>, respectively. Save also each of these two elements into a csv file.</li>
</ol>
<p>Well done! You just splitted a MAE into its components and stored them as csv files. <a href="https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl/blob/main/src/assets/XO_preprocess.R">This script</a> shows a possible approach.</p>
</section><section id="mae-reconstruction" class="level3" data-number="21.10.5"><h3 data-number="21.10.5" class="anchored" data-anchor-id="mae-reconstruction">
<span class="header-section-number">21.10.5</span> MAE reconstruction</h3>
<p>Next, we will try to reconstruct the same MAE from the files you created. Make sure you know their names and location! Alternatively, you can fetch or download the CSV files in <a href="https://github.com/microbiome/data/tree/main/HintikkaXOData">this directory</a> with the readily disassembled components of HintikkaXOData.</p>
<ol type="1">
<li>Read in the csv files containing assays with <code>read.csv</code> and save each of them into a variable named <code>&lt;assay name&gt;_assays</code>.</li>
<li>Create one TreeSE from each assays object with the <code>TreeSummarizedExperiment</code> function, as explained in <a href="#construct-TreeSE">this exercise</a>.</li>
<li>Read in the sampleData and the sampleMap and store them into the variables <code>sample_data</code> and <code>sample_map</code>, respectively.</li>
<li>Combine the components with <code>MultiAssayExperiment</code>, where the first argument is an <code>ExperimentList</code> (for now include only the microbiota and metabolites TreeSEs), the second is colData and the third is sampleMap.</li>
<li>Make sure that the MAE experiments are identical to the original TreeSEs. You can do that qualitatively by checking their <code>head</code> and quantitatively by looking at their <code>dim</code>.</li>
<li>
<strong>Extra</strong>: Add the biomarkers TreeSE as a new experiment to the MAE. Note that new experiments can be added to a MAE through simple concatenation with <code>c(mae, experiment)</code>.</li>
</ol>
<p>Good job! Now you are aware of how MAEs are built and we can proceed to some analytical exercises.</p>
</section><section id="cross-correlation-analysis" class="level3" data-number="21.10.6"><h3 data-number="21.10.6" class="anchored" data-anchor-id="cross-correlation-analysis">
<span class="header-section-number">21.10.6</span> Cross-correlation analysis</h3>
<p>Now we will perform a cross-correlation analysis between two of the experiments in the MAE.</p>
<ol type="1">
<li>Import the mia package, load HintikkaXOData with <code>data</code> and store it into a variable named <code>mae</code>.</li>
<li>Analyze correlations between the microbiota and the biomarkers experiments with <code>getExperimentCrossAssociation</code>. Don’t forget to specify the experiments you want to compare with the arguments <code>experiment1</code> and <code>experiment2</code>, and which of their assays with <code>assay.type1</code> and <code>assay.type2</code>.</li>
<li>What does the output look like? By default, correlation is measured in terms of Kendall tau coefficients. Repeat point 2, but this time change <code>method</code> to Spearman coefficients.</li>
<li>Are you able to infer significance from the output? In order to also obtain p-values from the cross-correlation analysis, repeat point 2 with the additional argument <code>test_significance = TRUE</code>.</li>
<li>Visualize results with a heatmap similarly to the example in section <a href="23_multi-assay_analyses.html#sec-cross-correlation" class="quarto-xref"><span>Section 12.1</span></a>. Do you see any significant correlations? Interpret your results.</li>
<li>
<strong>Extra</strong>: Perform cross-correlation analysis between the remaining experiments (microbiota vs metabolites and metabolites vs biomarkers) and visualize results with heatmaps.</li>
</ol>
<p>Great job! You performed a cross-correlation analysis between two experiments of a MAE and visualized the results with a heatmap. You are also able to customise the correlation method and significance testing used for the analysis.</p>


</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/95_resources.html" class="pagination-link" aria-label="Resources">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/97_extra_materials.html" class="pagination-link" aria-label="Extra material">
        <span class="nav-page-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>