<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 16&nbsp; Meta-analyses</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/MSEA.html" rel="next">
<link href="../pages/IntegratedLearner.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar docked nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../docs/pages/images/mia_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Orchestrating Microbiome Analysis</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">TEST</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-bi-github">
<li>
    <a class="dropdown-item" href="https://github.com/microbiome/OMA/">
 <span class="dropdown-text">Source Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://microbiome.github.io/OMA/docs/devel/">
 <span class="dropdown-text">Browse version <code>devel</code></span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/MMUPHin_meta_analysis.html"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/61_network_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Network comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#meta-analysis-of-cancer-microbiome-studies-using-relative-abundance-data" id="toc-meta-analysis-of-cancer-microbiome-studies-using-relative-abundance-data" class="nav-link active" data-scroll-target="#meta-analysis-of-cancer-microbiome-studies-using-relative-abundance-data"><span class="header-section-number">16.1</span> Meta-analysis of cancer microbiome studies using relative abundance data</a>
  <ul class="collapse">
<li><a href="#sec-batch-effect-adjustment" id="toc-sec-batch-effect-adjustment" class="nav-link" data-scroll-target="#sec-batch-effect-adjustment"><span class="header-section-number">16.1.1</span> Performing batch (study) effect adjustment with adjust_batch for relative abundance</a></li>
  <li><a href="#sec-batch-adjustment-evaluation" id="toc-sec-batch-adjustment-evaluation" class="nav-link" data-scroll-target="#sec-batch-adjustment-evaluation"><span class="header-section-number">16.1.2</span> Evaluation for batch effect adjustment</a></li>
  <li><a href="#sec-differential-abundance-testing" id="toc-sec-differential-abundance-testing" class="nav-link" data-scroll-target="#sec-differential-abundance-testing"><span class="header-section-number">16.1.3</span> Meta-analytical differential abundance testing with lm_meta</a></li>
  </ul>
</li>
  <li><a href="#sec-meta-analysis-pathway" id="toc-sec-meta-analysis-pathway" class="nav-link" data-scroll-target="#sec-meta-analysis-pathway"><span class="header-section-number">16.2</span> Meta-analysis for pathway abundance data</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/MMUPHin_meta_analysis.html"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-meta-analyses" class="quarto-section-identifier"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>Meta-analysis is a statistical analysis that combines the results of multiple scientific studies by appropriately weighting the individual study-specific effect sizes <span class="citation" data-cites="Lee2019">I. H (<a href="#ref-Israel2011" role="doc-biblioref">2011</a>)</span>. A common meta-analytic combination strategy is to take a weighted average of the study-specific summary measures. In fixed-effects meta-analysis, the weights are based on the assumption that there is a single true parameter underlying all the studies, while in random-effects meta-analysis, the weights are based on a model where the true parameter varies across studies according to a probability distribution.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Israel2011" class="csl-entry" role="listitem">
H, Israel. 2011. <span>“A Guide to Understanding Meta-Analysis.”</span> <em>Journal of Orthopaedic &amp; Sports Physical Therapy</em> 41 (7): 496–504. <a href="https://www.jospt.org/doi/10.2519/jospt.2011.3333?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed">https://www.jospt.org/doi/10.2519/jospt.2011.3333?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed</a>.
</div></div><p>Meta-analysis is particularly useful when individual studies are small, and a joint analysis of these studies can enhance the precision of effect size estimation and improve statistical power. Especially in microbiome studies, individual effect sizes from each study tend to be small <span class="citation" data-cites="Hoffman2019">S (<a href="#ref-Ma2022" role="doc-biblioref">2022</a>)</span>. Only through meta-analysis can these small effects reach statistical significance. Therefore, by jointly analyzing these studies, we amplify the collective voice of the data, offering a more comprehensive understanding of the effects under investigation.</p>
<p>In this chapter, our focus is on using the R/Bioconductor package [MMUPHin] (Meta-Analysis Methods with a Uniform Pipeline for Heterogeneity in microbiome studies) <span class="citation" data-cites="Ma2022">(<a href="#ref-Ma2022" role="doc-biblioref">S 2022</a>)</span> to perform the meta-analysis of microbiome data. We utilize publicly available cancer microbiome datasets from the R/Bioconductor package [curatedMetagenomicData] <span class="citation" data-cites="Pasolli2017">(<a href="#ref-Pasolli2017" role="doc-biblioref">E 2017</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Ma2022" class="csl-entry" role="listitem">
S, Ma. 2022. <span>“Population Structure Discovery in Meta-Analyzed Microbial Communities and Inflammatory Bowel Disease Using MMUPHin.”</span> <em>Genome Biology</em> 23 (1): 208. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02753-4">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02753-4</a>.
</div><div id="ref-Pasolli2017" class="csl-entry" role="listitem">
E, Pasolli. 2017. <span>“Accessible, Curated Metagenomic Data Through ExperimentHub.”</span> <em>Nature Methods</em> 14 (11): 1023–24. <a href="https://www.nature.com/articles/nmeth.4468">https://www.nature.com/articles/nmeth.4468</a>.
</div></div><section id="meta-analysis-of-cancer-microbiome-studies-using-relative-abundance-data" class="level2 page-columns page-full" data-number="16.1"><h2 data-number="16.1" class="anchored" data-anchor-id="meta-analysis-of-cancer-microbiome-studies-using-relative-abundance-data">
<span class="header-section-number">16.1</span> Meta-analysis of cancer microbiome studies using relative abundance data</h2>
<p>Recent work has indicated a possible involvement of the gut microbiome in influencing the efficacy of immune checkpoint inhibitor (ICI) treatment strategies for PD-1/PD-L1 (Programmed Cell Death Protein 1/Programmed Cell Death Ligand 1) targeting checkpoint inhibitors <span class="citation" data-cites="Frankel2017">BA (<a href="#ref-Peters2019" role="doc-biblioref">2019</a>)</span>. Several investigations into the microbiome have provided clinical data suggesting that the gut microbiome modulates the response to inhibitors of the PD-1/PD-L1 axis. However, published cancer microbiome studies face challenges, such as limited sample sizes, inconsistent data analysis methods, and a lack of functionally relevant consensus signatures.</p>
<div class="no-row-height column-margin column-container"></div><p>Elucidation of the mechanisms by which the gut microbiome alters the function of the immune system to enable or promote cancer development may reveal novel pathways to explore in cancer therapy. Consequently, a high-quality re-analysis of public cancer microbiome data through a systematic meta-analysis approach could provide valuable insights into the microbiome’s role in cancer development and progression in a rapid and cost-effective manner.</p>
<p>Using advanced melanoma as a model, extensively studied in the context of PD-1/PD-L1 immune checkpoint inhibitors, our aim is to identify functional biomarkers positively and negatively associated with ICI response by analyzing both taxonomic and functional profiles with ICI response.</p>
<p>In the following section, we provide detailed examples of how to perform batch (study) effect correction and meta-analytic differential abundance testing on publicly available cancer microbiome data from multiple studies. We use the R package MMUPHin for these tasks. MMUPHin is a recently developed R/Bioconductor package designed for meta-analysis of microbiome taxonomic and functional profiles. It is agnostic to the data type and leverages another R/Bioconductor package, MaAsLin2 <span class="citation" data-cites="Mallick2021">(<a href="#ref-Mallick2021" role="doc-biblioref">M. H 2021</a>)</span>, as a backend to conduct the meta-analysis. MaAsLin2 is particularly designed to be applicable to various microbial community data types (taxonomy or functional profiles) and environments (human or otherwise). It is modular, including implementations of alternative normalization/transformation schemes and statistical models (e.g., amplicon vs.&nbsp;shotgun metagenomic profiles). Leveraging this flexible framework under the hood, MMUPHin performs normalization, study (batch) effect correction, and meta-analysis using the default random effects meta-regression.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Mallick2021" class="csl-entry" role="listitem">
H, Mallick. 2021. <span>“Multivariable Association Discovery in Population-Scale Meta-Omics Studies.”</span> <em>PLOS Computational Biology</em> 17 (11): e1009442. <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442">https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442</a>.
</div><div id="ref-Frankel2017" class="csl-entry" role="listitem">
AE, Frankel. 2017. <span>“Metagenomic Shotgun Sequencing and Unbiased Metabolomic Profiling Identify Specific Human Gut Microbiota and Metabolites Associated with Immune Checkpoint Therapy Efficacy in Melanoma Patients.”</span> <em>Neoplasia</em> 19 (10): 848–55. <a href="https://www.clinicalkey.com/#!/content/playContent/1-s2.0-S1476558617302385?returnurl=https:%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1476558617302385%3Fshowall%3Dtrue&amp;referrer=https:%2F%2Fpubmed.ncbi.nlm.nih.gov%2F">https://www.clinicalkey.com/#!/content/playContent/1-s2.0-S1476558617302385?returnurl=https:%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1476558617302385%3Fshowall%3Dtrue&amp;referrer=https:%2F%2Fpubmed.ncbi.nlm.nih.gov%2F</a>.
</div><div id="ref-Lee2022" class="csl-entry" role="listitem">
KA, Lee. 2022. <span>“Cross-Cohort Gut Microbiome Associations with Immune Checkpoint Inhibitor Response in Advanced Melanoma.”</span> <em>Nature Medicine</em> 28 (3): 535–44. <a href="https://www.nature.com/articles/s41591-022-01695-5">https://www.nature.com/articles/s41591-022-01695-5</a>.
</div><div id="ref-Matson2018" class="csl-entry" role="listitem">
V, Matson. 2018. <span>“The Commensal Microbiome Is Associated with Anti-PD-1 Efficacy in Metastatic Melanoma Patients.”</span> <em>Science</em> 359 (6371): 104–8. <a href="https://www.science.org/doi/10.1126/science.aao3290?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed">https://www.science.org/doi/10.1126/science.aao3290?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed</a>.
</div><div id="ref-Peters2019" class="csl-entry" role="listitem">
BA, Peters. 2019. <span>“Relating the Gut Metagenome and Metatranscriptome to Immunotherapy Responses in Melanoma Patients.”</span> <em>Genome Medicine</em> 11 (1): 61. <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-019-0672-4">https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-019-0672-4</a>.
</div><div id="ref-Wind2020" class="csl-entry" role="listitem">
TT, Wind. 2020. <span>“Gut Microbial Species and Metabolic Pathways Associated with Response to Treatment with Immune Checkpoint Inhibitors in Metastatic Melanoma.”</span> <em>Melanoma Research</em> 30 (3): 235–46. <a href="https://journals.lww.com/melanomaresearch/fulltext/2020/06000/gut_microbial_species_and_metabolic_pathways.2.aspx">https://journals.lww.com/melanomaresearch/fulltext/2020/06000/gut_microbial_species_and_metabolic_pathways.2.aspx</a>.
</div></div><p>Here, we will first perform the batch effect correction analysis. We use both the relative abundance data and the pathway abundance data from 5 public datasets: <span class="citation" data-cites="Frankel2017">AE (<a href="#ref-Frankel2017" role="doc-biblioref">2017</a>)</span>, <span class="citation" data-cites="Lee2022">KA (<a href="#ref-Lee2022" role="doc-biblioref">2022</a>)</span>, <span class="citation" data-cites="Matson2018">V (<a href="#ref-Matson2018" role="doc-biblioref">2018</a>)</span>, <span class="citation" data-cites="Peters2019">BA (<a href="#ref-Peters2019" role="doc-biblioref">2019</a>)</span>, and <span class="citation" data-cites="Wind2020">TT (<a href="#ref-Wind2020" role="doc-biblioref">2020</a>)</span>. All these datasets are available from the curatedMetagenomicData package, with a total of 285 subjects included.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">######################</span></span>
<span><span class="co"># Load melanoma data #</span></span>
<span><span class="co">######################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/curatedMetagenomicData">curatedMetagenomicData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_relative</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FrankelAE_2017"</span>, <span class="st">"GopalakrishnanV_2018"</span>, <span class="st">"LeeKA_2022"</span>, </span>
<span>                         <span class="st">"MatsonV_2018"</span>, <span class="st">"PetersBA_2019"</span>, <span class="st">"WindTT_2020"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/returnSamples.html">returnSamples</a></span><span class="op">(</span><span class="st">"relative_abundance"</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_pathway</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FrankelAE_2017"</span>, <span class="st">"GopalakrishnanV_2018"</span>, <span class="st">"LeeKA_2022"</span>, </span>
<span>                         <span class="st">"MatsonV_2018"</span>, <span class="st">"PetersBA_2019"</span>, <span class="st">"WindTT_2020"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/returnSamples.html">returnSamples</a></span><span class="op">(</span><span class="st">"pathway_abundance"</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>First, let’s look into the relative abundance data.</p>
<section id="sec-batch-effect-adjustment" class="level3" data-number="16.1.1"><h3 data-number="16.1.1" class="anchored" data-anchor-id="sec-batch-effect-adjustment">
<span class="header-section-number">16.1.1</span> Performing batch (study) effect adjustment with adjust_batch for relative abundance</h3>
<p>In this analysis, we are using two input files. The first input file is the relative abundance data file, which consists of a feature-by-sample matrix. The second input is the metadata associated with the relative abundance file, which includes the study names and overall response rate (ORR) the response variable along with other metadata. By the end of this step, we obtain a batch-adjusted relative abundance matrix.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MMUPHin</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##########################</span></span>
<span><span class="co"># Create sample metadata #</span></span>
<span><span class="co">##########################</span></span>
<span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_relative</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"study_name"</span>, <span class="st">"ORR"</span>, <span class="st">"PFS12"</span>, <span class="st">"anti_PD_1"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define response variable</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">[</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span> <span class="op">==</span> <span class="st">"responder"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"yes"</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">[</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span> <span class="op">==</span> <span class="st">"non_responder"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"no"</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">PFS12</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">PFS12</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">PFS12</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">ORR</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">ORR</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">ORR</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Filter individuals without response variable</span></span>
<span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"study_name"</span>, <span class="st">"resvar"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Convert the "study_name" to factor variable</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###########################</span></span>
<span><span class="co"># Create Species Features #</span></span>
<span><span class="co">###########################</span></span>
<span><span class="co"># Transpose the abundance matrix and change the value of abundance data to proportion unit</span></span>
<span><span class="va">data_abd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_relative</span><span class="op">)</span></span>
<span><span class="va">data_abd</span> <span class="op">&lt;-</span> <span class="va">data_abd</span> <span class="op">/</span> <span class="fl">100</span></span>
<span></span>
<span><span class="co"># Match the individuals in the data_abd</span></span>
<span><span class="va">data_abd</span> <span class="op">&lt;-</span> <span class="va">data_abd</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#  Change the rownames names for variables of interest</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">'.*s__'</span>, <span class="st">''</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use adjust_batch to correct for differences in the five studies, while controlling for the effect of cases versus control (variable resvar in data_meta).</span></span>
<span><span class="va">fit_adjust_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MMUPHin/man/adjust_batch.html">adjust_batch</a></span><span class="op">(</span>feature_abd <span class="op">=</span> <span class="va">data_abd</span>,</span>
<span>                                 batch <span class="op">=</span> <span class="st">"study_name"</span>,</span>
<span>                                 covariates <span class="op">=</span> <span class="st">"resvar"</span>,</span>
<span>                                 data <span class="op">=</span> <span class="va">data_meta</span>,</span>
<span>                                 control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_abd_adj</span> <span class="op">&lt;-</span> <span class="va">fit_adjust_batch</span><span class="op">$</span><span class="va">feature_abd_adj</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="https://github.com/microbiome/OMA/blob/master/inst/pages/images/adjust_batch_diagnostic_relative_abundance.png" class="img-fluid"></p>
<p>The first figure shows the relationship between the original batch mean parameters (Gamma) and the shrunk (regularized) batch mean parameters (Gamma (shrunken)). The points represent the 5 different study groups within the dataset. The var_batch points represent the variance within each batch, with different colors indicating different levels or categories of variance. The scatter plot on the right represents the change in mean abundance between the original and adjusted ones, displaying an obvious decrease after the batch effect adjustment analysis.</p>
</section><section id="sec-batch-adjustment-evaluation" class="level3 page-columns page-full" data-number="16.1.2"><h3 data-number="16.1.2" class="anchored" data-anchor-id="sec-batch-adjustment-evaluation">
<span class="header-section-number">16.1.2</span> Evaluation for batch effect adjustment</h3>
<p>After obtaining the adjusted abundance matrix, we can evaluate the improvement of the batch effect adjustment. Here, the total variation from study difference will be assessed through the PERMANOVA test <span class="citation" data-cites="Tang2016">(<a href="#ref-Tang2016" role="doc-biblioref">ZZ 2016</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Tang2016" class="csl-entry" role="listitem">
ZZ, Tang. 2016. <span>“PERMANOVA-s: Association Test for Microbial Community Composition That Accommodates Confounders and Multiple Distances.”</span> <em>Bioinformatics</em> 32 (17): 2618–25. <a href="https://academic.oup.com/bioinformatics/article/32/17/2618/2450751?login=true">https://academic.oup.com/bioinformatics/article/32/17/2618/2450751?login=true</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vegandevs/vegan">vegan</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">D_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">D_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">data_abd_adj</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">fit_adonis_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegan-deprecated.html">adonis</a></span><span class="op">(</span><span class="va">D_before</span> <span class="op">~</span> <span class="va">study_name</span>, data <span class="op">=</span> <span class="va">data_meta</span><span class="op">)</span></span>
<span><span class="va">fit_adonis_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegan-deprecated.html">adonis</a></span><span class="op">(</span><span class="va">D_after</span> <span class="op">~</span> <span class="va">study_name</span>, data <span class="op">=</span> <span class="va">data_meta</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">fit_adonis_before</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">)</span></span>
<span><span class="co">##  Permutation: free</span></span>
<span><span class="co">##  Number of permutations: 999</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Terms added sequentially (first to last)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##              Df SumsOfSqs MeanSqs F.Model    R2 Pr(&gt;F)    </span></span>
<span><span class="co">##  study_name   4      11.1   2.767     9.5 0.119  0.001 ***</span></span>
<span><span class="co">##  Residuals  280      81.6   0.291         0.881           </span></span>
<span><span class="co">##  Total      284      92.6                 1.000           </span></span>
<span><span class="co">##  ---</span></span>
<span><span class="co">##  Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">fit_adonis_after</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">)</span></span>
<span><span class="co">##  Permutation: free</span></span>
<span><span class="co">##  Number of permutations: 999</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Terms added sequentially (first to last)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##              Df SumsOfSqs MeanSqs F.Model    R2 Pr(&gt;F)    </span></span>
<span><span class="co">##  study_name   4       2.5   0.629    2.02 0.028  0.001 ***</span></span>
<span><span class="co">##  Residuals  280      87.4   0.312         0.972           </span></span>
<span><span class="co">##  Total      284      89.9                 1.000           </span></span>
<span><span class="co">##  ---</span></span>
<span><span class="co">##  Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Based on the results, we can see that the study differences can explain a total of 11.922% (<span class="math inline">\(R^2\)</span>) of the variability in microbial abundance profiles before study effect adjustment, whereas after adjustment this was reduced to 2.806% (<span class="math inline">\(R^2\)</span>).</p>
<p>Let’s visualize the results of the PERMANOVA test.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############</span></span>
<span><span class="co"># Ordination #</span></span>
<span><span class="co">##############</span></span>
<span><span class="co"># Before</span></span>
<span><span class="va">R2_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">fit_adonis_before</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">]</span><span class="op">*</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pcoa_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html">cmdscale</a></span><span class="op">(</span><span class="va">D_before</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ord_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">pcoa_before</span><span class="op">$</span><span class="va">points</span><span class="op">)</span></span>
<span><span class="va">percent_var_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">pcoa_before</span><span class="op">$</span><span class="va">eig</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">pcoa_before</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">before_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 1 ('</span>, <span class="va">percent_var_before</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 2 ('</span>, <span class="va">percent_var_before</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">before_phrase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Before (PERMANOVA R2 = '</span>, <span class="va">R2_before</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ord_before</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'PC1'</span>, <span class="st">'PC2'</span><span class="op">)</span></span>
<span><span class="va">ord_before</span><span class="op">$</span><span class="va">Study</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span></span>
<span></span>
<span><span class="co"># After</span></span>
<span><span class="va">R2_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">fit_adonis_after</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">]</span><span class="op">*</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pcoa_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html">cmdscale</a></span><span class="op">(</span><span class="va">D_after</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ord_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">pcoa_after</span><span class="op">$</span><span class="va">points</span><span class="op">)</span></span>
<span><span class="va">percent_var_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">pcoa_after</span><span class="op">$</span><span class="va">eig</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">pcoa_after</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">after_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 1 ('</span>, <span class="va">percent_var_after</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 2 ('</span>, <span class="va">percent_var_after</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">after_phrase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'After (PERMANOVA R2 = '</span>, <span class="va">R2_after</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ord_after</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'PC1'</span>, <span class="st">'PC2'</span><span class="op">)</span></span>
<span><span class="va">ord_after</span><span class="op">$</span><span class="va">Study</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span></span>
<span></span>
<span><span class="co"># Ordination Plot</span></span>
<span><span class="va">p_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ord_before</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">Study</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="va">before_labels</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="va">before_labels</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">before_phrase</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ord_after</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">Study</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="va">after_labels</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="va">after_labels</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">after_phrase</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p_before</span>, <span class="va">p_after</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MMUPHin_meta_analysis_files/figure-html/visualization-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="sec-differential-abundance-testing" class="level3 page-columns page-full" data-number="16.1.3"><h3 data-number="16.1.3" class="anchored" data-anchor-id="sec-differential-abundance-testing">
<span class="header-section-number">16.1.3</span> Meta-analytical differential abundance testing with lm_meta</h3>
<p>Here, we illustrate the details of the meta-analytical differential abundance testing. We have several choices for the analysis methods. First, we apply linear models (LM) for individual study-specific analyses with various transformations.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://huttenhower.sph.harvard.edu/maaslin2">Maaslin2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">transform</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NONE"</span>, <span class="st">"AST"</span>, <span class="st">"LOGIT"</span>, <span class="st">"LOG"</span><span class="op">)</span></span>
<span><span class="va">num</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">transform</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">fit_lm_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MMUPHin/man/lm_meta.html">lm_meta</a></span><span class="op">(</span>feature_abd <span class="op">=</span> <span class="va">data_abd_adj</span>,</span>
<span>                       batch <span class="op">=</span> <span class="st">"study_name"</span>,</span>
<span>                       exposure <span class="op">=</span> <span class="st">"resvar"</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">data_meta</span>,</span>
<span>                       control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>analysis_method <span class="op">=</span> <span class="st">'LM'</span>, transform <span class="op">=</span> <span class="va">transform</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">num</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">fit_lm_meta</span><span class="op">$</span><span class="va">meta_fits</span><span class="op">$</span><span class="va">qval.fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">num</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0 0 0 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since none of the linear models are able to generate significant results, we next apply the Tweedie model (using the “CPLM” analysis method in <code>MaAsLin2</code>, also implemented in the R package <code>Tweedieverse</code> <span class="citation" data-cites="Mallick2022">(<a href="#ref-Mallick2022" role="doc-biblioref">M. H 2022</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Mallick2022" class="csl-entry" role="listitem">
———. 2022. <span>“Differential Expression of Single-Cell RNA-Seq Data Using Tweedie Models.”</span> <em>Statistics in Medicine</em> 41 (18): 3492–3510. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9288986/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9288986/</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit_lm_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MMUPHin/man/lm_meta.html">lm_meta</a></span><span class="op">(</span>feature_abd <span class="op">=</span> <span class="va">data_abd_adj</span>,</span>
<span>                       batch <span class="op">=</span> <span class="st">"study_name"</span>,</span>
<span>                       exposure <span class="op">=</span> <span class="st">"resvar"</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">data_meta</span>,</span>
<span>                       control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>analysis_method <span class="op">=</span> <span class="st">'CPLM'</span>, transform <span class="op">=</span> <span class="st">'NONE'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_fits</span> <span class="op">&lt;-</span> <span class="va">fit_lm_meta</span><span class="op">$</span><span class="va">meta_fits</span></span>
<span></span>
<span><span class="va">meta_fits</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">qval.fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">coef</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">feature</span>, levels <span class="op">=</span> <span class="va">feature</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">coef</span>, x <span class="op">=</span> <span class="va">feature</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MMUPHin_meta_analysis_files/figure-html/model-selection2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Based on the results, we see that there are 22 significant features in total, among which the has the strongest positive effect, while the species has the strongest negative effect.</p>
</section></section><section id="sec-meta-analysis-pathway" class="level2" data-number="16.2"><h2 data-number="16.2" class="anchored" data-anchor-id="sec-meta-analysis-pathway">
<span class="header-section-number">16.2</span> Meta-analysis for pathway abundance data</h2>
<p>In this section, we repeat all the analysis mentioned above using the pathway relative abundance data. We first prepare the input data for the analysis.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Prepare the meta data</span></span>
<span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_pathway</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"study_name"</span>, <span class="st">"ORR"</span>, <span class="st">"PFS12"</span>, <span class="st">"anti_PD_1"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define response variable</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">[</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span> <span class="op">==</span> <span class="st">"responder"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"yes"</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">[</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span> <span class="op">==</span> <span class="st">"non_responder"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"no"</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">anti_PD_1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">PFS12</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">PFS12</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">PFS12</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">ORR</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">ORR</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">ORR</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Filter individuals without response variable</span></span>
<span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">resvar</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"study_name"</span>, <span class="st">"resvar"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Convert the "study_name" to factor variable</span></span>
<span><span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare the abundance data</span></span>
<span><span class="co"># Transpose the abundance matrix and change the value of abundance data to proportion unit</span></span>
<span><span class="va">data_abd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_pathway</span><span class="op">)</span></span>
<span><span class="va">data_abd</span> <span class="op">&lt;-</span> <span class="va">data_abd</span><span class="op">/</span><span class="fl">100</span></span>
<span></span>
<span><span class="co"># Filter out stratified features</span></span>
<span><span class="va">data_abd</span><span class="op">&lt;-</span><span class="va">data_abd</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"\\|"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">data_abd</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span></span>
<span><span class="va">data_abd</span><span class="op">&lt;-</span><span class="va">data_abd</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Match the individuals in the data_abd</span></span>
<span><span class="va">data_abd</span> <span class="op">&lt;-</span> <span class="va">data_abd</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Use adjust_batch to correct for differences in the five studies, while controlling for the effect of cases versus control (variable resvar in data_meta).</span></span>
<span><span class="va">fit_adjust_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MMUPHin/man/adjust_batch.html">adjust_batch</a></span><span class="op">(</span>feature_abd <span class="op">=</span> <span class="va">data_abd</span>,</span>
<span>                                 batch <span class="op">=</span> <span class="st">"study_name"</span>,</span>
<span>                                 covariates <span class="op">=</span> <span class="st">"resvar"</span>,</span>
<span>                                 data <span class="op">=</span> <span class="va">data_meta</span>,</span>
<span>                                 control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_abd_adj</span> <span class="op">&lt;-</span> <span class="va">fit_adjust_batch</span><span class="op">$</span><span class="va">feature_abd_adj</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="https://github.com/microbiome/OMA/tree/master/inst/pages/images/adjust_batch_diagnostic_pathway_abundance.png" class="img-fluid"></p>
<p>Next, we will evaluate the improvement of the batch effect adjustment and apply the PERMANOVA test.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">D_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">data_abd</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">D_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">data_abd_adj</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">fit_adonis_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegan-deprecated.html">adonis</a></span><span class="op">(</span><span class="va">D_before</span> <span class="op">~</span> <span class="va">study_name</span>, data <span class="op">=</span> <span class="va">data_meta</span><span class="op">)</span></span>
<span><span class="va">fit_adonis_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegan-deprecated.html">adonis</a></span><span class="op">(</span><span class="va">D_after</span> <span class="op">~</span> <span class="va">study_name</span>, data <span class="op">=</span> <span class="va">data_meta</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">fit_adonis_before</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">)</span></span>
<span><span class="co">##  Permutation: free</span></span>
<span><span class="co">##  Number of permutations: 999</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Terms added sequentially (first to last)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##              Df SumsOfSqs MeanSqs F.Model    R2 Pr(&gt;F)    </span></span>
<span><span class="co">##  study_name   4      2.01   0.503    20.6 0.229  0.001 ***</span></span>
<span><span class="co">##  Residuals  278      6.78   0.024         0.771           </span></span>
<span><span class="co">##  Total      282      8.80                 1.000           </span></span>
<span><span class="co">##  ---</span></span>
<span><span class="co">##  Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">fit_adonis_after</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">)</span></span>
<span><span class="co">##  Permutation: free</span></span>
<span><span class="co">##  Number of permutations: 999</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Terms added sequentially (first to last)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##              Df SumsOfSqs MeanSqs F.Model    R2 Pr(&gt;F)    </span></span>
<span><span class="co">##  study_name   4      1.00  0.2510    10.1 0.127  0.001 ***</span></span>
<span><span class="co">##  Residuals  278      6.88  0.0248         0.873           </span></span>
<span><span class="co">##  Total      282      7.89                 1.000           </span></span>
<span><span class="co">##  ---</span></span>
<span><span class="co">##  Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Based on the results, we see that the study differences can explain a total of 15.508% (<span class="math inline">\(R^2\)</span>) of the variability in microbial pathway profiles before study effect adjustment, whereas after adjustment this was reduced to 5.037% (<span class="math inline">\(R^2\)</span>).</p>
<p>Let’s also visualize the results of the PERMANOVA test.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Ordination</span></span>
<span><span class="co"># Before</span></span>
<span><span class="va">R2_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">fit_adonis_before</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">]</span><span class="op">*</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pcoa_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html">cmdscale</a></span><span class="op">(</span><span class="va">D_before</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ord_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">pcoa_before</span><span class="op">$</span><span class="va">points</span><span class="op">)</span></span>
<span><span class="va">percent_var_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">pcoa_before</span><span class="op">$</span><span class="va">eig</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">pcoa_before</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">before_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 1 ('</span>, <span class="va">percent_var_before</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 2 ('</span>, <span class="va">percent_var_before</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">before_phrase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Before (PERMANOVA R2 = '</span>, <span class="va">R2_before</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ord_before</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'PC1'</span>, <span class="st">'PC2'</span><span class="op">)</span></span>
<span><span class="va">ord_before</span><span class="op">$</span><span class="va">Study</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span></span>
<span></span>
<span><span class="co"># After</span></span>
<span><span class="va">R2_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">fit_adonis_after</span><span class="op">$</span><span class="va">aov.tab</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">]</span><span class="op">*</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pcoa_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html">cmdscale</a></span><span class="op">(</span><span class="va">D_after</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ord_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">pcoa_after</span><span class="op">$</span><span class="va">points</span><span class="op">)</span></span>
<span><span class="va">percent_var_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">pcoa_after</span><span class="op">$</span><span class="va">eig</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">pcoa_after</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">after_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 1 ('</span>, <span class="va">percent_var_after</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'Axis 2 ('</span>, <span class="va">percent_var_after</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">after_phrase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'After (PERMANOVA R2 = '</span>, <span class="va">R2_after</span>, <span class="st">'%)'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ord_after</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'PC1'</span>, <span class="st">'PC2'</span><span class="op">)</span></span>
<span><span class="va">ord_after</span><span class="op">$</span><span class="va">Study</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">study_name</span></span>
<span></span>
<span><span class="co"># Ordination Plot</span></span>
<span><span class="va">p_before</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ord_before</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">Study</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="va">before_labels</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="va">before_labels</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">before_phrase</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_after</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ord_after</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">Study</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="va">after_labels</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="va">after_labels</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">after_phrase</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p_before</span>, <span class="va">p_after</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MMUPHin_meta_analysis_files/figure-html/visualization-pathway-abundance-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As before, we have several choices on the analysis methods. However, since we don’t have significance at the FDR level, we use the unadjusted p-values to showcase the ``top hits’’.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">transform</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NONE"</span>, <span class="st">"AST"</span>, <span class="st">"LOGIT"</span>, <span class="st">"LOG"</span><span class="op">)</span></span>
<span><span class="va">num</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">transform</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">fit_lm_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MMUPHin/man/lm_meta.html">lm_meta</a></span><span class="op">(</span>feature_abd <span class="op">=</span> <span class="va">data_abd_adj</span>,</span>
<span>                       batch <span class="op">=</span> <span class="st">"study_name"</span>,</span>
<span>                       exposure <span class="op">=</span> <span class="st">"resvar"</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">data_meta</span>,</span>
<span>                       control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>analysis_method <span class="op">=</span> <span class="st">'LM'</span>, transform <span class="op">=</span> <span class="va">transform</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">num</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">fit_lm_meta</span><span class="op">$</span><span class="va">meta_fits</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">num</span><span class="op">)</span></span>
<span><span class="co">##  [1] 11 10  9 11</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit_lm_meta2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MMUPHin/man/lm_meta.html">lm_meta</a></span><span class="op">(</span>feature_abd <span class="op">=</span> <span class="va">data_abd_adj</span>,</span>
<span>                       batch <span class="op">=</span> <span class="st">"study_name"</span>,</span>
<span>                       exposure <span class="op">=</span> <span class="st">"resvar"</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">data_meta</span>,</span>
<span>                       control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>analysis_method <span class="op">=</span> <span class="st">'CPLM'</span>, transform <span class="op">=</span> <span class="st">'NONE'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">fit_lm_meta2</span><span class="op">$</span><span class="va">meta_fits</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">num</span><span class="op">)</span></span>
<span><span class="co">##  [1] 15</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is the summary table for the number of significant pathways using various meta-analytic differential analysis methods and transformations:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">analysis_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  Analysis_Method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"LM"</span>, <span class="st">"CPLM"</span><span class="op">)</span>,</span>
<span>  Transform_Parameter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NONE"</span>, <span class="st">"AST"</span>, <span class="st">"LOGIT"</span>, <span class="st">"LOG"</span>, <span class="st">"NONE"</span><span class="op">)</span>,</span>
<span>  Number_Of_Significant_Pathways <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">11</span>, <span class="fl">10</span>, <span class="fl">9</span>, <span class="fl">11</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">analysis_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Analysis Method   "</span>, <span class="st">"   Transformation   "</span>, <span class="st">"   Number of Significant Pathways"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">analysis_data</span>, caption <span class="op">=</span> <span class="st">"Analysis Results"</span>, align <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'l'</span>,<span class="st">'l'</span>,<span class="st">'r'</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Analysis Results</caption>
<colgroup>
<col style="width: 25%">
<col style="width: 28%">
<col style="width: 45%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Analysis Method</th>
<th style="text-align: left;">Transformation</th>
<th style="text-align: right;">Number of Significant Pathways</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">LM</td>
<td style="text-align: left;">NONE</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="even">
<td style="text-align: left;">LM</td>
<td style="text-align: left;">AST</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LM</td>
<td style="text-align: left;">LOGIT</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">LM</td>
<td style="text-align: left;">LOG</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CPLM</td>
<td style="text-align: left;">NONE</td>
<td style="text-align: right;">15</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Finally, let’s visualize the results using the “CPLM” analysis method, which identified largest number of significant pathways.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Extract the results</span></span>
<span><span class="va">meta_fits2</span> <span class="op">&lt;-</span> <span class="va">fit_lm_meta2</span><span class="op">$</span><span class="va">meta_fits</span></span>
<span><span class="co"># meta_fits3 &lt;- meta_fits2[meta_fits2$pval &lt; 0.05 &amp; !is.na(meta_fits2$pval), ]</span></span>
<span></span>
<span><span class="co"># Create the figure</span></span>
<span><span class="va">meta_fits2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">coef</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">feature</span>, levels <span class="op">=</span> <span class="va">feature</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">coef</span>, x <span class="op">=</span> <span class="va">feature</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MMUPHin_meta_analysis_files/figure-html/visualization-pathway-abundance2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Based on the results, we observe 15 significant features in total. Among these, queuosine biosynthesis exhibits the strongest negative effects. Additionally, L-lysine degradation XI is the most positively significant pathway, while xylose degradation IV is the most negatively significant pathway.</p>



</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/IntegratedLearner.html" class="pagination-link" aria-label="Multi-omics Integration">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/MSEA.html" class="pagination-link" aria-label="**Microbe** Set **Enrichment Analysis** (MSEA)">
        <span class="nav-page-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>
  </div>
</footer>


</body></html>