<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Microbiome Analysis - 17&nbsp; Microbe Set Enrichment Analysis (MSEA)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/19_visualization_techniques.html" rel="next">
<link href="../pages/MMUPHin_meta_analysis.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/60_network_learning.html">Advanced topics</a></li><li class="breadcrumb-item"><a href="../pages/MSEA.html"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">**Microbe** Set **Enrichment Analysis** (MSEA)</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Microbiome Analysis</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/microbiome/OMA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://microbiome.github.io/OMA/docs/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">OMA ecosystem</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Focus topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/60_network_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Network learning and analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/IntegratedLearner.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Multi-omics Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MMUPHin_meta_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/MSEA.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Training</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Contributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Session info</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#input-data-for-msea-using-species-relative-abundance-data" id="toc-input-data-for-msea-using-species-relative-abundance-data" class="nav-link active" data-scroll-target="#input-data-for-msea-using-species-relative-abundance-data"><span class="header-section-number">17.1</span> Input data for MSEA using species relative abundance data</a></li>
  <li><a href="#performing-the-msea-analysis-with-species-relative-abundance-data" id="toc-performing-the-msea-analysis-with-species-relative-abundance-data" class="nav-link" data-scroll-target="#performing-the-msea-analysis-with-species-relative-abundance-data"><span class="header-section-number">17.2</span> Performing the MSEA analysis with species relative abundance data</a></li>
  <li><a href="#print-the-most-significant-modules-for-species-relative-abundance-data" id="toc-print-the-most-significant-modules-for-species-relative-abundance-data" class="nav-link" data-scroll-target="#print-the-most-significant-modules-for-species-relative-abundance-data"><span class="header-section-number">17.3</span> Print the most significant modules for species relative abundance data</a></li>
  <li><a href="#performing-the-msea-analysis-with-pathway-relative-abundance-data" id="toc-performing-the-msea-analysis-with-pathway-relative-abundance-data" class="nav-link" data-scroll-target="#performing-the-msea-analysis-with-pathway-relative-abundance-data"><span class="header-section-number">17.4</span> Performing the MSEA analysis with pathway relative abundance data</a></li>
  <li><a href="#print-the-most-significant-modules-for-pathway-abundance-data" id="toc-print-the-most-significant-modules-for-pathway-abundance-data" class="nav-link" data-scroll-target="#print-the-most-significant-modules-for-pathway-abundance-data"><span class="header-section-number">17.5</span> Print the most significant modules for pathway abundance data</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-MSEA" class="quarto-section-identifier"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title"><strong>Microbe</strong> Set <strong>Enrichment Analysis</strong> (MSEA)</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-layout-align="center">
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script><style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
</div>
<p>Similar to gene set enrichment analyses for genes <span class="citation" data-cites="Subramanian2005">(<a href="#ref-Subramanian2005" role="doc-biblioref">A 2005</a>)</span>, an obvious next step following differential abundance analysis in microbiome studies is to conduct enrichment analysis for microbe sets, known as microbe set enrichment analysis (MSEA) <span class="citation" data-cites="Kou2020">(<a href="#ref-Kou2020" role="doc-biblioref">Y 2020</a>)</span>. Similar to GSEA, the primary goal of MSEA is to detect the modest but coordinated changes in pre-specified sets of related microbial features. Such a set might include all the microbes in a specific pathway or microbial genes that have been shown to be co-regulated based on previously published studies. Like GSEA, MSEA aggregates the per-feature statistics across microbes within a microbe set. This corresponds to the hypothesis that many relevant phenotype differences are manifested by small but consistent changes in a set of features.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Subramanian2005" class="csl-entry" role="listitem">
A, Subramanian. 2005. <span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.”</span> <em>Proceedings of the National Academy of Sciences</em> 102 (43): 15545–50. <a href="https://www.pnas.org/doi/10.1073/pnas.0506580102?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed">https://www.pnas.org/doi/10.1073/pnas.0506580102?url_ver=Z39.88-2003&amp;rfr_id=ori:rid:crossref.org&amp;rfr_dat=cr_pub%20%200pubmed</a>.
</div><div id="ref-Kou2020" class="csl-entry" role="listitem">
Y, Kou. 2020. <span>“Microbe-Set Enrichment Analysis Facilitates Functional Interpretation of Microbiome Profiling Data.”</span> <em>Scientific Reports</em> 10 (1): 21466. <a href="https://www.nature.com/articles/s41598-020-78511-y">https://www.nature.com/articles/s41598-020-78511-y</a>.
</div></div><p>The goal of the MSEA approach is to determine if the members of S (microbe set) are randomly distributed throughout the ranked list of features (L) or primarily found at the top or bottom. We will use the R package <code>gsEasy</code> to conduct the MSEA test described by <span class="citation" data-cites="Subramanian2005">A (<a href="#ref-Subramanian2005" role="doc-biblioref">2005</a>)</span>.</p>
<section id="input-data-for-msea-using-species-relative-abundance-data" class="level2 page-columns page-full" data-number="17.1"><h2 data-number="17.1" class="anchored" data-anchor-id="input-data-for-msea-using-species-relative-abundance-data">
<span class="header-section-number">17.1</span> Input data for MSEA using species relative abundance data</h2>
<p>In this chapter, we will use the publicly available Inflammatory Bowel Diseases (IBD) microbiome data from the integrative Human Microbiome Project (iHMP) available from the <code>curatedMetagenomicData</code> package <span class="citation" data-cites="Lloyd-Price2019">(<a href="#ref-Lloyd-Price2019" role="doc-biblioref">J 2019</a>)</span>. We aim to conduct MSEA analysis based on both taxonomic profiles (species relative abundances) and functional profiles (pathway relative abundances).</p>
<div class="no-row-height column-margin column-container"><div id="ref-Lloyd-Price2019" class="csl-entry" role="listitem">
J, Lloyd-Price. 2019. <span>“<a href="">Multi-Omics of the Gut Microbial Ecosystem in Inflammatory Bowel Diseases</a>.”</span> <em>Nature</em> 569 (7758): 655–62.
</div></div></section><section id="performing-the-msea-analysis-with-species-relative-abundance-data" class="level2 page-columns page-full" data-number="17.2"><h2 data-number="17.2" class="anchored" data-anchor-id="performing-the-msea-analysis-with-species-relative-abundance-data">
<span class="header-section-number">17.2</span> Performing the MSEA analysis with species relative abundance data</h2>
<p>We will first prepare the input feature table and sample metadata for differential abundance analysis using <code>MaAsLin2</code> <span class="citation" data-cites="Mallick2021">(<a href="#ref-Mallick2021" role="doc-biblioref">H 2021</a>)</span>. The ranked feature list from the differential abundance analysis serves as an input for the MSEA.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Mallick2021" class="csl-entry" role="listitem">
H, Mallick. 2021. <span>“Multivariable Association Discovery in Population-Scale Meta-Omics Studies.”</span> <em>PLOS Computational Biology</em> 17 (11): e1009442. <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442">https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442</a>.
</div></div><div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/load-pkg-data_5ece69d4af3ce18def05369bde7b55a9">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">##################</span></span>
<span><span class="co"># Load iHMP data #</span></span>
<span><span class="co">##################</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/curatedMetagenomicData">curatedMetagenomicData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_relative</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"HMP_2019_ibdmdb"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/returnSamples.html">returnSamples</a></span><span class="op">(</span><span class="st">"relative_abundance"</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##########################</span></span>
<span><span class="co"># Create sample metadata #</span></span>
<span><span class="co">##########################</span></span>
<span></span>
<span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_relative</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">visit_number</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"age"</span>, <span class="st">"disease"</span>, <span class="st">"antibiotics_current_use"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#################</span></span>
<span><span class="co"># Set reference #</span></span>
<span><span class="co">#################</span></span>
<span></span>
<span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span><span class="op">)</span></span>
<span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span>, <span class="st">'healthy'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###########################</span></span>
<span><span class="co"># Create species features #</span></span>
<span><span class="co">###########################</span></span>
<span></span>
<span><span class="va">feature_species_t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_relative</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feature_species_t</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">'.*s__'</span>, <span class="st">''</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feature_species_t</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################</span></span>
<span><span class="co"># Subset to baseline samples #</span></span>
<span><span class="co">##############################</span></span>
<span></span>
<span><span class="va">feature_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">feature_species_t</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">feature_species</span> <span class="op">&lt;-</span> <span class="va">feature_species</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">feature_species</span> <span class="op">&lt;-</span> <span class="va">feature_species</span> <span class="op">/</span> <span class="fl">100</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">feature_species_t</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">se_relative</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the next step, we will use <code>MaAsLin2</code> to fit a multivariable regression model for testing the association between microbial species abundance versus IBD diagnosis. The analysis method we use here is “LM”, which is the default setting. We also adjust for age and antibiotic usage, following the original study.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/fit-model-Maaslin2_87c334924f3b6d531b0bff912ab80609">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://huttenhower.sph.harvard.edu/maaslin2">Maaslin2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Maaslin2/man/Maaslin2.html">Maaslin2</a></span><span class="op">(</span>input_data     <span class="op">=</span> <span class="va">feature_species</span>, </span>
<span>                    input_metadata <span class="op">=</span> <span class="va">sample_metadata</span>, </span>
<span>                    normalization  <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>                    output         <span class="op">=</span> <span class="st">"output_species"</span>, </span>
<span>                    fixed_effects  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"disease"</span>, <span class="st">"age"</span>, <span class="st">"antibiotics_current_use"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Creating output folder"</span></span>
<span><span class="co">##  [1] "Creating output feature tables folder"</span></span>
<span><span class="co">##  [1] "Creating output fits folder"</span></span>
<span><span class="co">##  [1] "Creating output figures folder"</span></span>
<span><span class="co">##  2024-04-26 15:42:19.186455 INFO::Writing function arguments to log file</span></span>
<span><span class="co">##  2024-04-26 15:42:19.200922 INFO::Verifying options selected are valid</span></span>
<span><span class="co">##  2024-04-26 15:42:19.227448 INFO::Determining format of input files</span></span>
<span><span class="co">##  2024-04-26 15:42:19.227997 INFO::Input format is data samples as rows and metadata samples as rows</span></span>
<span><span class="co">##  2024-04-26 15:42:19.233478 INFO::Formula for fixed effects: expr ~  disease + age + antibiotics_current_use</span></span>
<span><span class="co">##  2024-04-26 15:42:19.234131 INFO::Factor detected for categorial metadata 'disease'. Provide a reference argument or manually set factor ordering to change reference level.</span></span>
<span><span class="co">##  2024-04-26 15:42:19.234725 INFO::Filter data based on min abundance and min prevalence</span></span>
<span><span class="co">##  2024-04-26 15:42:19.235139 INFO::Total samples in data: 136</span></span>
<span><span class="co">##  2024-04-26 15:42:19.235539 INFO::Min samples required with min abundance for a feature not to be filtered: 13.600000</span></span>
<span><span class="co">##  2024-04-26 15:42:19.241186 INFO::Total filtered features: 452</span></span>
<span><span class="co">##  2024-04-26 15:42:19.241845 INFO::Filtered feature names from abundance and prevalence filtering: Clostridiales.bacterium.1_7_47FAA, Enterocloster.asparagiformis, Gordonibacter.pamelaeae, Firmicutes.bacterium.CAG.424, Adlercreutzia.equolifaciens, Anaerotruncus.sp..CAG.528, Parabacteroides.johnsonii, Erysipelothrix.larvae, Proteus.mirabilis, Bacteroides.sp..43_108, Blautia.producta, Tyzzerella.nexilis, Ruminococcus.obeum.CAG.39, X.Clostridium..scindens, Blautia.hydrogenotrophica, Lachnoclostridium.sp..An181, Candidatus.Stoquefichus.sp..KLE1796, Gemmiger.sp..An87, Lachnoclostridium.sp..An138, Harryflintia.acetispora, Streptococcus.thermophilus, Prevotella.buccae, Dorea.sp..CAG.317, Proteus.vulgaris, Proteus.hauseri, X.Clostridium..spiroforme, Blautia.hansenii, Eubacterium.sp..OM08.24, Parabacteroides.sp..CAG.409, Desulfovibrio.piger, Clostridium.sp..CAG.242, Pseudoflavonifractor.sp..An184, Clostridiales.bacterium.CHKCI006, Pseudoflavonifractor.sp..An85, Christensenella.minuta, Bifidobacterium.catenulatum, Slackia.isoflavoniconvertens, Alistipes.inops, Haemophilus.pittmaniae, Amedibacillus.dolichus, Eubacterium.dolichum.CAG.375, Turicibacter.sanguinis, Lactobacillus.delbrueckii, Aeriscardovia.aeriphila, Firmicutes.bacterium.CAG.145, Klebsiella.quasipneumoniae, Cutibacterium.granulosum, Methanobrevibacter.smithii, Porphyromonas.asaccharolytica, Clostridium.perfringens, Clostridium.paraputrificum, Limosilactobacillus.fermentum, Prevotella.sp..CAG.279, Fusobacterium.ulcerans, Bifidobacterium.breve, Prevotella.sp..CAG.891, Lactococcus.piscium, Klebsiella.oxytoca, Citrobacter.freundii, Citrobacter.braakii, Bifidobacterium.animalis, Citrobacter.youngae, Lactobacillus.acidophilus, Anaeroglobus.geminatus, Bifidobacterium.dentium, Clostridium.neonatale, Clostridioides.difficile, Enterococcus.faecium, Morganella.morganii, Streptococcus.oralis, Klebsiella.michiganensis, Clostridium.sp..chh4.2, Bacteroides.nordii, Klebsiella.aerogenes, Leclercia.adecarboxylata, Enterobacter.mori, Kluyvera.ascorbata, Pseudoflavonifractor.capillosus, Ruminococcaceae.bacterium.D5, Raoultella.planticola, Roseburia.sp..CAG.303, Alloscardovia.omnicolens, Eubacterium.sp..CAG.180, Eubacterium.sp..CAG.274, Cutibacterium.acnes, Dialister.sp..CAG.357, Bacteroides.oleiciplenus, Coprobacter.secundus, Kocuria.palustris, Acinetobacter.lwoffii, Micrococcus.lylae, Pararheinheimera.mesophila, Desulfovibrionaceae.bacterium, Massilimicrobiota.timonensis, Fretibacterium.fastidiosum, Streptococcus.vestibularis, Clostridium.sp..CAG.411, Megamonas.funiformis, Megamonas.hypermegale, Megamonas.funiformis.CAG.377, Streptococcus.gordonii, Coprobacillus.cateniformis, Prevotella.colorans, Prevotella.buccalis, Anaerofustis.stercorihominis, Latilactobacillus.sakei, Clostridium.disporicum, Ruminococcus.sp..CAG.330, Streptococcus.agalactiae, Aggregatibacter.segnis, Fusobacterium.nucleatum, Bifidobacterium.pseudolongum, Allisonella.histaminiformans, Prevotella.bivia, Clostridium.sp..CAG.590, Ruminococcus.callidus, Holdemanella.biformis, Escherichia.marmotae, Terrisporobacter.othiniensis, Porphyromonas.somerae, Prevotella.salivae, Citrobacter.amalonaticus, Ligilactobacillus.salivarius, Megasphaera.micronuciformis, Neisseria.sp..oral.taxon.014, Anaerostipes.caccae, Blautia.argi, Escherichia.albertii, Prevotella.disiens, Porphyromonas.uenonis, Peptococcus.niger, Bacteroides.sp..CAG.144, Eubacterium.sp..An11, Blautia.coccoides, Oxalobacter.formigenes, Tractidigestivibacter.scatoligenes, Dialister.pneumosintes, Megasphaera.sp..DISK.18, Clostridium.sp..CAG.167, Hafnia.paralvei, Alistipes.onderdonkii, Coprobacter.sp., Corynebacterium.matruchotii, Bacteroides.stercorirosoris, Staphylococcus.epidermidis, Parabacteroides.goldsteinii, Roseburia.sp..CAG.182, Oscillibacter.sp..PC13, Firmicutes.bacterium.CAG.110, Prevotella.sp..885, Parvimonas.micra, Peptostreptococcus.stomatis, Moraxella.osloensis, Enhydrobacter.aerosaccus, Micrococcus.luteus, Micrococcus.aloeverae, Dialister.micraerophilus, Campylobacter.concisus, Megasphaera.sp..MJR8396C, Cloacibacillus.porcorum, Desulfovibrio.fairfieldensis, Phascolarctobacterium.sp..CAG.266, Rikenella.microfusus, Roseburia.sp..CAG.309, Flavonifractor.sp..An10, Victivallis.vadensis, Firmicutes.bacterium.CAG.95, Enorma.massiliensis, Citrobacter.pasteurii, Streptococcus.salivarius.CAG.79, Veillonella.seminalis, Enterococcus.faecalis, Streptococcus.mitis, Butyricicoccus.pullicaecorum, Prevotella.corporis, Lachnoclostridium.sp..An131, Lachnoclostridium.sp..An118, X.Clostridium..methylpentosum, Haemophilus.parahaemolyticus, Neisseria.flavescens, Veillonella.rogosae, Neisseria.subflava, Sharpea.azabuensis, Anaerofilum.sp..An201, Romboutsia.ilealis, Fusobacterium.periodonticum, Lachnoclostridium.sp..An14, Flavonifractor.sp..An100, Clostridium.sp..CAG.253, Arthrospira.platensis, Lactobacillus.amylovorus, Limosilactobacillus.oris, Streptococcus.lutetiensis, Weissella.cibaria, Rothia.mucilaginosa, Streptococcus.pasteurianus, Prevotella.melaninogenica, Prevotella.jejuni, Veillonella.tobetsuensis, Streptococcus.sp..M334, Actinomyces.sp..ICM47, Streptococcus.sp..A12, Clostridium.sp..CAG.678, Candidatus.Methanomassiliicoccus.intestinalis, Candidatus.Gastranaerophilales.bacterium, Prevotella.denticola, Leuconostoc.lactis, Prevotella.oris, Prevotella.dentalis, X.Eubacterium..infirmum, Shuttleworthia.satelles, Prevotella.nigrescens, Enterococcus.hirae, Adlercreutzia.caecimuris, Citrobacter.portucalensis, Raoultella.ornithinolytica, Dysgonomonas.mossii, Firmicutes.bacterium.CAG.646, Streptococcus.gallolyticus, Christensenella.hongkongensis, Lactobacillus.paragasseri, Lactobacillus.gasseri, Peptostreptococcus.anaerobius, Acinetobacter.ursingii, Lactococcus.lactis, Odoribacter.laneus, Haemophilus.sputorum, Peptoniphilus.harei, Pediococcus.acidilactici, Fusobacterium.gonidiaformans, Fusobacterium.equinum, Leuconostoc.garlicum, Campylobacter.showae, Aeromonas.hydrophila, Aeromonas.dhakensis, Prevotella.stercorea, Phascolarctobacterium.succinatutens, Prevotella.sp..AM42.24, Blastocystis.sp..subtype.1, Bifidobacterium.angulatum, Prevotella.timonensis, Butyrivibrio.crossotus, Limosilactobacillus.vaginalis, Ligilactobacillus.ruminis, Weissella.confusa, Streptococcus.mutans, Clostridium.celatum, Streptococcus.australis, Streptococcus.sanguinis, Schaalia.turicensis, Peptoniphilus.duerdenii, Clostridium.baratii, Streptococcus.sp..F0442, Prevotella.sp..CAG.520, Cronobacter.sakazakii, Actinomyces.urogenitalis, Actinomyces.graevenitzii, Parvimonas.sp..oral.taxon.110, Eikenella.corrodens, Parvimonas.sp..oral.taxon.393, Anaerococcus.vaginalis, Gemella.asaccharolytica, Lachnospiraceae.bacterium.2_1_46FAA, Veillonella.sp..CAG.933, Lacticaseibacillus.rhamnosus, Campylobacter.upsaliensis, Lachnospiraceae.bacterium.oral.taxon.096, Campylobacter.gracilis, Gemmiger.sp..An50, Scardovia.wiggsiae, Limosilactobacillus.reuteri, Enterococcus.avium, Enterococcus.thailandicus, Paucilactobacillus.vaccinostercus, Enterococcus.casseliflavus, Bifidobacterium.pullorum, Actinomyces.sp..oral.taxon.180, Enterococcus.sp..3H8_DIV0648, Enterococcus.gallinarum, Varibaculum.cambriense, Veillonella.rodentium, Lactobacillus.crispatus, Lawsonella.clevelandensis, Porphyromonas.sp..HMSC065F10, Citrobacter.werkmanii, Lactobacillus.johnsonii, Sarcina.ventriculi, Streptococcus.macedonicus, Klebsiella.variicola.CAG.634, Phytobacter.palmae, Enterococcus.raffinosus, Prevotella.amnii, Catenibacterium.mitsuokai, Saccharomyces.cerevisiae, Eubacterium.limosum, Staphylococcus.hominis, Alloprevotella.tannerae, Haemophilus.paraphrohaemolyticus, X.Eubacterium..brachy, Ruminococcus.sp..CAG.403, Fannyhessea.vaginae, Cronobacter.malonaticus, Streptococcus.infantis, Actinomyces.sp..oral.taxon.181, Clostridium.butyricum, Abiotrophia.sp..HMSC24B09, Gemella.haemolysans, Firmicutes.bacterium.CAG.170, Solobacterium.moorei, Pseudomonas.guguanensis, Lactobacillus.jensenii, Clostridium.cadaveris, Aggregatibacter.sp..oral.taxon.458, Anaerocolumna.aminovalerica, Clostridium.sporogenes, Clostridium.botulinum, Campylobacter.hominis, Murimonas.intestini, Clostridium.sp..CAG.964, Pediococcus.pentosaceus, Corynebacterium.amycolatum, Parvimonas.sp..KA00067, Bacteroidales.bacterium.KA00251, Fusobacterium.mortiferum, Lacrimispora.celerecrescens, Butyrivibrio.sp..CAG.318, Cellulosilyticum.lentocellum, Eubacteriaceae.bacterium.CHKCI005, Eubacterium.coprostanoligenes, Firmicutes.bacterium.CAG.238, Dysgonomonas.gadei, Bacteroides.fluxus, Bacteroides.sp..D2, Citrobacter.farmeri, Gemella.sanguinis, Sanguibacteroides.justesenii, Porphyromonas.endodontalis, Streptococcus.peroris, Actinomyces.sp..HPA0247, Lactiplantibacillus.plantarum, Limosilactobacillus.mucosae, Providencia.alcalifaciens, Robinsoniella.sp..RHS, Fructilactobacillus.sanfranciscensis, Prevotella.sp..CAG.1185, Peptoniphilus.lacrimalis, Gemella.morbillorum, Frigoribacterium.sp..Leaf8, Clostridium.sp..MSTE9, X.Clostridium..hylemonae, Treponema.lecithinolyticum, Faecalicoccus.pleomorphus, Flavonifractor.sp..An306, Prevotella.oralis, Peptoniphilus.sp..BV3C26, Brachyspira.pilosicoli, Bacteroides.sp..OM08.11, Proteus.penneri, Corynebacterium.kroppenstedtii, Faecalitalea.cylindroides, Wohlfahrtiimonas.chitiniclastica, Bacteroides.clarus, Acidaminococcus.sp..CAG.542, Peptoniphilus.coxii, Alistipes.timonensis, Clostridium.sp..D5, Prevotella.bergensis, Ruminococcus.sp..CAG.563, Kluyvera.cryocrescens, Ezakiella.coagulans, X.Bacteroides..pectinophilus, Bacteroides.sp..CAG.661, Bacteroides.sp..CAG.598, Comamonas.kerstersii, Firmicutes.bacterium.CAG.534, Clostridium.sp..CAG.413, Schaalia.odontolytica, Streptococcus.cristatus, Actinomyces.sp..HMSC035G02, Lancefieldella.rimae, Streptococcus.milleri, Finegoldia.magna, Lancefieldella.parvula, Prevotella.oulorum, Dellaglioa.algida, Lactonifactor.longoviformis, Hafnia.alvei, Neisseria.cinerea, Corynebacterium.oculi, Ruminococcus.sp..CAG.488, Corynebacterium.accolens, Prevotella.sp..S7.1.8, Campylobacter.ureolyticus, Atopobium.deltae, Gleimia.europaea, Mitsuokella.jalaludinii, Mitsuokella.multacida, Peptoniphilus.sp..HMSC062D09, Anaerococcus.lactolyticus, Granulicatella.adiacens, Anaerosporobacter.mobilis, Clostridium.sp..7_2_43FAA, Kosakonia.sacchari, Staphylococcus.aureus, Ligilactobacillus.animalis, Weissella.viridescens, Dysgonomonas.sp..37.18, Serratia.liquefaciens, Streptococcus.sp..HPH0090, Anaerostipes.sp..494a, Lactococcus.petauri, Obesumbacterium.proteus, Fusobacterium.sp..oral.taxon.370, X.Butyribacterium..methylotrophicum, Phocaeicola.sartorii, Bacteroides.sp..CAG.530, Megasphaera.elsdenii, Prevotella.sp..CAG.1092, Fusobacterium.sp..CAG.439, Anaeromassilibacillus.sp..An172, Aggregatibacter.aphrophilus, Aerococcus.urinaeequi, Pseudomonas.fragi, Thermoleophilum.album, Bacteroides.sp..CAG.633, Serratia.marcescens, Actinotignum.timonense, Fusobacterium.naviforme, Haemophilus.influenzae, Prevotella.intermedia, Bacteroidetes.oral.taxon.274, Citrobacter.europaeus, Bifidobacterium.asteroides, Enterococcus.durans, Enterococcus.pseudoavium, Microvirgula.aerodenitrificans, Yersinia.frederiksenii, Faecalicatena.orotica, Enterococcus.asini, Paenibacillus.macerans, Bavariicoccus.seileri, Kluyvera.georgiana, Prevotella.histicola, Prevotella.pallens, Chlamydia.ibidis, Enterococcus.mundtii, Anaerostipes.sp..992a, Actinobaculum.sp..oral.taxon.183, Lachnoclostridium.sp..An298, Haemophilus.haemolyticus, Enterococcus.dispar, Atopobium.minutum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.244453 INFO::Total filtered features with variance filtering: 0</span></span>
<span><span class="co">##  2024-04-26 15:42:19.244944 INFO::Filtered feature names from variance filtering:</span></span>
<span><span class="co">##  2024-04-26 15:42:19.245341 INFO::Running selected normalization method: NONE</span></span>
<span><span class="co">##  2024-04-26 15:42:19.245801 INFO::Applying z-score to standardize continuous metadata</span></span>
<span><span class="co">##  2024-04-26 15:42:19.249876 INFO::Running selected transform method: LOG</span></span>
<span><span class="co">##  2024-04-26 15:42:19.252637 INFO::Running selected analysis method: LM</span></span>
<span><span class="co">##  2024-04-26 15:42:19.256168 INFO::Fitting model to feature number 1, Phocaeicola.vulgatus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.262338 INFO::Fitting model to feature number 2, Bacteroides.uniformis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.264696 INFO::Fitting model to feature number 3, Bacteroides.thetaiotaomicron</span></span>
<span><span class="co">##  2024-04-26 15:42:19.26686 INFO::Fitting model to feature number 4, Faecalibacterium.prausnitzii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.269037 INFO::Fitting model to feature number 5, Roseburia.faecis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.271161 INFO::Fitting model to feature number 6, Bacteroides.caccae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.273333 INFO::Fitting model to feature number 7, Enterocloster.clostridioformis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.27547 INFO::Fitting model to feature number 8, Bacteroides.fragilis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.277666 INFO::Fitting model to feature number 9, Fusicatenibacter.saccharivorans</span></span>
<span><span class="co">##  2024-04-26 15:42:19.279809 INFO::Fitting model to feature number 10, Flavonifractor.plautii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.282003 INFO::Fitting model to feature number 11, Dialister.invisus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.284149 INFO::Fitting model to feature number 12, Ruminococcus.bicirculans</span></span>
<span><span class="co">##  2024-04-26 15:42:19.286327 INFO::Fitting model to feature number 13, Blautia.sp..CAG.257</span></span>
<span><span class="co">##  2024-04-26 15:42:19.288487 INFO::Fitting model to feature number 14, X.Ruminococcus..gnavus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.290629 INFO::Fitting model to feature number 15, Dorea.longicatena</span></span>
<span><span class="co">##  2024-04-26 15:42:19.292797 INFO::Fitting model to feature number 16, Anaerobutyricum.hallii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.294929 INFO::Fitting model to feature number 17, Enterocloster.bolteae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.29708 INFO::Fitting model to feature number 18, Eubacterium.sp..CAG.38</span></span>
<span><span class="co">##  2024-04-26 15:42:19.299214 INFO::Fitting model to feature number 19, Lacrimispora.saccharolytica</span></span>
<span><span class="co">##  2024-04-26 15:42:19.301364 INFO::Fitting model to feature number 20, Intestinimonas.butyriciproducens</span></span>
<span><span class="co">##  2024-04-26 15:42:19.303517 INFO::Fitting model to feature number 21, Anaerostipes.hadrus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.305683 INFO::Fitting model to feature number 22, Anaerotignum.lactatifermentans</span></span>
<span><span class="co">##  2024-04-26 15:42:19.30781 INFO::Fitting model to feature number 23, Ruthenibacterium.lactatiformans</span></span>
<span><span class="co">##  2024-04-26 15:42:19.309987 INFO::Fitting model to feature number 24, Lawsonibacter.asaccharolyticus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.312128 INFO::Fitting model to feature number 25, Clostridium.bolteae.CAG.59</span></span>
<span><span class="co">##  2024-04-26 15:42:19.314282 INFO::Fitting model to feature number 26, X.Clostridium..symbiosum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.316407 INFO::Fitting model to feature number 27, Anaerotruncus.colihominis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.318591 INFO::Fitting model to feature number 28, Agathobaculum.butyriciproducens</span></span>
<span><span class="co">##  2024-04-26 15:42:19.320737 INFO::Fitting model to feature number 29, Escherichia.coli</span></span>
<span><span class="co">##  2024-04-26 15:42:19.322901 INFO::Fitting model to feature number 30, Eisenbergiella.massiliensis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.325092 INFO::Fitting model to feature number 31, Enterocloster.citroniae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.32721 INFO::Fitting model to feature number 32, Bacteroides.stercoris</span></span>
<span><span class="co">##  2024-04-26 15:42:19.329369 INFO::Fitting model to feature number 33, Monoglobus.pectinilyticus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.331514 INFO::Fitting model to feature number 34, Alistipes.putredinis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.333691 INFO::Fitting model to feature number 35, Parabacteroides.distasonis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.335853 INFO::Fitting model to feature number 36, Roseburia.hominis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.338041 INFO::Fitting model to feature number 37, Bifidobacterium.longum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.340195 INFO::Fitting model to feature number 38, X.Eubacterium..siraeum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.342367 INFO::Fitting model to feature number 39, Parabacteroides.merdae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.344515 INFO::Fitting model to feature number 40, Roseburia.inulinivorans</span></span>
<span><span class="co">##  2024-04-26 15:42:19.346697 INFO::Fitting model to feature number 41, Bacteroides.cellulosilyticus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.348883 INFO::Fitting model to feature number 42, Bacteroides.ovatus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.351033 INFO::Fitting model to feature number 43, Blautia.wexlerae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.353198 INFO::Fitting model to feature number 44, Bifidobacterium.adolescentis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.355338 INFO::Fitting model to feature number 45, Lachnospira.eligens</span></span>
<span><span class="co">##  2024-04-26 15:42:19.357568 INFO::Fitting model to feature number 46, X.Clostridium..leptum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.359747 INFO::Fitting model to feature number 47, Eggerthella.lenta</span></span>
<span><span class="co">##  2024-04-26 15:42:19.361911 INFO::Fitting model to feature number 48, Bacteroides.xylanisolvens</span></span>
<span><span class="co">##  2024-04-26 15:42:19.364031 INFO::Fitting model to feature number 49, Phocaeicola.dorei</span></span>
<span><span class="co">##  2024-04-26 15:42:19.366245 INFO::Fitting model to feature number 50, Enterocloster.lavalensis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.368408 INFO::Fitting model to feature number 51, X.Eubacterium..rectale</span></span>
<span><span class="co">##  2024-04-26 15:42:19.37056 INFO::Fitting model to feature number 52, Akkermansia.muciniphila</span></span>
<span><span class="co">##  2024-04-26 15:42:19.372718 INFO::Fitting model to feature number 53, Acidaminococcus.intestini</span></span>
<span><span class="co">##  2024-04-26 15:42:19.374861 INFO::Fitting model to feature number 54, X.Ruminococcus..torques</span></span>
<span><span class="co">##  2024-04-26 15:42:19.377035 INFO::Fitting model to feature number 55, Bacteroides.salyersiae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.379162 INFO::Fitting model to feature number 56, Sellimonas.intestinalis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.381314 INFO::Fitting model to feature number 57, Bacteroides.faecis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.383575 INFO::Fitting model to feature number 58, Roseburia.intestinalis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.38575 INFO::Fitting model to feature number 59, Collinsella.aerofaciens</span></span>
<span><span class="co">##  2024-04-26 15:42:19.387912 INFO::Fitting model to feature number 60, Bacteroides.intestinalis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.39008 INFO::Fitting model to feature number 61, Bacteroides.faecis.CAG.32</span></span>
<span><span class="co">##  2024-04-26 15:42:19.392217 INFO::Fitting model to feature number 62, X.Ruminococcus..lactaris</span></span>
<span><span class="co">##  2024-04-26 15:42:19.394387 INFO::Fitting model to feature number 63, Prevotella.copri</span></span>
<span><span class="co">##  2024-04-26 15:42:19.396545 INFO::Fitting model to feature number 64, Eisenbergiella.tayi</span></span>
<span><span class="co">##  2024-04-26 15:42:19.398757 INFO::Fitting model to feature number 65, Klebsiella.pneumoniae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.400956 INFO::Fitting model to feature number 66, Intestinibacter.bartlettii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.403095 INFO::Fitting model to feature number 67, Erysipelatoclostridium.ramosum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.405269 INFO::Fitting model to feature number 68, Barnesiella.intestinihominis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.407415 INFO::Fitting model to feature number 69, Blautia.obeum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.409568 INFO::Fitting model to feature number 70, Bacteroides.finegoldii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.411697 INFO::Fitting model to feature number 71, Parasutterella.excrementihominis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.413883 INFO::Fitting model to feature number 72, Phascolarctobacterium.faecium</span></span>
<span><span class="co">##  2024-04-26 15:42:19.416033 INFO::Fitting model to feature number 73, Lachnospira.pectinoschiza</span></span>
<span><span class="co">##  2024-04-26 15:42:19.418217 INFO::Fitting model to feature number 74, Clostridium.sp..CAG.58</span></span>
<span><span class="co">##  2024-04-26 15:42:19.420363 INFO::Fitting model to feature number 75, Bilophila.wadsworthia</span></span>
<span><span class="co">##  2024-04-26 15:42:19.422583 INFO::Fitting model to feature number 76, Turicimonas.muris</span></span>
<span><span class="co">##  2024-04-26 15:42:19.424797 INFO::Fitting model to feature number 77, Proteobacteria.bacterium.CAG.139</span></span>
<span><span class="co">##  2024-04-26 15:42:19.426949 INFO::Fitting model to feature number 78, Anaeromassilibacillus.sp..An250</span></span>
<span><span class="co">##  2024-04-26 15:42:19.429094 INFO::Fitting model to feature number 79, Hungatella.hathewayi</span></span>
<span><span class="co">##  2024-04-26 15:42:19.431283 INFO::Fitting model to feature number 80, Alistipes.finegoldii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.433509 INFO::Fitting model to feature number 81, Ruminococcus.bromii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.435681 INFO::Fitting model to feature number 82, Odoribacter.splanchnicus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.437879 INFO::Fitting model to feature number 83, Butyricimonas.virosa</span></span>
<span><span class="co">##  2024-04-26 15:42:19.440024 INFO::Fitting model to feature number 84, Oscillibacter.sp..CAG.241</span></span>
<span><span class="co">##  2024-04-26 15:42:19.442231 INFO::Fitting model to feature number 85, Alistipes.indistinctus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.444396 INFO::Fitting model to feature number 86, Coprococcus.comes</span></span>
<span><span class="co">##  2024-04-26 15:42:19.446597 INFO::Fitting model to feature number 87, Gemmiger.formicilis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.448794 INFO::Fitting model to feature number 88, Holdemania.filiformis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.450975 INFO::Fitting model to feature number 89, Firmicutes.bacterium.CAG.83</span></span>
<span><span class="co">##  2024-04-26 15:42:19.453164 INFO::Fitting model to feature number 90, Dorea.formicigenerans</span></span>
<span><span class="co">##  2024-04-26 15:42:19.455311 INFO::Fitting model to feature number 91, Collinsella.intestinalis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.457514 INFO::Fitting model to feature number 92, Oscillibacter.sp..57_20</span></span>
<span><span class="co">##  2024-04-26 15:42:19.459681 INFO::Fitting model to feature number 93, Firmicutes.bacterium.CAG.94</span></span>
<span><span class="co">##  2024-04-26 15:42:19.461878 INFO::Fitting model to feature number 94, Dielma.fastidiosa</span></span>
<span><span class="co">##  2024-04-26 15:42:19.464031 INFO::Fitting model to feature number 95, Roseburia.sp..CAG.471</span></span>
<span><span class="co">##  2024-04-26 15:42:19.466261 INFO::Fitting model to feature number 96, X.Clostridium..innocuum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.468399 INFO::Fitting model to feature number 97, Haemophilus.parainfluenzae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.470573 INFO::Fitting model to feature number 98, Veillonella.dispar</span></span>
<span><span class="co">##  2024-04-26 15:42:19.472758 INFO::Fitting model to feature number 99, Veillonella.parvula</span></span>
<span><span class="co">##  2024-04-26 15:42:19.474932 INFO::Fitting model to feature number 100, Veillonella.infantium</span></span>
<span><span class="co">##  2024-04-26 15:42:19.477107 INFO::Fitting model to feature number 101, Streptococcus.salivarius</span></span>
<span><span class="co">##  2024-04-26 15:42:19.479239 INFO::Fitting model to feature number 102, Enterocloster.aldenensis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.481458 INFO::Fitting model to feature number 103, Veillonella.atypica</span></span>
<span><span class="co">##  2024-04-26 15:42:19.483606 INFO::Fitting model to feature number 104, Phocaeicola.plebeius</span></span>
<span><span class="co">##  2024-04-26 15:42:19.485782 INFO::Fitting model to feature number 105, Paraprevotella.xylaniphila</span></span>
<span><span class="co">##  2024-04-26 15:42:19.48795 INFO::Fitting model to feature number 106, Bacteroides.eggerthii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.49018 INFO::Fitting model to feature number 107, Alistipes.shahii</span></span>
<span><span class="co">##  2024-04-26 15:42:19.492333 INFO::Fitting model to feature number 108, Butyricimonas.synergistica</span></span>
<span><span class="co">##  2024-04-26 15:42:19.494537 INFO::Fitting model to feature number 109, Haemophilus.sp..HMSC71H05</span></span>
<span><span class="co">##  2024-04-26 15:42:19.496729 INFO::Fitting model to feature number 110, Eubacterium.ramulus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.498899 INFO::Fitting model to feature number 111, Coprobacter.fastidiosus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.501078 INFO::Fitting model to feature number 112, Lactobacillus.rogosae</span></span>
<span><span class="co">##  2024-04-26 15:42:19.50335 INFO::Fitting model to feature number 113, Phocaeicola.massiliensis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.506095 INFO::Fitting model to feature number 114, Streptococcus.parasanguinis</span></span>
<span><span class="co">##  2024-04-26 15:42:19.508753 INFO::Fitting model to feature number 115, Ruminococcaceae.bacterium.D16</span></span>
<span><span class="co">##  2024-04-26 15:42:19.510998 INFO::Fitting model to feature number 116, Veillonella.sp..T11011.6</span></span>
<span><span class="co">##  2024-04-26 15:42:19.513233 INFO::Fitting model to feature number 117, Klebsiella.variicola</span></span>
<span><span class="co">##  2024-04-26 15:42:19.515407 INFO::Fitting model to feature number 118, Bacteroides.galacturonicus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.517602 INFO::Fitting model to feature number 119, Clostridium.sp..CAG.299</span></span>
<span><span class="co">##  2024-04-26 15:42:19.519759 INFO::Fitting model to feature number 120, Eubacterium.ventriosum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.521944 INFO::Fitting model to feature number 121, Bifidobacterium.bifidum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.524097 INFO::Fitting model to feature number 122, Bifidobacterium.pseudocatenulatum</span></span>
<span><span class="co">##  2024-04-26 15:42:19.526297 INFO::Fitting model to feature number 123, Coprococcus.eutactus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.528462 INFO::Fitting model to feature number 124, Eubacterium.sp..CAG.251</span></span>
<span><span class="co">##  2024-04-26 15:42:19.530668 INFO::Fitting model to feature number 125, Phocaeicola.coprocola</span></span>
<span><span class="co">##  2024-04-26 15:42:19.532842 INFO::Fitting model to feature number 126, Paraprevotella.clara</span></span>
<span><span class="co">##  2024-04-26 15:42:19.535021 INFO::Fitting model to feature number 127, Coprococcus.catus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.554808 INFO::Counting total values for each feature</span></span>
<span><span class="co">##  2024-04-26 15:42:19.566056 INFO::Writing filtered data to file output_species/features/filtered_data.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:19.578534 INFO::Writing filtered, normalized data to file output_species/features/filtered_data_norm.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:19.590648 INFO::Writing filtered, normalized, transformed data to file output_species/features/filtered_data_norm_transformed.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:19.608006 INFO::Writing residuals to file output_species/fits/residuals.rds</span></span>
<span><span class="co">##  2024-04-26 15:42:19.614069 INFO::Writing fitted values to file output_species/fits/fitted.rds</span></span>
<span><span class="co">##  2024-04-26 15:42:19.619381 INFO::Writing all results to file (ordered by increasing q-values): output_species/all_results.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:19.622227 INFO::Writing the significant results (those which are less than or equal to the threshold of 0.250000 ) to file (ordered by increasing q-values): output_species/significant_results.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:19.623025 INFO::Writing heatmap of significant results to file: output_species/heatmap.pdf</span></span>
<span><span class="co">##  2024-04-26 15:42:19.727497 INFO::Writing association plots (one for each significant association) to output folder: output_species</span></span>
<span><span class="co">##  2024-04-26 15:42:19.730246 INFO::Plotting associations from most to least significant, grouped by metadata</span></span>
<span><span class="co">##  2024-04-26 15:42:19.730787 INFO::Plotting data for metadata number 1, age</span></span>
<span><span class="co">##  2024-04-26 15:42:19.731724 INFO::Creating scatter plot for continuous data, age vs Alistipes.indistinctus</span></span>
<span><span class="co">##  2024-04-26 15:42:19.905355 INFO::Creating scatter plot for continuous data, age vs Ruminococcus.bicirculans</span></span>
<span><span class="co">##  2024-04-26 15:42:20.110852 INFO::Creating scatter plot for continuous data, age vs Lacrimispora.saccharolytica</span></span>
<span><span class="co">##  2024-04-26 15:42:20.27071 INFO::Creating scatter plot for continuous data, age vs Ruminococcus.bromii</span></span>
<span><span class="co">##  2024-04-26 15:42:20.417793 INFO::Creating scatter plot for continuous data, age vs Bifidobacterium.pseudocatenulatum</span></span>
<span><span class="co">##  2024-04-26 15:42:20.538164 INFO::Creating scatter plot for continuous data, age vs X.Ruminococcus..gnavus</span></span>
<span><span class="co">##  2024-04-26 15:42:20.659052 INFO::Creating scatter plot for continuous data, age vs Lachnospira.eligens</span></span>
<span><span class="co">##  2024-04-26 15:42:20.77854 INFO::Creating scatter plot for continuous data, age vs Haemophilus.parainfluenzae</span></span>
<span><span class="co">##  2024-04-26 15:42:20.906067 INFO::Creating scatter plot for continuous data, age vs Butyricimonas.virosa</span></span>
<span><span class="co">##  2024-04-26 15:42:21.026894 INFO::Creating scatter plot for continuous data, age vs Lawsonibacter.asaccharolyticus</span></span>
<span><span class="co">##  2024-04-26 15:42:21.161532 INFO::Creating scatter plot for continuous data, age vs X.Eubacterium..siraeum</span></span>
<span><span class="co">##  2024-04-26 15:42:21.279617 INFO::Creating scatter plot for continuous data, age vs Butyricimonas.synergistica</span></span>
<span><span class="co">##  2024-04-26 15:42:21.398479 INFO::Creating scatter plot for continuous data, age vs Monoglobus.pectinilyticus</span></span>
<span><span class="co">##  2024-04-26 15:42:21.517264 INFO::Creating scatter plot for continuous data, age vs Veillonella.atypica</span></span>
<span><span class="co">##  2024-04-26 15:42:23.098467 INFO::Plotting data for metadata number 2, disease</span></span>
<span><span class="co">##  2024-04-26 15:42:23.099674 INFO::Creating boxplot for categorical data, disease vs Alistipes.putredinis</span></span>
<span><span class="co">##  2024-04-26 15:42:23.24581 INFO::Creating boxplot for categorical data, disease vs Gemmiger.formicilis</span></span>
<span><span class="co">##  2024-04-26 15:42:23.367666 INFO::Creating boxplot for categorical data, disease vs X.Ruminococcus..torques</span></span>
<span><span class="co">##  2024-04-26 15:42:23.477417 INFO::Creating boxplot for categorical data, disease vs Ruminococcus.bicirculans</span></span>
<span><span class="co">##  2024-04-26 15:42:23.588362 INFO::Creating boxplot for categorical data, disease vs Sellimonas.intestinalis</span></span>
<span><span class="co">##  2024-04-26 15:42:23.699883 INFO::Creating boxplot for categorical data, disease vs X.Clostridium..leptum</span></span>
<span><span class="co">##  2024-04-26 15:42:23.810284 INFO::Creating boxplot for categorical data, disease vs Alistipes.shahii</span></span>
<span><span class="co">##  2024-04-26 15:42:24.859931 INFO::Plotting data for metadata number 3, antibiotics_current_use</span></span>
<span><span class="co">##  2024-04-26 15:42:24.861131 INFO::Creating boxplot for categorical data, antibiotics_current_use vs Coprobacter.fastidiosus</span></span>
<span><span class="co">##  2024-04-26 15:42:24.970189 INFO::Creating boxplot for categorical data, antibiotics_current_use vs X.Eubacterium..rectale</span></span>
<span><span class="co">##  2024-04-26 15:42:25.097372 INFO::Creating boxplot for categorical data, antibiotics_current_use vs Agathobaculum.butyriciproducens</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Unlike gene expression studies, we do not have well-defined signatures or modules for microbiome data. Here, we will construct data-driven modules using weighted gene co-expression network analysis (WGCNA) <span class="citation" data-cites="Langfelder2008">L (<a href="#ref-Geistlinger2023" role="doc-biblioref">2023</a>)</span>. We aim to ensure that the effect of disease and other covariates has been removed by working on the residuals. Following the WGCNA tutorial, our first step will be to check whether there are any outliers in our data.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Geistlinger2023" class="csl-entry" role="listitem">
L, Geistlinger. 2023. <span>“BugSigDB Captures Patterns of Differential Abundance Across a Broad Range of Host-Associated Microbial Signatures.”</span> <em>Nature Biotechnology</em>. <a href="https://doi.org/10.1038/s41587-023-01872-y">https://doi.org/10.1038/s41587-023-01872-y</a>.
</div></div><div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/extract-residuals_a126dd5a31d2aa9770af6a49bbe6a79e">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/">WGCNA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">datExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">fit_data</span><span class="op">$</span><span class="va">residuals</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gsg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/goodSamplesGenes.html">goodSamplesGenes</a></span><span class="op">(</span><span class="va">datExpr</span>, verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">##   Flagging genes and samples with too many missing values...</span></span>
<span><span class="co">##    ..step 1</span></span>
<span><span class="co">##    ..Excluding 6 samples from the calculation due to too many missing genes.</span></span>
<span><span class="co">##    ..step 2</span></span>
<span><span class="va">gsg</span><span class="op">$</span><span class="va">allOK</span></span>
<span><span class="co">##  [1] FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If the last statement returns TRUE, no outliers are identified. If not, we need to remove the outliers from the data.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/check-outliers1_ad24b9159ff00580c5ba4362969f57a0">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">allOK</span><span class="op">)</span></span>
<span><span class="op">{</span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodGenes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/dynamicTreeCut/man/printFlush.html">printFlush</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Removing genes:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">datExpr</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodGenes</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>;</span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodSamples</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dynamicTreeCut/man/printFlush.html">printFlush</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Removing samples:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">datExpr</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodSamples</span><span class="op">]</span>, collapse <span class="op">=</span><span class="st">", "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">datExpr</span> <span class="op">=</span> <span class="va">datExpr</span><span class="op">[</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodSamples</span>, <span class="va">gsg</span><span class="op">$</span><span class="va">goodGenes</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="co">##  Removing samples: CSM5MCVB_P, CSM79HNY_P, ESM5GEYY_P, ESM718U9_P, MSM6J2N6_P, MSM9VZLX_P</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After removing the outliers, we need to choose a suitable soft threshold parameter for creating the modules as part of the WGCNA algorithm. This power value must produce a graph similar to a scale-free network. We can use the mean connectivity graphic for the selection of this power parameter.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/choose-soft-threshold_485e61847e15b7c952dcada2541cb718">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Choose a set of soft threshold parameters</span></span>
<span><span class="va">powers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">22</span>, to<span class="op">=</span><span class="fl">30</span>, by<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sft</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/pickSoftThreshold.html">pickSoftThreshold</a></span><span class="op">(</span><span class="va">datExpr</span>, powerVector <span class="op">=</span> <span class="va">powers</span>, verbose <span class="op">=</span> <span class="fl">5</span>, dataIsExpr <span class="op">=</span> <span class="cn">TRUE</span>, RsquaredCut <span class="op">=</span> <span class="fl">0.30</span><span class="op">)</span> </span>
<span><span class="co">##  pickSoftThreshold: will use block size 127.</span></span>
<span><span class="co">##   pickSoftThreshold: calculating connectivity for given powers...</span></span>
<span><span class="co">##     ..working on genes 1 through 127 of 127</span></span>
<span><span class="co">##     Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.</span></span>
<span><span class="co">##  1      1   0.1790 -0.972        0.85700 15.00000  1.48e+01 23.0000</span></span>
<span><span class="co">##  2      2   0.3450 -0.729        0.63500  2.95000  2.85e+00  5.6900</span></span>
<span><span class="co">##  3      3   0.2400 -4.130        0.12700  0.80400  7.36e-01  2.6100</span></span>
<span><span class="co">##  4      4   0.3100 -4.870        0.19900  0.29200  2.14e-01  1.6700</span></span>
<span><span class="co">##  5      5   0.2150 -4.030       -0.00880  0.13700  7.02e-02  1.2200</span></span>
<span><span class="co">##  6      6   0.1800 -2.930       -0.05130  0.07910  2.46e-02  0.9570</span></span>
<span><span class="co">##  7      7   0.1380 -2.240       -0.10100  0.05320  9.33e-03  0.7900</span></span>
<span><span class="co">##  8      8   0.2200 -2.790       -0.00259  0.03940  3.71e-03  0.6690</span></span>
<span><span class="co">##  9      9   0.0872 -1.450       -0.09130  0.03100  1.50e-03  0.5760</span></span>
<span><span class="co">##  10    10   0.1190 -1.630       -0.07660  0.02540  6.73e-04  0.5020</span></span>
<span><span class="co">##  11    11   0.0562 -1.250       -0.03470  0.02130  2.66e-04  0.4410</span></span>
<span><span class="co">##  12    12   0.1620 -2.150       -0.07720  0.01810  1.12e-04  0.3910</span></span>
<span><span class="co">##  13    13   0.2120 -2.350       -0.00949  0.01560  4.84e-05  0.3480</span></span>
<span><span class="co">##  14    14   0.2480 -2.410        0.04950  0.01360  2.12e-05  0.3110</span></span>
<span><span class="co">##  15    15   0.2890 -2.470        0.15200  0.01190  9.53e-06  0.2810</span></span>
<span><span class="co">##  16    16   0.3030 -2.420        0.18400  0.01050  4.33e-06  0.2580</span></span>
<span><span class="co">##  17    17   0.2410 -2.380        0.02450  0.00933  1.98e-06  0.2380</span></span>
<span><span class="co">##  18    18   0.3190 -2.640        0.15400  0.00830  9.03e-07  0.2180</span></span>
<span><span class="co">##  19    19   0.3320 -2.600        0.17700  0.00741  4.03e-07  0.2010</span></span>
<span><span class="co">##  20    20   0.2370 -2.050        0.02180  0.00664  1.70e-07  0.1840</span></span>
<span><span class="co">##  21    22   0.2350 -2.000        0.07520  0.00536  3.40e-08  0.1560</span></span>
<span><span class="co">##  22    24   0.2430 -2.230        0.09560  0.00437  6.98e-09  0.1310</span></span>
<span><span class="co">##  23    26   0.2370 -2.120        0.26200  0.00358  1.44e-09  0.1110</span></span>
<span><span class="co">##  24    28   0.2490 -2.080        0.26200  0.00294  2.93e-10  0.0937</span></span>
<span><span class="co">##  25    30   0.2490 -2.290        0.30100  0.00243  5.67e-11  0.0791</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this step, we will conduct a one-step module detection based on the selected soft threshold parameter selected above.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/module-detection_cd14cb191311a9199893054f25e3fa3a">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">power</span> <span class="op">=</span> <span class="va">sft</span><span class="op">$</span><span class="va">powerEstimate</span></span>
<span><span class="va">net</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html">blockwiseModules</a></span><span class="op">(</span><span class="va">datExpr</span>, </span>
<span>                       power <span class="op">=</span> <span class="va">power</span>,</span>
<span>                       corFnc<span class="op">=</span><span class="st">"bicor"</span>, </span>
<span>                       corOptions<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>maxPOutliers<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>, </span>
<span>                       networkType <span class="op">=</span><span class="st">"unsigned"</span>, </span>
<span>                       maxBlockSize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">datExpr</span><span class="op">)</span>,</span>
<span>                       minModuleSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       TOMType <span class="op">=</span> <span class="st">"unsigned"</span>, </span>
<span>                       reassignThreshold <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                       mergeCutHeight <span class="op">=</span> <span class="fl">0</span>,                       </span>
<span>                       verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">##   Calculating module eigengenes block-wise from all genes</span></span>
<span><span class="co">##     Flagging genes and samples with too many missing values...</span></span>
<span><span class="co">##      ..step 1</span></span>
<span><span class="co">##   ..Working on block 1 .</span></span>
<span><span class="co">##      TOM calculation: adjacency..</span></span>
<span><span class="co">##      ..will not use multithreading.</span></span>
<span><span class="co">##       Fraction of slow calculations: 0.000000</span></span>
<span><span class="co">##      ..connectivity..</span></span>
<span><span class="co">##      ..matrix multiplication (system BLAS)..</span></span>
<span><span class="co">##      ..normalization..</span></span>
<span><span class="co">##      ..done.</span></span>
<span><span class="co">##   ....clustering..</span></span>
<span><span class="co">##   ....detecting modules..</span></span>
<span><span class="co">##   ....calculating module eigengenes..</span></span>
<span><span class="co">##   ....checking kME in modules..</span></span>
<span><span class="co">##       ..removing 3 genes from module 1 because their KME is too low.</span></span>
<span><span class="co">##       ..removing 3 genes from module 2 because their KME is too low.</span></span>
<span><span class="co">##       ..removing 2 genes from module 3 because their KME is too low.</span></span>
<span><span class="co">##       ..removing 1 genes from module 5 because their KME is too low.</span></span>
<span><span class="co">##       ..removing 1 genes from module 7 because their KME is too low.</span></span>
<span><span class="co">##   ..merging modules that are too close..</span></span>
<span><span class="co">##       mergeCloseModules: Merging modules whose distance is less than 0</span></span>
<span><span class="co">##         Calculating new MEs...</span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># How many modules #</span></span>
<span><span class="co">####################</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">MEs</span><span class="op">)</span></span>
<span><span class="co">##  [1] 14</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##        black        blue       brown       green greenyellow        grey </span></span>
<span><span class="co">##            7          14          13          12           5          10 </span></span>
<span><span class="co">##      magenta        pink      purple         red      salmon         tan </span></span>
<span><span class="co">##            6           6           5          11           3           5 </span></span>
<span><span class="co">##    turquoise      yellow </span></span>
<span><span class="co">##           18          12</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The WGCNA algorithm produced 14 modules which we can visualize as follows.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/visualization_9ef548ff7eaeff5bc5b1e6f1c1f074d4">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">##########################</span></span>
<span><span class="co"># Plot module dendrogram #</span></span>
<span><span class="co">##########################</span></span>
<span></span>
<span><span class="va">eigenGenes</span> <span class="op">&lt;-</span> <span class="va">net</span><span class="op">$</span><span class="va">MEs</span></span>
<span><span class="va">MEDiss</span> <span class="op">=</span> <span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/cor.html">cor</a></span><span class="op">(</span><span class="va">eigenGenes</span><span class="op">)</span></span>
<span><span class="va">METree</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/fastcluster/man/hclust.html">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="va">MEDiss</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">METree</span>, main <span class="op">=</span> <span class="st">"Clustering of module eigengenes"</span>, xlab <span class="op">=</span> <span class="st">""</span>, sub <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MSEA_files/figure-html/visualization-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next, we calculate hub genes for the modules and create the mapping files to proceed with the MSEA.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/create-mapping-files_3355f1378a2db022493c3400b274a965">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">###########################################</span></span>
<span><span class="co"># Re-calculate modules and find hub genes #</span></span>
<span><span class="co">###########################################</span></span>
<span></span>
<span><span class="va">moduleColors</span> <span class="op">&lt;-</span> <span class="va">net</span><span class="op">$</span><span class="va">colors</span></span>
<span><span class="va">MEs0</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/moduleEigengenes.html">moduleEigengenes</a></span><span class="op">(</span><span class="va">datExpr</span>, <span class="va">moduleColors</span><span class="op">)</span><span class="op">$</span><span class="va">eigengenes</span></span>
<span><span class="va">modules_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/orderMEs.html">orderMEs</a></span><span class="op">(</span><span class="va">MEs0</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#######################</span></span>
<span><span class="co"># Create mapping file #</span></span>
<span><span class="co">#######################</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_by_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span>
<span><span class="va">feature_by_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">feature_by_modules</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">feature_by_modules</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Feature'</span>, <span class="st">'Module'</span><span class="op">)</span></span>
<span><span class="va">features_mapping</span> <span class="op">&lt;-</span> <span class="va">feature_by_modules</span></span>
<span><span class="va">features_mapping</span><span class="op">$</span><span class="va">Module</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">'ME'</span>, <span class="va">features_mapping</span><span class="op">$</span><span class="va">Module</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we will run the MSEA analysis on the modules we constructed using WGCNA. Here, we first create a wrapper for the MSEA analysis using the <code>gsEasy</code> package.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/MSEA-wrapper_502588272a2505e9b09a82e2decd4aa3">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape">reshape2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gsEasy</span><span class="op">)</span></span>
<span></span>
<span><span class="co">################</span></span>
<span><span class="co"># MSEA Wrapper #</span></span>
<span><span class="co">################</span></span>
<span></span>
<span><span class="va">run_MSEA</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">microbeSet</span>, <span class="co"># A list</span></span>
<span>                     <span class="va">ranked_features</span>, <span class="co"># Ranked list of featured</span></span>
<span>                     <span class="va">filter.count</span> <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                     <span class="va">seed</span> <span class="op">=</span> <span class="fl">1234</span>,</span>
<span>                     <span class="va">fdr.correction</span> <span class="op">=</span> <span class="st">'BH'</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span></span>
<span>  <span class="co">###################</span></span>
<span>  <span class="co"># Filter out sets #</span></span>
<span>  <span class="co">##################</span></span>
<span>  </span>
<span>  <span class="va">microbeSet0</span> <span class="op">&lt;-</span> <span class="va">microbeSet</span></span>
<span>  <span class="va">cond</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">microbeSet0</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">filter.count</span><span class="op">)</span></span>
<span>  <span class="va">microbeSet</span> <span class="op">&lt;-</span> <span class="va">microbeSet0</span><span class="op">[</span><span class="va">cond</span><span class="op">]</span></span>
<span>  <span class="va">lengthmicrobeSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">microbeSet</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">lengthmicrobeSet</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Freq'</span>,<span class="st">'Set'</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">################</span></span>
<span>  <span class="co"># Classic MSEA #</span></span>
<span>  <span class="co">################</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span></span>
<span>  <span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">microbeSet</span>, <span class="kw">function</span><span class="op">(</span><span class="va">set</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/gsEasy/man/gset.html">gset</a></span><span class="op">(</span>S <span class="op">=</span> <span class="va">set</span>, r <span class="op">=</span> <span class="va">ranked_features</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">enrichment</span><span class="op">)</span><span class="op">&lt;-</span><span class="st">'ES'</span></span>
<span>  <span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">enrichment</span>, <span class="st">'Set'</span><span class="op">)</span></span>
<span>  <span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">enrichment</span>, <span class="va">lengthmicrobeSet</span>, <span class="st">'Set'</span><span class="op">)</span></span>
<span>  <span class="va">enrichment</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">enrichment</span><span class="op">$</span><span class="va">ES</span>, <span class="va">fdr.correction</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">##########</span></span>
<span>  <span class="co"># Return #</span></span>
<span>  <span class="co">##########</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">enrichment</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before running the MSEA, we also need to rank the differential analysis results from <code>MaAsLin2</code>. We use the <code>topGo</code> package to create a list of microbe sets from the mapping file created above.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/MSEA-processing_b22da03c3dfa5ef08953d99210515ac0">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">###################</span></span>
<span><span class="co"># Rank DA results #</span></span>
<span><span class="co">###################</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="va">fit_data</span><span class="op">$</span><span class="va">results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">metadata</span> <span class="op">==</span> <span class="st">'disease'</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">pval</span>, <span class="st">'BH'</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html">order</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">qval</span>, decreasing <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">###################</span></span>
<span><span class="co"># MSEA Processing #</span></span>
<span><span class="co">###################</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">topGO</span><span class="op">)</span></span>
<span><span class="va">module_map</span> <span class="op">&lt;-</span> <span class="va">features_mapping</span></span>
<span><span class="va">mod.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/tapply.html">tapply</a></span><span class="op">(</span><span class="va">module_map</span><span class="op">$</span><span class="va">Module</span>, <span class="va">module_map</span><span class="op">$</span><span class="va">Feature</span>, <span class="va">as.character</span><span class="op">)</span></span>
<span><span class="va">microbeSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/topGO/man/annFUN.html">inverseList</a></span><span class="op">(</span><span class="va">mod.gs</span><span class="op">)</span></span>
<span><span class="va">microbeSet</span></span>
<span><span class="co">##  $MEblack</span></span>
<span><span class="co">##  [1] "Anaeromassilibacillus.sp..An250" "Anaerotruncus.colihominis"      </span></span>
<span><span class="co">##  [3] "Blautia.wexlerae"                "Eisenbergiella.tayi"            </span></span>
<span><span class="co">##  [5] "Firmicutes.bacterium.CAG.94"     "Ruthenibacterium.lactatiformans"</span></span>
<span><span class="co">##  [7] "Sellimonas.intestinalis"        </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEblue</span></span>
<span><span class="co">##   [1] "Clostridium.sp..CAG.58"         "Erysipelatoclostridium.ramosum"</span></span>
<span><span class="co">##   [3] "Haemophilus.parainfluenzae"     "Intestinibacter.bartlettii"    </span></span>
<span><span class="co">##   [5] "Klebsiella.pneumoniae"          "Klebsiella.variicola"          </span></span>
<span><span class="co">##   [7] "Lawsonibacter.asaccharolyticus" "Streptococcus.parasanguinis"   </span></span>
<span><span class="co">##   [9] "Streptococcus.salivarius"       "Veillonella.atypica"           </span></span>
<span><span class="co">##  [11] "Veillonella.dispar"             "Veillonella.infantium"         </span></span>
<span><span class="co">##  [13] "Veillonella.parvula"            "Veillonella.sp..T11011.6"      </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEbrown</span></span>
<span><span class="co">##   [1] "Alistipes.finegoldii"        "Alistipes.indistinctus"     </span></span>
<span><span class="co">##   [3] "Alistipes.putredinis"        "Alistipes.shahii"           </span></span>
<span><span class="co">##   [5] "Bacteroides.xylanisolvens"   "Bilophila.wadsworthia"      </span></span>
<span><span class="co">##   [7] "Firmicutes.bacterium.CAG.83" "Odoribacter.splanchnicus"   </span></span>
<span><span class="co">##   [9] "Oscillibacter.sp..57_20"     "Oscillibacter.sp..CAG.241"  </span></span>
<span><span class="co">##  [11] "Phocaeicola.dorei"           "Ruminococcus.bromii"        </span></span>
<span><span class="co">##  [13] "X.Eubacterium..siraeum"     </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEgreen</span></span>
<span><span class="co">##   [1] "Blautia.sp..CAG.257"            "Clostridium.bolteae.CAG.59"    </span></span>
<span><span class="co">##   [3] "Eggerthella.lenta"              "Eisenbergiella.massiliensis"   </span></span>
<span><span class="co">##   [5] "Enterocloster.aldenensis"       "Enterocloster.bolteae"         </span></span>
<span><span class="co">##   [7] "Enterocloster.citroniae"        "Enterocloster.clostridioformis"</span></span>
<span><span class="co">##   [9] "Enterocloster.lavalensis"       "Flavonifractor.plautii"        </span></span>
<span><span class="co">##  [11] "Lacrimispora.saccharolytica"    "X.Clostridium..symbiosum"      </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEgreenyellow</span></span>
<span><span class="co">##  [1] "Coprobacter.fastidiosus" "Escherichia.coli"       </span></span>
<span><span class="co">##  [3] "Hungatella.hathewayi"    "X.Clostridium..innocuum"</span></span>
<span><span class="co">##  [5] "X.Ruminococcus..gnavus" </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEgrey</span></span>
<span><span class="co">##   [1] "Anaerotignum.lactatifermentans"   "Bacteroides.cellulosilyticus"    </span></span>
<span><span class="co">##   [3] "Bacteroides.fragilis"             "Collinsella.intestinalis"        </span></span>
<span><span class="co">##   [5] "Eubacterium.sp..CAG.38"           "Haemophilus.sp..HMSC71H05"       </span></span>
<span><span class="co">##   [7] "Intestinimonas.butyriciproducens" "Lachnospira.eligens"             </span></span>
<span><span class="co">##   [9] "Roseburia.intestinalis"           "Roseburia.sp..CAG.471"           </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEmagenta</span></span>
<span><span class="co">##  [1] "Bacteroides.finegoldii"     "Paraprevotella.clara"      </span></span>
<span><span class="co">##  [3] "Paraprevotella.xylaniphila" "Phocaeicola.coprocola"     </span></span>
<span><span class="co">##  [5] "Phocaeicola.plebeius"       "Prevotella.copri"          </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEpink</span></span>
<span><span class="co">##  [1] "Bacteroides.galacturonicus"        "Bifidobacterium.pseudocatenulatum"</span></span>
<span><span class="co">##  [3] "Eubacterium.sp..CAG.251"           "Lachnospira.pectinoschiza"        </span></span>
<span><span class="co">##  [5] "Lactobacillus.rogosae"             "Phocaeicola.massiliensis"         </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEpurple</span></span>
<span><span class="co">##  [1] "Barnesiella.intestinihominis" "Butyricimonas.synergistica"  </span></span>
<span><span class="co">##  [3] "Butyricimonas.virosa"         "Coprococcus.eutactus"        </span></span>
<span><span class="co">##  [5] "Ruminococcus.bicirculans"    </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEred</span></span>
<span><span class="co">##   [1] "Akkermansia.muciniphila"          "Bacteroides.intestinalis"        </span></span>
<span><span class="co">##   [3] "Clostridium.sp..CAG.299"          "Dialister.invisus"               </span></span>
<span><span class="co">##   [5] "Dielma.fastidiosa"                "Holdemania.filiformis"           </span></span>
<span><span class="co">##   [7] "Monoglobus.pectinilyticus"        "Parasutterella.excrementihominis"</span></span>
<span><span class="co">##   [9] "Proteobacteria.bacterium.CAG.139" "Turicimonas.muris"               </span></span>
<span><span class="co">##  [11] "X.Clostridium..leptum"           </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEsalmon</span></span>
<span><span class="co">##  [1] "Bacteroides.faecis"            "Bacteroides.faecis.CAG.32"    </span></span>
<span><span class="co">##  [3] "Phascolarctobacterium.faecium"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEtan</span></span>
<span><span class="co">##  [1] "Bacteroides.salyersiae"       "Bifidobacterium.adolescentis"</span></span>
<span><span class="co">##  [3] "Bifidobacterium.bifidum"      "Bifidobacterium.longum"      </span></span>
<span><span class="co">##  [5] "Collinsella.aerofaciens"     </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEturquoise</span></span>
<span><span class="co">##   [1] "Agathobaculum.butyriciproducens" "Anaerobutyricum.hallii"         </span></span>
<span><span class="co">##   [3] "Anaerostipes.hadrus"             "Bacteroides.eggerthii"          </span></span>
<span><span class="co">##   [5] "Blautia.obeum"                   "Coprococcus.catus"              </span></span>
<span><span class="co">##   [7] "Coprococcus.comes"               "Dorea.formicigenerans"          </span></span>
<span><span class="co">##   [9] "Dorea.longicatena"               "Eubacterium.ramulus"            </span></span>
<span><span class="co">##  [11] "Faecalibacterium.prausnitzii"    "Fusicatenibacter.saccharivorans"</span></span>
<span><span class="co">##  [13] "Gemmiger.formicilis"             "Roseburia.faecis"               </span></span>
<span><span class="co">##  [15] "Roseburia.hominis"               "Roseburia.inulinivorans"        </span></span>
<span><span class="co">##  [17] "X.Eubacterium..rectale"          "X.Ruminococcus..torques"        </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEyellow</span></span>
<span><span class="co">##   [1] "Acidaminococcus.intestini"     "Bacteroides.caccae"           </span></span>
<span><span class="co">##   [3] "Bacteroides.ovatus"            "Bacteroides.stercoris"        </span></span>
<span><span class="co">##   [5] "Bacteroides.thetaiotaomicron"  "Bacteroides.uniformis"        </span></span>
<span><span class="co">##   [7] "Eubacterium.ventriosum"        "Parabacteroides.distasonis"   </span></span>
<span><span class="co">##   [9] "Parabacteroides.merdae"        "Phocaeicola.vulgatus"         </span></span>
<span><span class="co">##  [11] "Ruminococcaceae.bacterium.D16" "X.Ruminococcus..lactaris"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are now ready to run the MSEA analysis. We run <span class="math inline">\(100,000\)</span> permutations to calculate the enrichment scores.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/calculate-enrichment-scores_af567075ca81e5547ec83d460e39fec8">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">MSEA</span> <span class="op">&lt;-</span> <span class="fu">run_MSEA</span><span class="op">(</span><span class="va">microbeSet</span>, <span class="va">results</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span></span>
<span><span class="va">MSEA</span> <span class="op">&lt;-</span> <span class="va">MSEA</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Set'</span>, <span class="st">'Freq'</span>, <span class="st">'ES'</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">MSEA</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Set'</span>, <span class="st">'Freq'</span>, <span class="st">'ES'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">MSEA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'ID'</span>, <span class="st">'Size'</span>, <span class="st">'pval'</span>, <span class="st">'qval'</span><span class="op">)</span></span>
<span><span class="va">MSEA</span><span class="op">$</span><span class="va">ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="va">MSEA</span><span class="op">$</span><span class="va">ID</span>, <span class="st">' ('</span>, <span class="va">MSEA</span><span class="op">$</span><span class="va">Size</span>, <span class="st">')'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can plot the enrichment scores to visualize the MSEA results.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/visualzation_c71e6a125473967b27e0ca01bc4cf3a1">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">MSEA</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">pval</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">ID</span>, levels <span class="op">=</span> <span class="va">ID</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span>, x <span class="op">=</span> <span class="va">ID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, fill <span class="op">=</span> <span class="st">'cornflowerblue'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Statistically significant modules associated with disease'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'MSEA enrichment score'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MSEA_files/figure-html/visualzation-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Based on the MSEA results, we obtain 13 enriched modules of microbial species. We can also examine the members of the top enriched modules.</p>
</section><section id="print-the-most-significant-modules-for-species-relative-abundance-data" class="level2" data-number="17.3"><h2 data-number="17.3" class="anchored" data-anchor-id="print-the-most-significant-modules-for-species-relative-abundance-data">
<span class="header-section-number">17.3</span> Print the most significant modules for species relative abundance data</h2>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/module-print_cf735322c166d55a83480495835f7afa">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">microbeSet</span><span class="op">[[</span><span class="st">"MEpurple"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  [1] "Barnesiella.intestinihominis" "Butyricimonas.synergistica"  </span></span>
<span><span class="co">##  [3] "Butyricimonas.virosa"         "Coprococcus.eutactus"        </span></span>
<span><span class="co">##  [5] "Ruminococcus.bicirculans"</span></span>
<span><span class="va">microbeSet</span><span class="op">[[</span><span class="st">"MEbrown"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##   [1] "Alistipes.finegoldii"        "Alistipes.indistinctus"     </span></span>
<span><span class="co">##   [3] "Alistipes.putredinis"        "Alistipes.shahii"           </span></span>
<span><span class="co">##   [5] "Bacteroides.xylanisolvens"   "Bilophila.wadsworthia"      </span></span>
<span><span class="co">##   [7] "Firmicutes.bacterium.CAG.83" "Odoribacter.splanchnicus"   </span></span>
<span><span class="co">##   [9] "Oscillibacter.sp..57_20"     "Oscillibacter.sp..CAG.241"  </span></span>
<span><span class="co">##  [11] "Phocaeicola.dorei"           "Ruminococcus.bromii"        </span></span>
<span><span class="co">##  [13] "X.Eubacterium..siraeum"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="performing-the-msea-analysis-with-pathway-relative-abundance-data" class="level2" data-number="17.4"><h2 data-number="17.4" class="anchored" data-anchor-id="performing-the-msea-analysis-with-pathway-relative-abundance-data">
<span class="header-section-number">17.4</span> Performing the MSEA analysis with pathway relative abundance data</h2>
<p>Next, we repeat the MSEA with the pathway relative abundance data from the iHMP project and follow the same steps as before.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/input-data_70a71d8a444a9368bb5a484b473ab362">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">##########################</span></span>
<span><span class="co"># Load HMP2 pathway data # </span></span>
<span><span class="co">##########################</span></span>
<span></span>
<span><span class="va">se_pathway</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"HMP_2019_ibdmdb"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/returnSamples.html">returnSamples</a></span><span class="op">(</span><span class="st">"pathway_abundance"</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##########################</span></span>
<span><span class="co"># Create sample metadata #</span></span>
<span><span class="co">##########################</span></span>
<span></span>
<span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_pathway</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">visit_number</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="st">"age"</span>, <span class="st">"disease"</span>, <span class="st">"antibiotics_current_use"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set reference</span></span>
<span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span><span class="op">)</span></span>
<span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">disease</span>, <span class="st">'healthy'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###########################</span></span>
<span><span class="co"># Create Pathway Features #</span></span>
<span><span class="co">###########################</span></span>
<span></span>
<span><span class="va">feature_pwys_t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_pathway</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">feature_pwys_t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">feature_pwys_t</span>, <span class="st">"ID"</span><span class="op">)</span></span>
<span><span class="va">feature_pwys_t</span> <span class="op">&lt;-</span> <span class="va">feature_pwys_t</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">"\\|"</span>, <span class="va">ID</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">ID</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'UNMAPPED'</span>, <span class="st">'UNINTEGRATED'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">'ID'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################</span></span>
<span><span class="co"># Subset to baseline samples #</span></span>
<span><span class="co">##############################</span></span>
<span></span>
<span><span class="va">feature_pwys</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html">t</a></span><span class="op">(</span><span class="va">feature_pwys_t</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">feature_pwys</span> <span class="op">&lt;-</span> <span class="va">feature_pwys</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">feature_pwys</span> <span class="op">&lt;-</span> <span class="va">feature_pwys</span> <span class="op">/</span> <span class="fl">100</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">feature_pwys_t</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">se_pathway</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As before, we first run a <code>MaAsLin2</code> analysis using default settings and construct the modules using residuals from the <code>MaAsLin2</code> models.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/MSEA-preparation_6c30e87ab71935ac88ae031b58a53af2">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Maaslin2/man/Maaslin2.html">Maaslin2</a></span><span class="op">(</span>input_data     <span class="op">=</span> <span class="va">feature_pwys</span>, </span>
<span>                    input_metadata <span class="op">=</span> <span class="va">sample_metadata</span>, </span>
<span>                    normalization  <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>                    output         <span class="op">=</span> <span class="st">"output_pwys"</span>, </span>
<span>                    fixed_effects  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"disease"</span>, <span class="st">"age"</span>, <span class="st">"antibiotics_current_use"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "Creating output folder"</span></span>
<span><span class="co">##  [1] "Creating output feature tables folder"</span></span>
<span><span class="co">##  [1] "Creating output fits folder"</span></span>
<span><span class="co">##  [1] "Creating output figures folder"</span></span>
<span><span class="co">##  2024-04-26 15:42:48.836878 INFO::Writing function arguments to log file</span></span>
<span><span class="co">##  2024-04-26 15:42:48.840605 INFO::Verifying options selected are valid</span></span>
<span><span class="co">##  2024-04-26 15:42:48.84109 INFO::Determining format of input files</span></span>
<span><span class="co">##  2024-04-26 15:42:48.841543 INFO::Input format is data samples as rows and metadata samples as rows</span></span>
<span><span class="co">##  2024-04-26 15:42:48.846767 INFO::Formula for fixed effects: expr ~  disease + age + antibiotics_current_use</span></span>
<span><span class="co">##  2024-04-26 15:42:48.847293 INFO::Factor detected for categorial metadata 'disease'. Provide a reference argument or manually set factor ordering to change reference level.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.847881 INFO::Filter data based on min abundance and min prevalence</span></span>
<span><span class="co">##  2024-04-26 15:42:48.848296 INFO::Total samples in data: 136</span></span>
<span><span class="co">##  2024-04-26 15:42:48.848728 INFO::Min samples required with min abundance for a feature not to be filtered: 13.600000</span></span>
<span><span class="co">##  2024-04-26 15:42:48.854263 INFO::Total filtered features: 113</span></span>
<span><span class="co">##  2024-04-26 15:42:48.854827 INFO::Filtered feature names from abundance and prevalence filtering: PWY.5044..purine.nucleotides.degradation.I..plants., PROPFERM.PWY..L.alanine.fermentation.to.propanoate.and.acetate, PWY.6596..adenosine.nucleotides.degradation.I, PWY.5004..superpathway.of.L.citrulline.metabolism, UDPNACETYLGALSYN.PWY..UDP.N.acetyl.D.glucosamine.biosynthesis.II, PWY66.367..ketogenesis, PWY.5392..reductive.TCA.cycle.II, PWY.101..photosynthesis.light.reactions, PWY.7031..protein.N.glycosylation..bacterial., PWY.5265..peptidoglycan.biosynthesis.II..staphylococci., PWY.7268..NAD.NADP.NADH.NADPH.cytosolic.interconversion..yeast., PWY.7165..L.ascorbate.biosynthesis.VI..engineered.pathway., CATECHOL.ORTHO.CLEAVAGE.PWY..catechol.degradation.to..beta..ketoadipate, PWY.5417..catechol.degradation.III..ortho.cleavage.pathway., PWY.5431..aromatic.compounds.degradation.via..beta..ketoadipate, PWY3DJ.35471..L.ascorbate.biosynthesis.IV, PWY.6185..4.methylcatechol.degradation..ortho.cleavage., PWY.7245..superpathway.NAD.NADP...NADH.NADPH.interconversion..yeast., PWY.5181..toluene.degradation.III..aerobic...via.p.cresol., PWY.6562..norspermidine.biosynthesis, PWY.7431..aromatic.biogenic.amine.degradation..bacteria., PWY.6307..L.tryptophan.degradation.X..mammalian..via.tryptamine., PWY.6313..serotonin.degradation, URSIN.PWY..ureide.biosynthesis, LIPASYN.PWY..phospholipases, DHGLUCONATE.PYR.CAT.PWY..glucose.degradation..oxidative., DENITRIFICATION.PWY..nitrate.reduction.I..denitrification., PWY.6662..superpathway.of.quinolone.and.alkylquinolone.biosynthesis, PWY.6660..2.heptyl.3.hydroxy.4.1H..quinolone.biosynthesis, PWY.6182..superpathway.of.salicylate.degradation, P165.PWY..superpathway.of.purines.degradation.in.plants, PWY66.388..fatty.acid..alpha..oxidation.III, TYRFUMCAT.PWY..L.tyrosine.degradation.I, PWY.5101..L.isoleucine.biosynthesis.II, PWY.6167..flavin.biosynthesis.II..archaea., PWY.6748..nitrate.reduction.VII..denitrification., PWY.7039..phosphatidate.metabolism..as.a.signaling.molecule, VALDEG.PWY..L.valine.degradation.I, PWY.5198..factor.420.biosynthesis, PWY.6215..4.chlorobenzoate.degradation, AEROBACTINSYN.PWY..aerobactin.biosynthesis, P562.PWY..myo.inositol.degradation.I, PWY.7409..phospholipid.remodeling..phosphatidylethanolamine..yeast., PWY.4722..creatinine.degradation.II, PWY.3801..sucrose.degradation.II..sucrose.synthase., PWY.7345..superpathway.of.anaerobic.sucrose.degradation, P125.PWY..superpathway.of..R.R..butanediol.biosynthesis, PWY.5994..palmitate.biosynthesis.I..animals.and.fungi., PWY.4321..L.glutamate.degradation.IV, PWY.7528..L.methionine.salvage.cycle.I..bacteria.and.plants., PWY.4361..S.methyl.5.thio..alpha..D.ribose.1.phosphate.degradation, PWY.1541..superpathway.of.taurine.degradation, PWY.5654..2.amino.3.carboxymuconate.semialdehyde.degradation.to.2.oxopentenoate, PWY.6210..2.aminophenol.degradation, PWY.6071..superpathway.of.phenylethylamine.degradation, PWY0.321..phenylacetate.degradation.I..aerobic., PWY.7200..superpathway.of.pyrimidine.deoxyribonucleoside.salvage, PWY.2221..Entner.Doudoroff.pathway.III..semi.phosphorylative., PWY.6992..1.5.anhydrofructose.degradation, PWY.7294..xylose.degradation.IV, PWY.6145..superpathway.of.sialic.acids.and.CMP.sialic.acids.biosynthesis, PWY.5180..toluene.degradation.I..aerobic...via.o.cresol., PWY.5182..toluene.degradation.II..aerobic...via.4.methylcatechol., PWY.5415..catechol.degradation.I..meta.cleavage.pathway., PWY.6785..hydrogen.production.VIII, PWY.5509..adenosylcobalamin.biosynthesis.from.cobyrinate.a.c.diamide.I, PWY.6641..superpathway.of.sulfolactate.degradation, PWY.7527..L.methionine.salvage.cycle.III, PWY.6396..superpathway.of.2.3.butanediol.biosynthesis, PWY.6467..Kdo.transfer.to.lipid.IVA.III..Chlamydia., X7ALPHADEHYDROX.PWY..cholate.degradation..bacteria..anaerobic., PWY.7374..1.4.dihydroxy.6.naphthoate.biosynthesis.I, PWY.6906..chitin.derivatives.degradation, PWY.5514..UDP.N.acetyl.D.galactosamine.biosynthesis.II, PWY.7317..superpathway.of.dTDP.glucose.derived.O.antigen.building.blocks.biosynthesis, PWY.7218..photosynthetic.3.hydroxybutanoate.biosynthesis..engineered., PHOTOALL.PWY..oxygenic.photosynthesis, PWY.6344..L.ornithine.degradation.II..Stickland.reaction., P621.PWY..nylon.6.oligomer.degradation, PWY.5028..L.histidine.degradation.II, CRNFORCAT.PWY..creatinine.degradation.I, PWY.6269..adenosylcobalamin.salvage.from.cobinamide.II, PWY.7389..superpathway.of.anaerobic.energy.metabolism..invertebrates., PWY.7384..anaerobic.energy.metabolism..invertebrates..mitochondrial., PWY.6165..chorismate.biosynthesis.II..archaea., PWY5F9.12..biphenyl.degradation, PWY.5647..2.nitrobenzoate.degradation.I, PWY.6138..CMP.N.acetylneuraminate.biosynthesis.I..eukaryotes., PWY.5910..superpathway.of.geranylgeranyldiphosphate.biosynthesis.I..via.mevalonate., PWY.922..mevalonate.pathway.I, PWY.6435..4.hydroxybenzoate.biosynthesis.V, PWY3O.1109..superpathway.of.4.hydroxybenzoate.biosynthesis..yeast., PWY.5754..4.hydroxybenzoate.biosynthesis.I..eukaryotes., PWY1G.0..mycothiol.biosynthesis, PWY.1501..mandelate.degradation.I, PWY.6107..chlorosalicylate.degradation, PWY.5534..propylene.degradation, PWY.7118..chitin.degradation.to.ethanol, PWY.7290..Escherichia.coli.serotype.O86.O.antigen.biosynthesis, PWY.181..photorespiration, PWY.1622..formaldehyde.assimilation.I..serine.pathway., CODH.PWY..reductive.acetyl.coenzyme.A.pathway, PWY.6349..CDP.archaeol.biosynthesis, PWY.6174..mevalonate.pathway.II..archaea., PWY.7286..7..3.amino.3.carboxypropyl..wyosine.biosynthesis, METHANOGENESIS.PWY..methanogenesis.from.H2.and.CO2, PWY.7391..isoprene.biosynthesis.II..engineered., PWY.5055..nicotinate.degradation.III, PWY.7399..methylphosphonate.degradation.II, P241.PWY..coenzyme.B.biosynthesis, PWY.5178..toluene.degradation.IV..aerobic...via.catechol., PWY.5420..catechol.degradation.II..meta.cleavage.pathway., PWY.5419..catechol.degradation.to.2.oxopent.4.enoate.II</span></span>
<span><span class="co">##  2024-04-26 15:42:48.861061 INFO::Total filtered features with variance filtering: 0</span></span>
<span><span class="co">##  2024-04-26 15:42:48.86156 INFO::Filtered feature names from variance filtering:</span></span>
<span><span class="co">##  2024-04-26 15:42:48.861974 INFO::Running selected normalization method: NONE</span></span>
<span><span class="co">##  2024-04-26 15:42:48.862389 INFO::Applying z-score to standardize continuous metadata</span></span>
<span><span class="co">##  2024-04-26 15:42:48.865797 INFO::Running selected transform method: LOG</span></span>
<span><span class="co">##  2024-04-26 15:42:48.873714 INFO::Running selected analysis method: LM</span></span>
<span><span class="co">##  2024-04-26 15:42:48.874308 INFO::Fitting model to feature number 1, PWY.1042..glycolysis.IV..plant.cytosol.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.876958 INFO::Fitting model to feature number 2, DTDPRHAMSYN.PWY..dTDP.L.rhamnose.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.879198 INFO::Fitting model to feature number 3, PWY.5695..urate.biosynthesis.inosine.5..phosphate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:48.881491 INFO::Fitting model to feature number 4, PWY.6936..seleno.amino.acid.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.883722 INFO::Fitting model to feature number 5, ILEUSYN.PWY..L.isoleucine.biosynthesis.I..from.threonine.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.885983 INFO::Fitting model to feature number 6, PWY.7111..pyruvate.fermentation.to.isobutanol..engineered.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.888219 INFO::Fitting model to feature number 7, VALSYN.PWY..L.valine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.890476 INFO::Fitting model to feature number 8, PWY.6609..adenine.and.adenosine.salvage.III</span></span>
<span><span class="co">##  2024-04-26 15:42:48.892718 INFO::Fitting model to feature number 9, PWY.6737..starch.degradation.V</span></span>
<span><span class="co">##  2024-04-26 15:42:48.894967 INFO::Fitting model to feature number 10, PWY.7219..adenosine.ribonucleotides.de.novo.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.897236 INFO::Fitting model to feature number 11, PWY0.1586..peptidoglycan.maturation..meso.diaminopimelate.containing.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.899467 INFO::Fitting model to feature number 12, ASPASN.PWY..superpathway.of.L.aspartate.and.L.asparagine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.901751 INFO::Fitting model to feature number 13, PWY.5973..cis.vaccenate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.903993 INFO::Fitting model to feature number 14, PWY.7221..guanosine.ribonucleotides.de.novo.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.906236 INFO::Fitting model to feature number 15, PWY.5686..UMP.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.908405 INFO::Fitting model to feature number 16, X1CMET2.PWY..N10.formyl.tetrahydrofolate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.910637 INFO::Fitting model to feature number 17, PWY.6122..5.aminoimidazole.ribonucleotide.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:48.912873 INFO::Fitting model to feature number 18, PWY.6277..superpathway.of.5.aminoimidazole.ribonucleotide.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.915071 INFO::Fitting model to feature number 19, PWY.6385..peptidoglycan.biosynthesis.III..mycobacteria.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.917323 INFO::Fitting model to feature number 20, PWY.3841..folate.transformations.II</span></span>
<span><span class="co">##  2024-04-26 15:42:48.919561 INFO::Fitting model to feature number 21, RIBOSYN2.PWY..flavin.biosynthesis.I..bacteria.and.plants.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.921801 INFO::Fitting model to feature number 22, PWY.6387..UDP.N.acetylmuramoyl.pentapeptide.biosynthesis.I..meso.diaminopimelate.containing.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.924015 INFO::Fitting model to feature number 23, THISYNARA.PWY..superpathway.of.thiamin.diphosphate.biosynthesis.III..eukaryotes.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.926313 INFO::Fitting model to feature number 24, PWY.7663..gondoate.biosynthesis..anaerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.928544 INFO::Fitting model to feature number 25, PWY.6386..UDP.N.acetylmuramoyl.pentapeptide.biosynthesis.II..lysine.containing.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.930797 INFO::Fitting model to feature number 26, PWY.6700..queuosine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.933038 INFO::Fitting model to feature number 27, PEPTIDOGLYCANSYN.PWY..peptidoglycan.biosynthesis.I..meso.diaminopimelate.containing.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.935306 INFO::Fitting model to feature number 28, TRNA.CHARGING.PWY..tRNA.charging</span></span>
<span><span class="co">##  2024-04-26 15:42:48.937585 INFO::Fitting model to feature number 29, PWY.6121..5.aminoimidazole.ribonucleotide.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.939858 INFO::Fitting model to feature number 30, HISTSYN.PWY..L.histidine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.942135 INFO::Fitting model to feature number 31, PWY.7229..superpathway.of.adenosine.nucleotides.de.novo.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.944351 INFO::Fitting model to feature number 32, PWY.7199..pyrimidine.deoxyribonucleosides.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:48.946595 INFO::Fitting model to feature number 33, PANTO.PWY..phosphopantothenate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.948838 INFO::Fitting model to feature number 34, PWY.2942..L.lysine.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:48.951081 INFO::Fitting model to feature number 35, PWY.7237..myo...chiro..and.scillo.inositol.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:48.953308 INFO::Fitting model to feature number 36, PWY.6168..flavin.biosynthesis.III..fungi.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.955534 INFO::Fitting model to feature number 37, COA.PWY.1..coenzyme.A.biosynthesis.II..mammalian.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.95779 INFO::Fitting model to feature number 38, PWY.5667..CDP.diacylglycerol.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.96001 INFO::Fitting model to feature number 39, PWY0.1319..CDP.diacylglycerol.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:48.962262 INFO::Fitting model to feature number 40, PWY.5097..L.lysine.biosynthesis.VI</span></span>
<span><span class="co">##  2024-04-26 15:42:48.964459 INFO::Fitting model to feature number 41, ANAGLYCOLYSIS.PWY..glycolysis.III..from.glucose.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.966733 INFO::Fitting model to feature number 42, PWY.6123..inosine.5..phosphate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.968968 INFO::Fitting model to feature number 43, ARGININE.SYN4.PWY..L.ornithine.de.novo..biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.971154 INFO::Fitting model to feature number 44, PWY.6163..chorismate.biosynthesis.from.3.dehydroquinate</span></span>
<span><span class="co">##  2024-04-26 15:42:48.973408 INFO::Fitting model to feature number 45, THRESYN.PWY..superpathway.of.L.threonine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:48.975619 INFO::Fitting model to feature number 46, PYRIDNUCSYN.PWY..NAD.biosynthesis.I..from.aspartate.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.977877 INFO::Fitting model to feature number 47, PWY.6124..inosine.5..phosphate.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:48.980092 INFO::Fitting model to feature number 48, PWY.6147..6.hydroxymethyl.dihydropterin.diphosphate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.982345 INFO::Fitting model to feature number 49, PWY.7539..6.hydroxymethyl.dihydropterin.diphosphate.biosynthesis.III..Chlamydia.</span></span>
<span><span class="co">##  2024-04-26 15:42:48.984541 INFO::Fitting model to feature number 50, PWY.6151..S.adenosyl.L.methionine.cycle.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.986765 INFO::Fitting model to feature number 51, SER.GLYSYN.PWY..superpathway.of.L.serine.and.glycine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.988991 INFO::Fitting model to feature number 52, PWY.6126..superpathway.of.adenosine.nucleotides.de.novo.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:48.991179 INFO::Fitting model to feature number 53, PANTOSYN.PWY..pantothenate.and.coenzyme.A.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.993405 INFO::Fitting model to feature number 54, PWY.7228..superpathway.of.guanosine.nucleotides.de.novo.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.99574 INFO::Fitting model to feature number 55, COA.PWY..coenzyme.A.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:48.997975 INFO::Fitting model to feature number 56, PWY.4242..pantothenate.and.coenzyme.A.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.000158 INFO::Fitting model to feature number 57, PWY.6703..preQ0.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.002379 INFO::Fitting model to feature number 58, THISYN.PWY..superpathway.of.thiamin.diphosphate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.004576 INFO::Fitting model to feature number 59, HISDEG.PWY..L.histidine.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.00679 INFO::Fitting model to feature number 60, PWY.5659..GDP.mannose.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.009001 INFO::Fitting model to feature number 61, PWY.5030..L.histidine.degradation.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.01118 INFO::Fitting model to feature number 62, PWY.5484..glycolysis.II..from.fructose.6.phosphate.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.013447 INFO::Fitting model to feature number 63, PWY.3001..superpathway.of.L.isoleucine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.015653 INFO::Fitting model to feature number 64, PWY.6125..superpathway.of.guanosine.nucleotides.de.novo.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.017888 INFO::Fitting model to feature number 65, TRPSYN.PWY..L.tryptophan.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.020069 INFO::Fitting model to feature number 66, GLYCOLYSIS..glycolysis.I..from.glucose.6.phosphate.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.022272 INFO::Fitting model to feature number 67, PWY.724..superpathway.of.L.lysine..L.threonine.and.L.methionine.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.024454 INFO::Fitting model to feature number 68, PWY.6897..thiamin.salvage.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.026695 INFO::Fitting model to feature number 69, BRANCHED.CHAIN.AA.SYN.PWY..superpathway.of.branched.amino.acid.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.02893 INFO::Fitting model to feature number 70, CALVIN.PWY..Calvin.Benson.Bassham.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.031124 INFO::Fitting model to feature number 71, PWY.841..superpathway.of.purine.nucleotides.de.novo.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.033348 INFO::Fitting model to feature number 72, COMPLETE.ARO.PWY..superpathway.of.aromatic.amino.acid.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.035551 INFO::Fitting model to feature number 73, PWY.7220..adenosine.deoxyribonucleotides.de.novo.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.03779 INFO::Fitting model to feature number 74, PWY.7222..guanosine.deoxyribonucleotides.de.novo.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.040065 INFO::Fitting model to feature number 75, DENOVOPURINE2.PWY..superpathway.of.purine.nucleotides.de.novo.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.042314 INFO::Fitting model to feature number 76, PWY.7357..thiamin.formation.from.pyrithiamine.and.oxythiamine..yeast.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.044537 INFO::Fitting model to feature number 77, PWY.5103..L.isoleucine.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.046775 INFO::Fitting model to feature number 78, ARO.PWY..chorismate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.048986 INFO::Fitting model to feature number 79, PWY.7282..4.amino.2.methyl.5.phosphomethylpyrimidine.biosynthesis..yeast.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.051172 INFO::Fitting model to feature number 80, PWY0.845..superpathway.of.pyridoxal.5..phosphate.biosynthesis.and.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:49.053452 INFO::Fitting model to feature number 81, ANAEROFRUCAT.PWY..homolactic.fermentation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.055645 INFO::Fitting model to feature number 82, PWY.6892..thiazole.biosynthesis.I..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.057898 INFO::Fitting model to feature number 83, RHAMCAT.PWY..L.rhamnose.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.060157 INFO::Fitting model to feature number 84, PYRIDOXSYN.PWY..pyridoxal.5..phosphate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.062469 INFO::Fitting model to feature number 85, PWY.1269..CMP.3.deoxy.D.manno.octulosonate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.065149 INFO::Fitting model to feature number 86, PHOSLIPSYN.PWY..superpathway.of.phospholipid.biosynthesis.I..bacteria.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.068436 INFO::Fitting model to feature number 87, CITRULBIO.PWY..L.citrulline.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.071484 INFO::Fitting model to feature number 88, NAGLIPASYN.PWY..lipid.IVA.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.074555 INFO::Fitting model to feature number 89, BIOTIN.BIOSYNTHESIS.PWY..biotin.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.077718 INFO::Fitting model to feature number 90, PWY.5154..L.arginine.biosynthesis.III..via.N.acetyl.L.citrulline.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.080777 INFO::Fitting model to feature number 91, FASYN.ELONG.PWY..fatty.acid.elongation....saturated</span></span>
<span><span class="co">##  2024-04-26 15:42:49.083789 INFO::Fitting model to feature number 92, PENTOSE.P.PWY..pentose.phosphate.pathway</span></span>
<span><span class="co">##  2024-04-26 15:42:49.086869 INFO::Fitting model to feature number 93, PWYG.321..mycolate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.089956 INFO::Fitting model to feature number 94, PWY.6519..8.amino.7.oxononanoate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.093017 INFO::Fitting model to feature number 95, PWY.7664..oleate.biosynthesis.IV..anaerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.095782 INFO::Fitting model to feature number 96, PWY.4981..L.proline.biosynthesis.II..from.arginine.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.098041 INFO::Fitting model to feature number 97, GLYCOGENSYNTH.PWY..glycogen.biosynthesis.I..from.ADP.D.Glucose.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.100264 INFO::Fitting model to feature number 98, PWY.7388..octanoyl..acyl.carrier.protein..biosynthesis..mitochondria..yeast.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.102509 INFO::Fitting model to feature number 99, NONOXIPENT.PWY..pentose.phosphate.pathway..non.oxidative.branch.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.104777 INFO::Fitting model to feature number 100, COBALSYN.PWY..adenosylcobalamin.salvage.from.cobinamide.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.107048 INFO::Fitting model to feature number 101, PWY.5989..stearate.biosynthesis.II..bacteria.and.plants.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.10937 INFO::Fitting model to feature number 102, PWY.6282..palmitoleate.biosynthesis.I..from..5Z..dodec.5.enoate.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.111568 INFO::Fitting model to feature number 103, PWY0.862...5Z..dodec.5.enoate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.11381 INFO::Fitting model to feature number 104, PWY.7400..L.arginine.biosynthesis.IV..archaebacteria.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.11603 INFO::Fitting model to feature number 105, ARGSYN.PWY..L.arginine.biosynthesis.I..via.L.ornithine.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.118291 INFO::Fitting model to feature number 106, PWY.4984..urea.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.120479 INFO::Fitting model to feature number 107, FASYN.INITIAL.PWY..superpathway.of.fatty.acid.biosynthesis.initiation..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.122766 INFO::Fitting model to feature number 108, PWY4FS.7..phosphatidylglycerol.biosynthesis.I..plastidic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.125003 INFO::Fitting model to feature number 109, PWY4FS.8..phosphatidylglycerol.biosynthesis.II..non.plastidic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.127204 INFO::Fitting model to feature number 110, ARGSYNBSUB.PWY..L.arginine.biosynthesis.II..acetyl.cycle.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.129449 INFO::Fitting model to feature number 111, PWY.6317..galactose.degradation.I..Leloir.pathway.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.131643 INFO::Fitting model to feature number 112, PWY66.422..D.galactose.degradation.V..Leloir.pathway.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.133862 INFO::Fitting model to feature number 113, GLUTORN.PWY..L.ornithine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.136071 INFO::Fitting model to feature number 114, PWY.2941..L.lysine.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.138335 INFO::Fitting model to feature number 115, PWY0.1296..purine.ribonucleosides.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.140528 INFO::Fitting model to feature number 116, NAD.BIOSYNTHESIS.II..NAD.salvage.pathway.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.142792 INFO::Fitting model to feature number 117, HSERMETANA.PWY..L.methionine.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.145064 INFO::Fitting model to feature number 118, PWY0.162..superpathway.of.pyrimidine.ribonucleotides.de.novo.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.147324 INFO::Fitting model to feature number 119, PWY.7208..superpathway.of.pyrimidine.nucleobases.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:49.149543 INFO::Fitting model to feature number 120, TEICHOICACID.PWY..teichoic.acid..poly.glycerol..biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.151734 INFO::Fitting model to feature number 121, PWY.6305..putrescine.biosynthesis.IV</span></span>
<span><span class="co">##  2024-04-26 15:42:49.154002 INFO::Fitting model to feature number 122, GLUCONEO.PWY..gluconeogenesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.156238 INFO::Fitting model to feature number 123, PWY.621..sucrose.degradation.III..sucrose.invertase.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.158494 INFO::Fitting model to feature number 124, PWY.6901..superpathway.of.glucose.and.xylose.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.160746 INFO::Fitting model to feature number 125, POLYISOPRENSYN.PWY..polyisoprenoid.biosynthesis..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.162976 INFO::Fitting model to feature number 126, PWY.7323..superpathway.of.GDP.mannose.derived.O.antigen.building.blocks.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.165191 INFO::Fitting model to feature number 127, PWY.5941..glycogen.degradation.II..eukaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.167394 INFO::Fitting model to feature number 128, PWY.6527..stachyose.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.169647 INFO::Fitting model to feature number 129, PWY0.1261..anhydromuropeptides.recycling</span></span>
<span><span class="co">##  2024-04-26 15:42:49.171844 INFO::Fitting model to feature number 130, PWY66.400..glycolysis.VI..metazoan.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.17412 INFO::Fitting model to feature number 131, PWY0.166..superpathway.of.pyrimidine.deoxyribonucleotides.de.novo.biosynthesis..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.176376 INFO::Fitting model to feature number 132, GLYCOLYSIS.E.D..superpathway.of.glycolysis.and.Entner.Doudoroff</span></span>
<span><span class="co">##  2024-04-26 15:42:49.178631 INFO::Fitting model to feature number 133, PWY.7187..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.180882 INFO::Fitting model to feature number 134, PWY.7456..mannan.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.183131 INFO::Fitting model to feature number 135, PWY.7184..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.185468 INFO::Fitting model to feature number 136, OANTIGEN.PWY..O.antigen.building.blocks.biosynthesis..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.187693 INFO::Fitting model to feature number 137, COLANSYN.PWY..colanic.acid.building.blocks.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.189924 INFO::Fitting model to feature number 138, PWY.6545..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.19211 INFO::Fitting model to feature number 139, PWY.6895..superpathway.of.thiamin.diphosphate.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.194358 INFO::Fitting model to feature number 140, PWY.6608..guanosine.nucleotides.degradation.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.196589 INFO::Fitting model to feature number 141, TCA..TCA.cycle.I..prokaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.198832 INFO::Fitting model to feature number 142, PWY.6859..all.trans.farnesol.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.201084 INFO::Fitting model to feature number 143, PWY.5100..pyruvate.fermentation.to.acetate.and.lactate.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.203319 INFO::Fitting model to feature number 144, PWY.6572..chondroitin.sulfate.degradation.I..bacterial.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.205585 INFO::Fitting model to feature number 145, PWY.6471..peptidoglycan.biosynthesis.IV..Enterococcus.faecium.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.207862 INFO::Fitting model to feature number 146, PWY.6470..peptidoglycan.biosynthesis.V...beta..lactam.resistance.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.210128 INFO::Fitting model to feature number 147, PWY.5188..tetrapyrrole.biosynthesis.I..from.glutamate.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.212517 INFO::Fitting model to feature number 148, GLUCUROCAT.PWY..superpathway.of..beta..D.glucuronide.and.D.glucuronate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.2148 INFO::Fitting model to feature number 149, PWY.6749..CMP.legionaminate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.217115 INFO::Fitting model to feature number 150, P164.PWY..purine.nucleobases.degradation.I..anaerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.219373 INFO::Fitting model to feature number 151, PWY.7198..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.IV</span></span>
<span><span class="co">##  2024-04-26 15:42:49.221644 INFO::Fitting model to feature number 152, PWY.7211..superpathway.of.pyrimidine.deoxyribonucleotides.de.novo.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.223882 INFO::Fitting model to feature number 153, PWY66.409..superpathway.of.purine.nucleotide.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:49.226144 INFO::Fitting model to feature number 154, GALACT.GLUCUROCAT.PWY..superpathway.of.hexuronide.and.hexuronate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.228367 INFO::Fitting model to feature number 155, UDPNAGSYN.PWY..UDP.N.acetyl.D.glucosamine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.230598 INFO::Fitting model to feature number 156, PWY.6353..purine.nucleotides.degradation.II..aerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.232851 INFO::Fitting model to feature number 157, PWY.7197..pyrimidine.deoxyribonucleotide.phosphorylation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.235064 INFO::Fitting model to feature number 158, PWY.7242..D.fructuronate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.237296 INFO::Fitting model to feature number 159, PWY.7196..superpathway.of.pyrimidine.ribonucleosides.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:49.239563 INFO::Fitting model to feature number 160, PWY.5690..TCA.cycle.II..plants.and.fungi.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.241851 INFO::Fitting model to feature number 161, PWY.5177..glutaryl.CoA.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.24409 INFO::Fitting model to feature number 162, PWY.6478..GDP.D.glycero..alpha..D.manno.heptose.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.246329 INFO::Fitting model to feature number 163, HEXITOLDEGSUPER.PWY..superpathway.of.hexitol.degradation..bacteria.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.248586 INFO::Fitting model to feature number 164, GALACTUROCAT.PWY..D.galacturonate.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.250856 INFO::Fitting model to feature number 165, PWY.5104..L.isoleucine.biosynthesis.IV</span></span>
<span><span class="co">##  2024-04-26 15:42:49.253119 INFO::Fitting model to feature number 166, PWY.5505..L.glutamate.and.L.glutamine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.25537 INFO::Fitting model to feature number 167, P441.PWY..superpathway.of.N.acetylneuraminate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.257687 INFO::Fitting model to feature number 168, PWY.6606..guanosine.nucleotides.degradation.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.259903 INFO::Fitting model to feature number 169, SALVADEHYPOX.PWY..adenosine.nucleotides.degradation.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.262187 INFO::Fitting model to feature number 170, PWY.5005..biotin.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.264455 INFO::Fitting model to feature number 171, PWY.6549..L.glutamine.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.266727 INFO::Fitting model to feature number 172, GLCMANNANAUT.PWY..superpathway.of.N.acetylglucosamine..N.acetylmannosamine.and.N.acetylneuraminate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.269187 INFO::Fitting model to feature number 173, PWY0.1241..ADP.L.glycero..beta..D.manno.heptose.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.271415 INFO::Fitting model to feature number 174, PWY.6507..4.deoxy.L.threo.hex.4.enopyranuronate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.273672 INFO::Fitting model to feature number 175, GALACTARDEG.PWY..D.galactarate.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.275892 INFO::Fitting model to feature number 176, GLUCARGALACTSUPER.PWY..superpathway.of.D.glucarate.and.D.galactarate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.278145 INFO::Fitting model to feature number 177, PWY.5676..acetyl.CoA.fermentation.to.butanoate.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.280396 INFO::Fitting model to feature number 178, PWY.5347..superpathway.of.L.methionine.biosynthesis..transsulfuration.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.282669 INFO::Fitting model to feature number 179, PWY.5138..unsaturated..even.numbered.fatty.acid..beta..oxidation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.285044 INFO::Fitting model to feature number 180, MET.SAM.PWY..superpathway.of.S.adenosyl.L.methionine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.287311 INFO::Fitting model to feature number 181, PWY66.399..gluconeogenesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.289594 INFO::Fitting model to feature number 182, PWY0.1297..superpathway.of.purine.deoxyribonucleosides.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.291841 INFO::Fitting model to feature number 183, METSYN.PWY..L.homoserine.and.L.methionine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.29413 INFO::Fitting model to feature number 184, PWY.7332..superpathway.of.UDP.N.acetylglucosamine.derived.O.antigen.building.blocks.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.296397 INFO::Fitting model to feature number 185, PWY.6588..pyruvate.fermentation.to.acetone</span></span>
<span><span class="co">##  2024-04-26 15:42:49.298673 INFO::Fitting model to feature number 186, PWY.6891..thiazole.biosynthesis.II..Bacillus.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.300946 INFO::Fitting model to feature number 187, GLUCARDEG.PWY..D.glucarate.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.303203 INFO::Fitting model to feature number 188, METH.ACETATE.PWY..methanogenesis.from.acetate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.305499 INFO::Fitting model to feature number 189, PWY.7013..L.1.2.propanediol.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.307742 INFO::Fitting model to feature number 190, PWY.6590..superpathway.of.Clostridium.acetobutylicum.acidogenic.fermentation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.31003 INFO::Fitting model to feature number 191, FAO.PWY..fatty.acid..beta..oxidation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.312297 INFO::Fitting model to feature number 192, PWY.6318..L.phenylalanine.degradation.IV..mammalian..via.side.chain.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.314588 INFO::Fitting model to feature number 193, CENTFERM.PWY..pyruvate.fermentation.to.butanoate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.316857 INFO::Fitting model to feature number 194, GOLPDLCAT.PWY..superpathway.of.glycerol.degradation.to.1.3.propanediol</span></span>
<span><span class="co">##  2024-04-26 15:42:49.319118 INFO::Fitting model to feature number 195, PWY.5136..fatty.acid..beta..oxidation.II..peroxisome.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.321381 INFO::Fitting model to feature number 196, PWY66.391..fatty.acid..beta..oxidation.VI..peroxisome.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.323616 INFO::Fitting model to feature number 197, HOMOSER.METSYN.PWY..L.methionine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.325938 INFO::Fitting model to feature number 198, METHGLYUT.PWY..superpathway.of.methylglyoxal.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.328206 INFO::Fitting model to feature number 199, PWY.7288..fatty.acid..beta..oxidation..peroxisome..yeast.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.330492 INFO::Fitting model to feature number 200, PWY.7383..anaerobic.energy.metabolism..invertebrates..cytosol.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.332754 INFO::Fitting model to feature number 201, P162.PWY..L.glutamate.degradation.V..via.hydroxyglutarate.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.334996 INFO::Fitting model to feature number 202, PWY.4041...gamma..glutamyl.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.337285 INFO::Fitting model to feature number 203, P108.PWY..pyruvate.fermentation.to.propanoate.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.339512 INFO::Fitting model to feature number 204, PWY.5022..4.aminobutanoate.degradation.V</span></span>
<span><span class="co">##  2024-04-26 15:42:49.341792 INFO::Fitting model to feature number 205, GLUDEG.I.PWY..GABA.shunt</span></span>
<span><span class="co">##  2024-04-26 15:42:49.344049 INFO::Fitting model to feature number 206, PWY.5677..succinate.fermentation.to.butanoate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.346322 INFO::Fitting model to feature number 207, P163.PWY..L.lysine.fermentation.to.acetate.and.butanoate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.348571 INFO::Fitting model to feature number 208, PWY.3781..aerobic.respiration.I..cytochrome.c.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.350861 INFO::Fitting model to feature number 209, PWY.5083..NAD.NADH.phosphorylation.and.dephosphorylation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.353178 INFO::Fitting model to feature number 210, DAPLYSINESYN.PWY..L.lysine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.355431 INFO::Fitting model to feature number 211, P461.PWY..hexitol.fermentation.to.lactate..formate..ethanol.and.acetate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.357746 INFO::Fitting model to feature number 212, PWY.1861..formaldehyde.assimilation.II..RuMP.Cycle.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.360018 INFO::Fitting model to feature number 213, PWY.7312..dTDP.D..beta..fucofuranose.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.362322 INFO::Fitting model to feature number 214, PWY.6630..superpathway.of.L.tyrosine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.364579 INFO::Fitting model to feature number 215, PWY.6595..superpathway.of.guanosine.nucleotides.degradation..plants.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.366847 INFO::Fitting model to feature number 216, ARG.POLYAMINE.SYN..superpathway.of.arginine.and.polyamine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.369129 INFO::Fitting model to feature number 217, PWY.7003..glycerol.degradation.to.butanol</span></span>
<span><span class="co">##  2024-04-26 15:42:49.371391 INFO::Fitting model to feature number 218, P185.PWY..formaldehyde.assimilation.III..dihydroxyacetone.cycle.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.373708 INFO::Fitting model to feature number 219, POLYAMSYN.PWY..superpathway.of.polyamine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.37597 INFO::Fitting model to feature number 220, PWY0.1298..superpathway.of.pyrimidine.deoxyribonucleosides.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.378233 INFO::Fitting model to feature number 221, LACTOSECAT.PWY..lactose.and.galactose.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.380471 INFO::Fitting model to feature number 222, PWY4LZ.257..superpathway.of.fermentation..Chlamydomonas.reinhardtii.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.38278 INFO::Fitting model to feature number 223, P161.PWY..acetylene.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.385064 INFO::Fitting model to feature number 224, RUMP.PWY..formaldehyde.oxidation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.387361 INFO::Fitting model to feature number 225, PPGPPMET.PWY..ppGpp.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.389636 INFO::Fitting model to feature number 226, PWY.5304..superpathway.of.sulfur.oxidation..Acidianus.ambivalens.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.39186 INFO::Fitting model to feature number 227, PWY.6531..mannitol.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.394133 INFO::Fitting model to feature number 228, FUC.RHAMCAT.PWY..superpathway.of.fucose.and.rhamnose.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.396376 INFO::Fitting model to feature number 229, PWY.7315..dTDP.N.acetylthomosamine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.398639 INFO::Fitting model to feature number 230, FUCCAT.PWY..fucose.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.400899 INFO::Fitting model to feature number 231, PWY.7210..pyrimidine.deoxyribonucleotides.biosynthesis.from.CTP</span></span>
<span><span class="co">##  2024-04-26 15:42:49.403136 INFO::Fitting model to feature number 232, FOLSYN.PWY..superpathway.of.tetrahydrofolate.biosynthesis.and.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:49.405455 INFO::Fitting model to feature number 233, PWY.6612..superpathway.of.tetrahydrofolate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.407705 INFO::Fitting model to feature number 234, PWY0.1061..superpathway.of.L.alanine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.409965 INFO::Fitting model to feature number 235, PWY.5189..tetrapyrrole.biosynthesis.II..from.glycine.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.412188 INFO::Fitting model to feature number 236, PWY.6876..isopropanol.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.414446 INFO::Fitting model to feature number 237, PWY0.781..aspartate.superpathway</span></span>
<span><span class="co">##  2024-04-26 15:42:49.416723 INFO::Fitting model to feature number 238, P4.PWY..superpathway.of.L.lysine..L.threonine.and.L.methionine.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.419006 INFO::Fitting model to feature number 239, PWY0.1479..tRNA.processing</span></span>
<span><span class="co">##  2024-04-26 15:42:49.421307 INFO::Fitting model to feature number 240, GLYCOLYSIS.TCA.GLYOX.BYPASS..superpathway.of.glycolysis..pyruvate.dehydrogenase..TCA..and.glyoxylate.bypass</span></span>
<span><span class="co">##  2024-04-26 15:42:49.423565 INFO::Fitting model to feature number 241, PYRIDNUCSAL.PWY..NAD.salvage.pathway.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.425863 INFO::Fitting model to feature number 242, GLYCOCAT.PWY..glycogen.degradation.I..bacterial.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.428098 INFO::Fitting model to feature number 243, PWY.5723..Rubisco.shunt</span></span>
<span><span class="co">##  2024-04-26 15:42:49.430414 INFO::Fitting model to feature number 244, TCA.GLYOX.BYPASS..superpathway.of.glyoxylate.bypass.and.TCA</span></span>
<span><span class="co">##  2024-04-26 15:42:49.432728 INFO::Fitting model to feature number 245, PWY.7385..1.3.propanediol.biosynthesis..engineered.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.434991 INFO::Fitting model to feature number 246, PWY.7254..TCA.cycle.VII..acetate.producers.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.437274 INFO::Fitting model to feature number 247, GLYOXYLATE.BYPASS..glyoxylate.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.439572 INFO::Fitting model to feature number 248, PWY.6731..starch.degradation.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.442128 INFO::Fitting model to feature number 249, PWY.7328..superpathway.of.UDP.glucose.derived.O.antigen.building.blocks.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.444974 INFO::Fitting model to feature number 250, POLYAMINSYN3.PWY..superpathway.of.polyamine.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.447769 INFO::Fitting model to feature number 251, PWY.2723..trehalose.degradation.V</span></span>
<span><span class="co">##  2024-04-26 15:42:49.450574 INFO::Fitting model to feature number 252, GLUCOSE1PMETAB.PWY..glucose.and.glucose.1.phosphate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.453457 INFO::Fitting model to feature number 253, PWY.5384..sucrose.degradation.IV..sucrose.phosphorylase.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.456258 INFO::Fitting model to feature number 254, PWY.7046..4.coumarate.degradation..anaerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.459081 INFO::Fitting model to feature number 255, PWY66.389..phytol.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.461982 INFO::Fitting model to feature number 256, PWY.5971..palmitate.biosynthesis.II..bacteria.and.plants.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.464826 INFO::Fitting model to feature number 257, PWY.6629..superpathway.of.L.tryptophan.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.467646 INFO::Fitting model to feature number 258, PWY.5345..superpathway.of.L.methionine.biosynthesis..by.sulfhydrylation.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.470505 INFO::Fitting model to feature number 259, SULFATE.CYS.PWY..superpathway.of.sulfate.assimilation.and.cysteine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.473342 INFO::Fitting model to feature number 260, PWY.5838..superpathway.of.menaquinol.8.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.476199 INFO::Fitting model to feature number 261, PRPP.PWY..superpathway.of.histidine..purine..and.pyrimidine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.479055 INFO::Fitting model to feature number 262, PWY.5845..superpathway.of.menaquinol.9.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.481934 INFO::Fitting model to feature number 263, PWY.6113..superpathway.of.mycolate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.484843 INFO::Fitting model to feature number 264, PWY.5897..superpathway.of.menaquinol.11.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.487704 INFO::Fitting model to feature number 265, PWY.5898..superpathway.of.menaquinol.12.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.490585 INFO::Fitting model to feature number 266, PWY.5899..superpathway.of.menaquinol.13.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.493469 INFO::Fitting model to feature number 267, PWY.7117..C4.photosynthetic.carbon.assimilation.cycle..PEPCK.type</span></span>
<span><span class="co">##  2024-04-26 15:42:49.496359 INFO::Fitting model to feature number 268, PWY.6969..TCA.cycle.V..2.oxoglutarate.ferredoxin.oxidoreductase.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.49922 INFO::Fitting model to feature number 269, PWY.6628..superpathway.of.L.phenylalanine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.502085 INFO::Fitting model to feature number 270, PWY.5861..superpathway.of.demethylmenaquinol.8.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.504964 INFO::Fitting model to feature number 271, PWY.5862..superpathway.of.demethylmenaquinol.9.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.507841 INFO::Fitting model to feature number 272, FERMENTATION.PWY..mixed.acid.fermentation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.510718 INFO::Fitting model to feature number 273, PWY.241..C4.photosynthetic.carbon.assimilation.cycle..NADP.ME.type</span></span>
<span><span class="co">##  2024-04-26 15:42:49.513636 INFO::Fitting model to feature number 274, SO4ASSIM.PWY..sulfate.reduction.I..assimilatory.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.516494 INFO::Fitting model to feature number 275, KETOGLUCONMET.PWY..ketogluconate.metabolism</span></span>
<span><span class="co">##  2024-04-26 15:42:49.51935 INFO::Fitting model to feature number 276, PWY.5840..superpathway.of.menaquinol.7.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.522221 INFO::Fitting model to feature number 277, PWY.5850..superpathway.of.menaquinol.6.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.525074 INFO::Fitting model to feature number 278, PWY.5896..superpathway.of.menaquinol.10.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.527933 INFO::Fitting model to feature number 279, P105.PWY..TCA.cycle.IV..2.oxoglutarate.decarboxylase.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.530822 INFO::Fitting model to feature number 280, PWY.5913..TCA.cycle.VI..obligate.autotrophs.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.533721 INFO::Fitting model to feature number 281, PWY.561..superpathway.of.glyoxylate.cycle.and.fatty.acid.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.536586 INFO::Fitting model to feature number 282, PWY.4702..phytate.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.539504 INFO::Fitting model to feature number 283, PWY.6803..phosphatidylcholine.acyl.editing</span></span>
<span><span class="co">##  2024-04-26 15:42:49.542392 INFO::Fitting model to feature number 284, PWY.7115..C4.photosynthetic.carbon.assimilation.cycle..NAD.ME.type</span></span>
<span><span class="co">##  2024-04-26 15:42:49.545253 INFO::Fitting model to feature number 285, PWY.5918..superpathay.of.heme.biosynthesis.from.glutamate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.548109 INFO::Fitting model to feature number 286, PWY.7234..inosine.5..phosphate.biosynthesis.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.550985 INFO::Fitting model to feature number 287, PWY.6285..superpathway.of.fatty.acids.biosynthesis..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.553936 INFO::Fitting model to feature number 288, PWY.5860..superpathway.of.demethylmenaquinol.6.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.556806 INFO::Fitting model to feature number 289, PWY.5863..superpathway.of.phylloquinol.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.559637 INFO::Fitting model to feature number 290, HEME.BIOSYNTHESIS.II..heme.biosynthesis.I..aerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.562522 INFO::Fitting model to feature number 291, PWY.821..superpathway.of.sulfur.amino.acid.biosynthesis..Saccharomyces.cerevisiae.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.5654 INFO::Fitting model to feature number 292, PWY.5675..nitrate.reduction.V..assimilatory.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.568259 INFO::Fitting model to feature number 293, PWY.5791..1.4.dihydroxy.2.naphthoate.biosynthesis.II..plants.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.571152 INFO::Fitting model to feature number 294, PWY.5837..1.4.dihydroxy.2.naphthoate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.574032 INFO::Fitting model to feature number 295, PWY.5920..superpathway.of.heme.biosynthesis.from.glycine</span></span>
<span><span class="co">##  2024-04-26 15:42:49.57694 INFO::Fitting model to feature number 296, PWY.7204..pyridoxal.5..phosphate.salvage.II..plants.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.609449 INFO::Fitting model to feature number 297, ENTBACSYN.PWY..enterobactin.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.612771 INFO::Fitting model to feature number 298, PWY0.1415..superpathway.of.heme.biosynthesis.from.uroporphyrinogen.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.615045 INFO::Fitting model to feature number 299, HEMESYN2.PWY..heme.biosynthesis.II..anaerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.617312 INFO::Fitting model to feature number 300, PWY.6284..superpathway.of.unsaturated.fatty.acids.biosynthesis..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.619551 INFO::Fitting model to feature number 301, P42.PWY..incomplete.reductive.TCA.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.621841 INFO::Fitting model to feature number 302, PWY.5656..mannosylglycerate.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.624111 INFO::Fitting model to feature number 303, PWY.5173..superpathway.of.acetyl.CoA.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.626412 INFO::Fitting model to feature number 304, ECASYN.PWY..enterobacterial.common.antigen.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.628704 INFO::Fitting model to feature number 305, AST.PWY..L.arginine.degradation.II..AST.pathway.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.630973 INFO::Fitting model to feature number 306, PWY3O.355..stearate.biosynthesis.III..fungi.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.633254 INFO::Fitting model to feature number 307, PWY0.42..2.methylcitrate.cycle.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.635503 INFO::Fitting model to feature number 308, PWY.7094..fatty.acid.salvage</span></span>
<span><span class="co">##  2024-04-26 15:42:49.637795 INFO::Fitting model to feature number 309, PWY.5747..2.methylcitrate.cycle.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.640204 INFO::Fitting model to feature number 310, PWY.5367..petroselinate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.642591 INFO::Fitting model to feature number 311, PWY0.1533..methylphosphonate.degradation.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.644901 INFO::Fitting model to feature number 312, PWY.6823..molybdenum.cofactor.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.647164 INFO::Fitting model to feature number 313, PWY.5705..allantoin.degradation.to.glyoxylate.III</span></span>
<span><span class="co">##  2024-04-26 15:42:49.649434 INFO::Fitting model to feature number 314, LPSSYN.PWY..superpathway.of.lipopolysaccharide.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.651647 INFO::Fitting model to feature number 315, REDCITCYC..TCA.cycle.VIII..helicobacter.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.653901 INFO::Fitting model to feature number 316, GLYCOL.GLYOXDEG.PWY..superpathway.of.glycol.metabolism.and.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.656128 INFO::Fitting model to feature number 317, P122.PWY..heterolactic.fermentation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.658365 INFO::Fitting model to feature number 318, KDO.NAGLIPASYN.PWY..superpathway.of..Kdo.2.lipid.A.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.660551 INFO::Fitting model to feature number 319, PWY.7446..sulfoglycolysis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.662785 INFO::Fitting model to feature number 320, PWY490.3..nitrate.reduction.VI..assimilatory.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.665077 INFO::Fitting model to feature number 321, PWY.7209..superpathway.of.pyrimidine.ribonucleosides.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.667306 INFO::Fitting model to feature number 322, UBISYN.PWY..superpathway.of.ubiquinol.8.biosynthesis..prokaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.669586 INFO::Fitting model to feature number 323, PWY0.41..allantoin.degradation.IV..anaerobic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.67181 INFO::Fitting model to feature number 324, PWY.5855..ubiquinol.7.biosynthesis..prokaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.674057 INFO::Fitting model to feature number 325, PWY.5856..ubiquinol.9.biosynthesis..prokaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.676279 INFO::Fitting model to feature number 326, PWY.5857..ubiquinol.10.biosynthesis..prokaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.678538 INFO::Fitting model to feature number 327, PWY.6708..ubiquinol.8.biosynthesis..prokaryotic.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.680781 INFO::Fitting model to feature number 328, PWY.5692..allantoin.degradation.to.glyoxylate.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.683004 INFO::Fitting model to feature number 329, URDEGR.PWY..superpathway.of.allantoin.degradation.in.plants</span></span>
<span><span class="co">##  2024-04-26 15:42:49.685271 INFO::Fitting model to feature number 330, PWY0.1338..polymyxin.resistance</span></span>
<span><span class="co">##  2024-04-26 15:42:49.687489 INFO::Fitting model to feature number 331, P221.PWY..octane.oxidation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.689737 INFO::Fitting model to feature number 332, THREOCAT.PWY..superpathway.of.L.threonine.metabolism</span></span>
<span><span class="co">##  2024-04-26 15:42:49.691944 INFO::Fitting model to feature number 333, NONMEVIPP.PWY..methylerythritol.phosphate.pathway.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.694217 INFO::Fitting model to feature number 334, PWY.7560..methylerythritol.phosphate.pathway.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.696427 INFO::Fitting model to feature number 335, PWY.5121..superpathway.of.geranylgeranyl.diphosphate.biosynthesis.II..via.MEP.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.698663 INFO::Fitting model to feature number 336, PWY.7392..taxadiene.biosynthesis..engineered.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.700927 INFO::Fitting model to feature number 337, PWY.6143..CMP.pseudaminate.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.703162 INFO::Fitting model to feature number 338, PWY.6270..isoprene.biosynthesis.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.705406 INFO::Fitting model to feature number 339, ALLANTOINDEG.PWY..superpathway.of.allantoin.degradation.in.yeast</span></span>
<span><span class="co">##  2024-04-26 15:42:49.707614 INFO::Fitting model to feature number 340, PWY.6263..superpathway.of.menaquinol.8.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.709835 INFO::Fitting model to feature number 341, PWY.7371..1.4.dihydroxy.6.naphthoate.biosynthesis.II</span></span>
<span><span class="co">##  2024-04-26 15:42:49.712055 INFO::Fitting model to feature number 342, ARGORNPROST.PWY..arginine..ornithine.and.proline.interconversion</span></span>
<span><span class="co">##  2024-04-26 15:42:49.714294 INFO::Fitting model to feature number 343, PWY66.398..TCA.cycle.III..animals.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.716533 INFO::Fitting model to feature number 344, PWY.5464..superpathway.of.cytosolic.glycolysis..plants...pyruvate.dehydrogenase.and.TCA.cycle</span></span>
<span><span class="co">##  2024-04-26 15:42:49.718786 INFO::Fitting model to feature number 345, ORNDEG.PWY..superpathway.of.ornithine.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.721035 INFO::Fitting model to feature number 346, P23.PWY..reductive.TCA.cycle.I</span></span>
<span><span class="co">##  2024-04-26 15:42:49.723251 INFO::Fitting model to feature number 347, PWY0.1277..3.phenylpropanoate.and.3..3.hydroxyphenyl.propanoate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.725507 INFO::Fitting model to feature number 348, ARGDEG.PWY..superpathway.of.L.arginine..putrescine..and.4.aminobutanoate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.727755 INFO::Fitting model to feature number 349, ORNARGDEG.PWY..superpathway.of.L.arginine.and.L.ornithine.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.729995 INFO::Fitting model to feature number 350, HCAMHPDEG.PWY..3.phenylpropanoate.and.3..3.hydroxyphenyl.propanoate.degradation.to.2.oxopent.4.enoate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.732249 INFO::Fitting model to feature number 351, PWY.6690..cinnamate.and.3.hydroxycinnamate.degradation.to.2.oxopent.4.enoate</span></span>
<span><span class="co">##  2024-04-26 15:42:49.734512 INFO::Fitting model to feature number 352, PROTOCATECHUATE.ORTHO.CLEAVAGE.PWY..protocatechuate.degradation.II..ortho.cleavage.pathway.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.736758 INFO::Fitting model to feature number 353, P124.PWY..Bifidobacterium.shunt</span></span>
<span><span class="co">##  2024-04-26 15:42:49.738966 INFO::Fitting model to feature number 354, PWY.622..starch.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.7412 INFO::Fitting model to feature number 355, PWY.7269..NAD.NADP.NADH.NADPH.mitochondrial.interconversion..yeast.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.743402 INFO::Fitting model to feature number 356, PWY.7316..dTDP.N.acetylviosamine.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.745636 INFO::Fitting model to feature number 357, PWY.7279..aerobic.respiration.II..cytochrome.c...yeast.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.747862 INFO::Fitting model to feature number 358, PWY.6837..fatty.acid.beta.oxidation.V..unsaturated..odd.number..di.isomerase.dependent.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.750102 INFO::Fitting model to feature number 359, X3.HYDROXYPHENYLACETATE.DEGRADATION.PWY..4.hydroxyphenylacetate.degradation</span></span>
<span><span class="co">##  2024-04-26 15:42:49.7523 INFO::Fitting model to feature number 360, PWY.6953..dTDP.3.acetamido.3.6.dideoxy..alpha..D.galactose.biosynthesis</span></span>
<span><span class="co">##  2024-04-26 15:42:49.754526 INFO::Fitting model to feature number 361, PWY.7616..methanol.oxidation.to.carbon.dioxide</span></span>
<span><span class="co">##  2024-04-26 15:42:49.756768 INFO::Fitting model to feature number 362, PWY.5088..L.glutamate.degradation.VIII..to.propanoate.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.758993 INFO::Fitting model to feature number 363, PWY0.881..superpathway.of.fatty.acid.biosynthesis.I..E..coli.</span></span>
<span><span class="co">##  2024-04-26 15:42:49.807666 INFO::Counting total values for each feature</span></span>
<span><span class="co">##  2024-04-26 15:42:49.839122 INFO::Writing filtered data to file output_pwys/features/filtered_data.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:49.87765 INFO::Writing filtered, normalized data to file output_pwys/features/filtered_data_norm.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:49.915844 INFO::Writing filtered, normalized, transformed data to file output_pwys/features/filtered_data_norm_transformed.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:49.965539 INFO::Writing residuals to file output_pwys/fits/residuals.rds</span></span>
<span><span class="co">##  2024-04-26 15:42:49.989157 INFO::Writing fitted values to file output_pwys/fits/fitted.rds</span></span>
<span><span class="co">##  2024-04-26 15:42:50.005504 INFO::Writing all results to file (ordered by increasing q-values): output_pwys/all_results.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:50.011709 INFO::Writing the significant results (those which are less than or equal to the threshold of 0.250000 ) to file (ordered by increasing q-values): output_pwys/significant_results.tsv</span></span>
<span><span class="co">##  2024-04-26 15:42:50.012444 INFO::Writing heatmap of significant results to file: output_pwys/heatmap.pdf</span></span>
<span><span class="co">##  [1] "There are no associations to plot!"</span></span>
<span><span class="co">##  2024-04-26 15:42:50.013389 INFO::Writing association plots (one for each significant association) to output folder: output_pwys</span></span>
<span><span class="co">##  2024-04-26 15:42:50.016604 INFO::Plotting associations from most to least significant, grouped by metadata</span></span>
<span><span class="co">##  2024-04-26 15:42:50.017167 INFO::Plotting data for metadata number 1, antibiotics_current_use</span></span>
<span><span class="co">##  2024-04-26 15:42:50.018186 INFO::Creating boxplot for categorical data, antibiotics_current_use vs POLYAMINSYN3.PWY..superpathway.of.polyamine.biosynthesis.II</span></span>
<span></span>
<span><span class="co">##########################</span></span>
<span><span class="co"># Extract the residuals #</span></span>
<span><span class="co">##########################</span></span>
<span></span>
<span><span class="va">datExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html">t</a></span><span class="op">(</span><span class="va">fit_data</span><span class="op">$</span><span class="va">residuals</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">########################</span></span>
<span><span class="co"># Create WGCNA modules #</span></span>
<span><span class="co">########################</span></span>
<span></span>
<span><span class="va">gsg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/goodSamplesGenes.html">goodSamplesGenes</a></span><span class="op">(</span><span class="va">datExpr</span>, verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">##   Flagging genes and samples with too many missing values...</span></span>
<span><span class="co">##    ..step 1</span></span>
<span><span class="co">##    ..Excluding 6 samples from the calculation due to too many missing genes.</span></span>
<span><span class="co">##    ..step 2</span></span>
<span><span class="va">gsg</span><span class="op">$</span><span class="va">allOK</span></span>
<span><span class="co">##  [1] FALSE</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">allOK</span><span class="op">)</span></span>
<span><span class="op">{</span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodGenes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/dynamicTreeCut/man/printFlush.html">printFlush</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"Removing genes:"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">datExpr</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodGenes</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>;</span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodSamples</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dynamicTreeCut/man/printFlush.html">printFlush</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"Removing samples:"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">datExpr</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodSamples</span><span class="op">]</span>, collapse <span class="op">=</span><span class="st">", "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">datExpr</span> <span class="op">=</span> <span class="va">datExpr</span><span class="op">[</span><span class="va">gsg</span><span class="op">$</span><span class="va">goodSamples</span>, <span class="va">gsg</span><span class="op">$</span><span class="va">goodGenes</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="co">##  Removing samples: CSM5MCVB_P, CSM79HNY_P, ESM5GEYY_P, ESM718U9_P, MSM6J2N6_P, MSM9VZLX_P</span></span>
<span></span>
<span><span class="va">gsg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/goodSamplesGenes.html">goodSamplesGenes</a></span><span class="op">(</span><span class="va">datExpr</span>, verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">##   Flagging genes and samples with too many missing values...</span></span>
<span><span class="co">##    ..step 1</span></span>
<span><span class="va">gsg</span><span class="op">$</span><span class="va">allOK</span> <span class="co"># TRUE</span></span>
<span><span class="co">##  [1] TRUE</span></span>
<span></span>
<span><span class="co">###################################</span></span>
<span><span class="co"># Choose soft threshold parameter #</span></span>
<span><span class="co">###################################</span></span>
<span></span>
<span><span class="va">powers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">22</span>, to<span class="op">=</span><span class="fl">30</span>, by<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sft</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/pickSoftThreshold.html">pickSoftThreshold</a></span><span class="op">(</span><span class="va">datExpr</span>, powerVector <span class="op">=</span> <span class="va">powers</span>, verbose <span class="op">=</span> <span class="fl">5</span>, dataIsExpr <span class="op">=</span> <span class="cn">TRUE</span>, RsquaredCut <span class="op">=</span> <span class="fl">0.30</span><span class="op">)</span></span>
<span><span class="co">##  pickSoftThreshold: will use block size 363.</span></span>
<span><span class="co">##   pickSoftThreshold: calculating connectivity for given powers...</span></span>
<span><span class="co">##     ..working on genes 1 through 363 of 363</span></span>
<span><span class="co">##     Power SFT.R.sq   slope truncated.R.sq mean.k. median.k. max.k.</span></span>
<span><span class="co">##  1      1    0.722  1.8100         0.7110  124.00   130.000 162.00</span></span>
<span><span class="co">##  2      2    0.596  0.6920         0.6290   63.10    68.700  95.70</span></span>
<span><span class="co">##  3      3    0.393  0.2300         0.2310   38.40    39.700  69.30</span></span>
<span><span class="co">##  4      4    0.124 -0.0902         0.0162   25.90    24.800  55.00</span></span>
<span><span class="co">##  5      5    0.561 -0.3550         0.5860   18.70    16.300  46.60</span></span>
<span><span class="co">##  6      6    0.673 -0.5190         0.6680   14.20    11.400  40.40</span></span>
<span><span class="co">##  7      7    0.781 -0.6230         0.8100   11.20     8.380  35.60</span></span>
<span><span class="co">##  8      8    0.815 -0.7040         0.8450    9.07     6.520  31.80</span></span>
<span><span class="co">##  9      9    0.861 -0.8180         0.9260    7.54     5.100  28.60</span></span>
<span><span class="co">##  10    10    0.847 -0.8620         0.9070    6.39     4.100  26.00</span></span>
<span><span class="co">##  11    11    0.835 -0.9150         0.9140    5.50     3.430  23.80</span></span>
<span><span class="co">##  12    12    0.847 -0.9540         0.9130    4.80     2.870  21.90</span></span>
<span><span class="co">##  13    13    0.868 -0.9730         0.9370    4.24     2.380  20.20</span></span>
<span><span class="co">##  14    14    0.864 -0.9990         0.9160    3.78     2.060  18.70</span></span>
<span><span class="co">##  15    15    0.890 -1.0300         0.9590    3.40     1.720  17.40</span></span>
<span><span class="co">##  16    16    0.898 -1.0400         0.9670    3.09     1.530  16.20</span></span>
<span><span class="co">##  17    17    0.888 -1.0600         0.9570    2.82     1.370  15.20</span></span>
<span><span class="co">##  18    18    0.877 -1.0600         0.9290    2.59     1.280  14.20</span></span>
<span><span class="co">##  19    19    0.864 -1.0500         0.9170    2.40     1.180  13.30</span></span>
<span><span class="co">##  20    20    0.844 -1.0600         0.8690    2.23     1.100  12.50</span></span>
<span><span class="co">##  21    22    0.854 -1.0700         0.8890    1.95     0.997  11.20</span></span>
<span><span class="co">##  22    24    0.820 -1.0600         0.8720    1.74     0.899   9.98</span></span>
<span><span class="co">##  23    26    0.797 -0.9790         0.8720    1.57     0.793   8.97</span></span>
<span><span class="co">##  24    28    0.829 -0.9390         0.8720    1.43     0.718   8.10</span></span>
<span><span class="co">##  25    30    0.807 -0.9130         0.8140    1.31     0.673   7.35</span></span>
<span></span>
<span><span class="co">##############################</span></span>
<span><span class="co">#  One-step module detection #</span></span>
<span><span class="co">##############################</span></span>
<span></span>
<span><span class="va">power</span> <span class="op">=</span> <span class="va">sft</span><span class="op">$</span><span class="va">powerEstimate</span></span>
<span><span class="va">net</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html">blockwiseModules</a></span><span class="op">(</span><span class="va">datExpr</span>, </span>
<span>                       power <span class="op">=</span> <span class="va">power</span>,</span>
<span>                       corFnc <span class="op">=</span> <span class="st">"bicor"</span>, </span>
<span>                       corOptions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>maxPOutliers <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>, </span>
<span>                       networkType <span class="op">=</span><span class="st">"unsigned"</span>, </span>
<span>                       maxBlockSize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html">ncol</a></span><span class="op">(</span><span class="va">datExpr</span><span class="op">)</span>,</span>
<span>                       minModuleSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       TOMType <span class="op">=</span> <span class="st">"unsigned"</span>, </span>
<span>                       reassignThreshold <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                       mergeCutHeight <span class="op">=</span> <span class="fl">0</span>,                       </span>
<span>                       verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">##   Calculating module eigengenes block-wise from all genes</span></span>
<span><span class="co">##     Flagging genes and samples with too many missing values...</span></span>
<span><span class="co">##      ..step 1</span></span>
<span><span class="co">##   ..Working on block 1 .</span></span>
<span><span class="co">##      TOM calculation: adjacency..</span></span>
<span><span class="co">##      ..will not use multithreading.</span></span>
<span><span class="co">##       Fraction of slow calculations: 0.000000</span></span>
<span><span class="co">##      ..connectivity..</span></span>
<span><span class="co">##      ..matrix multiplication (system BLAS)..</span></span>
<span><span class="co">##      ..normalization..</span></span>
<span><span class="co">##      ..done.</span></span>
<span><span class="co">##   ....clustering..</span></span>
<span><span class="co">##   ....detecting modules..</span></span>
<span><span class="co">##   ....calculating module eigengenes..</span></span>
<span><span class="co">##   ....checking kME in modules..</span></span>
<span><span class="co">##       ..removing 3 genes from module 1 because their KME is too low.</span></span>
<span><span class="co">##       ..removing 11 genes from module 2 because their KME is too low.</span></span>
<span><span class="co">##       ..removing 2 genes from module 3 because their KME is too low.</span></span>
<span><span class="co">##   ..merging modules that are too close..</span></span>
<span><span class="co">##       mergeCloseModules: Merging modules whose distance is less than 0</span></span>
<span><span class="co">##         Calculating new MEs...</span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># How many modules #</span></span>
<span><span class="co">####################</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html">ncol</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">MEs</span><span class="op">)</span></span>
<span><span class="co">##  [1] 4</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##       blue     brown      grey turquoise </span></span>
<span><span class="co">##        139        18        16       190</span></span>
<span></span>
<span><span class="co">##########################</span></span>
<span><span class="co"># Plot module dendrogram #</span></span>
<span><span class="co">##########################</span></span>
<span></span>
<span><span class="va">eigenGenes</span> <span class="op">&lt;-</span> <span class="va">net</span><span class="op">$</span><span class="va">MEs</span></span>
<span><span class="va">MEDiss</span> <span class="op">=</span> <span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/cor.html">cor</a></span><span class="op">(</span><span class="va">eigenGenes</span><span class="op">)</span></span>
<span><span class="va">METree</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/fastcluster/man/hclust.html">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="va">MEDiss</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">METree</span>, main <span class="op">=</span> <span class="st">"Clustering of module eigengenes"</span>, xlab <span class="op">=</span> <span class="st">""</span>, sub <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MSEA_files/figure-html/MSEA-preparation-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co">###########################################</span></span>
<span><span class="co"># Re-calculate modules and find hub genes #</span></span>
<span><span class="co">###########################################</span></span>
<span></span>
<span><span class="va">moduleColors</span> <span class="op">&lt;-</span> <span class="va">net</span><span class="op">$</span><span class="va">colors</span></span>
<span><span class="va">MEs0</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/moduleEigengenes.html">moduleEigengenes</a></span><span class="op">(</span><span class="va">datExpr</span>, <span class="va">moduleColors</span><span class="op">)</span><span class="op">$</span><span class="va">eigengenes</span></span>
<span><span class="va">modules_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/orderMEs.html">orderMEs</a></span><span class="op">(</span><span class="va">MEs0</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#######################</span></span>
<span><span class="co"># Create mapping file #</span></span>
<span><span class="co">#######################</span></span>
<span></span>
<span><span class="va">feature_by_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span>
<span><span class="va">feature_by_modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">feature_by_modules</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">feature_by_modules</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Feature'</span>, <span class="st">'Module'</span><span class="op">)</span></span>
<span><span class="va">features_mapping</span> <span class="op">&lt;-</span> <span class="va">feature_by_modules</span></span>
<span><span class="va">features_mapping</span><span class="op">$</span><span class="va">Module</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">'ME'</span>, <span class="va">features_mapping</span><span class="op">$</span><span class="va">Module</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We perform the MSEA as before using the modules from the WGCNA analysis on the pathways.</p>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/perform-MSEA-and-visualization_1ae05d08a7c05fd2b962a4cfd8148161">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">###################</span></span>
<span><span class="co"># Rank DA results #</span></span>
<span><span class="co">###################</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="va">fit_data</span><span class="op">$</span><span class="va">results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">==</span><span class="st">'disease'</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">pval</span>, <span class="st">'BH'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html">order</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">qval</span>, decreasing <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co">###################</span></span>
<span><span class="co"># MSEA Processing #</span></span>
<span><span class="co">##################</span></span>
<span></span>
<span><span class="va">module_map</span> <span class="op">&lt;-</span> <span class="va">features_mapping</span></span>
<span><span class="va">mod.gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/tapply.html">tapply</a></span><span class="op">(</span><span class="va">module_map</span><span class="op">$</span><span class="va">Module</span>, <span class="va">module_map</span><span class="op">$</span><span class="va">Feature</span>, <span class="va">as.character</span><span class="op">)</span></span>
<span><span class="va">microbeSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/topGO/man/annFUN.html">inverseList</a></span><span class="op">(</span><span class="va">mod.gs</span><span class="op">)</span></span>
<span><span class="va">microbeSet</span></span>
<span><span class="co">##  $MEblue</span></span>
<span><span class="co">##    [1] "ANAEROFRUCAT.PWY..homolactic.fermentation"                                                       </span></span>
<span><span class="co">##    [2] "ANAGLYCOLYSIS.PWY..glycolysis.III..from.glucose."                                                </span></span>
<span><span class="co">##    [3] "ARGININE.SYN4.PWY..L.ornithine.de.novo..biosynthesis"                                            </span></span>
<span><span class="co">##    [4] "ARGSYN.PWY..L.arginine.biosynthesis.I..via.L.ornithine."                                         </span></span>
<span><span class="co">##    [5] "ARGSYNBSUB.PWY..L.arginine.biosynthesis.II..acetyl.cycle."                                       </span></span>
<span><span class="co">##    [6] "ARO.PWY..chorismate.biosynthesis.I"                                                              </span></span>
<span><span class="co">##    [7] "ASPASN.PWY..superpathway.of.L.aspartate.and.L.asparagine.biosynthesis"                           </span></span>
<span><span class="co">##    [8] "BRANCHED.CHAIN.AA.SYN.PWY..superpathway.of.branched.amino.acid.biosynthesis"                     </span></span>
<span><span class="co">##    [9] "CALVIN.PWY..Calvin.Benson.Bassham.cycle"                                                         </span></span>
<span><span class="co">##   [10] "CITRULBIO.PWY..L.citrulline.biosynthesis"                                                        </span></span>
<span><span class="co">##   [11] "COA.PWY..coenzyme.A.biosynthesis.I"                                                              </span></span>
<span><span class="co">##   [12] "COA.PWY.1..coenzyme.A.biosynthesis.II..mammalian."                                               </span></span>
<span><span class="co">##   [13] "COBALSYN.PWY..adenosylcobalamin.salvage.from.cobinamide.I"                                       </span></span>
<span><span class="co">##   [14] "COLANSYN.PWY..colanic.acid.building.blocks.biosynthesis"                                         </span></span>
<span><span class="co">##   [15] "COMPLETE.ARO.PWY..superpathway.of.aromatic.amino.acid.biosynthesis"                              </span></span>
<span><span class="co">##   [16] "DENOVOPURINE2.PWY..superpathway.of.purine.nucleotides.de.novo.biosynthesis.II"                   </span></span>
<span><span class="co">##   [17] "DTDPRHAMSYN.PWY..dTDP.L.rhamnose.biosynthesis.I"                                                 </span></span>
<span><span class="co">##   [18] "GALACT.GLUCUROCAT.PWY..superpathway.of.hexuronide.and.hexuronate.degradation"                    </span></span>
<span><span class="co">##   [19] "GALACTUROCAT.PWY..D.galacturonate.degradation.I"                                                 </span></span>
<span><span class="co">##   [20] "GLUCONEO.PWY..gluconeogenesis.I"                                                                 </span></span>
<span><span class="co">##   [21] "GLUCUROCAT.PWY..superpathway.of..beta..D.glucuronide.and.D.glucuronate.degradation"              </span></span>
<span><span class="co">##   [22] "GLUTORN.PWY..L.ornithine.biosynthesis"                                                           </span></span>
<span><span class="co">##   [23] "GLYCOGENSYNTH.PWY..glycogen.biosynthesis.I..from.ADP.D.Glucose."                                 </span></span>
<span><span class="co">##   [24] "GLYCOLYSIS..glycolysis.I..from.glucose.6.phosphate."                                             </span></span>
<span><span class="co">##   [25] "HISDEG.PWY..L.histidine.degradation.I"                                                           </span></span>
<span><span class="co">##   [26] "HISTSYN.PWY..L.histidine.biosynthesis"                                                           </span></span>
<span><span class="co">##   [27] "HSERMETANA.PWY..L.methionine.biosynthesis.III"                                                   </span></span>
<span><span class="co">##   [28] "ILEUSYN.PWY..L.isoleucine.biosynthesis.I..from.threonine."                                       </span></span>
<span><span class="co">##   [29] "NAGLIPASYN.PWY..lipid.IVA.biosynthesis"                                                          </span></span>
<span><span class="co">##   [30] "NONOXIPENT.PWY..pentose.phosphate.pathway..non.oxidative.branch."                                </span></span>
<span><span class="co">##   [31] "OANTIGEN.PWY..O.antigen.building.blocks.biosynthesis..E..coli."                                  </span></span>
<span><span class="co">##   [32] "P108.PWY..pyruvate.fermentation.to.propanoate.I"                                                 </span></span>
<span><span class="co">##   [33] "PANTO.PWY..phosphopantothenate.biosynthesis.I"                                                   </span></span>
<span><span class="co">##   [34] "PANTOSYN.PWY..pantothenate.and.coenzyme.A.biosynthesis.I"                                        </span></span>
<span><span class="co">##   [35] "PENTOSE.P.PWY..pentose.phosphate.pathway"                                                        </span></span>
<span><span class="co">##   [36] "PEPTIDOGLYCANSYN.PWY..peptidoglycan.biosynthesis.I..meso.diaminopimelate.containing."            </span></span>
<span><span class="co">##   [37] "PHOSLIPSYN.PWY..superpathway.of.phospholipid.biosynthesis.I..bacteria."                          </span></span>
<span><span class="co">##   [38] "POLYISOPRENSYN.PWY..polyisoprenoid.biosynthesis..E..coli."                                       </span></span>
<span><span class="co">##   [39] "PWY.1042..glycolysis.IV..plant.cytosol."                                                         </span></span>
<span><span class="co">##   [40] "PWY.1269..CMP.3.deoxy.D.manno.octulosonate.biosynthesis.I"                                       </span></span>
<span><span class="co">##   [41] "PWY.2941..L.lysine.biosynthesis.II"                                                              </span></span>
<span><span class="co">##   [42] "PWY.2942..L.lysine.biosynthesis.III"                                                             </span></span>
<span><span class="co">##   [43] "PWY.3001..superpathway.of.L.isoleucine.biosynthesis.I"                                           </span></span>
<span><span class="co">##   [44] "PWY.3841..folate.transformations.II"                                                             </span></span>
<span><span class="co">##   [45] "PWY.4242..pantothenate.and.coenzyme.A.biosynthesis.III"                                          </span></span>
<span><span class="co">##   [46] "PWY.4981..L.proline.biosynthesis.II..from.arginine."                                             </span></span>
<span><span class="co">##   [47] "PWY.4984..urea.cycle"                                                                            </span></span>
<span><span class="co">##   [48] "PWY.5030..L.histidine.degradation.III"                                                           </span></span>
<span><span class="co">##   [49] "PWY.5097..L.lysine.biosynthesis.VI"                                                              </span></span>
<span><span class="co">##   [50] "PWY.5103..L.isoleucine.biosynthesis.III"                                                         </span></span>
<span><span class="co">##   [51] "PWY.5154..L.arginine.biosynthesis.III..via.N.acetyl.L.citrulline."                               </span></span>
<span><span class="co">##   [52] "PWY.5304..superpathway.of.sulfur.oxidation..Acidianus.ambivalens."                               </span></span>
<span><span class="co">##   [53] "PWY.5484..glycolysis.II..from.fructose.6.phosphate."                                             </span></span>
<span><span class="co">##   [54] "PWY.5505..L.glutamate.and.L.glutamine.biosynthesis"                                              </span></span>
<span><span class="co">##   [55] "PWY.5659..GDP.mannose.biosynthesis"                                                              </span></span>
<span><span class="co">##   [56] "PWY.5667..CDP.diacylglycerol.biosynthesis.I"                                                     </span></span>
<span><span class="co">##   [57] "PWY.5686..UMP.biosynthesis"                                                                      </span></span>
<span><span class="co">##   [58] "PWY.5695..urate.biosynthesis.inosine.5..phosphate.degradation"                                   </span></span>
<span><span class="co">##   [59] "PWY.5941..glycogen.degradation.II..eukaryotic."                                                  </span></span>
<span><span class="co">##   [60] "PWY.5973..cis.vaccenate.biosynthesis"                                                            </span></span>
<span><span class="co">##   [61] "PWY.6121..5.aminoimidazole.ribonucleotide.biosynthesis.I"                                        </span></span>
<span><span class="co">##   [62] "PWY.6122..5.aminoimidazole.ribonucleotide.biosynthesis.II"                                       </span></span>
<span><span class="co">##   [63] "PWY.6123..inosine.5..phosphate.biosynthesis.I"                                                   </span></span>
<span><span class="co">##   [64] "PWY.6124..inosine.5..phosphate.biosynthesis.II"                                                  </span></span>
<span><span class="co">##   [65] "PWY.6125..superpathway.of.guanosine.nucleotides.de.novo.biosynthesis.II"                         </span></span>
<span><span class="co">##   [66] "PWY.6126..superpathway.of.adenosine.nucleotides.de.novo.biosynthesis.II"                         </span></span>
<span><span class="co">##   [67] "PWY.6147..6.hydroxymethyl.dihydropterin.diphosphate.biosynthesis.I"                              </span></span>
<span><span class="co">##   [68] "PWY.6151..S.adenosyl.L.methionine.cycle.I"                                                       </span></span>
<span><span class="co">##   [69] "PWY.6163..chorismate.biosynthesis.from.3.dehydroquinate"                                         </span></span>
<span><span class="co">##   [70] "PWY.6168..flavin.biosynthesis.III..fungi."                                                       </span></span>
<span><span class="co">##   [71] "PWY.621..sucrose.degradation.III..sucrose.invertase."                                            </span></span>
<span><span class="co">##   [72] "PWY.6277..superpathway.of.5.aminoimidazole.ribonucleotide.biosynthesis"                          </span></span>
<span><span class="co">##   [73] "PWY.6305..putrescine.biosynthesis.IV"                                                            </span></span>
<span><span class="co">##   [74] "PWY.6317..galactose.degradation.I..Leloir.pathway."                                              </span></span>
<span><span class="co">##   [75] "PWY.6385..peptidoglycan.biosynthesis.III..mycobacteria."                                         </span></span>
<span><span class="co">##   [76] "PWY.6386..UDP.N.acetylmuramoyl.pentapeptide.biosynthesis.II..lysine.containing."                 </span></span>
<span><span class="co">##   [77] "PWY.6387..UDP.N.acetylmuramoyl.pentapeptide.biosynthesis.I..meso.diaminopimelate.containing."    </span></span>
<span><span class="co">##   [78] "PWY.6507..4.deoxy.L.threo.hex.4.enopyranuronate.degradation"                                     </span></span>
<span><span class="co">##   [79] "PWY.6545..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.III"                              </span></span>
<span><span class="co">##   [80] "PWY.6572..chondroitin.sulfate.degradation.I..bacterial."                                         </span></span>
<span><span class="co">##   [81] "PWY.6609..adenine.and.adenosine.salvage.III"                                                     </span></span>
<span><span class="co">##   [82] "PWY.6700..queuosine.biosynthesis"                                                                </span></span>
<span><span class="co">##   [83] "PWY.6703..preQ0.biosynthesis"                                                                    </span></span>
<span><span class="co">##   [84] "PWY.6737..starch.degradation.V"                                                                  </span></span>
<span><span class="co">##   [85] "PWY.6859..all.trans.farnesol.biosynthesis"                                                       </span></span>
<span><span class="co">##   [86] "PWY.6892..thiazole.biosynthesis.I..E..coli."                                                     </span></span>
<span><span class="co">##   [87] "PWY.6897..thiamin.salvage.II"                                                                    </span></span>
<span><span class="co">##   [88] "PWY.6901..superpathway.of.glucose.and.xylose.degradation"                                        </span></span>
<span><span class="co">##   [89] "PWY.6936..seleno.amino.acid.biosynthesis"                                                        </span></span>
<span><span class="co">##   [90] "PWY.7111..pyruvate.fermentation.to.isobutanol..engineered."                                      </span></span>
<span><span class="co">##   [91] "PWY.7184..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.I"                                </span></span>
<span><span class="co">##   [92] "PWY.7187..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.II"                               </span></span>
<span><span class="co">##   [93] "PWY.7197..pyrimidine.deoxyribonucleotide.phosphorylation"                                        </span></span>
<span><span class="co">##   [94] "PWY.7199..pyrimidine.deoxyribonucleosides.salvage"                                               </span></span>
<span><span class="co">##   [95] "PWY.7208..superpathway.of.pyrimidine.nucleobases.salvage"                                        </span></span>
<span><span class="co">##   [96] "PWY.7210..pyrimidine.deoxyribonucleotides.biosynthesis.from.CTP"                                 </span></span>
<span><span class="co">##   [97] "PWY.7211..superpathway.of.pyrimidine.deoxyribonucleotides.de.novo.biosynthesis"                  </span></span>
<span><span class="co">##   [98] "PWY.7219..adenosine.ribonucleotides.de.novo.biosynthesis"                                        </span></span>
<span><span class="co">##   [99] "PWY.7220..adenosine.deoxyribonucleotides.de.novo.biosynthesis.II"                                </span></span>
<span><span class="co">##  [100] "PWY.7221..guanosine.ribonucleotides.de.novo.biosynthesis"                                        </span></span>
<span><span class="co">##  [101] "PWY.7222..guanosine.deoxyribonucleotides.de.novo.biosynthesis.II"                                </span></span>
<span><span class="co">##  [102] "PWY.7228..superpathway.of.guanosine.nucleotides.de.novo.biosynthesis.I"                          </span></span>
<span><span class="co">##  [103] "PWY.7229..superpathway.of.adenosine.nucleotides.de.novo.biosynthesis.I"                          </span></span>
<span><span class="co">##  [104] "PWY.7234..inosine.5..phosphate.biosynthesis.III"                                                 </span></span>
<span><span class="co">##  [105] "PWY.724..superpathway.of.L.lysine..L.threonine.and.L.methionine.biosynthesis.II"                 </span></span>
<span><span class="co">##  [106] "PWY.7242..D.fructuronate.degradation"                                                            </span></span>
<span><span class="co">##  [107] "PWY.7282..4.amino.2.methyl.5.phosphomethylpyrimidine.biosynthesis..yeast."                       </span></span>
<span><span class="co">##  [108] "PWY.7323..superpathway.of.GDP.mannose.derived.O.antigen.building.blocks.biosynthesis"            </span></span>
<span><span class="co">##  [109] "PWY.7332..superpathway.of.UDP.N.acetylglucosamine.derived.O.antigen.building.blocks.biosynthesis"</span></span>
<span><span class="co">##  [110] "PWY.7357..thiamin.formation.from.pyrithiamine.and.oxythiamine..yeast."                           </span></span>
<span><span class="co">##  [111] "PWY.7383..anaerobic.energy.metabolism..invertebrates..cytosol."                                  </span></span>
<span><span class="co">##  [112] "PWY.7400..L.arginine.biosynthesis.IV..archaebacteria."                                           </span></span>
<span><span class="co">##  [113] "PWY.7456..mannan.degradation"                                                                    </span></span>
<span><span class="co">##  [114] "PWY.7539..6.hydroxymethyl.dihydropterin.diphosphate.biosynthesis.III..Chlamydia."                </span></span>
<span><span class="co">##  [115] "PWY.7663..gondoate.biosynthesis..anaerobic."                                                     </span></span>
<span><span class="co">##  [116] "PWY.841..superpathway.of.purine.nucleotides.de.novo.biosynthesis.I"                              </span></span>
<span><span class="co">##  [117] "PWY0.1296..purine.ribonucleosides.degradation"                                                   </span></span>
<span><span class="co">##  [118] "PWY0.1319..CDP.diacylglycerol.biosynthesis.II"                                                   </span></span>
<span><span class="co">##  [119] "PWY0.1586..peptidoglycan.maturation..meso.diaminopimelate.containing."                           </span></span>
<span><span class="co">##  [120] "PWY0.162..superpathway.of.pyrimidine.ribonucleotides.de.novo.biosynthesis"                       </span></span>
<span><span class="co">##  [121] "PWY0.166..superpathway.of.pyrimidine.deoxyribonucleotides.de.novo.biosynthesis..E..coli."        </span></span>
<span><span class="co">##  [122] "PWY0.845..superpathway.of.pyridoxal.5..phosphate.biosynthesis.and.salvage"                       </span></span>
<span><span class="co">##  [123] "PWY4FS.7..phosphatidylglycerol.biosynthesis.I..plastidic."                                       </span></span>
<span><span class="co">##  [124] "PWY4FS.8..phosphatidylglycerol.biosynthesis.II..non.plastidic."                                  </span></span>
<span><span class="co">##  [125] "PWY66.399..gluconeogenesis.III"                                                                  </span></span>
<span><span class="co">##  [126] "PWY66.422..D.galactose.degradation.V..Leloir.pathway."                                           </span></span>
<span><span class="co">##  [127] "PYRIDNUCSYN.PWY..NAD.biosynthesis.I..from.aspartate."                                            </span></span>
<span><span class="co">##  [128] "PYRIDOXSYN.PWY..pyridoxal.5..phosphate.biosynthesis.I"                                           </span></span>
<span><span class="co">##  [129] "RHAMCAT.PWY..L.rhamnose.degradation.I"                                                           </span></span>
<span><span class="co">##  [130] "RIBOSYN2.PWY..flavin.biosynthesis.I..bacteria.and.plants."                                       </span></span>
<span><span class="co">##  [131] "SER.GLYSYN.PWY..superpathway.of.L.serine.and.glycine.biosynthesis.I"                             </span></span>
<span><span class="co">##  [132] "TEICHOICACID.PWY..teichoic.acid..poly.glycerol..biosynthesis"                                    </span></span>
<span><span class="co">##  [133] "THISYN.PWY..superpathway.of.thiamin.diphosphate.biosynthesis.I"                                  </span></span>
<span><span class="co">##  [134] "THISYNARA.PWY..superpathway.of.thiamin.diphosphate.biosynthesis.III..eukaryotes."                </span></span>
<span><span class="co">##  [135] "THRESYN.PWY..superpathway.of.L.threonine.biosynthesis"                                           </span></span>
<span><span class="co">##  [136] "TRNA.CHARGING.PWY..tRNA.charging"                                                                </span></span>
<span><span class="co">##  [137] "TRPSYN.PWY..L.tryptophan.biosynthesis"                                                           </span></span>
<span><span class="co">##  [138] "VALSYN.PWY..L.valine.biosynthesis"                                                               </span></span>
<span><span class="co">##  [139] "X1CMET2.PWY..N10.formyl.tetrahydrofolate.biosynthesis"                                           </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEbrown</span></span>
<span><span class="co">##   [1] "GLCMANNANAUT.PWY..superpathway.of.N.acetylglucosamine..N.acetylmannosamine.and.N.acetylneuraminate.degradation"</span></span>
<span><span class="co">##   [2] "GLYCOLYSIS.E.D..superpathway.of.glycolysis.and.Entner.Doudoroff"                                               </span></span>
<span><span class="co">##   [3] "METH.ACETATE.PWY..methanogenesis.from.acetate"                                                                 </span></span>
<span><span class="co">##   [4] "PWY.5100..pyruvate.fermentation.to.acetate.and.lactate.II"                                                     </span></span>
<span><span class="co">##   [5] "PWY.5104..L.isoleucine.biosynthesis.IV"                                                                        </span></span>
<span><span class="co">##   [6] "PWY.5177..glutaryl.CoA.degradation"                                                                            </span></span>
<span><span class="co">##   [7] "PWY.5690..TCA.cycle.II..plants.and.fungi."                                                                     </span></span>
<span><span class="co">##   [8] "PWY.6470..peptidoglycan.biosynthesis.V...beta..lactam.resistance."                                             </span></span>
<span><span class="co">##   [9] "PWY.6527..stachyose.degradation"                                                                               </span></span>
<span><span class="co">##  [10] "PWY.6595..superpathway.of.guanosine.nucleotides.degradation..plants."                                          </span></span>
<span><span class="co">##  [11] "PWY.6608..guanosine.nucleotides.degradation.III"                                                               </span></span>
<span><span class="co">##  [12] "PWY.7196..superpathway.of.pyrimidine.ribonucleosides.salvage"                                                  </span></span>
<span><span class="co">##  [13] "PWY.7198..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.IV"                                             </span></span>
<span><span class="co">##  [14] "PWY0.1061..superpathway.of.L.alanine.biosynthesis"                                                             </span></span>
<span><span class="co">##  [15] "PWY0.1261..anhydromuropeptides.recycling"                                                                      </span></span>
<span><span class="co">##  [16] "PWY66.400..glycolysis.VI..metazoan."                                                                           </span></span>
<span><span class="co">##  [17] "TCA..TCA.cycle.I..prokaryotic."                                                                                </span></span>
<span><span class="co">##  [18] "UDPNAGSYN.PWY..UDP.N.acetyl.D.glucosamine.biosynthesis.I"                                                      </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEgrey</span></span>
<span><span class="co">##   [1] "NAD.BIOSYNTHESIS.II..NAD.salvage.pathway.II"                            </span></span>
<span><span class="co">##   [2] "P162.PWY..L.glutamate.degradation.V..via.hydroxyglutarate."             </span></span>
<span><span class="co">##   [3] "P163.PWY..L.lysine.fermentation.to.acetate.and.butanoate"               </span></span>
<span><span class="co">##   [4] "PWY.3781..aerobic.respiration.I..cytochrome.c."                         </span></span>
<span><span class="co">##   [5] "PWY.6143..CMP.pseudaminate.biosynthesis"                                </span></span>
<span><span class="co">##   [6] "PWY.622..starch.biosynthesis"                                           </span></span>
<span><span class="co">##   [7] "PWY.6263..superpathway.of.menaquinol.8.biosynthesis.II"                 </span></span>
<span><span class="co">##   [8] "PWY.6478..GDP.D.glycero..alpha..D.manno.heptose.biosynthesis"           </span></span>
<span><span class="co">##   [9] "PWY.6749..CMP.legionaminate.biosynthesis.I"                             </span></span>
<span><span class="co">##  [10] "PWY.6953..dTDP.3.acetamido.3.6.dideoxy..alpha..D.galactose.biosynthesis"</span></span>
<span><span class="co">##  [11] "PWY.7237..myo...chiro..and.scillo.inositol.degradation"                 </span></span>
<span><span class="co">##  [12] "PWY.7312..dTDP.D..beta..fucofuranose.biosynthesis"                      </span></span>
<span><span class="co">##  [13] "PWY.7316..dTDP.N.acetylviosamine.biosynthesis"                          </span></span>
<span><span class="co">##  [14] "PWY.7371..1.4.dihydroxy.6.naphthoate.biosynthesis.II"                   </span></span>
<span><span class="co">##  [15] "PWY490.3..nitrate.reduction.VI..assimilatory."                          </span></span>
<span><span class="co">##  [16] "RUMP.PWY..formaldehyde.oxidation.I"                                     </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $MEturquoise</span></span>
<span><span class="co">##    [1] "ALLANTOINDEG.PWY..superpathway.of.allantoin.degradation.in.yeast"                                           </span></span>
<span><span class="co">##    [2] "ARG.POLYAMINE.SYN..superpathway.of.arginine.and.polyamine.biosynthesis"                                     </span></span>
<span><span class="co">##    [3] "ARGDEG.PWY..superpathway.of.L.arginine..putrescine..and.4.aminobutanoate.degradation"                       </span></span>
<span><span class="co">##    [4] "ARGORNPROST.PWY..arginine..ornithine.and.proline.interconversion"                                           </span></span>
<span><span class="co">##    [5] "AST.PWY..L.arginine.degradation.II..AST.pathway."                                                           </span></span>
<span><span class="co">##    [6] "BIOTIN.BIOSYNTHESIS.PWY..biotin.biosynthesis.I"                                                             </span></span>
<span><span class="co">##    [7] "CENTFERM.PWY..pyruvate.fermentation.to.butanoate"                                                           </span></span>
<span><span class="co">##    [8] "DAPLYSINESYN.PWY..L.lysine.biosynthesis.I"                                                                  </span></span>
<span><span class="co">##    [9] "ECASYN.PWY..enterobacterial.common.antigen.biosynthesis"                                                    </span></span>
<span><span class="co">##   [10] "ENTBACSYN.PWY..enterobactin.biosynthesis"                                                                   </span></span>
<span><span class="co">##   [11] "FAO.PWY..fatty.acid..beta..oxidation.I"                                                                     </span></span>
<span><span class="co">##   [12] "FASYN.ELONG.PWY..fatty.acid.elongation....saturated"                                                        </span></span>
<span><span class="co">##   [13] "FASYN.INITIAL.PWY..superpathway.of.fatty.acid.biosynthesis.initiation..E..coli."                            </span></span>
<span><span class="co">##   [14] "FERMENTATION.PWY..mixed.acid.fermentation"                                                                  </span></span>
<span><span class="co">##   [15] "FOLSYN.PWY..superpathway.of.tetrahydrofolate.biosynthesis.and.salvage"                                      </span></span>
<span><span class="co">##   [16] "FUC.RHAMCAT.PWY..superpathway.of.fucose.and.rhamnose.degradation"                                           </span></span>
<span><span class="co">##   [17] "FUCCAT.PWY..fucose.degradation"                                                                             </span></span>
<span><span class="co">##   [18] "GALACTARDEG.PWY..D.galactarate.degradation.I"                                                               </span></span>
<span><span class="co">##   [19] "GLUCARDEG.PWY..D.glucarate.degradation.I"                                                                   </span></span>
<span><span class="co">##   [20] "GLUCARGALACTSUPER.PWY..superpathway.of.D.glucarate.and.D.galactarate.degradation"                           </span></span>
<span><span class="co">##   [21] "GLUCOSE1PMETAB.PWY..glucose.and.glucose.1.phosphate.degradation"                                            </span></span>
<span><span class="co">##   [22] "GLUDEG.I.PWY..GABA.shunt"                                                                                   </span></span>
<span><span class="co">##   [23] "GLYCOCAT.PWY..glycogen.degradation.I..bacterial."                                                           </span></span>
<span><span class="co">##   [24] "GLYCOL.GLYOXDEG.PWY..superpathway.of.glycol.metabolism.and.degradation"                                     </span></span>
<span><span class="co">##   [25] "GLYCOLYSIS.TCA.GLYOX.BYPASS..superpathway.of.glycolysis..pyruvate.dehydrogenase..TCA..and.glyoxylate.bypass"</span></span>
<span><span class="co">##   [26] "GLYOXYLATE.BYPASS..glyoxylate.cycle"                                                                        </span></span>
<span><span class="co">##   [27] "GOLPDLCAT.PWY..superpathway.of.glycerol.degradation.to.1.3.propanediol"                                     </span></span>
<span><span class="co">##   [28] "HCAMHPDEG.PWY..3.phenylpropanoate.and.3..3.hydroxyphenyl.propanoate.degradation.to.2.oxopent.4.enoate"      </span></span>
<span><span class="co">##   [29] "HEME.BIOSYNTHESIS.II..heme.biosynthesis.I..aerobic."                                                        </span></span>
<span><span class="co">##   [30] "HEMESYN2.PWY..heme.biosynthesis.II..anaerobic."                                                             </span></span>
<span><span class="co">##   [31] "HEXITOLDEGSUPER.PWY..superpathway.of.hexitol.degradation..bacteria."                                        </span></span>
<span><span class="co">##   [32] "HOMOSER.METSYN.PWY..L.methionine.biosynthesis.I"                                                            </span></span>
<span><span class="co">##   [33] "KDO.NAGLIPASYN.PWY..superpathway.of..Kdo.2.lipid.A.biosynthesis"                                            </span></span>
<span><span class="co">##   [34] "KETOGLUCONMET.PWY..ketogluconate.metabolism"                                                                </span></span>
<span><span class="co">##   [35] "LACTOSECAT.PWY..lactose.and.galactose.degradation.I"                                                        </span></span>
<span><span class="co">##   [36] "LPSSYN.PWY..superpathway.of.lipopolysaccharide.biosynthesis"                                                </span></span>
<span><span class="co">##   [37] "MET.SAM.PWY..superpathway.of.S.adenosyl.L.methionine.biosynthesis"                                          </span></span>
<span><span class="co">##   [38] "METHGLYUT.PWY..superpathway.of.methylglyoxal.degradation"                                                   </span></span>
<span><span class="co">##   [39] "METSYN.PWY..L.homoserine.and.L.methionine.biosynthesis"                                                     </span></span>
<span><span class="co">##   [40] "NONMEVIPP.PWY..methylerythritol.phosphate.pathway.I"                                                        </span></span>
<span><span class="co">##   [41] "ORNARGDEG.PWY..superpathway.of.L.arginine.and.L.ornithine.degradation"                                      </span></span>
<span><span class="co">##   [42] "ORNDEG.PWY..superpathway.of.ornithine.degradation"                                                          </span></span>
<span><span class="co">##   [43] "P105.PWY..TCA.cycle.IV..2.oxoglutarate.decarboxylase."                                                      </span></span>
<span><span class="co">##   [44] "P122.PWY..heterolactic.fermentation"                                                                        </span></span>
<span><span class="co">##   [45] "P124.PWY..Bifidobacterium.shunt"                                                                            </span></span>
<span><span class="co">##   [46] "P161.PWY..acetylene.degradation"                                                                            </span></span>
<span><span class="co">##   [47] "P164.PWY..purine.nucleobases.degradation.I..anaerobic."                                                     </span></span>
<span><span class="co">##   [48] "P185.PWY..formaldehyde.assimilation.III..dihydroxyacetone.cycle."                                           </span></span>
<span><span class="co">##   [49] "P221.PWY..octane.oxidation"                                                                                 </span></span>
<span><span class="co">##   [50] "P23.PWY..reductive.TCA.cycle.I"                                                                             </span></span>
<span><span class="co">##   [51] "P4.PWY..superpathway.of.L.lysine..L.threonine.and.L.methionine.biosynthesis.I"                              </span></span>
<span><span class="co">##   [52] "P42.PWY..incomplete.reductive.TCA.cycle"                                                                    </span></span>
<span><span class="co">##   [53] "P441.PWY..superpathway.of.N.acetylneuraminate.degradation"                                                  </span></span>
<span><span class="co">##   [54] "P461.PWY..hexitol.fermentation.to.lactate..formate..ethanol.and.acetate"                                    </span></span>
<span><span class="co">##   [55] "POLYAMINSYN3.PWY..superpathway.of.polyamine.biosynthesis.II"                                                </span></span>
<span><span class="co">##   [56] "POLYAMSYN.PWY..superpathway.of.polyamine.biosynthesis.I"                                                    </span></span>
<span><span class="co">##   [57] "PPGPPMET.PWY..ppGpp.biosynthesis"                                                                           </span></span>
<span><span class="co">##   [58] "PROTOCATECHUATE.ORTHO.CLEAVAGE.PWY..protocatechuate.degradation.II..ortho.cleavage.pathway."                </span></span>
<span><span class="co">##   [59] "PRPP.PWY..superpathway.of.histidine..purine..and.pyrimidine.biosynthesis"                                   </span></span>
<span><span class="co">##   [60] "PWY.1861..formaldehyde.assimilation.II..RuMP.Cycle."                                                        </span></span>
<span><span class="co">##   [61] "PWY.241..C4.photosynthetic.carbon.assimilation.cycle..NADP.ME.type"                                         </span></span>
<span><span class="co">##   [62] "PWY.2723..trehalose.degradation.V"                                                                          </span></span>
<span><span class="co">##   [63] "PWY.4041...gamma..glutamyl.cycle"                                                                           </span></span>
<span><span class="co">##   [64] "PWY.4702..phytate.degradation.I"                                                                            </span></span>
<span><span class="co">##   [65] "PWY.5005..biotin.biosynthesis.II"                                                                           </span></span>
<span><span class="co">##   [66] "PWY.5022..4.aminobutanoate.degradation.V"                                                                   </span></span>
<span><span class="co">##   [67] "PWY.5083..NAD.NADH.phosphorylation.and.dephosphorylation"                                                   </span></span>
<span><span class="co">##   [68] "PWY.5088..L.glutamate.degradation.VIII..to.propanoate."                                                     </span></span>
<span><span class="co">##   [69] "PWY.5121..superpathway.of.geranylgeranyl.diphosphate.biosynthesis.II..via.MEP."                             </span></span>
<span><span class="co">##   [70] "PWY.5136..fatty.acid..beta..oxidation.II..peroxisome."                                                      </span></span>
<span><span class="co">##   [71] "PWY.5138..unsaturated..even.numbered.fatty.acid..beta..oxidation"                                           </span></span>
<span><span class="co">##   [72] "PWY.5173..superpathway.of.acetyl.CoA.biosynthesis"                                                          </span></span>
<span><span class="co">##   [73] "PWY.5188..tetrapyrrole.biosynthesis.I..from.glutamate."                                                     </span></span>
<span><span class="co">##   [74] "PWY.5189..tetrapyrrole.biosynthesis.II..from.glycine."                                                      </span></span>
<span><span class="co">##   [75] "PWY.5345..superpathway.of.L.methionine.biosynthesis..by.sulfhydrylation."                                   </span></span>
<span><span class="co">##   [76] "PWY.5347..superpathway.of.L.methionine.biosynthesis..transsulfuration."                                     </span></span>
<span><span class="co">##   [77] "PWY.5367..petroselinate.biosynthesis"                                                                       </span></span>
<span><span class="co">##   [78] "PWY.5384..sucrose.degradation.IV..sucrose.phosphorylase."                                                   </span></span>
<span><span class="co">##   [79] "PWY.5464..superpathway.of.cytosolic.glycolysis..plants...pyruvate.dehydrogenase.and.TCA.cycle"              </span></span>
<span><span class="co">##   [80] "PWY.561..superpathway.of.glyoxylate.cycle.and.fatty.acid.degradation"                                       </span></span>
<span><span class="co">##   [81] "PWY.5656..mannosylglycerate.biosynthesis.I"                                                                 </span></span>
<span><span class="co">##   [82] "PWY.5675..nitrate.reduction.V..assimilatory."                                                               </span></span>
<span><span class="co">##   [83] "PWY.5676..acetyl.CoA.fermentation.to.butanoate.II"                                                          </span></span>
<span><span class="co">##   [84] "PWY.5677..succinate.fermentation.to.butanoate"                                                              </span></span>
<span><span class="co">##   [85] "PWY.5692..allantoin.degradation.to.glyoxylate.II"                                                           </span></span>
<span><span class="co">##   [86] "PWY.5705..allantoin.degradation.to.glyoxylate.III"                                                          </span></span>
<span><span class="co">##   [87] "PWY.5723..Rubisco.shunt"                                                                                    </span></span>
<span><span class="co">##   [88] "PWY.5747..2.methylcitrate.cycle.II"                                                                         </span></span>
<span><span class="co">##   [89] "PWY.5791..1.4.dihydroxy.2.naphthoate.biosynthesis.II..plants."                                              </span></span>
<span><span class="co">##   [90] "PWY.5837..1.4.dihydroxy.2.naphthoate.biosynthesis.I"                                                        </span></span>
<span><span class="co">##   [91] "PWY.5838..superpathway.of.menaquinol.8.biosynthesis.I"                                                      </span></span>
<span><span class="co">##   [92] "PWY.5840..superpathway.of.menaquinol.7.biosynthesis"                                                        </span></span>
<span><span class="co">##   [93] "PWY.5845..superpathway.of.menaquinol.9.biosynthesis"                                                        </span></span>
<span><span class="co">##   [94] "PWY.5850..superpathway.of.menaquinol.6.biosynthesis.I"                                                      </span></span>
<span><span class="co">##   [95] "PWY.5855..ubiquinol.7.biosynthesis..prokaryotic."                                                           </span></span>
<span><span class="co">##   [96] "PWY.5856..ubiquinol.9.biosynthesis..prokaryotic."                                                           </span></span>
<span><span class="co">##   [97] "PWY.5857..ubiquinol.10.biosynthesis..prokaryotic."                                                          </span></span>
<span><span class="co">##   [98] "PWY.5860..superpathway.of.demethylmenaquinol.6.biosynthesis.I"                                              </span></span>
<span><span class="co">##   [99] "PWY.5861..superpathway.of.demethylmenaquinol.8.biosynthesis"                                                </span></span>
<span><span class="co">##  [100] "PWY.5862..superpathway.of.demethylmenaquinol.9.biosynthesis"                                                </span></span>
<span><span class="co">##  [101] "PWY.5863..superpathway.of.phylloquinol.biosynthesis"                                                        </span></span>
<span><span class="co">##  [102] "PWY.5896..superpathway.of.menaquinol.10.biosynthesis"                                                       </span></span>
<span><span class="co">##  [103] "PWY.5897..superpathway.of.menaquinol.11.biosynthesis"                                                       </span></span>
<span><span class="co">##  [104] "PWY.5898..superpathway.of.menaquinol.12.biosynthesis"                                                       </span></span>
<span><span class="co">##  [105] "PWY.5899..superpathway.of.menaquinol.13.biosynthesis"                                                       </span></span>
<span><span class="co">##  [106] "PWY.5913..TCA.cycle.VI..obligate.autotrophs."                                                               </span></span>
<span><span class="co">##  [107] "PWY.5918..superpathay.of.heme.biosynthesis.from.glutamate"                                                  </span></span>
<span><span class="co">##  [108] "PWY.5920..superpathway.of.heme.biosynthesis.from.glycine"                                                   </span></span>
<span><span class="co">##  [109] "PWY.5971..palmitate.biosynthesis.II..bacteria.and.plants."                                                  </span></span>
<span><span class="co">##  [110] "PWY.5989..stearate.biosynthesis.II..bacteria.and.plants."                                                   </span></span>
<span><span class="co">##  [111] "PWY.6113..superpathway.of.mycolate.biosynthesis"                                                            </span></span>
<span><span class="co">##  [112] "PWY.6270..isoprene.biosynthesis.I"                                                                          </span></span>
<span><span class="co">##  [113] "PWY.6282..palmitoleate.biosynthesis.I..from..5Z..dodec.5.enoate."                                           </span></span>
<span><span class="co">##  [114] "PWY.6284..superpathway.of.unsaturated.fatty.acids.biosynthesis..E..coli."                                   </span></span>
<span><span class="co">##  [115] "PWY.6285..superpathway.of.fatty.acids.biosynthesis..E..coli."                                               </span></span>
<span><span class="co">##  [116] "PWY.6318..L.phenylalanine.degradation.IV..mammalian..via.side.chain."                                       </span></span>
<span><span class="co">##  [117] "PWY.6353..purine.nucleotides.degradation.II..aerobic."                                                      </span></span>
<span><span class="co">##  [118] "PWY.6471..peptidoglycan.biosynthesis.IV..Enterococcus.faecium."                                             </span></span>
<span><span class="co">##  [119] "PWY.6519..8.amino.7.oxononanoate.biosynthesis.I"                                                            </span></span>
<span><span class="co">##  [120] "PWY.6531..mannitol.cycle"                                                                                   </span></span>
<span><span class="co">##  [121] "PWY.6549..L.glutamine.biosynthesis.III"                                                                     </span></span>
<span><span class="co">##  [122] "PWY.6588..pyruvate.fermentation.to.acetone"                                                                 </span></span>
<span><span class="co">##  [123] "PWY.6590..superpathway.of.Clostridium.acetobutylicum.acidogenic.fermentation"                               </span></span>
<span><span class="co">##  [124] "PWY.6606..guanosine.nucleotides.degradation.II"                                                             </span></span>
<span><span class="co">##  [125] "PWY.6612..superpathway.of.tetrahydrofolate.biosynthesis"                                                    </span></span>
<span><span class="co">##  [126] "PWY.6628..superpathway.of.L.phenylalanine.biosynthesis"                                                     </span></span>
<span><span class="co">##  [127] "PWY.6629..superpathway.of.L.tryptophan.biosynthesis"                                                        </span></span>
<span><span class="co">##  [128] "PWY.6630..superpathway.of.L.tyrosine.biosynthesis"                                                          </span></span>
<span><span class="co">##  [129] "PWY.6690..cinnamate.and.3.hydroxycinnamate.degradation.to.2.oxopent.4.enoate"                               </span></span>
<span><span class="co">##  [130] "PWY.6708..ubiquinol.8.biosynthesis..prokaryotic."                                                           </span></span>
<span><span class="co">##  [131] "PWY.6731..starch.degradation.III"                                                                           </span></span>
<span><span class="co">##  [132] "PWY.6803..phosphatidylcholine.acyl.editing"                                                                 </span></span>
<span><span class="co">##  [133] "PWY.6823..molybdenum.cofactor.biosynthesis"                                                                 </span></span>
<span><span class="co">##  [134] "PWY.6837..fatty.acid.beta.oxidation.V..unsaturated..odd.number..di.isomerase.dependent."                    </span></span>
<span><span class="co">##  [135] "PWY.6876..isopropanol.biosynthesis"                                                                         </span></span>
<span><span class="co">##  [136] "PWY.6891..thiazole.biosynthesis.II..Bacillus."                                                              </span></span>
<span><span class="co">##  [137] "PWY.6895..superpathway.of.thiamin.diphosphate.biosynthesis.II"                                              </span></span>
<span><span class="co">##  [138] "PWY.6969..TCA.cycle.V..2.oxoglutarate.ferredoxin.oxidoreductase."                                           </span></span>
<span><span class="co">##  [139] "PWY.7003..glycerol.degradation.to.butanol"                                                                  </span></span>
<span><span class="co">##  [140] "PWY.7013..L.1.2.propanediol.degradation"                                                                    </span></span>
<span><span class="co">##  [141] "PWY.7046..4.coumarate.degradation..anaerobic."                                                              </span></span>
<span><span class="co">##  [142] "PWY.7094..fatty.acid.salvage"                                                                               </span></span>
<span><span class="co">##  [143] "PWY.7115..C4.photosynthetic.carbon.assimilation.cycle..NAD.ME.type"                                         </span></span>
<span><span class="co">##  [144] "PWY.7117..C4.photosynthetic.carbon.assimilation.cycle..PEPCK.type"                                          </span></span>
<span><span class="co">##  [145] "PWY.7204..pyridoxal.5..phosphate.salvage.II..plants."                                                       </span></span>
<span><span class="co">##  [146] "PWY.7209..superpathway.of.pyrimidine.ribonucleosides.degradation"                                           </span></span>
<span><span class="co">##  [147] "PWY.7254..TCA.cycle.VII..acetate.producers."                                                                </span></span>
<span><span class="co">##  [148] "PWY.7269..NAD.NADP.NADH.NADPH.mitochondrial.interconversion..yeast."                                        </span></span>
<span><span class="co">##  [149] "PWY.7279..aerobic.respiration.II..cytochrome.c...yeast."                                                    </span></span>
<span><span class="co">##  [150] "PWY.7288..fatty.acid..beta..oxidation..peroxisome..yeast."                                                  </span></span>
<span><span class="co">##  [151] "PWY.7315..dTDP.N.acetylthomosamine.biosynthesis"                                                            </span></span>
<span><span class="co">##  [152] "PWY.7328..superpathway.of.UDP.glucose.derived.O.antigen.building.blocks.biosynthesis"                       </span></span>
<span><span class="co">##  [153] "PWY.7385..1.3.propanediol.biosynthesis..engineered."                                                        </span></span>
<span><span class="co">##  [154] "PWY.7388..octanoyl..acyl.carrier.protein..biosynthesis..mitochondria..yeast."                               </span></span>
<span><span class="co">##  [155] "PWY.7392..taxadiene.biosynthesis..engineered."                                                              </span></span>
<span><span class="co">##  [156] "PWY.7446..sulfoglycolysis"                                                                                  </span></span>
<span><span class="co">##  [157] "PWY.7560..methylerythritol.phosphate.pathway.II"                                                            </span></span>
<span><span class="co">##  [158] "PWY.7616..methanol.oxidation.to.carbon.dioxide"                                                             </span></span>
<span><span class="co">##  [159] "PWY.7664..oleate.biosynthesis.IV..anaerobic."                                                               </span></span>
<span><span class="co">##  [160] "PWY.821..superpathway.of.sulfur.amino.acid.biosynthesis..Saccharomyces.cerevisiae."                         </span></span>
<span><span class="co">##  [161] "PWY0.1241..ADP.L.glycero..beta..D.manno.heptose.biosynthesis"                                               </span></span>
<span><span class="co">##  [162] "PWY0.1277..3.phenylpropanoate.and.3..3.hydroxyphenyl.propanoate.degradation"                                </span></span>
<span><span class="co">##  [163] "PWY0.1297..superpathway.of.purine.deoxyribonucleosides.degradation"                                         </span></span>
<span><span class="co">##  [164] "PWY0.1298..superpathway.of.pyrimidine.deoxyribonucleosides.degradation"                                     </span></span>
<span><span class="co">##  [165] "PWY0.1338..polymyxin.resistance"                                                                            </span></span>
<span><span class="co">##  [166] "PWY0.1415..superpathway.of.heme.biosynthesis.from.uroporphyrinogen.III"                                     </span></span>
<span><span class="co">##  [167] "PWY0.1479..tRNA.processing"                                                                                 </span></span>
<span><span class="co">##  [168] "PWY0.1533..methylphosphonate.degradation.I"                                                                 </span></span>
<span><span class="co">##  [169] "PWY0.41..allantoin.degradation.IV..anaerobic."                                                              </span></span>
<span><span class="co">##  [170] "PWY0.42..2.methylcitrate.cycle.I"                                                                           </span></span>
<span><span class="co">##  [171] "PWY0.781..aspartate.superpathway"                                                                           </span></span>
<span><span class="co">##  [172] "PWY0.862...5Z..dodec.5.enoate.biosynthesis"                                                                 </span></span>
<span><span class="co">##  [173] "PWY0.881..superpathway.of.fatty.acid.biosynthesis.I..E..coli."                                              </span></span>
<span><span class="co">##  [174] "PWY3O.355..stearate.biosynthesis.III..fungi."                                                               </span></span>
<span><span class="co">##  [175] "PWY4LZ.257..superpathway.of.fermentation..Chlamydomonas.reinhardtii."                                       </span></span>
<span><span class="co">##  [176] "PWY66.389..phytol.degradation"                                                                              </span></span>
<span><span class="co">##  [177] "PWY66.391..fatty.acid..beta..oxidation.VI..peroxisome."                                                     </span></span>
<span><span class="co">##  [178] "PWY66.398..TCA.cycle.III..animals."                                                                         </span></span>
<span><span class="co">##  [179] "PWY66.409..superpathway.of.purine.nucleotide.salvage"                                                       </span></span>
<span><span class="co">##  [180] "PWYG.321..mycolate.biosynthesis"                                                                            </span></span>
<span><span class="co">##  [181] "PYRIDNUCSAL.PWY..NAD.salvage.pathway.I"                                                                     </span></span>
<span><span class="co">##  [182] "REDCITCYC..TCA.cycle.VIII..helicobacter."                                                                   </span></span>
<span><span class="co">##  [183] "SALVADEHYPOX.PWY..adenosine.nucleotides.degradation.II"                                                     </span></span>
<span><span class="co">##  [184] "SO4ASSIM.PWY..sulfate.reduction.I..assimilatory."                                                           </span></span>
<span><span class="co">##  [185] "SULFATE.CYS.PWY..superpathway.of.sulfate.assimilation.and.cysteine.biosynthesis"                            </span></span>
<span><span class="co">##  [186] "TCA.GLYOX.BYPASS..superpathway.of.glyoxylate.bypass.and.TCA"                                                </span></span>
<span><span class="co">##  [187] "THREOCAT.PWY..superpathway.of.L.threonine.metabolism"                                                       </span></span>
<span><span class="co">##  [188] "UBISYN.PWY..superpathway.of.ubiquinol.8.biosynthesis..prokaryotic."                                         </span></span>
<span><span class="co">##  [189] "URDEGR.PWY..superpathway.of.allantoin.degradation.in.plants"                                                </span></span>
<span><span class="co">##  [190] "X3.HYDROXYPHENYLACETATE.DEGRADATION.PWY..4.hydroxyphenylacetate.degradation"</span></span>
<span></span>
<span><span class="co">############</span></span>
<span><span class="co"># Run MSEA #</span></span>
<span><span class="co">############</span></span>
<span></span>
<span><span class="va">MSEA</span> <span class="op">&lt;-</span> <span class="fu">run_MSEA</span><span class="op">(</span><span class="va">microbeSet</span>, <span class="va">results</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span></span>
<span><span class="va">MSEA</span> <span class="op">&lt;-</span> <span class="va">MSEA</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Set'</span>, <span class="st">'Freq'</span>, <span class="st">'ES'</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">MSEA</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Set'</span>, <span class="st">'Freq'</span>, <span class="st">'ES'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">MSEA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'ID'</span>, <span class="st">'Size'</span>, <span class="st">'pval'</span>, <span class="st">'qval'</span><span class="op">)</span></span>
<span><span class="va">MSEA</span><span class="op">$</span><span class="va">ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="va">MSEA</span><span class="op">$</span><span class="va">ID</span>, <span class="st">' ('</span>, <span class="va">MSEA</span><span class="op">$</span><span class="va">Size</span>, <span class="st">')'</span>, sep <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span></span>
<span><span class="co">########</span></span>
<span><span class="co"># Plot #</span></span>
<span><span class="co">########</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">MSEA</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">pval</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">ID</span>, levels <span class="op">=</span> <span class="va">ID</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span>, x <span class="op">=</span> <span class="va">ID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, fill <span class="op">=</span> <span class="st">'cornflowerblue'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Statistically significant modules associated with disease'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'MSEA enrichment score'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="MSEA_files/figure-html/perform-MSEA-and-visualization-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Based on the MSEA results, we obtain 4 enriched modules of microbial pathways. We can similarly examine the members of the top enriched modules.</p>
</section><section id="print-the-most-significant-modules-for-pathway-abundance-data" class="level2" data-number="17.5"><h2 data-number="17.5" class="anchored" data-anchor-id="print-the-most-significant-modules-for-pathway-abundance-data">
<span class="header-section-number">17.5</span> Print the most significant modules for pathway abundance data</h2>
<div class="cell" data-layout-align="center" data-hash="MSEA_cache/html/significant-results_3db61270f0fe3a0678a994a84b4affba">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Print the most significant modules</span></span>
<span><span class="va">microbeSet</span><span class="op">[[</span><span class="st">"MEpurple"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  NULL</span></span>
<span><span class="va">microbeSet</span><span class="op">[[</span><span class="st">"MEbrown"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##   [1] "GLCMANNANAUT.PWY..superpathway.of.N.acetylglucosamine..N.acetylmannosamine.and.N.acetylneuraminate.degradation"</span></span>
<span><span class="co">##   [2] "GLYCOLYSIS.E.D..superpathway.of.glycolysis.and.Entner.Doudoroff"                                               </span></span>
<span><span class="co">##   [3] "METH.ACETATE.PWY..methanogenesis.from.acetate"                                                                 </span></span>
<span><span class="co">##   [4] "PWY.5100..pyruvate.fermentation.to.acetate.and.lactate.II"                                                     </span></span>
<span><span class="co">##   [5] "PWY.5104..L.isoleucine.biosynthesis.IV"                                                                        </span></span>
<span><span class="co">##   [6] "PWY.5177..glutaryl.CoA.degradation"                                                                            </span></span>
<span><span class="co">##   [7] "PWY.5690..TCA.cycle.II..plants.and.fungi."                                                                     </span></span>
<span><span class="co">##   [8] "PWY.6470..peptidoglycan.biosynthesis.V...beta..lactam.resistance."                                             </span></span>
<span><span class="co">##   [9] "PWY.6527..stachyose.degradation"                                                                               </span></span>
<span><span class="co">##  [10] "PWY.6595..superpathway.of.guanosine.nucleotides.degradation..plants."                                          </span></span>
<span><span class="co">##  [11] "PWY.6608..guanosine.nucleotides.degradation.III"                                                               </span></span>
<span><span class="co">##  [12] "PWY.7196..superpathway.of.pyrimidine.ribonucleosides.salvage"                                                  </span></span>
<span><span class="co">##  [13] "PWY.7198..pyrimidine.deoxyribonucleotides.de.novo.biosynthesis.IV"                                             </span></span>
<span><span class="co">##  [14] "PWY0.1061..superpathway.of.L.alanine.biosynthesis"                                                             </span></span>
<span><span class="co">##  [15] "PWY0.1261..anhydromuropeptides.recycling"                                                                      </span></span>
<span><span class="co">##  [16] "PWY66.400..glycolysis.VI..metazoan."                                                                           </span></span>
<span><span class="co">##  [17] "TCA..TCA.cycle.I..prokaryotic."                                                                                </span></span>
<span><span class="co">##  [18] "UDPNAGSYN.PWY..UDP.N.acetyl.D.glucosamine.biosynthesis.I"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/MMUPHin_meta_analysis.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Meta-analyses</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/19_visualization_techniques.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Visualization</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>