<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Data Operations | Microbiome Analysis</title>
  <meta name="description" content="Chapter 2 Data Operations | Microbiome Analysis" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Data Operations | Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Data Operations | Microbiome Analysis" />
  
  
  

<meta name="author" content="" />


<meta name="date" content="2020-11-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro.html"/>
<link rel="next" href="bibliography.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="data-operations.html"><a href="data-operations.html"><i class="fa fa-check"></i><b>2</b> Data Operations</a><ul>
<li class="chapter" data-level="2.1" data-path="data-operations.html"><a href="data-operations.html#common-terminologies"><i class="fa fa-check"></i><b>2.1</b> Common terminologies</a></li>
<li class="chapter" data-level="2.2" data-path="data-operations.html"><a href="data-operations.html#retrieving-data-elements"><i class="fa fa-check"></i><b>2.2</b> Retrieving data elements</a></li>
<li class="chapter" data-level="2.3" data-path="data-operations.html"><a href="data-operations.html#extract-specific-elements"><i class="fa fa-check"></i><b>2.3</b> Extract specific elements</a><ul>
<li class="chapter" data-level="2.3.1" data-path="data-operations.html"><a href="data-operations.html#assays"><i class="fa fa-check"></i><b>2.3.1</b> Assays</a></li>
<li class="chapter" data-level="2.3.2" data-path="data-operations.html"><a href="data-operations.html#coldata"><i class="fa fa-check"></i><b>2.3.2</b> colData</a></li>
<li class="chapter" data-level="2.3.3" data-path="data-operations.html"><a href="data-operations.html#rowdata"><i class="fa fa-check"></i><b>2.3.3</b> rowData</a></li>
<li class="chapter" data-level="2.3.4" data-path="data-operations.html"><a href="data-operations.html#rowtree"><i class="fa fa-check"></i><b>2.3.4</b> rowTree</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-operations.html"><a href="data-operations.html#library-size"><i class="fa fa-check"></i><b>2.4</b> Library size</a></li>
<li class="chapter" data-level="2.5" data-path="data-operations.html"><a href="data-operations.html#agglomerate"><i class="fa fa-check"></i><b>2.5</b> Agglomerate</a></li>
<li class="chapter" data-level="2.6" data-path="data-operations.html"><a href="data-operations.html#get-unique"><i class="fa fa-check"></i><b>2.6</b> Get unique</a></li>
<li class="chapter" data-level="2.7" data-path="data-operations.html"><a href="data-operations.html#pick-specific"><i class="fa fa-check"></i><b>2.7</b> Pick specific</a><ul>
<li class="chapter" data-level="2.7.1" data-path="data-operations.html"><a href="data-operations.html#abundances-of-all-taxa-in-specific-sample"><i class="fa fa-check"></i><b>2.7.1</b> Abundances of all taxa in specific sample</a></li>
<li class="chapter" data-level="2.7.2" data-path="data-operations.html"><a href="data-operations.html#abundances-of-specific-taxa-in-all-samples"><i class="fa fa-check"></i><b>2.7.2</b> Abundances of specific taxa in all samples</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="data-operations.html"><a href="data-operations.html#join"><i class="fa fa-check"></i><b>2.8</b> Join</a></li>
<li class="chapter" data-level="2.9" data-path="data-operations.html"><a href="data-operations.html#barplot"><i class="fa fa-check"></i><b>2.9</b> Barplot</a><ul>
<li class="chapter" data-level="2.9.1" data-path="data-operations.html"><a href="data-operations.html#top-taxa"><i class="fa fa-check"></i><b>2.9.1</b> Top taxa</a></li>
<li class="chapter" data-level="2.9.2" data-path="data-operations.html"><a href="data-operations.html#convert-to-long-data.frame"><i class="fa fa-check"></i><b>2.9.2</b> Convert to long data.frame</a></li>
<li class="chapter" data-level="2.9.3" data-path="data-operations.html"><a href="data-operations.html#barplot-visualization"><i class="fa fa-check"></i><b>2.9.3</b> Barplot visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-operations" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Data Operations</h1>
<div id="common-terminologies" class="section level2">
<h2><span class="header-section-number">2.1</span> Common terminologies</h2>
<p>The <code>mia</code> package uses classes defined by <a href="https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html"><code>TreeSummarizedExperiment</code></a>, <a href="https://github.com/FelixErnst/MicrobiomeExperiment"><code>MicrobiomeExperiment</code></a> and in general <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html"><code>SummarizedExperiment</code></a>.</p>
<p><code>assays</code> : Similar to <code>otu_table</code> in <code>phyloseq</code>. An <code>SummarizedExperiment</code> object
can hold multiple assays, raw counts, transformed counts. See also
<a href="https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html"><code>MultiAssayExperiment</code></a><br />
<code>rowData</code> : Slot similar to <code>tax_table</code> in <code>phyloseq</code><br />
<code>colData</code> : Slot similar to <code>sample_data</code> in <code>phyloseq</code><br />
<code>rowTree</code> : Slot for phylogenetic tree. Similar to <code>phy_tree</code> in <code>phyloseq</code><br />
<code>FeatureIDs</code> : These are row names in <code>assays</code> and <code>rowData</code>. Synonymous to OTU/ASV ids.<br />
<code>SampleIDs</code> : These are column names in <code>assays</code> and row names in <code>colData</code>.</p>
</div>
<div id="retrieving-data-elements" class="section level2">
<h2><span class="header-section-number">2.2</span> Retrieving data elements</h2>
<p>Extract specific elements from <code>TreeSummarizedExperiment</code> and <code>MicrobiomeExperiment</code> object.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="data-operations.html#cb1-1"></a><span class="kw">library</span>(mia)</span></code></pre></div>
<pre><code>## Warning: multiple methods tables found for &#39;agglomerateByRank&#39;</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="data-operations.html#cb3-1"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb3-2"><a href="data-operations.html#cb3-2"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb3-3"><a href="data-operations.html#cb3-3"></a><span class="kw">library</span>(tibble)</span></code></pre></div>
<p>Example data</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="data-operations.html#cb4-1"></a><span class="kw">data</span>(GlobalPatterns, <span class="dt">package =</span> <span class="st">&quot;MicrobiomeExperiment&quot;</span>)</span>
<span id="cb4-2"><a href="data-operations.html#cb4-2"></a>me &lt;-<span class="st"> </span>GlobalPatterns </span></code></pre></div>
<p>Create a relative abundance assay</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="data-operations.html#cb5-1"></a>me &lt;-<span class="st"> </span><span class="kw">relAbundanceCounts</span>(GlobalPatterns)</span>
<span id="cb5-2"><a href="data-operations.html#cb5-2"></a>me</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 19216 26 
## metadata(0):
## assays(2): counts relabundance
## rownames(19216): 549322 522457 ... 200359 271582
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(7): X.SampleID Primer ... SampleType Description
## reducedDimNames(0):
## altExpNames(0):
## rowLinks: a LinkDataFrame (19216 rows)
## rowTree: a phylo (19216 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<p>Above there are now two <code>assays</code> 1] counts 2] relabundance</p>
</div>
<div id="extract-specific-elements" class="section level2">
<h2><span class="header-section-number">2.3</span> Extract specific elements</h2>
<div id="assays" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Assays</h3>
<p>Assay slots hold the counts and transformed data for <code>FeatureIDs</code></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="data-operations.html#cb7-1"></a>me_counts &lt;-<span class="st"> </span><span class="kw">assay</span>(me, <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb7-2"><a href="data-operations.html#cb7-2"></a>me_counts[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##        CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## 549322   0   0   0       0       0       0       0
## 522457   0   0   0       0       0       0       0
## 951      0   0   0       0       0       0       1
## 244423   0   0   0       0       0       0       0
## 586076   0   0   0       0       0       0       0</code></pre>
<p>Extract relative abundance data.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="data-operations.html#cb9-1"></a>me_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me, <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb9-2"><a href="data-operations.html#cb9-2"></a>me_reab[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##        CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr     M11Plmr
## 549322   0   0   0       0       0       0 0.00000e+00
## 522457   0   0   0       0       0       0 0.00000e+00
## 951      0   0   0       0       0       0 2.30471e-06
## 244423   0   0   0       0       0       0 0.00000e+00
## 586076   0   0   0       0       0       0 0.00000e+00</code></pre>
</div>
<div id="coldata" class="section level3">
<h3><span class="header-section-number">2.3.2</span> colData</h3>
<p>Get colData as a data.frame.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="data-operations.html#cb11-1"></a>me_sample_data &lt;-<span class="st"> </span><span class="kw">colData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-2"><a href="data-operations.html#cb11-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>()</span>
<span id="cb11-3"><a href="data-operations.html#cb11-3"></a><span class="co"># show select row and columns</span></span>
<span id="cb11-4"><a href="data-operations.html#cb11-4"></a>me_sample_data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##         X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T
## CL3            CL3 ILBC_01        AACGCA                   TGCGTT
## CC1            CC1 ILBC_02        AACTCG                   CGAGTT
## SV1            SV1 ILBC_03        AACTGT                   ACAGTT
## M31Fcsw    M31Fcsw ILBC_04        AAGAGA                   TCTCTT
## M11Fcsw    M11Fcsw ILBC_05        AAGCTG                   CAGCTT
##         Barcode_full_length
## CL3             CTAGCGTGCGT
## CC1             CATCGACGAGT
## SV1             GTACGCACAGT
## M31Fcsw         TCGACATCTCT
## M11Fcsw         CGACTGCAGCT</code></pre>
</div>
<div id="rowdata" class="section level3">
<h3><span class="header-section-number">2.3.3</span> rowData</h3>
<p>Get rowData as a data.frame</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="data-operations.html#cb13-1"></a>me_row_data &lt;-<span class="st"> </span><span class="kw">rowData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb13-2"><a href="data-operations.html#cb13-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>()</span>
<span id="cb13-3"><a href="data-operations.html#cb13-3"></a><span class="co"># show select row and columns</span></span>
<span id="cb13-4"><a href="data-operations.html#cb13-4"></a>me_row_data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##        Kingdom        Phylum        Class        Order        Family
## 549322 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 522457 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 951    Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 244423 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;
## 586076 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;</code></pre>
</div>
<div id="rowtree" class="section level3">
<h3><span class="header-section-number">2.3.4</span> rowTree</h3>
<p>Get rowTree</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="data-operations.html#cb15-1"></a>me_row_tree &lt;-<span class="st"> </span><span class="kw">rowTree</span>(me) </span>
<span id="cb15-2"><a href="data-operations.html#cb15-2"></a>me_row_tree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 19216 tips and 19215 internal nodes.
## 
## Tip labels:
##   549322, 522457, 951, 244423, 586076, 246140, ...
## Node labels:
##   , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>Check link between rownames (OTU/ASV ids)</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="data-operations.html#cb17-1"></a><span class="co"># extract the linkData</span></span>
<span id="cb17-2"><a href="data-operations.html#cb17-2"></a><span class="co"># on rows</span></span>
<span id="cb17-3"><a href="data-operations.html#cb17-3"></a>rowL &lt;-<span class="st"> </span><span class="kw">rowLinks</span>(<span class="dt">x =</span> me)</span></code></pre></div>
</div>
</div>
<div id="library-size" class="section level2">
<h2><span class="header-section-number">2.4</span> Library size</h2>
<p>Get total counts/sample</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="data-operations.html#cb18-1"></a>reads_sample &lt;-<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">assay</span>(me, <span class="st">&quot;counts&quot;</span>)) </span>
<span id="cb18-2"><a href="data-operations.html#cb18-2"></a><span class="co"># add it to colData</span></span>
<span id="cb18-3"><a href="data-operations.html#cb18-3"></a><span class="kw">colData</span>(me)<span class="op">$</span>reads_sample &lt;-<span class="st"> </span>reads_sample</span></code></pre></div>
<p>Visualize library size</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="data-operations.html#cb19-1"></a>me_sample_data &lt;-<span class="st"> </span><span class="kw">colData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-2"><a href="data-operations.html#cb19-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-3"><a href="data-operations.html#cb19-3"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;SampleId&quot;</span>)</span>
<span id="cb19-4"><a href="data-operations.html#cb19-4"></a></span>
<span id="cb19-5"><a href="data-operations.html#cb19-5"></a><span class="kw">ggplot</span>(me_sample_data, <span class="kw">aes</span>(reads_sample, SampleId)) <span class="op">+</span></span>
<span id="cb19-6"><a href="data-operations.html#cb19-6"></a><span class="st">    </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>SampleType)) <span class="op">+</span></span>
<span id="cb19-7"><a href="data-operations.html#cb19-7"></a><span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span><span class="st"> </span><span class="co"># prettify plot</span></span>
<span id="cb19-8"><a href="data-operations.html#cb19-8"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">10</span>), </span>
<span id="cb19-9"><a href="data-operations.html#cb19-9"></a>          <span class="dt">axis.line.x =</span> <span class="kw">element_line</span>(<span class="dt">color =</span> <span class="st">&quot;#242424&quot;</span>), </span>
<span id="cb19-10"><a href="data-operations.html#cb19-10"></a>          <span class="dt">axis.line.y =</span> <span class="kw">element_line</span>(<span class="dt">color =</span> <span class="st">&quot;#242424&quot;</span>), </span>
<span id="cb19-11"><a href="data-operations.html#cb19-11"></a>          <span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(), </span>
<span id="cb19-12"><a href="data-operations.html#cb19-12"></a>          <span class="dt">panel.border =</span> <span class="kw">element_rect</span>(<span class="dt">colour =</span> <span class="st">&quot;#242424&quot;</span>), </span>
<span id="cb19-13"><a href="data-operations.html#cb19-13"></a>          <span class="dt">legend.key =</span> <span class="kw">element_blank</span>(), <span class="dt">legend.text =</span> <span class="kw">element_text</span>(<span class="dt">color =</span> <span class="st">&quot;#353535&quot;</span>), </span>
<span id="cb19-14"><a href="data-operations.html#cb19-14"></a>          <span class="dt">legend.background =</span> <span class="kw">element_rect</span>(<span class="dt">colour =</span> <span class="ot">NA</span>, <span class="dt">fill =</span> <span class="ot">NA</span>))</span></code></pre></div>
<p><img src="02-data-operations_files/figure-html/viz-lib-size-1.png" width="768" /></p>
</div>
<div id="agglomerate" class="section level2">
<h2><span class="header-section-number">2.5</span> Agglomerate</h2>
<p>Agglomerate at a specific taxonomic rank. If multiple assays (counts and relabundance) exists, both will be agglomerated.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="data-operations.html#cb20-1"></a>me_fam &lt;-<span class="st"> </span><span class="kw">agglomerateByRank</span>(me, <span class="dt">rank =</span> <span class="st">&quot;Family&quot;</span>,</span>
<span id="cb20-2"><a href="data-operations.html#cb20-2"></a>                        <span class="dt">agglomerateTree =</span> <span class="ot">TRUE</span>)</span>
<span id="cb20-3"><a href="data-operations.html#cb20-3"></a></span>
<span id="cb20-4"><a href="data-operations.html#cb20-4"></a><span class="kw">print</span>(me_fam)</span></code></pre></div>
<pre><code>## class: TreeSummarizedExperiment 
## dim: 603 26 
## metadata(0):
## assays(2): counts relabundance
## rownames(603): Class::Thermoprotei Family::Sulfolobaceae ...
##   Family::Thermodesulfobiaceae Phylum::SR1
## rowData names(7): Kingdom Phylum ... Genus Species
## colnames(26): CL3 CC1 ... Even2 Even3
## colData names(8): X.SampleID Primer ... Description reads_sample
## reducedDimNames(0):
## altExpNames(0):
## rowLinks: a LinkDataFrame (603 rows)
## rowTree: a phylo (603 leaves)
## colLinks: NULL
## colTree: NULL</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="data-operations.html#cb22-1"></a>me_fam_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me_fam, <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb22-2"><a href="data-operations.html#cb22-2"></a>me_fam_reab[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                                CL3          CC1 SV1 M31Fcsw M11Fcsw M31Plmr
## Class::Thermoprotei   0.0000000000 0.000000e+00   0       0       0       0
## Family::Sulfolobaceae 0.0000000000 0.000000e+00   0       0       0       0
## Class::Sd-NA          0.0000000000 0.000000e+00   0       0       0       0
## Order::NRP-J          0.0001990563 2.069651e-04   0       0       0       0
## Family::SAGMA-X       0.0000000000 6.164919e-06   0       0       0       0
##                            M11Plmr
## Class::Thermoprotei   0.000000e+00
## Family::Sulfolobaceae 2.304710e-06
## Class::Sd-NA          0.000000e+00
## Order::NRP-J          6.914131e-06
## Family::SAGMA-X       0.000000e+00</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="data-operations.html#cb24-1"></a>me_fam_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me_fam, <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb24-2"><a href="data-operations.html#cb24-2"></a>me_fam_reab[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</span></code></pre></div>
<pre><code>##                       CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
## Class::Thermoprotei     0   0   0       0       0       0       0
## Family::Sulfolobaceae   0   0   0       0       0       0       1
## Class::Sd-NA            0   0   0       0       0       0       0
## Order::NRP-J          172 235   0       0       0       0       3
## Family::SAGMA-X         0   7   0       0       0       0       0</code></pre>
<p>These data can also be kept in</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="data-operations.html#cb26-1"></a><span class="kw">altExp</span>(me, <span class="st">&quot;family&quot;</span>) &lt;-<span class="st"> </span>me_fam</span></code></pre></div>
<p><code>altExpNames</code> now consists of <code>family</code> level data. This can be extended to use any level present in Kingdom, Phylum, Class, Order, Family, Genus, Species</p>
</div>
<div id="get-unique" class="section level2">
<h2><span class="header-section-number">2.6</span> Get unique</h2>
<p>Get which Phyla are present.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="data-operations.html#cb27-1"></a><span class="kw">head</span>(<span class="kw">unique</span>(<span class="kw">rowData</span>(me)[,<span class="st">&quot;Phylum&quot;</span>]))</span></code></pre></div>
<pre><code>## [1] &quot;Crenarchaeota&quot;  &quot;Euryarchaeota&quot;  &quot;Actinobacteria&quot; &quot;Spirochaetes&quot;  
## [5] &quot;MVP-15&quot;         &quot;Proteobacteria&quot;</code></pre>
</div>
<div id="pick-specific" class="section level2">
<h2><span class="header-section-number">2.7</span> Pick specific</h2>
<p>Retrieving of specific elements are required for specific analysis. For instance, extracting abundances for a specific taxa in all samples or all taxa in one sample.</p>
<div id="abundances-of-all-taxa-in-specific-sample" class="section level3">
<h3><span class="header-section-number">2.7.1</span> Abundances of all taxa in specific sample</h3>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="data-operations.html#cb29-1"></a><span class="co">#Note: getAbundanceSample()</span></span>
<span id="cb29-2"><a href="data-operations.html#cb29-2"></a>taxa.abundances.cc1 &lt;-<span class="st"> </span><span class="kw">assay</span>(me)[, <span class="st">&quot;CC1&quot;</span>]</span>
<span id="cb29-3"><a href="data-operations.html#cb29-3"></a>taxa.abundances.cc1[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 549322 522457    951 244423 586076 246140 143239 244960 255340 144887 
##      0      0      0      0      0      0      1      0    194      5</code></pre>
</div>
<div id="abundances-of-specific-taxa-in-all-samples" class="section level3">
<h3><span class="header-section-number">2.7.2</span> Abundances of specific taxa in all samples</h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="data-operations.html#cb31-1"></a><span class="co">#Note: getAbundanceFeature()</span></span>
<span id="cb31-2"><a href="data-operations.html#cb31-2"></a>taxa.abundances &lt;-<span class="st"> </span><span class="kw">assay</span>(me)[<span class="st">&quot;255340&quot;</span>,]</span>
<span id="cb31-3"><a href="data-operations.html#cb31-3"></a>taxa.abundances[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##     CL3     CC1     SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr F21Plmr M31Tong M11Tong 
##     153     194       0       0       0       0       0       0       0       0</code></pre>
</div>
</div>
<div id="join" class="section level2">
<h2><span class="header-section-number">2.8</span> Join</h2>
<p>Joining otu/asv table and taxonomy in one data frame</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="data-operations.html#cb33-1"></a>me_reab &lt;-<span class="st"> </span><span class="kw">assay</span>(me, <span class="st">&quot;relabundance&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb33-2"><a href="data-operations.html#cb33-2"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb33-3"><a href="data-operations.html#cb33-3"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;FeatureID&quot;</span>)</span>
<span id="cb33-4"><a href="data-operations.html#cb33-4"></a></span>
<span id="cb33-5"><a href="data-operations.html#cb33-5"></a>me_joined &lt;-<span class="st"> </span><span class="kw">rowData</span>(me) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb33-6"><a href="data-operations.html#cb33-6"></a><span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb33-7"><a href="data-operations.html#cb33-7"></a><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;FeatureID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb33-8"><a href="data-operations.html#cb33-8"></a><span class="st">    </span><span class="kw">left_join</span>(me_reab) </span></code></pre></div>
<pre><code>## Joining, by = &quot;FeatureID&quot;</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="data-operations.html#cb35-1"></a><span class="kw">head</span>(me_joined)</span></code></pre></div>
<pre><code>##   FeatureID Kingdom        Phylum        Class        Order        Family
## 1    549322 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 2    522457 Archaea Crenarchaeota Thermoprotei         &lt;NA&gt;          &lt;NA&gt;
## 3       951 Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
## 4    244423 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;
## 5    586076 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;
## 6    246140 Archaea Crenarchaeota        Sd-NA         &lt;NA&gt;          &lt;NA&gt;
##        Genus                  Species CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr
## 1       &lt;NA&gt;                     &lt;NA&gt;   0   0   0       0       0       0
## 2       &lt;NA&gt;                     &lt;NA&gt;   0   0   0       0       0       0
## 3 Sulfolobus Sulfolobusacidocaldarius   0   0   0       0       0       0
## 4       &lt;NA&gt;                     &lt;NA&gt;   0   0   0       0       0       0
## 5       &lt;NA&gt;                     &lt;NA&gt;   0   0   0       0       0       0
## 6       &lt;NA&gt;                     &lt;NA&gt;   0   0   0       0       0       0
##       M11Plmr F21Plmr M31Tong M11Tong LMEpi24M     SLEpi20M       AQC1cm
## 1 0.00000e+00       0       0       0        0 8.214821e-07 2.312143e-05
## 2 0.00000e+00       0       0       0        0 0.000000e+00 0.000000e+00
## 3 2.30471e-06       0       0       0        0 0.000000e+00 0.000000e+00
## 4 0.00000e+00       0       0       0        0 0.000000e+00 0.000000e+00
## 5 0.00000e+00       0       0       0        0 0.000000e+00 0.000000e+00
## 6 0.00000e+00       0       0       0        0 0.000000e+00 0.000000e+00
##         AQC4cm       AQC7cm          NP2 NP3 NP5 TRRsed1 TRRsed2 TRRsed3 TS28
## 1 4.242356e-05 7.650240e-05 1.909731e-06   0   0       0       0       0    0
## 2 8.484711e-07 3.530880e-06 0.000000e+00   0   0       0       0       0    0
## 3 0.000000e+00 0.000000e+00 0.000000e+00   0   0       0       0       0    0
## 4 9.333182e-06 1.706592e-05 0.000000e+00   0   0       0       0       0    0
## 5 8.484711e-07 5.884800e-07 0.000000e+00   0   0       0       0       0    0
## 6 4.242356e-07 1.765440e-06 0.000000e+00   0   0       0       0       0    0
##   TS29 Even1 Even2 Even3
## 1    0     0     0     0
## 2    0     0     0     0
## 3    0     0     0     0
## 4    0     0     0     0
## 5    0     0     0     0
## 6    0     0     0     0</code></pre>
</div>
<div id="barplot" class="section level2">
<h2><span class="header-section-number">2.9</span> Barplot</h2>
<p>As example we first <code>agglomerateByRank</code> to phylum level.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="data-operations.html#cb37-1"></a>me_phy &lt;-<span class="st"> </span><span class="kw">agglomerateByRank</span>(me, <span class="dt">rank =</span> <span class="st">&quot;Phylum&quot;</span>,</span>
<span id="cb37-2"><a href="data-operations.html#cb37-2"></a>                        <span class="dt">agglomerateTree =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div id="top-taxa" class="section level3">
<h3><span class="header-section-number">2.9.1</span> Top taxa</h3>
<p>Define a function to get top taxa.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="data-operations.html#cb38-1"></a>topTaxa &lt;-<span class="st"> </span><span class="cf">function</span>(x, <span class="dt">method=</span><span class="kw">c</span>(<span class="st">&quot;sum&quot;</span>,<span class="st">&quot;mean&quot;</span>,<span class="st">&quot;median&quot;</span>), <span class="dt">top=</span><span class="dv">5</span>){</span>
<span id="cb38-2"><a href="data-operations.html#cb38-2"></a>    taxs &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb38-3"><a href="data-operations.html#cb38-3"></a>    <span class="cf">if</span>(method<span class="op">==</span><span class="st">&quot;sum&quot;</span>){</span>
<span id="cb38-4"><a href="data-operations.html#cb38-4"></a>        taxs &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">rowSums</span>(<span class="kw">assay</span>(x)),<span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span>top]</span>
<span id="cb38-5"><a href="data-operations.html#cb38-5"></a>    } <span class="cf">else</span> <span class="cf">if</span> (method<span class="op">==</span><span class="st">&quot;mean&quot;</span>){</span>
<span id="cb38-6"><a href="data-operations.html#cb38-6"></a>        taxs &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">rowMeans</span>(<span class="kw">assay</span>(x)),<span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span>top]</span>
<span id="cb38-7"><a href="data-operations.html#cb38-7"></a>    } <span class="cf">else</span> <span class="cf">if</span> (method<span class="op">==</span><span class="st">&quot;median&quot;</span>) {</span>
<span id="cb38-8"><a href="data-operations.html#cb38-8"></a>        taxs &lt;-<span class="st"> </span><span class="kw">rowMedians</span>(<span class="kw">assay</span>(x))</span>
<span id="cb38-9"><a href="data-operations.html#cb38-9"></a>        <span class="kw">names</span>(taxs) &lt;-<span class="st"> </span><span class="kw">rownames</span>(<span class="kw">assay</span>(x))</span>
<span id="cb38-10"><a href="data-operations.html#cb38-10"></a>        taxs &lt;-<span class="st"> </span><span class="kw">sort</span>(taxs,<span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span>top]</span>
<span id="cb38-11"><a href="data-operations.html#cb38-11"></a>    }</span>
<span id="cb38-12"><a href="data-operations.html#cb38-12"></a>    <span class="kw">return</span>(<span class="kw">names</span>(taxs))</span>
<span id="cb38-13"><a href="data-operations.html#cb38-13"></a>}</span></code></pre></div>
<p>Get top 10 taxa</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="data-operations.html#cb39-1"></a>top_phy &lt;-<span class="st"> </span><span class="kw">topTaxa</span>(me_phy, <span class="dt">method=</span><span class="st">&quot;median&quot;</span>, <span class="dt">top=</span><span class="dv">10</span>)</span>
<span id="cb39-2"><a href="data-operations.html#cb39-2"></a>top_phy</span></code></pre></div>
<pre><code>##  [1] &quot;Phylum::Proteobacteria&quot;  &quot;Phylum::Bacteroidetes&quot;  
##  [3] &quot;Phylum::Actinobacteria&quot;  &quot;Phylum::Cyanobacteria&quot;  
##  [5] &quot;Phylum::Firmicutes&quot;      &quot;Phylum::Verrucomicrobia&quot;
##  [7] &quot;Phylum::Planctomycetes&quot;  &quot;Phylum::Acidobacteria&quot;  
##  [9] &quot;Phylum::Chloroflexi&quot;     &quot;Phylum::Tenericutes&quot;</code></pre>
</div>
<div id="convert-to-long-data.frame" class="section level3">
<h3><span class="header-section-number">2.9.2</span> Convert to long data.frame</h3>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="data-operations.html#cb41-1"></a>molten_se &lt;-<span class="st"> </span><span class="kw">meltAssay</span>(me_phy,</span>
<span id="cb41-2"><a href="data-operations.html#cb41-2"></a>  <span class="dt">add_row_data =</span> <span class="ot">TRUE</span>,</span>
<span id="cb41-3"><a href="data-operations.html#cb41-3"></a>  <span class="dt">add_col_data =</span> <span class="ot">TRUE</span>,</span>
<span id="cb41-4"><a href="data-operations.html#cb41-4"></a>  <span class="dt">abund_values =</span> <span class="st">&quot;relabundance&quot;</span></span>
<span id="cb41-5"><a href="data-operations.html#cb41-5"></a>)</span></code></pre></div>
</div>
<div id="barplot-visualization" class="section level3">
<h3><span class="header-section-number">2.9.3</span> Barplot visualization</h3>
<p>A barplot of composition for top 10 phyla</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="data-operations.html#cb42-1"></a><span class="co"># Chage all not in top_phy to &quot;Other&quot; category</span></span>
<span id="cb42-2"><a href="data-operations.html#cb42-2"></a>molten_se &lt;-<span class="st"> </span>molten_se <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb42-3"><a href="data-operations.html#cb42-3"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">TaxVar=</span> <span class="kw">ifelse</span>(FeatureID <span class="op">%in%</span><span class="st"> </span>top_phy, Phylum, <span class="st">&quot;Other&quot;</span>))</span>
<span id="cb42-4"><a href="data-operations.html#cb42-4"></a></span>
<span id="cb42-5"><a href="data-operations.html#cb42-5"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(molten_se, <span class="kw">aes_string</span>(<span class="dt">x=</span><span class="st">&quot;SampleID&quot;</span>, <span class="dt">y=</span><span class="st">&quot;relabundance&quot;</span>, <span class="dt">fill=</span><span class="st">&quot;TaxVar&quot;</span>)) <span class="op">+</span></span>
<span id="cb42-6"><a href="data-operations.html#cb42-6"></a><span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">position=</span><span class="st">&quot;stack&quot;</span>, <span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>) <span class="op">+</span></span>
<span id="cb42-7"><a href="data-operations.html#cb42-7"></a><span class="st">    </span><span class="kw">scale_x_discrete</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb42-8"><a href="data-operations.html#cb42-8"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">&quot;Relative Abundance&quot;</span>,<span class="dt">x=</span><span class="st">&quot;&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb42-9"><a href="data-operations.html#cb42-9"></a><span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb42-10"><a href="data-operations.html#cb42-10"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">90</span>, </span>
<span id="cb42-11"><a href="data-operations.html#cb42-11"></a>                                   <span class="dt">hjust=</span><span class="fl">0.95</span>,</span>
<span id="cb42-12"><a href="data-operations.html#cb42-12"></a>                                   <span class="dt">vjust=</span><span class="fl">0.2</span>),</span>
<span id="cb42-13"><a href="data-operations.html#cb42-13"></a>          <span class="dt">panel.grid =</span> <span class="kw">element_blank</span>(),</span>
<span id="cb42-14"><a href="data-operations.html#cb42-14"></a>          <span class="dt">legend.text =</span> <span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&quot;italic&quot;</span>)) <span class="op">+</span></span>
<span id="cb42-15"><a href="data-operations.html#cb42-15"></a><span class="st">    </span><span class="kw">scale_fill_brewer</span>(<span class="st">&quot;Phylum&quot;</span>,<span class="dt">palette =</span> <span class="st">&quot;Paired&quot;</span>) <span class="co"># this can be argument (fill.colors= &quot;Paired&quot;) or something similar</span></span>
<span id="cb42-16"><a href="data-operations.html#cb42-16"></a>p <span class="op">+</span><span class="st"> </span><span class="kw">scale_y_continuous</span>(<span class="dt">labels =</span> scales<span class="op">::</span>percent) <span class="co"># this can be optional (relative==TRUE) or something similar</span></span></code></pre></div>
<p><img src="02-data-operations_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bibliography.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MiaBook.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
